diff --git a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
index 7da4a0506d..49c945901c 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
@@ -1437,7 +1437,24 @@ void fn_800DA190(Fighter_GObj* gobj)
     }
 }
 
-/// #fn_800DA1D8
+void fn_800DA1D8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    HSD_JObj* jobj;
+    PAD_STACK(8);
+    fp->gr_vel = 0.0F;
+    Fighter_ChangeMotionState(gobj, ftCo_MS_CatchWait, 0, 0.0F, 1.0F, 0.0F,
+                              NULL);
+    fp->accessory1_cb = fn_800DA4A0;
+    fp->take_dmg_cb = fn_800DA490;
+    fp->x221B_b7 = false;
+    jobj = fp->parts[fp->ft_data->x8->x11].joint;
+    fp->mv.co.capturedamage.x18 = jobj;
+    efAsync_Spawn(gobj, &GET_FIGHTER(gobj)->x60C, 1, 0x41D, jobj);
+    ftCommon_8007EBAC(fp, 3, 0);
+    ftCommon_8007E2F4(fp, 0x1FF);
+    fn_800DB6C8(fp->victim_gobj);
+}
 
 void fn_800DA2B0(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftCommon/ftCo_Attack100.h b/src/melee/ft/chara/ftCommon/ftCo_Attack100.h
index 6381124a87..5807806fe6 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Attack100.h
+++ b/src/melee/ft/chara/ftCommon/ftCo_Attack100.h
@@ -120,7 +120,7 @@
 /* 0D9FDC */ void ftCo_CatchPull_Coll(Fighter_GObj* gobj);
 /* 0DA004 */ void fn_800DA004(Fighter_GObj* gobj);
 /* 0DA190 */ void fn_800DA190(Fighter_GObj* gobj);
-/* 0DA1D8 */ UNK_RET fn_800DA1D8(UNK_PARAMS);
+/* 0DA1D8 */ void fn_800DA1D8(Fighter_GObj* gobj);
 /* 0DA2B0 */ void fn_800DA2B0(Fighter_GObj*);
 /* 0DA318 */ void ftCo_CatchWait_Anim(Fighter_GObj* gobj);
 /* 0DA31C */ void ftCo_CatchWait_IASA(Fighter_GObj* gobj);
diff --git a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
index 57fc61216d..549676245d 100644
--- a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
+++ b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
@@ -31,6 +31,7 @@
 #include "gr/stage.h"
 #include "it/it_26B1.h"
 #include "it/items/itmasterhandlaser.h"
+#include "mp/mplib.h"
 #include "lb/lb_00B0.h"
 #include "lb/lbaudio_ax.h"
 #include "lb/lbvector.h"
@@ -931,7 +932,29 @@ void ftCh_Init_80157170(HSD_GObj* gobj)
     ft_PlaySFX(fp, 0x4E207, 127, 64);
 }
 
-/// #ftCh_Damage2_Anim
+void ftCh_Damage2_Anim(HSD_GObj* gobj)
+{
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        Vec3 pos;
+        mpFloorGetRight(0, &pos);
+        if (fp->cur_pos.x > pos.x) {
+            ftCh_Init_8015737C(gobj);
+        } else {
+            ftCrazyHand_DatAttrs* attrs;
+            fp = GET_FIGHTER(gobj);
+            attrs = fp->ft_data->ext_attr;
+            fp->mv.ch.unk0.xC.x = fp->cur_pos.x - attrs->x28;
+            fp->mv.ch.unk0.xC.y = attrs->x24;
+            Fighter_ChangeMotionState(
+                gobj, ftMh_MS_Damage, 0,
+                fp->mv.ch.unk0.xC.z = ftCh_Init_804DA070,
+                ftCh_Init_804DA074, ftCh_Init_804DA070, NULL);
+            ftAnim_8006EBA4(gobj);
+            ft_PlaySFX(fp, 0x4E207, 127, 64);
+        }
+    }
+}
 
 void ftCh_Damage2_IASA(HSD_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.c b/src/melee/ft/chara/ftKirby/ftKb_Init.c
index 1405389276..56e4ae5902 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.c
@@ -34,6 +34,7 @@
 #include "ft/ftcommon.h"
 #include "ft/ftdata.h"
 #include "ft/ftdynamics.h"
+#include "ft/ftanim.h"
 #include "ft/ftparts.h"
 #include "ft/ftwalkcommon.h"
 #include "ft/inlines.h"
@@ -3320,7 +3321,33 @@ void ftKb_SpecialN_800EF040(Fighter_GObj* gobj, int arg1, KirbyHatStruct* hat)
 
 /// #ftKb_SpecialN_800EF0E4
 
-/// #ftKb_SpecialN_800EF35C
+void ftKb_SpecialN_800EF35C(Fighter_GObj* gobj, int arg1, u8* arg2)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    struct {
+        HSD_Joint* joint;
+        HSD_MatAnimJoint* matanim;
+    }* costume_data = (void*) ftKb_Init_803C9FC8[arg1];
+    HSD_MatAnimJoint* matanimjoint =
+        costume_data[fp->x619_costume_id].matanim;
+    int idx = 0;
+    int i = 0;
+    while (matanimjoint != NULL) {
+        if (ftParts_8007506C(fp->kind, i) != 0) {
+            i++;
+            arg2++;
+        } else {
+            if (matanimjoint->matanim != NULL) {
+                HSD_DObjAddAnimAll(fp->fv.kb.hat.x14.data[*arg2],
+                                   matanimjoint->matanim, NULL);
+            }
+            i++;
+            arg2++;
+            ftAnim_GetNextMatAnimJointInTree(&matanimjoint, &idx);
+        }
+    }
+    PAD_STACK(8);
+}
 
 /// #ftKb_SpecialN_800EF438
 
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.h b/src/melee/ft/chara/ftKirby/ftKb_Init.h
index c8231afa66..e3c5ea4a75 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.h
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.h
@@ -45,7 +45,7 @@
 /* 0EF028 */ HSD_JObj* ftKb_Init_UnkMotionStates6(Fighter_GObj* gobj);
 /* 0EF040 */ void ftKb_SpecialN_800EF040(Fighter_GObj*, int, KirbyHatStruct*);
 /* 0EF0E4 */ UNK_RET ftKb_SpecialN_800EF0E4(Fighter_GObj*, int, u8*);
-/* 0EF35C */ UNK_RET ftKb_SpecialN_800EF35C(Fighter_GObj*, int, u8*);
+/* 0EF35C */ void ftKb_SpecialN_800EF35C(Fighter_GObj*, int, u8*);
 /* 0EF438 */ UNK_RET ftKb_SpecialN_800EF438(Fighter_GObj*, KirbyHatStruct*);
 /* 0EF69C */ void ftKb_SpecialN_800EF69C(Fighter_GObj*, int, KirbyHatStruct*);
 /* 0EF87C */ void ftKb_UnkIntBoolFunc0(Fighter*, int, bool);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index a9aacfd054..ad78c28e32 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -7,6 +7,8 @@
 #include "ef/efsync.h"
 #include "ft/chara/ftCommon/ftCo_Damage.h"
 #include "ft/chara/ftCommon/ftCo_FallSpecial.h"
+#include "ft/chara/ftCommon/ftCo_Throw.h"
+#include "ft/chara/ftCommon/ftCo_ThrownKirby.h"
 #include "ft/chara/ftCommon/ftCo_Jump.h"
 #include "ft/chara/ftCommon/ftCo_KneeBend.h"
 #include "ft/chara/ftCommon/ftCo_Wait.h"
@@ -17,6 +19,7 @@
 #include "ft/ft_081B.h"
 #include "ft/ft_0877.h"
 #include "ft/ft_0892.h"
+#include "ft/ftanim.h"
 #include "ft/ftcliffcommon.h"
 #include "ft/ftcommon.h"
 #include "ft/ftdata.h"
@@ -2023,13 +2026,63 @@ void ftKb_SpecialAirNCaptured_Anim(Fighter_GObj* gobj)
 
 /// #ftKb_SpecialAirNSpit_Anim
 
-/// #ftKb_SpecialNDrink0_Anim
+void ftKb_SpecialNDrink0_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = (0, GET_FIGHTER(gobj));
+    Item_GObj* item_gobj;
+    if (fp->cmd_vars[0] != 0 && (item_gobj = fp->target_item_gobj) != NULL) {
+        ftCommon_8007E2F4(fp, 0);
+        ft_PlaySFX(fp, 0x222F6, 0x7F, 0x40);
+        it_802F28C8(item_gobj, 0, 0.0F);
+        fp->x1A64 = NULL;
+        fp->target_item_gobj = NULL;
+        fp->cmd_vars[0] = 0;
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ft_8008A2BC(gobj);
+    }
+}
 
 /// #ftKb_SpecialNDrink_Anim
 
-/// #ftKb_SpecialNDrink1_Anim
+void ftKb_SpecialNDrink1_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = (0, GET_FIGHTER(gobj));
+    Item_GObj* item_gobj;
+    if (fp->cmd_vars[0] != 0 && (item_gobj = fp->target_item_gobj) != NULL) {
+        ftCommon_8007E2F4(fp, 0);
+        ft_PlaySFX(fp, 0x222F6, 0x7F, 0x40);
+        it_802F28C8(item_gobj, 0, 0.0F);
+        fp->x1A64 = NULL;
+        fp->target_item_gobj = NULL;
+        fp->cmd_vars[0] = 0;
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftCo_Fall_Enter(gobj);
+    }
+}
 
-/// #ftKb_SpecialAirNDrink_Anim
+void ftKb_SpecialAirNDrink_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp2;
+    Fighter_GObj* victim_gobj;
+    Fighter* fp = getFighter(gobj);
+    if (fp->cmd_vars[0] != 0) {
+        if ((victim_gobj = fp->victim_gobj) != NULL) {
+            ftCommon_8007E2F4(fp, 0);
+            ftCo_800DE2CC(gobj, victim_gobj);
+            ftCo_800BE000(victim_gobj, gobj);
+            fp2 = getFighter(gobj);
+            fp2->fv.kb.xE0 = ftCo_800BD9E0(gobj, victim_gobj);
+            ftKb_SpecialN_800F1BAC(gobj, fp2->fv.kb.xE0, true);
+            fp->cmd_vars[0] = 0;
+        }
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftCo_Fall_Enter(gobj);
+    }
+    PAD_STACK(32);
+}
 
 void ftKb_EatTurn_Anim(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
index 8c12684d8d..f2e84b06ec 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
@@ -986,7 +986,29 @@ void ftKb_SpecialNPr_80101618(Fighter_GObj* gobj)
 
 /// #ftKb_PrSpecialNEnd_Anim
 
-/// #ftKb_PrSpecialAirNStart_Anim
+void ftKb_PrSpecialAirNStart_Anim(Fighter_GObj* gobj)
+{
+    static u32 const mf = (1 << 1) | (1 << 4) | (1 << 18);
+    Fighter* fp = GET_FIGHTER(gobj);
+    PAD_STACK(8);
+    fp->mv.pr.specialn.facing_dir = 0;
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        Fighter_ChangeMotionState(gobj, ftKb_MS_PrSpecialAirNFull, mf, 0,
+                                  0, 0, NULL);
+        {
+            Fighter* fp = GET_FIGHTER(gobj);
+            fp->death2_cb = ftKb_Init_800EE74C;
+            fp->take_dmg_cb = ftKb_Init_800EE7B8;
+            fp->deal_dmg_cb = fn_80100E0C;
+            fp->x21F8 = fn_80105978;
+        }
+        fp->cur_anim_frame = 0;
+        ftAnim_SetAnimRate(gobj, 0);
+        fp->self_vel.x = fp->facing_dir * 0.0001f;
+        fp->x74_anim_vel.x = 0;
+        ftPartSetRotY(fp, 0, M_PI_2);
+    }
+}
 
 /// #ftKb_PrSpecialAirNLoop_Anim
 
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index dc57c8f86f..1517c50cb0 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -84,11 +84,13 @@ extern f32 ftKb_Init_804D9558;
 /// Forward declarations for functions called before definition
 void fn_80105AB0(Fighter_GObj*);
 void fn_80105A34(Fighter_GObj*);
+void fn_80106DB0(Fighter_GObj*);
 void fn_801090D4(Fighter_GObj*);
 static void fn_801095DC(HSD_GObj*);
 static void fn_80109680(HSD_GObj*);
 static void fn_80109714(HSD_GObj*);
 static void fn_801097B8(HSD_GObj*);
+static void fn_80106DB0(HSD_GObj*);
 void fn_8010A930(Fighter_GObj*, Fighter_GObj*);
 
 inline void ftKirbyDmgInline(Fighter_GObj* gobj)
@@ -490,7 +492,21 @@ void ftKb_SkSpecialAirNEnd_Anim(Fighter_GObj* gobj)
 
 void ftKb_SkSpecialNStart_IASA(Fighter_GObj* gobj) {}
 
-/// #ftKb_SkSpecialNLoop_IASA
+void ftKb_SkSpecialNLoop_IASA(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (!(fp->input.held_inputs & HSD_PAD_B)) {
+        fp->mv.kb.specialhi.x0 = 0;
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SkSpecialNEnd, Ft_MF_None,
+                                  0.0F, 1.0F, 0.0F, NULL);
+        ftKirbyDmgInline(gobj);
+        fp->accessory4_cb = fn_80106DB0;
+    } else if (fp->input.x668 & HSD_PAD_LR) {
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SkSpecialNCancel, Ft_MF_None,
+                                  0.0F, 1.0F, 0.0F, NULL);
+        ftKirbyDmgInline(gobj);
+    }
+}
 
 void ftKb_SkSpecialNCancel_IASA(Fighter_GObj* gobj) {}
 
@@ -498,7 +514,21 @@ void ftKb_SkSpecialNEnd_IASA(Fighter_GObj* gobj) {}
 
 void ftKb_SkSpecialAirNStart_IASA(Fighter_GObj* gobj) {}
 
-/// #ftKb_SkSpecialAirNLoop_IASA
+void ftKb_SkSpecialAirNLoop_IASA(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (!(fp->input.held_inputs & HSD_PAD_B)) {
+        fp->mv.kb.specialhi.x0 = 0;
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SkSpecialAirNEnd, Ft_MF_None,
+                                  0, 1, 0, NULL);
+        ftKirbyDmgInline(gobj);
+        fp->accessory4_cb = fn_80106DB0;
+    } else if (fp->input.x668 & HSD_PAD_LR) {
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SkSpecialAirNCancel,
+                                  Ft_MF_None, 0, 1, 0, NULL);
+        ftKirbyDmgInline(gobj);
+    }
+}
 
 void ftKb_SkSpecialAirNCancel_IASA(Fighter_GObj* gobj) {}
 
@@ -807,7 +837,48 @@ void ftKb_SpecialNMt_80107130(Fighter_GObj* gobj)
 
 /// #ftKb_SpecialNMt_80107410
 
-/// #ftKb_SpecialNMt_80107568
+static inline void ftKb_SpecialNMt_ChangeAction(Fighter_GObj* gobj)
+{
+    Fighter* fp = getFighter(gobj);
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    s32 timer;
+
+    Fighter_ChangeMotionState(gobj, ftKb_MS_MtSpecialNStart, 0, 0.0F, 1.0F,
+                              0.0F, NULL);
+
+    timer = 0;
+    fp->cmd_vars[3] = 0;
+    fp->cmd_vars[2] = 0;
+    fp->cmd_vars[1] = 0;
+    fp->cmd_vars[0] = 0;
+
+    ftCommon_8007D7FC(fp);
+
+    fp->self_vel.y = 0.0F;
+
+    {
+        Fighter* fp2 = gobj->user_data;
+        fp2->death2_cb = ftKb_Init_800EE74C;
+        fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+    }
+
+    fp->mv.kb.specialhi.x0 = 0;
+    fp->mv.kb.specialhi.x4 = 0;
+
+    if (fp->fv.kb.x9C == 0) {
+        timer = da->specialn_mt_frames_to_transition;
+    }
+    fp->mv.kb.specialhi.xC = timer;
+    fp->mv.kb.specialhi.x10.f = 0.0F;
+
+    ftAnim_8006EBA4(gobj);
+}
+
+void ftKb_SpecialNMt_80107568(Fighter_GObj* gobj)
+{
+    ftKb_SpecialNMt_ChangeAction(gobj);
+    PAD_STACK(8);
+}
 
 void ftKb_SpecialNMt_80107638(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
index 7fcebaf774..75119a87e9 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
@@ -1021,7 +1021,46 @@ void ftPp_SpecialHiThrow_0_Anim(Fighter_GObj* gobj)
     }
 }
 
-/// #ftPp_SpecialAirHiThrow_0_Anim
+void ftPp_SpecialAirHiThrow_0_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(40);
+
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
+    }
+
+    {
+        Fighter* fp = gobj->user_data;
+        if (fp->cmd_vars[1] != 0) {
+            fp->cmd_vars[1] = 0;
+            {
+                int found;
+                Fighter_GObj* nn_gobj =
+                    Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id,
+                                            1);
+                if (nn_gobj != NULL &&
+                    ftNn_Init_8012309C(nn_gobj) == 1)
+                {
+                    found = 1;
+                } else {
+                    found = 0;
+                }
+                if (found == 1) {
+                    ftPp_SpecialHi_8012280C(gobj);
+                    return;
+                }
+            }
+        }
+    }
+
+    {
+        Fighter* fp = GET_FIGHTER(gobj);
+        ++fp->mv.pp.unk_80123954.x0;
+        ftPp_SpecialS_80120FE0(gobj);
+    }
+}
 
 void ftPp_SpecialHiThrow_0_IASA(Fighter_GObj* gobj) {}
 
diff --git a/src/melee/ft/ftcolanim.c b/src/melee/ft/ftcolanim.c
index a1260b2044..c29dee48fd 100644
--- a/src/melee/ft/ftcolanim.c
+++ b/src/melee/ft/ftcolanim.c
@@ -125,7 +125,31 @@ void ft_800C0098(Fighter* fp)
 }
 #pragma pop
 
-/// #ftCo_800C0134
+#pragma push
+#pragma auto_inline off
+#pragma global_optimizer off
+void ftCo_800C0134(Fighter* fp)
+{
+    lb_80014498(&fp->x488);
+    if (ftData_UnkMotionStates4[fp->kind] != NULL) {
+        ftData_UnkMotionStates4[fp->kind](fp->gobj);
+    }
+    if (ftCo_800C53E4(fp) != 0) {
+        s32 arg1 = 0x6A;
+        if (arg1 >= 0x7B) {
+            s32 temp = arg1 - 0x7B;
+            lb_800144C8(&fp->x508, Fighter_804D6538, temp, 0);
+        } else {
+            Fighter_804D653C_t* entry = &Fighter_804D653C[arg1];
+            if (entry->unk5 != 0) {
+                lb_800144C8(&fp->x488, Fighter_804D653C, arg1, 0);
+            } else {
+                lb_800144C8(&fp->x408, Fighter_804D653C, arg1, 0);
+            }
+        }
+    }
+}
+#pragma pop
 
 #pragma push
 #pragma dont_inline on
diff --git a/src/melee/gm/gm_16AE.h b/src/melee/gm/gm_16AE.h
index 763e626128..436774f11e 100644
--- a/src/melee/gm/gm_16AE.h
+++ b/src/melee/gm/gm_16AE.h
@@ -114,6 +114,6 @@
 /* 16F088 */ void gm_8016F088(struct StartMeleeData*); /// Checks rumble?
 /* 16F120 */ int gm_8016F120(int);
 /* 16F140 */ void fn_8016F140(int);
-/* 16F160 */ void fn_8016F160(int, int);
+/* 16F160 */ int fn_8016F160(int, int);
 
 #endif
diff --git a/src/melee/gm/gm_16F1.h b/src/melee/gm/gm_16F1.h
index 9368481a5a..1f7950b8c8 100644
--- a/src/melee/gm/gm_16F1.h
+++ b/src/melee/gm/gm_16F1.h
@@ -16,7 +16,7 @@ struct lbl_804D65A8_t;
 
 /* 16F180 */ int fn_8016F180(int);
 /* 16F1B8 */ Gm_DecType gmDecisionGetType(int kind);
-/* 16F1F0 */ short fn_8016F1F0(int idx);
+/* 16F1F0 */ int fn_8016F1F0(int idx);
 /* 16F208 */ int gm_8016F208(int);
 /* 16F280 */ int fn_8016F280(int);
 /* 16F2F8 */ int gm_8016F2F8(int, unsigned char);
@@ -27,7 +27,7 @@ struct lbl_804D65A8_t;
 /* 16F870 */ UNK_RET fn_8016F870(UNK_PARAMS);
 /* 16F9A8 */ UNK_RET fn_8016F9A8(UNK_PARAMS);
 /* 16FAD4 */ int fn_8016FAD4(UNK_T, u16, int, u8);
-/* 16FFD4 */ UNK_RET fn_8016FFD4(UNK_PARAMS);
+/* 16FFD4 */ void fn_8016FFD4(struct lbl_8046B6A0_24C_t*, int, int);
 /* 170110 */ int fn_80170110(void*, int, int, u8);
 /* 1701A0 */ UNK_RET gm_801701A0(UNK_PARAMS);
 /* 1701AC */ UNK_RET fn_801701AC(UNK_PARAMS);
diff --git a/src/melee/gm/gmregclear.c b/src/melee/gm/gmregclear.c
index dfcef75289..7e4a8182dc 100644
--- a/src/melee/gm/gmregclear.c
+++ b/src/melee/gm/gmregclear.c
@@ -1129,7 +1129,32 @@ s32 fn_8017F14C(void* arg0)
     return 0;
 }
 
-/// #fn_8017F1B8
+void fn_8017F1B8(void)
+{
+    int i;
+    int mask;
+
+    for (i = 0; i < 256; i++) {
+        if (fn_8016F1F0(i) < 0xD7) {
+            mask = fn_8017F008();
+            if (mask & (u8) fn_8016F180(i)) {
+                if ((u32) fn_8016F160(0, i) != 0) {
+                    fn_8016F140(i);
+                }
+            }
+        } else {
+            mask = fn_8017F008();
+            if (mask & (u8) fn_8016F180(i)) {
+                if ((u32) fn_801701C0(gm_8016B774(), 0, i) != 0) {
+                    fn_8016F140(i);
+                }
+            }
+        }
+    }
+
+    mask = (u8) fn_8017F008();
+    fn_8016FFD4(gm_8016B774(), mask, 0);
+}
 
 int fn_8017F294(void)
 {
diff --git a/src/melee/gm/gmregclear.h b/src/melee/gm/gmregclear.h
index b2459d7e26..e1f37cb462 100644
--- a/src/melee/gm/gmregclear.h
+++ b/src/melee/gm/gmregclear.h
@@ -79,7 +79,7 @@
 /* 17F008 */ int fn_8017F008(void);
 /* 17F09C */ s32 fn_8017F09C(void);
 /* 17F14C */ s32 fn_8017F14C(void* arg0);
-/* 17F1B8 */ UNK_RET fn_8017F1B8(UNK_PARAMS);
+/* 17F1B8 */ void fn_8017F1B8(void);
 /* 17F2A4 */ UNK_RET fn_8017F2A4(UNK_PARAMS);
 /* 17F47C */ UNK_RET fn_8017F47C(UNK_PARAMS);
 /* 17F608 */ UNK_RET fn_8017F608(UNK_PARAMS);
diff --git a/src/melee/gr/granime.c b/src/melee/gr/granime.c
index e54370a02b..738de56305 100644
--- a/src/melee/gr/granime.c
+++ b/src/melee/gr/granime.c
@@ -10,6 +10,7 @@
 #include "lb/lbarchive.h"
 #include "lb/lbheap.h"
 
+#include <baselib/aobj.h>
 #include <baselib/dobj.h>
 #include <baselib/mobj.h>
 #include <baselib/robj.h>
@@ -757,7 +758,14 @@ HSD_AObj* grAnime_801C8318(HSD_GObj* gobj, int arg1, u32 arg2)
     return sp14;
 }
 
-/// #grAnime_801C83D0
+bool grAnime_801C83D0(HSD_GObj* gobj, bool arg1, enum_t arg2)
+{
+    HSD_AObj* aobj = grAnime_801C8318(gobj, arg1, arg2);
+    if (HSD_AObjGetFlags(aobj) & 0x40000000) {
+        return true;
+    }
+    return false;
+}
 
 /// #grAnime_801C84A4
 
diff --git a/src/melee/gr/grcorneria.c b/src/melee/gr/grcorneria.c
index 3a67f9250b..9ec173feca 100644
--- a/src/melee/gr/grcorneria.c
+++ b/src/melee/gr/grcorneria.c
@@ -300,7 +300,59 @@ void grCorneria_801DDD4C(Vec3* vec)
     vec->z = 0.0f;
 }
 
-/// #grCorneria_801DDDA8
+typedef struct {
+    int data[3];
+} grCn_Entry;
+
+extern grCn_Entry grCn_803E2204[][5];
+
+typedef struct grCn_Data {
+    /* 0x000 */ u8 pad[0x4CC];
+    /* 0x4CC */ grCn_Entry entries[][5];
+} grCn_Data;
+
+extern grCn_Data grCn_803E1D38;
+
+void grCorneria_801DDDA8(HSD_GObj* gobj, Vec3* vec)
+{
+    grCn_Data* data = &grCn_803E1D38;
+    Ground* gp;
+    Ground* corn_gp;
+    HSD_JObj* jobj;
+    u8 unused[12];
+    Vec3 pos;
+    int idx;
+    PAD_STACK(4);
+
+    if (gobj != NULL) {
+        gp = gobj->user_data;
+        corn_gp = Ground_801C2BA4(3)->user_data;
+        jobj = Ground_801C3FA4(gobj, 4);
+        lb_8000B1CC(jobj, NULL, &pos);
+        idx = ((int*) &data->pad[0x3C])[gp->gv.corneria.xC8];
+        vec->x =
+            corn_gp->gv.corneria.offset_x +
+            (-pos.z +
+             *(f32*) &data->pad[0x330 +
+                                idx * (int) sizeof(grCn_Entry)]);
+        vec->y =
+            pos.y +
+            *(f32*) &data->pad[0x334 +
+                               idx * (int) sizeof(grCn_Entry)];
+        vec->z =
+            pos.x +
+            *(f32*) &data->pad[0x338 +
+                               idx * (int) sizeof(grCn_Entry)];
+    } else {
+        vec->z = 0.0f;
+        vec->y = 0.0f;
+        vec->x = 0.0f;
+    }
+
+    if (vec->y <= 100.0f) {
+        vec->y = 100.0f;
+    }
+}
 
 /// #grCorneria_801DDE88
 
@@ -709,19 +761,6 @@ void smashTaunt_801E2550(Ground_GObj* gobj, struct grSmashTaunt_GroundVars* gv)
     gobj->render_cb = smashTaunt_801E24AC;
 }
 
-typedef struct {
-    int data[3];
-} grCn_Entry;
-
-extern grCn_Entry grCn_803E2204[][5];
-
-typedef struct grCn_Data {
-    /* 0x000 */ u8 pad[0x4CC];
-    /* 0x4CC */ grCn_Entry entries[][5];
-} grCn_Data;
-
-extern grCn_Data grCn_803E1D38;
-
 s32 grCorneria_801E2598(u32 arg0, u32 arg1)
 {
     int val;
diff --git a/src/melee/gr/grcorneria.h b/src/melee/gr/grcorneria.h
index fd500b83aa..066ef860da 100644
--- a/src/melee/gr/grcorneria.h
+++ b/src/melee/gr/grcorneria.h
@@ -29,7 +29,7 @@
 /* 1DDCE8 */ bool grCorneria_801DDCE8(Ground_GObj*);
 /* 1DDCF0 */ void grCorneria_801DDCF0(Vec3*);
 /* 1DDD4C */ void grCorneria_801DDD4C(Vec3*);
-/* 1DDDA8 */ UNK_RET grCorneria_801DDDA8(UNK_PARAMS);
+/* 1DDDA8 */ void grCorneria_801DDDA8(HSD_GObj*, Vec3*);
 /* 1DDE88 */ UNK_RET grCorneria_801DDE88(UNK_PARAMS);
 /* 1DE024 */ void grCorneria_801DE024(Ground_GObj*);
 /* 1DE4C0 */ void grCorneria_Arwing_801DE4C0(Ground_GObj*);
diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index e21290f9ee..af3be585a4 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -12,6 +12,7 @@
 #include "lb/lb_00B0.h"
 #include "lb/lblanguage.h"
 
+f32 grHr_804D6AD8;
 int grHr_804D6ADC;
 f32 grHr_804D6AE0;
 f32 grHr_804D6AE4;
@@ -480,7 +481,35 @@ void grHomeRun_8021ED74(void)
     it_8026BE84(&bobomb_rain);
 }
 
-/// #grHomeRun_8021EDD4
+void grHomeRun_8021EDD4(int arg0)
+{
+    f32 scale;
+    f32 max_dist;
+    f32 ratio;
+    f32 num_ticks;
+    f32 dist;
+
+    if (lbLang_IsSavedLanguageJP()) {
+        scale = 0.1F;
+        max_dist = 50.0F;
+    } else {
+        scale = 0.3280969F;
+        max_dist = 100.0F;
+    }
+
+    ratio = max_dist / scale;
+    num_ticks = (int) (0.5F + ratio / (160.0F * Ground_801C0498()));
+
+    if (num_ticks < 1.0F) {
+        num_ticks = 1.0F;
+    }
+
+    dist = 160.0F * Ground_801C0498();
+    grHr_804D6AD8 = scale;
+    grHr_804D6AE0 = max_dist;
+    grHr_804D6ADC = num_ticks;
+    grHr_804D6AE4 = ratio / (num_ticks * dist);
+}
 
 DynamicsDesc* grHomeRun_8021EEB4(enum_t arg)
 {
diff --git a/src/melee/gr/grkongo.c b/src/melee/gr/grkongo.c
index 185ecbf0f4..b06b26d03a 100644
--- a/src/melee/gr/grkongo.c
+++ b/src/melee/gr/grkongo.c
@@ -1411,7 +1411,37 @@ void fn_801D542C(HSD_GObj* arg0)
 
 /// #grKongo_801D7134
 
-/// #fn_801D7700
+static void fn_801D7700(Ground* gp, s32 arg1, CollData* cd, s32 arg3,
+                        mpLib_GroundEnum arg4, float arg8)
+{
+    Vec3* cur = &cd->cur_pos;
+    s32 type = cd->x34_flags.b1234;
+
+    if (type == 1 || type == 3) {
+        if (arg4 == mpLib_GroundEnum_Unk1) {
+            f32 x1, y1, x2, y2;
+
+            mpVtxGetPos(0x1D, &x1, &y1);
+            mpVtxGetPos(0x1A, &x2, &y2);
+
+            {
+                f32 segment_size = (x2 - x1) / 15.0F;
+                f32 ratio = (cur->x - x1) / segment_size;
+                s32 idx = (s32) ratio;
+                s32 idx2 = (s32) ratio;
+
+                if (idx < 0) {
+                    idx2 = 0;
+                } else if ((u32) idx2 >= 15) {
+                    idx2 = 14;
+                }
+
+                grKg_803E188C[idx2].unk10 +=
+                    0.017453292F * grKg_804D6980->unkA4;
+            }
+        }
+    }
+}
 
 /// #grKongo_801D77E0
 
@@ -1427,7 +1457,39 @@ void fn_801D7E60(Ground* gp, s32 arg1, CollData* arg2, s32 arg3,
 
 /// #grKongo_801D7E78
 
-/// #grKongo_801D7F78
+bool grKongo_801D7F78(HSD_GObj* gobj)
+{
+    Vec3 pos1;
+    Vec3 pos2;
+    HSD_GObj* cur;
+    PAD_STACK(16);
+
+    if (grKongo_801D7E78(gobj, &pos1) == NULL) {
+        return false;
+    }
+
+    for (cur = HSD_GObj_Entities->x14; cur != NULL; cur = cur->next) {
+        if (cur == gobj) {
+            continue;
+        }
+        if (grKongo_801D7E78(cur, &pos2) == NULL) {
+            continue;
+        }
+        {
+            f32 dz = pos2.z - pos1.z;
+            if (dz < -100.0F) {
+                if (dz > -150.0F) {
+                    f32 dx = pos2.x - pos1.x;
+                    if (dx * dx < 6400.0F) {
+                        return true;
+                    }
+                }
+            }
+        }
+    }
+
+    return false;
+}
 
 void grKongo_801D8058(Ground_GObj* arg)
 {
diff --git a/src/melee/gr/grkongo.h b/src/melee/gr/grkongo.h
index 77664a612f..bcc21fc777 100644
--- a/src/melee/gr/grkongo.h
+++ b/src/melee/gr/grkongo.h
@@ -52,7 +52,7 @@
 /* 1D7134 */ void grKongo_801D7134(HSD_GObj*, s32);
 /* 1D77E0 */ void grKongo_801D77E0(HSD_GObj*, s32);
 /* 1D7BBC */ void grKongo_801D7BBC(HSD_GObj*);
-/* 1D7E78 */ UNK_RET grKongo_801D7E78(UNK_PARAMS);
+/* 1D7E78 */ Vec3* grKongo_801D7E78(HSD_GObj*, Vec3*);
 /* 1D7F78 */ bool grKongo_801D7F78(HSD_GObj*);
 /* 1D8058 */ void grKongo_801D8058(Ground_GObj*);
 /* 1D8078 */ HSD_GObj* grKongo_801D8078(HSD_GObj*);
diff --git a/src/melee/gr/grkongo.static.h b/src/melee/gr/grkongo.static.h
index 275da089f2..86f092a127 100644
--- a/src/melee/gr/grkongo.static.h
+++ b/src/melee/gr/grkongo.static.h
@@ -61,6 +61,9 @@ static struct {
     f32 unk90;
     f32 unk94;
     f32 unk98;
+    f32 unk9C;
+    f32 unkA0;
+    f32 unkA4;
 }* grKg_804D6980;
 
 extern StageInfo stage_info;
diff --git a/src/melee/gr/groldpupupu.c b/src/melee/gr/groldpupupu.c
index 6be94088de..d2d147b081 100644
--- a/src/melee/gr/groldpupupu.c
+++ b/src/melee/gr/groldpupupu.c
@@ -22,6 +22,10 @@ static struct {
 
 static void* grOp_804D6A9C;
 
+static int grOp_804D48A0[1][2] = { { 4, 0 } };
+static int grOp_804D48A8[1][2] = { { 5, 1 } };
+static int grOp_804D48B0[1][2] = { { 2, 3 } };
+
 static int grOp_804D48E0[1][2] = { { 0, 0 } };
 
 StageCallbacks grOp_803E6688[9] = {
@@ -111,7 +115,31 @@ bool grOldPupupu_80210A1C(Ground_GObj* gobj)
     return false;
 }
 
-/// #grOldPupupu_80210A24
+void grOldPupupu_80210A24(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+
+    if ((int) grOp_804D6A9C == 0 && gp->gv.unk.xD0 == 1) {
+        gp->gv.unk.xD0 = 0;
+        switch (gp->gv.unk.xC8) {
+        case 0:
+            grAnime_801C8138(
+                gobj, gp->map_id,
+                grOp_804D48A0[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+            break;
+        case 1:
+            grAnime_801C8138(
+                gobj, gp->map_id,
+                grOp_804D48B0[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+            break;
+        case 2:
+            grAnime_801C8138(
+                gobj, gp->map_id,
+                grOp_804D48A8[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+            break;
+        }
+    }
+}
 
 void grOldPupupu_80210B00(Ground_GObj* arg) {}
 
diff --git a/src/melee/gr/grrcruise.c b/src/melee/gr/grrcruise.c
index fac04be647..b85f3622de 100644
--- a/src/melee/gr/grrcruise.c
+++ b/src/melee/gr/grrcruise.c
@@ -1,5 +1,6 @@
 #include "grrcruise.h"
 
+#include "gr/granime.h"
 #include "grzakogenerator.h"
 
 #include <platform.h>
@@ -44,6 +45,7 @@ StageCallbacks grRc_803E4E34[7] = {
 };
 
 extern s16 grRc_803E4FF0[];
+extern s16 grRc_804D4790[4];
 
 static struct {
     int x0;
@@ -291,7 +293,25 @@ bool grRCruise_801FFAD4(Ground_GObj* arg)
 
 void grRCruise_80200070(Ground_GObj* arg) {}
 
-/// #grRCruise_80200074
+void grRCruise_80200074(Ground_GObj* gobj)
+{
+    int i;
+    Ground* gp = gobj->user_data;
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    Ground_801C2ED0(jobj, gp->map_id);
+    grAnime_801C8138(gobj, gp->map_id, 0);
+    for (i = 0; i < 3; i++) {
+        grAnime_801C8098(gobj, grRc_804D4790[i], 7, 1, 0.0F, 0.0F);
+        gp->gv.rcruise.x3C[i].x02 =
+            Ground_801C32D4(gp->map_id, grRc_804D4790[i]);
+        mpJointSetCb1(gp->gv.rcruise.x3C[i].x02, gobj,
+                      (mpLib_Callback) fn_80200460);
+        gp->gv.rcruise.x3C[i].x0C = Ground_801C3FA4(gobj, grRc_804D4790[i]);
+        gp->gv.rcruise.x3C[i].x08 = 0;
+        gp->gv.rcruise.x3C[i].x04 = 0;
+        gp->gv.rcruise.x3C[i].x00 = 0;
+    }
+}
 
 bool grRCruise_8020014C(Ground_GObj* arg)
 {
@@ -302,7 +322,39 @@ bool grRCruise_8020014C(Ground_GObj* arg)
 
 void grRCruise_8020045C(Ground_GObj* arg) {}
 
-/// #fn_80200460
+inline Ground* fn_80200460_inline(Ground_GObj* gobj)
+{
+    return GET_GROUND(gobj);
+}
+
+void fn_80200460(Ground_GObj* gobj, s32 id, CollData* coll)
+{
+    Ground* gp = fn_80200460_inline(gobj);
+    s32 i;
+    PAD_STACK(24);
+
+    if ((s32) coll->x34_flags.b1234 != 1) {
+        return;
+    }
+
+    for (i = 0; i < 3; i++) {
+        if (gp->gv.rcruise.x3C[i].x02 == id) {
+            if (gp->gv.rcruise.x3C[i].x00 == 1 ||
+                gp->gv.rcruise.x3C[i].x00 == 3 ||
+                gp->gv.rcruise.x3C[i].x00 == 4)
+            {
+                gp->gv.rcruise.x3C[i].x04 = 0;
+                grRCruise_80201B60(
+                    gp->gv.rcruise.x3C[i].x0C->child, 1);
+                grAnime_801C7A94(gobj, grRc_804D4790[i], 1,
+                                 1.0F);
+                gp->gv.rcruise.x3C[i].x00 = 2;
+            }
+            gp->gv.rcruise.x3C[i].x08++;
+            return;
+        }
+    }
+}
 
 void grRCruise_80200540(Ground_GObj* gobj)
 {
diff --git a/src/melee/gr/grrcruise.h b/src/melee/gr/grrcruise.h
index 6965344361..b1dc31b635 100644
--- a/src/melee/gr/grrcruise.h
+++ b/src/melee/gr/grrcruise.h
@@ -42,7 +42,7 @@
 /* 20014C */ bool grRCruise_8020014C(Ground_GObj*);
 /* 200154 */ void grRCruise_80200154(Ground_GObj*);
 /* 20045C */ void grRCruise_8020045C(Ground_GObj*);
-/* 200460 */ UNK_RET fn_80200460(UNK_PARAMS);
+/* 200460 */ void fn_80200460(Ground_GObj*, s32, CollData*);
 /* 200540 */ void grRCruise_80200540(Ground_GObj*);
 /* 200578 */ void grRCruise_80200578(Ground*, s32, CollData*, s32,
                                      mpLib_GroundEnum, f32);
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index fe46d02252..0e7bf77616 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -738,6 +738,15 @@ struct grZebes_GroundVars5 {
     /* +3C gp+100 */ u32 x100;
 };
 
+struct grRCruise_SubEntry {
+    /* 0x00 */ u8 x00;
+    /* 0x01 */ u8 pad_01;
+    /* 0x02 */ s16 x02;
+    /* 0x04 */ s32 x04;
+    /* 0x08 */ s32 x08;
+    /* 0x0C */ HSD_JObj* x0C;
+};
+
 struct grRCruise_Entry {
     /* 0x00 */ u8 x00;
     /* 0x01 */ u8 pad_01;
@@ -749,6 +758,15 @@ struct grRCruise_Entry {
     /* 0x14 */ HSD_JObj* x14;
 };
 
+struct grRCruise_SubEntry {
+    /* 0x00 */ u8 x00;
+    /* 0x01 */ u8 pad_01;
+    /* 0x02 */ s16 x02;
+    /* 0x04 */ s32 x04;
+    /* 0x08 */ s32 x08;
+    /* 0x0C */ HSD_JObj* x0C;
+};
+
 struct grRCruise_GroundVars {
     struct {
         u8 b0 : 1;
@@ -773,7 +791,7 @@ struct grRCruise_GroundVars {
     /* +30 gp+F4 */ s32 x30;
     /* +34 gp+F8 */ s32 x34;
     /* +38 gp+FC */ s32 x38;
-    /* +3C gp+100 */ char pad_3C[0x30];
+    /* +3C gp+100 */ struct grRCruise_SubEntry x3C[3];
     /* +6C gp+130 */ struct grRCruise_Entry* entries;
 };
 
diff --git a/src/melee/it/items/itkamex.c b/src/melee/it/items/itkamex.c
index 41dffcff21..6578b6dd2e 100644
--- a/src/melee/it/items/itkamex.c
+++ b/src/melee/it/items/itkamex.c
@@ -96,7 +96,37 @@ void it_802CA6A0(Item_GObj* gobj)
     ip->on_accessory = it_802CA8DC;
 }
 
-/// #itKamex_UnkMotion1_Anim
+bool itKamex_UnkMotion1_Anim(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    PAD_STACK(8);
+
+    if (!it_80272C6C(gobj)) {
+        if (ip->xDD4_itemVar.pokemon.timer <= 0) {
+            ip->xDD4_itemVar.pokemon.timer = 0;
+            ip->x40_vel.x = 0.0f;
+            it_802CA938(gobj);
+        } else {
+            ip->xDD4_itemVar.pokemon.timer--;
+            {
+                Item* ip = GET_ITEM(gobj);
+                itKamexAttributes* attrs =
+                    ip->xC4_article_data->x4_specialAttributes;
+
+                if (ip->xDD4_itemVar.pokemon.timer == -1) {
+                    ip->xDD4_itemVar.pokemon.timer = attrs->timer;
+                    ip->xDD4_itemVar.pokemon.x68 = attrs->x18;
+                    ip->xDD4_itemVar.pokemon.x6C = attrs->x1C;
+                }
+            }
+            Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
+            ip->entered_hitlag = efLib_PauseAll;
+            ip->exited_hitlag = efLib_ResumeAll;
+            ip->on_accessory = it_802CA8DC;
+        }
+    }
+    return false;
+}
 
 void itKamex_UnkMotion1_Phys(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itkirbycutterbeam.c b/src/melee/it/items/itkirbycutterbeam.c
index a2447b916b..08adfc1f75 100644
--- a/src/melee/it/items/itkirbycutterbeam.c
+++ b/src/melee/it/items/itkirbycutterbeam.c
@@ -1,14 +1,38 @@
 #include "itkirbycutterbeam.h"
 
+#include "db/db.h"
+#include "it/it_26B1.h"
+#include "it/it_2725.h"
+#include "it/item.h"
+#include "it/types.h"
+
 #include <placeholder.h>
 #include <platform.h>
 
-#include "it/forward.h"
+void it_8029BAB8(HSD_GObj* gobj, Vec3* pos, float facing_dir)
+{
+    Item_GObj* spawned_gobj;
 
-#include "it/it_26B1.h"
-#include "it/it_2725.h"
+    SpawnItem spawn;
+    pos->z = 0.0f;
+    spawn.kind = It_Kind_Kirby_CBeam;
+    spawn.prev_pos = *pos;
+    it_8026BB68(gobj, &spawn.pos);
+    spawn.facing_dir = facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.x = spawn.vel.y = spawn.vel.z = 0.0f;
+    spawn.x0_parent_gobj = gobj;
+    spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+    spawn.x44_flag.b0 = true;
+    spawn.x40 = 0;
 
-/// #it_8029BAB8
+    spawned_gobj = Item_80268B18(&spawn);
+    if (spawned_gobj != NULL) {
+        it_8029BB90(spawned_gobj, gobj);
+        db_80225DD8(spawned_gobj, gobj);
+        it_802750F8(spawned_gobj);
+    }
+}
 
 /// #it_8029BB90
 
diff --git a/src/melee/it/items/itkirbycutterbeam.h b/src/melee/it/items/itkirbycutterbeam.h
index 98d0576ba4..cc334541f0 100644
--- a/src/melee/it/items/itkirbycutterbeam.h
+++ b/src/melee/it/items/itkirbycutterbeam.h
@@ -8,7 +8,7 @@
 #include "it/items/types.h"
 
 /* 29BAB8 */ void it_8029BAB8(HSD_GObj* gobj, Vec3* pos, float facing_dir);
-/* 29BB90 */ UNK_RET it_8029BB90(UNK_PARAMS);
+/* 29BB90 */ void it_8029BB90(Item_GObj* gobj, HSD_GObj* owner);
 /* 29BC88 */ bool itKirbycutterbeam_UnkMotion0_Anim(Item_GObj* gobj);
 /* 29BCA8 */ void itKirbycutterbeam_UnkMotion0_Phys(Item_GObj* gobj);
 /* 29BDD4 */ bool itKirbycutterbeam_UnkMotion0_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/itlugia.c b/src/melee/it/items/itlugia.c
index 7254ddc1a2..791b22443b 100644
--- a/src/melee/it/items/itlugia.c
+++ b/src/melee/it/items/itlugia.c
@@ -267,7 +267,23 @@ bool it_802D1E64(Item_GObj* gobj)
     return it_8027A118(gobj, (void (*)(HSD_GObj*)) it_802D14D0);
 }
 
-/// #it_802D1E8C
+Item_GObj* it_802D1E8C(Item_GObj* gobj, ItemKind kind, f32 param)
+{
+    SpawnItem spawn;
+    Item* ip = GET_ITEM(gobj);
+
+    spawn.prev_pos = *(Vec3*) ip->xDD4_itemVar.lugia.x70_pad;
+    it_8026BB88(gobj, &spawn.pos);
+    spawn.facing_dir = ip->facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel = it_802D1F64(gobj, param);
+    spawn.kind = kind;
+    spawn.x0_parent_gobj = ip->owner;
+    spawn.x4_parent_gobj2 = gobj;
+    spawn.x44_flag.b0 = 1;
+    spawn.x40 = 0;
+    return Item_80268B18(&spawn);
+}
 
 /// #it_802D1F64
 
diff --git a/src/melee/it/items/itlugia.h b/src/melee/it/items/itlugia.h
index 94986ad0e1..569d55ffca 100644
--- a/src/melee/it/items/itlugia.h
+++ b/src/melee/it/items/itlugia.h
@@ -34,8 +34,8 @@
 /* 2D1DB4 */ bool it_802D1DB4(Item_GObj* gobj);
 /* 2D1DD8 */ void it_802D1DD8(Item_GObj* gobj);
 /* 2D1E64 */ bool it_802D1E64(Item_GObj* gobj);
-/* 2D1E8C */ UNK_RET it_802D1E8C(UNK_PARAMS);
-/* 2D1F64 */ UNK_RET it_802D1F64(UNK_PARAMS);
+/* 2D1E8C */ Item_GObj* it_802D1E8C(Item_GObj*, ItemKind, f32);
+/* 2D1F64 */ Vec3 it_802D1F64(Item_GObj*, f32);
 /* 2D208C */ UNK_RET it_802D208C(UNK_PARAMS);
 /* 2D2308 */ void itLugia_Logic39_Spawned(Item_GObj*);
 /* 2D234C */ void itLugia_Logic40_Spawned(Item_GObj*);
diff --git a/src/melee/it/items/itoldkuri.c b/src/melee/it/items/itoldkuri.c
index c8c1bfb04a..d74644a096 100644
--- a/src/melee/it/items/itoldkuri.c
+++ b/src/melee/it/items/itoldkuri.c
@@ -153,7 +153,28 @@ bool itOldkuri_UnkMotion2_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itOldkuri_UnkMotion2_Phys
+void itOldkuri_UnkMotion2_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(8);
+    if (ip->xDD4_itemVar.oldkuri.xDFC != 0 && it_802750E8(gobj, 2) != 0) {
+        Item* ip = GET_ITEM(gobj);
+        itOldkuriAttributes* attr =
+            ip->xC4_article_data->x4_specialAttributes;
+        ip->xDD4_itemVar.oldkuri.xDF4 =
+            ip->facing_dir * ((f32*) attr->x0)[1];
+        ip->x40_vel.x = ip->xDD4_itemVar.oldkuri.xDF4;
+        ip->x40_vel.z = 0.0f;
+        ip->x40_vel.y = 0.0f;
+        ip->facing_dir = -1.0f;
+        ip->xDD4_itemVar.oldkuri.xDFC = 0;
+        it_802D848C(gobj, 1, ITEM_ANIM_UPDATE);
+    }
+    ip->x40_vel.x = ip->xDD4_itemVar.oldkuri.xDF4;
+    it_8027C8D0(&ip->x40_vel, &ip->x378_itemColl.floor.normal,
+                ip->facing_dir);
+    it_8027C0A8(gobj, 0.0f, 5.0f);
+}
 
 bool itOldkuri_UnkMotion2_Coll(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itpikachuthunder.c b/src/melee/it/items/itpikachuthunder.c
index 7d79466f24..60aee13126 100644
--- a/src/melee/it/items/itpikachuthunder.c
+++ b/src/melee/it/items/itpikachuthunder.c
@@ -4,6 +4,7 @@
 
 #include "ft/chara/ftPikachu/ftPk_SpecialLw.h"
 #include "it/inlines.h"
+#include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
@@ -90,7 +91,39 @@ bool itPikachuthunder_UnkMotion1_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itPikachuthunder_UnkMotion1_Coll
+bool itPikachuthunder_UnkMotion1_Coll(Item_GObj* gobj)
+{
+    if (GET_ITEM(gobj)->xDD4_itemVar.pikachuthunder.x0 == 0) {
+        if (it_8026DA70(gobj)) {
+            Item* ip = GET_ITEM(gobj);
+            itPikachuthunderAttributes* attrs =
+                ip->xC4_article_data->x4_specialAttributes;
+            Vec3 pos;
+            pos.z = 0.0f;
+            pos.x = 0.0f;
+            pos.y = -attrs->x8;
+            it_80272AC4(gobj, &pos);
+            it_802B22B8(gobj);
+            if (ip->xDD4_itemVar.pikachuthunder.x38 != NULL) {
+                if (!ftPk_SpecialLw_CheckProperty(
+                        ip->xDD4_itemVar.pikachuthunder.x38))
+                {
+                    ftPk_SpecialLw_80127608(
+                        ip->xDD4_itemVar.pikachuthunder.x38);
+                }
+            }
+        }
+    } else {
+        Item* ip = GET_ITEM(gobj);
+        if (ip->xDD4_itemVar.pikachuthunder.x4 != 0) {
+            if (ip->pos.y <= ip->xDD4_itemVar.pikachuthunder.x28.y) {
+                it_802B22B8(gobj);
+            }
+        }
+    }
+    PAD_STACK(8);
+    return false;
+}
 
 s32 it_802B22B8(Item_GObj* gobj)
 {
diff --git a/src/melee/lb/lbmthp.c b/src/melee/lb/lbmthp.c
index e6d44b7fa0..8add150cb3 100644
--- a/src/melee/lb/lbmthp.c
+++ b/src/melee/lb/lbmthp.c
@@ -94,7 +94,40 @@ s32 fn_8001EBF0(THPDecComp* data)
 
 /// #fn_8001EF5C
 
-/// #fn_8001F06C
+s32 fn_8001F06C(THPDecComp* data)
+{
+    BOOL intr;
+
+    intr = OSDisableInterrupts();
+
+    if (data->unk_108 > 0 || data->unk_6C == 0) {
+        data->unk_78++;
+
+        if (data->unk_78 == data->unk_40) {
+            if (data->unk_68 != 0) {
+                data->unk_78 = 0;
+            } else {
+                data->unk_78--;
+                goto end;
+            }
+        }
+
+        data->unk_88++;
+        data->unk_108--;
+
+        if (data->unk_88 >= data->unk_104) {
+            data->unk_88 = 0;
+        }
+    }
+
+    if (data->unk_108 < data->unk_10C) {
+        data->unk_10C = data->unk_108;
+    }
+
+end:
+    OSRestoreInterrupts(intr);
+    return 1;
+}
 
 /// #fn_8001F13C
 
diff --git a/src/melee/lb/lbmthp.h b/src/melee/lb/lbmthp.h
index 3f409e4b9a..b4ed3145ed 100644
--- a/src/melee/lb/lbmthp.h
+++ b/src/melee/lb/lbmthp.h
@@ -17,8 +17,9 @@ typedef struct THPDecComp {
     /* 0x44 */ u32 width;
     /* 0x48 */ u32 height;
     /* 0x4C */ u8 pad4C[0x68 - 0x4C];
-    /* 0x68 */ u32 unk_68;
-    /* 0x6C */ u8 pad6C[0x78 - 0x6C];
+    /* 0x68 */ s32 unk_68;
+    /* 0x6C */ s32 unk_6C;
+    /* 0x70 */ u8 pad70[0x78 - 0x70];
     /* 0x78 */ u32 unk_78;
     /* 0x7C */ u32 unk_7C;
     /* 0x80 */ u32 unk_80;
@@ -39,6 +40,8 @@ typedef struct THPDecComp {
     /* 0xAD */ u8 padAD[0x100 - 0xAD];
     /* 0x100 */ u32 unk_100;
     /* 0x104 */ u32 unk_104;
+    /* 0x108 */ s32 unk_108;
+    /* 0x10C */ s32 unk_10C;
 } THPDecComp;
 
 /* 01E910 */ UNK_RET fn_8001E910(UNK_PARAMS);
@@ -46,7 +49,7 @@ typedef struct THPDecComp {
 /* 01EBF0 */ s32 fn_8001EBF0(THPDecComp* data);
 /* 01ECF4 */ UNK_RET fn_8001ECF4(UNK_PARAMS);
 /* 01EF5C */ UNK_RET fn_8001EF5C(UNK_PARAMS);
-/* 01F06C */ UNK_RET fn_8001F06C(UNK_PARAMS);
+/* 01F06C */ s32 fn_8001F06C(THPDecComp* data);
 /* 01F13C */ UNK_RET fn_8001F13C(UNK_PARAMS);
 /* 01F294 */ s32 fn_8001F294(void);
 /* 01F2A4 */ UNK_RET fn_8001F2A4(UNK_PARAMS);
diff --git a/src/melee/mn/mnname.c b/src/melee/mn/mnname.c
index 6b368ee4f1..d525a0ceeb 100644
--- a/src/melee/mn/mnname.c
+++ b/src/melee/mn/mnname.c
@@ -245,7 +245,37 @@ f32 mnName_80238964(u8 index, u8 target, u8 flag)
     }
 }
 
-/// #mnName_80238A04
+#pragma dont_inline on
+void mnName_80238A04(HSD_GObj* gobj, u8 target, u8 flag)
+{
+    AnimLoopSettings* base = mnName_803ED538;
+    HSD_JObj* jobj;
+    HSD_JObj* jobj2;
+    int i;
+
+    for (i = 0x18; i < 0x1B; i++) {
+        jobj = mnName_802388D4(gobj, i);
+        HSD_JObjReqAnimAll(jobj, mnName_80238964(i, target, flag));
+        HSD_JObjAnimAll(jobj);
+    }
+
+    jobj2 = (HSD_JObj*) gobj->prev_gx;
+
+    if ((u8) target == 0x18) {
+        if ((u8) flag) {
+            HSD_JObjReqAnimAll(jobj2, base[5].start_frame);
+        } else {
+            HSD_JObjReqAnimAll(jobj2, base[4].start_frame);
+        }
+    } else {
+        if ((u8) flag) {
+            HSD_JObjReqAnimAll(jobj2, base[8].start_frame);
+        } else {
+            HSD_JObjReqAnimAll(jobj2, base[6].start_frame);
+        }
+    }
+    HSD_JObjAnimAll(jobj2);
+}
 
 /// #mnName_80238AE0
 
diff --git a/src/melee/mn/mnname.h b/src/melee/mn/mnname.h
index 2a308f8591..8bf8dfa2fb 100644
--- a/src/melee/mn/mnname.h
+++ b/src/melee/mn/mnname.h
@@ -37,7 +37,7 @@ typedef struct MnName_GObj {
 /* 238754 */ void mnName_80238754(HSD_GObj* gobj);
 /* 2388D4 */ HSD_JObj* mnName_802388D4(HSD_GObj* gobj, u8 index);
 /* 238964 */ f32 mnName_80238964(u8 index, u8 target, u8 flag);
-/* 238A04 */ UNK_RET mnName_80238A04(UNK_PARAMS);
+/* 238A04 */ void mnName_80238A04(HSD_GObj* gobj, u8 target, u8 flag);
 /* 238AE0 */ UNK_RET mnName_80238AE0(UNK_PARAMS);
 /* 238C34 */ UNK_RET mnName_80238C34(UNK_PARAMS);
 /* 239574 */ UNK_RET fn_80239574(UNK_PARAMS);
diff --git a/src/melee/pl/pltrick.c b/src/melee/pl/pltrick.c
index baf1bc90dc..41c4ea6ef4 100644
--- a/src/melee/pl/pltrick.c
+++ b/src/melee/pl/pltrick.c
@@ -36,7 +36,30 @@ void pl_80037BC0(struct plAttackStats* stats, union Struct2070* ev)
 
 /// #pl_80037C60
 
-/// #pl_80037DF4
+void pl_80037DF4(HSD_GObj* gobj, union Struct2070* ev)
+{
+    Fighter* ft = GET_FIGHTER(gobj);
+    union Struct2070 local_ev;
+    plActionStats* acp = Player_GetActionStats(ft->player_id);
+    local_ev = *ev;
+    acp->attacks.total++;
+    acp->attacks.by_attack_counts[local_ev.x2073]++;
+    if (local_ev.count_thrown_items) {
+        acp->attacks.thrown_item_count++;
+    }
+    if (local_ev.count_aerials) {
+        acp->attacks.aerials_count++;
+    }
+    if (local_ev.count_specials) {
+        acp->attacks.specials_count++;
+    }
+    if (local_ev.count_x1A0) {
+        acp->attacks.x1A0_count++;
+    }
+    if (local_ev.count_x1A4) {
+        acp->attacks.x1A4_count++;
+    }
+}
 
 void pl_80037ECC(HSD_GObj* gobj)
 {
diff --git a/src/melee/ty/toy.c b/src/melee/ty/toy.c
index c8246914f7..0337cacb23 100644
--- a/src/melee/ty/toy.c
+++ b/src/melee/ty/toy.c
@@ -99,6 +99,11 @@ typedef struct TyCameraData_ {
     f32 x18;
     f32 x1C;
     f32 x20;
+    f32 x24;
+    f32 x28;
+    f32 x2C;
+    u8 pad30[0x58 - 0x30];
+    s32 x58;
 } TyCameraData_;
 
 typedef struct TyLightGObj_ {
@@ -260,6 +265,7 @@ extern void* un_804D6ED4;
 extern char un_803FE3B8[];
 extern char un_803FE3DC[];
 extern HSD_FogDesc un_803B8844;
+extern Vec3 un_803B8858;
 
 typedef struct PosArray {
     s32 xy[2];
@@ -1605,7 +1611,33 @@ void un_803075E8(s32 arg0)
     }
 }
 
-/// #un_80307828
+void un_80307828(int arg0)
+{
+    Vec3 interest;
+    TyCameraData_* data;
+    TyLightArray_* data2;
+    HSD_CObj* cobj;
+
+    interest = un_803B8858;
+    data = (void*) un_804D6E68;
+    data2 = un_804D6ED4;
+    cobj = data->x8->x28;
+
+    if (arg0 == 0) {
+        data->x18 = data->x1C = (f32)(data->x58 = 0);
+        data2->x18 = 0.0F;
+        data2->x14 = 0.0F;
+    } else {
+        data->x18 = 0.0F;
+        data2->x14 = 0.0F;
+    }
+
+    data->x20 = 38.0F;
+    data->x24 = data->x28 = data->x2C = 0.0F;
+
+    HSD_CObjSetInterest(cobj, &interest);
+    PAD_STACK(12);
+}
 
 /* 96.8% match */
 
diff --git a/src/melee/ty/toy.h b/src/melee/ty/toy.h
index a30d09e81f..cd6a99173f 100644
--- a/src/melee/ty/toy.h
+++ b/src/melee/ty/toy.h
@@ -44,7 +44,7 @@
 /* 30715C */ void un_8030715C(f32, f32);
 /* 307470 */ void un_80307470(s32);
 /* 3075E8 */ void un_803075E8(s32 arg0);
-/* 307828 */ UNK_RET un_80307828(int);
+/* 307828 */ void un_80307828(int);
 /* 3078E4 */ void un_803078E4(void);
 /* 307BA0 */ HSD_JObj* un_80307BA0(HSD_JObj*, s16);
 /* 307E84 */ void fn_80307E84(HSD_GObj* gobj);
