diff --git a/src/melee/gm/gmregclear.c b/src/melee/gm/gmregclear.c
index 6f276422b8..54addbe7ae 100644
--- a/src/melee/gm/gmregclear.c
+++ b/src/melee/gm/gmregclear.c
@@ -48,16 +48,21 @@ lbl_804706D8_t lbl_804706D8[12];
 struct lbl_80472D28_t {
     /*   +0 */ char pad_0[0x104];
     /* +104 */ int x104;
-    /* +108 */ char pad_108[0x2];
+    /* +108 */ s16 x108;
     /* +10A */ s16 x10A;
     /* +10C */ char pad_10C[0xA];
     /* +116 */ u8 x116;
+    /* +117 */ char pad_117[1];
+    /* +118 */ u8 x118;
+    /* +119 */ char pad_119[1];
+    /* +11A */ u8 x11A;
+    /* +11B */ u8 x11B;
 };
 
 struct lbl_80472E48_t {
     /* 0x00 */ char pad_0[4];
-    /* 0x04 */ s32 unk_4;        /* inferred */
-    /* 0x08 */ s32 unk_8;        /* inferred */
+    /* 0x04 */ s32 unk_4; /* inferred */
+    /* 0x08 */ s32 unk_8; /* inferred */
     /* 0x0C */ char pad_C[0x8];
     /* 0x14 */ s32 x14[0x1B];
 }; /* size = 0x80 */
@@ -108,7 +113,7 @@ static struct lbl_80472E48_t lbl_80472E48;
 static int lbl_80472EC8[4];
 
 struct {
-    u8 pad[0x398];
+    u8 x0[0x394];
     // offsets relative to .bss
     u8 x6BC;
     u16 x6BE;
@@ -802,7 +807,17 @@ s8 gm_8017E280(u16 arg0, u32 arg1)
     return -1;
 }
 
-/// #fn_8017E318
+int fn_8017E318(void)
+{
+    int i;
+    int sum = 0;
+    for (i = 0; i < 12; i++) {
+        if (lbl_804706D8[i].x0 != -1) {
+            sum += lbl_804706D8[i].x2;
+        }
+    }
+    return sum;
+}
 
 void fn_8017E3C8(void)
 {
@@ -931,7 +946,29 @@ bool gm_8017E7E0(void)
     return lbl_80472C30.x7C == 0x14;
 }
 
-/// #gm_8017E7FC
+void gm_8017E7FC(u8 arg0)
+{
+    UnkAdventureData* r31 = &lbl_80472C30;
+    bool cond;
+
+    if (gm_801A4310() == 4 && r31->x0.cpu_level >= 2 &&
+        r31->x0.xC.x20 + gm_8016AEDC() < 0xFD20U)
+    {
+        cond = true;
+    } else {
+        cond = false;
+    }
+
+    if (!cond) {
+        struct StartMeleeRules* rules = gm_8016AE50();
+        rules->x4_5 = 1;
+        r31->x77 = 0;
+        gm_SetScenePendingMinor(0x5A);
+    } else {
+        r31->x77 = 1;
+        gm_SetScenePendingMinor(0x5A);
+    }
+}
 
 /// #fn_8017E8A4
 
@@ -1099,7 +1136,16 @@ Fighter_GObj* gm_80180AF4(void)
     return Player_GetEntity(1);
 }
 
-/// #gm_80180B18
+void gm_80180B18(void)
+{
+    int i;
+
+    for (i = 0; i < 0x1B; i++) {
+        u8 idx = gm_80164024((u8) i);
+        lbl_80472E48.x14[gm_80164024((u8) i)] =
+            (u32) *gmMainLib_8015D06C(idx) / 10;
+    }
+}
 
 void gm_80180BA0(void)
 {
@@ -1107,8 +1153,7 @@ void gm_80180BA0(void)
 
     for (i = 0; i < 0x1B; i++) {
         u8 idx = gm_80164024((u8) i);
-        *gmMainLib_8015D06C(gm_80164024((u8) i)) =
-            lbl_80472E48.x14[idx] * 0xA;
+        *gmMainLib_8015D06C(gm_80164024((u8) i)) = lbl_80472E48.x14[idx] * 0xA;
     }
 }
 
@@ -1171,10 +1216,36 @@ s32 gm_80181A34(void)
     return lbl_80473594.x4;
 }
 
-/// #gm_80181A44
-
 extern u8 lbl_803D8D08[];
 
+void gm_80181A44(int c_kind, int arg1, bool arg2)
+{
+    u8* base;
+
+    base = lbl_803D8D08;
+
+    switch (arg1) {
+    case 0x21:
+        (base + 0x6C)[c_kind] = arg2;
+        break;
+    case 0x22:
+        (base + 0x12C)[c_kind] = arg2;
+        break;
+    case 0x23:
+        (base + 0x1EC)[c_kind] = arg2;
+        break;
+    case 0x24:
+        (base + 0x2AC)[c_kind] = arg2;
+        break;
+    case 0x25:
+        (base + 0x36C)[c_kind] = arg2;
+        break;
+    case 0x26:
+        (base + 0x42C)[c_kind] = arg2;
+        break;
+    }
+}
+
 void gm_80181AC8(int c_kind, int arg1, u16 arg2)
 {
     u8* base;
diff --git a/src/melee/gm/gmregclear.h b/src/melee/gm/gmregclear.h
index 712ed8b480..b2920f2d3e 100644
--- a/src/melee/gm/gmregclear.h
+++ b/src/melee/gm/gmregclear.h
@@ -40,7 +40,7 @@
 /* 17E160 */ bool fn_8017E160(void);
 /* 17E21C */ void fn_8017E21C(void);
 /* 17E280 */ s8 gm_8017E280(u16, u32);
-/* 17E318 */ UNK_RET fn_8017E318(UNK_PARAMS);
+/* 17E318 */ int fn_8017E318(void);
 /* 17E3C8 */ void fn_8017E3C8(void);
 /* 17E424 */ UnkAdventureData* gm_8017E424(void);
 /* 17E430 */ u8 gm_8017E430(void); ///< returns a player slot
@@ -92,7 +92,7 @@
 /* 180AC0 */ int fn_80180AC0(void);
 /* 180AE4 */ s32 gm_80180AE4(void);
 /* 180AF4 */ Fighter_GObj* gm_80180AF4(void);
-/* 180B18 */ UNK_RET gm_80180B18(UNK_PARAMS);
+/* 180B18 */ void gm_80180B18(void);
 /* 180BA0 */ void gm_80180BA0(void);
 /* 180C14 */ void fn_80180C14(HSD_GObj* gobj);
 /* 180C60 */ UNK_RET fn_80180C60(UNK_PARAMS);
@@ -103,7 +103,7 @@
 /* 181A14 */ int gm_80181A14(void);
 /* 181A24 */ u32 gm_80181A24(void);
 /* 181A34 */ s32 gm_80181A34(void);
-/* 181A44 */ UNK_RET gm_80181A44(int c_kind, int, bool);
+/* 181A44 */ void gm_80181A44(int c_kind, int, bool);
 /* 181AC8 */ void gm_80181AC8(int, int, u16);
 /* 181B64 */ void gm_80181B64(int c_kind, int, s32);
 /* 181BFC */ int fn_80181BFC(int*);
diff --git a/src/melee/gr/grcastle.c b/src/melee/gr/grcastle.c
index d96e55e36e..aaf5a8108f 100644
--- a/src/melee/gr/grcastle.c
+++ b/src/melee/gr/grcastle.c
@@ -3,6 +3,7 @@
 #include "placeholder.h"
 #include "types.h"
 
+#include "cm/camera.h"
 #include "ft/ftdevice.h"
 #include "ft/ftlib.h"
 #include "gr/grdisplay.h"
@@ -396,7 +397,24 @@ bool grCastle_801CE858(Ground_GObj* gobj)
     return false;
 }
 
-/// #grCastle_801CE860
+void grCastle_801CE860(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+
+    if ((s32) gp->gv.castle.xC4 < 3) {
+        s16 timer = gp->gv.castle.xC8;
+        gp->gv.castle.xC8 = timer - 1;
+        if (timer < 0) {
+            u8* ptr;
+            gp->gv.castle.xC4++;
+            grAnime_801C8138(gobj, gp->map_id, gp->gv.castle.xC4);
+            ptr = (u8*) grCs_804D6970;
+            ptr += gp->gv.castle.xC4 * 2;
+            gp->gv.castle.xC8 = *(s16*) (ptr + 0x12C);
+        }
+    }
+    Camera_80030E44(1, NULL);
+}
 
 void grCastle_801CE8E4(Ground_GObj* gobj) {}
 
@@ -583,7 +601,26 @@ void grCastle_801D08AC(void* arg0, unkCastle* arg1, Ground_GObj* gobj)
     grCastle_801D06CC_sub(arg1, gobj, 4);
 }
 
-/// #fn_801D0924
+void fn_801D0924(HSD_GObj* gobj, int renderpass)
+{
+    Ground* gp = gobj->user_data;
+    int i;
+
+    grDisplay_801C5DB0(gobj, renderpass);
+
+    if (gp->x10_flags.b7) {
+        GXColor color1 = { 0xFF, 0xFF, 0xFF, 0xFE };
+        GXColor color2 = { 0xFF, 0xFF, 0xFF, 0xFF };
+        PAD_STACK(4);
+
+        for (i = 0; i < 12; i++) {
+            if (gp->gv.castle3.x1C[i].data != NULL) {
+                lb_800117F4(&gp->gv.castle3.x1C[i], &color1, &color2,
+                            999, renderpass);
+            }
+        }
+    }
+}
 
 /// #grCastle_801D09B8
 
diff --git a/src/melee/gr/grcastle.h b/src/melee/gr/grcastle.h
index 714a83ed64..cac62015f0 100644
--- a/src/melee/gr/grcastle.h
+++ b/src/melee/gr/grcastle.h
@@ -86,7 +86,7 @@ struct unkCastle {
 /* 1D07BC */ void grCastle_801D07BC(UNK_T, unkCastle*, Ground_GObj*);
 /* 1D0834 */ void grCastle_801D0834(UNK_T, unkCastle*, Ground_GObj*);
 /* 1D08AC */ void grCastle_801D08AC(UNK_T, unkCastle*, Ground_GObj*);
-/* 1D0924 */ UNK_RET fn_801D0924(UNK_PARAMS);
+/* 1D0924 */ void fn_801D0924(HSD_GObj*, int);
 /* 1D09B8 */ UNK_RET grCastle_801D09B8(UNK_PARAMS);
 /* 1D0A9C */ void grCastle_801D0A9C(Vec3* arg0, f32 arg8);
 /* 1D0B04 */ DynamicsDesc* grCastle_801D0B04(enum_t);
diff --git a/src/melee/gr/gronett.c b/src/melee/gr/gronett.c
index 5fceaa65c0..259a1112b7 100644
--- a/src/melee/gr/gronett.c
+++ b/src/melee/gr/gronett.c
@@ -165,7 +165,26 @@ void grOnett_801E3C60(Ground_GObj* gobj)
 
 void grOnett_801E3CE0(Ground_GObj* gobj) {}
 
-/// #grOnett_801E3CE4
+void grOnett_801E3CE4(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    PAD_STACK(8);
+
+    Ground_801C2ED0(GET_JOBJ(gobj), gp->map_id);
+    grAnime_801C7FF8(gobj, 0, 7, 0, 0.0f, 0.0f);
+
+    gp->gv.kongo3.xC4 = -1;
+    gp->gv.kongo3.xC6 = 0;
+    gp->gv.unk.xC8 = 0;
+    gp->gv.unk.xCC = 0;
+
+    gp->x8_callback = NULL;
+    gp->xC_callback = NULL;
+    gp->x10_flags.b5 = 1;
+
+    mpJointSetCb1(3, gp, grOnett_801E40E4);
+    mpJointSetCb1(4, gp, grOnett_801E40E4);
+}
 
 bool grOnett_801E3D98(Ground_GObj* gobj)
 {
diff --git a/src/melee/gr/gronett.h b/src/melee/gr/gronett.h
index 072c81739f..0d4213c1e4 100644
--- a/src/melee/gr/gronett.h
+++ b/src/melee/gr/gronett.h
@@ -30,7 +30,8 @@
 /* 1E3D98 */ bool grOnett_801E3D98(Ground_GObj*);
 /* 1E3DA0 */ void grOnett_801E3DA0(Ground_GObj*);
 /* 1E40E0 */ void grOnett_801E40E0(Ground_GObj*);
-/* 1E40E4 */ UNK_RET grOnett_801E40E4(UNK_PARAMS);
+/* 1E40E4 */ void grOnett_801E40E4(Ground*, s32, CollData*, s32,
+                                   mpLib_GroundEnum, f32);
 /* 1E41B0 */ void grOnett_801E41B0(HSD_GObj*);
 /* 1E41C8 */ void grOnett_801E41C8(Ground_GObj*);
 /* 1E43D8 */ bool grOnett_801E43D8(Ground_GObj*);
