diff --git a/src/melee/mn/mnname.c b/src/melee/mn/mnname.c
index 850a92c321..38298fed5d 100644
--- a/src/melee/mn/mnname.c
+++ b/src/melee/mn/mnname.c
@@ -1,5 +1,9 @@
 #include "mnname.h"
 
+#include "mnmain.h"
+#include "mnnamenew.h"
+
+#include <baselib/gobj.h>
 #include <baselib/jobj.h>
 #include <melee/gm/gmmain_lib.h>
 
@@ -76,7 +80,20 @@ void CreateNameAtIndex(s32 slot)
 
 /// #mnName_MainInput
 
-/// #fn_80238540
+void fn_80238540(HSD_GObj* gobj)
+{
+    switch ((s32) mn_804A04F0.x10) {
+    case 0:
+        mnName_MainInput(gobj);
+        break;
+    case 1:
+        mnNameNew_MainInput(gobj);
+        break;
+    case 2:
+        mnName_ConfirmNameDeleteInput(gobj);
+        break;
+    }
+}
 
 void mnName_802385A0(HSD_GObj* gobj)
 {
@@ -141,7 +158,24 @@ void mnName_80239F5C(HSD_JObj* jobj, f32 x)
     HSD_JObjSetTranslateX(jobj, x);
 }
 
-/// #mnName_80239FFC
+void mnName_80239FFC(HSD_GObj* gobj)
+{
+    MnName_GObj* mn = (MnName_GObj*) gobj;
+    HSD_JObj* jobj = (HSD_JObj*) mn->gobj.user_data_remove_func;
+    HSD_JObj* child;
+
+    if (jobj == NULL) {
+        child = NULL;
+    } else {
+        child = jobj->child;
+    }
+    HSD_JObjRemoveAll(child);
+
+    if (mn->text != NULL) {
+        HSD_SisLib_803A5CC4(mn->text);
+        mn->text = NULL;
+    }
+}
 
 void mnName_8023A058(HSD_GObj* gobj)
 {
diff --git a/src/melee/mn/mnname.h b/src/melee/mn/mnname.h
index 3c4293d3fb..2efcf9be1e 100644
--- a/src/melee/mn/mnname.h
+++ b/src/melee/mn/mnname.h
@@ -4,9 +4,18 @@
 #include <placeholder.h>
 
 #include <baselib/forward.h>
+#include <baselib/gobj.h>
+#include <baselib/jobj.h>
+#include <baselib/sislib.h>
 
 #include <stdbool.h>
 
+typedef struct MnName_GObj {
+    /* +00 */ HSD_GObj gobj;
+    /* +38 */ void* x38;
+    /* +3C */ HSD_Text* text;
+} MnName_GObj;
+
 /* 23749C */ UNK_RET mnName_8023749C(UNK_PARAMS);
 /* 23754C */ char* GetNameText(int slot);
 /* 237594 */ int GetNameCount(void);
@@ -19,9 +28,9 @@
 /* 237A04 */ void CreateNameAtIndex(s32 slot);
 /* 237A68 */ UNK_RET mnName_SortNames(UNK_PARAMS);
 /* 237D94 */ UNK_RET mnName_80237D94(UNK_PARAMS);
-/* 237F78 */ UNK_RET mnName_ConfirmNameDeleteInput(UNK_PARAMS);
-/* 23817C */ UNK_RET mnName_MainInput(UNK_PARAMS);
-/* 238540 */ UNK_RET fn_80238540(UNK_PARAMS);
+/* 237F78 */ UNK_RET mnName_ConfirmNameDeleteInput(HSD_GObj*);
+/* 23817C */ UNK_RET mnName_MainInput(HSD_GObj*);
+/* 238540 */ void fn_80238540(HSD_GObj* gobj);
 /* 2385A0 */ void mnName_802385A0(HSD_GObj* gobj);
 /* 2385D4 */ UNK_RET mnName_GetPageCount(UNK_PARAMS);
 /* 238698 */ UNK_RET mnName_GetColumnCount(UNK_PARAMS);
@@ -36,7 +45,7 @@
 /* 239A24 */ void mnName_80239A24(HSD_GObj* gobj);
 /* 239EBC */ void mnName_80239EBC(HSD_JObj* jobj, f32 y);
 /* 239F5C */ void mnName_80239F5C(HSD_JObj* jobj, f32 x);
-/* 239FFC */ UNK_RET mnName_80239FFC(UNK_PARAMS);
+/* 239FFC */ void mnName_80239FFC(HSD_GObj* gobj);
 /* 23A058 */ void mnName_8023A058(HSD_GObj* gobj);
 /* 23A0BC */ UNK_RET fn_8023A0BC(UNK_PARAMS);
 /* 23A290 */ UNK_RET mnName_8023A290(UNK_PARAMS);
diff --git a/src/melee/mn/mnnamenew.h b/src/melee/mn/mnnamenew.h
index b620a821d5..49a7d99f6a 100644
--- a/src/melee/mn/mnnamenew.h
+++ b/src/melee/mn/mnnamenew.h
@@ -3,6 +3,8 @@
 
 #include <placeholder.h>
 
+#include <baselib/forward.h>
+
 /* 23B0F8 */ UNK_RET mnNameNew_8023B0F8(UNK_PARAMS);
 /* 23B224 */ UNK_RET mnNameNew_8023B224(UNK_PARAMS);
 /* 23B314 */ UNK_RET mnNameNew_8023B314(UNK_PARAMS);
@@ -13,7 +15,7 @@
 /* 23BFE4 */ UNK_RET WriteCharactersForNameAtIndex(UNK_PARAMS);
 /* 23C148 */ UNK_RET AddCharacterToName(UNK_PARAMS);
 /* 23C290 */ UNK_RET mnNameNew_GlyphVariantInput(UNK_PARAMS);
-/* 23C54C */ UNK_RET mnNameNew_MainInput(UNK_PARAMS);
+/* 23C54C */ UNK_RET mnNameNew_MainInput(HSD_GObj*);
 /* 23CE4C */ UNK_RET mnNameNew_8023CE4C(UNK_PARAMS);
 /* 23CFC8 */ UNK_RET fn_8023CFC8(UNK_PARAMS);
 /* 23D0F8 */ void fn_8023D0F8(void*);
