diff --git a/src/melee/gm/gm_1B03.c b/src/melee/gm/gm_1B03.c
index 4c60c21127..e5c4904472 100644
--- a/src/melee/gm/gm_1B03.c
+++ b/src/melee/gm/gm_1B03.c
@@ -685,7 +685,41 @@ void gm_801B0CF0(MinorScene* arg0)
     }
 }
 
-/// #gm_801B0DD0
+void gm_801B0DD0(MinorScene* arg0)
+{
+    u64 sfx_result = 0;
+    int i;
+    struct DebugResultsData* data = gm_801A427C(arg0);
+    MatchEnd* match_end = &data->match_end;
+
+    data->x0_0 = un_803FA258[0x5E];
+    data->x0_1 = un_803FA258[0x5F];
+    data->x1 = un_803FA258[0x5A];
+    data->x2 = un_803FA258[0x5B];
+    data->x3 = un_803FA258[0x5C];
+    data->x4 = un_803FA258[0x5D];
+
+    gm_80166A98(match_end, un_803FA258[0x60] & 0xFF, (s8) un_803FA258[0x51],
+                un_803FA258[0x56] - 1, (s8) un_803FA258[0x52],
+                un_803FA258[0x57] - 1, (s8) un_803FA258[0x53],
+                un_803FA258[0x58] - 1, (s8) un_803FA258[0x54],
+                un_803FA258[0x59] - 1);
+
+    for (i = 0; i < 4; i++) {
+        if (match_end->player_standings[i].slot_type != 3 &&
+            match_end->player_standings[i].is_big_loser == 0)
+        {
+            sfx_result |= lbAudioAx_80026E84(
+                match_end->player_standings[i].character_kind);
+        }
+    }
+
+    lbAudioAx_80026F2C(0x14);
+    lbAudioAx_8002702C(0x4, sfx_result);
+    gm_80168FC4();
+    gm_801701A0();
+    PAD_STACK(8);
+}
 
 void gm_801B0F1C(MinorScene* arg0)
 {
diff --git a/src/melee/gm/gm_1B03.static.h b/src/melee/gm/gm_1B03.static.h
index 1b88fd444e..70ea4cb5b4 100644
--- a/src/melee/gm/gm_1B03.static.h
+++ b/src/melee/gm/gm_1B03.static.h
@@ -5,7 +5,15 @@
 #include "gm/types.h"
 
 struct DebugResultsData {
-    u8 pad_x0[0x2284];
+    u8 x0_0 : 1;
+    u8 x0_1 : 1;
+    u8 x0_other : 6;
+    u8 x1;
+    u8 x2;
+    u8 x3;
+    u8 x4;
+    u8 pad_x5[3];
+    MatchEnd match_end;
 }; /// data used by gm_801B0DD0
 
 struct DebugMemcardData {
