diff --git a/src/melee/ef/efsync.c b/src/melee/ef/efsync.c
index 26f73d3bc0..c7ebd79694 100644
--- a/src/melee/ef/efsync.c
+++ b/src/melee/ef/efsync.c
@@ -106,7 +106,8 @@ void* efSync_Spawn(s32 arg_gfx_id, HSD_GObj* arg_gobj, ...)
         return efAsync_80063930(gfx_id, arg_gobj, vlist);
     }
     if (gfx_id < 0x4BA) {
-        // return efSpecial_SpawnSpecial(gfx_id, arg_gobj, va_arg(vlist, void*));
+        // return efSpecial_SpawnSpecial(gfx_id, arg_gobj, va_arg(vlist,
+        // void*));
         return efSpecial_SpawnSpecial(gfx_id, arg_gobj, vlist);
     }
     efLib_804D64E8 = 1;
diff --git a/src/melee/gr/grfigure1.c b/src/melee/gr/grfigure1.c
index b8af9b20b3..e0c0b32256 100644
--- a/src/melee/gr/grfigure1.c
+++ b/src/melee/gr/grfigure1.c
@@ -106,8 +106,7 @@ static HSD_GObj* grFigure1_8020E080(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 194,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 194, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grfigure2.c b/src/melee/gr/grfigure2.c
index 920bd7fea4..d4e28780b8 100644
--- a/src/melee/gr/grfigure2.c
+++ b/src/melee/gr/grfigure2.c
@@ -6,8 +6,8 @@
 #include "gr/grdisplay.h"
 #include "gr/grfigure1.h"
 #include "gr/ground.h"
-#include "gr/inlines.h"
 #include "gr/grzakogenerator.h"
+#include "gr/inlines.h"
 #include "gr/types.h"
 
 #include "lb/forward.h"
@@ -103,8 +103,7 @@ static HSD_GObj* grFigure2_8020E314(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 196,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 196, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grfigure3.c b/src/melee/gr/grfigure3.c
index 783846ff2e..336709453a 100644
--- a/src/melee/gr/grfigure3.c
+++ b/src/melee/gr/grfigure3.c
@@ -5,8 +5,8 @@
 #include "gr/granime.h"
 #include "gr/grdisplay.h"
 #include "gr/ground.h"
-#include "gr/inlines.h"
 #include "gr/grzakogenerator.h"
+#include "gr/inlines.h"
 #include "gr/types.h"
 
 #include "lb/forward.h"
@@ -105,8 +105,7 @@ static HSD_GObj* grFigure3_8020E5A8(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grfigureget.c b/src/melee/gr/grfigureget.c
index 9b2d8afbf9..0f8cb3bb96 100644
--- a/src/melee/gr/grfigureget.c
+++ b/src/melee/gr/grfigureget.c
@@ -99,8 +99,7 @@ HSD_GObj* grFigureGet_802195CC(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 187,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 187, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grgarden.c b/src/melee/gr/grgarden.c
index 4da8ae50f9..a940f7aedd 100644
--- a/src/melee/gr/grgarden.c
+++ b/src/melee/gr/grgarden.c
@@ -19,6 +19,7 @@
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
 
+#include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjobject.h>
@@ -26,7 +27,6 @@
 #include <baselib/gobjuserdata.h>
 #include <baselib/jobj.h>
 #include <baselib/random.h>
-#include <dolphin/os/OSError.h>
 
 StageCallbacks grGd_803E5248[] = {
     { grGarden_80202D60, grGarden_80202D8C, grGarden_80202D94,
@@ -109,8 +109,7 @@ HSD_GObj* grGarden_80202C78(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xE3,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xE3, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grpushon.c b/src/melee/gr/grpushon.c
index 8c5649a664..a025150e80 100644
--- a/src/melee/gr/grpushon.c
+++ b/src/melee/gr/grpushon.c
@@ -20,11 +20,11 @@
 #include "lb/lb_00F9.h"
 #include "lb/types.h"
 #include "mp/mplib.h"
+#include "sysdolphin/baselib/lobj.h"
 
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
 #include <baselib/random.h>
-#include "sysdolphin/baselib/lobj.h"
 
 struct grPushOn_Entry {
     s32 x0;
diff --git a/src/melee/gr/grpushon.h b/src/melee/gr/grpushon.h
index 754648e688..537b1e1409 100644
--- a/src/melee/gr/grpushon.h
+++ b/src/melee/gr/grpushon.h
@@ -28,7 +28,8 @@
 /* 218ED0 */ void grPushOn_80218ED0(Ground_GObj*);
 /* 218ED4 */ HSD_LObj* grPushOn_80218ED4(HSD_GObj*);
 /* 218FC0 */ HSD_LObj* grPushOn_80218FC0(HSD_GObj*);
-/* 2190A0 */ void fn_802190A0(Ground*, s32, CollData*, s32, mpLib_GroundEnum, f32);
+/* 2190A0 */ void fn_802190A0(Ground*, s32, CollData*, s32, mpLib_GroundEnum,
+                              f32);
 /* 2190D0 */ void grPushOn_802190D0(HSD_GObj*);
 /* 219204 */ void grPushOn_80219204(int arg0, int* out1, int* out2);
 /* 219230 */ int grPushOn_80219230(int);
diff --git a/src/melee/gr/grshrine.c b/src/melee/gr/grshrine.c
index 9d241ec308..f56aaee39b 100644
--- a/src/melee/gr/grshrine.c
+++ b/src/melee/gr/grshrine.c
@@ -5,8 +5,8 @@
 #include "gr/granime.h"
 #include "gr/grdisplay.h"
 #include "gr/ground.h"
-#include "gr/inlines.h"
 #include "gr/grzakogenerator.h"
+#include "gr/inlines.h"
 #include "gr/types.h"
 
 #include "lb/forward.h"
@@ -115,8 +115,7 @@ static HSD_GObj* grShrine_80201D20(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 205,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 205, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grstory.c b/src/melee/gr/grstory.c
index e0c3f45188..7951db7911 100644
--- a/src/melee/gr/grstory.c
+++ b/src/melee/gr/grstory.c
@@ -2,6 +2,7 @@
 
 #include <platform.h>
 
+#include <dolphin/os/OSError.h>
 #include <sysdolphin/baselib/gobjgxlink.h>
 #include <sysdolphin/baselib/gobjproc.h>
 #include <sysdolphin/baselib/random.h>
@@ -18,8 +19,6 @@
 #include <melee/lb/lb_00B0.h>
 #include <melee/lb/lb_00F9.h>
 
-#include <dolphin/os/OSError.h>
-
 /* 1E302C */ static void grStory_801E302C(bool);
 /* 1E36D0 */ static DynamicsDesc* grStory_801E36D0(enum_t);
 
@@ -97,8 +96,7 @@ Ground_GObj* grStory_801E30D8(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 220,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 220, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtcaptain.c b/src/melee/gr/grtcaptain.c
index 292df8feeb..82028db39b 100644
--- a/src/melee/gr/grtcaptain.c
+++ b/src/melee/gr/grtcaptain.c
@@ -104,8 +104,7 @@ static HSD_GObj* grTCaptain_8021FD04(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 215,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 215, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtclink.c b/src/melee/gr/grtclink.c
index 18d797d74f..b5eb9ae81f 100644
--- a/src/melee/gr/grtclink.c
+++ b/src/melee/gr/grtclink.c
@@ -8,10 +8,10 @@
 #include "lb/lb_00F9.h"
 #include "lb/types.h"
 
+#include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
-#include <dolphin/os/OSError.h>
 
 static StageCallbacks grTCLink_803E86D0[] = {
     {
diff --git a/src/melee/gr/grtdonkey.c b/src/melee/gr/grtdonkey.c
index 3181e83aad..004a08e02e 100644
--- a/src/melee/gr/grtdonkey.c
+++ b/src/melee/gr/grtdonkey.c
@@ -106,8 +106,7 @@ static HSD_GObj* grTDonkey_802202CC(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xc3,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xc3, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtdrmario.c b/src/melee/gr/grtdrmario.c
index 426e45dc2d..8a47d96bb7 100644
--- a/src/melee/gr/grtdrmario.c
+++ b/src/melee/gr/grtdrmario.c
@@ -101,8 +101,7 @@ HSD_GObj* grtDrMario_802205B0(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xCC,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xCC, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtemblem.c b/src/melee/gr/grtemblem.c
index 2c2134dcea..277417fedb 100644
--- a/src/melee/gr/grtemblem.c
+++ b/src/melee/gr/grtemblem.c
@@ -116,8 +116,7 @@ static HSD_GObj* grTRoy_80224498(s32 gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtest.c b/src/melee/gr/grtest.c
index 1d8d1499ba..ab27f9dc91 100644
--- a/src/melee/gr/grtest.c
+++ b/src/melee/gr/grtest.c
@@ -12,10 +12,10 @@
 #include "gr/stage.h"
 #include "gr/types.h"
 
+#include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/jobj.h>
-#include <dolphin/os/OSError.h>
 #include <sysdolphin/baselib/controller.h>
 #include <sysdolphin/baselib/dobj.h>
 #include <sysdolphin/baselib/gobjproc.h>
@@ -96,8 +96,7 @@ HSD_GObj* grTest_80207044(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 209,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 209, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtfox.c b/src/melee/gr/grtfox.c
index 64fa56e89d..61c86d2547 100644
--- a/src/melee/gr/grtfox.c
+++ b/src/melee/gr/grtfox.c
@@ -116,8 +116,7 @@ HSD_GObj* grTFox_80220C2C(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xCB,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xCB, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtgamewatch.c b/src/melee/gr/grtgamewatch.c
index 276e896673..787e1be2d2 100644
--- a/src/melee/gr/grtgamewatch.c
+++ b/src/melee/gr/grtgamewatch.c
@@ -9,10 +9,10 @@
 #include "lb/lb_00F9.h"
 #include "lb/types.h"
 
+#include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
-#include <dolphin/os/OSError.h>
 
 StageCallbacks grTGw_803E96F8[] = {
     { grTGameWatch_8022429C, grTGameWatch_802242C8, grTGameWatch_802242D0,
diff --git a/src/melee/gr/grtganon.c b/src/melee/gr/grtganon.c
index 965a487f4c..95dee28e6d 100644
--- a/src/melee/gr/grtganon.c
+++ b/src/melee/gr/grtganon.c
@@ -12,10 +12,10 @@
 
 #include "mp/mplib.h"
 
+#include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
-#include <dolphin/os/OSError.h>
 
 StageCallbacks grTGn_803E9880[] = {
     { grTGanon_8022486C, grTGanon_80224898, grTGanon_802248A0,
diff --git a/src/melee/gr/grtkirby.c b/src/melee/gr/grtkirby.c
index e4d776147a..3b1b0bcb60 100644
--- a/src/melee/gr/grtkirby.c
+++ b/src/melee/gr/grtkirby.c
@@ -102,8 +102,7 @@ HSD_GObj* grTKirby_80221408(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtkoopa.c b/src/melee/gr/grtkoopa.c
index ab31b76ed6..85dd710bdf 100644
--- a/src/melee/gr/grtkoopa.c
+++ b/src/melee/gr/grtkoopa.c
@@ -5,8 +5,8 @@
 #include "gr/granime.h"
 #include "gr/grdisplay.h"
 #include "gr/ground.h"
-#include "gr/inlines.h"
 #include "gr/grzakogenerator.h"
+#include "gr/inlines.h"
 #include "gr/types.h"
 
 #include "lb/forward.h"
@@ -104,8 +104,7 @@ static HSD_GObj* grTKoopa_802216EC(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtlink.c b/src/melee/gr/grtlink.c
index 380d038305..5f5dbd1692 100644
--- a/src/melee/gr/grtlink.c
+++ b/src/melee/gr/grtlink.c
@@ -103,8 +103,7 @@ HSD_GObj* grTLink_802219D0(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtluigi.c b/src/melee/gr/grtluigi.c
index dfc6fd3acf..7f371cba51 100644
--- a/src/melee/gr/grtluigi.c
+++ b/src/melee/gr/grtluigi.c
@@ -88,8 +88,7 @@ HSD_GObj* grTLuigi_80221CB4(int arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtmario.c b/src/melee/gr/grtmario.c
index 141e47156a..caf641c5ac 100644
--- a/src/melee/gr/grtmario.c
+++ b/src/melee/gr/grtmario.c
@@ -124,8 +124,7 @@ HSD_GObj* grTMario_8021F8E4(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xCD,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xCD, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtmars.c b/src/melee/gr/grtmars.c
index d9fbe56e24..4de67fe58a 100644
--- a/src/melee/gr/grtmars.c
+++ b/src/melee/gr/grtmars.c
@@ -3,8 +3,8 @@
 #include "gr/granime.h"
 #include "gr/grdisplay.h"
 #include "gr/ground.h"
-#include "gr/inlines.h"
 #include "gr/grzakogenerator.h"
+#include "gr/inlines.h"
 #include "gr/types.h"
 
 #include "lb/forward.h"
@@ -116,8 +116,7 @@ static HSD_GObj* grTMars_80221F98(int id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtmewtwo.c b/src/melee/gr/grtmewtwo.c
index 7b599dd8b8..a8e1364baf 100644
--- a/src/melee/gr/grtmewtwo.c
+++ b/src/melee/gr/grtmewtwo.c
@@ -142,8 +142,7 @@ Ground_GObj* grTMewtwo_80222284(int index)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 201,
-                 index);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 201, index);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtness.c b/src/melee/gr/grtness.c
index 7889e60d75..33e3cfad4d 100644
--- a/src/melee/gr/grtness.c
+++ b/src/melee/gr/grtness.c
@@ -3,8 +3,8 @@
 #include "gr/granime.h"
 #include "gr/grdisplay.h"
 #include "gr/ground.h"
-#include "gr/inlines.h"
 #include "gr/grzakogenerator.h"
+#include "gr/inlines.h"
 #include "gr/types.h"
 
 #include "lb/forward.h"
@@ -118,8 +118,7 @@ static HSD_GObj* grTNess_80222674(int id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtpeach.c b/src/melee/gr/grtpeach.c
index f76f09dda9..9b0898a457 100644
--- a/src/melee/gr/grtpeach.c
+++ b/src/melee/gr/grtpeach.c
@@ -126,8 +126,7 @@ Ground_GObj* grTPeach_80222958(int index)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195,
-                 index);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195, index);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtpichu.c b/src/melee/gr/grtpichu.c
index 7388f49395..b430b09095 100644
--- a/src/melee/gr/grtpichu.c
+++ b/src/melee/gr/grtpichu.c
@@ -119,8 +119,7 @@ static HSD_GObj* grTPichu_80222C3C(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xc2,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xc2, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtpikachu.c b/src/melee/gr/grtpikachu.c
index 3c87c6e594..f8f0e816df 100644
--- a/src/melee/gr/grtpikachu.c
+++ b/src/melee/gr/grtpikachu.c
@@ -8,10 +8,10 @@
 #include "lb/lb_00F9.h"
 #include "lb/types.h"
 
+#include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
-#include <dolphin/os/OSError.h>
 
 StageCallbacks grTPk_803E9270[] = {
     { grTPikachu_80223008, grTPikachu_80223034, grTPikachu_8022303C,
diff --git a/src/melee/gr/grtpurin.c b/src/melee/gr/grtpurin.c
index 8af279e95e..910f25cb03 100644
--- a/src/melee/gr/grtpurin.c
+++ b/src/melee/gr/grtpurin.c
@@ -10,10 +10,10 @@
 #include "lb/types.h"
 #include "mp/mplib.h"
 
+#include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
-#include <dolphin/os/OSError.h>
 #include <baselib/jobj.h>
 
 StageCallbacks grTPr_803E9338[] = {
diff --git a/src/melee/gr/grtsamus.c b/src/melee/gr/grtsamus.c
index 2981a2e89e..60a81a0ba8 100644
--- a/src/melee/gr/grtsamus.c
+++ b/src/melee/gr/grtsamus.c
@@ -116,8 +116,7 @@ HSD_GObj* grTSamus_80223624(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtseak.c b/src/melee/gr/grtseak.c
index 5600392fae..b4a6498a66 100644
--- a/src/melee/gr/grtseak.c
+++ b/src/melee/gr/grtseak.c
@@ -116,8 +116,7 @@ HSD_GObj* grTSeak_80223908(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, arg0);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtyoshi.c b/src/melee/gr/grtyoshi.c
index bca1ced49d..76408747f6 100644
--- a/src/melee/gr/grtyoshi.c
+++ b/src/melee/gr/grtyoshi.c
@@ -117,8 +117,7 @@ static HSD_GObj* grTYoshi_80223BEC(int gobj_id)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195,
-                 gobj_id);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 195, gobj_id);
     }
 
     return gobj;
diff --git a/src/melee/gr/grtzelda.c b/src/melee/gr/grtzelda.c
index a062ba2897..452ac8f803 100644
--- a/src/melee/gr/grtzelda.c
+++ b/src/melee/gr/grtzelda.c
@@ -103,8 +103,7 @@ HSD_GObj* grTZelda_80223ED0(s32 arg0)
     if (gobj != NULL) {
         Ground_SetupStageCallbacks(gobj, callbacks);
     } else {
-        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3,
-                 arg0);
+        OSReport("%s:%d: couldn t get gobj(id=%d)\n", __FILE__, 0xC3, arg0);
     }
 
     return gobj;
diff --git a/src/melee/mn/mnruleplus.c b/src/melee/mn/mnruleplus.c
index 4b82d9bf07..6794691b90 100644
--- a/src/melee/mn/mnruleplus.c
+++ b/src/melee/mn/mnruleplus.c
@@ -63,12 +63,8 @@ AnimLoopSettings mn_803ED270[3] = {
 };
 
 AnimLoopSettings mn_803ED294[7] = {
-    { 0.0f, 0.0f, 0.0f },
-    { 0.0f, 0.0f, 0.0f },
-    { 0.0f, 0.0f, 0.0f },
-    { 0.0f, 0.0f, 0.0f },
-    { 0.0f, 0.0f, 0.0f },
-    { 0.0f, 0.0f, 0.0f },
+    { 0.0f, 0.0f, 0.0f }, { 0.0f, 0.0f, 0.0f }, { 0.0f, 0.0f, 0.0f },
+    { 0.0f, 0.0f, 0.0f }, { 0.0f, 0.0f, 0.0f }, { 0.0f, 0.0f, 0.0f },
     { 0.0f, 0.0f, 0.0f },
 };
 
