diff --git a/config/GALE01/splits.txt b/config/GALE01/splits.txt
index 7a85b87e6a..a7ebffd5ff 100644
--- a/config/GALE01/splits.txt
+++ b/config/GALE01/splits.txt
@@ -340,8 +340,12 @@ melee/ft/ft_0819.c:
 	.sdata2     start:0x804D83A0 end:0x804D83A8
 
 melee/ft/ft_081B.c:
-	.text       start:0x80081B38 end:0x8008521C
-	.sdata2     start:0x804D83A8 end:0x804D83C0
+	.text       start:0x80081B38 end:0x80084E1C
+	.sdata2     start:0x804D83A8 end:0x804D83B8
+
+melee/ft/ft_084E.c:
+	.text       start:0x80084E1C end:0x8008521C
+	.sdata2     start:0x804D83B8 end:0x804D83C0
 
 melee/ft/ft_0852.c:
 	.text       start:0x8008521C end:0x80085560
diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 875d16d5f0..bea0e471b3 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -29349,11 +29349,11 @@ ftDrawCommon_804D8370 = .sdata2:0x804D8370; // type:object size:0x4 scope:global
 @277 = .sdata2:0x804D8398; // type:object size:0x8 scope:local align:8 data:double
 @273 = .sdata2:0x804D83A0; // type:object size:0x4 scope:local align:4 data:float
 @274 = .sdata2:0x804D83A4; // type:object size:0x4 scope:local align:4 data:float
-ft_804D83A8 = .sdata2:0x804D83A8; // type:object size:0x4 scope:local align:4 data:float
-ft_804D83AC = .sdata2:0x804D83AC; // type:object size:0x4 scope:local align:4 data:float
-ft_804D83B0 = .sdata2:0x804D83B0; // type:object size:0x4 scope:local align:4 data:float
-ft_804D83B4 = .sdata2:0x804D83B4; // type:object size:0x4 scope:local align:4 data:float
-ft_804D83B8 = .sdata2:0x804D83B8; // type:object size:0x4 scope:local align:4 data:float
+@195 = .sdata2:0x804D83A8; // type:object size:0x4 scope:local align:4 data:float
+@333 = .sdata2:0x804D83AC; // type:object size:0x4 scope:local align:4 data:float
+@1111 = .sdata2:0x804D83B0; // type:object size:0x4 scope:local align:4 data:float
+@1153 = .sdata2:0x804D83B4; // type:object size:0x4 scope:local align:4 data:float
+@204 = .sdata2:0x804D83B8; // type:object size:0x4 scope:local align:4 data:float
 @274 = .sdata2:0x804D83C0; // type:object size:0x8 scope:local align:8 data:double
 @295 = .sdata2:0x804D83C8; // type:object size:0x4 scope:local align:4 data:float
 @321 = .sdata2:0x804D83CC; // type:object size:0x4 scope:local align:4 data:float
diff --git a/configure.py b/configure.py
index 4167af3da7..e4120c304d 100755
--- a/configure.py
+++ b/configure.py
@@ -525,7 +525,8 @@ def MatchingFor(*versions):
             Object(Matching, "melee/ft/ftcliffcommon.c"),
             Object(Matching, "melee/ft/ftwalljump.c"),
             Object(Matching, "melee/ft/ft_0819.c"),
-            Object(NonMatching, "melee/ft/ft_081B.c"),
+            Object(Matching, "melee/ft/ft_081B.c"),
+            Object(Matching, "melee/ft/ft_084E.c"),
             Object(NonMatching, "melee/ft/ft_0852.c"),
             Object(Matching, "melee/ft/ftdata.c"),
             Object(Matching, "melee/ft/ftmotionstates.c"),
diff --git a/src/melee/ft/ft_081B.c b/src/melee/ft/ft_081B.c
index 97a054ac56..a40081ddba 100644
--- a/src/melee/ft/ft_081B.c
+++ b/src/melee/ft/ft_081B.c
@@ -1380,156 +1380,3 @@ void ft_80084DB0(Fighter_GObj* gobj)
     }
     ftCommon_8007D268(fp);
 }
-
-void ft_80084E1C(Fighter_GObj* gobj, float threshold, float drift_max,
-                 float target_max)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCo_DatAttrs* co_attrs = getFtAttrs(fp);
-    float drift;
-    float target_vel;
-
-    ftCommon_CheckFallFast(fp);
-    if (fp->fall_fast) {
-        ftCommon_FallFast(fp);
-    } else {
-        ftCommon_Fall(fp, co_attrs->grav, co_attrs->terminal_vel);
-    }
-
-    if (ABS(fp->input.lstick.x) >= threshold) {
-        drift = fp->input.lstick.x * drift_max;
-        target_vel = fp->input.lstick.x * target_max;
-    } else {
-        target_vel = 0.0F;
-        drift = 0.0F;
-    }
-    ftCommon_8007D140(fp, drift, target_vel, co_attrs->aerial_friction);
-}
-
-void ft_80084EEC(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCo_DatAttrs* co_attrs = getFtAttrs(fp);
-
-    ftCommon_Fall(fp, co_attrs->grav, co_attrs->terminal_vel);
-    ftCommon_ApplyFrictionAir(fp, co_attrs->aerial_friction);
-}
-
-void ft_80084F3C(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCo_DatAttrs* co = &fp->co_attrs;
-    f32 friction = co->gr_friction;
-    PAD_STACK(8);
-    if (ABS(fp->gr_vel) > co->walk_max_vel) {
-        friction *= p_ftCommonData->x6C;
-    }
-    ftCommon_ApplyFrictionGround(fp, friction);
-    ftCommon_ApplyGroundMovement(gobj);
-}
-
-void ft_80084FA8(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCo_DatAttrs* co = &fp->co_attrs;
-    f32 var_f2 = fp->gr_vel;
-    f32 var_f1 = co->gr_friction;
-    if (var_f2 < 0.0F) {
-        var_f2 = -var_f2;
-    }
-    if (var_f2 > co->walk_max_vel) {
-        var_f1 *= p_ftCommonData->x6C;
-    }
-    ft_80085030(gobj, var_f1, fp->facing_dir);
-}
-
-void ft_80085004(Fighter_GObj* gobj)
-{
-    ftCo_DatAttrs* co_attrs;
-    Fighter* fp = GET_FIGHTER(gobj);
-    co_attrs = &fp->co_attrs;
-
-    ft_80085030(gobj, co_attrs->gr_friction, fp->facing_dir1);
-}
-
-void ft_80085030(Fighter_GObj* gobj, float gr_friction, float facing_dir)
-{
-    Fighter* fp = gobj->user_data;
-    if (fp->x594_b0) {
-        fp->xE4_ground_accel_1 =
-            fp->x6A4_transNOffset.z * facing_dir - fp->gr_vel;
-    } else {
-        ftCommon_ApplyFrictionGround(fp, gr_friction);
-    }
-    ftCommon_ApplyGroundMovement(gobj);
-}
-
-void ft_80085088(Fighter_GObj* gobj)
-{
-    ftCo_DatAttrs* co_attrs;
-    Fighter* fp = GET_FIGHTER(gobj);
-    co_attrs = &fp->co_attrs;
-
-    ft_800850E0(gobj, co_attrs->gr_friction, fp->facing_dir);
-}
-
-void ft_800850B4(Fighter_GObj* gobj)
-{
-    ftCo_DatAttrs* co_attrs;
-    Fighter* fp = GET_FIGHTER(gobj);
-    co_attrs = &fp->co_attrs;
-
-    ft_800850E0(gobj, co_attrs->gr_friction, fp->facing_dir1);
-}
-
-void ft_800850E0(Fighter_GObj* gobj, float arg8, float arg9)
-{
-    Fighter* fp = gobj->user_data;
-    if (fp->x594_b0) {
-        fp->gr_vel = fp->x6A4_transNOffset.z * arg9;
-    } else {
-        ftCommon_ApplyFrictionGround(fp, arg8);
-    }
-    ftCommon_ApplyGroundMovement(gobj);
-}
-
-void ft_80085134(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    fp->self_vel.x = fp->x6A4_transNOffset.z * fp->facing_dir;
-    fp->self_vel.y = fp->x6A4_transNOffset.y;
-}
-
-void ft_80085154(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    f32 lstick_x = cosf(fp->lstick_angle);
-    f32 lstick_y = sinf(fp->lstick_angle);
-    f32 temp_f0 = fp->x6A4_transNOffset.y;
-    f32 temp_f3 = fp->x6A4_transNOffset.z * fp->facing_dir;
-    fp->self_vel.x = (temp_f3 * lstick_x) - (temp_f0 * lstick_y);
-    fp->self_vel.y = (temp_f3 * lstick_y) + (temp_f0 * lstick_x);
-}
-
-void ft_800851C0(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    fp->self_vel.y = fp->x6A4_transNOffset.y;
-}
-
-void ft_800851D0(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    if (fp->x197C != NULL) {
-        fp->self_vel.y =
-            fp->x6A4_transNOffset.y * ((f32*) Fighter_804D6520)[5];
-    } else {
-        fp->self_vel.y = fp->x6A4_transNOffset.y;
-    }
-}
-
-void ft_80085204(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    fp->self_vel.x = fp->x6A4_transNOffset.z * fp->facing_dir;
-}
diff --git a/src/melee/ft/ft_081B.h b/src/melee/ft/ft_081B.h
index c2e8176961..66dc33c60f 100644
--- a/src/melee/ft/ft_081B.h
+++ b/src/melee/ft/ft_081B.h
@@ -9,6 +9,9 @@
 #include "ftCommon/forward.h"
 #include <baselib/forward.h>
 
+// TODO transitive include hack, should be included directly by dependers
+#include <melee/ft/ft_084E.h>
+
 /* 081B38 */ void ft_80081B38(Fighter_GObj* gobj);
 /* 081C88 */ void ft_80081C88(Fighter_GObj* dst_gobj, float scl_y);
 /* 081D0C */ GroundOrAir ft_80081D0C(Fighter_GObj* gobj);
@@ -81,20 +84,5 @@
 /* 084CB0 */ void ft_80084CB0(Fighter*, ftCollisionBox*);
 /* 084CE4 */ bool ft_80084CE4(Fighter* attacker, Fighter* victim);
 /* 084DB0 */ void ft_80084DB0(Fighter_GObj*);
-/* 084E1C */ void ft_80084E1C(Fighter_GObj* gobj, float, float, float);
-/* 084EEC */ void ft_80084EEC(Fighter_GObj* gobj);
-/* 084F3C */ void ft_80084F3C(Fighter_GObj* gobj);
-/* 084FA8 */ void ft_80084FA8(Fighter_GObj* gobj);
-/* 085004 */ void ft_80085004(Fighter_GObj* gobj);
-/* 085030 */ void ft_80085030(Fighter_GObj*, float gr_friction,
-                              float facing_dir);
-/* 085088 */ void ft_80085088(Fighter_GObj*);
-/* 0850B4 */ void ft_800850B4(Fighter_GObj*);
-/* 0850E0 */ void ft_800850E0(Fighter_GObj*, float, float);
-/* 085134 */ void ft_80085134(Fighter_GObj* gobj);
-/* 085154 */ void ft_80085154(Fighter_GObj* gobj);
-/* 0851C0 */ void ft_800851C0(Fighter_GObj*);
-/* 0851D0 */ void ft_800851D0(Fighter_GObj*);
-/* 085204 */ void ft_80085204(Fighter_GObj* gobj);
 
 #endif
diff --git a/src/melee/ft/ft_084E.c b/src/melee/ft/ft_084E.c
new file mode 100644
index 0000000000..6b720a2688
--- /dev/null
+++ b/src/melee/ft/ft_084E.c
@@ -0,0 +1,158 @@
+#include "ft_084E.h"
+
+#include <melee/ft/fighter.h>
+#include <melee/ft/ftcommon.h>
+#include <math.h>
+
+void ft_80084E1C(Fighter_GObj* gobj, float threshold, float drift_max,
+                 float target_max)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCo_DatAttrs* co_attrs = getFtAttrs(fp);
+    float drift;
+    float target_vel;
+
+    ftCommon_CheckFallFast(fp);
+    if (fp->fall_fast) {
+        ftCommon_FallFast(fp);
+    } else {
+        ftCommon_Fall(fp, co_attrs->grav, co_attrs->terminal_vel);
+    }
+
+    if (ABS(fp->input.lstick.x) >= threshold) {
+        drift = fp->input.lstick.x * drift_max;
+        target_vel = fp->input.lstick.x * target_max;
+    } else {
+        target_vel = 0.0F;
+        drift = 0.0F;
+    }
+    ftCommon_8007D140(fp, drift, target_vel, co_attrs->aerial_friction);
+}
+
+void ft_80084EEC(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCo_DatAttrs* co_attrs = getFtAttrs(fp);
+
+    ftCommon_Fall(fp, co_attrs->grav, co_attrs->terminal_vel);
+    ftCommon_ApplyFrictionAir(fp, co_attrs->aerial_friction);
+}
+
+void ft_80084F3C(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCo_DatAttrs* co = &fp->co_attrs;
+    f32 friction = co->gr_friction;
+    PAD_STACK(8);
+    if (ABS(fp->gr_vel) > co->walk_max_vel) {
+        friction *= p_ftCommonData->x6C;
+    }
+    ftCommon_ApplyFrictionGround(fp, friction);
+    ftCommon_ApplyGroundMovement(gobj);
+}
+
+void ft_80084FA8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCo_DatAttrs* co = &fp->co_attrs;
+    f32 var_f2 = fp->gr_vel;
+    f32 var_f1 = co->gr_friction;
+    if (var_f2 < 0.0F) {
+        var_f2 = -var_f2;
+    }
+    if (var_f2 > co->walk_max_vel) {
+        var_f1 *= p_ftCommonData->x6C;
+    }
+    ft_80085030(gobj, var_f1, fp->facing_dir);
+}
+
+void ft_80085004(Fighter_GObj* gobj)
+{
+    ftCo_DatAttrs* co_attrs;
+    Fighter* fp = GET_FIGHTER(gobj);
+    co_attrs = &fp->co_attrs;
+
+    ft_80085030(gobj, co_attrs->gr_friction, fp->facing_dir1);
+}
+
+void ft_80085030(Fighter_GObj* gobj, float gr_friction, float facing_dir)
+{
+    Fighter* fp = gobj->user_data;
+    if (fp->x594_b0) {
+        fp->xE4_ground_accel_1 =
+            fp->x6A4_transNOffset.z * facing_dir - fp->gr_vel;
+    } else {
+        ftCommon_ApplyFrictionGround(fp, gr_friction);
+    }
+    ftCommon_ApplyGroundMovement(gobj);
+}
+
+void ft_80085088(Fighter_GObj* gobj)
+{
+    ftCo_DatAttrs* co_attrs;
+    Fighter* fp = GET_FIGHTER(gobj);
+    co_attrs = &fp->co_attrs;
+
+    ft_800850E0(gobj, co_attrs->gr_friction, fp->facing_dir);
+}
+
+void ft_800850B4(Fighter_GObj* gobj)
+{
+    ftCo_DatAttrs* co_attrs;
+    Fighter* fp = GET_FIGHTER(gobj);
+    co_attrs = &fp->co_attrs;
+
+    ft_800850E0(gobj, co_attrs->gr_friction, fp->facing_dir1);
+}
+
+void ft_800850E0(Fighter_GObj* gobj, float arg8, float arg9)
+{
+    Fighter* fp = gobj->user_data;
+    if (fp->x594_b0) {
+        fp->gr_vel = fp->x6A4_transNOffset.z * arg9;
+    } else {
+        ftCommon_ApplyFrictionGround(fp, arg8);
+    }
+    ftCommon_ApplyGroundMovement(gobj);
+}
+
+void ft_80085134(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->self_vel.x = fp->x6A4_transNOffset.z * fp->facing_dir;
+    fp->self_vel.y = fp->x6A4_transNOffset.y;
+}
+
+void ft_80085154(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    f32 lstick_x = cosf(fp->lstick_angle);
+    f32 lstick_y = sinf(fp->lstick_angle);
+    f32 temp_f0 = fp->x6A4_transNOffset.y;
+    f32 temp_f3 = fp->x6A4_transNOffset.z * fp->facing_dir;
+    fp->self_vel.x = (temp_f3 * lstick_x) - (temp_f0 * lstick_y);
+    fp->self_vel.y = (temp_f3 * lstick_y) + (temp_f0 * lstick_x);
+}
+
+void ft_800851C0(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->self_vel.y = fp->x6A4_transNOffset.y;
+}
+
+void ft_800851D0(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->x197C != NULL) {
+        fp->self_vel.y =
+            fp->x6A4_transNOffset.y * ((f32*) Fighter_804D6520)[5];
+    } else {
+        fp->self_vel.y = fp->x6A4_transNOffset.y;
+    }
+}
+
+void ft_80085204(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->self_vel.x = fp->x6A4_transNOffset.z * fp->facing_dir;
+}
diff --git a/src/melee/ft/ft_084E.h b/src/melee/ft/ft_084E.h
new file mode 100644
index 0000000000..123e697a16
--- /dev/null
+++ b/src/melee/ft/ft_084E.h
@@ -0,0 +1,22 @@
+#ifndef MELEE_FT_084E_H
+#define MELEE_FT_084E_H
+
+#include <melee/ft/forward.h>
+
+/* 084E1C */ void ft_80084E1C(Fighter_GObj* gobj, float, float, float);
+/* 084EEC */ void ft_80084EEC(Fighter_GObj* gobj);
+/* 084F3C */ void ft_80084F3C(Fighter_GObj* gobj);
+/* 084FA8 */ void ft_80084FA8(Fighter_GObj* gobj);
+/* 085004 */ void ft_80085004(Fighter_GObj* gobj);
+/* 085030 */ void ft_80085030(Fighter_GObj*, float gr_friction,
+                              float facing_dir);
+/* 085088 */ void ft_80085088(Fighter_GObj*);
+/* 0850B4 */ void ft_800850B4(Fighter_GObj*);
+/* 0850E0 */ void ft_800850E0(Fighter_GObj*, float, float);
+/* 085134 */ void ft_80085134(Fighter_GObj* gobj);
+/* 085154 */ void ft_80085154(Fighter_GObj* gobj);
+/* 0851C0 */ void ft_800851C0(Fighter_GObj*);
+/* 0851D0 */ void ft_800851D0(Fighter_GObj*);
+/* 085204 */ void ft_80085204(Fighter_GObj* gobj);
+
+#endif
