diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index dd107ea2c1..5991b8199f 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -357,7 +357,7 @@ lbArchive_80016EFC = .text:0x80016EFC; // type:function size:0x84 scope:global
 lbArchive_80016F80 = .text:0x80016F80; // type:function size:0xC0 scope:global
 lbArchive_80017040 = .text:0x80017040; // type:function size:0x18C scope:global
 lbArchive_800171CC = .text:0x800171CC; // type:function size:0x174 scope:global
-lbArchive_80017340 = .text:0x80017340; // type:function size:0x17C scope:global
+lbArchiveRelocate = .text:0x80017340; // type:function size:0x17C scope:global
 lbDvd_800174BC = .text:0x800174BC; // type:function size:0x2C scope:global
 lbDvd_800174E8 = .text:0x800174E8; // type:function size:0xB0 scope:global
 lbDvd_80017598 = .text:0x80017598; // type:function size:0xAC scope:global
diff --git a/src/melee/ft/ftdata.c b/src/melee/ft/ftdata.c
index 38d6a01908..98ab4e81d2 100644
--- a/src/melee/ft/ftdata.c
+++ b/src/melee/ft/ftdata.c
@@ -1921,7 +1921,7 @@ void ftData_80085CD8(Fighter* fp, Fighter* arg1, int msid)
                 {
                     memcpy(fp->x59C, temp_r3_3->x59C, temp_r3->x8);
                     temp_r4 = fp->x59C;
-                    temp_ret = lbArchive_80017340(
+                    temp_ret = lbArchiveRelocate(
                         &sp14, temp_r4->x0, temp_r3->x8,
                         (intptr_t) temp_r4 - (intptr_t) temp_r3_3->x59C);
                     if (temp_ret == -1) {
@@ -1974,7 +1974,7 @@ FigaTree* ftData_80085E50(Fighter* arg0, int msid)
                 {
                     memcpy(arg0->x59C, temp_r3_3->x59C, temp_r3->x8);
                     temp_r4 = arg0->x59C;
-                    temp_ret = lbArchive_80017340(
+                    temp_ret = lbArchiveRelocate(
                         &sp10, temp_r4->x0, temp_r3->x8,
                         (intptr_t) temp_r4 - (intptr_t) temp_r3_3->x59C);
                     if (temp_ret == -1) {
diff --git a/src/melee/lb/lbarchive.c b/src/melee/lb/lbarchive.c
index 606162839c..493ce94d0d 100644
--- a/src/melee/lb/lbarchive.c
+++ b/src/melee/lb/lbarchive.c
@@ -33,23 +33,6 @@ void lbArchive_InitializeDAT(HSD_Archive* archive, void* data, size_t length)
 }
 #pragma pop
 
-void lbArchive_LoadSections(HSD_Archive* archive, void** symbol, ...)
-{
-    const char* symbol_name;
-    va_list symbols;
-
-    va_start(symbols, symbol);
-    for (; symbol != NULL; symbol = va_arg(symbols, void**)) {
-        symbol_name = va_arg(symbols, const char*);
-        *symbol = NULL;
-        *symbol = HSD_ArchiveGetPublicAddress(archive, symbol_name);
-        if (*symbol == NULL) {
-            OSReport("Cannot find symbol %s.\n", symbol_name);
-        }
-    }
-    va_end(symbols);
-}
-
 static inline HSD_Archive* lbArchive_LoadArchive_inline(const char* filename)
 {
     HSD_Archive* archive;
@@ -74,7 +57,7 @@ static inline void lbArchive_vLoadSectionsFatal(HSD_Archive* archive,
     const char* symbol_name;
 
     for (; symbol != NULL; symbol = va_arg(symbols, void**)) {
-        const char* symbol_name = va_arg(symbols, const char*);
+        symbol_name = va_arg(symbols, const char*);
         *symbol = NULL;
         *symbol = HSD_ArchiveGetPublicAddress(archive, symbol_name);
         if (*symbol == NULL) {
@@ -180,7 +163,7 @@ bool lbArchive_80017040(HSD_Archive** dst, const char* filename, void* symbols,
     bool preloaded;
     va_list args;
 
-    va_start(args, filename);
+    va_start(args, symbols);
 
     archive = lbDvd_8001819C(filename);
     if (archive != NULL) {
@@ -207,7 +190,7 @@ bool lbArchive_800171CC(HSD_Archive** dst, const char* filename, void* symbols,
     bool preloaded;
     va_list args;
 
-    va_start(args, filename);
+    va_start(args, symbols);
 
     archive = lbDvd_8001819C(filename);
     if (archive != NULL) {
@@ -238,8 +221,8 @@ inline void Locate(HSD_Archive* archive, intptr_t base_addr)
     }
 }
 
-int lbArchive_80017340(HSD_Archive* archive, u8* src, size_t file_size,
-                       intptr_t base_addr)
+int lbArchiveRelocate(HSD_Archive* archive, u8* src, size_t file_size,
+                      intptr_t base_addr)
 {
     size_t file_offset;
 
diff --git a/src/melee/lb/lbarchive.h b/src/melee/lb/lbarchive.h
index efc7775598..e9562bb114 100644
--- a/src/melee/lb/lbarchive.h
+++ b/src/melee/lb/lbarchive.h
@@ -18,7 +18,6 @@ bool lbArchive_80017040(HSD_Archive** dst, const char* filename, void* symbols,
                         ...);
 bool lbArchive_800171CC(HSD_Archive** dst, const char* filename, void* symbols,
                         ...);
-int lbArchive_80017340(HSD_Archive*, u8*, size_t file_size,
-                       intptr_t base_addr);
+int lbArchiveRelocate(HSD_Archive*, u8*, size_t file_size, intptr_t base_addr);
 
 #endif
