diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 42b6633273..34227166a9 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -15029,15 +15029,15 @@ it_802AE8A8 = .text:0x802AE8A8; // type:function size:0xEC scope:global
 it_802AE994 = .text:0x802AE994; // type:function size:0x120 scope:global
 it_802AEAB4 = .text:0x802AEAB4; // type:function size:0x4C scope:global
 itFoxBlaster_Logic96_PickedUp = .text:0x802AEB00; // type:function size:0xCC scope:global
-itFoxblaster_UnkMotion8_Anim = .text:0x802AEBCC; // type:function size:0x308 scope:global
-itFoxblaster_UnkMotion8_Phys = .text:0x802AEED4; // type:function size:0x34 scope:global
-itFoxblaster_UnkMotion8_Coll = .text:0x802AEF08; // type:function size:0x8 scope:global
-itFoxblaster_UnkMotion9_Anim = .text:0x802AEF10; // type:function size:0xA8 scope:global
-itFoxblaster_UnkMotion9_Phys = .text:0x802AEFB8; // type:function size:0x4 scope:global
-itFoxblaster_UnkMotion9_Coll = .text:0x802AEFBC; // type:function size:0xA8 scope:global
-itFoxblaster_UnkMotion10_Anim = .text:0x802AF064; // type:function size:0x114 scope:global
-itFoxblaster_UnkMotion10_Phys = .text:0x802AF178; // type:function size:0x4 scope:global
-itFoxblaster_UnkMotion10_Coll = .text:0x802AF17C; // type:function size:0x8 scope:global
+itFoxblaster_UnkMotion8_Anim = .text:0x802AEBCC; // type:function size:0x308 scope:local
+itFoxblaster_UnkMotion8_Phys = .text:0x802AEED4; // type:function size:0x34 scope:local
+itFoxblaster_UnkMotion8_Coll = .text:0x802AEF08; // type:function size:0x8 scope:local
+itFoxblaster_UnkMotion9_Anim = .text:0x802AEF10; // type:function size:0xA8 scope:local
+itFoxblaster_UnkMotion9_Phys = .text:0x802AEFB8; // type:function size:0x4 scope:local
+itFoxblaster_UnkMotion9_Coll = .text:0x802AEFBC; // type:function size:0xA8 scope:local
+itFoxblaster_UnkMotion10_Anim = .text:0x802AF064; // type:function size:0x114 scope:local
+itFoxblaster_UnkMotion10_Phys = .text:0x802AF178; // type:function size:0x4 scope:local
+itFoxblaster_UnkMotion10_Coll = .text:0x802AF17C; // type:function size:0x8 scope:local
 itFoxBlaster_Logic96_EvtUnk = .text:0x802AF184; // type:function size:0x20 scope:global
 it_802AF1A4 = .text:0x802AF1A4; // type:function size:0xF4 scope:global
 itLinkBow_Logic100_Destroyed = .text:0x802AF298; // type:function size:0x6C scope:global
@@ -20177,8 +20177,8 @@ it_803B8680 = .rodata:0x803B8680; // type:object size:0xC scope:global data:4byt
 it_803B868C = .rodata:0x803B868C; // type:object size:0xC scope:global data:4byte
 it_803B8698 = .rodata:0x803B8698; // type:object size:0xC scope:global data:4byte
 it_803B86A4 = .rodata:0x803B86A4; // type:object size:0xC scope:global data:4byte
-it_803B86B0 = .rodata:0x803B86B0; // type:object size:0xC scope:global data:4byte
-it_803B86BC = .rodata:0x803B86BC; // type:object size:0xC scope:global data:4byte
+@187 = .rodata:0x803B86B0; // type:object size:0xC scope:local data:4byte
+@435 = .rodata:0x803B86BC; // type:object size:0xC scope:local data:4byte
 it_803B86C8 = .rodata:0x803B86C8; // type:object size:0xC scope:local data:float
 zero_vec = .rodata:0x803B86D8; // type:object size:0xC scope:global data:4byte
 it_803B86E8 = .rodata:0x803B86E8; // type:object size:0xC scope:global data:4byte
@@ -23629,7 +23629,8 @@ it_803F57B0 = .data:0x803F57B0; // type:object size:0x50 scope:global
 it_803F5800 = .data:0x803F5800; // type:object size:0x50 scope:global
 it_803F5850 = .data:0x803F5850; // type:object size:0x90 scope:global
 it_803F58E0 = .data:0x803F58E0; // type:object size:0x80 scope:global
-it_803F5960 = .data:0x803F5960; // type:object size:0x25 scope:global data:string
+...data.0 = .data:0x803F58E0; // type:label scope:local
+@327 = .data:0x803F5960; // type:object size:0x25 scope:local data:string
 it_803F5988 = .data:0x803F5988; // type:object size:0x70 scope:global
 it_803F59F8 = .data:0x803F59F8; // type:object size:0x90 scope:global
 it_803F5A88 = .data:0x803F5A88; // type:object size:0x25 scope:global data:string
@@ -23711,11 +23712,14 @@ it_803F6C68 = .data:0x803F6C68; // type:object size:0x20 scope:global
 it_803F6C88 = .data:0x803F6C88; // type:object size:0x10 scope:global
 it_803F6C98 = .data:0x803F6C98; // type:object size:0x10 scope:global
 it_803F6CA8 = .data:0x803F6CA8; // type:object size:0xB0 scope:global
-lbl_803F6D58 = .data:0x803F6D58; // type:object size:0xA8
-lbl_803F6E00 = .data:0x803F6E00; // type:object size:0x25 data:string
+...data.0 = .data:0x803F6CA8; // type:label scope:local
+it_803F6D58 = .data:0x803F6D58; // type:object size:0x38 scope:global
+it_803F6D90 = .data:0x803F6D90; // type:object size:0x38 scope:global
+it_803F6DC8 = .data:0x803F6DC8; // type:object size:0x38 scope:global
+@284 = .data:0x803F6E00; // type:object size:0x25 scope:local data:string
 it_803F6E28 = .data:0x803F6E28; // type:object size:0x14 scope:global
 it_803F6E3C = .data:0x803F6E3C; // type:object size:0x2C scope:global
-it_803F6E68 = .data:0x803F6E68; // type:object size:0x30 scope:global
+it_803F6E68 = .data:0x803F6E68; // type:object size:0x2C scope:global
 it_803F6E98 = .data:0x803F6E98; // type:object size:0x70 scope:global
 it_803F6F08 = .data:0x803F6F08; // type:object size:0x1C scope:local
 it_803F6F24 = .data:0x803F6F24; // type:object size:0x1C scope:local
@@ -23727,7 +23731,7 @@ it_803F70B8 = .data:0x803F70B8; // type:object size:0x10 scope:global
 it_803F70C8 = .data:0x803F70C8; // type:object size:0x30 scope:global
 it_803F70F8 = .data:0x803F70F8; // type:object size:0x20 scope:global
 it_803F7118 = .data:0x803F7118; // type:object size:0x30 scope:global
-it_803F7148 = .data:0x803F7148; // type:object size:0xA scope:global data:string
+@164 = .data:0x803F7148; // type:object size:0xA scope:local data:string
 it_803F7158 = .data:0x803F7158; // type:object size:0x10 scope:global
 @240 = .data:0x803F7168; // type:object size:0x25 scope:local data:string
 it_803F7190 = .data:0x803F7190; // type:object size:0x20 scope:global
@@ -27081,8 +27085,8 @@ it_804D5208 = .sdata:0x804D5208; // type:object size:0x5 scope:global data:strin
 @237 = .sdata:0x804D5228; // type:object size:0x5 scope:local data:string
 @254 = .sdata:0x804D5230; // type:object size:0x7 scope:local data:string
 @255 = .sdata:0x804D5238; // type:object size:0x5 scope:local data:string
-it_804D5240 = .sdata:0x804D5240; // type:object size:0x7 scope:global data:string
-it_804D5248 = .sdata:0x804D5248; // type:object size:0x5 scope:global data:string
+@323 = .sdata:0x804D5240; // type:object size:0x7 scope:local data:string
+@324 = .sdata:0x804D5248; // type:object size:0x5 scope:local data:string
 it_804D5250 = .sdata:0x804D5250; // type:object size:0x7 scope:global data:string
 it_804D5258 = .sdata:0x804D5258; // type:object size:0x5 scope:global data:string
 @247 = .sdata:0x804D5260; // type:object size:0x7 scope:local data:string
@@ -27142,8 +27146,8 @@ it_804D53C0 = .sdata:0x804D53C0; // type:object size:0x7 scope:global data:strin
 @280 = .sdata:0x804D5410; // type:object size:0x5 scope:local data:string
 @272 = .sdata:0x804D5418; // type:object size:0x7 scope:local data:string
 @273 = .sdata:0x804D5420; // type:object size:0x5 scope:local data:string
-it_804D5428 = .sdata:0x804D5428; // type:object size:0x7 scope:global data:string
-it_804D5430 = .sdata:0x804D5430; // type:object size:0x5 scope:global data:string
+@282 = .sdata:0x804D5428; // type:object size:0x7 scope:local data:string
+@283 = .sdata:0x804D5430; // type:object size:0x5 scope:local data:string
 @278 = .sdata:0x804D5438; // type:object size:0x7 scope:local data:string
 @279 = .sdata:0x804D5440; // type:object size:0x5 scope:local data:string
 @206 = .sdata:0x804D5448; // type:object size:0x7 scope:local data:string
@@ -27154,8 +27158,8 @@ it_804D5460 = .sdata:0x804D5460; // type:object size:0x5 scope:global data:strin
 @225 = .sdata:0x804D5470; // type:object size:0x5 scope:local data:string
 it_804D5478 = .sdata:0x804D5478; // type:object size:0x7 scope:global data:string
 it_804D5480 = .sdata:0x804D5480; // type:object size:0x5 scope:global data:string
-it_804D5488 = .sdata:0x804D5488; // type:object size:0x7 scope:global data:string
-it_804D5490 = .sdata:0x804D5490; // type:object size:0x5 scope:global data:string
+@162 = .sdata:0x804D5488; // type:object size:0x7 scope:local data:string
+@163 = .sdata:0x804D5490; // type:object size:0x5 scope:local data:string
 @238 = .sdata:0x804D5498; // type:object size:0x7 scope:local data:string
 @239 = .sdata:0x804D54A0; // type:object size:0x5 scope:local data:string
 @181 = .sdata:0x804D54A8; // type:object size:0x7 scope:local data:string
@@ -27177,8 +27181,8 @@ it_804D5520 = .sdata:0x804D5520; // type:object size:0x7 scope:global data:strin
 it_804D5528 = .sdata:0x804D5528; // type:object size:0x5 scope:global data:string
 @208 = .sdata:0x804D5530; // type:object size:0x7 scope:local data:string
 @209 = .sdata:0x804D5538; // type:object size:0x5 scope:local data:string
-it_804D5540 = .sdata:0x804D5540; // type:object size:0x7 scope:global data:string
-it_804D5548 = .sdata:0x804D5548; // type:object size:0x5 scope:global data:string
+@259 = .sdata:0x804D5540; // type:object size:0x7 scope:local data:string
+@260 = .sdata:0x804D5548; // type:object size:0x5 scope:local data:string
 @243 = .sdata:0x804D5550; // type:object size:0x7 scope:local data:string
 @244 = .sdata:0x804D5558; // type:object size:0x5 scope:local data:string
 @191 = .sdata:0x804D5560; // type:object size:0x7 scope:local data:string
@@ -27189,8 +27193,8 @@ it_804D5578 = .sdata:0x804D5578; // type:object size:0x5 scope:global data:strin
 @210 = .sdata:0x804D5588; // type:object size:0x5 scope:local data:string
 @194 = .sdata:0x804D5590; // type:object size:0x7 scope:local data:string
 @195 = .sdata:0x804D5598; // type:object size:0x5 scope:local data:string
-it_804D55A0 = .sdata:0x804D55A0; // type:object size:0x7 scope:global data:string
-it_804D55A8 = .sdata:0x804D55A8; // type:object size:0x5 scope:global data:string
+@253 = .sdata:0x804D55A0; // type:object size:0x7 scope:local data:string
+@254 = .sdata:0x804D55A8; // type:object size:0x5 scope:local data:string
 @163 = .sdata:0x804D55B0; // type:object size:0x7 scope:local data:string
 @164 = .sdata:0x804D55B8; // type:object size:0x5 scope:local data:string
 @202 = .sdata:0x804D55C0; // type:object size:0x7 scope:local data:string
@@ -27246,9 +27250,9 @@ it_804D5730 = .sdata:0x804D5730; // type:object size:0x5 scope:global data:strin
 @196 = .sdata:0x804D5750; // type:object size:0x5 scope:local data:string
 @282 = .sdata:0x804D5758; // type:object size:0x7 scope:local data:string
 @283 = .sdata:0x804D5760; // type:object size:0x5 scope:local data:string
-it_804D5768 = .sdata:0x804D5768; // type:object size:0x7 scope:global data:string
-it_804D5770 = .sdata:0x804D5770; // type:object size:0x5 scope:global data:string
-it_804D5778 = .sdata:0x804D5778; // type:object size:0x6 scope:global data:string
+@182 = .sdata:0x804D5768; // type:object size:0x7 scope:local data:string
+@183 = .sdata:0x804D5770; // type:object size:0x5 scope:local data:string
+@184 = .sdata:0x804D5778; // type:object size:0x6 scope:local data:string
 @216 = .sdata:0x804D5780; // type:object size:0x6 scope:local data:string
 @152 = .sdata:0x804D5788; // type:object size:0x6 scope:local data:string
 @260 = .sdata:0x804D5790; // type:object size:0x1 scope:local
@@ -32905,23 +32909,23 @@ it_804DC8A0 = .sdata2:0x804DC8A0; // type:object size:0x4 scope:global align:4 d
 @474 = .sdata2:0x804DC950; // type:object size:0x4 scope:local align:4 data:float
 @475 = .sdata2:0x804DC958; // type:object size:0x8 scope:local align:8 data:double
 @476 = .sdata2:0x804DC960; // type:object size:0x4 scope:local align:4 data:float
-it_804DC968 = .sdata2:0x804DC968; // type:object size:0x4 scope:global align:4 data:float
-it_804DC96C = .sdata2:0x804DC96C; // type:object size:0x4 scope:global align:4 data:float
-it_804DC970 = .sdata2:0x804DC970; // type:object size:0x4 scope:global align:4 data:float
-it_804DC974 = .sdata2:0x804DC974; // type:object size:0x4 scope:global align:4 data:float
-it_804DC978 = .sdata2:0x804DC978; // type:object size:0x4 scope:global align:4 data:float
-it_804DC980 = .sdata2:0x804DC980; // type:object size:0x8 scope:global align:8 data:double
-it_804DC988 = .sdata2:0x804DC988; // type:object size:0x4 scope:global align:4 data:float
-it_804DC98C = .sdata2:0x804DC98C; // type:object size:0x4 scope:global align:4 data:float
-it_804DC990 = .sdata2:0x804DC990; // type:object size:0x4 scope:global align:4 data:float
-it_804DC994 = .sdata2:0x804DC994; // type:object size:0x4 scope:global align:4 data:float
-it_804DC998 = .sdata2:0x804DC998; // type:object size:0x4 scope:global align:4 data:float
-it_804DC99C = .sdata2:0x804DC99C; // type:object size:0x4 scope:global align:4 data:float
-it_804DC9A0 = .sdata2:0x804DC9A0; // type:object size:0x4 scope:global align:4 data:float
-it_804DC9A4 = .sdata2:0x804DC9A4; // type:object size:0x4 scope:global align:4 data:float
-it_804DC9A8 = .sdata2:0x804DC9A8; // type:object size:0x4 scope:global align:4 data:float
-it_804DC9AC = .sdata2:0x804DC9AC; // type:object size:0x4 scope:global align:4 data:float
-it_804DC9B0 = .sdata2:0x804DC9B0; // type:object size:0x8 scope:global align:8 data:double
+@149 = .sdata2:0x804DC968; // type:object size:0x4 scope:local align:4 data:float
+@150 = .sdata2:0x804DC96C; // type:object size:0x4 scope:local align:4 data:float
+@151 = .sdata2:0x804DC970; // type:object size:0x4 scope:local align:4 data:float
+@152 = .sdata2:0x804DC974; // type:object size:0x4 scope:local align:4 data:float
+@153 = .sdata2:0x804DC978; // type:object size:0x4 scope:local align:4 data:float
+@159 = .sdata2:0x804DC980; // type:object size:0x8 scope:local align:8 data:double
+@211 = .sdata2:0x804DC988; // type:object size:0x4 scope:local align:4 data:float
+@212 = .sdata2:0x804DC98C; // type:object size:0x4 scope:local align:4 data:float
+@213 = .sdata2:0x804DC990; // type:object size:0x4 scope:local align:4 data:float
+@322 = .sdata2:0x804DC994; // type:object size:0x4 scope:local align:4 data:float
+@325 = .sdata2:0x804DC998; // type:object size:0x4 scope:local align:4 data:float
+@326 = .sdata2:0x804DC99C; // type:object size:0x4 scope:local align:4 data:float
+@328 = .sdata2:0x804DC9A0; // type:object size:0x4 scope:local align:4 data:float
+@500 = .sdata2:0x804DC9A4; // type:object size:0x4 scope:local align:4 data:float
+@504 = .sdata2:0x804DC9A8; // type:object size:0x4 scope:local align:4 data:float
+@505 = .sdata2:0x804DC9AC; // type:object size:0x4 scope:local align:4 data:float
+@632 = .sdata2:0x804DC9B0; // type:object size:0x8 scope:local align:8 data:double
 @208 = .sdata2:0x804DC9B8; // type:object size:0x4 scope:local align:4 data:float
 it_804DC9C0 = .sdata2:0x804DC9C0; // type:object size:0x4 scope:global align:4 data:float
 it_804DC9C4 = .sdata2:0x804DC9C4; // type:object size:0x4 scope:global align:4 data:float
@@ -33174,7 +33178,7 @@ it_804DCE50 = .sdata2:0x804DCE50; // type:object size:0x8 scope:global align:8 d
 @203 = .sdata2:0x804DCF28; // type:object size:0x4 scope:local align:4 data:float
 @211 = .sdata2:0x804DCF2C; // type:object size:0x4 scope:local align:4 data:float
 @223 = .sdata2:0x804DCF30; // type:object size:0x4 scope:local align:4 data:float
-it_804DCF38 = .sdata2:0x804DCF38; // type:object size:0x4 scope:global align:4 data:float
+@196 = .sdata2:0x804DCF38; // type:object size:0x4 scope:local align:4 data:float
 @203 = .sdata2:0x804DCF40; // type:object size:0x4 scope:local align:4 data:float
 @331 = .sdata2:0x804DCF44; // type:object size:0x4 scope:local align:4 data:float
 @156 = .sdata2:0x804DCF48; // type:object size:0x4 scope:local align:4 data:float
@@ -33200,8 +33204,8 @@ it_804DCFA4 = .sdata2:0x804DCFA4; // type:object size:0x4 scope:global align:4 d
 it_804DCFA8 = .sdata2:0x804DCFA8; // type:object size:0x8 scope:global align:8 data:double
 it_804DCFB0 = .sdata2:0x804DCFB0; // type:object size:0x4 scope:global align:4 data:float
 @151 = .sdata2:0x804DCFB8; // type:object size:0x4 scope:local align:4 data:float
-it_804DCFC0 = .sdata2:0x804DCFC0; // type:object size:0x4 scope:global align:4 data:float
-it_804DCFC4 = .sdata2:0x804DCFC4; // type:object size:0x4 scope:global align:4 data:float
+@172 = .sdata2:0x804DCFC0; // type:object size:0x4 scope:local align:4 data:float
+@182 = .sdata2:0x804DCFC4; // type:object size:0x4 scope:local align:4 data:float
 @214 = .sdata2:0x804DCFC8; // type:object size:0x4 scope:local align:4 data:float
 @237 = .sdata2:0x804DCFCC; // type:object size:0x4 scope:local align:4 data:float
 @184 = .sdata2:0x804DCFD0; // type:object size:0x4 scope:local align:4 data:float
@@ -33299,10 +33303,10 @@ it_804DD180 = .sdata2:0x804DD180; // type:object size:0x8 scope:global align:8 d
 @174 = .sdata2:0x804DD1F4; // type:object size:0x4 scope:local align:4 data:float
 @175 = .sdata2:0x804DD1F8; // type:object size:0x8 scope:local align:8 data:double
 @176 = .sdata2:0x804DD200; // type:object size:0x8 scope:local align:8 data:double
-it_804DD208 = .sdata2:0x804DD208; // type:object size:0x4 scope:global align:4 data:float
-it_804DD20C = .sdata2:0x804DD20C; // type:object size:0x4 scope:global align:4 data:float
-it_804DD210 = .sdata2:0x804DD210; // type:object size:0x4 scope:global align:4 data:float
-it_804DD214 = .sdata2:0x804DD214; // type:object size:0x4 scope:global align:4 data:float
+@241 = .sdata2:0x804DD208; // type:object size:0x4 scope:local align:4 data:float
+@385 = .sdata2:0x804DD20C; // type:object size:0x4 scope:local align:4 data:float
+@417 = .sdata2:0x804DD210; // type:object size:0x4 scope:local align:4 data:float
+@513 = .sdata2:0x804DD214; // type:object size:0x4 scope:local align:4 data:float
 @171 = .sdata2:0x804DD218; // type:object size:0x4 scope:local align:4 data:float
 @198 = .sdata2:0x804DD21C; // type:object size:0x4 scope:local align:4 data:float
 @199 = .sdata2:0x804DD220; // type:object size:0x8 scope:local align:8 data:double
@@ -33346,8 +33350,8 @@ it_804DD2BC = .sdata2:0x804DD2BC; // type:object size:0x4 scope:global align:4 d
 @165 = .sdata2:0x804DD300; // type:object size:0x4 scope:local align:4 data:float
 @197 = .sdata2:0x804DD304; // type:object size:0x4 scope:local align:4 data:float
 @277 = .sdata2:0x804DD308; // type:object size:0x4 scope:local align:4 data:float
-it_804DD310 = .sdata2:0x804DD310; // type:object size:0x4 scope:global align:4 data:float
-it_804DD314 = .sdata2:0x804DD314; // type:object size:0x4 scope:global align:4 data:float
+@159 = .sdata2:0x804DD310; // type:object size:0x4 scope:local align:4 data:float
+@171 = .sdata2:0x804DD314; // type:object size:0x4 scope:local align:4 data:float
 @177 = .sdata2:0x804DD318; // type:object size:0x4 scope:local align:4 data:float
 @187 = .sdata2:0x804DD31C; // type:object size:0x4 scope:local align:4 data:float
 @188 = .sdata2:0x804DD320; // type:object size:0x4 scope:local align:4 data:float
@@ -33657,9 +33661,9 @@ it_804DD988 = .sdata2:0x804DD988; // type:object size:0x8 scope:global align:8 d
 @336 = .sdata2:0x804DD9C4; // type:object size:0x4 scope:local align:4 data:float
 @338 = .sdata2:0x804DD9C8; // type:object size:0x8 scope:local align:8 data:double
 @415 = .sdata2:0x804DD9D0; // type:object size:0x4 scope:local align:4 data:float
-it_804DD9D8 = .sdata2:0x804DD9D8; // type:object size:0x4 scope:global align:4 data:float
-it_804DD9DC = .sdata2:0x804DD9DC; // type:object size:0x4 scope:global align:4 data:float
-it_804DD9E0 = .sdata2:0x804DD9E0; // type:object size:0x8 scope:global align:8 data:double
+@252 = .sdata2:0x804DD9D8; // type:object size:0x4 scope:local align:4 data:float
+@253 = .sdata2:0x804DD9DC; // type:object size:0x4 scope:local align:4 data:float
+@255 = .sdata2:0x804DD9E0; // type:object size:0x8 scope:local align:8 data:double
 @226 = .sdata2:0x804DD9E8; // type:object size:0x4 scope:local align:4 data:float
 @227 = .sdata2:0x804DD9EC; // type:object size:0x4 scope:local align:4 data:float
 @232 = .sdata2:0x804DD9F0; // type:object size:0x4 scope:local align:4 data:float
diff --git a/configure.py b/configure.py
index a3f848eac5..b5a71a253e 100755
--- a/configure.py
+++ b/configure.py
@@ -1116,7 +1116,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itbat.c"),
             Object(Matching, "melee/it/items/itsword.c"),
             Object(Matching, "melee/it/items/itbox.c"),
-            Object(NonMatching, "melee/it/items/ittaru.c"),
+            Object(Matching, "melee/it/items/ittaru.c"),
             Object(Matching, "melee/it/items/itegg.c"),
             Object(NonMatching, "melee/it/items/itkusudama.c"),
             Object(Matching, "melee/it/items/itparasol.c"),
@@ -1168,7 +1168,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itnessbat.c"),
             Object(Matching, "melee/it/items/it_2ADA.c"),
             Object(Matching, "melee/it/items/itkirbyhammer.c"),
-            Object(NonMatching, "melee/it/items/itfoxblaster.c"),
+            Object(Matching, "melee/it/items/itfoxblaster.c"),
             Object(Matching, "melee/it/items/itlinkbow.c"),
             Object(Matching, "melee/it/items/itnesspkflashexplode.c"),
             Object(NonMatching, "melee/it/items/itseakneedlethrown.c"),
@@ -1176,7 +1176,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itseakvanish.c"),
             Object(NonMatching, "melee/it/items/itpikachuthunder.c"),
             Object(Matching, "melee/it/items/itmariocape.c"),
-            Object(NonMatching, "melee/it/items/ityoshieggthrow.c"),
+            Object(Matching, "melee/it/items/ityoshieggthrow.c"),
             Object(Matching, "melee/it/items/ityoshistar.c"),
             Object(Matching, "melee/it/items/itpikachutjoltground.c"),
             Object(Matching, "melee/it/items/itpikachutjoltair.c"),
@@ -1195,7 +1195,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itdrmariopill.c"),
             Object(Matching, "melee/it/items/itclimbersice.c"),
             Object(Matching, "melee/it/items/itclimbersblizzard.c"),
-            Object(NonMatching, "melee/it/items/itclimbersstring.c"),
+            Object(Matching, "melee/it/items/itclimbersstring.c"),
             Object(Matching, "melee/it/items/itzeldadinfire.c"),
             Object(Matching, "melee/it/items/itzeldadinfireexplode.c"),
             Object(Matching, "melee/it/items/itmewtwodisable.c"),
@@ -1211,7 +1211,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itgamewatchpanic.c"),
             Object(Matching, "melee/it/items/itgamewatchrescue.c"),
             Object(Matching, "melee/it/items/itgamewatchchef.c"),
-            Object(NonMatching, "melee/it/items/itclinkmilk.c"),
+            Object(Matching, "melee/it/items/itclinkmilk.c"),
             # Pokémon
             Object(Matching, "melee/it/items/ittosakinto.c"),
             Object(Matching, "melee/it/items/itchicorita.c"),
@@ -1272,7 +1272,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itmasterhandbullet.c"),
             Object(Matching, "melee/it/items/itcrazyhandbomb.c"),
             Object(Matching, "melee/it/items/itcoin.c"),
-            Object(NonMatching, "melee/it/items/itkirby_2F23.c"),
+            Object(Matching, "melee/it/items/itkirby_2F23.c"),
             Object(Matching, "melee/it/items/it_2F28.c"),
             Object(Matching, "melee/it/items/ityoshitongue.c"),
             Object(Matching, "melee/it/items/itkirbyyoshispecialn.c"),
diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index 2886bdc4e5..0e4c8b76ae 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -143,40 +143,21 @@ typedef struct {
                        // ____but not yet invisible)
                        // 0: when xDD8 <= 0 or xDD8 >= 5
                        // 1: when cmd_var3 = 1 (when gun shooting sfx starts)
-    /* xC */ bool xDE0; // false when not shot yet; true after shot (in this
-                        // spawn instance of the blaster - not put away)
-    /* x10 */ s32 xDE4; // group 1; gets set to 0 in func (it_802ADF10) that
-                        // sets item joint locations from corresponding fighter
-                        // joint locations
-    /* x14 */ s32 xDE8; // group 1
-    /* x18 */ s32 xDEC; // group 1
-    /* x1C */ s32 xDF0; // group 1
-    /* x20 */ s32 xDF4; // group 1
-    /* x24 */ s32 xDF8; // group 1
-    /* x28 */ s32 xDFC; // group 2; gets set to 0 in func (it_802ADF10) that
-                        // sets item joint locations from corresponding fighter
-                        // joint locations
-    /* x2C */ s32 xE00; // group 2
-    /* x30 */ s32 xE04; // group 2
-    /* x34 */ s32 xE08; // group 2
-    /* x38 */ s32 xE0C; // group 2
-    /* x3C */ s32 xE10; // group 2
-    /* x40 */ Vec3 xE14; // group 3; gets set as the vector from fighter's
-                         // current position to the joint holding the blaster
-                         // on the frame blaster its shot
-    /* x4C */ Vec3 xE20; // group 3
-    /* x58 */ Vec3 xE2C; // group 3
-    /* x64 */ Vec3 xE38; // group 3
-    /* x70 */ Vec3 xE44; // group 3
-    /* x7C */ Vec3 xE50; // group 3
-    /* x88 */ f32 angle; // xE5C group 4; gets set to an angle value in func
-                         // (it_802ADF10) that sets item joint locations from
-                         // corresponding fighter joint locations
-    /* x8C */ f32 xE60;  // group 4
-    /* x90 */ f32 xE64;  // group 4
-    /* x94 */ f32 xE68;  // group 4
-    /* x98 */ f32 xE6C;  // group 4
-    /* x9C */ f32 xE70;  // group 4
+    /* xC */ bool xDE0;    // false when not shot yet; true after shot (in this
+                           // spawn instance of the blaster - not put away)
+    /* x10 */ s32 xDE4[6]; // group 1; gets set to 0 in func (it_802ADF10)
+                           // that sets item joint locations from
+                           // corresponding fighter joint locations
+    /* x28 */ s32 xDFC[6]; // group 2; gets set to 0 in func (it_802ADF10)
+                           // that sets item joint locations from
+                           // corresponding fighter joint locations
+    /* x40 */ Vec3 xE14[6]; // group 3; gets set as the vector from fighter's
+                            // current position to the joint holding the
+                            // blaster on the frame blaster its shot
+    /* x88 */ f32 angle[6]; // xE5C group 4; angle[0] gets set to an angle
+                            // value in func (it_802ADF10) that sets item
+                            // joint locations from corresponding fighter
+                            // joint locations
     /* x100 */ bool gfx_spawn_var; // xE74 Signals to spawn shoot gfx from
                                    // blaster when set to true, which is done
                                    // from subaction funcs on frame of shot
diff --git a/src/melee/it/items/itclimbersstring.c b/src/melee/it/items/itclimbersstring.c
index 9abb8ec5d7..1c672530de 100644
--- a/src/melee/it/items/itclimbersstring.c
+++ b/src/melee/it/items/itclimbersstring.c
@@ -1,31 +1,16 @@
 #include "itclimbersstring.h"
 
-#include "placeholder.h"
-
-#include "dolphin/types.h"
 #include "ft/chara/ftPopo/ftPp_SpecialS.h"
-#include "ft/chara/ftPopo/types.h"
 #include "ft/ftlib.h"
 #include "ft/inlines.h"
 #include "it/inlines.h"
-#include "it/it_26B1.h"
 #include "it/it_2725.h"
-#include "it/itCharItems.h"
 #include "it/item.h"
 #include "it/items/itlinkhookshot.h"
 #include "lb/lbvector.h"
 
-#include <dolphin/mtx.h>
-#include <dolphin/os/OSError.h>
-#include <baselib/gobj.h>
-#include <baselib/gobjgxlink.h>
 #include <baselib/gobjobject.h>
 #include <baselib/gobjplink.h>
-#include <baselib/gobjuserdata.h>
-#include <baselib/jobj.h>
-
-extern const Vec3 it_803B86B0;
-extern const Vec3 it_803B86BC;
 
 ItemStateTable it_803F76B8[] = {
     { -1, itClimbersstring_UnkMotion3_Anim, NULL, NULL },
@@ -34,6 +19,28 @@ ItemStateTable it_803F76B8[] = {
     { -1, itClimbersstring_UnkMotion3_Anim, NULL, NULL },
 };
 
+static inline void it_802C248C_setupGObj(HSD_GObj* link_gobj,
+                                         HSD_JObj* loaded_jobj,
+                                         GObj_RenderFunc callback)
+{
+    HSD_GObjObject_80390A70(link_gobj, HSD_GObj_804D7849, loaded_jobj);
+    GObj_SetupGXLink(link_gobj, callback, 6, 0);
+}
+
+static inline HSD_Joint* it_802C248C_joint24(Item* ip)
+{
+    return (((itClimbersStringAttributes*)
+                 ip->xC4_article_data->x4_specialAttributes)
+                ->x24_joint);
+}
+
+static inline HSD_Joint* it_802C248C_joint28(Item* ip)
+{
+    return (((itClimbersStringAttributes*)
+                 ip->xC4_article_data->x4_specialAttributes)
+                ->x28_joint);
+}
+
 static HSD_GObj* it_802C248C(Item* ip, HSD_JObj* jobj)
 {
     itClimbersStringAttributes* attrs =
@@ -42,13 +49,15 @@ static HSD_GObj* it_802C248C(Item* ip, HSD_JObj* jobj)
     ItemLink* head_link;
     ItemLink* tail_link;
     HSD_GObj* result;
-    Vec3 zero_vel = it_803B86B0;
+    HSD_GObj* link_gobj;
+    ItemLink* link;
+    HSD_JObj* joint;
+    Vec3 zero_vel = { 0.0F, 0.0F, 0.0F };
     int i;
     PAD_STACK(4);
 
     for (i = 0; i < attrs->x0_count; i++) {
-        HSD_GObj* link_gobj = GObj_Create(7, 0xA, 0);
-        ItemLink* link;
+        link_gobj = GObj_Create(7, 0xA, 0);
 
         if (link_gobj == NULL) {
             while (prev_link != NULL) {
@@ -70,13 +79,9 @@ static HSD_GObj* it_802C248C(Item* ip, HSD_JObj* jobj)
             link->vel = zero_vel;
             link->pos = ip->pos;
             link->x2C_b0 = false;
-            HSD_GObjObject_80390A70(
-                link_gobj, HSD_GObj_804D7849,
-                HSD_JObjLoadJoint(
-                    ((itClimbersStringAttributes*)
-                         ip->xC4_article_data->x4_specialAttributes)
-                        ->x24_joint));
-            GObj_SetupGXLink(link_gobj, it_802A24A0, 6, 0);
+            it_802C248C_setupGObj(link_gobj,
+                                  HSD_JObjLoadJoint(it_802C248C_joint24(ip)),
+                                  it_802A24A0);
         } else if (i == attrs->x0_count - 1) {
             prev_link->prev = link;
             link->prev = NULL;
@@ -87,17 +92,12 @@ static HSD_GObj* it_802C248C(Item* ip, HSD_JObj* jobj)
             link->vel = zero_vel;
             link->pos = ip->pos;
             link->x2C_b0 = true;
-            {
-                HSD_JObj* tail_jobj = HSD_JObjLoadJoint(
-                    ((itClimbersStringAttributes*)
-                         ip->xC4_article_data->x4_specialAttributes)
-                        ->x28_joint);
-                HSD_GObjObject_80390A70(link_gobj, HSD_GObj_804D7849,
-                                        tail_jobj);
-                GObj_SetupGXLink(link_gobj, HSD_GObj_JObjCallback, 6, 0);
-                ip->xDD4_itemVar.climbersstring.x14 = tail_jobj;
-                result = link_gobj->hsd_obj;
-            }
+
+            joint = HSD_JObjLoadJoint(it_802C248C_joint28(ip));
+            HSD_GObjObject_80390A70(link_gobj, HSD_GObj_804D7849, joint);
+            GObj_SetupGXLink(link_gobj, HSD_GObj_JObjCallback, 6, 0);
+            ip->xDD4_itemVar.climbersstring.x14 = joint;
+            result = link_gobj->hsd_obj;
         } else {
             prev_link->prev = link;
             link->next = prev_link;
@@ -106,13 +106,9 @@ static HSD_GObj* it_802C248C(Item* ip, HSD_JObj* jobj)
             link->vel = zero_vel;
             link->pos = ip->pos;
             link->x2C_b0 = false;
-            HSD_GObjObject_80390A70(
-                link_gobj, HSD_GObj_804D7849,
-                HSD_JObjLoadJoint(
-                    ((itClimbersStringAttributes*)
-                         ip->xC4_article_data->x4_specialAttributes)
-                        ->x24_joint));
-            GObj_SetupGXLink(link_gobj, it_802A24A0, 6, 0);
+            it_802C248C_setupGObj(link_gobj,
+                                  HSD_JObjLoadJoint(it_802C248C_joint24(ip)),
+                                  it_802A24A0);
         }
         prev_link = link;
     }
@@ -172,27 +168,6 @@ Item_GObj* it_802C27D4(Fighter_GObj* owner, Vec3* pos, int msid, float dir)
     return item_gobj;
 }
 
-static void it_802C33B8(Item* ip)
-{
-    Mtx m;
-    ItemLink* link = ip->xDD4_itemVar.climbersstring.x4;
-    HSD_JObj* jobj = GET_JOBJ(link->gobj);
-    Vec3 zero = it_803B86BC;
-
-    link->x2C_b0 = false;
-    HSD_JObjSetupMatrix(link->jobj);
-
-    PSMTXIdentity(m);
-    m[0][3] = zero.x;
-    m[1][3] = zero.y;
-    m[2][3] = zero.z;
-    PSMTXConcat(link->jobj->mtx, m, m);
-
-    HSD_JObjCopyMtx(jobj, m);
-    jobj->flags |= 0x03800000;
-    HSD_JObjSetMtxDirty(jobj);
-}
-
 void fn_802C28B8(Item_GObj* gobj)
 {
     it_802C33B8(GET_ITEM(gobj));
@@ -272,6 +247,49 @@ static void fn_802C2AF4(HSD_GObj* gobj)
     it_802C3520(ip, &pos);
 }
 
+static inline void itClimbersstring_Cleanup(Item_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Item* ip = GET_ITEM(gobj);
+        if (ip->owner != NULL) {
+            ftPp_SpecialS_8012114C(ip->owner);
+        }
+        ip->owner = NULL;
+        ip->xDD4_itemVar.climbersstring.xC = NULL;
+        {
+            ItemLink* cur;
+            for (cur = ip->xDD4_itemVar.climbersstring.x4; cur != NULL;) {
+                HSD_GObj* cur_gobj = cur->gobj;
+                cur = cur->next;
+                HSD_GObjPLink_80390228(cur_gobj);
+            }
+        }
+    }
+}
+
+bool itClimbersstring_UnkMotion3_Anim(Item_GObj* gobj)
+{
+    s32 should_cleanup;
+    Item* ip = GET_ITEM(gobj);
+
+    if (ip->owner != NULL) {
+        enum_t action = ftLib_80086C0C(ip->owner);
+        if (action >= 0x15B && action <= 0x164) {
+            should_cleanup = 0;
+        } else {
+            should_cleanup = 1;
+        }
+    } else {
+        should_cleanup = 1;
+    }
+    if (should_cleanup != 0) {
+        itClimbersstring_Cleanup(gobj);
+        return true;
+    }
+    return false;
+    PAD_STACK(8);
+}
+
 void it_802C2CA8(ItemLink* link, Vec3* target,
                  itClimbersStringAttributes* attrs, f32 length)
 {
@@ -321,65 +339,54 @@ void it_802C2DB0(ItemLink* cur, Vec3* target,
     }
 }
 
-void it_802C3520(Item* ip, Vec3* target)
+s32 it_802C2EC4(ItemLink* link, Vec3* target,
+                itClimbersStringAttributes* attrs, Item* ip)
 {
-    HSD_JObj* jobj;
-    ItemLink* link = ip->xDD4_itemVar.climbersstring.x8;
-    itClimbersStringAttributes* attrs;
-    Vec3 next_pos;
-    Vec3 prev_pos;
-    Vec3 link_pos;
+    u8 _pad[8];
     Vec3 dir;
-    Mtx m;
-    // PAD_STACK(8);
+    Vec3 dir2;
+    ItemLink* cur = link;
+    ItemLink* iter = link->next;
+    Fighter* fp = GET_FIGHTER(ip->xDD4_itemVar.climbersstring.xC);
+    PAD_STACK(4);
 
-    while (!link->x2C_b0) {
-        link = link->prev;
+    if (fp->fv.pp.x2240.x != 0.0f || fp->fv.pp.x2240.y != 0.0f) {
+        link->pos = fp->fv.pp.x2240;
     }
+    link->vel.x = link->vel.y = link->vel.z = 0.0f;
 
-    while (link != NULL) {
-        if (link->next != NULL) {
-            if (link->next->x2C_b0) {
-                next_pos = link->next->pos;
-            } else {
-                next_pos = *target;
+    while (iter != NULL) {
+        if (iter->x2C_b0) {
+            f32 dist;
+            iter->pos = *target;
+            dist = it_802A3C98(&iter->pos, &cur->pos, &dir);
+            if (dist > attrs->x8) {
+                iter->pos.x = (dir.x * attrs->x8) + cur->pos.x;
+                iter->pos.y = (dir.y * attrs->x8) + cur->pos.y;
+                iter->pos.z = (dir.z * attrs->x8) + cur->pos.z;
+            } else if (dist < attrs->xC) {
+                if (it_802A3C98(&iter->pos, target, &dir2) <= 0.1f) {
+                    iter->x2C_b0 = false;
+                } else {
+                    iter->pos.x = (dir.x * attrs->xC) + cur->pos.x;
+                    iter->pos.y = (dir.y * attrs->xC) + cur->pos.y;
+                    iter->pos.z = (dir.z * attrs->xC) + cur->pos.z;
+                }
             }
         } else {
-            next_pos = *target;
-        }
-
-        if (link->prev != NULL) {
-            prev_pos = link->prev->pos;
-        } else {
-            prev_pos = link->pos;
+            if (it_802A3C98(target, &cur->pos, &dir) > attrs->x8) {
+                iter->pos.x = (dir.x * attrs->x8) + cur->pos.x;
+                iter->pos.y = (dir.y * attrs->x8) + cur->pos.y;
+                iter->pos.z = (dir.z * attrs->x8) + cur->pos.z;
+                iter->x2C_b0 = true;
+            } else {
+                return 0;
+            }
         }
-
-        jobj = GET_JOBJ(link->gobj);
-        link_pos = link->pos;
-        HSD_JObjSetTranslate(jobj, &link_pos);
-
-        dir.x = prev_pos.x - next_pos.x;
-        dir.y = prev_pos.y - next_pos.y;
-        dir.z = prev_pos.z - next_pos.z;
-        lbVector_Normalize(&dir);
-
-        it_802A6DC8(jobj, &link_pos, &dir);
-
-        attrs = ip->xC4_article_data->x4_specialAttributes;
-        PSMTXCopy(jobj->mtx, m);
-        m[0][0] *= attrs->x8;
-        m[1][0] *= attrs->x8;
-        m[2][0] *= attrs->x8;
-        m[0][1] *= attrs->x8;
-        m[1][1] *= attrs->x8;
-        m[2][1] *= attrs->x8;
-        m[0][2] *= attrs->x8;
-        m[1][2] *= attrs->x8;
-        m[2][2] *= attrs->x8;
-        PSMTXCopy(m, jobj->mtx);
-
-        link = link->prev;
+        cur = iter;
+        iter = iter->next;
     }
+    return 2;
 }
 
 static inline void it_802C2CA8_outline(ItemLink* link, Vec3* target,
@@ -442,56 +449,6 @@ s32 it_802C30E8(ItemLink* link, Vec3* target,
     return 2;
 }
 
-s32 it_802C2EC4(ItemLink* link, Vec3* target,
-                itClimbersStringAttributes* attrs, Item* ip)
-{
-    u8 _pad[8];
-    Vec3 dir;
-    Vec3 dir2;
-    ItemLink* cur = link;
-    ItemLink* iter = link->next;
-    Fighter* fp = GET_FIGHTER(ip->xDD4_itemVar.climbersstring.xC);
-    PAD_STACK(4);
-
-    if (fp->fv.pp.x2240.x != 0.0f || fp->fv.pp.x2240.y != 0.0f) {
-        link->pos = fp->fv.pp.x2240;
-    }
-    link->vel.x = link->vel.y = link->vel.z = 0.0f;
-
-    while (iter != NULL) {
-        if (iter->x2C_b0) {
-            f32 dist;
-            iter->pos = *target;
-            dist = it_802A3C98(&iter->pos, &cur->pos, &dir);
-            if (dist > attrs->x8) {
-                iter->pos.x = (dir.x * attrs->x8) + cur->pos.x;
-                iter->pos.y = (dir.y * attrs->x8) + cur->pos.y;
-                iter->pos.z = (dir.z * attrs->x8) + cur->pos.z;
-            } else if (dist < attrs->xC) {
-                if (it_802A3C98(&iter->pos, target, &dir2) <= 0.1f) {
-                    iter->x2C_b0 = false;
-                } else {
-                    iter->pos.x = (dir.x * attrs->xC) + cur->pos.x;
-                    iter->pos.y = (dir.y * attrs->xC) + cur->pos.y;
-                    iter->pos.z = (dir.z * attrs->xC) + cur->pos.z;
-                }
-            }
-        } else {
-            if (it_802A3C98(target, &cur->pos, &dir) > attrs->x8) {
-                iter->pos.x = (dir.x * attrs->x8) + cur->pos.x;
-                iter->pos.y = (dir.y * attrs->x8) + cur->pos.y;
-                iter->pos.z = (dir.z * attrs->x8) + cur->pos.z;
-                iter->x2C_b0 = true;
-            } else {
-                return 0;
-            }
-        }
-        cur = iter;
-        iter = iter->next;
-    }
-    return 2;
-}
-
 bool it_802C32D4(ItemLink* link, Vec3* pos, itClimbersStringAttributes* attrs,
                  Item* ip, f32 dist)
 {
@@ -527,56 +484,96 @@ bool it_802C32D4(ItemLink* link, Vec3* pos, itClimbersStringAttributes* attrs,
     return true;
 }
 
-static inline void itClimbersstring_Cleanup(Item_GObj* gobj)
+void it_802C33B8(Item* ip)
 {
-    if (gobj != NULL) {
-        Item* ip = GET_ITEM(gobj);
-        if (ip->owner != NULL) {
-            ftPp_SpecialS_8012114C(ip->owner);
-        }
-        ip->owner = NULL;
-        ip->xDD4_itemVar.climbersstring.xC = NULL;
-        {
-            ItemLink* cur;
-            for (cur = ip->xDD4_itemVar.climbersstring.x4; cur != NULL;) {
-                HSD_GObj* cur_gobj = cur->gobj;
-                cur = cur->next;
-                HSD_GObjPLink_80390228(cur_gobj);
-            }
-        }
-    }
+    Mtx m;
+    ItemLink* link = ip->xDD4_itemVar.climbersstring.x4;
+    HSD_JObj* jobj = GET_JOBJ(link->gobj);
+    Vec3 zero = { 0.0F, 0.0F, 0.0F };
+
+    link->x2C_b0 = false;
+    HSD_JObjSetupMatrix(link->jobj);
+
+    PSMTXIdentity(m);
+    m[0][3] = zero.x;
+    m[1][3] = zero.y;
+    m[2][3] = zero.z;
+    PSMTXConcat(link->jobj->mtx, m, m);
+
+    HSD_JObjCopyMtx(jobj, m);
+    jobj->flags |= 0x03800000;
+    HSD_JObjSetMtxDirty(jobj);
 }
 
-bool itClimbersstring_UnkMotion3_Anim(Item_GObj* gobj)
+void it_802C3520(Item* ip, Vec3* target)
 {
-    s32 should_cleanup;
-    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* jobj;
+    ItemLink* link = ip->xDD4_itemVar.climbersstring.x8;
+    itClimbersStringAttributes* attrs;
+    Vec3 next_pos;
+    Vec3 prev_pos;
+    Vec3 link_pos;
+    Vec3 dir;
+    Mtx m;
 
-    if (ip->owner != NULL) {
-        enum_t action = ftLib_80086C0C(ip->owner);
-        if (action >= 0x15B && action <= 0x164) {
-            should_cleanup = 0;
+    while (!link->x2C_b0) {
+        link = link->prev;
+    }
+
+    while (link != NULL) {
+        if (link->next != NULL) {
+            if (link->next->x2C_b0) {
+                next_pos = link->next->pos;
+            } else {
+                next_pos = *target;
+            }
         } else {
-            should_cleanup = 1;
+            next_pos = *target;
         }
-    } else {
-        should_cleanup = 1;
-    }
-    if (should_cleanup != 0) {
-        itClimbersstring_Cleanup(gobj);
-        return true;
+
+        if (link->prev != NULL) {
+            prev_pos = link->prev->pos;
+        } else {
+            prev_pos = link->pos;
+        }
+
+        jobj = GET_JOBJ(link->gobj);
+        link_pos = link->pos;
+        HSD_JObjSetTranslate(jobj, &link_pos);
+
+        dir.x = prev_pos.x - next_pos.x;
+        dir.y = prev_pos.y - next_pos.y;
+        dir.z = prev_pos.z - next_pos.z;
+        lbVector_Normalize(&dir);
+
+        it_802A6DC8(jobj, &link_pos, &dir);
+
+        attrs = ip->xC4_article_data->x4_specialAttributes;
+        PSMTXCopy(jobj->mtx, m);
+        m[0][0] *= attrs->x8;
+        m[1][0] *= attrs->x8;
+        m[2][0] *= attrs->x8;
+        m[0][1] *= attrs->x8;
+        m[1][1] *= attrs->x8;
+        m[2][1] *= attrs->x8;
+        m[0][2] *= attrs->x8;
+        m[1][2] *= attrs->x8;
+        m[2][2] *= attrs->x8;
+        PSMTXCopy(m, jobj->mtx);
+
+        link = link->prev;
     }
-    return false;
-    PAD_STACK(8);
 }
 
-void it_2725_Logic70_EvtUnk(Item_GObj* arg0, Item_GObj* arg1)
+void it_2725_Logic70_PickedUp(Item_GObj* gobj)
 {
-    Item* ip = GET_ITEM((HSD_GObj*) arg0);
-    it_8026B894(arg0, (HSD_GObj*) arg1);
-    if (ip->xDD4_itemVar.climbersstring.xC == arg1) {
-        ip->xDD4_itemVar.climbersstring.xC = NULL;
-    }
+    short pad;
+    void (*new_var)(Item_GObj* gobj);
+    Item* ip = gobj->user_data;
+    Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    it_8026BB44(gobj);
+    new_var = &fn_802C28B8;
+    ip->on_accessory = *new_var;
 }
 
 void it_802C3810(Item_GObj* gobj)
@@ -641,13 +638,11 @@ void it_802C3950(Item_GObj* gobj)
     ip->on_accessory = fn_802C28DC;
 }
 
-void it_2725_Logic70_PickedUp(Item_GObj* gobj)
+void it_2725_Logic70_EvtUnk(Item_GObj* arg0, Item_GObj* arg1)
 {
-    short pad;
-    void (*new_var)(Item_GObj* gobj);
-    Item* ip = gobj->user_data;
-    Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
-    it_8026BB44(gobj);
-    new_var = &fn_802C28B8;
-    ip->on_accessory = *new_var;
+    Item* ip = GET_ITEM((HSD_GObj*) arg0);
+    it_8026B894(arg0, (HSD_GObj*) arg1);
+    if (ip->xDD4_itemVar.climbersstring.xC == arg1) {
+        ip->xDD4_itemVar.climbersstring.xC = NULL;
+    }
 }
diff --git a/src/melee/it/items/itclinkmilk.c b/src/melee/it/items/itclinkmilk.c
index bc20af888a..4e70d50464 100644
--- a/src/melee/it/items/itclinkmilk.c
+++ b/src/melee/it/items/itclinkmilk.c
@@ -1,15 +1,17 @@
 #include "itclinkmilk.h"
 
-#include <placeholder.h>
-#include <platform.h>
-
 #include "ft/chara/ftCLink/ftCl_Init.h"
 #include "ft/ftlib.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
 
-#include <baselib/jobj.h>
+ItemStateTable it_803F7A28[] = {
+    { 0, itClinkmilk_UnkMotion1_Anim, itClinkmilk_UnkMotion1_Phys,
+      itClinkmilk_UnkMotion1_Coll },
+    { 1, itClinkmilk_UnkMotion1_Anim, itClinkmilk_UnkMotion1_Phys,
+      itClinkmilk_UnkMotion1_Coll },
+};
 
 HSD_GObj* it_802C8B28(Item_GObj* parent, Vec3* pos, u32 bone, float facing_dir)
 {
@@ -70,10 +72,12 @@ void it_2725_Logic80_PickedUp(Item_GObj* gobj)
 
 static inline void itCLinkMilk_NotifyParent(Item_GObj* gobj)
 {
+    Item* ip2;
     if (gobj) {
         Item* ip = GET_ITEM(gobj);
         if (gobj && ip) {
-            Fighter_GObj* parent = ip->xDD4_itemVar.clinkmilk.x0;
+            // permuterslop
+            Fighter_GObj* parent = (ip2 = ip)->xDD4_itemVar.clinkmilk.x0;
             if (parent != NULL && ip->owner == parent) {
                 ftCl_Init_801492C4(parent);
             }
@@ -100,14 +104,15 @@ bool itClinkmilk_UnkMotion1_Anim(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     HSD_JObj* child = HSD_JObjGetChild(gobj->hsd_obj);
-    HSD_JObj* grandchild = HSD_JObjGetChild(child);
+    HSD_JObj* grandchild = HSD_JObjGetChild(gobj->hsd_obj);
+    u8 _pad[4];
     Vec3 sv;
-    PAD_STACK(16);
+    PAD_STACK(12);
 
     if (ip->xDAC_itcmd_var0 == 1) {
-        HSD_JObjClearFlags(child, JOBJ_HIDDEN);
+        HSD_JObjClearFlags(grandchild, JOBJ_HIDDEN);
     } else {
-        HSD_JObjSetFlags(child, JOBJ_HIDDEN);
+        HSD_JObjSetFlags(grandchild, JOBJ_HIDDEN);
     }
 
     if (ip->xDD4_itemVar.clinkmilk.x0 != NULL) {
@@ -120,7 +125,7 @@ bool itClinkmilk_UnkMotion1_Anim(Item_GObj* gobj)
             ip->xDAC_itcmd_var0 =
                 ftCl_Init_801492F4(ip->xDD4_itemVar.clinkmilk.x0);
             sv.x = sv.y = sv.z = ftLib_800869D4(ip->xDD4_itemVar.clinkmilk.x0);
-            HSD_JObjSetScale(grandchild, &sv);
+            HSD_JObjSetScale(child, &sv);
         }
     } else {
         itCLinkMilk_NotifyParent(gobj);
diff --git a/src/melee/it/items/itfoxblaster.c b/src/melee/it/items/itfoxblaster.c
index 705aa9b103..1f173be925 100644
--- a/src/melee/it/items/itfoxblaster.c
+++ b/src/melee/it/items/itfoxblaster.c
@@ -139,9 +139,6 @@ void it_802ADF10(HSD_GObj* item_gobj)
 {
     Vec3 ft_cur_pos;
     Vec3 ft_hold_joint_pos;
-    // TODO: probably a fake-improvement? idiomatic way is
-    // with the for loop, but can't get it to 100% match yet
-    int new_var2;
     Vec3 it_hold_joint_pos;
     Item* item = GET_ITEM(item_gobj);
 
@@ -158,119 +155,69 @@ void it_802ADF10(HSD_GObj* item_gobj)
         case It_Kind_Falco_Blaster:
             ftFx_SpecialN_FtGetHoldJoint(item->xDD4_itemVar.foxblaster.owner,
                                          &ft_hold_joint_pos);
-            item->xDD4_itemVar.foxblaster.xE14.x =
+            item->xDD4_itemVar.foxblaster.xE14[0].x =
                 ft_hold_joint_pos.x - ft_cur_pos.x;
-            item->xDD4_itemVar.foxblaster.xE14.y =
+            item->xDD4_itemVar.foxblaster.xE14[0].y =
                 ft_hold_joint_pos.y - ft_cur_pos.y;
-            item->xDD4_itemVar.foxblaster.xE14.z =
+            item->xDD4_itemVar.foxblaster.xE14[0].z =
                 ft_hold_joint_pos.z - ft_cur_pos.z;
             ftFx_SpecialN_ItGetHoldJoint(item->xDD4_itemVar.foxblaster.owner,
                                          &it_hold_joint_pos);
 
-            item->xDD4_itemVar.foxblaster.angle =
+            item->xDD4_itemVar.foxblaster.angle[0] =
                 atan2f(ft_hold_joint_pos.y - it_hold_joint_pos.y,
                        ft_hold_joint_pos.x - it_hold_joint_pos.x);
-            item->xDD4_itemVar.foxblaster.xDFC = 0;
+            item->xDD4_itemVar.foxblaster.xDFC[0] = 0;
 
-            item->xDD4_itemVar.foxblaster.xDE4 =
+            item->xDD4_itemVar.foxblaster.xDE4[0] =
                 (s32) efSync_Spawn(1166, item_gobj, &ft_hold_joint_pos,
-                                   &item->xDD4_itemVar.foxblaster.angle);
+                                   &item->xDD4_itemVar.foxblaster.angle[0]);
             break;
         case It_Kind_Kirby_FoxBlaster:
         case It_Kind_Kirby_FalcoBlaster:
             ftKb_SpecialNFx_800FDC00(item->xDD4_itemVar.foxblaster.owner,
                                      &ft_hold_joint_pos);
-            new_var2 = 0;
-            item->xDD4_itemVar.foxblaster.xE14.x =
+            item->xDD4_itemVar.foxblaster.xE14[0].x =
                 ft_hold_joint_pos.x - ft_cur_pos.x;
-            item->xDD4_itemVar.foxblaster.xE14.y =
+            item->xDD4_itemVar.foxblaster.xE14[0].y =
                 ft_hold_joint_pos.y - ft_cur_pos.y;
-            item->xDD4_itemVar.foxblaster.xE14.z =
+            item->xDD4_itemVar.foxblaster.xE14[0].z =
                 ft_hold_joint_pos.z - ft_cur_pos.z;
             ftKb_SpecialNFx_800FDC70(item->xDD4_itemVar.foxblaster.owner,
                                      &it_hold_joint_pos);
 
-            item->xDD4_itemVar.foxblaster.angle =
+            item->xDD4_itemVar.foxblaster.angle[0] =
                 atan2f(ft_hold_joint_pos.y - it_hold_joint_pos.y,
                        ft_hold_joint_pos.x - it_hold_joint_pos.x);
-            item->xDD4_itemVar.foxblaster.xDFC = new_var2;
+            item->xDD4_itemVar.foxblaster.xDFC[0] = 0;
 
-            item->xDD4_itemVar.foxblaster.xDE4 =
+            item->xDD4_itemVar.foxblaster.xDE4[0] =
                 (s32) efSync_Spawn(1196, item_gobj, &ft_hold_joint_pos,
-                                   &item->xDD4_itemVar.foxblaster.angle);
+                                   &item->xDD4_itemVar.foxblaster.angle[0]);
             break;
         }
 
-#if 0
         {
             int i;
-            for (i = 0; i < 5; i++) {
-                *(&item->xDD4_itemVar.foxblaster.xDF8 - i) =
-                    *(&item->xDD4_itemVar.foxblaster.xDF4 - i);
-                *(&item->xDD4_itemVar.foxblaster.xE10 - i) =
-                    *(&item->xDD4_itemVar.foxblaster.xE0C - i);
-
-                *(&item->xDD4_itemVar.foxblaster.xE50 - i) =
-                    *(&item->xDD4_itemVar.foxblaster.xE44 - i);
-
-                *(&item->xDD4_itemVar.foxblaster.xE70 - i) =
-                    *(&item->xDD4_itemVar.foxblaster.xE6C - i);
+            for (i = 5; i > 0; i--) {
+                item->xDD4_itemVar.foxblaster.xDE4[i] =
+                    item->xDD4_itemVar.foxblaster.xDE4[i - 1];
+                item->xDD4_itemVar.foxblaster.xDFC[i] =
+                    item->xDD4_itemVar.foxblaster.xDFC[i - 1];
+                item->xDD4_itemVar.foxblaster.xE14[i] =
+                    item->xDD4_itemVar.foxblaster.xE14[i - 1];
+                item->xDD4_itemVar.foxblaster.angle[i] =
+                    item->xDD4_itemVar.foxblaster.angle[i - 1];
             }
         }
-#else
-        item->xDD4_itemVar.foxblaster.xDF8 =
-            item->xDD4_itemVar.foxblaster.xDF4;
-        item->xDD4_itemVar.foxblaster.xE10 =
-            item->xDD4_itemVar.foxblaster.xE0C;
-        item->xDD4_itemVar.foxblaster.xE50 =
-            item->xDD4_itemVar.foxblaster.xE44;
-        item->xDD4_itemVar.foxblaster.xE70 =
-            item->xDD4_itemVar.foxblaster.xE6C;
-
-        item->xDD4_itemVar.foxblaster.xDF4 =
-            item->xDD4_itemVar.foxblaster.xDF0;
-        item->xDD4_itemVar.foxblaster.xE0C =
-            item->xDD4_itemVar.foxblaster.xE08;
-        item->xDD4_itemVar.foxblaster.xE44 =
-            item->xDD4_itemVar.foxblaster.xE38;
-        item->xDD4_itemVar.foxblaster.xE6C =
-            item->xDD4_itemVar.foxblaster.xE68;
-
-        item->xDD4_itemVar.foxblaster.xDF0 =
-            item->xDD4_itemVar.foxblaster.xDEC;
-        item->xDD4_itemVar.foxblaster.xE08 =
-            item->xDD4_itemVar.foxblaster.xE04;
-        item->xDD4_itemVar.foxblaster.xE38 =
-            item->xDD4_itemVar.foxblaster.xE2C;
-        item->xDD4_itemVar.foxblaster.xE68 =
-            item->xDD4_itemVar.foxblaster.xE64;
-
-        item->xDD4_itemVar.foxblaster.xDEC =
-            item->xDD4_itemVar.foxblaster.xDE8;
-        item->xDD4_itemVar.foxblaster.xE04 =
-            item->xDD4_itemVar.foxblaster.xE00;
-        item->xDD4_itemVar.foxblaster.xE2C =
-            item->xDD4_itemVar.foxblaster.xE20;
-        item->xDD4_itemVar.foxblaster.xE64 =
-            item->xDD4_itemVar.foxblaster.xE60;
-
-        item->xDD4_itemVar.foxblaster.xDE8 =
-            item->xDD4_itemVar.foxblaster.xDE4;
-        item->xDD4_itemVar.foxblaster.xE00 =
-            item->xDD4_itemVar.foxblaster.xDFC;
-        item->xDD4_itemVar.foxblaster.xE20 =
-            item->xDD4_itemVar.foxblaster.xE14;
-        item->xDD4_itemVar.foxblaster.xE60 =
-            item->xDD4_itemVar.foxblaster.angle;
-#endif
 
         item = GET_ITEM(item_gobj);
-        item->xDD4_itemVar.foxblaster.xDE4 = 0;
-        item->xDD4_itemVar.foxblaster.xDFC = new_var2;
-        item->xDD4_itemVar.foxblaster.angle =
-            item->xDD4_itemVar.foxblaster.xE14.x =
-                item->xDD4_itemVar.foxblaster.xE14.y =
-                    item->xDD4_itemVar.foxblaster.xE14.z = 0.0F;
+        item->xDD4_itemVar.foxblaster.xDE4[0] = 0;
+        item->xDD4_itemVar.foxblaster.xDFC[0] = 0;
+        item->xDD4_itemVar.foxblaster.angle[0] =
+            item->xDD4_itemVar.foxblaster.xE14[0].x =
+                item->xDD4_itemVar.foxblaster.xE14[0].y =
+                    item->xDD4_itemVar.foxblaster.xE14[0].z = 0.0F;
     }
 }
 
@@ -501,57 +448,57 @@ void it_802AE7B8(Item_GObj* item_gobj)
 
     item = GET_ITEM(item_gobj);
 
-    item->xDD4_itemVar.foxblaster.xDE4 = 0.0F;
-    item->xDD4_itemVar.foxblaster.xDFC = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE14.z = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE14.y = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE14.x = 0.0F;
-    item->xDD4_itemVar.foxblaster.angle = 0.0F;
+    item->xDD4_itemVar.foxblaster.xDE4[0] = 0.0F;
+    item->xDD4_itemVar.foxblaster.xDFC[0] = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[0].z = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[0].y = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[0].x = 0.0F;
+    item->xDD4_itemVar.foxblaster.angle[0] = 0.0F;
 
     item = GET_ITEM(item_gobj);
 
-    item->xDD4_itemVar.foxblaster.xDE8 = 0;
-    item->xDD4_itemVar.foxblaster.xE00 = 0;
-    item->xDD4_itemVar.foxblaster.xE20.z = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE20.y = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE20.x = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE60 = 0.0F;
+    item->xDD4_itemVar.foxblaster.xDE4[1] = 0;
+    item->xDD4_itemVar.foxblaster.xDFC[1] = 0;
+    item->xDD4_itemVar.foxblaster.xE14[1].z = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[1].y = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[1].x = 0.0F;
+    item->xDD4_itemVar.foxblaster.angle[1] = 0.0F;
 
     item = GET_ITEM(item_gobj);
 
-    item->xDD4_itemVar.foxblaster.xDEC = 0;
-    item->xDD4_itemVar.foxblaster.xE04 = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE2C.z = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE2C.y = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE2C.x = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE64 = 0.0F;
+    item->xDD4_itemVar.foxblaster.xDE4[2] = 0;
+    item->xDD4_itemVar.foxblaster.xDFC[2] = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[2].z = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[2].y = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[2].x = 0.0F;
+    item->xDD4_itemVar.foxblaster.angle[2] = 0.0F;
 
     item = GET_ITEM(item_gobj);
 
-    item->xDD4_itemVar.foxblaster.xDF0 = 0;
-    item->xDD4_itemVar.foxblaster.xE08 = 0;
-    item->xDD4_itemVar.foxblaster.xE38.z = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE38.y = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE38.x = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE68 = 0.0F;
+    item->xDD4_itemVar.foxblaster.xDE4[3] = 0;
+    item->xDD4_itemVar.foxblaster.xDFC[3] = 0;
+    item->xDD4_itemVar.foxblaster.xE14[3].z = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[3].y = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[3].x = 0.0F;
+    item->xDD4_itemVar.foxblaster.angle[3] = 0.0F;
 
     item = GET_ITEM(item_gobj);
 
-    item->xDD4_itemVar.foxblaster.xDF4 = 0;
-    item->xDD4_itemVar.foxblaster.xE0C = 0;
-    item->xDD4_itemVar.foxblaster.xE44.z = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE44.y = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE44.x = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE6C = 0.0F;
+    item->xDD4_itemVar.foxblaster.xDE4[4] = 0;
+    item->xDD4_itemVar.foxblaster.xDFC[4] = 0;
+    item->xDD4_itemVar.foxblaster.xE14[4].z = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[4].y = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[4].x = 0.0F;
+    item->xDD4_itemVar.foxblaster.angle[4] = 0.0F;
 
     item = GET_ITEM(item_gobj);
 
-    item->xDD4_itemVar.foxblaster.xDF8 = 0;
-    item->xDD4_itemVar.foxblaster.xE10 = 0;
-    item->xDD4_itemVar.foxblaster.xE50.z = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE50.y = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE50.x = 0.0F;
-    item->xDD4_itemVar.foxblaster.xE70 = 0.0F;
+    item->xDD4_itemVar.foxblaster.xDE4[5] = 0;
+    item->xDD4_itemVar.foxblaster.xDFC[5] = 0;
+    item->xDD4_itemVar.foxblaster.xE14[5].z = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[5].y = 0.0F;
+    item->xDD4_itemVar.foxblaster.xE14[5].x = 0.0F;
+    item->xDD4_itemVar.foxblaster.angle[5] = 0.0F;
 }
 
 /// @brief Spawns blaster item at location arg2 and gives to fighter at
diff --git a/src/melee/it/items/itkirby_2F23.c b/src/melee/it/items/itkirby_2F23.c
index 0ed8df5962..1c171a027d 100644
--- a/src/melee/it/items/itkirby_2F23.c
+++ b/src/melee/it/items/itkirby_2F23.c
@@ -1,18 +1,20 @@
 #include "itkirby_2F23.h"
 
-#include <placeholder.h>
-#include <platform.h>
-
 #include "ft/chara/ftKirby/ftKb_Init.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
-#include "MetroTRK/intrinsics.h"
-#include "MSL/math.h"
 
-#include <baselib/gobj.h>
-#include <baselib/jobj.h>
+#include <MetroTRK/intrinsics.h>
+#include <MSL/math.h>
+
+ItemStateTable it_803F9450[] = {
+    { -1, itKirby_2F23_UnkMotion0_Anim, itKirby_2F23_UnkMotion0_Phys,
+      itKirby_2F23_UnkMotion0_Coll },
+    { -1, itKirby_2F23_UnkMotion1_Anim, itKirby_2F23_UnkMotion1_Phys,
+      itKirby_2F23_UnkMotion1_Coll },
+};
 
 f64 it_802F23AC(Item_GObj* gobj, Vec3* pos)
 {
@@ -38,12 +40,6 @@ static inline void it_802F23EC_inline(Item_GObj* gobj)
     ip->xDD4_itemVar.kirby2f23.x1DC = limits.y;
     ip->xDD4_itemVar.kirby2f23.x1F4 = ftKb_SpecialN_800F5AF0(ip->grab_victim);
     HSD_JObjGetScale(jobj, &ip->xDD4_itemVar.kirby2f23.x1E8);
-    it_802756D0(gobj);
-    it_80274ECC(gobj, 1);
-    it_80274C88(gobj);
-    it_8026BD54(gobj);
-    it_80273408(gobj);
-    ip->on_accessory = it_802F258C;
 }
 
 void it_802F23EC(Item_GObj* gobj, Fighter_GObj* owner_gobj, f32 facing_dir)
@@ -59,6 +55,12 @@ void it_802F23EC(Item_GObj* gobj, Fighter_GObj* owner_gobj, f32 facing_dir)
     Item_80268E5C(gobj, 0, 0);
     it_802762BC(ip);
     it_802F23EC_inline(gobj);
+    it_802756D0(gobj);
+    it_80274ECC(gobj, 1);
+    it_80274C88(gobj);
+    it_8026BD54(gobj);
+    it_80273408(gobj);
+    ip->on_accessory = it_802F258C;
 }
 
 bool itKirby_2F23_UnkMotion0_Anim(Item_GObj* gobj)
@@ -81,10 +83,10 @@ static inline void it_802F258C_scale(Item_GObj* gobj, Vec3* offset)
 
     if (dist < ip->xDD4_itemVar.kirby2f23.x1D0) {
         Vec3 scale;
-        scale.x = scale.y = scale.z =
-            (1.0f - ip->xDD4_itemVar.kirby2f23.x1D4) +
-            ((dist / ip->xDD4_itemVar.kirby2f23.x1D0) *
-             ip->xDD4_itemVar.kirby2f23.x1D4);
+        f32 f = dist / ip->xDD4_itemVar.kirby2f23.x1D0;
+        dist = ip->xDD4_itemVar.kirby2f23.x1D4;
+        f *= dist;
+        scale.x = scale.y = scale.z = (1.0f - dist) + f;
 
         scale.x *= ip->xDD4_itemVar.kirby2f23.x1E8.x;
         scale.y *= ip->xDD4_itemVar.kirby2f23.x1E8.y;
diff --git a/src/melee/it/items/itkusudama.c b/src/melee/it/items/itkusudama.c
index 671b3f5676..8ffce6a48c 100644
--- a/src/melee/it/items/itkusudama.c
+++ b/src/melee/it/items/itkusudama.c
@@ -208,11 +208,17 @@ void it_80289BE8(Item_GObj* gobj, s32 arg1, s32 arg2, s32 arg3)
             goto food_or_random;
         }
         for (i = 0; i < count; i++) {
+            Item_GObj* spawned_gobj;
             if (kind == It_Kind_M_Ball && it_8026C704() == true) {
                 break;
             }
-            it_80289BE8_spawn(gobj, kind, 1.4f, &pos, &vel);
-            spawned[i] = kind;
+            it_80289BE8_inline(gobj, 1.4f, &pos, &vel);
+            spawned_gobj = it_8026F5C8(gobj, kind, &pos);
+            if (spawned_gobj != NULL) {
+                it_8026F53C(spawned_gobj, &vel, true);
+                it_80274ED8();
+                spawned[i] = kind;
+            }
         }
         for (; i < count; i++) {
             ItemKind rand_kind = it_8026F3AC();
@@ -220,8 +226,16 @@ void it_80289BE8(Item_GObj* gobj, s32 arg1, s32 arg2, s32 arg3)
             Vec3 pos;
             PAD_STACK(8);
             if (rand_kind != -1) {
-                prev_kind =
-                    it_80289BE8_spawn(gobj, rand_kind, 1.2f, &pos, &vel);
+                Item_GObj* spawned_gobj;
+                it_80289BE8_inline(gobj, 1.2f, &pos, &vel);
+                spawned_gobj = it_8026F5C8(gobj, rand_kind, &pos);
+                if (spawned_gobj != NULL) {
+                    it_8026F53C(spawned_gobj, &vel, true);
+                    it_80274ED8();
+                    prev_kind = rand_kind;
+                } else {
+                    prev_kind = -1;
+                }
             }
             spawned[i] = prev_kind;
         }
@@ -231,8 +245,14 @@ void it_80289BE8(Item_GObj* gobj, s32 arg1, s32 arg2, s32 arg3)
         if (i < arg2) {
             count = HSD_Randi(5) + 10;
             for (i = 0; i < count; i++) {
-                it_80289BE8_spawn(gobj, It_Kind_Foods, 1.8f, &pos, &vel);
-                spawned[i] = It_Kind_Foods;
+                Item_GObj* spawned_gobj;
+                it_80289BE8_inline(gobj, 1.8f, &pos, &vel);
+                spawned_gobj = it_8026F5C8(gobj, It_Kind_Foods, &pos);
+                if (spawned_gobj != NULL) {
+                    it_8026F53C(spawned_gobj, &vel, true);
+                    it_80274ED8();
+                    spawned[i] = It_Kind_Foods;
+                }
             }
         } else {
             ItemKind prev_kind;
@@ -242,8 +262,16 @@ void it_80289BE8(Item_GObj* gobj, s32 arg1, s32 arg2, s32 arg3)
             for (i = 0; i < count; i++) {
                 ItemKind rand_kind = it_8026F3AC();
                 if (rand_kind != -1) {
-                    prev_kind =
-                        it_80289BE8_spawn(gobj, rand_kind, 1.2f, &pos, &vel);
+                    Item_GObj* spawned_gobj;
+                    it_80289BE8_inline(gobj, 1.2f, &pos, &vel);
+                    spawned_gobj = it_8026F5C8(gobj, rand_kind, &pos);
+                    if (spawned_gobj != NULL) {
+                        it_8026F53C(spawned_gobj, &vel, true);
+                        it_80274ED8();
+                        prev_kind = rand_kind;
+                    } else {
+                        prev_kind = -1;
+                    }
                 }
                 spawned[i] = prev_kind;
             }
diff --git a/src/melee/it/items/itkyasarin.c b/src/melee/it/items/itkyasarin.c
index 89f6487583..12a6437243 100644
--- a/src/melee/it/items/itkyasarin.c
+++ b/src/melee/it/items/itkyasarin.c
@@ -274,8 +274,10 @@ static inline void itKyasarin_UnkMotion4_Coll_inline(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     itKyasarinAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
-    ip->xDD4_itemVar.kyasarin.x2C = attr->x10 + HSD_Randi(attr->x14);
-    ip->xDD4_itemVar.kyasarin.x28 = attr->x18 + HSD_Randi(attr->x1C);
+    s32 rand = HSD_Randi(attr->x14);
+    ip->xDD4_itemVar.kyasarin.x2C = attr->x10 + rand;
+    rand = HSD_Randi(attr->x1C);
+    ip->xDD4_itemVar.kyasarin.x28 = attr->x18 + rand;
     it_80273454(gobj);
     if (HSD_Randi(2) != 0) {
         ip->xDD4_itemVar.kyasarin.x34 = attr->x48;
@@ -516,7 +518,7 @@ bool it_802EDDC0(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     itKyasarinAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
-    PAD_STACK(40);
+    PAD_STACK(32);
 
     ip->xDD4_itemVar.kyasarin.x38 += ip->xCA0;
     if (ip->xDD4_itemVar.kyasarin.x38 < attr->x40) {
diff --git a/src/melee/it/items/itpikachuthunder.c b/src/melee/it/items/itpikachuthunder.c
index e34c5bf4d9..bfde2f4ecc 100644
--- a/src/melee/it/items/itpikachuthunder.c
+++ b/src/melee/it/items/itpikachuthunder.c
@@ -48,13 +48,15 @@ s32 it_802B1DEC(Item_GObj* arg0)
 Item_GObj* it_802B1DF8(Item_GObj* owner, Vec3* pos, Vec3* vel, s32 count,
                        s32 delay, s32 kind)
 {
+    int i;
+    Item* ip;
+    itPikachuthunderAttributes* attrs;
     SpawnItem spawn;
     u8 _pad[4];
     u32 x40 = Item_8026AE60();
     s32 cur_delay = 0;
     Item_GObj* prev = NULL;
-    Item_GObj* first = NULL;
-    int i;
+    Item_GObj* first;
 
     spawn.kind = kind;
     spawn.prev_pos = *pos;
@@ -74,9 +76,8 @@ Item_GObj* it_802B1DF8(Item_GObj* owner, Vec3* pos, Vec3* vel, s32 count,
             first = item_gobj;
         }
         if (item_gobj != NULL) {
-            Item* ip = GET_ITEM(item_gobj);
-            itPikachuthunderAttributes* attrs =
-                ip->xC4_article_data->x4_specialAttributes;
+            ip = GET_ITEM(item_gobj);
+            attrs = ip->xC4_article_data->x4_specialAttributes;
             if (prev != NULL) {
                 GET_ITEM(prev)->xDD4_itemVar.pikachuthunder.x34 = item_gobj;
             }
@@ -231,8 +232,9 @@ void it_802B22B8(Item_GObj* gobj)
 
 static inline f32 pika_scale(f32 a, f32 b)
 {
-    s32 n = 10000.0f * a / b + 1.0f;
-    return n / 10000.0f;
+    f32 f = 10000.0f * a / b;
+    f = ((s32) (f + 1)) / 10000.0f;
+    return f;
 }
 
 static inline void itPikachuthunder_UnkMotion2_ScaleCall(Item_GObj* gobj)
@@ -267,7 +269,7 @@ bool itPikachuthunder_UnkMotion2_Anim(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     HSD_JObj* jobj = GET_JOBJ(gobj);
-    PAD_STACK(4);
+    PAD_STACK(12);
 
     ip->xD44_lifeTimer--;
     if (ip->xD44_lifeTimer <= 0.0f) {
diff --git a/src/melee/it/items/ittaru.c b/src/melee/it/items/ittaru.c
index 6facf637a7..6bf316956e 100644
--- a/src/melee/it/items/ittaru.c
+++ b/src/melee/it/items/ittaru.c
@@ -1,10 +1,6 @@
 #include "ittaru.h"
 
-#include <placeholder.h>
-#include <platform.h>
-
 #include "baselib/jobj.h"
-#include "baselib/random.h"
 #include "cm/camera.h"
 #include "ef/efsync.h"
 #include "it/inlines.h"
@@ -12,12 +8,32 @@
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/itcoll.h"
-#include "it/itCommonItems.h"
 #include "it/item.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
 #include "lb/lbvector.h"
 
+#include <string.h>
+#include <baselib/random.h>
+
+ItemStateTable it_803F58E0[] = {
+    { -1, itTaru_UnkMotion0_Anim, itTaru_UnkMotion0_Phys,
+      itTaru_UnkMotion0_Coll },
+    { -1, itTaru_UnkMotion1_Anim, itTaru_UnkMotion1_Phys,
+      itTaru_UnkMotion1_Coll },
+    { -1, itTaru_UnkMotion2_Anim, itTaru_UnkMotion2_Phys, NULL },
+    { 0, itTaru_UnkMotion3_Anim, itTaru_UnkMotion3_Phys,
+      itTaru_UnkMotion3_Coll },
+    { -1, itTaru_UnkMotion4_Anim, itTaru_UnkMotion4_Phys,
+      itTaru_UnkMotion4_Coll },
+    { 1, itTaru_UnkMotion5_Anim, itTaru_UnkMotion5_Phys,
+      itTaru_UnkMotion5_Coll },
+    { 2, itTaru_UnkMotion6_Anim, itTaru_UnkMotion6_Phys,
+      itTaru_UnkMotion6_Coll },
+    { 3, itTaru_UnkMotion7_Anim, itTaru_UnkMotion7_Phys,
+      itTaru_UnkMotion7_Coll },
+};
+
 void it_3F14_Logic2_Spawned(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -30,8 +46,8 @@ void it_3F14_Logic2_Spawned(Item_GObj* gobj)
     ip->xDB0_itcmd_var1 = 0;
     ip->xDD4_itemVar.taru.xDD4 = 0;
     /// @todo float regswap. same as in it_3F14_Logic5_Spawned
-    temp = 0.34906584f * HSD_Randf();
-    ip->xDD4_itemVar.taru.xDDC = (new_var = 1.3089969f) + temp;
+    ip->xDD4_itemVar.taru.xDDC =
+        (new_var = 1.3089969f) + (temp = 0.34906584f * HSD_Randf());
     ip->xDD4_itemVar.taru.xDE0 = 0.0f;
     ip->xDD4_itemVar.taru.xDE4 = 0.0f;
     ip->xDD4_itemVar.taru.xDE8.x = 0.0f;
@@ -49,6 +65,12 @@ static inline f32 inline_fabsf(f32 x)
     }
 }
 
+// ensure U32_TO_F32 is in the right place
+static f32 sdata_ordering(u32 u)
+{
+    return u;
+}
+
 void it_802874F0(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -141,12 +163,13 @@ void it_80287690(Item_GObj* gobj)
         }
         HSD_JObjSetRotationZ(child, var_f31);
 
-        var_f31 = ip->xDD4_itemVar.taru.xDDC;
         rot = HSD_JObjGetRotationX(child);
-        if (rot > (0.017453292f + var_f31 + step)) {
+        if (rot > ip->xDD4_itemVar.taru.xDDC + 0.017453292f + step) {
             var_f31 = rot - step;
-        } else if (rot < ((var_f31 - 0.017453292f) - step)) {
+        } else if (rot < ip->xDD4_itemVar.taru.xDDC - 0.017453292f - step) {
             var_f31 = rot + step;
+        } else {
+            var_f31 = ip->xDD4_itemVar.taru.xDDC;
         }
         HSD_JObjSetRotationX(child, var_f31);
     } else {
@@ -519,15 +542,12 @@ bool itTaru_UnkMotion6_Coll(Item_GObj* gobj)
     return false;
 }
 
-static inline void it_802886C4_inline(Item_GObj* gobj)
+static inline void it_802886C4_inline(Item_GObj* gobj, Vec3* zero)
 {
     Item* ip = GET_ITEM(gobj);
-    Vec3 zero;
-
-    /// @todo 4 byte gap for these accesses...
-    zero.x = zero.y = zero.z = 0.0f;
-    if (!it_8026F8B4(gobj, &ip->pos, &zero, false) &&
-        it_8026F6BC(gobj, &ip->pos, &zero, false) == NULL)
+    zero->x = zero->y = zero->z = 0.0f;
+    if (!it_8026F8B4(gobj, &ip->pos, zero, false) &&
+        it_8026F6BC(gobj, &ip->pos, zero, false) == NULL)
     {
         it_8026F3D4(gobj, NULL, 1, false);
     }
@@ -535,12 +555,14 @@ static inline void it_802886C4_inline(Item_GObj* gobj)
 
 void it_802886C4(Item_GObj* gobj)
 {
+    Vec3 zero;
     Item* ip = GET_ITEM(gobj);
     HSD_JObj* jobj = GET_JOBJ(gobj);
+    PAD_STACK(4);
 
     Item_8026AE84(ip, 0xFB, 0x7F, 0x40);
     Camera_80030E44(2, &ip->pos);
-    it_802886C4_inline(gobj);
+    it_802886C4_inline(gobj, &zero);
     HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
     it_802756D0(gobj);
     ip->x40_vel.x = 0.0f;
diff --git a/src/melee/it/items/itwstar.c b/src/melee/it/items/itwstar.c
index 4fc4fb97c2..fcf8075aff 100644
--- a/src/melee/it/items/itwstar.c
+++ b/src/melee/it/items/itwstar.c
@@ -1,5 +1,7 @@
 #include "itwstar.h"
 
+#include "placeholder.h"
+
 #include "ef/efasync.h"
 #include "ef/eflib.h"
 #include "it/inlines.h"
@@ -15,33 +17,23 @@
 
 HSD_AnimJoint* it_80294364(Item_GObj* gobj)
 {
+    Item* ip = GET_ITEM(gobj);
+    itWstarAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
     s32 candidates[8];
-    s32 i;
-    Item* ip;
-    s32 last;
-    itWstarAttributes* attr;
-    s32 count;
-    s32 total;
-    s32* p;
+    s32* p = candidates;
+    s32 i = 0;
+    s32 count = i;
     s32 picked;
-
-    i = 0;
-    p = candidates;
-    ip = GET_ITEM(gobj);
-    last = it_804D6D00;
-    attr = ip->xC4_article_data->x4_specialAttributes;
-    count = 0;
-    total = attr->x24_count;
-
-    for (; i < total; i++) {
-        if (i != last) {
+    s32 end = it_804D6D00;
+    s32 var_ctr = attr->x24_count;
+    for (; i < var_ctr; i++) {
+        if (i != end) {
             *p++ = i;
             count++;
         }
     }
-
     picked = candidates[HSD_Randi(count)];
-    it_804D6D00 = (s8) picked;
+    it_804D6D00 = picked;
     ip->xDD4_itemVar.wstar.xDDC = picked;
     Item_8026AE84(ip, attr->x28_entries[ip->xDD4_itemVar.wstar.xDDC].x4_sfx,
                   0x7f, 0x40);
@@ -76,8 +68,8 @@ void itWStar_Logic29_Spawned(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     ip->xDCE_flag.b7 = 0;
-    ip->xDD4_itemVar.wstar.xDD8 = 0.0F;
-    ip->xDD4_itemVar.wstar.xDD4 = 0.0F;
+    ip->xDD4_itemVar.wstar.xDD8 = 1.0f;
+    ip->xDD4_itemVar.wstar.xDD4 = 1.0f;
     it_80294364(gobj);
 }
 
@@ -167,14 +159,16 @@ void itWStar_Logic29_Dropped(Item_GObj* gobj)
 
 void it_802947CC(Item_GObj* gobj, Vec3* pos)
 {
-    HSD_JObj* jobj = GET_JOBJ(gobj);
-    Item* ip = GET_ITEM(gobj);
+    u8 _pad[8];
+    HSD_JObj* jobj = gobj->hsd_obj;
+    Item* ip = gobj->user_data;
+    Vec3 item_pos;
     union Struct2070 saved_xD90;
-    Vec2 saved_xD94;
+    volatile u32 saved_xDA4;
     S32Vec2 saved_xD9C;
-    u32 saved_xDA4;
+    Vec2 saved_xD94;
+    u8 _pad2[4];
     u16 saved_xDA8;
-    Vec3 item_pos;
 
     itResetVelocity(ip);
     ip->xDAC_itcmd_var0 = 0;
diff --git a/src/melee/it/items/ityoshieggthrow.c b/src/melee/it/items/ityoshieggthrow.c
index e48e1ce885..6c135b6b42 100644
--- a/src/melee/it/items/ityoshieggthrow.c
+++ b/src/melee/it/items/ityoshieggthrow.c
@@ -5,14 +5,8 @@
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
-#include "it/itCommonItems.h"
 #include "it/item.h"
 
-#include "it/items/forward.h"
-
-#include <baselib/gobj.h>
-#include <baselib/jobj.h>
-
 ItemStateTable it_803F7118[] = {
     {
         -1,
@@ -59,11 +53,6 @@ void it_802B28C8(Item_GObj* gobj, Vec3* pos, Vec3* offset, f32 facing_dir,
     it_80274658(gobj, facing_dir);
 }
 
-void itYoshiEggThrow_Logic43_PickedUp(Item_GObj* gobj)
-{
-    Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
-}
-
 Item_GObj* it_802B2A10(Fighter_GObj* parent, Vec3* pos, s32 part,
                        f32 facing_dir)
 {
@@ -90,6 +79,11 @@ Item_GObj* it_802B2A10(Fighter_GObj* parent, Vec3* pos, s32 part,
     return gobj;
 }
 
+void itYoshiEggThrow_Logic43_PickedUp(Item_GObj* gobj)
+{
+    Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+}
+
 void it_802B2B08(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -112,45 +106,39 @@ bool itYoshieggthrow_UnkMotion1_Anim(Item_GObj* gobj)
     return false;
 }
 
-bool itYoshieggthrow_UnkMotion1_Coll(Item_GObj* gobj)
-{
-    it_8026E5A0(gobj, it_802B2C38);
-    return 0;
-}
-
-void it_802B2FA8(Item_GObj* gobj, Item_GObj* ref_gobj)
-{
-    it_8026B894(gobj, ref_gobj);
-}
-
-bool itYoshieggthrow_UnkMotion2_Anim(Item_GObj* gobj)
+void itYoshieggthrow_UnkMotion1_Phys(Item_GObj* gobj)
 {
-    return it_802751D8(gobj);
+    Item* ip = GET_ITEM(gobj);
+    ItemAttr* attrs = ip->xCC_item_attr;
+    it_80272860(gobj, attrs->x10_fall_speed, attrs->x14_fall_speed_max);
 }
 
-bool it_802B2E5C(Item_GObj* gobj)
+bool itYoshieggthrow_UnkMotion1_Coll(Item_GObj* gobj)
 {
-    return it_80273030(gobj);
+    it_8026E5A0(gobj, it_802B2C38);
+    return 0;
 }
 
-bool it_802B2F88(Item_GObj* gobj)
+bool it_802B2C04(Item_GObj* gobj)
 {
-    return itColl_BounceOffShield(gobj);
+    if (((Item*) gobj->user_data)->msid != 2) {
+        it_802B2C38(gobj);
+    }
+    return false;
 }
 
-void itYoshieggthrow_UnkMotion1_Phys(Item_GObj* gobj)
+static inline void spawn1(HSD_GObj* gobj, HSD_JObj* jobj)
 {
     Item* ip = GET_ITEM(gobj);
-    ItemAttr* attrs = ip->xCC_item_attr;
-    it_80272860(gobj, attrs->x10_fall_speed, attrs->x14_fall_speed_max);
+    efAsync_Spawn(gobj, &ip->xBC0, 1, 0x4CE, jobj);
 }
 
-bool it_802B2C04(Item_GObj* gobj)
+static inline void spawn2(HSD_GObj* gobj, HSD_JObj* jobj)
 {
-    if (((Item*) gobj->user_data)->msid != 2) {
-        it_802B2C38(gobj);
-    }
-    return false;
+    Item* ip = GET_ITEM(gobj);
+    f32 scale;
+    scale = 1.0f;
+    efAsync_Spawn(gobj, &GET_ITEM(gobj)->xBC0, 4, 0x4CF, jobj, &scale);
 }
 
 void it_802B2C38(HSD_GObj* gobj)
@@ -158,9 +146,7 @@ void it_802B2C38(HSD_GObj* gobj)
     Item* ip = GET_ITEM(gobj);
     itYoshiEggThrowAttributes* attrs =
         ip->xC4_article_data->x4_specialAttributes;
-    HSD_JObj* jobj = GET_JOBJ(gobj);
-    f32 scale;
-    PAD_STACK(8);
+    HSD_JObj* jobj = gobj->hsd_obj;
     Item_80268E5C(gobj, 2, 0x12);
     it_8026BB44(gobj);
     it_80273598(gobj, 0xC, 0xA);
@@ -168,28 +154,45 @@ void it_802B2C38(HSD_GObj* gobj)
     it_8027518C(gobj);
     it_80273454(gobj);
     ip->xD44_lifeTimer = attrs->x4;
-    efAsync_Spawn(gobj, &GET_ITEM(gobj)->xBC0, 1, 0x4CE, jobj);
-    scale = 1.0f;
-    efAsync_Spawn(gobj, &GET_ITEM(gobj)->xBC0, 4, 0x4CF, jobj, &scale);
+    spawn1(gobj, jobj);
+    spawn2(gobj, jobj);
     Item_8026AE84(ip, 0x44618, 0x7F, 0x40);
 }
 
+bool itYoshieggthrow_UnkMotion2_Anim(Item_GObj* gobj)
+{
+    return it_802751D8(gobj);
+}
+
 bool it_2725_Logic43_Clanked(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
-    PAD_STACK(8);
     if (ip->msid != 2) {
         it_802B2C38(gobj);
     }
     return false;
 }
 
+bool it_802B2E5C(Item_GObj* gobj)
+{
+    return it_80273030(gobj);
+}
+
 bool it_802B2E7C(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
-    PAD_STACK(4);
     if (ip->msid != 2) {
         it_802B2C38(gobj);
     }
     return false;
 }
+
+bool it_802B2F88(Item_GObj* gobj)
+{
+    return itColl_BounceOffShield(gobj);
+}
+
+void it_802B2FA8(Item_GObj* gobj, Item_GObj* ref_gobj)
+{
+    it_8026B894(gobj, ref_gobj);
+}
