diff --git a/config/GALE01/splits.txt b/config/GALE01/splits.txt
index 02be5a615e..00ae7234cf 100644
--- a/config/GALE01/splits.txt
+++ b/config/GALE01/splits.txt
@@ -252,7 +252,7 @@ melee/ef/efsync.c:
 	.sdata      start:0x804D39C8 end:0x804D39D8
 	.sdata2     start:0x804D81A0 end:0x804D81D0
 
-melee/ef/ef_061D.c:
+melee/ef/efspecial.c:
 	.text       start:0x80061D70 end:0x80063930
 	.data       start:0x803BF9D0 end:0x803BFAFC
 	.sdata      start:0x804D39D8 end:0x804D39E5
diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 7533ecf3cb..ae8ec3b552 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -1374,7 +1374,7 @@ efLib_8005FAB8 = .text:0x8005FAB8; // type:function size:0x12C scope:global
 fn_8005FBE4 = .text:0x8005FBE4; // type:function size:0xF4
 efLib_8005FCD8 = .text:0x8005FCD8; // type:function size:0x104 scope:global
 efSync_Spawn = .text:0x8005FDDC; // type:function size:0x1F94 scope:global
-ef_80061D70 = .text:0x80061D70; // type:function size:0x1BC0 scope:global
+efSpecial_SpawnSpecial = .text:0x80061D70; // type:function size:0x1BC0 scope:global
 efAsync_80063930 = .text:0x80063930; // type:function size:0x396C scope:global
 efAsync_8006729C = .text:0x8006729C; // type:function size:0x70 scope:global
 efAsync_8006730C = .text:0x8006730C; // type:function size:0x70 scope:global
diff --git a/configure.py b/configure.py
index d5ec9182bc..472e107295 100755
--- a/configure.py
+++ b/configure.py
@@ -504,7 +504,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/ef/eflib_alloc.c"),
             Object(Matching, "melee/ef/eflib.c"),
             Object(NonMatching, "melee/ef/efsync.c"),
-            Object(Matching, "melee/ef/ef_061D.c"),
+            Object(Matching, "melee/ef/efspecial.c"),
             Object(NonMatching, "melee/ef/efasync.c"),
         ],
     ),
diff --git a/src/melee/ef/ef_061D.c b/src/melee/ef/efspecial.c
similarity index 99%
rename from src/melee/ef/ef_061D.c
rename to src/melee/ef/efspecial.c
index 887e900398..f734625b6b 100644
--- a/src/melee/ef/ef_061D.c
+++ b/src/melee/ef/efspecial.c
@@ -1,4 +1,4 @@
-#include "ef_061D.h"
+#include "efspecial.h"
 
 #include "eflib.h"
 #include "types.h"
@@ -16,7 +16,7 @@
 extern volatile u32 efLib_804D64E8;
 extern volatile s32 efLib_804D64F0;
 
-void* ef_80061D70(s32 gfx_id, HSD_GObj* gobj, va_list vlist)
+void* efSpecial_SpawnSpecial(s32 gfx_id, HSD_GObj* gobj, va_list vlist)
 {
     Effect* effect;
     HSD_JObj* jobj;
diff --git a/src/melee/ef/ef_061D.h b/src/melee/ef/efspecial.h
similarity index 63%
rename from src/melee/ef/ef_061D.h
rename to src/melee/ef/efspecial.h
index 8d24d17675..6445e70ca8 100644
--- a/src/melee/ef/ef_061D.h
+++ b/src/melee/ef/efspecial.h
@@ -3,6 +3,6 @@
 
 #include <baselib/forward.h>
 
-void* ef_80061D70(s32, HSD_GObj*, va_list);
+void* efSpecial_SpawnSpecial(s32, HSD_GObj*, va_list);
 
 #endif
diff --git a/src/melee/ef/efsync.c b/src/melee/ef/efsync.c
index c75d822ae1..26f73d3bc0 100644
--- a/src/melee/ef/efsync.c
+++ b/src/melee/ef/efsync.c
@@ -20,7 +20,7 @@
 #include "ft/inlines.h"
 
 extern ef_Symbols efAsync_803C025C[51];
-extern Effect* ef_80061D70(s32, HSD_GObj*, va_list);
+extern Effect* efSpecial_SpawnSpecial(s32, HSD_GObj*, va_list);
 
 extern s32 efLib_804D64E8;
 extern s32 efLib_804D64F0;
@@ -106,8 +106,8 @@ void* efSync_Spawn(s32 arg_gfx_id, HSD_GObj* arg_gobj, ...)
         return efAsync_80063930(gfx_id, arg_gobj, vlist);
     }
     if (gfx_id < 0x4BA) {
-        // return ef_80061D70(gfx_id, arg_gobj, va_arg(vlist, void*));
-        return ef_80061D70(gfx_id, arg_gobj, vlist);
+        // return efSpecial_SpawnSpecial(gfx_id, arg_gobj, va_arg(vlist, void*));
+        return efSpecial_SpawnSpecial(gfx_id, arg_gobj, vlist);
     }
     efLib_804D64E8 = 1;
     switch (gfx_id) {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index b6ffe1908b..f5c84eff3a 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -176,8 +176,8 @@ void ftKb_SpecialHi1_Anim(Fighter_GObj* gobj)
 
     if (!ftAnim_IsFramesRemaining(gobj)) {
         ftCommon_8007D5D4(fp);
-        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirHi2, 0xA, 0.0f,
-                                  1.0f, 0.0f, NULL);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirHi2, 0xA, 0.0f, 1.0f,
+                                  0.0f, NULL);
         fp->pre_hitlag_cb = efLib_PauseAll;
         fp->post_hitlag_cb = efLib_ResumeAll;
         fp->mv.kb.specialhi.x0 = 0;
@@ -1222,8 +1222,8 @@ void ftKb_SpecialAirS_Enter(Fighter_GObj* gobj)
         fp->self_vel.y = da->specials_aerial_vertical_momentum;
         fp->fv.kb.x64 = true;
     }
-    Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirS, 0, 0.0f, 1.0f,
-                              0.0f, NULL);
+    Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirS, 0, 0.0f, 1.0f, 0.0f,
+                              NULL);
     ftAnim_8006EBA4(gobj);
     fp->accessory4_cb = &fn_800F53AC;
 }
diff --git a/src/melee/it/items/ithouou.c b/src/melee/it/items/ithouou.c
index d8c6c9e69c..11d7d52929 100644
--- a/src/melee/it/items/ithouou.c
+++ b/src/melee/it/items/ithouou.c
@@ -268,10 +268,8 @@ void it_802D2B4C(Item_GObj* gobj)
 
     if (ip->xDAC_itcmd_var0 != 0) {
         ip->xDAC_itcmd_var0 = 0;
-        it_8027ADEC(0x46F, gobj, ip->xBBC_dynamicBoneTable->bones[16],
-                    1.3F);
-        it_8027ADEC(0x46F, gobj, ip->xBBC_dynamicBoneTable->bones[17],
-                    1.3F);
+        it_8027ADEC(0x46F, gobj, ip->xBBC_dynamicBoneTable->bones[16], 1.3F);
+        it_8027ADEC(0x46F, gobj, ip->xBBC_dynamicBoneTable->bones[17], 1.3F);
     }
 
     if (ip->xDB0_itcmd_var1 != 0) {
diff --git a/src/melee/it/items/itoldkuri.c b/src/melee/it/items/itoldkuri.c
index 123999af9c..c8c1bfb04a 100644
--- a/src/melee/it/items/itoldkuri.c
+++ b/src/melee/it/items/itoldkuri.c
@@ -162,10 +162,8 @@ bool itOldkuri_UnkMotion2_Coll(Item_GObj* gobj)
     it_8026D62C(gobj, it_802D7AF0);
     if (it_80276308(gobj) == 8 && ip->xDD4_itemVar.oldkuri.xDFC != 0) {
         Item* ip = GET_ITEM(gobj);
-        itOldkuriAttributes* attr =
-            ip->xC4_article_data->x4_specialAttributes;
-        ip->xDD4_itemVar.oldkuri.xDF4 =
-            ip->facing_dir * ((f32*) attr->x0)[1];
+        itOldkuriAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+        ip->xDD4_itemVar.oldkuri.xDF4 = ip->facing_dir * ((f32*) attr->x0)[1];
         ip->x40_vel.x = ip->xDD4_itemVar.oldkuri.xDF4;
         ip->x40_vel.z = 0.0f;
         ip->x40_vel.y = 0.0f;
diff --git a/src/melee/it/items/itpikachuthunder.c b/src/melee/it/items/itpikachuthunder.c
index 2ce7a7f448..7d79466f24 100644
--- a/src/melee/it/items/itpikachuthunder.c
+++ b/src/melee/it/items/itpikachuthunder.c
@@ -1,5 +1,7 @@
 #include "itpikachuthunder.h"
 
+#include <placeholder.h>
+
 #include "ft/chara/ftPikachu/ftPk_SpecialLw.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
@@ -8,7 +10,6 @@
 #include "it/types.h"
 
 #include <baselib/gobj.h>
-#include <placeholder.h>
 
 int it_802B1DEC(Item_GObj* arg0)
 {
