diff --git a/src/melee/gm/gm_1601.c b/src/melee/gm/gm_1601.c
index 321e5ef80b..fb0c4874cb 100644
--- a/src/melee/gm/gm_1601.c
+++ b/src/melee/gm/gm_1601.c
@@ -14,6 +14,7 @@
 
 #include "gm/forward.h"
 
+#include "gm/types.h"
 #include "gr/ground.h"
 #include "gr/stage.h"
 #include "if/ifstatus.h"
@@ -1026,7 +1027,6 @@ long gm_80162A4C(s32 amount)
 struct gmm_x1868_1A8_t* gm_80162A98(s32 arg0)
 {
     struct gmm_x1868_1A8_t* temp_r3;
-    u32 var_r31_2;
     u32 var_r31;
     s32* temp_r3_2 = gmMainLib_8015CDB0();
 
@@ -1036,10 +1036,10 @@ struct gmm_x1868_1A8_t* gm_80162A98(s32 arg0)
 
     gmMainLib_8015EDBC()->xC = var_r31;
 
-    var_r31_2 = MAX((u32) (arg0 + gmMainLib_8015EDC8()->x0), -1U);
+    var_r31 = MAX((u32) (arg0 + gmMainLib_8015EDC8()->x0), -1U);
 
     temp_r3 = gmMainLib_8015EDC8();
-    temp_r3->x0 = var_r31_2;
+    temp_r3->x0 = var_r31;
     return temp_r3;
 }
 
@@ -1318,25 +1318,26 @@ u16 gm_80163274(u8 i)
 
 bool gm_80163298(s8 c_kind, u16 arg1)
 {
-    u16 v1;
-    s8 v2;
-    s16* v3;
-    u32 v4;
-    u32 v5;
+    u8 index;
+    s16* record;
+    u32 save_data;
+    u32 offset;
+    u16 score;
 
-    v2 = gm_80164024((u8) c_kind);
-    v3 = gmMainLib_8015D7EC((u8) v2);
-    v4 = (u32) gmMainLib_8015EDBC();
-    v5 = ((v2 << 2) & 0x3FC) + 0x114;
-    v1 = (u16) arg1;
+    index = gm_80164024((u8) c_kind);
+    record = gmMainLib_8015D7EC((u8) index);
+    save_data = (u32) gmMainLib_8015EDBC();
+    offset = ((index << 2) & 0x3FC) + 0x114;
+    score = (u16) arg1;
 
-    if (*(u32*) (v4 + v5) < (u32) v1) {
-        v4 = (u32) gmMainLib_8015EDBC();
-        *(u32*) (v4 + v5) = (u32) v1;
+    // POTENTIAL TODO: get rid of pointer math: seems hard with the addi + lwzx
+    if (*(u32*) (save_data + offset) < (u32) score) {
+        save_data = (u32) gmMainLib_8015EDBC();
+        *(u32*) (save_data + offset) = (u32) score;
     }
 
-    if (*(u16*) v3 < (u16) arg1) {
-        *v3 = arg1;
+    if (*(u16*) record < (u16) arg1) {
+        *record = arg1;
         return true;
     }
     return false;
@@ -2201,34 +2202,35 @@ void gm_80167320(int slot, bool arg1)
 
 void gm_80167470(s32 arg0, s32 arg1)
 {
-    if ((s32) gm_801A4310() == 0x1F) {
-        return;
+    switch ((s32) gm_801A4310()) {
+    case 0x1F:
+        gm_801B97C4(arg0, arg1);
+        break;
     }
-    gm_801B97C4(arg0, arg1);
 }
 
 void gm_801674C4(s8 arg0, u8 arg1, s8 arg2, s8 arg3, s32 arg4)
 {
     s8 temp_r31;
-    struct lbl_8046B488_t* temp_r3;
+    struct lbl_8046B488_t *new_var;
     struct lbl_8046B488_t* temp_ptr;
 
-    temp_r3 = fn_80169364();
-    temp_r3->x0 = arg0;
-    temp_r3->x1 = 0x21;
-    temp_r3->x2 = 0x21;
-    temp_ptr = temp_r3;
+    temp_ptr = fn_80169364();
+    temp_ptr->x0 = arg0;
+    temp_ptr->x1 = 0x21;
+    temp_ptr->x2 = 0x21;
     temp_ptr->x7 = arg1;
     temp_ptr->x8 = arg1;
     temp_ptr->x9 = arg2;
     temp_ptr->xA = arg3;
     PAD_STACK(10);
-    temp_ptr->unk_10_b2 = 1;
-    fn_80169574((s32) temp_ptr->x7, temp_ptr->x20);
-    fn_80169900(temp_ptr->xD, temp_ptr, temp_ptr->xA2, temp_ptr->x20);
+    new_var = temp_ptr;
+    new_var->unk_10_b2 = 1;
+    fn_80169574((s32) new_var->x7, new_var->x20);
+    fn_80169900(new_var->xD, new_var, temp_ptr->xA2, new_var->x20);
     temp_r31 = Player_GetCostumeId(0);
-    fn_8016989C((u8*) temp_ptr, (s32) Player_GetPlayerCharacter(0),
-                (s32) temp_r31, (u8*) temp_ptr->xA2, (u8*) temp_ptr->x20);
+    fn_8016989C((u8*) new_var, (s8) Player_GetPlayerCharacter(0),
+                (s32) temp_r31, (u8*) new_var->xA2, (u8*) temp_ptr->x20);
     fn_8016A09C();
     fn_80169434(arg4);
 }
@@ -3018,10 +3020,10 @@ void fn_80169550(int slot)
     lbl_8046B488.x20[idx] = -1;
 }
 
-void fn_80169574(ssize_t size, void* buf)
+void fn_80169574(ssize_t size, s8* buf)
 {
     memzero(buf, size);
-    ((s8*) buf)[size] = -2;
+    buf[size] = -2;
 }
 
 s32 fn_801695BC(u8 arg0, s32 arg1, s32 arg2, u8* arg3, u8* arg4)
@@ -3421,7 +3423,7 @@ void fn_8016A46C(void)
 
 void fn_8016A488(s32 arg0)
 {
-    if (gm_8016AE44()->hud_enabled) {
+    if (gm_8016AE44()->hud_enabled == true) {
         Player_80031848(arg0);
     }
 }
diff --git a/src/melee/gm/gm_1601.h b/src/melee/gm/gm_1601.h
index fe55e2bfea..09b96d1b81 100644
--- a/src/melee/gm/gm_1601.h
+++ b/src/melee/gm/gm_1601.h
@@ -223,7 +223,7 @@ gm_80167BC8(struct VsModeData*); // inits VsModeData from game rules
 /* 169530 */ void* gm_80169530(void);
 /* 169540 */ void* gm_80169540(void);
 /* 169550 */ void fn_80169550(int slot);
-/* 169574 */ void fn_80169574(ssize_t size, void* buf);
+/* 169574 */ void fn_80169574(ssize_t size, s8* buf);
 /* 1695BC */ long fn_801695BC(u8, s32, s32, u8*, u8*);
 /* 1697FC */ void fn_801697FC(signed char, signed char, signed char, signed char,
                             signed char*);
diff --git a/src/melee/it/it_266F.c b/src/melee/it/it_266F.c
index 66ce598894..eaf1fea1f9 100644
--- a/src/melee/it/it_266F.c
+++ b/src/melee/it/it_266F.c
@@ -69,11 +69,12 @@ static U8Vec4 it_804D5168 = { 0xFF, 0x40, 0x80, 0x80 };
 void it_8026C47C(struct it_8026C47C_arg0_t* arg_struct)
 {
     u32 it_kind;
+    u32 new_var;
     s32 var_r29; // Hold Kind??
     s32* var_r28;
     PAD_STACK(8);
 
-    it_kind = It_Kind_Capsule;
+    it_kind = (new_var = It_Kind_Capsule);
     var_r29 = 0;
     var_r28 = &arg_struct->unk0;
     arg_struct->unk0 = 0;
@@ -100,93 +101,73 @@ void it_8026C47C(struct it_8026C47C_arg0_t* arg_struct)
 s32 it_8026C530(s32 arg0, HSD_ObjAllocUnk6* arg1, s32 arg2, s32 arg3)
 {
     u16* temp_r10;
-    s32 temp_r5;
     s32 temp_r6;
     s32 temp_r7;
-    // s32 temp_r7_2;
-    s32 temp_r7_3;
-    // s32 temp_r8;
-    s32 temp_r9;
-    s32 temp_r9_2;
+    s32 mid;
+    u16* temp_r9;
     s32 var_r5;
 
     var_r5 = arg2;
-    // temp_r8 = arg3 - 1;
     if (var_r5 == (arg3 - 1)) {
         return var_r5;
     }
     temp_r10 = arg1->xC;
-    temp_r9 = (var_r5 + arg3) / 2;
-    // temp_r7_2 = temp_r9 * 2;
-    // temp_r7_2 = temp_r9;
-    // if ((s32) *(temp_r10 + temp_r7_2) > arg0) {
-    // if (((s32) *temp_r10[temp_r7_2])->x0 > arg0) {
-    if (temp_r10[temp_r9] > arg0) {
-        if (var_r5 == (temp_r9 - 1)) {
-        } else {
-            temp_r6 = (var_r5 + temp_r9) / 2;
-            temp_r7_3 = temp_r6 * 1;
-            // if ((s32) *(temp_r10 + temp_r7_3) > arg0) {
-            // if (temp_r10[temp_r7_3]->x0 > arg0) {
-            if (temp_r10[temp_r7_3] > arg0) {
+    mid = (var_r5 + arg3) / 2;
+    if (temp_r10[mid] > arg0) {
+        if (var_r5 != (mid - 1)) {
+            temp_r6 = (var_r5 + mid) / 2;
+            if (temp_r10[temp_r6] > arg0) {
                 var_r5 = it_8026C530(arg0, arg1, var_r5, temp_r6);
-                // } else if ((s32) (temp_r10 + temp_r7_3)->unk2 > arg0) {
-                // } else if (temp_r10[temp_r7_3]->x2 > arg0) {
-            } else if (temp_r10[temp_r7_3] > arg0) {
+            } else if (temp_r10[temp_r6 + 1] > arg0) {
                 var_r5 = temp_r6;
             } else {
-                var_r5 = it_8026C530(arg0, arg1, temp_r6, temp_r9);
+                var_r5 = it_8026C530(arg0, arg1, temp_r6, mid);
             }
         }
         return var_r5;
     }
-    temp_r9_2 = (s32) temp_r10 + 2;
-    // temp_r9_2 = temp_r10->x2;
-    // if ((s32) *(temp_r9_2 + temp_r7_2) > arg0) {
-    if ((temp_r9_2 + temp_r9) > arg0) {
-        return temp_r9;
+    temp_r9 = temp_r10 + 1;
+    if (temp_r9[mid] > arg0) {
+        return mid;
     }
-    if (temp_r9 == arg3 - 1) {
-        return temp_r9;
+    if (mid == (arg3 - 1)) {
+        return mid;
     }
-    temp_r7 = (s32) (temp_r9 + arg3) / 2;
-    // temp_r5 = temp_r7 * 2;
-    temp_r5 = temp_r7;
-    // if ((s32) *(temp_r10 + temp_r5) > arg0) {
-    if (temp_r10[temp_r5] > arg0) {
-        return it_8026C530(arg0, arg1, temp_r9, temp_r7);
+    temp_r7 = (mid + arg3) / 2;
+    if (temp_r10[temp_r7] > arg0) {
+        return it_8026C530(arg0, arg1, mid, temp_r7);
     }
-    // if ((s32) *(temp_r9_2 + temp_r5) > arg0) {
-    if ((temp_r9_2 + temp_r5) > arg0) {
+    if (temp_r9[temp_r7] > arg0) {
         return temp_r7;
     }
     return it_8026C530(arg0, arg1, temp_r7, arg3);
 }
 
 /// argument is definitely size 0x10
-u8 it_8026C65C(HSD_ObjAllocUnk6* arg_struct)
+int it_8026C65C(HSD_ObjAllocUnk6* arg_struct)
 {
     s32 rand_int;
     s32 temp_r6;
     s32 var_r3;
     u16* temp_r5;
+    s32 count;
+    temp_r6 = (s32) arg_struct->x8;
+
 
-    rand_int = HSD_Randi((s32) arg_struct->x8);
-    if ((arg_struct->x0 - 1) == 0) {
+    count = arg_struct->x0;
+    rand_int = HSD_Randi(temp_r6);
+    if ((count - 1) == 0) {
         var_r3 = 0;
     } else {
         temp_r5 = arg_struct->xC;
-        temp_r6 = arg_struct->x0 / 2;
+        temp_r6 = count / 2;
         if (temp_r5[temp_r6] > rand_int) {
             var_r3 = it_8026C530(rand_int, arg_struct, 0, temp_r6);
-            // } else if (temp_r5[temp_r6].unk2 > rand_int) {
-            // } else if (temp_r5[temp_r6] > rand_int) {
-            // } else if ((&temp_r5[temp_r6] + 2) > rand_int) {
-        } else if ((arg_struct->x0 << 2) > rand_int) {
+        } else if (temp_r5[temp_r6 + 1] > rand_int) {
             var_r3 = temp_r6;
         } else {
             var_r3 =
-                it_8026C530(rand_int, arg_struct, temp_r6, arg_struct->x0);
+                it_8026C530(rand_int, arg_struct, temp_r6, count);
         }
     }
     return arg_struct->x4[var_r3];
@@ -1097,12 +1078,12 @@ void it_8026E0F4(Item_GObj* gobj)
 
 inline bool it_8026E_inline(Item_GObj* gobj)
 {
+    Item* ip = GET_ITEM((HSD_GObj*) gobj);
     bool cond;
     CollData* coll;
-    Item* ip;
-    PAD_STACK(26);
+    bool cond2;
+    PAD_STACK(24);
 
-    ip = GET_ITEM((HSD_GObj*) gobj);
     coll = &ip->x378_itemColl;
     it_80276214(gobj);
     cond = mpColl_800471F8(coll);
@@ -1110,11 +1091,22 @@ inline bool it_8026E_inline(Item_GObj* gobj)
     if (cond) {
         ip->xC30 = coll->floor.index;
     }
-    cond |= it_80276308(gobj);
-    cond |= it_802763E0(gobj);
+    cond2 = cond;
+    cond2 |= it_80276308(gobj);
+    cond2 |= it_802763E0(gobj);
+    cond = cond2;
     return cond & 0xF;
 }
 
+bool it_8026E15C_inline1(Item_GObj* gobj)
+{
+    return it_8026DDFC(gobj);
+}
+bool it_8026E15C_inline2(Item_GObj* gobj)
+{
+    return it_8026DD5C(gobj);
+}
+
 void it_8026E15C(Item_GObj* gobj, HSD_GObjEvent arg1)
 {
     bool res2;
@@ -1124,9 +1116,9 @@ void it_8026E15C(Item_GObj* gobj, HSD_GObjEvent arg1)
     if (res) {
         it_80276FC4(gobj, res);
         if (res & 1) {
-            if (it_8026DDFC(gobj) && it_8026DC24(gobj)) {
+            if (it_8026E15C_inline1(gobj) && it_8026DC24(gobj)) {
                 it_802725D4(gobj);
-                res2 = it_8026DD5C(gobj);
+                res2 = it_8026E15C_inline2(gobj);
             } else {
                 res2 = false;
             }
@@ -1137,6 +1129,11 @@ void it_8026E15C(Item_GObj* gobj, HSD_GObjEvent arg1)
     }
 }
 
+bool it_8026E248_inline(Item_GObj* gobj)
+{
+    return it_8026DD5C(gobj);
+}
+
 void it_8026E248(Item_GObj* gobj, HSD_GObjEvent arg1)
 {
     bool res2;
@@ -1147,7 +1144,7 @@ void it_8026E248(Item_GObj* gobj, HSD_GObjEvent arg1)
         it_80276FC4(gobj, res);
         if (res & 1) {
             if (it_8026DE98(gobj) && it_8026DC24(gobj)) {
-                res2 = it_8026DD5C(gobj);
+                res2 = it_8026E248_inline(gobj);
             } else {
                 res2 = false;
             }
@@ -1324,13 +1321,19 @@ void it_8026E8C4(Item_GObj* item_gobj, HSD_GObjEvent arg1, HSD_GObjEvent arg2) {
     }
 }
 #else
+
+bool it_8026E32C_inline(Item_GObj* gobj)
+{
+    return it_8026DD5C(gobj);
+}
 bool it_8026E32C(Item_GObj* item_gobj, HSD_GObjEvent arg1)
 {
     CollData* coll;
     Item* item;
     bool chk;
     bool chk2;
-    PAD_STACK(38);
+    bool new_var;
+    PAD_STACK(42);
 
     item = item_gobj->user_data;
     coll = &item->x378_itemColl;
@@ -1342,18 +1345,19 @@ bool it_8026E32C(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     }
     chk = chk | it_80276308(item_gobj);
     chk = chk | it_802763E0(item_gobj);
-    if (chk & 0xF) {
-        it_80276FC4(item_gobj, chk);
-        if (chk & 1) {
+    new_var = chk;
+    if (new_var & 0xF) {
+        it_80276FC4(item_gobj, new_var);
+        if (new_var & 1) {
             item = item_gobj->user_data;
             item->xD50_landNum += 1;
             if (it_8026DC24(item_gobj)) {
-                chk2 = it_8026DD5C(item_gobj);
+                chk2 = it_8026E32C_inline(item_gobj);
             } else {
                 chk2 = false;
             }
             if (chk2) {
-                arg1((HSD_GObj*) item_gobj);
+                arg1(item_gobj);
             }
         }
     }
@@ -1365,6 +1369,7 @@ void it_8026E414(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     CollData* coll;
     Item* item;
     bool chk;
+    bool new_var;
     PAD_STACK(28);
 
     item = item_gobj->user_data;
@@ -1377,10 +1382,11 @@ void it_8026E414(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     }
     chk = chk | it_80276308(item_gobj);
     chk = chk | it_802763E0(item_gobj);
-    if (chk & 0xE) {
-        it_80276FC4(item_gobj, chk);
+    new_var = chk;
+    if (new_var & 0xE) {
+        it_80276FC4(item_gobj, new_var);
     }
-    if (chk & 1) {
+    if (new_var & 1) {
         it_80275DFC(item_gobj);
         it_802762B0(item);
         arg1((HSD_GObj*) item_gobj);
@@ -1393,6 +1399,7 @@ void it_8026E4D0(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     HSD_JObj* item_jobj;
     Item* item;
     bool chk;
+    bool new_var;
     PAD_STACK(34);
 
     item = item_gobj->user_data;
@@ -1405,14 +1412,15 @@ void it_8026E4D0(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     }
     chk = chk | it_80276308(item_gobj);
     chk = chk | it_802763E0(item_gobj);
-    if (chk & 0xE) {
-        it_80276FC4(item_gobj, chk);
+    new_var = chk;
+    if (new_var & 0xE) {
+        it_80276FC4(item_gobj, new_var);
     }
-    if (chk & 1) {
+    if (new_var & 1) {
         item_jobj = item_gobj->hsd_obj;
         it_80275DFC(item_gobj);
         it_802762B0(item);
-        arg1((HSD_GObj*) item_gobj);
+        arg1(item_gobj);
         it_80272F7C(item_jobj, item->xCC_item_attr->x60_scale);
     }
 }
@@ -1422,6 +1430,7 @@ void it_8026E5A0(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     CollData* coll;
     Item* item;
     bool chk;
+    bool new_val;
     PAD_STACK(28);
 
     item = item_gobj->user_data;
@@ -1434,15 +1443,16 @@ void it_8026E5A0(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     }
     chk = chk | it_80276308(item_gobj);
     chk = chk | it_802763E0(item_gobj);
-    if (chk & 0xF) {
-        if (chk & 0xE) {
-            it_80276FC4(item_gobj, chk);
+    new_val = chk;
+    if (new_val & 0xF) {
+        if (new_val & 0xE) {
+            it_80276FC4(item_gobj, new_val);
         }
-        if (chk & 1) {
+        if (new_val & 1) {
             it_80275DFC(item_gobj);
             it_802762B0(item);
         }
-        arg1((HSD_GObj*) item_gobj);
+        arg1(item_gobj);
     }
 }
 
@@ -1451,6 +1461,8 @@ void it_8026E664(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     CollData* coll;
     Item* item;
     bool chk;
+    bool new_var;
+
     PAD_STACK(28);
 
     item = item_gobj->user_data;
@@ -1462,11 +1474,12 @@ void it_8026E664(Item_GObj* item_gobj, HSD_GObjEvent arg1)
         item->xC30 = coll->floor.index;
     }
     chk = chk | it_80276308(item_gobj);
-    if (chk & 0xD) {
-        if (chk & 0xC) {
-            it_80276FC4(item_gobj, chk);
+    new_var = chk;
+    if (new_var & 0xD) {
+        if (new_var & 0xC) {
+            it_80276FC4(item_gobj, new_var);
         }
-        if (chk & 1) {
+        if (new_var & 1) {
             it_80275DFC(item_gobj);
             it_802762B0(item);
         }
@@ -1479,6 +1492,7 @@ void it_8026E71C(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     CollData* coll;
     Item* item;
     bool chk;
+    bool new_var;
     PAD_STACK(28);
 
     item = item_gobj->user_data;
@@ -1491,24 +1505,28 @@ void it_8026E71C(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     }
     chk = chk | it_80276308(item_gobj);
     chk = chk | it_802763E0(item_gobj);
-    if (chk & 0xF) {
-        if (chk & 0xC) {
-            it_80276D9C(item_gobj, chk);
+    new_var = chk;
+    if (new_var & 0xF) {
+        if (new_var & 0xC) {
+            it_80276D9C(item_gobj, new_var);
         }
-        if (chk & 1) {
+        if (new_var & 1) {
             it_80275DFC(item_gobj);
             it_802762B0(item);
         }
-        arg1((HSD_GObj*) item_gobj);
+        arg1(item_gobj);
     }
 }
 
+#pragma push
+#pragma dont_inline on
 void it_8026E7E0(Item_GObj* item_gobj, HSD_GObjEvent arg1)
 {
     CollData* coll;
     Item* item;
     bool chk;
     bool chk2;
+    bool new_var;
     PAD_STACK(30);
 
     item = item_gobj->user_data;
@@ -1521,21 +1539,22 @@ void it_8026E7E0(Item_GObj* item_gobj, HSD_GObjEvent arg1)
     }
     chk = chk | it_80276308(item_gobj);
     chk = chk | it_802763E0(item_gobj);
-    // chk &= 0xC;
-    if (chk & 0xC) {
-        it_80276D9C(item_gobj, chk);
+    new_var = chk;
+    if (new_var & 0xC) {
+        it_80276D9C(item_gobj, new_var);
     }
-    if (chk & 1) {
+    if (new_var & 1) {
         if (it_8026DE98(item_gobj) && it_8026DC24(item_gobj)) {
-            chk2 = it_8026DD5C(item_gobj);
+            new_var = it_8026DD5C(item_gobj);
         } else {
-            chk2 = false;
+            new_var = false;
         }
-        if (chk2) {
-            arg1((HSD_GObj*) item_gobj);
+        if (new_var) {
+            arg1(item_gobj);
         }
     }
 }
+#pragma pop
 
 void it_8026E8C4(Item_GObj* item_gobj, HSD_GObjEvent arg1, HSD_GObjEvent arg2)
 {
@@ -1583,14 +1602,14 @@ bool it_8026E9A4(HSD_GObj* gobj, Vec3* arg1, Vec3* arg2, Vec3* arg3)
     return false;
 }
 
-bool it_8026EA20(HSD_GObj* gobj, Vec3* arg1, Vec3* arg2, Vec3* arg3,
-                 UNK_T* arg4, u32* flags_out)
+bool it_8026EA20(HSD_GObj* gobj, Vec3* arg1, Vec3* arg2, Vec3* arg3, int* arg4,
+                 u32* flags_out)
 {
+    UNUSED int _;
     Vec3 p;
-    // PAD_STACK(4);
-    PAD_STACK(8);
+    PAD_STACK(4);
 
-    if (mpCheckAllRemap(&p, *arg4, flags_out, arg3, -1, -1, arg1->x, arg1->y,
+    if (mpCheckAllRemap(&p, arg4, flags_out, arg3, -1, -1, arg1->x, arg1->y,
                         arg2->x, arg2->y) == true)
     {
         *arg2 = p;
diff --git a/src/melee/it/it_266F.h b/src/melee/it/it_266F.h
index 943b556733..89dfaae277 100644
--- a/src/melee/it/it_266F.h
+++ b/src/melee/it/it_266F.h
@@ -12,7 +12,7 @@
 
 /* 26C47C */ void it_8026C47C(struct it_8026C47C_arg0_t*);
 /* 26C530 */ s32 it_8026C530(s32, HSD_ObjAllocUnk6*, s32, s32);
-/* 26C65C */ u8 it_8026C65C(HSD_ObjAllocUnk6*);
+/* 26C65C */ int it_8026C65C(HSD_ObjAllocUnk6*);
 /* 26C704 */ bool it_8026C704(void);
 /* 26C75C */ ItemKind it_8026C75C(HSD_ObjAllocUnk6*);
 /* 26C88C */ void fn_8026C88C(HSD_GObj*);
@@ -60,7 +60,7 @@
 /* 26E7E0 */ void it_8026E7E0(Item_GObj*, HSD_GObjEvent);
 /* 26E8C4 */ void it_8026E8C4(Item_GObj*, HSD_GObjEvent, HSD_GObjEvent);
 /* 26E9A4 */ bool it_8026E9A4(HSD_GObj*, Vec3*, Vec3*, Vec3*);
-/* 26EA20 */ bool it_8026EA20(HSD_GObj*, Vec3*, Vec3*, Vec3*, UNK_T*,
+/* 26EA20 */ bool it_8026EA20(HSD_GObj*, Vec3*, Vec3*, Vec3*, int*,
                               u32* flags_out);
 /* 26EA9C */ bool it_8026EA9C(HSD_GObj*, Vec3*, Vec3*, Vec3*, int*,
                               u32* flags_out, s32);
@@ -79,4 +79,10 @@
 /* 26F8B4 */ bool it_8026F8B4(Item_GObj*, Vec3*, Vec3*, bool);
 /* 26F9A0 */ void it_8026F9A0(void);
 
+
+
+bool it_8026E15C_inline1(Item_GObj* gobj);
+bool it_8026E15C_inline2(Item_GObj* gobj);
+bool it_8026E248_inline(Item_GObj* gobj);
+bool it_8026E32C_inline(Item_GObj* gobj);
 #endif
diff --git a/src/melee/it/itcoll.c b/src/melee/it/itcoll.c
index 121290ac1c..cbc7428ae5 100644
--- a/src/melee/it/itcoll.c
+++ b/src/melee/it/itcoll.c
@@ -162,6 +162,8 @@ void it_8026FC00(Item* arg_item, HitCapsule* arg_hit, s32 arg2, Fighter* arg3)
 
     if (arg_item->xAC4_ignoreItemID != 0) {
         HSD_GObj* item_gobj = HSD_GObj_Entities->items;
+        if (item_gobj->next) {
+        }
         while (item_gobj != NULL) {
             Item* item = GET_ITEM(item_gobj);
             if (item->xAC4_ignoreItemID == arg_item->xAC4_ignoreItemID) {
@@ -179,105 +181,30 @@ void it_8026FCF8(Item* arg_item, HitCapsule* arg_hit)
     HSD_GObj* item_gobj;
     HitCapsule* hit;
     bool chk;
-    u32 index;
     Item* item;
-    PAD_STACK(4);
+    PAD_STACK(8);
 
-#if 0
-    if (arg_item->xAC4_ignoreItemID != 0U) {
-        item_gobj = HSD_GObj_Entities->items;
-        while (item_gobj != NULL) {
-            item = GET_ITEM(item_gobj);
-            if (item->xAC4_ignoreItemID == arg_item->xAC4_ignoreItemID) {
-                index = 0U;
-                chk = false;
-                while (!chk && index < 4U) {
-                    hit = &item->x5D4_hitboxes[index].hit;
-                    if ((hit != arg_hit) && (hit->state != HitCapsule_Disabled) && (hit->x4 == arg_hit->x4)) {
-                        lbColl_CopyHitCapsule(hit, arg_hit);
-                        chk = true;
-                        return;
-                    }
-                    index++;
-                }
-                if (chk) {
-                    return;
-                }
-            }
-            item_gobj = item_gobj->next;
-        }
-    }
-    lbColl_80008440(arg_hit);
-    return;
-#elif 0
     if (arg_item->xAC4_ignoreItemID != 0U) {
         item_gobj = HSD_GObj_Entities->items;
         while (item_gobj != NULL) {
             item = GET_ITEM(item_gobj);
             if (item->xAC4_ignoreItemID == arg_item->xAC4_ignoreItemID) {
+                u32 i;
+
                 chk = false;
-                for (index = 0U; index < 4U; index++) {
-                    hit = &item->x5D4_hitboxes[index].hit;
-                    if ((hit != arg_hit) &&
-                        (hit->state != HitCapsule_Disabled) &&
-                        (hit->x4 == arg_hit->x4))
-                    {
-                        lbColl_CopyHitCapsule(hit, arg_hit);
-                        return;
-                    }
-                }
-            }
-            item_gobj = item_gobj->next;
-        }
-    }
-    lbColl_80008440(arg_hit);
-    return;
-#else
-    if (arg_item->xAC4_ignoreItemID != 0U) {
-        item_gobj = HSD_GObj_Entities->items;
-        while (item_gobj != NULL) {
-            item = GET_ITEM(item_gobj);
-            if (item->xAC4_ignoreItemID == arg_item->xAC4_ignoreItemID) {
-                index = 0U;
-                hit = &item->x5D4_hitboxes[index].hit;
-                if ((hit != arg_hit) && (hit->state != HitCapsule_Disabled) &&
-                    (hit->x4 == arg_hit->x4))
-                {
-                    lbColl_CopyHitCapsule(hit, arg_hit);
-                    chk = true;
-                } else {
-                    index++;
-                    hit = &item->x5D4_hitboxes[index].hit;
+                for (i = 0; i < ARRAY_SIZE(item->x5D4_hitboxes); i++) {
+                    hit = &item->x5D4_hitboxes[i].hit;
                     if ((hit != arg_hit) &&
                         (hit->state != HitCapsule_Disabled) &&
                         (hit->x4 == arg_hit->x4))
                     {
                         lbColl_CopyHitCapsule(hit, arg_hit);
                         chk = true;
-                    } else {
-                        index++;
-                        hit = &item->x5D4_hitboxes[index].hit;
-                        if ((hit != arg_hit) &&
-                            (hit->state != HitCapsule_Disabled) &&
-                            (hit->x4 == arg_hit->x4))
-                        {
-                            lbColl_CopyHitCapsule(hit, arg_hit);
-                            chk = true;
-                        } else {
-                            index++;
-                            hit = &item->x5D4_hitboxes[index].hit;
-                            if ((hit != arg_hit) &&
-                                (hit->state != HitCapsule_Disabled) &&
-                                (hit->x4 == arg_hit->x4))
-                            {
-                                lbColl_CopyHitCapsule(hit, arg_hit);
-                                chk = true;
-                            } else {
-                                chk = false;
-                            }
-                        }
+                        break;
                     }
+                    chk = false;
                 }
+
                 if (chk) {
                     return;
                 }
@@ -286,8 +213,6 @@ void it_8026FCF8(Item* arg_item, HitCapsule* arg_hit)
         }
     }
     lbColl_80008440(arg_hit);
-    return;
-#endif
 }
 
 void it_8026FE68(Item* arg_item0, HitCapsule* hit1, Item* arg_item2,
@@ -1428,11 +1353,12 @@ void it_80271A58(Item_GObj* item_gobj)
     Item* item;
     HSD_JObj* jobj;
 
-    item = item_gobj->user_data;
+    item = GET_ITEM(item_gobj);
     item->xBEC = item->xBDC;
     if ((item->facing_dir == 1.0f) && (item->xDC8_word.flags.x19 == 1)) {
-        item->xBEC.left = -item->xBEC.left;
-        item->xBEC.right = -item->xBEC.right;
+        f32 temp = -item->xBEC.right;
+        item->xBEC.right = -item->xBEC.left;
+        item->xBEC.left = temp;
     }
     jobj = it_802746F8(item_gobj);
     if (item->xDC8_word.flags.x17 == 0) {
diff --git a/src/melee/it/items/itfoxblaster.c b/src/melee/it/items/itfoxblaster.c
index 63e78509f3..8d93d1b87c 100644
--- a/src/melee/it/items/itfoxblaster.c
+++ b/src/melee/it/items/itfoxblaster.c
@@ -150,6 +150,9 @@ void it_802ADF10(HSD_GObj* item_gobj)
 {
     Vec3 ft_cur_pos;
     Vec3 ft_hold_joint_pos;
+    // TODO: probably a fake-improvement? idiomatic way is
+    // with the for loop, but can't get it to 100% match yet
+    int new_var2;
     Vec3 it_hold_joint_pos;
     Item* item = GET_ITEM(item_gobj);
 
@@ -188,6 +191,7 @@ void it_802ADF10(HSD_GObj* item_gobj)
         case It_Kind_Kirby_FalcoBlaster:
             ftKb_SpecialNFx_800FDC00(item->xDD4_itemVar.foxblaster.owner,
                                      &ft_hold_joint_pos);
+            new_var2 = 0;
             item->xDD4_itemVar.foxblaster.xE14.x =
                 ft_hold_joint_pos.x - ft_cur_pos.x;
             item->xDD4_itemVar.foxblaster.xE14.y =
@@ -200,7 +204,7 @@ void it_802ADF10(HSD_GObj* item_gobj)
             item->xDD4_itemVar.foxblaster.angle =
                 atan2f(ft_hold_joint_pos.y - it_hold_joint_pos.y,
                        ft_hold_joint_pos.x - it_hold_joint_pos.x);
-            item->xDD4_itemVar.foxblaster.xDFC = 0;
+            item->xDD4_itemVar.foxblaster.xDFC = new_var2;
 
             item->xDD4_itemVar.foxblaster.xDE4 =
                 (s32) efSync_Spawn(1196, item_gobj, &ft_hold_joint_pos,
@@ -273,7 +277,7 @@ void it_802ADF10(HSD_GObj* item_gobj)
 
         item = GET_ITEM(item_gobj);
         item->xDD4_itemVar.foxblaster.xDE4 = 0;
-        item->xDD4_itemVar.foxblaster.xDFC = 0;
+        item->xDD4_itemVar.foxblaster.xDFC = new_var2;
         item->xDD4_itemVar.foxblaster.angle =
             item->xDD4_itemVar.foxblaster.xE14.x =
                 item->xDD4_itemVar.foxblaster.xE14.y =
