diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index 5f4988ffbc..af3f8a6240 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -1621,7 +1621,7 @@ static void fn_800F6318(HSD_GObj* gobj)
 void ftKb_SpecialN_800F6388(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
-    Fighter_ChangeMotionState(gobj, ftKb_MS_EatWait, 0x10, 0.0F, 1.0F, 0.0F,
+    Fighter_ChangeMotionState(gobj, ftKb_MS_Eat, 0x10, 0.0F, 1.0F, 0.0F,
                               NULL);
     ftAnim_8006EBA4(gobj);
     ftCommon_8007E2F4(fp, 0x1FF);
@@ -1630,7 +1630,7 @@ void ftKb_SpecialN_800F6388(Fighter_GObj* gobj)
 void ftKb_SpecialN_800F63EC(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
-    Fighter_ChangeMotionState(gobj, ftKb_MS_EatFall, 0x10, 0.0F, 1.0F, 0.0F,
+    Fighter_ChangeMotionState(gobj, ftKb_MS_EatAir, 0x10, 0.0F, 1.0F, 0.0F,
                               NULL);
     ftAnim_8006EBA4(gobj);
     ftCommon_8007E2F4(fp, 0x1FF);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
index 3dcc8ab65f..20d43072a4 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
@@ -345,6 +345,12 @@ void ftKb_KpSpecialNStart_Anim(Fighter_GObj* gobj)
     }
 }
 
+inline ftKb_DatAttrs* ftKb_KpSpecialN_Anim_inline(Fighter *arg0)
+{
+  return arg0->dat_attrs;
+}
+
+
 void ftKb_KpSpecialN_Anim(Fighter_GObj* gobj)
 {
     PAD_STACK(8);
@@ -359,7 +365,7 @@ void ftKb_KpSpecialN_Anim(Fighter_GObj* gobj)
     }
     {
         Fighter* fp = GET_FIGHTER(gobj);
-        ftKb_DatAttrs* da = fp->dat_attrs;
+        ftKb_DatAttrs* da = ftKb_KpSpecialN_Anim_inline(fp);
         if (fp->mv.kb.specialn_kp.x18 == 0) {
             Camera_80030E44(2, &fp->cur_pos);
         }
@@ -409,7 +415,7 @@ void ftKb_KpSpecialAirN_Anim(Fighter_GObj* gobj)
     }
     {
         Fighter* fp = GET_FIGHTER(gobj);
-        ftKb_DatAttrs* da = fp->dat_attrs;
+        ftKb_DatAttrs* da = ftKb_KpSpecialN_Anim_inline(fp);
         if (fp->mv.kb.specialn_kp.x18 == 0) {
             Camera_80030E44(2, &fp->cur_pos);
         }
@@ -974,6 +980,7 @@ void ftKb_SpecialNSs_800FCD04(Fighter_GObj* gobj)
 
 void ftKb_SpecialNSs_800FCD60(Fighter_GObj* gobj)
 {
+    PAD_STACK(8);
     if (gobj != NULL) {
         Fighter* fp = GET_FIGHTER(gobj);
         /// FAKE MATCH: redundant NULL checks required for regalloc
@@ -983,7 +990,7 @@ void ftKb_SpecialNSs_800FCD60(Fighter_GObj* gobj)
                 fp->fv.kb.xA4 = 0;
             }
             if (gobj != NULL) {
-                Fighter* fp2 = GET_FIGHTER(gobj);
+                Fighter *fp2 = HSD_GObjGetUserData(gobj);
                 efLib_DestroyAll(gobj);
                 fp2->fv.kb.xAC = 0;
             }
@@ -997,6 +1004,8 @@ void ftKb_SpecialNSs_800FCD60(Fighter_GObj* gobj)
 void ftKb_SpecialNSs_800FCF74(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
+    Fighter* fp2;
+    PAD_STACK(8);
     Fighter_ChangeMotionState(gobj, ftKb_MS_SsSpecialNStart, 0, 0.0F, 1.0F,
                               0.0F, NULL);
     fp->cmd_vars[3] = 0;
@@ -1005,9 +1014,9 @@ void ftKb_SpecialNSs_800FCF74(Fighter_GObj* gobj)
     fp->cmd_vars[0] = 0;
     ftCommon_8007D7FC(fp);
     fp->self_vel.y = 0.0F;
-    fp = GET_FIGHTER(gobj);
-    fp->death2_cb = ftKb_Init_800EE74C;
-    fp->take_dmg_cb = ftKb_Init_800EE7B8;
+    fp2 = GET_FIGHTER(gobj);
+    fp2->death2_cb = ftKb_Init_800EE74C;
+    fp2->take_dmg_cb = ftKb_Init_800EE7B8;
     fp->mv.kb.specialhi.x0 = 0;
     fp->mv.kb.specialhi.x4 = 0;
     fp->mv.kb.specialhi.x8.f = 0.0F;
@@ -1017,15 +1026,17 @@ void ftKb_SpecialNSs_800FCF74(Fighter_GObj* gobj)
 void ftKb_SpecialNSs_800FD020(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
+    Fighter* fp2;
+    PAD_STACK(8);
     Fighter_ChangeMotionState(gobj, ftKb_MS_SsSpecialAirNStart, 0, 0.0F, 1.0F,
                               0.0F, NULL);
     fp->cmd_vars[3] = 0;
     fp->cmd_vars[2] = 0;
     fp->cmd_vars[1] = 0;
     fp->cmd_vars[0] = 0;
-    fp = GET_FIGHTER(gobj);
-    fp->death2_cb = ftKb_Init_800EE74C;
-    fp->take_dmg_cb = ftKb_Init_800EE7B8;
+    fp2 = GET_FIGHTER(gobj);
+    fp2->death2_cb = ftKb_Init_800EE74C;
+    fp2->take_dmg_cb = ftKb_Init_800EE7B8;
     fp->mv.kb.specialhi.x0 = 1;
     fp->mv.kb.specialhi.x4 = 0;
     fp->mv.kb.specialhi.x8.f = 0.0F;
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index a22e59f00f..ab1f7f36de 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -227,10 +227,12 @@ void ftKb_SpecialNSk_80105E8C(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da;
+    void *new_var; // Permuter slop
+    PAD_STACK(8);
     ftCommon_8007D5D4(fp);
     Fighter_ChangeMotionState(gobj, ftKb_MS_ZdSpecialAirN, 0x0C4C508E,
                               fp->cur_anim_frame, 1.0F, 0.0F, NULL);
-    fp = GET_FIGHTER(gobj);
+    fp = (Fighter *) (new_var = HSD_GObjGetUserData(gobj));
     da = fp->dat_attrs;
     if (fp->x2219_b0 == true) {
         fp->pre_hitlag_cb = efLib_PauseAll;
@@ -246,10 +248,12 @@ void ftKb_SpecialNSk_80105F3C(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da;
+    void *new_var; // Permuter slop
+    PAD_STACK(8);
     ftCommon_8007D7FC(fp);
     Fighter_ChangeMotionState(gobj, ftKb_MS_ZdSpecialN, 0x0C4C508E,
                               fp->cur_anim_frame, 1.0F, 0.0F, NULL);
-    fp = GET_FIGHTER(gobj);
+    fp = (Fighter *) (new_var = HSD_GObjGetUserData(gobj));
     da = fp->dat_attrs;
     if (fp->x2219_b0 == true) {
         fp->pre_hitlag_cb = efLib_PauseAll;
diff --git a/src/melee/gm/gm_1601.c b/src/melee/gm/gm_1601.c
index 5dbbc15436..2b33c62057 100644
--- a/src/melee/gm/gm_1601.c
+++ b/src/melee/gm/gm_1601.c
@@ -1323,24 +1323,18 @@ u16 gm_80163274(u8 i)
 bool gm_80163298(s8 c_kind, u16 arg1)
 {
     u8 index;
-    s16* record;
-    u32 save_data;
-    u32 offset;
+    u16* record;
     u16 score;
 
-    index = gm_80164024((u8) c_kind);
-    record = gmMainLib_8015D7EC((u8) index);
-    save_data = (u32) gmMainLib_8015EDBC();
-    offset = ((index << 2) & 0x3FC) + 0x114;
-    score = (u16) arg1;
+    index = gm_80164024(c_kind);
+    record = gmMainLib_8015D7EC(index);
+    score = arg1;
 
-    // POTENTIAL TODO: get rid of pointer math: seems hard with the addi + lwzx
-    if (*(u32*) (save_data + offset) < (u32) score) {
-        save_data = (u32) gmMainLib_8015EDBC();
-        *(u32*) (save_data + offset) = (u32) score;
+    if ((u32) gmMainLib_8015EDBC()->x114[index] < (u32) score) {
+        gmMainLib_8015EDBC()->x114[index] = (int) score;
     }
 
-    if (*(u16*) record < (u16) arg1) {
+    if (*record < arg1) {
         *record = arg1;
         return true;
     }
diff --git a/src/melee/gm/gm_18A5.c b/src/melee/gm/gm_18A5.c
index 21e63e464b..ff5cc36a91 100644
--- a/src/melee/gm/gm_18A5.c
+++ b/src/melee/gm/gm_18A5.c
@@ -7182,7 +7182,7 @@ void fn_8019AF50(s32* arg0, u32 arg1, u32 arg2)
                     dest += 2;
                 }
             }
-            HSD_SisLib_803A70A0((void*) tm->x524[3], 0, &sp_buf);
+            HSD_SisLib_803A70A0(tm->x524[3], 0, (char*) &sp_buf);
         } else {
             *counter += 2;
             if (lbl_804799D8.x4D != 1) {
@@ -7190,7 +7190,7 @@ void fn_8019AF50(s32* arg0, u32 arg1, u32 arg2)
                     *counter = 0xFA;
                 }
             }
-            HSD_SisLib_803A70A0((void*) tm->x524[3], 0, (u8*) counter + 0x4E);
+            HSD_SisLib_803A70A0(tm->x524[3], 0, (char*) counter + 0x4E);
         }
     } else {
         if (*counter < 0xFAU) {
diff --git a/src/melee/gm/gmmain_lib.c b/src/melee/gm/gmmain_lib.c
index d1b9f24f51..9b49401255 100644
--- a/src/melee/gm/gmmain_lib.c
+++ b/src/melee/gm/gmmain_lib.c
@@ -597,7 +597,7 @@ s32* gmMainLib_8015D7D4(u8 arg0)
     return &base[arg0].x7C.xA8;
 }
 
-s16* gmMainLib_8015D7EC(u8 arg0)
+u16* gmMainLib_8015D7EC(u8 arg0)
 {
     struct FighterData* base = &gmMainLib_804D3EE0->thing.x1F2C[0];
     return &base[arg0].x7C.x7E;
diff --git a/src/melee/gm/gmmain_lib.h b/src/melee/gm/gmmain_lib.h
index 1d64f53e70..80318e1d1e 100644
--- a/src/melee/gm/gmmain_lib.h
+++ b/src/melee/gm/gmmain_lib.h
@@ -94,7 +94,7 @@
 /* 15D7A4 */ u16* gmMainLib_8015D7A4(u8);
 /* 15D7BC */ s32* gmMainLib_8015D7BC(u8);
 /* 15D7D4 */ s32* gmMainLib_8015D7D4(u8);
-/* 15D7EC */ s16* gmMainLib_8015D7EC(u8);
+/* 15D7EC */ u16* gmMainLib_8015D7EC(u8);
 /* 15D804 */ u32* gmMainLib_8015D804(s32);
 /* 15D818 */ s32 gmMainLib_8015D818(u32);
 /* 15D8B0 */ void gmMainLib_8015D8B0(u32);
diff --git a/src/melee/gm/types.h b/src/melee/gm/types.h
index 654721002e..56ac1b22c6 100644
--- a/src/melee/gm/types.h
+++ b/src/melee/gm/types.h
@@ -145,7 +145,7 @@ struct FighterData {
         u16 b789 : 3;      ///< stocks
         u16 b10_to_12 : 3; ///< stocks
         u16 b13_to_15 : 3; ///< stocks
-        s16 x7E;
+        u16 x7E;
         u8 x80;
         u8 x81;
         u8 x82;
diff --git a/src/melee/gr/grcastle.c b/src/melee/gr/grcastle.c
index 1559568ea5..d96e55e36e 100644
--- a/src/melee/gr/grcastle.c
+++ b/src/melee/gr/grcastle.c
@@ -102,13 +102,15 @@ StageData grCs_803E11A4 = {
 };
 
 typedef struct grCastleParams {
-    s16 x0;
-    s16 x2;
-    s16 x4;
-    s16 x6;
-    s16 x8;
-    s16 xA;
-    s32 xC;
+    /* 0x000 */ s16 x0;
+    /* 0x002 */ s16 x2;
+    /* 0x004 */ s16 x4;
+    /* 0x006 */ s16 x6;
+    /* 0x008 */ s16 x8;
+    /* 0x00A */ s16 xA;
+    /* 0x00C */ s32 xC;
+    /* 0x010 */ u8 x10[0x12C - 0x10];
+    /* 0x12C */ s16 x12C[]; // TODO: How big is this array?
 } grCastleParams;
 
 static u8 grCs_804D45E4 = 1;
@@ -382,10 +384,10 @@ void grCastle_801CE7E4(Ground_GObj* gobj) {}
 void grCastle_801CE7E8(Ground_GObj* gobj)
 {
     Ground* gp = GET_GROUND(gobj);
+    PAD_STACK(8);
     Ground_801C2ED0(GET_JOBJ(gobj), gp->map_id);
     gp->gv.castle.xC4 = 0;
-    gp->gv.castle.xC8 =
-        *(s16*) ((u8*) grCs_804D6970 + gp->gv.castle.xC4 * 2 + 0x12C);
+    gp->gv.castle.xC8 = grCs_804D6970->x12C[gp->gv.castle.xC4];
     grAnime_801C8138(gobj, gp->map_id, gp->gv.castle.xC4);
 }
 
diff --git a/src/melee/gr/grmaterial.c b/src/melee/gr/grmaterial.c
index 565f8ffb94..142ffd0e03 100644
--- a/src/melee/gr/grmaterial.c
+++ b/src/melee/gr/grmaterial.c
@@ -322,9 +322,15 @@ void grMaterial_801C95C4(HSD_GObj* gobj)
     gp->x10_flags.b4 = 1;
 }
 
+// TODO: is this GET_GROUND? calling it directly didn't work.
+inline Ground* grMaterial_801C9604_inline(HSD_GObj *arg0)
+{
+  return arg0->user_data;
+}
+
 void grMaterial_801C9604(HSD_GObj* gobj, int arg1, bool arg2)
 {
-    Ground* gp = gobj->user_data;
+    Ground* gp = grMaterial_801C9604_inline(gobj);
     ColorOverlay* co = grMaterial_GetOverlay(gp);
     co->x4_pri = arg2;
     co->x8_ptr1 = (union ColorOverlay_x8_t*) arg1;
diff --git a/src/melee/gr/grpstadium.c b/src/melee/gr/grpstadium.c
index 01475769d3..ec07930047 100644
--- a/src/melee/gr/grpstadium.c
+++ b/src/melee/gr/grpstadium.c
@@ -1278,16 +1278,20 @@ void fn_801D2FD0(Ground_GObj* gobj, int unused)
 
 void fn_801D3084(HSD_GObj* gobj, int unused)
 {
+    void* new_var2; // Permuter slop
     ImageDescWrapper* wrapper = GET_WRAPPER(gobj);
     HSD_GObj* vision_gobj;
     Ground* gp2;
+    ImageDescWrapper* new_var; // Permuter slop
 
-    if (!wrapper->flag) {
-        lb_800122C8(&wrapper->desc, wrapper->x1A, wrapper->x1C, 0);
-        wrapper->flag = true;
+    new_var = wrapper;
+    if (!new_var->flag) {
+        lb_800122C8(&new_var->desc, new_var->x1A, new_var->x1C, 0);
+        new_var->flag = true;
         vision_gobj = Ground_801C2BA4(PsType_Display);
         HSD_ASSERT(0x703, vision_gobj);
-        gp2 = GET_GROUND(vision_gobj);
+        new_var2 = HSD_GObjGetUserData(vision_gobj);
+        gp2 = (Ground*) new_var2;
         HSD_ASSERT(0x704, gp2);
         gp2->u.display.xF8_0 = true;
     }
diff --git a/src/melee/gr/grrcruise.c b/src/melee/gr/grrcruise.c
index 2e80c358b2..949b848805 100644
--- a/src/melee/gr/grrcruise.c
+++ b/src/melee/gr/grrcruise.c
@@ -166,23 +166,29 @@ void grRCruise_80200540(Ground_GObj* gobj)
 
 /// #grRCruise_80200C04
 
-void fn_802010A4(Ground_GObj* gobj, s32 id, u8* arg2)
+// TODO: is this GET_GROUND? calling it directly didn't work.
+inline Ground* fn_802010A4_inline(Ground_GObj *arg0)
 {
-    Ground* gp = gobj->user_data;
+  return arg0->user_data;
+}
+
+void fn_802010A4(Ground_GObj* gobj, s32 id, CollData* coll)
+{
+    Ground* gp = fn_802010A4_inline(gobj);
     s32 i;
 
-    if (((arg2[0x34] >> 3) & 0xF) != 1) {
+    if ((s32) coll->x34_flags.b1234 != 1) {
         return;
     }
 
     for (i = 0; i < 17; i++) {
-        u8* entry = *(u8**) (gp->gv.pad_0 + 0x6C) + i * 0x18;
-        if (*(s16*) (entry + 2) == id) {
-            if (entry[0] == 0) {
-                *(s32*) (entry + 4) = 0;
-                entry[0] = 1;
+        struct grRCruise_Entry* entry = &gp->gv.rcruise.entries[i];
+        if (entry->x02 == id) {
+            if (entry->x00 == 0) {
+                entry->x04 = 0;
+                entry->x00 = 1;
             }
-            (*(s32*) (entry + 8))++;
+            entry->x08++;
             return;
         }
     }
diff --git a/src/melee/gr/grrcruise.h b/src/melee/gr/grrcruise.h
index a1ac975034..2f81895819 100644
--- a/src/melee/gr/grrcruise.h
+++ b/src/melee/gr/grrcruise.h
@@ -48,7 +48,7 @@
 /* 20071C */ void grRCruise_8020071C(Ground_GObj*);
 /* 200B48 */ UNK_RET grRCruise_80200B48(UNK_PARAMS);
 /* 200C04 */ void grRCruise_80200C04(Ground_GObj*);
-/* 2010A4 */ void fn_802010A4(Ground_GObj*, s32, u8*);
+/* 2010A4 */ void fn_802010A4(Ground_GObj*, s32, CollData*);
 /* 201110 */ void grRCruise_80201110(Ground_GObj*);
 /* 201288 */ UNK_RET grRCruise_80201288(UNK_PARAMS);
 /* 201410 */ UNK_RET grRCruise_80201410(UNK_PARAMS);
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 62c75bfb7f..f6a5f8a301 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -730,6 +730,17 @@ struct grZebes_GroundVars5 {
     /* +3C gp+100 */ u32 x100;
 };
 
+struct grRCruise_Entry {
+    /* 0x00 */ u8 x00;
+    /* 0x01 */ u8 pad_01;
+    /* 0x02 */ s16 x02;
+    /* 0x04 */ s32 x04;
+    /* 0x08 */ s32 x08;
+    /* 0x0C */ f32 x0C;
+    /* 0x10 */ u8 pad_10[4];
+    /* 0x14 */ HSD_JObj* x14;
+};
+
 struct grRCruise_GroundVars {
     /* +00 gp+C4 */ char pad_0[0x14];
     /* +14 gp+D8 */ f32 x14;
@@ -742,6 +753,8 @@ struct grRCruise_GroundVars {
     /* +30 gp+F4 */ s32 x30;
     /* +34 gp+F8 */ s32 x34;
     /* +38 gp+FC */ s32 x38;
+    /* +3C gp+100 */ char pad_3C[0x30];
+    /* +6C gp+130 */ struct grRCruise_Entry* entries;
 };
 
 struct grFigureGet_GroundVars {
diff --git a/src/melee/if/ifcoget.c b/src/melee/if/ifcoget.c
index b284eddd9f..118c6a0236 100644
--- a/src/melee/if/ifcoget.c
+++ b/src/melee/if/ifcoget.c
@@ -42,7 +42,9 @@
     int x4;
     int x8;
     HSD_GObj* xC;
-} un_803F9E08 = { 0 };
+    char x10[12];
+    char x1C[20];
+} un_803F9E08 = { {0}, 0, 0, 0, 0, 0, 0, "IfCoGet.dat", "ScInfCgt_scene_data" };
 /* 3F9E60 */ static HSD_CObjDesc un_803F9E60 = { 0 };
 /* 3F9ED4 */ static HSD_LightDesc un_803F9ED4 = { 0 };
 
@@ -52,11 +54,14 @@ struct un_804A1F58_x8_t {
     HSD_Text* x4;
     unsigned int x8;
 };
-/* 4A1F58 */ static struct {
+
+// TODO: sizeof(un_804A1F58) should be 0x80, see un_802FF498
+// change 0x80 to sizeof(un_804A1F58) when done
+/* 4A1F58 */ static struct un_804A1F58_t {
     unsigned int x0;
-    unsigned int x4;
+    unsigned char x4;
     struct un_804A1F58_x8_t x8;
-} un_804A1F58[6];
+} un_804A1F58[7];
 
 /// .sbss
 /* 4D6DA0 */ static void* un_804D6DA0;
@@ -64,7 +69,7 @@ struct un_804A1F58_x8_t {
 
 void fn_802FED14(HSD_GObj* gobj)
 {
-    HSD_JObj* jobj = gobj->hsd_obj;
+    HSD_JObj* jobj = HSD_GObjGetHSDObj(gobj);
     if (!un_803F9E08.x0.b1) {
         HSD_JObjSetFlagsAll(jobj, 0x10);
         return;
@@ -100,12 +105,12 @@ void un_802FEFAC(void)
     HSD_JObj* jobj_ui;
     gobj_camera = GObj_Create(HSD_GOBJ_CLASS_CAMERA, 21, 0);
     HSD_GObjObject_80390A70(gobj_camera, HSD_GObj_804D784B,
-                            HSD_CObjLoadDesc(un_804D6DA4->cameras[0].desc));
+        (0, HSD_CObjLoadDesc(un_804D6DA4->cameras[0].desc)));
     GObj_SetupGXLinkMax(gobj_camera, HSD_GObj_803910D8, 9);
     gobj_camera->gxlink_prios = 0x8400;
     gobj_light = GObj_Create(HSD_GOBJ_CLASS_LIGHT, 3, 0);
     HSD_GObjObject_80390A70(gobj_light, HSD_GObj_804D784A,
-                            lb_80011AC4(un_804D6DA4->lights));
+                            (0, lb_80011AC4(un_804D6DA4->lights)));
     GObj_SetupGXLink(gobj_light, HSD_GObj_LObjCallback, 10, 0);
     gobj_ui = GObj_Create(HSD_GOBJ_CLASS_UI, 14, 0);
     jobj_ui = HSD_JObjLoadJoint(un_804D6DA4->models[0]->joint);
@@ -146,25 +151,27 @@ void un_802FF1B4(void)
 {
     un_803F9E08.x0.b0 = 1;
     un_803F9E08.x0.b1 = 0;
-    un_804D6DA0 = lbArchive_80016DBC("IfCoGet.dat", &un_804D6DA4,
-                                     "ScInfCgt_scene_data", 0);
+    un_804D6DA0 = lbArchive_80016DBC(un_803F9E08.x10, &un_804D6DA4,
+                                     un_803F9E08.x1C, 0);
     un_802FEFAC();
 }
 
 void fn_802FF218(HSD_GObj* arg0)
 {
     int x;
-    int y = -1;
+    int y;
     for (x = 0; x < 6; x++) {
         if (un_804A1F58[x].x8.x0 == arg0) {
             y = x;
-            break;
+            goto _done;
         }
     }
+    y = -1;
+    _done:
     if (y >= 0) {
         if (un_804A1F58[y].x0 == 1) {
             HSD_SisLib_803A70A0(un_804A1F58[y].x8.x4,
-                                (void*) un_804A1F58[y].x4, 0);
+                                un_804A1F58[y].x8.x8, "");
         } else {
             int s;
             gm_8016B774();
@@ -174,7 +181,8 @@ void fn_802FF218(HSD_GObj* arg0)
             }
             if (un_804A1F58[y].x0 != s) {
                 HSD_SisLib_803A70A0(un_804A1F58[y].x8.x4,
-                                    (void*) un_804A1F58[y].x4, 0);
+                                    un_804A1F58[y].x8.x8,
+                                    "%d", s);
                 un_804A1F58[y].x0 = s;
             }
         }
@@ -189,16 +197,19 @@ void un_802FF364(int slot)
     Vec3* ifAll;
     struct un_804A1F58_x8_t* thing;
     HSD_GObj* gobj;
-    thing = &un_804A1F58[slot].x8;
-    gobj = thing->x0;
+    struct un_804A1F58_t* base = un_804A1F58;
+    PAD_STACK(0x10);
+    thing = &base[slot].x8;
     ifAll = ifAll_802F3424(slot);
+    gobj = thing->x0;
+    if ((thing && thing) && thing){}
     if (gobj) {
         HSD_GObjPLink_80390228(gobj);
     }
     if (thing->x4) {
         HSD_SisLib_803A5CC4(thing->x4);
     }
-    thing->x4 = HSD_SisLib_803A6754(2, un_804A1F58->x0);
+    thing->x4 = HSD_SisLib_803A6754(2, base->x0);
     thing->x4->default_alignment = 1;
     thing->x4->default_kerning = 1;
     gm_8016B774();
@@ -216,7 +227,8 @@ void un_802FF364(int slot)
 
 void un_802FF498(void)
 {
-    memzero(un_804A1F58, sizeof(un_804A1F58));
+    PAD_STACK(8);
+    memzero(un_804A1F58, 0x80); // TODO: change to sizeof(un_802FF498) when size fixed
     un_804A1F58->x0 =
         HSD_SisLib_803A611C(2, ifAll_802F3404(), 14, 15, 0, 11, 0, 19);
 }
@@ -238,6 +250,7 @@ void un_802FF570(void)
 {
     int i;
     for (i = 0; i < 6; i++) {
+        un_804A1F58[i + 1].x4 = 1;
         if (un_804A1F58[i].x8.x4) {
             un_804A1F58[i].x8.x4->hidden = 1;
         }
@@ -247,13 +260,16 @@ void un_802FF570(void)
 void un_802FF620(void)
 {
     int i;
+    struct un_804A1F58_t* base = un_804A1F58;
+    PAD_STACK(8);
     for (i = 0; i < 6; i++) {
-        un_804A1F58[i].x0 = 0;
-        if (un_804A1F58[i].x8.x4) {
+        base[i + 1].x4 = 0;
+        if (base[i].x8.x4) {
             un_802FF364(i);
-            un_804A1F58[i].x8.x4->hidden = 0;
+            base[i].x8.x4->hidden = 0;
         }
     }
+    base[i].x8 = base[i].x8;
 }
 
 void un_802FF6A0(void)
diff --git a/src/melee/if/textdraw.c b/src/melee/if/textdraw.c
index e90bda4d56..79992adc6b 100644
--- a/src/melee/if/textdraw.c
+++ b/src/melee/if/textdraw.c
@@ -131,15 +131,18 @@ void DevText_InitPool(void)
 void DevText_Remove(DevText** ptext)
 {
     DevText* text = *ptext;
-    if (text->next) {
-        text->next->prev = text->prev;
+    struct DevText *new_var; // Permuter slop
+    DevText *new_var3; // Permuter slop
+    new_var = text->next;
+    if (new_var) {
+        new_var->prev = text->prev;
     }
-    text = *ptext;
-    if (text->prev) {
-        text->prev->next = text->next;
+    new_var3 = *ptext;
+    if ((*ptext)->prev) {
+        (*ptext)->prev->next = (*ptext)->next;
     } else {
-        if (text->next) {
-            *ptext = text->next;
+        if (new_var3->next != 0) {
+            *ptext = (*ptext)->next;
         } else {
             *ptext = NULL;
         }
diff --git a/src/melee/lb/lbsnap.c b/src/melee/lb/lbsnap.c
index 6b9e0c311c..35c9ecbe66 100644
--- a/src/melee/lb/lbsnap.c
+++ b/src/melee/lb/lbsnap.c
@@ -284,6 +284,7 @@ int lbSnap_8001DE8C(void* arg0)
 
 int lbSnap_8001DF20(void)
 {
+    // This is probably an offsetof call, but I don't know what pointer is in xC
     lbSnap_803BACC8.x14 =
         lbSnap_80433380.x0->xC +
         ((int) &lbSnap_80433380.x0->x38 - (int) &lbSnap_80433380.x0->x0);
@@ -305,6 +306,7 @@ int lbSnap_8001DF6C(int chan)
     if (ret == 0) {
         lbSnap_80433380.x48[chan].unk0 = 8;
         lbSnap_8001D4A4(chan, text);
+        // This is probably an offsetof call, but I don't know what pointer is in xC
         lbSnap_803BACC8.x14 =
             lbSnap_80433380.x0->xC +
             ((int) &lbSnap_80433380.x0->x38 - (int) &lbSnap_80433380.x0->x0);
diff --git a/src/melee/mn/mncharsel.c b/src/melee/mn/mncharsel.c
index a510bea708..adbaa95b00 100644
--- a/src/melee/mn/mncharsel.c
+++ b/src/melee/mn/mncharsel.c
@@ -256,7 +256,7 @@ void mnCharSel_8025C020(int arg0)
                                         temp_r29);
                 } else {
                     HSD_SisLib_803A70A0(mnCharSel_804D6CDC, 0,
-                                        &mnCharSel_804D5108, temp_r29);
+                                        "%d", temp_r29);
                 }
             } else {
                 gm_80163374(temp_r31, &sp7B, &sp7A, &sp79, &sp78);
diff --git a/src/melee/mn/mncharsel.static.h b/src/melee/mn/mncharsel.static.h
index 09b8176365..ba87951976 100644
--- a/src/melee/mn/mncharsel.static.h
+++ b/src/melee/mn/mncharsel.static.h
@@ -3,7 +3,6 @@
 #include <melee/mn/types.h>
 
 extern u8 mnCharSel_804D5100;
-extern u8 mnCharSel_804D5108;
 
 static CSSData* volatile mnCharSel_804D6CB0;
 static void* mnCharSel_804D6CB4;
diff --git a/src/melee/mn/mnmain.c b/src/melee/mn/mnmain.c
index 136e593002..18eafadd03 100644
--- a/src/melee/mn/mnmain.c
+++ b/src/melee/mn/mnmain.c
@@ -936,7 +936,8 @@ void mn_80229F60(HSD_GObj* gp, HSD_JObj* root, MainMenuSelection selection)
     HSD_JObj* jobj;
 
     HSD_JObj* sp54[7];
-    u8 pad50[4];
+    float new_var; // Permuter slop
+    u8 pad_50[4];
     Vec3 sp44;
     Vec3 sp38;
     Vec3 sp2C;
@@ -946,6 +947,7 @@ void mn_80229F60(HSD_GObj* gp, HSD_JObj* root, MainMenuSelection selection)
     sp44 = mn_803B84E8;
     data = gp->user_data;
     lb_8001204C(root, sp54, mn_803EAE7C, 7);
+    mn_803EB360[0].start_frame = mn_803EB360[0].start_frame; // Permuter slop
     jobj = sp54[1];
     flow = &mn_804A04F0;
     HSD_JObjReqAnimAll(jobj, 1.0F);
@@ -961,13 +963,15 @@ void mn_80229F60(HSD_GObj* gp, HSD_JObj* root, MainMenuSelection selection)
     mn_8022F3D8(jobj, 0x12, TOBJ_MASK);
     mn_8022F3D8(jobj, 0x13, TOBJ_MASK);
     HSD_JObjAnim(jobj);
+    new_var = mn_8022F298(sp54[0]);
     HSD_JObjReqAnimAll(
         sp54[0], mn_803EB360[1].start_frame +
-                     (mn_8022F298(sp54[0]) - mn_803EB360[0].start_frame));
+                     (new_var - mn_803EB360[0].start_frame));
     HSD_JObjAnimAll(sp54[0]);
+    new_var = mn_8022F298(sp54[2]);
     HSD_JObjReqAnimAll(
         sp54[2], mn_803EB378[1].start_frame +
-                     (mn_8022F298(sp54[2]) - mn_803EB378[0].start_frame));
+                     (new_var - mn_803EB378[0].start_frame));
     HSD_JObjAnimAll(sp54[2]);
     HSD_JObjReqAnimAll(sp54[3], mn_803EB390.start_frame);
     HSD_JObjAnimAll(sp54[3]);
@@ -975,6 +979,7 @@ void mn_80229F60(HSD_GObj* gp, HSD_JObj* root, MainMenuSelection selection)
     HSD_JObjClearFlagsAll(sp54[4], JOBJ_HIDDEN);
 
     HSD_JObjSetScaleY(sp54[5], 2.0F);
+    mn_803EB6B0[flow->cur_menu].start_frame += 0; // Permuter slop
     HSD_JObjSetScaleY(sp54[6], 2.0F);
 
     HSD_JObjGetTranslation(sp54[5], &sp20);
@@ -1538,12 +1543,14 @@ void mn_8022BA1C(HSD_GObj* gp)
     float var_f31;
     float var_f30;
     float y;
+    HSD_PadStatus *new_var; // Permuter slop
     int i;
 
     HSD_CObj* cobj = GET_COBJ(gp);
+    new_var = HSD_PadCopyStatus;
     for (i = 0; i < 4; i++) {
-        x = HSD_PadCopyStatus[(u8) i].nml_subStickX;
-        y = HSD_PadCopyStatus[(u8) i].nml_subStickY;
+        x = new_var[(u8) i].nml_subStickX;
+        y = new_var[(u8) i].nml_subStickY;
         if (!(-0.4F < x && x < 0.4f && -0.4F < y && y < 0.4f)) {
             break;
         }
diff --git a/src/sysdolphin/baselib/particle.c b/src/sysdolphin/baselib/particle.c
index 10f7ba3f50..2131af3f3d 100644
--- a/src/sysdolphin/baselib/particle.c
+++ b/src/sysdolphin/baselib/particle.c
@@ -7319,8 +7319,7 @@ void hsd_8039CEAC(u32 mask)
 
     bins = hsd_804D0908;
     bits = mask;
-    i = 0;
-    do {
+    for (i = 0; i < 16; i++) {
         if (!(bits & 0x10000)) {
             cur = *bins;
             prev = NULL;
@@ -7338,10 +7337,9 @@ void hsd_8039CEAC(u32 mask)
                 cur = next;
             }
         }
-        i++;
         bits >>= 1;
         bins++;
-    } while (i < 16);
+    }
 }
 
 // @TODO: Currently 96.59% match - instruction scheduling in address
diff --git a/src/sysdolphin/baselib/sislib.h b/src/sysdolphin/baselib/sislib.h
index ccf0f4d292..5d63ac09b2 100644
--- a/src/sysdolphin/baselib/sislib.h
+++ b/src/sysdolphin/baselib/sislib.h
@@ -128,7 +128,7 @@ void HSD_SisLib_803A75E0(HSD_Text*, s32);
 void HSD_SisLib_803A7548(HSD_Text*, int, float, float);
 void HSD_SisLib_803A74F0(HSD_Text*, s32 arg1, u8* arg2);
 void HSD_SisLib_803A746C(HSD_Text*, s32, f32, f32);
-s32 HSD_SisLib_803A70A0(HSD_Text*, ...);
+s32 HSD_SisLib_803A70A0(HSD_Text*, s32 entry_idx, char* fmt, ...);
 u32 lbl_803A6FEC(s32, s32, bool);
 int HSD_SisLib_803A6B98(HSD_Text* arg0, float, float, const char*, ...);
 s32 HSD_SisLib_803A67EC(u8*, u8*);
