diff --git a/src/melee/gr/groldyoshi.c b/src/melee/gr/groldyoshi.c
index 16df388523..9c0f014838 100644
--- a/src/melee/gr/groldyoshi.c
+++ b/src/melee/gr/groldyoshi.c
@@ -7,26 +7,95 @@
 #include "forward.h"
 
 #include "gr/grdisplay.h"
+#include "gr/grlib.h"
 #include "gr/grmaterial.h"
 #include "gr/ground.h"
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
-#include "mp/forward.h"
+#include "mp/mplib.h"
 
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
-
-extern StageCallbacks grOy_803E6488[];
-typedef struct { s16 x; s16 y; } grOy_JointPair;
-extern grOy_JointPair grOy_803E6574[];
+#include "baselib/random.h"
+
+StageCallbacks grOy_803E6488[] = {
+    {
+        grOldYoshi_8020E93C,
+        grOldYoshi_8020E968,
+        grOldYoshi_8020E970,
+        grOldYoshi_8020E974,
+        0
+    },
+    {
+        grOldYoshi_8020E978,
+        grOldYoshi_8020E9E0,
+        grOldYoshi_8020E9E8,
+        grOldYoshi_8020E9EC,
+        0
+    },
+    {
+        grOldYoshi_8020EAFC,
+        grOldYoshi_8020EC08,
+        grOldYoshi_8020EC10,
+        grOldYoshi_8020EFC8,
+        0
+    },
+    {
+        grOldYoshi_8020EFCC,
+        grOldYoshi_8020F080,
+        grOldYoshi_8020F088,
+        grOldYoshi_8020F2A4,
+        0
+    },
+    {
+        grOldYoshi_8020E9F0,
+        grOldYoshi_8020EA5C,
+        grOldYoshi_8020EA64,
+        grOldYoshi_8020EA88,
+        0xC0000000
+    },
+    {
+        grOldYoshi_8020EA8C,
+        grOldYoshi_8020EAEC,
+        grOldYoshi_8020EAF4,
+        grOldYoshi_8020EAF8,
+        0},
+};
+
+StageData grOy_803E650C = {
+    29,
+    grOy_803E6488,
+    "/GrOy.dat",
+    grOldYoshi_8020E79C,
+    grOldYoshi_8020E798,
+    grOldYoshi_8020E824,
+    grOldYoshi_8020E828,
+    grOldYoshi_8020E84C,
+    grOldYoshi_8020F404,
+    grOldYoshi_8020F40C,
+    1,
+    0,
+    0,
+};
 
 static struct {
-    int x0;
+    s16 x0;
+    s16 x2;
+    float x4;
+    float x8;
+    float xC;
+    s16 x10;
+    s16 x12;
+    s16 x14;
+    s16 x16;
+    s16 x18;
 }* grOy_804D6A88;
 
+
+
 void grOldYoshi_8020E798(bool arg) {}
 
 void grOldYoshi_8020E79C(void)
@@ -167,57 +236,257 @@ void grOldYoshi_8020EAF4(Ground_GObj* arg) {}
 
 void grOldYoshi_8020EAF8(Ground_GObj* arg) {}
 
+
+s16 grOy_803E6574[] = {
+    0,1,1,5,2,9
+};
 /// #grOldYoshi_8020EAFC
+void grOldYoshi_8020EAFC(Ground_GObj* arg) {
+    Ground* gp = GET_GROUND(arg);
+    HSD_JObj* jobj = arg->hsd_obj;
+    int i;
+    PAD_STACK(4);
+    Ground_801C2ED0(jobj,gp->map_id);
+    grAnime_801C8138(arg,gp->map_id,0);
+    for (i = 0; i < 3; i++) {
+        gp->gv.oldyoshicloud.cloud[i].xC4_0123 = 0;
+        gp->gv.oldyoshicloud.cloud[i].xC4_4 = 0;
+        gp->gv.oldyoshicloud.cloud[i].xC4_567 = 0;
+        gp->gv.oldyoshicloud.cloud[i].xD4 = 0.0f;
+        gp->gv.oldyoshicloud.cloud[i].xD0 = 0.0f;
+        gp->gv.oldyoshicloud.cloud[i].xC8 = Ground_801C3FA4(arg,grOy_803E6574[i*2+1]);
+        gp->gv.oldyoshicloud.cloud[i].xCC = HSD_JObjGetTranslationY(gp->gv.oldyoshicloud.cloud[i].xC8);
+        mpJointSetCb1(grOy_803E6574[i*2],gp,fn_8020F2A8);
+    }
+    gp->x11_flags.b012 = 1;
+}
 
 bool grOldYoshi_8020EC08(Ground_GObj* arg)
 {
     return false;
 }
 
-/// #grOldYoshi_8020EC10
+void grOldYoshi_8020EC10(Ground_GObj* arg)
+{
+    u16 uVar3;
+    int i = 0;
+    Ground* gp = arg->user_data;
+    HSD_JObj* jobj = arg->user_data;
+    do {
+        //int test = HSD_Randi(3);
+        switch (gp->gv.oldyoshicloud.cloud[i].xC4_0123) {
+            case 0:
+                if (gp->gv.oldyoshicloud.cloud[i].xC4_4 != 0) {
+                    //grOldYoshi_8020F31C(0,0,0,0,0,0);
+                    gp->gv.oldyoshicloud.cloud[i].xD4 = grOldYoshi_8020F31C(gp->gv.oldyoshicloud.cloud[i].xD4,grOy_804D6A88->x8,gp->gv.oldyoshicloud.cloud[i].xD0,grOy_804D6A88->xC,grOy_804D6A88->x4,-grOy_804D6A88->x4);
+                    gp->gv.oldyoshicloud.cloud[i].xD0+= gp->gv.oldyoshicloud.cloud[i].xD4;
+                    if (gp->gv.oldyoshicloud.cloud[i].xD0 > grOy_804D6A88->xC) {
+                        gp->gv.oldyoshicloud.cloud[i].xD0 = grOy_804D6A88->xC;
+                    }
+                    uVar3 =  gp->gv.oldyoshicloud.cloud[i].xC4_567++;
+                    if (uVar3 > grOy_804D6A88->x0) {
+                        gp->gv.oldyoshicloud.cloud[i].xC4_0123 = 1;
+                        grAnime_801C7FF8(arg,grOy_803E6574[i*2+1],2,1,0.0f,1.0);
+                        grLib_801C97DC(30000,30,gp->gv.oldyoshicloud.cloud[i].xC8); //spawn effect
+                        gp->gv.oldyoshicloud.cloud[i].xC4_567 = 0;
+                    }
+                } else {
+                    gp->gv.oldyoshicloud.cloud[i].xD4 = grOldYoshi_8020F31C(gp->gv.oldyoshicloud.cloud[i].xD4,grOy_804D6A88->x8,gp->gv.oldyoshicloud.cloud[i].xD0,0.0f,grOy_804D6A88->x4,-grOy_804D6A88->x4);
+                    gp->gv.oldyoshicloud.cloud[i].xD0+= gp->gv.oldyoshicloud.cloud[i].xD4;
+                    if (gp->gv.oldyoshicloud.cloud[i].xD0 < 0.0f) {
+                        gp->gv.oldyoshicloud.cloud[i].xD0 = 0.0f;
+                    }
+                    gp->gv.oldyoshicloud.cloud[i].xC4_567 = 0;
+                }
+                break;
+            case 1:
+                if (gp->gv.oldyoshicloud.cloud[i].xC4_567 == grOy_804D6A88->x10) {
+                    mpLib_80057BC0(grOy_803E6574[i*2]);
+                }
+                gp->gv.oldyoshicloud.cloud[i].xC4_567++;
+                if (grAnime_801C83D0(arg,grOy_803E6574[i*2+1],2)) {
+                    gp->gv.oldyoshicloud.cloud[i].xC4_0123 = 2;
+                    mpLib_80057BC0(grOy_803E6574[i*2]);
+                    HSD_JObjSetFlagsAll(gp->gv.oldyoshicloud.cloud[i].xC8,16);
+                    gp->gv.oldyoshicloud.cloud[i].xC4_567 = 0;
+                }
+                break;
+            case 2:
+//                gp->gv.oldyoshicloud.cloud[i].xC4_5++;
+                uVar3 = gp->gv.oldyoshicloud.cloud[i].xC4_567++;
+                if (uVar3 > grOy_804D6A88->x2) {
+                    gp->gv.oldyoshicloud.cloud[i].xC4_0123 = 3;
+                    grAnime_801C7FF8(arg,grOy_803E6574[i*2+1],2,2,0.0f,1.0);
+                    gp->gv.oldyoshicloud.cloud[i].xC4_567 = 0;
+                }
+                break;
+            case 3:
+                if (gp->gv.oldyoshicloud.cloud[i].xC4_567 == 0) {
+                    HSD_JObjClearFlagsAll(gp->gv.oldyoshicloud.cloud[i].xC8,16);
+                }
+                if (gp->gv.oldyoshicloud.cloud[i].xC4_567 == grOy_804D6A88->x12) {
+                    mpJointListAdd(grOy_803E6574[i*2]);
+                }
+                gp->gv.oldyoshicloud.cloud[i].xC4_567++;
+                if (grAnime_801C83D0(arg,grOy_803E6574[i*2+1],2)) {
+                    gp->gv.oldyoshicloud.cloud[i].xC4_0123 = 0;
+                    gp->gv.oldyoshicloud.cloud[i].xC4_567 = 0;
+                }
+                break;
+        }
+        HSD_JObjSetTranslateY(gp->gv.oldyoshicloud.cloud[i].xC8,gp->gv.oldyoshicloud.cloud[i].xCC-gp->gv.oldyoshicloud.cloud[i].xD0);
+        gp->gv.oldyoshicloud.cloud[i].xC4_4 = 0;
+        i++;
+    } while (i < 3);
+    Ground_801C2FE0(arg);
+    return;
+}
 
 void grOldYoshi_8020EFC8(Ground_GObj* arg) {}
 
+static inline s32 randi_between(s32 min, s32 max)
+{
+    if (max > min) {
+        if (min - max != 0) {
+            return max + HSD_Randi(max - min);
+        } else {
+            return max;
+        }
+    } else if (max < min) {
+        if (min - max != 0) {
+            return min + HSD_Randi(max - min);
+        } else {
+            return min;
+        }
+    }// else {
+    //    return min;
+    //}
+}
+
+static inline int rand_inline(int a, int b)
+{
+    if (a > b) {
+        return b + (a - b != 0 ? HSD_Randi(a - b) : 0);
+    } else if (a < b) {
+        return a + (b - a != 0 ? HSD_Randi(b - a) : 0);
+    } else {
+        return a;
+    }
+}
+
 /// #grOldYoshi_8020EFCC
+void grOldYoshi_8020EFCC(Ground_GObj* arg) {
+    HSD_JObj* jobj = arg->hsd_obj;
+    Ground* gp = arg->user_data;
+    HSD_JObjSetFlagsAll(jobj,0x10);
+    gp->gv.oldyoshiguest.xC4 = rand_inline(grOy_804D6A88->x16,grOy_804D6A88->x14);
+    gp->gv.oldyoshiguest.xC6 = -1;
+    gp->x11_flags.b012 = 2;
+    //oldyoshiguest;
+}
 
 bool grOldYoshi_8020F080(Ground_GObj* arg)
 {
     return false;
 }
-
+const int grOy_803B83F0[5] = {1,2,3,4,5};
 /// #grOldYoshi_8020F088
+void grOldYoshi_8020F088(Ground_GObj* arg) {
+    HSD_JObj* jobj = arg->hsd_obj;
+    Ground* gp = arg->user_data;
+    s16 sVar5;
+    float dVar9;
+    float dVar10;
+    int local34[5];
+    if (gp->gv.oldyoshiguest.xC6 == -1) {
+        sVar5 = gp->gv.oldyoshiguest.xC4;
+        gp->gv.oldyoshiguest.xC4 = sVar5-1;
+        if (sVar5 < 0) {
+            local34[0] = grOy_803B83F0[0];
+            local34[1] = grOy_803B83F0[1];
+            local34[2] = grOy_803B83F0[2];
+            local34[3] = grOy_803B83F0[3];
+            local34[4] = grOy_803B83F0[4];
+            gp->gv.oldyoshiguest.xC6 = local34[HSD_Randi(5)];
+        }
+        grAnime_801C8138(arg,gp->map_id,0);
+        dVar10 = HSD_Randf();
+        jobj = arg->hsd_obj;
+        dVar9 = grOy_804D6A88->x18*(dVar10*2.0f-1.0f);
+        HSD_JObjSetTranslateY(jobj,dVar9);
+    } else {
+        jobj = Ground_801C3FA4(arg,gp->gv.oldyoshiguest.xC6);
+        if (jobj == NULL) return;
+        if ((HSD_JObjGetFlags(jobj)&0x10) != 0) {
+            HSD_JObjClearFlagsAll(jobj,0x10);
+        }
+        if (grAnime_801C83D0(arg,0,7) != 0) {
+            HSD_JObjSetFlagsAll(jobj,0x10);
+            gp->gv.oldyoshiguest.xC4 = rand_inline(grOy_804D6A88->x16,grOy_804D6A88->x14);
+            gp->gv.oldyoshiguest.xC6 = -1;
+        }
+    }
+}
 
 void grOldYoshi_8020F2A4(Ground_GObj* arg) {}
 
-void fn_8020F2A8(Ground* gp, s32 joint_id, CollData* cd, s32 arg3,
-                 mpLib_GroundEnum arg4, f32 arg5)
-{
-    s32 idx;
-    grOy_JointPair* p;
-
-    if ((s32) cd->x34_flags.b1234 != 1) {
-        return;
+void fn_8020F2A8(Ground* gp, s32 param2, CollData* coll, s32 param4, mpLib_GroundEnum param5,float param6) {
+    int i;
+    if ((s32)coll->x34_flags.b1234 != 1) return;
+    for (i = 0; i < 3;i++) {
+        if (param2 == grOy_803E6574[i*2]) {
+            break;
+        }
     }
+    if (i == 3) return;
+    gp->gv.oldyoshicloud.cloud[i].xC4_4 = 1;
+    return;
+}
 
-    p = grOy_803E6574;
-    idx = 0;
-    if (joint_id != p[0].x) {
-        idx = 1;
-        if (joint_id != (++p)->x) {
-            idx = 2;
-            if (joint_id != (p + 1)->x) {
-                idx = 3;
-            }
+// Move related code from Context tab to here
+float grOldYoshi_8020F31C(float param1,float param2,float param3,float param4,float param5,float param6) {
+    float fVar1;
+    float fVar2;
+    float fVar3;
+    //fVar2 = param1;
+    if (param1 < 0.0f) {
+        fVar2 = -param1;
+    } else {
+        fVar2 = param1;
+    }
+    fVar3 = (fVar2/param2);
+    //fVar2 = param1;
+    if (param1 < 0.0f) {
+        fVar2 = -param1;
+    } else {
+        fVar2 = param1;
+    }
+    fVar1 = param4 - param3;
+    if (fVar1 < 0.0f) {
+        fVar1 = -fVar1;
+    }
+    if (fVar3*(-param2*0.5f*fVar3)+(fVar3*fVar2) < fVar1 && ((param4 > param3 && param1 > 0.0f) || (param4 < param3 && param1 < 0.0f))) {
+         if (param1 > 0.0f) {
+            param1 = param1 - param2;
+        } else if (param1 < 0.0f) {
+            param1 = param1 + param2;
+        }
+    } else {
+        if (param4 > param3) {
+            param1 = param1 + param2;
+        } else if (param4 < param3) {
+            param1 = param1 - param2;
         }
     }
-
-    if (idx != 3) {
-        gp->gv.oldyoshi.entries[idx].b4 = 1;
+    if (param1 > param5) {
+        return param5;
     }
+    if (param1 > param6) {
+        return param1;
+    }
+    return param6;
 }
 
-/// #grOldYoshi_8020F31C
-
 DynamicsDesc* grOldYoshi_8020F404(enum_t arg)
 {
     return NULL;
diff --git a/src/melee/gr/groldyoshi.h b/src/melee/gr/groldyoshi.h
index ef75a35908..4c7930a2f9 100644
--- a/src/melee/gr/groldyoshi.h
+++ b/src/melee/gr/groldyoshi.h
@@ -4,8 +4,6 @@
 #include <placeholder.h>
 
 #include "it/inlines.h"
-#include "lb/forward.h"
-#include "mp/forward.h"
 
 /* 20E798 */ void grOldYoshi_8020E798(bool);
 /* 20E79C */ void grOldYoshi_8020E79C(void);
@@ -38,8 +36,8 @@
 /* 20F088 */ void grOldYoshi_8020F088(Ground_GObj*);
 /* 20F2A4 */ void grOldYoshi_8020F2A4(Ground_GObj*);
 /* 20F2A8 */ void fn_8020F2A8(Ground*, s32, CollData*, s32, mpLib_GroundEnum,
-                              f32);
-/* 20F31C */ UNK_RET grOldYoshi_8020F31C(UNK_PARAMS);
+                              float);
+/* 20F31C */ float grOldYoshi_8020F31C(float,float,float,float,float,float);
 /* 20F404 */ DynamicsDesc* grOldYoshi_8020F404(enum_t);
 /* 20F40C */ bool grOldYoshi_8020F40C(Vec3*, int arg, HSD_JObj* jobj);
 
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 02e65c5c71..cfb656b8bd 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -877,17 +877,6 @@ struct grOnett_AwningData {
     /* +0x14 */ u8 pad14[8];
 };
 
-struct grOldYoshi_Entry {
-    u8 b0123 : 4;
-    u8 b4 : 1;
-    u8 b567 : 3;
-    u8 pad[0x13];
-};
-
-struct grOldYoshi_GroundVars {
-    struct grOldYoshi_Entry entries[3];
-};
-
 struct grOnett_GroundVars {
     /*  +0 gp+C4:0 */ u8 x0_b0 : 1;
     u8 pad[0xCC - 0xC5];
@@ -1018,6 +1007,25 @@ struct Map_GroundVars {
     /* +40 gp+104   */ float x104;
 };
 
+struct grOldYoshi_Cloud {
+    u32 xC4_0123:4;
+    u32 xC4_4:1;
+    u32 xC4_567:8;
+    HSD_JObj* xC8;
+    float xCC;
+    float xD0;
+    float xD4;
+};
+
+struct grOldYoshi_Cloud_GroundVars {
+    struct grOldYoshi_Cloud cloud [3];
+};
+
+struct grOldYoshi_Guest_GroundVars {
+    s16 xC4;
+    s16 xC6;
+};
+
 struct Ground {
     int x0;         // 0x0
     HSD_GObj* gobj; // 0x4
@@ -1089,6 +1097,8 @@ struct Ground {
             struct grInishie2_GroundVars2 inishie22;
             struct grInishie2_GroundVars3 inishie23;
             struct grOldKongo_GroundVars oldkongo;
+            struct grOldYoshi_Cloud_GroundVars oldyoshicloud;
+            struct grOldYoshi_Guest_GroundVars oldyoshiguest;
             struct GroundVars_izumi izumi;
             struct GroundVars_izumi2 izumi2;
             struct GroundVars_izumi3 izumi3;
@@ -1099,7 +1109,6 @@ struct Ground {
             struct grKraid_GroundVars2 kraid2;
             struct grMuteCity_GroundVars mutecity;
             struct grMuteCity_GroundVars2 mutecity2;
-            struct grOldYoshi_GroundVars oldyoshi;
             struct grOnett_GroundVars onett;
             struct grPura_GroundVars pura;
             struct grPura_GroundVars2 pura2;
