diff --git a/src/melee/gr/grcastle.c b/src/melee/gr/grcastle.c
index 83651c0943..2670ab0ab4 100644
--- a/src/melee/gr/grcastle.c
+++ b/src/melee/gr/grcastle.c
@@ -264,7 +264,16 @@ void grCastle_801CD8A8(Ground_GObj* gobj)
     }
 }
 
-/// #grCastle_801CD960
+void grCastle_801CD960(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    s32 i = 0;
+
+    do {
+        grLib_801C9B6C((u8*) gp + 0xE0 + i * 0x14);
+        i++;
+    } while (i < 12);
+}
 
 void grCastle_801CD9B4(Ground_GObj* gobj)
 {
@@ -425,7 +434,23 @@ bool grCastle_801CF300(Ground_GObj* gobj)
 
 void grCastle_801CF74C(Ground_GObj* gobj) {}
 
-/// #grCastle_801CF750
+void grCastle_801CF750(Ground* gp, s32 arg1, CollData* cd, s32 arg3,
+                       mpLib_GroundEnum arg4, f32 arg5)
+{
+    s32 idx;
+    PAD_STACK(16);
+
+    if (arg1 == 4) {
+        idx = 0;
+    } else {
+        idx = 1;
+    }
+
+    if ((s32) cd->x34_flags.b1234 == 1) {
+        gp = (Ground*) ((u8*) gp + idx * 16);
+        *(f32*) ((u8*) gp + 0xD0) += (f32) arg3;
+    }
+}
 
 /// #grCastle_801CF7B0
 void grCastle_801CF7B0(Ground_GObj* gobj)
@@ -448,7 +473,14 @@ void grCastle_801CF7B0(Ground_GObj* gobj)
 
 /// #grCastle_801CF868
 
-/// #fn_801CFAFC
+void fn_801CFAFC(Item_GObj* item, Ground* gp, Vec3* pos, HSD_GObj* gobj,
+                 f32 arg4)
+{
+    *(s16*) &gp->gv.castle2.xC4 = 4;
+    if (ftLib_80086960(gobj)) {
+        ftLib_80086A4C(gobj, (f32) grCs_804D6970->x4);
+    }
+}
 
 void fn_801CFB68(Item_GObj* item_gobj, Ground* gp, HSD_GObj* gobj)
 {
diff --git a/src/melee/gr/grcastle.h b/src/melee/gr/grcastle.h
index 8a4c44297c..714a83ed64 100644
--- a/src/melee/gr/grcastle.h
+++ b/src/melee/gr/grcastle.h
@@ -8,6 +8,7 @@
 #include "lb/forward.h"
 #include "mp/forward.h"
 #include <baselib/forward.h>
+#include <melee/it/forward.h>
 
 #include <dolphin/mtx.h>
 
@@ -69,7 +70,7 @@ struct unkCastle {
                                     mpLib_GroundEnum, f32);
 /* 1CF7B0 */ void grCastle_801CF7B0(Ground_GObj*);
 /* 1CF868 */ UNK_RET grCastle_801CF868(UNK_PARAMS);
-/* 1CFAFC */ UNK_RET fn_801CFAFC(UNK_PARAMS);
+/* 1CFAFC */ void fn_801CFAFC(Item_GObj*, Ground*, Vec3*, HSD_GObj*, f32);
 /* 1CFB68 */ void fn_801CFB68(Item_GObj*, Ground*, HSD_GObj*);
 /* 1CFBD4 */ UNK_RET grCastle_801CFBD4(UNK_PARAMS);
 /* 1D0298 */ UNK_RET grCastle_801D0298(UNK_PARAMS);
