diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 3c91c944e9..555cef7309 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -32920,7 +32920,7 @@ it_804DCB4C = .sdata2:0x804DCB4C; // type:object size:0x4 scope:global align:4 d
 @203 = .sdata2:0x804DCB58; // type:object size:0x4 scope:local align:4 data:float
 @204 = .sdata2:0x804DCB5C; // type:object size:0x4 scope:local align:4 data:float
 @201 = .sdata2:0x804DCB60; // type:object size:0x4 scope:local align:4 data:float
-it_804DCB68 = .sdata2:0x804DCB68; // type:object size:0x4 scope:global align:4 data:float
+@158 = .sdata2:0x804DCB68; // type:object size:0x4 scope:local align:4 data:float
 @201 = .sdata2:0x804DCB70; // type:object size:0x4 scope:local align:4 data:float
 it_804DCB78 = .sdata2:0x804DCB78; // type:object size:0x8 scope:global align:8 data:double
 it_804DCB80 = .sdata2:0x804DCB80; // type:object size:0x8 scope:global align:8 data:double
diff --git a/configure.py b/configure.py
index a8c806bbfd..5d21653449 100755
--- a/configure.py
+++ b/configure.py
@@ -1138,7 +1138,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itscball.c"),
             Object(Matching, "melee/it/items/itrabbitc.c"),
             Object(Matching, "melee/it/items/itmetalb.c"),
-            Object(NonMatching, "melee/it/items/itlipstick.c"),
+            Object(Matching, "melee/it/items/itlipstick.c"),
             Object(Matching, "melee/it/items/itspycloak.c"),
             Object(NonMatching, "melee/it/items/ittarucann.c"),
             Object(Matching, "melee/it/items/itmball.c"),
diff --git a/src/melee/it/items/itlipstick.c b/src/melee/it/items/itlipstick.c
index 0d989ea0f9..2c7104ba98 100644
--- a/src/melee/it/items/itlipstick.c
+++ b/src/melee/it/items/itlipstick.c
@@ -5,11 +5,28 @@
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
+
+#include "it/items/forward.h"
+
 #include "it/items/itlipstickspore.h"
 
 #define GET_ATTRS(ip)                                                         \
     ((itLipstickAttributes*) (ip)->xC4_article_data->x4_specialAttributes)
 
+ItemStateTable it_803F6310[] = {
+    { 0, itLipstick_UnkMotion0_Anim, itLipstick_UnkMotion0_Phys,
+      itLipstick_UnkMotion0_Coll },
+    { 0, itLipstick_UnkMotion4_Anim, itLipstick_UnkMotion1_Phys,
+      itLipstick_UnkMotion1_Coll },
+    { 0, itLipstick_UnkMotion2_Anim, itLipstick_UnkMotion2_Phys, NULL },
+    { 1, itLipstick_UnkMotion4_Anim, itLipstick_UnkMotion4_Phys,
+      itLipstick_UnkMotion3_Coll },
+    { 1, itLipstick_UnkMotion4_Anim, itLipstick_UnkMotion4_Phys,
+      itLipstick_UnkMotion4_Coll },
+    { 0, itLipstick_UnkMotion5_Anim, itLipstick_UnkMotion5_Phys,
+      itLipstick_UnkMotion5_Coll },
+};
+
 void itLipstick_Logic23_Spawned(Item_GObj* gobj)
 {
     Item* ip = gobj->user_data;
