diff --git a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
index 99d8baabf9..39c34fab30 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
@@ -2299,32 +2299,36 @@ void ftCo_CaptureDamageLw_Anim(Fighter_GObj* gobj)
 }
 #pragma pop
 
-#pragma push
-#pragma dont_inline on
 void ftCo_CaptureDamageLw_IASA(Fighter_GObj* gobj) {}
 
-void ftCo_CaptureDamageLw_Phys(Fighter_GObj* gobj)
+static inline void ftCo_CaptureDamageLw_Phys_inline(Fighter_GObj* gobj)
 {
-    Fighter_GObj* temp_r30;
-    Fighter* temp_r31;
+    Fighter* fp2;
     Vec3* pos;
 
-    temp_r30 = gobj;
-    if (fn_800DAD18(temp_r30)) {
-        temp_r31 = GET_FIGHTER(temp_r30);
-        Fighter_ChangeMotionState(temp_r30, ftCo_MS_CaptureDamageHi,
-                                  Ft_MF_UpdateCmd, temp_r31->cur_anim_frame,
-                                  1.0F, 0.0F, NULL);
-        fn_800DB5D8(temp_r30);
-        fn_800DAA40(temp_r30, temp_r31->victim_gobj);
-        if (!GET_FIGHTER(temp_r30)->x2226_b2) {
-            ft_80083C00(temp_r30, fn_800DC384);
+    if (fn_800DAD18(gobj)) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        Fighter_ChangeMotionState(gobj, ftCo_MS_CaptureDamageHi,
+                                  Ft_MF_UpdateCmd, fp->cur_anim_frame, 1.0F,
+                                  0.0F, NULL);
+        fn_800DB5D8(gobj);
+        fn_800DAA40(gobj, fp->victim_gobj);
+
+        fp2 = GET_FIGHTER(gobj);
+        if (!fp2->x2226_b2) {
+            ft_80083C00(gobj, fn_800DC384);
         }
-        pos = &temp_r31->cur_pos;
-        HSD_JObjSetTranslate(GET_JOBJ(temp_r30), pos);
+
+        pos = &fp->cur_pos;
+        HSD_JObjSetTranslate(GET_JOBJ(gobj), pos);
     }
 }
 
+void ftCo_CaptureDamageLw_Phys(Fighter_GObj* gobj)
+{
+    ftCo_CaptureDamageLw_Phys_inline(gobj);
+}
+
 void ftCo_CaptureDamageLw_Coll(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
@@ -2333,24 +2337,21 @@ void ftCo_CaptureDamageLw_Coll(Fighter_GObj* gobj)
     }
 }
 
-void fn_800DC624(HSD_GObj* gobj)
+static inline void fn_800DC624_inline(HSD_GObj* gobj)
 {
-    Fighter_GObj* temp_r30;
-    Fighter* temp_r31;
-    Vec3* pos;
-
-    temp_r30 = gobj;
-    temp_r31 = GET_FIGHTER(temp_r30);
-    Fighter_ChangeMotionState(temp_r30, ftCo_MS_CaptureDamageHi,
-                              Ft_MF_UpdateCmd, temp_r31->cur_anim_frame, 1.0F,
-                              0.0F, NULL);
-    fn_800DB5D8(temp_r30);
-    fn_800DAA40(temp_r30, temp_r31->victim_gobj);
-    if (!GET_FIGHTER(temp_r30)->x2226_b2) {
-        ft_80083C00(temp_r30, fn_800DC384);
+    Fighter* temp_r31 = GET_FIGHTER(gobj);
+    Fighter_ChangeMotionState(gobj, ftCo_MS_CaptureDamageHi, Ft_MF_UpdateCmd,
+                              temp_r31->cur_anim_frame, 1.0F, 0.0F, NULL);
+    fn_800DB5D8(gobj);
+    fn_800DAA40(gobj, temp_r31->victim_gobj);
+    if (!GET_FIGHTER(gobj)->x2226_b2) {
+        ft_80083C00(gobj, fn_800DC384);
     }
-    pos = &temp_r31->cur_pos;
-    HSD_JObjSetTranslate(GET_JOBJ(temp_r30), pos);
+    HSD_JObjSetTranslate(GET_JOBJ(gobj), &temp_r31->cur_pos);
 }
 
-#pragma pop
+void fn_800DC624(HSD_GObj* gobj)
+{
+    PAD_STACK(4);
+    fn_800DC624_inline(gobj);
+}
