diff --git a/configure.py b/configure.py
index 2c2504ffbb..e0d63223dd 100755
--- a/configure.py
+++ b/configure.py
@@ -1272,7 +1272,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itwhispyapple.c"),
             Object(Matching, "melee/it/items/ittools.c"),
             Object(Matching, "melee/it/items/itkyasarinegg.c"),
-            Object(NonMatching, "melee/it/items/itmasterhandlaser.c"),
+            Object(Matching, "melee/it/items/itmasterhandlaser.c"),
             Object(Matching, "melee/it/items/itmasterhandbullet.c"),
             Object(Matching, "melee/it/items/itcrazyhandbomb.c"),
             Object(Matching, "melee/it/items/itcoin.c"),
diff --git a/src/melee/it/items/itmasterhandlaser.c b/src/melee/it/items/itmasterhandlaser.c
index 52a090e2c1..86ea504c39 100644
--- a/src/melee/it/items/itmasterhandlaser.c
+++ b/src/melee/it/items/itmasterhandlaser.c
@@ -28,6 +28,15 @@
 #include <math.h>
 #include <baselib/gobj.h>
 
+ItemStateTable it_803F9378[] = {
+    {
+        0,
+        itMasterhandlaser_UnkMotion0_Anim,
+        itMasterhandlaser_UnkMotion0_Phys,
+        itMasterhandlaser_UnkMotion0_Coll,
+    },
+};
+
 static inline float my_sqrtf(float x)
 {
     static const double _half = .5;
