diff --git a/src/melee/gr/grkraid.c b/src/melee/gr/grkraid.c
index 4364af71f3..831242a821 100644
--- a/src/melee/gr/grkraid.c
+++ b/src/melee/gr/grkraid.c
@@ -4,9 +4,10 @@
 
 #include <platform.h>
 
+#include "cm/camera.h"
+
 #include "forward.h"
 
-#include "cm/camera.h"
 #include "ft/ftlib.h"
 #include "gr/granime.h"
 #include "gr/grdisplay.h"
@@ -22,7 +23,7 @@
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
 #include <baselib/jobj.h>
-#include "baselib/random.h"
+#include <baselib/random.h>
 
 S16Vec3 grKr_803E4C78[] = { { 0, 3, 12 }, { 1, 3, 12 }, { 2, 3, 12 },
                             { 3, 3, 12 }, { 4, 3, 12 }, { 5, 3, 12 } };
@@ -229,60 +230,53 @@ static inline int rand_inline(int a, int b)
     }
 }
 
-const Vec grKr_803B8278 = {0,0,0};
-
 void grKraid_801FE440(Ground_GObj* gobj)
 {
     float fVar2;
     float fVar3;
     float fVar4;
     Ground* gp = GET_GROUND(gobj);
-    HSD_JObj* jobj = Ground_801C3FA4(gobj,0);
-    Vec pos = grKr_803B8278;
-    //            pos.x = 0.0f;
-    //            pos.y = 0.0f;
-    //            pos.z = 0.0f;
-    switch(gp->gv.kraid.x0) {
-        case 0:
-            if (gp->gv.kraid.x4 != 0.0f) {
-                gp->gv.kraid.x10 = rand_inline(grKr_804D6A08->map_time_max,grKr_804D6A08->map_time_min);
-                gp->gv.kraid.x0 = 1;
-                grAnime_801C7FF8(gobj,18,7,1,0.0f,1.0f);
-                Ground_801C53EC(420007);
-                gp->gv.kraid.x1 = 1;
-            }
-            break;
-        case 1:
-            if (gp->gv.kraid.x10 == 0) {
-                gp->gv.kraid.x8 -= gp->gv.kraid.xC;
-            } else {
-                gp->gv.kraid.x10 += -1;
-            }
-            fVar2 = gp->gv.kraid.xC;
-            if (gp->gv.kraid.xC < 0.0f) {
-                fVar2 = -fVar2;
-            }
-            fVar3 = gp->gv.kraid.x8;
-            //fVar4 = fVar3;
-            if (fVar3 < 0.0f) {
-                fVar4 = -fVar3;
-            } else {
-                fVar4 = fVar3;
-            }
-            if (fVar4 < fVar2) {
-                gp->gv.kraid.xC = 0.0f;
-                gp->gv.kraid.x8 = 0.0f;
-                gp->gv.kraid.x4 = 0.0f;
-                gp->gv.kraid.x0 = 0;
-            } else {
-                fVar3 = fVar3 *(3.14159265f/180);
-                HSD_JObjAddRotationZ(jobj,fVar3);
-                Camera_80030E44(1,&pos);
-            }
-            break;
+    HSD_JObj* jobj = Ground_801C3FA4(gobj, 0);
+    Vec pos = { 0.0f, 0.0f, 0.0f };
+    switch (gp->gv.kraid.x0) {
+    case 0:
+        if (gp->gv.kraid.x4)
+        { /// @remark Explicit != 0.0f comparison leads to incorrect ordering
+            gp->gv.kraid.x10 = rand_inline(grKr_804D6A08->map_time_max,
+                                           grKr_804D6A08->map_time_min);
+            gp->gv.kraid.x0 = 1;
+            grAnime_801C7FF8(gobj, 18, 7, 1, 0.0f, 1.0f);
+            Ground_801C53EC(420007);
+            gp->gv.kraid.x1 = 1;
+        }
+        break;
+    case 1:
+        if (gp->gv.kraid.x10 == 0) {
+            gp->gv.kraid.x8 -= gp->gv.kraid.xC;
+        } else {
+            gp->gv.kraid.x10 += -1;
+        }
+        fVar2 = gp->gv.kraid.xC < 0.0f ? -gp->gv.kraid.xC : gp->gv.kraid.xC;
+        fVar3 = gp->gv.kraid.x8;
+        if (fVar3 < 0.0f) {
+            fVar4 = -fVar3;
+        } else {
+            fVar4 = fVar3;
+        }
+        if (fVar4 < fVar2) {
+            gp->gv.kraid.xC = 0.0f;
+            gp->gv.kraid.x8 = 0.0f;
+            gp->gv.kraid.x4 = 0.0f;
+            gp->gv.kraid.x0 = 0;
+        } else {
+            fVar3 = fVar3 * (3.14159265f / 180);
+            HSD_JObjAddRotationZ(jobj, fVar3);
+            Camera_80030E44(1, &pos);
+        }
+        break;
     }
-    if (gp->gv.kraid.x1 != 0 && grAnime_801C83D0(gobj,18,7) != 0) {
-        grAnime_801C7FF8(gobj,18,7,0,0.0f,1.0f);
+    if (gp->gv.kraid.x1 != 0 && grAnime_801C83D0(gobj, 18, 7) != 0) {
+        grAnime_801C7FF8(gobj, 18, 7, 0, 0.0f, 1.0f);
         gp->gv.kraid.x1 = 0;
     }
     lb_800115F4();
@@ -290,26 +284,31 @@ void grKraid_801FE440(Ground_GObj* gobj)
     grKraid_801FF150(gobj);
 }
 
-
 void grKraid_801FE6D4(Ground_GObj* gobj)
 {
     return;
 }
 
-void grKraid_801FE6D8(HSD_JObj* hand, float param2) {
+static f32 getMapRotation()
+{
+    return ((grKr_804D6A08->map_rot_spd_max / grKr_804D6A08->map_rot_spd_min) /
+            132.0f) *
+           grKr_804D6A08->map_rot_spd_min;
+}
+
+void grKraid_801FE6D8(HSD_JObj* hand, float param2)
+{
     Ground* map = Ground_801C2BA4(3)->user_data;
     Vec handpos;
-    int useless;
+    f32 rot;
+
     if (map->gv.kraid.x4 == 0.0f) {
         map->gv.kraid.x4 = param2;
-        lb_8000B1CC(hand,NULL,&handpos);
-        OSReport("Kraid Hand Pos = %f\n",handpos.x);
+        lb_8000B1CC(hand, NULL, &handpos);
+        OSReport("Kraid Hand Pos = %f\n", handpos.x);
 
-        map->gv.kraid.x8 = ((grKr_804D6A08->map_rot_spd_max/grKr_804D6A08->map_rot_spd_min)/132.0f)
-        *grKr_804D6A08->map_rot_spd_max*handpos.x;
-        if (handpos.x < 0.0f) {
-            handpos.x = -handpos.x;
-        }
+        rot = getMapRotation();
+        map->gv.kraid.x8 = rot * ((handpos.x < 0.0f) ? -handpos.x : handpos.x);
         if (map->gv.kraid.x8 < grKr_804D6A08->map_rot_spd_min) {
             map->gv.kraid.x8 = grKr_804D6A08->map_rot_spd_min;
         }
@@ -317,9 +316,9 @@ void grKraid_801FE6D8(HSD_JObj* hand, float param2) {
             map->gv.kraid.x8 = grKr_804D6A08->map_rot_spd_max;
         }
         map->gv.kraid.x8 = map->gv.kraid.x8 * map->gv.kraid.x4;
-        map->gv.kraid.xC = map->gv.kraid.x8/grKr_804D6A08->map_time_acl;
+        map->gv.kraid.xC = map->gv.kraid.x8 / grKr_804D6A08->map_time_acl;
         Ground_801C53EC(420005);
-        Camera_80030E44(3,&handpos);
+        Camera_80030E44(3, &handpos);
     }
 }
 
@@ -329,12 +328,14 @@ void grKraid_801FE818(Ground_GObj* gobj)
     float fVar2;
     float fVar3;
     int iVar5;
+
     Ground* gp = GET_GROUND(gobj);
-    gp->gv.kraid2.x10 = Ground_801C3FA4(gobj,31);
-    gp->gv.kraid2.x14 = Ground_801C3FA4(gobj,47);
+    gp->gv.kraid2.x10 = Ground_801C3FA4(gobj, 31);
+    gp->gv.kraid2.x14 = Ground_801C3FA4(gobj, 47);
     fVar1 = 0.0f;
     gp->gv.kraid2.x4 = 0;
     gp->gv.kraid2.x3 = 0;
+
     fVar2 = grKr_804D6A08->kraid_pos_x[0];
     if (fVar2 < 0.0f) {
         gp->gv.kraid2.x3 = 0;
@@ -402,7 +403,10 @@ void grKraid_801FE818(Ground_GObj* gobj)
         }
     }
     gp->gv.kraid2.x0 = 0;
-    gp->gv.kraid2.xC = grKr_804D6A08->kraid_wait_time + (grKr_804D6A08->kraid_wait_time_add != 0 ? HSD_Randi(grKr_804D6A08->kraid_wait_time_add) : 0);
+    gp->gv.kraid2.xC = grKr_804D6A08->kraid_wait_time +
+                       ((s32) grKr_804D6A08->kraid_wait_time_add != 0
+                            ? HSD_Randi(grKr_804D6A08->kraid_wait_time_add)
+                            : 0);
     gp->gv.kraid2.x8 = 0.0f;
     gp->gv.kraid2.x2 = 0;
     gp->gv.kraid2.x1 = 0;
@@ -414,104 +418,117 @@ bool grKraid_801FE9F8(Ground_GObj* gobj)
     return false;
 }
 
-void grKraid_801FEA00(Ground_GObj* gobj) {
+void grKraid_801FEA00(Ground_GObj* gobj)
+{
     s8 cVar3;
     Ground* gp = GET_GROUND(gobj);
     HSD_JObj* jobj = GET_JOBJ(gobj);
     PAD_STACK(0x70);
     switch (gp->gv.kraid2.x0) {
-        case 0:
-            if (gp->gv.kraid2.xC <= 0) {
-                Ground* map = Ground_801C2BA4(3)->user_data;
-                if ((int)map->gv.kraid.x4 == 0) {
-                    grKraid_801FEE54(gobj);
-                    HSD_JObjClearFlagsAll(jobj,0x10);
-                    HSD_JObjClearFlagsAll(Ground_801C2BA4(1)->hsd_obj,0x10);
-                    grKraid_801FF068(gobj,3);
-                    Ground_801C53EC(420001);
-                    Ground_801C53EC(420008);
+    case 0:
+        if (gp->gv.kraid2.xC <= 0) {
+            Ground* map = Ground_801C2BA4(3)->user_data;
+            if ((int) map->gv.kraid.x4 == 0) {
+                grKraid_801FEE54(gobj);
+                HSD_JObjClearFlagsAll(jobj, 0x10);
+                HSD_JObjClearFlagsAll(Ground_801C2BA4(1)->hsd_obj, 0x10);
+                grKraid_801FF068(gobj, 3);
+                Ground_801C53EC(420001);
+                Ground_801C53EC(420008);
                 gp->gv.kraid2.x0 = 1;
-                }
-            } else {
-                gp->gv.kraid2.xC += -1;
-            }
-            break;
-        case 1:
-            if (grAnime_801C83D0(gobj,0,7) != 0) {
-                grKraid_801FF068(gobj,HSD_Randi(3));
-                gp->gv.kraid2.xC = 0;
-                gp->gv.kraid2.x0 = 2;
-            }
-            break;
-        case 2:
-            cVar3 = gp->gv.kraid2.x1;
-            if (cVar3 == 0 && gp->gv.kraid2.xC == 80) {
-                Ground_801C53EC(420000);
-            } else if (cVar3 == 1 && gp->gv.kraid2.xC == 45) {
-                Ground_801C53EC(420000);
-            } else if (cVar3 == 2 && gp->gv.kraid2.xC == 55) {
-                Ground_801C5440(gp,0,420000);
             }
-            cVar3 = gp->gv.kraid2.x1;
-            if (cVar3 == 2 && gp->gv.kraid2.xC == 161) {
-                Ground_801C5544(gp,0);
-                Ground_801C53EC(420004);
-            } else if (cVar3 == 2 && gp->gv.kraid2.xC == 246) {
-                Ground_801C53EC(420004);
-            }
-            if (grAnime_801C83D0(gobj,0,7) != 0) {
-                grKraid_801FF068(gobj,gp->gv.kraid2.x8+5.0f);
-                gp->gv.kraid2.xC = 0;
-                gp->gv.kraid2.x0 = 3;
+        } else {
+            gp->gv.kraid2.xC += -1;
+        }
+        break;
+    case 1:
+        if (grAnime_801C83D0(gobj, 0, 7) != 0) {
+            grKraid_801FF068(gobj, HSD_Randi(3));
+            gp->gv.kraid2.xC = 0;
+            gp->gv.kraid2.x0 = 2;
+        }
+        break;
+    case 2:
+        cVar3 = gp->gv.kraid2.x1;
+        if (cVar3 == 0 && gp->gv.kraid2.xC == 80) {
+            Ground_801C53EC(420000);
+        } else if (cVar3 == 1 && gp->gv.kraid2.xC == 45) {
+            Ground_801C53EC(420000);
+        } else if (cVar3 == 2 && gp->gv.kraid2.xC == 55) {
+            Ground_801C5440(gp, 0, 420000);
+        }
+        cVar3 = gp->gv.kraid2.x1;
+        if (cVar3 == 2 && gp->gv.kraid2.xC == 161) {
+            Ground_801C5544(gp, 0);
+            Ground_801C53EC(420004);
+        } else if (cVar3 == 2 && gp->gv.kraid2.xC == 246) {
+            Ground_801C53EC(420004);
+        }
+        if (grAnime_801C83D0(gobj, 0, 7) != 0) {
+            grKraid_801FF068(gobj, gp->gv.kraid2.x8 + 5.0f);
+            gp->gv.kraid2.xC = 0;
+            gp->gv.kraid2.x0 = 3;
+        }
+        gp->gv.kraid2.xC++;
+        break;
+    case 3:
+        if (gp->gv.kraid2.x1 == 5) {
+            Ground* map = Ground_801C2BA4(3)->user_data;
+            if ((int) map->gv.kraid.x4 == 0 && gp->gv.kraid2.xC >= 147 &&
+                gp->gv.kraid2.xC <= 169)
+            {
+                if (gp->gv.kraid2.xC == 147) {
+                    Ground_801C53EC(420003);
+                }
+                if (grKraid_801FF0E0(gobj, 1)) {
+                    gp->gv.kraid2.x5 = 1;
+                    gp->gv.kraid2.x8 = 1.0f;
+                    grKraid_801FE6D8(gp->gv.kraid2.x14, -1.0f);
+                }
             }
             gp->gv.kraid2.xC++;
-            break;
-        case 3:
-            if (gp->gv.kraid2.x1 == 5) {
-                Ground* map = Ground_801C2BA4(3)->user_data;
-                if ((int)map->gv.kraid.x4 == 0 && gp->gv.kraid2.xC >= 147 && gp->gv.kraid2.xC <= 169) {
-                    if (gp->gv.kraid2.xC == 147) Ground_801C53EC(420003);
-                    if (grKraid_801FF0E0(gobj,1)) {
-                        gp->gv.kraid2.x5 = 1;
-                        gp->gv.kraid2.x8 = 1.0f;
-                        grKraid_801FE6D8(gp->gv.kraid2.x14,-1.0f);
-                    }
+        } else {
+            Ground* map = Ground_801C2BA4(3)->user_data;
+            if ((int) map->gv.kraid.x4 == 0 && gp->gv.kraid2.xC >= 123 &&
+                gp->gv.kraid2.xC <= 138)
+            {
+                if (gp->gv.kraid2.xC == 123) {
+                    Ground_801C53EC(420003);
                 }
-                gp->gv.kraid2.xC++;
-            } else {
-                Ground* map = Ground_801C2BA4(3)->user_data;
-                if ((int)map->gv.kraid.x4 == 0 && gp->gv.kraid2.xC >= 123 && gp->gv.kraid2.xC <= 138) {
-                    if (gp->gv.kraid2.xC == 123) Ground_801C53EC(420003);
-                    if (grKraid_801FF0E0(gobj,0)) {
-                        gp->gv.kraid2.x5 = 1;
-                        gp->gv.kraid2.x8 = 1.0f;
-                        grKraid_801FE6D8(gp->gv.kraid2.x10,1.0f);
-                    }
+                if (grKraid_801FF0E0(gobj, 0)) {
+                    gp->gv.kraid2.x5 = 1;
+                    gp->gv.kraid2.x8 = 1.0f;
+                    grKraid_801FE6D8(gp->gv.kraid2.x10, 1.0f);
                 }
-                gp->gv.kraid2.xC++;
-            }
-            if (grAnime_801C83D0(gobj,0,7) != 0) {
-                grKraid_801FF068(gobj,4);
-                Ground_801C53EC(420002);
-                Ground_801C53EC(420008);
-                gp->gv.kraid2.x0 = 4;
             }
-            break;
-        case 4:
-            if (grAnime_801C83D0(gobj,0,7) != 0) {
-                gp->gv.kraid2.xC = grKr_804D6A08->kraid_wait_time
-                + grKr_804D6A08->kraid_wait_time_add;
-                HSD_JObjSetFlagsAll(jobj,0x10);
-                HSD_JObjSetFlagsAll(Ground_801C2BA4(1)->hsd_obj,0x10);
-                gp->gv.kraid2.xC = grKr_804D6A08->kraid_wait_time
-                + (grKr_804D6A08->kraid_wait_time_add != 0 ? HSD_Randi(grKr_804D6A08->kraid_wait_time_add) : 0);
-                gp->gv.kraid2.x0 = 0;
-            }
-            break;
+            gp->gv.kraid2.xC++;
+        }
+        if (grAnime_801C83D0(gobj, 0, 7) != 0) {
+            grKraid_801FF068(gobj, 4);
+            Ground_801C53EC(420002);
+            Ground_801C53EC(420008);
+            gp->gv.kraid2.x0 = 4;
+        }
+        break;
+    case 4:
+        if (grAnime_801C83D0(gobj, 0, 7) != 0) {
+            gp->gv.kraid2.xC = grKr_804D6A08->kraid_wait_time +
+                               grKr_804D6A08->kraid_wait_time_add;
+            HSD_JObjSetFlagsAll(jobj, 0x10);
+            HSD_JObjSetFlagsAll(Ground_801C2BA4(1)->hsd_obj, 0x10);
+            gp->gv.kraid2.xC =
+                grKr_804D6A08->kraid_wait_time +
+                ((s32) grKr_804D6A08->kraid_wait_time_add != 0
+                     ? HSD_Randi(grKr_804D6A08->kraid_wait_time_add)
+                     : 0);
+            gp->gv.kraid2.x0 = 0;
+        }
+        break;
     }
 }
 
-void grKraid_801FEE54(HSD_GObj* gobj) {
+void grKraid_801FEE54(HSD_GObj* gobj)
+{
     s32 iVar3;
     Ground* gp = GET_GROUND(gobj);
     HSD_JObj* jobj = GET_JOBJ(gobj);
@@ -521,16 +538,13 @@ void grKraid_801FEE54(HSD_GObj* gobj) {
             iVar3 = HSD_Randi(6);
         } while (gp->gv.kraid2.x2 == iVar3);
     } else {
-        if (HSD_Randi(2) != 0) {
-            iVar3 = gp->gv.kraid2.x3;
-        } else {
-            iVar3 = gp->gv.kraid2.x4;
-        }
+        iVar3 = HSD_Randi(2) != 0 ? gp->gv.kraid2.x3 : gp->gv.kraid2.x4;
     }
     gp->gv.kraid2.x5 = 0;
     gp->gv.kraid2.x2 = iVar3;
-    HSD_JObjSetTranslateX(jobj,grKr_804D6A08->kraid_pos_x[iVar3]);
-    HSD_JObjSetTranslateX(Ground_801C2BA4(1)->hsd_obj,grKr_804D6A08->kraid_pos_x[iVar3]);
+    HSD_JObjSetTranslateX(jobj, grKr_804D6A08->kraid_pos_x[iVar3]);
+    HSD_JObjSetTranslateX(Ground_801C2BA4(1)->hsd_obj,
+                          grKr_804D6A08->kraid_pos_x[iVar3]);
     if (grKr_804D6A08->kraid_pos_x[iVar3] == 0.0f) {
         if (HSD_Randi(2) != 0) {
             iVar3 = 1;
diff --git a/src/melee/gr/grkraid.h b/src/melee/gr/grkraid.h
index 779f09e45d..f2aa77d8dd 100644
--- a/src/melee/gr/grkraid.h
+++ b/src/melee/gr/grkraid.h
@@ -43,7 +43,7 @@ typedef struct grKr_804D6A08_t {
 /* 1FE438 */ bool grKraid_801FE438(Ground_GObj*);
 /* 1FE440 */ void grKraid_801FE440(Ground_GObj*);
 /* 1FE6D4 */ void grKraid_801FE6D4(Ground_GObj*);
-/* 1FE6D8 */ void grKraid_801FE6D8(HSD_JObj*,float);
+/* 1FE6D8 */ void grKraid_801FE6D8(HSD_JObj*, float);
 /* 1FE818 */ void grKraid_801FE818(Ground_GObj*);
 /* 1FE9F8 */ bool grKraid_801FE9F8(Ground_GObj*);
 /* 1FEA00 */ void grKraid_801FEA00(Ground_GObj*);
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 003394c7b7..d3ff06d4de 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -288,7 +288,7 @@ struct grKongo_GroundVars3 {
 };
 
 struct grKraid_GroundVars {
-    /*  + gp+C4 */ u8 x0;
+    /*  + gp+C4 */ s8 x0;
     /*  + gp+C5 */ s8 x1;
     /*  + gp+C8 */ f32 x4;
     /*  + gp+CC */ f32 x8;
@@ -300,7 +300,7 @@ struct grKraid_GroundVars2 {
     /*  + gp+C4 */ s8 x0;
     /*  + gp+C5 */ s8 x1;
     /*  + gp+C6 */ s8 x2;
-    /*  + gp+C7 */ u8 x3;
+    /*  + gp+C7 */ s8 x3;
     /*  + gp+C8 */ s8 x4;
     /*  + gp+C9 */ s8 x5;
     /*  + gp+CC */ f32 x8;
@@ -1017,9 +1017,9 @@ struct Map_GroundVars {
 };
 
 struct grOldYoshi_Cloud {
-    u32 xC4_0123:4;
-    u32 xC4_4:1;
-    u32 xC4_567:8;
+    u32 xC4_0123 : 4;
+    u32 xC4_4 : 1;
+    u32 xC4_567 : 8;
     HSD_JObj* xC8;
     float xCC;
     float xD0;
@@ -1027,7 +1027,7 @@ struct grOldYoshi_Cloud {
 };
 
 struct grOldYoshi_Cloud_GroundVars {
-    struct grOldYoshi_Cloud cloud [3];
+    struct grOldYoshi_Cloud cloud[3];
 };
 
 struct grOldYoshi_Guest_GroundVars {
