diff --git a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
index 5d3784b0af..dfad22be34 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
@@ -1104,6 +1104,45 @@ void ftCo_ItemScopeAirFire_Coll(Fighter_GObj* gobj)
     ft_80082C74(gobj, fn_800D8268);
 }
 
+static void fn_800D84D4(Fighter_GObj* gobj, int arg1)
+{
+    extern int ftCo_804D9030;
+    Fighter* fp = gobj->user_data;
+    Vec3 sp20;
+    Vec3 sp14;
+    PAD_STACK(4);
+
+    if (fp->item_gobj != NULL) {
+        if (ftCheckThrowB0(fp)) {
+            it_80291CCC(fp->item_gobj, &sp14);
+            ftCo_800CDE94(fp, &sp14, &sp20);
+            if (arg1 == 0) {
+                it_80291FA8(fp->item_gobj, &sp20, arg1, fp->facing_dir);
+                sp20.z = *(f32*) &ftCo_804D9030;
+                efSync_Spawn(0x430, fp->gobj, &sp20, &fp->facing_dir);
+                ft_PlaySFX(fp, 0xFC, 0x7F, 0x40);
+            } else if (arg1 >= 1 && arg1 <= 8) {
+                it_80291FA8(fp->item_gobj, &sp20, arg1, fp->facing_dir);
+                sp20.z = *(f32*) &ftCo_804D9030;
+                efSync_Spawn(0x430, fp->gobj, &sp20, &fp->facing_dir);
+                if (arg1 <= 4) {
+                    ft_PlaySFX(fp, 0xFF, 0x7F, 0x40);
+                } else {
+                    ft_PlaySFX(fp, 0x100, 0x7F, 0x40);
+                }
+            } else if (arg1 == 9) {
+                it_80291FA8(fp->item_gobj, &sp20, arg1, fp->facing_dir);
+                sp20.z = *(f32*) &ftCo_804D9030;
+                efSync_Spawn(0x430, fp->gobj, &sp20, &fp->facing_dir);
+                ft_PlaySFX(fp, 0x100, 0x7F, 0x40);
+            } else {
+                efSync_Spawn(0x405, gobj, &sp20);
+                ft_PlaySFX(fp, 0x101, 0x7F, 0x40);
+            }
+        }
+    }
+}
+
 void fn_800D86B8(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.c b/src/melee/ft/chara/ftKirby/ftKb_Init.c
index 76e2e84f53..3d1fb54e5c 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.c
@@ -4345,7 +4345,7 @@ void ftKb_SpecialN_800F1D24(Fighter_GObj* gobj)
         if (x60 != 0) {
             fp->fv.kb.x60 = x60 - 1;
         } else {
-            ftKb_SpecialN_800F1DAC();
+            ftKb_SpecialN_800F1DAC(gobj);
         }
         if ((coll->env_flags & 0x800) || (coll->env_flags & 0x20) ||
             (coll->env_flags & 0x4000) || (coll->env_flags & 0x10000))
@@ -4355,7 +4355,45 @@ void ftKb_SpecialN_800F1D24(Fighter_GObj* gobj)
     }
 }
 
-/// #ftKb_SpecialN_800F1DAC
+void ftKb_SpecialN_800F1DAC(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    CollData* coll = &fp->coll_data;
+    Vec3 vec;
+    PAD_STACK(16);
+
+    if ((coll->env_flags & 0x800) && !(coll->prev_env_flags & 0xFC0)) {
+        vec.x = coll->ecb.left.x;
+        vec.y = coll->ecb.left.y;
+        vec.z = 0.0f;
+        efAsync_Spawn(gobj, &GET_FIGHTER(gobj)->x60C, 2, 0x49E,
+                      fp->parts[0].joint, &vec);
+        return;
+    }
+    if ((coll->env_flags & 0x20) && !(coll->prev_env_flags & 0x3F)) {
+        vec.x = coll->ecb.right.x;
+        vec.y = coll->ecb.right.y;
+        vec.z = 0.0f;
+        efAsync_Spawn(gobj, &GET_FIGHTER(gobj)->x60C, 2, 0x49E,
+                      fp->parts[0].joint, &vec);
+        return;
+    }
+    if ((coll->env_flags & 0x4000) && !(coll->prev_env_flags & 0x6000)) {
+        vec.x = 0.0f;
+        vec.y = coll->ecb.top.y;
+        vec.z = 0.0f;
+        efAsync_Spawn(gobj, &GET_FIGHTER(gobj)->x60C, 2, 0x49E,
+                      fp->parts[0].joint, &vec);
+        return;
+    }
+    if ((coll->env_flags & 0x10000) && !(coll->prev_env_flags & 0x18000)) {
+        vec.x = 0.0f;
+        vec.y = coll->ecb.bottom.y;
+        vec.z = 0.0f;
+        efAsync_Spawn(gobj, &GET_FIGHTER(gobj)->x60C, 2, 0x49E,
+                      fp->parts[0].joint, &vec);
+    }
+}
 
 void ftKb_SpecialN_800F1F1C(Fighter_GObj* gobj, Vec3* pos)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.h b/src/melee/ft/chara/ftKirby/ftKb_Init.h
index 0f11b4dbb6..a62f1ce2a0 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.h
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.h
@@ -115,7 +115,7 @@
 /* 0F1BAC */ void ftKb_SpecialN_800F1BAC(Fighter_GObj* gobj, s32, bool);
 /* 0F1CD8 */ s32 ftKb_SpecialN_800F1CD8(HSD_GObj* gobj);
 /* 0F1D24 */ void ftKb_SpecialN_800F1D24(Fighter_GObj* gobj);
-/* 0F1DAC */ UNK_RET ftKb_SpecialN_800F1DAC(UNK_PARAMS);
+/* 0F1DAC */ void ftKb_SpecialN_800F1DAC(HSD_GObj* gobj);
 /* 0F1F1C */ void ftKb_SpecialN_800F1F1C(Fighter_GObj* gobj, Vec3*);
 /* 0F1F68 */ void ftKb_SpecialN_800F1F68(Fighter_GObj* gobj);
 /* 0F1FDC */ void fn_800F1FDC(Fighter_GObj* gobj);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index 2a5df29746..890969a652 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -1223,7 +1223,87 @@ void ftKb_SpecialAirLwEnd_Phys(Fighter_GObj* gobj)
 
 /// #ftKb_SpecialLw1_Coll
 
-/// #ftKb_SpecialLw_Coll
+struct ftKb_Init_803CB490_layout {
+    char pad[0x74];
+    Vec3 vec;
+};
+
+void ftKb_SpecialLw_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter* fp2;
+    Fighter* fp3;
+    s32 temp;
+    struct ftKb_Init_803CB490_layout* p =
+        (struct ftKb_Init_803CB490_layout*) ftKb_Init_803CB490;
+    PAD_STACK(16);
+
+    if (ft_80082708(gobj) == GA_Ground) {
+        fp2 = GET_FIGHTER(gobj);
+        fp2->mv.kb.speciallw.x24 = p->vec;
+        fp2->mv.kb.speciallw.x54 = fp2->mv.kb.speciallw.x24;
+        fp2->mv.kb.speciallw.x88[0] = 0.0f;
+        fp2->mv.kb.speciallw.x88[4] = 0.0f;
+        fp2->mv.kb.speciallw.x30 = p->vec;
+        fp2->mv.kb.speciallw.x60 = fp2->mv.kb.speciallw.x30;
+        fp2->mv.kb.speciallw.x88[1] = 0.0f;
+        fp2->mv.kb.speciallw.x88[5] = 0.0f;
+        fp2->mv.kb.speciallw.x3C = p->vec;
+        fp2->mv.kb.speciallw.x6C = fp2->mv.kb.speciallw.x3C;
+        fp2->mv.kb.speciallw.x88[2] = 0.0f;
+        fp2->mv.kb.speciallw.x88[6] = 0.0f;
+        fp2->mv.kb.speciallw.x48 = p->vec;
+        fp2->mv.kb.speciallw.x78 = fp2->mv.kb.speciallw.x48;
+        fp2->mv.kb.speciallw.x88[3] = 0.0f;
+        fp2->mv.kb.speciallw.x88[7] = 0.0f;
+        fp2->mv.kb.speciallw.x18 = p->vec;
+        fp2->mv.kb.speciallw.x84 = 0.0f;
+        fp->self_vel.x = 0.0f;
+        ftCommon_8007D5D4(fp);
+        temp = fp->x221C_b4;
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirLw, 0x0C4C1092U,
+                                  0.0f, 0.0f, 0.0f, NULL);
+        fp->x221C_b4 = temp;
+        ftAnim_SetAnimRate(gobj, 0.0f);
+        fp3 = GET_FIGHTER(gobj);
+        fp3->take_dmg_cb = ftKb_Init_800EE7B8;
+        fp3->death2_cb = ftKb_Init_800EE74C;
+        ftPartSetRotX(fp, 0, 0.0f);
+    } else {
+        ftKb_SpecialHi_800F3570(gobj);
+        ftKb_SpecialHi_800F37EC(gobj);
+    }
+}
+
+void ftKb_SpecialLwEnd_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter* fp2;
+    struct ftKb_Init_803CB490_layout* p =
+        (struct ftKb_Init_803CB490_layout*) ftKb_Init_803CB490;
+    PAD_STACK(8);
+    ft_80081D0C(gobj);
+    fp2 = GET_FIGHTER(gobj);
+    fp2->mv.kb.speciallw.x24 = p->vec;
+    fp2->mv.kb.speciallw.x54 = fp2->mv.kb.speciallw.x24;
+    fp2->mv.kb.speciallw.x88[0] = 0.0f;
+    fp2->mv.kb.speciallw.x88[4] = 0.0f;
+    fp2->mv.kb.speciallw.x30 = p->vec;
+    fp2->mv.kb.speciallw.x60 = fp2->mv.kb.speciallw.x30;
+    fp2->mv.kb.speciallw.x88[1] = 0.0f;
+    fp2->mv.kb.speciallw.x88[5] = 0.0f;
+    fp2->mv.kb.speciallw.x3C = p->vec;
+    fp2->mv.kb.speciallw.x6C = fp2->mv.kb.speciallw.x3C;
+    fp2->mv.kb.speciallw.x88[2] = 0.0f;
+    fp2->mv.kb.speciallw.x88[6] = 0.0f;
+    fp2->mv.kb.speciallw.x48 = p->vec;
+    fp2->mv.kb.speciallw.x78 = fp2->mv.kb.speciallw.x48;
+    fp2->mv.kb.speciallw.x88[3] = 0.0f;
+    fp2->mv.kb.speciallw.x88[7] = 0.0f;
+    fp2->mv.kb.speciallw.x18 = p->vec;
+    fp2->mv.kb.speciallw.x84 = 0.0f;
+    ftPartSetRotX(fp, 0, 0.0f);
+}
 
 /// #ftKb_SpecialAirLwStart_Coll
 
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
index 648bfa8db7..12db4fa48f 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
@@ -735,7 +735,36 @@ void ftKb_SpecialNYs_8010AA2C(Fighter_GObj* gobj)
     ftColl_8007B62C(gobj, 2);
 }
 
-/// #fn_8010AA64
+static void fn_8010AA64(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    HSD_JObj* jobj = gobj->hsd_obj;
+    if (fp->mv.co.yoshiegg.x14 <= 0.0f) {
+        Fighter_UpdateModelScale(gobj);
+        fp->accessory4_cb = NULL;
+        return;
+    }
+    fp->mv.co.yoshiegg.x14 -= 1.0f;
+    {
+        Vec3 vec1, vec0;
+        f32 temp0 = (fp->mv.co.yoshiegg.x10 - fp->mv.co.yoshiegg.x14) /
+                    fp->mv.co.yoshiegg.x10;
+        temp0 *= fp->mv.co.yoshiegg.xC;
+        {
+            f32 temp1 = (1.0f - fp->mv.co.yoshiegg.xC) + temp0;
+            vec1.x = vec1.y = vec1.z = temp1;
+            vec0 = vec1;
+            vec1.x = vec1.x * fp->mv.co.yoshiegg.x18.x;
+            vec1.y *= fp->mv.co.yoshiegg.x18.y;
+            vec1.z *= fp->mv.co.yoshiegg.x18.z;
+            HSD_JObjSetScale(jobj, &vec1);
+            vec0.x = vec0.x * fp->mv.co.yoshiegg.scale.x;
+            vec0.y *= fp->mv.co.yoshiegg.scale.y;
+            vec0.z *= fp->mv.co.yoshiegg.scale.z;
+            HSD_JObjSetScale(fp->x20A0_accessory, &vec0);
+        }
+    }
+}
 
 /// #ftKb_SpecialNYs_8010AC78
 
diff --git a/src/melee/gm/gm_1601.c b/src/melee/gm/gm_1601.c
index 38d91c18b8..7b1408d5db 100644
--- a/src/melee/gm/gm_1601.c
+++ b/src/melee/gm/gm_1601.c
@@ -826,19 +826,77 @@ f32 fn_80160F58(u8 ckind)
     return result;
 }
 
-/// #fn_80161004
+s32 fn_80161004(MatchEnd* match_end)
+{
+    s32 max = 0;
+    s32 i;
+
+    if (match_end->is_teams == 1) {
+        for (i = 0; i < 5; i++) {
+            struct MatchTeamData* t = &match_end->team_standings[i];
+            if (match_end->team_standings[i].active != 0 &&
+                (s32) t->is_big_loser > max)
+            {
+                max = t->is_big_loser;
+            }
+        }
+    } else {
+        for (i = 0; i < 4; i++) {
+            struct MatchPlayerData* p = &match_end->player_standings[i];
+            if (match_end->player_standings[i].slot_type != 3 &&
+                (s32) p->is_big_loser > max)
+            {
+                max = p->is_big_loser;
+            }
+        }
+    }
+    return max;
+}
 
 /// #fn_80161154
 
 /// #fn_80161C90
 
-/// #fn_80162068
+void fn_80162068(MatchEnd* match_end)
+{
+    s32 i;
+    s32 j;
+    struct MatchPlayerData* pdata_i;
+    struct MatchPlayerData* pdata_j;
+    struct FighterData* fd;
+    s32 sum;
+
+    for (i = 0; i < 4; i++) {
+        pdata_i = &match_end->player_standings[i];
+        if (pdata_i->slot_type == 3) {
+            continue;
+        }
+        fd = GetPersistentFighterData(gm_80164024(pdata_i->character_kind));
+        for (j = 0; j < 4; j++) {
+            pdata_j = &match_end->player_standings[j];
+            if (i == j || pdata_j->slot_type == 3) {
+                continue;
+            }
+            if (pdata_i->kills[j] +
+                    fd->fighter_kos[gm_80164024(pdata_j->character_kind)] >
+                0xFFFF)
+            {
+                sum = 0xFFFF;
+            } else {
+                sum = pdata_i->kills[j] +
+                      fd->fighter_kos[gm_80164024(pdata_j->character_kind)];
+            }
+            fd->fighter_kos[gm_80164024(pdata_j->character_kind)] = (u16) sum;
+        }
+        fn_80161C90(match_end, i, &fd->sd_count);
+    }
+}
 
 /// #fn_80162170
 
 s32 gm_801623A4(MatchEnd* arg0)
 {
-    fn_80162068();
+    fn_80162068(arg0);
     return fn_80162170(arg0);
 }
 
@@ -2180,7 +2238,30 @@ void fn_80165E7C(MatchEnd* arg0)
 
 /// #fn_80165FA4
 
-/// #fn_801661E0
+s32 fn_801661E0(MatchEnd* arg0)
+{
+    s32 i;
+    s32 j;
+
+    for (i = 0; i < 5; i++) {
+        if ((u8) arg0->team_standings[i].active != 0) {
+            arg0->team_standings[i].is_small_loser =
+                arg0->team_standings[i].is_big_loser;
+            for (j = 0; j < 5; j++) {
+                if ((u8) arg0->team_standings[j].active != 0 && i != j &&
+                    (u8) arg0->team_standings[i].is_big_loser ==
+                        (u8) arg0->team_standings[j].is_big_loser &&
+                    (u32) arg0->team_standings[i].subscore <
+                        (u32) arg0->team_standings[j].subscore)
+                {
+                    arg0->team_standings[i].is_small_loser++;
+                }
+            }
+        }
+    }
+
+    return (s32) arg0;
+}
 
 /// #gm_80166378
 
@@ -3439,7 +3520,54 @@ void fn_80169900(u8 arg0, struct lbl_8046B488_t* arg1, s8* arg2, s8* arg3)
 
 /// #fn_80169C54
 
-/// #fn_80169F50
+void fn_80169F50(s8 arg0, s8 arg1)
+{
+    struct lbl_8046B488_t* gp = &lbl_8046B488;
+    s32 i;
+    s32 flag;
+    s8 v;
+    s8* p;
+
+    flag = 0;
+    if (gp->x0 == 4) {
+        flag = 1;
+    } else {
+        p = &gp->x1;
+        if (gp->x1 == 4) {
+            flag = 1;
+        } else if (p[1] == 4) {
+            flag = 1;
+        }
+    }
+
+    if (flag == 1 && gp->xE != 0) {
+        for (i = 0; gp->x20[i] != -2; i++) {
+            v = gp->x20[i];
+            if (v == -1) {
+                continue;
+            }
+            if (gp->xA2[i] != 4) {
+                continue;
+            }
+            if (gp->x124[i] == -1) {
+                continue;
+            }
+            Player_80031DA8(gp->x124[i], v);
+        }
+    }
+
+    if (arg0 == 4 && gp->xE != 0) {
+        for (i = 0; gp->x20[i] != -2; i++) {
+            if (gp->x20[i] == -1) {
+                continue;
+            }
+            if (gp->x124[i] == -1) {
+                continue;
+            }
+            Player_80031DA8(gp->x124[i], arg1);
+        }
+    }
+}
 
 void fn_8016A09C(void)
 {
diff --git a/src/melee/gm/gm_1601.h b/src/melee/gm/gm_1601.h
index 1cb2eb6ca7..69dcbc37c0 100644
--- a/src/melee/gm/gm_1601.h
+++ b/src/melee/gm/gm_1601.h
@@ -53,10 +53,10 @@ struct lbl_8046B488_t;
 /* 160C90 */ void gm_80160C90(HSD_Text* text, u8 fighter_id, bool arg2);
 /* 160DE8 */ void fn_80160DE8(HSD_JObj*, u8, s32, u8, f32, f32);
 /* 160F58 */ float fn_80160F58(u8 ckind);
-/* 161004 */ UNK_RET fn_80161004(UNK_PARAMS);
+/* 161004 */ s32 fn_80161004(MatchEnd*);
 /* 161154 */ UNK_RET fn_80161154(UNK_PARAMS);
-/* 161C90 */ UNK_RET fn_80161C90(UNK_PARAMS);
-/* 162068 */ void fn_80162068(void);
+/* 161C90 */ void fn_80161C90(MatchEnd*, s32, u16*);
+/* 162068 */ void fn_80162068(MatchEnd*);
 /* 162170 */ s32 fn_80162170(MatchEnd*);
 /* 1623A4 */ s32 gm_801623A4(MatchEnd*);
 /* 1623D8 */ int gm_801623D8(void);
diff --git a/src/melee/gm/gm_1BA8.c b/src/melee/gm/gm_1BA8.c
index 380ffd4658..a6f1e68730 100644
--- a/src/melee/gm/gm_1BA8.c
+++ b/src/melee/gm/gm_1BA8.c
@@ -153,7 +153,60 @@ void gm_801BAAD0(MinorScene* arg0)
     temp_r31->xA = -1;
 }
 
-/// #gm_801BAB40
+typedef struct gm_801BAB40_src {
+    /* 0x00 */ u8 c_kind;
+    /* 0x01 */ u8 slot_type;
+    /* 0x02 */ u8 stocks;
+    /* 0x03 */ u8 color;
+    /* 0x04 */ u8 x5;
+    /* 0x05 */ u8 sub_color;
+    /* 0x06 */ u8 team;
+    /* 0x07 */ u8 xB;
+    /* 0x08 */ u8 flags;
+    /* 0x09 */ u8 xE;
+    /* 0x0A */ u8 cpu_level;
+    /* 0x0B */ u8 pad;
+    /* 0x0C */ u16 x12;
+    /* 0x0E */ u16 hp;
+    /* 0x10 */ f32 x18;
+    /* 0x14 */ f32 x1C;
+    /* 0x18 */ f32 x20;
+} gm_801BAB40_src;
+
+void gm_801BAB40(PlayerInitData* arg0, int arg1)
+{
+    gm_801BAB40_src* src = (gm_801BAB40_src*) arg1;
+    gm_8016795C(arg0);
+    arg0->c_kind = src->c_kind;
+    arg0->slot_type = src->slot_type;
+    arg0->stocks = src->stocks;
+    arg0->color = src->color;
+    arg0->slot = 0;
+    arg0->x5 = src->x5;
+    arg0->sub_color = src->sub_color;
+    arg0->handicap = 9;
+    arg0->team = src->team;
+    arg0->xA = 0x78;
+    arg0->xB = src->xB;
+    arg0->xC_b0 = 0;
+    arg0->xC_b1 = (src->flags & 0x80) >> 7;
+    arg0->xC_b2 = (src->flags & 0x40) >> 6;
+    arg0->xC_b3 = (src->flags & 0x20) >> 5;
+    arg0->xC_b4 = (src->flags & 0x10) >> 4;
+    arg0->xC_b6 = (src->flags & 0x08) >> 3;
+    arg0->xC_b7 = (src->flags & 0x04) >> 2;
+    arg0->xD_b1 = 0;
+    arg0->xD_b2 = 0;
+    arg0->xD_b4 = 0;
+    arg0->xE = src->xE;
+    arg0->cpu_level = src->cpu_level;
+    arg0->x10 = 0;
+    arg0->x12 = src->x12;
+    arg0->hp = src->hp;
+    arg0->x18 = src->x18;
+    arg0->x1C = src->x1C;
+    arg0->x20 = src->x20;
+}
 
 /// #gm_801BAC9C
 
@@ -760,7 +813,77 @@ void gm_801BD44C(HSD_GObj* gobj)
     gm_801BCC9C(gobj);
 }
 
-/// #gm_801BD46C
+void gm_801BD46C(HSD_GObj* gobj)
+{
+    lbl_8046B6A0_t* temp_r3;
+    s32 var_r0;
+    struct EventData* temp_r31;
+    s32 var_r30;
+    int i;
+    int count;
+    HSD_GObj* p;
+    PAD_STACK(0x38);
+
+    count = 0;
+    for (i = 1; i < 3; i++) {
+        p = Player_GetEntityAtIndex(i, 1);
+        if (p != NULL) {
+            var_r30 = ftLib_8008731C(p);
+        }
+        if (var_r30 != 0) {
+            count++;
+        }
+    }
+    if (count == 2) {
+        gm_801BC4F4(gobj);
+        return;
+    }
+    for (i = 1; i < 3; i++) {
+        if (Player_GetStocks(i) <= 0) {
+            gmMainLib_804D3EE0->unk_530.xB_1 = false;
+            lbAudioAx_80028B90();
+            gm_SetGameSpeed(1.0F);
+            gm_8016B33C(6);
+            gm_8016B364(0x148);
+            gm_8016B378(0x28);
+            gm_8016B328();
+            HSD_GObjPLink_80390228(gobj);
+            return;
+        }
+    }
+    if (Player_GetP1Stock() <= 0) {
+        gmMainLib_804D3EE0->unk_530.xB_1 = false;
+        lbAudioAx_80028B90();
+        gm_SetGameSpeed(1.0F);
+        gm_8016B33C(6);
+        gm_8016B364(0x148);
+        gm_8016B378(0x28);
+        gm_8016B328();
+        HSD_GObjPLink_80390228(gobj);
+        return;
+    }
+    temp_r31 = &gmMainLib_804D3EE0->unk_530;
+    temp_r3 = gm_8016AE38();
+    if (temp_r31->xB_0) {
+        var_r0 = 0;
+    } else if (temp_r3->x24C8.x0_6 && gm_8016AEEC() == 0 &&
+               gm_8016AEFC() == 0x3B)
+    {
+        var_r0 = 1;
+    } else {
+        var_r0 = 0;
+    }
+    if (var_r0 != 0) {
+        gmMainLib_804D3EE0->unk_530.xB_1 = false;
+        lbAudioAx_80028B90();
+        gm_SetGameSpeed(1.0F);
+        gm_8016B33C(6);
+        gm_8016B364(0x148);
+        gm_8016B378(0x28);
+        gm_8016B328();
+        HSD_GObjPLink_80390228(gobj);
+    }
+}
 
 void gm_801BD658(HSD_GObj* gobj)
 {
diff --git a/src/melee/gr/granime.c b/src/melee/gr/granime.c
index 7732327e03..38171d0c7b 100644
--- a/src/melee/gr/granime.c
+++ b/src/melee/gr/granime.c
@@ -720,7 +720,71 @@ void grAnime_801C8098(Ground_GObj* gobj, int arg1, int arg2, int arg3,
     }
 }
 
-/// #grAnime_801C8138
+void grAnime_801C8138(HSD_GObj* gobj, enum_t arg1, bool arg2)
+{
+    HSD_JObj* jobj = gobj->hsd_obj;
+    UnkArchiveStruct* archive;
+    HSD_Joint* joint;
+    HSD_AnimJoint** ajp;
+    HSD_MatAnimJoint** mjp;
+    HSD_ShapeAnimJoint** sap;
+    HSD_AnimJoint* aj;
+    HSD_MatAnimJoint* ma;
+    HSD_ShapeAnimJoint* sa;
+    u8* flags;
+    s32 flag;
+
+    if (arg2 < 0) {
+        return;
+    }
+    archive = grDatFiles_801C6330(arg1);
+    if (archive == NULL) {
+        __assert("granime.c", 0x556, "archive");
+    }
+    joint = archive->unk4->unk8[arg1].unk0;
+    if (HSD_JObjGetChild(jobj) != NULL) {
+        jobj = HSD_JObjGetChild(jobj);
+    }
+    if (joint != NULL) {
+        lb_8000B804(jobj, joint);
+    }
+    grMaterial_801C8B68(jobj, 0);
+    HSD_JObjRemoveAnimAll(jobj);
+    sap = *(HSD_ShapeAnimJoint***) ((u8*) &archive->unk4->unk8[arg1] + 0xC);
+    ajp = *(HSD_AnimJoint***) ((u8*) &archive->unk4->unk8[arg1] + 0x4);
+    mjp = *(HSD_MatAnimJoint***) ((u8*) &archive->unk4->unk8[arg1] + 0x8);
+    if (sap != NULL) {
+        sa = sap[arg2];
+    } else {
+        sa = NULL;
+    }
+    if (mjp != NULL) {
+        ma = mjp[arg2];
+    } else {
+        ma = NULL;
+    }
+    if (ajp != NULL) {
+        aj = ajp[arg2];
+    } else {
+        aj = NULL;
+    }
+    HSD_JObjAddAnimAll(jobj, aj, ma, sa);
+    HSD_JObjReqAnimAll(jobj, 0.0f);
+    archive = grDatFiles_801C6330(arg1);
+    if (archive == NULL) {
+        __assert("granime.c", 0x148, "archive");
+    }
+    flags = (u8*) archive->unk4->unk8[arg1].x28;
+    if (flags != NULL) {
+        flag = flags[arg2];
+    } else {
+        flag = 0;
+    }
+    if ((s32) flag != 0) {
+        grAnime_801C752C(jobj, 1, 0x77A4, HSD_AObjSetFlags, 3, 0x20000000);
+    }
+    HSD_JObjAnimAll(jobj);
+}
 
 extern __jmp_buf grAnime_8049EE40;
 
diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index e664597f19..21f4288ab4 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -91,7 +91,17 @@ HSD_GObj* grHomeRun_8021C82C(int gobj_id)
     return gobj;
 }
 
-/// #grHomeRun_8021C914
+void grHomeRun_8021C914(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    HSD_JObj* jobj = (HSD_JObj*) gobj->hsd_obj;
+
+    grAnime_801C8138((HSD_GObj*) gobj, gp->map_id, 0);
+
+    HSD_JObjSetScaleX(jobj, grHr_804D6AE4 * HSD_JObjGetScaleX(jobj));
+    HSD_JObjSetScaleY(jobj, grHr_804D6AE4 * HSD_JObjGetScaleY(jobj));
+    HSD_JObjSetScaleZ(jobj, grHr_804D6AE4 * HSD_JObjGetScaleZ(jobj));
+}
 
 bool grHomeRun_8021CB10(Ground_GObj* arg)
 {
@@ -458,7 +468,47 @@ void fn_8021EB10(HSD_GObj* target_cam_gobj)
     }
 }
 
-/// #grHomeRun_8021EC58
+extern const float grHr_804DBC38;
+extern const float grHr_804DBC78;
+extern const float grHr_804DBC7C;
+extern const float grHr_804DBC80;
+extern const float grHr_804DBC84;
+extern const GXColor grHr_804DBC74;
+
+HSD_Text* grHomeRun_8021EC58(u32 score)
+{
+    HSD_Text* text;
+    GXColor color = grHr_804DBC74;
+    f32 scale;
+    f32 box_y;
+    f32 ofs_y;
+
+    text = HSD_SisLib_803A6754(1, 0);
+    text->pos_z = grHr_804DBC38;
+
+    scale = grHr_804D6AE4 * Ground_801C0498();
+    box_y = grHr_804DBC78 * scale;
+
+    scale = grHr_804D6AE4 * Ground_801C0498();
+    text->box_size_x = grHr_804DBC7C * scale;
+    text->box_size_y = box_y;
+    text->default_kerning = 0;
+
+    scale = grHr_804D6AE4 * Ground_801C0498();
+    ofs_y = grHr_804DBC80 * scale;
+
+    scale = grHr_804D6AE4 * Ground_801C0498();
+    text->x34.x = grHr_804DBC80 * scale;
+    text->x34.y = ofs_y;
+    text->text_color = color;
+    text->default_alignment = 1;
+    text->default_fitting = 1;
+    text->x4C = 1;
+
+    HSD_SisLib_803A6B98(text, grHr_804DBC38, grHr_804DBC84, "%d",
+                        (s32) (f32) (s32) score);
+    return text;
+}
 
 void grHomeRun_8021ED74(void)
 {
diff --git a/src/melee/gr/grhomerun.h b/src/melee/gr/grhomerun.h
index 0b554038b3..599a866bf1 100644
--- a/src/melee/gr/grhomerun.h
+++ b/src/melee/gr/grhomerun.h
@@ -68,7 +68,7 @@
 /* 21EA30 */ void grHomeRun_8021EA30(f32*);
 /* 21EAF8 */ UNK_RET grHomeRun_8021EAF8(UNK_PARAMS);
 /* 21EB10 */ void fn_8021EB10(HSD_GObj*);
-/* 21EC58 */ void* grHomeRun_8021EC58(u16);
+/* 21EC58 */ HSD_Text* grHomeRun_8021EC58(u32);
 /* 21ED74 */ void grHomeRun_8021ED74(void);
 /* 21EDD4 */ void grHomeRun_8021EDD4(int);
 /* 21EEB4 */ DynamicsDesc* grHomeRun_8021EEB4(enum_t);
diff --git a/src/melee/gr/gricemt.c b/src/melee/gr/gricemt.c
index f44c9f20c4..745b4266a7 100644
--- a/src/melee/gr/gricemt.c
+++ b/src/melee/gr/gricemt.c
@@ -7,6 +7,8 @@
 #include "baselib/random.h"
 #include "cm/camera.h"
 #include "ef/efsync.h"
+#include "ft/ftlib.h"
+#include "gm/gm_1601.h"
 
 #include "gr/forward.h"
 
@@ -460,13 +462,24 @@ void grIceMt_801F7728(Ground_GObj* gobj)
 
 void grIceMt_801F77AC(Ground_GObj* arg0) {}
 
+extern s16 grIm_803E4544[218];
+
 /// #fn_801F77B0
 void fn_801F77B0(HSD_GObj* arg0)
 {
     Ground* gp = GET_GROUND(arg0);
-    // 803E4544
-    // mpJointGetCb1();
-    mpJointSetCb1(4, gp, grIceMt_801FA7F0);
+    u32 i;
+    mpLib_Callback cb;
+    Ground* gp_out;
+
+    ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b0 = 0;
+
+    for (i = 0; i < 0xD9; i++) {
+        mpJointGetCb1(grIm_803E4544[i], &cb, &gp_out);
+        if (cb == NULL) {
+            mpJointSetCb1(grIm_803E4544[i], gp, grIceMt_801FA7F0);
+        }
+    }
 }
 
 /// #grIceMt_801F785C
@@ -526,6 +539,112 @@ bool grIceMt_801F796C(Ground_GObj* arg0)
 }
 
 /// #grIceMt_801F7A2C
+void grIceMt_801F7A2C(Ground_GObj* arg0)
+{
+    Ground* gp = arg0->user_data;
+    f32 sp30;
+    Vec3 sp24;
+    HSD_GObj* fighter;
+    HSD_GObj* mgobj;
+    HSD_JObj* jobj;
+    f32 dist;
+    f32 cur;
+    f32 step;
+    f32 ratio;
+    s32 var_r30;
+    s32 r;
+    PAD_STACK(8);
+
+    if (((UnkFlagStruct*) &gp->gv.icemt.xD8)->b0) {
+        return;
+    }
+    if (!((UnkFlagStruct*) &gp->gv.icemt.xD8)->b2) {
+        var_r30 = grIceMt_801FA364(&gp->gv.corneria.xC8, &sp30,
+                                   (HSD_GObjEvent) fn_801F9150, arg0);
+        if (((UnkFlagStruct*) &gp->gv.icemt.xD8)->b4) {
+            fighter = Ground_801C57A4();
+            if (fighter != NULL) {
+                ftLib_80086644(fighter, &sp24);
+                if (!((UnkFlagStruct*) &gp->gv.icemt.xD8)->b1 &&
+                    !ftLib_80086EC0(fighter) &&
+                    sp24.y > (f32) ((s16*) grIm_804D69F4)[0x98 / 2])
+                {
+                    ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b1 = 1;
+                    gp->gv.icemt.xDA = ((s16*) grIm_804D69F4)[0xA4 / 2];
+                }
+            }
+        }
+        if (((UnkFlagStruct*) &gp->gv.icemt.xD8)->b1) {
+            gp->gv.icemt.xE4 += grIm_804D69F4->xA0;
+            if (gp->gv.icemt.xE4 > grIm_804D69F4->x9C) {
+                gp->gv.icemt.xE4 = grIm_804D69F4->x9C;
+            }
+            gp->gv.icemt.xDA -= 1;
+            if (gp->gv.icemt.xDA == 0) {
+                ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b1 = 0;
+            }
+        } else {
+            gp->gv.icemt.xE4 -= grIm_804D69F4->xA0;
+            if (gp->gv.icemt.xE4 < grIm_804DB574) {
+                gp->gv.icemt.xE4 = grIm_804DB574;
+            }
+        }
+        sp30 += gp->gv.icemt.xE4;
+    } else {
+        mgobj = Ground_801C2BA4(gp->gv.icemt.xC6);
+        if (mgobj == NULL) {
+            __assert("gricemt.c", 0x4C3, "mgobj");
+        }
+        jobj = mgobj->hsd_obj;
+        if (jobj == NULL) {
+            __assert("gricemt.c", 0x4C5, "jobj");
+        }
+        dist = HSD_JObjGetTranslationY(jobj) -
+               (f32) ((s16*) grIm_804D69F4)[0xA8 / 2];
+        if (dist < grIm_804DB574) {
+            gp->gv.icemt.xD4 = grIm_804DB574;
+        } else {
+            ratio = gp->gv.icemt.xD4 / grIm_804D69F4->xA0;
+            if ((gp->gv.icemt.xD4 * ratio) -
+                    (ratio * (0.5f * grIm_804D69F4->xA0 * ratio)) >
+                dist)
+            {
+                gp->gv.icemt.xD4 -= grIm_804D69F4->xA0;
+                if (gp->gv.icemt.xD4 < grIm_804D69F4->xA0) {
+                    gp->gv.icemt.xD4 = grIm_804D69F4->xA0;
+                }
+            }
+        }
+        sp30 = gp->gv.icemt.xD4;
+    }
+    grIceMt_801F9ACC((Ground_GObj*) &gp->gv.icemt.xC4,
+                     grIceMt_801F96E0((HSD_GObj*) &gp->gv.icemt.xC4, -sp30),
+                     (HSD_GObjEvent) fn_801F91A8, arg0);
+    grIceMt_801F9668(sp30);
+    if (gp->gv.icemt.xC4 == -1) {
+        ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b2 = 1;
+    }
+    if (!((UnkFlagStruct*) &gp->gv.icemt.xD8)->b3) {
+        if (var_r30 != 0 &&
+            gp->gv.icemt.xDE == ((s16*) grIm_804D69F4)[0xA6 / 2])
+        {
+            ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b3 = 1;
+            Ground_801C5750();
+            gm_801674C4(0xE, 2, 2, 0, (bool (*)(s32)) fn_801FA4CC);
+            grZakoGenerator_801CAF08();
+        }
+    } else {
+        r = Ground_801C5764();
+        switch (r) {
+        case 2:
+            break;
+        case 1:
+            Ground_801C1D6C(0x10);
+            break;
+        }
+    }
+    ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b4 = 0;
+}
 
 void grIceMt_801F7D90(Ground_GObj* arg0) {}
 
@@ -997,9 +1116,19 @@ void grIceMt_801F91EC(HSD_GObj* param_1, s16* param_2, int param_3,
 /// #grIceMt_801F929C
 void grIceMt_801F929C(HSD_GObj* arg0, void* arg1)
 {
-    mpLib_80057BC0(2);
-    mpJointListAdd(2);
-    grAnime_801C83D0(arg0, 2, 7);
+    s16* p = arg1;
+
+    if (p[0] != 0) {
+        p[1] += 1;
+        if (p[1] == p[5]) {
+            mpLib_80057BC0(p[4]);
+        } else if (p[1] == p[6]) {
+            mpJointListAdd(p[4]);
+        }
+        if (grAnime_801C83D0(arg0, p[2], 7)) {
+            p[0] = 0;
+        }
+    }
 }
 
 /// #fn_801F9338
@@ -1130,7 +1259,60 @@ void grIceMt_801F9ACC(Ground_GObj* gobj, float y, HSD_GObjEvent ev,
 
 /// #grIceMt_801FA0BC
 
-/// #grIceMt_801FA364
+typedef struct grIceMt_FA364_State {
+    /* 0x0 */ s16 phase;
+    /* 0x2 */ s16 delay;
+    /* 0x4 */ s16 lerp_count;
+    /* 0x6 */ s16 burst_count;
+    /* 0x8 */ s16 idx;
+    /* 0xA */ s16 pad;
+    /* 0xC */ f32 cur;
+} grIceMt_FA364_State;
+
+bool grIceMt_801FA364(void* state_, f32* out, HSD_GObjEvent cb_,
+                      Ground_GObj* gobj)
+{
+    grIceMt_FA364_State* state = state_;
+    s32 (*cb)(HSD_GObj*, s32*) = (s32(*)(HSD_GObj*, s32*)) cb_;
+    bool ret = true;
+    f32 result;
+    s32 next_delay;
+    s16 tmp;
+
+    switch (state->phase) {
+    case 0:
+        tmp = state->delay;
+        state->delay = tmp - 1;
+        if (tmp < 0) {
+            state->idx = cb((HSD_GObj*) gobj, &next_delay);
+            state->delay = (s16) next_delay;
+            state->phase = 1;
+            state->lerp_count = ((s16*) grIm_804D69F4)[0x34 / 2];
+        }
+        break;
+    case 1:
+        state->lerp_count = state->lerp_count - 1;
+        tmp = state->lerp_count;
+        if (tmp != 0) {
+            state->cur +=
+                (((f32*) ((u8*) grIm_804D69F4 + 4))[state->idx] - state->cur) /
+                (f32) tmp;
+            ret = false;
+        } else {
+            state->phase = 0;
+            state->cur = ((f32*) ((u8*) grIm_804D69F4 + 4))[state->idx];
+        }
+        break;
+    }
+
+    result = state->cur;
+    if (state->burst_count != 0) {
+        result = grIm_804D69F4->x3C * Ground_801C0498();
+        state->burst_count--;
+    }
+    *out = result;
+    return ret;
+}
 
 int fn_801FA4CC(int num)
 { // https://decomp.me/scratch/pSJNA
diff --git a/src/melee/gr/gricemt.h b/src/melee/gr/gricemt.h
index 7e685b276e..682242bbb9 100644
--- a/src/melee/gr/gricemt.h
+++ b/src/melee/gr/gricemt.h
@@ -145,7 +145,8 @@ extern f32 grIm_804DB570;
 /* 1F9ACC */ void grIceMt_801F9ACC(HSD_GObj*, float, HSD_GObjEvent,
                                    Ground_GObj*);
 /* 1FA0BC */ UNK_RET grIceMt_801FA0BC(s16*);
-/* 1FA364 */ void grIceMt_801FA364(UNK_T, float*, HSD_GObjEvent, Ground_GObj*);
+/* 1FA364 */ bool grIceMt_801FA364(void* state, float* out, HSD_GObjEvent cb,
+                                   Ground_GObj* gobj);
 /* 1FA4CC */ int fn_801FA4CC(int num);
 /* 1FA500 */ int grIceMt_801FA500(HSD_GObj*);
 /* 1FA6D8 */ void grIceMt_801FA6D8(HSD_GObj*);
diff --git a/src/melee/mn/mndatadel.c b/src/melee/mn/mndatadel.c
index fd89c62e76..4c36fafab3 100644
--- a/src/melee/mn/mndatadel.c
+++ b/src/melee/mn/mndatadel.c
@@ -12,6 +12,7 @@
 #include "baselib/jobj.h"
 #include "lb/lb_00F9.h"
 #include "lb/lblanguage.h"
+#include "mn/inlines.h"
 #include "mn/mnmain.h"
 #include "mn/mnmainrule.h"
 #include "sc/types.h"
@@ -20,8 +21,35 @@
 
 /// #mnDataDel_8024EA6C
 
-/// #mnDataDel_8024EBC8                 /* unable to generate initializer:
-/// unknown type */
+void mnDataDel_8024EBC8(HSD_JObj* root, u8 unused, u8 a, u8 b)
+{
+    HSD_JObj* j2;
+    HSD_JObj* j1;
+    f32 frame;
+
+    lb_80011E24(root, &j1, WARN_JOINT_PANEL_TEXT_BOTTOM, -1);
+    HSD_JObjReqAnimAll(j1, a ? 1.0f : 0.0f);
+    mn_8022F3D8(j1, 0xff, (HSD_TypeMask) 0x400);
+    HSD_JObjAnimAll(j1);
+
+    lb_80011E24(root, &j2, WARN_JOINT_BACKGROUND, -1);
+    if (a) {
+        if (b) {
+            frame = mnDataDel_803EF870.x18.end_frame;
+        } else {
+            frame = mnDataDel_803EF870.x18.start_frame;
+        }
+    } else {
+        if (b) {
+            frame = mnDataDel_803EF870.x24.end_frame;
+        } else {
+            frame = mnDataDel_803EF870.x24.start_frame;
+        }
+    }
+    HSD_JObjReqAnimAll(j2, frame);
+    mn_8022F3D8(j2, 0xff, (HSD_TypeMask) 0x400);
+    HSD_JObjAnimAll(j2);
+}
 
 /// @brief animates the warning modal
 void fn_8024ECCC(HSD_GObj* arg0)
@@ -144,7 +172,42 @@ void mnDataDel_8024EEC0(void)
     }
 }
 
-/// #fn_8024F1D4
+void fn_8024F1D4(HSD_GObj* gobj)
+{
+    HSD_GObjProc* proc;
+    struct WarnCmnData* data;
+    u32 input;
+    PAD_STACK(16);
+
+    data = mnDataDel_804D6C68->user_data;
+    input = Menu_GetAllInputs();
+    if (input & 0x10) {
+        if (data->cursor_idx != 0) {
+            mnDataDel_8024EA6C();
+        } else {
+            sfxBack();
+        }
+        mn_804D6BC8.cooldown = 5;
+        data->visible = 0;
+        HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+        proc = HSD_GObj_SetupProc(gobj, fn_8024F840, 0);
+        proc->flags_3 = HSD_GObj_804D783C;
+    } else if (input & 0x20) {
+        sfxBack();
+        mn_804D6BC8.cooldown = 5;
+        data->visible = 0;
+        HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+        proc = HSD_GObj_SetupProc(gobj, fn_8024F840, 0);
+        proc->flags_3 = HSD_GObj_804D783C;
+    } else if ((input & 8) || (input & 4)) {
+        sfxMove();
+        if (data->cursor_idx != 0) {
+            data->cursor_idx = 0;
+        } else {
+            data->cursor_idx = 1;
+        }
+    }
+}
 
 /// #fn_8024F318
 
diff --git a/src/melee/mn/mndatadel.h b/src/melee/mn/mndatadel.h
index fbf68aa600..d614c976e5 100644
--- a/src/melee/mn/mndatadel.h
+++ b/src/melee/mn/mndatadel.h
@@ -7,7 +7,7 @@
 
 /* 24E940 */ void mnDataDel_8024E940(void);
 /* 24EA6C */ void mnDataDel_8024EA6C(void);
-/* 24EBC8 */ void mnDataDel_8024EBC8(UNK_PARAMS);
+/* 24EBC8 */ void mnDataDel_8024EBC8(HSD_JObj*, u8, u8, u8);
 /* 24ECCC */ void fn_8024ECCC(HSD_GObj*);
 /* 24EEC0 */ void mnDataDel_8024EEC0(void);
 /* 24F1D4 */ void fn_8024F1D4(HSD_GObj*);
diff --git a/src/melee/mn/mnmainrule.c b/src/melee/mn/mnmainrule.c
index 15812b3222..cb466bbfbc 100644
--- a/src/melee/mn/mnmainrule.c
+++ b/src/melee/mn/mnmainrule.c
@@ -9,6 +9,7 @@
 #include <sysdolphin/baselib/gobjproc.h>
 #include <sysdolphin/baselib/sislib.h>
 #include <melee/gm/gmmain_lib.h>
+#include <melee/lb/lb_00F9.h>
 #include <melee/lb/lbarchive.h>
 #include <melee/lb/lbaudio_ax.h>
 #include <melee/sc/types.h>
@@ -51,6 +52,7 @@ extern StaticModelDesc MenMainCursorSs_Top;
 
 extern u8 mn_804D4B96;
 extern u8 mn_803EC818[];
+extern s32 mn_804D6BD4;
 
 void mn_802307F8(struct mn_802307F8_t* data, s32 mode, s32 index)
 {
@@ -120,7 +122,57 @@ void mn_802308F0(HSD_GObj* gobj, int arg1, int arg2)
 
 /// #fn_802309F0
 
-/// #mn_80230D18
+struct mn_80230D18_t {
+    /* 0x000 */ u8 unk0;
+    /* 0x001 */ u8 unk1;
+    /* 0x002 */ u8 unk2;
+    /* 0x003 */ u8 unk3;
+    /* 0x004 */ u8 unk4;
+    /* 0x005 */ u8 unk5;
+    /* 0x006 */ u8 unk6;
+    /* 0x007 */ u8 pad7[2];
+    /* 0x009 */ u8 unk9;
+    /* 0x00A */ u8 unkA;
+    /* 0x00B */ u8 padB;
+    /* 0x00C */ HSD_JObj* slots[10];
+    /* 0x034 */ u8 pad34[0x130 - 0x34];
+    /* 0x130 */ s32 unk130;
+}; /* size = 0x134 */
+
+s32 mn_80230D18(struct mn_80230D18_t* arg0, HSD_JObj* arg1, s8 arg2)
+{
+    GameRules* rules;
+    s32 ret;
+    s32 i;
+
+    arg0->unk0 = mn_804A04F0.cur_menu;
+    arg0->unk1 = mn_804A04F0.hovered_selection;
+
+    rules = gmMainLib_8015CC34();
+    arg0->unk4 = rules->handicap;
+    if (mn_804D6BD4 == 0 && arg0->unk4 == 1) {
+        arg0->unk4 = 0;
+    }
+
+    arg0->unk5 = gmMainLib_8015CC34()->damage_ratio;
+    arg0->unk6 = gmMainLib_8015CC34()->unk_x7;
+    arg0->unk2 = gmMainLib_8015CC34()->mode;
+    arg0->unk9 = gmMainLib_8015CC34()->stock_count;
+    arg0->unk3 = gmMainLib_8015CC34()->time_limit;
+
+    if (gm_801A4310() == 0x1B && arg0->unk3 == 0) {
+        arg0->unk3 = 0x63;
+    }
+
+    arg0->unkA = arg2;
+    arg0->unk130 = 0;
+
+    ret = 0;
+    for (i = 0; i < 10; i++) {
+        ret = lb_80011E24(arg1, &arg0->slots[i], i, -1);
+    }
+    return ret;
+}
 
 /// #mn_80230E38
 
diff --git a/src/melee/mn/mnmainrule.h b/src/melee/mn/mnmainrule.h
index e046607da1..fcb6c02ef1 100644
--- a/src/melee/mn/mnmainrule.h
+++ b/src/melee/mn/mnmainrule.h
@@ -29,7 +29,8 @@ struct mn_80231634_t {
 /* 2307F8 */ void mn_802307F8(struct mn_802307F8_t*, s32, s32);
 /* 2308F0 */ void mn_802308F0(HSD_GObj*, int, int);
 /* 2309F0 */ UNK_RET fn_802309F0(UNK_PARAMS);
-/* 230D18 */ UNK_RET mn_80230D18(UNK_PARAMS);
+struct mn_80230D18_t;
+/* 230D18 */ s32 mn_80230D18(struct mn_80230D18_t*, HSD_JObj*, s8);
 /* 230E38 */ HSD_GObj* mn_80230E38(int);
 /* 231634 */ int mn_80231634(struct mn_80231634_t*);
 /* 23164C */ void mn_8023164C(void);
diff --git a/src/melee/vi/vi1101.c b/src/melee/vi/vi1101.c
index 8f71a68874..034d6c6277 100644
--- a/src/melee/vi/vi1101.c
+++ b/src/melee/vi/vi1101.c
@@ -2,13 +2,117 @@
 
 #include "vi.h"
 
+#include "cm/camera.h"
+#include "ef/efasync.h"
+#include "ef/eflib.h"
+#include "ft/ftdemo.h"
 #include "gm/gm_1601.h"
 #include "gm/gm_1A45.h"
+#include "gr/grlib.h"
+#include "gr/ground.h"
+#include "gr/stage.h"
+#include "it/item.h"
 #include "lb/lb_00F9.h"
 #include "lb/lbaudio_ax.h"
 #include "lb/lbshadow.h"
+#include "mp/mpcoll.h"
+#include "pl/player.h"
+#include "sc/types.h"
+#include "vi/types.h"
 
-/// #un_8031F294
+#include <baselib/aobj.h>
+#include <baselib/cobj.h>
+#include <baselib/gobj.h>
+#include <baselib/gobjgxlink.h>
+#include <baselib/gobjobject.h>
+#include <baselib/gobjproc.h>
+#include <baselib/jobj.h>
+
+static SceneDesc* un_804D6FC0;
+static SceneDesc* un_804D6FC4;
+static HSD_Archive* un_804D6FC8;
+static HSD_Archive* un_804D6FCC;
+
+static Vec3 un_80400200[3] = {
+    { 0.0f, 0.0f, 0.0f },
+    { 15.8512f, 67.0f, 0.0f },
+    { -15.7657f, 67.0f, 0.0f },
+};
+
+static struct {
+    HSD_GObj* p1;
+    HSD_GObj* p2;
+} un_804D6FD0;
+
+void un_8031F294(int char_index, int costume_index)
+{
+    HSD_JObj* jobj;
+    Vec3* pmtx;
+    char pad[40];
+
+    Camera_80028B9C(6);
+    lb_8000FCDC();
+    mpColl_80041C78();
+    Ground_801C0378(0x40);
+    Stage_802251E8(KINOKOROUTE, 0);
+    Item_80266FA8();
+    Item_80266FCC();
+    Stage_8022524C();
+    Stage_8022532C(KINOKOROUTE, 0);
+    ftDemo_ObjAllocInit();
+    Player_InitAllPlayers();
+
+    Player_80036E20(char_index, un_804D6FC8, 3);
+    Player_SetPlayerCharacter(0, char_index);
+    Player_SetCostumeId(0, costume_index);
+    Player_SetPlayerId(0, 0);
+    Player_SetSlottype(0, Gm_PKind_Demo);
+    Player_SetFacingDirection(0, 1.0f);
+    Player_80032768(0, un_80400200);
+    Player_80036F34(0, 8);
+
+    pmtx = grLib_801C9A10();
+
+    Player_80036E20(CKIND_MARIO, un_804D6FCC, 5);
+    Player_SetPlayerCharacter(1, CKIND_MARIO);
+    Player_SetCostumeId(1, 0);
+    Player_SetPlayerId(1, 0);
+    Player_SetSlottype(1, Gm_PKind_Demo);
+    Player_SetFacingDirection(1, 0.0f);
+    Player_80032768(1, un_80400200);
+    Player_SetFlagsBit5(1, 1);
+    Player_80036F34(1, 0xA);
+    un_804D6FD0.p1 = Player_GetEntity(1);
+    jobj = un_804D6FD0.p1->hsd_obj;
+    HSD_JObjReqAnimAll(jobj, 432.0f);
+    HSD_JObjAnimAll(jobj);
+    pmtx[1] = un_80400200[1];
+    HSD_JObjReqAnimAll(jobj, 0.0f);
+
+    if (gm_80164840(7) != 0) {
+        Player_80036E20(CKIND_LUIGI, un_804D6FCC, 5);
+        Player_SetPlayerCharacter(2, CKIND_LUIGI);
+        Player_SetCostumeId(2, 0);
+        Player_SetPlayerId(2, 0);
+        Player_SetSlottype(2, Gm_PKind_Demo);
+        Player_SetFacingDirection(2, 0.0f);
+        Player_80032768(2, un_80400200);
+        Player_SetFlagsBit5(2, 1);
+        Player_80036F34(2, 0xA);
+        un_804D6FD0.p2 = Player_GetEntity(2);
+        jobj = un_804D6FD0.p2->hsd_obj;
+        HSD_JObjReqAnimAll(jobj, 432.0f);
+        HSD_JObjAnimAll(jobj);
+        pmtx[2] = un_80400200[2];
+        HSD_JObjReqAnimAll(jobj, 0.0f);
+    }
+
+    lbAudioAx_80026F2C(0x1C);
+    lbAudioAx_8002702C(0xC,
+                       lbAudioAx_80026E84(char_index) | 0x0020000000000000ULL);
+    lbAudioAx_80027168();
+    lbAudioAx_80027648();
+}
 
 void fn_8031F548(HSD_GObj* gobj)
 {
diff --git a/src/melee/vi/vi1101.h b/src/melee/vi/vi1101.h
index 7b75daba0f..e002db4415 100644
--- a/src/melee/vi/vi1101.h
+++ b/src/melee/vi/vi1101.h
@@ -7,7 +7,7 @@
 
 /* 4D5B08 */ extern u8 un_804D5B08;
 
-/* 31F294 */ UNK_RET un_8031F294(UNK_PARAMS);
+/* 31F294 */ void un_8031F294(int char_index, int costume_index);
 /* 31F548 */ void fn_8031F548(HSD_GObj* gobj);
 /* 31F600 */ void fn_8031F600(HSD_GObj* gobj);
 /* 31F714 */ void un_8031F714_OnEnter(UNK_T);
diff --git a/src/sysdolphin/baselib/synth.c b/src/sysdolphin/baselib/synth.c
index 4fc32bed5b..952b4acd25 100644
--- a/src/sysdolphin/baselib/synth.c
+++ b/src/sysdolphin/baselib/synth.c
@@ -933,7 +933,67 @@ void HSD_Synth_8038AD74(u32 offset, uintptr_t src)
 
 /// #HSD_Synth_8038ADD0(void)
 
-/// #HSD_Synth_8038B120
+void HSD_Synth_8038B120(void)
+{
+    AXPBVE ve;
+    int i;
+    bool enabled;
+    struct HSD_SynthSFXNode* node;
+
+    PAD_STACK(0x10);
+
+    node = getNode(HSD_Synth_804D7760);
+    if (node != NULL) {
+        if (!(node->flags & 2)) {
+            ve.currentVolume =
+                (32767.0F *
+                 (node->user_vol[0].x8_float *
+                  (node->unk28 * (HSD_Synth_804D6030 *
+                                  HSD_Synth_804C28E0_1784[node->xB].x1784))));
+        } else {
+            ve.currentVolume = 0;
+        }
+        ve.currentDelta = 0;
+        node->x24 = ve.currentVolume;
+        for (i = 0; i < node->voice_count; i++) {
+            AXSetVoiceVe(node->voice[i], &ve);
+            if (node->flags & 4) {
+                *(u32*) &HSD_Synth_80407FD8.ratioHi = 0;
+            } else {
+                *(u32*) &HSD_Synth_80407FD8.ratioHi =
+                    (u32) (65536.0F *
+                           (node->x14 * node->x18[0] * node->x18[1]));
+            }
+            AXSetVoiceSrc(node->voice[i], &HSD_Synth_80407FD8);
+            AXSetVoiceCurrentAddr(
+                node->voice[i],
+                (HSD_Synth_804D7780 + ((u32) HSD_Synth_804D7768 << 16)) * 2 +
+                    i * lbl_804C4540[HSD_Synth_804D7768].x0 + 2);
+            AXSetVoiceEndAddr(
+                node->voice[i],
+                (HSD_Synth_804D7780 + ((u32) HSD_Synth_804D7768 << 16)) * 2 +
+                    i * lbl_804C4540[HSD_Synth_804D7768].x0 +
+                    lbl_804C4540[HSD_Synth_804D7768].x4);
+            AXSetVoiceLoopAddr(
+                node->voice[i],
+                (HSD_Synth_804D7780 + ((u32) HSD_Synth_804D7768 << 16)) * 2 +
+                    i * lbl_804C4540[HSD_Synth_804D7768].x0 + 2);
+            AXSetVoiceState(node->voice[i], 1);
+        }
+        node->flags &= ~8;
+        enabled = OSDisableInterrupts();
+        if (!node->volume_update_pending && !(node->flags & 8)) {
+            node->volume_update_pending = 1;
+            node->x20 = HSD_Synth_804D774C;
+            HSD_Synth_804D774C = node;
+        }
+        OSRestoreInterrupts(enabled);
+        HSD_SynthSFXUpdateMix(node, 0);
+        HSD_Synth_804D7778 = 0;
+    } else {
+        HSD_Synth_804D7778 = 0;
+    }
+}
 
 void HSD_SynthPStreamFirstHakoHeaderCallback(void)
 {
@@ -943,7 +1003,43 @@ void HSD_SynthPStreamFirstHakoHeaderCallback(void)
                       (HSD_DevComCallback) HSD_Synth_8038B120, 0);
 }
 
-/// #HSD_SynthPStreamHeaderCallback
+extern s32 HSD_Synth_804D7770;
+extern u32 HSD_Synth_804D7774;
+
+void HSD_SynthPStreamHeaderCallback(int arg0, int arg1, void* arg2,
+                                    int cancelflag)
+{
+    u32* entry = arg2;
+    struct HSD_SynthSFXNode* node;
+    int i;
+
+    node = getNode(HSD_Synth_804D7760);
+    if (node != NULL) {
+        node->voice_count = entry[3];
+        if (node->voice_count == 2) {
+            node->voice[1] = AXAcquireVoice(0x1D, dropcallback, 0);
+            if (node->voice[1] == NULL) {
+                __assert("synth.c", 0x5CF, "entry->voice[1]");
+            }
+        }
+        node->x14 = 0.00003125f * (f32) entry[2];
+        for (i = 0; i < node->voice_count; i++) {
+            *(u32*) &HSD_Synth_80407FD8.ratioHi = (u32) (65536.0f * node->x14);
+            AXSetVoiceAddr(node->voice[i], (AXPBADDR*) &entry[i * 14 + 4]);
+            AXSetVoiceAdpcm(node->voice[i], (AXPBADPCM*) &entry[i * 14 + 8]);
+        }
+        HSD_Synth_804D7774 = (HSD_Synth_804D7774 + 2) % 3;
+        HSD_Synth_804D776C = HSD_Synth_804D7770 = HSD_Synth_804D7768 =
+            HSD_Synth_804D7774;
+        HSD_DevComRequest(
+            HSD_Synth_804D7764, 0x80,
+            (uintptr_t) &lbl_804C4540[HSD_Synth_804D7768], 0x20, 0x21, 0,
+            (HSD_DevComCallback) HSD_SynthPStreamFirstHakoHeaderCallback,
+            NULL);
+    } else {
+        HSD_Synth_804D7778 = 0;
+    }
+}
 
 static inline void HSD_Synth_8038B5AC_inline(void)
 {
diff --git a/src/sysdolphin/baselib/synth.static.h b/src/sysdolphin/baselib/synth.static.h
index e6ac2f1978..ad816562dd 100644
--- a/src/sysdolphin/baselib/synth.static.h
+++ b/src/sysdolphin/baselib/synth.static.h
@@ -59,9 +59,12 @@ static struct {
 
 static int HSD_Synth_804C28E0_1844[HSD_SYNTHSFXGROUP_MAX];
 
+static u8 lbl_804C4524[0x1C];
+
 static struct {
     /* 00 */ s32 x0;
-    /* 04 */ char pad[0x1C];
+    /* 04 */ s32 x4;
+    /* 08 */ char pad[0x18];
 } lbl_804C4540[3];
 
 static int hsd_SynthSFXBankNum;
