diff --git a/configure.py b/configure.py
index 7ee04d1ac2..18f597af81 100755
--- a/configure.py
+++ b/configure.py
@@ -1206,7 +1206,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itgamewatchgreenhouse.c"),
             Object(Matching, "melee/it/items/itgamewatchmanhole.c"),
             Object(Matching, "melee/it/items/itgamewatchfire.c"),
-            Object(NonMatching, "melee/it/items/itgamewatchparachute.c"),
+            Object(Matching, "melee/it/items/itgamewatchparachute.c"),
             Object(Matching, "melee/it/items/itgamewatchturtle.c"),
             Object(Matching, "melee/it/items/itgamewatchbreath.c"),
             Object(Matching, "melee/it/items/itkirbygamewatchchefpan.c"),
diff --git a/src/melee/it/items/itgamewatchparachute.c b/src/melee/it/items/itgamewatchparachute.c
index b6adbe4b5f..e787a62881 100644
--- a/src/melee/it/items/itgamewatchparachute.c
+++ b/src/melee/it/items/itgamewatchparachute.c
@@ -10,6 +10,21 @@
 #include "it/it_2725.h"
 #include "it/item.h"
 
+ItemStateTable it_803F78F8[] = {
+    {
+        0,
+        itGamewatchparachute_UnkMotion1_Anim,
+        NULL,
+        NULL,
+    },
+    {
+        1,
+        itGamewatchparachute_UnkMotion1_Anim,
+        NULL,
+        NULL,
+    },
+};
+
 HSD_GObj* it_802C6C38(Item_GObj* parent, Vec3* pos, enum_t part,
                       float facing_dir)
 {
@@ -41,6 +56,14 @@ HSD_GObj* it_802C6C38(Item_GObj* parent, Vec3* pos, enum_t part,
     return NULL;
 }
 
+void itGameWatchParachute_Logic74_Destroyed(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (ip->owner != NULL) {
+        ftGw_AttackAirN_ItemParachuteSetFlag(ip->owner);
+    }
+}
+
 void it_802C6D6C(Item_GObj* item_gobj)
 {
     int pad[1];
@@ -54,11 +77,6 @@ void it_802C6D6C(Item_GObj* item_gobj)
     }
 }
 
-void itGameWatchParachute_Logic74_EvtUnk(Item_GObj* gobj, Item_GObj* ref_gobj)
-{
-    it_8026B894(gobj, ref_gobj);
-}
-
 void it_802C6DB8(Item_GObj* gobj)
 {
     it_8026B724(gobj);
@@ -69,14 +87,6 @@ void it_802C6DD8(Item_GObj* gobj)
     it_8026B73C(gobj);
 }
 
-void itGameWatchParachute_Logic74_Destroyed(Item_GObj* gobj)
-{
-    Item* ip = GET_ITEM(gobj);
-    if (ip->owner != NULL) {
-        ftGw_AttackAirN_ItemParachuteSetFlag(ip->owner);
-    }
-}
-
 void itGameWatchParachute_Logic74_PickedUp(Item_GObj* item_gobj)
 {
     Item* temp_r3;
@@ -124,3 +134,8 @@ bool itGamewatchparachute_UnkMotion1_Anim(Item_GObj* gobj)
 
     return false;
 }
+
+void itGameWatchParachute_Logic74_EvtUnk(Item_GObj* gobj, Item_GObj* ref_gobj)
+{
+    it_8026B894(gobj, ref_gobj);
+}
