diff --git a/config/GALE01/splits.txt b/config/GALE01/splits.txt
index e7113cd3a3..6c73082b8e 100644
--- a/config/GALE01/splits.txt
+++ b/config/GALE01/splits.txt
@@ -3309,8 +3309,8 @@ melee/it/items/itseakneedlethrown.c:
 melee/it/items/itseakneedleheld.c:
 	.text       start:0x802B18B0 end:0x802B1C60
 	.data       start:0x803F70A8 end:0x803F70B8
-	.sdata      start:0x804D5468 end:0x804D5478
-	.sdata2     start:0x804DCF88 end:0x804DCF90
+	.sdata      start:0x804D5468 end:0x804D5475
+	.sdata2     start:0x804DCF88 end:0x804DCF8C
 
 melee/it/items/itseakvanish.c:
 	.text       start:0x802B1C60 end:0x802B1DEC
diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 09125637b1..44d6797c1d 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -27061,8 +27061,8 @@ it_804D5448 = .sdata:0x804D5448; // type:object size:0x7 scope:global data:strin
 it_804D5450 = .sdata:0x804D5450; // type:object size:0x5 scope:global data:string
 it_804D5458 = .sdata:0x804D5458; // type:object size:0x7 scope:global data:string
 it_804D5460 = .sdata:0x804D5460; // type:object size:0x5 scope:global data:string
-it_804D5468 = .sdata:0x804D5468; // type:object size:0x7 scope:global data:string
-it_804D5470 = .sdata:0x804D5470; // type:object size:0x5 scope:global data:string
+@224 = .sdata:0x804D5468; // type:object size:0x7 scope:local data:string
+@225 = .sdata:0x804D5470; // type:object size:0x5 scope:local data:string
 it_804D5478 = .sdata:0x804D5478; // type:object size:0x7 scope:global data:string
 it_804D5480 = .sdata:0x804D5480; // type:object size:0x5 scope:global data:string
 it_804D5488 = .sdata:0x804D5488; // type:object size:0x7 scope:global data:string
@@ -33101,7 +33101,7 @@ it_804DCF70 = .sdata2:0x804DCF70; // type:object size:0x4 scope:global align:4 d
 it_804DCF74 = .sdata2:0x804DCF74; // type:object size:0x4 scope:global align:4 data:float
 it_804DCF78 = .sdata2:0x804DCF78; // type:object size:0x8 scope:global align:8 data:double
 it_804DCF80 = .sdata2:0x804DCF80; // type:object size:0x8 scope:global align:8 data:double
-it_804DCF88 = .sdata2:0x804DCF88; // type:object size:0x4 scope:global align:4 data:float
+@186 = .sdata2:0x804DCF88; // type:object size:0x4 scope:local align:4 data:float
 @202 = .sdata2:0x804DCF90; // type:object size:0x4 scope:local align:4 data:float
 @206 = .sdata2:0x804DCF94; // type:object size:0x4 scope:local align:4 data:float
 it_804DCF98 = .sdata2:0x804DCF98; // type:object size:0x4 scope:global align:4 data:float
diff --git a/configure.py b/configure.py
index 3aa19a051f..a93eaf1d31 100755
--- a/configure.py
+++ b/configure.py
@@ -1170,7 +1170,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itlinkbow.c"),
             Object(NonMatching, "melee/it/items/itnesspkflushexplode.c"),
             Object(NonMatching, "melee/it/items/itseakneedlethrown.c"),
-            Object(NonMatching, "melee/it/items/itseakneedleheld.c"),
+            Object(Matching, "melee/it/items/itseakneedleheld.c"),
             Object(Matching, "melee/it/items/itseakvanish.c"),
             Object(NonMatching, "melee/it/items/itpikachuthunder.c"),
             Object(Matching, "melee/it/items/itmariocape.c"),
diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index 2e58ba1bff..1cc4d2a19d 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -599,6 +599,10 @@ typedef struct itSamusMissile_ItemVars {
     /* +10 +DE4 */ HSD_GObj* x10;
 } itSamusMissile_ItemVars;
 
+typedef struct itSeakNeedleHeld_ItemVars {
+    /* +0 ip+DD4 */ Fighter_GObj* owner;
+} itSeakNeedleHeld_ItemVars;
+
 typedef struct itSeakNeedleThrown_ItemVars {
     /*  +0 ip+DD4 */ char pad_0[0x4];
     /*  +4 ip+DD8 */ f32 xDD8;
diff --git a/src/melee/it/items/itseakneedleheld.c b/src/melee/it/items/itseakneedleheld.c
index 86c8c10fcc..3cda5dd4e3 100644
--- a/src/melee/it/items/itseakneedleheld.c
+++ b/src/melee/it/items/itseakneedleheld.c
@@ -1,18 +1,146 @@
 #include "itseakneedleheld.h"
 
+#include "ft/chara/ftKirby/ftKb_Init.h"
+#include "ft/chara/ftSeak/ftSk_SpecialS.h"
+
+#include "ft/forward.h"
+
+#include "ft/ftlib.h"
+#include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
 
-/// #it_802B18B0
+#include <baselib/jobj.h>
+
+ItemStateTable it_803F70A8[] = {
+    {
+        -1,
+        itSeakneedleheld_UnkMotion0_Anim,
+        itSeakneedleheld_UnkMotion0_Phys,
+        itSeakneedleheld_UnkMotion0_Coll,
+    },
+};
+
+void it_802B18B0(Item_GObj* gobj)
+{
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* child = HSD_JObjGetChild(jobj);
+    HSD_JObj* needle = HSD_JObjGetChild(child);
+    Fighter_GObj* owner = ip->xDD4_itemVar.seakneedleheld.owner;
+    int count;
+    int i;
+
+    if (owner == NULL) {
+        return;
+    }
+    if (ip->owner != owner) {
+        return;
+    }
+
+    switch (ip->kind) {
+    case It_Kind_Seak_NeedleHeld:
+        count = ftSk_SpecialS_80111FA0(owner);
+        break;
+    case It_Kind_Kirby_SeakNeedleHeld:
+        count = ftKb_SpecialNSk_80106020(owner);
+        break;
+    default:
+        break;
+    }
+
+    for (i = 0; i < 6; i++) {
+        if (i < count) {
+            HSD_JObjClearFlagsAll(needle, JOBJ_HIDDEN);
+        } else {
+            HSD_JObjSetFlagsAll(needle, JOBJ_HIDDEN);
+        }
+        needle = HSD_JObjGetNext(needle);
+    }
+}
+
+Item_GObj* it_802B19AC(Fighter_GObj* parent_gobj, Vec3* cur_pos,
+                       Fighter_Part part, ItemKind kind, float facing_dir)
+{
+    SpawnItem spawn;
 
-/// #it_802B19AC
+    spawn.kind = kind;
+    spawn.prev_pos = *cur_pos;
+    spawn.prev_pos.z = 0.0f;
+    spawn.pos = spawn.prev_pos;
+    spawn.facing_dir = facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.x = spawn.vel.y = spawn.vel.z = 0.0f;
+    spawn.x0_parent_gobj = parent_gobj;
+    spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+    spawn.x44_flag.b0 = true;
+    spawn.x40 = 0;
+
+    {
+        Item_GObj* gobj = Item_80268B18(&spawn);
+        if (gobj != NULL) {
+            Item* ip = GET_ITEM(gobj);
+            PAD_STACK(2 * 4);
+            ip->xDAC_itcmd_var0 = ip->xDB0_itcmd_var1 = ip->xDB4_itcmd_var2 =
+                ip->xDB8_itcmd_var3 = 0;
+            ip->xDCC_flag.b3 = false;
+            ip->xDD4_itemVar.seakneedleheld.owner = parent_gobj;
+            Item_8026AB54(gobj, parent_gobj, part);
+        }
+        return gobj;
+    }
+}
 
 void itSeakNeedleHeld_Logic110_PickedUp(Item_GObj* gobj)
 {
     Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
 }
 
-/// #itSeakneedleheld_UnkMotion0_Anim
+bool itSeakneedleheld_UnkMotion0_Anim(Item_GObj* gobj)
+{
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* needle = HSD_JObjGetChild(jobj);
+    Vec3 scale;
+    PAD_STACK(3 * 4);
+
+    if (ip->xDD4_itemVar.seakneedleheld.owner == NULL) {
+        goto ret_true;
+    }
+    if (ip->owner != ip->xDD4_itemVar.seakneedleheld.owner) {
+        goto ret_false;
+    }
+
+    switch (ip->kind) {
+    case It_Kind_Seak_NeedleHeld:
+        if (ftSk_SpecialS_80111F70(ip->xDD4_itemVar.seakneedleheld.owner) == 1)
+        {
+            return true;
+        }
+        break;
+    case It_Kind_Kirby_SeakNeedleHeld:
+        if (ftKb_SpecialNSk_80105FF0(ip->xDD4_itemVar.seakneedleheld.owner) ==
+            1)
+        {
+            return true;
+        }
+        break;
+    default:
+        break;
+    }
+
+    it_802B18B0(gobj);
+
+    scale.x = scale.y = scale.z =
+        ftLib_800869D4(ip->xDD4_itemVar.seakneedleheld.owner);
+    HSD_JObjSetScale(needle, &scale);
+    goto ret_false;
+
+ret_true:
+    return true;
+ret_false:
+    return false;
+}
 
 void itSeakneedleheld_UnkMotion0_Phys(Item_GObj* gobj) {}
 
diff --git a/src/melee/it/items/itseakneedleheld.h b/src/melee/it/items/itseakneedleheld.h
index 4e182658d5..dda3eb0b6f 100644
--- a/src/melee/it/items/itseakneedleheld.h
+++ b/src/melee/it/items/itseakneedleheld.h
@@ -9,8 +9,9 @@
 
 #include <dolphin/mtx.h>
 
-/* 2B18B0 */ UNK_RET it_802B18B0(UNK_PARAMS);
-/* 2B19AC */ Item_GObj* it_802B19AC(HSD_GObj*, Vec3* cur_pos, u32, u32,
+/* 2B18B0 */ void it_802B18B0(Item_GObj* gobj);
+/* 2B19AC */ Item_GObj* it_802B19AC(Fighter_GObj* parent_gobj, Vec3* cur_pos,
+                                    Fighter_Part part, ItemKind kind,
                                     float facing_dir);
 /* 2B1AAC */ void itSeakNeedleHeld_Logic110_PickedUp(Item_GObj*);
 /* 2B1AD4 */ bool itSeakneedleheld_UnkMotion0_Anim(Item_GObj* gobj);
diff --git a/src/melee/it/types.h b/src/melee/it/types.h
index 892a10046c..588cbd038f 100644
--- a/src/melee/it/types.h
+++ b/src/melee/it/types.h
@@ -617,6 +617,7 @@ struct Item {
         itSamusGrapple_ItemVars samusgrapple;
         itSamusMissile_ItemVars samusmissile;
         itSeakChain_ItemVars seakchain;
+        itSeakNeedleHeld_ItemVars seakneedleheld;
         itSeakNeedleThrown_ItemVars seakneedlethrown;
         itSonans_ItemVars sonans;
         itStar_ItemVars star;
