diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 12d78435b5..65b7d484c1 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -26998,8 +26998,8 @@ it_804D5250 = .sdata:0x804D5250; // type:object size:0x7 scope:global data:strin
 it_804D5258 = .sdata:0x804D5258; // type:object size:0x5 scope:global data:string
 @247 = .sdata:0x804D5260; // type:object size:0x7 scope:local data:string
 @248 = .sdata:0x804D5268; // type:object size:0x5 scope:local data:string
-it_804D5270 = .sdata:0x804D5270; // type:object size:0x7 scope:global data:string
-it_804D5278 = .sdata:0x804D5278; // type:object size:0x5 scope:global data:string
+@265 = .sdata:0x804D5270; // type:object size:0x7 scope:local data:string
+@266 = .sdata:0x804D5278; // type:object size:0x5 scope:local data:string
 it_804D5280 = .sdata:0x804D5280; // type:object size:0x7 scope:global data:string
 it_804D5288 = .sdata:0x804D5288; // type:object size:0x5 scope:global data:string
 it_804D5290 = .sdata:0x804D5290; // type:object size:0x7 scope:global data:string
@@ -32862,10 +32862,10 @@ it_804DCA20 = .sdata2:0x804DCA20; // type:object size:0x8 scope:global align:8 d
 @204 = .sdata2:0x804DCA28; // type:object size:0x4 scope:local align:4 data:float
 @249 = .sdata2:0x804DCA2C; // type:object size:0x4 scope:local align:4 data:float
 @296 = .sdata2:0x804DCA30; // type:object size:0x4 scope:local align:4 data:float
-it_804DCA38 = .sdata2:0x804DCA38; // type:object size:0x4 scope:global align:4 data:float
-it_804DCA3C = .sdata2:0x804DCA3C; // type:object size:0x4 scope:global align:4 data:float
-it_804DCA40 = .sdata2:0x804DCA40; // type:object size:0x8 scope:global align:8 data:double
-it_804DCA48 = .sdata2:0x804DCA48; // type:object size:0x8 scope:global align:8 data:double
+@168 = .sdata2:0x804DCA38; // type:object size:0x4 scope:local align:4 data:float
+@169 = .sdata2:0x804DCA3C; // type:object size:0x4 scope:local align:4 data:float
+@207 = .sdata2:0x804DCA40; // type:object size:0x8 scope:local align:8 data:double
+@329 = .sdata2:0x804DCA48; // type:object size:0x8 scope:local align:8 data:double
 it_804DCA50 = .sdata2:0x804DCA50; // type:object size:0x4 scope:global align:4 data:float
 it_804DCA54 = .sdata2:0x804DCA54; // type:object size:0x4 scope:global align:4 data:float
 it_804DCA58 = .sdata2:0x804DCA58; // type:object size:0x8 scope:global align:8 data:double
diff --git a/configure.py b/configure.py
index f7b89e40e7..1d16d91e33 100755
--- a/configure.py
+++ b/configure.py
@@ -1118,7 +1118,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itegg.c"),
             Object(NonMatching, "melee/it/items/itkusudama.c"),
             Object(Matching, "melee/it/items/itparasol.c"),
-            Object(NonMatching, "melee/it/items/itgshell.c"),
+            Object(Matching, "melee/it/items/itgshell.c"),
             Object(NonMatching, "melee/it/items/itrshell.c"),
             Object(Matching, "melee/it/items/itlgun.c"),
             Object(NonMatching, "melee/it/items/itfreeze.c"),
diff --git a/src/melee/it/items/itgshell.c b/src/melee/it/items/itgshell.c
index 7f9444a8ff..9a21bc1964 100644
--- a/src/melee/it/items/itgshell.c
+++ b/src/melee/it/items/itgshell.c
@@ -65,13 +65,16 @@ void it_8028B8D8(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     itGShell_Attrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    f32 temp;
     Vec v;
     HSD_JObj* jobj;
     PAD_STACK(4);
+
     if (ip->xDD4_itemVar.gshell.xDDC <= 0.0f) {
         jobj = GET_JOBJ(gobj);
         v = attrs->x34;
-        v.x *= -ip->facing_dir;
+        temp = -ip->facing_dir;
+        v.x *= temp;
         efAsync_Spawn(gobj, &GET_ITEM(gobj)->xBC0, 2, 1029, jobj, &v);
         ip->xDD4_itemVar.gshell.xDDC = attrs->x30;
     } else {
@@ -163,8 +166,9 @@ void it_8028BC2C(Item_GObj* gobj)
         ip->x40_vel.x = -ip->x40_vel.x * attrs->xC * HSD_Randf();
         ip->x40_vel.y = attrs->x10;
         if (ip->xDD4_itemVar.gshell.xDEC_b0) {
-            ip->xDD4_itemVar.gshell.xDE8 =
-                HSD_Randi(it_804D6D28->x48_byte & 0xF);
+            u8 rand_max = it_804D6D28->x48_byte;
+            rand_max &= 0xF;
+            ip->xDD4_itemVar.gshell.xDE8 = HSD_Randi(rand_max);
         } else {
             ip->xDD4_itemVar.gshell.xDEC_b0 = 1;
         }
@@ -414,7 +418,6 @@ void it_8028C3A8(Item_GObj* gobj)
 bool itGshell_UnkMotion6_Anim(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
-    PAD_STACK(8);
     if (ip->xDD4_itemVar.gshell.xDD8 <= 0.0f) {
         if (!ip->xDCD_flag.b5) {
             it_80275444(gobj);
@@ -595,18 +598,18 @@ bool itGShell_Logic14_Reflected(Item_GObj* gobj)
     return false;
 }
 
-bool itGShell_Logic14_Clanked(Item_GObj* gobj)
+static inline void shellHit(HSD_GObj* gobj)
 {
-    itGShell_Attrs* attrs;
     Item* ip = GET_ITEM(gobj);
-    attrs = ip->xC4_article_data->x4_specialAttributes;
-    PAD_STACK(8);
+    itGShell_Attrs* attrs = ip->xC4_article_data->x4_specialAttributes;
     it_802756D0(gobj);
     it_80275444(gobj);
     ip->x40_vel.x = -ip->x40_vel.x * attrs->xC * HSD_Randf();
     ip->x40_vel.y = attrs->x10;
     if (ip->xDD4_itemVar.gshell.xDEC_b0) {
-        ip->xDD4_itemVar.gshell.xDE8 = HSD_Randi(it_804D6D28->x48_byte & 0xF);
+        u8 rand_max = it_804D6D28->x48_byte;
+        rand_max &= 0xF;
+        ip->xDD4_itemVar.gshell.xDE8 = HSD_Randi(rand_max);
     } else {
         ip->xDD4_itemVar.gshell.xDEC_b0 = 1;
     }
@@ -615,33 +618,22 @@ bool itGShell_Logic14_Clanked(Item_GObj* gobj)
     it_80274C88(gobj);
     ip->xDD4_itemVar.gshell.xDEC_b3 = 1;
     ip->xDD4_itemVar.gshell.xDE4 = attrs->x2C;
+}
+
+bool itGShell_Logic14_Clanked(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    shellHit(gobj);
     return false;
 }
 
 bool itGShell_Logic14_HitShield(Item_GObj* gobj)
 {
-    itGShell_Attrs* attrs;
     Item* ip = GET_ITEM(gobj);
-    PAD_STACK(8);
     if (ip->msid == 3 || ip->msid == 4) {
         itColl_BounceOffVictim(gobj);
     } else if (ip->msid - 5u <= 3) {
-        attrs = ip->xC4_article_data->x4_specialAttributes;
-        it_802756D0(gobj);
-        it_80275444(gobj);
-        ip->x40_vel.x = -ip->x40_vel.x * attrs->xC * HSD_Randf();
-        ip->x40_vel.y = attrs->x10;
-        if (ip->xDD4_itemVar.gshell.xDEC_b0) {
-            ip->xDD4_itemVar.gshell.xDE8 =
-                HSD_Randi(it_804D6D28->x48_byte & 0xF);
-        } else {
-            ip->xDD4_itemVar.gshell.xDEC_b0 = 1;
-        }
-        it_802762BC(ip);
-        Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
-        it_80274C88(gobj);
-        ip->xDD4_itemVar.gshell.xDEC_b3 = 1;
-        ip->xDD4_itemVar.gshell.xDE4 = attrs->x2C;
+        shellHit(gobj);
     }
     return false;
 }
