diff --git a/src/melee/lb/lbmemory.c b/src/melee/lb/lbmemory.c
index 946440e451..9c098ec153 100644
--- a/src/melee/lb/lbmemory.c
+++ b/src/melee/lb/lbmemory.c
@@ -199,6 +199,47 @@ void lbMemory_800150F0(Handle* h, void* arg1)
     __assert("lbmemory.c", 283, "0");
 }
 
+static s8 lbl_803BA2C0[0xB] = "lbmemory.c";
+static s8 lbMemory_804D3794[8] = "p->size";
+
+struct LBMgr {
+    OSAlarm alarm;                  // 0x00
+    u8* src;                        // 0x28
+    u8* dst;                        // 0x2C
+    u32 size;                       // 0x30
+    u32 offset;                     // 0x34
+    u32 cb_arg;                     // 0x38
+    void (*cb)(u32, u32, u32, u32); // 0x3C
+};
+
+void fn_80015184(OSAlarm* alarm, OSContext* context)
+{
+    struct LBMgr* temp_r3;
+    u32 temp_r3_2;
+    u32 temp_r6;
+    u32 var_r30;
+
+    temp_r3 = (struct LBMgr*) ((u8*) &g_alloc + 0x6A0);
+    if ((u32) temp_r3->size == 0U) {
+        __assert("lbmemory.c", 0x127U, "p->size");
+    }
+    temp_r6 = temp_r3->offset;
+    temp_r3_2 = temp_r3->size - temp_r6;
+    var_r30 = temp_r3_2;
+    if (temp_r3_2 > 0x19000U) {
+        var_r30 = 0x19000;
+    }
+    memcpy(temp_r3->dst + temp_r6, temp_r3->src + temp_r6, var_r30);
+    temp_r3->offset = temp_r3->offset + var_r30;
+    if (temp_r3->offset == temp_r3->size) {
+        temp_r3->size = 0U;
+        temp_r3->cb(0, temp_r3->cb_arg, 0, 0);
+        return;
+    }
+    OSCreateAlarm(&temp_r3->alarm);
+    OSSetAlarm(&temp_r3->alarm, 0x10624DD3, NULL);
+}
+
 u32 lbMemory_8001529C(Handle* h, void* arg1, u32 arg2)
 {
     void* lo;
diff --git a/src/melee/lb/lbmemory.h b/src/melee/lb/lbmemory.h
index 8635452d63..714c410444 100644
--- a/src/melee/lb/lbmemory.h
+++ b/src/melee/lb/lbmemory.h
@@ -5,6 +5,8 @@
 
 #include <baselib/forward.h>
 
+#include <dolphin/os/OSAlarm.h>
+
 typedef struct _Handle {
     struct _Handle* x0_next;
     void* x4_lo;
@@ -18,6 +20,7 @@ typedef struct _Handle {
 /* 014F7C */ u32 lbMemory_80014F7C(Handle*);
 /* 014FC8 */ Handle* lbMemory_80014FC8(Handle*, u32);
 /* 0150F0 */ void lbMemory_800150F0(Handle*, void*);
+/* 015184 */ void fn_80015184(OSAlarm* alarm, OSContext* context);
 /* 01529C */ u32 lbMemory_8001529C(Handle*, void*, u32);
 /* 0154BC */ void lbMemory_800154BC(uintptr_t*, uintptr_t*);
 /* 0154D4 */ Handle* lbMemory_800154D4(void* lo, void* hi);
diff --git a/src/sysdolphin/baselib/hsd_3B34.h b/src/sysdolphin/baselib/hsd_3B34.h
index cc47601b01..b4464a4fd5 100644
--- a/src/sysdolphin/baselib/hsd_3B34.h
+++ b/src/sysdolphin/baselib/hsd_3B34.h
@@ -3,16 +3,16 @@
 
 #include <platform.h>
 
-void hsd_803B3408(u8*, s32, s32, s32, s32);
-void hsd_803B3CD8(s32);
-void hsd_803B46D4(void);
-void hsd_803B4A2C(void);
-void hsd_803B4D64(u32, u32);
-s32 hsd_803B51C8(s32, s32, s32, char*, s32);
-void hsd_803B5C2C(s32);
-s32 hsd_803B5C4C(s32);
-s32 hsd_803B5D70(s32, s32);
-void hsd_803B5EA0(s32);
-s32 hsd_803B6BE4(char*, s32, void*);
+/* 3B3408 */ void hsd_803B3408(u8*, s32, s32, s32, s32);
+/* 3B3CD8 */ void hsd_803B3CD8(s32);
+/* 3B46D4 */ void hsd_803B46D4(void);
+/* 3B4A2C */ void hsd_803B4A2C(void);
+/* 3B4D64 */ void hsd_803B4D64(u32, u32);
+/* 3B51C8 */ s32 hsd_803B51C8(s32, s32, s32, char*, s32);
+/* 3B5C2C */ void hsd_803B5C2C(s32);
+/* 3B5C4C */ s32 hsd_803B5C4C(s32);
+/* 3B5D70 */ s32 hsd_803B5D70(s32, s32);
+/* 3B5EA0 */ void hsd_803B5EA0(s32);
+/* 3B6BE4 */ s32 hsd_803B6BE4(char*, s32, void*);
 
 #endif
