diff --git a/src/melee/ft/chara/ftCommon/ftCo_Damage.c b/src/melee/ft/chara/ftCommon/ftCo_Damage.c
index 56c54a7f2c..5bdeb632c9 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Damage.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_Damage.c
@@ -566,49 +566,27 @@ bool ftCo_Damage_CheckAirMotion(Fighter* fp)
     }
 }
 
-#define SOLUTION 1
 void ftCo_Damage_OnEveryHitlag(Fighter_GObj* gobj)
 {
-    Fighter* fp = gobj->user_data;
-    if (fp->x221A_b2) {
-        float lstick_x = fp->input.lstick.x;
-        float lstick_y = fp->input.lstick.y;
-        float cd_x4B0 = p_ftCommonData->x4B0;
-        if (lstick_x * lstick_x + lstick_y * lstick_y == cd_x4B0 * cd_x4B0) {
-            s32 cd_x4B4 = p_ftCommonData->x4B4;
-#if SOLUTION == 0
-            if (((M2C_FIELD(fp, u8*, 0x670) < cd_x4B4) ||
-                 M2C_FIELD(fp, u8*, 0x671) < cd_x4B4))
-#else
-            if (((fp->x670_timer_lstick_tilt_x < cd_x4B4) ||
-                 fp->x671_timer_lstick_tilt_y < cd_x4B4))
-#endif
-            {
-                float cd_x4B8 = p_ftCommonData->x4B8;
-                float scaled_lstick_x = lstick_x * cd_x4B8;
-                float scaled_lstick_y = lstick_y * cd_x4B8;
-                fp->cur_pos.x += scaled_lstick_x;
-                fp->cur_pos.y += scaled_lstick_y;
-#if SOLUTION == 0
-                M2C_FIELD(fp, u8*, 0x670) = 0xFE;
-                M2C_FIELD(fp, u8*, 0x671) = 0xFE;
-#else
-                fp->x670_timer_lstick_tilt_x = 0xFE;
-                fp->x671_timer_lstick_tilt_y = 0xFE;
-#endif
-#if SOLUTION == 0
-                pl_800401F0(fp->player_id,
-                            (M2C_FIELD(fp, u8*, 0x221F) >> 3) & 1,
-                            scaled_lstick_x, scaled_lstick_y);
-#else
-                pl_800401F0(fp->player_id, fp->x221F_b4, scaled_lstick_x,
-                            scaled_lstick_y);
-#endif
-            }
-        }
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->allow_sdi &&
+        VEC2_SQ_LEN(fp->input.lstick) >=
+            SQ(p_ftCommonData->sdi_min_stick_mag) &&
+        (fp->x670_timer_lstick_tilt_x < p_ftCommonData->sdi_stick_window ||
+         fp->x671_timer_lstick_tilt_y < p_ftCommonData->sdi_stick_window))
+    {
+        float scaled_lstick_x =
+            fp->input.lstick.x * p_ftCommonData->sdi_pos_scale;
+        float scaled_lstick_y =
+            fp->input.lstick.y * p_ftCommonData->sdi_pos_scale;
+        fp->cur_pos.x += scaled_lstick_x;
+        fp->cur_pos.y += scaled_lstick_y;
+        fp->x670_timer_lstick_tilt_x = 254;
+        fp->x671_timer_lstick_tilt_y = 254;
+        pl_800401F0(fp->player_id, fp->x221F_b4, scaled_lstick_x,
+                    scaled_lstick_y);
     }
 }
-#undef SOLUTION
 
 void ftCo_8008E5A4(Fighter* fp)
 {
@@ -660,7 +638,7 @@ void ftCo_Damage_OnExitHitlag(Fighter_GObj* gobj)
 {
     Fighter* fp = gobj->user_data;
     if (isPointInCircle(SQ(fp->input.lstick.x), SQ(fp->input.lstick.y),
-                        SQ(p_ftCommonData->x4B0)) ||
+                        SQ(p_ftCommonData->sdi_min_stick_mag)) ||
         ftCo_800DF608(fp))
     {
         float x, y;
@@ -714,7 +692,7 @@ void ftCo_8008E908(Fighter_GObj* gobj, float facing_dir)
 bool ftCo_8008E984(Fighter* fp)
 {
     if (fp->dmg.kb_applied == 0 ||
-        (fp->x221A_b2 && fp->x221A_b3 &&
+        (fp->allow_sdi && fp->x221A_b3 &&
          fp->dmg.kb_applied < fp->dmg.x18A8 + p_ftCommonData->x140))
     {
         return true;
@@ -757,7 +735,7 @@ static bool inlineB0(Fighter_GObj* gobj)
     if (kb_applied != 0) {
         return true;
     }
-    if (fp->x221A_b2 && fp->x221A_b3 &&
+    if (fp->allow_sdi && fp->x221A_b3 &&
         fp->dmg.kb_applied < fp->dmg.x18A8 + p_ftCommonData->x140)
     {
         return true;
@@ -854,7 +832,7 @@ void ftCo_8008EC90(Fighter_GObj* gobj)
                     other_fp->dmg.x195c_hitlag_frames = ftCommon_CalcHitlag(
                         fp->dmg.x183C_applied, other_fp->motion_id,
                         other_fp->x1960_vibrateMult);
-                    other_fp->x221A_b2 = true;
+                    other_fp->allow_sdi = true;
                     if (!other_fp->x2219_b5) {
                         if (other_fp->pre_hitlag_cb != NULL) {
                             other_fp->pre_hitlag_cb(gobj);
diff --git a/src/melee/ft/chara/ftCommon/ftCo_Guard.c b/src/melee/ft/chara/ftCommon/ftCo_Guard.c
index 34639f0cf2..8c83fbe554 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Guard.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_Guard.c
@@ -621,14 +621,14 @@ void ftCo_80092F2C(HSD_GObj* gobj, bool arg1)
 void ftCo_80093240(Fighter_GObj* gobj)
 {
     Fighter* fp = gobj->user_data;
-    if (fp->x221A_b2 && fp->ground_or_air == GA_Ground) {
+    if (fp->allow_sdi && fp->ground_or_air == GA_Ground) {
         if ((fp->input.lstick.x < 0
                  ? -fp->input.lstick.x
-                 : fp->input.lstick.x) >= p_ftCommonData->x4B0 &&
-            fp->x670_timer_lstick_tilt_x < p_ftCommonData->x4B4)
+                 : fp->input.lstick.x) >= p_ftCommonData->sdi_min_stick_mag &&
+            fp->x670_timer_lstick_tilt_x < p_ftCommonData->sdi_stick_window)
         {
             float scl = p_ftCommonData->x4C0 *
-                        (fp->input.lstick.x * p_ftCommonData->x4B8);
+                        (fp->input.lstick.x * p_ftCommonData->sdi_pos_scale);
             fp->cur_pos.x += fp->coll_data.floor.normal.y * scl;
             fp->cur_pos.y += -fp->coll_data.floor.normal.x * scl;
             fp->x670_timer_lstick_tilt_x = 254;
@@ -642,7 +642,7 @@ void ftCo_800932DC(Fighter_GObj* gobj)
     if (fp->ground_or_air == GA_Ground) {
         if ((fp->input.lstick.x < 0
                  ? -fp->input.lstick.x
-                 : fp->input.lstick.x) >= p_ftCommonData->x4B0)
+                 : fp->input.lstick.x) >= p_ftCommonData->sdi_min_stick_mag)
         {
             float scl = p_ftCommonData->x4C0 *
                         (fp->input.lstick.x * p_ftCommonData->x4BC);
@@ -681,7 +681,7 @@ void ftCo_GuardSetOff_Phys(Fighter_GObj* gobj)
 void ftCo_GuardSetOff_Coll(Fighter_GObj* gobj)
 {
     Fighter* fp = gobj->user_data;
-    if (!(fp->x221A_b2)) {
+    if (!(fp->allow_sdi)) {
         ft_800845B4(gobj);
     } else {
         ft_80084104(gobj);
diff --git a/src/melee/ft/fighter.c b/src/melee/ft/fighter.c
index 7e1a3f76fe..664bb6d508 100644
--- a/src/melee/ft/fighter.c
+++ b/src/melee/ft/fighter.c
@@ -347,7 +347,7 @@ void Fighter_UnkInitReset_80067C98(Fighter* fp)
     fp->dmg.x1954 = 0;
     fp->dmg.x1958 = 0;
 
-    fp->x221A_b2 = 0;
+    fp->allow_sdi = 0;
 
     fp->dmg.x195c_hitlag_frames = 0;
 
@@ -1408,7 +1408,7 @@ void Fighter_8006A1BC(Fighter_GObj* gobj)
             fp->dmg.x1954 -= 1.0f;
             if (fp->dmg.x1954 <= 0.0f) {
                 fp->dmg.x1954 = 0.0f;
-                if (!fp->x221A_b2 && !fp->x2219_b7) {
+                if (!fp->allow_sdi && !fp->x2219_b7) {
                     Fighter_8006D10C(gobj);
                 }
             }
@@ -1430,7 +1430,7 @@ void Fighter_8006A1BC(Fighter_GObj* gobj)
                 if ((!fp->dmg.x1954) && !fp->x2219_b7) {
                     Fighter_8006D10C(gobj);
                 }
-                fp->x221A_b2 = 0;
+                fp->allow_sdi = 0;
             }
         }
         ftCo_800C37A0(gobj);
@@ -2713,7 +2713,7 @@ void Fighter_8006CFE0(Fighter_GObj* gobj)
     Fighter* fp = GET_FIGHTER(gobj);
 
     if (fp->x2219_b7) {
-        if (!fp->x221A_b2) {
+        if (!fp->allow_sdi) {
             if (!fp->dmg.x1954) {
                 Fighter_8006D10C(gobj);
             }
@@ -2760,7 +2760,7 @@ static void Fighter_8006D10C_Inline1(Fighter_GObj* gobj)
     Fighter* fp = GET_FIGHTER(gobj);
 
     if (fp->x2219_b7) {
-        if (!fp->x221A_b2 && !fp->dmg.x1954) {
+        if (!fp->allow_sdi && !fp->dmg.x1954) {
             if (fp->post_hitlag_cb) {
                 fp->post_hitlag_cb(gobj);
             }
@@ -2958,7 +2958,7 @@ void Fighter_ProcessHit_8006D1EC(Fighter_GObj* gobj)
                     fp->dmg.x195c_hitlag_frames =
                         p_ftCommonData->x194_unkHitLagFrames;
                 }
-                fp->x221A_b2 = 1;
+                fp->allow_sdi = 1;
                 if (bool2) {
                     fp->x221A_b3 = 1;
                 }
diff --git a/src/melee/ft/ft_081B.c b/src/melee/ft/ft_081B.c
index f16525e7d0..7a5ad6c419 100644
--- a/src/melee/ft/ft_081B.c
+++ b/src/melee/ft/ft_081B.c
@@ -140,7 +140,7 @@ bool ft_80081DD4(Fighter_GObj* gobj)
     coll = &fp->coll_data;
     fp->coll_data.last_pos = fp->coll_data.cur_pos;
     fp->coll_data.cur_pos = fp->cur_pos;
-    if (fp->x221A_b2) {
+    if (fp->allow_sdi) {
         var_r28 = mpColl_800477E0(coll);
     } else if (fp->x2064_ledgeCooldown != 0 || fp->x2224_b2) {
         var_r28 = mpColl_800471F8(coll);
@@ -184,7 +184,7 @@ bool ft_80081F2C(Fighter_GObj* gobj)
     coll = &fp->coll_data;
     fp->coll_data.last_pos = fp->coll_data.cur_pos;
     fp->coll_data.cur_pos = fp->cur_pos;
-    if (fp->x221A_b2) {
+    if (fp->allow_sdi) {
         var_r28 = mpColl_80048274(coll);
     } else if (fp->x2064_ledgeCooldown != 0 || fp->x2224_b2) {
         var_r28 = mpColl_80048160(coll);
@@ -227,7 +227,7 @@ bool ft_80082084(Fighter_GObj* gobj)
     coll = &fp->coll_data;
     fp->coll_data.last_pos = fp->coll_data.cur_pos;
     fp->coll_data.cur_pos = fp->cur_pos;
-    if (fp->x221A_b2) {
+    if (fp->allow_sdi) {
         var_r28 = mpColl_80048388(coll);
     } else if (fp->x2064_ledgeCooldown != 0 || fp->x2224_b2) {
         var_r28 = mpColl_80048768(coll);
diff --git a/src/melee/ft/ft_0DF1.c b/src/melee/ft/ft_0DF1.c
index 152477c707..32f6a69aa1 100644
--- a/src/melee/ft/ft_0DF1.c
+++ b/src/melee/ft/ft_0DF1.c
@@ -132,7 +132,7 @@ bool ftCo_800DF50C(Fighter* fp)
 bool ftCo_800DF608(Fighter* fp)
 {
     if (SQ(fp->input.cstick.x) + SQ(fp->input.cstick.y) >=
-        SQ(p_ftCommonData->x4B0))
+        SQ(p_ftCommonData->sdi_min_stick_mag))
     {
         return true;
     }
diff --git a/src/melee/ft/types.h b/src/melee/ft/types.h
index 29443feb67..7c44f6a5e9 100644
--- a/src/melee/ft/types.h
+++ b/src/melee/ft/types.h
@@ -350,9 +350,9 @@ struct ftCommonData {
     /* +4A4 */ float x4A4;
     /* +4A8 */ float x4A8;
     /* +4AC */ float x4AC;
-    /* +4B0 */ float x4B0;
-    /* +4B4 */ int x4B4;
-    /* +4B8 */ float x4B8;
+    /* +4B0 */ float sdi_min_stick_mag;
+    /* +4B4 */ int sdi_stick_window;
+    /* +4B8 */ float sdi_pos_scale;
     /* +4BC */ float x4BC;
     /* +4C0 */ float x4C0;
     /* +4C4 */ int x4C4;
@@ -1596,7 +1596,7 @@ struct Fighter {
 
     /* fp+221A:0 */ u8 x221A_b0 : 1;
     /* fp+221A:1 */ u8 x221A_b1 : 1;
-    /* fp+221A:2 */ u8 x221A_b2 : 1;
+    /* fp+221A:2 */ u8 allow_sdi : 1;
     /* fp+221A:3 */ u8 x221A_b3 : 1;
     /* fp+221A:4 */ u8 fall_fast : 1;
     /* fp+221A:5 */ u8 x221A_b5 : 1;
