diff --git a/src/melee/gr/grcorneria.c b/src/melee/gr/grcorneria.c
index 99f4c9c27b..3a67f9250b 100644
--- a/src/melee/gr/grcorneria.c
+++ b/src/melee/gr/grcorneria.c
@@ -540,7 +540,14 @@ bool grCorneria_801E0D28(Ground_GObj* arg)
     return false;
 }
 
-/// #grCorneria_801E0D30
+void grCorneria_801E0D30(Ground_GObj* gobj)
+{
+    HSD_JObj* jobj;
+
+    if ((jobj = gobj->hsd_obj) != NULL) {
+        HSD_JObjSetTranslateY(jobj, -grCorneria_801E2EA0());
+    }
+}
 
 void grCorneria_801E0DE0(Ground_GObj* arg) {}
 
@@ -586,7 +593,33 @@ bool grCorneria_801E0F64(Ground_GObj* arg)
     return false;
 }
 
-/// #grCorneria_801E0F6C
+void grCorneria_801E0F6C(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+
+    ifStatus_802F6898();
+    un_802FF570();
+
+    if (Ground_801C2BA4(5) == NULL) {
+        if (grCorneria_801E2598(gp->gv.arwing.xC4,
+                                gp->gv.arwing.xC8))
+        {
+            if ((s32) gp->gv.arwing.xCC-- < 0) {
+                HSD_GObj* gobj5 = grCorneria_801DD534(5);
+                Ground* gp5 = GET_GROUND(gobj5);
+                grCorneria_801E2738(gobj5, &gp5->gv,
+                                    gp->gv.arwing.xC4,
+                                    gp->gv.arwing.xC8);
+                gp->gv.arwing.xC8++;
+                gp->gv.arwing.xCC = 0;
+            }
+        } else {
+            if ((s32) gp->gv.arwing.xCC-- < 0) {
+                Ground_801C4A08(gobj);
+            }
+        }
+    }
+}
 
 void grCorneria_801E1030(Ground_GObj* gobj)
 {
@@ -825,7 +858,21 @@ bool grCorneria_801E2D14(void)
     return false;
 }
 
-/// #grCorneria_801E2D90
+bool grCorneria_801E2D90(enum_t line_id)
+{
+    if (stage_info.internal_stage_id == CORNERIA && line_id != -1) {
+        s32 joint = mpJointFromLine(line_id);
+        s32 result;
+
+        result = joint == 0 || joint == 1 || joint == 2 || joint == 5 ||
+                 joint == 6 || joint == 7;
+
+        if (result) {
+            return true;
+        }
+    }
+    return false;
+}
 
 bool grCorneria_801E2E50(int line_id)
 {
diff --git a/src/melee/gr/grgreatbay.c b/src/melee/gr/grgreatbay.c
index 5b64c84e48..5509fc40af 100644
--- a/src/melee/gr/grgreatbay.c
+++ b/src/melee/gr/grgreatbay.c
@@ -390,7 +390,24 @@ void grGreatBay_801F5D4C(Ground_GObj* gobj)
 
 /// #grGreatBay_801F63F4
 
-/// #grGreatBay_801F660C
+s16 grGb_803E4048[][2] = {
+    { 32, 20 }, { 23, 18 }, { 22, 15 }, { 29, 16 },
+    { 28, 17 }, { 27, 19 }, { 39, 21 }, { 0, 0 },
+};
+
+void grGreatBay_801F660C(Ground_GObj* gobj)
+{
+    u32 i;
+    HSD_JObj* jobj;
+    Vec3 pos;
+
+    for (i = 0; i < 7; i++) {
+        jobj = Ground_801C3FA4(gobj, grGb_803E4048[i][1]);
+        HSD_ASSERT(1398, jobj);
+        lb_8000B1CC(jobj, NULL, &pos);
+        mpVtxSetPos(grGb_803E4048[i][0], pos.x, pos.y);
+    }
+}
 
 bool grGreatBay_801F66A4(void)
 {
diff --git a/src/melee/gr/grgreatbay.h b/src/melee/gr/grgreatbay.h
index 4032bee7a5..1e02647423 100644
--- a/src/melee/gr/grgreatbay.h
+++ b/src/melee/gr/grgreatbay.h
@@ -58,7 +58,7 @@
                                       mpLib_GroundEnum, f32);
 /* 1F62F8 */ UNK_RET grGreatBay_801F62F8(UNK_PARAMS);
 /* 1F63F4 */ UNK_RET grGreatBay_801F63F4(UNK_PARAMS);
-/* 1F660C */ UNK_RET grGreatBay_801F660C(UNK_PARAMS);
+/* 1F660C */ void grGreatBay_801F660C(Ground_GObj*);
 /* 1F66A4 */ bool grGreatBay_801F66A4(void);
 /* 1F6708 */ bool grGreatBay_801F6708(u32, HSD_GObj*);
 /* 1F67A4 */ void grGreatBay_801F67A4(Vec3*, f32);
diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index 04fd401cce..3fe1a3b9d0 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -11,6 +11,7 @@
 #include "it/it_26B1.h"
 #include "lb/lb_00B0.h"
 
+f32 grHr_804D6AE4;
 static void* grHr_804D6AE8;
 StageCallbacks grHr_803E8140[11] = {
     { grHomeRun_8021C914, grHomeRun_8021CB10, grHomeRun_8021CB18,
@@ -136,7 +137,35 @@ bool grHomeRun_8021DF48(Ground_GObj* arg)
     return false;
 }
 
-/// #grHomeRun_8021DF50
+void grHomeRun_8021DF50(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    if ((u32) gp->gv.unk.xCC != 0) {
+        if ((u32) gp->gv.unk.xC8 == 0) {
+            Vec3 pos;
+            f32 scale;
+            f32 y, z;
+            HSD_Text* text;
+
+            gp->gv.unk.xC8 =
+                (int) grHomeRun_8021EC58(gp->gv.homerun.xC6);
+            lb_8000B1CC((HSD_JObj*) gp->gv.unk.xCC, NULL, &pos);
+
+            scale = Ground_801C0498();
+            z = pos.z + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            y = -pos.y + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            text = (HSD_Text*) gp->gv.unk.xC8;
+            text->pos_x =
+                pos.x + (-1.0F) * (grHr_804D6AE4 * scale);
+            text->pos_y = y;
+            text->pos_z = z;
+        }
+    }
+}
 
 void grHomeRun_8021E008(Ground_GObj* gobj)
 {
@@ -174,7 +203,35 @@ bool grHomeRun_8021E0CC(Ground_GObj* arg)
     return false;
 }
 
-/// #grHomeRun_8021E0D4
+void grHomeRun_8021E0D4(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    if ((u32) gp->gv.unk.xCC != 0) {
+        if ((u32) gp->gv.unk.xC8 == 0) {
+            Vec3 pos;
+            f32 scale;
+            f32 y, z;
+            HSD_Text* text;
+
+            gp->gv.unk.xC8 =
+                (int) grHomeRun_8021EC58(gp->gv.homerun.xC6);
+            lb_8000B1CC((HSD_JObj*) gp->gv.unk.xCC, NULL, &pos);
+
+            scale = Ground_801C0498();
+            z = pos.z + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            y = -pos.y + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            text = (HSD_Text*) gp->gv.unk.xC8;
+            text->pos_x =
+                pos.x + (-1.0F) * (grHr_804D6AE4 * scale);
+            text->pos_y = y;
+            text->pos_z = z;
+        }
+    }
+}
 
 void grHomeRun_8021E18C(Ground_GObj* gobj)
 {
@@ -212,7 +269,35 @@ bool grHomeRun_8021E250(Ground_GObj* arg)
     return false;
 }
 
-/// #grHomeRun_8021E258
+void grHomeRun_8021E258(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    if ((u32) gp->gv.unk.xCC != 0) {
+        if ((u32) gp->gv.unk.xC8 == 0) {
+            Vec3 pos;
+            f32 scale;
+            f32 y, z;
+            HSD_Text* text;
+
+            gp->gv.unk.xC8 =
+                (int) grHomeRun_8021EC58(gp->gv.homerun.xC6);
+            lb_8000B1CC((HSD_JObj*) gp->gv.unk.xCC, NULL, &pos);
+
+            scale = Ground_801C0498();
+            z = pos.z + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            y = -pos.y + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            text = (HSD_Text*) gp->gv.unk.xC8;
+            text->pos_x =
+                pos.x + (-1.0F) * (grHr_804D6AE4 * scale);
+            text->pos_y = y;
+            text->pos_z = z;
+        }
+    }
+}
 
 void grHomeRun_8021E310(Ground_GObj* gobj)
 {
@@ -250,7 +335,35 @@ bool grHomeRun_8021E3D4(Ground_GObj* arg)
     return false;
 }
 
-/// #grHomeRun_8021E3DC
+void grHomeRun_8021E3DC(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    if ((u32) gp->gv.unk.xCC != 0) {
+        if ((u32) gp->gv.unk.xC8 == 0) {
+            Vec3 pos;
+            f32 scale;
+            f32 y, z;
+            HSD_Text* text;
+
+            gp->gv.unk.xC8 =
+                (int) grHomeRun_8021EC58(gp->gv.homerun.xC6);
+            lb_8000B1CC((HSD_JObj*) gp->gv.unk.xCC, NULL, &pos);
+
+            scale = Ground_801C0498();
+            z = pos.z + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            y = -pos.y + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            text = (HSD_Text*) gp->gv.unk.xC8;
+            text->pos_x =
+                pos.x + (-1.0F) * (grHr_804D6AE4 * scale);
+            text->pos_y = y;
+            text->pos_z = z;
+        }
+    }
+}
 
 void grHomeRun_8021E494(Ground_GObj* gobj)
 {
diff --git a/src/melee/gr/grhomerun.h b/src/melee/gr/grhomerun.h
index bab72ef497..d3c74d0ee6 100644
--- a/src/melee/gr/grhomerun.h
+++ b/src/melee/gr/grhomerun.h
@@ -4,6 +4,7 @@
 #include <placeholder.h>
 
 #include "baselib/forward.h"
+#include "lb/forward.h"
 
 #include "it/inlines.h"
 
@@ -66,7 +67,7 @@
 /* 21EA30 */ UNK_RET grHomeRun_8021EA30(UNK_PARAMS);
 /* 21EAF8 */ UNK_RET grHomeRun_8021EAF8(UNK_PARAMS);
 /* 21EB10 */ void fn_8021EB10(HSD_GObj*);
-/* 21EC58 */ UNK_RET grHomeRun_8021EC58(UNK_PARAMS);
+/* 21EC58 */ void* grHomeRun_8021EC58(u16);
 /* 21ED74 */ void grHomeRun_8021ED74(void);
 /* 21EDD4 */ void grHomeRun_8021EDD4(int);
 /* 21EEB4 */ DynamicsDesc* grHomeRun_8021EEB4(enum_t);
diff --git a/src/melee/gr/grinishie1.c b/src/melee/gr/grinishie1.c
index fe2a9fadf2..26ea1433b8 100644
--- a/src/melee/gr/grinishie1.c
+++ b/src/melee/gr/grinishie1.c
@@ -719,7 +719,34 @@ void grInishie1_801FBCEC(HSD_GObj* gobj, u32 index)
     Camera_80030E44(2, &effect_pos);
 }
 
-/// #fn_801FBF6C
+void fn_801FBF6C(Item_GObj* item_gobj, Ground* gp, Vec3* pos,
+                 HSD_GObj* arg3, f32 arg4)
+{
+    s32 i;
+    HSD_GObj* gobj;
+    Item* ip = item_gobj->user_data;
+
+    for (i = 0; i < 19; i++) {
+        if (gp->gv.inishie1.blocks[i].jobj2 ==
+            (HSD_JObj*) ip->xDD4_itemVar.it_266F.x4)
+        {
+            break;
+        }
+    }
+
+    if ((u32) i == 19) {
+        return;
+    }
+
+    gobj = Ground_801C2BA4(3);
+    if (gobj == NULL) {
+        __assert("grinishie1.c", 0x425, "new_gobj");
+    }
+
+    grInishie1_801FB0AC(gobj, i);
+    grInishie1_801FBCEC(gobj, i);
+    PAD_STACK(16);
+}
 
 /// #grInishie1_801FC018
 
diff --git a/src/melee/gr/groldkongo.c b/src/melee/gr/groldkongo.c
index 98a76ddf07..2c0cda3c35 100644
--- a/src/melee/gr/groldkongo.c
+++ b/src/melee/gr/groldkongo.c
@@ -14,9 +14,12 @@
 #include "lb/lb_00F9.h"
 
 #include <baselib/gobj.h>
+#include <baselib/jobj.h>
+#include <baselib/random.h>
 
 static struct {
-    int x0;
+    s16 x0;
+    s16 x2;
 }* grOk_804D6A90;
 
 StageCallbacks grOk_803E658C[4] = {
@@ -126,7 +129,25 @@ bool grOldKongo_8020F880(Ground_GObj* gobj)
 
 void grOldKongo_80210058(Ground_GObj* arg) {}
 
-/// #grOldKongo_8021005C
+static inline int rand_inline(int a, int b)
+{
+    if (a > b) {
+        return b + (a - b != 0 ? HSD_Randi(a - b) : 0);
+    } else if (a < b) {
+        return a + (b - a != 0 ? HSD_Randi(b - a) : 0);
+    } else {
+        return a;
+    }
+}
+
+void grOldKongo_8021005C(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    HSD_JObjSetFlagsAll(jobj, 0x10);
+    gp->gv.unk.xC4 =
+        rand_inline(grOk_804D6A90->x2, grOk_804D6A90->x0);
+}
 
 bool grOldKongo_802100F4(Ground_GObj* gobj)
 {
diff --git a/src/melee/gr/groldpupupu.c b/src/melee/gr/groldpupupu.c
index ccadb40e5b..6be94088de 100644
--- a/src/melee/gr/groldpupupu.c
+++ b/src/melee/gr/groldpupupu.c
@@ -13,9 +13,11 @@
 
 #include <baselib/gobj.h>
 #include <baselib/jobj.h>
+#include <baselib/random.h>
 
 static struct {
-    int x0;
+    s16 x0;
+    s16 x2;
 }* grOp_804D6A98;
 
 static void* grOp_804D6A9C;
@@ -173,7 +175,20 @@ void grOldPupupu_80210C34(Ground_GObj* gobj)
 
 void grOldPupupu_80210C78(Ground_GObj* arg) {}
 
-/// #grOldPupupu_80210C7C
+void grOldPupupu_80210C7C(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    s16 x2, x0;
+
+    x2 = grOp_804D6A98->x2;
+    x0 = grOp_804D6A98->x0;
+    *(s16*) &gp->gv.unk.xC4 =
+        x2 > x0
+            ? x0 + (x2 - x0 != 0 ? HSD_Randi(x2 - x0) : 0)
+        : x2 < x0
+            ? x2 + (x0 - x2 != 0 ? HSD_Randi(x0 - x2) : 0)
+            : x2;
+}
 
 bool grOldPupupu_80210D08(Ground_GObj* gobj)
 {
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 39407cf087..7c16ba0e82 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -1052,6 +1052,13 @@ struct grPushOn_GroundVars {
     /* +AC gp+170 */ HSD_LObj* spot_light;
 };
 
+struct grHomeRun_GroundVars {
+    u16 xC4;
+    u16 xC6;
+    int xC8;
+    int xCC;
+};
+
 struct Map_GroundVars {
     /* +00 gp+C4:0  */ u32 xC4_b0 : 1;
     /* +00 gp+C4:1  */ u32 xC4_b1 : 1;
@@ -1188,6 +1195,7 @@ struct Ground {
             struct grShrineroute_GroundVars2 shrineroute2;
             struct grSmashTaunt_GroundVars smashtaunt;
             struct GroundVars_unk unk;
+            struct grHomeRun_GroundVars homerun;
             struct grVenom_GroundVars venom;
             struct grYorster_GroundVars yorster;
             struct grZebes_GroundVars zebes;
