diff --git a/src/melee/ft/ft_0D31.c b/src/melee/ft/ft_0D31.c
index c05681f2a0..cf1cff665b 100644
--- a/src/melee/ft/ft_0D31.c
+++ b/src/melee/ft/ft_0D31.c
@@ -483,7 +483,15 @@ void ftCo_800D4F24(Fighter_GObj* gobj, int index)
 
 /// #ftCo_800D4FF4
 
-/// #ftCo_Rebirth_Anim
+void ftCo_Rebirth_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftCo_8008A7A8(gobj, fp->ft_data->x24);
+    fp->mv.co.common.x0 -= 1;
+    if (fp->mv.co.common.x0 == 0) {
+        ftCo_800D5600(gobj);
+    }
+}
 
 void ftCo_Rebirth_IASA(Fighter_GObj* gobj) {}
 
diff --git a/src/melee/ft/ft_0D31.h b/src/melee/ft/ft_0D31.h
index 956ed099db..1d72678bc2 100644
--- a/src/melee/ft/ft_0D31.h
+++ b/src/melee/ft/ft_0D31.h
@@ -39,7 +39,7 @@
 /* 0D535C */ void ftCo_Rebirth_Phys(Fighter_GObj* gobj);
 /* 0D5470 */ void ftCo_Rebirth_Coll(Fighter_GObj* gobj);
 /* 0D55B4 */ void fn_800D55B4(Fighter_GObj* gobj);
-/* 0D5600 */ UNK_RET ftCo_800D5600(UNK_PARAMS);
+/* 0D5600 */ void ftCo_800D5600(Fighter_GObj* gobj);
 /* 0D56EC */ void ftCo_RebirthWait_Anim(Fighter_GObj* gobj);
 /* 0D575C */ void ftCo_RebirthWait_IASA(Fighter_GObj* gobj);
 /* 0D58F4 */ void ftCo_RebirthWait_Phys(Fighter_GObj* gobj);
