diff --git a/src/melee/ty/toy.c b/src/melee/ty/toy.c
index 34a869a599..e889842acc 100644
--- a/src/melee/ty/toy.c
+++ b/src/melee/ty/toy.c
@@ -271,55 +271,6 @@ typedef struct {
     s8 x61;
 } Toy6E68;
 
-typedef struct ToyGlobalsS_ {
-    HSD_GObj* x0;
-    u8 pad4[0x4];
-    HSD_GObj* x8;
-    HSD_GObj* xC;
-    u8 pad10[0x20];
-    void* x30;
-    u8 pad34[0x1C];
-    void* x50;
-    void* x54;
-} ToyGlobalsS_;
-
-typedef struct ToyGlobals2S_ {
-    HSD_GObj* x0;
-    u8 x4;
-} ToyGlobals2S_;
-
-typedef struct ToyGlobals3S_ {
-    u8 pad0[0x10];
-    s32 x10;
-} ToyGlobals3S_;
-
-typedef struct ToyGlobals4S_ {
-    u8 pad0[0x58];
-    s32 x58;
-} ToyGlobals4S_;
-
-typedef struct ToySubStructS_ {
-    u8 pad0[0x10];
-    s16 x10;
-} ToySubStructS_;
-
-typedef struct ToyGlobals5S_ {
-    u8 pad0[0x140];
-    ToySubStructS_* x140;
-    u8 pad144[0x10];
-    s16 x154;
-} ToyGlobals5S_;
-
-typedef struct TyUnk25 {
-    u8 pad[0x140];
-    ToySubStructS_* x140;
-    void* x144;
-    void* x148;
-    void* x14C;
-    void* x150;
-    s16 x154;
-} TyUnk25;
-
 typedef struct {
     u8 pad[0x14];
     void* x14;
@@ -5225,10 +5176,10 @@ void un_80310324(void)
     char* data;
     ToyGlobalsS_* tg;
     ToyGlobalsS_* tg2;
-    ToyGlobals2S_* tg3;
-    ToyGlobals3S_* tg4;
-    ToyGlobals4S_* tg5;
-    ToyGlobals5S_* tg6;
+    ToyGlobalsS_* tg3;
+    ToyGlobalsS_* tg4;
+    ToyGlobalsS_* tg5;
+    ToyGlobalsS_* tg6;
     ToySubStructS_* sub;
     s32 sp[4];
     s32 i;
@@ -5350,7 +5301,7 @@ void un_80310324(void)
 
     un_80307828(0);
 
-    tg5 = (ToyGlobals4S_*) un_804D6E68;
+    tg5 = (ToyGlobalsS_*) un_804D6E68;
     tg5->x58 = 0x95E;
 
     un_8030715C(un_804DDCD8, un_804DDCD8);
@@ -5372,7 +5323,7 @@ void un_80310660(s32 arg0)
     TyCleanupObj* ty28;
     u8* ty27;
     void* ty26;
-    TyUnk25* ty25;
+    ToyGlobalsS_* ty25;
     tyLightData* ty30;
 
     state = (u8*) un_804A26B8;
diff --git a/src/melee/ty/tylist.c b/src/melee/ty/tylist.c
index e930a18c9d..c4803eceb3 100644
--- a/src/melee/ty/tylist.c
+++ b/src/melee/ty/tylist.c
@@ -1,5 +1,7 @@
 #include "tylist.h"
 
+#include "stddef.h"
+
 #include "gm/gmmain_lib.h"
 #include "if/textlib.h"
 #include "lb/lb_00B0.h"
@@ -44,8 +46,15 @@ typedef struct {
     void* data;
 } TyArchiveData;
 
+typedef struct {
+    u8 pad[0x20];
+    u32 x20;
+    s32 x24;
+} TyListWaitData;
+
 void un_803124BC(void)
 {
+    char* data = un_803FDD18;
     u16* table1;
     s16* list;
     u16* table2;
@@ -57,15 +66,15 @@ void un_803124BC(void)
     if (un_804D6ED0 == NULL) {
         char* archiveName;
         if (lbLang_IsSavedLanguageJP()) {
-            archiveName = "TyDatai.dat";
+            archiveName = data + 0x608;
         } else {
-            archiveName = "TyDatai.usd";
+            archiveName = data + 0x614;
         }
         un_804D6ED0 = lbArchive_LoadSymbols(
-            archiveName, &un_804D6EC4, "tyInitModelTbl", &un_804D6EC0,
-            "tyInitModelDTbl", &un_804D6EBC, "tyModelSortTbl", &un_804D6EB8,
-            "tyExpDifferentTbl", &un_804D6EB4, "tyNoGetUsTbl", &un_804D6EB0,
-            "tyDisplayModelTbl", &un_804D6EAC, "tyDisplayModelUsTbl", NULL);
+            archiveName, &un_804D6EC4, data + 0x9DC, &un_804D6EC0,
+            data + 0x9EC, &un_804D6EBC, data + 0x9FC, &un_804D6EB8,
+            data + 0xA0C, &un_804D6EB4, data + 0xA20, &un_804D6EB0,
+            data + 0xA30, &un_804D6EAC, data + 0xA44, NULL);
     }
 
     i = 0;
@@ -101,6 +110,7 @@ loop: {
 
 void un_8031263C(void)
 {
+    char* data = un_803FDD18;
     s32 i;
     u16* table1;
     u16* table2;
@@ -112,15 +122,15 @@ void un_8031263C(void)
     if (un_804D6ED0 == NULL) {
         char* archiveName;
         if (lbLang_IsSavedLanguageJP()) {
-            archiveName = "TyDatai.dat";
+            archiveName = data + 0x608;
         } else {
-            archiveName = "TyDatai.usd";
+            archiveName = data + 0x614;
         }
         un_804D6ED0 = lbArchive_LoadSymbols(
-            archiveName, &un_804D6EC4, "tyInitModelTbl", &un_804D6EC0,
-            "tyInitModelDTbl", &un_804D6EBC, "tyModelSortTbl", &un_804D6EB8,
-            "tyExpDifferentTbl", &un_804D6EB4, "tyNoGetUsTbl", &un_804D6EB0,
-            "tyDisplayModelTbl", &un_804D6EAC, "tyDisplayModelUsTbl", NULL);
+            archiveName, &un_804D6EC4, data + 0x9DC, &un_804D6EC0,
+            data + 0x9EC, &un_804D6EBC, data + 0x9FC, &un_804D6EB8,
+            data + 0xA0C, &un_804D6EB4, data + 0xA20, &un_804D6EB0,
+            data + 0xA30, &un_804D6EAC, data + 0xA44, NULL);
     }
 
     i = 0;
@@ -138,13 +148,14 @@ void un_8031263C(void)
     un_804A284C[3] |= 4;
 
     if (un_804D6ECC == NULL) {
-        un_804D6ECC = lbArchive_LoadSymbols(str_TyDataf_dat, &un_804D6EA8,
-                                            str_tyModelFileTbl, &un_804D6EA4,
-                                            str_tyModelFileUsTbl, NULL);
+        un_804D6ECC =
+            lbArchive_LoadSymbols(data + 0xA58, &un_804D6EA8, data + 0xA64,
+                                  &un_804D6EA4, data + 0xA74, NULL);
     }
 
     un_8031234C(0);
 }
+
 void un_803127D4(void)
 {
     un_804D6ED0 = NULL;
@@ -406,7 +417,7 @@ void un_80312BAC(TyListState* state, s8 arg1)
             }
         }
     done_inner_dirty:
-        row = row->x4;
+        row = row->links[1];
         var_r28++;
         if (var_r28 >= un_GetTrophyTotal()) {
             var_r28 = 0;
@@ -453,7 +464,7 @@ void un_80312E88(TyListArg* arg, float delta)
     data = un_804A2AC0;
     i = 0;
 loop:
-    if ((jobj = ptr->xC) == NULL) {
+    if ((jobj = ptr->jobjs[0]) == NULL) {
         goto next;
     }
     if (i == 2) {
@@ -526,14 +537,15 @@ void un_80312E88(TyListArg* arg, float delta)
     un_80312904(arg, 0x63);
     un_80313464(arg);
 }
-s8 un_8031305C(void* a, TyListState* state, s8 movedFlag)
+
+s32 un_8031305C(void* a, TyListState* state, s8 movedFlag)
 {
     TyListArg* entry;
     f32 delta;
     s32 i;
-    PAD_STACK(8);
-
-    if ((s8) state->x29F > 0) {
+    PAD_STACK(0xC);
+    i = ((s8) state->x29F) > 0;
+    if (i) {
         delta = state->x2A4;
         if ((s8) state->x2A1 == 0) {
             delta *= un_804DDE50;
@@ -567,7 +579,7 @@ s8 un_8031305C(void* a, TyListState* state, s8 movedFlag)
                         entry->x24 = (s8) -1;
                     }
                 }
-                entry = entry->x4;
+                entry = entry->links[1];
                 i++;
             }
             if ((s8) state->x2A1 == 0) {
@@ -576,20 +588,20 @@ s8 un_8031305C(void* a, TyListState* state, s8 movedFlag)
                     state->x270->idx = 0;
                 }
                 un_80312904(state->x270, un_804A2D6C.x0C);
-                state->x278 = state->x278->x4;
+                state->x278 = state->x278->links[1];
                 state->selectedIdx = state->x278->idx;
-                state->x270 = state->x270->x4;
-                state->x274 = state->x274->x4;
+                state->x270 = state->x270->links[1];
+                state->x274 = state->x274->links[1];
             } else {
                 state->x274->idx = (s16) (state->x270->idx - 1);
                 if (state->x274->idx < 0) {
                     state->x274->idx = (s16) (un_GetTrophyTotal() - 1);
                 }
                 un_80312904(entry, un_804A2D6C.x0C);
-                state->x278 = state->x278->x0;
+                state->x278 = state->x278->links[0];
                 state->selectedIdx = state->x278->idx;
-                state->x270 = state->x270->x0;
-                state->x274 = state->x274->x0;
+                state->x270 = state->x270->links[0];
+                state->x274 = state->x274->links[0];
             }
             if ((s8) state->x29E > 0) {
                 state->x29E = (u8) (state->x29E - 1);
@@ -626,20 +638,20 @@ void un_80313358(TyListState* state, s8 arg2, s8 arg3, s8 arg4)
     if (state->x2A1 == 0) {
         for (i = 0; i < state->entryCount; i++) {
             TyListArg* entry = &state->entries[i];
-            TyListArg* sub = entry->x0;
+            TyListArg* sub = entry->links[0];
             entry->x2C = sub->x30;
             un_80312904(entry, state->entryCount + 1);
         }
     } else {
         for (i = 0; i < state->entryCount; i++) {
             TyListArg* entry = &state->entries[i];
-            TyListArg* sub = entry->x4;
+            TyListArg* sub = entry->links[1];
             entry->x2C = sub->x30;
             un_80312904(entry, state->entryCount + 1);
         }
     }
 }
-
+extern s32 un_804D6EE8;
 void un_80313464(TyListArg* arg)
 {
     char* data = un_804A2AC0;
@@ -648,18 +660,19 @@ void un_80313464(TyListArg* arg)
 
     val = un_804D6EDC[arg->idx];
 
-    un_803083D8(arg->x14, val);
+    un_803083D8(arg->jobjs[2], val);
 
-    if (arg->x10 != NULL) {
-        HSD_JObjUnref(arg->x10);
-        arg->x10 = NULL;
+    if (arg->jobjs[1] != NULL) {
+        HSD_JObjUnref(arg->jobjs[1]);
+        arg->jobjs[1] = NULL;
     }
 
     if (un_80304924(val) != 0) {
-        arg->x10 = un_80313508(((TyListState*) data)->gobj, un_803FE8D0,
-                               un_804DDE60, arg->x30, un_804DDE48);
+        arg->jobjs[1] = un_80313508(((TyListState*) data)->gobj, un_803FE8D0,
+                                    un_804DDE60, arg->x30, un_804DDE48);
     }
 }
+
 HSD_JObj* un_80313508(void* parent, void* symbol, float x, float y, float z)
 {
     HSD_JObj* jobj;
@@ -672,111 +685,136 @@ HSD_JObj* un_80313508(void* parent, void* symbol, float x, float y, float z)
 
     if (joint == NULL) {
         OSPanic(un_803FE8F0, 0x337, un_803FE8FC);
-        return NULL;
-    }
-
-    jobj = HSD_JObjLoadJoint(joint);
+    } else {
+        jobj = HSD_JObjLoadJoint(joint);
 
-    if (x != un_804DDE48 || y != un_804DDE48 || z != un_804DDE48) {
-        // Set X
-        if (jobj == NULL) {
-            __assert(&un_804D5A78, 0x3A4, &un_804D5A80);
+        if (x != un_804DDE48 || y != un_804DDE48 || z != un_804DDE48) {
+            HSD_JObjSetTranslateX(jobj, x);
+            HSD_JObjSetTranslateY(jobj, y);
+            HSD_JObjSetTranslateZ(jobj, z);
         }
-        jobj->translate.x = x;
 
-        if ((jobj->flags & 0x02000000) == 0) {
-            if (jobj == NULL) {
-                goto skip_x;
-            }
-            if (jobj == NULL) {
-                __assert(&un_804D5A78, 0x234, &un_804D5A80);
-            }
-            {
-                u32 flags = jobj->flags;
-                s32 skip = 0;
-                if ((flags & 0x800000) == 0 && (flags & 0x40)) {
-                    skip = 1;
-                }
-                if (skip == 0) {
-                    HSD_JObjSetMtxDirtySub(jobj);
-                }
+        if (parent != NULL) {
+            HSD_JObj* parentJobj = ((HSD_GObj*) parent)->hsd_obj;
+            HSD_JObj* child;
+            HSD_JObjAddChild(parentJobj, jobj);
+            child = parentJobj->child;
+            while (child->next != NULL) {
+                child = child->next;
             }
+            return child;
+        } else {
+            HSD_GObj* gobj;
+            gobj = GObj_Create(6, 7, 0);
+            HSD_GObjObject_80390A70(gobj, (u8) HSD_GObj_804D7849, jobj);
+            GObj_SetupGXLink(gobj, HSD_GObj_JObjCallback, 0x39, 0);
+            return (HSD_JObj*) gobj;
         }
-    skip_x:
+    }
+}
 
-        // Set Y
-        if (jobj == NULL) {
-            __assert(&un_804D5A78, 0x3B3, &un_804D5A80);
-        }
-        jobj->translate.y = y;
+void un_80313774(void)
+{
+    TyListState* state = (TyListState*) un_804A2AC0;
+    TyArchiveData* archive = un_804D6ED8;
+    ToyGlobalsS_* disp = un_804D6EE0;
+    TyListArg* entry;
+    HSD_JObj* root_jobj;
+    f32 step;
+    f32 pos;
+    s16 idx;
+    s32 i;
 
-        if ((jobj->flags & 0x02000000) == 0) {
-            if (jobj == NULL) {
-                goto skip_y;
-            }
-            if (jobj == NULL) {
-                __assert(&un_804D5A78, 0x234, &un_804D5A80);
-            }
-            {
-                u32 flags = jobj->flags;
-                s32 skip = 0;
-                if ((flags & 0x800000) == 0 && (flags & 0x40)) {
-                    skip = 1;
-                }
-                if (skip == 0) {
-                    HSD_JObjSetMtxDirtySub(jobj);
-                }
-            }
-        }
-    skip_y:
+    state->selectedIdx = un_804A284C[0x12A];
+    state->x29B = un_804A284C[1];
+    state->x29C = un_804A284C[2];
+    state->x2B8 = un_804A284C[3];
+    *(u8*) ((u8*) state + 0x2B9) = un_804A284C[3];
 
-        // Set Z
-        if (jobj == NULL) {
-            __assert(&un_804D5A78, 0x3C2, &un_804D5A80);
+    if (un_GetTrophyTotal() <= 0xA) {
+        state->entryCount = un_GetTrophyTotal() + 2;
+    } else {
+        state->entryCount = 0xC;
+    }
+
+    idx = un_803062BC(disp->x140->x10);
+    for (i = 0; i < state->entryCount; i++) {
+        entry = &state->entries[i];
+        if (i == 0) {
+            entry->links[0] = &state->entries[state->entryCount - 1];
+        } else {
+            entry->links[0] = &state->entries[i - 1];
+        }
+        if (i == state->entryCount - 1) {
+            entry->links[1] = &state->entries[0];
+        } else {
+            entry->links[1] = &state->entries[i + 1];
         }
-        jobj->translate.z = z;
+    }
 
-        if ((jobj->flags & 0x02000000) == 0) {
-            if (jobj == NULL) {
-                goto skip_z;
-            }
-            if (jobj == NULL) {
-                __assert(&un_804D5A78, 0x234, &un_804D5A80);
-            }
-            {
-                u32 flags = jobj->flags;
-                s32 skip = 0;
-                if ((flags & 0x800000) == 0 && (flags & 0x40)) {
-                    skip = 1;
-                }
-                if (skip == 0) {
-                    HSD_JObjSetMtxDirtySub(jobj);
-                }
-            }
+    state->x278 = &state->entries[1];
+    state->x270 = &state->entries[0];
+    state->x274 = &state->entries[state->entryCount - 1];
+
+    state->gobj =
+        (HSD_GObj*) un_80313508(NULL, un_803FE880 + 0xAC, 0.0f, 0.0f, 0.0f);
+    root_jobj = (HSD_JObj*) state->gobj;
+    HSD_ASSERTMSG(0x3E1, root_jobj != NULL, "jobj");
+
+    step = 5.11f - root_jobj->translate.y;
+    pos = -step;
+    state->x2A8 = step;
+
+    entry = &state->entries[0];
+    for (i = 0; i < state->entryCount; i++) {
+        entry->x28 = i;
+        entry->jobjs[0] =
+            un_80313508(state->gobj, un_803FE880 + 0xCC, 0.0f, pos, 0.0f);
+        entry->jobjs[2] =
+            entry->jobjs[0] != NULL ? entry->jobjs[0]->child : NULL;
+        un_80306A48(entry->jobjs[0], NULL, un_803FE880 + 0xE8, NULL,
+                    archive->data, 0);
+        entry->texts[0] = HSD_SisLib_803A5ACC(0, un_804D6EE8, 0.0f, 0.0f,
+                                              17.2f, 640.0f, 64.0f);
+        entry->texts[1] = HSD_SisLib_803A5ACC(0, un_804D6EE8, 0.0f, 0.0f,
+                                              17.2f, 64.0f, 64.0f);
+        entry->texts[2] = HSD_SisLib_803A5ACC(0, un_804D6EE8, 0.0f, 0.0f,
+                                              17.2f, 192.0f, 64.0f);
+        entry->x30 = pos;
+        entry->x24 = i - 1;
+        entry->idx = idx;
+        idx++;
+        if (idx >= un_GetTrophyTotal()) {
+            idx = 0;
         }
-    skip_z:;
+        pos += step;
+        entry = entry->links[1];
     }
 
-    if (parent != NULL) {
-        HSD_JObj* parentJobj = ((HSD_GObj*) parent)->hsd_obj;
-        HSD_JObj* child;
-        HSD_JObjAddChild(parentJobj, jobj);
-        child = parentJobj->child;
-        while (child->next != NULL) {
-            child = child->next;
+    state->selectedIdx = state->x278->idx;
+    state->jobj = un_80313508(state->gobj, un_803FE880 + 0x10C, 0.0f,
+                              state->x278->x30, 0.0f);
+    state->x288 = un_80313508(state->gobj, un_803FE880 + 0x12C, 0.0f,
+                              state->entries[state->x2B8].x30, 0.0f);
+
+    for (i = 0; i < state->x2B8; i++) {
+        un_80313358(state, 0, 1, 1);
+        un_8031305C(NULL, state, 0);
+    }
+
+    entry = &state->entries[0];
+    for (i = 0; i < state->entryCount; i++) {
+        if (entry->x24 == state->x2B8) {
+            state->selectedIdx = entry->idx;
+            state->x278 = entry;
+            break;
         }
-    } else {
-        HSD_GObj* gobj;
-        gobj = GObj_Create(6, 7, 0);
-        HSD_GObjObject_80390A70(gobj, (u8) HSD_GObj_804D7849, jobj);
-        GObj_SetupGXLink(gobj, HSD_GObj_JObjCallback, 0x39, 0);
+        entry++;
     }
 
-    return jobj;
+    un_80312BAC(state, state->x2B8);
 }
 
-/// #un_80313774
-
 extern s32 un_804D6EE8;
 extern s32 un_804D6EEC;
 extern f32 un_804DDE44;
@@ -795,7 +833,6 @@ extern f32 un_804DDE98;
 extern char un_803FE5E8[];
 extern char un_804D5A88[3];
 
-s8 un_8031305C(void* a, TyListState* state, s8 movedFlag);
 void un_80312BAC(TyListState* state, s8 arg1);
 
 void fn_80313BD8(HSD_GObj* gobj)
@@ -945,7 +982,7 @@ void fn_80313BD8(HSD_GObj* gobj)
         if (un_80305C44() & 0x400) {
             HSD_JObjSetFlagsAll(state->jobj, 0x10);
             if (state->entries[0].idx == 0 ||
-                state->entries[0].x0->idx + 9 < un_GetTrophyTotal())
+                state->entries[0].links[0]->idx + 9 < un_GetTrophyTotal())
             {
                 un_80313358(state, 9, 4, 0);
             } else {
@@ -961,7 +998,7 @@ void fn_80313BD8(HSD_GObj* gobj)
             HSD_JObjSetFlagsAll(state->jobj, 0x10);
             idx = state->entries[0].idx;
             if (idx == un_GetTrophyTotal() - 1 ||
-                state->entries[0].x4->idx - 9 > 0)
+                state->entries[0].links[1]->idx - 9 > 0)
             {
                 un_80313358(state, 9, 4, 1);
             } else {
@@ -995,7 +1032,7 @@ void fn_80313BD8(HSD_GObj* gobj)
         for (; i < (s8) state->entryCount; i++, p++) {
             if (((s8*) p)[0x24] == g[0xC]) {
                 state->selectedIdx = p->idx;
-                state->entries[10].x0 = (TyListArg*) p;
+                state->entries[10].links[0] = (TyListArg*) p;
                 lbAudioAx_80024030(2);
                 jobj = state->jobj;
                 ftmp = p->x30;
@@ -1066,36 +1103,35 @@ void fn_8031438C(HSD_GObj* arg0)
     TyListGobjEntry* entry = (TyListGobjEntry*) &state->gobj_2AC;
     TyListArchive* archive = un_804D6ED8;
     s32 i;
-    HSD_GObj* tmp;
+    HSD_Text* new_var;
+
     PAD_STACK(0x18);
 
-    if (((s8*) state)[0x2C2] != 0) {
+    if (state->x2C2 != 0) {
         if (entry->x16 > 1) {
             for (i = 0; i < 3; i++) {
-                u8* base_p = (u8*) archive + i * 4;
                 if (i == (s8) state->x29B) {
-                    HSD_JObjReqAnim(*(HSD_JObj**) (base_p + 0x18),
-                                    un_804DDE44);
+                    HSD_JObjReqAnim(archive->jobjs[i], un_804DDE44);
                 } else {
-                    HSD_JObjReqAnim(*(HSD_JObj**) (base_p + 0x18),
-                                    un_804DDE48);
+                    HSD_JObjReqAnim(archive->jobjs[i], un_804DDE48);
                 }
                 HSD_AObjSetRate(archive->jobjs[0]->u.dobj->mobj->tobj->aobj,
                                 un_804DDE48);
             }
             HSD_JObjAnimAll((HSD_JObj*) archive->x0[10]);
         } else {
-            tmp = entry->x4;
-            if (tmp != NULL) {
-                ((s32*) tmp)[9] = 0;
-                ((s32*) tmp)[8] = 0x42100000;
+            TyListWaitData* wait_data = (TyListWaitData*) entry->x4;
+            if (wait_data != NULL) {
+                wait_data->x24 = 0;
+                wait_data->x20 = 0x42100000;
             }
             state->x290 = HSD_SisLib_803A6754(3, un_804D6EEC);
             state->x290->pos_z = un_804DDE68;
+            new_var = state->x290;
             state->x290->font_size.x = un_804DDE40;
-            state->x290->font_size.y = un_804DDE34;
+            new_var->font_size.y = un_804DDE34;
             state->x290->default_kerning = 1;
-            state->x290->default_alignment = 2;
+            state->x290->default_alignment = (double) 2;
             HSD_SisLib_803A6B98(state->x290, un_804DDE88, un_804DDE8C,
                                 un_804D5A88, un_GetTrophyTotal());
         }
@@ -1181,14 +1217,17 @@ void un_8031457C(void)
 
 void un_803147C4(void)
 {
-    TyListState* state = (TyListState*) un_804A2AC0;
+    char* data = un_804A2AC0;
     char* strs = un_803FE880;
     TyArchiveData* archive;
+    LightList** jobj;
+    HSD_GObj** gobj;
     PAD_STACK(8);
 
-    memzero(&state->gobj_2AC, 0x18);
+    memzero(data + 0x2AC, 0x18);
     un_8031457C();
-    memzero(&state->gobj_2C4, 0x14);
+    gobj = (HSD_GObj**) (data + 0x2C4);
+    memzero(gobj, 0x14);
 
     archive = un_804D6ED8;
 
@@ -1197,20 +1236,17 @@ void un_803147C4(void)
         OSPanic(strs + 0x70, 0x636, un_804D5A8C);
     }
 
-    {
-        LightList** jobj =
-            HSD_ArchiveGetPublicAddress(archive->data, strs + 0x170);
-        if (jobj != NULL) {
-            state->gobj_2C4 = GObj_Create(2, 3, 0);
-            HSD_GObjObject_80390A70(state->gobj_2C4, (u8) HSD_GObj_804D784A,
-                                    Toy_LoadLObjList(jobj, 0));
-            GObj_SetupGXLink(state->gobj_2C4, HSD_GObj_LObjCallback, 0x34, 0);
-        }
+    jobj = HSD_ArchiveGetPublicAddress(archive->data, strs + 0x170);
+    if (jobj != NULL) {
+        *gobj = GObj_Create(2, 3, 0);
+        HSD_GObjObject_80390A70(*gobj, (u8) HSD_GObj_804D784A,
+                                Toy_LoadLObjList(jobj, 0));
+        GObj_SetupGXLink(*gobj, HSD_GObj_LObjCallback, 0x34, 0);
     }
 
     un_80307470(0);
     if (un_GetTrophyTotal() != 0) {
-        memzero(state, 0x2AC);
+        memzero(data, 0x2AC);
         un_80313774();
     }
     HSD_PadRenewStatus();
diff --git a/src/melee/ty/tylist.h b/src/melee/ty/tylist.h
index 44be0a74c6..b02a6cfa46 100644
--- a/src/melee/ty/tylist.h
+++ b/src/melee/ty/tylist.h
@@ -18,7 +18,7 @@ struct TyListState;
 /* 312904 */ void un_80312904(void*, s8);
 /* 312BAC */ void un_80312BAC(struct TyListState* state, s8 arg1);
 /* 312E88 */ void un_80312E88(struct TyListArg* arg, float delta);
-/* 31305C */ s8 un_8031305C(void* a, struct TyListState* state, s8 movedFlag);
+/* 31305C */ s32 un_8031305C(void* a, struct TyListState* state, s8 movedFlag);
 /* 313358 */ void un_80313358(struct TyListState* state, s8 arg2, s8 arg3,
                               s8 arg4);
 /* 313464 */ void un_80313464(struct TyListArg* arg);
diff --git a/src/melee/ty/types.h b/src/melee/ty/types.h
index 24cd2ee9dc..849d9fd53c 100644
--- a/src/melee/ty/types.h
+++ b/src/melee/ty/types.h
@@ -61,17 +61,14 @@ typedef struct TrophyData {
 
 /* Trophy list entry. Size: 0x34 bytes. */
 typedef struct TyListArg {
-    /* 0x00 */ struct TyListArg* x0;
-    /* 0x04 */ struct TyListArg* x4;
-    /* 0x08 */ void* x8;
-    /* 0x0C */ struct HSD_JObj* xC;
-    /* 0x10 */ struct HSD_JObj* x10;
-    /* 0x14 */ struct HSD_JObj* x14;
-    /* 0x18 */ u8 pad_18[0x24 - 0x18];
+    /* 0x00 */ struct TyListArg* links[3];
+    /* 0x0C */ struct HSD_JObj* jobjs[3];
+    /* 0x18 */ struct HSD_Text* texts[3];
     /* 0x24 */ s8 x24;
-    /* 0x25 */ u8 pad_25;
+    /* 0x25 */ u8 x25;
     /* 0x26 */ s16 idx;
-    /* 0x28 */ u8 pad_28[0x2C - 0x28];
+    /* 0x28 */ s8 x28;
+    /* 0x29 */ u8 pad_29[0x2C - 0x29];
     /* 0x2C */ float x2C;
     /* 0x30 */ float x30;
 } TyListArg;
@@ -104,7 +101,9 @@ typedef struct TyListState {
     /* 0x2B0 */ struct HSD_GObj* gobj_2B0;
     /* 0x2B4 */ u8 pad_2B4[4];
     /* 0x2B8 */ u8 x2B8;
-    /* 0x2B9 */ u8 pad_2B9[0xB];
+    /* 0x2B9 */ u8 pad_2B9[9];
+    /* 0x2C2 */ s8 x2C2;
+    /* 0x2C3 */ u8 pad_2C3;
     /* 0x2C4 */ struct HSD_GObj* gobj_2C4;
     /* 0x2C8 */ u8 pad_2C8[0x10];
 } TyListState;
@@ -137,4 +136,30 @@ typedef struct TyDspEntry {
     /* 0x0C */ f32 x0C;
 } TyDspEntry;
 STATIC_ASSERT(sizeof(struct TyDspEntry) == 0x10);
+
+typedef struct ToySubStructS_ {
+    u8 pad0[0x10];
+    s16 x10;
+} ToySubStructS_;
+
+typedef struct ToyGlobalsS_ {
+    HSD_GObj* x0;
+    u8 x4;
+    HSD_GObj* x8;
+    HSD_GObj* xC;
+    s32 x10;
+    u8 pad14[0x1C];
+    void* x30;
+    u8 pad34[0x1C];
+    void* x50;
+    void* x54;
+    s32 x58;
+    u8 pad0[0x140 - 0x5C];
+    ToySubStructS_* x140;
+    void* x144;
+    void* x148;
+    void* x14C;
+    void* x150;
+    s16 x154;
+} ToyGlobalsS_;
 #endif
