diff --git a/src/melee/gm/gmregclear.c b/src/melee/gm/gmregclear.c
index 813d1f5ade..033736e8f7 100644
--- a/src/melee/gm/gmregclear.c
+++ b/src/melee/gm/gmregclear.c
@@ -48,7 +48,9 @@ lbl_804706D8_t lbl_804706D8[12];
 struct lbl_80472D28_t {
     /*   +0 */ char pad_0[0x104];
     /* +104 */ int x104;
-    /* +108 */ char pad_108[0xE];
+    /* +108 */ char pad_108[0x2];
+    /* +10A */ s16 x10A;
+    /* +10C */ char pad_10C[0xA];
     /* +116 */ u8 x116;
 };
 
@@ -655,7 +657,33 @@ u8 fn_8017DD7C(PlayerInitData* arg0, Unk1PData_x24* arg1)
     return index - 1;
 }
 
-/// #fn_8017DE54
+s32 fn_8017DE54(u8 arg0, u8* arg1)
+{
+    u8* p;
+    s32 count;
+
+    if (arg0 & 0x20) {
+        return 0;
+    }
+    if (arg0 & 0x10) {
+        count = 0;
+        if ((s32) arg1[0] != 0x21) {
+            count = 1;
+        }
+        p = &arg1[1];
+        if ((s32) *p != 0x21) {
+            count += 1;
+        }
+        if ((s32) p[1] != 0x21) {
+            count += 1;
+        }
+        return 3 - count;
+    }
+    if (arg0 & 2) {
+        return 2;
+    }
+    return 0;
+}
 
 Unk1PData* fn_8017DEC8(int arg0)
 {
@@ -1008,7 +1036,19 @@ int fn_8017F008(void)
 
 /// #fn_8017F09C
 
-/// #fn_8017F14C
+s32 fn_8017F14C(void* arg0)
+{
+    struct lbl_80472D28_t* data = &lbl_80472D28;
+    if (gm_8016AE38()->x24C8.x4_7) {
+        s32 val = *(s32*) ((u8*) arg0 + 0x98);
+        s32 ret = val;
+        if (val > 9999) {
+            ret = 9999;
+        }
+        return ret * data->x10A;
+    }
+    return 0;
+}
 
 /// #fn_8017F1B8
 
diff --git a/src/melee/gm/gmregclear.h b/src/melee/gm/gmregclear.h
index 1616aaf702..9b857cf32d 100644
--- a/src/melee/gm/gmregclear.h
+++ b/src/melee/gm/gmregclear.h
@@ -30,7 +30,7 @@
 /* 17DB78 */ u8 gm_8017DB78(gm_8017DB6C_arg0_t* arg0, int index);
 /* 17DB88 */ UNK_RET gm_8017DB88(UNK_PARAMS);
 /* 17DD7C */ u8 fn_8017DD7C(PlayerInitData* arg0, Unk1PData_x24* arg1);
-/* 17DE54 */ UNK_RET fn_8017DE54(UNK_PARAMS);
+/* 17DE54 */ s32 fn_8017DE54(u8, u8*);
 /* 17DEC8 */ Unk1PData* fn_8017DEC8(int);
 /* 17DF28 */ Unk1PData* fn_8017DF28(void);
 /* 17DF90 */ u8 fn_8017DF90(void);
@@ -78,7 +78,7 @@
 /* 17EE40 */ void fn_8017EE40(int);
 /* 17F008 */ int fn_8017F008(void);
 /* 17F09C */ UNK_RET fn_8017F09C(UNK_PARAMS);
-/* 17F14C */ UNK_RET fn_8017F14C(UNK_PARAMS);
+/* 17F14C */ s32 fn_8017F14C(void* arg0);
 /* 17F1B8 */ UNK_RET fn_8017F1B8(UNK_PARAMS);
 /* 17F2A4 */ UNK_RET fn_8017F2A4(UNK_PARAMS);
 /* 17F47C */ UNK_RET fn_8017F47C(UNK_PARAMS);
