diff --git a/src/melee/gr/gricemt.c b/src/melee/gr/gricemt.c
index f0c2b06697..95e36e5198 100644
--- a/src/melee/gr/gricemt.c
+++ b/src/melee/gr/gricemt.c
@@ -29,10 +29,10 @@
 
 /* 1F8C64 */ static void fn_801F8C64(Item_GObj* gobj, Ground* u1, Vec3* u2,
                                      HSD_GObj* u3, f32 u4);
-/* 1F91EC */ static void FUN_801f91ec(HSD_GObj* param_1, s16* param_2,
-                                      int param_3, int param_4, int param_5,
-                                      int param_6, int param_7,
-                                      mpLib_Callback param_8);
+/* 1F91EC */ void grIceMt_801F91EC(HSD_GObj* param_1, s16* param_2,
+                                   int param_3, int param_4, int param_5,
+                                   int param_6, int param_7,
+                                   mpLib_Callback param_8);
 
 /// StageData grIm_803E8974 = {
 ///     45,
@@ -985,10 +985,9 @@ HSD_GObj* fn_801F91A8(HSD_GObj* gobj)
     return (HSD_GObj*) result;
 }
 
-/// #grIceMt_801F91EC
-void FUN_801f91ec(HSD_GObj* param_1, s16* param_2, int param_3, int param_4,
-                  int param_5, int param_6, int param_7,
-                  mpLib_Callback param_8)
+void grIceMt_801F91EC(HSD_GObj* param_1, s16* param_2, int param_3,
+                      int param_4, int param_5, int param_6, int param_7,
+                      mpLib_Callback param_8)
 {
     Ground* gp = GET_GROUND(param_1);
     param_2[0] = 0;
@@ -998,9 +997,9 @@ void FUN_801f91ec(HSD_GObj* param_1, s16* param_2, int param_3, int param_4,
     param_2[4] = param_5;
     param_2[5] = param_6;
     param_2[6] = param_7;
-    grAnime_801C7A04(param_1, param_3, 7, -20);
+    grAnime_801C7A04(param_1, param_3, 7, 0);
     if (param_4 != -1) {
-        grAnime_801C7A04(param_1, param_4, 7, -20);
+        grAnime_801C7A04(param_1, param_4, 7, 0);
     }
     mpJointSetCb1(param_5, gp, param_8);
 }
diff --git a/src/melee/gr/gricemt.h b/src/melee/gr/gricemt.h
index 2c46c1665f..0910cc2aa3 100644
--- a/src/melee/gr/gricemt.h
+++ b/src/melee/gr/gricemt.h
@@ -125,7 +125,8 @@ extern f32 grIm_804DB570;
 /* 1F9150 */ IceMountainParams* fn_801F9150(HSD_GObj*);
 /* 1F91A4 */ void fn_801F91A4(void);
 /* 1F91A8 */ HSD_GObj* fn_801F91A8(HSD_GObj*);
-/* 1F91EC */ UNK_RET grIceMt_801F91EC(HSD_GObj*);
+/* 1F91EC */ void grIceMt_801F91EC(HSD_GObj*, s16*, int, int, int, int, int,
+                                   mpLib_Callback);
 /* 1F929C */ void grIceMt_801F929C(HSD_GObj* arg0, void* arg1);
 /* 1F9338 */ void fn_801F9338(Ground*, int, CollData*, s32, mpLib_GroundEnum,
                               float);
diff --git a/src/melee/gr/grkongo.c b/src/melee/gr/grkongo.c
index 1ee2bc47c5..185ecbf0f4 100644
--- a/src/melee/gr/grkongo.c
+++ b/src/melee/gr/grkongo.c
@@ -13,6 +13,7 @@
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
 #include "gr/stage.h"
+#include "it/it_26B1.h"
 #include "it/items/itklap.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
@@ -1433,7 +1434,37 @@ void grKongo_801D8058(Ground_GObj* arg)
     Ground_801C4A08(arg);
 }
 
-/// #grKongo_801D8078
+HSD_GObj* grKongo_801D8078(HSD_GObj* gobj)
+{
+    Vec3 pos;
+    f32 unk;
+    Vec3 item_pos;
+    HSD_GObj* cur;
+
+    Ground_801C4DA0(&pos, &unk);
+
+    for (cur = HSD_GObj_Entities->items; cur != NULL; cur = cur->next) {
+        if (itGetKind(cur) == 0xDA) {
+            f32 dx, dy, dz, dx2, dy2, dz2, r;
+
+            it_8026B294(cur, &item_pos);
+
+            dx = pos.x - item_pos.x;
+            dy = pos.y - item_pos.y;
+            dz = pos.z - item_pos.z;
+            dx2 = dx * dx;
+            dy2 = dy * dy;
+            dz2 = dz * dz;
+            r = grKg_804D6980->unk28;
+
+            if (dx2 + dy2 + dz2 < r * r) {
+                return cur;
+            }
+        }
+    }
+
+    return NULL;
+}
 
 /// #fn_801D8134
 
