diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index 559bf2af75..fd7b85cff9 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -1264,8 +1264,8 @@ void ftKb_SpecialLw1_Coll(Fighter_GObj* gobj)
         fp->self_vel.x = 0.0f;
         ftCommon_8007D5D4(fp);
         temp = fp->x221C_b4;
-        Fighter_ChangeMotionState(gobj, 0x18C, 0x0C4C5092U,
-                                  fp->cur_anim_frame, 1.0f, 0.0f, NULL);
+        Fighter_ChangeMotionState(gobj, 0x18C, 0x0C4C5092U, fp->cur_anim_frame,
+                                  1.0f, 0.0f, NULL);
         fp->x221C_b4 = temp;
         fp3 = GET_FIGHTER(gobj);
         fp3->take_dmg_cb = ftKb_Init_800EE7B8;
@@ -1371,8 +1371,8 @@ void ftKb_SpecialAirLwStart_Coll(Fighter_GObj* gobj)
         fp->self_vel.y = 0.0f;
         ftCommon_8007D7FC(fp);
         temp = fp->x221C_b4;
-        Fighter_ChangeMotionState(gobj, 0x189, 0x0C4C5092U,
-                                  fp->cur_anim_frame, 1.0f, 0.0f, NULL);
+        Fighter_ChangeMotionState(gobj, 0x189, 0x0C4C5092U, fp->cur_anim_frame,
+                                  1.0f, 0.0f, NULL);
         fp->x221C_b4 = temp;
         fp2 = GET_FIGHTER(gobj);
         fp2->take_dmg_cb = ftKb_Init_800EE7B8;
@@ -1404,8 +1404,7 @@ void ftKb_SpecialAirLwStart_Coll(Fighter_GObj* gobj)
             angle = -angle;
         }
         if (angle >= da->speciallw_min_slant_angle_slide) {
-            fp->gr_vel =
-                fp->mv.co.common.x18 * da->speciallw_slide_max_speed;
+            fp->gr_vel = fp->mv.co.common.x18 * da->speciallw_slide_max_speed;
         }
     } else {
         fp2 = GET_FIGHTER(gobj);
@@ -1482,8 +1481,7 @@ void ftKb_SpecialAirLw_Coll(Fighter_GObj* gobj)
             angle = -angle;
         }
         if (angle >= da->speciallw_min_slant_angle_slide) {
-            fp->gr_vel =
-                fp->mv.co.common.x18 * da->speciallw_slide_max_speed;
+            fp->gr_vel = fp->mv.co.common.x18 * da->speciallw_slide_max_speed;
         }
     } else {
         fp2 = GET_FIGHTER(gobj);
@@ -2600,9 +2598,8 @@ void ftKb_SpecialNSpit0_Anim(Fighter_GObj* gobj)
 
     if (fp->cmd_vars[0] != 0 && (item_gobj = fp->target_item_gobj) != NULL) {
         ftCommon_8007E2F4(fp, 0);
-        lb_8000B1CC(
-            fp->parts[ftParts_GetBoneIndex(fp, FtPart_TransN2)].joint, NULL,
-            &attr.pos);
+        lb_8000B1CC(fp->parts[ftParts_GetBoneIndex(fp, FtPart_TransN2)].joint,
+                    NULL, &attr.pos);
         {
             f32 neg_facing = -fp->facing_dir;
             da = getFighter(gobj)->dat_attrs;
@@ -2661,19 +2658,18 @@ void ftKb_SpecialNSpit1_Anim(Fighter_GObj* gobj)
 
     if (fp->cmd_vars[0] != 0 && (item_gobj = fp->target_item_gobj) != NULL) {
         ftCommon_8007E2F4(fp, 0);
-        lb_8000B1CC(
-            fp->parts[ftParts_GetBoneIndex(fp, FtPart_TransN2)].joint, NULL,
-            &attr.pos);
+        lb_8000B1CC(fp->parts[ftParts_GetBoneIndex(fp, FtPart_TransN2)].joint,
+                    NULL, &attr.pos);
         {
             f32 neg_facing = -fp->facing_dir;
             da = getFighter(gobj)->dat_attrs;
             attr.vel.x = -neg_facing *
-                        da->specialn_ground_spit_initial_horizontal_velocity;
+                         da->specialn_ground_spit_initial_horizontal_velocity;
             attr.vel.z = 0.0f;
             attr.vel.y = 0.0f;
             attr.float2 = da->specialn_spit_deceleration_rate;
             attr.float1 = ((ftKb_DatAttrs*) getFighter(gobj)->dat_attrs)
-                            ->specialn_star_base_duration;
+                              ->specialn_star_base_duration;
             it_802F28C8(item_gobj, 0, it_802F295C(gobj, &attr));
         }
         fp->x1A64 = NULL;
@@ -2887,7 +2883,8 @@ void ftKb_EatWait_IASA(Fighter_GObj* gobj)
         if (((fp->input.x668 & 0x200) && fp->target_item_gobj != NULL) ||
             ((fp->input.lstick.y <
               -((ftKb_DatAttrs*) fp->dat_attrs)->specialn_y_axis_range_jump) &&
-             fp->target_item_gobj != NULL)) {
+             fp->target_item_gobj != NULL))
+        {
             Fighter_ChangeMotionState(gobj, 0x170, 2, 0.0f, 1.0f, 0.0f, NULL);
             r29 = 1;
             fp->x2222_b2 = 1;
@@ -2919,7 +2916,8 @@ void ftKb_EatWait_IASA(Fighter_GObj* gobj)
         if (((fp->input.x668 & 0x200) && fp->victim_gobj != NULL) ||
             ((fp->input.lstick.y <
               -((ftKb_DatAttrs*) fp->dat_attrs)->specialn_y_axis_range_jump) &&
-             fp->victim_gobj != NULL)) {
+             fp->victim_gobj != NULL))
+        {
             Fighter_ChangeMotionState(gobj, 0x16F, 2, 0.0f, 1.0f, 0.0f, NULL);
             r29_3 = 1;
             fp->x2222_b2 = 1;
@@ -2951,11 +2949,14 @@ void ftKb_EatWait_IASA(Fighter_GObj* gobj)
                 } else {
                     absX = stickX;
                 }
-                if (absX < ((ftKb_DatAttrs*) fp2->dat_attrs)->specialn_x_axis_range_walk) {
+                if (absX < ((ftKb_DatAttrs*) fp2->dat_attrs)
+                               ->specialn_x_axis_range_walk)
+                {
                     stickX = 0.0f;
                 }
                 if (((stickX < 0.0f) && (fp2->facing_dir == 1.0f)) ||
-                    ((stickX > 0.0f) && (fp2->facing_dir == -1.0f))) {
+                    ((stickX > 0.0f) && (fp2->facing_dir == -1.0f)))
+                {
                     Fighter_ChangeMotionState(gobj, 0x16B, 0x92, 0.0f, 1.0f,
                                               0.0f, NULL);
                     ftKb_SpecialN_800F9070(gobj);
@@ -2982,13 +2983,14 @@ void ftKb_EatWait_IASA(Fighter_GObj* gobj)
                     }
                     if (r0_2 == 0 && ftWalkCommon_800DFC70(gobj)) {
                         Fighter* fp4 = GET_FIGHTER(gobj);
-                        ftWalkCommon_800DFCA4(
-                            gobj, 0x168, 0x10, 0.0f, fp4->fv.kb.xE8,
-                            fp4->fv.kb.xEC, fp4->fv.kb.xF0,
-                            fp4->co_attrs.slow_walk_max,
-                            fp4->co_attrs.mid_walk_point,
-                            fp4->co_attrs.fast_walk_min,
-                            ((ftKb_DatAttrs*) fp4->dat_attrs)->specialn_walk_speed);
+                        ftWalkCommon_800DFCA4(gobj, 0x168, 0x10, 0.0f,
+                                              fp4->fv.kb.xE8, fp4->fv.kb.xEC,
+                                              fp4->fv.kb.xF0,
+                                              fp4->co_attrs.slow_walk_max,
+                                              fp4->co_attrs.mid_walk_point,
+                                              fp4->co_attrs.fast_walk_min,
+                                              ((ftKb_DatAttrs*) fp4->dat_attrs)
+                                                  ->specialn_walk_speed);
                         ftCommon_8007E2F4(fp4, 0x1FF);
                     }
                 }
@@ -3012,7 +3014,8 @@ void ftKb_SpecialAirNCaptureWait_IASA(Fighter_GObj* gobj)
         if (((fp->input.x668 & 0x200) && fp->target_item_gobj != NULL) ||
             ((fp->input.lstick.y <
               -((ftKb_DatAttrs*) fp->dat_attrs)->specialn_y_axis_range_jump) &&
-             fp->target_item_gobj != NULL)) {
+             fp->target_item_gobj != NULL))
+        {
             Fighter_ChangeMotionState(gobj, 0x17B, 2, 0.0f, 1.0f, 0.0f, NULL);
             r29 = 1;
             fp->x2222_b2 = 1;
@@ -3044,7 +3047,8 @@ void ftKb_SpecialAirNCaptureWait_IASA(Fighter_GObj* gobj)
         if (((fp->input.x668 & 0x200) && fp->victim_gobj != NULL) ||
             ((fp->input.lstick.y <
               -((ftKb_DatAttrs*) fp->dat_attrs)->specialn_y_axis_range_jump) &&
-             fp->victim_gobj != NULL)) {
+             fp->victim_gobj != NULL))
+        {
             Fighter_ChangeMotionState(gobj, 0x17A, 2, 0.0f, 1.0f, 0.0f, NULL);
             r29_3 = 1;
             fp->x2222_b2 = 1;
@@ -3077,11 +3081,13 @@ void ftKb_SpecialAirNCaptureWait_IASA(Fighter_GObj* gobj)
                     absX = stickX;
                 }
                 if (absX < ((ftKb_DatAttrs*) fp2->dat_attrs)
-                               ->specialn_x_axis_range_walk) {
+                               ->specialn_x_axis_range_walk)
+                {
                     stickX = 0.0f;
                 }
                 if (((stickX < 0.0f) && (fp2->facing_dir == 1.0f)) ||
-                    ((stickX > 0.0f) && (fp2->facing_dir == -1.0f))) {
+                    ((stickX > 0.0f) && (fp2->facing_dir == -1.0f)))
+                {
                     Fighter_ChangeMotionState(gobj, 0x17E, 0x92, 0.0f, 1.0f,
                                               0.0f, NULL);
                     ftKb_SpecialN_800F9070(gobj);
@@ -3548,11 +3554,11 @@ void fn_800F9260(HSD_GObj* gobj)
         }
         if (found) {
             lb_8000B1CC(
-                fp->parts[ftParts_GetBoneIndex(fp, FtPart_LHandN)].joint,
-                NULL, &sp44);
+                fp->parts[ftParts_GetBoneIndex(fp, FtPart_LHandN)].joint, NULL,
+                &sp44);
             if (fp->fv.kb.hat.kind == FTKIND_MARIO) {
-                it_8029B6F8((Item_GObj*) gobj, &sp44,
-                            It_Kind_Kirby_MarioFire, fp->facing_dir);
+                it_8029B6F8((Item_GObj*) gobj, &sp44, It_Kind_Kirby_MarioFire,
+                            fp->facing_dir);
                 bone = ftParts_GetBoneIndex(fp, FtPart_LHandN);
                 joint = fp->parts[bone].joint;
                 efSync_Spawn(0x49F, gobj, joint, fp->facing_dir);
diff --git a/src/melee/ft/ftdynamics.c b/src/melee/ft/ftdynamics.c
index 8c8ec63e3c..8490848c4a 100644
--- a/src/melee/ft/ftdynamics.c
+++ b/src/melee/ft/ftdynamics.c
@@ -362,9 +362,9 @@ void ftCo_8009D704(Fighter* fp)
 
 void ftCo_8009D81C(Fighter* fp)
 {
-    KirbyHatStruct* data = ft_80459B88.hats[FTKIND_YOSHI];
-    PAD_STACK(8);
-    fp->dynamics_num = data->hat_dynamics[3]->dynamicsNum;
+    KirbyHatStruct* hat = ft_80459B88.hats[FTKIND_YOSHI];
+    PAD_STACK(2 * 4);
+    fp->dynamics_num = hat->hat_dynamics[3]->dynamicsNum;
     if (fp->dynamics_num >= Ft_Dynamics_NumMax) {
         OSReport("fighter dynamics num over!\n");
         __assert(__FILE__, 364, "fp->dynamics_num < Ft_Dynamics_NumMax");
@@ -372,15 +372,15 @@ void ftCo_8009D81C(Fighter* fp)
     {
         ssize_t i;
         for (i = 0; i < fp->dynamics_num; i++) {
-            BoneDynamicsDesc* desc =
-                &data->hat_dynamics[3]->ftDynamicBones->array[i];
-            fp->parts[desc->bone_id].flags_b0 = true;
-            lb_8000FD48(fp->parts[desc->bone_id].joint,
-                        &fp->dynamic_bone_sets[i].dyn_desc,
-                        desc->dyn_desc.count);
-            desc->bone_id = FtPart_TopN;
+            s32 bone_id =
+                hat->hat_dynamics[3]->ftDynamicBones->array[i].bone_id;
+            fp->parts[bone_id].flags_b0 = true;
+            lb_8000FD48(
+                fp->parts[bone_id].joint, &fp->dynamic_bone_sets[i].dyn_desc,
+                hat->hat_dynamics[3]->ftDynamicBones->array[i].dyn_desc.count);
+            fp->dynamic_bone_sets[i].bone_id = FtPart_TopN;
             lb_80011710(
-                &data->hat_dynamics[3]->ftDynamicBones->array[i].dyn_desc,
+                &hat->hat_dynamics[3]->ftDynamicBones->array[i].dyn_desc,
                 &fp->dynamic_bone_sets[i].dyn_desc);
         }
     }
diff --git a/src/melee/gm/gm_1601.c b/src/melee/gm/gm_1601.c
index 2f4aadea51..4ee6f013f3 100644
--- a/src/melee/gm/gm_1601.c
+++ b/src/melee/gm/gm_1601.c
@@ -2183,31 +2183,34 @@ void gm_80163A3C(u8 arg0, u8* arg1, u8* arg2, u8* arg3, u8* arg4)
 
 bool gm_80163B9C(u8* arg0, u8* arg1, u8* arg2, u8* arg3)
 {
-    s32 i;
-    u32 total_frames;
+    u32 total_frames = 0;
     s32 frames;
-    total_frames = 0;
-    for (i = 0; i < 25; i++) {
-        if (gmMainLib_8015D710((u8) i) != 0) {
-            total_frames = *gmMainLib_8015D6F8((u8) i);
-
-            frames = total_frames / 60;
-            if (arg0 != NULL) {
-                *arg0 = (u8) ((frames / 60) / 60);
-            }
-            if (arg1 != NULL) {
-                *arg1 = (u8) ((frames / 60) % 60);
-            }
-            if (arg2 != NULL) {
-                *arg2 = (u8) (frames % 60);
-            }
-            if (arg3 != NULL) {
-                *arg3 = (u8) ((99.0f * (f32) (total_frames % 60)) / 59.0f);
-            }
-            return true;
+    s32 i;
+
+    for (i = 0; i < 0x19; i++) {
+        if (gmMainLib_8015D710(i) != 0) {
+            total_frames += *gmMainLib_8015D6F8((u8) i);
+        } else {
+            return false;
         }
     }
-    return false;
+
+    frames = total_frames / 60;
+
+    if (arg0 != NULL) {
+        *arg0 = (u8) ((frames / 60) / 60);
+    }
+    if (arg1 != NULL) {
+        *arg1 = (u8) ((frames / 60) % 60);
+    }
+    if (arg2 != NULL) {
+        *arg2 = (u8) (frames % 60);
+    }
+    if (arg3 != NULL) {
+        *arg3 = (u8) ((99.0f * (f32) (total_frames % 60)) / 59.0f);
+    }
+
+    return true;
 }
 
 bool fn_80163D24(void)
diff --git a/src/melee/gm/gmmain_lib.c b/src/melee/gm/gmmain_lib.c
index c191c272bf..4ddd4459a6 100644
--- a/src/melee/gm/gmmain_lib.c
+++ b/src/melee/gm/gmmain_lib.c
@@ -680,8 +680,8 @@ bool gmMainLib_8015D984(u32 arg0)
 
     if (gmMainLib_8015DA90(arg0) == 0) {
         u32* base = &gmMainLib_804D3EE0->unk_6C[0];
-        u32* qwe = &base[arg0];
-        *qwe = lbTime_8000AFBC();
+        u32* temp_r31 = &base[arg0];
+        *temp_r31 = lbTime_8000AFBC();
 
         gmMainLib_8015D9F4(arg0);
         gmMainLib_8015DA40(arg0);
diff --git a/src/melee/gr/granime.c b/src/melee/gr/granime.c
index dad7e9541a..d01b797249 100644
--- a/src/melee/gr/granime.c
+++ b/src/melee/gr/granime.c
@@ -1046,32 +1046,9 @@ bool grAnime_801C83D0(HSD_GObj* gobj, bool arg1, enum_t arg2)
 
 bool grAnime_801C84A4(HSD_GObj* gobj, s32 arg1, s32 arg2)
 {
-    enum _HSD_TypeMask mask = 0;
-    HSD_JObj* jobj;
-    u8 _[8];
-    HSD_AObj* sp14 = NULL;
-    HSD_AObj* result;
-
-    jobj = Ground_801C3FA4(gobj, arg1);
-    if (jobj == NULL) {
-        result = NULL;
-    } else {
-        if (arg2 & 1) {
-            mask |= 0x220;
-        }
-        if (arg2 & 2) {
-            mask |= 0x7484;
-        }
-        if (arg2 & 4) {
-            mask |= 0x100;
-        }
-        if (!__setjmp(&grAnime_8049EE40)) {
-            HSD_ForeachAnim(jobj, JOBJ_TYPE, mask, fn_801C82E8, AOBJ_ARG_AV,
-                            &sp14);
-        }
-        result = sp14;
-    }
-    if (HSD_AObjGetFlags(result) & 0x04000000) {
+    HSD_AObj* aobj = grAnime_801C8318(gobj, arg1, arg2);
+    PAD_STACK(8);
+    if (HSD_AObjGetFlags(aobj) & 0x04000000) {
         return true;
     }
     return false;
diff --git a/src/melee/gr/grkongo.c b/src/melee/gr/grkongo.c
index c9b5f0d951..a2c6c6f743 100644
--- a/src/melee/gr/grkongo.c
+++ b/src/melee/gr/grkongo.c
@@ -177,6 +177,7 @@ void grKongo_801D55D8(Ground_GObj* arg0)
     Vec3 sp14;
     Ground* temp_r31;
     void* temp_r28;
+    PAD_STACK(8);
 
     temp_r31 = arg0->user_data;
     temp_r28 = arg0->hsd_obj;
@@ -194,9 +195,9 @@ void grKongo_801D55D8(Ground_GObj* arg0)
     temp_r31->gv.kongo3.xE0 = 0.0f;
     temp_r31->gv.kongo.xE8 = grKg_804D6980->unk60;
     temp_r31->gv.kongo2.xCC =
-        random_adder_f(grKg_804D6980->unk2C, grKg_804D6980->unk30);
+        rand_range((s32) grKg_804D6980->unk30, (s32) grKg_804D6980->unk2C);
     temp_r31->gv.kongo2.xCE =
-        random_adder(grKg_804D6980->unk68, grKg_804D6980->unk64);
+        rand_range(grKg_804D6980->unk68, grKg_804D6980->unk64);
     lb_8000B1CC(Ground_801C3FA4(arg0, 1), NULL, &sp14);
     Ground_801C4D70(arg0, &sp14, temp_r31->gv.kongo.xD8);
 #if 0
diff --git a/src/melee/it/itcoll.c b/src/melee/it/itcoll.c
index 79ed0a75cc..dbf3cc6a85 100644
--- a/src/melee/it/itcoll.c
+++ b/src/melee/it/itcoll.c
@@ -1,6 +1,8 @@
 #include "it/itcoll.h"
 
 #include "inlines.h"
+#include "it_2725.h"
+#include "it_279C.h"
 #include "placeholder.h"
 #include "types.h"
 
@@ -18,11 +20,9 @@
 #include "gm/gm_unsplit.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
-#include "it/it_3F14.h"
 #include "it/it_279C.h"
+#include "it/it_3F14.h"
 #include "it/item.h"
-#include "it_2725.h"
-#include "it_279C.h"
 #include "lb/lb_00B0.h"
 #include "lb/lbaudio_ax.h"
 #include "lb/lbcollision.h"
@@ -425,335 +425,268 @@ static inline void it_8026F9AC_outline(s32 arg0, void* fighter,
 
 void it_802703E8(Item_GObj* arg_item_gobj)
 {
+    Item* arg_item;
     HSD_GObj* fighter_gobj;
+    Fighter* fighter;
+    u32 hit_index;
     HitCapsule* hit;
-    HurtCapsule* hurt;
-    s32 dmg;
     ItemKind kind;
-    HSD_GObj* arg_item_owner_gobj;
+    s32 dmg;
     u32 hurt_index;
-    u32 hit_index;
     u32 ft_team;
-    Item* arg_item;
-    Fighter* fighter;
     PAD_STACK(16);
 
     arg_item = arg_item_gobj->user_data;
-    if (arg_item->xAC8_hurtboxNum != 0) {
-        fighter_gobj = HSD_GObj_Entities->fighters;
-        while (fighter_gobj != NULL) {
-            arg_item_owner_gobj = arg_item->owner;
-            if (((arg_item_owner_gobj != fighter_gobj) ||
-                 arg_item->xDCE_flag.b0))
+    if (arg_item->xAC8_hurtboxNum == 0) {
+        return;
+    }
+    for (fighter_gobj = HSD_GObj_Entities->fighters; fighter_gobj != NULL;
+         fighter_gobj = fighter_gobj->next)
+    {
+        if (((arg_item->owner == fighter_gobj) && !arg_item->xDCE_flag.b0)) {
+            continue;
+        }
+        fighter = fighter_gobj->user_data;
+        if (!((fighter->x1064_thrownHitbox.owner == NULL) ||
+              (fighter->x1064_thrownHitbox.owner != arg_item->owner) ||
+              arg_item->xDCE_flag.b0))
+        {
+            continue;
+        }
+        if (fighter->x1064_thrownHitbox.owner != NULL) {
+            ft_team = ftLib_80086EB4(fighter->x1064_thrownHitbox.owner) & 0xFF;
+        } else {
+            ft_team = fighter->team;
+        }
+        if (gm_8016B168() && !gm_8016B0D4() && !arg_item->xDCE_flag.b1 &&
+            (ft_team == arg_item->x20_team_id))
+        {
+            continue;
+        }
+        for (hit_index = 0; hit_index < 4; hit_index++) {
+            hit = &fighter->x914[hit_index];
+            if ((hit->state == HitCapsule_Disabled) ||
+                (hit->element == HitElement_Catch) || (hit->x42_b7 != 1) ||
+                ((!hit->x40_b2 || (arg_item->ground_or_air != GA_Air)) &&
+                 (!hit->x40_b3 || (arg_item->ground_or_air != GA_Ground))) ||
+                lbColl_8000ACFC(arg_item, hit) || (arg_item->xD0C == 2))
             {
-                fighter = fighter_gobj->user_data;
-                if (!((fighter->x1064_thrownHitbox.owner == NULL) ||
-                      (fighter->x1064_thrownHitbox.owner !=
-                       arg_item_owner_gobj) ||
-                      arg_item->xDCE_flag.b0))
+                continue;
+            }
+            for (hurt_index = 0; hurt_index < arg_item->xAC8_hurtboxNum;
+                 hurt_index++)
+            {
+                if (!lbColl_8000805C(hit,
+                                     &arg_item->xACC_itemHurtbox[hurt_index],
+                                     NULL, 0, fighter->x34_scale.y,
+                                     arg_item->scl, 0.0f))
                 {
-                    goto next_fighter;
+                    continue;
                 }
-                if (fighter->x1064_thrownHitbox.owner != NULL) {
-                    ft_team =
-                        ftLib_80086EB4(fighter->x1064_thrownHitbox.owner);
+                if (hit->element != HitElement_Inert) {
+                    ftColl_80076808(fighter, hit, 0, arg_item, 0);
+                    dmg = hit->damage;
+                    fighter->dmg.x1914 = dmg;
+                    arg_item->xCA0 += dmg;
+                    if (dmg > arg_item->xCA4) {
+                        arg_item->xCA4 = dmg;
+                    }
+                    it_8026F9AC_outline(
+                        1, fighter, hit, arg_item,
+                        &arg_item->xACC_itemHurtbox[hurt_index]);
+                    it_8027B378(fighter->gobj, arg_item->entity, dmg);
+                    kind = arg_item->kind;
+                    if ((kind == Pokemon_Random) &&
+                        (arg_item->xDD4_itemVar.pokemon.x0 == 7) &&
+                        ((hit->sfx_kind == 1U) || (hit->sfx_kind == 2U)))
+                    {
+                        lbAudioAx_800237A8(0x61A87, 0x7FU, 0x40U);
+                    } else if ((kind != Pokemon_Random) ||
+                               (arg_item->xDD4_itemVar.pokemon.x0 != 8) ||
+                               ((hit->sfx_kind != 1U) &&
+                                (hit->sfx_kind != 2U)))
+                    {
+                        lbColl_80005BB0(hit, -1);
+                    }
                 } else {
-                    ft_team = fighter->team;
+                    fighter->unk_gobj = arg_item_gobj;
                 }
-                if (!gm_8016B168() || gm_8016B0D4() ||
-                    arg_item->xDCE_flag.b1 ||
-                    (ft_team != arg_item->x20_team_id))
+                break;
+            }
+        }
+    }
+}
+
+static inline bool it_802706D0_sub(Item_GObj* arg_item_gobj, Item* item,
+                                   Item* arg_item, HitCapsule* hit)
+{
+    bool chk2 = false;
+    u32 i;
+    for (i = 0; i < 4U; i++) {
+        if (it_804D6D1C[i] != 0) {
+            HitCapsule* arg_hit = &arg_item->x5D4_hitboxes[i].hit;
+            if ((hit->element == HitElement_Inert) ||
+                (arg_hit->element == HitElement_Inert))
+            {
+                if ((hit->element != arg_hit->element) &&
+                    lbColl_80007AFC(hit, arg_hit, item->scl, arg_item->scl))
                 {
-                    hit_index = 0;
-                    while (hit_index < 4) {
-                        hit = &fighter->x914[hit_index];
-                        if ((hit->state != HitCapsule_Disabled) &&
-                            (hit->element != HitElement_Catch) &&
-                            (hit->x42_b7 == 1) &&
-                            ((hit->x40_b2 &&
-                              (arg_item->ground_or_air == GA_Air)) ||
-                             (hit->x40_b3 &&
-                              (arg_item->ground_or_air == GA_Ground))) &&
-                            !lbColl_8000ACFC(arg_item, hit) &&
-                            (arg_item->xD0C != 2))
-                        {
-                            hurt_index = 0;
-                            hurt = &arg_item->xACC_itemHurtbox[hurt_index];
-                            while (hurt_index < arg_item->xAC8_hurtboxNum) {
-                                if (lbColl_8000805C(hit, hurt, NULL, 0,
-                                                    fighter->x34_scale.y,
-                                                    arg_item->scl, 0.0f))
-                                {
-                                    if (hit->element != HitElement_Inert) {
-                                        ftColl_80076808(fighter, hit, 0,
-                                                        arg_item, 0);
-                                        dmg = hit->damage;
-                                        fighter->dmg.x1914 = dmg;
-                                        arg_item->xCA0 += dmg;
-                                        if (dmg > arg_item->xCA4) {
-                                            arg_item->xCA4 = dmg;
-                                        }
-                                        it_8026F9AC_outline(1, fighter, hit,
-                                                            arg_item,
-                                                            &arg_item
-                                                                 ->xACC_itemHurtbox
-                                                                     [hurt_index]);
-                                        it_8027B378(fighter->gobj,
-                                                    arg_item->entity,
-                                                    (f32) dmg);
-                                        kind = arg_item->kind;
-                                        if ((kind == Pokemon_Random) &&
-                                            (arg_item->xDD4_itemVar.pokemon
-                                                 .x0 == 7) &&
-                                            ((hit->sfx_kind == 1U) ||
-                                             (hit->sfx_kind == 2U)))
-                                        {
-                                            lbAudioAx_800237A8(0x61A87, 0x7FU,
-                                                               0x40U);
-                                        } else if ((kind != Pokemon_Random) ||
-                                                   (arg_item->xDD4_itemVar
-                                                        .pokemon.x0 != 8) ||
-                                                   ((hit->sfx_kind != 1U) &&
-                                                    (hit->sfx_kind != 2U)))
-                                        {
-                                            lbColl_80005BB0(hit, -1);
-                                        }
-                                    } else {
-                                        fighter->unk_gobj = arg_item_gobj;
-                                    }
-                                    goto block_cf0;
-                                } else {
-                                    hurt++;
-                                    hurt_index++;
-                                }
-                            }
-                        }
-                    block_cf0:
-                        hit_index++;
+                    if (hit->element == HitElement_Inert) {
+                        item->xDCE_flag.b6 = 1;
+                        item->toucher = arg_item_gobj;
+                    } else {
+                        item->xDCE_flag.b6 = 1;
+                        arg_item->toucher = item->entity;
                     }
+                    chk2 = true;
+                    break;
                 }
+            } else if ((hit->x40_b0 == 1) && (arg_hit->x40_b0 == 1) &&
+                       lbColl_80007AFC(hit, arg_hit, item->scl, arg_item->scl))
+            {
+                it_8026FE68(item, hit, arg_item, arg_hit);
+                chk2 = true;
+                break;
             }
-        next_fighter:
-            fighter_gobj = fighter_gobj->next;
         }
     }
+    return chk2;
 }
 
-void it_802706D0(Item_GObj* arg_item_gobj)
+static inline s32 it_802706D0_sub2(Item* item, Item* arg_item)
+{
+    s32 n = 0;
+    u32 i;
+    for (i = 0; i < 4; i++) {
+        HitCapsule* arg_hit = &arg_item->x5D4_hitboxes[i].hit;
+        if ((arg_hit->state != HitCapsule_Disabled) &&
+            (arg_hit->element != HitElement_Catch) &&
+            ((arg_hit->x40_b2 && (item->ground_or_air == GA_Air)) ||
+             (arg_hit->x40_b3 && (item->ground_or_air == GA_Ground))) &&
+            !lbColl_8000ACFC(item, arg_hit))
+        {
+            it_804D6D1C[i] = 1;
+            n++;
+        } else {
+            it_804D6D1C[i] = 0;
+        }
+    }
+    return n;
+}
+
+static inline void it_802706D0_sub3(Item* item, Item* arg_item,
+                                    HitCapsule* hit, HurtCapsule* arg_hurt)
 {
-    Item* arg_item;
-    HSD_GObj* item_gobj;
-    HitCapsule* hit;
-    HitCapsule* arg_hit;
-    f32 pos_x;
     f32 dir;
-    f32 pos_x_mag;
     s32 dmg;
     ItemKind kind;
-    bool chk2;
-    s32 var_r28;
+    it_8026FAC4_outline(item, hit, (hit->x41_b4) ? 8 : 0, arg_item, 0);
+    if (ABS(item->x40_vel.x) < it_804D6D28->xD4) {
+        if (item->pos.x > arg_item->pos.x) {
+            dir = -1.0f;
+        } else {
+            dir = 1.0f;
+        }
+    } else if (item->x40_vel.x < 0.0f) {
+        dir = -1.0f;
+    } else {
+        dir = 1.0f;
+    }
+    item->xCB8_outDamageDirection = dir;
+    dmg = hit->damage;
+    item->xC34_damageDealt = dmg;
+    item->xCF4_fighterGObjUnk = NULL;
+    item->xC38 = arg_item->kind;
+    item->xCD4 = arg_item->pos;
+    arg_item->xCA0 = arg_item->xCA0 + dmg;
+    if (dmg > arg_item->xCA4) {
+        arg_item->xCA4 = dmg;
+    }
+    it_8026F9AC_outline(2, item, hit, arg_item, arg_hurt);
+    it_8027B408(item->entity, arg_item->entity, dmg);
+    kind = arg_item->kind;
+    if ((kind == Pokemon_Random) && (arg_item->xDD4_itemVar.pokemon.x0 == 7) &&
+        ((hit->sfx_kind == 1U) || (hit->sfx_kind == 2U)))
+    {
+        lbAudioAx_800237A8(0x61A87, 0x7FU, 0x40U);
+    } else if ((kind != Pokemon_Random) ||
+               (arg_item->xDD4_itemVar.pokemon.x0 != 8) ||
+               ((hit->sfx_kind != 1U) && (hit->sfx_kind != 2U)))
+    {
+        lbColl_80005BB0(hit, -1);
+    }
+}
+
+void it_802706D0(Item_GObj* arg_item_gobj)
+{
+    Item* arg_item;
+    HSD_GObj* item_gobj;
+    s32 count;
     bool chk;
-    s32 var_r5;
-    u32 arg_hurt_index;
-    u32 arg_hit_index;
     u32 hit_index;
-    u8* cnt;
-    Item* item;
-    HurtCapsule* arg_hurt;
     PAD_STACK(8);
 
     chk = false;
     arg_item = GET_ITEM(arg_item_gobj);
-    item_gobj = HSD_GObj_Entities->items;
-    while (item_gobj != NULL) {
-        item = item_gobj->user_data;
+    for (item_gobj = HSD_GObj_Entities->items; item_gobj != NULL;
+         item_gobj = item_gobj->next)
+    {
+        Item* item = item_gobj->user_data;
         if (arg_item_gobj == item_gobj) {
             chk = true;
-        } else if (((arg_item->owner != NULL) || (item->owner != NULL) ||
-                    item->xDCD_flag.b7 || arg_item->xDCE_flag.b2) &&
-                   (!ftLib_80086960(item->owner) ||
-                    !ftLib_80086960(arg_item->owner) ||
-                    !ftLib_80086FD4(arg_item->owner, item->owner) ||
-                    item->xDCD_flag.b7 || arg_item->xDCE_flag.b2) &&
-                   (!gm_8016B168() || gm_8016B0D4() || item->xDCD_flag.b6 ||
-                    arg_item->xDCE_flag.b1 ||
-                    (item->x20_team_id != arg_item->x20_team_id)))
+        } else if (((arg_item->owner == NULL) && (item->owner == NULL) &&
+                    !item->xDCD_flag.b7 && !arg_item->xDCE_flag.b2) ||
+                   (ftLib_80086960(item->owner) &&
+                    ftLib_80086960(arg_item->owner) &&
+                    ftLib_80086FD4(arg_item->owner, item->owner) &&
+                    !item->xDCD_flag.b7 && !arg_item->xDCE_flag.b2) ||
+                   (gm_8016B168() && !gm_8016B0D4() && !item->xDCD_flag.b6 &&
+                    !arg_item->xDCE_flag.b1 &&
+                    (item->x20_team_id == arg_item->x20_team_id)))
         {
-            if (chk && !arg_item->xDD0_flag.b1) {
-                var_r28 = 0;
-                arg_hit_index = 0U;
-                cnt = it_804D6D1C;
-                while (arg_hit_index < 4U) {
-                    arg_hit = &arg_item->x5D4_hitboxes[arg_hit_index].hit;
-                    if ((arg_hit->state != HitCapsule_Disabled) &&
-                        (arg_hit->element != HitElement_Catch) &&
-                        ((arg_hit->x40_b2 &&
-                          (item->ground_or_air == GA_Air)) ||
-                         (arg_hit->x40_b3 &&
-                          (item->ground_or_air == GA_Ground))) &&
-                        !lbColl_8000ACFC(item, arg_hit))
+            continue;
+        }
+        if (chk && !arg_item->xDD0_flag.b1) {
+            count = it_802706D0_sub2(item, arg_item);
+        }
+        for (hit_index = 0; hit_index < 4; hit_index++) {
+            HitCapsule* hit = &item->x5D4_hitboxes[hit_index].hit;
+            if ((hit->state == HitCapsule_Disabled) || (hit->x42_b7 != 1) ||
+                ((!hit->x40_b2 || (arg_item->ground_or_air != GA_Air)) &&
+                 (!hit->x40_b3 || (arg_item->ground_or_air != GA_Ground))) ||
+                lbColl_8000ACFC(arg_item, hit))
+            {
+                continue;
+            }
+            if (!chk || arg_item->xDD0_flag.b1 || (count == 0) ||
+                it_802706D0_sub(arg_item_gobj, item, arg_item, hit) ||
+                (arg_item->xAC8_hurtboxNum == 0))
+            {
+                continue;
+            }
+            if (hit->element == HitElement_Inert) {
+                u32 i;
+                for (i = 0; i < arg_item->xAC8_hurtboxNum; i++) {
+                    if (lbColl_80008248(hit, &arg_item->xACC_itemHurtbox[i],
+                                        NULL, item->scl, arg_item->scl, 0.0f))
                     {
-                        *cnt = 1;
-                        var_r28++;
-                    } else {
-                        *cnt = 0;
+                        item->xDCE_flag.b6 = 1;
+                        item->toucher = arg_item_gobj;
                     }
-                    arg_hit_index++;
-                    cnt++;
                 }
-            }
-            hit_index = 0U;
-            while (hit_index < 4U) {
-                hit = &item->x5D4_hitboxes[hit_index].hit;
-                if ((hit->state != HitCapsule_Disabled) &&
-                    (hit->x42_b7 == 1) &&
-                    ((hit->x40_b2 && (arg_item->ground_or_air == GA_Air)) ||
-                     (hit->x40_b3 &&
-                      (arg_item->ground_or_air == GA_Ground))) &&
-                    !lbColl_8000ACFC(arg_item, hit))
-                {
-                    if (chk && !arg_item->xDD0_flag.b1 && (var_r28 != 0)) {
-                        chk2 = false;
-                        arg_hit_index = 0U;
-                        cnt = it_804D6D1C;
-                        while (arg_hit_index < 4U) {
-                            if (*cnt != 0) {
-                                arg_hit =
-                                    &arg_item->x5D4_hitboxes[arg_hit_index]
-                                         .hit;
-                                if ((hit->element == HitElement_Inert) ||
-                                    (arg_hit->element == HitElement_Inert))
-                                {
-                                    if ((hit->element != arg_hit->element) &&
-                                        lbColl_80007AFC(hit, arg_hit,
-                                                        item->scl,
-                                                        arg_item->scl))
-                                    {
-                                        if (hit->element == HitElement_Inert) {
-                                            item->xDCE_flag.b6 = 1;
-                                            item->toucher = arg_item_gobj;
-                                        } else {
-                                            item->xDCE_flag.b6 = 1;
-                                            arg_item->toucher = item->entity;
-                                        }
-                                        chk2 = true;
-                                        goto block_1080;
-                                    }
-                                } else if ((hit->x40_b0 == 1) &&
-                                           (arg_hit->x40_b0 == 1) &&
-                                           lbColl_80007AFC(hit, arg_hit,
-                                                           item->scl,
-                                                           arg_item->scl))
-                                {
-                                    it_8026FE68(item, hit, arg_item, arg_hit);
-                                    chk2 = true;
-                                    goto block_1080;
-                                }
-                            }
-                            arg_hit_index++;
-                            cnt++;
-                        }
-                    block_1080:
-                        if (!chk2) {
-                            goto block_57;
-                        }
-                    } else {
-                    block_57:
-                        if (arg_item->xAC8_hurtboxNum != 0) {
-                            if (hit->element == HitElement_Inert) {
-                                arg_hurt_index = 0U;
-                                while (arg_hurt_index <
-                                       arg_item->xAC8_hurtboxNum)
-                                {
-                                    arg_hurt = &arg_item->xACC_itemHurtbox
-                                                    [arg_hurt_index];
-                                    if (lbColl_80008248(hit, arg_hurt, NULL,
-                                                        item->scl,
-                                                        arg_item->scl, 0.0f))
-                                    {
-                                        item->xDCE_flag.b6 = 1;
-                                        item->toucher = arg_item_gobj;
-                                    }
-                                    arg_hurt_index++;
-                                }
-                            } else if (arg_item->xD0C != 2) {
-                                arg_hurt_index = 0;
-                                while (arg_hurt_index <
-                                       arg_item->xAC8_hurtboxNum)
-                                {
-                                    arg_hurt = &arg_item->xACC_itemHurtbox
-                                                    [arg_hurt_index];
-                                    if (lbColl_8000805C(hit, arg_hurt, NULL, 0,
-                                                        item->scl,
-                                                        arg_item->scl, 0.0f))
-                                    {
-                                        if (hit->x41_b4) {
-                                            var_r5 = 8;
-                                        } else {
-                                            var_r5 = 0;
-                                        }
-                                        it_8026FAC4_outline(item, hit, var_r5,
-                                                            arg_item, 0);
-                                        pos_x = item->x40_vel.x;
-                                        pos_x_mag = ABS(pos_x);
-                                        if (pos_x_mag < it_804D6D28->xD4) {
-                                            if (item->pos.x > arg_item->pos.x)
-                                            {
-                                                dir = -1.0f;
-                                            } else {
-                                                dir = 1.0f;
-                                            }
-                                        } else if (pos_x < 0.0f) {
-                                            dir = -1.0f;
-                                        } else {
-                                            dir = 1.0f;
-                                        }
-                                        item->xCB8_outDamageDirection = dir;
-                                        dmg = hit->damage;
-                                        item->xC34_damageDealt = dmg;
-                                        item->xCF4_fighterGObjUnk = NULL;
-                                        item->xC38 = arg_item->kind;
-                                        item->xCD4 = arg_item->pos;
-                                        arg_item->xCA0 = arg_item->xCA0 + dmg;
-                                        if (dmg > arg_item->xCA4) {
-                                            arg_item->xCA4 = dmg;
-                                        }
-                                        it_8026F9AC_outline(
-                                            2, item, hit, arg_item, arg_hurt);
-                                        it_8027B408(item->entity,
-                                                    arg_item->entity, dmg);
-                                        kind = arg_item->kind;
-                                        // sfx_kind = hit->sfx_kind;
-                                        if ((kind == Pokemon_Random) &&
-                                            (arg_item->xDD4_itemVar.pokemon
-                                                 .x0 == 7) &&
-                                            ((hit->sfx_kind == 1U) ||
-                                             (hit->sfx_kind == 2U)))
-                                        {
-                                            lbAudioAx_800237A8(0x61A87, 0x7FU,
-                                                               0x40U);
-                                        } else if ((kind != Pokemon_Random) ||
-                                                   (arg_item->xDD4_itemVar
-                                                        .pokemon.x0 != 8) ||
-                                                   ((hit->sfx_kind != 1U) &&
-                                                    (hit->sfx_kind != 2U)))
-                                        {
-                                            lbColl_80005BB0(hit, -1);
-                                        }
-                                        goto block_12f4;
-                                    } else {
-                                        arg_hurt_index++;
-                                    }
-                                }
-                            }
-                        }
+            } else if (arg_item->xD0C != 2) {
+                u32 i;
+                for (i = 0; i < arg_item->xAC8_hurtboxNum; i++) {
+                    if (lbColl_8000805C(hit, &arg_item->xACC_itemHurtbox[i],
+                                        NULL, 0, item->scl, arg_item->scl, 0.0f))
+                    {
+                        it_802706D0_sub3(item, arg_item, hit,
+                                         &arg_item->xACC_itemHurtbox[i]);
+                        break;
                     }
                 }
-            block_12f4:
-                hit_index++;
             }
         }
-        item_gobj = item_gobj->next;
     }
 }
 
@@ -780,12 +713,11 @@ f32 it_80270CD8(Item* ip, HitCapsule* hit)
         x24_f = 0.01f;
         x24_f *= hit->x24;
         f0 = ip->xC9C + (f32) ip->xCA0;
-        f1 = (x24_f *
-              ((it_804D6D28->x80_float[11] *
-                (attr->x1C_damage_mul *
-                 ((it_804D6D28->x80_float[8] * f0) +
-                  (it_804D6D28->x80_float[9] * (hit->damage * f0))))) +
-               it_804D6D28->x80_float[12])) +
+        f1 = (x24_f * ((it_804D6D28->x80_float[11] *
+                        (attr->x1C_damage_mul *
+                         ((it_804D6D28->x80_float[8] * f0) +
+                          (it_804D6D28->x80_float[9] * (hit->damage * f0))))) +
+                       it_804D6D28->x80_float[12])) +
              x2C_f;
     }
     if (f1 >= it_804D6D28->x80_float[7]) {
diff --git a/src/melee/it/items/itlinkbomb.c b/src/melee/it/items/itlinkbomb.c
index f3e0d1379a..5a58208fd0 100644
--- a/src/melee/it/items/itlinkbomb.c
+++ b/src/melee/it/items/itlinkbomb.c
@@ -15,10 +15,10 @@
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/it_3F14.h"
-#include "it/ithitbox.h"
-#include "it/itmaplib.h"
 #include "it/itCharItems.h"
 #include "it/item.h"
+#include "it/ithitbox.h"
+#include "it/itmaplib.h"
 #include "lb/lb_00F9.h"
 #include "mp/mpcoll.h"
 #include "MSL/math.h"
@@ -433,16 +433,14 @@ bool itLinkbomb_UnkMotion3_Anim(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     itLinkBombAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
-    bool ok = it_80272C6C(gobj);
-    PAD_STACK(8);
-
-    if (!ok) {
+    PAD_STACK(12);
+    if (!it_80272C6C(gobj)) {
         itLinkbomb_UnkMotion3_Anim_inline1(gobj);
     }
     // permuterslop
     if (attrs != NULL) {
     }
-    if (ok != 0) {
+    if (ip != NULL) {
     }
     itLinkbomb_UnkMotion3_Anim_inline2(gobj);
     return false;
diff --git a/src/melee/it/items/itlinkhookshot.c b/src/melee/it/items/itlinkhookshot.c
index 47c5657a70..cc69478de3 100644
--- a/src/melee/it/items/itlinkhookshot.c
+++ b/src/melee/it/items/itlinkhookshot.c
@@ -465,6 +465,22 @@ static void inline it_802A2EE4_inline_alt(ItemLink* item_link, Vec3* pos)
     pos->z = m[2][3];
 }
 
+static void inline it_802A2EE4_inline_alt_pad(ItemLink* item_link, Vec3* pos)
+{
+    f32 zero = 0.0F;
+    UNUSED f32 pad = zero;
+    Mtx m;
+    PSMTXIdentity((MtxPtr) &m);
+    m[0][3] = zero;
+    m[1][3] = zero;
+    m[2][3] = it_804D6D48;
+    HSD_JObjSetupMatrix(item_link->jobj);
+    PSMTXConcat(item_link->jobj->mtx, (MtxPtr) &m, (MtxPtr) &m);
+    pos->x = m[0][3];
+    pos->y = m[1][3];
+    pos->z = m[2][3];
+}
+
 void it_802A2EE4(Item_GObj* arg0)
 {
     Vec3 pos;
@@ -1206,31 +1222,34 @@ s32 it_802A4BFC(ItemLink* link_0, Vec3* arg1, itLinkHookshotAttributes* attr,
     ItemLink* link_1;
     Vec3 vec;
     f64 len;
+    UNUSED u8 frame_pad[8];
+    Vec3 pos0;
+    UNUSED u8 pad0[8];
+    Vec3 pos1;
+    UNUSED u8 pad1[8];
+    Vec3 pos2;
     ftLk_DatAttrs* lk_attr = fp->ft_data->ext_attr;
-    PAD_STACK(14 * 4);
+    PAD_STACK(7 * 4);
     if (fp->motion_id == 0xD4) {
         if (fp->mv.ca.specials.grav == lk_attr->x88) {
-            Vec3 pos;
-            it_802A2EE4_inline_alt(link_0, &pos);
-            link_0->pos = pos;
+            it_802A2EE4_inline_alt(link_0, &pos0);
+            link_0->pos = pos0;
             link_0->coll_data.cur_pos = link_0->pos;
             link_0->coll_data.last_pos = link_0->coll_data.cur_pos;
             link_0->x2C_b0 = 1;
         }
     } else if (fp->motion_id == 0xD6) {
         if (fp->mv.ca.specials.grav == lk_attr->x98) {
-            Vec3 pos;
-            it_802A2EE4_inline_alt(link_0, &pos);
-            link_0->pos = pos;
+            it_802A2EE4_inline_alt(link_0, &pos1);
+            link_0->pos = pos1;
             link_0->coll_data.cur_pos = link_0->pos;
             link_0->coll_data.last_pos = link_0->coll_data.cur_pos;
             link_0->x2C_b0 = 1;
         }
     } else if (fp->motion_id == 0x168) {
         if (fp->mv.ca.specials.grav == lk_attr->xA8) {
-            Vec3 pos;
-            it_802A2EE4_inline_alt(link_0, &pos);
-            link_0->pos = pos;
+            it_802A2EE4_inline_alt_pad(link_0, &pos2);
+            link_0->pos = pos2;
             link_0->coll_data.cur_pos = link_0->pos;
             link_0->coll_data.last_pos = link_0->coll_data.cur_pos;
             link_0->x2C_b0 = 1;
