diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 42f12af7cb..338b4d6d64 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -31288,12 +31288,12 @@ gm_804DAB88 = .sdata2:0x804DAB88; // type:object size:0x8 scope:global align:8 d
 @151 = .sdata2:0x804DAC04; // type:object size:0x4 scope:local align:4 data:float
 @152 = .sdata2:0x804DAC08; // type:object size:0x4 scope:local align:4 data:float
 @154 = .sdata2:0x804DAC10; // type:object size:0x8 scope:local align:8 data:double
-gm_804DAC18 = .sdata2:0x804DAC18; // type:object size:0x3 scope:local data:byte
-gm_804DAC1C = .sdata2:0x804DAC1C; // type:object size:0x3 scope:local data:byte
-gm_804DAC20 = .sdata2:0x804DAC20; // type:object size:0x4 scope:local align:4 data:float
-gm_804DAC24 = .sdata2:0x804DAC24; // type:object size:0x3 scope:local
-gm_804DAC28 = .sdata2:0x804DAC28; // type:object size:0x3 scope:local
-gm_804DAC2C = .sdata2:0x804DAC2C; // type:object size:0x4 scope:local align:4 data:float
+@217 = .sdata2:0x804DAC18; // type:object size:0x3 scope:local data:byte
+@218 = .sdata2:0x804DAC1C; // type:object size:0x3 scope:local data:byte
+@244 = .sdata2:0x804DAC20; // type:object size:0x4 scope:local align:4 data:float
+@306 = .sdata2:0x804DAC24; // type:object size:0x3 scope:local
+@307 = .sdata2:0x804DAC28; // type:object size:0x3 scope:local
+@330 = .sdata2:0x804DAC2C; // type:object size:0x4 scope:local align:4 data:float
 @91 = .sdata2:0x804DAC30; // type:object size:0x4 scope:local align:4 data:float
 @93 = .sdata2:0x804DAC34; // type:object size:0x4 scope:local align:4 data:float
 @22 = .sdata2:0x804DAC38; // type:object size:0x4 scope:local align:4 data:float
diff --git a/configure.py b/configure.py
index cc097e0a20..8a20e45afe 100755
--- a/configure.py
+++ b/configure.py
@@ -945,7 +945,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/gm/gm_1AED.c"),
             Object(NonMatching, "melee/gm/gm_1B03.c"),
             Object(NonMatching, "melee/gm/gmclassic.c"),
-            Object(NonMatching, "melee/gm/gmadventure.c"),
+            Object(Matching, "melee/gm/gmadventure.c"),
             Object(NonMatching, "melee/gm/gmallstar.c"),
             Object(Matching, "melee/gm/gmmultiman.c"),
             Object(Matching, "melee/gm/gmsupersudden.c"),
diff --git a/src/melee/gm/gmadventure.c b/src/melee/gm/gmadventure.c
index dac13d5733..a80cb05aa1 100644
--- a/src/melee/gm/gmadventure.c
+++ b/src/melee/gm/gmadventure.c
@@ -1467,12 +1467,25 @@ void gm_801B4860(MinorScene* scene)
     gm_801B4860_inline1(scene);
 }
 
+static void gm_8016A22C_inline(MinorScene* arg0, UnkAdventureData* temp_r29)
+{
+    u8 count = gm_8017E48C(arg0);
+    struct gm_803DE650_t* temp_r25_2 = gm_8017E4C4(arg0->idx);
+    gm_8016A22C(temp_r25_2->xA[0], temp_r25_2->xA[1], temp_r25_2->xA[2],
+                temp_r25_2->xD, temp_r25_2->xE, temp_r25_2->xF, 1, 0, 1,
+                temp_r29->x0.ckind, temp_r29->x0.color,
+                temp_r29->x4C(count, temp_r29->x0.cpu_level, 0),
+                temp_r25_2->x4, gm_8017BE8C(temp_r25_2->xA), 1, 0, 1,
+                temp_r29->x64(count, temp_r29->x0.cpu_level),
+                temp_r29->x68(count, temp_r29->x0.cpu_level));
+    temp_r29->x0.xC.x11 = 1;
+}
+
 void gm_801B4974(MinorScene* arg0)
 {
     u8 count;
     UnkAdventureData* temp_r29;
     UnkAdventureData* temp_r3;
-    struct gm_803DE650_t* temp_r25_2;
     struct gm_803DE650_t* temp_r30;
     struct gmm_x0_528_t* temp_r24;
     u8 temp_r3_2;
@@ -1496,17 +1509,9 @@ void gm_801B4974(MinorScene* arg0)
     temp_r30 = gm_8017E4C4(arg0->idx);
     temp_r24_2 = gm_80169520();
     temp_r31 = gm_80169540();
-    count = gm_8017E48C(arg0);
-    temp_r25_2 = gm_8017E4C4(arg0->idx);
-    gm_8016A22C(temp_r25_2->xA[0], temp_r25_2->xA[1], temp_r25_2->xA[2],
-                temp_r25_2->xD, temp_r25_2->xE, temp_r25_2->xF, 1, 0, 1,
-                temp_r29->x0.ckind, temp_r29->x0.color,
-                temp_r29->x4C(count, temp_r29->x0.cpu_level, 0),
-                temp_r25_2->x4, gm_8017BE8C(temp_r25_2->xA), 1, 0, 1,
-                temp_r29->x64(count, temp_r29->x0.cpu_level),
-                temp_r29->x68(count, temp_r29->x0.cpu_level));
 
-    temp_r29->x0.xC.x11 = 1;
+    gm_8016A22C_inline(arg0, temp_r29);
+
     temp_r28->x3 = *temp_r24_2;
     temp_r28->x4 = temp_r31[temp_r30->x4 - 2];
     temp_r28->x5 = temp_r31[temp_r30->x4 - 1];
@@ -1519,16 +1524,7 @@ static inline void inline0(MinorScene* scene)
     if (adv_data->x0.xC.x11 != 1) {
         UNK_T unused1 = gm_80169520();
         UNK_T unused2 = gm_80169540();
-        u8 count = gm_8017E48C(scene);
-        struct gm_803DE650_t* temp_r29 = gm_8017E4C4(scene->idx);
-        gm_8016A22C(temp_r29->xA[0], temp_r29->xA[1], temp_r29->xA[2],
-                    temp_r29->xD, temp_r29->xE, temp_r29->xF, 1, 0, 1,
-                    adv_data->x0.ckind, adv_data->x0.color,
-                    adv_data->x4C(count, adv_data->x0.cpu_level, 0),
-                    temp_r29->x4, gm_8017BE8C(temp_r29->xA), 1, 0, 1,
-                    adv_data->x64(count, adv_data->x0.cpu_level),
-                    adv_data->x68(count, adv_data->x0.cpu_level));
-        adv_data->x0.xC.x11 = 1;
+        gm_8016A22C_inline(scene, adv_data);
     }
 }
 
