diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index dc57c8f86f..3b2f8e5e5a 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -637,7 +637,7 @@ void ftKb_SkSpecialAirNEnd_Coll(Fighter_GObj* gobj)
     if (ft_80081D0C(gobj)) {
         fp->fv.kb.xB4 = 0;
         fp->mv.kb.specialhi.x4 = 0;
-        ftKb_SpecialNSk_80105E8C(gobj);
+        ftCo_Landing_Enter_Basic(gobj);
     }
 }
 
