diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
index 04e085cbb9..80c5525089 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
@@ -1028,7 +1028,16 @@ void ftPp_SpecialHiStart_1_Phys(Fighter_GObj* gobj)
     ft_80084F3C(gobj);
 }
 
-/// #ftPp_SpecialAirHiStart_1_Phys
+void ftPp_SpecialAirHiStart_1_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(8);
+    ftCommon_Fall(fp, da->xA8, da->xAC);
+    if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+}
 
 void ftPp_SpecialHiStart_1_Coll(Fighter_GObj* gobj)
 {
@@ -1112,7 +1121,16 @@ void ftPp_SpecialHiThrow_1_Phys(Fighter_GObj* gobj)
     ft_80084F3C(gobj);
 }
 
-/// #ftPp_SpecialAirHiThrow_1_Phys
+void ftPp_SpecialAirHiThrow_1_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(8);
+    ftCommon_Fall(fp, da->xA8, da->xAC);
+    if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+}
 
 void ftPp_SpecialHiThrow_1_Coll(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftPopo/types.h b/src/melee/ft/chara/ftPopo/types.h
index f9858bcc59..b2fd1179ed 100644
--- a/src/melee/ft/chara/ftPopo/types.h
+++ b/src/melee/ft/chara/ftPopo/types.h
@@ -58,7 +58,8 @@ typedef struct ftIceClimberAttributes {
     float x98;
     u8 _9C[0xA4 - 0x9C];
     float xA4;
-    u8 _A8[0xB0 - 0xA8];
+    float xA8;
+    float xAC;
     float xB0;
     float xB4;
     u8 _B8[0xC4 - 0xB8];
