diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index f5c84eff3a..a9aacfd054 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -36,6 +36,7 @@
 #include "it/items/itkirby_2F23.h"
 #include "it/items/itkirbycutterbeam.h"
 #include "it/items/itkirbyhammer.h"
+#include "it/items/itluigifireball.h"
 #include "lb/lb_00B0.h"
 #include "lb/lbanim.h"
 #include "lb/lbvector.h"
@@ -2696,7 +2697,22 @@ void ftKb_LgSpecialAirN_Coll(Fighter_GObj* gobj)
     }
 }
 
-/// #fn_800F98F4
+void fn_800F98F4(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    if (ftCheckThrowB0(fp)) {
+        Vec3 pos;
+        Fighter_Part bone;
+        PAD_STACK(4);
+        bone = ftParts_GetBoneIndex(fp, FtPart_LHandN);
+        lb_8000B1CC(fp->parts[bone].joint, NULL, &pos);
+        it_802C01AC(gobj, &pos, It_Kind_Kirby_LuigiFire,
+                    fp->facing_dir);
+        bone = ftParts_GetBoneIndex(fp, FtPart_LHandN);
+        efSync_Spawn(0x4B1, gobj, fp->parts[bone].joint,
+                     &fp->facing_dir);
+    }
+}
 
 void ftKb_SpecialNCa_800F99BC(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
index e67e06c721..7fcebaf774 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
@@ -1470,7 +1470,28 @@ void ftPp_SpecialAirLw_Phys(Fighter_GObj* gobj)
     ft_80084EEC(gobj);
 }
 
-/// #fn_80122B54
+void fn_80122B54(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    PAD_STACK(8);
+    ftCommon_8007D7FC(fp);
+    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialLw, 0x0C4C5282,
+                              fp->cur_anim_frame, 1.0F, 0.0F, NULL);
+    {
+        Fighter* fp2 = gobj->user_data;
+        fp2->death2_cb = ftPp_Init_8011F060;
+        fp2->take_dmg_cb = ftPp_Init_8011F060;
+    }
+    fp = gobj->user_data;
+    if (fp->cmd_vars[3] != 0) {
+        ftPartSetRotX(fp, 0,
+                      fp->facing_dir * atan2f(fp->coll_data.floor.normal.x,
+                                              fp->coll_data.floor.normal.y));
+    } else {
+        ftPartSetRotX(fp, 0, 0.0F);
+    }
+    fp->accessory4_cb = fn_80122D2C;
+}
 
 /// #ftPp_SpecialLw_Coll
 
diff --git a/src/melee/gm/gm_1601.c b/src/melee/gm/gm_1601.c
index 38b9097824..06d62da0b2 100644
--- a/src/melee/gm/gm_1601.c
+++ b/src/melee/gm/gm_1601.c
@@ -2777,7 +2777,32 @@ void fn_801689E4(HSD_JObj* arg0, DynamicModelDesc* arg1, int idx)
     HSD_JObjAddAnimAll(arg0, anim, matanim, shapeanim);
 }
 
-/// #fn_80168A6C
+s32 fn_80168A6C(void* arg0, void* arg1, s32 idx)
+{
+    struct {
+        /* 0x00 */ s32** x0;
+        /* 0x04 */ s32* x4;
+        /* 0x08 */ s32 x8;
+        /* 0x0C */ s32 xC;
+    }* src = arg0;
+
+    memzero(arg1, 0x20);
+
+    if (src->x0[idx] != NULL) {
+        ((s32*)arg1)[0] = src->x0[idx][0];
+        ((s32*)arg1)[1] = src->x0[idx][1];
+        ((s32*)arg1)[2] = src->x0[idx][2];
+        ((s32*)arg1)[3] = src->x0[idx][3];
+    }
+
+    if (src->x4 != NULL) {
+        ((s32*)arg1)[5] = src->x4[0];
+        ((s32*)arg1)[6] = src->x4[1];
+    }
+
+    ((s32*)arg1)[4] = src->x8;
+    ((s32*)arg1)[7] = src->xC;
+}
 
 f32 gm_80168B34(CharacterKind ckind, int arg1, int arg2)
 {
diff --git a/src/melee/gm/gmregclear.c b/src/melee/gm/gmregclear.c
index bde07c53a1..01c7f3148e 100644
--- a/src/melee/gm/gmregclear.c
+++ b/src/melee/gm/gmregclear.c
@@ -46,11 +46,15 @@ typedef struct lbl_804706D8_t {
 lbl_804706D8_t lbl_804706D8[12];
 
 struct lbl_80472D28_t {
-    /*   +0 */ char pad_0[0x104];
+    /*   +0 */ char pad_0[0x2C];
+    /* +2C */ HSD_GObj* x2C;
+    /* +30 */ char pad_30[0xD4];
     /* +104 */ int x104;
     /* +108 */ s16 x108;
     /* +10A */ s16 x10A;
-    /* +10C */ char pad_10C[0xA];
+    /* +10C */ f32 x10C;
+    /* +110 */ u32 x110;
+    /* +114 */ char pad_114[2];
     /* +116 */ u8 x116;
     /* +117 */ char pad_117[1];
     /* +118 */ u8 x118;
@@ -97,6 +101,12 @@ STATIC_ASSERT(sizeof(AllstarStageEntry) == 0x1A);
 extern AdventureStageEntry lbl_803D7AC0[110];
 extern AllstarStageEntry lbl_803D85F0[55];
 
+typedef struct RegClearEv {
+    /* 0x00 */ char pad_0[0x1C];
+    /* 0x1C */ HSD_ImageDesc* x1C;
+    /* 0x20 */ f32 x20;
+} RegClearEv;
+
 static struct lbl_804706C0_t {
     int x0;
     int x4;
@@ -1111,7 +1121,23 @@ int fn_8017F294(void)
 
 /// #fn_8017FBA4
 
-/// #fn_8017FE54
+void fn_8017FE54(HSD_GObj* gobj)
+{
+    RegClearEv* ev = gobj->user_data;
+
+    lb_800122C8(ev->x1C, 0, 0, 1);
+    lb_800138D8(lbl_80472D28.x2C,
+                (int)(120.0F * lbl_80472D28.x10C) + 1);
+
+    ev->x20 = 0.0225F * (f32) lbl_80472D28.x110 - 0.175F;
+
+    if (ev->x20 < 0.05F) {
+        ev->x20 = 0.0F;
+    }
+    if (ev->x20 > 1.0F) {
+        ev->x20 = 1.0F;
+    }
+}
 
 /// #fn_8017FF1C
 
diff --git a/src/melee/gm/gmregclear.h b/src/melee/gm/gmregclear.h
index b2920f2d3e..bba2c86737 100644
--- a/src/melee/gm/gmregclear.h
+++ b/src/melee/gm/gmregclear.h
@@ -85,7 +85,7 @@
 /* 17F608 */ UNK_RET fn_8017F608(UNK_PARAMS);
 /* 17FA1C */ UNK_RET fn_8017FA1C(UNK_PARAMS);
 /* 17FBA4 */ UNK_RET fn_8017FBA4(UNK_PARAMS);
-/* 17FE54 */ UNK_RET fn_8017FE54(UNK_PARAMS);
+/* 17FE54 */ void fn_8017FE54(HSD_GObj* gobj);
 /* 17FF1C */ UNK_RET fn_8017FF1C(UNK_PARAMS);
 /* 1803FC */ UNK_RET fn_801803FC(UNK_PARAMS);
 /* 180630 */ void fn_80180630(int, int, int, bool, lbl_8046B6A0_24C_t*);
diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index 3fe1a3b9d0..e21290f9ee 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -10,7 +10,10 @@
 #include "gr/stage.h"
 #include "it/it_26B1.h"
 #include "lb/lb_00B0.h"
+#include "lb/lblanguage.h"
 
+int grHr_804D6ADC;
+f32 grHr_804D6AE0;
 f32 grHr_804D6AE4;
 static void* grHr_804D6AE8;
 StageCallbacks grHr_803E8140[11] = {
@@ -392,7 +395,29 @@ void grHomeRun_8021E4FC(Ground_GObj* arg) {}
 
 /// #fn_8021E994
 
-/// #grHomeRun_8021EA30
+void grHomeRun_8021EA30(f32* pos)
+{
+    f32 result;
+    f32 time;
+
+    time = Ground_801C0498();
+    result = *pos - 70.0F * (grHr_804D6AE4 * time);
+
+    time = Ground_801C0498();
+    result /= grHr_804D6AE4 * ((f32) grHr_804D6ADC * (160.0F * time));
+    result *= grHr_804D6AE0;
+
+    if (result < 0.0F) {
+        result = 0.0F;
+    }
+
+    if (lbLang_IsSavedLanguageUS()) {
+        result = (f32)(result * 0.304788);
+    }
+
+    stage_info.x6E0 = 100.0F * result;
+    PAD_STACK(8);
+}
 
 void grHomeRun_8021EAF8(void)
 {
diff --git a/src/melee/gr/grhomerun.h b/src/melee/gr/grhomerun.h
index d3c74d0ee6..f202b144f0 100644
--- a/src/melee/gr/grhomerun.h
+++ b/src/melee/gr/grhomerun.h
@@ -64,7 +64,7 @@
 /* 21E4FC */ void grHomeRun_8021E4FC(Ground_GObj*);
 /* 21E500 */ UNK_RET grHomeRun_8021E500(UNK_PARAMS);
 /* 21E994 */ UNK_RET fn_8021E994(UNK_PARAMS);
-/* 21EA30 */ UNK_RET grHomeRun_8021EA30(UNK_PARAMS);
+/* 21EA30 */ void grHomeRun_8021EA30(f32*);
 /* 21EAF8 */ UNK_RET grHomeRun_8021EAF8(UNK_PARAMS);
 /* 21EB10 */ void fn_8021EB10(HSD_GObj*);
 /* 21EC58 */ void* grHomeRun_8021EC58(u16);
diff --git a/src/melee/gr/grrcruise.c b/src/melee/gr/grrcruise.c
index 2385db8275..fac04be647 100644
--- a/src/melee/gr/grrcruise.c
+++ b/src/melee/gr/grrcruise.c
@@ -43,6 +43,8 @@ StageCallbacks grRc_803E4E34[7] = {
       grRCruise_801FF920, 0 },
 };
 
+extern s16 grRc_803E4FF0[];
+
 static struct {
     int x0;
 }* grRc_804D6A10;
@@ -321,7 +323,22 @@ void grRCruise_80200540(Ground_GObj* gobj)
 
 /// #grRCruise_8020071C
 
-/// #grRCruise_80200B48
+void grRCruise_80200B48(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    s32 i;
+
+    for (i = 0; i < 17; i++) {
+        struct grRCruise_Entry* entry = &gp->gv.rcruise.entries[i];
+        entry->x02 =
+            Ground_801C32D4(gp->map_id, grRc_803E4FF0[i]);
+        entry->x14 = Ground_801C3FA4(gobj, grRc_803E4FF0[i]);
+        entry->x08 = 0;
+        entry->x04 = 0;
+        entry->x0C = HSD_JObjGetTranslationY(entry->x14);
+        entry->x00 = 0;
+    }
+}
 
 /// #grRCruise_80200C04
 
diff --git a/src/melee/gr/grrcruise.h b/src/melee/gr/grrcruise.h
index 71960b8e4d..6965344361 100644
--- a/src/melee/gr/grrcruise.h
+++ b/src/melee/gr/grrcruise.h
@@ -47,7 +47,7 @@
 /* 200578 */ void grRCruise_80200578(Ground*, s32, CollData*, s32,
                                      mpLib_GroundEnum, f32);
 /* 20071C */ void grRCruise_8020071C(Ground_GObj*);
-/* 200B48 */ UNK_RET grRCruise_80200B48(UNK_PARAMS);
+/* 200B48 */ void grRCruise_80200B48(Ground_GObj*);
 /* 200C04 */ void grRCruise_80200C04(Ground_GObj*);
 /* 2010A4 */ void grRCruise_802010A4(Ground_GObj*, s32, CollData*);
 /* 201110 */ void grRCruise_80201110(Ground_GObj*);
diff --git a/src/melee/gr/gryorster.c b/src/melee/gr/gryorster.c
index c4a5682e61..703e4c3315 100644
--- a/src/melee/gr/gryorster.c
+++ b/src/melee/gr/gryorster.c
@@ -4,6 +4,7 @@
 
 #include "forward.h"
 
+#include "ft/ftlib.h"
 #include "gr/grdisplay.h"
 #include "gr/ground.h"
 #include "gr/grzakogenerator.h"
@@ -38,8 +39,13 @@ StageData grYt_803E51CC = {
     0,
 };
 
+typedef struct YorsterParams {
+    u8 pad_00[0x14];
+    int x14;
+} YorsterParams;
+
 typedef struct grYt_804D6A20_t {
-    void* x0;
+    YorsterParams* x0;
 } grYt_804D6A20_t;
 grYt_804D6A20_t grYt_804D6A20;
 
@@ -165,7 +171,22 @@ void grYorster_802022A0(HSD_GObj* gobj) {}
 
 void grYorster_802022A4(void) {}
 
-/// #grYorster_80202428
+void grYorster_80202428(HSD_GObj* item_gobj, Ground* gp, Vec3* pos,
+                        HSD_GObj* fighter_gobj, f32 value)
+{
+    int i;
+
+    if (ftLib_80086960(fighter_gobj)) {
+        ftLib_80086A4C(fighter_gobj, (f32) grYt_804D6A20.x0->x14);
+    }
+
+    for (i = 0; i < 9; i++) {
+        if (gp->gv.yorster.elements[i].x1C == item_gobj) {
+            gp->gv.yorster.elements[i].x4 += value;
+            break;
+        }
+    }
+}
 
 /// #grYorster_802024F0
 
diff --git a/src/melee/gr/gryorster.h b/src/melee/gr/gryorster.h
index 24e1f62349..8b65da7311 100644
--- a/src/melee/gr/gryorster.h
+++ b/src/melee/gr/gryorster.h
@@ -25,7 +25,8 @@
 /* 202254 */ void grYorster_80202254(Ground_GObj*);
 /* 2022A0 */ void grYorster_802022A0(Ground_GObj*);
 /* 2022A4 */ void grYorster_802022A4(void);
-/* 202428 */ UNK_RET grYorster_80202428(UNK_PARAMS);
+/* 202428 */ void grYorster_80202428(HSD_GObj*, Ground*, Vec3*,
+                                     HSD_GObj*, f32);
 /* 2024F0 */ UNK_RET grYorster_802024F0(UNK_PARAMS);
 /* 20266C */ void grYorster_8020266C(HSD_GObj*);
 /* 202B5C */ DynamicsDesc* grYorster_80202B5C(enum_t);
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 7c16ba0e82..fe46d02252 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -670,8 +670,16 @@ struct grStadium_type9_GroundVars {
     /* D4   */ HSD_JObj* xD4_jobj;
 };
 
+struct grYorster_TrackElement {
+    /* 0x00 */ u8 pad_00[4];
+    /* 0x04 */ f32 x4;
+    /* 0x08 */ u8 pad_08[0x14];
+    /* 0x1C */ HSD_GObj* x1C;
+};
+
 struct grYorster_GroundVars {
     int xC4;
+    struct grYorster_TrackElement elements[9];
 };
 
 struct grZebes_GroundVars {
diff --git a/src/melee/if/soundtest.c b/src/melee/if/soundtest.c
index 9ca2316504..75582844c3 100644
--- a/src/melee/if/soundtest.c
+++ b/src/melee/if/soundtest.c
@@ -304,7 +304,16 @@ bool fn_802FFCC8(void)
     return false;
 }
 
-/// #un_802FFCD0
+#pragma dont_inline on
+void un_802FFCD0(int count, void* data)
+{
+    int* arr = data;
+    int i;
+    for (i = 0; i < count; i++) {
+        arr[i] = arr[count];
+    }
+}
+#pragma dont_inline reset
 
 void un_802FFD94(int arg0, void* arg1, void* arg2)
 {
diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index 7316b220d0..af8a8329e2 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -155,9 +155,12 @@ typedef struct itKyasarinAttributes {
     /* +0  */ s32* x0;
     /* +4  */ f32 x4;
     /* +8  */ f32 x8;
-    /* +C  */ u8 xC_pad[0x28 - 0xC];
+    /* +C  */ f32 xC;
+    /* +10 */ u8 x10_pad[0x28 - 0x10];
     /* +28 */ f32 x28;
     /* +2C */ f32 x2C;
+    /* +30 */ f32 x30;
+    /* +34 */ f32 x34;
 } itKyasarinAttributes;
 
 typedef struct itKyasarinEgg_ItemVars {
diff --git a/src/melee/it/items/itclimbersice.c b/src/melee/it/items/itclimbersice.c
index 5c9447b557..afba86670d 100644
--- a/src/melee/it/items/itclimbersice.c
+++ b/src/melee/it/items/itclimbersice.c
@@ -146,9 +146,45 @@ bool itClimbersice_UnkMotion2_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itClimbersice_UnkMotion2_Phys
+void itClimbersice_UnkMotion2_Phys(Item_GObj* gobj)
+{
+    if (it_802C1854()) {
+        Item* ip = GET_ITEM(gobj);
+        itClimbersIceAttributes* sa =
+            ip->xC4_article_data->x4_specialAttributes;
+        u32 dmg = (u32) ABS(ip->x40_vel.x * sa->x30) + sa->x2C;
+        it_80272460(&ip->x5D4_hitboxes[0].hit, dmg, gobj);
+    }
+}
+
+bool itClimbersice_UnkMotion2_Coll(Item_GObj* gobj)
+{
+    Item* ip;
+    itClimbersIceAttributes* sa;
+
+    it_8026D62C(gobj, (HSD_GObjEvent) fn_802C1D44);
+    it_80276CB8(gobj);
 
-/// #itClimbersice_UnkMotion2_Coll
+    ip = GET_ITEM(gobj);
+    sa = ip->xC4_article_data->x4_specialAttributes;
+
+    if (it_80276308(gobj) != 0) {
+        f32 vel_x = ip->x40_vel.x;
+        if (vel_x < 0.0f) {
+            vel_x = -vel_x;
+        }
+        if (vel_x <= sa->xC) {
+            it_8027770C(gobj);
+            it_80272980(gobj);
+            ip = GET_ITEM(gobj);
+            sa = ip->xC4_article_data->x4_specialAttributes;
+            ip->xD44_lifeTimer -= sa->x4;
+        } else {
+            return true;
+        }
+    }
+    return false;
+}
 
 void fn_802C1D44(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itcrazyhandbomb.c b/src/melee/it/items/itcrazyhandbomb.c
index 03a6858dab..eef55e1e8d 100644
--- a/src/melee/it/items/itcrazyhandbomb.c
+++ b/src/melee/it/items/itcrazyhandbomb.c
@@ -1,8 +1,8 @@
 #include "itcrazyhandbomb.h"
 
-#include <placeholder.h>
 #include <platform.h>
 
+#include "db/db.h"
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
@@ -11,7 +11,27 @@
 
 #include <baselib/jobj.h>
 
-/// #it_802F0F6C
+void it_802F0F6C(Fighter_GObj* owner, Vec3* prev_pos, Vec3* pos,
+                  ItemKind kind, f32 facing_dir)
+{
+    SpawnItem spawnitem;
+    Item_GObj* gobj;
+
+    spawnitem.kind = kind;
+    spawnitem.prev_pos = *prev_pos;
+    spawnitem.pos = *pos;
+    spawnitem.facing_dir = facing_dir;
+    spawnitem.x3C_damage = 0;
+    spawnitem.vel.x = spawnitem.vel.y = spawnitem.vel.z = 0.0F;
+    spawnitem.x0_parent_gobj = owner;
+    spawnitem.x4_parent_gobj2 = spawnitem.x0_parent_gobj;
+    spawnitem.x44_flag.b0 = true;
+    spawnitem.x40 = 0;
+    gobj = Item_80268B18(&spawnitem);
+    it_802F10F8(gobj);
+    db_80225DD8(gobj, owner);
+    it_802750F8(gobj);
+}
 
 void it_802F1030(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itcrazyhandbomb.h b/src/melee/it/items/itcrazyhandbomb.h
index 8458322b20..c0f7f04326 100644
--- a/src/melee/it/items/itcrazyhandbomb.h
+++ b/src/melee/it/items/itcrazyhandbomb.h
@@ -7,7 +7,8 @@
 
 #include "it/items/types.h"
 
-/* 2F0F6C */ UNK_RET it_802F0F6C(UNK_PARAMS);
+/* 2F0F6C */ void it_802F0F6C(Fighter_GObj* owner, Vec3* prev_pos, Vec3* pos,
+                              ItemKind kind, f32 facing_dir);
 /* 2F1030 */ void it_802F1030(Item_GObj* gobj);
 /* 2F1094 */ void itCrazyHandBomb_Logic86_EvtUnk(Item_GObj*, Item_GObj*);
 /* 2F10B4 */ bool itCrazyHandBomb_Logic86_DmgDealt(Item_GObj* gobj);
diff --git a/src/melee/it/items/itkyasarin.c b/src/melee/it/items/itkyasarin.c
index 717482c2a6..d479eacb23 100644
--- a/src/melee/it/items/itkyasarin.c
+++ b/src/melee/it/items/itkyasarin.c
@@ -121,7 +121,25 @@ bool itKyasarin_UnkMotion2_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itKyasarin_UnkMotion2_Coll
+bool itKyasarin_UnkMotion2_Coll(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itKyasarinAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+    if (ip->facing_dir == 1.0F) {
+        if (ip->pos.x <= attr->x34) {
+            Item_80268E5C(gobj, 3, ITEM_ANIM_UPDATE);
+            it_80273454(gobj);
+            ip->xDD4_itemVar.kyasarin.x24 = attr->xC;
+        }
+    } else {
+        if (ip->pos.x >= attr->x30) {
+            Item_80268E5C(gobj, 3, ITEM_ANIM_UPDATE);
+            it_80273454(gobj);
+            ip->xDD4_itemVar.kyasarin.x24 = attr->xC;
+        }
+    }
+    return false;
+}
 
 void it_802ED25C(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ittincle.c b/src/melee/it/items/ittincle.c
index 9bbe2e538f..812a4044fb 100644
--- a/src/melee/it/items/ittincle.c
+++ b/src/melee/it/items/ittincle.c
@@ -6,12 +6,31 @@
 #include "it/it_2725.h"
 #include "it/itCommonItems.h"
 #include "it/item.h"
+#include "lb/lb_00B0.h"
 
+#include <MSL/math.h>
 #include <baselib/random.h>
 
 /* 2EC870 */ static void it_802EC870(Item_GObj*, int);
 
-/// #it_802EB5C8
+void it_802EB5C8(Item_GObj* gobj)
+{
+    Vec3 pos;
+    Item* ip = GET_ITEM(gobj);
+
+    it_8027B730(gobj);
+    ip->xD5C = 0;
+    ip->xDCC_flag.b3 = 0;
+    ip->xDC8_word.flags.x15 = 0;
+    it_8027542C(gobj);
+
+    lb_8000B1CC(ip->xBBC_dynamicBoneTable->bones[11], NULL, &pos);
+
+    ip->xDD4_itemVar.tincle.x50 = ABS(pos.y - ip->pos.y);
+
+    ip->xDD4_itemVar.tincle.x68 = ip->xC1C;
+    it_802EB6DC(gobj);
+}
 
 bool itTincle_Logic13_DmgReceived(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ittincle.h b/src/melee/it/items/ittincle.h
index 335a67d26f..d578e792f3 100644
--- a/src/melee/it/items/ittincle.h
+++ b/src/melee/it/items/ittincle.h
@@ -10,7 +10,7 @@
 /* 2EB5C8 */ void it_802EB5C8(Item_GObj*);
 /* 2EB684 */ bool itTincle_Logic13_DmgReceived(Item_GObj*);
 /* 2EB6A8 */ bool itTincle_Logic13_DmgDealt(Item_GObj*);
-/* 2EB6DC */ UNK_RET it_802EB6DC(UNK_PARAMS);
+/* 2EB6DC */ void it_802EB6DC(Item_GObj*);
 /* 2EB808 */ bool itTincle_UnkMotion0_Anim(Item_GObj* gobj);
 /* 2EB810 */ void itTincle_UnkMotion0_Phys(Item_GObj* gobj);
 /* 2EB84C */ bool itTincle_UnkMotion0_Coll(Item_GObj* gobj);
