diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index 2ab4934c69..e9e1528e49 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -152,9 +152,12 @@ typedef struct itKyasarin_ItemVars {
 } itKyasarin_ItemVars;
 
 typedef struct itKyasarinAttributes {
-    /* +0 */ s32* x0;
-    /* +4 */ f32 x4;
-    /* +8 */ f32 x8;
+    /* +0  */ s32* x0;
+    /* +4  */ f32 x4;
+    /* +8  */ f32 x8;
+    /* +C  */ u8 xC_pad[0x28 - 0xC];
+    /* +28 */ f32 x28;
+    /* +2C */ f32 x2C;
 } itKyasarinAttributes;
 
 typedef struct itKyasarinEgg_ItemVars {
@@ -173,7 +176,9 @@ typedef struct itKyasarinEggAttributes {
 
 typedef struct itHououAttr {
     /* +00 */ f32 timer;
-    /* +04 */ u8 x4_pad[0x14];
+    /* +04 */ u8 x4_pad[0xC];
+    /* +10 */ f32 x10;
+    /* +14 */ u8 x14_pad[0x4];
     /* +18 */ f32 x18;
     /* +1C */ f32 x1C;
 } itHououAttr;
@@ -1014,7 +1019,7 @@ typedef struct itMewtwoShadowball_ItemVars {
     /* ip+E00 */ f32 x2C[6];
     /* ip+E18 */ f32 x44;
     /* ip+E1C */ f32 x48;
-    /* ip+E20 */ f32 x4C;
+    /* ip+E20 */ s32 x4C;
     /* ip+E24 */ f32 x50;
     /* ip+E28 */ s32 x54;
     /* ip+E2C */ s32 x58;
diff --git a/src/melee/it/items/itdosei.c b/src/melee/it/items/itdosei.c
index fe777b5b27..c3398791b8 100644
--- a/src/melee/it/items/itdosei.c
+++ b/src/melee/it/items/itdosei.c
@@ -43,7 +43,23 @@ bool itDosei_UnkMotion0_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #fn_80281734
+void fn_80281734(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    if (ip->facing_dir == 1.0f) {
+        if (ip->xDD4_itemVar.dosei.xDF8.x < 0.0f) {
+            it_80281C6C(gobj);
+            return;
+        }
+        it_802817A0(gobj);
+        return;
+    }
+    if (ip->xDD4_itemVar.dosei.xDF8.x > 0.0f) {
+        it_80281C6C(gobj);
+        return;
+    }
+    it_802817A0(gobj);
+}
 
 /// #it_802817A0
 
@@ -209,7 +225,23 @@ void itDosei_UnkMotion5_Phys(Item_GObj* gobj)
     it_80274658(gobj, it_804D6D28->x68_float);
 }
 
-/// #it_3F14_Logic7_EnteredAir
+void it_3F14_Logic7_EnteredAir(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    s32 motion_id;
+    if (!ip->xDC8_word.flags.x19) {
+        motion_id = 0xA;
+    } else {
+        motion_id = 6;
+    }
+    Item_80268E5C(gobj, motion_id, 3);
+    {
+        HSD_JObj* jobj = gobj->hsd_obj;
+        ip->x5D0_animFrameSpeed = 1.0f;
+        lb_8000BA0C(jobj, 1.0f);
+    }
+    ip->owner = NULL;
+}
 
 bool itDosei_UnkMotion6_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itgamewatchchef.c b/src/melee/it/items/itgamewatchchef.c
index 1cc0a9cb0e..4849074d92 100644
--- a/src/melee/it/items/itgamewatchchef.c
+++ b/src/melee/it/items/itgamewatchchef.c
@@ -44,7 +44,21 @@ void it_802C875C(Item_GObj* gobj)
     Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
 }
 
-/// #itGamewatchchef_UnkMotion1_Anim
+bool itGamewatchchef_UnkMotion1_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    ip->xD44_lifeTimer -= 1.0f;
+    if (ip->xD44_lifeTimer <= 0.0f) {
+        ip->xD44_lifeTimer = 0.0f;
+        return true;
+    }
+    if ((s32) ip->xD44_lifeTimer & 2) {
+        it_8026BB20(gobj);
+    } else {
+        it_8026BB44(gobj);
+    }
+    return false;
+}
 
 void itGamewatchchef_UnkMotion1_Phys(Item_GObj* gobj) {}
 
diff --git a/src/melee/it/items/ithitodeman.c b/src/melee/it/items/ithitodeman.c
index 8253feadc7..9b355f5d7d 100644
--- a/src/melee/it/items/ithitodeman.c
+++ b/src/melee/it/items/ithitodeman.c
@@ -136,7 +136,17 @@ bool itHitodeman_UnkMotion2_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itHitodeman_UnkMotion2_Phys
+void itHitodeman_UnkMotion2_Phys(Item_GObj* gobj)
+{
+    if (it_8027A09C(gobj)) {
+        Item* ip;
+        it_80273454(gobj);
+        ip = GET_ITEM(gobj);
+        Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+        ip->entered_hitlag = efLib_PauseAll;
+        ip->exited_hitlag = efLib_ResumeAll;
+    }
+}
 
 bool itHitodeman_UnkMotion2_Coll(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ithouou.c b/src/melee/it/items/ithouou.c
index 52ccd57a1d..6dfcc1649b 100644
--- a/src/melee/it/items/ithouou.c
+++ b/src/melee/it/items/ithouou.c
@@ -10,6 +10,7 @@
 #include "it/it_2725.h"
 #include "it/itCommonItems.h"
 #include "it/item.h"
+#include "mp/mplib.h"
 
 /// #it_2725_Logic18_Spawned
 
@@ -115,7 +116,25 @@ bool itHouou_UnkMotion3_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itHouou_UnkMotion3_Phys
+void itHouou_UnkMotion3_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itHououAttr* attr = ip->xC4_article_data->x4_specialAttributes;
+    s32 timer;
+
+    it_8027A344(gobj);
+
+    timer = ip->xDD4_itemVar.pokemon.timer;
+    ip->xDD4_itemVar.pokemon.timer = timer - 1;
+
+    if (timer != 0) {
+        ip->x40_vel.z = attr->x10;
+        return;
+    }
+
+    ip->x40_vel.z = 0.0f;
+    it_802D290C(gobj);
+}
 
 bool itHouou_UnkMotion3_Coll(Item_GObj* gobj)
 {
@@ -276,7 +295,18 @@ bool it_802D2F3C(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802D2F70
+void it_802D2F70(Item_GObj* gobj)
+{
+    Vec3 sp10;
+    Item* ip = GET_ITEM(gobj);
+
+    if (mpLib_80054ED8(ip->xDD4_itemVar.pokemon.timer)) {
+        mpGetSpeed(ip->xDD4_itemVar.pokemon.timer, &ip->pos, &sp10);
+        ip->pos.x += sp10.x;
+        ip->pos.y += sp10.y;
+        ip->pos.z += sp10.z;
+    }
+}
 
 bool it_802D2FE8(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itkyasarin.c b/src/melee/it/items/itkyasarin.c
index 3357128107..0f9ea58bf1 100644
--- a/src/melee/it/items/itkyasarin.c
+++ b/src/melee/it/items/itkyasarin.c
@@ -38,9 +38,35 @@ void it_802ECEB0(Item_GObj* gobj)
     Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
 }
 
-/// #itKyasarin_UnkMotion0_Anim
+bool itKyasarin_UnkMotion0_Anim(Item_GObj* gobj)
+{
+    if (!it_80272C6C(gobj)) {
+        Item* ip = GET_ITEM(gobj);
+        itOldottoseaAttributes* attr =
+            ip->xC4_article_data->x4_specialAttributes;
+        ip->x40_vel.x = attr->x4 * ip->facing_dir;
+        ip->x40_vel.z = 0.0F;
+        ip->x40_vel.y = 0.0F;
+        Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    }
+    return false;
+}
 
-/// #itKyasarin_UnkMotion0_Coll
+bool itKyasarin_UnkMotion0_Coll(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itKyasarinAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+    if (ip->facing_dir == 1.0f) {
+        if (ip->pos.x >= attr->x2C) {
+            it_802ECFE0(gobj);
+        }
+    } else {
+        if (ip->pos.x <= attr->x28) {
+            it_802ECFE0(gobj);
+        }
+    }
+    return false;
+}
 
 void it_802ECFE0(Item_GObj* gobj)
 {
@@ -81,7 +107,18 @@ void it_802ED0D0(Item_GObj* gobj)
     Item_80268E5C(gobj, 2, ITEM_ANIM_UPDATE);
 }
 
-/// #itKyasarin_UnkMotion2_Anim
+bool itKyasarin_UnkMotion2_Anim(Item_GObj* gobj)
+{
+    if (!it_80272C6C(gobj)) {
+        Item* ip = GET_ITEM(gobj);
+        itKyasarinAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+        ip->x40_vel.x = -attr->x4 * ip->facing_dir;
+        ip->x40_vel.z = 0.0F;
+        ip->x40_vel.y = 0.0F;
+        Item_80268E5C(gobj, 2, ITEM_ANIM_UPDATE);
+    }
+    return false;
+}
 
 /// #itKyasarin_UnkMotion2_Coll
 
@@ -95,7 +132,19 @@ void it_802ED25C(Item_GObj* gobj)
     Item_80268E5C(gobj, 4, ITEM_ANIM_UPDATE);
 }
 
-/// #itKyasarin_UnkMotion4_Anim
+bool itKyasarin_UnkMotion4_Anim(Item_GObj* gobj)
+{
+    if (!it_80272C6C(gobj)) {
+        Item* ip = GET_ITEM(gobj);
+        itKyasarinAttributes* attr =
+            ip->xC4_article_data->x4_specialAttributes;
+        ip->x40_vel.x = attr->x4 * ip->facing_dir;
+        ip->x40_vel.z = 0.0F;
+        ip->x40_vel.y = 0.0F;
+        Item_80268E5C(gobj, 4, ITEM_ANIM_UPDATE);
+    }
+    return false;
+}
 
 /// #itKyasarin_UnkMotion4_Coll
 
diff --git a/src/melee/it/items/itmatadogas.c b/src/melee/it/items/itmatadogas.c
index 53d87c2dc8..05a95990cd 100644
--- a/src/melee/it/items/itmatadogas.c
+++ b/src/melee/it/items/itmatadogas.c
@@ -4,6 +4,7 @@
 #include <platform.h>
 
 #include "ef/eflib.h"
+#include "ef/efsync.h"
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
@@ -104,9 +105,27 @@ bool itMatadogas_UnkMotion2_Coll(Item_GObj* gobj)
 
 /// #it_802CB4F0
 
-/// #it_2725_Logic32_Spawned
+void it_2725_Logic32_Spawned(Item_GObj* gobj)
+{
+    HSD_JObj* jobj = gobj->hsd_obj;
+    f32 scale = 1.0f;
+    PAD_STACK(4);
+    efSync_Spawn(0x45B, gobj, jobj, &scale);
+    efSync_Spawn(0x45D, gobj, jobj, &scale);
+    it_802CB798(gobj);
+}
 
-/// #it_2725_Logic33_Spawned
+void it_2725_Logic33_Spawned(Item_GObj* gobj)
+{
+    UNUSED u32 _pad;
+    f32 scale;
+    HSD_JObj* jobj = gobj->hsd_obj;
+    PAD_STACK(4);
+    scale = 1.0f;
+    efSync_Spawn(0x45C, gobj, jobj, &scale);
+    efSync_Spawn(0x45E, gobj, jobj, &scale);
+    it_802CB798(gobj);
+}
 
 void itMatadogas_Logic33_EvtUnk(Item_GObj* gobj, Item_GObj* ref_gobj)
 {
diff --git a/src/melee/it/items/itmatadogas.h b/src/melee/it/items/itmatadogas.h
index 9dc68f3f9e..596ed0e6ce 100644
--- a/src/melee/it/items/itmatadogas.h
+++ b/src/melee/it/items/itmatadogas.h
@@ -23,7 +23,7 @@
 /* 2CB690 */ void it_2725_Logic32_Spawned(Item_GObj*);
 /* 2CB704 */ void it_2725_Logic33_Spawned(Item_GObj*);
 /* 2CB778 */ void itMatadogas_Logic33_EvtUnk(Item_GObj*, Item_GObj*);
-/* 2CB798 */ UNK_RET it_802CB798(UNK_PARAMS);
+/* 2CB798 */ void it_802CB798(Item_GObj*);
 /* 2CB810 */ bool it_802CB810(Item_GObj* gobj);
 /* 2CB844 */ void it_802CB844(Item_GObj* gobj);
 /* 2CB8A4 */ bool it_802CB8A4(Item_GObj* gobj);
diff --git a/src/melee/it/items/itmewtwoshadowball.c b/src/melee/it/items/itmewtwoshadowball.c
index 9a2bdb0857..113250d6d4 100644
--- a/src/melee/it/items/itmewtwoshadowball.c
+++ b/src/melee/it/items/itmewtwoshadowball.c
@@ -7,6 +7,8 @@
 #include "it/it_2725.h"
 #include "it/item.h"
 
+#include <math.h>
+
 /// #it_802C4D10
 
 bool it_802C4F50(Item_GObj* gobj, CollData* cd)
@@ -70,7 +72,17 @@ bool itMewtwoshadowball_UnkMotion0_Coll(Item_GObj* gobj)
 
 /// #itMewtwoshadowball_UnkMotion8_Anim
 
-/// #itMewtwoshadowball_UnkMotion8_Phys
+void itMewtwoshadowball_UnkMotion8_Phys(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    if (ip->xDD4_itemVar.mewtwoshadowball.x4C > 0) {
+        ip->x40_vel.x =
+            ip->xDD4_itemVar.mewtwoshadowball.x4.y * cosf(ip->xDD4_itemVar.mewtwoshadowball.x4.x);
+        ip->x40_vel.y =
+            ip->xDD4_itemVar.mewtwoshadowball.x4.y * sinf(ip->xDD4_itemVar.mewtwoshadowball.x4.x);
+    }
+    it_802C4D10(gobj);
+}
 
 bool itMewtwoshadowball_UnkMotion8_Coll(Item_GObj* gobj)
 {
@@ -136,7 +148,18 @@ bool itMewtwoShadowball_Logic101_Absorbed(Item_GObj* arg0)
     return true;
 }
 
-/// #it_2725_Logic101_Reflected
+bool it_2725_Logic101_Reflected(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    ip->xDD4_itemVar.mewtwoshadowball.x4.x += M_PI;
+    while (ip->xDD4_itemVar.mewtwoshadowball.x4.x < 0.0f) {
+        ip->xDD4_itemVar.mewtwoshadowball.x4.x += M_TAU;
+    }
+    while (ip->xDD4_itemVar.mewtwoshadowball.x4.x > M_TAU) {
+        ip->xDD4_itemVar.mewtwoshadowball.x4.x -= M_TAU;
+    }
+    return false;
+}
 
 bool itMewtwoShadowball_Logic101_HitShield(Item_GObj* arg0)
 {
diff --git a/src/melee/it/items/itmewtwoshadowball.h b/src/melee/it/items/itmewtwoshadowball.h
index 91cc79e20a..0ec76afee9 100644
--- a/src/melee/it/items/itmewtwoshadowball.h
+++ b/src/melee/it/items/itmewtwoshadowball.h
@@ -9,7 +9,7 @@
 
 #include <dolphin/mtx.h>
 
-/* 2C4D10 */ UNK_RET it_802C4D10(UNK_PARAMS);
+/* 2C4D10 */ void it_802C4D10(Item_GObj*);
 /* 2C4F50 */ bool it_802C4F50(Item_GObj*, CollData*);
 /* 2C5000 */ HSD_GObj* it_802C5000(Item_GObj*, Vec3*, s32, s32, float);
 /* 2C519C */ void it_802C519C(Item_GObj*, Vec3*, s32, s32, float, float);
diff --git a/src/melee/it/items/itoldkuri.c b/src/melee/it/items/itoldkuri.c
index c71cc0f09b..c54ac9fcbd 100644
--- a/src/melee/it/items/itoldkuri.c
+++ b/src/melee/it/items/itoldkuri.c
@@ -32,7 +32,20 @@ bool itOldkuri_UnkMotion0_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itOldkuri_UnkMotion0_Phys
+void itOldkuri_UnkMotion0_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (ip->xDD4_itemVar.oldkuri.xDF8 == 0) {
+        if (ip->facing_dir == 1.0f) {
+            it_802D775C(gobj);
+        } else {
+            it_802D758C(gobj);
+        }
+    } else {
+        ip->xDD4_itemVar.oldkuri.xDF8--;
+    }
+    it_8027C0A8(gobj, 0.0f, 5.0f);
+}
 
 bool itOldkuri_UnkMotion0_Coll(Item_GObj* gobj)
 {
@@ -70,7 +83,16 @@ bool itOldkuri_UnkMotion1_Anim(Item_GObj* gobj)
 
 /// #itOldkuri_UnkMotion1_Phys
 
-/// #itOldkuri_UnkMotion1_Coll
+bool itOldkuri_UnkMotion1_Coll(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(8);
+    it_8026D62C(gobj, it_802D7AF0);
+    if (it_80276308(gobj) == 4 && ip->xDD4_itemVar.oldkuri.xDFC != 0) {
+        it_802D775C(gobj);
+    }
+    return it_8027C794(gobj);
+}
 
 void it_802D775C(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itoldottosea.c b/src/melee/it/items/itoldottosea.c
index 1b6f1804d7..d040c03085 100644
--- a/src/melee/it/items/itoldottosea.c
+++ b/src/melee/it/items/itoldottosea.c
@@ -157,7 +157,19 @@ void it_802E2E30(Item_GObj* gobj)
     }
 }
 
-/// #itOldottosea_UnkMotion7_Anim
+bool itOldottosea_UnkMotion7_Anim(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    PAD_STACK(8);
+    if (!it_80272C6C(gobj)) {
+        ip->xDD4_itemVar.oldottosea.x28 = 0;
+        ip = gobj->user_data;
+        it_8027CAD8(gobj);
+        it_8027C0A8(gobj, ip->facing_dir, 1.0f);
+        Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    }
+    return false;
+}
 
 /// #itOldottosea_UnkMotion7_Phys
 
diff --git a/src/melee/it/items/ittincle.c b/src/melee/it/items/ittincle.c
index 92e7d462c5..fb020e8148 100644
--- a/src/melee/it/items/ittincle.c
+++ b/src/melee/it/items/ittincle.c
@@ -220,7 +220,19 @@ bool itTincle_UnkMotion8_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802EC3F4
+void it_802EC3F4(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    ip->x40_vel.z = 0.0f;
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->xDC8_word.flags.x1A = 0;
+    ip->x70_nudge.z = 0.0f;
+    ip->x70_nudge.y = 0.0f;
+    ip->x70_nudge.x = 0.0f;
+    Item_80268E5C(gobj, 9, ITEM_ANIM_UPDATE);
+    it_802EC9E8(gobj);
+}
 
 bool itTincle_UnkMotion9_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itwhitebea.c b/src/melee/it/items/itwhitebea.c
index 31d63f9ba2..3538b1c0bf 100644
--- a/src/melee/it/items/itwhitebea.c
+++ b/src/melee/it/items/itwhitebea.c
@@ -75,7 +75,15 @@ bool itOldottosea_UnkMotion9_Coll(Item_GObj* gobj)
     return it_8027C824(gobj, NULL);
 }
 
-/// #it_802E3400
+void it_802E3400(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    ip->xDCC_flag.b3 = 1;
+    it_802762BC(ip);
+    it_8027B964(gobj, 0);
+    ip->xDD4_itemVar.whitebea.x18 = 0;
+    Item_80268E5C(gobj, 0xA, ITEM_ANIM_UPDATE);
+}
 
 bool itOldottosea_UnkMotion10_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itwhitebea.h b/src/melee/it/items/itwhitebea.h
index 3483546135..b03d8a486d 100644
--- a/src/melee/it/items/itwhitebea.h
+++ b/src/melee/it/items/itwhitebea.h
@@ -15,7 +15,7 @@
 /* 2E3364 */ bool itOldottosea_UnkMotion9_Anim(Item_GObj* gobj);
 /* 2E33AC */ void itOldottosea_UnkMotion9_Phys(Item_GObj* gobj);
 /* 2E33DC */ bool itOldottosea_UnkMotion9_Coll(Item_GObj* gobj);
-/* 2E3400 */ UNK_RET it_802E3400(UNK_PARAMS);
+/* 2E3400 */ void it_802E3400(Item_GObj*);
 /* 2E3470 */ bool itOldottosea_UnkMotion10_Anim(Item_GObj* gobj);
 /* 2E34B8 */ void itOldottosea_UnkMotion10_Phys(Item_GObj* gobj);
 /* 2E34BC */ bool itOldottosea_UnkMotion10_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/itzgshell.c b/src/melee/it/items/itzgshell.c
index 4243e3b4fe..5f55641c99 100644
--- a/src/melee/it/items/itzgshell.c
+++ b/src/melee/it/items/itzgshell.c
@@ -358,7 +358,15 @@ void itZrshell_UnkMotion11_Phys(Item_GObj* gobj)
 
 /// #itZrshell_UnkMotion11_Coll
 
-/// #it_802DF9F8
+void it_802DF9F8(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    it_802762BC(ip);
+    ip->xDC8_word.flags.x1F = 0;
+    it_802DFF14(gobj, 0);
+    it_802756D0(gobj);
+    Item_80268E5C((HSD_GObj*) gobj, 0xA, ITEM_ANIM_UPDATE);
+}
 
 /// #itZrshell_UnkMotion10_Anim
 
diff --git a/src/melee/it/items/itzgshell.h b/src/melee/it/items/itzgshell.h
index 00042c1b1b..2bd2198f7f 100644
--- a/src/melee/it/items/itzgshell.h
+++ b/src/melee/it/items/itzgshell.h
@@ -51,7 +51,7 @@
 /* 2DF724 */ bool itZrshell_UnkMotion11_Anim(Item_GObj*);
 /* 2DF93C */ void itZrshell_UnkMotion11_Phys(Item_GObj*);
 /* 2DF970 */ bool itZrshell_UnkMotion11_Coll(Item_GObj*);
-/* 2DF9F8 */ UNK_RET it_802DF9F8(UNK_PARAMS);
+/* 2DF9F8 */ void it_802DF9F8(Item_GObj*);
 /* 2DFA68 */ bool itZrshell_UnkMotion10_Anim(Item_GObj*);
 /* 2DFB90 */ void itZrshell_UnkMotion10_Phys(Item_GObj*);
 /* 2DFBAC */ bool itZrshell_UnkMotion10_Coll(Item_GObj*);
@@ -64,7 +64,7 @@
 /* 2DFE7C */ int fn_802DFE7C(Item_GObj* gobj);
 /* 2DFED4 */ void it_802DFED4(Item_GObj*, Item_GObj*);
 /* 2DFEF4 */ void itZGShell_Logic11_Destroyed(Item_GObj*);
-/* 2DFF14 */ UNK_RET it_802DFF14(UNK_PARAMS);
+/* 2DFF14 */ void it_802DFF14(Item_GObj*, s32);
 /* 2DFFB8 */ void it_802DFFB8(HSD_JObj*, Item*);
 /* 2E0100 */ Item_GObj* it_802E0100(s32, Vec3*, s32);
 /* 2E0408 */ bool itZRShell_Logic12_Clanked(Item_GObj*);
