diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index e9e1528e49..18593dab03 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -184,8 +184,11 @@ typedef struct itHououAttr {
 } itHououAttr;
 
 typedef struct itLugiaAttributes {
-    /* +0 */ f32 x0;
-    /* +4 */ f32 x4;
+    /* +00 */ f32 x0;
+    /* +04 */ f32 x4;
+    /* +08 */ f32 x8;
+    /* +0C */ f32 xC;
+    /* +10 */ f32 x10;
 } itLugiaAttributes;
 
 /// Box/Crate item attributes loaded from .dat file
@@ -639,7 +642,8 @@ typedef struct itHitodeman_ItemVars {
     /* +60 ip+E34 */ float x60;
     /* +64 ip+E38 */ float x64;
     /* +68 ip+E3C */ s32 x68;
-    /* +6C ip+E40 */ char pad_6C[0x1C];
+    /* +6C ip+E40 */ char pad_6C[0x18];
+    /* +84 ip+E58 */ float x84;
     /* +88 ip+E5C */ s32 x88;
     /* +8C ip+E60 */ char pad_8C[0x4];
     /* +90 ip+E64 */ HSD_GObj* x90;
@@ -1229,6 +1233,8 @@ typedef struct itWhiteBea_ItemVars {
 typedef struct itWhiteBeaAttributes {
     /* +0 */ s32* x0;
     /* +4 */ f32 x4;
+    /* +8 */ char pad_x8[0xC];
+    /* +14 */ s16 x14;
 } itWhiteBeaAttributes;
 
 typedef struct itOctarock_ItemVars {
@@ -1476,7 +1482,9 @@ typedef struct itCrazyHandBombAttributes {
 } itCrazyHandBombAttributes;
 
 typedef struct itLugia_ItemVars {
-    /* +00 ip+DD4 */ u8 x0_pad[0x7C];
+    /* +00 ip+DD4 */ u8 x0_pad[0x60];
+    /* +60 ip+E34 */ s32 x60;
+    /* +64 ip+E38 */ u8 x64_pad[0x18];
     /* +7C ip+E50 */ Vec3 xE50;
 } itLugia_ItemVars;
 
diff --git a/src/melee/it/items/itgreatfoxlaser.c b/src/melee/it/items/itgreatfoxlaser.c
index 33cc5b3589..4a3781d0ba 100644
--- a/src/melee/it/items/itgreatfoxlaser.c
+++ b/src/melee/it/items/itgreatfoxlaser.c
@@ -2,6 +2,7 @@
 
 #include "it/inlines.h"
 #include "it/it_26B1.h"
+#include "it/it_2725.h"
 #include "it/item.h"
 
 UNK_T it_802EAF28(Item_GObj* item_gobj)
@@ -11,7 +12,19 @@ UNK_T it_802EAF28(Item_GObj* item_gobj)
 
 /// #it_802EAF34
 
-/// #it_802EB1EC
+void it_802EB1EC(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(16);
+    it_8027B730(gobj);
+    it_8027C56C(gobj, ip->facing_dir);
+    ip->xDCC_flag.b3 = 0;
+    ip->xD5C = 0;
+    ip->xDC8_word.flags.x15 = 0;
+    it_8027542C(gobj);
+    ip->xDD4_itemVar.greatfoxlaser.x36 = 1;
+    it_802EB268(gobj);
+}
 
 void it_802EB268(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ithitodeman.c b/src/melee/it/items/ithitodeman.c
index 9b355f5d7d..26af100534 100644
--- a/src/melee/it/items/ithitodeman.c
+++ b/src/melee/it/items/ithitodeman.c
@@ -22,7 +22,9 @@ typedef struct {
     float x8;
     float xC;
     float x10;
-    char pad_0[0x40 - 0x14];
+    float x14;
+    float x18;
+    char pad_0[0x40 - 0x1C];
     s32 x40;
     s32 x44;
     u8 _pad2[0x4];
@@ -70,7 +72,15 @@ void it_802D43EC(Item_GObj* gobj)
     }
 }
 
-/// #it_802D4494
+void it_802D4494(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itHitodemanAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    s32 randi;
+    PAD_STACK(8);
+    randi = HSD_Randi((s32) (attrs->x14 - attrs->x18));
+    ip->xDD4_itemVar.hitodeman.x84 = attrs->x18 + (f32) randi;
+}
 
 void it_802D4510(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ithitodeman.h b/src/melee/it/items/ithitodeman.h
index 5b750a43c9..45cbd34b4b 100644
--- a/src/melee/it/items/ithitodeman.h
+++ b/src/melee/it/items/ithitodeman.h
@@ -11,7 +11,7 @@
 /* 2D43AC */ UNK_RET it_802D43AC(UNK_PARAMS);
 /* 2D43B0 */ void it_802D43B0(Item_GObj*, Item_GObj*);
 /* 2D43EC */ void it_802D43EC(Item_GObj* gobj);
-/* 2D4494 */ UNK_RET it_802D4494(UNK_PARAMS);
+/* 2D4494 */ void it_802D4494(Item_GObj* gobj);
 /* 2D4510 */ void it_802D4510(Item_GObj* gobj);
 /* 2D4564 */ bool it_802D4564(Item_GObj* gobj);
 /* 2D472C */ void it_802D472C(Item_GObj* gobj);
diff --git a/src/melee/it/items/itkirby_2F23.c b/src/melee/it/items/itkirby_2F23.c
index 8b0be0eb97..19b66b7f1d 100644
--- a/src/melee/it/items/itkirby_2F23.c
+++ b/src/melee/it/items/itkirby_2F23.c
@@ -5,6 +5,9 @@
 
 #include "ft/chara/ftKirby/ftKb_Init.h"
 #include "it/inlines.h"
+#include "it/it_26B1.h"
+#include "it/it_2725.h"
+#include "it/item.h"
 
 double it_802F23AC(int* obj, float* pos)
 {
@@ -47,7 +50,17 @@ bool itKirby_2F23_UnkMotion0_Coll(Item_GObj* gobj)
 
 /// #it_802F258C
 
-/// #it_802F2810
+void it_802F2810(HSD_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    Item_80268E5C(gobj, 1, 0);
+    ip->on_accessory = it_802F289C;
+    it_8026BB44((Item_GObj*) gobj);
+    it_802756D0((Item_GObj*) gobj);
+    it_802762BC(ip);
+    it_80274ECC((Item_GObj*) gobj, 1);
+    it_80273408((Item_GObj*) gobj);
+}
 
 bool itKirby_2F23_UnkMotion1_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itlugia.c b/src/melee/it/items/itlugia.c
index 36d57e6172..2cfd4bac67 100644
--- a/src/melee/it/items/itlugia.c
+++ b/src/melee/it/items/itlugia.c
@@ -112,7 +112,18 @@ bool itLugia_UnkMotion3_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itLugia_UnkMotion3_Phys
+void itLugia_UnkMotion3_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itLugiaAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    it_8027A344(gobj);
+    if (ip->xDD4_itemVar.lugia.x60-- != 0) {
+        ip->x40_vel.z = attrs->x10;
+    } else {
+        ip->x40_vel.z = 0.0F;
+        it_802D1830(gobj);
+    }
+}
 
 bool itLugia_UnkMotion3_Coll(Item_GObj* gobj)
 {
@@ -234,7 +245,17 @@ void it_802D23D4(Item_GObj* gobj, Item_GObj* ref_gobj)
     it_8026B894(gobj, ref_gobj);
 }
 
-/// #it_802D23F4
+void it_802D23F4(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itLugiaAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    ip->xD44_lifeTimer = attrs->x0;
+    it_80274740(gobj);
+    Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    ip->entered_hitlag = efLib_PauseAll;
+    ip->exited_hitlag = efLib_ResumeAll;
+    it_8026B3A8(gobj);
+}
 
 bool it_802D246C(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itlugia.h b/src/melee/it/items/itlugia.h
index e9f37883a0..86b63e45a7 100644
--- a/src/melee/it/items/itlugia.h
+++ b/src/melee/it/items/itlugia.h
@@ -21,7 +21,7 @@
 /* 2D1720 */ bool itLugia_UnkMotion3_Anim(Item_GObj* gobj);
 /* 2D17B0 */ void itLugia_UnkMotion3_Phys(Item_GObj* gobj);
 /* 2D1828 */ bool itLugia_UnkMotion3_Coll(Item_GObj* gobj);
-/* 2D1830 */ UNK_RET it_802D1830(UNK_PARAMS);
+/* 2D1830 */ void it_802D1830(Item_GObj* gobj);
 /* 2D1930 */ bool itLugia_UnkMotion4_Anim(Item_GObj* gobj);
 /* 2D19C0 */ void itLugia_UnkMotion4_Phys(Item_GObj* gobj);
 /* 2D1A3C */ bool itLugia_UnkMotion4_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/itmatadogas.c b/src/melee/it/items/itmatadogas.c
index 05a95990cd..6f0b052b8b 100644
--- a/src/melee/it/items/itmatadogas.c
+++ b/src/melee/it/items/itmatadogas.c
@@ -132,7 +132,18 @@ void itMatadogas_Logic33_EvtUnk(Item_GObj* gobj, Item_GObj* ref_gobj)
     it_8026B894(gobj, ref_gobj);
 }
 
-/// #it_802CB798
+void it_802CB798(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    itMatadogasAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(8);
+    ip->xD44_lifeTimer = attrs->x0;
+    it_80274740(gobj);
+    Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    ip->entered_hitlag = efLib_PauseAll;
+    ip->exited_hitlag = efLib_ResumeAll;
+    it_8026B3A8(gobj);
+}
 
 bool it_802CB810(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itnessbat.c b/src/melee/it/items/itnessbat.c
index 43cf986b6d..17ff12cd5c 100644
--- a/src/melee/it/items/itnessbat.c
+++ b/src/melee/it/items/itnessbat.c
@@ -21,7 +21,18 @@ void it_802AD6B8(Item_GObj* gobj)
     }
 }
 
-/// #it_2725_Logic58_PickedUp
+void it_2725_Logic58_PickedUp(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if ((ip->xDAC_itcmd_var0 | ip->xDB0_itcmd_var1 | ip->xDB4_itcmd_var2 |
+         ip->xDB8_itcmd_var3) != 0)
+    {
+        Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
+    } else {
+        Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    }
+    Item_802694CC(gobj);
+}
 
 /// #itNessbat_UnkMotion0_Anim
 
diff --git a/src/melee/it/items/itnesspkthunderball.c b/src/melee/it/items/itnesspkthunderball.c
index 8ec99bd16e..d65f7a2b18 100644
--- a/src/melee/it/items/itnesspkthunderball.c
+++ b/src/melee/it/items/itnesspkthunderball.c
@@ -21,7 +21,35 @@ static inline void normalizeAngle(f32* angle)
     }
 }
 
-/// #it_802AB3F0
+void it_802AB3F0(Item_GObj* gobj, void* out, s32 idx)
+{
+    Item* ip;
+
+    if (gobj != NULL) {
+        ip = gobj->user_data;
+        if (ip != NULL) {
+            if (out == NULL) {
+                return;
+            }
+            ip = (Item*) ((u8*) ip + idx * 0x18);
+            *(Vec3*) out = ip->xDD4_itemVar.pkthunder.xDEC;
+            return;
+        }
+        if (out == NULL) {
+            return;
+        }
+        ((Vec3*) out)->z = 0.0F;
+        ((Vec3*) out)->y = 0.0F;
+        ((Vec3*) out)->x = 0.0F;
+        return;
+    }
+    if (out == NULL) {
+        return;
+    }
+    ((Vec3*) out)->z = 0.0F;
+    ((Vec3*) out)->y = 0.0F;
+    ((Vec3*) out)->x = 0.0F;
+}
 
 void it_802AB468(Item_GObj* gobj, f32* out, s32 unused)
 {
diff --git a/src/melee/it/items/itoldkuri.c b/src/melee/it/items/itoldkuri.c
index c54ac9fcbd..144e8fe09f 100644
--- a/src/melee/it/items/itoldkuri.c
+++ b/src/melee/it/items/itoldkuri.c
@@ -81,7 +81,17 @@ bool itOldkuri_UnkMotion1_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itOldkuri_UnkMotion1_Phys
+void itOldkuri_UnkMotion1_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(8);
+    if (ip->xDD4_itemVar.oldkuri.xDFC != 0 && it_802750E8(gobj, 2) != 0) {
+        it_802D775C(gobj);
+    }
+    it_8027C8D0(&ip->x40_vel, &ip->x378_itemColl.floor.normal,
+                ip->facing_dir);
+    it_8027C0A8(gobj, 0.0F, 5.0F);
+}
 
 bool itOldkuri_UnkMotion1_Coll(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itwhitebea.c b/src/melee/it/items/itwhitebea.c
index 3538b1c0bf..25cad9f3a9 100644
--- a/src/melee/it/items/itwhitebea.c
+++ b/src/melee/it/items/itwhitebea.c
@@ -360,9 +360,31 @@ bool itWhitebea_UnkMotion6_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-/// #it_802E4558
+void grIceMt_801FA6D8(void);
 
-/// #itWhitebea_UnkMotion7_Anim
+void it_802E4558(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    it_802762B0(ip);
+    Item_8026AE84(ip, 0x46, 0x7F, 0x40);
+    Item_8026AE84(ip, 0x13C, 0x7F, 0x40);
+    Item_80268E5C(gobj, 7, ITEM_ANIM_UPDATE);
+    grIceMt_801FA6D8();
+}
+
+bool itWhitebea_UnkMotion7_Anim(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    itWhiteBeaAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(16);
+    if (!it_80272C6C(gobj)) {
+        ip->xDD4_itemVar.whitebea.x3C = 1;
+        ip->xDD4_itemVar.whitebea.x40 = attrs->x14;
+        it_8027CAD8(gobj);
+        Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    }
+    return false;
+}
 
 void itWhitebea_UnkMotion7_Phys(Item_GObj* gobj) {}
 
