diff --git a/config/GALE01/splits.txt b/config/GALE01/splits.txt
index 2dceab2233..e7113cd3a3 100644
--- a/config/GALE01/splits.txt
+++ b/config/GALE01/splits.txt
@@ -3538,12 +3538,8 @@ melee/it/items/ittosakinto.c:
 	.sdata2     start:0x804DD318 end:0x804DD330
 
 melee/it/items/itchicorita.c:
-	.text       start:0x802C9588 end:0x802C9B20
-	.data       start:0x803F7A98 end:0x803F7AC8
-
-melee/it/items/itchicoritaleaf.c:
-	.text       start:0x802C9B20 end:0x802C9D40
-	.data       start:0x803F7AC8 end:0x803F7AD8
+	.text       start:0x802C9588 end:0x802C9D40
+	.data       start:0x803F7A98 end:0x803F7AD8
 	.sdata2     start:0x804DD330 end:0x804DD338
 
 melee/it/items/itkabigon.c:
diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 5810925865..10426b50d4 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -33264,8 +33264,8 @@ it_804DD314 = .sdata2:0x804DD314; // type:object size:0x4 scope:global align:4 d
 @188 = .sdata2:0x804DD320; // type:object size:0x4 scope:local align:4 data:float
 @237 = .sdata2:0x804DD324; // type:object size:0x4 scope:local align:4 data:float
 @238 = .sdata2:0x804DD328; // type:object size:0x4 scope:local align:4 data:float
-it_804DD330 = .sdata2:0x804DD330; // type:object size:0x4 scope:global align:4 data:float
-it_804DD334 = .sdata2:0x804DD334; // type:object size:0x4 scope:global align:4 data:float
+@153 = .sdata2:0x804DD330; // type:object size:0x4 scope:local align:4 data:float
+@274 = .sdata2:0x804DD334; // type:object size:0x4 scope:local align:4 data:float
 it_804DD338 = .sdata2:0x804DD338; // type:object size:0x4 scope:global align:4 data:float
 it_804DD33C = .sdata2:0x804DD33C; // type:object size:0x4 scope:global align:4 data:float
 it_804DD340 = .sdata2:0x804DD340; // type:object size:0x4 scope:global align:4 data:float
diff --git a/configure.py b/configure.py
index a59f10ec7d..79f8514d1d 100755
--- a/configure.py
+++ b/configure.py
@@ -1214,8 +1214,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/itclinkmilk.c"),
             # Pokémon
             Object(Matching, "melee/it/items/ittosakinto.c"),
-            Object(NonMatching, "melee/it/items/itchicorita.c"),
-            Object(NonMatching, "melee/it/items/itchicoritaleaf.c"),
+            Object(Matching, "melee/it/items/itchicorita.c"),
             Object(NonMatching, "melee/it/items/itkabigon.c"),
             Object(NonMatching, "melee/it/items/itkamex.c"),
             Object(NonMatching, "melee/it/items/itmatadogas.c"),
diff --git a/src/melee/it/it_2725.c b/src/melee/it/it_2725.c
index c96e89e5ea..954cc553bc 100644
--- a/src/melee/it/it_2725.c
+++ b/src/melee/it/it_2725.c
@@ -35,7 +35,6 @@
 #include "it/items/itcapsule.h"
 #include "it/items/itcerebi.h"
 #include "it/items/itchicorita.h"
-#include "it/items/itchicoritaleaf.h"
 #include "it/items/itclimbersblizzard.h"
 #include "it/items/itclimbersice.h"
 #include "it/items/itclimbersstring.h"
diff --git a/src/melee/it/items/itchicorita.c b/src/melee/it/items/itchicorita.c
index 161a8420ac..6a932a439c 100644
--- a/src/melee/it/items/itchicorita.c
+++ b/src/melee/it/items/itchicorita.c
@@ -8,7 +8,6 @@
 #include "it/it_2725.h"
 #include "it/itCommonItems.h"
 #include "it/item.h"
-#include "it/items/itchicoritaleaf.h"
 
 ItemStateTable it_803F7A98[] = {
     { 0, itChicorita_UnkMotion0_Anim, itChicorita_UnkMotion0_Phys,
@@ -19,6 +18,11 @@ ItemStateTable it_803F7A98[] = {
       itChicorita_UnkMotion2_Coll },
 };
 
+ItemStateTable it_803F7AC8[] = {
+    { 0, itChicoritaleaf_UnkMotion0_Anim, itChicoritaleaf_UnkMotion0_Phys,
+      itChicoritaleaf_UnkMotion0_Coll },
+};
+
 void it_802C9588(Item_GObj* item_gobj)
 {
     Article* article;
@@ -104,41 +108,31 @@ bool itChicorita_UnkMotion0_Anim(HSD_GObj* item_gobj)
     return false;
 }
 
-void itChicorita_UnkMotion0_Phys(HSD_GObj* item_gobj)
+static inline void itChicorita_Phys(HSD_GObj* item_gobj)
 {
     Item* item;
-    ItemAttr* attr;
-    f32 fall_speed;
+    f32 chicorita_fall_speed;
     s32 fall_speed_dir;
-    f32 item_vel_y;
     s32 item_vel_y_dir;
 
     item = item_gobj->user_data;
-    it_8027A344((Item_GObj*) item_gobj);
+    it_8027A344(item_gobj);
     if (item->ground_or_air == GA_Air) {
-        attr = item->xCC_item_attr;
-        fall_speed = attr->x10_fall_speed;
-        if (fall_speed < 0.0f) {
-            fall_speed_dir = -1;
-        } else {
-            fall_speed_dir = 1;
-        }
-        item_vel_y = item->xDD4_itemVar.chicorita.x64;
-        if (item_vel_y < 0.0f) {
-            item_vel_y_dir = -1;
-        } else {
-            item_vel_y_dir = 1;
-        }
+        fall_speed_dir = item->xCC_item_attr->x10_fall_speed < 0.0f ? -1 : 1;
+        chicorita_fall_speed = item->xDD4_itemVar.chicorita.x64;
+        item_vel_y_dir = chicorita_fall_speed < 0.0f ? -1 : 1;
         if (item_vel_y_dir != fall_speed_dir) {
-            if (item_vel_y < 0.0f) {
-                item_vel_y = -item_vel_y;
+            if (chicorita_fall_speed < 0.0f) {
+                chicorita_fall_speed = -chicorita_fall_speed;
             }
-            if (item_vel_y < attr->x14_fall_speed_max) {
-                goto block_11;
+            if (chicorita_fall_speed < item->xCC_item_attr->x14_fall_speed_max)
+            {
+            block_11:
+                item->xDD4_itemVar.chicorita.x64 -=
+                    item->xCC_item_attr->x10_fall_speed;
             }
         } else {
-        block_11:
-            item->xDD4_itemVar.chicorita.x64 -= fall_speed;
+            goto block_11;
         }
         item->x40_vel.y = item->xDD4_itemVar.chicorita.x64;
         return;
@@ -146,6 +140,11 @@ void itChicorita_UnkMotion0_Phys(HSD_GObj* item_gobj)
     item->xDD4_itemVar.chicorita.x64 = 0.0f;
 }
 
+void itChicorita_UnkMotion0_Phys(HSD_GObj* item_gobj)
+{
+    itChicorita_Phys(item_gobj);
+}
+
 bool itChicorita_UnkMotion0_Coll(HSD_GObj* item_gobj)
 {
     if (((Item*) item_gobj->user_data)->ground_or_air == GA_Air) {
@@ -191,44 +190,7 @@ bool itChicorita_UnkMotion1_Anim(HSD_GObj* item_gobj)
 
 void itChicorita_UnkMotion1_Phys(HSD_GObj* item_gobj)
 {
-    Item* item;
-    ItemAttr* attr;
-    f32 fall_speed;
-    s32 fall_speed_dir;
-    f32 item_vel_y;
-    s32 item_vel_y_dir;
-
-    item = item_gobj->user_data;
-    it_8027A344((Item_GObj*) item_gobj);
-    if (item->ground_or_air == GA_Air) {
-        attr = item->xCC_item_attr;
-        fall_speed = attr->x10_fall_speed;
-        if (fall_speed < 0.0f) {
-            fall_speed_dir = -1;
-        } else {
-            fall_speed_dir = 1;
-        }
-        item_vel_y = item->xDD4_itemVar.chicorita.x64;
-        if (item_vel_y < 0.0f) {
-            item_vel_y_dir = -1;
-        } else {
-            item_vel_y_dir = 1;
-        }
-        if (item_vel_y_dir != fall_speed_dir) {
-            if (item_vel_y < 0.0f) {
-                item_vel_y = -item_vel_y;
-            }
-            if (item_vel_y < attr->x14_fall_speed_max) {
-                goto block_11;
-            }
-        } else {
-        block_11:
-            item->xDD4_itemVar.chicorita.x64 -= fall_speed;
-        }
-        item->x40_vel.y = item->xDD4_itemVar.chicorita.x64;
-        return;
-    }
-    item->xDD4_itemVar.chicorita.x64 = 0.0f;
+    itChicorita_Phys(item_gobj);
 }
 
 bool itChicorita_UnkMotion1_Coll(HSD_GObj* item_gobj)
@@ -262,3 +224,83 @@ bool itChicorita_UnkMotion2_Coll(HSD_GObj* item_gobj)
 {
     return it_8027A118((Item_GObj*) item_gobj, it_802C963C);
 }
+
+void it_802C9B20(Item_GObj* chicorita_gobj)
+{
+    SpawnItem spawn;
+    Item* chicorita;
+    itChicoritaLeafAttr* attr;
+
+    chicorita = GET_ITEM((HSD_GObj*) chicorita_gobj);
+    attr = chicorita->xC4_article_data->x4_specialAttributes;
+    spawn.prev_pos = chicorita->pos;
+    spawn.prev_pos.x += attr->x8 * chicorita->facing_dir;
+    spawn.prev_pos.y += attr->xC;
+    it_8026BB88(chicorita_gobj, &spawn.pos);
+    spawn.facing_dir = chicorita->facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.x = attr->x10 * chicorita->facing_dir;
+    spawn.vel.z = spawn.vel.y = 0.0f;
+    spawn.x40 = 0;
+    spawn.kind = Pokemon_Chicorita_Leaf;
+    spawn.x0_parent_gobj = chicorita->owner;
+    spawn.x4_parent_gobj2 = (HSD_GObj*) chicorita_gobj;
+    spawn.x44_flag.b0 = 1;
+    if (Item_80268B18(&spawn)) {
+        Item_8026AE84(chicorita, 0x2710, 0x7F, 0x40);
+    }
+}
+
+void itChicoritaLeaf_Logic30_Spawned(Item_GObj* item_gobj)
+{
+    Item* item;
+    PAD_STACK(12);
+
+    item = GET_ITEM((HSD_GObj*) item_gobj);
+    item->xD44_lifeTimer =
+        ((itChicoritaLeafAttr*) item->xC4_article_data->x4_specialAttributes)
+            ->timer;
+    it_80274740(item_gobj);
+    it_802C9CC0(item_gobj);
+    it_8026B3A8(item_gobj);
+    it_80274C60(item_gobj);
+}
+
+bool itChicoritaLeaf_Logic30_HitShield(Item_GObj* item_gobj)
+{
+    return false;
+}
+
+void itChicoritaLeaf_Logic30_EvtUnk(Item_GObj* item_gobj, HSD_GObj* ref_gobj)
+{
+    it_8026B894(item_gobj, ref_gobj);
+}
+
+bool itChicoritaLeaf_Logic30_Reflected(Item_GObj* item_gobj)
+{
+    return it_80273030(item_gobj);
+}
+
+void it_802C9CC0(Item_GObj* item_gobj)
+{
+    Item_80268E5C((HSD_GObj*) item_gobj, 0, ITEM_ANIM_UPDATE);
+}
+
+bool itChicoritaleaf_UnkMotion0_Anim(HSD_GObj* item_gobj)
+{
+    Item* item;
+
+    item = item_gobj->user_data;
+    if (item->xD44_lifeTimer <= 0.0f) {
+        return true;
+    }
+    item->xD44_lifeTimer--;
+    return false;
+}
+
+void itChicoritaleaf_UnkMotion0_Phys(HSD_GObj* item_gobj) {}
+
+bool itChicoritaleaf_UnkMotion0_Coll(HSD_GObj* item_gobj)
+{
+    return it_8026DFB0((Item_GObj*) item_gobj);
+}
diff --git a/src/melee/it/items/itchicorita.h b/src/melee/it/items/itchicorita.h
index 97b6294996..fd8936ca3a 100644
--- a/src/melee/it/items/itchicorita.h
+++ b/src/melee/it/items/itchicorita.h
@@ -25,6 +25,16 @@
 /* 2C9AB4 */ bool itChicorita_UnkMotion2_Anim(HSD_GObj*);
 /* 2C9AD8 */ void itChicorita_UnkMotion2_Phys(HSD_GObj*);
 /* 2C9AF8 */ bool itChicorita_UnkMotion2_Coll(HSD_GObj*);
+/* 2C9B20 */ void it_802C9B20(Item_GObj*);
+/* 2C9C20 */ void itChicoritaLeaf_Logic30_Spawned(Item_GObj*);
+/* 2C9C78 */ bool itChicoritaLeaf_Logic30_HitShield(Item_GObj*);
+/* 2C9C80 */ void itChicoritaLeaf_Logic30_EvtUnk(Item_GObj*, HSD_GObj*);
+/* 2C9CA0 */ bool itChicoritaLeaf_Logic30_Reflected(Item_GObj*);
+/* 2C9CC0 */ void it_802C9CC0(Item_GObj*);
+/* 2C9CE8 */ bool itChicoritaleaf_UnkMotion0_Anim(HSD_GObj*);
+/* 2C9D1C */ void itChicoritaleaf_UnkMotion0_Phys(HSD_GObj*);
+/* 2C9D20 */ bool itChicoritaleaf_UnkMotion0_Coll(HSD_GObj*);
 /* 3F7A98 */ extern ItemStateTable it_803F7A98[];
+/* 3F7AC8 */ extern ItemStateTable it_803F7AC8[];
 
 #endif
diff --git a/src/melee/it/items/itchicoritaleaf.c b/src/melee/it/items/itchicoritaleaf.c
deleted file mode 100644
index 1d437e7a22..0000000000
--- a/src/melee/it/items/itchicoritaleaf.c
+++ /dev/null
@@ -1,95 +0,0 @@
-#include "itchicoritaleaf.h"
-
-#include "it/forward.h"
-
-#include "it/inlines.h"
-#include "it/it_266F.h"
-#include "it/it_26B1.h"
-#include "it/it_2725.h"
-#include "it/itCommonItems.h"
-#include "it/item.h"
-
-ItemStateTable it_803F7AC8[] = {
-    { 0, itChicoritaleaf_UnkMotion0_Anim, itChicoritaleaf_UnkMotion0_Phys,
-      itChicoritaleaf_UnkMotion0_Coll },
-};
-
-void it_802C9B20(Item_GObj* chicorita_gobj)
-{
-    SpawnItem spawn;
-    Item* chicorita;
-    itChicoritaLeafAttr* attr;
-
-    chicorita = GET_ITEM((HSD_GObj*) chicorita_gobj);
-    attr = chicorita->xC4_article_data->x4_specialAttributes;
-    spawn.prev_pos = chicorita->pos;
-    spawn.prev_pos.x += attr->x8 * chicorita->facing_dir;
-    spawn.prev_pos.y += attr->xC;
-    it_8026BB88(chicorita_gobj, &spawn.pos);
-    spawn.facing_dir = chicorita->facing_dir;
-    spawn.x3C_damage = 0;
-    spawn.vel.x = attr->x10 * chicorita->facing_dir;
-    spawn.vel.z = spawn.vel.y = 0.0f;
-    spawn.x40 = 0;
-    spawn.kind = Pokemon_Chicorita_Leaf;
-    spawn.x0_parent_gobj = chicorita->owner;
-    spawn.x4_parent_gobj2 = (HSD_GObj*) chicorita_gobj;
-    spawn.x44_flag.b0 = 1;
-    if (Item_80268B18(&spawn)) {
-        Item_8026AE84(chicorita, 0x2710, 0x7F, 0x40);
-    }
-}
-
-void itChicoritaLeaf_Logic30_Spawned(Item_GObj* item_gobj)
-{
-    Item* item;
-    PAD_STACK(12);
-
-    item = GET_ITEM((HSD_GObj*) item_gobj);
-    item->xD44_lifeTimer =
-        ((itChicoritaLeafAttr*) item->xC4_article_data->x4_specialAttributes)
-            ->timer;
-    it_80274740(item_gobj);
-    it_802C9CC0(item_gobj);
-    it_8026B3A8(item_gobj);
-    it_80274C60(item_gobj);
-}
-
-bool itChicoritaLeaf_Logic30_HitShield(Item_GObj* item_gobj)
-{
-    return false;
-}
-
-void itChicoritaLeaf_Logic30_EvtUnk(Item_GObj* item_gobj, HSD_GObj* ref_gobj)
-{
-    it_8026B894(item_gobj, ref_gobj);
-}
-
-bool itChicoritaLeaf_Logic30_Reflected(Item_GObj* item_gobj)
-{
-    return it_80273030(item_gobj);
-}
-
-void it_802C9CC0(Item_GObj* item_gobj)
-{
-    Item_80268E5C((HSD_GObj*) item_gobj, 0, ITEM_ANIM_UPDATE);
-}
-
-bool itChicoritaleaf_UnkMotion0_Anim(HSD_GObj* item_gobj)
-{
-    Item* item;
-
-    item = item_gobj->user_data;
-    if (item->xD44_lifeTimer <= 0.0f) {
-        return true;
-    }
-    item->xD44_lifeTimer--;
-    return false;
-}
-
-void itChicoritaleaf_UnkMotion0_Phys(HSD_GObj* item_gobj) {}
-
-bool itChicoritaleaf_UnkMotion0_Coll(HSD_GObj* item_gobj)
-{
-    return it_8026DFB0((Item_GObj*) item_gobj);
-}
diff --git a/src/melee/it/items/itchicoritaleaf.h b/src/melee/it/items/itchicoritaleaf.h
deleted file mode 100644
index 3570c0a2d6..0000000000
--- a/src/melee/it/items/itchicoritaleaf.h
+++ /dev/null
@@ -1,22 +0,0 @@
-#ifndef GALE01_2C9B20
-#define GALE01_2C9B20
-
-#include <placeholder.h>
-#include <platform.h>
-
-#include "it/forward.h"
-
-#include "it/items/types.h"
-
-/* 2C9B20 */ void it_802C9B20(Item_GObj*);
-/* 2C9C20 */ void itChicoritaLeaf_Logic30_Spawned(Item_GObj*);
-/* 2C9C78 */ bool itChicoritaLeaf_Logic30_HitShield(Item_GObj*);
-/* 2C9C80 */ void itChicoritaLeaf_Logic30_EvtUnk(Item_GObj*, HSD_GObj*);
-/* 2C9CA0 */ bool itChicoritaLeaf_Logic30_Reflected(Item_GObj*);
-/* 2C9CC0 */ void it_802C9CC0(Item_GObj*);
-/* 2C9CE8 */ bool itChicoritaleaf_UnkMotion0_Anim(HSD_GObj*);
-/* 2C9D1C */ void itChicoritaleaf_UnkMotion0_Phys(HSD_GObj*);
-/* 2C9D20 */ bool itChicoritaleaf_UnkMotion0_Coll(HSD_GObj*);
-/* 3F7AC8 */ extern ItemStateTable it_803F7AC8[];
-
-#endif
