diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 65b7d484c1..d156645dd3 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -33494,10 +33494,10 @@ it_804DD824 = .sdata2:0x804DD824; // type:object size:0x4 scope:global align:4 d
 it_804DD828 = .sdata2:0x804DD828; // type:object size:0x4 scope:global align:4 data:float
 it_804DD82C = .sdata2:0x804DD82C; // type:object size:0x4 scope:global align:4 data:float
 it_804DD830 = .sdata2:0x804DD830; // type:object size:0x4 scope:global align:4 data:float
-it_804DD838 = .sdata2:0x804DD838; // type:object size:0x4 scope:global align:4 data:float
-it_804DD83C = .sdata2:0x804DD83C; // type:object size:0x4 scope:global align:4 data:float
-it_804DD840 = .sdata2:0x804DD840; // type:object size:0x4 scope:global align:4 data:float
-it_804DD848 = .sdata2:0x804DD848; // type:object size:0x8 scope:global align:8 data:double
+@189 = .sdata2:0x804DD838; // type:object size:0x4 scope:local align:4 data:float
+@190 = .sdata2:0x804DD83C; // type:object size:0x4 scope:local align:4 data:float
+@191 = .sdata2:0x804DD840; // type:object size:0x4 scope:local align:4 data:float
+@196 = .sdata2:0x804DD848; // type:object size:0x8 scope:local align:8 data:double
 @158 = .sdata2:0x804DD850; // type:object size:0x4 scope:local align:4 data:float
 @201 = .sdata2:0x804DD854; // type:object size:0x4 scope:local align:4 data:float
 @202 = .sdata2:0x804DD858; // type:object size:0x4 scope:local align:4 data:float
@@ -33578,10 +33578,10 @@ it_804DD9E0 = .sdata2:0x804DD9E0; // type:object size:0x8 scope:global align:8 d
 @252 = .sdata2:0x804DDA00; // type:object size:0x8 scope:local align:8 data:double
 @207 = .sdata2:0x804DDA08; // type:object size:0x4 scope:local align:4 data:float
 @207 = .sdata2:0x804DDA10; // type:object size:0x4 scope:local align:4 data:float
-it_804DDA18 = .sdata2:0x804DDA18; // type:object size:0x4 scope:global align:4 data:float
-it_804DDA1C = .sdata2:0x804DDA1C; // type:object size:0x4 scope:global align:4 data:float
-it_804DDA20 = .sdata2:0x804DDA20; // type:object size:0x4 scope:global align:4 data:float
-it_804DDA28 = .sdata2:0x804DDA28; // type:object size:0x8 scope:global align:8 data:double
+@156 = .sdata2:0x804DDA18; // type:object size:0x4 scope:local align:4 data:float
+@168 = .sdata2:0x804DDA1C; // type:object size:0x4 scope:local align:4 data:float
+@188 = .sdata2:0x804DDA20; // type:object size:0x4 scope:local align:4 data:float
+@196 = .sdata2:0x804DDA28; // type:object size:0x8 scope:local align:8 data:double
 @184 = .sdata2:0x804DDA30; // type:object size:0x4 scope:local align:4 data:float
 ifTime_804DDA38 = .sdata2:0x804DDA38; // type:object size:0x4 scope:global align:4 data:float
 ifTime_804DDA40 = .sdata2:0x804DDA40; // type:object size:0x8 scope:global align:8 data:double
diff --git a/configure.py b/configure.py
index 1d16d91e33..8318655a07 100755
--- a/configure.py
+++ b/configure.py
@@ -1258,7 +1258,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/it_2E5A.c"),
             Object(Matching, "melee/it/items/it_2E6A.c"),
             Object(NonMatching, "melee/it/items/itarwinglaser.c"),
-            Object(NonMatching, "melee/it/items/itoctarockstone.c"),
+            Object(Matching, "melee/it/items/itoctarockstone.c"),
             Object(Matching, "melee/it/items/itleadead.c"),
             Object(NonMatching, "melee/it/items/itgreatfoxlaser.c"),
             Object(NonMatching, "melee/it/items/ittincle.c"),
@@ -1272,9 +1272,9 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itcoin.c"),
             Object(NonMatching, "melee/it/items/itkirby_2F23.c"),
             Object(Matching, "melee/it/items/it_2F28.c"),
-            Object(Matching, "melee/it/items/it_2F2B.c"),
+            Object(Matching, "melee/it/items/ityoshitongue.c"),
             Object(Matching, "melee/it/items/itkirbyyoshispecialn.c"),
-            Object(NonMatching, "melee/it/items/it_27CF.c"),
+            Object(Matching, "melee/it/items/ityoshiegglay.c"),
         ],
     ),
     MeleeLib(
diff --git a/src/melee/ft/chara/ftFox/ftFx_SpecialN.c b/src/melee/ft/chara/ftFox/ftFx_SpecialN.c
index ba980e64c4..78e8c6a6ae 100644
--- a/src/melee/ft/chara/ftFox/ftFx_SpecialN.c
+++ b/src/melee/ft/chara/ftFox/ftFx_SpecialN.c
@@ -21,7 +21,7 @@
 #include "ftCommon/ftpickupitem.h"
 #include "ftFox/types.h"
 #include "it/it_26B1.h"
-#include "it/items/it_27CF.h"
+#include "it/items/itfoxblaster.h"
 #include "lb/lb_00B0.h"
 #include "lb/lbrefract.h"
 
diff --git a/src/melee/ft/chara/ftGameWatch/ftGw_AttackAir.c b/src/melee/ft/chara/ftGameWatch/ftGw_AttackAir.c
index a8b3d3c544..e9b49f179f 100644
--- a/src/melee/ft/chara/ftGameWatch/ftGw_AttackAir.c
+++ b/src/melee/ft/chara/ftGameWatch/ftGw_AttackAir.c
@@ -10,13 +10,12 @@
 
 #include "ftCommon/forward.h"
 
-#include "ftCommon/ftCo_Attack100.h"
 #include "ftCommon/ftCo_AttackAir.h"
 #include "ftCommon/ftCo_Landing.h"
 #include "ftCommon/ftCo_LandingAir.h"
 #include "ftGameWatch/types.h"
-#include "it/items/it_27CF.h"
 #include "it/items/itgamewatchbreath.h"
+#include "it/items/itgamewatchparachute.h"
 #include "it/items/itgamewatchturtle.h"
 #include "lb/lb_00B0.h"
 
diff --git a/src/melee/ft/chara/ftGameWatch/ftGw_AttackLw3.c b/src/melee/ft/chara/ftGameWatch/ftGw_AttackLw3.c
index d87da4a614..88ce91a650 100644
--- a/src/melee/ft/chara/ftGameWatch/ftGw_AttackLw3.c
+++ b/src/melee/ft/chara/ftGameWatch/ftGw_AttackLw3.c
@@ -10,7 +10,6 @@
 #include "ft/ftcommon.h"
 #include "ft/types.h"
 #include "ftCommon/ftCo_Attack1.h"
-#include "ftCommon/ftCo_Attack100.h"
 #include "ftCommon/ftCo_AttackHi3.h"
 #include "ftCommon/ftCo_AttackHi4.h"
 #include "ftCommon/ftCo_AttackLw3.h"
@@ -26,7 +25,7 @@
 #include "ftCommon/ftpickupitem.h"
 #include "ftGameWatch/types.h"
 #include "it/it_26B1.h"
-#include "it/items/it_27CF.h"
+#include "it/items/itgamewatchmanhole.h"
 #include "lb/lb_00B0.h"
 
 #include <dolphin/mtx.h>
diff --git a/src/melee/ft/chara/ftGameWatch/ftGw_SpecialS.c b/src/melee/ft/chara/ftGameWatch/ftGw_SpecialS.c
index 519be992eb..534f902909 100644
--- a/src/melee/ft/chara/ftGameWatch/ftGw_SpecialS.c
+++ b/src/melee/ft/chara/ftGameWatch/ftGw_SpecialS.c
@@ -13,8 +13,8 @@
 #include "ft/ftparts.h"
 #include "ft/types.h"
 #include "ftCommon/ftCo_Fall.h"
-#include "it/items/it_27CF.h"
 #include "it/items/itfoods.h"
+#include "it/items/itgamewatchjudge.h"
 #include "lb/lb_00B0.h"
 
 #include <dolphin/mtx.h>
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.c b/src/melee/ft/chara/ftKirby/ftKb_Init.c
index fa8dd9cc1e..a5187a002e 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.c
@@ -48,7 +48,6 @@
 #include "ftLink/types.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
-#include "it/items/it_27CF.h"
 #include "it/items/it_2ADA.h"
 #include "it/items/it_2F28.h"
 #include "it/items/itclimbersice.h"
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index f9a3f1040f..a1e57d13a6 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -3,76 +3,45 @@
 
 #include <placeholder.h>
 
-#include "cm/camera.h"
-#include "ef/efasync.h"
 #include "ef/eflib.h"
 #include "ef/efsync.h"
-#include "ft/chara/ftCommon/ftCo_CaptureWaitKirby.h"
 #include "ft/chara/ftCommon/ftCo_Damage.h"
-#include "ft/chara/ftCommon/ftCo_Escape.h"
 #include "ft/chara/ftCommon/ftCo_FallSpecial.h"
 #include "ft/chara/ftCommon/ftCo_Jump.h"
 #include "ft/chara/ftCommon/ftCo_KneeBend.h"
-#include "ft/chara/ftCommon/ftCo_Lift.h"
-#include "ft/chara/ftCommon/ftCo_Throw.h"
 #include "ft/chara/ftCommon/ftCo_Wait.h"
-#include "ft/chara/ftCommon/ftpickupitem.h"
 #include "ft/fighter.h"
 
 #include "ft/forward.h"
 
 #include "ft/ft_081B.h"
 #include "ft/ft_0877.h"
-#include "ft/ft_0881.h"
 #include "ft/ft_0892.h"
-#include "ft/ft_0C31.h"
-#include "ft/ftcamera.h"
 #include "ft/ftcliffcommon.h"
-#include "ft/ftcolanim.h"
-#include "ft/ftcoll.h"
 #include "ft/ftcommon.h"
 #include "ft/ftdata.h"
-#include "ft/ftdynamics.h"
 #include "ft/ftparts.h"
 #include "ft/ftwalkcommon.h"
 #include "ft/inlines.h"
 #include "ft/types.h"
-#include "ftCommon/ftCo_Attack100.h"
 #include "ftCommon/ftCo_CaptureKirby.h"
 #include "ftCommon/ftCo_Fall.h"
 #include "ftCommon/ftCo_Landing.h"
 
 #include "ftKirby/forward.h"
 
-#include "ftLink/types.h"
-#include "it/it_26B1.h"
-#include "it/item.h"
-#include "it/items/it_27CF.h"
 #include "it/items/it_2ADA.h"
 #include "it/items/it_2F28.h"
-#include "it/items/itclimbersice.h"
-#include "it/items/itgamewatchchef.h"
+#include "it/items/itfoxblaster.h"
 #include "it/items/itkirby_2F23.h"
 #include "it/items/itkirbycutterbeam.h"
-#include "it/items/itkirbygamewatchchefpan.h"
 #include "it/items/itkirbyhammer.h"
-#include "it/items/itkirbyyoshispecialn.h"
-#include "it/items/itlinkarrow.h"
-#include "it/items/itlinkbow.h"
-#include "it/items/itmewtwoshadowball.h"
-#include "it/items/itnesspkflush.h"
-#include "it/items/itpeachtoad.h"
-#include "it/items/itpeachtoadspore.h"
-#include "it/items/itpikachutjoltground.h"
-#include "it/items/itsamuschargeshot.h"
-#include "it/items/itseakneedleheld.h"
 #include "lb/lb_00B0.h"
 #include "lb/lbanim.h"
 #include "lb/lbvector.h"
 #include "melee/lb/lbrefract.h"
 #include "mp/mpcoll.h"
 #include "mp/mplib.h"
-#include "pl/player.h"
 
 #include <common_structs.h>
 #include <stddef.h>
@@ -257,8 +226,8 @@ void ftKb_SpecialAirHi1_Anim(Fighter_GObj* gobj)
     Fighter* fp = GET_FIGHTER(gobj);
 
     if (!ftAnim_IsFramesRemaining(gobj)) {
-        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirHi2, 0xA,
-                                  0.0f, 1.0f, 0.0f, NULL);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirHi2, 0xA, 0.0f, 1.0f,
+                                  0.0f, NULL);
         fp->pre_hitlag_cb = efLib_PauseAll;
         fp->post_hitlag_cb = efLib_ResumeAll;
         fp->mv.kb.specialn_pe.facing_dir = 0;
@@ -436,8 +405,8 @@ void ftKb_SpecialAirHi3_Phys(Fighter_GObj* gobj)
 
     /// FAKE MATCH: comma operator required for regalloc
     ftCommon_8007D3A8(fp, 0.0f,
-                      ((0, fp->co_attrs.air_drift_stick_mul)) *
-                          dat_attr->specialhi_horizontal_momentum,
+                      ((0, fp->co_attrs.air_drift_stick_mul)) *dat_attr
+                          ->specialhi_horizontal_momentum,
                       fp->co_attrs.air_drift_max);
 }
 
@@ -2086,7 +2055,8 @@ void ftKb_SpecialNLoop_IASA(Fighter_GObj* gobj)
     }
 
     if ((fp->input.held_inputs & 0x200) == 0) {
-        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialNEnd, 0, 0.0f, 1.0f, 0.0f, NULL);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialNEnd, 0, 0.0f, 1.0f,
+                                  0.0f, NULL);
     }
 }
 
@@ -2102,7 +2072,8 @@ void ftKb_SpecialAirNLoop_IASA(Fighter_GObj* gobj)
     }
 
     if ((fp->input.held_inputs & 0x200) == 0) {
-        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirNEnd, 0, 0.0f, 1.0f, 0.0f, NULL);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirNEnd, 0, 0.0f, 1.0f,
+                                  0.0f, NULL);
     }
 }
 
@@ -2531,8 +2502,8 @@ void ftKb_SpecialNLg_800F9598(Fighter_GObj* gobj)
         motion_id = 0x200;
     }
 
-    Fighter_ChangeMotionState(gobj, motion_id, 0x5000,
-                              anim_frame, 1.0f, 0.0f, NULL);
+    Fighter_ChangeMotionState(gobj, motion_id, 0x5000, anim_frame, 1.0f, 0.0f,
+                              NULL);
     fp->accessory4_cb = fn_800F9260;
 }
 
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
index 3cf973b1e4..a99ff8fd4c 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
@@ -3,76 +3,24 @@
 
 #include <placeholder.h>
 
-#include "cm/camera.h"
-#include "ef/efasync.h"
-#include "ef/eflib.h"
-#include "ef/efsync.h"
-#include "ft/chara/ftCommon/ftCo_CaptureWaitKirby.h"
-#include "ft/chara/ftCommon/ftCo_Damage.h"
-#include "ft/chara/ftCommon/ftCo_Escape.h"
 #include "ft/chara/ftCommon/ftCo_FallSpecial.h"
-#include "ft/chara/ftCommon/ftCo_Jump.h"
-#include "ft/chara/ftCommon/ftCo_KneeBend.h"
-#include "ft/chara/ftCommon/ftCo_Lift.h"
-#include "ft/chara/ftCommon/ftCo_Throw.h"
-#include "ft/chara/ftCommon/ftCo_Wait.h"
 #include "ft/chara/ftCommon/ftpickupitem.h"
 #include "ft/fighter.h"
 
 #include "ft/forward.h"
 
 #include "ft/ft_081B.h"
-#include "ft/ft_0877.h"
 #include "ft/ft_0881.h"
 #include "ft/ft_0892.h"
-#include "ft/ft_0C31.h"
-#include "ft/ftcamera.h"
-#include "ft/ftcliffcommon.h"
-#include "ft/ftcolanim.h"
-#include "ft/ftcoll.h"
-#include "ft/ftcommon.h"
-#include "ft/ftdata.h"
-#include "ft/ftdynamics.h"
-#include "ft/ftparts.h"
-#include "ft/ftwalkcommon.h"
 #include "ft/inlines.h"
 #include "ft/types.h"
-#include "ftCommon/ftCo_Attack100.h"
-#include "ftCommon/ftCo_CaptureKirby.h"
 #include "ftCommon/ftCo_Fall.h"
-#include "ftCommon/ftCo_Landing.h"
 
 #include "ftKirby/forward.h"
 
-#include "ftLink/types.h"
-#include "it/it_26B1.h"
-#include "it/item.h"
-#include "it/items/it_27CF.h"
-#include "it/items/it_2ADA.h"
-#include "it/items/it_2F28.h"
-#include "it/items/itclimbersice.h"
-#include "it/items/itgamewatchchef.h"
-#include "it/items/itkirby_2F23.h"
-#include "it/items/itkirbycutterbeam.h"
-#include "it/items/itkirbygamewatchchefpan.h"
-#include "it/items/itkirbyhammer.h"
-#include "it/items/itkirbyyoshispecialn.h"
-#include "it/items/itlinkarrow.h"
-#include "it/items/itlinkbow.h"
-#include "it/items/itmewtwoshadowball.h"
+#include "it/items/itfoxblaster.h"
 #include "it/items/itnesspkflush.h"
-#include "it/items/itpeachtoad.h"
-#include "it/items/itpeachtoadspore.h"
-#include "it/items/itpikachutjoltground.h"
-#include "it/items/itsamuschargeshot.h"
-#include "it/items/itseakneedleheld.h"
-#include "lb/lb_00B0.h"
-#include "lb/lbanim.h"
-#include "lb/lbvector.h"
 #include "melee/lb/lbrefract.h"
-#include "mp/mpcoll.h"
-#include "mp/mplib.h"
-#include "pl/player.h"
 
 #include <common_structs.h>
 #include <stddef.h>
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
index 423696a72b..186d3b3f6d 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
@@ -47,7 +47,6 @@
 #include "ftLink/types.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
-#include "it/items/it_27CF.h"
 #include "it/items/it_2ADA.h"
 #include "it/items/it_2F28.h"
 #include "it/items/itclimbersice.h"
@@ -854,7 +853,7 @@ long ftKb_SpecialNSs_800FCC14(Fighter_GObj* gobj, long* out1, long* out2)
     }
 
     *out1 = fp->fv.kb.xA8;
-    *out2 = (long)da->specialn_ss_charge_time;
+    *out2 = (long) da->specialn_ss_charge_time;
 
     return 0;
 
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
index cbb5a3a1c2..2a9ee5241f 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
@@ -3,76 +3,33 @@
 
 #include <placeholder.h>
 
-#include "cm/camera.h"
-#include "ef/efasync.h"
 #include "ef/eflib.h"
 #include "ef/efsync.h"
-#include "ft/chara/ftCommon/ftCo_CaptureWaitKirby.h"
-#include "ft/chara/ftCommon/ftCo_Damage.h"
-#include "ft/chara/ftCommon/ftCo_Escape.h"
-#include "ft/chara/ftCommon/ftCo_FallSpecial.h"
-#include "ft/chara/ftCommon/ftCo_Jump.h"
-#include "ft/chara/ftCommon/ftCo_KneeBend.h"
 #include "ft/chara/ftCommon/ftCo_Lift.h"
 #include "ft/chara/ftCommon/ftCo_Throw.h"
-#include "ft/chara/ftCommon/ftCo_Wait.h"
-#include "ft/chara/ftCommon/ftpickupitem.h"
 #include "ft/fighter.h"
 
 #include "ft/forward.h"
 
 #include "ft/ft_081B.h"
 #include "ft/ft_0877.h"
-#include "ft/ft_0881.h"
 #include "ft/ft_0892.h"
-#include "ft/ft_0C31.h"
-#include "ft/ftcamera.h"
-#include "ft/ftcliffcommon.h"
 #include "ft/ftcolanim.h"
 #include "ft/ftcoll.h"
 #include "ft/ftcommon.h"
-#include "ft/ftdata.h"
-#include "ft/ftdynamics.h"
 #include "ft/ftparts.h"
-#include "ft/ftwalkcommon.h"
 #include "ft/inlines.h"
 #include "ft/types.h"
 #include "ftCommon/ftCo_Attack100.h"
-#include "ftCommon/ftCo_CaptureKirby.h"
 #include "ftCommon/ftCo_Fall.h"
-#include "ftCommon/ftCo_Landing.h"
 
 #include "ftKirby/forward.h"
 
-#include "ftLink/types.h"
-#include "it/it_26B1.h"
-#include "it/item.h"
-#include "it/items/it_27CF.h"
-#include "it/items/it_2ADA.h"
-#include "it/items/it_2F28.h"
-#include "it/items/itclimbersice.h"
-#include "it/items/itgamewatchchef.h"
-#include "it/items/itkirby_2F23.h"
-#include "it/items/itkirbycutterbeam.h"
 #include "it/items/itkirbygamewatchchefpan.h"
-#include "it/items/itkirbyhammer.h"
 #include "it/items/itkirbyyoshispecialn.h"
-#include "it/items/itlinkarrow.h"
-#include "it/items/itlinkbow.h"
-#include "it/items/itmewtwoshadowball.h"
-#include "it/items/itnesspkflush.h"
 #include "it/items/itpeachtoad.h"
 #include "it/items/itpeachtoadspore.h"
-#include "it/items/itpikachutjoltground.h"
-#include "it/items/itsamuschargeshot.h"
-#include "it/items/itseakneedleheld.h"
 #include "lb/lb_00B0.h"
-#include "lb/lbanim.h"
-#include "lb/lbvector.h"
-#include "melee/lb/lbrefract.h"
-#include "mp/mpcoll.h"
-#include "mp/mplib.h"
-#include "pl/player.h"
 
 #include <common_structs.h>
 #include <stddef.h>
@@ -867,7 +824,7 @@ void ftKb_MsSpecialAirNLoop_Anim(Fighter_GObj* gobj)
     }
     fp->mv.kb.specialn_ms.cur_frame++;
     new_var = 1;
-    if (fp->mv.kb.specialn_ms.cur_frame > (s32)(*charge * 30)) {
+    if (fp->mv.kb.specialn_ms.cur_frame > (s32) (*charge * 30)) {
         fp->cmd_vars[0] = new_var;
         ftAnim_IsFramesRemaining(gobj);
     }
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index c5aec20d76..a2e96b8f7e 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -47,7 +47,6 @@
 #include "ftLink/types.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
-#include "it/items/it_27CF.h"
 #include "it/items/it_2ADA.h"
 #include "it/items/it_2F28.h"
 #include "it/items/itclimbersice.h"
@@ -262,9 +261,7 @@ void ftKb_SpecialNSk_80105F3C(Fighter_GObj* gobj)
     }
 }
 
-void fn_80105FEC(Fighter_GObj* gobj)
-{
-}
+void fn_80105FEC(Fighter_GObj* gobj) {}
 
 bool ftKb_SpecialNSk_80105FF0(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftLink/ftLk_SpecialLw.c b/src/melee/ft/chara/ftLink/ftLk_SpecialLw.c
index 899fcdaeeb..920e24e570 100644
--- a/src/melee/ft/chara/ftLink/ftLk_SpecialLw.c
+++ b/src/melee/ft/chara/ftLink/ftLk_SpecialLw.c
@@ -17,7 +17,6 @@
 #include "it/forward.h"
 
 #include "it/it_26B1.h"
-#include "it/items/it_27CF.h"
 #include "it/items/itlinkbomb.h"
 #include "lb/lb_00B0.h"
 
diff --git a/src/melee/ft/chara/ftLink/ftLk_SpecialN.c b/src/melee/ft/chara/ftLink/ftLk_SpecialN.c
index 3fcf6b0150..afb1a5d34c 100644
--- a/src/melee/ft/chara/ftLink/ftLk_SpecialN.c
+++ b/src/melee/ft/chara/ftLink/ftLk_SpecialN.c
@@ -23,7 +23,6 @@
 
 #include "it/forward.h"
 
-#include "it/items/it_27CF.h"
 #include "it/items/itlinkarrow.h"
 #include "it/items/itlinkbow.h"
 #include "lb/lb_00B0.h"
diff --git a/src/melee/ft/chara/ftMario/ftMr_SpecialS.c b/src/melee/ft/chara/ftMario/ftMr_SpecialS.c
index 9263a430bb..d01995284d 100644
--- a/src/melee/ft/chara/ftMario/ftMr_SpecialS.c
+++ b/src/melee/ft/chara/ftMario/ftMr_SpecialS.c
@@ -18,7 +18,6 @@
 #include "ft/ftparts.h"
 #include "ft/types.h"
 #include "ftCommon/ftCo_Fall.h"
-#include "it/items/it_27CF.h"
 #include "it/items/itmariocape.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
diff --git a/src/melee/ft/chara/ftMewtwo/ftMt_SpecialLw.c b/src/melee/ft/chara/ftMewtwo/ftMt_SpecialLw.c
index 0bc487b846..25920902cb 100644
--- a/src/melee/ft/chara/ftMewtwo/ftMt_SpecialLw.c
+++ b/src/melee/ft/chara/ftMewtwo/ftMt_SpecialLw.c
@@ -14,7 +14,6 @@
 #include "ft/types.h"
 #include "ftCommon/ftCo_Fall.h"
 #include "ftMewtwo/types.h"
-#include "it/items/it_27CF.h"
 #include "it/items/itmewtwodisable.h"
 #include "lb/lb_00B0.h"
 
diff --git a/src/melee/ft/chara/ftMewtwo/ftMt_SpecialN.c b/src/melee/ft/chara/ftMewtwo/ftMt_SpecialN.c
index 88a93bf73d..33bb592f45 100644
--- a/src/melee/ft/chara/ftMewtwo/ftMt_SpecialN.c
+++ b/src/melee/ft/chara/ftMewtwo/ftMt_SpecialN.c
@@ -27,7 +27,6 @@
 
 #include "it/forward.h"
 
-#include "it/items/it_27CF.h"
 #include "it/items/itmewtwoshadowball.h"
 #include "lb/lb_00B0.h"
 
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialN.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialN.c
index 7d1e6711ee..57fef8bc66 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialN.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialN.c
@@ -12,14 +12,12 @@
 #include "ft/ft_0892.h"
 #include "ft/ftanim.h"
 #include "ft/types.h"
-#include "ftCommon/ftCo_Attack100.h"
 #include "ftCommon/ftCo_Fall.h"
 #include "ftCommon/ftCo_Landing.h"
 #include "ftPopo/types.h"
 
 #include "it/forward.h"
 
-#include "it/items/it_27CF.h"
 #include "it/items/itclimbersice.h"
 #include "lb/lb_00B0.h"
 
diff --git a/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c b/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c
index aed277dcae..e8e6783b08 100644
--- a/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c
+++ b/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c
@@ -17,8 +17,8 @@
 
 #include "ftYoshi/ftYs_Init.h"
 #include "ftYoshi/types.h"
-#include "it/items/it_2F2B.h"
 #include "it/items/ityoshieggthrow.h"
+#include "it/items/ityoshitongue.h"
 
 #include <math.h>
 #include <dolphin/mtx.h>
diff --git a/src/melee/ft/chara/ftYoshi/ftYs_SpecialN.c b/src/melee/ft/chara/ftYoshi/ftYs_SpecialN.c
index b028bc767a..c5789d7463 100644
--- a/src/melee/ft/chara/ftYoshi/ftYs_SpecialN.c
+++ b/src/melee/ft/chara/ftYoshi/ftYs_SpecialN.c
@@ -16,16 +16,14 @@
 #include "ft/ftparts.h"
 #include "ft/inlines.h"
 #include "ft/types.h"
-#include "ftCommon/ftCo_Attack100.h"
-#include "ftCommon/ftCo_CaptureCut.h"
 #include "ftCommon/ftCo_CaptureYoshi.h"
 #include "ftCommon/ftCo_Fall.h"
 #include "ftCommon/ftCo_Throw.h"
 #include "ftCommon/ftCo_YoshiEgg.h"
 #include "ftYoshi/ftYs_Init.h"
 #include "ftYoshi/types.h"
-#include "it/items/it_27CF.h"
-#include "it/items/it_2F2B.h"
+#include "it/items/ityoshiegglay.h"
+#include "it/items/ityoshitongue.h"
 #include "lb/lb_00B0.h"
 
 #include <dolphin/mtx.h>
diff --git a/src/melee/it/it_26B1.c b/src/melee/it/it_26B1.c
index 6015ce1602..507c6870fe 100644
--- a/src/melee/it/it_26B1.c
+++ b/src/melee/it/it_26B1.c
@@ -10,18 +10,20 @@
 #include "it/it_2725.h"
 #include "it/itCommonItems.h"
 #include "it/item.h"
-#include "it/items/it_27CF.h"
 #include "it/items/itbat.h"
 #include "it/items/itbombhei.h"
 #include "it/items/itbox.h"
 #include "it/items/itfflower.h"
 #include "it/items/itflipper.h"
+#include "it/items/itheart.h"
 #include "it/items/itkusudama.h"
+#include "it/items/itlinkbomb.h"
 #include "it/items/itmarumine.h"
 #include "it/items/itmsbomb.h"
 #include "it/items/itrabbitc.h"
 #include "it/items/itsscope.h"
 #include "it/items/itsword.h"
+#include "it/items/ittomato.h"
 #include "it/types.h"
 
 #include <baselib/forward.h>
@@ -902,7 +904,7 @@ HSD_GObj* it_8026BE84(BobOmbRain* bobOmbRain)
     Item* item_data_4;
     Item* item_data_5;
 
-    HSD_GObj* bobOmbGObj = NULL;
+    HSD_GObj* gobj = NULL;
 
     if (gm_8016B238() && bobOmbRain->x14 != 6) {
         return NULL;
@@ -913,48 +915,48 @@ HSD_GObj* it_8026BE84(BobOmbRain* bobOmbRain)
     switch (bobomb_id) {
     case 1:
         if (it_8026D324(true)) {
-            bobOmbGObj = it_80286088(bobOmbRain->x0);
+            gobj = it_80286088(bobOmbRain->x0);
         }
 
         break;
 
     case 20:
-        bobOmbGObj = it_80290938(bobOmbRain->x4);
+        gobj = it_80290938(bobOmbRain->x4);
         break;
 
     case 8:
-        bobOmbGObj =
+        gobj =
             it_80283AE4(bobOmbRain->x0, &bobOmbRain->x8_vec, bobOmbRain->x18);
-        if (bobOmbGObj != NULL) {
+        if (gobj != NULL) {
             u8 _[8];
 
-            ip = GET_ITEM(bobOmbGObj);
+            ip = GET_ITEM(gobj);
             kind0 = ip->kind;
             switch (kind0) {
             case It_Kind_Heart:
-                it_80283BD4(bobOmbGObj);
+                it_80283BD4(gobj);
                 break;
             case It_Kind_Tomato:
-                it_8028428C(bobOmbGObj);
+                it_8028428C(gobj);
                 break;
             }
         }
         break;
 
     case 9:
-        bobOmbGObj =
+        gobj =
             it_802841B4(bobOmbRain->x0, &bobOmbRain->x8_vec, bobOmbRain->x18);
-        if (bobOmbGObj != NULL) {
-            ip = GetItemData(bobOmbGObj);
+        if (gobj != NULL) {
+            ip = GetItemData(gobj);
             kind1 = ip->kind;
 
             switch (kind1) {
             case It_Kind_Heart:
-                it_80283BD4(bobOmbGObj);
+                it_80283BD4(gobj);
                 break;
 
             case It_Kind_Tomato:
-                it_8028428C(bobOmbGObj);
+                it_8028428C(gobj);
                 break;
             }
         }
@@ -962,46 +964,46 @@ HSD_GObj* it_8026BE84(BobOmbRain* bobOmbRain)
 
     case 4:
         if (it_8026D324(4)) {
-            bobOmbGObj = it_802896CC(&bobOmbRain->x8_vec);
-            if (bobOmbGObj != NULL) {
-                it_80274F10(bobOmbGObj);
+            gobj = it_802896CC(&bobOmbRain->x8_vec);
+            if (gobj != NULL) {
+                it_80274F10(gobj);
             }
         }
         break;
 
     case 12:
-        bobOmbGObj = itSword_Spawn(&bobOmbRain->x8_vec);
-        if (bobOmbGObj != NULL) {
-            it_80274F10(bobOmbGObj);
+        gobj = itSword_Spawn(&bobOmbRain->x8_vec);
+        if (gobj != NULL) {
+            it_80274F10(gobj);
         }
         break;
 
     case 21:
-        bobOmbGObj = it_80291BE0(&bobOmbRain->x8_vec);
-        if (bobOmbGObj != NULL) {
-            it_80274F10(bobOmbGObj);
+        gobj = it_80291BE0(&bobOmbRain->x8_vec);
+        if (gobj != NULL) {
+            it_80274F10(gobj);
         }
 
         break;
 
     case 25:
-        bobOmbGObj = it_80292D48(&bobOmbRain->x8_vec);
-        if (bobOmbGObj != NULL) {
-            it_80274F10(bobOmbGObj);
+        gobj = it_80292D48(&bobOmbRain->x8_vec);
+        if (gobj != NULL) {
+            it_80274F10(gobj);
         }
         break;
 
     case 31:
-        bobOmbGObj = it_80294DC0(&bobOmbRain->x8_vec);
-        if (bobOmbGObj != NULL) {
-            it_80274F10(bobOmbGObj);
+        gobj = it_80294DC0(&bobOmbRain->x8_vec);
+        if (gobj != NULL) {
+            it_80274F10(gobj);
         }
         break;
 
     case 11:
-        bobOmbGObj = it_80284854(&bobOmbRain->x8_vec);
-        if (bobOmbGObj != NULL) {
-            it_80274F10(bobOmbGObj);
+        gobj = it_80284854(&bobOmbRain->x8_vec);
+        if (gobj != NULL) {
+            it_80274F10(gobj);
         }
         break;
 
@@ -1009,28 +1011,28 @@ HSD_GObj* it_8026BE84(BobOmbRain* bobOmbRain)
         it_8027D670(&bobOmbRain->x8_vec);
 
     default:
-        bobOmbGObj = NULL;
+        gobj = NULL;
         break;
     }
 
-    if (bobOmbGObj != NULL && bobOmbRain->x1C.b0) {
-        ip = bobOmbGObj->user_data;
+    if (gobj != NULL && bobOmbRain->x1C.b0) {
+        ip = gobj->user_data;
         ip->xDD0_flag.b3 = true;
 
-        item_data_2 = bobOmbGObj->user_data;
+        item_data_2 = gobj->user_data;
         item_data_2->xDCC_flag.b3 = false;
 
-        item_data_3 = bobOmbGObj->user_data;
+        item_data_3 = gobj->user_data;
         item_data_3->xDCD_flag.b3 = true;
 
-        item_data_4 = bobOmbGObj->user_data;
+        item_data_4 = gobj->user_data;
         item_data_4->xDCD_flag.b4 = true;
 
-        item_data_5 = bobOmbGObj->user_data;
+        item_data_5 = gobj->user_data;
         item_data_5->xDC8_word.flags.x1A = false;
     }
 
-    return bobOmbGObj;
+    return gobj;
 }
 
 extern CollData* ftLib_80086984(HSD_GObj*);
diff --git a/src/melee/it/it_2725.c b/src/melee/it/it_2725.c
index a4caaa741b..0519f00139 100644
--- a/src/melee/it/it_2725.c
+++ b/src/melee/it/it_2725.c
@@ -23,7 +23,6 @@
 
 #include "it/it_26B1.h"
 #include "it/itcoll.h"
-#include "it/items/it_27CF.h"
 #include "it/items/it_2ADA.h"
 #include "it/items/it_2E5A.h"
 #include "it/items/it_2E6A.h"
@@ -179,6 +178,7 @@
 #include "it/items/itwhispyapple.h"
 #include "it/items/itwhitebea.h"
 #include "it/items/itwstar.h"
+#include "it/items/ityoshiegglay.h"
 #include "it/items/ityoshieggthrow.h"
 #include "it/items/ityoshistar.h"
 #include "it/items/itzeldadinfire.h"
@@ -186,7 +186,6 @@
 #include "it/items/itzgshell.h"
 #include "it/items/itzrshell.h"
 #include "it/types.h"
-#include "items/it_27CF.h"
 #include "items/itbat.h"
 #include "items/itbombhei.h"
 #include "items/itbox.h"
@@ -234,6 +233,7 @@
 #include "items/ittarucann.h"
 #include "items/ittomato.h"
 #include "items/itwstar.h"
+#include "items/ityoshiegglay.h"
 #include "lb/inlines.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
@@ -5128,8 +5128,8 @@ HSD_JObj* it_80272CC0(Item_GObj* item_gobj, enum_t idx)
 /// Check if the HSD_GObj* class is an item
 bool it_80272D1C(Item_GObj* item_gobj)
 {
-    if ((item_gobj != NULL) &&
-        (item_gobj->classifier == HSD_GOBJ_CLASS_ITEM)) { // ITEM_UNK_ENEMY?
+    if ((item_gobj != NULL) && (item_gobj->classifier == HSD_GOBJ_CLASS_ITEM))
+    { // ITEM_UNK_ENEMY?
         return true;
     }
     return false;
@@ -5142,8 +5142,8 @@ s32 it_80272D40(Item_GObj* item_gobj)
     if (ftLib_80086960((HSD_GObj*) item_gobj)) {
         return 0;
     }
-    if ((item_gobj != NULL) &&
-        (item_gobj->classifier == HSD_GOBJ_CLASS_ITEM)) { // ITEM_UNK_ENEMY?
+    if ((item_gobj != NULL) && (item_gobj->classifier == HSD_GOBJ_CLASS_ITEM))
+    { // ITEM_UNK_ENEMY?
         chk = true;
     } else {
         chk = false;
@@ -9919,8 +9919,8 @@ bool it_8027B798(Item_GObj* item_gobj, Vec3* arg1)
         sp14.x = -temp_f30 * item->facing_dir;
         sp14.y = temp_f2;
         sp14.z = 0.0f;
-        if (lbVector_Angle(temp_r31_2, &sp14) <
-            1.5708f) { // Should this be M_PI_2?
+        if (lbVector_Angle(temp_r31_2, &sp14) < 1.5708f)
+        { // Should this be M_PI_2?
             ret_chk = true;
             arg1->x = sp14.x;
             arg1->y = sp14.y;
diff --git a/src/melee/it/items/it_27CF.dox b/src/melee/it/items/it_27CF.dox
deleted file mode 100644
index 5ce4c8abac..0000000000
--- a/src/melee/it/items/it_27CF.dox
+++ /dev/null
@@ -1,190 +0,0 @@
-#ifndef GALE01_27CF30
-#define GALE01_27CF30
-
-#include <platform.h>
-#include "ft/forward.h"
-#include <dolphin/mtx.h>
-#include <baselib/forward.h>
-
-#include "ftSeak/types.h"
-#include "it/types.h"
-
-#include <placeholder.h>
-#include <baselib/gobj.h>
-#include <baselib/jobj.h>
-
-void it_8027D730(Item_GObj*);
-HSD_GObj* it_80283AE4(s32, Vec3*, s32);
-HSD_GObj* it_802841B4(s32, Vec3*, s32);
-HSD_GObj* it_802896CC(Vec3*);
-void it_8028FE90(Item_GObj*);
-
-void it_8028428C(Item_GObj*);
-void it_80283BD4(Item_GObj*);
-HSD_GObj* it_80286088(s32);
-HSD_GObj* it_80290938(HSD_JObj*);
-HSD_GObj* it_80291BE0(Vec3*);
-HSD_GObj* it_80292D48(Vec3*);
-void it_8029D968(Item_GObj*);
-void it_802D09D0(Item_GObj*);
-
-/// @attention Above this line, symbols are placed correctly.
-
-void it_80294E78(Item_GObj*, f32);
-HSD_GObj* it_8029A748(Item_GObj*, Vec3*, u8, f32);
-void it_8029A89C(Item_GObj*, f32);
-void it_8029A8F4(Item_GObj*, Vec3*, f32);
-s32 it_802B1DEC(u32);
-u32 it_802B1DF8(Item_GObj*, Vec3*, Vec3*, s32, s32, s32);
-s32 it_802B1FC8(u32);
-void it_802B1FE8(u32, Vec3*);
-void it_802B56E4(Item_GObj*, Vec3*, f32, f32, f32);
-HSD_GObj* it_802C8B28(Item_GObj*, Vec3*, u32, f32);
-void it_802D8618(s32, Vec3*, s32, s32);
-u32 it_802F0340(Item_GObj*, Vec3*, Vec3*, u32, u32, f32);
-void it_802F046C(s32);
-void it_802F0AE0(Item_GObj* fighter_gobj, Vec3*, Vec3*, enum_t, bool,
-                 f32 facing_dir, f32, f32);
-
-void it_802C8C34(Item_GObj*);
-
-void it_802AE1D0(Item_GObj*);
-void it_802AEAB4(Item_GObj*);
-HSD_GObj* it_802AE8A8(f32, HSD_GObj*, Vec3*, s32, s32);
-void it_802ADDD0(Item_GObj*, s32);
-void it_802AE538(Item_GObj*);
-HSD_GObj* it_802C6C38(Item_GObj*, Vec3*, enum_t part, f32);
-void it_802C7158(Item_GObj*);
-void it_802C7074(Item_GObj*);
-void it_802C7424(Item_GObj*);
-void it_802C6E50(Item_GObj*);
-void it_802C6D6C(Item_GObj*);
-HSD_GObj* it_802C6F40(Item_GObj*, Vec3*, s32, f32);
-HSD_GObj* it_802C65E4(Item_GObj*, Vec3*, s32, f32);
-HSD_GObj* it_802C8038(Item_GObj*, Vec3*, s32, s32, f32, f32);
-HSD_GObj* it_802C7D60(Item_GObj*, Vec3*, s32, f32);
-void it_802C7E94(Item_GObj*);
-void it_802C7EE0(Item_GObj*);
-void it_802C7F00(Item_GObj*);
-HSD_GObj* it_802C837C(Item_GObj*, Vec3*, enum_t, u32, f32);
-HSD_GObj* it_802C7774(f32, HSD_GObj*, Vec3*, s32, u32);
-void it_8028FAF4(Item_GObj*, Vec3*);
-void it_802C7A84(Item_GObj*);
-void it_802C7AD0(Item_GObj*);
-void it_802C7AF0(Item_GObj*);
-HSD_GObj* it_802B2560(Item_GObj*, f32, Vec3*, long, u32);
-void it_802B2674(Item_GObj*);
-void it_802B26C0(Item_GObj*);
-void it_802B26E0(Item_GObj*);
-void it_802C49E0(Item_GObj*);
-void it_802BFE5C(Item_GObj*, Vec3*, f32);
-void it_802C0010(Item_GObj*, void*);
-void it_802BE5D8(void*, f32);
-void it_802BE958(void*);
-void it_802BFEC4(void*);
-HSD_GObj* it_802BE9D8(f32, HSD_GObj*, Vec3*, s32);
-void it_802BE598(Item_GObj*);
-void it_802BE5B8(Item_GObj*);
-HSD_GObj* it_802AD478(Item_GObj*, Vec3*, s32, f32);
-void it_802AD6B8(Item_GObj*);
-void it_802AB3F0(Item_GObj*, void*, s32);
-
-/// Spawn an item
-HSD_GObj* it_802AB58C(Item_GObj* owner, Vec3* pos, f32 facing_dir);
-
-void it_802AB9C0(Item_GObj*);
-HSD_GObj* it_802AB568(Item_GObj*);
-void it_802AAA50(Item_GObj*);
-HSD_GObj* it_802AA8C0(Item_GObj*, Vec3*, s32, f32);
-HSD_GObj* it_802AA7E4(Item_GObj*);
-s32 it_802AA7F0(Item_GObj*);
-void it_802AA054(Item_GObj*, Vec3*, Vec3*, f32, f32);
-void it_802B338C(Item_GObj*, Vec3*, f32 facing_dir, u32);
-void it_802B5974(Item_GObj*);
-HSD_GObj* it_802B55C8(Item_GObj* owner, Vec3*, s32, s32, f32 facing_dir);
-void it_802B62D0(Item_GObj*, Vec3* pos, bool, f32 facing_dir);
-
-/// Destroys an item?
-void it_802C3D44(Item_GObj*);
-HSD_GObj* it_802C3BAC(Item_GObj*, Vec3*, f32 facing_dir, f32);
-
-/// Clones an entity?
-HSD_GObj* it_802C3AF0(Item_GObj*);
-void it_802AE608(Item_GObj*);
-void it_802C7340(Item_GObj*);
-void it_802C6DB8(Item_GObj*);
-void it_802C70C0(Item_GObj*);
-void it_802C738C(Item_GObj*);
-void it_802C6DD8(Item_GObj*);
-void it_802C70E0(Item_GObj*);
-void it_802C73AC(Item_GObj*);
-void it_802C6718(Item_GObj*);
-void it_802C6764(Item_GObj*);
-void it_802C6784(Item_GObj*);
-void it_802C8158(Item_GObj*);
-void it_802C81C8(Item_GObj*);
-void it_802C81E8(Item_GObj*);
-HSD_GObj* it_802C4A40(Item_GObj*, Vec3*, f32);
-void it_802C573C(Item_GObj*);
-HSD_GObj* it_802C5000(Item_GObj*, Vec3*, s32, s32, f32);
-void it_802C519C(Item_GObj*, Vec3*, s32, s32, f32, f32);
-void it_802B4AC8(Item_GObj* gobj, Vec3*, f32 facing_dir);
-void it_802C53F0(Item_GObj*, Vec3*, f32, f32, f32);
-s32 it_802E5F8C(Item_GObj*, Vec3*, s32, bool, f32, f32);
-void it_802950D4(Item_GObj*, bool);
-void it_80294EB0(Item_GObj*, Vec3*, Vec3*);
-HSD_GObj* it_80294DC0(Vec3*);
-void it_8027D670(Vec3* const);
-void it_802C17DC(uint);
-void it_802C16F8(Item_GObj*);
-void it_802C1590(Item_GObj*);
-void it_802BB20C(Item_GObj*);
-void it_802BAEEC(Item_GObj*);
-void it_802BAF0C(Item_GObj*);
-void it_802BCFC4(Item_GObj* gobj, f32* arg1, itChainSegment* chainSegment,
-                 Item* item_data, f32 arg4);
-HSD_GObj* itSeakChain_Spawn(Item_GObj* gobj, Vec3* arg1, f32 facing_dir);
-void it_802BCF84(Item_GObj*);
-void it_802BD158(Item_GObj* gobj, Vec3*, int, float);
-
-/// @returns An #Item #HSD_GObj.
-HSD_GObj* it_802BD4AC(Item_GObj*, Vec3*, u8, s32, f32);
-
-void it_802BD45C(Item_GObj* gobj);
-void it_802BDB94(Item_GObj* gobj);
-void it_802BDBF8(Item_GObj* gobj);
-void it_802BDC18(Item_GObj* gobj);
-bool it_802BDC38(Item_GObj* gobj);
-HSD_GObj* it_802BDE18(Item_GObj*, Vec3*, Fighter_Part, enum_t, float);
-void it_802BE214(Item_GObj* gobj, Vec3* pos, enum_t, float facing_dir);
-void it_802BDF40(Item_GObj* gobj);
-void it_802BDFA0(Item_GObj* gobj);
-void it_802BDFC0(Item_GObj* gobj);
-void it_802BE100(Item_GObj* gobj);
-/* 2A07B4 */ void it_802A07B4(Item_GObj* gobj);
-/* 2A0534 */ void it_802A0534(Item_GObj* gobj, Vec3*);
-/* 29FDBC */ bool it_8029FDBC(Item_GObj* gobj);
-
-/// @returns An #Item #HSD_GObj.
-/* 2A013C */ HSD_GObj* it_802A013C(float facing_dir, Fighter_GObj* owner_gobj,
-                                   Vec3* pos, Fighter_Part, int);
-
-/* 2A8A7C */ void it_802A8A7C(Item_GObj* gobj);
-/* 2AF304 */ void it_802AF304(Item_GObj* gobj);
-
-/// @returns An #Item #HSD_GObj.
-/* 29DD58 */ HSD_GObj* it_8029DD58(Item_GObj* fighter_gobj, Vec3*, u8, int,
-                                   float);
-
-/* 2AF1A4 */ HSD_GObj* it_802AF1A4(float facing_dir, Fighter_GObj* owner_gobj,
-                                   Vec3*, Fighter_Part, int);
-/* 2A8398 */ void it_802A8398(Item_GObj*, Vec3*);
-/* 2A83E0 */ HSD_GObj* it_802A83E0(float facing_dir, Fighter_GObj*, Vec3*,
-                                   Fighter_Part, int);
-/* 2A850C */ void it_802A850C(Item_GObj*, Vec3*, Vec3*, float, float, float);
-
-/* 2F3330 */ bool it_802F3330(Item_GObj*);
-/* 2F3374 */ void it_802F3374(Item_GObj*, Item_GObj*);
-/* 3F94A0 */ extern ItemStateTable it_803F94A0[];
-
-#endif
diff --git a/src/melee/it/items/it_27CF.h b/src/melee/it/items/it_27CF.h
deleted file mode 100644
index 672a4ae462..0000000000
--- a/src/melee/it/items/it_27CF.h
+++ /dev/null
@@ -1,67 +0,0 @@
-#ifndef GALE01_27CF30
-#define GALE01_27CF30
-
-#include <platform.h>
-
-#include "ft/forward.h"
-#include "it/forward.h"
-
-#include "it/itCommonItems.h"
-#include "it/items/types.h"
-
-#include <baselib/forward.h>
-
-#include <dolphin/mtx.h>
-
-/* 283AE4 */ Item_GObj* it_80283AE4(Item_GObj*, Vec3*, s32);
-/* 283BD4 */ void it_80283BD4(Item_GObj*);
-/* 2841B4 */ HSD_GObj* it_802841B4(HSD_GObj*, Vec3*, s32);
-/* 28428C */ void it_8028428C(Item_GObj*);
-/* 29D968 */ void it_8029D968(Item_GObj*);
-/* 2A850C */ void it_802A850C(Item_GObj*, Vec3*, Vec3*, float, float, float);
-/* 2A8A7C */ void it_802A8A7C(Item_GObj* gobj);
-/* 2ADDD0 */ void it_802ADDD0(Item_GObj*, s32);
-/* 2AE1D0 */ void it_802AE1D0(Item_GObj*);
-/* 2AE538 */ void it_802AE538(Item_GObj*);
-/* 2AE608 */ void it_802AE608(Item_GObj*);
-/* 2AE8A8 */ Item_GObj* it_802AE8A8(float, Fighter_GObj*, Vec3*, Fighter_Part,
-                                    ItemKind);
-/* 2AEAB4 */ void it_802AEAB4(Item_GObj*);
-/* 2B2674 */ void it_802B2674(Item_GObj*);
-/* 2B26C0 */ void it_802B26C0(Item_GObj*);
-/* 2B26E0 */ void it_802B26E0(Item_GObj*);
-/* 2BDC18 */ void it_802BDC18(Item_GObj* gobj);
-/* 2BDC38 */ bool it_802BDC38(Item_GObj* gobj);
-/* 2C65E4 */ HSD_GObj* it_802C65E4(Item_GObj*, Vec3*, s32, float);
-/* 2C6718 */ void it_802C6718(Item_GObj*);
-/* 2C6764 */ void it_802C6764(Item_GObj*);
-/* 2C6784 */ void it_802C6784(Item_GObj*);
-/* 2C6C38 */ HSD_GObj* it_802C6C38(Item_GObj*, Vec3*, enum_t part, float);
-/* 2C6D6C */ void it_802C6D6C(Item_GObj*);
-/* 2C6DB8 */ void it_802C6DB8(Item_GObj*);
-/* 2C6DD8 */ void it_802C6DD8(Item_GObj*);
-/* 2C6E50 */ void it_802C6E50(Item_GObj*);
-/* 2C7340 */ void it_802C7340(Item_GObj*);
-/* 2C738C */ void it_802C738C(Item_GObj*);
-/* 2C73AC */ void it_802C73AC(Item_GObj*);
-/* 2C7424 */ void it_802C7424(Item_GObj*);
-/* 2C7774 */ HSD_GObj* it_802C7774(float, HSD_GObj*, Vec3*, s32, u32);
-/* 2C7A84 */ void it_802C7A84(Item_GObj*);
-/* 2C7AD0 */ void it_802C7AD0(Item_GObj*);
-/* 2C7AF0 */ void it_802C7AF0(Item_GObj*);
-/* 2F2F34 */ void it_802F2F34(HSD_GObj*, it_27CF_DatAttrs*);
-/* 2F3020 */ void it_802F3020(HSD_GObj*);
-/* 2F3078 */ bool it_27CF_UnkMotion1_Anim(HSD_GObj*);
-/* 2F3120 */ void it_27CF_UnkMotion1_Phys(HSD_GObj*);
-/* 2F3150 */ bool it_27CF_UnkMotion1_Coll(HSD_GObj*);
-/* 2F317C */ void it_802F317C(HSD_GObj*);
-/* 2F31B8 */ bool it_27CF_UnkMotion0_Anim(HSD_GObj*);
-/* 2F3260 */ void it_27CF_UnkMotion0_Phys(HSD_GObj*);
-/* 2F3264 */ bool it_27CF_UnkMotion0_Coll(HSD_GObj*);
-/* 2F3290 */ void it_802F3290(Item_GObj*);
-/* 2F32F8 */ bool it_27CF_UnkMotion2_Anim(HSD_GObj*);
-/* 2F3330 */ bool it_27CF_Logic114_DmgReceived(Item_GObj*);
-/* 2F3374 */ void it_27CF_Logic114_EvtUnk(Item_GObj*, HSD_GObj*);
-/* 3F94A0 */ extern ItemStateTable it_803F94A0[];
-
-#endif
diff --git a/src/melee/it/items/itfoxblaster.c b/src/melee/it/items/itfoxblaster.c
index 8d93d1b87c..705aa9b103 100644
--- a/src/melee/it/items/itfoxblaster.c
+++ b/src/melee/it/items/itfoxblaster.c
@@ -17,20 +17,9 @@
 #include <melee/it/item.h>
 #include <melee/lb/lbrefract.h>
 
-/* 2ADDD0 */ static void it_802ADDD0(Item_GObj* item_gobj, s32 visibility);
-/* 2ADEF0 */ static void it_802ADEF0(HSD_GObj* item_gobj);
-/* 2ADF10 */ static void it_802ADF10(HSD_GObj* item_gobj);
-/* 2AE1D0 */ static void it_802AE1D0(Item_GObj* item_gobj);
-/* 2AE200 */ void it_802AE200(Item_GObj* item_gobj);
-/* 2AE538 */ static void it_802AE538(Item_GObj* item_gobj);
-/* 2AE608 */ static void it_802AE608(Item_GObj* item_gobj);
+/* 2AE200 */ static void it_802AE200(Item_GObj* item_gobj);
 /* 2AE63C */ static void it_802AE63C(Item_GObj* item_gobj);
 /* 2AE7B8 */ static void it_802AE7B8(Item_GObj* item_gobj);
-/* 2AE8A8 */ static Item_GObj* it_802AE8A8(f32 facing_dir,
-                                           Fighter_GObj* fighter_gobj,
-                                           Vec3* arg2, Fighter_Part ft_part,
-                                           ItemKind it_kind);
-/* 2AEAB4 */ static void it_802AEAB4(Item_GObj* item_gobj);
 /* 2AEBCC */ static bool itFoxblaster_UnkMotion8_Anim(HSD_GObj* item_gobj);
 /* 2AEED4 */ static void itFoxblaster_UnkMotion8_Phys(HSD_GObj* item_gobj);
 /* 2AEF08 */ static bool itFoxblaster_UnkMotion8_Coll(HSD_GObj* item_gobj);
diff --git a/src/melee/it/items/itfoxblaster.h b/src/melee/it/items/itfoxblaster.h
index 43463ce00b..68428d82e9 100644
--- a/src/melee/it/items/itfoxblaster.h
+++ b/src/melee/it/items/itfoxblaster.h
@@ -6,7 +6,17 @@
 
 #include "it/items/types.h"
 
+/* 2ADDD0 */ void it_802ADDD0(Item_GObj* item_gobj, s32 visibility);
+/* 2ADEF0 */ void it_802ADEF0(HSD_GObj* item_gobj);
+/* 2ADF10 */ void it_802ADF10(HSD_GObj* item_gobj);
+/* 2AE1D0 */ void it_802AE1D0(Item_GObj* item_gobj);
+/* 2AE538 */ void it_802AE538(Item_GObj* item_gobj);
+/* 2AE608 */ void it_802AE608(Item_GObj* item_gobj);
+/* 2AE8A8 */ Item_GObj* it_802AE8A8(f32 facing_dir, Fighter_GObj* fighter_gobj,
+                                    Vec3* arg2, Fighter_Part ft_part,
+                                    ItemKind it_kind);
 /* 2AE994 */ Item_GObj* it_802AE994(Fighter_GObj*, Fighter_Part, ItemKind);
+/* 2AEAB4 */ void it_802AEAB4(Item_GObj* item_gobj);
 /* 2AEB00 */ void itFoxBlaster_Logic96_PickedUp(Item_GObj*);
 /* 2AF184 */ void itFoxBlaster_Logic96_EvtUnk(Item_GObj*, HSD_GObj*);
 /* 3F6CA8 */ extern ItemStateTable it_803F6CA8[];
diff --git a/src/melee/it/items/itgamewatchmanhole.c b/src/melee/it/items/itgamewatchmanhole.c
index 4fc69cdba9..c47c90df19 100644
--- a/src/melee/it/items/itgamewatchmanhole.c
+++ b/src/melee/it/items/itgamewatchmanhole.c
@@ -12,11 +12,6 @@
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
 
-HSD_GObj* it_802C65E4(Item_GObj* gobj, Vec* vec, enum Fighter_Part arg2,
-                      float arg3);
-void it_802C6764(Item_GObj* gobj);
-void it_802C6784(Item_GObj* gobj);
-void it_802C6718(Item_GObj* gobj);
 bool itGamewatchmanhole_UnkMotion0_Anim(Item_GObj* gobj);
 
 ItemStateTable it_803F78D8[] = { { 0, itGamewatchmanhole_UnkMotion0_Anim, 0,
diff --git a/src/melee/it/items/itgamewatchmanhole.h b/src/melee/it/items/itgamewatchmanhole.h
index a71d766361..acf45436e1 100644
--- a/src/melee/it/items/itgamewatchmanhole.h
+++ b/src/melee/it/items/itgamewatchmanhole.h
@@ -6,7 +6,12 @@
 #include "it/items/types.h"
 #include "it/types.h"
 
+/* 2C65E4 */ HSD_GObj* it_802C65E4(Item_GObj* gobj, Vec* vec,
+                                   enum Fighter_Part arg2, float arg3);
 /* 2C66E8 */ void itGameWatchManhole_Logic72_Destroyed(Item_GObj*);
+/* 2C6718 */ void it_802C6718(Item_GObj* gobj);
+/* 2C6764 */ void it_802C6764(Item_GObj* gobj);
+/* 2C6784 */ void it_802C6784(Item_GObj* gobj);
 /* 2C67A4 */ void itGameWatchManhole_Logic72_PickedUp(Item_GObj*);
 /* 2C68D8 */ void itGameWatchManhole_Logic72_EvtUnk(Item_GObj*, Item_GObj*);
 /* 3F78D8 */ extern ItemStateTable it_803F78D8[];
diff --git a/src/melee/it/items/itgamewatchparachute.h b/src/melee/it/items/itgamewatchparachute.h
index 696aabf6d9..257260a780 100644
--- a/src/melee/it/items/itgamewatchparachute.h
+++ b/src/melee/it/items/itgamewatchparachute.h
@@ -5,6 +5,7 @@
 
 #include "it/items/types.h"
 
+/* 2C6C38 */ HSD_GObj* it_802C6C38(Item_GObj*, Vec3*, enum_t part, float);
 /* 2C6D3C */ void itGameWatchParachute_Logic74_Destroyed(Item_GObj*);
 /* 2C6D6C */ void it_802C6D6C(Item_GObj*);
 /* 2C6DB8 */ void it_802C6DB8(Item_GObj*);
diff --git a/src/melee/it/items/itoctarockstone.c b/src/melee/it/items/itoctarockstone.c
index e39047ddcb..106c74125a 100644
--- a/src/melee/it/items/itoctarockstone.c
+++ b/src/melee/it/items/itoctarockstone.c
@@ -1,5 +1,7 @@
 #include "itoctarockstone.h"
 
+#include "ft/ftlib.h"
+
 #include "it/forward.h"
 
 #include "it/inlines.h"
@@ -8,7 +10,7 @@
 #include "it/it_2725.h"
 #include "it/item.h"
 
-#include "it/items/forward.h"
+#include <baselib/random.h>
 
 ItemStateTable it_803F8E90[] = {
     { 0, itOctarockstone_UnkMotion0_Anim, itOctarockstone_UnkMotion0_Phys,
@@ -117,7 +119,48 @@ void itOctarockstone_802E89B0(Item_GObj* gobj, Item_GObj* ref_gobj)
     it_8026B894(gobj, ref_gobj);
 }
 
-/// #it_802E89D0
+static inline f32 getX(f32 x1, f32 x2)
+{
+    f32 res = x1 - x2;
+    if (res < 0.0f) {
+        return -res;
+    }
+    return res;
+}
+
+void it_802E89D0(Item_GObj* gobj, f32 horiz_speed, f32 min_vy, f32 max_vy)
+{
+    HSD_GObj* gp;
+    Item* ip = GET_ITEM(gobj);
+    Vec3 pos;
+    f32 x;
+    f32 speed;
+    float new_var;
+
+    ip->x40_vel.x = horiz_speed * ip->facing_dir;
+    ip->x40_vel.z = 0.0f;
+
+    gp = ftLib_8008627C(&ip->pos, NULL);
+    if (gp != NULL) {
+        ftLib_800866DC(gp, &pos);
+        x = getX(ip->pos.x, pos.x);
+        speed = x / horiz_speed;
+        x = (new_var = ip->xCC_item_attr->x10_fall_speed * (speed * speed));
+        x *= 0.5f;
+        ip->x40_vel.y = x / speed;
+        ip->x40_vel.y =
+            HSD_Randf() * ((pos.y - ip->pos.y) / 10.0f) + ip->x40_vel.y;
+
+        if (ip->x40_vel.y > max_vy) {
+            ip->x40_vel.y = max_vy;
+        }
+        if (ip->x40_vel.y < min_vy) {
+            ip->x40_vel.y = min_vy;
+        }
+        return;
+    }
+    ip->x40_vel.y = 0.0f;
+}
 
 /// @brief Spawn an Octarock stone projectile.
 /// @param gobj The parent Octarock gobj.
diff --git a/src/melee/it/items/it_27CF.c b/src/melee/it/items/ityoshiegglay.c
similarity index 91%
rename from src/melee/it/items/it_27CF.c
rename to src/melee/it/items/ityoshiegglay.c
index f0650f3f35..24654be04f 100644
--- a/src/melee/it/items/it_27CF.c
+++ b/src/melee/it/items/ityoshiegglay.c
@@ -1,4 +1,4 @@
-#include "it_27CF.h"
+#include "ityoshiegglay.h"
 
 #include "ef/efasync.h"
 
@@ -182,33 +182,3 @@ void it_27CF_Logic114_EvtUnk(Item_GObj* item_gobj, HSD_GObj* ref_gobj)
 {
     it_8026B894(item_gobj, ref_gobj);
 }
-
-void it_802B26C0(Item_GObj* gobj)
-{
-    it_8026B724(gobj);
-}
-
-void it_802B26E0(Item_GObj* gobj)
-{
-    it_8026B73C(gobj);
-}
-
-void it_802C6DB8(Item_GObj* gobj)
-{
-    it_8026B724(gobj);
-}
-
-void it_802C6DD8(Item_GObj* gobj)
-{
-    it_8026B73C(gobj);
-}
-
-void it_802C7AD0(Item_GObj* gobj)
-{
-    it_8026B724(gobj);
-}
-
-void it_802C7AF0(Item_GObj* gobj)
-{
-    it_8026B73C(gobj);
-}
diff --git a/src/melee/it/items/ityoshiegglay.h b/src/melee/it/items/ityoshiegglay.h
new file mode 100644
index 0000000000..d0c6f5c911
--- /dev/null
+++ b/src/melee/it/items/ityoshiegglay.h
@@ -0,0 +1,30 @@
+#ifndef GALE01_27CF30
+#define GALE01_27CF30
+
+#include <platform.h>
+
+#include "it/forward.h"
+
+#include "it/itCommonItems.h"
+#include "it/items/types.h"
+
+#include <baselib/forward.h>
+
+#include <dolphin/mtx.h>
+
+/* 2F2F34 */ void it_802F2F34(HSD_GObj*, it_27CF_DatAttrs*);
+/* 2F3020 */ void it_802F3020(HSD_GObj*);
+/* 2F3078 */ bool it_27CF_UnkMotion1_Anim(HSD_GObj*);
+/* 2F3120 */ void it_27CF_UnkMotion1_Phys(HSD_GObj*);
+/* 2F3150 */ bool it_27CF_UnkMotion1_Coll(HSD_GObj*);
+/* 2F317C */ void it_802F317C(HSD_GObj*);
+/* 2F31B8 */ bool it_27CF_UnkMotion0_Anim(HSD_GObj*);
+/* 2F3260 */ void it_27CF_UnkMotion0_Phys(HSD_GObj*);
+/* 2F3264 */ bool it_27CF_UnkMotion0_Coll(HSD_GObj*);
+/* 2F3290 */ void it_802F3290(Item_GObj*);
+/* 2F32F8 */ bool it_27CF_UnkMotion2_Anim(HSD_GObj*);
+/* 2F3330 */ bool it_27CF_Logic114_DmgReceived(Item_GObj*);
+/* 2F3374 */ void it_27CF_Logic114_EvtUnk(Item_GObj*, HSD_GObj*);
+/* 3F94A0 */ extern ItemStateTable it_803F94A0[];
+
+#endif
diff --git a/src/melee/it/items/it_2F2B.c b/src/melee/it/items/ityoshitongue.c
similarity index 94%
rename from src/melee/it/items/it_2F2B.c
rename to src/melee/it/items/ityoshitongue.c
index 9151b8563d..9782937461 100644
--- a/src/melee/it/items/it_2F2B.c
+++ b/src/melee/it/items/ityoshitongue.c
@@ -1,15 +1,10 @@
-#include "it_2F2B.h"
+#include "ityoshitongue.h"
 
 #include <platform.h>
 
-#include "ft/chara/ftYoshi/ftYs_Guard.h"
 #include "ft/ftlib.h"
-
-#include "ftYoshi/forward.h"
-
 #include "ftYoshi/ftYs_SpecialN.h"
 #include "it/inlines.h"
-#include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
diff --git a/src/melee/it/items/it_2F2B.h b/src/melee/it/items/ityoshitongue.h
similarity index 100%
rename from src/melee/it/items/it_2F2B.h
rename to src/melee/it/items/ityoshitongue.h
