diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.c
index 8d30ad6f87..4764b24808 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.c
@@ -1,17 +1,17 @@
-#include "ft/chara/ftPopo/ftPp_SpecialS.h"
+#include "ftPopo/ftPp_SpecialLw.h"
 
 #include "ef/eflib.h"
 #include "ef/efsync.h"
 #include "ft/chara/ftCommon/ftCo_Fall.h"
+#include "ft/chara/ftPopo/ftPp_SpecialS.h"
 #include "ft/fighter.h"
 #include "ft/ft_081B.h"
 #include "ft/ft_0881.h"
 #include "ft/ft_0892.h"
 #include "ft/ftcommon.h"
 #include "ft/ftparts.h"
-#include "ftPopo/ftPp_SpecialHi.h"
-#include "ftPopo/ftPp_SpecialLw.h"
 #include "ftPopo/ftPp_Init.h"
+#include "ftPopo/ftPp_SpecialHi.h"
 #include "it/items/itclimbersblizzard.h"
 #include "lb/lb_00B0.h"
 #include "pl/player.h"
diff --git a/src/melee/gm/gm_18A5.c b/src/melee/gm/gm_18A5.c
index c7b5c338c1..666403e049 100644
--- a/src/melee/gm/gm_18A5.c
+++ b/src/melee/gm/gm_18A5.c
@@ -785,11 +785,10 @@ void fn_8018DF68(void* arg0, s32 arg1, s32 arg2, s32 arg3, s32 arg4, s32 arg5,
     s32 half;
     GXColor c0, c1, c2, c3, c4, c5, c6, c7, c8, c9, c10;
 
-
-    c0 = (GXColor){ 255, 255, 0, 255 };
-    c10 = (GXColor){ 255, 255, 0, 255 };
+    c0 = (GXColor) { 255, 255, 0, 255 };
+    c10 = (GXColor) { 255, 255, 0, 255 };
     thickness = ((BracketEntry*) data)->x1C;
-    c1 = (GXColor){ 255, 255, 0, 255 };
+    c1 = (GXColor) { 255, 255, 0, 255 };
     DrawRectangle((f32) arg1, (f32) arg2, thickness, (f32) arg4, &c1);
 
     right = arg1 + arg3;
diff --git a/src/melee/gm/gmclassic.c b/src/melee/gm/gmclassic.c
index 6a484fd57e..934e367e35 100644
--- a/src/melee/gm/gmclassic.c
+++ b/src/melee/gm/gmclassic.c
@@ -706,9 +706,9 @@ void gmClassic_801B3500(MinorScene* arg0)
 
     gm_8017DB88(ad->x0.xC.x24, entry->x1, ad->x0.cpu_level,
                 (u8) gm_8017BE84(arg0->idx), sd->x10, sd->x0D[0],
-                (u8(*)(s32, s32, u8)) ad->x58, (u8(*)(s32, s32, u8)) ad->x5C,
-                (u8(*)(s32, s32, u8)) ad->x60, (f32(*)(s32, s32)) ad->x6C,
-                (f32(*)(s32, s32)) ad->x70);
+                (u8 (*)(s32, s32, u8)) ad->x58, (u8 (*)(s32, s32, u8)) ad->x5C,
+                (u8 (*)(s32, s32, u8)) ad->x60, (f32 (*)(s32, s32)) ad->x6C,
+                (f32 (*)(s32, s32)) ad->x70);
 
     for (i = 1; i < 3; i++) {
         sd->x0D[i] = gm_8017DB6C((gm_8017DB6C_arg0_t*) ad->x0.xC.x24, i - 1);
diff --git a/src/melee/gm/gmresultplayer.c b/src/melee/gm/gmresultplayer.c
index a49b6d64af..bcd5e2450a 100644
--- a/src/melee/gm/gmresultplayer.c
+++ b/src/melee/gm/gmresultplayer.c
@@ -1096,7 +1096,6 @@ void fn_80179990(HSD_GObj* arg0, int arg1, int arg2)
     int lookup;
     PAD_STACK(16);
 
-
     fn_801795D4();
     fn_801796F0(arg2);
 
diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index 21f4288ab4..c187c35af9 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -1,22 +1,69 @@
 #include "grhomerun.h"
 
+#include "placeholder.h"
+
 #include <platform.h>
 
+#include "baselib/archive.h"
+#include "baselib/debug.h"
+
+#include "baselib/forward.h"
+
+#include "baselib/gobj.h"
+#include "baselib/gobjgxlink.h"
+#include "baselib/gobjobject.h"
+#include "baselib/jobj.h"
+#include "baselib/memory.h"
 #include "baselib/sislib.h"
 #include "cm/camera.h"
+#include "ft/ftlib.h"
+#include "gm/gmregclear.h"
+#include "gr/grdatfiles.h"
 #include "gr/ground.h"
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
 #include "gr/stage.h"
 #include "it/it_26B1.h"
 #include "lb/lb_00B0.h"
+#include "lb/lb_00F9.h"
 #include "lb/lblanguage.h"
+#include "lb/types.h"
+#include "mp/mplib.h"
 
 f32 grHr_804D6AD8;
 int grHr_804D6ADC;
 f32 grHr_804D6AE0;
 f32 grHr_804D6AE4;
 static void* grHr_804D6AE8;
+static char grHr_804D49D0[] = "%d";
+extern GXColor grHr_804DBC74;
+extern f32 grHr_804DBC78;
+extern f32 grHr_804DBC7C;
+extern f32 grHr_804DBC80;
+extern f32 grHr_804DBC84;
+
+typedef struct grHomeRun_MainGroundVars {
+    HSD_GObj** gobjs;
+    HSD_GObj** jobj_gobjs;
+    HSD_Text* xCC;
+    HSD_JObj* xD0;
+    HSD_GObj* xD4;
+    HSD_GObj* rear_gobj;
+    HSD_GObj* rear2_gobj;
+    HSD_GObj* rear3_gobj;
+    HSD_GObj* rear4_gobj;
+    struct {
+        u8 b0 : 1;
+        u8 b1 : 1;
+        u8 b2 : 1;
+        u8 b3 : 1;
+        u8 b4 : 1;
+        u8 b5 : 1;
+        u8 b6 : 1;
+        u8 b7 : 1;
+    } xE8_flags;
+} grHomeRun_MainGroundVars;
+
 StageCallbacks grHr_803E8140[11] = {
     { grHomeRun_8021C914, grHomeRun_8021CB10, grHomeRun_8021CB18,
       grHomeRun_8021CB1C, 0 },
@@ -51,7 +98,7 @@ void grHomeRun_8021C754(void)
     grHr_804D6AE8 = Ground_801C49F8();
     stage_info.unk8C.b4 = false;
     stage_info.unk8C.b5 = true;
-    grHomeRun_8021EDD4(1);
+    grHomeRun_8021EDD4();
     grHomeRun_8021C82C(0);
     grHomeRun_8021C82C(0xA);
     grHomeRun_8021ED74();
@@ -73,6 +120,16 @@ bool grHomeRun_8021C824(void)
 {
     return false;
 }
+#pragma push
+#pragma dont_inline on
+HSD_GObj* grHomeRun_8021C82C_noinline(int gobj_id);
+
+HSD_GObj* grHomeRun_8021C82C_noinline(int gobj_id)
+{
+    return grHomeRun_8021C82C(gobj_id);
+}
+#pragma dont_inline reset
+#pragma pop
 
 HSD_GObj* grHomeRun_8021C82C(int gobj_id)
 {
@@ -112,14 +169,285 @@ void grHomeRun_8021CB18(Ground_GObj* arg) {}
 
 void grHomeRun_8021CB1C(Ground_GObj* arg) {}
 
-/// #grHomeRun_8021CB20
+void grHomeRun_8021CB20(Ground_GObj* gobj)
+{
+    static HSD_CameraDescPerspective cobj_desc = {
+        NULL,
+        0,
+        1,
+        { 0, 640, 0, 480 },
+        { 0, 640, 0, 480 },
+        NULL,
+        NULL,
+        0.0F,
+        NULL,
+        0.1F,
+        32768.0F,
+        30.0F,
+        1.4F,
+    };
+
+    Ground* gp = GET_GROUND(gobj);
+    grHomeRun_MainGroundVars* vars = (grHomeRun_MainGroundVars*) &gp->gv.homerun;
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    UnkArchiveStruct* archive;
+    HSD_CObj* cobj;
+    s16 i;
+    f32 gobjs_num;
+    HSD_GObj** gobjs;
+    HSD_GObj** jobjs;
+    PAD_STACK(8);
+
+    Ground_801C2ED0(jobj, gp->map_id);
+
+    gobjs = HSD_MemAlloc(0x100);
+    HSD_ASSERT(0x17A, gobjs);
+    vars->gobjs = gobjs;
+
+    jobjs = HSD_MemAlloc(0x40);
+    HSD_ASSERT(0x17B, jobjs);
+    vars->jobj_gobjs = jobjs;
+
+    HSD_JObjSetScaleX(jobj, grHr_804D6AE4 * HSD_JObjGetScaleX(jobj));
+    HSD_JObjSetScaleY(jobj, grHr_804D6AE4 * HSD_JObjGetScaleY(jobj));
+    HSD_JObjSetScaleZ(jobj, grHr_804D6AE4 * HSD_JObjGetScaleZ(jobj));
+
+    *(u8*) &gp->x10_flags |= 4;
+    grAnime_801C8138(gobj, gp->map_id, false);
+    mpJointSetCb1(0, gp, (mpLib_Callback) fn_8021E994);
+
+    archive = grDatFiles_801C6324();
+    vars->xD4 = GObj_Create(0x11, 0x13, 0);
+    cobj = lb_80013B14(&cobj_desc);
+    HSD_CObjSetPerspective(cobj, 30.0F, 1.4F);
+    HSD_GObjObject_80390A70(vars->xD4, HSD_GObj_804D784B, cobj);
+    GObj_SetupGXLinkMax(vars->xD4, (GObj_RenderFunc) fn_8021EB10, 7);
+    vars->xD4->gxlink_prios = 2;
+    HSD_SisLib_803A611C(1, vars->xD4, 9, 0xD, 0, 1, 0, 7);
+    HSD_SisLib_804D1124[1] =
+        HSD_ArchiveGetPublicAddress(archive->unk0, "SIS_GrHomerunData");
+    gobjs_num = 2.0F * (1.0F + (2400.0F / (160.0F * Ground_801C0498())));
+    HSD_ASSERT(0x1A2, gobjs_num < 64);
+
+    for (i = 0; (f32) i < 1.0F + (2400.0F / (160.0F * Ground_801C0498())); i++) {
+        gobjs[i] = grHomeRun_8021E500(i);
+    }
+    for (; i < 64; i++) {
+        gobjs[i] = NULL;
+    }
+
+    for (i = 0; i < 16; i++) {
+        HSD_JObj* child;
+        HSD_GObj* gobj;
+        jobjs[i] = grHomeRun_8021C82C_noinline(4);
+        gobj = jobjs[i];
+        HSD_ASSERT(0x1AB, gobj);
+
+        jobj = GET_JOBJ(gobj);
+        HSD_ASSERT(0x1AC, jobj);
+
+        HSD_JObjSetScaleX(jobj, grHr_804D6AE4 * HSD_JObjGetScaleX(jobj));
+        HSD_JObjSetScaleY(jobj, grHr_804D6AE4 * HSD_JObjGetScaleY(jobj));
+        HSD_JObjSetScaleZ(jobj, grHr_804D6AE4 * HSD_JObjGetScaleZ(jobj));
+
+        child = HSD_JObjGetChild(jobj);
+        HSD_JObjSetTranslateX(child, 0.0F);
+        HSD_JObjSetTranslateX(
+            jobj, grHr_804D6AE4 *
+                      ((4.0F + (f32) i) * -(160.0F * Ground_801C0498())));
+    }
+
+    {
+        HSD_GObj* rear_gobj = grHomeRun_8021C82C_noinline(3);
+        HSD_ASSERT(0x1BA, rear_gobj);
+        vars->rear_gobj = rear_gobj;
+        jobj = GET_JOBJ(rear_gobj);
+        HSD_ASSERT(0x1BB, jobj);
+    }
+    HSD_JObjSetTranslateX(jobj, 1.5F * -(2150.99F * Ground_801C0498()));
+
+    {
+        HSD_GObj* rear2_gobj = grHomeRun_8021C82C_noinline(3);
+        HSD_ASSERT(0x1BD, rear2_gobj);
+        vars->rear2_gobj = rear2_gobj;
+        jobj = GET_JOBJ(rear2_gobj);
+        HSD_ASSERT(0x1BE, jobj);
+    }
+
+    HSD_JObjSetTranslateX(jobj, 0.5F * -(2150.99F * Ground_801C0498()));
+
+    {
+        HSD_GObj* rear3_gobj = grHomeRun_8021C82C_noinline(3);
+        HSD_ASSERT(0x1C0, rear3_gobj);
+        vars->rear3_gobj = rear3_gobj;
+        jobj = GET_JOBJ(rear3_gobj);
+        HSD_ASSERT(0x1C1, jobj);
+    }
+
+    HSD_JObjSetTranslateX(jobj, 0.5F * (2150.99F * Ground_801C0498()));
+
+    {
+        HSD_GObj* rear4_gobj = grHomeRun_8021C82C_noinline(3);
+        HSD_ASSERT(0x1C3, rear4_gobj);
+        vars->rear4_gobj = rear4_gobj;
+        jobj = GET_JOBJ(rear4_gobj);
+        HSD_ASSERT(0x1C4, jobj);
+    }
+
+    HSD_JObjSetTranslateX(jobj, 1.5F * (2150.99F * Ground_801C0498()));
+
+    vars->xCC = NULL;
+    vars->xD0 = Ground_801C3FA4(gobj, 1);
+    vars->xE8_flags.b0 = 0;
+}
 
 bool grHomeRun_8021D678(Ground_GObj* arg)
 {
     return false;
 }
 
-/// #grHomeRun_8021D680
+void grHomeRun_8021D680(Ground_GObj* gobj)
+{
+    Vec3 pos0;
+    Vec3 pos1;
+    Vec3 pos2;
+    Vec3 cam_interest;
+    CmSubject* subject;
+    Ground* gp;
+    Ground* gp2;
+    grHomeRun_MainGroundVars* vars;
+    HSD_GObj* gobj2;
+    HSD_JObj* jobj;
+    HSD_Text* text;
+    s32 x0;
+    s32 x1;
+    s32 i;
+    f32 scale;
+    f32 y;
+    f32 z;
+    f32 x;
+
+    gp = GET_GROUND(gobj);
+    vars = (grHomeRun_MainGroundVars*) &gp->gv.homerun;
+    gobj2 = Ground_801C57A4();
+    if (gobj2 != NULL) {
+        ftLib_80086644(gobj2, &pos0);
+        subject = ftLib_80086B74(gobj2);
+        if (subject != NULL) {
+            subject->x8 = 1;
+            subject->xC_b1 = 1;
+        }
+        gobj2 = Ground_801C57C8();
+        if (gobj2 != NULL) {
+            subject = ftLib_80086B74(gobj2);
+            if (subject != NULL) {
+                subject->x8 = 1;
+                subject->xC_b1 = 1;
+            }
+        }
+        gobj2 = gm_80180AF4();
+        if (gobj2 != NULL) {
+            ftLib_80086644(gobj2, &pos1);
+            grHomeRun_8021EA30(&pos1.x);
+            if (pos1.y < 4.0F && vars->xE8_flags.b0 != 0) {
+                grHomeRun_8021EAF8();
+            }
+        }
+        vars->xE8_flags.b0 = 0;
+        if (vars->xD0 != NULL && vars->xCC == NULL) {
+            vars->xCC = grHomeRun_8021EC58(0);
+            lb_8000B1CC(vars->xD0, NULL, &pos2);
+
+            scale = Ground_801C0498();
+            z = pos2.z + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            y = -pos2.y + 0.0F * (grHr_804D6AE4 * scale);
+
+            scale = Ground_801C0498();
+            text = vars->xCC;
+            text->pos_x = pos2.x + (-1.0F) * (grHr_804D6AE4 * scale);
+            text->pos_y = y;
+            text->pos_z = z;
+        }
+
+        Camera_GetTransformInterest(&cam_interest);
+        x0 = (s32) (((cam_interest.x - 2400.0F) /
+                     (160.0F * Ground_801C0498())) -
+                    1.0F);
+        if (x0 < 0) {
+            x0 = 0;
+        }
+
+        gp2 = GET_GROUND(vars->gobjs[0]);
+        x1 = (s32) (1.0F + ((2400.0F + cam_interest.x) /
+                            (160.0F * Ground_801C0498())));
+        if (gp2->gv.homerun.xC4 < x0) {
+            Ground_801C4A08(vars->gobjs[0]);
+            for (i = 0; i < 63; i++) {
+                vars->gobjs[i] = vars->gobjs[i + 1];
+            }
+            vars->gobjs[63] = NULL;
+        } else if (gp2->gv.homerun.xC4 > x0) {
+            gobj2 = grHomeRun_8021E500(gp2->gv.homerun.xC4 - 1);
+            for (i = 63; i > 0; i--) {
+                vars->gobjs[i] = vars->gobjs[i - 1];
+            }
+            vars->gobjs[0] = gobj2;
+        }
+
+        for (i = 63; i >= 0; i--) {
+            gobj2 = vars->gobjs[i];
+            if (gobj2 != NULL) {
+                break;
+            }
+        }
+        HSD_ASSERT(0x239, gobj2);
+        gp2 = GET_GROUND(gobj2);
+        if (gp2->gv.homerun.xC4 > x1) {
+            vars->gobjs[i] = NULL;
+            Ground_801C4A08(gobj2);
+        } else if (gp2->gv.homerun.xC4 < x1) {
+            vars->gobjs[i + 1] = grHomeRun_8021E500(gp2->gv.homerun.xC4 + 1);
+        }
+
+        x = 0.85F * cam_interest.x;
+        while (x - cam_interest.x < 0.5F * -(2150.99F * Ground_801C0498())) {
+            x += 2150.99F * Ground_801C0498();
+        }
+        while (x - cam_interest.x > 0.5F * (2150.99F * Ground_801C0498())) {
+            x -= 2150.99F * Ground_801C0498();
+        }
+
+        jobj = GET_JOBJ(vars->rear_gobj);
+        HSD_ASSERT(0x257, jobj);
+        HSD_JObjSetTranslateX(
+            jobj, x - 1.5F * (2150.99F * Ground_801C0498()));
+
+        jobj = GET_JOBJ(vars->rear2_gobj);
+        HSD_ASSERT(0x259, jobj);
+
+        HSD_JObjSetTranslateX(
+            jobj, x - 0.5F * (2150.99F * Ground_801C0498()));
+
+        jobj = GET_JOBJ(vars->rear3_gobj);
+        HSD_ASSERT(0x25B, jobj);
+
+        HSD_JObjSetTranslateX(
+            jobj, x + 0.5F * (2150.99F * Ground_801C0498()));
+
+        jobj = GET_JOBJ(vars->rear4_gobj);
+        HSD_ASSERT(0x25D, jobj);
+
+        HSD_JObjSetTranslateX(
+            jobj, x + 1.5F * (2150.99F * Ground_801C0498()));
+
+        lb_800115F4();
+        Ground_801C2FE0(gobj);
+        mpLib_80056758(3, 0.0F, 0.0F, 20000.0F, 0.0F);
+        mpJointUpdateBounding(0);
+    }
+}
 
 void grHomeRun_8021DEB0(Ground_GObj* arg) {}
 
@@ -394,12 +722,97 @@ void grHomeRun_8021E4F8(Ground_GObj* arg) {}
 
 void grHomeRun_8021E4FC(Ground_GObj* arg) {}
 
-/// #grHomeRun_8021E500
+HSD_GObj* grHomeRun_8021E500(s16 arg0)
+{
+    HSD_GObj* gobj;
+    HSD_JObj* jobj;
+    HSD_JObj* child;
+    Ground* gp;
+    s32 gobj_id;
+    s32 idx;
+    f32 offset;
+
+    idx = (arg0 / grHr_804D6ADC) % 4;
+    if (idx == 0) {
+        if (grHr_804D6ADC - 1 != arg0 % grHr_804D6ADC) {
+            gobj_id = 4;
+        } else {
+            gobj_id = 5;
+        }
+    } else if (idx == 1) {
+        if (grHr_804D6ADC - 1 != arg0 % grHr_804D6ADC) {
+            gobj_id = 1;
+        } else {
+            gobj_id = 2;
+        }
+    } else if (idx == 2) {
+        if (grHr_804D6ADC - 1 != arg0 % grHr_804D6ADC) {
+            gobj_id = 6;
+        } else {
+            gobj_id = 7;
+        }
+    } else if (idx == 3) {
+        if (grHr_804D6ADC - 1 != arg0 % grHr_804D6ADC) {
+            gobj_id = 8;
+        } else {
+            gobj_id = 9;
+        }
+    } else {
+        HSD_ASSERT(0x3D2, NULL);
+        gobj_id = 0;
+    }
+
+    gobj = grHomeRun_8021C82C_noinline(gobj_id);
+    HSD_ASSERT(0x3D5, gobj);
+    jobj = GET_JOBJ(gobj);
+    HSD_ASSERT(0x3D6, jobj);
+    gp = GET_GROUND(gobj);
+    HSD_ASSERT(0x3D7, gp);
+
+    HSD_JObjSetScaleX(jobj, grHr_804D6AE4 * HSD_JObjGetScaleX(jobj));
+    HSD_JObjSetScaleY(jobj, grHr_804D6AE4 * HSD_JObjGetScaleY(jobj));
+    HSD_JObjSetScaleZ(jobj, grHr_804D6AE4 * HSD_JObjGetScaleZ(jobj));
+
+    child = HSD_JObjGetChild(jobj);
+    HSD_JObjSetTranslateX(child, 0.0F);
 
-/// #fn_8021E994
+    offset = grHr_804D6AE4 * ((f32) arg0 * (160.0F * Ground_801C0498()));
+    gp->gv.homerun.xD0 = (160.0F * (grHr_804D6AE4 * Ground_801C0498())) + offset;
+    HSD_JObjSetTranslateX(jobj, gp->gv.homerun.xD0);
+
+    gp->gv.homerun.xC6 =
+        (s16) ((s32) grHr_804D6AE0 * ((arg0 + 1) / grHr_804D6ADC));
+    gp->gv.homerun.xC4 = arg0;
+    return gobj;
+}
+
+void fn_8021E994(Ground* arg0, s32 arg1, CollData* arg2, s32 arg3,
+                 mpLib_GroundEnum arg4, f32 arg5)
+{
+    HSD_GObj* gobj;
+    Ground* gp;
+    s32 collKind = arg2->x34_flags.b1234;
+    Ground* new_var;
+    gobj = Ground_801C2BA4(0xA);
+    if (gobj != NULL) {
+        gp = (new_var = GET_GROUND(gobj));
+        if (gp == NULL) {
+            arg2->x34_flags = arg2->x34_flags;
+            return;
+        }
+        if (collKind == 1) {
+            gobj = gm_80180AF4();
+            if (arg2->x0_gobj == gobj && arg4 == 1) {
+                ((grHomeRun_MainGroundVars*) &gp->gv.homerun)->xE8_flags.b0 =
+                    1;
+            }
+        }
+    }
+}
 
 void grHomeRun_8021EA30(f32* pos)
 {
+    float new_var;
     f32 result;
     f32 time;
 
@@ -407,7 +820,9 @@ void grHomeRun_8021EA30(f32* pos)
     result = *pos - 70.0F * (grHr_804D6AE4 * time);
 
     time = Ground_801C0498();
-    result /= grHr_804D6AE4 * ((f32) grHr_804D6ADC * (160.0F * time));
+    new_var = 160.0F * time;
+    new_var = ((f32) grHr_804D6ADC) * new_var;
+    result /= grHr_804D6AE4 * new_var;
     result *= grHr_804D6AE0;
 
     if (result < 0.0F) {
@@ -415,7 +830,7 @@ void grHomeRun_8021EA30(f32* pos)
     }
 
     if (lbLang_IsSavedLanguageUS()) {
-        result = (f32) (result * 0.304788);
+        result *= 0.304788;
     }
 
     stage_info.x6E0 = 100.0F * result;
@@ -468,45 +883,28 @@ void fn_8021EB10(HSD_GObj* target_cam_gobj)
     }
 }
 
-extern const float grHr_804DBC38;
-extern const float grHr_804DBC78;
-extern const float grHr_804DBC7C;
-extern const float grHr_804DBC80;
-extern const float grHr_804DBC84;
-extern const GXColor grHr_804DBC74;
-
-HSD_Text* grHomeRun_8021EC58(u32 score)
+HSD_Text* grHomeRun_8021EC58(s32 arg0)
 {
     HSD_Text* text;
+    f32 temp_f31;
+    f32 temp_f31_2;
     GXColor color = grHr_804DBC74;
-    f32 scale;
-    f32 box_y;
-    f32 ofs_y;
 
     text = HSD_SisLib_803A6754(1, 0);
-    text->pos_z = grHr_804DBC38;
-
-    scale = grHr_804D6AE4 * Ground_801C0498();
-    box_y = grHr_804DBC78 * scale;
-
-    scale = grHr_804D6AE4 * Ground_801C0498();
-    text->box_size_x = grHr_804DBC7C * scale;
-    text->box_size_y = box_y;
+    text->pos_z = 0.0F;
+    temp_f31 = grHr_804DBC78 * (grHr_804D6AE4 * Ground_801C0498());
+    text->box_size_x = grHr_804DBC7C * (grHr_804D6AE4 * Ground_801C0498());
+    text->box_size_y = temp_f31;
     text->default_kerning = 0;
-
-    scale = grHr_804D6AE4 * Ground_801C0498();
-    ofs_y = grHr_804DBC80 * scale;
-
-    scale = grHr_804D6AE4 * Ground_801C0498();
-    text->x34.x = grHr_804DBC80 * scale;
-    text->x34.y = ofs_y;
+    temp_f31_2 = grHr_804DBC80 * (grHr_804D6AE4 * Ground_801C0498());
+    text->x34.x = grHr_804DBC80 * (grHr_804D6AE4 * Ground_801C0498());
+    text->x34.y = temp_f31_2;
     text->text_color = color;
     text->default_alignment = 1;
     text->default_fitting = 1;
     text->x4C = 1;
-
-    HSD_SisLib_803A6B98(text, grHr_804DBC38, grHr_804DBC84, "%d",
-                        (s32) (f32) (s32) score);
+    HSD_SisLib_803A6B98(text, 0.0F, grHr_804DBC84, grHr_804D49D0,
+                        (s32) (f32) arg0);
     return text;
 }
 
@@ -523,7 +921,7 @@ void grHomeRun_8021ED74(void)
     it_8026BE84(&bobomb_rain);
 }
 
-void grHomeRun_8021EDD4(int arg0)
+void grHomeRun_8021EDD4(void)
 {
     f32 scale;
     f32 max_dist;
diff --git a/src/melee/gr/grhomerun.h b/src/melee/gr/grhomerun.h
index 599a866bf1..13c89e0e21 100644
--- a/src/melee/gr/grhomerun.h
+++ b/src/melee/gr/grhomerun.h
@@ -8,6 +8,7 @@
 #include "it/inlines.h"
 
 #include "lb/forward.h"
+#include "mp/forward.h"
 
 #include <baselib/cobj.h>
 #include <baselib/fog.h>
@@ -63,14 +64,15 @@
 /* 21E4F0 */ bool grHomeRun_8021E4F0(Ground_GObj*);
 /* 21E4F8 */ void grHomeRun_8021E4F8(Ground_GObj*);
 /* 21E4FC */ void grHomeRun_8021E4FC(Ground_GObj*);
-/* 21E500 */ UNK_RET grHomeRun_8021E500(UNK_PARAMS);
-/* 21E994 */ UNK_RET fn_8021E994(UNK_PARAMS);
+/* 21E500 */ HSD_GObj* grHomeRun_8021E500(s16);
+/* 21E994 */ void fn_8021E994(Ground*, s32, CollData*, s32, mpLib_GroundEnum,
+                              f32);
 /* 21EA30 */ void grHomeRun_8021EA30(f32*);
 /* 21EAF8 */ UNK_RET grHomeRun_8021EAF8(UNK_PARAMS);
 /* 21EB10 */ void fn_8021EB10(HSD_GObj*);
-/* 21EC58 */ HSD_Text* grHomeRun_8021EC58(u32);
+/* 21EC58 */ HSD_Text* grHomeRun_8021EC58(s32 arg0);
 /* 21ED74 */ void grHomeRun_8021ED74(void);
-/* 21EDD4 */ void grHomeRun_8021EDD4(int);
+/* 21EDD4 */ void grHomeRun_8021EDD4(void);
 /* 21EEB4 */ DynamicsDesc* grHomeRun_8021EEB4(enum_t);
 /* 21EEBC */ bool grHomeRun_8021EEBC(Vec3*, int arg, HSD_JObj* jobj);
 /* 21EF10 */ float grHomeRun_8021EF10(void);
diff --git a/src/melee/gr/gricemt.c b/src/melee/gr/gricemt.c
index dfca12a3cd..6496bc085d 100644
--- a/src/melee/gr/gricemt.c
+++ b/src/melee/gr/gricemt.c
@@ -1283,7 +1283,7 @@ bool grIceMt_801FA364(void* state_, f32* out, HSD_GObjEvent cb_,
                       Ground_GObj* gobj)
 {
     grIceMt_FA364_State* state = state_;
-    s32 (*cb)(HSD_GObj*, s32*) = (s32(*)(HSD_GObj*, s32*)) cb_;
+    s32 (*cb)(HSD_GObj*, s32*) = (s32 (*)(HSD_GObj*, s32*)) cb_;
     bool ret = true;
     f32 result;
     s32 next_delay;
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index c9df12dc3b..39dc4a498d 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -1497,6 +1497,7 @@ struct grHomeRun_GroundVars {
     u16 xC6;
     int xC8;
     int xCC;
+    f32 xD0;
 };
 
 struct Map_GroundVars {
diff --git a/src/melee/lb/lbbgflash.c b/src/melee/lb/lbbgflash.c
index 3e8d064cb4..5d1a613efa 100644
--- a/src/melee/lb/lbbgflash.c
+++ b/src/melee/lb/lbbgflash.c
@@ -151,7 +151,6 @@ void fn_8001FEC4(HSD_GObj* gobj, s32 code)
     s32 y;
     PAD_STACK(24);
 
-
     if (data->state.active) {
         return;
     }
diff --git a/src/melee/mn/mnnamenew.c b/src/melee/mn/mnnamenew.c
index 16d512e7b5..fdd67c6ec5 100644
--- a/src/melee/mn/mnnamenew.c
+++ b/src/melee/mn/mnnamenew.c
@@ -1212,7 +1212,6 @@ s32 mnNameNew_8023D130(GlyphVariantEntry* arg0, u8 arg1, u8 arg2, s32 arg3)
     GlyphRow* table_lower;
     GXColor* sp2C_ptr;
 
-
     jobj14 = arg0->jobjs[4];
     text = HSD_SisLib_803A6754(0, (s32) mn_804D6BB4);
     jobj18 = arg0->jobjs[5];
diff --git a/src/melee/ty/tydisplay.c b/src/melee/ty/tydisplay.c
index ab84ebd059..6bb09e97fd 100644
--- a/src/melee/ty/tydisplay.c
+++ b/src/melee/ty/tydisplay.c
@@ -896,7 +896,6 @@ void un_80319EF0(void)
     f32 scale;
     PAD_STACK(16);
 
-
     cobj = (HSD_CObj*) cfg->x00->hsd_obj;
 
     range = grid->x0C_max_x - grid->x04_min_x;
