diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index 59f4ab780e..25428456b8 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -28,6 +28,7 @@
 #include "ft/inlines.h"
 #include "ft/types.h"
 #include "ftCommon/ftCo_CaptureKirby.h"
+#include "ftCommon/ftCo_CaptureWaitKirby.h"
 #include "ftCommon/ftCo_Fall.h"
 #include "ftCommon/ftCo_Landing.h"
 
@@ -170,9 +171,44 @@ void ftKb_AttackDashAir_800F22D4(Fighter_GObj* gobj)
     ftPartSetRotX(ft, 0, 0.0F);
 }
 
-/// #ftKb_SpecialHi_Enter
+void ftKb_SpecialHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->cmd_vars[3] = 0;
+    fp->cmd_vars[2] = 0;
+    fp->cmd_vars[1] = 0;
+    fp->cmd_vars[0] = 0;
+    fp->mv.kb.specialhi.x0 = 0;
+    fp->mv.kb.specialhi.x4 = 0;
+    fp->mv.kb.specialhi.x8.i = 0;
+    fp->mv.kb.specialhi.xC = 0;
+    Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialHi1, 0, 0.0f, 1.0f, 0.0f,
+                              NULL);
+    ftAnim_8006EBA4(gobj);
+    fp = GET_FIGHTER(gobj);
+    efSync_Spawn(0x494, gobj);
+    fp->x2219_b0 = 1;
+    fp->pre_hitlag_cb = efLib_PauseAll;
+    fp->post_hitlag_cb = efLib_ResumeAll;
+}
 
-/// #ftKb_SpecialAirHi_Enter
+void ftKb_SpecialAirHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->cmd_vars[0] = fp->cmd_vars[1] = fp->cmd_vars[2] = fp->cmd_vars[3] = 0;
+    fp->mv.kb.specialhi.x0 = 0;
+    fp->mv.kb.specialhi.x4 = 0;
+    fp->mv.kb.specialhi.x8.i = 0;
+    fp->mv.kb.specialhi.xC = 0;
+    Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirHi1, 0, 0.0f, 1.0f, 0.0f,
+                              NULL);
+    ftAnim_8006EBA4(gobj);
+    fp = GET_FIGHTER(gobj);
+    efSync_Spawn(0x494, gobj);
+    fp->x2219_b0 = 1;
+    fp->pre_hitlag_cb = efLib_PauseAll;
+    fp->post_hitlag_cb = efLib_ResumeAll;
+}
 
 void ftKb_SpecialHi1_Anim(Fighter_GObj* gobj)
 {
@@ -1548,9 +1584,57 @@ void ftKb_SpecialN_800F5D04(Fighter_GObj* gobj, bool arg1)
     new_var->fv.kb.hat.kind = 4;
 }
 
-/// #ftKb_SpecialN_800F5DE8
+#pragma dont_inline on
+void ftKb_SpecialN_800F5DE8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    Vec3 pos;
+
+    pos = fp->cur_pos;
+    pos.x += da->specialn_x_offset_inhaled * fp->facing_dir;
+    pos.y += da->specialn_y_offset_inhaled;
+
+    {
+        f64 dist =
+            it_802F23AC((int*) fp->target_item_gobj, (float*) &pos);
+        if (dist <
+            da->specialn_inhale_velocity * da->specialn_inhale_velocity)
+        {
+            it_802F2810(fp->target_item_gobj);
+            if (fp->ground_or_air == GA_Air) {
+                ftKb_SpecialN_800F63EC(gobj);
+            } else {
+                ftKb_SpecialN_800F6388(gobj);
+            }
+        }
+    }
+}
+#pragma dont_inline off
+
+void ftKb_SpecialN_800F5EA8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    Vec3 pos;
 
-/// #ftKb_SpecialN_800F5EA8
+    pos = fp->cur_pos;
+    pos.x += da->specialn_x_offset_inhaled * fp->facing_dir;
+    pos.y += da->specialn_y_offset_inhaled;
+
+    {
+        f32 dist = ftCo_800BD19C(fp->victim_gobj, &pos);
+        if (dist < da->specialn_inhale_velocity * da->specialn_inhale_velocity)
+        {
+            ftCo_800BD620(fp->victim_gobj);
+            if (fp->ground_or_air == GA_Air) {
+                ftKb_SpecialN_800F63EC(gobj);
+            } else {
+                ftKb_SpecialN_800F6388(gobj);
+            }
+        }
+    }
+}
 
 void ftKb_SpecialN_800F5F68(HSD_GObj* gobj)
 {
@@ -1982,6 +2066,7 @@ void ftKb_SpecialAirNEnd_Anim(Fighter_GObj* gobj)
 
 /// #ftKb_SpecialNCapture_Anim
 
+#pragma dont_inline on
 void ftKb_SpecialNCapture1_Anim(Fighter_GObj* gobj)
 {
     ftKb_SpecialN_800F5DE8(gobj);
@@ -1991,6 +2076,7 @@ void ftKb_SpecialAirNCapture_Anim(Fighter_GObj* gobj)
 {
     ftKb_SpecialN_800F5EA8(gobj);
 }
+#pragma dont_inline off
 
 void ftKb_Eat_Anim(HSD_GObj* gobj)
 {
@@ -2759,11 +2845,9 @@ void fn_800F98F4(Fighter_GObj* gobj)
         PAD_STACK(4);
         bone = ftParts_GetBoneIndex(fp, FtPart_LHandN);
         lb_8000B1CC(fp->parts[bone].joint, NULL, &pos);
-        it_802C01AC(gobj, &pos, It_Kind_Kirby_LuigiFire,
-                    fp->facing_dir);
+        it_802C01AC(gobj, &pos, It_Kind_Kirby_LuigiFire, fp->facing_dir);
         bone = ftParts_GetBoneIndex(fp, FtPart_LHandN);
-        efSync_Spawn(0x4B1, gobj, fp->parts[bone].joint,
-                     &fp->facing_dir);
+        efSync_Spawn(0x4B1, gobj, fp->parts[bone].joint, &fp->facing_dir);
     }
 }
 
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
index f2e84b06ec..dafb6252c4 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
@@ -35,6 +35,7 @@
 
 /* 105A34 */ static void fn_80105A34(Fighter_GObj* gobj);
 
+extern ftCollisionBox ftKb_Init_803CB758;
 extern f32 ftKb_Init_804D94F0;
 extern f32 ftKb_Init_804D94F4;
 
@@ -1180,7 +1181,29 @@ void ftKb_PrSpecialNEnd_Coll(Fighter_GObj* gobj)
     }
 }
 
-/// #ftKb_PrSpecialAirNStart_Coll
+void ftKb_PrSpecialAirNStart_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (ft_80081D0C(gobj) != GA_Ground) {
+        FtMotionId msid;
+        ftCommon_8007D7FC(fp);
+        if (1.0F == fp->mv.pr.specialn.x34.x) {
+            msid = ftKb_MS_PrSpecialNStartL;
+        } else {
+            msid = ftKb_MS_PrSpecialNLoop;
+        }
+        Fighter_ChangeMotionState(gobj, msid, 0x0C4C5092U,
+                                  fp->cur_anim_frame, 1.0F, 0.0F,
+                                  NULL);
+        {
+            Fighter* fp2 = GET_FIGHTER(gobj);
+            fp2->death2_cb = ftKb_Init_800EE74C;
+            fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+            fp2->deal_dmg_cb = fn_80100E0C;
+            fp2->x21F8 = fn_80105978;
+        }
+    }
+}
 
 /// #ftKb_PrSpecialAirNLoop_Coll
 
@@ -1188,7 +1211,26 @@ void ftKb_PrSpecialNEnd_Coll(Fighter_GObj* gobj)
 
 /// #ftKb_PrSpecialAirN_Coll
 
-/// #ftKb_PrSpecialN_Coll
+void ftKb_PrSpecialN_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (ft_800824A0(gobj, &ftKb_Init_803CB758)) {
+        ftCommon_8007D7FC(fp);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_PrSpecialNEndR,
+                                  0x0C4C5092U, fp->cur_anim_frame,
+                                  0.0F, 0.0F, NULL);
+        {
+            Fighter* fp2 = GET_FIGHTER(gobj);
+            fp2->death2_cb = ftKb_Init_800EE74C;
+            fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+            fp2->deal_dmg_cb = fn_80100E0C;
+            fp2->x21F8 = fn_80105978;
+        }
+        fp->gr_vel = fp->self_vel.x;
+        fp->self_vel.y = 0;
+        fp->self_vel.z = 0;
+    }
+}
 
 void ftKb_PrSpecialAirNEndR_Coll(Fighter_GObj* gobj)
 {
@@ -1204,7 +1246,21 @@ void ftKb_PrSpecialAirNEndR_Coll(Fighter_GObj* gobj)
 
 /// #ftKb_PrSpecialNHit_Coll
 
-/// #fn_80105978
+void fn_80105978(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftPartSetRotY(fp, 0, M_PI_2 * fp->facing_dir);
+    fp->self_vel.x = -fp->self_vel.x;
+    fp->gr_vel = -fp->gr_vel;
+    fp->xE4_ground_accel_1 = -fp->xE4_ground_accel_1;
+    fp->mv.pr.specialn.x10 = -fp->mv.pr.specialn.x10;
+    fp->mv.pr.specialn.x14 = -fp->mv.pr.specialn.x14;
+    fp->mv.pr.specialn.x18 = -fp->mv.pr.specialn.x18;
+    fp->mv.pr.specialn.x1C = -fp->mv.pr.specialn.x1C;
+    fp->mv.pr.specialn.facing_dir = -fp->mv.pr.specialn.facing_dir;
+    fp->mv.pr.specialn.x34.x = -fp->mv.pr.specialn.x34.x;
+    fp->mv.pr.specialn.x34.y = -fp->mv.pr.specialn.x34.y;
+}
 
 void fn_80105A34(Fighter_GObj* gobj)
 {
