diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index db0eefd64b..875d16d5f0 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -24700,7 +24700,8 @@ un_804000D0 = .data:0x804000D0; // type:object size:0x38 scope:global
 un_80400128 = .data:0x80400128; // type:object size:0xB8 scope:global data:4byte
 @187 = .data:0x804001E0; // type:object size:0xB scope:local data:string
 @188 = .data:0x804001EC; // type:object size:0x10 scope:local data:string
-un_80400200 = .data:0x80400200; // type:object size:0x58 scope:global
+un_80400200 = .data:0x80400200; // type:object size:0x54 scope:global
+...data.0 = .data:0x80400200; // type:label scope:local
 un_80400258 = .data:0x80400258; // type:object size:0xC scope:global
 un_80400264 = .data:0x80400264; // type:object size:0x25 scope:global data:string
 lbl_8040028C = .data:0x8040028C; // type:object size:0xD scope:global data:string
@@ -27384,7 +27385,7 @@ un_804D5AE8 = .sdata:0x804D5AE8; // type:object size:0x7 scope:global data:strin
 un_804D5AF0 = .sdata:0x804D5AF0; // type:object size:0x5 scope:global data:string
 @203 = .sdata:0x804D5AF8; // type:object size:0x7 scope:local data:string
 @204 = .sdata:0x804D5B00; // type:object size:0x5 scope:local data:string
-un_804D5B08 = .sdata:0x804D5B08; // type:object size:0x8 scope:global data:byte
+un_804D5B08 = .sdata:0x804D5B08; // type:object size:0x4 scope:global data:byte
 un_804D5B10 = .sdata:0x804D5B10; // type:object size:0x7 scope:global data:string
 un_804D5B18 = .sdata:0x804D5B18; // type:object size:0x5 scope:global data:string
 un_804D5B20 = .sdata:0x804D5B20; // type:object size:0x7 scope:global data:string
@@ -34046,13 +34047,13 @@ un_804DE0A4 = .sdata2:0x804DE0A4; // type:object size:0x4 scope:global align:4 d
 @183 = .sdata2:0x804DE0C8; // type:object size:0x4 scope:local align:4 data:float
 @184 = .sdata2:0x804DE0CC; // type:object size:0x4 scope:local align:4 data:float
 @189 = .sdata2:0x804DE0D0; // type:object size:0x4 scope:local align:4 data:float
-un_804DE0D8 = .sdata2:0x804DE0D8; // type:object size:0x4 scope:global align:4 data:float
-un_804DE0DC = .sdata2:0x804DE0DC; // type:object size:0x4 scope:global align:4 data:float
-un_804DE0E0 = .sdata2:0x804DE0E0; // type:object size:0x4 scope:global align:4 data:float
-un_804DE0E4 = .sdata2:0x804DE0E4; // type:object size:0x4 scope:global align:4 data:float
-un_804DE0E8 = .sdata2:0x804DE0E8; // type:object size:0x4 scope:global align:4 data:float
-un_804DE0EC = .sdata2:0x804DE0EC; // type:object size:0x4 scope:global align:4 data:float
-un_804DE0F0 = .sdata2:0x804DE0F0; // type:object size:0x4 scope:global align:4 data:float
+@166 = .sdata2:0x804DE0D8; // type:object size:0x4 scope:local align:4 data:float
+@167 = .sdata2:0x804DE0DC; // type:object size:0x4 scope:local align:4 data:float
+@168 = .sdata2:0x804DE0E0; // type:object size:0x4 scope:local align:4 data:float
+@192 = .sdata2:0x804DE0E4; // type:object size:0x4 scope:local align:4 data:float
+@193 = .sdata2:0x804DE0E8; // type:object size:0x4 scope:local align:4 data:float
+@194 = .sdata2:0x804DE0EC; // type:object size:0x4 scope:local align:4 data:float
+@195 = .sdata2:0x804DE0F0; // type:object size:0x4 scope:local align:4 data:float
 un_804DE0F8 = .sdata2:0x804DE0F8; // type:object size:0x4 scope:global align:4 data:float
 un_804DE0FC = .sdata2:0x804DE0FC; // type:object size:0x4 scope:global align:4 data:float
 un_804DE100 = .sdata2:0x804DE100; // type:object size:0x4 scope:global align:4 data:float
diff --git a/configure.py b/configure.py
index f0b03a0ab6..4167af3da7 100755
--- a/configure.py
+++ b/configure.py
@@ -1320,7 +1320,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/vi/vi0502.c"),
             Object(Matching, "melee/vi/vi0601.c"),
             Object(Matching, "melee/vi/vi0801.c"),
-            Object(NonMatching, "melee/vi/vi1101.c"),
+            Object(Matching, "melee/vi/vi1101.c"),
             Object(NonMatching, "melee/vi/vi1201v1.c"),
             Object(NonMatching, "melee/vi/vi1201v2.c"),
             Object(NonMatching, "melee/vi/vi1202.c"),
diff --git a/src/melee/gr/grmutecity.c b/src/melee/gr/grmutecity.c
index 84e527155b..6388a16a50 100644
--- a/src/melee/gr/grmutecity.c
+++ b/src/melee/gr/grmutecity.c
@@ -36,6 +36,7 @@
 #include <baselib/lobj.h>
 #include <baselib/psappsrt.h>
 #include <baselib/psstructs.h>
+#include <MSL/math_ppc.h>
 
 extern s16 grMc_803E34A4[];
 extern Vec3 grMc_803B81B8;
diff --git a/src/melee/gr/grrcruise.c b/src/melee/gr/grrcruise.c
index 7b259bf576..f77c028706 100644
--- a/src/melee/gr/grrcruise.c
+++ b/src/melee/gr/grrcruise.c
@@ -36,6 +36,7 @@
 #include <baselib/gobjproc.h>
 #include <baselib/jobj.h>
 #include <baselib/random.h>
+#include <MSL/math_ppc.h>
 #include <MSL/trigf.h>
 
 S16Vec3 grRc_803E4DA8[] = {
diff --git a/src/melee/if/ifmagnify.c b/src/melee/if/ifmagnify.c
index 2210e7da44..f85bf4fe00 100644
--- a/src/melee/if/ifmagnify.c
+++ b/src/melee/if/ifmagnify.c
@@ -58,9 +58,8 @@ static GXColor* (*ifMagnify_803F9828[])(void) = {
 };
 
 static s32 ifMagnify_803F984C[0x10] = {
-    0,          0x10001,   0x01020102, 0x02020202,
-    0x303,      0x10304,   0x01020405, 0x02020505,
-    0x03030606, 0x03040607, 0x04050708, 0x05050808,
+    0,          0x10001,    0x01020102, 0x02020202, 0x303,      0x10304,
+    0x01020405, 0x02020505, 0x03030606, 0x03040607, 0x04050708, 0x05050808,
     0x06060606, 0x06070607, 0x07080708, 0x08080808,
 };
 
@@ -72,7 +71,8 @@ s32 ifMagnify_802FB6E8(s32 slot)
     return 0;
 }
 
-ifMagnifyPlayer* ifMagnify_802FB73C(ifMagnifyPlayer* arg0, Vec2* arg1, Vec2* arg2)
+ifMagnifyPlayer* ifMagnify_802FB73C(ifMagnifyPlayer* arg0, Vec2* arg1,
+                                    Vec2* arg2)
 {
     f32 temp_f1;
     f32 temp_f1_2;
@@ -181,14 +181,16 @@ void ifMagnify_802FB8C0(HSD_GObj* arg0, s32 arg1)
             translate.x = 0.09125f * out.x;
             translate.y = 0.1f * out.y;
             translate.z = 0.0f;
-            HSD_JObjSetTranslate((HSD_JObj*) player->gobj->hsd_obj, &translate);
+            HSD_JObjSetTranslate((HSD_JObj*) player->gobj->hsd_obj,
+                                 &translate);
 
             HSD_GObj_JObjCallback(arg0, arg1);
             if ((player->state.unk == 4) || (player->state.unk == 2)) {
                 slot_type = Player_GetPlayerSlotType(slot);
                 teams_enabled = gm_8016B168();
-                color = gm_80160968(gm_80160854((u8) slot, Player_GetTeam(slot),
-                                                teams_enabled, slot_type));
+                color =
+                    gm_80160968(gm_80160854((u8) slot, Player_GetTeam(slot),
+                                            teams_enabled, slot_type));
                 if (player->state.unk == 2) {
                     arrow_kind = 1;
                 } else {
@@ -285,20 +287,18 @@ void ifMagnify_802FBBDC(HSD_GObj* arg0)
             }
             if (((s32) x_class - 1) == 0) {
                 x_blend =
-                    1.0f -
-                    ((world_pos.x - Stage_GetCamBoundsLeftOffset()) /
-                     ((0.5f * (Stage_GetCamBoundsLeftOffset() +
-                               Stage_GetCamBoundsRightOffset())) -
-                      Stage_GetCamBoundsLeftOffset()));
+                    1.0f - ((world_pos.x - Stage_GetCamBoundsLeftOffset()) /
+                            ((0.5f * (Stage_GetCamBoundsLeftOffset() +
+                                      Stage_GetCamBoundsRightOffset())) -
+                             Stage_GetCamBoundsLeftOffset()));
             } else {
                 x_blend =
-                    1.0f -
-                    (-((0.5f * (Stage_GetCamBoundsLeftOffset() +
-                                Stage_GetCamBoundsRightOffset())) -
-                       world_pos.x) /
-                     -((0.5f * (Stage_GetCamBoundsLeftOffset() +
-                                 Stage_GetCamBoundsRightOffset())) -
-                       Stage_GetCamBoundsRightOffset()));
+                    1.0f - (-((0.5f * (Stage_GetCamBoundsLeftOffset() +
+                                       Stage_GetCamBoundsRightOffset())) -
+                              world_pos.x) /
+                            -((0.5f * (Stage_GetCamBoundsLeftOffset() +
+                                       Stage_GetCamBoundsRightOffset())) -
+                              Stage_GetCamBoundsRightOffset()));
             }
         }
         x_inv = 1.0f - x_blend;
@@ -321,20 +321,18 @@ void ifMagnify_802FBBDC(HSD_GObj* arg0)
             }
             if (((s32) y_class - 1) == 0) {
                 y_blend =
-                    1.0f -
-                    ((Stage_GetCamBoundsTopOffset() - world_pos.y) /
-                     -((0.5f * (Stage_GetCamBoundsTopOffset() +
-                                Stage_GetCamBoundsBottomOffset())) -
-                       Stage_GetCamBoundsTopOffset()));
+                    1.0f - ((Stage_GetCamBoundsTopOffset() - world_pos.y) /
+                            -((0.5f * (Stage_GetCamBoundsTopOffset() +
+                                       Stage_GetCamBoundsBottomOffset())) -
+                              Stage_GetCamBoundsTopOffset()));
             } else {
                 y_blend =
-                    1.0f -
-                    (((0.5f * (Stage_GetCamBoundsTopOffset() +
-                               Stage_GetCamBoundsBottomOffset())) -
-                      world_pos.y) /
-                     ((0.5f * (Stage_GetCamBoundsTopOffset() +
-                               Stage_GetCamBoundsBottomOffset())) -
-                      Stage_GetCamBoundsBottomOffset()));
+                    1.0f - (((0.5f * (Stage_GetCamBoundsTopOffset() +
+                                      Stage_GetCamBoundsBottomOffset())) -
+                             world_pos.y) /
+                            ((0.5f * (Stage_GetCamBoundsTopOffset() +
+                                      Stage_GetCamBoundsBottomOffset())) -
+                             Stage_GetCamBoundsBottomOffset()));
             }
         }
         y_inv = 1.0f - y_blend;
@@ -342,9 +340,8 @@ void ifMagnify_802FBBDC(HSD_GObj* arg0)
             y_class = 0.0f;
         } else if (world_pos.y < Stage_GetCamBoundsBottomOffset()) {
             y_class = 3.0f;
-        } else if (world_pos.y >
-                   (0.5f * (Stage_GetCamBoundsTopOffset() +
-                            Stage_GetCamBoundsBottomOffset())))
+        } else if (world_pos.y > (0.5f * (Stage_GetCamBoundsTopOffset() +
+                                          Stage_GetCamBoundsBottomOffset())))
         {
             y_class = 1.0f;
         } else {
@@ -354,15 +351,15 @@ void ifMagnify_802FBBDC(HSD_GObj* arg0)
             x_class = 0.0f;
         } else if (world_pos.x > Stage_GetCamBoundsRightOffset()) {
             x_class = 3.0f;
-        } else if (world_pos.x <
-                   (0.5f * (Stage_GetCamBoundsLeftOffset() +
-                            Stage_GetCamBoundsRightOffset())))
+        } else if (world_pos.x < (0.5f * (Stage_GetCamBoundsLeftOffset() +
+                                          Stage_GetCamBoundsRightOffset())))
         {
             x_class = 1.0f;
         } else {
             x_class = 2.0f;
         }
-        color_ids = (u8*) &ifMagnify_803F984C[(s32) x_class + ((s32) y_class * 4)];
+        color_ids =
+            (u8*) &ifMagnify_803F984C[(s32) x_class + ((s32) y_class * 4)];
         for (j = 0; j < 4; j++) {
             colors[j] = *ifMagnify_803F9828[color_ids[j]]();
         }
@@ -435,12 +432,11 @@ void ifMagnify_802FC3C0(s32 slot)
         image_descs = (HSD_ImageDesc*) ifMagnify_804A1DE0.image_descs;
         image_descs[slot - 1] = *ifMagnify_804A1DE0.player[0].idesc;
         player->idesc = &image_descs[slot - 1];
-        player->idesc->image_ptr = HSD_MemAlloc((GXGetTexBufferSize(
-                                                     player->idesc->width,
-                                                     player->idesc->height,
-                                                     player->idesc->format, 0, 0) +
-                                                 0x1F) &
-                                                ~0x1F);
+        player->idesc->image_ptr = HSD_MemAlloc(
+            (GXGetTexBufferSize(player->idesc->width, player->idesc->height,
+                                player->idesc->format, 0, 0) +
+             0x1F) &
+            ~0x1F);
         child->u.dobj->next->mobj->tobj->imagedesc = player->idesc;
     }
 
@@ -448,8 +444,8 @@ void ifMagnify_802FC3C0(s32 slot)
 
     slot_type = Player_GetPlayerSlotType(slot);
     teams_enabled = gm_8016B168();
-    color = gm_80160968(
-        gm_80160854((u8) slot, Player_GetTeam(slot), teams_enabled, slot_type));
+    color = gm_80160968(gm_80160854((u8) slot, Player_GetTeam(slot),
+                                    teams_enabled, slot_type));
 
     dobj = player->jobj->u.dobj;
     mobj = dobj->mobj;
diff --git a/src/melee/if/ifmagnify.h b/src/melee/if/ifmagnify.h
index db9af766fe..9aef58690a 100644
--- a/src/melee/if/ifmagnify.h
+++ b/src/melee/if/ifmagnify.h
@@ -8,7 +8,8 @@
 #include <dolphin/mtx.h>
 
 /* 2FB6E8 */ s32 ifMagnify_802FB6E8(s32);
-/* 2FB73C */ ifMagnifyPlayer* ifMagnify_802FB73C(ifMagnifyPlayer*, Vec2*, Vec2*);
+/* 2FB73C */ ifMagnifyPlayer* ifMagnify_802FB73C(ifMagnifyPlayer*, Vec2*,
+                                                 Vec2*);
 /* 2FB8C0 */ void ifMagnify_802FB8C0(HSD_GObj*, s32);
 /* 2FBBDC */ void ifMagnify_802FBBDC(HSD_GObj*);
 /* 2FC3BC */ void ifMagnify_802FC3BC(void);
diff --git a/src/melee/lb/lbcollision.c b/src/melee/lb/lbcollision.c
index e697515d48..e235fc69bb 100644
--- a/src/melee/lb/lbcollision.c
+++ b/src/melee/lb/lbcollision.c
@@ -24,6 +24,7 @@
 #include <baselib/state.h>
 #include <baselib/tev.h>
 #include <MetroTRK/intrinsics.h>
+#include <MSL/math_ppc.h>
 
 /* 006E58 */ static bool lbColl_80006E58(Vec3* arg0, Vec3* arg1, Vec3* arg2,
                                          Vec3* arg3, Vec3* arg4, Vec3* arg5,
diff --git a/src/melee/mn/mndatadel.c b/src/melee/mn/mndatadel.c
index 26e236e70d..4fff80dce1 100644
--- a/src/melee/mn/mndatadel.c
+++ b/src/melee/mn/mndatadel.c
@@ -100,9 +100,9 @@ void mnDataDel_8024EA6C(void)
     temp_r31 = mnDataDel_804D6C68->user_data;
     do {
         temp_ret = mn_80231634(
-            (struct mn_80231634_t*)
-                ((struct MnDataDelGObjUserData*) mnDataDel_804D6C68->user_data)
-                    ->x10[*var_r30]);
+            (struct mn_80231634_t*) ((struct MnDataDelGObjUserData*)
+                                         mnDataDel_804D6C68->user_data)
+                ->x10[*var_r30]);
         lb_80011E24((HSD_JObj*) temp_ret, &sp18, 1, -1);
         temp_f31 = mn_8022F298(sp18);
         HSD_JObjReqAnimAll(sp18, temp_f30);
@@ -355,10 +355,9 @@ void fn_8024F318(HSD_GObj* gobj)
                 } else {
                     sis_id = 0x13F;
                 }
-                text = HSD_SisLib_803A5ACC(0, 1, mnDataDel_803EF870.x64,
-                                           mnDataDel_803EF870.x68,
-                                           mnDataDel_803EF870.x6C, 250.0f,
-                                           5.0f);
+                text = HSD_SisLib_803A5ACC(
+                    0, 1, mnDataDel_803EF870.x64, mnDataDel_803EF870.x68,
+                    mnDataDel_803EF870.x6C, 250.0f, 5.0f);
                 mnDataDel_804D6C6C = text;
                 text->font_size.x = 0.05f;
                 text->font_size.y = 0.05f;
@@ -521,9 +520,9 @@ void fn_8024F840(HSD_GObj* gobj)
     }
     if (buttons & 1) {
         lbAudioAx_80024030(2);
-        joint = (HSD_JObj*) mn_80231634((struct mn_80231634_t*)
-                                            user_data->x10[(
-                                                (s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
+        joint = (HSD_JObj*) mn_80231634(
+            (struct mn_80231634_t*) user_data
+                ->x10[((s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
         mnDataDel_8024EBC8(joint, user_data->x0, 0U, 0U);
         cursor = user_data->x0;
         if (cursor != 0) {
@@ -531,9 +530,9 @@ void fn_8024F840(HSD_GObj* gobj)
         } else {
             user_data->x0 = 5U;
         }
-        joint = (HSD_JObj*) mn_80231634((struct mn_80231634_t*)
-                                            user_data->x10[(
-                                                (s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
+        joint = (HSD_JObj*) mn_80231634(
+            (struct mn_80231634_t*) user_data
+                ->x10[((s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
         mnDataDel_8024EBC8(joint, user_data->x0, 1U, 0U);
         if (user_data->xC != NULL) {
             HSD_SisLib_803A5CC4(user_data->xC);
@@ -549,9 +548,9 @@ void fn_8024F840(HSD_GObj* gobj)
     }
     if (buttons & 2) {
         lbAudioAx_80024030(2);
-        joint = (HSD_JObj*) mn_80231634((struct mn_80231634_t*)
-                                            user_data->x10[(
-                                                (s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
+        joint = (HSD_JObj*) mn_80231634(
+            (struct mn_80231634_t*) user_data
+                ->x10[((s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
         mnDataDel_8024EBC8(joint, user_data->x0, 0U, 0U);
         cursor = user_data->x0;
         if (cursor == 5) {
@@ -559,9 +558,9 @@ void fn_8024F840(HSD_GObj* gobj)
         } else {
             user_data->x0 = cursor + 1;
         }
-        joint = (HSD_JObj*) mn_80231634((struct mn_80231634_t*)
-                                            user_data->x10[(
-                                                (s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
+        joint = (HSD_JObj*) mn_80231634(
+            (struct mn_80231634_t*) user_data
+                ->x10[((s32*) &mnDataDel_803EF870.x3C)[user_data->x0]]);
         mnDataDel_8024EBC8(joint, user_data->x0, 1U, 0U);
         if (user_data->xC != NULL) {
             HSD_SisLib_803A5CC4(user_data->xC);
@@ -744,7 +743,8 @@ void mnDataDel_8024FE4C(u8 arg0)
         HSD_SisLib_803A5CC4(text);
     }
     sis_id = ((u16*) &mnDataDel_803EF870.x58)[user_data->x0];
-    text = HSD_SisLib_803A5ACC(0, 0, -9.5f, 9.1f, 17.0f, 364.68332f, 38.38772f);
+    text =
+        HSD_SisLib_803A5ACC(0, 0, -9.5f, 9.1f, 17.0f, 364.68332f, 38.38772f);
     user_data->xC = text;
     text->font_size.x = 0.0521f;
     text->font_size.y = 0.0521f;
@@ -764,8 +764,7 @@ void mnDataDel_80250170(void)
         mn_804D6BB8, (void**) &mnDataDel_804A0918.joint,
         "MenMainConDl_Top_joint", &mnDataDel_804A0918.animjoint,
         "MenMainConDl_Top_animjoint", &mnDataDel_804A0918.matanim_joint,
-        "MenMainConDl_Top_matanim_joint",
-        &mnDataDel_804A0918.shapeanim_joint,
+        "MenMainConDl_Top_matanim_joint", &mnDataDel_804A0918.shapeanim_joint,
         "MenMainConDl_Top_shapeanim_joint", &mnDataDel_804A0928.joint,
         "MenMainCursorDl_Top_joint", &mnDataDel_804A0928.animjoint,
         "MenMainCursorDl_Top_animjoint", &mnDataDel_804A0928.matanim_joint,
@@ -774,8 +773,7 @@ void mnDataDel_80250170(void)
         "MenMainCursorDl_Top_shapeanim_joint", &mnDataDel_804A0938.joint,
         "MenMainWarCmn_Top_joint", &mnDataDel_804A0938.animjoint,
         "MenMainWarCmn_Top_animjoint", &mnDataDel_804A0938.matanim_joint,
-        "MenMainWarCmn_Top_matanim_joint",
-        &mnDataDel_804A0938.shapeanim_joint,
+        "MenMainWarCmn_Top_matanim_joint", &mnDataDel_804A0938.shapeanim_joint,
         "MenMainWarCmn_Top_shapeanim_joint", 0);
     mnDataDel_8024FE4C(0U);
     proc = HSD_GObj_SetupProc(GObj_Create(0U, 1U, 0x80U), fn_8024F840, 0U);
diff --git a/src/melee/mn/mninfo.c b/src/melee/mn/mninfo.c
index c1a0008066..ef11a76490 100644
--- a/src/melee/mn/mninfo.c
+++ b/src/melee/mn/mninfo.c
@@ -2,8 +2,8 @@
 
 #include "placeholder.h"
 
-#include "gm/gmmain_lib.h"
 #include "gm/gm_1601.h"
+#include "gm/gmmain_lib.h"
 #include "gm/types.h"
 #include "if/ifprize.h"
 #include "lb/lb_00F9.h"
@@ -515,8 +515,7 @@ s32 mnInfo_80252758(void)
     mnInfo_80252720(temp_r3_3);
     GObj_InitUserData(temp_r3_2, 0, HSD_Free, temp_r3_3);
 
-    temp_r3_4 =
-        HSD_GObj_SetupProc(temp_r3_2, (HSD_GObjEvent) fn_80252548, 0);
+    temp_r3_4 = HSD_GObj_SetupProc(temp_r3_2, (HSD_GObjEvent) fn_80252548, 0);
     temp_r3_4->flags_3 = HSD_GObj_804D783C;
 
     temp_r29 = temp_r3_2->user_data;
diff --git a/src/melee/vi/vi1101.c b/src/melee/vi/vi1101.c
index 034d6c6277..ac84607c95 100644
--- a/src/melee/vi/vi1101.c
+++ b/src/melee/vi/vi1101.c
@@ -3,6 +3,7 @@
 #include "vi.h"
 
 #include "cm/camera.h"
+#include "dolphin/gx/GXStruct.h"
 #include "ef/efasync.h"
 #include "ef/eflib.h"
 #include "ft/ftdemo.h"
@@ -13,42 +14,53 @@
 #include "gr/stage.h"
 #include "it/item.h"
 #include "lb/lb_00F9.h"
+#include "lb/lbarchive.h"
 #include "lb/lbaudio_ax.h"
 #include "lb/lbshadow.h"
 #include "mp/mpcoll.h"
 #include "pl/player.h"
 #include "sc/types.h"
-#include "vi/types.h"
 
-#include <baselib/aobj.h>
-#include <baselib/cobj.h>
-#include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjobject.h>
 #include <baselib/gobjproc.h>
-#include <baselib/jobj.h>
+u8 un_804D6FD8[8];
+GXColor un_804D5B08 = { 0, 0, 0, 0xff };
+
+typedef struct Vi1101Data {
+    Vec3 spawns[3];
+    char vi1101_dat[0xC];
+    char visual1101_scene[0x10];
+    char visual1101_cam2_scene[0x14];
+} Vi1101Data;
+
+static f32 const un_804DE0D8 = 1.0f;
+static f32 const un_804DE0DC = 0.0f;
+static f32 const un_804DE0E0 = 432.0f;
+
+static Vi1101Data un_80400200 = {
+    { { 0.0f, 0.0f, 0.0f },
+      { 15.8512f, 67.0f, 0.0f },
+      { -15.7657f, 67.0f, 0.0f } },
+    "Vi1101.dat",
+    "visual1101Scene",
+    "visual1101Cam2Scene",
+};
 
 static SceneDesc* un_804D6FC0;
 static SceneDesc* un_804D6FC4;
 static HSD_Archive* un_804D6FC8;
 static HSD_Archive* un_804D6FCC;
-
-static Vec3 un_80400200[3] = {
-    { 0.0f, 0.0f, 0.0f },
-    { 15.8512f, 67.0f, 0.0f },
-    { -15.7657f, 67.0f, 0.0f },
-};
-
 static struct {
-    HSD_GObj* p1;
-    HSD_GObj* p2;
+    HSD_GObj* unk0;
+    HSD_GObj* unk4;
 } un_804D6FD0;
 
-void un_8031F294(int char_index, int costume_index)
+void un_8031F294(s32 arg0, s32 arg1)
 {
     HSD_JObj* jobj;
-    Vec3* pmtx;
-    char pad[40];
+    VecMtxPtr pmtx;
+    char pad[16];
 
     Camera_80028B9C(6);
     lb_8000FCDC();
@@ -61,55 +73,49 @@ void un_8031F294(int char_index, int costume_index)
     Stage_8022532C(KINOKOROUTE, 0);
     ftDemo_ObjAllocInit();
     Player_InitAllPlayers();
-
-    Player_80036E20(char_index, un_804D6FC8, 3);
-    Player_SetPlayerCharacter(0, char_index);
-    Player_SetCostumeId(0, costume_index);
+    Player_80036E20(arg0, un_804D6FC8, 3);
+    Player_SetPlayerCharacter(0, arg0);
+    Player_SetCostumeId(0, arg1);
     Player_SetPlayerId(0, 0);
     Player_SetSlottype(0, Gm_PKind_Demo);
-    Player_SetFacingDirection(0, 1.0f);
-    Player_80032768(0, un_80400200);
+    Player_SetFacingDirection(0, un_804DE0D8);
+    Player_80032768(0, un_80400200.spawns);
     Player_80036F34(0, 8);
-
     pmtx = grLib_801C9A10();
-
     Player_80036E20(CKIND_MARIO, un_804D6FCC, 5);
     Player_SetPlayerCharacter(1, CKIND_MARIO);
     Player_SetCostumeId(1, 0);
     Player_SetPlayerId(1, 0);
     Player_SetSlottype(1, Gm_PKind_Demo);
-    Player_SetFacingDirection(1, 0.0f);
-    Player_80032768(1, un_80400200);
+    Player_SetFacingDirection(1, un_804DE0DC);
+    Player_80032768(1, un_80400200.spawns);
     Player_SetFlagsBit5(1, 1);
     Player_80036F34(1, 0xA);
-    un_804D6FD0.p1 = Player_GetEntity(1);
-    jobj = un_804D6FD0.p1->hsd_obj;
-    HSD_JObjReqAnimAll(jobj, 432.0f);
+    un_804D6FD0.unk0 = Player_GetEntity(1);
+    jobj = GET_JOBJ(un_804D6FD0.unk0);
+    HSD_JObjReqAnimAll(jobj, un_804DE0E0);
     HSD_JObjAnimAll(jobj);
-    pmtx[1] = un_80400200[1];
-    HSD_JObjReqAnimAll(jobj, 0.0f);
-
+    pmtx[1] = un_80400200.spawns[1];
+    HSD_JObjReqAnimAll(jobj, un_804DE0DC);
     if (gm_80164840(7) != 0) {
         Player_80036E20(CKIND_LUIGI, un_804D6FCC, 5);
         Player_SetPlayerCharacter(2, CKIND_LUIGI);
         Player_SetCostumeId(2, 0);
         Player_SetPlayerId(2, 0);
         Player_SetSlottype(2, Gm_PKind_Demo);
-        Player_SetFacingDirection(2, 0.0f);
-        Player_80032768(2, un_80400200);
+        Player_SetFacingDirection(2, un_804DE0DC);
+        Player_80032768(2, un_80400200.spawns);
         Player_SetFlagsBit5(2, 1);
         Player_80036F34(2, 0xA);
-        un_804D6FD0.p2 = Player_GetEntity(2);
-        jobj = un_804D6FD0.p2->hsd_obj;
-        HSD_JObjReqAnimAll(jobj, 432.0f);
+        un_804D6FD0.unk4 = Player_GetEntity(2);
+        jobj = GET_JOBJ(un_804D6FD0.unk4);
+        HSD_JObjReqAnimAll(jobj, un_804DE0E0);
         HSD_JObjAnimAll(jobj);
-        pmtx[2] = un_80400200[2];
-        HSD_JObjReqAnimAll(jobj, 0.0f);
+        pmtx[2] = un_80400200.spawns[2];
+        HSD_JObjReqAnimAll(jobj, un_804DE0DC);
     }
-
     lbAudioAx_80026F2C(0x1C);
-    lbAudioAx_8002702C(0xC,
-                       lbAudioAx_80026E84(char_index) | 0x0020000000000000ULL);
+    lbAudioAx_8002702C(0xC, ((u64) 0x20 << 48) | lbAudioAx_80026E84(arg0));
     lbAudioAx_80027168();
     lbAudioAx_80027648();
 }
@@ -119,7 +125,7 @@ void fn_8031F548(HSD_GObj* gobj)
     HSD_JObjAnimAll(GET_JOBJ(gobj));
 }
 
-static void fn_8031F56C(HSD_GObj* gobj)
+static void fn_8031F56C(HSD_GObj* gobj, int unused)
 {
     HSD_CObj* cobj;
     char pad[8];
@@ -172,7 +178,70 @@ void fn_8031F600(HSD_GObj* gobj)
     }
 }
 
-/// #un_8031F714_OnEnter
+void un_8031F714_OnEnter(void* arg)
+{
+    HSD_Joint* new_var2;
+    SceneDesc* var_r28;
+    HSD_CObj* temp_r3;
+    HSD_GObj* temp_r28;
+    HSD_GObj* temp_r31;
+    HSD_JObj* temp_r3_2;
+    HSD_GObj* temp_r26;
+    u8 new_var3;
+    s32 var_r31;
+    u8 temp_r29;
+    u8* input = arg;
+    char* data = (char*) &un_80400200;
+    HSD_JObj* new_var;
+    lbAudioAx_800236DC();
+    efLib_Init();
+    temp_r29 = 0;
+    efAsync_LoadSync(temp_r29);
+    lbAudioAx_80023F28(0x55);
+    lbAudioAx_80024E50(1);
+
+    temp_r29 = input[0];
+    un_804D6FCC = lbArchive_LoadSymbols(data + 0x24, &un_804D6FC0, data + 0x30,
+                                        &un_804D6FC4, data + 0x40, NULL);
+    un_804D6FC8 = lbArchive_LoadSymbols(viGetCharAnimByIndex(temp_r29), NULL);
+
+    temp_r28 = GObj_Create(0xB, 3, 0);
+    HSD_GObjObject_80390A70(temp_r28, HSD_GObj_804D784A & 0xFFFF,
+                            lb_80011AC4(un_804D6FC0->lights));
+    GObj_SetupGXLink(temp_r28, HSD_GObj_LObjCallback, 0, 0);
+
+    if (gm_80164840(7) != 0) {
+        var_r28 = un_804D6FC0;
+    } else {
+        var_r28 = un_804D6FC4;
+    }
+
+    temp_r31 = GObj_Create(0x13, 0x14, 0);
+    temp_r3 = lb_80013B14((HSD_CameraDescPerspective*) var_r28->cameras->desc);
+    HSD_GObjObject_80390A70(temp_r31, HSD_GObj_804D784B, temp_r3);
+    GObj_SetupGXLinkMax(temp_r31, fn_8031F56C, 5);
+    HSD_CObjAddAnim(temp_r3, var_r28->cameras->anims[0]);
+    HSD_CObjReqAnim(temp_r3, un_804DE0DC);
+    HSD_CObjAnim(temp_r3);
+    HSD_GObj_SetupProc(temp_r31, fn_8031F600, 0);
+
+    for (var_r31 = 0; un_804D6FC0->models[var_r31] != NULL; var_r31++) {
+        temp_r26 = GObj_Create(0xE, 0xF, 0);
+        temp_r3_2 =
+            HSD_JObjLoadJoint(new_var2 = un_804D6FC0->models[var_r31]->joint);
+        HSD_GObjObject_80390A70(temp_r26, new_var3 = HSD_GObj_804D7849,
+                                temp_r3_2);
+        GObj_SetupGXLink(temp_r26, HSD_GObj_JObjCallback, 9, 0);
+        gm_8016895C(temp_r3_2, un_804D6FC0->models[var_r31], 0);
+        HSD_JObjReqAnimAll(temp_r3_2, un_804DE0DC);
+        new_var = temp_r3_2;
+        HSD_JObjAnimAll(new_var);
+        HSD_GObj_SetupProc(temp_r26, fn_8031F548, 0x17);
+    }
+
+    un_8031F294(input[0], input[1]);
+    lbAudioAx_80024E50(0);
+}
 
 void un_8031F960_OnFrame(void)
 {
diff --git a/src/melee/vi/vi1101.h b/src/melee/vi/vi1101.h
index e002db4415..de3ca6894a 100644
--- a/src/melee/vi/vi1101.h
+++ b/src/melee/vi/vi1101.h
@@ -1,17 +1,12 @@
 #ifndef GALE01_31F294
 #define GALE01_31F294
 
-#include <placeholder.h>
-
 #include <baselib/forward.h>
 
-/* 4D5B08 */ extern u8 un_804D5B08;
-
-/* 31F294 */ void un_8031F294(int char_index, int costume_index);
+/* 31F294 */ void un_8031F294(s32 char_index, s32 costume_index);
 /* 31F548 */ void fn_8031F548(HSD_GObj* gobj);
 /* 31F600 */ void fn_8031F600(HSD_GObj* gobj);
-/* 31F714 */ void un_8031F714_OnEnter(UNK_T);
+/* 31F714 */ void un_8031F714_OnEnter(void*);
 /* 31F960 */ void un_8031F960_OnFrame(void);
-/* 4D6FD8 */ extern u8 un_804D6FD8[8];
-
+extern u8 un_804D6FD8[8];
 #endif
diff --git a/src/sysdolphin/baselib/cobj.c b/src/sysdolphin/baselib/cobj.c
index 41a8262465..c38c429dfb 100644
--- a/src/sysdolphin/baselib/cobj.c
+++ b/src/sysdolphin/baselib/cobj.c
@@ -18,6 +18,7 @@
 #include <dolphin/mtx.h>
 #include <dolphin/vi.h>
 #include <MetroTRK/intrinsics.h>
+#include <MSL/math_ppc.h>
 #include <MSL/trigf.h>
 
 static HSD_ClassInfo* default_class;
diff --git a/src/sysdolphin/baselib/displayfunc.c b/src/sysdolphin/baselib/displayfunc.c
index cfbd58b35e..e8f7944af6 100644
--- a/src/sysdolphin/baselib/displayfunc.c
+++ b/src/sysdolphin/baselib/displayfunc.c
@@ -13,6 +13,7 @@
 
 #include <dolphin/gx.h>
 #include <dolphin/mtx.h>
+#include <MSL/math_ppc.h>
 #include <Runtime/__mem.h>
 
 #define FLT_EPSILON 1.00000001335e-10F
@@ -89,6 +90,12 @@ Vec3 zOne = { 0, 0, 1 };
 Vec3 yOne = { 0, 1, 0 };
 Vec3 zOne2 = { 0, 0, 1 };
 
+inline f32 HSD_MtxColMagFloat(MtxPtr mtx, int col)
+{
+    return sqrtf((mtx[0][col] * mtx[0][col]) + (mtx[1][col] * mtx[1][col]) +
+                 (mtx[2][col] * mtx[2][col]));
+}
+
 static void mkVBillBoardMtx(HSD_JObj* jobj, MtxPtr src, MtxPtr dst)
 {
     Vec3 pos, ax, ay, az;
diff --git a/src/sysdolphin/baselib/jobj.c b/src/sysdolphin/baselib/jobj.c
index 3b2a4b0cea..57ddc40c78 100644
--- a/src/sysdolphin/baselib/jobj.c
+++ b/src/sysdolphin/baselib/jobj.c
@@ -19,6 +19,7 @@
 #include <trigf.h>
 #include <dolphin/mtx.h>
 #include <dolphin/os.h>
+#include <MSL/math_ppc.h>
 
 void JObjInfoInit(void);
 HSD_JObjInfo hsdJObj = { JObjInfoInit };
diff --git a/src/sysdolphin/baselib/mtx.c b/src/sysdolphin/baselib/mtx.c
index 11702cc6ce..6b91d64ace 100644
--- a/src/sysdolphin/baselib/mtx.c
+++ b/src/sysdolphin/baselib/mtx.c
@@ -3,6 +3,7 @@
 #include "debug.h"
 #include "math.h"
 
+#include <MSL/math_ppc.h>
 #include <MSL/trigf.h>
 
 #define EPSILON 0.0000000001f
diff --git a/src/sysdolphin/baselib/mtx.h b/src/sysdolphin/baselib/mtx.h
index 8d73fa553f..3fe9dcb7fb 100644
--- a/src/sysdolphin/baselib/mtx.h
+++ b/src/sysdolphin/baselib/mtx.h
@@ -7,7 +7,6 @@
 
 #include <dolphin/mtx.h>
 #include <MSL/math.h>
-#include <MSL/math_ppc.h>
 
 #define VEC2_SQ_LEN(v) ((SQ((v).x) + SQ((v).y)))
 #define VEC3_SQ_LEN(v) ((SQ((v).x) + SQ((v).y) + SQ((v).z)))
@@ -63,10 +62,4 @@ static inline f32 HSD_MtxColMag(MtxPtr mtx, int col)
                      (mtx[2][col] * mtx[2][col]));
 }
 
-static inline f32 HSD_MtxColMagFloat(MtxPtr mtx, int col)
-{
-    return sqrtf((mtx[0][col] * mtx[0][col]) + (mtx[1][col] * mtx[1][col]) +
-                 (mtx[2][col] * mtx[2][col]));
-}
-
 #endif
diff --git a/src/sysdolphin/baselib/particle.c b/src/sysdolphin/baselib/particle.c
index eed37ab21b..0bbfb7b7ef 100644
--- a/src/sysdolphin/baselib/particle.c
+++ b/src/sysdolphin/baselib/particle.c
@@ -23,6 +23,7 @@
 #include <baselib/state.h>
 #include <baselib/video.h>
 #include <MetroTRK/ppc_reg.h>
+#include <MSL/math_ppc.h>
 #include <MSL/trigf.h>
 
 typedef struct {
diff --git a/src/sysdolphin/baselib/robj.c b/src/sysdolphin/baselib/robj.c
index a2c21fa4c6..29bc865134 100644
--- a/src/sysdolphin/baselib/robj.c
+++ b/src/sysdolphin/baselib/robj.c
@@ -19,6 +19,7 @@
 #include <math_ppc.h>
 #include <dolphin/mtx.h>
 #include <dolphin/os.h>
+#include <MSL/math_ppc.h>
 
 HSD_ObjAllocData robj_alloc_data;   // robj_alloc_data
 HSD_ObjAllocData rvalue_alloc_data; // rvalue_alloc_data
@@ -343,6 +344,12 @@ static inline HSD_RObj* inlined_HSD_RObjGetByType(HSD_RObj* robj, u32 type,
     return NULL;
 }
 
+inline f32 HSD_MtxColMagFloat(MtxPtr mtx, int col)
+{
+    return sqrtf((mtx[0][col] * mtx[0][col]) + (mtx[1][col] * mtx[1][col]) +
+                 (mtx[2][col] * mtx[2][col]));
+}
+
 static void resolveCnsOrientation(HSD_RObj* robj, void* obj,
                                   void (*update_func)(void*, int,
                                                       HSD_ObjData*))
