diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 0b9f4b52a2..f61ecf1a60 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -34218,18 +34218,18 @@ HSD_CObj_804DE4C8 = .sdata2:0x804DE4C8; // type:object size:0x8 scope:global ali
 @361 = .sdata2:0x804DE608; // type:object size:0x8 scope:local align:8 data:double
 @362 = .sdata2:0x804DE610; // type:object size:0x8 scope:local align:8 data:double
 @384 = .sdata2:0x804DE618; // type:object size:0x8 scope:local align:8 data:double
-HSD_Spline_804DE620 = .sdata2:0x804DE620; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE624 = .sdata2:0x804DE624; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE628 = .sdata2:0x804DE628; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE62C = .sdata2:0x804DE62C; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE630 = .sdata2:0x804DE630; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE634 = .sdata2:0x804DE634; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE638 = .sdata2:0x804DE638; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE640 = .sdata2:0x804DE640; // type:object size:0x8 scope:global align:8 data:double
-HSD_Spline_804DE648 = .sdata2:0x804DE648; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE64C = .sdata2:0x804DE64C; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE650 = .sdata2:0x804DE650; // type:object size:0x4 scope:global align:4 data:float
-HSD_Spline_804DE654 = .sdata2:0x804DE654; // type:object size:0x4 scope:global align:4 data:float
+@13 = .sdata2:0x804DE620; // type:object size:0x4 scope:global align:4 data:float
+@14 = .sdata2:0x804DE624; // type:object size:0x4 scope:global align:4 data:float
+@15 = .sdata2:0x804DE628; // type:object size:0x4 scope:global align:4 data:float
+@17 = .sdata2:0x804DE62C; // type:object size:0x4 scope:global align:4 data:float
+@18 = .sdata2:0x804DE630; // type:object size:0x4 scope:global align:4 data:float
+@19 = .sdata2:0x804DE634; // type:object size:0x4 scope:global align:4 data:float
+@80 = .sdata2:0x804DE638; // type:object size:0x4 scope:global align:4 data:float
+@82 = .sdata2:0x804DE640; // type:object size:0x8 scope:global align:8 data:double
+@91 = .sdata2:0x804DE648; // type:object size:0x4 scope:global align:4 data:float
+@101 = .sdata2:0x804DE64C; // type:object size:0x4 scope:global align:4 data:float
+@178 = .sdata2:0x804DE650; // type:object size:0x4 scope:global align:4 data:float
+@179 = .sdata2:0x804DE654; // type:object size:0x4 scope:global align:4 data:float
 @104 = .sdata2:0x804DE658; // type:object size:0x4 scope:local align:4 data:float
 @105 = .sdata2:0x804DE65C; // type:object size:0x4 scope:local align:4 data:float
 @171 = .sdata2:0x804DE660; // type:object size:0x4 scope:local align:4 data:float
diff --git a/configure.py b/configure.py
index 26c70be7ed..08628c4620 100755
--- a/configure.py
+++ b/configure.py
@@ -1603,7 +1603,7 @@ def MatchingFor(*versions):
             Object(Matching, "sysdolphin/baselib/video.c"),
             Object(Matching, "sysdolphin/baselib/controller.c"),
             Object(Matching, "sysdolphin/baselib/rumble.c"),
-            Object(NonMatching, "sysdolphin/baselib/spline.c"),
+            Object(Matching, "sysdolphin/baselib/spline.c"),
             Object(Matching, "sysdolphin/baselib/mtx.c"),
             Object(Matching, "sysdolphin/baselib/util.c"),
             Object(Matching, "sysdolphin/baselib/objalloc.c"),
diff --git a/src/sysdolphin/baselib/spline.c b/src/sysdolphin/baselib/spline.c
index 27db6cbd84..4d76dba8eb 100644
--- a/src/sysdolphin/baselib/spline.c
+++ b/src/sysdolphin/baselib/spline.c
@@ -28,7 +28,7 @@ f32 splGetHelmite(f32 fterm, f32 time, f32 p0, f32 p1, f32 d0, f32 d1)
                                      (p1 * (-_2t3_T3 + _3t2_T2))));
 }
 
-void splGetCardinalPoint(Vec3* p, Vec3* cp, f32 tension, f32 u)
+inline void splGetCardinalPoint(Vec3* p, Vec3* cp, f32 tension, f32 u)
 {
     f32 u2 = u * u;
     f32 u3 = u2 * u;
@@ -46,7 +46,7 @@ void splGetCardinalPoint(Vec3* p, Vec3* cp, f32 tension, f32 u)
            (cp[3].z * car3);
 }
 
-void splGetBSplinePoint(Vec3* p, Vec3* cp, f32 u)
+static void splGetBSplinePoint(Vec3* p, Vec3* cp, f32 u)
 {
     f32 u2 = u * u;
     f32 u3 = u2 * u;
@@ -62,7 +62,7 @@ void splGetBSplinePoint(Vec3* p, Vec3* cp, f32 u)
     p->z = (cp[0].z * b0) + (cp[1].z * b1) + (cp[2].z * b2) + (cp[3].z * b3);
 }
 
-void splGetBezierPoint(Vec3* p, Vec3* cp, f32 u)
+inline void splGetBezierPoint(Vec3* p, Vec3* cp, f32 u)
 {
     f32 u_1 = 1.0F - u;
     f32 u2 = u * u;
@@ -135,7 +135,7 @@ void splGetSplinePoint(Vec3* p, HSD_Spline* spline, f32 u)
     }
 }
 
-inline f32 splArcLengthPolynomial(const f32 coeffs[5], f32 t)
+static f32 splArcLengthPolynomial(const f32 coeffs[5], f32 t)
 {
     f32 t2 = t * t;
     f32 t3 = t2 * t;
@@ -170,12 +170,17 @@ inline f32 spl_IterateSimpsonsMiddle(const f32 coeffs[5], const f32 dx, f32 t)
     return var_f24;
 }
 
+static f32 spl_GetArcLengthDx(f32 start, f32 midpoint)
+{
+    return (midpoint - start) / 8.0F;
+}
+
 f32 splArcLengthGetParameter(HSD_Spline* spl, f32 arg1)
 {
     s32 idx = 0;
-    f32 result;
     f32 start = 0.0F;
     f32 end = 1.0F;
+    f32 result;
 
     if (arg1 <= 0.0F) {
         return start;
@@ -203,20 +208,21 @@ f32 splArcLengthGetParameter(HSD_Spline* spl, f32 arg1)
 
         while (ABS(start - end) >= 0.00001F) {
             const f32* coeffs = spl_GetCoeffs(spl, idx);
-            const f32 midpoint = (start + end) / 2.0F;
-            const f32 dx = (midpoint - start) / 8.0F;
-            f32 middle = spl_IterateSimpsonsMiddle(coeffs, dx, start + dx);
+            f32 dx;
+            f32 middle;
             f32 simpsons;
 
-            result = midpoint;
+            result = (start + end) / 2.0F;
+            dx = spl_GetArcLengthDx(start, result);
+            middle = spl_IterateSimpsonsMiddle(coeffs, dx, start + dx);
             simpsons = dx *
                        (middle + splArcLengthPolynomial(coeffs, start) +
-                        splArcLengthPolynomial(coeffs, midpoint)) /
+                        splArcLengthPolynomial(coeffs, result)) /
                        3.0F;
             if (var_f22 < (0.00001F + simpsons)) {
-                end = midpoint;
+                end = result;
             } else {
-                start = midpoint;
+                start = result;
                 var_f22 -= simpsons;
             }
         }
diff --git a/src/sysdolphin/baselib/spline.h b/src/sysdolphin/baselib/spline.h
index d7fc1d8006..4cfb9c9d11 100644
--- a/src/sysdolphin/baselib/spline.h
+++ b/src/sysdolphin/baselib/spline.h
@@ -17,7 +17,6 @@ struct HSD_Spline {
 
 f32 splGetHelmite(f32, f32, f32, f32, f32, f32);
 void splGetCardinalPoint(Vec3*, Vec3*, f32, f32);
-void splGetBSplinePoint(Vec3*, Vec3*, f32);
 void splGetBezierPoint(Vec3*, Vec3*, f32);
 void splGetSplinePoint(Vec3*, HSD_Spline*, f32);
 f32 splArcLengthGetParameter(HSD_Spline*, f32);
