diff --git a/src/melee/vi/vi1201v2.c b/src/melee/vi/vi1201v2.c
index afe241d0de..8b5a8acc25 100644
--- a/src/melee/vi/vi1201v2.c
+++ b/src/melee/vi/vi1201v2.c
@@ -2,20 +2,29 @@
 
 #include "vi.h"
 
+#include "baselib/forward.h"
+
 #include "cm/camera.h"
+#include "ef/efasync.h"
+#include "ef/eflib.h"
+#include "ft/ft_0C31.h"
 #include "ft/ftdemo.h"
 #include "gm/gm_1601.h"
 #include "gm/gm_unsplit.h"
 #include "gr/ground.h"
 #include "gr/stage.h"
 #include "it/item.h"
+#include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
+#include "lb/lbarchive.h"
 #include "lb/lbaudio_ax.h"
 #include "lb/lbshadow.h"
 #include "mn/mnmain.h"
 #include "mp/mpcoll.h"
 #include "pl/player.h"
 #include "sc/types.h"
+#include "ty/toy.h"
+#include "ty/tylist.h"
 
 #include "vi/forward.h"
 
@@ -29,13 +38,39 @@
 #include <baselib/gobjproc.h>
 #include <baselib/jobj.h>
 
+typedef struct un_80400304_t {
+    Vec3 player_spawn;
+    char use_quaternion_assert[0x28];
+    char vi1201v2_dat[0x0C];
+    char visual1201v2_scene[0x14];
+    char tykoopa_dat[0x0C];
+    char toykoopa_model_topn_joint[0x1C];
+    char gmrgstnd_dat[0x10];
+    char stand_scene[0x0C];
+} un_80400304_t;
+
 Vec3 un_804002F8 = { 0.0f, 0.0f, 0.0f };
-Vec3 un_80400304 = { 0.0f, 0.0f, 0.0f };
+un_80400304_t un_80400304 = {
+    { 0.0f, 0.0f, 0.0f },
+    "!(jobj->flags & JOBJ_USE_QUATERNION)",
+    "Vi1201v2.dat",
+    "visual1201v2Scene",
+    "TyKoopa.dat",
+    "ToyKoopaModel_TopN_joint",
+    "GmRgStnd.dat",
+    "standScene",
+};
 
 static SceneDesc* un_804D7010;
+static SceneDesc* un_804D7014;
 static HSD_Archive* un_804D7018;
 static HSD_Archive* un_804D701C;
+static HSD_Joint* un_804D7020;
 static HSD_JObj* un_804D7024;
+GXColor un_804D7028;
+static s32 un_804D702C;
+HSD_GObj* un_804D7030;
+HSD_GObj* un_804D7034;
 static un_804D7004_t un_804D7038;
 
 void un_803204B0(int arg0, int arg1)
@@ -100,7 +135,7 @@ void un_803205F4(void)
     Player_SetPlayerId(1, 0);
     Player_SetSlottype(1, 2);
     Player_SetFacingDirection(1, -1.0f);
-    Player_80032768(1, &un_80400304);
+    Player_80032768(1, &un_80400304.player_spawn);
     Player_80036F34(1, 0xF);
     Player_SetScale(1, 1.0f / Player_80032BB0(1));
 
@@ -163,12 +198,12 @@ void un_803207C4(void)
 
 void un_803208F0(HSD_GObj* gobj)
 {
-    u8* colors;
+    GXColor* colors;
     char pad[8];
     lbShadow_8000F38C(0);
     if (HSD_CObjSetCurrent(GET_COBJ(gobj)) != 0) {
         colors = &un_804D7028;
-        HSD_SetEraseColor(colors[0], colors[1], colors[2], colors[3]);
+        HSD_SetEraseColor(colors->r, colors->g, colors->b, colors->a);
         HSD_CObjEraseScreen(GET_COBJ(gobj), 1, 0, 1);
         vi_8031CA04(gobj);
         gobj->gxlink_prios = 0x881;
@@ -198,6 +233,169 @@ void fn_80320A1C(HSD_GObj* gobj)
     HSD_FogInterpretAnim(gobj->hsd_obj);
 }
 
+static void HSD_JObjSetRotationY_2(HSD_JObj* jobj, f32 y)
+{
+    ((jobj) ? ((void) 0) : __assert("jobj.h", 660, "jobj"));
+    ((!(jobj->flags & JOBJ_USE_QUATERNION))
+         ? ((void) 0)
+         : __assert("jobj.h", 661, (char*) &un_804002F8 + 0x18));
+    jobj->rotate.y = y;
+    if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+        ftCo_800C6AFC(jobj);
+    }
+}
+
+static void HSD_JObjSetScaleX_2(HSD_JObj* jobj, f32 x)
+{
+    ((jobj) ? ((void) 0) : __assert("jobj.h", 776, "jobj"));
+    jobj->scale.x = x;
+    if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+        ftCo_800C6AFC(jobj);
+    }
+}
+
+static void HSD_JObjSetScaleY_2(HSD_JObj* jobj, f32 y)
+{
+    ((jobj) ? ((void) 0) : __assert("jobj.h", 791, "jobj"));
+    jobj->scale.y = y;
+    if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+        ftCo_800C6AFC(jobj);
+    }
+}
+
+static void HSD_JObjSetScaleZ_2(HSD_JObj* jobj, f32 z)
+{
+    ((jobj) ? ((void) 0) : __assert("jobj.h", 806, "jobj"));
+    jobj->scale.z = z;
+    if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+        ftCo_800C6AFC(jobj);
+    }
+}
+
+static void HSD_JObjSetTranslateX_2(HSD_JObj* jobj, f32 x)
+{
+    ((jobj) ? ((void) 0) : __assert("jobj.h", 932, "jobj"));
+    jobj->translate.x = x;
+    if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+        ftCo_800C6AFC(jobj);
+    }
+}
+
+static void HSD_JObjSetTranslateY_2(HSD_JObj* jobj, f32 y)
+{
+    ((jobj) ? ((void) 0) : __assert("jobj.h", 947, "jobj"));
+    jobj->translate.y = y;
+    if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+        ftCo_800C6AFC(jobj);
+    }
+}
+
+static void HSD_JObjSetTranslateZ_2(HSD_JObj* jobj, f32 z)
+{
+    ((jobj) ? ((void) 0) : __assert("jobj.h", 962, "jobj"));
+    jobj->translate.z = z;
+    if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+        ftCo_800C6AFC(jobj);
+    }
+}
+
+void un_80320A40_OnEnter(void* arg)
+{
+    u8* input = arg;
+    u8 char_index;
+    char* data = (char*) &un_804002F8;
+    HSD_CObj* cobj;
+    HSD_Fog* fog;
+    HSD_GObj* gobj;
+    HSD_JObj* child;
+    HSD_JObj* jobj;
+    f32 scale;
+    PAD_STACK(24);
+
+    efLib_Init();
+    efAsync_LoadSync(0);
+    lbAudioAx_80023F28(0x5C);
+    lbAudioAx_80024E50(1);
+
+    char_index = input[0];
+
+    un_804D701C =
+        lbArchive_LoadSymbols(data + 0x40, &un_804D7010, data + 0x50, NULL);
+    lbArchive_LoadSymbols(data + 0x64, &un_804D7020, data + 0x70, NULL);
+    lbArchive_LoadSymbols(data + 0x8C, &un_804D7014, data + 0x9C, NULL);
+    un_803124BC();
+    un_804D7018 = lbArchive_LoadSymbols(viGetCharAnimByIndex(char_index), NULL);
+
+    gobj = GObj_Create(0x13, 0x14, 0);
+    cobj = lb_80013B14((HSD_CameraDescPerspective*) un_804D7010->cameras->desc);
+    HSD_GObjObject_80390A70(gobj, HSD_GObj_804D784B, cobj);
+    GObj_SetupGXLinkMax(gobj, (void (*)(HSD_GObj*, int)) un_803208F0, 8);
+    HSD_CObjAddAnim(cobj, un_804D7010->cameras->anims[0]);
+    HSD_CObjReqAnim(cobj, 0.0f);
+    HSD_CObjAnim(cobj);
+    HSD_GObj_SetupProc(gobj, un_80320984, 0);
+
+    un_803207C4();
+    un_80320508(input[0], input[1]);
+
+    gobj = GObj_Create(0xE, 0xF, 0);
+    un_804D7030 = gobj;
+    jobj = HSD_JObjLoadJoint(un_804D7020);
+    HSD_GObjObject_80390A70(gobj, HSD_GObj_804D7849, jobj);
+    GObj_SetupGXLink(gobj, HSD_GObj_JObjCallback, 0xB, 0);
+    HSD_JObjSetScaleX(jobj, 0.55f);
+    HSD_JObjSetScaleY(jobj, 0.55f);
+    HSD_JObjSetScaleZ(jobj, 0.55f);
+    lb_8000C1C0(jobj, un_804D7024);
+    lb_8000C290(jobj, un_804D7024);
+    HSD_GObj_SetupProc(gobj, un_803204E4, 0x17);
+
+    gobj = GObj_Create(0xE, 0xF, 0);
+    un_804D7034 = gobj;
+    jobj = HSD_JObjLoadJoint(un_804D7014->models[0]->joint);
+    HSD_GObjObject_80390A70(gobj, HSD_GObj_804D7849, jobj);
+    GObj_SetupGXLink(gobj, HSD_GObj_JObjCallback, 0xB, 0);
+    HSD_GObj_SetupProc(gobj, un_803204C0, 0x17);
+
+    if (jobj == NULL) {
+        child = NULL;
+    } else {
+        child = jobj->child;
+    }
+
+    HSD_JObjSetTranslateX_2(child, -un_803060BC(0x1E, 0));
+    HSD_JObjSetTranslateY_2(child, -un_803060BC(0x1E, 1));
+    HSD_JObjSetTranslateZ_2(child, -un_803060BC(0x1E, 2));
+    HSD_JObjSetRotationY_2(child, -un_803060BC(0x1E, 5));
+
+    scale = 0.55f * (un_803060BC(0x1E, 4) * (1.0f / un_803060BC(0x1E, 3)));
+
+    HSD_JObjSetScaleX_2(child, scale);
+    HSD_JObjSetScaleY_2(child, scale);
+    HSD_JObjSetScaleZ_2(child, scale);
+
+    lb_8000C1C0(jobj, un_804D7024);
+    lb_8000C290(jobj, un_804D7024);
+
+    gobj = GObj_Create(0xB, 3, 0);
+    fog = HSD_FogLoadDesc(un_804D7010->fogs->desc);
+    HSD_GObjObject_80390A70(gobj, HSD_GObj_804D7848, fog);
+    GObj_SetupGXLink(gobj, HSD_GObj_FogCallback, 0, 0);
+    HSD_Fog_8037DE7C(fog, un_804D7010->fogs->anims[0]->aobjdesc);
+    HSD_FogReqAnim(fog, 0.0f);
+    HSD_FogInterpretAnim(fog);
+    HSD_GObj_SetupProc(gobj, fn_80320A1C, 0x17);
+    un_804D7028 = fog->color;
+    un_804D702C = 0;
+
+    gobj = GObj_Create(0xB, 3, 0);
+    HSD_GObjObject_80390A70(gobj, HSD_GObj_804D784A,
+                            lb_80011AC4(un_804D7010->lights));
+    GObj_SetupGXLink(gobj, HSD_GObj_LObjCallback, 0, 0);
+
+    lbAudioAx_80024E50(0);
+}
+
 void un_803210EC_OnFrame(void)
 {
     vi_8031CAAC();
diff --git a/src/melee/vi/vi1201v2.h b/src/melee/vi/vi1201v2.h
index 160583e976..2e9e66629d 100644
--- a/src/melee/vi/vi1201v2.h
+++ b/src/melee/vi/vi1201v2.h
@@ -3,10 +3,12 @@
 
 #include <placeholder.h>
 
+#include "dolphin/gx/GXStruct.h"
+
 #include "ft/forward.h"
 #include <baselib/forward.h>
 
-/* 4D7028 */ extern u8 un_804D7028;
+/* 4D7028 */ extern GXColor un_804D7028;
 /* 4D7030 */ extern HSD_GObj* un_804D7030;
 /* 4D7034 */ extern HSD_GObj* un_804D7034;
 
@@ -20,7 +22,7 @@
 /* 3208F0 */ void un_803208F0(HSD_GObj*);
 /* 320984 */ void un_80320984(HSD_GObj*);
 /* 320A1C */ void fn_80320A1C(HSD_GObj*);
-/* 320A40 */ void un_80320A40_OnEnter(UNK_T);
+/* 320A40 */ void un_80320A40_OnEnter(void*);
 /* 3210EC */ void un_803210EC_OnFrame(void);
 
 #endif
