diff --git a/config/GALE01/splits.txt b/config/GALE01/splits.txt
index 45a023c0f3..f77a9a53d7 100644
--- a/config/GALE01/splits.txt
+++ b/config/GALE01/splits.txt
@@ -2493,15 +2493,14 @@ melee/gr/grpushon.c:
 	.rodata     start:0x803B8440 end:0x803B8470
 	.data       start:0x803E7AC8 end:0x803E7D00
 	.sdata      start:0x804D4930 end:0x804D4950
-	.sbss       start:0x804D6AB8 end:0x804D6AC0
+	.sbss       start:0x804D6AB8 end:0x804D6AC8
 	.sdata2     start:0x804DBAF0 end:0x804DBB40
 
 melee/gr/grfigureget.c:
 	.text       start:0x80219530 end:0x80219C98
 	.rodata     start:0x803B8470 end:0x803B8480
 	.data       start:0x803E7D00 end:0x803E7DA0
-	.sbss       start:0x804D6AC0 end:0x804D6AC8
-	.sdata2     start:0x804DBB40 end:0x804DBB68
+	.sdata2     start:0x804DBB40 end:0x804DBB60
 
 melee/gr/grbattle.c:
 	.text       start:0x80219C98 end:0x8021A620
diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 34227166a9..03618e139c 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -11699,8 +11699,8 @@ grFigureGet_80219B0C = .text:0x80219B0C; // type:function size:0x4 scope:global
 grFigureGet_80219B10 = .text:0x80219B10; // type:function size:0x124
 grFigureGet_80219C34 = .text:0x80219C34; // type:function size:0x1C scope:global
 grFigureGet_80219C50 = .text:0x80219C50; // type:function size:0x38 scope:global
-grFigureGet_OnTouchLine = .text:0x80219C88; // type:function size:0x8 scope:global
-grFigureGet_OnCheckShadowRender = .text:0x80219C90; // type:function size:0x8 scope:global
+grFigureGet_OnTouchLine = .text:0x80219C88; // type:function size:0x8 scope:local
+grFigureGet_OnCheckShadowRender = .text:0x80219C90; // type:function size:0x8 scope:local
 grBattle_OnDemoInit = .text:0x80219C98; // type:function size:0xC scope:local
 grBattle_OnInit = .text:0x80219CA4; // type:function size:0xB0 scope:local
 grBattle_OnLoad = .text:0x80219D54; // type:function size:0x4 scope:local
@@ -20143,7 +20143,7 @@ grPushOn_803B8440 = .rodata:0x803B8440; // type:object size:0xC scope:global dat
 grPushOn_803B844C = .rodata:0x803B844C; // type:object size:0xC scope:global data:4byte
 grPushOn_803B8458 = .rodata:0x803B8458; // type:object size:0xC scope:global data:4byte
 grPushOn_803B8464 = .rodata:0x803B8464; // type:object size:0xC scope:global data:4byte
-grFigureGet_803B8470 = .rodata:0x803B8470; // type:object size:0x10 scope:global data:4byte
+grFigureGet_803B8470 = .rodata:0x803B8470; // type:object size:0xC scope:global data:4byte
 grLast_803B8480 = .rodata:0x803B8480; // type:object size:0xC scope:global data:4byte
 grLast_803B848C = .rodata:0x803B848C; // type:object size:0xC scope:global data:4byte
 grLast_803B8498 = .rodata:0x803B8498; // type:object size:0xC scope:global data:4byte
@@ -22863,6 +22863,7 @@ grPushOn_803E7B74 = .data:0x803E7B74; // type:object size:0x1C scope:global
 grPushOn_803E7B90 = .data:0x803E7B90; // type:object size:0x13C scope:global
 grPushOn_803E7CCC = .data:0x803E7CCC; // type:object size:0x34 scope:global align:4 data:float
 grFigureGet_803E7D00 = .data:0x803E7D00; // type:object size:0x28 scope:local
+...data.0 = .data:0x803E7D00; // type:label scope:local
 @169 = .data:0x803E7D28; // type:object size:0xB scope:local data:string
 grFigureGet_803E7D34 = .data:0x803E7D34; // type:object size:0x34 scope:global
 @189 = .data:0x803E7D68; // type:object size:0x21 scope:local data:string
@@ -32162,13 +32163,13 @@ grPushOn_804DBB24 = .sdata2:0x804DBB24; // type:object size:0x4 scope:global ali
 grPushOn_804DBB28 = .sdata2:0x804DBB28; // type:object size:0x4 scope:global align:4 data:float
 grPushOn_804DBB30 = .sdata2:0x804DBB30; // type:object size:0x8 scope:global align:8 data:double
 grPushOn_804DBB38 = .sdata2:0x804DBB38; // type:object size:0x4 scope:global align:4 data:float
-grFigureGet_804DBB40 = .sdata2:0x804DBB40; // type:object size:0x4 scope:global align:4 data:float
-grFigureGet_804DBB44 = .sdata2:0x804DBB44; // type:object size:0x4 scope:global align:4 data:float
-grFigureGet_804DBB48 = .sdata2:0x804DBB48; // type:object size:0x4 scope:global align:4 data:float
-grFigureGet_804DBB4C = .sdata2:0x804DBB4C; // type:object size:0x4 scope:global align:4 data:float
-grFigureGet_804DBB50 = .sdata2:0x804DBB50; // type:object size:0x4 scope:global align:4 data:float
-grFigureGet_804DBB54 = .sdata2:0x804DBB54; // type:object size:0x4 scope:global align:4 data:float
-grFigureGet_804DBB58 = .sdata2:0x804DBB58; // type:object size:0x8 scope:global align:8 data:double
+@215 = .sdata2:0x804DBB40; // type:object size:0x4 scope:local align:4 data:float
+@216 = .sdata2:0x804DBB44; // type:object size:0x4 scope:local align:4 data:float
+@259 = .sdata2:0x804DBB48; // type:object size:0x4 scope:local align:4 data:float
+@260 = .sdata2:0x804DBB4C; // type:object size:0x4 scope:local align:4 data:float
+@261 = .sdata2:0x804DBB50; // type:object size:0x4 scope:local align:4 data:float
+@262 = .sdata2:0x804DBB54; // type:object size:0x4 scope:local align:4 data:float
+@264 = .sdata2:0x804DBB58; // type:object size:0x8 scope:local align:8 data:double
 grFigureGet_804DBB60 = .sdata2:0x804DBB60; // type:object size:0x4 scope:global align:4 data:float
 grFigureGet_804DBB64 = .sdata2:0x804DBB64; // type:object size:0x4 scope:global align:4 data:float
 @299 = .sdata2:0x804DBB68; // type:object size:0x6 scope:local
diff --git a/configure.py b/configure.py
index b5a71a253e..6664222274 100755
--- a/configure.py
+++ b/configure.py
@@ -1015,7 +1015,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/gr/grgreens.c"),
             Object(NonMatching, "melee/gr/grflatzone.c"),
             Object(NonMatching, "melee/gr/grpushon.c"),
-            Object(NonMatching, "melee/gr/grfigureget.c"),
+            Object(Matching, "melee/gr/grfigureget.c"),
             Object(Matching, "melee/gr/grbattle.c"),
             Object(Matching, "melee/gr/grlast.c"),
             Object(NonMatching, "melee/gr/grhomerun.c"),
diff --git a/src/melee/gr/grfigureget.c b/src/melee/gr/grfigureget.c
index 0cc045e603..8b2da36b3d 100644
--- a/src/melee/gr/grfigureget.c
+++ b/src/melee/gr/grfigureget.c
@@ -6,12 +6,14 @@
 #include "gr/forward.h"
 
 #include "gr/granime.h"
-#include "gr/grdisplay.h"
 #include "gr/ground.h"
+#include "gr/stage.h"
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
 #include "gr/types.h"
 
+#include "ft/ftlib.h"
+#include "gm/gmregclear.h"
 #include "it/forward.h"
 
 #include "it/inlines.h"
@@ -20,9 +22,15 @@
 
 #include "lb/forward.h"
 
+#include "lb/lb_00B0.h"
+#include "lb/lb_00F9.h"
 #include "lb/types.h"
 
 #include "mp/forward.h"
+#include "mp/mplib.h"
+
+#include "ty/tydisplay.h"
+#include "ty/toy.h"
 
 #include <dolphin/mtx.h>
 #include <dolphin/os/OSError.h>
@@ -30,6 +38,17 @@
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
 
+typedef struct grFigureGet_Data {
+    s32 x0;
+    s32 x4;
+    s32 x8;
+    f32 xC;
+    f32 x10;
+    f32 x14;
+} grFigureGet_Data;
+
+static Vec3 const grFigureGet_803B8470 = { 0.0f, 0.0f, 0.0f };
+
 /* 219530 */ static void grFigureGet_OnDemoInit(int);
 /* 219534 */ static void grFigureGet_OnInit(void);
 /* 21959C */ static void grFigureGet_OnLoad(void);
@@ -71,7 +90,9 @@ StageData grFigureGet_803E7D34 = {
     0,
 };
 
-extern UNK_T grFigureGet_804D6AC0;
+extern grFigureGet_Data* grFigureGet_804D6AC0;
+extern f32 grFigureGet_804DBB60;
+extern f32 grFigureGet_804DBB64;
 
 extern StageInfo stage_info;
 
@@ -131,23 +152,152 @@ void grFigureGet_802196E8(Ground_GObj* gobj) {}
 
 void grFigureGet_802196EC(Ground_GObj* gobj) {}
 
-/// #grFigureGet_802196F0
+void grFigureGet_802196F0(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    s32 i0;
+    s32 i1;
+    s32 i2;
+    s32 temp_r3;
+    s32 temp_r3_2;
+    int* var_r30;
+    s32 sp14;
+    s32 var_r29_3;
+    s32 var_r6;
+
+    grAnime_801C8138(gobj, gp->map_id, 0);
+    gp->x8_callback = NULL;
+    gp->xC_callback = NULL;
+    tyDisplay_8031C2CC();
+    i0 = HSD_Randi(3);
+    gp->gv.figureget.x10[i0] = un_80305058(0, 0x63, 0, 65.0f);
+    i1 = i0 + 1;
+    if (i1 >= 3) {
+        i1 = 0;
+    }
+    gp->gv.figureget.x10[i1] = Ground_801C5840();
+    if (un_8031C354(gp->gv.figureget.x10[i1], &sp14, 1, 2) == 0) {
+        do {
+            temp_r3 = un_80305058(2, 0x63, 1, 75.0f);
+            sp14 = temp_r3;
+        } while (gp->gv.figureget.x10[i1] == temp_r3);
+    }
+    i1 = i1 + 1;
+    if (i1 >= 3) {
+        i1 = 0;
+    }
+    gp->gv.figureget.x10[i1] = sp14;
+    for (var_r29_3 = 0; var_r29_3 < 3; var_r29_3++) {
+        un_8031C454(gp->gv.figureget.x10[var_r29_3]);
+    }
+    PAD_STACK(8);
+    temp_r3_2 = grFigureGet_804D6AC0->x4;
+    if (temp_r3_2 != 0) {
+        var_r6 = HSD_Randi(temp_r3_2);
+    } else {
+        var_r6 = 0;
+    }
+    gp->gv.figureget.x0 = grFigureGet_804D6AC0->x0 + var_r6;
+    gp->gv.figureget.x4 = 0;
+    gp->gv.figureget.x8 = 0;
+    gp->gv.figureget.xC = 0;
+    memzero(gp->gv.figureget.x1C, 0xC);
+    memzero(gp->gv.figureget.x28, 0xC);
+    memzero(gp->gv.figureget.x34, 0xC);
+    mpJointSetCb1(0, gp, grFigureGet_80219B10);
+}
 
 bool grFigureGet_80219890(Ground_GObj* gobj)
 {
     return false;
 }
 
-/// #grFigureGet_80219898
+void grFigureGet_80219898(Ground_GObj* gobj)
+{
+    u32 pad2;
+    HSD_GObj* var_r28;
+    Item_GObj* temp_r3_2;
+    Point3d pos;
+    f32 temp_f1;
+    f32 var_f2;
+    s32 temp_r0;
+    s32 temp_r3;
+    s32 temp_r6;
+    s32 var_r30;
+    s32 var_r3;
+    Ground* gp = GET_GROUND(gobj);
+    int* var_r29;
+    u32 pad;
+
+    for (var_r28 = HSD_GObj_Entities->fighters; var_r28 != NULL;
+         var_r28 = var_r28->next)
+    {
+        ftLib_80086984(var_r28)->joint_id_skip = 0;
+    }
+    if (gp->gv.figureget.x4 < grFigureGet_804D6AC0->x8) {
+        temp_r3 = gp->gv.figureget.x0;
+        if (temp_r3 == 0) {
+            pos = grFigureGet_803B8470;
+            gp->gv.figureget.x4 += 1;
+            var_f2 = grFigureGet_804D6AC0->x10;
+            var_f2 = var_f2 - grFigureGet_804D6AC0->xC;
+            if (var_f2 < 0.0f) {
+                var_f2 = -var_f2;
+            }
+            temp_f1 = 0.5f;
+            temp_f1 = var_f2 * temp_f1;
+            temp_f1 -= 20.0f;
+            if ((s32) temp_f1 != 0) {
+                var_r3 = HSD_Randi((s32) temp_f1);
+            } else {
+                var_r3 = 0;
+            }
+            pos.x = 20.0f + (f32) var_r3;
+            var_r30 = 2;
+            if (HSD_Randi(var_r30) == 0) {
+                pos.x *= -1.0f;
+            }
+            pos.y = grFigureGet_804D6AC0->x14 + Stage_GetCamBoundsTopOffset();
+            temp_r6 = gp->gv.figureget.x4;
+            temp_r3_2 = it_802F2094(gobj, &pos,
+                                    gp->gv.figureget.x10[temp_r6 - 1], temp_r6);
+            gp->gv.figureget.x34[gp->gv.figureget.x4 - 1] = temp_r3_2;
+            it_802F2014(temp_r3_2, 0);
+            temp_r3 = grFigureGet_804D6AC0->x4;
+            if (temp_r3 != 0) {
+                var_r3 = HSD_Randi(temp_r3);
+            } else {
+                var_r3 = 0;
+            }
+            gp->gv.figureget.x0 = grFigureGet_804D6AC0->x0 + var_r3;
+        } else {
+            gp->gv.figureget.x0 = temp_r3 - 1;
+        }
+    } else if (grFigureGet_80219C50(gobj) != 0) {
+        for (var_r30 = 0; var_r30 < gp->gv.figureget.x8; var_r30++) {
+            gm_8017E280((u16) gp->gv.figureget.x1C[var_r30], 1U);
+        }
+        if (gp->gv.figureget.x8 != 0) {
+            if (gp->gv.figureget.x8 == 3) {
+                stage_info.flags |= 0x20;
+            }
+            stage_info.flags |= 0x100;
+        } else {
+            stage_info.flags |= 0x200;
+        }
+    }
+    lb_800115F4();
+}
 
 void grFigureGet_80219B0C(Ground_GObj* gobj) {}
 
 void grFigureGet_80219B10(Ground* gp, s32 arg1, CollData* cd, s32 arg3,
                           mpLib_GroundEnum arg4, f32 arg5)
 {
+    ECBFlagStruct* tmp;
     HSD_GObj* item_gobj = cd->x0_gobj;
-    if ((s32) cd->x34_flags.b1234 == 5 && item_gobj &&
-        item_gobj->classifier == HSD_GOBJ_CLASS_ITEM)
+    if (((((s32) (*(tmp = &cd->x34_flags)).b1234) == 5) && item_gobj) &&
+        (item_gobj->classifier == 0x6))
     {
         Item* ip = GET_ITEM(cd->x0_gobj);
         if (ip->kind == It_Kind_Coin && gp->gv.figureget.x8 < 3) {
@@ -161,7 +311,9 @@ void grFigureGet_80219B10(Ground* gp, s32 arg1, CollData* cd, s32 arg3,
                 }
             }
             if (count == 0) {
-                if (cd->cur_pos.x <= 4.5f && cd->cur_pos.x >= -4.5f) {
+                if (cd->cur_pos.x <= grFigureGet_804DBB60 &&
+                    cd->cur_pos.x >= grFigureGet_804DBB64)
+                {
                     it_802F202C(cd->x0_gobj);
                     gp->gv.figureget.x28[gp->gv.figureget.x8] = cd->x0_gobj;
                     gp->gv.figureget.x1C[gp->gv.figureget.x8] = value;
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 2230e38e84..1b0caf7759 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -981,15 +981,14 @@ struct grRCruise_GroundVars2 {
 };
 
 struct grFigureGet_GroundVars {
-    /*  +0 gp+C4 */ UNK_T x0;
-    /*  +4 gp+C8 */ UNK_T x4;
+    /*  +0 gp+C4 */ s32 x0;
+    /*  +4 gp+C8 */ s32 x4;
     /*  +8 gp+CC */ int x8;
     /*  +C gp+D0 */ int xC;
-    /* +10 gp+D4 */ int x10;
-    /* +14 gp+D8 */ int x14;
-    /* +18 gp+DC */ int x18;
+    /* +10 gp+D4 */ int x10[3];
     /* +1C gp+E0 */ int x1C[3];
     /* +28 gp+EC */ HSD_GObj* x28[3];
+    /* +34 gp+F8 */ Item_GObj* x34[3];
 };
 
 struct grFourside_GroundVars {
