diff --git a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
index 308cc6e5af..9f1bb91cdf 100644
--- a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
+++ b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
@@ -1270,7 +1270,7 @@ void ftCh_Damage2_Anim(HSD_GObj* gobj)
             attrs = fp->ft_data->ext_attr;
             fp->mv.ch.unk0.xC.x = fp->cur_pos.x - attrs->x28;
             fp->mv.ch.unk0.xC.y = attrs->x24;
-            Fighter_ChangeMotionState(gobj, ftMh_MS_Damage, 0,
+            Fighter_ChangeMotionState(gobj, ftMh_MS_Damage2, 0,
                                       fp->mv.ch.unk0.xC.z = ftCh_Init_804DA070,
                                       ftCh_Init_804DA074, ftCh_Init_804DA070,
                                       NULL);
diff --git a/src/melee/gr/gricemt.c b/src/melee/gr/gricemt.c
index 95e36e5198..af26940daa 100644
--- a/src/melee/gr/gricemt.c
+++ b/src/melee/gr/gricemt.c
@@ -576,15 +576,20 @@ void grIceMt_801F7EE8(Ground_GObj* arg0)
 }
 
 /// #grIceMt_801F7F1C
-void grIceMt_801F7F1C(Ground_GObj* arg0)
+void grIceMt_801F7F1C(HSD_GObj* gobj)
 {
-    int i = 0;
-    Ground* gp = GET_GROUND(arg0);
+    u8 _[8];
+    void* ptr;
+    s32 i;
+
+    i = 0;
+    ptr = gobj->user_data;
     do {
-        if (42 != 0) {
-            grMaterial_801C8CDC(arg0);
+        if (*(void**)((u8*)ptr + 0xF8) != NULL) {
+            grMaterial_801C8CDC(*(void**)((u8*)ptr + 0xF8));
         }
-        i++;
+        i += 1;
+        ptr = (void*)((u8*)ptr + 4);
     } while (i < 2);
 }
 
@@ -631,13 +636,19 @@ void grIceMt_801F815C(Ground_GObj* param1)
 /// #grIceMt_801F81B4
 void grIceMt_801F81B4(Ground_GObj* arg0)
 {
-    int i = 0;
-    Ground* gp = GET_GROUND(arg0);
+    Ground* gp;
+    int i;
+    PAD_STACK(8);
+
+    i = 0;
+    gp = arg0->user_data;
+
     do {
-        if (42 != 0) {
-            grMaterial_801C8CDC(arg0);
+        if (*(HSD_GObj**)((u8*)gp + 0xF8) != NULL) {
+            grMaterial_801C8CDC(*(HSD_GObj**)((u8*)gp + 0xF8));
         }
         i++;
+        gp = (Ground*)((u8*)gp + 4);
     } while (i < 2);
 }
 
@@ -1007,18 +1018,47 @@ void grIceMt_801F91EC(HSD_GObj* param_1, s16* param_2, int param_3,
 /// #grIceMt_801F929C
 void grIceMt_801F929C(HSD_GObj* arg0, void* arg1)
 {
-    mpLib_80057BC0(2);
-    mpJointListAdd(2);
-    grAnime_801C83D0(arg0, 2, 7);
+    s16 temp_r3;
+
+    if (*(s16*)((u8*)arg1 + 0x0) != 0) {
+        *(s16*)((u8*)arg1 + 0x2) = (s16) (*(s16*)((u8*)arg1 + 0x2) + 1);
+        temp_r3 = *(s16*)((u8*)arg1 + 0x2);
+        if (temp_r3 == *(s16*)((u8*)arg1 + 0xA)) {
+            mpLib_80057BC0(*(s16*)((u8*)arg1 + 0x8));
+        } else if (temp_r3 == *(s16*)((u8*)arg1 + 0xC)) {
+            mpJointListAdd(*(s16*)((u8*)arg1 + 0x8));
+        }
+        if (grAnime_801C83D0(arg0, *(s16*)((u8*)arg1 + 0x4), 7) != 0) {
+            *(s16*)((u8*)arg1 + 0x0) = 0;
+        }
+    }
 }
 
 /// #fn_801F9338
 void fn_801F9338(Ground* gp, int arg1, CollData* arg2, s32 arg3,
                  mpLib_GroundEnum arg4, float arg8)
 {
-    // mpLib_80057BC0(2);
-    // mpJointListAdd(2);
-    // grAnime_801C83D0(arg0,2,7);
+    HSD_GObj* temp_r3;
+    s16* temp_r30;
+    s16 temp_r4;
+    u8 flag_byte;
+
+    temp_r30 = (s16*)((u8*)gp + 0x100);
+    flag_byte = *(u8*)((u8*)arg2 + 0x34);
+    if (((flag_byte >> 3) & 0xF) == 1 && *temp_r30 == 0) {
+        temp_r3 = Ground_801C2BA4(2);
+        *temp_r30 = 1;
+        temp_r30[1] = 0;
+        grAnime_801C7A04(temp_r3, temp_r30[2], 7, 1.0f);
+        grAnime_801C7B24(temp_r3, temp_r30[2], 7, 0.0f);
+        grAnime_801C78FC(temp_r3, temp_r30[2], 7);
+        temp_r4 = temp_r30[3];
+        if (temp_r4 != -1) {
+            grAnime_801C7A04(temp_r3, temp_r4, 7, 1.0f);
+            grAnime_801C7B24(temp_r3, temp_r30[3], 7, 0.0f);
+            grAnime_801C78FC(temp_r3, temp_r30[3], 7);
+        }
+    }
     grIceMt_801FA7F0(gp, arg1, arg2, arg3, arg4, arg8);
 }
 
@@ -1026,9 +1066,25 @@ void fn_801F9338(Ground* gp, int arg1, CollData* arg2, s32 arg3,
 void fn_801F9448(Ground* gp, int arg1, CollData* arg2, s32 arg3,
                  mpLib_GroundEnum arg4, float arg8)
 {
-    // mpLib_80057BC0(2);
-    // mpJointListAdd(2);
-    // grAnime_801C83D0(arg0,2,7);
+    HSD_GObj* gobj;
+    s16* icemt_ptr = &gp->gv.icemt.x10E;
+    s16 temp_r4;
+    u8 flag_byte = *(u8*)((u8*)arg2 + 0x34);
+
+    if (((flag_byte >> 3) & 0xF) == 1 && icemt_ptr[0] == 0) {
+        gobj = Ground_801C2BA4(2);
+        icemt_ptr[0] = 1;
+        icemt_ptr[1] = 0;
+        grAnime_801C7A04(gobj, icemt_ptr[2], 7, 1.0f);
+        grAnime_801C7B24(gobj, icemt_ptr[2], 7, 0.0f);
+        grAnime_801C78FC(gobj, icemt_ptr[2], 7);
+        temp_r4 = icemt_ptr[3];
+        if (temp_r4 != -1) {
+            grAnime_801C7A04(gobj, temp_r4, 7, 1.0f);
+            grAnime_801C7B24(gobj, icemt_ptr[3], 7, 0.0f);
+            grAnime_801C78FC(gobj, icemt_ptr[3], 7);
+        }
+    }
     grIceMt_801FA7F0(gp, arg1, arg2, arg3, arg4, arg8);
 }
 
diff --git a/src/melee/gr/grpura.c b/src/melee/gr/grpura.c
index 78cb8fe65b..dc6a37c423 100644
--- a/src/melee/gr/grpura.c
+++ b/src/melee/gr/grpura.c
@@ -239,14 +239,18 @@ void grPura_8021228C(Ground_GObj* arg0) {}
 /// #grPura_80212290
 void grPura_80212290(Ground_GObj* arg0)
 {
-    Ground* gp = GET_GROUND(arg0);
-    HSD_JObj* jobj = arg0->hsd_obj;
-    HSD_ImageDesc* image = grPu_803E6E20;
-    HSD_MObjSetToonTextureImage(image);
-    lb_80011C18(jobj, 0x1000);
-    grPura_80213250(jobj);
-    HSD_MObjSetToonTextureImage(0);
-    grAnime_801C8138(arg0, gp->map_id, 0);
+    Ground* temp_r31;
+    HSD_JObj* temp_r30;
+    PAD_STACK(8);
+
+    temp_r31 = arg0->user_data;
+    temp_r30 = arg0->hsd_obj;
+    *(void (**)(HSD_GObj*, int))((u8*)arg0 + 0x1C) = fn_802130D0;
+    HSD_MObjSetToonTextureImage(&grPu_803E7620);
+    lb_80011C18(temp_r30, 0x1000U);
+    grPura_80213250(temp_r30);
+    HSD_MObjSetToonTextureImage(NULL);
+    grAnime_801C8138((HSD_GObj* ) arg0, temp_r31->map_id, 0);
 }
 
 bool grPura_80212314(Ground_GObj* arg0)
diff --git a/src/melee/it/items/itoldkuri.c b/src/melee/it/items/itoldkuri.c
index d74644a096..9b323a7ceb 100644
--- a/src/melee/it/items/itoldkuri.c
+++ b/src/melee/it/items/itoldkuri.c
@@ -38,7 +38,14 @@ void it_802D73F0(Item_GObj* gobj)
     it_802D747C(gobj);
 }
 
-void it_802D747C(Item_GObj* gobj) {}
+void it_802D747C(Item_GObj* gobj)
+{
+    u8 _[8];
+    it_8027CAD8(gobj);
+    it_8027C0A8(gobj, 0.0f, 5.0f);
+    it_802756E0(gobj);
+    it_802D848C(gobj, 0, 2);
+}
 
 bool itOldkuri_UnkMotion0_Anim(Item_GObj* gobj)
 {
