diff --git a/configure.py b/configure.py
index 84210d7294..1aa415100a 100755
--- a/configure.py
+++ b/configure.py
@@ -1111,7 +1111,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/itbombhei.c"),
             Object(NonMatching, "melee/it/items/itdosei.c"),
             Object(Matching, "melee/it/items/itheart.c"),
-            Object(NonMatching, "melee/it/items/ittomato.c"),
+            Object(Matching, "melee/it/items/ittomato.c"),
             Object(Matching, "melee/it/items/itstar.c"),
             Object(Matching, "melee/it/items/itbat.c"),
             Object(Matching, "melee/it/items/itsword.c"),
diff --git a/src/melee/it/items/ittomato.c b/src/melee/it/items/ittomato.c
index 0718a99b46..753709f0b4 100644
--- a/src/melee/it/items/ittomato.c
+++ b/src/melee/it/items/ittomato.c
@@ -9,9 +9,6 @@
 #include "it/itCommonItems.h"
 #include "it/item.h"
 
-/* 2841B4 */ static Item_GObj* it_802841B4(Item_GObj* gobj, Vec3* arg1,
-                                           s32 arg2);
-/* 28428C */ static void it_8028428C(Item_GObj* gobj);
 /* 284358 */ static void it_80284358(Item_GObj* gobj);
 /* 284380 */ static bool itTomato_UnkMotion0_Anim(HSD_GObj* gobj);
 /* 284388 */ static void itTomato_UnkMotion0_Phys(Item_GObj* gobj);
diff --git a/src/melee/it/items/ittomato.h b/src/melee/it/items/ittomato.h
index 3e37412628..a6dc284c55 100644
--- a/src/melee/it/items/ittomato.h
+++ b/src/melee/it/items/ittomato.h
@@ -7,6 +7,8 @@
 
 #include "it/items/types.h"
 
+/* 2841B4 */ Item_GObj* it_802841B4(Item_GObj* gobj, Vec3* arg1, s32 arg2);
+/* 28428C */ void it_8028428C(Item_GObj* gobj);
 /* 2842A4 */ void itTomato_Logic9_Spawned(Item_GObj*);
 /* 284324 */ void itTomato_Logic9_Destroyed(Item_GObj*);
 /* 2844E4 */ void itTomato_Logic9_PickedUp(Item_GObj*);
