diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 750d0eb36e..adf6e9889d 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -22533,8 +22533,8 @@ Ground_803E06C8 = .data:0x803E06C8; // type:object size:0xC scope:global
 lbl_803E06D4 = .data:0x803E06D4; // type:object size:0x1A data:string
 lbl_803E06F0 = .data:0x803E06F0; // type:object size:0x1A data:string
 lbl_803E070C = .data:0x803E070C; // type:object size:0x24
-lbl_803E0730 = .data:0x803E0730; // type:object size:0xC data:string
-lbl_803E073C = .data:0x803E073C; // type:object size:0xE data:string
+@209 = .data:0x803E0730; // type:object size:0xC data:string
+@260 = .data:0x803E073C; // type:object size:0xE data:string
 @202 = .data:0x803E0750; // type:object size:0x9 scope:local data:string
 ...data.0 = .data:0x803E0750; // type:label scope:local
 @203 = .data:0x803E075C; // type:object size:0x9 scope:local data:string
@@ -26453,9 +26453,9 @@ Ground_804D4514 = .sdata:0x804D4514; // type:object size:0x8 scope:global
 Ground_804D451C = .sdata:0x804D451C; // type:object size:0x8 scope:global
 Ground_804D4524 = .sdata:0x804D4524; // type:object size:0x6 scope:global data:string
 Ground_804D452C = .sdata:0x804D452C; // type:object size:0x4 scope:global data:string
-lbl_804D4530 = .sdata:0x804D4530; // type:object size:0x7 data:string
-lbl_804D4538 = .sdata:0x804D4538; // type:object size:0x5 data:string
-lbl_804D4540 = .sdata:0x804D4540; // type:object size:0x5 data:string
+@207 = .sdata:0x804D4530; // type:object size:0x7 data:string
+@208 = .sdata:0x804D4538; // type:object size:0x5 data:string
+@210 = .sdata:0x804D4540; // type:object size:0x5 data:string
 @256 = .sdata:0x804D4548; // type:object size:0x2 scope:local data:string
 @296 = .sdata:0x804D454C; // type:object size:0x4 scope:local data:string
 grAnime_804D4550 = .sdata:0x804D4550; // type:object size:0x8 scope:global data:string
diff --git a/configure.py b/configure.py
index 7e720e1bad..ed06017d1a 100755
--- a/configure.py
+++ b/configure.py
@@ -969,7 +969,7 @@ def MatchingFor(*versions):
         [
             # Main
             Object(NonMatching, "melee/gr/ground.c"),
-            Object(NonMatching, "melee/gr/grdisplay.c"),
+            Object(Matching, "melee/gr/grdisplay.c"),
             Object(Matching, "melee/gr/grdatfiles.c"),
             Object(NonMatching, "melee/gr/granime.c"),
             Object(NonMatching, "melee/gr/grmaterial.c"),
diff --git a/src/melee/gr/grdisplay.c b/src/melee/gr/grdisplay.c
index 41d432bc4d..099a54d3c6 100644
--- a/src/melee/gr/grdisplay.c
+++ b/src/melee/gr/grdisplay.c
@@ -33,8 +33,11 @@ void grDisplay_801C5B90(HSD_JObj* jobj, Mtx vmtx, u32 flags, u32 rendermode)
     if (jobj != NULL) {
         if (jobj->flags & JOBJ_INSTANCE) {
             if (!(jobj->flags & JOBJ_HIDDEN)) {
+                HSD_JObj* child;
                 HSD_JObjSetupMatrix(jobj);
-                HSD_JObjSetupMatrix(jobj->child);
+                if ((child = jobj->child) != NULL && HSD_JObjMtxIsDirty(child)) {
+                    HSD_JObjSetupMatrixSub(child);
+                }
 
                 HSD_MtxInverseConcat(jobj->child->mtx, jobj->mtx, mtx);
                 cobj = HSD_CObjGetCurrent();
@@ -74,37 +77,37 @@ void grDisplay_801C5DB0(HSD_GObj* gobj, int code)
     HSD_GObj* foggobj;
     HSD_Fog* fog;
     int i;
+    HSD_GObj* fighter;
+    u32 unused[14];
 
     gp = GET_GROUND(gobj);
     if (gp->x11_flags.b012 == Camera_8003108C()) {
         if (gp->x18 != NULL) {
             if (((intptr_t) gp->x18 & ~0x7FFFFFFF) == 0) {
-                OSReport("oioi... %08x\n");
-            }
-            if (HSD_GObj_804D7818->hsd_obj == gp->x18->hsd_obj) {
-                goto skip;
-            }
-            if (gp->x10_flags.b3 == 0) {
-                return;
-            }
-            camgobj = Camera_80030A50();
-            if (camgobj == NULL) {
-                return;
-            }
-            if (GET_COBJ(camgobj) == NULL) {
-                return;
+                OSReport("oioi... %08x\n", gp->x18);
             }
+            if (HSD_GObj_804D7818->hsd_obj != gp->x18->hsd_obj) {
+                HSD_CObj* cobj;
+                if (gp->x10_flags.b3 == 0) {
+                    return;
+                }
+                camgobj = Camera_80030A50();
+                if (camgobj == NULL) {
+                    return;
+                }
+
+                if ((cobj = GET_COBJ(camgobj)) == NULL) {
+                    return;
+                }
 
-            if (HSD_CObjGetCurrent() == GET_COBJ(camgobj)) {
-                goto skip;
+                if (HSD_CObjGetCurrent() != cobj) {
+                    return;
+                }
             }
-            return;
-        }
-        if (Ground_801C2C8C(HSD_GObj_804D7818->hsd_obj) != false) {
+        } else if (Ground_801C2C8C(HSD_GObj_804D7818->hsd_obj) != false) {
             return;
         }
 
-    skip:
         if (Camera_80030A78() == false && Camera_80030AC4() != false) {
             if (gp->x10_flags.b2 == 0) {
                 HSD_FogSet(0);
@@ -112,10 +115,9 @@ void grDisplay_801C5DB0(HSD_GObj* gobj, int code)
             HSD_StateInvalidate(-1);
 
             if (gp->x10_flags.b5 != 0) {
-                grDisplay_801C5B90(GET_JOBJ(gobj), NULL,
-                                   HSD_GObj_80390EB8(code), 0);
+                HSD_JObj* jobj = GET_JOBJ(gobj);
+                grDisplay_801C5B90(jobj, NULL, HSD_GObj_80390EB8(code), 0);
             } else {
-                HSD_GObj* fighter;
                 for (fighter = HSD_GObj_Entities->fighters; fighter != NULL;
                      fighter = fighter->next)
                 {
