diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index ad6c7852cb..cc453be0e6 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -7,6 +7,7 @@
 #include "gr/ground.h"
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
+#include "it/it_26B1.h"
 #include "lb/lb_00B0.h"
 
 void grHomeRun_8021C750(bool arg) {}
@@ -270,7 +271,18 @@ void fn_8021EB10(HSD_GObj* target_cam_gobj)
 
 /// #grHomeRun_8021EC58
 
-/// #grHomeRun_8021ED74
+void grHomeRun_8021ED74(void)
+{
+    BobOmbRain bobomb_rain;
+    HSD_JObj* jobj = Ground_801C2CF4(0x7F);
+    bobomb_rain.x0 = NULL;
+    bobomb_rain.x4 = NULL;
+    bobomb_rain.x14 = 0xB;
+    bobomb_rain.x18 = 0;
+    bobomb_rain.x1C.b0 = 1;
+    lb_8000B1CC(jobj, NULL, &bobomb_rain.x8_vec);
+    it_8026BE84(&bobomb_rain);
+}
 
 /// #grHomeRun_8021EDD4
 
diff --git a/src/melee/gr/grhomerun.h b/src/melee/gr/grhomerun.h
index 409e910be8..802abf4e1f 100644
--- a/src/melee/gr/grhomerun.h
+++ b/src/melee/gr/grhomerun.h
@@ -67,7 +67,7 @@
 /* 21EAF8 */ UNK_RET grHomeRun_8021EAF8(UNK_PARAMS);
 /* 21EB10 */ void fn_8021EB10(HSD_GObj*);
 /* 21EC58 */ UNK_RET grHomeRun_8021EC58(UNK_PARAMS);
-/* 21ED74 */ UNK_RET grHomeRun_8021ED74(UNK_PARAMS);
+/* 21ED74 */ void grHomeRun_8021ED74(void);
 /* 21EDD4 */ UNK_RET grHomeRun_8021EDD4(UNK_PARAMS);
 /* 21EEB4 */ DynamicsDesc* grHomeRun_8021EEB4(enum_t);
 /* 21EEBC */ bool grHomeRun_8021EEBC(Vec3*, int arg, HSD_JObj* jobj);
