diff --git a/extern/dolphin/include/charPipeline/structures/dolphinString.h b/extern/dolphin/include/charPipeline/structures/dolphinString.h
index 1efcc778d9..2686ec20eb 100644
--- a/extern/dolphin/include/charPipeline/structures/dolphinString.h
+++ b/extern/dolphin/include/charPipeline/structures/dolphinString.h
@@ -7,8 +7,8 @@
 extern "C" {
 #endif
 
-u8 Strcat(char *str1, char *str2, char *dst);
-void Strcpy(char *dst, char *src);
+u8 Strcat(char *str1, char *str2, char *dst, u32 dst_size);
+void Strcpy(char *dst, char *src, u32 dst_size);
 s8 Strcmp(char *str1, char *str2);
 u32 Strlen(char *str);
 
