diff --git a/src/melee/mn/mncount.c b/src/melee/mn/mncount.c
index 184e1f6eec..de4114f01f 100644
--- a/src/melee/mn/mncount.c
+++ b/src/melee/mn/mncount.c
@@ -4,9 +4,12 @@
 
 #include "gm/gm_1601.h"
 #include "gm/gmmain_lib.h"
+#include "mn/inlines.h"
+#include "mn/mnmain.h"
 #include "sysdolphin/baselib/gobjplink.h"
 
 #include <baselib/gobj.h>
+#include <baselib/jobj.h>
 
 extern s32 mnDiagram_GetFighterTotalKOs(s32);
 extern s32 mnDiagram_GetFighterTotalFalls(s32);
@@ -238,7 +241,46 @@ s32 mnCount_8025072C(CountEntry* entries, s32 start_idx, s32 mode)
     return entries[best_idx].id;
 }
 
+extern AnimLoopSettings mnCount_803EFAA0;
+
+void mnCount_802513F4(HSD_GObj* gobj)
+{
+    HSD_JObj* jobj;
+    MnCountData* data = gobj->user_data;
+    HSD_JObj* tree = gobj->hsd_obj;
+    FORCE_PAD_STACK_4;
+
+    lb_80011E24(tree, &jobj, 2, -1);
+    if (data->cursor != 0) {
+        HSD_JObjClearFlagsAll(jobj, JOBJ_HIDDEN);
+    } else {
+        HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
+    }
+
+    lb_80011E24(tree, &jobj, 1, -1);
+    if (data->cursor < 0x14) {
+        HSD_JObjClearFlagsAll(jobj, JOBJ_HIDDEN);
+    } else {
+        HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
+    }
+
+    mn_8022ED6C(tree, &mnCount_803EFAA0);
+}
+
 void fn_802514B8(HSD_GObj* gobj)
 {
     HSD_GObjPLink_80390228(gobj);
 }
+
+void mnCount_802517E0(MnCountData* data)
+{
+    int i;
+
+    data->cursor = 0;
+    data->page_size = 0xA;
+
+    for (i = 0; i < 0xA; i++) {
+        data->labels[i] = NULL;
+        data->values[i] = NULL;
+    }
+}
diff --git a/src/melee/mn/mncount.h b/src/melee/mn/mncount.h
index e300ce4885..9a1907dc54 100644
--- a/src/melee/mn/mncount.h
+++ b/src/melee/mn/mncount.h
@@ -6,6 +6,16 @@
 
 #include "mn/forward.h"
 #include <baselib/forward.h>
+#include <baselib/sislib.h>
+
+typedef struct MnCountData {
+    /* 0x00 */ u8 cursor;
+    /* 0x01 */ u8 pad_01[3];
+    /* 0x04 */ s32 page_size;
+    /* 0x08 */ HSD_Text* labels[10];
+    /* 0x30 */ HSD_Text* values[10];
+    /* 0x58 */ HSD_Text* title;
+} MnCountData; // size 0x5C
 
 /* 2502CC */ s32 mnCount_802502CC(s32);
 /* 2502F0 */ s32 fn_802502F0(u8);
@@ -23,11 +33,11 @@
 /* 250DE4 */ UNK_RET mnCount_80250DE4(UNK_PARAMS);
 /* 250FF8 */ UNK_RET mnCount_80250FF8(UNK_PARAMS);
 /* 251278 */ UNK_RET fn_80251278(UNK_PARAMS);
-/* 2513F4 */ UNK_RET mnCount_802513F4(UNK_PARAMS);
+/* 2513F4 */ void mnCount_802513F4(HSD_GObj*);
 /* 2514B8 */ void fn_802514B8(HSD_GObj* gobj);
 /* 2514D8 */ UNK_RET fn_802514D8(UNK_PARAMS);
 /* 251640 */ UNK_RET fn_80251640(UNK_PARAMS);
-/* 2517E0 */ UNK_RET mnCount_802517E0(UNK_PARAMS);
+/* 2517E0 */ void mnCount_802517E0(MnCountData*);
 /* 25186C */ void mnCount_8025186C(void);
 
 #endif
diff --git a/src/melee/mn/mndeflicker.c b/src/melee/mn/mndeflicker.c
index 8a61c52162..40cf92fc6d 100644
--- a/src/melee/mn/mndeflicker.c
+++ b/src/melee/mn/mndeflicker.c
@@ -81,7 +81,7 @@ void mnDeflicker_8024A2E8(HSD_GObj* gobj)
 {
     int x[2];
     float f =
-        mn_8022EC18(GET_JOBJ(gobj), &mnDeflicker_803EEFE4.start_frame, 0x400);
+        mn_8022EC18(GET_JOBJ(gobj), &mnDeflicker_803EEFE4, 0x400);
     if (f == mnDeflicker_803EEFE4.end_frame) {
         HSD_GObjPLink_80390228(gobj);
     }
@@ -115,7 +115,7 @@ void mnDeflicker_8024A3E8(HSD_GObj* gobj)
         p->flags_3 = HSD_GObj_804D783C;
         HSD_SisLib_803A5CC4(menu->text);
     } else {
-        float f = mn_8022EC18(jobj, &mnDeflicker_803EEFD8.start_frame, 0x400);
+        float f = mn_8022EC18(jobj, &mnDeflicker_803EEFD8, 0x400);
         if (f == mnDeflicker_803EEFD8.end_frame) {
             HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
             p = HSD_GObj_SetupProc(gobj, &mnDeflicker_8024A344, 0);
diff --git a/src/melee/mn/mnevent.c b/src/melee/mn/mnevent.c
index 8b13789179..ff5dd89405 100644
--- a/src/melee/mn/mnevent.c
+++ b/src/melee/mn/mnevent.c
@@ -1 +1,272 @@
+#include "mnevent.h"
 
+#include "db/db.h"
+#include "gm/gm_1601.h"
+#include "gm/gm_1BA8.h"
+#include "gm/gmmain_lib.h"
+#include "lb/lb_00F9.h"
+#include "lb/lbarchive.h"
+#include "lb/lbaudio_ax.h"
+#include "mn/inlines.h"
+#include "mn/mnmain.h"
+#include "mn/types.h"
+
+#include <baselib/gobj.h>
+#include <baselib/gobjgxlink.h>
+#include <baselib/gobjobject.h>
+#include <baselib/gobjplink.h>
+#include <baselib/gobjproc.h>
+#include <baselib/gobjuserdata.h>
+#include <baselib/jobj.h>
+#include <baselib/memory.h>
+#include <baselib/sislib.h>
+
+s32 mnEvent_8024CE74(void)
+{
+    int count = 0;
+    int i;
+    PAD_STACK(8);
+
+    if (g_debugLevel > 2) {
+        return 0x2A;
+    }
+
+    count = 0;
+    for (i = 0; i < 0x33; i++) {
+        if (gmMainLib_8015CEFC(i)) {
+            count += 1;
+        }
+    }
+
+    if (count <= 5) {
+        return 1;
+    }
+    if (count <= 9) {
+        return 6;
+    }
+    if (count <= 0xF) {
+        return 0xB;
+    }
+    if (count <= 0x15) {
+        return 0x10;
+    }
+    if (count <= 0x1A) {
+        return 0x14;
+    }
+    if (count >= 0x32) {
+        if (gm_80162EC8()) {
+            return 0x2A;
+        }
+        return 0x29;
+    }
+    if (gm_80164840(0x14) && gm_80164840(7) && gm_80164840(0xF) &&
+        gm_80164840(0x15))
+    {
+        if (gm_80164840(0xA) && gm_80164840(9) && gm_80164840(3) &&
+            gm_80164840(0x18) && gm_80164840(0x19) && gm_80164840(0x16) &&
+            gm_80164840(0x17))
+        {
+            return 0x29;
+        }
+        return 0x1E;
+    }
+    return 0x15;
+}
+
+static AnimLoopSettings mnEvent_803EF740 = { 0, 0, -0.1f };
+static AnimLoopSettings mnEvent_803EF74C = { 0, 0, -0.1f };
+static AnimLoopSettings mnEvent_803EF758 = { 0, 199.0f, 0 };
+
+void mnEvent_8024D4E0(HSD_JObj* jobj, Vec3* translate)
+{
+    HSD_JObjSetTranslate(jobj, translate);
+}
+
+void fn_8024E1B4(HSD_GObj* gobj)
+{
+    int i;
+    MnEventData* data = gobj->user_data;
+    if (mn_8022EC18(gobj->hsd_obj, &mnEvent_803EF74C, 0x80) >=
+        mnEvent_803EF74C.end_frame)
+    {
+        for (i = 0; i < 9; i++) {
+            if (data->gobjs[i] != NULL) {
+                HSD_GObjPLink_80390228(data->gobjs[i]);
+                data->gobjs[i] = NULL;
+            }
+            if (data->texts[i] != NULL) {
+                HSD_SisLib_803A5CC4(data->texts[i]);
+                data->texts[i] = NULL;
+            }
+            if (data->icons[i] != NULL) {
+                HSD_SisLib_803A5CC4(data->icons[i]);
+                data->icons[i] = NULL;
+            }
+        }
+        HSD_GObjPLink_80390228(gobj);
+    }
+}
+
+void fn_8024E2A0(HSD_GObj* gobj)
+{
+    HSD_GObjProc* proc;
+    HSD_JObj* jobj;
+    HSD_JObj* tree;
+    MnEventData* data = gobj->user_data;
+    PAD_STACK(8);
+
+    tree = gobj->hsd_obj;
+
+    if (mn_804A04F0.cur_menu != 7) {
+        HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+        proc = HSD_GObj_SetupProc(gobj, fn_8024E1B4, 0);
+        proc->flags_3 = HSD_GObj_804D783C;
+        HSD_SisLib_803A5CC4(data->desc_text);
+        HSD_SisLib_803A5CC4(data->name_text);
+    } else {
+        lb_80011E24(tree, &jobj, 1, -1);
+        mn_8022ED6C(jobj, &mnEvent_803EF758);
+    }
+}
+
+void fn_8024E34C(HSD_GObj* gobj)
+{
+    HSD_GObjProc* proc;
+    HSD_JObj* tree = gobj->hsd_obj;
+    MnEventData* data = gobj->user_data;
+    PAD_STACK(16);
+
+    if (mn_804A04F0.cur_menu != 7) {
+        HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+        proc = HSD_GObj_SetupProc(gobj, fn_8024E1B4, 0);
+        proc->flags_3 = HSD_GObj_804D783C;
+        HSD_SisLib_803A5CC4(data->desc_text);
+        HSD_SisLib_803A5CC4(data->name_text);
+    } else {
+        float frame = mn_8022EC18(tree, &mnEvent_803EF740, 0x80);
+        if (frame == mnEvent_803EF740.end_frame) {
+            HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+            proc = HSD_GObj_SetupProc(gobj, fn_8024E2A0, 0);
+            proc->flags_3 = HSD_GObj_804D783C;
+        }
+    }
+}
+
+void mnEvent_8024D014(HSD_GObj* gobj)
+{
+    HSD_JObj* jobj;
+    MnEventData* data = gobj->user_data;
+    HSD_JObj* tree = gobj->hsd_obj;
+    PAD_STACK(8);
+
+    lb_80011E24(tree, &jobj, 3, -1);
+    if (data->first_event == 0) {
+        HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
+    } else {
+        HSD_JObjClearFlagsAll(jobj, JOBJ_HIDDEN);
+    }
+
+    lb_80011E24(tree, &jobj, 2, -1);
+    if (data->first_event == mnEvent_8024CE74()) {
+        HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
+    } else {
+        HSD_JObjClearFlagsAll(jobj, JOBJ_HIDDEN);
+    }
+}
+
+void mnEvent_8024D0CC(HSD_GObj* gobj, s32 event)
+{
+    HSD_JObj* tree = gobj->hsd_obj;
+    HSD_JObj* jobj;
+    f32 frame;
+    FORCE_PAD_STACK_4;
+
+    if (event == 0x21) {
+        frame = 25.0f;
+    } else {
+        frame = gm_80164024(event);
+    }
+
+    lb_80011E24(tree, &jobj, 4, -1);
+    HSD_JObjReqAnimAll(jobj, frame);
+    HSD_JObjAnimAll(jobj);
+}
+
+void mnEvent_8024D7E0(HSD_GObj* gobj, s32 idx)
+{
+    MnEventData* data = gobj->user_data;
+    HSD_Text* text;
+    s32 sis_idx;
+    PAD_STACK(8);
+
+    if (data->desc_text != NULL) {
+        HSD_SisLib_803A5CC4(data->desc_text);
+    }
+
+    sis_idx = (idx * 2) + 0x155;
+    text =
+        HSD_SisLib_803A5ACC(0, 1, -9.5f, 8.0f, 17.0f, 364.68332f, 38.38772f);
+    data->desc_text = text;
+    text->font_size.x = 0.0521f;
+    text->font_size.y = 0.0521f;
+    HSD_SisLib_803A6368(text, sis_idx);
+}
+
+void mnEvent_8024E420(MnEventData* data, s32 event_idx)
+{
+    s32 max_events;
+    int i;
+
+    max_events = mnEvent_8024CE74();
+
+    if (event_idx < max_events) {
+        data->first_event = event_idx;
+        data->page = 0;
+    } else {
+        data->first_event = max_events;
+        data->page = event_idx - max_events;
+    }
+
+    data->desc_text = NULL;
+    data->name_text = NULL;
+
+    for (i = 0; i < 9; i++) {
+        data->gobjs[i] = NULL;
+        data->texts[i] = NULL;
+        data->icons[i] = NULL;
+    }
+}
+
+void mnEvent_8024E838(int event_idx, int first_time)
+{
+    HSD_GObjProc* proc;
+    void** arr = mnEvent_804A08F8;
+    char* base = (char*) &mnEvent_803EF740;
+
+    mn_804D6BC8.cooldown = 5;
+    mn_804A04F0.prev_menu = mn_804A04F0.cur_menu;
+    mn_804A04F0.cur_menu = 7;
+    mn_804A04F0.hovered_selection = 0;
+    mnEvent_804D6C65 = event_idx;
+
+    if (first_time) {
+        mnEvent_804D6C64 = 0x14;
+    } else {
+        mnEvent_804D6C64 = 0;
+    }
+
+    mnEvent_804D6C60 = NULL;
+    {
+        HSD_Archive* archive = mn_804D6BB8;
+        lbArchive_LoadSections(archive, arr, base + 0xA0, arr + 1, base + 0xB8,
+                               arr + 2, base + 0xD4, arr + 3, base + 0xF4,
+                               arr + 4, base + 0x118, 0);
+    }
+
+    if (first_time == 0) {
+        mnEvent_8024E524(event_idx);
+    }
+
+    proc = HSD_GObj_SetupProc(GObj_Create(0, 1, 0x80), fn_8024D864, 0);
+    proc->flags_3 = HSD_GObj_804D783C;
+}
diff --git a/src/melee/mn/mnevent.h b/src/melee/mn/mnevent.h
index 87d1920628..4672765876 100644
--- a/src/melee/mn/mnevent.h
+++ b/src/melee/mn/mnevent.h
@@ -3,19 +3,41 @@
 
 #include <placeholder.h>
 
-/* 24CE74 */ UNK_RET mnEvent_8024CE74(UNK_PARAMS);
-/* 24D014 */ UNK_RET mnEvent_8024D014(UNK_PARAMS);
-/* 24D0CC */ UNK_RET mnEvent_8024D0CC(UNK_PARAMS);
-/* 24D15C */ UNK_RET mnEvent_8024D15C(UNK_PARAMS);
-/* 24D4E0 */ UNK_RET mnEvent_8024D4E0(UNK_PARAMS);
-/* 24D5B0 */ UNK_RET mnEvent_8024D5B0(UNK_PARAMS);
-/* 24D7E0 */ UNK_RET mnEvent_8024D7E0(UNK_PARAMS);
-/* 24D864 */ UNK_RET fn_8024D864(UNK_PARAMS);
-/* 24E1B4 */ UNK_RET fn_8024E1B4(UNK_PARAMS);
-/* 24E2A0 */ UNK_RET fn_8024E2A0(UNK_PARAMS);
-/* 24E34C */ UNK_RET fn_8024E34C(UNK_PARAMS);
-/* 24E420 */ UNK_RET mnEvent_8024E420(UNK_PARAMS);
-/* 24E524 */ UNK_RET mnEvent_8024E524(UNK_PARAMS);
+#include <dolphin/mtx.h>
+#include <baselib/gobj.h>
+#include <baselib/jobj.h>
+#include <baselib/sislib.h>
+
+typedef struct MnEventData {
+    /* 0x00 */ u8 page;
+    /* 0x01 */ u8 pad_01[3];
+    /* 0x04 */ s32 first_event;
+    /* 0x08 */ HSD_GObj* gobjs[9];
+    /* 0x2C */ HSD_Text* texts[9];
+    /* 0x50 */ HSD_Text* icons[9];
+    /* 0x74 */ HSD_Text* name_text;
+    /* 0x78 */ HSD_Text* desc_text;
+} MnEventData;
+
+/* 24CE74 */ s32 mnEvent_8024CE74(void);
+/* 24D014 */ void mnEvent_8024D014(HSD_GObj*);
+/* 24D0CC */ void mnEvent_8024D0CC(HSD_GObj*, s32);
+/* 24D15C */ void mnEvent_8024D15C(s32, s32);
+/* 24D4E0 */ void mnEvent_8024D4E0(HSD_JObj*, Vec3*);
+/* 24D5B0 */ void mnEvent_8024D5B0(HSD_GObj*, u8);
+/* 24D7E0 */ void mnEvent_8024D7E0(HSD_GObj*, s32);
+/* 24D864 */ void fn_8024D864(HSD_GObj*);
+/* 24E1B4 */ void fn_8024E1B4(HSD_GObj*);
+/* 24E2A0 */ void fn_8024E2A0(HSD_GObj*);
+/* 24E34C */ void fn_8024E34C(HSD_GObj*);
+/* 24E420 */ void mnEvent_8024E420(MnEventData*, s32);
+/* 24E524 */ void mnEvent_8024E524(s32);
 /* 24E838 */ void mnEvent_8024E838(int, int);
 
+extern void* mnEvent_804A08F8[];
+extern void* mnEvent_804A0908;
+extern HSD_GObj* mnEvent_804D6C60;
+extern u8 mnEvent_804D6C64;
+extern u8 mnEvent_804D6C65;
+
 #endif
diff --git a/src/melee/mn/mnhyaku.c b/src/melee/mn/mnhyaku.c
index f319ad49e4..054f8fcbc1 100644
--- a/src/melee/mn/mnhyaku.c
+++ b/src/melee/mn/mnhyaku.c
@@ -14,8 +14,8 @@ static HSD_GObj* mnHyaku_804D6C58;
 
 void gm_801677E8(void);
 
-static f32 mnHyaku_803EF668[3] = { 0.0f, 19.0f, -0.1f };
-static f32 mnHyaku_803EF674[3] = { 20.0f, 29.0f, -0.1f };
+static AnimLoopSettings mnHyaku_803EF668 = { 0.0f, 19.0f, -0.1f };
+static AnimLoopSettings mnHyaku_803EF674 = { 20.0f, 29.0f, -0.1f };
 
 static f32 mnHyaku_803EF680[6] = {
     0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 5.0f,
@@ -116,8 +116,8 @@ void mnHyaku_8024C68C(HSD_GObj* arg0)
 void mnHyaku_8024C9F0(HSD_GObj* gobj)
 {
     HSD_JObj* jobj = (HSD_JObj*) gobj->hsd_obj;
-    f32 ret = mn_8022EC18(jobj, &mnHyaku_803EF674[0], 0x80);
-    if (ret >= mnHyaku_803EF674[1]) {
+    f32 ret = mn_8022EC18(jobj, &mnHyaku_803EF674, 0x80);
+    if (ret >= mnHyaku_803EF674.end_frame) {
         HSD_GObjPLink_80390228(gobj);
     }
 }
@@ -151,8 +151,8 @@ void mnHyaku_8024CAC8(HSD_GObj* gobj)
         HSD_SisLib_803A5CC4(menu->text);
         return;
     }
-    res = mn_8022EC18(jobj, &mnHyaku_803EF668[0], 0x80);
-    if (res == mnHyaku_803EF668[1]) {
+    res = mn_8022EC18(jobj, &mnHyaku_803EF668, 0x80);
+    if (res == mnHyaku_803EF668.end_frame) {
         HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
         gobj_proc = HSD_GObj_SetupProc(gobj, mnHyaku_8024CA50, 0U);
         gobj_proc->flags_3 = HSD_GObj_804D783C;
diff --git a/src/melee/mn/mnitemsw.c b/src/melee/mn/mnitemsw.c
index 8b13789179..dea0ee70f7 100644
--- a/src/melee/mn/mnitemsw.c
+++ b/src/melee/mn/mnitemsw.c
@@ -1 +1,45 @@
+#include "mnitemsw.h"
 
+#include "mn/inlines.h"
+#include "mn/mnmain.h"
+
+#include <baselib/gobj.h>
+#include <baselib/gobjproc.h>
+#include <baselib/jobj.h>
+
+HSD_JObj* mnItemSw_8023405C(MnItemSwData* data, u8 idx)
+{
+    HSD_JObj* cur;
+    u8 i;
+
+    if (idx >= 16) {
+        cur = HSD_JObjGetChild(data->jobjs[6]);
+        for (i = 16; i < idx; i++) {
+            cur = HSD_JObjGetNext(cur);
+        }
+    } else {
+        if (data->jobjs[4] == NULL) {
+            cur = NULL;
+        } else {
+            cur = data->jobjs[4]->child;
+        }
+        for (i = 0; i < idx; i++) {
+            cur = HSD_JObjGetNext(cur);
+        }
+    }
+    return cur;
+}
+
+void mnItemSw_802358C0(void)
+{
+    HSD_GObjProc* proc;
+
+    mn_804A04F0.prev_menu = mn_804A04F0.cur_menu;
+    mn_804A04F0.cur_menu = 0x10;
+    mn_804A04F0.hovered_selection = 0;
+    mnItemSw_804D6BEC = 1;
+    mnItemSw_802351A0(1);
+
+    proc = HSD_GObj_SetupProc(GObj_Create(0, 1, 0x80), fn_80233E10, 0);
+    proc->flags_3 = HSD_GObj_804D783C;
+}
diff --git a/src/melee/mn/mnitemsw.h b/src/melee/mn/mnitemsw.h
index 8b13789179..ccf88bb14d 100644
--- a/src/melee/mn/mnitemsw.h
+++ b/src/melee/mn/mnitemsw.h
@@ -1 +1,32 @@
+#ifndef MELEE_MN_ITEMSW_H
+#define MELEE_MN_ITEMSW_H
 
+#include <placeholder.h>
+
+#include <baselib/gobj.h>
+#include <baselib/jobj.h>
+
+typedef struct MnItemSwData {
+    /* 0x00 */ u8 menu_kind;
+    /* 0x01 */ u8 cursor;
+    /* 0x02 */ u8 items[0x1F];
+    /* 0x21 */ u8 x21;
+    /* 0x22 */ u8 pad;
+    /* 0x23 */ u8 x23;
+    /* 0x24 */ HSD_JObj* jobjs[9];
+} MnItemSwData; // size 0x48
+
+/* 233A98 */ s32 mnItemSw_80233A98(s32);
+/* 233B68 */ UNK_RET mnItemSw_80233B68(UNK_PARAMS);
+/* 233E10 */ void fn_80233E10(HSD_GObj*);
+/* 23405C */ HSD_JObj* mnItemSw_8023405C(MnItemSwData*, u8);
+/* 234104 */ UNK_RET mnItemSw_80234104(UNK_PARAMS);
+/* 23453C */ UNK_RET mnItemSw_8023453C(UNK_PARAMS);
+/* 234C24 */ UNK_RET fn_80234C24(UNK_PARAMS);
+/* 235020 */ UNK_RET mnItemSw_80235020(UNK_PARAMS);
+/* 2351A0 */ void mnItemSw_802351A0(s32);
+/* 2358C0 */ void mnItemSw_802358C0(void);
+
+extern u8 mnItemSw_804D6BEC;
+
+#endif
diff --git a/src/melee/mn/mnlanguage.c b/src/melee/mn/mnlanguage.c
index 980e95aba4..6dff592818 100644
--- a/src/melee/mn/mnlanguage.c
+++ b/src/melee/mn/mnlanguage.c
@@ -91,7 +91,7 @@ void fn_8024C210(HSD_GObj* gobj)
 {
     u8 _[4];
     HSD_JObj* jobj = GET_JOBJ(gobj);
-    if (mn_8022EC18(jobj, &mnLanguage_803EF5AC.x, 0x80) >=
+    if (mn_8022EC18(jobj, (AnimLoopSettings*) &mnLanguage_803EF5AC, 0x80) >=
         mnLanguage_803EF5AC.y)
     {
         HSD_GObjPLink_80390228(gobj);
@@ -127,7 +127,7 @@ void fn_8024C2E8(HSD_GObj* gobj)
         return;
     }
 
-    tmp = mn_8022EC18(jobj, &mnLanguage_803EF5A0.x, 0x80);
+    tmp = mn_8022EC18(jobj, (AnimLoopSettings*) &mnLanguage_803EF5A0, 0x80);
     if (tmp == mnLanguage_803EF5A0.y) {
         HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
         gobjproc = HSD_GObj_SetupProc(gobj, fn_8024C270, 0);
diff --git a/src/melee/mn/mnmain.c b/src/melee/mn/mnmain.c
index 5edeb29510..fd249d9e1a 100644
--- a/src/melee/mn/mnmain.c
+++ b/src/melee/mn/mnmain.c
@@ -2953,23 +2953,25 @@ void mn_8022EBDC(void)
     }
 }
 
-float mn_8022EC18(HSD_JObj* arg0, float* arg1, HSD_TypeMask arg2)
+float mn_8022EC18(HSD_JObj* arg0, AnimLoopSettings* arg1, HSD_TypeMask arg2)
 {
     float temp_f1;
     float var_f31;
 
     temp_f1 = mn_8022F298(arg0);
-    if (!(arg1[0] <= temp_f1) || !(temp_f1 <= arg1[1])) {
-        HSD_JObjReqAnimAll(arg0, *(volatile float*) &arg1[0]);
+    if (!(arg1->start_frame <= temp_f1) ||
+        !(temp_f1 <= arg1->end_frame))
+    {
+        HSD_JObjReqAnimAll(arg0, *(volatile float*) &arg1->start_frame);
         mn_8022F3D8(arg0, 0xFFU, arg2);
     }
-    if (arg1[2] == -0.1f) {
+    if (arg1->loop_frame == -0.1f) {
         var_f31 = mn_8022F298(arg0);
-        if (var_f31 < arg1[1]) {
+        if (var_f31 < arg1->end_frame) {
             HSD_JObjAnimAll(arg0);
             var_f31 = mn_8022F298(arg0);
-            if (var_f31 > arg1[1]) {
-                var_f31 = arg1[1];
+            if (var_f31 > arg1->end_frame) {
+                var_f31 = arg1->end_frame;
                 HSD_JObjReqAnimAll(arg0, var_f31);
                 mn_8022F3D8(arg0, 0xFFU, arg2);
                 HSD_JObjAnimAll(arg0);
@@ -2978,8 +2980,8 @@ float mn_8022EC18(HSD_JObj* arg0, float* arg1, HSD_TypeMask arg2)
     } else {
         HSD_JObjAnimAll(arg0);
         var_f31 = mn_8022F298(arg0);
-        if (var_f31 >= arg1[1]) {
-            var_f31 = arg1[2] + (var_f31 - arg1[1]);
+        if (var_f31 >= arg1->end_frame) {
+            var_f31 = arg1->loop_frame + (var_f31 - arg1->end_frame);
             HSD_JObjReqAnimAll(arg0, var_f31);
             mn_8022F3D8(arg0, 0xFFU, arg2);
             HSD_JObjAnimAll(arg0);
diff --git a/src/melee/mn/mnmain.h b/src/melee/mn/mnmain.h
index 872ea0c994..95a37a4f31 100644
--- a/src/melee/mn/mnmain.h
+++ b/src/melee/mn/mnmain.h
@@ -96,7 +96,7 @@ typedef struct _MenuInputState {
 /* 22EB24 */ s32 mn_GetDigitAt(s32, s32);
 /* 22EB78 */ s32 mn_GetDigitCount(s32);
 /* 22EBDC */ void mn_8022EBDC(void);
-/* 22EC18 */ float mn_8022EC18(HSD_JObj*, float*, HSD_TypeMask);
+/* 22EC18 */ float mn_8022EC18(HSD_JObj*, AnimLoopSettings*, HSD_TypeMask);
 /* 22ED6C */ float mn_8022ED6C(HSD_JObj*, AnimLoopSettings*);
 /* 22EE84 */ float mn_8022EE84(HSD_JObj* arg0, AnimLoopSettings* arg1,
                                HSD_TypeMask arg2);
diff --git a/src/melee/mn/mnmainrule.c b/src/melee/mn/mnmainrule.c
index 85f09d1d9f..013cea1145 100644
--- a/src/melee/mn/mnmainrule.c
+++ b/src/melee/mn/mnmainrule.c
@@ -2,6 +2,9 @@
 
 #include "mnmain.h"
 
+#include "gm/gm_1A3F.h"
+
+#include <sysdolphin/baselib/gobj.h>
 #include <sysdolphin/baselib/gobjplink.h>
 #include <sysdolphin/baselib/gobjproc.h>
 #include <sysdolphin/baselib/sislib.h>
@@ -64,16 +67,71 @@ int mn_80231634(struct mn_80231634_t* arg0)
     return arg0->x10;
 }
 
-/// #mn_8023164C
+void mn_8023164C(void)
+{
+    HSD_GObjProc* proc;
+    s32 mode;
+
+    mn_804A04F0.prev_menu = mn_804A04F0.cur_menu;
+    mn_804A04F0.cur_menu = 0xD;
+
+    switch (mn_804A04F0.prev_menu) {
+    case 15:
+        mn_804A04F0.hovered_selection = 6;
+        mode = 3;
+        break;
+    case 16:
+        mn_804A04F0.hovered_selection = 5;
+        mode = 3;
+        break;
+    default:
+        mn_804A04F0.hovered_selection = 0;
+        mode = 1;
+        break;
+    }
+
+    proc = HSD_GObj_SetupProc(GObj_Create(0, 1, 0x80), fn_8022F538, 0);
+    proc->flags_3 = HSD_GObj_804D783C;
+    HSD_GObj_80390CD4(mn_80230E38(mode));
+}
+
+extern s32 mn_804D6BD4;
 
-/// #mn_80231714
+void mn_80231714(void)
+{
+    HSD_GObjProc* proc;
+    s32 mode;
+
+    mn_804A04F0.prev_menu = mn_804A04F0.cur_menu;
+    mn_804A04F0.cur_menu = 0xD;
+
+    switch (mn_804A04F0.prev_menu) {
+    case 15:
+        mn_804A04F0.hovered_selection = 6;
+        mode = 3;
+        break;
+    case 16:
+        mn_804A04F0.hovered_selection = 5;
+        mode = 3;
+        break;
+    default:
+        mn_804A04F0.hovered_selection = 0;
+        mode = 1;
+        break;
+    }
+
+    proc = HSD_GObj_SetupProc(GObj_Create(0, 1, 0x80), fn_8022F538, 0);
+    proc->flags_3 = HSD_GObj_804D783C;
+    HSD_GObj_80390CD4(mn_80230E38(mode));
+    mn_804D6BD4 = 1;
+}
 
 void mn_802317E4(HSD_Archive* archive, int arg1)
 {
     mn_80231804(archive, arg1);
 }
 
-extern int mn_804D6BD4;
+extern s32 mn_804D6BD4;
 
 void mn_80231804(HSD_Archive* archive, int arg1)
 {
@@ -226,4 +284,22 @@ void mn_80231804(HSD_Archive* archive, int arg1)
     lbAudioAx_80023F28(gmMainLib_8015ECB0());
 }
 
-/// #mn_80231F80
+bool mn_80231F80(u8 arg0)
+{
+    if (gm_801A4310() == 0x1B && arg0 == 1) {
+        return false;
+    }
+    if (arg0 == 3) {
+        if (gmMainLib_8015EE0C()) {
+            return true;
+        }
+        return false;
+    }
+    if (arg0 == 5) {
+        if (gmMainLib_8015EE44()) {
+            return true;
+        }
+        return false;
+    }
+    return true;
+}
diff --git a/src/melee/mn/mnmainrule.h b/src/melee/mn/mnmainrule.h
index 44158e5bc3..54a19122cc 100644
--- a/src/melee/mn/mnmainrule.h
+++ b/src/melee/mn/mnmainrule.h
@@ -22,10 +22,10 @@ struct mn_80231634_t {
 /* 230D18 */ UNK_RET mn_80230D18(UNK_PARAMS);
 /* 230E38 */ HSD_GObj* mn_80230E38(int);
 /* 231634 */ int mn_80231634(struct mn_80231634_t*);
-/* 23164C */ UNK_RET mn_8023164C(UNK_PARAMS);
+/* 23164C */ void mn_8023164C(void);
 /* 231714 */ UNK_RET mn_80231714(UNK_PARAMS);
 /* 2317E4 */ void mn_802317E4(HSD_Archive*, int);
 /* 231804 */ void mn_80231804(HSD_Archive*, int);
-/* 231F80 */ UNK_RET mn_80231F80(UNK_PARAMS);
+/* 231F80 */ bool mn_80231F80(u8);
 
 #endif
diff --git a/src/melee/mn/mnsound.c b/src/melee/mn/mnsound.c
index 88e495a1cc..2356631a3c 100644
--- a/src/melee/mn/mnsound.c
+++ b/src/melee/mn/mnsound.c
@@ -182,7 +182,7 @@ void fn_80249A1C(HSD_GObj* arg0)
         } else {
             lb_80011E24(jobj, &sp34, 9, -1);
         }
-        arg8 = mn_8022EC18(sp34, (f32*) mnSound_803EEED8, 0x80);
+        arg8 = mn_8022EC18(sp34, &mnSound_803EEED8[0], 0x80);
     }
     lb_80011E24((HSD_JObj*) arg0->hsd_obj, &sp20, 6, -1, arg8);
     temp_f1 = mn_8022F298(sp20);
diff --git a/src/melee/mn/mnsoundtest.c b/src/melee/mn/mnsoundtest.c
index dba89a1565..66cf4655df 100644
--- a/src/melee/mn/mnsoundtest.c
+++ b/src/melee/mn/mnsoundtest.c
@@ -443,3 +443,4 @@ void fn_8024AED0(HSD_GObj* arg0)
         mnSoundTest_8024A790(mnSoundTest_804D6C40);
     }
 }
+
diff --git a/src/melee/mn/mnsoundtest.h b/src/melee/mn/mnsoundtest.h
index c022a649c7..013632c560 100644
--- a/src/melee/mn/mnsoundtest.h
+++ b/src/melee/mn/mnsoundtest.h
@@ -49,7 +49,7 @@ typedef struct {
 /* 24AD58 */ void mnSoundTest_8024AD58(HSD_GObj*, u8);
 /* 24AED0 */ void fn_8024AED0(HSD_GObj*);
 /* 24B2B0 */ void fn_8024B2B0(HSD_GObj*);
-/* 24B7E4 */ UNK_RET fn_8024B7E4(UNK_PARAMS);
+/* 24B7E4 */ void fn_8024B7E4(HSD_GObj*);
 /* 24B8B4 */ UNK_RET fn_8024B8B4(UNK_PARAMS);
 /* 24BAF0 */ UNK_RET fn_8024BAF0(UNK_PARAMS);
 /* 24BCA0 */ UNK_RET mnSoundTest_8024BCA0(UNK_PARAMS);
diff --git a/src/melee/mn/mnvibration.c b/src/melee/mn/mnvibration.c
index bc0eab7c2e..44d61d9599 100644
--- a/src/melee/mn/mnvibration.c
+++ b/src/melee/mn/mnvibration.c
@@ -624,7 +624,7 @@ void fn_80248748(HSD_GObj* gobj)
     }
 }
 
-static f32 mnVibration_803EECF8[3] = { 0.0f, 14.0f, -0.1f };
+static AnimLoopSettings mnVibration_803EECF8 = { 0.0f, 14.0f, -0.1f };
 
 void fn_802487A8(HSD_GObj* gobj)
 {
@@ -669,7 +669,7 @@ void fn_802487A8(HSD_GObj* gobj)
         }
         lb_80011E24(var_r3_2, &sp44, 1, -1);
         if (GetRumbleSettingOfPort(var_r24) != 0) {
-            mn_8022EC18(sp44, mnVibration_803EECF8, MOBJ_MASK);
+            mn_8022EC18(sp44, &mnVibration_803EECF8, MOBJ_MASK);
         } else {
             HSD_JObjReqAnimAll(sp44, 0.0f);
             mn_8022F3D8(sp44, 0xFF, MOBJ_MASK);
