diff --git a/tools/project.py b/tools/project.py
index a2a700fb0f..ac352bc224 100644
--- a/tools/project.py
+++ b/tools/project.py
@@ -285,7 +285,7 @@ def compiler_wrapper(self) -> Optional[Path]:
     def use_wibo(self) -> bool:
         return (
             self.wibo_tag is not None
-            and (sys.platform == "linux" or sys.platform == "darwin")
+            and sys.platform == "linux"
             and platform.machine() in ("i386", "x86_64", "aarch64", "arm64")
             and self.wrapper is None
         )
