diff --git a/src/melee/gr/granime.c b/src/melee/gr/granime.c
index 738de56305..d33ebc6bdd 100644
--- a/src/melee/gr/granime.c
+++ b/src/melee/gr/granime.c
@@ -32,8 +32,7 @@
                                           HSD_AnimJoint* animjoint,
                                           HSD_MatAnimJoint* matanimjoint,
                                           HSD_ShapeAnimJoint* shapeanimjoint);
-/* 1C6C0C */ void grAnime_801C6C0C(HSD_JObj* jobj,
-                                   HSD_AnimJoint* animjoint,
+/* 1C6C0C */ void grAnime_801C6C0C(HSD_JObj* jobj, HSD_AnimJoint* animjoint,
                                    HSD_MatAnimJoint* matanimjoint,
                                    HSD_ShapeAnimJoint* shapeanimjoint);
 /* 1C6EE4 */ static void fn_801C6EE4(HSD_AObj* aobj);
@@ -767,7 +766,38 @@ bool grAnime_801C83D0(HSD_GObj* gobj, bool arg1, enum_t arg2)
     return false;
 }
 
-/// #grAnime_801C84A4
+bool grAnime_801C84A4(HSD_GObj* gobj, s32 arg1, s32 arg2)
+{
+    enum _HSD_TypeMask mask = 0;
+    HSD_JObj* jobj;
+    u8 _[8];
+    HSD_AObj* sp14 = NULL;
+    HSD_AObj* result;
+
+    jobj = Ground_801C3FA4(gobj, arg1);
+    if (jobj == NULL) {
+        result = NULL;
+    } else {
+        if (arg2 & 1) {
+            mask |= 0x220;
+        }
+        if (arg2 & 2) {
+            mask |= 0x7484;
+        }
+        if (arg2 & 4) {
+            mask |= 0x100;
+        }
+        if (!__setjmp(&grAnime_8049EE40)) {
+            HSD_ForeachAnim(jobj, JOBJ_TYPE, mask, fn_801C82E8, AOBJ_ARG_AV,
+                            &sp14);
+        }
+        result = sp14;
+    }
+    if (HSD_AObjGetFlags(result) & 0x04000000) {
+        return true;
+    }
+    return false;
+}
 
 /// #grAnime_801C8578
 
diff --git a/src/melee/gr/grbattle.c b/src/melee/gr/grbattle.c
index 1e2ccfe8f7..cf6480407f 100644
--- a/src/melee/gr/grbattle.c
+++ b/src/melee/gr/grbattle.c
@@ -169,7 +169,7 @@ static void grBattle_OnLoad(void) {}
 
 static void grBattle_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grBattle_80219D7C(void)
diff --git a/src/melee/gr/grbigblue.c b/src/melee/gr/grbigblue.c
index feb0f589d6..a9b56fd696 100644
--- a/src/melee/gr/grbigblue.c
+++ b/src/melee/gr/grbigblue.c
@@ -177,7 +177,7 @@ void grBigBlue_801E59C8(void) {}
 
 void grBigBlue_801E59CC(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grBigBlue_801E59F0(void)
diff --git a/src/melee/gr/grcastle.c b/src/melee/gr/grcastle.c
index 5160d28d2d..604ea507e3 100644
--- a/src/melee/gr/grcastle.c
+++ b/src/melee/gr/grcastle.c
@@ -305,7 +305,7 @@ void grCastle_801CD4A0(void) {}
 
 void grCastle_801CD4A4(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grCastle_801CD4C8(void)
diff --git a/src/melee/gr/grcorneria.c b/src/melee/gr/grcorneria.c
index 3e0cef162f..37fc0fbbf0 100644
--- a/src/melee/gr/grcorneria.c
+++ b/src/melee/gr/grcorneria.c
@@ -19,6 +19,7 @@
 #include "if/ifcoget.h"
 #include "if/ifstatus.h"
 #include "lb/lb_00B0.h"
+#include "lb/lb_00F9.h"
 #include "mp/mplib.h"
 
 #include <baselib/controller.h>
@@ -96,7 +97,7 @@ void grCorneria_801DD478(void)
 
 void grCorneria_801DD508(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grCorneria_801DD52C(void)
@@ -196,8 +197,8 @@ void grCorneria_801DD65C(Ground_GObj* gobj)
     GET_GROUND(gobj)->gv.corneria.xC4.flags.b0 = false;
 }
 
-extern int grCn_803E1FE8[];
-extern int grCn_803E2000[];
+extern grZakoGenerator_Spawn grCn_803E1FE8[];
+extern grZakoGenerator_Spawn grCn_803E2000[];
 
 void grCorneria_801DD674(Ground_GObj* ground_gobj)
 {
@@ -267,7 +268,43 @@ bool grCorneria_801DD9A0(Ground_GObj* arg)
     return false;
 }
 
-/// #grCorneria_801DD9A8
+extern Vec3 grCn_803E1F70[5];
+extern Vec3 grCn_803E1FAC[5];
+
+void grCorneria_801DD9A8(Ground_GObj* gobj)
+{
+    Ground* gp = HSD_GObjGetUserData(gobj);
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    int i;
+
+    if (!gp->gv.corneria.xC4.flags.b0) {
+        f32 speed;
+
+        grCorneria_801E1970(gobj);
+        grCorneria_801E2228(gobj);
+
+        speed = -35.0f * Ground_801C0498() - gp->gv.corneria.xD0;
+        grZakoGenerator_801CA43C(gp->gv.corneria.xC8, Ground_801C3FA4(gobj, 0),
+                                 speed);
+
+        speed = -35.0f * Ground_801C0498() - gp->gv.corneria.xD0;
+        grZakoGenerator_801CA43C(gp->gv.corneria.xCC, Ground_801C3FA4(gobj, 0),
+                                 speed);
+
+        grCorneria_801E1348(gobj);
+        grCorneria_801E0E40();
+        grCorneria_801DCE1C();
+    }
+
+    for (i = 0; i < 5; i++) {
+        lb_8000B1CC(jobj, &grCn_803E1F70[i], &grCn_803E1FAC[i]);
+    }
+
+    Ground_801C2FE0(gobj);
+    grCorneria_801E1878(gobj);
+    grCorneria_801E2110();
+    lb_800115F4();
+}
 
 void grCorneria_801DDAC0(Ground_GObj* arg) {}
 
@@ -311,7 +348,11 @@ typedef struct {
 extern grCn_Entry grCn_803E2204[][5];
 
 typedef struct grCn_Data {
-    /* 0x000 */ u8 pad[0x4CC];
+    /* 0x000 */ u8 pad0[0x3C];
+    /* 0x03C */ s32 indices[3];
+    /* 0x048 */ u8 pad1[0x2E8];
+    /* 0x330 */ Vec3 positions[34];
+    /* +4C8 */ UNK_T pad2;
     /* 0x4CC */ grCn_Entry entries[][5];
 } grCn_Data;
 
@@ -320,8 +361,8 @@ extern grCn_Data grCn_803E1D38;
 void grCorneria_801DDDA8(HSD_GObj* gobj, Vec3* vec)
 {
     grCn_Data* data = &grCn_803E1D38;
+    Ground* gp2;
     Ground* gp;
-    Ground* corn_gp;
     HSD_JObj* jobj;
     u8 unused[12];
     Vec3 pos;
@@ -330,23 +371,13 @@ void grCorneria_801DDDA8(HSD_GObj* gobj, Vec3* vec)
 
     if (gobj != NULL) {
         gp = gobj->user_data;
-        corn_gp = Ground_801C2BA4(3)->user_data;
+        gp2 = Ground_801C2BA4(3)->user_data;
         jobj = Ground_801C3FA4(gobj, 4);
         lb_8000B1CC(jobj, NULL, &pos);
-        idx = ((int*) &data->pad[0x3C])[gp->gv.corneria.xC8];
-        vec->x =
-            corn_gp->gv.corneria.offset_x +
-            (-pos.z +
-             *(f32*) &data->pad[0x330 +
-                                idx * (int) sizeof(grCn_Entry)]);
-        vec->y =
-            pos.y +
-            *(f32*) &data->pad[0x334 +
-                               idx * (int) sizeof(grCn_Entry)];
-        vec->z =
-            pos.x +
-            *(f32*) &data->pad[0x338 +
-                               idx * (int) sizeof(grCn_Entry)];
+        idx = data->indices[gp->gv.arwing.xC8];
+        vec->x = gp2->gv.arwing.xDC + (-pos.z + data->positions[idx].x);
+        vec->y = pos.y + data->positions[idx].y;
+        vec->z = pos.x + data->positions[idx].z;
     } else {
         vec->z = 0.0f;
         vec->y = 0.0f;
@@ -697,14 +728,11 @@ void grCorneria_801E0F6C(Ground_GObj* gobj)
     un_802FF570();
 
     if (Ground_801C2BA4(5) == NULL) {
-        if (grCorneria_801E2598(gp->gv.arwing.xC4,
-                                gp->gv.arwing.xC8))
-        {
+        if (grCorneria_801E2598(gp->gv.arwing.xC4, gp->gv.arwing.xC8)) {
             if ((s32) gp->gv.arwing.xCC-- < 0) {
                 HSD_GObj* gobj5 = grCorneria_801DD534(5);
                 Ground* gp5 = GET_GROUND(gobj5);
-                grCorneria_801E2738(gobj5, &gp5->gv,
-                                    gp->gv.arwing.xC4,
+                grCorneria_801E2738(gobj5, &gp5->gv, gp->gv.arwing.xC4,
                                     gp->gv.arwing.xC8);
                 gp->gv.arwing.xC8++;
                 gp->gv.arwing.xCC = 0;
@@ -774,12 +802,12 @@ void grCorneria_801E1878(Ground_GObj* gobj)
 
 /// #grCorneria_801E2228
 
-HSD_Generator* grCorneria_801E2454(Vec3* vec)
+HSD_Generator* grCorneria_801E2454(Vec3* vec, s32 arg1)
 {
     return grLib_801C96F8(0x7534, 0x1E, vec);
 }
 
-HSD_Generator* grCorneria_801E2480(Vec3* vec)
+HSD_Generator* grCorneria_801E2480(Vec3* vec, s32 arg1)
 {
     return grLib_801C96F8(0x7530, 0x1E, vec);
 }
diff --git a/src/melee/gr/grcorneria.h b/src/melee/gr/grcorneria.h
index f1d4cb7023..e71d03ba5d 100644
--- a/src/melee/gr/grcorneria.h
+++ b/src/melee/gr/grcorneria.h
@@ -73,14 +73,14 @@
 /* 1E1060 */ void grCorneria_801E1060(Ground_GObj*);
 /* 1E12D0 */ void fn_801E12D0(Item_GObj*, Ground*);
 /* 1E12D4 */ void fn_801E12D4(Item_GObj*, Ground*, Vec3*, HSD_GObj*, f32);
-/* 1E1348 */ UNK_RET grCorneria_801E1348(UNK_PARAMS);
+/* 1E1348 */ void grCorneria_801E1348(Ground_GObj*);
 /* 1E1878 */ void grCorneria_801E1878(Ground_GObj*);
-/* 1E1970 */ UNK_RET grCorneria_801E1970(UNK_PARAMS);
+/* 1E1970 */ void grCorneria_801E1970(Ground_GObj*);
 /* 1E1BF0 */ HSD_GObj* grCorneria_801E1BF0(void);
-/* 1E2110 */ UNK_RET grCorneria_801E2110(UNK_PARAMS);
-/* 1E2228 */ UNK_RET grCorneria_801E2228(UNK_PARAMS);
-/* 1E2454 */ HSD_Generator* grCorneria_801E2454(Vec3*);
-/* 1E2480 */ HSD_Generator* grCorneria_801E2480(Vec3*);
+/* 1E2110 */ void grCorneria_801E2110(void);
+/* 1E2228 */ void grCorneria_801E2228(Ground_GObj*);
+/* 1E2454 */ HSD_Generator* grCorneria_801E2454(Vec3*, s32);
+/* 1E2480 */ HSD_Generator* grCorneria_801E2480(Vec3*, s32);
 /* 1E24AC */ void smashTaunt_801E24AC(Ground_GObj*, int);
 /* 1E2550 */ void smashTaunt_801E2550(Ground_GObj*,
                                       struct grSmashTaunt_GroundVars*);
diff --git a/src/melee/gr/grfigure1.c b/src/melee/gr/grfigure1.c
index e0c0b32256..fad788de88 100644
--- a/src/melee/gr/grfigure1.c
+++ b/src/melee/gr/grfigure1.c
@@ -88,7 +88,7 @@ static void grFigure1_OnLoad(void) {}
 
 static void grFigure1_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grFigure1_8020E078(void)
diff --git a/src/melee/gr/grfigure2.c b/src/melee/gr/grfigure2.c
index d4e28780b8..93d0e2c6da 100644
--- a/src/melee/gr/grfigure2.c
+++ b/src/melee/gr/grfigure2.c
@@ -85,7 +85,7 @@ static void grFigure2_OnLoad(void) {}
 
 static void grFigure2_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grFigure2_8020E30C(void)
diff --git a/src/melee/gr/grfigure3.c b/src/melee/gr/grfigure3.c
index 336709453a..5cf40d92b5 100644
--- a/src/melee/gr/grfigure3.c
+++ b/src/melee/gr/grfigure3.c
@@ -87,7 +87,7 @@ static void grFigure3_OnLoad(void) {}
 
 static void grFigure3_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grFigure3_8020E5A0(void)
diff --git a/src/melee/gr/grfigureget.c b/src/melee/gr/grfigureget.c
index 0f8cb3bb96..0cc045e603 100644
--- a/src/melee/gr/grfigureget.c
+++ b/src/melee/gr/grfigureget.c
@@ -2,6 +2,7 @@
 
 #include <platform.h>
 
+#include "baselib/forward.h"
 #include "gr/forward.h"
 
 #include "gr/granime.h"
@@ -11,8 +12,18 @@
 #include "gr/inlines.h"
 #include "gr/types.h"
 
+#include "it/forward.h"
+
+#include "it/inlines.h"
+#include "it/items/itcoin.h"
+#include "it/types.h"
+
 #include "lb/forward.h"
 
+#include "lb/types.h"
+
+#include "mp/forward.h"
+
 #include <dolphin/mtx.h>
 #include <dolphin/os/OSError.h>
 #include <baselib/gobj.h>
@@ -81,7 +92,7 @@ void grFigureGet_OnLoad(void) {}
 
 void grFigureGet_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grFigureGet_802195C4(void)
@@ -131,7 +142,35 @@ bool grFigureGet_80219890(Ground_GObj* gobj)
 
 void grFigureGet_80219B0C(Ground_GObj* gobj) {}
 
-/// #grFigureGet_80219B10
+void grFigureGet_80219B10(Ground* gp, s32 arg1, CollData* cd, s32 arg3,
+                          mpLib_GroundEnum arg4, f32 arg5)
+{
+    HSD_GObj* item_gobj = cd->x0_gobj;
+    if ((s32) cd->x34_flags.b1234 == 5 && item_gobj &&
+        item_gobj->classifier == HSD_GOBJ_CLASS_ITEM)
+    {
+        Item* ip = GET_ITEM(cd->x0_gobj);
+        if (ip->kind == It_Kind_Coin && gp->gv.figureget.x8 < 3) {
+            HSD_GObj* item_gobj = cd->x0_gobj;
+            int value = it_802F2020(item_gobj);
+            int count = 0;
+            int i;
+            for (i = 0; i < gp->gv.figureget.x8; i++) {
+                if (gp->gv.figureget.x28[i] == cd->x0_gobj) {
+                    count++;
+                }
+            }
+            if (count == 0) {
+                if (cd->cur_pos.x <= 4.5f && cd->cur_pos.x >= -4.5f) {
+                    it_802F202C(cd->x0_gobj);
+                    gp->gv.figureget.x28[gp->gv.figureget.x8] = cd->x0_gobj;
+                    gp->gv.figureget.x1C[gp->gv.figureget.x8] = value;
+                    gp->gv.figureget.x8++;
+                }
+            }
+        }
+    }
+}
 
 void grFigureGet_80219C34(HSD_GObj* gobj)
 {
diff --git a/src/melee/gr/grfigureget.h b/src/melee/gr/grfigureget.h
index 94415930f0..0933b5d518 100644
--- a/src/melee/gr/grfigureget.h
+++ b/src/melee/gr/grfigureget.h
@@ -4,9 +4,15 @@
 #include <placeholder.h>
 #include <platform.h>
 
+#include <melee/gr/forward.h>
+#include <melee/lb/forward.h>
+#include <melee/mp/forward.h>
+
 #include <baselib/forward.h>
 
-/* 219B10 */ UNK_RET grFigureGet_80219B10(UNK_PARAMS);
+/* 219B10 */ void grFigureGet_80219B10(Ground* gp, s32 arg1, CollData* cd,
+                                       s32 arg3, mpLib_GroundEnum arg4,
+                                       f32 arg5);
 /* 219C34 */ void grFigureGet_80219C34(HSD_GObj*);
 /* 219C50 */ bool grFigureGet_80219C50(HSD_GObj*);
 
diff --git a/src/melee/gr/grfourside.c b/src/melee/gr/grfourside.c
index ef9030b7d1..985425fe56 100644
--- a/src/melee/gr/grfourside.c
+++ b/src/melee/gr/grfourside.c
@@ -108,7 +108,7 @@ void grFourside_801F2DA0(void) {}
 
 void grFourside_801F2DA4(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grFourside_801F2DC8(void)
diff --git a/src/melee/gr/grgarden.c b/src/melee/gr/grgarden.c
index a940f7aedd..d7ebff969a 100644
--- a/src/melee/gr/grgarden.c
+++ b/src/melee/gr/grgarden.c
@@ -91,7 +91,7 @@ void grGarden_80202C24(void)
 
 void grGarden_80202C4C(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grGarden_80202C70(void)
diff --git a/src/melee/gr/grgreatbay.c b/src/melee/gr/grgreatbay.c
index 1e9e057aeb..2a7edb9d7c 100644
--- a/src/melee/gr/grgreatbay.c
+++ b/src/melee/gr/grgreatbay.c
@@ -103,7 +103,7 @@ void grGreatBay_801F42D0(void) {}
 
 void grGreatBay_801F42D4(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grGreatBay_801F42F8(void)
diff --git a/src/melee/gr/grgreens.c b/src/melee/gr/grgreens.c
index 168d4586b2..5a95160734 100644
--- a/src/melee/gr/grgreens.c
+++ b/src/melee/gr/grgreens.c
@@ -172,7 +172,7 @@ void grGreens_802134F4(void) {}
 
 void grGreens_802134F8(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grGreens_8021351C(void)
diff --git a/src/melee/gr/grheal.c b/src/melee/gr/grheal.c
index 5dcce2786c..a4dfe1b413 100644
--- a/src/melee/gr/grheal.c
+++ b/src/melee/gr/grheal.c
@@ -17,10 +17,12 @@
 #include "gr/ground.h"
 #include "gr/grzakogenerator.h"
 #include "gr/stage.h"
+#include "gm/gm_1832.h"
 #include "it/it_26B1.h"
 #include "it/types.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
+#include "gr/inlines.h"
 
 struct {
     int x0;
@@ -98,6 +100,7 @@ static char grHeal_803E851C[0x2C] = {
 };
 
 extern struct yaku* grHeal_804D6AF0;
+extern s16 grHeal_804D49D8[4];
 
 void grHeal_8021EF38(bool arg0) {}
 
@@ -119,7 +122,7 @@ void grHeal_8021EFBC(void) {}
 
 void grHeal_8021EFC0(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grHeal_8021EFE4(void)
@@ -159,7 +162,19 @@ Ground_GObj* grHeal_8021EFEC(u32 idx)
     return gobj;
 }
 
-/// #grHeal_8021F0D8
+void grHeal_8021F0D8(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    s32 i;
+
+    grAnime_801C8138(gobj, gp->map_id, 0);
+
+    for (i = 0; i < *(s32*) grHeal_804D6AF0; i++) {
+        if ((int) gm_80473A18.x90[i] != 0) {
+            grHeal_8021F79C(8, grHeal_804D49D8[i], i);
+        }
+    }
+}
 
 bool grHeal_8021F170(Ground_GObj* gobj)
 {
diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index af3be585a4..29626c007b 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -66,7 +66,7 @@ void grHomeRun_8021C7FC(void) {}
 
 void grHomeRun_8021C800(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grHomeRun_8021C824(void)
diff --git a/src/melee/gr/gricemt.c b/src/melee/gr/gricemt.c
index 95e36e5198..26bf2f0bde 100644
--- a/src/melee/gr/gricemt.c
+++ b/src/melee/gr/gricemt.c
@@ -3,13 +3,11 @@
 #include <placeholder.h>
 #include <platform.h>
 
-#include "baselib/gobjgxlink.h"
 #include "baselib/gobjproc.h"
 #include "baselib/random.h"
 #include "cm/camera.h"
 #include "ef/efsync.h"
 #include "gr/grdatfiles.h"
-#include "gr/grdisplay.h"
 #include "gr/grlib.h"
 #include "gr/grmaterial.h"
 #include "gr/ground.h"
@@ -273,12 +271,14 @@ void grIceMt_801F686C(void)
     Ground_801C3BB4();
 }
 
+static inline bool rand_zero(s32 n)
+{
+    return ((n) ? HSD_Randi(n) : 0) == 0;
+}
+
 void grIceMt_801F7080(void)
 {
-    // Ground_801C2BA4(5);
-    int iVar1;
     HSD_GObj* gobj;
-
     Vec3 x;
     if (grIm_804D69E8) {
         Ground_801C4A08(grIm_804D69E8);
@@ -289,48 +289,36 @@ void grIceMt_801F7080(void)
     if (grIm_804D69F0) {
         Ground_801C4A08(grIm_804D69F0);
     }
-    gobj = Ground_801C2BA4(1);
-    if (gobj) {
+    if ((gobj = Ground_801C2BA4(1))) {
         Ground_801C2FE0(gobj);
     }
-    gobj = Ground_801C2BA4(2);
-    if (gobj) {
+    if ((gobj = Ground_801C2BA4(2))) {
         Ground_801C2FE0(gobj);
     }
-    gobj = Ground_801C2BA4(3);
-    if (gobj) {
+    if ((gobj = Ground_801C2BA4(3))) {
         Ground_801C2FE0(gobj);
     }
-    gobj = Ground_801C2BA4(4);
-    if (gobj) {
+    if ((gobj = Ground_801C2BA4(4))) {
         Ground_801C2FE0(gobj);
     }
-    gobj = Ground_801C2BA4(5);
-    if (gobj) {
+    if ((gobj = Ground_801C2BA4(5))) {
         Ground_801C2FE0(gobj);
     }
-    gobj = Ground_801C2BA4(6);
-    if (gobj) {
+    if ((gobj = Ground_801C2BA4(6))) {
         Ground_801C2FE0(gobj);
     }
     if (Stage_80225194() == 76) {
         grZakoGenerator_801CAE04(&grIm_804D69F4->xBC);
-        if (grIm_804D69F4->xB8 != 0) {
-            iVar1 = HSD_Randi(grIm_804D69F4->xB8);
-        } else {
-            iVar1 = 0;
-        }
-        if (iVar1 == 0) {
-            grZakoGenerator_801CAEB0(Ground_801C5940(), Ground_801C5840());
+        if (rand_zero(grIm_804D69F4->xB8)) {
+            grZakoGenerator_801CAEB0(Ground_801C5840(), Ground_801C5940());
         }
 
         grZakoGenerator_801CADE0();
     } else {
-        grZakoGenerator_801CAE04(0);
+        grZakoGenerator_801CAE04(NULL);
     }
     x = grIm_803B8220[0];
-    grLib_801C96F8(0x7534, 0x1E, &x);
-    ////grLib_801C96F8(30000,30,{0,0,0});
+    grLib_801C96F8(0x7530, 0x1E, &x);
 }
 
 void grIceMt_801F71DC(void) {}
diff --git a/src/melee/gr/gricemt.h b/src/melee/gr/gricemt.h
index 0910cc2aa3..5280b864d3 100644
--- a/src/melee/gr/gricemt.h
+++ b/src/melee/gr/gricemt.h
@@ -5,6 +5,9 @@
 #include <platform.h>
 
 #include "gr/forward.h"
+
+#include "gr/types.h"
+
 #include "lb/forward.h"
 #include "mp/forward.h"
 #include <baselib/forward.h>
@@ -60,7 +63,8 @@ struct IceMountainParams {
     float xB0;
     float xB4;
     s16 xB8;
-    bool xBC;
+    s16 pad;
+    grZakoGenerator_SpawnDesc xBC;
     float xC0;
     float xC4;
     float xC8;
diff --git a/src/melee/gr/grinishie1.c b/src/melee/gr/grinishie1.c
index 68c9166219..85a80e2736 100644
--- a/src/melee/gr/grinishie1.c
+++ b/src/melee/gr/grinishie1.c
@@ -166,7 +166,7 @@ void grInishie1_801FA984(void) {}
 
 void grInishie1_801FA988(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grInishie1_801FA9AC(void)
diff --git a/src/melee/gr/grinishie2.c b/src/melee/gr/grinishie2.c
index 025dabf496..fd22d80ff0 100644
--- a/src/melee/gr/grinishie2.c
+++ b/src/melee/gr/grinishie2.c
@@ -217,7 +217,7 @@ void grInishie2_801FCCAC(void) {}
 
 void grInishie2_801FCCB0(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grInishie2_801FCCD4(void)
diff --git a/src/melee/gr/grizumi.c b/src/melee/gr/grizumi.c
index 869985ba54..ca779086ec 100644
--- a/src/melee/gr/grizumi.c
+++ b/src/melee/gr/grizumi.c
@@ -214,7 +214,7 @@ void grIzumi_OnLoad(void)
 
 void grIzumi_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grIzumi_801CBCE0(void)
diff --git a/src/melee/gr/grkinokoroute.c b/src/melee/gr/grkinokoroute.c
index 445e5eae20..bb3d188d13 100644
--- a/src/melee/gr/grkinokoroute.c
+++ b/src/melee/gr/grkinokoroute.c
@@ -24,7 +24,7 @@
 
 static struct {
     int x0;
-    int x4;
+    grZakoGenerator_SpawnDesc x4;
 }* grNKr_804D6A50;
 
 StageCallbacks grNKr_803E57F0[4] = {
diff --git a/src/melee/gr/grkongo.c b/src/melee/gr/grkongo.c
index 604d598974..acea9aba68 100644
--- a/src/melee/gr/grkongo.c
+++ b/src/melee/gr/grkongo.c
@@ -1292,7 +1292,7 @@ void grKongo_801D52FC(void)
 {
     HSD_GObj* gobj = Ground_801C2BA4(5);
     ftCo_800C0764(gobj, 2, fn_801D8134);
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grKongo_801D5338(void)
diff --git a/src/melee/gr/grkraid.c b/src/melee/gr/grkraid.c
index 03eb2ce05d..5535263504 100644
--- a/src/melee/gr/grkraid.c
+++ b/src/melee/gr/grkraid.c
@@ -77,7 +77,7 @@ void grKraid_OnLoad(void) {}
 
 void grKraid_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grKraid_801FE0BC(void)
diff --git a/src/melee/gr/grlast.c b/src/melee/gr/grlast.c
index 1c477539b5..03795b6849 100644
--- a/src/melee/gr/grlast.c
+++ b/src/melee/gr/grlast.c
@@ -183,7 +183,7 @@ static void grLast_OnLoad(void) {}
 
 static void grLast_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grLast_8021A7EC(void)
diff --git a/src/melee/gr/grmutecity.c b/src/melee/gr/grmutecity.c
index 3928d8b5a4..f9c117ae72 100644
--- a/src/melee/gr/grmutecity.c
+++ b/src/melee/gr/grmutecity.c
@@ -355,7 +355,7 @@ void grMuteCity_801EFCDC(void) {}
 
 void grMuteCity_801EFCE0(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grMuteCity_801EFD04(void)
diff --git a/src/melee/gr/groldkongo.c b/src/melee/gr/groldkongo.c
index 2c0cda3c35..fcddf0bcdf 100644
--- a/src/melee/gr/groldkongo.c
+++ b/src/melee/gr/groldkongo.c
@@ -20,6 +20,13 @@
 static struct {
     s16 x0;
     s16 x2;
+    u8 x4[0x10];
+    f32 x14;
+    f32 x18;
+    u8 x1C[0x2C];
+    f32 x48;
+    s32 x4C;
+    s32 x50;
 }* grOk_804D6A90;
 
 StageCallbacks grOk_803E658C[4] = {
@@ -35,7 +42,6 @@ StageCallbacks grOk_803E658C[4] = {
 
 void grOldKongo_8020F468(bool arg) {}
 
-/// #grOldKongo_8020F46C
 void grOldKongo_8020F46C(void)
 {
     grOk_804D6A90 = Ground_801C49F8();
@@ -56,7 +62,7 @@ void grOldKongo_8020F4E8(void)
     HSD_GObj* gobj;
     gobj = Ground_801C2BA4(1);
     ftCo_800C0764(gobj, 2, grOldKongo_80210454);
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grOldKongo_8020F524(void)
@@ -118,17 +124,6 @@ void grOldKongo_8020F6BC(Ground_GObj* gobj)
 
 void grOldKongo_8020F6E0(Ground_GObj* arg) {}
 
-/// #grOldKongo_8020F6E4
-
-bool grOldKongo_8020F880(Ground_GObj* gobj)
-{
-    return false;
-}
-
-/// #grOldKongo_8020F888
-
-void grOldKongo_80210058(Ground_GObj* arg) {}
-
 static inline int rand_inline(int a, int b)
 {
     if (a > b) {
@@ -140,6 +135,42 @@ static inline int rand_inline(int a, int b)
     }
 }
 
+void grOldKongo_8020F6E4(Ground_GObj* gobj)
+{
+    Vec3 sp14;
+    Ground* gp = GET_GROUND(gobj);
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+
+    grAnime_801C8138(gobj, gp->map_id, 0);
+    grMaterial_801C94D8(jobj);
+    gp->gv.oldkongo.xE8 = 1.0f;
+    gp->gv.oldkongo.xEC = 1.0f;
+    gp->gv.oldkongo.xD8 = 0.0f;
+    gp->gv.oldkongo.xDC = 0.0f;
+    gp->gv.oldkongo.xC4 = 0;
+    gp->gv.oldkongo.xC6 = 0;
+    gp->gv.oldkongo.xC8 = 2;
+    gp->gv.oldkongo.xD4 = NULL;
+    gp->gv.oldkongo.xE0 = 0.0f;
+    gp->gv.oldkongo.xE4 = 0.0f;
+    gp->gv.oldkongo.xEC = grOk_804D6A90->x48;
+    gp->gv.oldkongo.xCC =
+        rand_inline(grOk_804D6A90->x18, grOk_804D6A90->x14);
+    gp->gv.oldkongo.xCE =
+        rand_inline(grOk_804D6A90->x50, grOk_804D6A90->x4C);
+    lb_8000B1CC(Ground_801C3FA4(gobj, 1), NULL, &sp14);
+    Ground_801C4D70(gobj, &sp14, gp->gv.oldkongo.xDC);
+}
+
+bool grOldKongo_8020F880(Ground_GObj* gobj)
+{
+    return false;
+}
+
+/// #grOldKongo_8020F888
+
+void grOldKongo_80210058(Ground_GObj* arg) {}
+
 void grOldKongo_8021005C(Ground_GObj* gobj)
 {
     Ground* gp = GET_GROUND(gobj);
diff --git a/src/melee/gr/groldpupupu.c b/src/melee/gr/groldpupupu.c
index 1266ae3838..bd4f27d4db 100644
--- a/src/melee/gr/groldpupupu.c
+++ b/src/melee/gr/groldpupupu.c
@@ -15,6 +15,7 @@
 #include <baselib/gobj.h>
 #include <baselib/jobj.h>
 #include <baselib/random.h>
+#include "ft/ftdevice.h"
 
 static struct {
     s16 x0;
@@ -273,7 +274,41 @@ void grOldPupupu_80211194(Ground_GObj* arg) {}
 
 void grOldPupupu_80211198(Ground_GObj* arg) {}
 
-/// #grOldPupupu_8021119C
+void grOldPupupu_8021119C(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    s32 max;
+    s32 min;
+
+    ftCo_800C06E8(gobj, 0xA, fn_802112F4);
+    gp->gv.unk.xC4 = 0;
+    gp->gv.unk.xC8 = 0;
+    gp->gv.unk.xCC = 0;
+
+    max = grOp_804D6A98->xC;
+    min = grOp_804D6A98->x8;
+
+    if (max > min) {
+        s32 range = max - min;
+        if (range != 0) {
+            max = min + HSD_Randi(range);
+        } else {
+            max = min;
+        }
+    } else if (max < min) {
+        s32 range = min - max;
+        if (range != 0) {
+            max += HSD_Randi(range);
+        }
+    }
+
+    gp->gv.unk.xD0 = max;
+    gp->gv.unk.xD4 = 1;
+    gp->gv.unk.xE0 = 0;
+    gp->gv.unk.xD8 = 1;
+    grAnime_801C8138(gobj, gp->map_id, 0);
+    gp->x11_flags.b012 = 1;
+}
 
 bool grOldPupupu_80211284(Ground_GObj* gobj)
 {
diff --git a/src/melee/gr/groldyoshi.c b/src/melee/gr/groldyoshi.c
index 3a4c70bda7..bb6c5a7d3a 100644
--- a/src/melee/gr/groldyoshi.c
+++ b/src/melee/gr/groldyoshi.c
@@ -87,7 +87,7 @@ void grOldYoshi_8020E824(void) {}
 
 void grOldYoshi_8020E828(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grOldYoshi_8020E84C(void)
@@ -207,15 +207,14 @@ void grOldYoshi_8020EAF4(Ground_GObj* arg) {}
 void grOldYoshi_8020EAF8(Ground_GObj* arg) {}
 
 s16 grOy_803E6574[] = { 0, 1, 1, 5, 2, 9 };
-/// #grOldYoshi_8020EAFC
-void grOldYoshi_8020EAFC(Ground_GObj* arg)
+
+void grOldYoshi_8020EAFC(Ground_GObj* gobj)
 {
-    Ground* gp = GET_GROUND(arg);
-    HSD_JObj* jobj = arg->hsd_obj;
+    Ground* gp = GET_GROUND(gobj);
+    HSD_JObj* jobj = GET_JOBJ(gobj);
     int i;
-    PAD_STACK(4);
     Ground_801C2ED0(jobj, gp->map_id);
-    grAnime_801C8138(arg, gp->map_id, 0);
+    grAnime_801C8138(gobj, gp->map_id, 0);
     for (i = 0; i < 3; i++) {
         gp->gv.oldyoshicloud.cloud[i].xC4_0123 = 0;
         gp->gv.oldyoshicloud.cloud[i].xC4_4 = 0;
@@ -223,7 +222,7 @@ void grOldYoshi_8020EAFC(Ground_GObj* arg)
         gp->gv.oldyoshicloud.cloud[i].xD4 = 0.0f;
         gp->gv.oldyoshicloud.cloud[i].xD0 = 0.0f;
         gp->gv.oldyoshicloud.cloud[i].xC8 =
-            Ground_801C3FA4(arg, grOy_803E6574[i * 2 + 1]);
+            Ground_801C3FA4(gobj, grOy_803E6574[i * 2 + 1]);
         gp->gv.oldyoshicloud.cloud[i].xCC =
             HSD_JObjGetTranslationY(gp->gv.oldyoshicloud.cloud[i].xC8);
         mpJointSetCb1(grOy_803E6574[i * 2], gp, fn_8020F2A8);
diff --git a/src/melee/gr/gronett.c b/src/melee/gr/gronett.c
index 67af624e9e..4b61fe860a 100644
--- a/src/melee/gr/gronett.c
+++ b/src/melee/gr/gronett.c
@@ -38,6 +38,9 @@ static struct grOnett_StageParam {
     /* 0x00 */ float awning_initial;
     /* 0x04 */ u8 pad04[0x1C];
     /* 0x20 */ float awning_delta;
+    /* 0x24 */ u8 pad24[0x10];
+    /* 0x34 */ float x34;
+    /* 0x38 */ float x38;
 }* grOt_804D69C0;
 
 void grOnett_801E3734(bool arg) {}
@@ -195,7 +198,31 @@ bool grOnett_801E3D98(Ground_GObj* gobj)
 
 void grOnett_801E40E0(Ground_GObj* gobj) {}
 
-/// #grOnett_801E40E4
+void grOnett_801E40E4(Ground* gp, s32 arg1, CollData* cd, s32 arg3,
+                      mpLib_GroundEnum arg4, f32 arg5)
+{
+    PAD_STACK(8);
+
+    if (arg4 == 1) {
+        gp->gv.unk.xC8 += 1;
+        switch (gp->gv.kongo3.xC4) {
+        case 0:
+            if (gp->gv.unk.xC8 < grOt_804D69C0->x34) {
+                gp->gv.kongo3.xC6 = 1;
+                return;
+            }
+            gp->gv.kongo3.xC6 = 2;
+            return;
+        case 3:
+            if (gp->gv.unk.xC8 < grOt_804D69C0->x38) {
+                gp->gv.kongo3.xC6 = 4;
+                return;
+            }
+            gp->gv.kongo3.xC6 = 5;
+            break;
+        }
+    }
+}
 
 void grOnett_801E41B0(HSD_GObj* gobj)
 {
diff --git a/src/melee/gr/grpstadium.c b/src/melee/gr/grpstadium.c
index bf829ebadf..86426942f2 100644
--- a/src/melee/gr/grpstadium.c
+++ b/src/melee/gr/grpstadium.c
@@ -187,7 +187,7 @@ void grStadium_OnLoad(void) {}
 
 void grStadium_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grStadium_801D10F0(void)
diff --git a/src/melee/gr/grpura.c b/src/melee/gr/grpura.c
index 78cb8fe65b..cc612a4b44 100644
--- a/src/melee/gr/grpura.c
+++ b/src/melee/gr/grpura.c
@@ -128,7 +128,7 @@ void grPura_80211DD8(void) {}
 /// #grPura_80211DDC
 void grPura_80211DDC(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grPura_80211E00(void)
diff --git a/src/melee/gr/grpura.h b/src/melee/gr/grpura.h
index b0b43b288b..fe20fa72f7 100644
--- a/src/melee/gr/grpura.h
+++ b/src/melee/gr/grpura.h
@@ -39,7 +39,7 @@
 /* 2125F0 */ UNK_RET grPura_802125F0(HSD_GObj*);
 /* 212CD4 */ UNK_RET grPura_80212CD4(HSD_GObj*);
 /* 212EF4 */ UNK_RET grPura_80212EF4(HSD_GObj*);
-/* 212FC0 */ UNK_RET grPura_80212FC0(HSD_GObj*);
+/* 212FC0 */ void grPura_80212FC0(HSD_GObj*);
 /* 2130C0 */ DynamicsDesc* grPura_802130C0(enum_t);
 /* 2130C8 */ bool grPura_802130C8(Vec3* a, int, HSD_JObj*);
 /* 2130D0 */ UNK_RET fn_802130D0(HSD_GObj*, int);
diff --git a/src/melee/gr/grrcruise.c b/src/melee/gr/grrcruise.c
index 878b6a2743..4e7a9b0944 100644
--- a/src/melee/gr/grrcruise.c
+++ b/src/melee/gr/grrcruise.c
@@ -19,6 +19,7 @@
 
 #include "lb/lb_00F9.h"
 #include "mp/mplib.h"
+#include "sysdolphin/baselib/memory.h"
 
 #include <baselib/archive.h>
 #include <baselib/dobj.h>
@@ -43,6 +44,7 @@ StageCallbacks grRc_803E4E34[7] = {
       grRCruise_801FF920, 0 },
 };
 
+extern Vec3 grRc_803B8288;
 extern s16 grRc_803E4FF0[];
 extern s16 grRc_804D4790[4];
 
@@ -96,7 +98,7 @@ void grRCruise_801FF298(void) {}
 
 void grRCruise_801FF29C(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grRCruise_801FF2C0(void)
@@ -187,7 +189,31 @@ void grRCruise_801FF444(Ground_GObj* gobj)
     gp->gv.rcruise.x10 = 0;
 }
 
-/// #grRCruise_801FF5B4
+void grRCruise_801FF5B4(Ground_GObj* gobj)
+{
+    Vec3 pos;
+    Ground* gp = GET_GROUND(gobj);
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+
+    Ground_801C2ED0(jobj, gp->map_id);
+    grAnime_801C8138(gobj, gp->map_id, 0);
+    gp->gv.rcruise.x10 = 1;
+    gp->gv.rcruise.entries = HSD_MemAlloc(0x198);
+    if (gp->gv.rcruise.entries == NULL) {
+        __assert("grrcruise.c", 0x19A, "gp->u.map.chikuwa");
+    }
+    grRCruise_80201410(gobj);
+    Ground_801C10B8(gobj, grRCruise_801FF444);
+    grRCruise_80200540(gobj);
+    grRCruise_80200B48(gobj);
+    Ground_801C2FE0(gobj);
+    mpLib_80058560();
+    pos = grRc_803B8288;
+    gp->gv.rcruise.x04 = lb_80011A50(&pos, -1, 0.4f, 0.0f, 1.0471976f,
+                                     -10000.0f, 10000.0f, 10000.0f, -10000.0f);
+    gp->gv.rcruise.x08 = 0.017453292f;
+    gp->gv.rcruise.x0C = 1;
+}
 
 bool grRCruise_801FF6CC(Ground_GObj* arg)
 {
@@ -321,7 +347,31 @@ bool grRCruise_8020014C(Ground_GObj* arg)
 
 void grRCruise_8020045C(Ground_GObj* arg) {}
 
-/// #fn_80200460
+void fn_80200460(Ground* gp_arg, s32 joint_id, CollData* cd, s32 arg3,
+                 mpLib_GroundEnum arg4, f32 arg5)
+{
+    HSD_GObj* gobj = (HSD_GObj*) gp_arg;
+    Ground* gp = HSD_GObjGetUserData(gobj);
+    int i;
+    PAD_STACK(16);
+
+    if ((s32) cd->x34_flags.b1234 == 1) {
+        for (i = 0; i < 3; i++) {
+            if (gp->gv.rcruise.x3C[i].x02 == joint_id) {
+                u8 state = gp->gv.rcruise.x3C[i].x00;
+                if (state == 1 || (u8)(state - 3) <= 1u) {
+                    gp->gv.rcruise.x3C[i].x04 = 0;
+                    grRCruise_80201B60(gp->gv.rcruise.x3C[i].x0C->child,
+                                       1);
+                    grAnime_801C7A94(gobj, grRc_804D4790[i], 1, 1.0f);
+                    gp->gv.rcruise.x3C[i].x00 = 2;
+                }
+                gp->gv.rcruise.x3C[i].x08++;
+                return;
+            }
+        }
+    }
+}
 
 void grRCruise_80200540(Ground_GObj* gobj)
 {
@@ -349,8 +399,7 @@ void grRCruise_80200B48(Ground_GObj* gobj)
 
     for (i = 0; i < 17; i++) {
         struct grRCruise_Entry* entry = &gp->gv.rcruise.entries[i];
-        entry->x02 =
-            Ground_801C32D4(gp->map_id, grRc_803E4FF0[i]);
+        entry->x02 = Ground_801C32D4(gp->map_id, grRc_803E4FF0[i]);
         entry->x14 = Ground_801C3FA4(gobj, grRc_803E4FF0[i]);
         entry->x08 = 0;
         entry->x04 = 0;
diff --git a/src/melee/gr/grrcruise.h b/src/melee/gr/grrcruise.h
index 6965344361..8699aeb563 100644
--- a/src/melee/gr/grrcruise.h
+++ b/src/melee/gr/grrcruise.h
@@ -42,7 +42,8 @@
 /* 20014C */ bool grRCruise_8020014C(Ground_GObj*);
 /* 200154 */ void grRCruise_80200154(Ground_GObj*);
 /* 20045C */ void grRCruise_8020045C(Ground_GObj*);
-/* 200460 */ UNK_RET fn_80200460(UNK_PARAMS);
+/* 200460 */ void fn_80200460(Ground*, s32, CollData*, s32, mpLib_GroundEnum,
+                              f32);
 /* 200540 */ void grRCruise_80200540(Ground_GObj*);
 /* 200578 */ void grRCruise_80200578(Ground*, s32, CollData*, s32,
                                      mpLib_GroundEnum, f32);
@@ -52,7 +53,7 @@
 /* 2010A4 */ void grRCruise_802010A4(Ground_GObj*, s32, CollData*);
 /* 201110 */ void grRCruise_80201110(Ground_GObj*);
 /* 201288 */ UNK_RET grRCruise_80201288(UNK_PARAMS);
-/* 201410 */ UNK_RET grRCruise_80201410(UNK_PARAMS);
+/* 201410 */ void grRCruise_80201410(Ground_GObj*);
 /* 201588 */ void grRCruise_80201588(Ground_GObj*);
 /* 201918 */ void grRCruise_80201918(Vec3*);
 /* 201988 */ UNK_RET grRCruise_80201988(UNK_PARAMS);
diff --git a/src/melee/gr/grshrine.c b/src/melee/gr/grshrine.c
index f56aaee39b..f98966d77c 100644
--- a/src/melee/gr/grshrine.c
+++ b/src/melee/gr/grshrine.c
@@ -97,7 +97,7 @@ static void grShrine_UnkStage0_OnLoad(void) {}
 
 static void grShrine_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grShrine_80201D18(void)
diff --git a/src/melee/gr/grshrineroute.c b/src/melee/gr/grshrineroute.c
index 53b7b8c0de..40affbd2b7 100644
--- a/src/melee/gr/grshrineroute.c
+++ b/src/melee/gr/grshrineroute.c
@@ -8,6 +8,7 @@
 #include "baselib/forward.h"
 
 #include "baselib/jobj.h"
+#include "ef/efsync.h"
 #include "ft/ftlib.h"
 #include "gr/grdisplay.h"
 #include "gr/grmaterial.h"
@@ -102,12 +103,12 @@ static struct {
     int x8;
     int xC;
     int x10;
-    int x14;
-    int x18;
+    f32 x14;
+    f32 x18;
     int x1C;
     int x20;
     int x24;
-    int x28;
+    grZakoGenerator_SpawnDesc x28;
 }* grSh_Route_804D6A58;
 
 extern float grNKr_804DB868;
diff --git a/src/melee/gr/grshrineroute.h b/src/melee/gr/grshrineroute.h
index 823daf04c7..030049283b 100644
--- a/src/melee/gr/grshrineroute.h
+++ b/src/melee/gr/grshrineroute.h
@@ -49,7 +49,7 @@
 /* 20AD24 */ bool grShrineRoute_8020AD24(bool);
 /* 20AD58 */ void grShrineRoute_8020AD58(Ground*, int, CollData*, int, int);
 /* 20AE08 */ UNK_RET grShrineRoute_8020AE08(UNK_PARAMS);
-/* 20AF38 */ UNK_RET grShrineRoute_8020AF38(UNK_PARAMS);
+/* 20AF38 */ void grShrineRoute_8020AF38(HSD_GObj*, s32);
 /* 20B020 */ void grShrineRoute_8020B020(HSD_GObj* gobj, int r4, bool hide);
 /* 20B0AC */ void grShrineRoute_8020B0AC(void);
 /* 20B104 */ DynamicsDesc* grShrineRoute_OnTouchLine(enum_t);
diff --git a/src/melee/gr/grstory.c b/src/melee/gr/grstory.c
index 7951db7911..d4ba084f32 100644
--- a/src/melee/gr/grstory.c
+++ b/src/melee/gr/grstory.c
@@ -78,7 +78,7 @@ void grStory_UnkStage0_OnLoad(void) {}
 
 void grStory_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grStory_801E30D0(void)
diff --git a/src/melee/gr/grtcaptain.c b/src/melee/gr/grtcaptain.c
index 82028db39b..73c2176943 100644
--- a/src/melee/gr/grtcaptain.c
+++ b/src/melee/gr/grtcaptain.c
@@ -86,7 +86,7 @@ static void grTCaptain_OnLoad(void) {}
 
 static void grTCaptain_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTCaptain_8021FCFC(void)
diff --git a/src/melee/gr/grtclink.c b/src/melee/gr/grtclink.c
index b5eb9ae81f..13d7c2e81c 100644
--- a/src/melee/gr/grtclink.c
+++ b/src/melee/gr/grtclink.c
@@ -84,7 +84,7 @@ void grTclink_UnkStage0_OnLoad(void)
 
 void grTclink_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTCLink_8021FFE0(void)
diff --git a/src/melee/gr/grtdonkey.c b/src/melee/gr/grtdonkey.c
index 004a08e02e..8c25afa65b 100644
--- a/src/melee/gr/grtdonkey.c
+++ b/src/melee/gr/grtdonkey.c
@@ -88,7 +88,7 @@ static void grTdonkey_UnkStage0_OnLoad(void) {}
 
 static void grTdonkey_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTDonkey_802202C4(void)
diff --git a/src/melee/gr/grtdrmario.c b/src/melee/gr/grtdrmario.c
index 8a47d96bb7..4d5fb97079 100644
--- a/src/melee/gr/grtdrmario.c
+++ b/src/melee/gr/grtdrmario.c
@@ -83,7 +83,7 @@ void grTdrmario_UnkStage0_OnLoad(void) {}
 
 void grTdrmario_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grtDrMario_802205A8(void)
diff --git a/src/melee/gr/grtemblem.c b/src/melee/gr/grtemblem.c
index 277417fedb..080fc1c948 100644
--- a/src/melee/gr/grtemblem.c
+++ b/src/melee/gr/grtemblem.c
@@ -98,7 +98,7 @@ static void grTemblem_UnkStage0_OnLoad(void) {}
 
 static void grTemblem_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTRoy_80224490(void)
diff --git a/src/melee/gr/grtest.c b/src/melee/gr/grtest.c
index ab27f9dc91..fecf76879d 100644
--- a/src/melee/gr/grtest.c
+++ b/src/melee/gr/grtest.c
@@ -78,7 +78,7 @@ void grTest_UnkStage0_OnLoad(void) {}
 
 void grTest_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTest_8020703C(void)
diff --git a/src/melee/gr/grtfalco.c b/src/melee/gr/grtfalco.c
index f6b23eaf8e..28f09de730 100644
--- a/src/melee/gr/grtfalco.c
+++ b/src/melee/gr/grtfalco.c
@@ -86,7 +86,7 @@ void grTfalco_UnkStage0_OnLoad(void) {}
 
 void grTfalco_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTFalco_80220894(void)
diff --git a/src/melee/gr/grtfox.c b/src/melee/gr/grtfox.c
index 61c86d2547..8e380bf234 100644
--- a/src/melee/gr/grtfox.c
+++ b/src/melee/gr/grtfox.c
@@ -98,7 +98,7 @@ void grTfox_UnkStage0_OnLoad(void) {}
 
 void grTfox_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTFox_80220C24(void)
diff --git a/src/melee/gr/grtgamewatch.c b/src/melee/gr/grtgamewatch.c
index 787e1be2d2..ce74b4c18b 100644
--- a/src/melee/gr/grtgamewatch.c
+++ b/src/melee/gr/grtgamewatch.c
@@ -64,7 +64,7 @@ void grTgamewatch_UnkStage0_OnLoad(void)
 
 void grTgamewatch_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTGameWatch_802241AC(void)
diff --git a/src/melee/gr/grtganon.c b/src/melee/gr/grtganon.c
index 95dee28e6d..dea1c65d27 100644
--- a/src/melee/gr/grtganon.c
+++ b/src/melee/gr/grtganon.c
@@ -76,7 +76,7 @@ void grTganon_UnkStage0_OnLoad(void)
 
 void grTganon_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTGanon_8022477C(void)
diff --git a/src/melee/gr/grticeclimber.c b/src/melee/gr/grticeclimber.c
index 2628af9bb1..ce49b8e960 100644
--- a/src/melee/gr/grticeclimber.c
+++ b/src/melee/gr/grticeclimber.c
@@ -80,7 +80,7 @@ void grTiceclimber_UnkStage0_OnLoad(void) {}
 
 void grTiceclimber_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTIceClimber_80220FAC(void)
diff --git a/src/melee/gr/grtkirby.c b/src/melee/gr/grtkirby.c
index 3b1b0bcb60..4445e1db3a 100644
--- a/src/melee/gr/grtkirby.c
+++ b/src/melee/gr/grtkirby.c
@@ -84,7 +84,7 @@ void grTkirby_UnkStage0_OnLoad(void) {}
 
 void grTkirby_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTKirby_80221400(void)
diff --git a/src/melee/gr/grtkoopa.c b/src/melee/gr/grtkoopa.c
index 85dd710bdf..a9d679cc6a 100644
--- a/src/melee/gr/grtkoopa.c
+++ b/src/melee/gr/grtkoopa.c
@@ -86,7 +86,7 @@ static void grTkoopa_UnkStage0_OnLoad(void) {}
 
 static void grTkoopa_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTKoopa_802216E4(void)
diff --git a/src/melee/gr/grtlink.c b/src/melee/gr/grtlink.c
index 5f5dbd1692..ca12213ac6 100644
--- a/src/melee/gr/grtlink.c
+++ b/src/melee/gr/grtlink.c
@@ -85,7 +85,7 @@ void grTlink_UnkStage0_OnLoad(void) {}
 
 void grTlink_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTLink_802219C8(void)
diff --git a/src/melee/gr/grtluigi.c b/src/melee/gr/grtluigi.c
index 7f371cba51..c65dc057ba 100644
--- a/src/melee/gr/grtluigi.c
+++ b/src/melee/gr/grtluigi.c
@@ -70,7 +70,7 @@ void grTluigi_UnkStage0_OnLoad(void) {}
 
 void grTluigi_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTLuigi_80221CAC(void)
diff --git a/src/melee/gr/grtmario.c b/src/melee/gr/grtmario.c
index caf641c5ac..c43b7685d7 100644
--- a/src/melee/gr/grtmario.c
+++ b/src/melee/gr/grtmario.c
@@ -106,7 +106,7 @@ void grTmario_UnkStage0_OnLoad(void) {}
 
 void grTmario_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTMario_8021F8DC(void)
diff --git a/src/melee/gr/grtmars.c b/src/melee/gr/grtmars.c
index 4de67fe58a..7db05dd292 100644
--- a/src/melee/gr/grtmars.c
+++ b/src/melee/gr/grtmars.c
@@ -98,7 +98,7 @@ static void grTmars_UnkStage0_OnLoad(void)
 
 static void grTmars_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTMars_80221F90(void)
diff --git a/src/melee/gr/grtmewtwo.c b/src/melee/gr/grtmewtwo.c
index a8e1364baf..b0f15d6c6e 100644
--- a/src/melee/gr/grtmewtwo.c
+++ b/src/melee/gr/grtmewtwo.c
@@ -124,7 +124,7 @@ void grTmewtwo_UnkStage0_OnLoad(void) {}
 
 void grTmewtwo_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTMewtwo_8022227C(void)
diff --git a/src/melee/gr/grtness.c b/src/melee/gr/grtness.c
index 33e3cfad4d..fb44376e57 100644
--- a/src/melee/gr/grtness.c
+++ b/src/melee/gr/grtness.c
@@ -100,7 +100,7 @@ static void grTness_UnkStage0_OnLoad(void)
 
 static void grTness_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTNess_8022266C(void)
diff --git a/src/melee/gr/grtpeach.c b/src/melee/gr/grtpeach.c
index 9b0898a457..d87d77a701 100644
--- a/src/melee/gr/grtpeach.c
+++ b/src/melee/gr/grtpeach.c
@@ -108,7 +108,7 @@ void grTpeach_UnkStage0_OnLoad(void) {}
 
 void grTpeach_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTPeach_80222950(void)
diff --git a/src/melee/gr/grtpichu.c b/src/melee/gr/grtpichu.c
index b430b09095..ac6ee05fe5 100644
--- a/src/melee/gr/grtpichu.c
+++ b/src/melee/gr/grtpichu.c
@@ -101,7 +101,7 @@ static void grTpichu_UnkStage0_OnLoad(void) {}
 
 static void grTpichu_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTPichu_80222C34(void)
diff --git a/src/melee/gr/grtpikachu.c b/src/melee/gr/grtpikachu.c
index f8f0e816df..e245205ce7 100644
--- a/src/melee/gr/grtpikachu.c
+++ b/src/melee/gr/grtpikachu.c
@@ -63,7 +63,7 @@ void grTpikachu_UnkStage0_OnLoad(void)
 
 void grTpikachu_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTPikachu_80222F18(void)
diff --git a/src/melee/gr/grtpurin.c b/src/melee/gr/grtpurin.c
index 910f25cb03..c6956a7bce 100644
--- a/src/melee/gr/grtpurin.c
+++ b/src/melee/gr/grtpurin.c
@@ -73,7 +73,7 @@ void grTpurin_UnkStage0_OnLoad(void)
 
 void grTpurin_UnkStage0_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTPurin_80223204(void)
diff --git a/src/melee/gr/grtsamus.c b/src/melee/gr/grtsamus.c
index 60a81a0ba8..2b3b827974 100644
--- a/src/melee/gr/grtsamus.c
+++ b/src/melee/gr/grtsamus.c
@@ -98,7 +98,7 @@ void grTSamus_OnLoad(void) {}
 
 void grTSamus_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTSamus_8022361C(void)
diff --git a/src/melee/gr/grtseak.c b/src/melee/gr/grtseak.c
index b4a6498a66..33e8df5871 100644
--- a/src/melee/gr/grtseak.c
+++ b/src/melee/gr/grtseak.c
@@ -98,7 +98,7 @@ void grTseak_OnLoad(void) {}
 
 void grTseak_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTSeak_80223900(void)
diff --git a/src/melee/gr/grtyoshi.c b/src/melee/gr/grtyoshi.c
index 76408747f6..191871603a 100644
--- a/src/melee/gr/grtyoshi.c
+++ b/src/melee/gr/grtyoshi.c
@@ -99,7 +99,7 @@ static void grTYoshi_OnLoad(void) {}
 
 static void grTYoshi_OnStart(void)
 {
-    grZakoGenerator_801CAE04(false);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 static bool grTYoshi_80223BE4(void)
diff --git a/src/melee/gr/grtzelda.c b/src/melee/gr/grtzelda.c
index 452ac8f803..2f826587e0 100644
--- a/src/melee/gr/grtzelda.c
+++ b/src/melee/gr/grtzelda.c
@@ -85,7 +85,7 @@ void grTZelda_OnLoad(void) {}
 
 void grTZelda_OnStart(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grTZelda_80223EC8(void)
diff --git a/src/melee/gr/grvenom.c b/src/melee/gr/grvenom.c
index a37ed2300f..daf601f699 100644
--- a/src/melee/gr/grvenom.c
+++ b/src/melee/gr/grvenom.c
@@ -215,7 +215,7 @@ void grVenom_80203DD0(void)
 
 void grVenom_80203E80(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grVenom_80203EA4(void)
diff --git a/src/melee/gr/gryorster.c b/src/melee/gr/gryorster.c
index 703e4c3315..58b8269041 100644
--- a/src/melee/gr/gryorster.c
+++ b/src/melee/gr/gryorster.c
@@ -73,7 +73,7 @@ void grYorster_8020200C(void)
 
 void grYorster_80202010(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grYorster_80202034(void)
diff --git a/src/melee/gr/grzakogenerator.c b/src/melee/gr/grzakogenerator.c
index 0f15173fa6..976015b01f 100644
--- a/src/melee/gr/grzakogenerator.c
+++ b/src/melee/gr/grzakogenerator.c
@@ -2,85 +2,340 @@
 
 #include "grzakogenerator.static.h"
 
+#include "placeholder.h"
+
+#include "baselib/forward.h"
+
+#include "gr/grlib.h"
+#include "gr/ground.h"
+#include "gr/stage.h"
+#include "gr/types.h"
+#include "it/inlines.h"
 #include "it/it_26B1.h"
+#include "it/it_2725.h"
+#include "it/item.h"
+#include "it/items/itleadead.h"
+#include "it/items/itlikelike.h"
+#include "it/items/itnokonoko.h"
+#include "it/items/itpatapata.h"
+#include "it/types.h"
+#include "lb/lb_00B0.h"
 
 #include <baselib/gobj.h>
+#include <baselib/gobjproc.h>
 #include <baselib/memory.h>
+#include <baselib/psstructs.h>
+#include <sysdolphin/baselib/random.h>
 
-int grZakoGenerator_801CA394(void* arg0, int arg1, void* arg2, f32 arg3)
+grZakoGenerator_Config*
+grZakoGenerator_801CA394(grZakoGenerator_Spawn* spawn_descs, int count,
+                         grZakoGenerator_SpawnFunc callback, f32 arg4)
 {
-    u8* result;
-    void* arr;
-
-    result = (u8*) HSD_MemAlloc(0x1C);
-    if (result != NULL) {
-        arr = HSD_MemAlloc(arg1 * 0x18);
-        *(void**) (result + 0x04) = arr;
-        if (arr != NULL) {
-            *(void**) (result + 0x00) = arg0;
-            *(int*) (result + 0x08) = arg1;
-            *(void**) (result + 0x0C) = arg2;
-            *(f32*) (result + 0x10) = arg3;
-            *(int*) (result + 0x14) = 0;
-            *(int*) (result + 0x18) = 0;
+    grZakoGenerator_Config* config =
+        HSD_MemAlloc(sizeof(grZakoGenerator_Config));
+    if (config != NULL) {
+        grZakoGenerator_Spawn* spawns =
+            HSD_MemAlloc(count * sizeof(grZakoGenerator_Spawn));
+        config->spawns = spawns;
+        if (spawns != NULL) {
+            config->spawn_descs = spawn_descs;
+            config->count = count;
+            config->callback = callback;
+            config->x10 = arg4;
+            config->gen = NULL;
+            config->x18 = 0;
         } else {
-            HSD_Free(result);
-            result = NULL;
+            HSD_Free(config);
+            config = NULL;
+        }
+    }
+    return config;
+}
+
+static inline void
+grZakoGenerator_801CA43C_inline(f32 t, grZakoGenerator_Spawn* spawns,
+                                f32 height, Vec3* pos)
+{
+    pos->x = t * (spawns->pos1.x - spawns->pos0.x) + spawns->pos0.x;
+    pos->y = height,
+    pos->z = t * (spawns->pos1.z - spawns->pos0.z) + spawns->pos0.z;
+}
+
+s32 grZakoGenerator_801CA43C(grZakoGenerator_Config* config, HSD_JObj* jobj,
+                             f32 height)
+{
+    Vec3 pos;
+    s32 spawned = 0;
+    f32 t;
+    int i;
+    grZakoGenerator_Spawn* descs;
+    grZakoGenerator_Spawn* spawns;
+    PAD_STACK(4);
+
+    if (config == NULL || jobj == NULL) {
+        return 0;
+    }
+
+    descs = config->spawn_descs;
+    spawns = config->spawns;
+    for (i = 0; i < config->count; i++, descs++, spawns++) {
+        lb_8000B1CC(jobj, &descs->pos0, &spawns->pos0);
+        lb_8000B1CC(jobj, &descs->pos1, &spawns->pos1);
+    }
+
+    spawns = config->spawns;
+    for (i = 0; i < config->count; i++, spawns++) {
+        if (!((spawns->pos0.y > height && spawns->pos1.y > height) ||
+              (spawns->pos0.y < height && spawns->pos1.y < height)))
+        {
+            if (config->x10 < 0.0f || HSD_Randf() < config->x10) {
+                f32 dy = spawns->pos1.y - spawns->pos0.y;
+                if (ABS(dy) > 0.0001f) {
+                    t = (height - spawns->pos0.y) / dy;
+                    grZakoGenerator_801CA43C_inline(t, spawns, height, &pos);
+                } else {
+                    grZakoGenerator_801CA43C_inline(HSD_Randf(), spawns,
+                                                    height, &pos);
+                }
+
+                if (config->x10 < 0.0f) {
+                    HSD_Generator* gen = config->gen;
+                    if (gen == NULL) {
+                        config->gen = config->callback(&pos, i);
+                    } else {
+                        gen->pos.x = pos.x;
+                        gen->pos.y = pos.y;
+                        gen->pos.z = pos.z;
+                    }
+                } else {
+                    config->callback(&pos, i);
+                }
+                spawned++;
+            }
+        } else if (config->x10 < 0.0f) {
+            if (config->gen != NULL) {
+                grLib_801C9874(config->gen);
+                config->gen = NULL;
+            }
         }
     }
-    return (int) result;
+    return spawned;
+}
+
+grZakoGenerator_Data* grZakoGenerator_801CA67C(void)
+{
+    grZakoGenerator_Data* data = HSD_MemAlloc(sizeof(grZakoGenerator_Data));
+    int i;
+    PAD_STACK(8);
+
+    if (data == NULL) {
+        __assert("grzakogenerator.c", 0x52, "pointp");
+    }
+
+    for (i = 0; i < 80; i++) {
+        data->entries[i].x0 = i + 0x20;
+        data->entries[i].x2 = 0;
+        data->entries[i].x4 = 0;
+    }
+    // NOTE: this actually intializes data->sentinel
+    for (; i < 81; i++) {
+        data->entries[i].x0 = -1;
+        data->entries[i].x2 = 0;
+        data->entries[i].x4 = 0;
+    }
+
+    lbl_8049F030.xA_b0 = true;
+    lbl_8049F030.x8 = 0;
+    return data;
 }
 
-/// #grZakoGenerator_801CA43C
+void grZakoGenerator_801CA8B4(int arg0)
+{
+    if (lbl_8049F030.x4 != NULL && lbl_8049F030.xA_b0) {
+        f32 right = Stage_GetBlastZoneRightOffset();
+        f32 left = Stage_GetBlastZoneLeftOffset();
+        f32 top = Stage_GetBlastZoneTopOffset();
+        f32 bottom = Stage_GetBlastZoneBottomOffset();
+        f32 cam_right = 0.5f * (right + Stage_GetCamBoundsRightOffset());
+        f32 cam_left = 0.5f * (left + Stage_GetCamBoundsLeftOffset());
+        f32 cam_top = 0.5f * (top + Stage_GetCamBoundsTopOffset());
+        f32 cam_bottom = 0.5f * (bottom + Stage_GetCamBoundsBottomOffset());
+        int start = lbl_8049F030.x8;
+        int idx = start;
+        int max_iter;
+        Vec3 pos;
+
+        for (max_iter = 0xFF; max_iter > 0; max_iter--) {
+            grZakoGenerator_Entry* entry = &lbl_8049F030.x4->entries[idx];
 
-/// #grZakoGenerator_801CA67C
+            if (entry->x0 != -1 && entry->x4 == 0) {
+                if (lbl_8049F030.x4->entries[idx].x2 != -1) {
+                    if (lbl_8049F030.x4->entries[idx].x2 > 0) {
+                        lbl_8049F030.x4->entries[idx].x2--;
+                    } else if (Ground_801C2D24(entry->x0, &pos)) {
+                        if (left < pos.x && pos.x < right && bottom < pos.y &&
+                            pos.y < top &&
+                            (arg0 != 0 || !(cam_left < pos.x) ||
+                             !(pos.x < cam_right) || !(cam_bottom < pos.y) ||
+                             !(pos.y < cam_top)))
+                        {
+                            Item_GObj* gobj = NULL;
+                            grZakoGenerator_Entry* e2 =
+                                &lbl_8049F030.x4->entries[idx];
 
-/// #grZakoGenerator_801CA8B4
+                            s16 k = e2->x0;
+                            if (k >= 0xDC && k <= 0xFB) {
+                                if (e2->x8 != -1) {
+                                    gobj = Ground_801C58E0(e2->x8, k);
+                                    if (gobj != NULL) {
+                                        lbl_8049F030.x4->entries[idx].x4 =
+                                            gobj;
+                                    }
+                                }
+                            } else if (lbl_8049F030.x0 != NULL) {
+                                s32 desc_idx = k - 0x20;
+                                switch (lbl_8049F030.x0[desc_idx].kind) {
+                                case 0x2B:
+                                    gobj = it_8027B5B0(It_Kind_Kuriboh, &pos,
+                                                       NULL, NULL, 1);
+                                    break;
+                                case 0x2C:
+                                    gobj = it_802EA9FC(&pos, -1);
+                                    break;
+                                case 0x2D:
+                                    gobj = it_8027B5B0(It_Kind_Octarock, &pos,
+                                                       NULL, NULL, 1);
+                                    break;
+                                case 0x2E:
+                                    gobj = it_8027B5B0(It_Kind_Ottosea, &pos,
+                                                       NULL, NULL, 1);
+                                    break;
+                                case 0xD3:
+                                    gobj = it_802DD7F0(
+                                        lbl_8049F030.x0[desc_idx].x2, &pos,
+                                        NULL, -1);
+                                    break;
+                                case 0xD4:
+                                    gobj = it_802E16F8(
+                                        lbl_8049F030.x0[desc_idx].x2, &pos,
+                                        -1);
+                                    break;
+                                case 0xD5:
+                                    gobj = it_802DC4BC(
+                                        lbl_8049F030.x0[desc_idx].x2, &pos,
+                                        -1);
+                                    break;
+                                case 0xD9:
+                                    gobj = it_8027B5B0(It_Kind_Whitebea, &pos,
+                                                       NULL, NULL, 1);
+                                    break;
+                                }
+                                if (gobj != NULL) {
+                                    Item* ip = GET_ITEM(gobj);
+                                    lbl_8049F030.x4->entries[idx].x4 = gobj;
+                                    ip->xDD4_itemVar.zako.idx = idx;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            idx++;
+            if (idx >= 0x51) {
+                idx = 0;
+            }
+            if (idx == start) {
+                break;
+            }
+        }
+
+        lbl_8049F030.x8 = idx;
+    }
+}
 
 void grZakoGenerator_801CAC14(HSD_GObj* gobj)
 {
-    void* ip = gobj->user_data;
+    Item* ip = GET_ITEM(gobj);
     s32 kind = itGetKind(gobj);
 
     if (kind == 0x9F) {
-        struct grZakoGenerator_Data* data = lbl_8049F030.x4;
-        if (data->x3C4 == gobj) {
-            data->x3C4 = NULL;
+        grZakoGenerator_Data* data = lbl_8049F030.x4;
+        if (data->sentinel.x4 == gobj) {
+            data->sentinel.x4 = NULL;
         }
     } else {
-        s32 idx = *(s32*) ((u8*) ip + 0xDD8);
+        s32 idx = ip->xDD4_itemVar.zako.idx;
         if (idx != -1) {
-            lbl_8049F030.x4->entries[idx].x4 = 0;
+            lbl_8049F030.x4->entries[idx].x4 = NULL;
             lbl_8049F030.x4->entries[idx].x2 = 2;
         }
     }
 }
 
-/// #grZakoGenerator_801CACB8
+void grZakoGenerator_801CACB8(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    s32 kind = itGetKind(gobj);
+
+    it_8027CE18(gobj);
 
-s32 fn_801CADBC(void)
+    if (kind == 0x9F) {
+        if (lbl_8049F030.x4->sentinel.x4 == gobj) {
+            lbl_8049F030.x4->sentinel.x0 = -1;
+            lbl_8049F030.x4->sentinel.x4 = NULL;
+        }
+    } else {
+        s32 idx = ip->xDD4_itemVar.zako.idx;
+        if (idx != -1) {
+            int off = lbl_8049F030.x4->entries[idx].x0 - 0x20;
+            lbl_8049F030.x4->entries[idx].x4 = NULL;
+            if (lbl_8049F030.x0[off].respawn == 1) {
+                lbl_8049F030.x4->entries[idx].x2 = 0x708;
+            } else {
+                lbl_8049F030.x4->entries[idx].x2 = -1;
+            }
+            ip->xDD4_itemVar.zako.idx = -1;
+        }
+    }
+}
+
+void fn_801CADBC(HSD_GObj* gobj)
 {
-    return grZakoGenerator_801CA8B4(0);
+    grZakoGenerator_801CA8B4(0);
 }
 
-s32 grZakoGenerator_801CADE0(void)
+void grZakoGenerator_801CADE0(void)
 {
-    return grZakoGenerator_801CA8B4(1);
+    grZakoGenerator_801CA8B4(1);
 }
 
-/// #grZakoGenerator_801CAE04
+HSD_GObj* grZakoGenerator_801CAE04(grZakoGenerator_SpawnDesc* arg0)
+{
+    HSD_GObj* gobj;
+    grZakoGenerator_Data* data = grZakoGenerator_801CA67C();
+    gobj = GObj_Create(2, 4, 0);
+
+    if (gobj == NULL) {
+        OSReport("%s:%d: couldn t get gobj!\n", "grzakogenerator.c", 0x166);
+        HSD_Free(data);
+        return NULL;
+    }
+
+    HSD_GObj_SetupProc(gobj, fn_801CADBC, 0);
+    lbl_8049F030.x0 = arg0;
+    lbl_8049F030.x4 = data;
+    return gobj;
+}
 
 void grZakoGenerator_801CAEB0(int arg0, int arg1)
 {
-    struct grZakoGenerator_Data* data = lbl_8049F030.x4;
-    s16 val = (s16) arg1;
-
-    if (data->x3C0 != -1) {
-        return;
+    s16 val = arg1;
+    if (lbl_8049F030.x4->sentinel.x0 == -1) {
+        lbl_8049F030.x4->sentinel.x0 = val;
+        lbl_8049F030.x4->sentinel.x8 = arg0;
+        lbl_8049F030.x4->sentinel.x4 = NULL;
     }
-    data->x3C0 = val;
-    lbl_8049F030.x4->x3C8 = arg0;
-    lbl_8049F030.x4->x3C4 = NULL;
 }
 
 void grZakoGenerator_801CAEF0(bool arg0)
@@ -88,4 +343,28 @@ void grZakoGenerator_801CAEF0(bool arg0)
     lbl_8049F030.xA_b0 = arg0;
 }
 
-/// #grZakoGenerator_801CAF08
+void grZakoGenerator_801CAF08(void)
+{
+    HSD_GObj* next;
+    HSD_GObj* cur;
+    PAD_STACK(8);
+
+    lbl_8049F030.xA_b0 = false;
+    cur = HSD_GObj_Entities->items;
+
+    while (cur != NULL) {
+        next = cur->next;
+        {
+            s32 kind = itGetKind(cur);
+            if ((kind >= 0x2B && kind < 0x30) || (kind >= 0xD0 && kind < 0xE8))
+            {
+                if (it_8026C1E8(cur)) {
+                    Item_8026A8EC(cur);
+                }
+            } else if (kind == 0x9F) {
+                Item_8026A8EC(cur);
+            }
+        }
+        cur = next;
+    }
+}
diff --git a/src/melee/gr/grzakogenerator.h b/src/melee/gr/grzakogenerator.h
index aa70bf1e01..5ffe2b2d9d 100644
--- a/src/melee/gr/grzakogenerator.h
+++ b/src/melee/gr/grzakogenerator.h
@@ -5,17 +5,20 @@
 
 #include "gr/types.h"
 
-/* 1CA394 */ int grZakoGenerator_801CA394(UNK_T, int, UNK_T, f32);
-/* 1CA43C */ s32 grZakoGenerator_801CA43C(void*, HSD_JObj*, f32);
-/* 1CA67C */ UNK_RET grZakoGenerator_801CA67C(UNK_PARAMS);
-/* 1CA8B4 */ long grZakoGenerator_801CA8B4(int);
+/* 1CA394 */ grZakoGenerator_Config*
+grZakoGenerator_801CA394(grZakoGenerator_Spawn*, int,
+                         grZakoGenerator_SpawnFunc, f32);
+/* 1CA43C */ s32 grZakoGenerator_801CA43C(grZakoGenerator_Config*, HSD_JObj*,
+                                          f32);
+/* 1CA67C */ grZakoGenerator_Data* grZakoGenerator_801CA67C(void);
+/* 1CA8B4 */ void grZakoGenerator_801CA8B4(int);
 /* 1CAC14 */ void grZakoGenerator_801CAC14(HSD_GObj* gobj);
-/* 1CACB8 */ void grZakoGenerator_801CACB8(Ground_GObj* gobj);
-/* 1CADBC */ s32 fn_801CADBC(void);
-/* 1CADE0 */ s32 grZakoGenerator_801CADE0(void);
-/* 1CAE04 */ void grZakoGenerator_801CAE04(UNK_T);
+/* 1CACB8 */ void grZakoGenerator_801CACB8(Item_GObj* gobj);
+/* 1CADBC */ void fn_801CADBC(HSD_GObj* gobj);
+/* 1CADE0 */ void grZakoGenerator_801CADE0(void);
+/* 1CAE04 */ HSD_GObj* grZakoGenerator_801CAE04(grZakoGenerator_SpawnDesc*);
 /* 1CAEB0 */ void grZakoGenerator_801CAEB0(int, int);
 /* 1CAEF0 */ UNK_RET grZakoGenerator_801CAEF0(bool);
-/* 1CAF08 */ UNK_RET grZakoGenerator_801CAF08(UNK_PARAMS);
+/* 1CAF08 */ void grZakoGenerator_801CAF08(void);
 
 #endif
diff --git a/src/melee/gr/grzakogenerator.static.h b/src/melee/gr/grzakogenerator.static.h
index 4508a49637..adda9a00f7 100644
--- a/src/melee/gr/grzakogenerator.static.h
+++ b/src/melee/gr/grzakogenerator.static.h
@@ -1,28 +1,11 @@
 #ifndef __GALE01_1CA394
 #define __GALE01_1CA394
 
-#include <platform.h>
+#include "gr/types.h"
 
-#include "baselib/forward.h"
-
-struct grZakoGenerator_Entry {
-    /* +0x0 */ s16 x0;
-    /* +0x2 */ s16 x2;
-    /* +0x4 */ s32 x4;
-    /* +0x8 */ s32 x8;
-};
-
-struct grZakoGenerator_Data {
-    /* +0x000 */ struct grZakoGenerator_Entry entries[80];
-    /* +0x3C0 */ s16 x3C0;
-    /* +0x3C2 */ s16 x3C2;
-    /* +0x3C4 */ HSD_GObj* x3C4;
-    /* +0x3C8 */ s32 x3C8;
-};
-
-struct lbl_8049F030_t {
-    /* +0 */ void* x0;
-    /* +4 */ struct grZakoGenerator_Data* x4;
+static struct {
+    /* +0 */ grZakoGenerator_SpawnDesc* x0;
+    /* +4 */ grZakoGenerator_Data* x4;
     /* +8 */ s16 x8;
     /* +A:0 */ u8 xA_b0 : 1;
     /* +A:1 */ u8 xA_b1 : 1;
@@ -32,8 +15,6 @@ struct lbl_8049F030_t {
     /* +A:5 */ u8 xA_b5 : 1;
     /* +A:6 */ u8 xA_b6 : 1;
     /* +A:7 */ u8 xA_b7 : 1;
-};
-
-static struct lbl_8049F030_t lbl_8049F030;
+} lbl_8049F030;
 
 #endif
diff --git a/src/melee/gr/grzebes.c b/src/melee/gr/grzebes.c
index 5e3559bdbf..0e0ff5d713 100644
--- a/src/melee/gr/grzebes.c
+++ b/src/melee/gr/grzebes.c
@@ -187,7 +187,7 @@ void grZebes_801D8528(void) {}
 
 void grZebes_801D852C(void)
 {
-    grZakoGenerator_801CAE04(0);
+    grZakoGenerator_801CAE04(NULL);
 }
 
 bool grZebes_801D8550(void)
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 0d6772d337..79aea44cc8 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -174,6 +174,7 @@ struct GroundVars_unk {
     int xD4;
     int xD8;
     int xDC;
+    int xE0;
 };
 
 struct GroundVars_izumi {
@@ -310,6 +311,41 @@ struct grKraid_GroundVars2 {
     /*  + gp+D8 */ HSD_JObj* x14;
 };
 
+typedef struct grZakoGenerator_SpawnDesc {
+    /* +0x0 */ u16 kind;
+    /* +0x2 */ u8 x2;
+    /* +0x3 */ u8 respawn;
+} grZakoGenerator_SpawnDesc;
+
+typedef struct grZakoGenerator_Spawn {
+    /* +0x00 */ Vec3 pos0;
+    /* +0x0C */ Vec3 pos1;
+} grZakoGenerator_Spawn;
+
+typedef HSD_Generator* (*grZakoGenerator_SpawnFunc)(Vec3*, s32);
+
+typedef struct grZakoGenerator_Config {
+    /* +0x00 */ grZakoGenerator_Spawn* spawn_descs;
+    /* +0x04 */ grZakoGenerator_Spawn* spawns;
+    /* +0x08 */ int count;
+    /* +0x0C */ grZakoGenerator_SpawnFunc callback;
+    /* +0x10 */ f32 x10;
+    /* +0x14 */ HSD_Generator* gen;
+    /* +0x18 */ int x18;
+} grZakoGenerator_Config;
+
+typedef struct grZakoGenerator_Entry {
+    /* +0x0 */ s16 x0;
+    /* +0x2 */ s16 x2;
+    /* +0x4 */ Item_GObj* x4;
+    /* +0x8 */ s32 x8;
+} grZakoGenerator_Entry;
+
+typedef struct grZakoGenerator_Data {
+    /* +0x000 */ grZakoGenerator_Entry entries[80];
+    /* +0x3C0 */ grZakoGenerator_Entry sentinel;
+} grZakoGenerator_Data;
+
 struct grCorneria_GroundVars {
     union {
         struct {
@@ -326,8 +362,8 @@ struct grCorneria_GroundVars {
         u8 value;
     } xC6;
     u8 xC7;
-    u32 xC8;
-    u32 xCC;
+    grZakoGenerator_Config* xC8;
+    grZakoGenerator_Config* xCC;
     f32 xD0;
     f32 base_x;
     f32 base_y;
@@ -580,8 +616,7 @@ struct grInishie2_GroundVars {
 };
 
 struct grOldKongo_GroundVars {
-    u8 xC4;
-    u8 xC5;
+    s16 xC4;
     s16 xC6;
     s16 xC8;
     s16 xCA;
@@ -589,6 +624,12 @@ struct grOldKongo_GroundVars {
     s16 xCE;
     void* xD0;
     void* xD4;
+    f32 xD8;
+    f32 xDC;
+    f32 xE0;
+    f32 xE4;
+    f32 xE8;
+    f32 xEC;
 };
 
 /// likely for Cathrine (Birdo)
@@ -769,7 +810,8 @@ struct grRCruise_GroundVars {
         u8 b6 : 1;
         u8 b7 : 1;
     } xC4;
-    /* +04 gp+C8 */ char pad_0[0x8];
+    /* +04 gp+C8 */ struct lb_80011A50_t* x04;
+    /* +08 gp+CC */ f32 x08;
     /* +0C gp+D0 */ u32 x0C;
     /* +10 gp+D4 */ u32 x10;
     /* +14 gp+D8 */ f32 x14;
@@ -797,6 +839,11 @@ struct grFigureGet_GroundVars {
     /*  +4 gp+C8 */ UNK_T x4;
     /*  +8 gp+CC */ int x8;
     /*  +C gp+D0 */ int xC;
+    /* +10 gp+D4 */ int x10;
+    /* +14 gp+D8 */ int x14;
+    /* +18 gp+DC */ int x18;
+    /* +1C gp+E0 */ int x1C[3];
+    /* +28 gp+EC */ HSD_GObj* x28[3];
 };
 
 struct grFourside_GroundVars {
diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index ae838a0fb2..9992d2bd4c 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -728,6 +728,11 @@ typedef struct itSeakChain_Attrs {
 } itSeakChain_Attrs;
 STATIC_ASSERT(sizeof(struct itSeakChain_Attrs) == 0x5C);
 
+typedef struct itZako_ItemVars {
+    /* +0 ip+DD4 */ s32 pad;
+    /* +4 ip+DD8 */ s32 idx;
+} itZako_ItemVars;
+
 typedef struct itZeldaDinFireExplode_ItemVars {
     /* +0 ip+DD4 */ f32 xDD4;
     /* +4 ip+DD8 */ f32 xDD8;
diff --git a/src/melee/it/it_2725.c b/src/melee/it/it_2725.c
index aa5a3e4f37..66c6d77ef6 100644
--- a/src/melee/it/it_2725.c
+++ b/src/melee/it/it_2725.c
@@ -10442,7 +10442,7 @@ void it_8027CE18(Item_GObj* item_gobj)
 
 void it_8027CE44(Item_GObj* item_gobj)
 {
-    grZakoGenerator_801CACB8((Ground_GObj*) item_gobj);
+    grZakoGenerator_801CACB8(item_gobj);
 }
 
 #pragma push
diff --git a/src/melee/it/items/itlikelike.c b/src/melee/it/items/itlikelike.c
index b6c1116da0..f5389024d1 100644
--- a/src/melee/it/items/itlikelike.c
+++ b/src/melee/it/items/itlikelike.c
@@ -1323,7 +1323,7 @@ void itLikeLike_Logic5_Destroyed(Item_GObj* gobj)
     it_2725_Logic9_Destroyed(gobj);
 }
 
-Item_GObj* it_802DC4BC(s32 arg0, Vec3* arg1)
+Item_GObj* it_802DC4BC(s32 arg0, Vec3* arg1, s32 arg2)
 {
     // one r29-r30 regswap away from being solved
     Vec3 sp50;
diff --git a/src/melee/it/items/itlikelike.h b/src/melee/it/items/itlikelike.h
index c156d43d44..712cba2570 100644
--- a/src/melee/it/items/itlikelike.h
+++ b/src/melee/it/items/itlikelike.h
@@ -96,7 +96,7 @@
 /* 2DC444 */ void itLikelike_UnkMotion21_Phys(Item_GObj* gobj);
 /* 2DC460 */ bool itLikelike_UnkMotion21_Coll(Item_GObj* gobj);
 /* 2DC480 */ void itLikeLike_Logic5_Destroyed(Item_GObj*);
-/* 2DC4BC */ Item_GObj* it_802DC4BC(s32 arg0, Vec3* arg1);
+/* 2DC4BC */ Item_GObj* it_802DC4BC(s32 arg0, Vec3* arg1, s32 arg2);
 /* 3F8468 */ extern ItemStateTable it_803F8468[];
 /* 4D6D40 */ extern f32* it_804D6D40;
 #endif
diff --git a/src/melee/it/items/itpatapata.h b/src/melee/it/items/itpatapata.h
index f12498c56a..27fa87d844 100644
--- a/src/melee/it/items/itpatapata.h
+++ b/src/melee/it/items/itpatapata.h
@@ -37,7 +37,7 @@
 /* 2E1648 */ void it_802E1648(Item_GObj*, int, int);
 /* 2E1694 */ void it_802E1694(Item_GObj*);
 /* 2E16D8 */ void it_802E16D8(Item_GObj*, Item_GObj*);
-/* 2E16F8 */ UNK_RET it_802E16F8(UNK_PARAMS);
+/* 2E16F8 */ Item_GObj* it_802E16F8(s32, Vec3*, s32);
 extern ItemStateTable it_803F88C0[];
 
 #endif
diff --git a/src/melee/it/items/itzgshell.c b/src/melee/it/items/itzgshell.c
index 61d342d903..eee1648bd8 100644
--- a/src/melee/it/items/itzgshell.c
+++ b/src/melee/it/items/itzgshell.c
@@ -507,7 +507,7 @@ void it_802DFED4(Item_GObj* gobj, Item_GObj* ref_gobj)
 
 void itZGShell_Logic11_Destroyed(Item_GObj* gobj)
 {
-    grZakoGenerator_801CACB8((Ground_GObj*) gobj);
+    grZakoGenerator_801CACB8(gobj);
 }
 
 void it_802DFF14(Item_GObj* gobj, s32 arg1)
diff --git a/src/melee/it/types.h b/src/melee/it/types.h
index fe2158b806..1ca272cb73 100644
--- a/src/melee/it/types.h
+++ b/src/melee/it/types.h
@@ -645,6 +645,7 @@ struct Item {
         itWstar_ItemVars wstar;
         itZeldaDinFire_ItemVars zeldadinfire;
         itZeldaDinFireExplode_ItemVars zeldadinfireexplode;
+        itZako_ItemVars zako;
         u8 _[0xFCC - 0xDD4];
     } xDD4_itemVar;
 };
