diff --git a/src/melee/gm/gm_16F1.c b/src/melee/gm/gm_16F1.c
index 3258fd43e4..8ac10f46cd 100644
--- a/src/melee/gm/gm_16F1.c
+++ b/src/melee/gm/gm_16F1.c
@@ -163,7 +163,30 @@ void fn_80171AD4(void)
 
 /// #fn_8016FFD4
 
-/// #fn_80170110
+int fn_80170110(void* arg0, int idx, int mask, u8 player_id)
+{
+    struct lbl_803D5A4C_t* curr = lbl_803D5A4C;
+    int kind = lbl_803D5A4C[idx].kind;
+    u8 flags;
+
+    while (curr->kind != kind) {
+        if (curr->kind == 0x29A) {
+            flags = 0;
+            goto check;
+        }
+        curr++;
+    }
+    flags = curr->x4;
+
+check:
+    if ((u8) mask & flags) {
+        if (kind < 0xD7) {
+            return pl_80039418(player_id, kind);
+        }
+        return fn_801701C0(arg0, player_id, kind);
+    }
+    return 0;
+}
 
 void gm_801701A0(void)
 {
diff --git a/src/melee/gm/gm_16F1.h b/src/melee/gm/gm_16F1.h
index 9116283858..9368481a5a 100644
--- a/src/melee/gm/gm_16F1.h
+++ b/src/melee/gm/gm_16F1.h
@@ -28,11 +28,11 @@ struct lbl_804D65A8_t;
 /* 16F9A8 */ UNK_RET fn_8016F9A8(UNK_PARAMS);
 /* 16FAD4 */ int fn_8016FAD4(UNK_T, u16, int, u8);
 /* 16FFD4 */ UNK_RET fn_8016FFD4(UNK_PARAMS);
-/* 170110 */ int fn_80170110(UNK_T, int, int, u8);
+/* 170110 */ int fn_80170110(void*, int, int, u8);
 /* 1701A0 */ UNK_RET gm_801701A0(UNK_PARAMS);
 /* 1701AC */ UNK_RET fn_801701AC(UNK_PARAMS);
 /* 1701B8 */ int fn_801701B8(void);
-/* 1701C0 */ UNK_RET fn_801701C0(UNK_PARAMS);
+/* 1701C0 */ int fn_801701C0(void*, int, int);
 /* 171A88 */ int fn_80171A88(void);
 /* 171AD4 */ UNK_RET fn_80171AD4(UNK_PARAMS);
 /* 171B00 */ bool fn_80171B00(int);
diff --git a/src/melee/gm/gm_1BA8.c b/src/melee/gm/gm_1BA8.c
index 325b4d0346..380ffd4658 100644
--- a/src/melee/gm/gm_1BA8.c
+++ b/src/melee/gm/gm_1BA8.c
@@ -1430,7 +1430,23 @@ void gm_801BED3C(MinorScene* arg0)
     }
 }
 
-/// #gm_801BEDA8
+void gm_801BEDA8(MinorScene* arg0)
+{
+    CSSData* css = gm_801A427C(arg0);
+    VsModeData* vs = gm_801A5244();
+
+    if (css->pending_scene_change == 2) {
+        if (css->match_type != 0) {
+            css->match_type--;
+        } else {
+            css->match_type = 0x17;
+        }
+    } else {
+        css->match_type = (css->match_type + 1) % 24;
+    }
+
+    *vs = css->data;
+}
 
 void gm_801BEE58(MinorScene* arg0)
 {
diff --git a/src/melee/gm/gmregclear.c b/src/melee/gm/gmregclear.c
index 01c7f3148e..dfcef75289 100644
--- a/src/melee/gm/gmregclear.c
+++ b/src/melee/gm/gmregclear.c
@@ -1082,7 +1082,32 @@ int fn_8017F008(void)
     }
 }
 
-/// #fn_8017F09C
+s32 fn_8017F09C(void)
+{
+    struct lbl_80472D28_t* data = &lbl_80472D28;
+    lbl_8046B6A0_t* p = gm_8016AE38();
+    if (p->x24C8.x4_6) {
+        lbl_8046B6A0_t* p2 = gm_8016AE38();
+        int flag;
+        if (data->x118 != 0 ||
+            (data->x11A != 0 && data->x11B == 0) ||
+            p2->x24C8.x0_7)
+        {
+            flag = 0;
+        } else {
+            flag = 1;
+        }
+        if (flag != 0) {
+            s32 product = p->timer_seconds * data->x108;
+            s32 ret = product;
+            if (product < 0) {
+                ret = 0;
+            }
+            return ret;
+        }
+    }
+    return 0;
+}
 
 typedef struct fn_8017F14C_arg {
     /* 0x00 */ char pad_0[0x98];
diff --git a/src/melee/gm/gmregclear.h b/src/melee/gm/gmregclear.h
index bba2c86737..b2459d7e26 100644
--- a/src/melee/gm/gmregclear.h
+++ b/src/melee/gm/gmregclear.h
@@ -77,7 +77,7 @@
 /* 17EDDC */ bool fn_8017EDDC(void);
 /* 17EE40 */ void fn_8017EE40(int);
 /* 17F008 */ int fn_8017F008(void);
-/* 17F09C */ UNK_RET fn_8017F09C(UNK_PARAMS);
+/* 17F09C */ s32 fn_8017F09C(void);
 /* 17F14C */ s32 fn_8017F14C(void* arg0);
 /* 17F1B8 */ UNK_RET fn_8017F1B8(UNK_PARAMS);
 /* 17F2A4 */ UNK_RET fn_8017F2A4(UNK_PARAMS);
