diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.h b/src/melee/ft/chara/ftKirby/ftKb_Init.h
index 80f50d4b7a..7396f609d5 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.h
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.h
@@ -128,7 +128,7 @@
 /* 0F2160 */ void ftKb_AttackDashAir_IASA(Fighter_GObj* gobj);
 /* 0F2164 */ void ftKb_AttackDashAir_Phys(Fighter_GObj* gobj);
 /* 0F2198 */ void ftKb_AttackDashAir_Coll(Fighter_GObj* gobj);
-/* 0F21C0 */ void ftKb_AttackDashAir_800F21C0(Fighter_GObj* gobj);
+/* 0F21C0 */ float ftKb_AttackDashAir_800F21C0(Fighter_GObj* gobj);
 /* 0F21E8 */ void fn_800F21E8(HSD_GObj* gobj);
 /* 0F22D4 */ void ftKb_AttackDashAir_800F22D4(Fighter_GObj* gobj);
 /* 0F2300 */ void ftKb_SpecialHi_Enter(Fighter_GObj* gobj);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index 3b232c3a34..75374e7001 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -158,9 +158,9 @@ void ftKb_AttackDashAir_Coll(Fighter_GObj* gobj)
     ft_80082C74(gobj, fn_800F1FDC);
 }
 
-void ftKb_AttackDashAir_800F21C0(Fighter_GObj* gobj)
+float ftKb_AttackDashAir_800F21C0(Fighter_GObj* gobj)
 {
-    ftPartGetRotX(GET_FIGHTER(gobj), 0);
+    return ftPartGetRotX(GET_FIGHTER(gobj), 0);
 }
 
 void fn_800F21E8(Fighter_GObj* gobj)
diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index 97141c8b3b..ec6d36e8be 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -42,11 +42,12 @@ typedef struct itClimbersStringAttributes {
 } itClimbersStringAttributes;
 
 typedef struct itClimbersIceAttributes {
-    /* +00 */ u8 pad_00[0x4];
+    /* +00 */ f32 x0;
     /* +04 */ f32 x4;
-    /* +08 */ u8 pad_08[0x4];
+    /* +08 */ f32 x8;
     /* +0C */ f32 xC;
-    /* +10 */ u8 pad_10[0x18];
+    /* +10 */ f32 x10;
+    /* +14 */ u8 pad_14[0x14];
     /* +28 */ f32 x28;
     /* +2C */ s32 x2C;
     /* +30 */ u32 x30;
@@ -261,6 +262,13 @@ typedef struct {
     u32 x0;
 } itKirbyHammer_ItemVars;
 
+typedef struct {
+    /*  +0 ip+DD4 */ f32 angle;
+    /*  +4 ip+DD8 */ f32 speed;
+    /*  +8 ip+DDC */ Vec3 dir;
+    /* +14 ip+DE8 */ Vec3 init_pos;
+} itKirbyCutterBeam_ItemVars;
+
 /// user_data struct of GObj class 7
 typedef struct ItemLink {
     /*   +0 */ struct ItemLink* prev;
@@ -691,7 +699,7 @@ typedef struct itSeakNeedleHeld_ItemVars {
 } itSeakNeedleHeld_ItemVars;
 
 typedef struct itSeakNeedleThrown_ItemVars {
-    /*  +0 ip+DD4 */ char pad_0[0x4];
+    /*  +0 ip+DD4 */ f32 xDD4;
     /*  +4 ip+DD8 */ f32 xDD8;
     /*  +8 ip+DDC */ f32 xDDC;
     /*  +C ip+DE0 */ f32 xDE0;
diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index af8a8329e2..344a138bbd 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -196,6 +196,8 @@ typedef struct itLugiaAttributes {
     /* +10 */ f32 x10;
     /* +14 */ s32 x14;
     /* +18 */ f32 x18;
+    /* +1C */ u8 x1C_pad[0x24];
+    /* +40 */ f32 x40;
 } itLugiaAttributes;
 
 /// Box/Crate item attributes loaded from .dat file
@@ -709,10 +711,15 @@ typedef struct itHitodeman_ItemVars {
     /* +60 ip+E34 */ float x60;
     /* +64 ip+E38 */ float x64;
     /* +68 ip+E3C */ s32 x68;
-    /* +6C ip+E40 */ char pad_6C[0x18];
+    /* +6C ip+E40 */ float x6C;
+    /* +70 ip+E44 */ float x70;
+    /* +74 ip+E48 */ float x74;
+    /* +78 ip+E4C */ float x78;
+    /* +7C ip+E50 */ float x7C;
+    /* +80 ip+E54 */ float x80;
     /* +84 ip+E58 */ float x84;
     /* +88 ip+E5C */ s32 x88;
-    /* +8C ip+E60 */ char pad_8C[0x4];
+    /* +8C ip+E60 */ float x8C;
     /* +90 ip+E64 */ HSD_GObj* x90;
 } itHitodeman_ItemVars;
 
@@ -1082,7 +1089,9 @@ typedef struct it_279D_ItemVars {
 typedef struct itMewtwoShadowball_ItemVars {
     /* ip+DD4 */ s32 x0;
     /* ip+DD8 */ Vec3 x4;
-    /* ip+DE4 */ s32 x10[3];
+    /* ip+DE4 */ f32 x10;
+    /* ip+DE8 */ s32 x14;
+    /* ip+DEC */ s32 x18;
     /* ip+DF0 */ s32 x1C;
     /* ip+DF4 */ s32 x20;
     /* ip+DF8 */ f32 x24;
@@ -1453,6 +1462,9 @@ typedef struct itPatapataAttributes {
     /* 0x24 */ s32 x24;
     /* 0x28 */ s32 x28;
     /* 0x2C */ s32 x2C;
+    /* 0x30 */ u8 pad_30[0x38 - 0x30];
+    /* 0x38 */ f32 x38;
+    /* 0x3C */ f32 x3C;
 } itPatapataAttributes;
 
 typedef struct itOldottoseaAttributes {
@@ -1564,8 +1576,12 @@ typedef struct itLugia_ItemVars {
     /* +00 ip+DD4 */ u8 x0_pad[0x60];
     /* +60 ip+E34 */ s32 x60;
     /* +64 ip+E38 */ Vec3 x64;
-    /* +70 ip+E44 */ u8 x70_pad[0xC];
+    /* +70 ip+E44 */ Vec3 x70;
     /* +7C ip+E50 */ Vec3 xE50;
+    /* +88 ip+E5C */ f32 x88;
+    /* +8C ip+E60 */ Vec3 x8C;
+    /* +98 ip+E6C */ u8 x98_pad[0xC];
+    /* +A4 ip+E78 */ s32 xA4;
 } itLugia_ItemVars;
 
 typedef struct itGreatFoxLaser_ItemVars {
diff --git a/src/melee/it/items/itclimbersice.c b/src/melee/it/items/itclimbersice.c
index afba86670d..8e9db35a89 100644
--- a/src/melee/it/items/itclimbersice.c
+++ b/src/melee/it/items/itclimbersice.c
@@ -22,7 +22,47 @@
 
 /// #it_802C1590
 
-/// #it_802C16F8
+void it_802C16F8(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itClimbersIceAttributes* sa = ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(24);
+
+    ip->x40_vel.x = sa->x10 * ip->facing_dir;
+    ip->x40_vel.z = 0.0f;
+    ip->x40_vel.y = 0.0f;
+
+    it_80275158(gobj, sa->x0);
+
+    ip->xDD4_itemVar.climbersice.x8_b0 = 1;
+
+    it_80272980(gobj);
+
+    {
+        HSD_JObj* child;
+        HSD_JObj* jobj = gobj->hsd_obj;
+
+        if (jobj == NULL) {
+            child = NULL;
+        } else {
+            child = jobj->child;
+        }
+
+        {
+            Item* ip = GET_ITEM(gobj);
+
+            if (ip->kind == 0x6A) {
+                f32* facing = &ip->facing_dir;
+                efAsync_Spawn(gobj, &ip->xBC0, 3, 0x4EA, child, facing);
+            } else {
+                f32* facing = &ip->facing_dir;
+                efAsync_Spawn(gobj, &ip->xBC0, 3, 0x4AF, child, facing);
+            }
+        }
+    }
+
+    it_802C1AE4(gobj);
+}
 
 void it_802C17DC(Item_GObj* gobj)
 {
@@ -202,7 +242,21 @@ bool itClimbersice_UnkMotion3_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itClimbersice_UnkMotion3_Phys
+void itClimbersice_UnkMotion3_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    ItemAttr* attrs = ip->xCC_item_attr;
+    itClimbersIceAttributes* sa;
+    u32 dmg;
+
+    it_80272860(gobj, attrs->x10_fall_speed, attrs->x14_fall_speed_max);
+    it_80274658(gobj, it_804D6D28->x68_float);
+
+    ip = GET_ITEM(gobj);
+    sa = ip->xC4_article_data->x4_specialAttributes;
+    dmg = (u32) ABS(ip->x40_vel.x * sa->x30) + sa->x2C;
+    it_80272460(&ip->x5D4_hitboxes[0].hit, dmg, gobj);
+}
 
 /// #itClimbersice_UnkMotion3_Coll
 
@@ -221,7 +275,41 @@ bool itClimbersIce_Logic90_Clanked(Item_GObj* arg0)
     return true;
 }
 
-/// #it_2725_Logic90_HitShield
+bool it_2725_Logic90_HitShield(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    f32 vel_x = ip->x40_vel.x;
+    ItemAttr* attr = ip->xCC_item_attr;
+    itClimbersIceAttributes* sa =
+        ip->xC4_article_data->x4_specialAttributes;
+
+    if (vel_x < 0.0f) {
+        vel_x = -vel_x;
+    }
+
+    if (vel_x <= sa->xC) {
+        itColl_BounceOffVictim(gobj);
+        it_80272980(gobj);
+        {
+            f32 scale = attr->x58;
+            ip->x40_vel.x *= scale;
+            ip->x40_vel.y *= scale;
+            ip->x40_vel.z *= scale;
+        }
+        {
+            Item* ip2 = GET_ITEM(gobj);
+            itClimbersIceAttributes* sa2 =
+                ip2->xC4_article_data->x4_specialAttributes;
+            ip2->xD44_lifeTimer -= sa2->x4;
+        }
+        if (ip->xD44_lifeTimer < sa->x8) {
+            return true;
+        }
+        return false;
+    }
+
+    return true;
+}
 
 bool itClimbersIce_Logic90_Absorbed(Item_GObj* arg0)
 {
diff --git a/src/melee/it/items/itflipper.c b/src/melee/it/items/itflipper.c
index 078f6769fe..aec91bfdbe 100644
--- a/src/melee/it/items/itflipper.c
+++ b/src/melee/it/items/itflipper.c
@@ -14,7 +14,34 @@
 
 /* 0x802910B8 */ void itFlipper_Logic20_Thrown(Item_GObj* gobj);
 
-/// #it_80290938
+HSD_GObj* it_80290938(HSD_JObj* jobj)
+{
+    SpawnItem spawn;
+    Vec3 pos;
+    Item_GObj* gobj = NULL;
+
+    if (jobj != NULL) {
+        lb_8000B1CC(jobj, NULL, &pos);
+        spawn.kind = It_Kind_Flipper;
+        spawn.prev_pos = pos;
+        spawn.pos = spawn.prev_pos;
+        spawn.facing_dir = 1.0f;
+        spawn.x3C_damage = 0;
+        spawn.vel.x = spawn.vel.y = spawn.vel.z = 0.0f;
+        spawn.x0_parent_gobj = NULL;
+        spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+        spawn.x44_flag.b0 = 0;
+        spawn.x40 = 0;
+        gobj = Item_80268B18(&spawn);
+        if (gobj != NULL) {
+            Item* ip = GET_ITEM(gobj);
+            ip->xDD4_itemVar.flipper.xDE8 = 1;
+            ip->xDD4_itemVar.flipper.xDEC = jobj;
+            it_80291254(gobj);
+        }
+    }
+    return gobj;
+}
 
 void itFlipper_Logic20_Spawned(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ithitodeman.c b/src/melee/it/items/ithitodeman.c
index 26af100534..f522134d3f 100644
--- a/src/melee/it/items/ithitodeman.c
+++ b/src/melee/it/items/ithitodeman.c
@@ -1,5 +1,6 @@
 #include "ithitodeman.h"
 
+#include <math.h>
 #include <placeholder.h>
 #include <platform.h>
 
@@ -17,21 +18,49 @@
 #include <baselib/random.h>
 
 typedef struct {
-    UNK_T x0;
+    f32 x0;
     float x4;
     float x8;
     float xC;
     float x10;
     float x14;
     float x18;
-    char pad_0[0x40 - 0x1C];
+    char pad_0[0x3C - 0x1C];
+    s32 x3C;
     s32 x40;
     s32 x44;
     u8 _pad2[0x4];
     f32 x4C;
 } itHitodemanAttributes;
 
-/// #it_2725_Logic24_Spawned
+void it_2725_Logic24_Spawned(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itHitodemanAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+
+    ip->facing_dir = 0.0f;
+    ip->xDBC_itcmd_var4.flags.x0 = false;
+
+    it_80279CDC(gobj, attrs->x0);
+    it_802D43EC(gobj);
+    it_802D4494(gobj);
+    it_802D4510(gobj);
+
+    ip->xDD4_itemVar.hitodeman.x8C = (f32) attrs->x3C;
+
+    ip->xDD4_itemVar.hitodeman.x6C = 0.0f;
+    ip->xDD4_itemVar.hitodeman.x70 = 0.0f;
+    ip->xDD4_itemVar.hitodeman.x74 = 0.0f;
+    ip->xDD4_itemVar.hitodeman.x78 = 0.0f;
+    ip->xDD4_itemVar.hitodeman.x7C = 0.0f;
+    ip->xDD4_itemVar.hitodeman.x80 = 0.0f;
+
+    ip->xDAC_itcmd_var0 = 0;
+    ip->xDB0_itcmd_var1 = 0;
+
+    it_802D4B54(gobj);
+    Item_8026AE84(ip, 0x2738, 0x7F, 0x40);
+}
 
 void it_802D43AC(void) {}
 
@@ -100,7 +129,38 @@ bool it_802D48A8(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802D48B0
+void it_802D48B0(f32* value, f32 target, f32 max_val, f32 accel,
+                 f32 decel)
+{
+    f32 current = *value;
+    f32 diff = target - current;
+
+    if (ABS(diff) > ABS(accel)) {
+        if (diff > 0.0F) {
+            target = current + accel;
+        } else {
+            target = current - accel;
+        }
+    } else if (ABS(diff) < ABS(decel)) {
+        if (diff > 0.0F) {
+            target = current + decel;
+        } else {
+            target = current - decel;
+        }
+    }
+
+    if (target > 0.0F) {
+        if (target > max_val) {
+            target = max_val;
+        }
+    } else {
+        if (target < -max_val) {
+            target = -max_val;
+        }
+    }
+
+    *value = target;
+}
 
 void it_802D4990(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ithitodeman.h b/src/melee/it/items/ithitodeman.h
index 45cbd34b4b..72f2298d87 100644
--- a/src/melee/it/items/ithitodeman.h
+++ b/src/melee/it/items/ithitodeman.h
@@ -16,7 +16,8 @@
 /* 2D4564 */ bool it_802D4564(Item_GObj* gobj);
 /* 2D472C */ void it_802D472C(Item_GObj* gobj);
 /* 2D48A8 */ bool it_802D48A8(Item_GObj* gobj);
-/* 2D48B0 */ UNK_RET it_802D48B0(UNK_PARAMS);
+/* 2D48B0 */ void it_802D48B0(f32* value, f32 target, f32 max_val,
+                              f32 accel, f32 decel);
 /* 2D4990 */ void it_802D4990(Item_GObj*);
 /* 2D49E8 */ bool itHitodeman_UnkMotion1_Anim(Item_GObj* gobj);
 /* 2D4B14 */ void itHitodeman_UnkMotion1_Phys(Item_GObj* gobj);
diff --git a/src/melee/it/items/itkamex.c b/src/melee/it/items/itkamex.c
index 6578b6dd2e..0ccb22b84b 100644
--- a/src/melee/it/items/itkamex.c
+++ b/src/melee/it/items/itkamex.c
@@ -13,6 +13,8 @@
 #include "it/it_2725.h"
 #include "it/item.h"
 
+#include <baselib/jobj.h>
+
 void it_802CA49C(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -227,7 +229,23 @@ bool itKamex_UnkMotion3_Coll(Item_GObj* gobj)
 
 /// #it_802CAB10
 
-/// #it_2725_Logic31_Spawned
+void it_2725_Logic31_Spawned(Item_GObj* gobj)
+{
+    HSD_JObj* jobj;
+    Vec3 scale;
+    Item* ip = gobj->user_data;
+    itKamexAttributes* sa = ip->xC4_article_data->x4_specialAttributes;
+
+    jobj = gobj->hsd_obj;
+    ip->xD44_lifeTimer = sa->x0;
+    it_80274740(gobj);
+    it_802CADF0(gobj);
+    it_8026B3A8(gobj);
+    scale.x = 0.2f;
+    scale.y = 0.2f;
+    scale.z = 0.2f;
+    HSD_JObjSetScale(jobj, &scale);
+}
 
 bool itKamex_Logic31_HitShield(Item_GObj* arg0)
 {
diff --git a/src/melee/it/items/itkirbycutterbeam.c b/src/melee/it/items/itkirbycutterbeam.c
index 08adfc1f75..46adf62aee 100644
--- a/src/melee/it/items/itkirbycutterbeam.c
+++ b/src/melee/it/items/itkirbycutterbeam.c
@@ -1,11 +1,14 @@
 #include "itkirbycutterbeam.h"
 
 #include "db/db.h"
+#include "ft/chara/ftKirby/ftKb_Init.h"
+#include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
 #include "it/types.h"
 
+#include <MSL/trigf.h>
 #include <placeholder.h>
 #include <platform.h>
 
@@ -34,7 +37,30 @@ void it_8029BAB8(HSD_GObj* gobj, Vec3* pos, float facing_dir)
     }
 }
 
-/// #it_8029BB90
+void it_8029BB90(Item_GObj* gobj, HSD_GObj* owner)
+{
+    Item* ip = GET_ITEM(gobj);
+    f32* attr = ip->xC4_article_data->x4_specialAttributes;
+
+    ip->xDD4_itemVar.kirbycutterbeam.angle =
+        ftKb_AttackDashAir_800F21C0(owner);
+    ip->xDD4_itemVar.kirbycutterbeam.speed = attr[0];
+    ip->x40_vel.x = ip->facing_dir *
+                     ip->xDD4_itemVar.kirbycutterbeam.speed *
+                     cosf(ip->xDD4_itemVar.kirbycutterbeam.angle);
+    ip->x40_vel.y = -ip->xDD4_itemVar.kirbycutterbeam.speed *
+                     sinf(ip->xDD4_itemVar.kirbycutterbeam.angle);
+    ip->x40_vel.z = 0.0F;
+    it_80275158(gobj, attr[2]);
+    Item_80268E5C(gobj, 0, 2);
+    it_802762B0(ip);
+    ip->xDD4_itemVar.kirbycutterbeam.dir.x =
+        ip->facing_dir * sinf(ip->xDD4_itemVar.kirbycutterbeam.angle);
+    ip->xDD4_itemVar.kirbycutterbeam.dir.y =
+        cosf(ip->xDD4_itemVar.kirbycutterbeam.angle);
+    ip->xDD4_itemVar.kirbycutterbeam.dir.z = 0.0F;
+    ip->xDD4_itemVar.kirbycutterbeam.init_pos = ip->pos;
+}
 
 bool itKirbycutterbeam_UnkMotion0_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itlugia.c b/src/melee/it/items/itlugia.c
index 791b22443b..98fb0802f2 100644
--- a/src/melee/it/items/itlugia.c
+++ b/src/melee/it/items/itlugia.c
@@ -10,6 +10,7 @@
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
+#include "lb/lb_00B0.h"
 
 void it_2725_Logic17_Spawned(Item_GObj* gobj)
 {
@@ -201,7 +202,26 @@ bool itLugia_UnkMotion4_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802D1A44
+void it_802D1A44(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itLugiaAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    Vec3 pos;
+
+    Item_80268E5C(gobj, 5, ITEM_ANIM_UPDATE);
+    ip->entered_hitlag = efLib_PauseAll;
+    ip->exited_hitlag = efLib_ResumeAll;
+    ip->xDD4_itemVar.lugia.xE50.y = it_804DD46C;
+    ip->xDD4_itemVar.lugia.xE50.z = it_804DD470;
+    ip->xDD4_itemVar.lugia.x88 = it_804DD474;
+    lb_8000B1CC(ip->xBBC_dynamicBoneTable->bones[25], NULL, &pos);
+    ip->xDD4_itemVar.lugia.x8C = ip->xDD4_itemVar.lugia.x64;
+    ip->xDD4_itemVar.lugia.x8C.y += attrs->x40;
+    ip->xDD4_itemVar.lugia.x70 = pos;
+    ip->xDD4_itemVar.lugia.xA4 = 0;
+    ip->on_accessory = (HSD_GObjEvent) it_802D1BBC;
+    ip->xDCC_flag.b3 = 1;
+}
 
 bool itLugia_UnkMotion5_Anim(Item_GObj* gobj)
 {
@@ -272,7 +292,7 @@ Item_GObj* it_802D1E8C(Item_GObj* gobj, ItemKind kind, f32 param)
     SpawnItem spawn;
     Item* ip = GET_ITEM(gobj);
 
-    spawn.prev_pos = *(Vec3*) ip->xDD4_itemVar.lugia.x70_pad;
+    spawn.prev_pos = ip->xDD4_itemVar.lugia.x70;
     it_8026BB88(gobj, &spawn.pos);
     spawn.facing_dir = ip->facing_dir;
     spawn.x3C_damage = 0;
diff --git a/src/melee/it/items/itlugia.h b/src/melee/it/items/itlugia.h
index 569d55ffca..d9863e95f7 100644
--- a/src/melee/it/items/itlugia.h
+++ b/src/melee/it/items/itlugia.h
@@ -49,6 +49,9 @@
 /* 3F7F48 */ extern ItemStateTable it_803F7F48[];
 
 extern f32 it_804DD450;
+extern f32 it_804DD46C;
+extern f32 it_804DD470;
+extern f32 it_804DD474;
 extern f32 it_804DD490;
 
 #endif
diff --git a/src/melee/it/items/itmatadogas.c b/src/melee/it/items/itmatadogas.c
index 6f0b052b8b..ccedbdd4eb 100644
--- a/src/melee/it/items/itmatadogas.c
+++ b/src/melee/it/items/itmatadogas.c
@@ -5,12 +5,15 @@
 
 #include "ef/eflib.h"
 #include "ef/efsync.h"
+#include "ft/ft_0C31.h"
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
 
+#include <baselib/debug.h>
+#include <baselib/jobj.h>
 #include <baselib/random.h>
 
 void it_802CAFD4(Item_GObj* gobj)
@@ -96,7 +99,32 @@ bool itMatadogas_UnkMotion2_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itMatadogas_UnkMotion2_Phys
+void itMatadogas_UnkMotion2_Phys(Item_GObj* gobj)
+{
+    Item* item;
+    Item* item2;
+    HSD_JObj* jobj;
+
+    if (it_8027A09C(gobj)) {
+        item = gobj->user_data;
+        it_80273454(gobj);
+        item2 = gobj->user_data;
+        jobj = gobj->hsd_obj;
+        Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
+        item2->entered_hitlag = efLib_PauseAll;
+        item2->exited_hitlag = efLib_ResumeAll;
+        ((jobj) ? ((void) 0) : __assert("jobj.h", 660, "jobj"));
+        ((!(jobj->flags & JOBJ_USE_QUATERNION)) ? ((void) 0)
+            : __assert("jobj.h", 661,
+                       "!(jobj->flags & JOBJ_USE_QUATERNION)"));
+        jobj->rotate.y = 0.0f;
+        if (!(jobj->flags & JOBJ_MTX_INDEP_SRT)) {
+            ftCo_800C6AFC(jobj);
+        }
+        item2->on_accessory = (HSD_GObjEvent) it_802CB2B0;
+        item->xDD1_flag.b1 = 1;
+    }
+}
 
 bool itMatadogas_UnkMotion2_Coll(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itmewtwoshadowball.c b/src/melee/it/items/itmewtwoshadowball.c
index b37674e317..7ada090d51 100644
--- a/src/melee/it/items/itmewtwoshadowball.c
+++ b/src/melee/it/items/itmewtwoshadowball.c
@@ -4,6 +4,8 @@
 #include "ft/chara/ftKirby/ftKb_Init.h"
 #include "ft/chara/ftMewtwo/ftMt_SpecialN.h"
 #include "it/inlines.h"
+
+#include <baselib/jobj.h>
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
@@ -48,7 +50,7 @@ void it_2725_Logic101_Destroyed(Item_GObj* gobj)
     if (gobj != NULL) {
         Item* ip = gobj->user_data;
         efLib_DestroyAll(gobj);
-        if (ip->xDD4_itemVar.mewtwoshadowball.x10[1] == 0) {
+        if (ip->xDD4_itemVar.mewtwoshadowball.x14 == 0) {
             if (ip->xDD4_itemVar.mewtwoshadowball.x2C != NULL &&
                 ip->owner == ip->xDD4_itemVar.mewtwoshadowball.x2C)
             {
@@ -97,9 +99,42 @@ bool itMewtwoshadowball_UnkMotion0_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802C5B18
+void it_802C5B18(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    it_279D_DatAttrs* attr = ip->xC4_article_data->x4_specialAttributes;
+    it_80275158(gobj, attr->x0);
+    efLib_DestroyAll(gobj);
+    ip->xDD4_itemVar.mewtwoshadowball.x28 = 0;
+    if (ip->xDD4_itemVar.mewtwoshadowball.x18 < 0) {
+        ip->xDD4_itemVar.mewtwoshadowball.x18 = 0;
+    }
+    if (ip->xDD4_itemVar.mewtwoshadowball.x18 >=
+        ip->xDD4_itemVar.mewtwoshadowball.x1C)
+    {
+        ip->xDD4_itemVar.mewtwoshadowball.x18 =
+            ip->xDD4_itemVar.mewtwoshadowball.x1C;
+    }
+    Item_80268E5C(gobj,
+                  ip->xDD4_itemVar.mewtwoshadowball.x18 + 1,
+                  ITEM_ANIM_UPDATE);
+    ip->on_accessory = fn_802C5E18;
+    ip->xDD4_itemVar.mewtwoshadowball.x50 =
+        0.5F * ((f32) ip->xDD4_itemVar.mewtwoshadowball.x18 /
+                (f32) ip->xDD4_itemVar.mewtwoshadowball.x1C) +
+        0.5F;
+    PAD_STACK(8);
+}
 
-/// #itMewtwoshadowball_UnkMotion8_Anim
+bool itMewtwoshadowball_UnkMotion8_Anim(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    HSD_JObj* jobj = itGetJObjGrandchild(gobj);
+    Vec3 scale;
+    scale.x = scale.y = scale.z = ip->xDD4_itemVar.mewtwoshadowball.x10;
+    HSD_JObjSetScale(jobj, &scale);
+    return it_80273130(gobj);
+}
 
 void itMewtwoshadowball_UnkMotion8_Phys(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itmewtwoshadowball.h b/src/melee/it/items/itmewtwoshadowball.h
index 0ec76afee9..eeb54350ee 100644
--- a/src/melee/it/items/itmewtwoshadowball.h
+++ b/src/melee/it/items/itmewtwoshadowball.h
@@ -20,7 +20,7 @@
 /* 2C57C8 */ bool itMewtwoshadowball_UnkMotion0_Anim(Item_GObj* gobj);
 /* 2C5B0C */ void itMewtwoshadowball_UnkMotion0_Phys(Item_GObj* gobj);
 /* 2C5B10 */ bool itMewtwoshadowball_UnkMotion0_Coll(Item_GObj* gobj);
-/* 2C5B18 */ UNK_RET it_802C5B18(UNK_PARAMS);
+/* 2C5B18 */ void it_802C5B18(Item_GObj*);
 /* 2C5BF8 */ bool itMewtwoshadowball_UnkMotion8_Anim(Item_GObj* gobj);
 /* 2C5CF4 */ void itMewtwoshadowball_UnkMotion8_Phys(Item_GObj* gobj);
 /* 2C5D64 */ bool itMewtwoshadowball_UnkMotion8_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/itoldottosea.c b/src/melee/it/items/itoldottosea.c
index 92a119325e..e687482e64 100644
--- a/src/melee/it/items/itoldottosea.c
+++ b/src/melee/it/items/itoldottosea.c
@@ -53,7 +53,36 @@ bool itOldottosea_UnkMotion0_Coll(Item_GObj* gobj)
 
 /// #itOldottosea_UnkMotion2_Anim
 
-/// #itOldottosea_UnkMotion2_Phys
+void itOldottosea_UnkMotion2_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itOldottoseaAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+    s32 jp_offset;
+
+    if (lbLang_IsSettingJP()) {
+        jp_offset = 10;
+    } else {
+        jp_offset = 0;
+    }
+
+    if (ip->xDD4_itemVar.oldottosea.x24 == jp_offset) {
+        ip->x40_vel.x = ip->facing_dir * ((f32*)attr->x0)[1];
+    }
+
+    ip->xDD4_itemVar.oldottosea.x24++;
+
+    if (ip->x40_vel.x != 0.0f) {
+        if ((u32) ip->xDD4_itemVar.oldottosea.x20 == 0) {
+            if (it_802750E8(gobj, 2)) {
+                it_802E3098(gobj);
+            }
+        }
+    }
+
+    it_8027C8D0(&ip->x40_vel, &ip->x378_itemColl.floor.normal,
+                ip->facing_dir);
+    it_8027C0A8(gobj, ip->facing_dir, 5.0f);
+}
 
 bool itOldottosea_UnkMotion2_Coll(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itpatapata.c b/src/melee/it/items/itpatapata.c
index 8542a9f1a3..d11868cb24 100644
--- a/src/melee/it/items/itpatapata.c
+++ b/src/melee/it/items/itpatapata.c
@@ -261,7 +261,14 @@ bool itPatapata_UnkMotion4_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itPatapata_UnkMotion4_Phys
+void itPatapata_UnkMotion4_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itPatapataAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    HSD_JObjAddTranslationZ(ip->xBBC_dynamicBoneTable->bones[18],
+                             ip->xDD4_itemVar.patapata.x2C);
+    it_80272860(gobj, attrs->x38, attrs->x3C);
+}
 
 bool itPatapata_UnkMotion4_Coll(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itseakneedlethrown.c b/src/melee/it/items/itseakneedlethrown.c
index 1648208ee0..9ce0fd7e16 100644
--- a/src/melee/it/items/itseakneedlethrown.c
+++ b/src/melee/it/items/itseakneedlethrown.c
@@ -6,6 +6,8 @@
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 
+#include <baselib/jobj.h>
+
 /// #it_802AFD8C
 
 Item_GObj* it_802AFEA8(Item_GObj* gobj, Fighter_GObj* owner_gobj, u32 arg2)
@@ -42,7 +44,14 @@ bool itSeakneedlethrown_UnkMotion0_Anim(Item_GObj* gobj)
     return it_80273130(gobj);
 }
 
-/// #itSeakneedlethrown_UnkMotion1_Anim
+bool itSeakneedlethrown_UnkMotion1_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* jobj = HSD_JObjGetChild(GET_JOBJ(gobj));
+    ip->xDD4_itemVar.seakneedlethrown.xDE4 = ip->pos;
+    HSD_JObjAddRotationX(jobj, ip->xDD4_itemVar.seakneedlethrown.xDD4);
+    return it_80273130(gobj);
+}
 
 /// #itSeakneedlethrown_UnkMotion2_Anim
 
@@ -51,7 +60,14 @@ bool itSeakneedlethrown_UnkMotion3_Anim(Item_GObj* gobj)
     return it_80273130(gobj);
 }
 
-/// #itSeakneedlethrown_UnkMotion4_Anim
+bool itSeakneedlethrown_UnkMotion4_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* jobj = HSD_JObjGetChild(GET_JOBJ(gobj));
+    ip->xDD4_itemVar.seakneedlethrown.xDE4 = ip->pos;
+    HSD_JObjAddRotationX(jobj, ip->xDD4_itemVar.seakneedlethrown.xDD4);
+    return false;
+}
 
 void itSeakneedlethrown_UnkMotion0_Phys(Item_GObj* gobj) {}
 
diff --git a/src/melee/it/items/ittincle.c b/src/melee/it/items/ittincle.c
index 812a4044fb..ed62761941 100644
--- a/src/melee/it/items/ittincle.c
+++ b/src/melee/it/items/ittincle.c
@@ -90,7 +90,38 @@ bool itTincle_UnkMotion1_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802EBA00
+void it_802EBA00(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itTincleAttributes* sa = ip->xC4_article_data->x4_specialAttributes;
+    f32 dur_f = (f32) sa->x28;
+    f32 height = sa->x24;
+
+    ip->x40_vel.z = 0.0f;
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+
+    ip->xDD4_itemVar.tincle.x20 = sa->x2C;
+    ip->x40_vel.y = 0.0f;
+
+    ip->xDD4_itemVar.tincle.x38 =
+        (2.0f * height) / (dur_f * dur_f);
+    ip->xDD4_itemVar.tincle.x38 *= 0.5f;
+
+    ip->xDD4_itemVar.tincle.x2C = sa->x28 * 2;
+    ip->xDD4_itemVar.tincle.x3C = 2.0f;
+    ip->xDD4_itemVar.tincle.x24 = 0;
+
+    it_802756E0(gobj);
+
+    ip->xDC8_word.flags.x1A = 1;
+
+    HSD_JObjClearFlagsAll(GET_JOBJ(gobj), 0x10);
+
+    Item_80268E5C(gobj, 2, ITEM_UNK_0x1);
+
+    it_802EC9E8(gobj);
+}
 
 bool itTincle_UnkMotion2_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/types.h b/src/melee/it/types.h
index 8b06eabfcb..fe2158b806 100644
--- a/src/melee/it/types.h
+++ b/src/melee/it/types.h
@@ -573,6 +573,7 @@ struct Item {
         itKabigon_ItemVars kabigon;
         itKinoko_ItemVars kinoko;
         itKirbyHammer_ItemVars kirbyhammer;
+        itKirbyCutterBeam_ItemVars kirbycutterbeam;
         itkireihana_ItemVars kireihana;
         itKlap_ItemVars klap;
         itKoopaFlame_ItemVars koopaflame;
