diff --git a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
index 7b820175f0..930bc70454 100644
--- a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
+++ b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
@@ -1032,7 +1032,15 @@ void ftCh_Init_8015764C(HSD_GObj* gobj)
     fp->accessory4_cb = fn_801577B4;
 }
 
-/// #ftCh_Slap_Anim
+void ftCh_Slap_Anim(HSD_GObj* gobj)
+{
+    if (ftAnim_IsFramesRemaining(gobj) == 0) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        Fighter_ChangeMotionState(gobj, 0x15D, 0, 0.0f, 1.0f, 0.0f, NULL);
+        ftAnim_8006EBA4(gobj);
+        fp->accessory4_cb = fn_801577B4;
+    }
+}
 
 void ftCh_Slap_IASA(HSD_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index 4d29105dd2..f6bd4a5b0c 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -2603,9 +2603,27 @@ void ftKb_LgSpecialAirN_Phys(Fighter_GObj* gobj)
     ft_80084DB0(gobj);
 }
 
-/// #ftKb_LgSpecialN_Coll
+void ftKb_LgSpecialN_Coll(Fighter_GObj* gobj)
+{
+    if (ft_80082708(gobj) == GA_Ground) {
+        Fighter* fp = gobj->user_data;
+        ftCommon_8007D5D4(fp);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_LgSpecialAirN, 0x5000,
+                                  fp->cur_anim_frame, 1.0f, 0.0f, NULL);
+        fp->accessory4_cb = fn_800F98F4;
+    }
+}
 
-/// #ftKb_LgSpecialAirN_Coll
+void ftKb_LgSpecialAirN_Coll(Fighter_GObj* gobj)
+{
+    if (ft_80081D0C(gobj) != GA_Ground) {
+        Fighter* fp = gobj->user_data;
+        ftCommon_8007D7FC(fp);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_LgSpecialN, 0x5000,
+                                  fp->cur_anim_frame, 1.0f, 0.0f, NULL);
+        fp->accessory4_cb = fn_800F98F4;
+    }
+}
 
 /// #fn_800F98F4
 
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
index 903f824e8f..3dcc8ab65f 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
@@ -1202,4 +1202,20 @@ void ftKb_SsSpecialAirN_Coll(Fighter_GObj* gobj)
     }
 }
 
-/// #ftKb_SpecialNFx_800FDC00
+void ftKb_SpecialNFx_800FDC00(Fighter_GObj* gobj, Vec3* vec)
+{
+    if (gobj != NULL) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        if (fp != NULL) {
+            Vec3 offset;
+            offset.x = 0.0f;
+            offset.y = 1.45f;
+            offset.z = 5.016f;
+            lb_8000B1CC(fp->parts[44].joint, &offset, vec);
+            return;
+        }
+    }
+    vec->z = 0.0f;
+    vec->y = 0.0f;
+    vec->x = 0.0f;
+}
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index 9ec78ede9d..a22e59f00f 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -645,7 +645,26 @@ bool ftKb_SpecialNMt_80106F9C(Fighter_GObj* gobj)
     return true;
 }
 
-/// #ftKb_SpecialNMt_80106FEC
+bool ftKb_SpecialNMt_80106FEC(Fighter_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Fighter* fp = gobj->user_data;
+        switch (fp->motion_id) {
+        case ftKb_MS_MtSpecialNStart:
+        case ftKb_MS_MtSpecialNLoop:
+        case ftKb_MS_MtSpecialNLoopFull:
+        case ftKb_MS_MtSpecialNEnd:
+        case ftKb_MS_MtSpecialAirNStart:
+        case ftKb_MS_MtSpecialAirNLoop:
+        case ftKb_MS_MtSpecialAirNLoopFull:
+        case ftKb_MS_MtSpecialAirNEnd:
+            return false;
+        default:
+            return true;
+        }
+    }
+    return true;
+}
 
 /// @brief Clears Mewtwo copy ability state and effects.
 void ftKb_SpecialNMt_80107040(Fighter_GObj* gobj)
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
index 80c5525089..6a8dfcb919 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
@@ -1256,7 +1256,19 @@ void ftPp_SpecialHi_801227AC(Fighter_GObj* gobj)
 
 /// #ftPp_SpecialHi_8012280C
 
-/// #ftPp_SpecialHi_80122898
+void ftPp_SpecialHi_80122898(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    PAD_STACK(16);
+    if (fp->fv.pp.x2230_b0) {
+        efLib_DestroyAll(gobj);
+        fp = gobj->user_data;
+        fp->fv.pp.x2230_b0 = false;
+        fp->death2_cb = NULL;
+        fp->take_dmg_cb = NULL;
+        ftPartSetRotX(gobj->user_data, 0, 0.0f);
+    }
+}
 
 /// #ftPp_SpecialLw_Enter
 
