diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index b3f8f5f267..daa9ffca3a 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -190,6 +190,25 @@ typedef struct {
     /*  +C ip+DE0 */ Vec3 pos;
 } itFoxLaser_ItemVars;
 
+typedef struct {
+    /* padding from 0xDD4 to 0xDF4 */ char _pad0[0xDF4 - 0xDD4];
+    /* item+0xDF4 */ Item_GObj* xDF4;
+    /* padding from 0xDF8 to 0xE18 */ char _pad1[0xE18 - 0xDF8];
+    /* item+0xE18 */ s32 xE18;
+    /* item+0xE1C */ s32 xE1C;
+    /* item+0xE20 */ s32 xE20;
+    /* item+0xE24 */ s32 xE24;
+    /* item+0xE28 */ s32 xE28;
+    /* item+0xE2C */ s32 xE2C;
+    /* item+0xE30 */ s16 xE30;
+    /* item+0xE32 */ s16 xE32_pad;
+    /* item+0xE34 */ s32 xE34_pad;
+    /* item+0xE38 */ s32 xE38;
+    /* item+0xE3C */ s32 xE3C_pad;
+    /* item+0xE40 */ Item_GObj* xE40;
+    /* item+0xE44 */ Item_GObj* xE44;
+} itArwingLaser_ItemVars;
+
 typedef struct FoxLaserAttr {
     /* +0 */ float lifetime; // [35]
     /* +4 */ float scale;    // [3]
diff --git a/src/melee/it/items/itarwinglaser.c b/src/melee/it/items/itarwinglaser.c
index a7bf526d8c..1f9d209b42 100644
--- a/src/melee/it/items/itarwinglaser.c
+++ b/src/melee/it/items/itarwinglaser.c
@@ -3,6 +3,7 @@
 #include <placeholder.h>
 #include <platform.h>
 
+#include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
 
@@ -11,9 +12,9 @@ void it_802E7A4C(Item_GObj* gobj)
     Item* item;
     int unused[2];
 
-    item = *(Item**) ((u8*) gobj + 0x2C);
+    item = GET_ITEM(gobj);
     it_8026B3A8(gobj);
-    Item_80268E5C(gobj, *(int*) ((u8*) item + 0xE38), 2);
+    Item_80268E5C(gobj, item->xDD4_itemVar.arwinglaser.xE38, 2);
 }
 
 bool it_802E8418(Item_GObj* gobj)
diff --git a/src/melee/it/types.h b/src/melee/it/types.h
index 4bccab9118..8202a849ab 100644
--- a/src/melee/it/types.h
+++ b/src/melee/it/types.h
@@ -556,6 +556,7 @@ struct Item {
         itFoods_ItemVars foods;
         itFoxBlaster_ItemVars foxblaster;
         itFoxIllusion_ItemVars foxillusion;
+        itArwingLaser_ItemVars arwinglaser;
         itFoxLaser_ItemVars foxlaser;
         itFreeze_ItemVars freeze;
         itFreezer_ItemVars freezer;
