diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index adf6e9889d..38c6f29967 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -10278,8 +10278,8 @@ grStadium_801D2BEC = .text:0x801D2BEC; // type:function size:0x18C scope:global
 grStadium_801D2D78 = .text:0x801D2D78; // type:function size:0xAC scope:global
 grStadium_801D2E24 = .text:0x801D2E24; // type:function size:0xAC scope:global
 fn_801D2ED0 = .text:0x801D2ED0; // type:function size:0x100
-fn_801D2FD0 = .text:0x801D2FD0; // type:function size:0xB4
-fn_801D3084 = .text:0x801D3084; // type:function size:0xB4
+grStadium_801D2FD0 = .text:0x801D2FD0; // type:function size:0xB4
+grStadium_801D3084 = .text:0x801D3084; // type:function size:0xB4
 grStadium_801D3138 = .text:0x801D3138; // type:function size:0x198 scope:global
 grStadium_801D32D0 = .text:0x801D32D0; // type:function size:0x190 scope:global
 grStadium_801D3460 = .text:0x801D3460; // type:function size:0x3EC scope:global
diff --git a/src/melee/gr/grfzerocar.c b/src/melee/gr/grfzerocar.c
index 85782c5ed0..0e55041e70 100644
--- a/src/melee/gr/grfzerocar.c
+++ b/src/melee/gr/grfzerocar.c
@@ -8,9 +8,11 @@
 #include "gr/ground.h"
 #include "gr/types.h"
 #include "lb/lb_00F9.h"
+#include "sc/types.h"
 
 #include <dolphin/mtx.h>
 #include <baselib/aobj.h>
+#include <baselib/gobj.h>
 #include <baselib/jobj.h>
 #include <baselib/object.h>
 
@@ -61,13 +63,11 @@ static inline void setup_car_child(HSD_JObj* parent, s16 ext_count, s32 offset,
 
     jobj = Ground_801C13D0(ext_count, 0);
     if (jobj != NULL) {
-        u8* entry = (u8*) archive->unk4->unk8 + offset;
-        void* anm_p = *(void**) (entry + 4);
-        if (anm_p != NULL) {
-            void* mat_p = *(void**) (entry + 8);
-            if (mat_p != NULL) {
-                grAnime_801C6C0C(jobj, *(HSD_AnimJoint**) anm_p,
-                                 *(HSD_MatAnimJoint**) mat_p, NULL);
+        DynamicModelDesc* entry =
+            (DynamicModelDesc*) ((u8*) archive->unk4->unk8 + offset);
+        if (entry->anims != NULL) {
+            if (entry->matanims != NULL) {
+                grAnime_801C6C0C(jobj, *entry->anims, *entry->matanims, NULL);
                 HSD_JObjReqAnimAllByFlags(jobj, 0x497, 0.0f);
                 HSD_ForeachAnim(jobj, JOBJ_TYPE, 0x76a4, HSD_AObjSetRate,
                                 AOBJ_ARG_AF, 1.0);
@@ -80,24 +80,30 @@ static inline void setup_car_child(HSD_JObj* parent, s16 ext_count, s32 offset,
     HSD_JObjSetScaleY(parent, scale_factor);
 }
 
+static inline void multiplyScale(HSD_JObj* jobj, f32 scale)
+{
+    Vec3 scl;
+    HSD_JObjGetScale(jobj, &scl);
+    scl.x *= scale;
+    scl.y *= scale;
+    scl.z *= scale;
+    HSD_JObjSetScale(jobj, &scl);
+}
+
 void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
 {
     HSD_JObj* root;
     HSD_JObj* child;
     HSD_JObj* tra_jobj;
     f32 ground_scale;
-    s16 ext_count;
     s32 offset;
     s16* data_ptr;
-    grFZeroCarEntry* tbl;
     int i;
     Vec3 scl_local;
     Vec3 trans_local;
     Quaternion rot_local;
 
-    PAD_STACK(0x18);
-
-    root = gobj->hsd_obj;
+    root = GET_JOBJ(gobj);
     ground_scale = Ground_801C0498();
 
     scl_local = grFZeroCar_803B7E50.scale;
@@ -115,12 +121,9 @@ void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
     tra_jobj->scl = NULL;
     HSD_JObjAddNext(child, tra_jobj);
 
-    ext_count = (s16) count;
-    offset = ext_count * 0x34;
     data_ptr = (s16*) data;
-    tbl = grFZeroCar_803E0BD8;
 
-    for (i = 0; i < 30; i++, data_ptr++, tbl++) {
+    for (i = 0; i < 30; i++) {
         s32 idx;
         s32 active;
         f32 scale_factor;
@@ -130,7 +133,7 @@ void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
         HSD_JObj* c3;
 
         if (i != 0) {
-            child = Ground_801C13D0(*data_ptr, 0);
+            child = Ground_801C13D0(data_ptr[i], 0);
             if (child != NULL) {
                 tra_jobj = HSD_JObjAlloc();
                 HSD_ASSERT(186, tra_jobj);
@@ -150,34 +153,26 @@ void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
         rot_local = grFZeroCar_803B7E50.rotate;
         HSD_JObjSetTranslate(child, &trans_local);
         HSD_JObjSetRotation(child, &rot_local);
+        multiplyScale(tra_jobj, ground_scale);
 
-        {
-            Vec3 scl;
-            HSD_JObjGetScale(tra_jobj, &scl);
-            scl.x *= ground_scale;
-            scl.y *= ground_scale;
-            scl.z *= ground_scale;
-            HSD_JObjSetScale(tra_jobj, &scl);
-        }
-
-        idx = tbl->children[0];
+        idx = grFZeroCar_803E0BD8[i].children[0];
         scale_factor = 1.0f;
+        active = 0;
         c0 = NULL;
         c1 = NULL;
         c2 = NULL;
         c3 = NULL;
 
-        active = 0;
         if (idx != -1) {
             active = 1;
         }
-        if (tbl->children[1] != -1) {
+        if (grFZeroCar_803E0BD8[i].children[1] != -1) {
             active++;
         }
-        if (tbl->children[2] != -1) {
+        if (grFZeroCar_803E0BD8[i].children[2] != -1) {
             active++;
         }
-        if (tbl->children[3] != -1) {
+        if (grFZeroCar_803E0BD8[i].children[3] != -1) {
             active++;
         }
 
@@ -200,7 +195,7 @@ void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
             c0 = j;
         }
 
-        idx = tbl->children[1];
+        idx = grFZeroCar_803E0BD8[i].children[1];
         if (idx != -1) {
             HSD_JObj* j = child;
             while (j != NULL && idx != 0) {
@@ -210,7 +205,7 @@ void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
             c1 = j;
         }
 
-        idx = tbl->children[2];
+        idx = grFZeroCar_803E0BD8[i].children[2];
         if (idx != -1) {
             HSD_JObj* j = child;
             while (j != NULL && idx != 0) {
@@ -220,7 +215,7 @@ void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
             c2 = j;
         }
 
-        idx = tbl->children[3];
+        idx = grFZeroCar_803E0BD8[i].children[3];
         if (idx != -1) {
             HSD_JObj* j = child;
             while (j != NULL && idx != 0) {
@@ -231,19 +226,19 @@ void grFZeroCar_801CAFBC(HSD_GObj* gobj, void* data, s32 count, s32 mode)
         }
 
         if (c0 != NULL) {
-            setup_car_child(c0, ext_count, offset, scale_factor);
+            setup_car_child(c0, count, (s16) count * 0x34, scale_factor);
         }
 
         if (c1 != NULL) {
-            setup_car_child(c1, ext_count, offset, scale_factor);
+            setup_car_child(c1, count, (s16) count * 0x34, scale_factor);
         }
 
         if (c2 != NULL) {
-            setup_car_child(c2, ext_count, offset, scale_factor);
+            setup_car_child(c2, count, (s16) count * 0x34, scale_factor);
         }
 
         if (c3 != NULL) {
-            setup_car_child(c3, ext_count, offset, scale_factor);
+            setup_car_child(c3, count, (s16) count * 0x34, scale_factor);
         }
     }
 }
diff --git a/src/melee/gr/grpstadium.c b/src/melee/gr/grpstadium.c
index c17291b7cf..bf829ebadf 100644
--- a/src/melee/gr/grpstadium.c
+++ b/src/melee/gr/grpstadium.c
@@ -1,5 +1,7 @@
 #include "grpstadium.h"
 
+#include "baselib/forward.h"
+
 #include "dolphin/gx/GXStruct.h"
 
 #include <sysdolphin/baselib/archive.h>
@@ -1209,7 +1211,7 @@ HSD_GObj* grStadium_801D2D78(void)
     ImageDescWrapper* wrapper;
 
     temp_r3 = GObj_Create(0x11, 0x12, 0);
-    GObj_SetupGXLinkMax(temp_r3, fn_801D2FD0, 3);
+    GObj_SetupGXLinkMax(temp_r3, grStadium_801D2FD0, 3);
     wrapper = HSD_MemAlloc(0x1C);
     GObj_InitUserData(temp_r3, 3, HSD_Free, wrapper);
     memzero(&wrapper->desc, sizeof(wrapper->desc));
@@ -1222,7 +1224,7 @@ HSD_GObj* grStadium_801D2E24(void)
 {
     HSD_GObj* gobj = GObj_Create(0x11, 0x12, 0);
     ImageDescWrapper* wrapper;
-    GObj_SetupGXLinkMax(gobj, fn_801D3084, 3);
+    GObj_SetupGXLinkMax(gobj, grStadium_801D3084, 3);
     wrapper = HSD_MemAlloc(sizeof(ImageDescWrapper));
     GObj_InitUserData(gobj, 3, HSD_Free, wrapper);
     memzero(&wrapper->desc, sizeof(HSD_ImageDesc));
@@ -1256,27 +1258,27 @@ void fn_801D2ED0(HSD_GObj* gobj, int unused)
     gp2->u.display.xF8_0 = true;
 }
 
-void fn_801D2FD0(Ground_GObj* gobj, int unused)
+void grStadium_801D2FD0(Ground_GObj* gobj, int unused)
 {
-    ImageDescWrapper* wrapper = GET_WRAPPER(gobj);
     HSD_GObj* vision_gobj;
     Ground* gp2;
+    ImageDescWrapper* wrapper = GET_WRAPPER(gobj);
+    ImageDescWrapper* copy;
 
-    if (wrapper->flag) {
-        return;
-    }
-
-    lb_800122C8(&wrapper->desc, 0, 0x24, 0);
-    wrapper->flag = true;
+    copy = wrapper;
+    if (!wrapper->flag) {
+        lb_800122C8(&copy->desc, 0, 36, 0);
+        wrapper->flag = true;
 
-    vision_gobj = Ground_801C2BA4(PsType_Display);
-    HSD_ASSERT(0x6EA, vision_gobj);
-    gp2 = GET_GROUND(vision_gobj);
-    HSD_ASSERT(0x6EB, gp2);
-    gp2->u.display.xF8_0 = true;
+        vision_gobj = Ground_801C2BA4(PsType_Display);
+        HSD_ASSERT(0x6EA, vision_gobj);
+        gp2 = GET_GROUND(vision_gobj);
+        HSD_ASSERT(0x6EB, gp2);
+        gp2->u.display.xF8_0 = true;
+    }
 }
 
-void fn_801D3084(HSD_GObj* gobj, int unused)
+void grStadium_801D3084(HSD_GObj* gobj, int unused)
 {
     void* new_var2; // Permuter slop
     ImageDescWrapper* wrapper = GET_WRAPPER(gobj);
@@ -1911,6 +1913,8 @@ void fn_801D4220(int un0, int un1, UNK_T un2, bool cancelflag)
     gp->u.stadium.xC4_b1 = false;
 }
 
+#pragma push
+#pragma dont_inline on
 bool grStadium_801D42B8(void)
 {
     HSD_GObj* map_gobj;
@@ -1927,11 +1931,7 @@ bool grStadium_801D42B8(void)
         grDatFiles_801C6478(gp->u.stadium.xCC, gp->u.stadium.xC8);
     return true;
 }
-
-inline bool grStadium_801D42B8_dontinline(void)
-{
-    return grStadium_801D42B8();
-}
+#pragma pop
 
 Ground* grStadium_801D4354(Ground_GObj* gobj)
 {
@@ -2153,7 +2153,7 @@ void grStadium_801D4548(Ground_GObj* gobj)
         }
         break;
     case 1:
-        if (grStadium_801D42B8_dontinline()) {
+        if (grStadium_801D42B8()) {
             temp_r31->u.stadium.xDC = 2;
             return;
         }
diff --git a/src/melee/gr/grpstadium.h b/src/melee/gr/grpstadium.h
index 1fb3d408a7..66cf8feaf1 100644
--- a/src/melee/gr/grpstadium.h
+++ b/src/melee/gr/grpstadium.h
@@ -64,8 +64,8 @@ typedef enum {
 /* 1D2D78 */ HSD_GObj* grStadium_801D2D78(void);
 /* 1D2E24 */ HSD_GObj* grStadium_801D2E24(void);
 /* 1D2ED0 */ void fn_801D2ED0(HSD_GObj*, int);
-/* 1D2FD0 */ void fn_801D2FD0(HSD_GObj*, int);
-/* 1D3084 */ void fn_801D3084(HSD_GObj*, int);
+/* 1D2FD0 */ void grStadium_801D2FD0(HSD_GObj*, int);
+/* 1D3084 */ void grStadium_801D3084(HSD_GObj*, int);
 /* 1D3138 */ HSD_TObj* grStadium_801D3138(Ground_GObj*, HSD_ImageDesc*,
                                           HSD_MObj**);
 /* 1D32D0 */ bool grStadium_801D32D0(Ground_GObj*);
