diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index f6bd4a5b0c..5f4988ffbc 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -234,7 +234,17 @@ void ftKb_SpecialAirHi1_Anim(Fighter_GObj* gobj)
     }
 }
 
-/// #ftKb_SpecialAirHi2_Anim
+void ftKb_SpecialAirHi2_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirHi3, 0xA, 0.0F, 1.0F,
+                                  0.0F, NULL);
+        fp->pre_hitlag_cb = efLib_PauseAll;
+        fp->post_hitlag_cb = efLib_ResumeAll;
+    }
+}
 
 void ftKb_SpecialAirHi3_Anim(Fighter_GObj* gobj) {}
 
@@ -2517,7 +2527,27 @@ void ftKb_MrSpecialAirN_Coll(Fighter_GObj* gobj)
 
 static void fn_800F98F4(Fighter_GObj*);
 
-/// #ftKb_SpecialNLg_800F951C
+void ftKb_SpecialNLg_800F951C(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter* fp2;
+    s32 motion_id;
+    f32 anim_frame;
+
+    ftCommon_8007D5D4(fp);
+
+    fp2 = gobj->user_data;
+    motion_id = 0x190;
+    anim_frame = fp->cur_anim_frame;
+
+    if (fp2->fv.kb.hat.kind != 0) {
+        motion_id = 0x201;
+    }
+
+    Fighter_ChangeMotionState(gobj, motion_id, 0x5000, anim_frame, 1.0F, 0.0F,
+                              NULL);
+    fp->accessory4_cb = fn_800F9260;
+}
 
 void ftKb_SpecialNLg_800F9598(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c b/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c
index e8e6783b08..69615172a9 100644
--- a/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c
+++ b/src/melee/ft/chara/ftYoshi/ftYs_SpecialHi.c
@@ -86,9 +86,41 @@ void ftYs_SpecialS_8012E270(Fighter_GObj* gobj)
     fp->death2_cb = NULL;
 }
 
-/// #ftYs_SpecialHi_Enter
+void ftYs_SpecialHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+
+    u8 _[24];
+
+    fp->mv.ys.specialhi.x0 = 0;
+    fp->mv.ys.specialhi.x4 = 0;
+    fp->fv.ys.x2238 = NULL;
+    fp->throw_flags = 0;
+    fp->cmd_vars[0] = 0;
+    Fighter_ChangeMotionState(gobj, ftYs_MS_SpecialHi, 0, 0.0F, 1.0F, 0.0F,
+                              NULL);
+    fp = GET_FIGHTER(gobj);
+    fp->accessory4_cb = fn_8012E110;
+    ftAnim_8006EBA4(gobj);
+}
 
-/// #ftYs_SpecialAirHi_Enter
+void ftYs_SpecialAirHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+
+    u8 _[24];
+
+    fp->mv.ys.specialhi.x0 = 0;
+    fp->mv.ys.specialhi.x4 = 0;
+    fp->fv.ys.x2238 = NULL;
+    fp->throw_flags = 0;
+    fp->cmd_vars[0] = 0;
+    Fighter_ChangeMotionState(gobj, ftYs_MS_SpecialAirHi, 0, 0.0F, 1.0F,
+                              0.0F, NULL);
+    fp = GET_FIGHTER(gobj);
+    fp->accessory4_cb = fn_8012E110;
+    ftAnim_8006EBA4(gobj);
+}
 
 static void setAccessory4Callback(Fighter_GObj* gobj)
 {
