diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index b992be1fc9..801c58a12a 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -33,8 +33,14 @@ typedef struct itClimbersIce_ItemVars {
 } itClimbersIce_ItemVars;
 
 typedef struct itClimbersIceAttributes {
-    /* +00 */ u8 pad_00[0x28];
+    /* +00 */ u8 pad_00[0x4];
+    /* +04 */ f32 x4;
+    /* +08 */ u8 pad_08[0x4];
+    /* +0C */ f32 xC;
+    /* +10 */ u8 pad_10[0x18];
     /* +28 */ f32 x28;
+    /* +2C */ s32 x2C;
+    /* +30 */ u32 x30;
 } itClimbersIceAttributes;
 
 typedef struct itNessbat_ItemVars {
@@ -207,12 +213,8 @@ typedef struct {
     /* padding from 0xDD4 to 0xDF4 */ char _pad0[0xDF4 - 0xDD4];
     /* item+0xDF4 */ Item_GObj* xDF4;
     /* padding from 0xDF8 to 0xE18 */ char _pad1[0xE18 - 0xDF8];
-    /* item+0xE18 */ s32 xE18;
-    /* item+0xE1C */ s32 xE1C;
-    /* item+0xE20 */ s32 xE20;
-    /* item+0xE24 */ s32 xE24;
-    /* item+0xE28 */ s32 xE28;
-    /* item+0xE2C */ s32 xE2C;
+    /* item+0xE18 */ Vec3 xE18;
+    /* item+0xE24 */ Vec3 xE24;
     /* item+0xE30 */ s16 xE30;
     /* item+0xE32 */ s16 xE32_pad;
     /* item+0xE34 */ s32 xE34_pad;
@@ -701,6 +703,23 @@ typedef struct itGamewatchchefAttributes {
     /* +10 */ itGamewatchchefAttrEntry entries[1];
 } itGamewatchchefAttributes;
 
+typedef struct itGamewatchparachuteAttributes {
+    /* +0 */ void* x0;
+} itGamewatchparachuteAttributes;
+
+typedef struct itHinoarashi_ItemVars {
+    /* xDD4 */ u8 _0[0x60 - 0x0];
+    /* xE34 */ UnkFlagStruct x60;
+    /* xE35 */ u8 x61[0x64 - 0x61];
+    /* xE38 */ f32 x64;
+} itHinoarashi_ItemVars;
+
+typedef struct itHinoarashiAttributes {
+    /* +0 */ f32 x0;
+    /* +4 */ f32 x4;
+    /* +8 */ f32 x8;
+} itHinoarashiAttributes;
+
 typedef struct itTools_ItemVars {
     /* +0 ip+DD4 */ s32 x0;
 } itTools_ItemVars;
diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index 5b9d9e365d..7316b220d0 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -382,8 +382,16 @@ typedef struct itFlipper_ItemVars {
 } itFlipper_ItemVars;
 
 typedef struct itFlipper_DatAttrs {
-    /* +0 */ s32 x0;
-    /* +4 */ s32 x4;
+    /* +0  */ s32 x0;
+    /* +4  */ s32 x4;
+    /* +8  */ s32 x8;
+    /* +C  */ f32 xC;
+    /* +10 */ f32 x10;
+    /* +14 */ s32 x14;
+    /* +18 */ f32 x18;
+    /* +1C */ f32 x1C;
+    /* +20 */ f32 x20;
+    /* +24 */ f32 x24;
 } itFlipper_DatAttrs;
 
 typedef struct itHarisen_DatAttrs {
@@ -508,12 +516,13 @@ typedef struct {
 typedef struct itTincle_ItemVars {
     /*  +0 ip+DD4 */ u8 pad[0x20];
     /* +20 ip+DF4 */ s32 x20;
-    /* +24 ip+DF8 */ u8 pad1[0x8];
+    /* +24 ip+DF8 */ s32 x24;
+    /* +28 ip+DFC */ u8 pad1[0x4];
     /* +2C ip+E00 */ s32 x2C;
     /* +30 ip+E04 */ f32 x30;
     /* +34 ip+E08 */ u8 pad1b[0x4];
     /* +38 ip+E0C */ f32 x38;
-    /* +3C ip+E10 */ u8 pad1c[0x4];
+    /* +3C ip+E10 */ f32 x3C;
     /* +40 ip+E14 */ f32 x40;
     /* +44 ip+E18 */ f32 x44;
     /* +48 ip+E1C */ f32 x48;
@@ -527,7 +536,11 @@ typedef struct itTincle_ItemVars {
 } itTincle_ItemVars;
 
 typedef struct itTincleAttributes {
-    /* 0x00 */ u8 pad[0x40];
+    /* 0x00 */ u8 pad[0x24];
+    /* 0x24 */ f32 x24;
+    /* 0x28 */ s32 x28;
+    /* 0x2C */ s32 x2C;
+    /* 0x30 */ u8 pad1b[0x10];
     /* 0x40 */ s32 x40;
     /* 0x44 */ f32 x44;
     /* 0x48 */ f32 x48;
@@ -575,7 +588,8 @@ typedef struct {
     /* ip+E09 */ char pad_E09[0xE0C - 0xE09];
     /* ip+E0C */ int xE0C;
     /* ip+E10 */ int xE10;
-    /* ip+E14 */ char pad_E14[0xE1C - 0xE14];
+    /* ip+E14 */ int xE14;
+    /* ip+E18 */ int xE18;
     /* ip+E1C */ unsigned char xE1C_b0 : 1;
     /* ip+E1C */ unsigned char xE1C_b1 : 1;
     /* ip+E1C */ unsigned char xE1C_b2 : 1;
@@ -1070,7 +1084,8 @@ typedef struct itMewtwoShadowball_ItemVars {
     /* ip+DF4 */ s32 x20;
     /* ip+DF8 */ f32 x24;
     /* ip+DFC */ s32 x28;
-    /* ip+E00 */ f32 x2C[6];
+    /* ip+E00 */ HSD_GObj* x2C;
+    /* ip+E04 */ f32 x30[5];
     /* ip+E18 */ f32 x44;
     /* ip+E1C */ f32 x48;
     /* ip+E20 */ s32 x4C;
@@ -1284,7 +1299,10 @@ typedef struct itWhiteBea_ItemVars {
 typedef struct itWhiteBeaAttributes {
     /* +0 */ s32* x0;
     /* +4 */ f32 x4;
-    /* +8 */ char pad_x8[0x8];
+    /* +8 */ s16 x8;
+    /* +A */ s16 xA;
+    /* +C */ s16 xC;
+    /* +E */ s16 xE;
     /* +10 */ f32 x10;
     /* +14 */ s16 x14;
 } itWhiteBeaAttributes;
diff --git a/src/melee/it/items/itarwinglaser.c b/src/melee/it/items/itarwinglaser.c
index 1f9d209b42..58d1556bb2 100644
--- a/src/melee/it/items/itarwinglaser.c
+++ b/src/melee/it/items/itarwinglaser.c
@@ -3,9 +3,49 @@
 #include <placeholder.h>
 #include <platform.h>
 
+#include "ef/efsync.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
+#include "it/it_2725.h"
 #include "it/item.h"
+#include "lb/lb_00B0.h"
+
+static void itArwinglaser_UnkMotion2_Anim(Item_GObj*);
+static void itArwinglaser_UnkMotion3_Anim(Item_GObj*);
+
+void it_802E79C8(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+
+    switch (ip->xDD4_itemVar.arwinglaser.xE38) {
+    case 0:
+    case 2: {
+        Item* ip2;
+        Item* ip3;
+
+        ip->xDD4_itemVar.arwinglaser.xDF4 = NULL;
+
+        if (ip->xDD4_itemVar.arwinglaser.xE40 != NULL) {
+            ip2 = GET_ITEM(ip->xDD4_itemVar.arwinglaser.xE40);
+            if (ip2->xDD4_itemVar.arwinglaser.xDF4 == gobj) {
+                ip2->xDD4_itemVar.arwinglaser.xDF4 = NULL;
+            }
+        }
+        ip->xDD4_itemVar.arwinglaser.xE40 = NULL;
+
+        if (ip->xDD4_itemVar.arwinglaser.xE44 != NULL) {
+            ip3 = GET_ITEM(ip->xDD4_itemVar.arwinglaser.xE44);
+            if (ip3->xDD4_itemVar.arwinglaser.xDF4 == gobj) {
+                ip3->xDD4_itemVar.arwinglaser.xDF4 = NULL;
+            }
+        }
+        ip->xDD4_itemVar.arwinglaser.xE44 = NULL;
+        return;
+    }
+    case 1:
+        return;
+    }
+}
 
 void it_802E7A4C(Item_GObj* gobj)
 {
@@ -17,6 +57,38 @@ void it_802E7A4C(Item_GObj* gobj)
     Item_80268E5C(gobj, item->xDD4_itemVar.arwinglaser.xE38, 2);
 }
 
+static void itArwinglaser_UnkMotion2_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+
+    ip->xDD4_itemVar.arwinglaser.xE24 = ip->xDD4_itemVar.arwinglaser.xE18;
+    if (it_80272C6C(gobj) == 0) {
+        ip->xDD4_itemVar.arwinglaser.xE30 = 1;
+    }
+    it_80273130(gobj);
+}
+
+static void itArwinglaser_UnkMotion3_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+
+    ip->xDD4_itemVar.arwinglaser.xE24 = ip->xDD4_itemVar.arwinglaser.xE18;
+    ip->xDD4_itemVar.arwinglaser.xE30 = 1;
+    it_80273130(gobj);
+}
+
+bool it_802E838C(Item_GObj* gobj)
+{
+    Vec3 sp10;
+    Item* ip = GET_ITEM(gobj);
+
+    lb_8000B1CC(ip->xBBC_dynamicBoneTable->bones[4], NULL, &sp10);
+    ip->xDD4_itemVar.arwinglaser.xE18 = sp10;
+    efSync_Spawn(0x40E, gobj, &sp10);
+    Item_8026AE84(ip, 0x74, 0x7FU, 0x40U);
+    return true;
+}
+
 bool it_802E8418(Item_GObj* gobj)
 {
     return true;
diff --git a/src/melee/it/items/itclimbersice.c b/src/melee/it/items/itclimbersice.c
index 116d3a7a68..5c9447b557 100644
--- a/src/melee/it/items/itclimbersice.c
+++ b/src/melee/it/items/itclimbersice.c
@@ -15,6 +15,8 @@
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
+#include "MSL/math.h"
+#include "it/itcoll.h"
 
 /* 2C23D4 */ static bool itClimbersBlizzard_UnkMotion0_Coll(Item_GObj* gobj);
 
diff --git a/src/melee/it/items/itclimbersice.h b/src/melee/it/items/itclimbersice.h
index 574a869dd7..09974f1eb4 100644
--- a/src/melee/it/items/itclimbersice.h
+++ b/src/melee/it/items/itclimbersice.h
@@ -14,7 +14,7 @@
 /* 2C16F8 */ void it_802C16F8(Item_GObj*);
 /* 2C17DC */ void it_802C17DC(Item_GObj*);
 /* 2C17FC */ void it_2725_Logic90_Destroyed(Item_GObj*);
-/* 2C1854 */ UNK_RET it_802C1854(UNK_PARAMS);
+/* 2C1854 */ bool it_802C1854(void);
 /* 2C1950 */ void it_802C1950(Item_GObj*);
 /* 2C198C */ bool itClimbersice_UnkMotion0_Anim(Item_GObj* gobj);
 /* 2C1A14 */ void itClimbersice_UnkMotion0_Phys(Item_GObj* gobj);
diff --git a/src/melee/it/items/itclimbersstring.c b/src/melee/it/items/itclimbersstring.c
index e51ee5f82a..3a7aeea477 100644
--- a/src/melee/it/items/itclimbersstring.c
+++ b/src/melee/it/items/itclimbersstring.c
@@ -1,12 +1,43 @@
 #include "itclimbersstring.h"
 
+#include "ft/chara/ftPopo/ftPp_SpecialS.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
+#include "it/itCharItems.h"
 #include "it/item.h"
 
+#include <dolphin/os/OSError.h>
+#include <baselib/gobjplink.h>
+
 /* 2C33B8 */ static void it_802C33B8(Item* item);
 /* 2C29E8 */ static void fn_802C29E8(Item_GObj* gobj);
 
+void it_802C2750(Item_GObj* gobj)
+{
+    int unused[2];
+    Item* ip = GET_ITEM(gobj);
+    if (ip != NULL) {
+        if (gobj != NULL) {
+            if (ip->owner != NULL) {
+                ftPp_SpecialS_8012114C(ip->owner);
+            }
+            ip->owner = NULL;
+            ip->xDD4_itemVar.it_266F.xC = NULL;
+            {
+                ItemLink* cur;
+                for (cur = (ItemLink*) ip->xDD4_itemVar.it_266F.x4;
+                     cur != NULL;)
+                {
+                    HSD_GObj* cur_gobj = cur->gobj;
+                    cur = cur->next;
+                    HSD_GObjPLink_80390228(cur_gobj);
+                }
+            }
+        }
+        Item_8026A8EC(gobj);
+    }
+}
+
 void fn_802C28B8(Item_GObj* gobj)
 {
     it_802C33B8(GET_ITEM(gobj));
diff --git a/src/melee/it/items/itflipper.c b/src/melee/it/items/itflipper.c
index 880fb647f0..078f6769fe 100644
--- a/src/melee/it/items/itflipper.c
+++ b/src/melee/it/items/itflipper.c
@@ -33,7 +33,33 @@ void itFlipper_Logic20_Spawned(Item_GObj* gobj)
 
 /// #it_80290A7C
 
-/// #it_80290C38
+void it_80290C38(Item_GObj* gobj, Vec3* pos, f32 angle)
+{
+    Item* ip = GET_ITEM(gobj);
+    itFlipper_DatAttrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    f32 speed = angle + attrs->x1C;
+
+    if (pos->x > ip->pos.x) {
+        if (pos->y > ip->pos.y) {
+            ip->xDD4_itemVar.flipper.xDE4 += speed;
+        } else {
+            ip->xDD4_itemVar.flipper.xDE4 -= speed;
+        }
+    } else {
+        if (pos->y > ip->pos.y) {
+            ip->xDD4_itemVar.flipper.xDE4 -= speed;
+        } else {
+            ip->xDD4_itemVar.flipper.xDE4 += speed;
+        }
+    }
+
+    if (ip->xDD4_itemVar.flipper.xDE4 > attrs->x24) {
+        ip->xDD4_itemVar.flipper.xDE4 = attrs->x24;
+    }
+    if (ip->xDD4_itemVar.flipper.xDE4 < -attrs->x24) {
+        ip->xDD4_itemVar.flipper.xDE4 = -attrs->x24;
+    }
+}
 
 /// #it_80290CE8
 
@@ -146,11 +172,64 @@ bool itFlipper_UnkMotion3_Anim(Item_GObj* gobj)
 
 /// #itFlipper_UnkMotion3_Anim
 
-/// #itFlipper_UnkMotion3_Phys
+void itFlipper_UnkMotion3_Phys(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    ItemAttr* item_attr = ip->xCC_item_attr;
+    it_80272860(gobj, item_attr->x10_fall_speed,
+                item_attr->x14_fall_speed_max);
+    if (ip->xDD4_itemVar.flipper.xDD4 <= 0xA) {
+        ip->x40_vel.x *=
+            (f32) ip->xDD4_itemVar.flipper.xDD4 /
+            (1.0f + (f32) ip->xDD4_itemVar.flipper.xDD4);
+        ip->x40_vel.y *=
+            (f32) ip->xDD4_itemVar.flipper.xDD4 /
+            (1.0f + (f32) ip->xDD4_itemVar.flipper.xDD4);
+    }
+    it_80274658(gobj, it_804D6D28->x68_float);
+}
 
-/// #itFlipper_UnkMotion3_Coll
+bool itFlipper_UnkMotion3_Coll(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itFlipper_DatAttrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    s32 coll = it_8026DAA8(gobj);
+    if (coll != 0) {
+        if (coll & 0xC) {
+            ip->x40_vel.x *= -attrs->xC;
+        }
+        ip->x40_vel.y *= attrs->xC;
+        if (coll & 2) {
+            ip->x40_vel.x *= attrs->xC;
+        }
+        ip->x40_vel.y *= -attrs->xC;
+        if (coll & 1) {
+            ip->x40_vel.x *= attrs->x10;
+        }
+        ip->x40_vel.y *= -attrs->x10;
+    }
+    return false;
+}
 
-/// #it_80291254
+void it_80291254(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itFlipper_DatAttrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    ip->xD5C = 0;
+    ip->x40_vel.x = 0.0f;
+    ip->x40_vel.y = 0.0f;
+    ip->xDD1_flag.b1 = 1;
+    ip->xDD4_itemVar.flipper.xDD8 = 1;
+    it_80275444(gobj);
+    it_802725D4(gobj);
+    if (ip->xDD4_itemVar.flipper.xDE8 != 0) {
+        it_8027542C(gobj);
+    }
+    ip->xD44_lifeTimer = (f32) attrs->x8;
+    it_8026B3A8(gobj);
+    it_802756E0(gobj);
+    it_8029131C(gobj);
+}
 
 void it_8029131C(Item_GObj* gobj)
 {
@@ -198,10 +277,60 @@ bool itFlipper_UnkMotion6_Coll(Item_GObj* gobj)
 
 /// #it_3F14_Logic20_DmgDealt
 
-/// #it_3F14_Logic20_Clanked
+#pragma dont_inline on
+bool it_3F14_Logic20_Clanked(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    itFlipper_DatAttrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(8);
+    if (ip->xDD4_itemVar.flipper.xDD8 == 0) {
+        if (ip->xDD4_itemVar.flipper.xDD4 >= 6) {
+            itColl_BounceOffVictim(gobj);
+        } else {
+            ip->x40_vel.x = -1.0f * ip->x40_vel.x;
+            ip->x40_vel.y =
+                -1.0f * ip->x40_vel.y + it_804D6D28->x60_float;
+        }
+    } else {
+        ip->xDD4_itemVar.flipper.xDDC = attrs->x14;
+        it_80290CE8(gobj);
+        it_80272560(gobj, 0);
+        it_80272560(gobj, 1);
+        it_802756D0(gobj);
+        if (ip->xDD4_itemVar.flipper.xDE8 == 0) {
+            it_80290DD4(gobj, ip->xC38, &ip->xCD4);
+        }
+    }
+    return false;
+}
 
-/// #it_3F14_Logic20_HitShield
+bool it_3F14_Logic20_HitShield(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    itFlipper_DatAttrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(8);
+    if (ip->xDD4_itemVar.flipper.xDD8 == 0) {
+        if (ip->xDD4_itemVar.flipper.xDD4 >= 6) {
+            itColl_BounceOffVictim(gobj);
+        } else {
+            ip->x40_vel.x = -1.0f * ip->x40_vel.x;
+            ip->x40_vel.y =
+                -1.0f * ip->x40_vel.y + it_804D6D28->x60_float;
+        }
+    } else {
+        ip->xDD4_itemVar.flipper.xDDC = attrs->x14;
+        it_80290CE8(gobj);
+        it_80272560(gobj, 0);
+        it_80272560(gobj, 1);
+        it_802756D0(gobj);
+        if (ip->xDD4_itemVar.flipper.xDE8 == 0) {
+            it_80290DD4(gobj, ip->xC38, &ip->xCD4);
+        }
+    }
+    return false;
+}
 
+#pragma dont_inline reset
 bool itFlipper_Logic20_Reflected(Item_GObj* gobj)
 {
     return it_80273030(gobj);
diff --git a/src/melee/it/items/itflipper.h b/src/melee/it/items/itflipper.h
index 797973f630..6d1f1a3a3f 100644
--- a/src/melee/it/items/itflipper.h
+++ b/src/melee/it/items/itflipper.h
@@ -10,8 +10,8 @@
 /* 290938 */ HSD_GObj* it_80290938(HSD_JObj*);
 /* 290A20 */ void itFlipper_Logic20_Spawned(Item_GObj*);
 /* 290A7C */ UNK_RET it_80290A7C(UNK_PARAMS);
-/* 290C38 */ UNK_RET it_80290C38(UNK_PARAMS);
-/* 290CE8 */ UNK_RET it_80290CE8(UNK_PARAMS);
+/* 290C38 */ void it_80290C38(Item_GObj*, Vec3*, f32);
+/* 290CE8 */ void it_80290CE8(Item_GObj*);
 /* 290DD4 */ void it_80290DD4(Item_GObj* gobj, s32 kind, Vec3* pos);
 /* 290E78 */ void it_80290E78(Item_GObj*);
 /* 290EC8 */ bool itFlipper_UnkMotion0_Anim(Item_GObj* gobj);
diff --git a/src/melee/it/items/itfreeze.h b/src/melee/it/items/itfreeze.h
index 9418698860..7070094233 100644
--- a/src/melee/it/items/itfreeze.h
+++ b/src/melee/it/items/itfreeze.h
@@ -7,7 +7,7 @@
 
 #include "it/items/types.h"
 
-/* 28EB88 */ UNK_RET it_8028EB88(UNK_PARAMS);
+/* 28EB88 */ Item_GObj* it_8028EB88(Item_GObj*, Vec3*, Item*, f32, f32);
 /* 28EC98 */ void it_8028EC98(Item_GObj* gobj, f32 vel);
 /* 28ECE0 */ void it_8028ECE0(Item_GObj* gobj);
 /* 28ECF0 */ Item* it_8028ECF0(Item_GObj* gobj, Vec3* v);
diff --git a/src/melee/it/items/itgamewatchchef.c b/src/melee/it/items/itgamewatchchef.c
index ad3d1aa003..1c84885e55 100644
--- a/src/melee/it/items/itgamewatchchef.c
+++ b/src/melee/it/items/itgamewatchchef.c
@@ -9,8 +9,39 @@
 #include "it/item.h"
 
 #include <baselib/jobj.h>
+#include "db/db.h"
 
-/// #it_802C837C
+HSD_GObj* it_802C837C(Item_GObj* parent, Vec3* pos, enum_t kind, u32 arg3,
+                       float facing_dir)
+{
+    Item_GObj* gobj;
+    SpawnItem spawn;
+    PAD_STACK(4);
+
+    spawn.kind = kind;
+    spawn.prev_pos = *pos;
+    spawn.prev_pos.z = 0.0f;
+    it_8026BB68(parent, &spawn.pos);
+    spawn.facing_dir = facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.z = 0.0f;
+    spawn.vel.y = 0.0f;
+    spawn.vel.x = 0.0f;
+    spawn.x0_parent_gobj = parent;
+    spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+    spawn.x44_flag.b0 = true;
+    spawn.x40 = 0;
+    gobj = Item_80268B18(&spawn);
+    if (gobj != NULL) {
+        itGamewatchchefAttributes* attrs =
+            GET_ITEM(gobj)->xC4_article_data->x4_specialAttributes;
+        it_802C84A0(gobj, arg3);
+        db_80225DD8(gobj, parent);
+        it_8027CE64(gobj, parent, attrs->x0);
+        return gobj;
+    }
+    return NULL;
+}
 
 bool itGameWatchChef_Logic112_DmgDealt(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itgamewatchparachute.c b/src/melee/it/items/itgamewatchparachute.c
index 67cccd608f..45bdcde202 100644
--- a/src/melee/it/items/itgamewatchparachute.c
+++ b/src/melee/it/items/itgamewatchparachute.c
@@ -7,6 +7,39 @@
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
+#include "db/db.h"
+#include "it/it_2725.h"
+
+HSD_GObj* it_802C6C38(Item_GObj* parent, Vec3* pos, enum_t part,
+                       float facing_dir)
+{
+    Item_GObj* gobj;
+    SpawnItem spawn;
+    PAD_STACK(4);
+
+    spawn.kind = 0x75;
+    spawn.prev_pos = *pos;
+    spawn.pos = spawn.prev_pos;
+    spawn.facing_dir = facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.z = 0.0f;
+    spawn.vel.y = 0.0f;
+    spawn.vel.x = 0.0f;
+    spawn.x0_parent_gobj = parent;
+    spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+    spawn.x44_flag.b0 = true;
+    spawn.x40 = 0;
+    gobj = Item_80268B18(&spawn);
+    if (gobj != NULL) {
+        itGamewatchparachuteAttributes* attrs =
+            GET_ITEM(gobj)->xC4_article_data->x4_specialAttributes;
+        Item_8026AB54(gobj, parent, part);
+        db_80225DD8(gobj, parent);
+        it_8027CE64(gobj, parent, attrs->x0);
+        return gobj;
+    }
+    return NULL;
+}
 
 void it_802C6D6C(Item_GObj* item_gobj)
 {
@@ -62,3 +95,32 @@ void it_802C6E50(Item_GObj* gobj)
     it_8026BB44(gobj);
     Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
 }
+
+bool itGamewatchparachute_UnkMotion1_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    bool remove;
+    PAD_STACK(8);
+
+    if (ip->x5CC_currentAnimFrame == 30.0f) {
+        if (ip->owner != NULL) {
+            ftGw_AttackAirN_ItemParachuteSetFlag(ip->owner);
+        }
+        return true;
+    }
+
+    if (ip->owner != NULL) {
+        remove = ftGw_AttackAirN_ItemCheckParachuteRemove(ip->owner);
+    } else {
+        remove = true;
+    }
+
+    if (remove) {
+        if (GET_ITEM(gobj)->owner != NULL) {
+            ftGw_AttackAirN_ItemParachuteSetFlag(GET_ITEM(gobj)->owner);
+        }
+        return true;
+    }
+
+    return false;
+}
diff --git a/src/melee/it/items/itgamewatchparachute.h b/src/melee/it/items/itgamewatchparachute.h
index 257260a780..ce62722bbb 100644
--- a/src/melee/it/items/itgamewatchparachute.h
+++ b/src/melee/it/items/itgamewatchparachute.h
@@ -12,6 +12,7 @@
 /* 2C6DD8 */ void it_802C6DD8(Item_GObj*);
 /* 2C6DF8 */ void itGameWatchParachute_Logic74_PickedUp(Item_GObj*);
 /* 2C6E50 */ void it_802C6E50(Item_GObj*);
+/* 2C6E8C */ bool itGamewatchparachute_UnkMotion1_Anim(Item_GObj*);
 /* 2C6F20 */ void itGameWatchParachute_Logic74_EvtUnk(Item_GObj*, Item_GObj*);
 /* 3F78F8 */ extern ItemStateTable it_803F78F8[];
 
diff --git a/src/melee/it/items/ithinoarashi.c b/src/melee/it/items/ithinoarashi.c
index 744ee9aa59..3abff008c8 100644
--- a/src/melee/it/items/ithinoarashi.c
+++ b/src/melee/it/items/ithinoarashi.c
@@ -1,14 +1,32 @@
 #include "ithinoarashi.h"
 
+#include "placeholder.h"
+
+#include "baselib/gobj.h"
 #include "ef/eflib.h"
+#include "ef/efsync.h"
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
 
+/* 2D5D7C */ void it_802D5D7C(HSD_GObj*);
 /* 2D60C8 */ static bool itHinoarashi_UnkMotion2_Anim(Item_GObj* gobj);
 
+void it_802D5CF8(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    ip->xDAC_itcmd_var0 = 0;
+    ip->xDB0_itcmd_var1 = 0;
+    ip->xDD4_itemVar.hinoarashi.x60.b0 = false;
+    ip->xDD4_itemVar.hinoarashi.x60.b1 = false;
+    ip->xDD4_itemVar.hinoarashi.x64 = 0.0f;
+    it_80279CDC(gobj, ip->scl);
+    Item_8026AE84(ip, 0x2734, 0x7FU, 0x40U);
+    it_802D5E4C(gobj);
+}
+
 void it_802D5E4C(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -52,6 +70,55 @@ void it_802D5F0C(Item_GObj* gobj)
     it_8027A118(gobj, it_802D5F34);
 }
 
+void it_802D5F34(HSD_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
+    ip->entered_hitlag = efLib_PauseAll;
+    ip->exited_hitlag = efLib_ResumeAll;
+    it_802756E0((Item_GObj*) gobj);
+    it_802754D4((Item_GObj*) gobj);
+    it_802754BC((Item_GObj*) gobj);
+    it_80273454((Item_GObj*) gobj);
+    ip->on_accessory = it_802D5D7C;
+    ip->xDD1_flag.b1 = true;
+}
+
+bool itHinoarashi_UnkMotion1_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itHinoarashiAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+    if (!it_80272C6C(gobj)) {
+        return true;
+    }
+    if (ip->xDD4_itemVar.hinoarashi.x60.b0 &&
+        !ip->xDD4_itemVar.hinoarashi.x60.b1)
+    {
+        if (!ip->xDD4_itemVar.hinoarashi.x64) {
+            ip->xDD4_itemVar.hinoarashi.x64 = attr->x8;
+            it_802752D8(gobj, 3, 0x2735);
+        }
+        ip->xDD4_itemVar.hinoarashi.x64--;
+    }
+    it_8027A160(ip->xBBC_dynamicBoneTable->bones[1], ip);
+    return false;
+}
+
+bool it_2725_Logic27_DmgReceived(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itHinoarashiAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+    if (ip->xC9C >= attr->x4) {
+        it_80279D38(gobj);
+        it_802762BC(gobj->user_data);
+        it_802756D0(gobj);
+        efLib_DestroyAll(gobj);
+        Item_80268E5C(gobj, 2, ITEM_ANIM_UPDATE);
+        it_80273670(gobj, 0, 0.0f);
+    }
+    return false;
+}
+
 void it_802D61A8(Item_GObj* gobj, Item_GObj* ref_gobj)
 {
     it_8026B894(gobj, ref_gobj);
@@ -63,3 +130,13 @@ void itHinoarashi_UnkMotion2_Phys(Item_GObj* gobj)
     ItemAttr* attrs = ip->xCC_item_attr;
     it_80272860(gobj, attrs->x10_fall_speed, attrs->x14_fall_speed_max);
 }
+
+void it_802D6674(Item_GObj* gobj, s32 arg1)
+{
+    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    f32 scale = 1.0f;
+    Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    efSync_Spawn(arg1 + 0x44E, gobj, jobj, &ip->facing_dir, &scale);
+    it_8026B3A8(gobj);
+}
diff --git a/src/melee/it/items/ithinoarashi.h b/src/melee/it/items/ithinoarashi.h
index 5f481eb133..9acf68c8f9 100644
--- a/src/melee/it/items/ithinoarashi.h
+++ b/src/melee/it/items/ithinoarashi.h
@@ -13,11 +13,13 @@
 /* 2D5EEC */ void it_802D5EEC(Item_GObj*);
 /* 2D5F0C */ void it_802D5F0C(Item_GObj*);
 /* 2D5F34 */ void it_802D5F34(HSD_GObj*);
+/* 2D5FC8 */ bool itHinoarashi_UnkMotion1_Anim(Item_GObj*);
 /* 2D6084 */ void itHinoarashi_UnkMotion1_Phys(Item_GObj*);
 /* 2D60A4 */ bool itHinoarashi_UnkMotion1_Coll(Item_GObj*);
 /* 2D60D0 */ void itHinoarashi_UnkMotion2_Phys(Item_GObj* gobj);
 /* 2D6100 */ bool it_2725_Logic27_DmgReceived(Item_GObj*);
 /* 2D61A8 */ void it_802D61A8(Item_GObj*, Item_GObj*);
+/* 2D6674 */ void it_802D6674(Item_GObj*, s32);
 /* 3F8270 */ extern ItemStateTable it_803F8270[];
 /* 3F82A0 */ extern ItemStateTable it_803F82A0[];
 
diff --git a/src/melee/it/items/itklap.c b/src/melee/it/items/itklap.c
index 2508a3e909..4e3868a30b 100644
--- a/src/melee/it/items/itklap.c
+++ b/src/melee/it/items/itklap.c
@@ -86,7 +86,22 @@ void it_802E1C4C(Item_GObj* gobj)
     Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
 }
 
-/// #it_802E1C84
+void it_802E1C84(Item_GObj* gobj)
+{
+    Vec3 pos;
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* child;
+
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->xDCC_flag.b3 = true;
+
+    child = HSD_JObjGetChild(jobj);
+    lb_8000B1CC(child, NULL, &pos);
+    ip->pos = pos;
+    Item_80268E5C(gobj, 2, ITEM_ANIM_UPDATE);
+}
 
 bool itKlap_UnkMotion2_Anim(Item_GObj* gobj)
 {
@@ -108,7 +123,20 @@ bool it_2725_Logic10_DmgReceived(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802E1E94
+void it_802E1E94(Item_GObj* gobj)
+{
+    f32 rand_val;
+    Item* ip = GET_ITEM(gobj);
+    it_802762BC(ip);
+    if (it_8027B798(gobj, &ip->x40_vel)) {
+        it_802762BC(ip);
+    }
+    rand_val = HSD_Randf();
+    ip->xDD4_itemVar.klap.x24 = (rand_val - 0.5f) * 0.17453292f;
+    rand_val = HSD_Randf();
+    ip->xDD4_itemVar.klap.x28 = (rand_val - 0.5f) * 0.17453292f;
+    Item_80268E5C(gobj, 4, ITEM_ANIM_UPDATE);
+}
 
 bool itKlap_UnkMotion4_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itklap.h b/src/melee/it/items/itklap.h
index ecf80c43f8..7cbd32cd7d 100644
--- a/src/melee/it/items/itklap.h
+++ b/src/melee/it/items/itklap.h
@@ -15,7 +15,7 @@
 /* 2E1970 */ void itKlap_UnkMotion1_Phys(Item_GObj* gobj);
 /* 2E19FC */ bool itKlap_UnkMotion1_Coll(Item_GObj* gobj);
 /* 2E1C4C */ void it_802E1C4C(Item_GObj* gobj);
-/* 2E1C84 */ UNK_RET it_802E1C84(UNK_PARAMS);
+/* 2E1C84 */ void it_802E1C84(Item_GObj* gobj);
 /* 2E1D24 */ bool itKlap_UnkMotion2_Anim(Item_GObj* gobj);
 /* 2E1D2C */ void itKlap_UnkMotion2_Phys(Item_GObj* gobj);
 /* 2E1E54 */ bool itKlap_UnkMotion2_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/itlipstickspore.c b/src/melee/it/items/itlipstickspore.c
index 632e114bfa..eebb87168e 100644
--- a/src/melee/it/items/itlipstickspore.c
+++ b/src/melee/it/items/itlipstickspore.c
@@ -14,6 +14,8 @@
 
 #include <trigf.h>
 #include <baselib/gobj.h>
+#include "db/db.h"
+#include "ef/efsync.h"
 
 /* 29A5F8 */ static bool itLipstickspore_UnkMotion1_Coll(Item_GObj* gobj);
 
diff --git a/src/melee/it/items/itlipstickspore.h b/src/melee/it/items/itlipstickspore.h
index 41395a143d..9c50079f66 100644
--- a/src/melee/it/items/itlipstickspore.h
+++ b/src/melee/it/items/itlipstickspore.h
@@ -10,7 +10,7 @@
 /* 29A114 */ void it_8029A114(Fighter_GObj*, Vec*, int, float);
 /* 29A218 */ void it_8029A218(HSD_GObj* owner, Vec3* pos, f32 facing_dir,
                               s32 arg4);
-/* 29A31C */ UNK_RET it_8029A31C(UNK_PARAMS);
+/* 29A31C */ void it_8029A31C(Item_GObj*, Fighter_GObj*, Item*);
 /* 29A498 */ void it_8029A498(Item_GObj* gobj);
 /* 29A504 */ bool itLipstickspore_UnkMotion0_Anim(Item_GObj* gobj);
 /* 29A524 */ void itLipstickspore_UnkMotion0_Phys(Item_GObj* gobj);
diff --git a/src/melee/it/items/itmewtwoshadowball.c b/src/melee/it/items/itmewtwoshadowball.c
index 1ce85f1310..b37674e317 100644
--- a/src/melee/it/items/itmewtwoshadowball.c
+++ b/src/melee/it/items/itmewtwoshadowball.c
@@ -1,13 +1,17 @@
 #include "itmewtwoshadowball.h"
 
 #include "ef/eflib.h"
+#include "ft/chara/ftKirby/ftKb_Init.h"
+#include "ft/chara/ftMewtwo/ftMt_SpecialN.h"
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
+#include "lb/lbvector.h"
 
 #include <math.h>
+#include "MSL/trigf.h"
 
 /// #it_802C4D10
 
@@ -39,7 +43,32 @@ bool it_802C4F50(Item_GObj* gobj, CollData* cd)
 
 /// #it_802C53F0
 
-/// #it_2725_Logic101_Destroyed
+void it_2725_Logic101_Destroyed(Item_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Item* ip = gobj->user_data;
+        efLib_DestroyAll(gobj);
+        if (ip->xDD4_itemVar.mewtwoshadowball.x10[1] == 0) {
+            if (ip->xDD4_itemVar.mewtwoshadowball.x2C != NULL &&
+                ip->owner == ip->xDD4_itemVar.mewtwoshadowball.x2C)
+            {
+                switch (ip->kind) {
+                case It_Kind_Mewtwo_ShadowBall:
+                    ftMt_SpecialN_SetNULL(
+                        ip->xDD4_itemVar.mewtwoshadowball.x2C);
+                    break;
+                case It_Kind_Kirby_MewtwoShadowBall:
+                    ftKb_SpecialNMt_80107040(
+                        ip->xDD4_itemVar.mewtwoshadowball.x2C);
+                    break;
+                }
+            }
+            ip->xDD4_itemVar.mewtwoshadowball.x2C = NULL;
+        }
+        ip->owner = NULL;
+        ip->xDC8_word.flags.x13 = 0;
+    }
+}
 
 void it_802C573C(Item_GObj* gobj)
 {
@@ -166,7 +195,21 @@ bool itMewtwoShadowball_Logic101_HitShield(Item_GObj* arg0)
     return true;
 }
 
-/// #it_2725_Logic101_ShieldBounced
+bool it_2725_Logic101_ShieldBounced(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    PAD_STACK(8);
+    lbVector_Mirror(&ip->x40_vel, &ip->xC58);
+    ip->xDD4_itemVar.mewtwoshadowball.x4.x =
+        atan2f(ip->x40_vel.y, ip->x40_vel.x);
+    while (ip->xDD4_itemVar.mewtwoshadowball.x4.x < 0.0f) {
+        ip->xDD4_itemVar.mewtwoshadowball.x4.x += M_TAU;
+    }
+    while (ip->xDD4_itemVar.mewtwoshadowball.x4.x > M_TAU) {
+        ip->xDD4_itemVar.mewtwoshadowball.x4.x -= M_TAU;
+    }
+    return false;
+}
 
 void itMewtwoShadowball_Logic101_EvtUnk(Item_GObj* gobj, Item_GObj* ref_gobj)
 {
diff --git a/src/melee/it/items/itpatapata.c b/src/melee/it/items/itpatapata.c
index 7c32700837..8542a9f1a3 100644
--- a/src/melee/it/items/itpatapata.c
+++ b/src/melee/it/items/itpatapata.c
@@ -109,7 +109,30 @@ void it_802E0D9C(Item_GObj* gobj)
     ip->facing_dir = -ip->facing_dir;
 }
 
-/// #itPatapata_UnkMotion3_Anim
+bool itPatapata_UnkMotion3_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (!it_80272C6C(gobj)) {
+        s32 facing;
+        if (ip->facing_dir == -1.0f) {
+            facing = -1;
+        } else {
+            facing = 1;
+        }
+        mpCollSetFacingDir(&ip->x378_itemColl, facing);
+        it_8027C56C(gobj, ip->facing_dir);
+        {
+            f32 vel_x = ip->x40_vel.x;
+            if ((vel_x < 0.0f && ip->facing_dir == 1.0f) ||
+                (vel_x > 0.0f && ip->facing_dir == -1.0f))
+            {
+                ip->x40_vel.x *= -1.0f;
+            }
+        }
+        it_802E1648(gobj, 2, 2);
+    }
+    return false;
+}
 
 bool it_2725_Logic4_DmgReceived(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itsscope.c b/src/melee/it/items/itsscope.c
index adac4d0b5a..b844fff352 100644
--- a/src/melee/it/items/itsscope.c
+++ b/src/melee/it/items/itsscope.c
@@ -68,6 +68,27 @@ s32 it_80291CF4(Item_GObj* gobj, s32 arg1)
     return 9;
 }
 
+s32 it_80291D38(Item_GObj* gobj, s32 arg1)
+{
+    Item* ip = GET_ITEM(gobj);
+    itSScopeAttributes* attrs = GET_ATTRS(ip);
+    switch (arg1) {
+    case 0:
+        return attrs->xC[0];
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+    case 8:
+        return attrs->xC[arg1];
+    case 9:
+        return attrs->xC[9];
+    }
+}
+
 /// #it_80291DAC
 
 void it_80291F14(Item_GObj* gobj, int charge_level)
@@ -103,25 +124,27 @@ void it_80291F14(Item_GObj* gobj, int charge_level)
     PAD_STACK(16);
 }
 
+static s32 it_80291D38_outline(Item_GObj* gobj, int charge_level)
+{
+    return it_80291D38(gobj, charge_level);
+}
+
 void it_80291FA8(Item_GObj* gobj, Vec3* pos, int charge_level, float scale)
 {
     Item* ip = GET_ITEM(gobj);
-    int cost = it_80291D38(gobj, charge_level);
+    s32 cost = it_80291D38_outline(gobj, charge_level);
+    PAD_STACK(8);
     ip->xD4C -= cost;
     if (ip->xD4C < 0) {
         ip->xD4C = 0;
     }
     it_80298DEC(ip->owner, pos, charge_level, scale);
-
-    PAD_STACK(8);
 }
 
 void it_80292030(Item_GObj* gobj)
 {
-    Item* temp_r4 = GET_ITEM(gobj);
-    temp_r4->x40_vel.z = 0.0f;
-    temp_r4->x40_vel.y = 0.0f;
-    temp_r4->x40_vel.x = 0.0f;
+    Item* ip = GET_ITEM(gobj);
+    itResetVelocity(ip);
     it_8026B390(gobj);
     Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
 }
diff --git a/src/melee/it/items/itsscope.h b/src/melee/it/items/itsscope.h
index e30821720a..00b8311145 100644
--- a/src/melee/it/items/itsscope.h
+++ b/src/melee/it/items/itsscope.h
@@ -14,7 +14,7 @@
 /* 291C98 */ void itSScope_Logic21_Spawned(Item_GObj*);
 /* 291CCC */ itSScopeAttributes* it_80291CCC(Item_GObj* gobj, Vec3* pos);
 /* 291CF4 */ s32 it_80291CF4(Item_GObj*, s32);
-/* 291D38 */ int it_80291D38(Item_GObj*, int);
+/* 291D38 */ s32 it_80291D38(Item_GObj*, s32);
 /* 291DAC */ int it_80291DAC(Item_GObj*, int);
 /* 291F14 */ void it_80291F14(Item_GObj*, int);
 /* 291FA8 */ void it_80291FA8(Item_GObj*, Vec3*, int, float);
diff --git a/src/melee/it/items/ittincle.c b/src/melee/it/items/ittincle.c
index e5c36228bb..9bbe2e538f 100644
--- a/src/melee/it/items/ittincle.c
+++ b/src/melee/it/items/ittincle.c
@@ -95,7 +95,22 @@ bool itTincle_UnkMotion2_Coll(Item_GObj* gobj)
 
 /// #itTincle_UnkMotion2_Coll
 
-/// #it_802EBD14
+void it_802EBD14(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(8);
+    ip->xDD4_itemVar.tincle.x30 = (f32) ip->msid;
+    ip->xDD4_itemVar.tincle.x44 = ip->x40_vel.y;
+    ip->xDD4_itemVar.tincle.x48 = ip->xDD4_itemVar.tincle.x38;
+    ip->xDD4_itemVar.tincle.x40 = (f32) ip->xDD4_itemVar.tincle.x2C;
+    ip->x40_vel.z = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->x40_vel.y = -ip->xDD4_itemVar.tincle.x54;
+    ip->xDD4_itemVar.tincle.x38 = ip->xDD4_itemVar.tincle.x54 / 10.0f;
+    ip->xDD4_itemVar.tincle.x20 = 0x13;
+    ip->xDD4_itemVar.tincle.x4C = ip->pos.y;
+    Item_80268E5C(gobj, 3, ITEM_ANIM_UPDATE);
+}
 
 bool itTincle_UnkMotion3_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ittincle.h b/src/melee/it/items/ittincle.h
index fbfe98fcce..335a67d26f 100644
--- a/src/melee/it/items/ittincle.h
+++ b/src/melee/it/items/ittincle.h
@@ -22,7 +22,7 @@
 /* 2EBAF0 */ bool itTincle_UnkMotion2_Anim(Item_GObj* gobj);
 /* 2EBB38 */ void itTincle_UnkMotion2_Phys(Item_GObj* gobj);
 /* 2EBCD0 */ bool itTincle_UnkMotion2_Coll(Item_GObj* gobj);
-/* 2EBD14 */ UNK_RET it_802EBD14(UNK_PARAMS);
+/* 2EBD14 */ void it_802EBD14(Item_GObj*);
 /* 2EBDC8 */ bool itTincle_UnkMotion3_Anim(Item_GObj* gobj);
 /* 2EBDD0 */ void itTincle_UnkMotion3_Phys(Item_GObj* gobj);
 /* 2EBE30 */ bool itTincle_UnkMotion3_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/ittools.c b/src/melee/it/items/ittools.c
index 86d9d0b4e5..32ea063464 100644
--- a/src/melee/it/items/ittools.c
+++ b/src/melee/it/items/ittools.c
@@ -6,10 +6,35 @@
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
+#include "it/item.h"
 
 #include <baselib/jobj.h>
 
-/// #it_802EEFA8
+Item_GObj* it_802EEFA8(Vec3* pos, s32 arg1, f32 facing_dir)
+{
+    Item_GObj* gobj;
+    SpawnItem spawn;
+    PAD_STACK(12);
+
+    spawn.kind = It_Kind_Tools;
+    spawn.prev_pos = *pos;
+    spawn.pos = *pos;
+    spawn.facing_dir = facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.z = 0.0f;
+    spawn.vel.y = 0.0f;
+    spawn.vel.x = 0.0f;
+    spawn.x0_parent_gobj = NULL;
+    spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+    spawn.x44_flag.b0 = true;
+    spawn.x40 = 0;
+    gobj = Item_80268B18(&spawn);
+    if (gobj != NULL) {
+        it_802EF098(gobj, arg1);
+        return gobj;
+    }
+    return NULL;
+}
 
 bool itTools_Logic22_DmgDealt(Item_GObj* gobj)
 {
@@ -17,7 +42,19 @@ bool itTools_Logic22_DmgDealt(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802EF098
+void it_802EF098(Item_GObj* gobj, s32 arg1)
+{
+    Item* ip = GET_ITEM(gobj);
+    itToolsAttrEntry* attrs =
+        (itToolsAttrEntry*) ip->xC4_article_data->x4_specialAttributes;
+    ip->xDD4_itemVar.tools.x0 = arg1;
+    it_80275158(gobj, attrs->x0);
+    ip->xD5C = 0;
+    ip->x40_vel.x = 0.0f;
+    ip->x40_vel.y = -attrs[ip->xDD4_itemVar.tools.x0].x10;
+    it_8026B3A8(gobj);
+    Item_80268E5C(gobj, ip->xDD4_itemVar.tools.x0, ITEM_ANIM_UPDATE);
+}
 
 /// #itTools_UnkMotion4_Anim
 
@@ -42,7 +79,19 @@ bool itTools_UnkMotion4_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802EF320
+void it_802EF320(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itToolsAttrEntry* attrs =
+        (itToolsAttrEntry*) ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(8);
+    it_802756D0(gobj);
+    it_8026B3A8(gobj);
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->xD44_lifeTimer = (s32) attrs->x4 & 0xFFFE;
+    Item_80268E5C(gobj, ip->xDD4_itemVar.tools.x0 + 5, ITEM_ANIM_UPDATE);
+}
 
 bool itTools_UnkMotion9_Anim(Item_GObj* gobj)
 {
@@ -63,7 +112,29 @@ bool itTools_UnkMotion9_Anim(Item_GObj* gobj)
 
 void itTools_UnkMotion9_Phys(Item_GObj* gobj) {}
 
-/// #itTools_UnkMotion9_Coll
+static inline void itTools_UnkMotion9_Coll_inline(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itToolsAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    it_802756D0(gobj);
+    it_8026B3A8(gobj);
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->xD44_lifeTimer = (f32) ((s32) attrs->x4 & 0xFFFE);
+    Item_80268E5C(gobj, ip->xDD4_itemVar.tools.x0 + 5, ITEM_ANIM_UPDATE);
+}
+
+bool itTools_UnkMotion9_Coll(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(8);
+    if ((ip->x40_vel.x != 0.0f || ip->x40_vel.y != 0.0f) &&
+        (it_8026DAA8(gobj) & 3))
+    {
+        itTools_UnkMotion9_Coll_inline(gobj);
+    }
+    return false;
+}
 
 /// #it_802EF548
 
@@ -73,11 +144,51 @@ bool itTools_Logic22_DmgReceived(Item_GObj* gobj)
     return false;
 }
 
-/// #it_2725_Logic22_Clanked
+bool it_2725_Logic22_Clanked(Item_GObj* gobj)
+{
+    itToolsAttrEntry* attrs;
+    Item* ip = GET_ITEM(gobj);
+    attrs = (itToolsAttrEntry*) ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(16);
+    it_802756D0(gobj);
+    it_8026B3A8(gobj);
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->xD44_lifeTimer = (s32) attrs->x4 & 0xFFFE;
+    Item_80268E5C(gobj, ip->xDD4_itemVar.tools.x0 + 5, ITEM_ANIM_UPDATE);
+    return false;
+}
 
-/// #it_2725_Logic22_HitShield
+bool it_2725_Logic22_HitShield(Item_GObj* gobj)
+{
+    itToolsAttrEntry* attrs;
+    Item* ip = GET_ITEM(gobj);
+    attrs = (itToolsAttrEntry*) ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(16);
+    it_802756D0(gobj);
+    it_8026B3A8(gobj);
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->xD44_lifeTimer = (s32) attrs->x4 & 0xFFFE;
+    Item_80268E5C(gobj, ip->xDD4_itemVar.tools.x0 + 5, ITEM_ANIM_UPDATE);
+    return false;
+}
 
-/// #it_2725_Logic22_Absorbed
+bool it_2725_Logic22_Absorbed(Item_GObj* gobj)
+{
+    itToolsAttrEntry* attrs;
+    Item* ip = GET_ITEM(gobj);
+    attrs = (itToolsAttrEntry*) ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(24);
+    it_802756D0(gobj);
+    it_8026B3A8(gobj);
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+    ip->xD44_lifeTimer = (s32) attrs->x4 & 0xFFFE;
+    Item_80268E5C(gobj, ip->xDD4_itemVar.tools.x0 + 5, ITEM_ANIM_UPDATE);
+    ip->xD44_lifeTimer = 0.0f;
+    return false;
+}
 
 bool itTools_Logic22_ShieldBounced(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ittools.h b/src/melee/it/items/ittools.h
index abd44ca77d..3f543219c3 100644
--- a/src/melee/it/items/ittools.h
+++ b/src/melee/it/items/ittools.h
@@ -7,9 +7,9 @@
 
 #include "it/items/types.h"
 
-/* 2EEFA8 */ UNK_RET it_802EEFA8(UNK_PARAMS);
+/* 2EEFA8 */ Item_GObj* it_802EEFA8(Vec3*, s32, f32);
 /* 2EF074 */ bool itTools_Logic22_DmgDealt(Item_GObj*);
-/* 2EF098 */ UNK_RET it_802EF098(UNK_PARAMS);
+/* 2EF098 */ void it_802EF098(Item_GObj*, s32);
 /* 2EF128 */ bool itTools_UnkMotion4_Anim(Item_GObj* gobj);
 /* 2EF27C */ void itTools_UnkMotion4_Phys(Item_GObj* gobj);
 /* 2EF2BC */ bool itTools_UnkMotion4_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/itwhitebea.c b/src/melee/it/items/itwhitebea.c
index 6732075820..aa1f5af90c 100644
--- a/src/melee/it/items/itwhitebea.c
+++ b/src/melee/it/items/itwhitebea.c
@@ -11,6 +11,7 @@
 #include "mp/mpcoll.h"
 #include "mp/mplib.h"
 #include "sysdolphin/baselib/random.h"
+#include "MSL/math.h"
 
 void it_802E31F8(Item_GObj* gobj)
 {
@@ -153,7 +154,29 @@ void it_802E37A4(Item_GObj* gobj)
     }
 }
 
-/// #it_802E37BC
+void it_802E37BC(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    s32 facing;
+    it_8027B730(gobj);
+    ip->facing_dir = it_8026B684(&ip->pos);
+    it_8027C56C(gobj, ip->facing_dir);
+    if (ip->facing_dir == -1.0f) {
+        facing = -1;
+    } else {
+        facing = 1;
+    }
+    mpCollSetFacingDir(&ip->x378_itemColl, facing);
+    ip->xD5C = 0;
+    ip->xDC8_word.flags.x15 = 0;
+    it_8027542C(gobj);
+    it_80275270(gobj);
+    it_80274740(gobj);
+    ip->xDC8_word.flags.x19 = 1;
+    ip->xDD4_itemVar.whitebea.x3C = 1;
+    ip->xDD4_itemVar.whitebea.x40 = 0;
+    it_802E3DA0(gobj);
+}
 
 bool it_802E3884(Item_GObj* gobj)
 {
@@ -218,7 +241,32 @@ bool itWhitebea_UnkMotion0_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-/// #it_802E3AC8
+void it_802E3AC8(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itWhiteBeaAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    s32 facing;
+    s32 timer;
+    PAD_STACK(8);
+    ip->x40_vel.x = ip->facing_dir * ((f32*) attrs->x0)[1];
+    ip->x40_vel.z = 0.0f;
+    ip->x40_vel.y = 0.0f;
+    it_8027C56C(gobj, ip->facing_dir);
+    if (ip->facing_dir == -1.0f) {
+        facing = -1;
+    } else {
+        facing = 1;
+    }
+    mpCollSetFacingDir(&ip->x378_itemColl, facing);
+    it_802756E0(gobj);
+    timer = ip->xDD4_itemVar.whitebea.x3C;
+    if (timer < 2 || timer > attrs->xE) {
+        s32 range = ABS(attrs->xE - attrs->xC);
+        ip->xDD4_itemVar.whitebea.x3C = attrs->xC + HSD_Randi(range);
+    }
+    ip->xDD4_itemVar.whitebea.x44 = 4;
+    Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
+}
 
 bool itWhitebea_UnkMotion1_Anim(Item_GObj* gobj)
 {
@@ -242,7 +290,27 @@ bool itWhitebea_UnkMotion1_Anim(Item_GObj* gobj)
     return false;
 }
 
-/// #itWhitebea_UnkMotion1_Phys
+void itWhitebea_UnkMotion1_Phys(Item_GObj* gobj)
+{
+    Item* ip = gobj->user_data;
+    itWhiteBeaAttributes* attrs =
+        ip->xC4_article_data->x4_specialAttributes;
+    s32 range;
+    PAD_STACK(16);
+
+    if (ip->xDD4_itemVar.whitebea.x3C == 0) {
+        range = ABS(attrs->xA - attrs->x8);
+        ip->xDD4_itemVar.whitebea.x40 = attrs->x8 + HSD_Randi(range);
+        it_8027CAD8(gobj);
+        Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
+    } else {
+        ip->xDD4_itemVar.whitebea.x3C -= 1;
+    }
+    if (it_802750E8(gobj, 2) != 0) {
+        it_802E4190(gobj);
+    }
+    it_8027C8D0(&ip->x40_vel, &ip->x378_itemColl.floor.normal, ip->facing_dir);
+}
 
 bool itWhitebea_UnkMotion1_Coll(Item_GObj* gobj)
 {
@@ -298,7 +366,32 @@ bool itWhitebea_UnkMotion3_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-/// #it_802E3ED0
+void it_802E3ED0(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    s32 facing;
+    PAD_STACK(8);
+    it_802762BC(ip);
+    if (it_8027B798(gobj, &ip->x40_vel)) {
+        if (ip->x40_vel.y <= 0.2) {
+            ip->x40_vel.y = 0.2f;
+        } else {
+            ip->x40_vel.y *= 0.2f;
+        }
+        it_802762BC(ip);
+    } else {
+        it_802762B0(ip);
+        ip->x40_vel.x *= 0.2f;
+    }
+    ip->facing_dir = ip->init_facing_dir;
+    if (ip->facing_dir == -1.0f) {
+        facing = -1;
+    } else {
+        facing = 1;
+    }
+    mpCollSetFacingDir(&ip->x378_itemColl, facing);
+    Item_80268E5C(gobj, 4, 3);
+}
 
 bool itWhitebea_UnkMotion4_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/ityoshieggthrow.c b/src/melee/it/items/ityoshieggthrow.c
index 7dad500890..fdf56a71e1 100644
--- a/src/melee/it/items/ityoshieggthrow.c
+++ b/src/melee/it/items/ityoshieggthrow.c
@@ -5,6 +5,7 @@
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
+#include "ef/efasync.h"
 
 void it_802B2890(Item_GObj* gobj)
 {
@@ -19,6 +20,32 @@ void itYoshiEggThrow_Logic43_PickedUp(Item_GObj* gobj)
     Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
 }
 
+Item_GObj* it_802B2A10(Fighter_GObj* parent, Vec3* pos, s32 part,
+                        f32 facing_dir)
+{
+    SpawnItem spawn;
+    Item_GObj* gobj;
+    PAD_STACK(8);
+
+    spawn.kind = It_Kind_Yoshi_EggThrow;
+    spawn.prev_pos = *pos;
+    it_8026BB68(parent, &spawn.pos);
+    spawn.facing_dir = facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.z = 0.0f;
+    spawn.vel.y = 0.0f;
+    spawn.vel.x = 0.0f;
+    spawn.x0_parent_gobj = parent;
+    spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+    spawn.x44_flag.b0 = true;
+    spawn.x40 = 0;
+    gobj = Item_80268B18(&spawn);
+    if (gobj != NULL) {
+        Item_8026AB54(gobj, parent, part);
+    }
+    return gobj;
+}
+
 void it_802B2B08(Item_GObj* item_gobj)
 {
     Item* item = GET_ITEM((HSD_GObj*) item_gobj);
@@ -81,3 +108,27 @@ bool it_802B2C04(Item_GObj* gobj)
     }
     return false;
 }
+
+void it_802B2C38(HSD_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    f32* attrs = ip->xC4_article_data->x4_specialAttributes;
+    HSD_JObj* jobj = gobj->hsd_obj;
+    f32 scale;
+    PAD_STACK(12);
+    Item_80268E5C(gobj, 2, 0x12);
+    it_8026BB44(gobj);
+    it_80273598(gobj, 0xC, 0xA);
+    it_8026BD24(gobj);
+    it_8027518C(gobj);
+    it_80273454(gobj);
+    ip->xD44_lifeTimer = attrs[1];
+    efAsync_Spawn(gobj, &GET_ITEM(gobj)->xBC0, 1, 0x4CE, jobj);
+    scale = 1.0f;
+    efAsync_Spawn(gobj, &GET_ITEM(gobj)->xBC0, 4, 0x4CF, jobj, &scale);
+    Item_8026AE84(ip, 0x44618, 0x7F, 0x40);
+}
+
+/// #it_802B2D50
+
+/// #it_802B2E7C
diff --git a/src/melee/it/items/itzgshell.c b/src/melee/it/items/itzgshell.c
index 29bab09bf5..61d342d903 100644
--- a/src/melee/it/items/itzgshell.c
+++ b/src/melee/it/items/itzgshell.c
@@ -1,6 +1,7 @@
 #include "itzgshell.h"
 
 #include "baselib/jobj.h"
+#include "baselib/random.h"
 #include "ef/efasync.h"
 #include "gr/grzakogenerator.h"
 
@@ -95,7 +96,23 @@ void fn_802DDC8C(Item_GObj* gobj)
 
 /// #it_802DDEB4
 
-/// #it_802DE040
+void it_802DE040(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itGShell_Attrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    it_8027B730(gobj);
+    it_80275210(gobj);
+    ip->xDD4_itemVar.zgshell.xE08_b0 = 0;
+    ip->xDD4_itemVar.zgshell.xDFC = attrs->x28;
+    ip->xDD4_itemVar.zgshell.xE08_b1 = 1;
+    ip->xDD4_itemVar.zgshell.xE10 = 0;
+    ip->xDD4_itemVar.zgshell.xE14 = 0;
+    ip->xDD4_itemVar.zgshell.xE0C = 0;
+    ip->xDD4_itemVar.zgshell.xE1C_b0 = 0;
+    ip->xDD4_itemVar.zgshell.xE04 = attrs->x34.x;
+    it_80271590(gobj, 0, (HurtCapsule*) &ip->xDD4_itemVar.zgshell.x58_hurtInit);
+    it_802DE320(gobj);
+}
 
 void it_802DE0F0(Item_GObj* gobj)
 {
@@ -446,7 +463,22 @@ bool itZGShell_Logic11_Reflected(Item_GObj* gobj)
     return it_80273030(gobj);
 }
 
-/// #it_2725_Logic11_Clanked
+bool it_2725_Logic11_Clanked(Item_GObj* gobj)
+{
+    itGShell_Attrs* attrs;
+    Item* ip = GET_ITEM(gobj);
+    attrs = ip->xC4_article_data->x4_specialAttributes;
+    it_802756D0(gobj);
+    it_80275444(gobj);
+    ip->x40_vel.x = -ip->x40_vel.x * attrs->xC * HSD_Randf();
+    ip->x40_vel.y = attrs->x10;
+    it_802762BC(ip);
+    ip->xDD4_itemVar.zgshell.xE1C_b0 = 1;
+    ip->xDD4_itemVar.zgshell.xE04 = attrs->x34.x;
+    Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
+    it_80274C88(gobj);
+    return false;
+}
 
 /// #it_2725_Logic11_HitShield
 
diff --git a/src/melee/it/types.h b/src/melee/it/types.h
index 99096c157b..8b06eabfcb 100644
--- a/src/melee/it/types.h
+++ b/src/melee/it/types.h
@@ -568,6 +568,7 @@ struct Item {
         itHassam_ItemVars hassam;
         itHeart_ItemVars heart;
         itHeiho_ItemVars heiho;
+        itHinoarashi_ItemVars hinoarashi;
         itHitodeman_ItemVars hitodeman;
         itKabigon_ItemVars kabigon;
         itKinoko_ItemVars kinoko;
