diff --git a/src/melee/it/items/itnessyoyo.c b/src/melee/it/items/itnessyoyo.c
index 351e503e67..9fb6dc8b9b 100644
--- a/src/melee/it/items/itnessyoyo.c
+++ b/src/melee/it/items/itnessyoyo.c
@@ -796,38 +796,12 @@ void it_802C0010(Item_GObj* gobj, Vec3* vel)
     Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
 }
 
-#ifdef MWERKS_GEKKO
-
-#pragma push
-asm void it_2725_Logic59_EvtUnk(Item_GObj* gobj, Item_GObj* ref_gobj)
-{
-    // clang-format off
-    nofralloc
-    mflr r0
-    stw r0, 0x4(r1)
-    stwu r1, -0x20(r1)
-    stw r31, 0x1c(r1)
-    stw r30, 0x18(r1)
-    mr r30, r4
-    lwz r31, 0x2c(r3)
-    bl it_8026B894
-    lwz r0, 0xde4(r31)
-    cmplw r0, r30
-    lwz r0, 0x24(r1)
-    lwz r31, 0x1c(r1)
-    lwz r30, 0x18(r1)
-    addi r1, r1, 0x20
-    mtlr r0
-    blr
-    // clang-format on
-}
-#pragma pop
-
-#else
-
 void it_2725_Logic59_EvtUnk(Item_GObj* gobj, Item_GObj* ref_gobj)
 {
-    NOT_IMPLEMENTED;
+    Item* ip = GET_ITEM(gobj);
+    it_8026B894(gobj, ref_gobj);
+    {
+        HSD_GObj* volatile* owner_ptr = &ip->xDD4_itemVar.nessyoyo.x10;
+        (void) ((*owner_ptr == ref_gobj) ? 0 : 0);
+    }
 }
-
-#endif
