diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index c276b0d26b..4d29105dd2 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -2081,7 +2081,41 @@ void ftKb_SpecialAirNLoop_IASA(Fighter_GObj* gobj)
 
 /// #ftKb_SpecialAirNCaptureWait_IASA
 
-/// #ftKb_EatWalk_IASA
+void ftKb_EatWalk_IASA(Fighter_GObj* gobj)
+{
+    ftCo_JumpInput jump_input;
+    s32 var;
+    PAD_STACK(8);
+
+    jump_input = ftCo_Jump_GetInput(gobj);
+    if (jump_input != JumpInput_None) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        fp->mv.kb.specialhi.x4 = jump_input;
+        fp->mv.kb.specialhi.x0 = 0;
+        Fighter_ChangeMotionState(gobj, ftKb_MS_EatJump1,
+                                  Ft_MF_KeepGfx | Ft_MF_SkipModel |
+                                      Ft_MF_SkipMatAnim,
+                                  0, 1, 0, NULL);
+        ftKb_SpecialN_800F9070(gobj);
+        ftAnim_8006EBA4(gobj);
+        ftCommon_8007E2F4(fp, 0x1FF);
+        var = 1;
+    } else {
+        var = 0;
+    }
+    if (var == 0) {
+        if (ft_8008A1FC(gobj)) {
+            Fighter* fp = GET_FIGHTER(gobj);
+            Fighter_ChangeMotionState(gobj, ftKb_MS_EatWait,
+                                      Ft_MF_SkipModel, 0, 1, 0, NULL);
+            ftKb_SpecialN_800F9070(gobj);
+            ftAnim_8006EBA4(gobj);
+            ftCommon_8007E2F4(fp, 0x1FF);
+            return;
+        }
+        ftWalkCommon_800DFEC8(gobj, fn_800F64C8);
+    }
+}
 
 void ftKb_EatJump1_IASA(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
index 186d3b3f6d..903f824e8f 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
@@ -779,7 +779,30 @@ void ftKb_LkSpecialNEnd_Anim(Fighter_GObj* gobj)
 
 /// #ftKb_LkSpecialAirNLoop_Anim
 
-/// #ftKb_LkSpecialAirNEnd_Anim
+void ftKb_LkSpecialAirNEnd_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    float freefall;
+    PAD_STACK(24);
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftKb_SpecialNLk_UnsetArrow(gobj);
+        ftKb_SpecialNLk_UnsetBow(gobj);
+        switch (fp->fv.kb.hat.kind) {
+        case FTKIND_LINK:
+            freefall = da->specialn_lk_freefall_toggle;
+            break;
+        case FTKIND_CLINK:
+            freefall = da->specialn_cl_freefall_toggle;
+            break;
+        }
+        if (freefall == 0.0f) {
+            ftCo_Fall_Enter(gobj);
+        } else {
+            ftCo_80096900(gobj, 1, 0, true, 1.0f, freefall);
+        }
+    }
+}
 
 /// #ftKb_LkSpecialNStart_IASA
 
@@ -861,7 +884,34 @@ long ftKb_SpecialNSs_800FCC14(Fighter_GObj* gobj, long* out1, long* out2)
     return -1;
 }
 
-/// #ftKb_SpecialNSs_800FCC6C
+bool ftKb_SpecialNSs_800FCC6C(Fighter_GObj* gobj)
+{
+    Fighter* fp;
+
+    if (!gobj) {
+        goto end_true;
+    }
+
+    fp = GET_FIGHTER(gobj);
+
+    switch (fp->motion_id) {
+    case 0x197:
+    case 0x198:
+    case 0x199:
+    case 0x19A:
+    case 0x19B:
+    case 0x19C:
+        if (fp->x2070.x2071_b6) {
+            return true;
+        }
+        return false;
+    default:
+        return true;
+    }
+
+end_true:
+    return true;
+}
 
 bool ftKb_SpecialNSs_800FCCBC(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
index 621fb166ba..1dc928e3a0 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
@@ -5,6 +5,7 @@
 
 #include "ef/eflib.h"
 #include "ef/efsync.h"
+#include "ft/chara/ftCommon/ftCo_Bury.h"
 #include "ft/chara/ftCommon/ftCo_Lift.h"
 #include "ft/chara/ftCommon/ftCo_Throw.h"
 #include "ft/fighter.h"
@@ -636,7 +637,18 @@ void fn_8010B148(Fighter_GObj* gobj)
     ftCommon_8007D5D4(GET_FIGHTER(gobj));
 }
 
-/// #fn_8010B16C
+#pragma dont_inline on
+void fn_8010B16C(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    fp->grab_timer -=
+        fp->dmg.x1838_percentTemp * ftKb_SpecialNYs_80109380(gobj);
+    if ((s32) fp->dmg.x18CC == 3 && ftCo_800C0C88(fp->dmg.x18D0) != 0) {
+        fp->grab_timer = 0.0f;
+    }
+    fp->x1828 = 4;
+}
+#pragma dont_inline reset
 
 void fn_8010B1D4(Fighter_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index 097bc4a198..9ec78ede9d 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -618,7 +618,32 @@ s32 ftKb_SpecialNMt_80106F44(Fighter_GObj* gobj, s32* out_charge, s32* out_max)
     return -1;
 }
 
-/// #ftKb_SpecialNMt_80106F9C
+bool ftKb_SpecialNMt_80106F9C(Fighter_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Fighter* fp = gobj->user_data;
+        s32 msid = fp->motion_id;
+
+        switch (msid) {
+        case ftKb_MS_MtSpecialNStart:
+        case ftKb_MS_MtSpecialNLoop:
+        case ftKb_MS_MtSpecialNLoopFull:
+        case ftKb_MS_MtSpecialNCancel:
+        case ftKb_MS_MtSpecialNEnd:
+        case ftKb_MS_MtSpecialAirNStart:
+        case ftKb_MS_MtSpecialAirNLoop:
+        case ftKb_MS_MtSpecialAirNLoopFull:
+        case ftKb_MS_MtSpecialAirNCancel:
+        case ftKb_MS_MtSpecialAirNEnd:
+            if (fp->x2070.x2071_b6) {
+                return true;
+            }
+            return false;
+        }
+        return true;
+    }
+    return true;
+}
 
 /// #ftKb_SpecialNMt_80106FEC
 
