diff --git a/src/melee/gr/granime.c b/src/melee/gr/granime.c
index e54370a02b..738de56305 100644
--- a/src/melee/gr/granime.c
+++ b/src/melee/gr/granime.c
@@ -10,6 +10,7 @@
 #include "lb/lbarchive.h"
 #include "lb/lbheap.h"
 
+#include <baselib/aobj.h>
 #include <baselib/dobj.h>
 #include <baselib/mobj.h>
 #include <baselib/robj.h>
@@ -757,7 +758,14 @@ HSD_AObj* grAnime_801C8318(HSD_GObj* gobj, int arg1, u32 arg2)
     return sp14;
 }
 
-/// #grAnime_801C83D0
+bool grAnime_801C83D0(HSD_GObj* gobj, bool arg1, enum_t arg2)
+{
+    HSD_AObj* aobj = grAnime_801C8318(gobj, arg1, arg2);
+    if (HSD_AObjGetFlags(aobj) & 0x40000000) {
+        return true;
+    }
+    return false;
+}
 
 /// #grAnime_801C84A4
 
diff --git a/src/melee/gr/grcorneria.c b/src/melee/gr/grcorneria.c
index 3a67f9250b..9ec173feca 100644
--- a/src/melee/gr/grcorneria.c
+++ b/src/melee/gr/grcorneria.c
@@ -300,7 +300,59 @@ void grCorneria_801DDD4C(Vec3* vec)
     vec->z = 0.0f;
 }
 
-/// #grCorneria_801DDDA8
+typedef struct {
+    int data[3];
+} grCn_Entry;
+
+extern grCn_Entry grCn_803E2204[][5];
+
+typedef struct grCn_Data {
+    /* 0x000 */ u8 pad[0x4CC];
+    /* 0x4CC */ grCn_Entry entries[][5];
+} grCn_Data;
+
+extern grCn_Data grCn_803E1D38;
+
+void grCorneria_801DDDA8(HSD_GObj* gobj, Vec3* vec)
+{
+    grCn_Data* data = &grCn_803E1D38;
+    Ground* gp;
+    Ground* corn_gp;
+    HSD_JObj* jobj;
+    u8 unused[12];
+    Vec3 pos;
+    int idx;
+    PAD_STACK(4);
+
+    if (gobj != NULL) {
+        gp = gobj->user_data;
+        corn_gp = Ground_801C2BA4(3)->user_data;
+        jobj = Ground_801C3FA4(gobj, 4);
+        lb_8000B1CC(jobj, NULL, &pos);
+        idx = ((int*) &data->pad[0x3C])[gp->gv.corneria.xC8];
+        vec->x =
+            corn_gp->gv.corneria.offset_x +
+            (-pos.z +
+             *(f32*) &data->pad[0x330 +
+                                idx * (int) sizeof(grCn_Entry)]);
+        vec->y =
+            pos.y +
+            *(f32*) &data->pad[0x334 +
+                               idx * (int) sizeof(grCn_Entry)];
+        vec->z =
+            pos.x +
+            *(f32*) &data->pad[0x338 +
+                               idx * (int) sizeof(grCn_Entry)];
+    } else {
+        vec->z = 0.0f;
+        vec->y = 0.0f;
+        vec->x = 0.0f;
+    }
+
+    if (vec->y <= 100.0f) {
+        vec->y = 100.0f;
+    }
+}
 
 /// #grCorneria_801DDE88
 
@@ -709,19 +761,6 @@ void smashTaunt_801E2550(Ground_GObj* gobj, struct grSmashTaunt_GroundVars* gv)
     gobj->render_cb = smashTaunt_801E24AC;
 }
 
-typedef struct {
-    int data[3];
-} grCn_Entry;
-
-extern grCn_Entry grCn_803E2204[][5];
-
-typedef struct grCn_Data {
-    /* 0x000 */ u8 pad[0x4CC];
-    /* 0x4CC */ grCn_Entry entries[][5];
-} grCn_Data;
-
-extern grCn_Data grCn_803E1D38;
-
 s32 grCorneria_801E2598(u32 arg0, u32 arg1)
 {
     int val;
diff --git a/src/melee/gr/grcorneria.h b/src/melee/gr/grcorneria.h
index fd500b83aa..066ef860da 100644
--- a/src/melee/gr/grcorneria.h
+++ b/src/melee/gr/grcorneria.h
@@ -29,7 +29,7 @@
 /* 1DDCE8 */ bool grCorneria_801DDCE8(Ground_GObj*);
 /* 1DDCF0 */ void grCorneria_801DDCF0(Vec3*);
 /* 1DDD4C */ void grCorneria_801DDD4C(Vec3*);
-/* 1DDDA8 */ UNK_RET grCorneria_801DDDA8(UNK_PARAMS);
+/* 1DDDA8 */ void grCorneria_801DDDA8(HSD_GObj*, Vec3*);
 /* 1DDE88 */ UNK_RET grCorneria_801DDE88(UNK_PARAMS);
 /* 1DE024 */ void grCorneria_801DE024(Ground_GObj*);
 /* 1DE4C0 */ void grCorneria_Arwing_801DE4C0(Ground_GObj*);
diff --git a/src/melee/gr/grhomerun.c b/src/melee/gr/grhomerun.c
index e21290f9ee..af3be585a4 100644
--- a/src/melee/gr/grhomerun.c
+++ b/src/melee/gr/grhomerun.c
@@ -12,6 +12,7 @@
 #include "lb/lb_00B0.h"
 #include "lb/lblanguage.h"
 
+f32 grHr_804D6AD8;
 int grHr_804D6ADC;
 f32 grHr_804D6AE0;
 f32 grHr_804D6AE4;
@@ -480,7 +481,35 @@ void grHomeRun_8021ED74(void)
     it_8026BE84(&bobomb_rain);
 }
 
-/// #grHomeRun_8021EDD4
+void grHomeRun_8021EDD4(int arg0)
+{
+    f32 scale;
+    f32 max_dist;
+    f32 ratio;
+    f32 num_ticks;
+    f32 dist;
+
+    if (lbLang_IsSavedLanguageJP()) {
+        scale = 0.1F;
+        max_dist = 50.0F;
+    } else {
+        scale = 0.3280969F;
+        max_dist = 100.0F;
+    }
+
+    ratio = max_dist / scale;
+    num_ticks = (int) (0.5F + ratio / (160.0F * Ground_801C0498()));
+
+    if (num_ticks < 1.0F) {
+        num_ticks = 1.0F;
+    }
+
+    dist = 160.0F * Ground_801C0498();
+    grHr_804D6AD8 = scale;
+    grHr_804D6AE0 = max_dist;
+    grHr_804D6ADC = num_ticks;
+    grHr_804D6AE4 = ratio / (num_ticks * dist);
+}
 
 DynamicsDesc* grHomeRun_8021EEB4(enum_t arg)
 {
diff --git a/src/melee/gr/grkongo.c b/src/melee/gr/grkongo.c
index 185ecbf0f4..b06b26d03a 100644
--- a/src/melee/gr/grkongo.c
+++ b/src/melee/gr/grkongo.c
@@ -1411,7 +1411,37 @@ void fn_801D542C(HSD_GObj* arg0)
 
 /// #grKongo_801D7134
 
-/// #fn_801D7700
+static void fn_801D7700(Ground* gp, s32 arg1, CollData* cd, s32 arg3,
+                        mpLib_GroundEnum arg4, float arg8)
+{
+    Vec3* cur = &cd->cur_pos;
+    s32 type = cd->x34_flags.b1234;
+
+    if (type == 1 || type == 3) {
+        if (arg4 == mpLib_GroundEnum_Unk1) {
+            f32 x1, y1, x2, y2;
+
+            mpVtxGetPos(0x1D, &x1, &y1);
+            mpVtxGetPos(0x1A, &x2, &y2);
+
+            {
+                f32 segment_size = (x2 - x1) / 15.0F;
+                f32 ratio = (cur->x - x1) / segment_size;
+                s32 idx = (s32) ratio;
+                s32 idx2 = (s32) ratio;
+
+                if (idx < 0) {
+                    idx2 = 0;
+                } else if ((u32) idx2 >= 15) {
+                    idx2 = 14;
+                }
+
+                grKg_803E188C[idx2].unk10 +=
+                    0.017453292F * grKg_804D6980->unkA4;
+            }
+        }
+    }
+}
 
 /// #grKongo_801D77E0
 
@@ -1427,7 +1457,39 @@ void fn_801D7E60(Ground* gp, s32 arg1, CollData* arg2, s32 arg3,
 
 /// #grKongo_801D7E78
 
-/// #grKongo_801D7F78
+bool grKongo_801D7F78(HSD_GObj* gobj)
+{
+    Vec3 pos1;
+    Vec3 pos2;
+    HSD_GObj* cur;
+    PAD_STACK(16);
+
+    if (grKongo_801D7E78(gobj, &pos1) == NULL) {
+        return false;
+    }
+
+    for (cur = HSD_GObj_Entities->x14; cur != NULL; cur = cur->next) {
+        if (cur == gobj) {
+            continue;
+        }
+        if (grKongo_801D7E78(cur, &pos2) == NULL) {
+            continue;
+        }
+        {
+            f32 dz = pos2.z - pos1.z;
+            if (dz < -100.0F) {
+                if (dz > -150.0F) {
+                    f32 dx = pos2.x - pos1.x;
+                    if (dx * dx < 6400.0F) {
+                        return true;
+                    }
+                }
+            }
+        }
+    }
+
+    return false;
+}
 
 void grKongo_801D8058(Ground_GObj* arg)
 {
diff --git a/src/melee/gr/grkongo.h b/src/melee/gr/grkongo.h
index 77664a612f..bcc21fc777 100644
--- a/src/melee/gr/grkongo.h
+++ b/src/melee/gr/grkongo.h
@@ -52,7 +52,7 @@
 /* 1D7134 */ void grKongo_801D7134(HSD_GObj*, s32);
 /* 1D77E0 */ void grKongo_801D77E0(HSD_GObj*, s32);
 /* 1D7BBC */ void grKongo_801D7BBC(HSD_GObj*);
-/* 1D7E78 */ UNK_RET grKongo_801D7E78(UNK_PARAMS);
+/* 1D7E78 */ Vec3* grKongo_801D7E78(HSD_GObj*, Vec3*);
 /* 1D7F78 */ bool grKongo_801D7F78(HSD_GObj*);
 /* 1D8058 */ void grKongo_801D8058(Ground_GObj*);
 /* 1D8078 */ HSD_GObj* grKongo_801D8078(HSD_GObj*);
diff --git a/src/melee/gr/grkongo.static.h b/src/melee/gr/grkongo.static.h
index 275da089f2..86f092a127 100644
--- a/src/melee/gr/grkongo.static.h
+++ b/src/melee/gr/grkongo.static.h
@@ -61,6 +61,9 @@ static struct {
     f32 unk90;
     f32 unk94;
     f32 unk98;
+    f32 unk9C;
+    f32 unkA0;
+    f32 unkA4;
 }* grKg_804D6980;
 
 extern StageInfo stage_info;
diff --git a/src/melee/gr/groldpupupu.c b/src/melee/gr/groldpupupu.c
index 6be94088de..d2d147b081 100644
--- a/src/melee/gr/groldpupupu.c
+++ b/src/melee/gr/groldpupupu.c
@@ -22,6 +22,10 @@ static struct {
 
 static void* grOp_804D6A9C;
 
+static int grOp_804D48A0[1][2] = { { 4, 0 } };
+static int grOp_804D48A8[1][2] = { { 5, 1 } };
+static int grOp_804D48B0[1][2] = { { 2, 3 } };
+
 static int grOp_804D48E0[1][2] = { { 0, 0 } };
 
 StageCallbacks grOp_803E6688[9] = {
@@ -111,7 +115,31 @@ bool grOldPupupu_80210A1C(Ground_GObj* gobj)
     return false;
 }
 
-/// #grOldPupupu_80210A24
+void grOldPupupu_80210A24(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+
+    if ((int) grOp_804D6A9C == 0 && gp->gv.unk.xD0 == 1) {
+        gp->gv.unk.xD0 = 0;
+        switch (gp->gv.unk.xC8) {
+        case 0:
+            grAnime_801C8138(
+                gobj, gp->map_id,
+                grOp_804D48A0[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+            break;
+        case 1:
+            grAnime_801C8138(
+                gobj, gp->map_id,
+                grOp_804D48B0[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+            break;
+        case 2:
+            grAnime_801C8138(
+                gobj, gp->map_id,
+                grOp_804D48A8[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+            break;
+        }
+    }
+}
 
 void grOldPupupu_80210B00(Ground_GObj* arg) {}
 
diff --git a/src/melee/gr/grrcruise.c b/src/melee/gr/grrcruise.c
index fac04be647..878b6a2743 100644
--- a/src/melee/gr/grrcruise.c
+++ b/src/melee/gr/grrcruise.c
@@ -44,6 +44,7 @@ StageCallbacks grRc_803E4E34[7] = {
 };
 
 extern s16 grRc_803E4FF0[];
+extern s16 grRc_804D4790[4];
 
 static struct {
     int x0;
@@ -291,7 +292,25 @@ bool grRCruise_801FFAD4(Ground_GObj* arg)
 
 void grRCruise_80200070(Ground_GObj* arg) {}
 
-/// #grRCruise_80200074
+void grRCruise_80200074(Ground_GObj* gobj)
+{
+    int i;
+    Ground* gp = gobj->user_data;
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    Ground_801C2ED0(jobj, gp->map_id);
+    grAnime_801C8138(gobj, gp->map_id, 0);
+    for (i = 0; i < 3; i++) {
+        grAnime_801C8098(gobj, grRc_804D4790[i], 7, 1, 0.0F, 0.0F);
+        gp->gv.rcruise.x3C[i].x02 =
+            Ground_801C32D4(gp->map_id, grRc_804D4790[i]);
+        mpJointSetCb1(gp->gv.rcruise.x3C[i].x02, gobj,
+                      (mpLib_Callback) fn_80200460);
+        gp->gv.rcruise.x3C[i].x0C = Ground_801C3FA4(gobj, grRc_804D4790[i]);
+        gp->gv.rcruise.x3C[i].x08 = 0;
+        gp->gv.rcruise.x3C[i].x04 = 0;
+        gp->gv.rcruise.x3C[i].x00 = 0;
+    }
+}
 
 bool grRCruise_8020014C(Ground_GObj* arg)
 {
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index fe46d02252..ab41344348 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -749,6 +749,15 @@ struct grRCruise_Entry {
     /* 0x14 */ HSD_JObj* x14;
 };
 
+struct grRCruise_SubEntry {
+    /* 0x00 */ u8 x00;
+    /* 0x01 */ u8 pad_01;
+    /* 0x02 */ s16 x02;
+    /* 0x04 */ s32 x04;
+    /* 0x08 */ s32 x08;
+    /* 0x0C */ HSD_JObj* x0C;
+};
+
 struct grRCruise_GroundVars {
     struct {
         u8 b0 : 1;
@@ -773,7 +782,7 @@ struct grRCruise_GroundVars {
     /* +30 gp+F4 */ s32 x30;
     /* +34 gp+F8 */ s32 x34;
     /* +38 gp+FC */ s32 x38;
-    /* +3C gp+100 */ char pad_3C[0x30];
+    /* +3C gp+100 */ struct grRCruise_SubEntry x3C[3];
     /* +6C gp+130 */ struct grRCruise_Entry* entries;
 };
 
