diff --git a/config/GALE01/splits.txt b/config/GALE01/splits.txt
index f77a9a53d7..7a85b87e6a 100644
--- a/config/GALE01/splits.txt
+++ b/config/GALE01/splits.txt
@@ -1346,10 +1346,18 @@ melee/ft/chara/ftPopo/ftPp_SpecialN.c:
 	.sdata2     start:0x804D9838 end:0x804D9848
 
 melee/ft/chara/ftPopo/ftPp_SpecialS.c:
-	.text       start:0x8011F68C end:0x80122EDC
+	.text       start:0x8011F68C end:0x80120E68
 	.data       start:0x803CD7F8 end:0x803CD838
 	.sdata      start:0x804D3E10 end:0x804D3E20
-	.sdata2     start:0x804D9848 end:0x804D9898
+	.sdata2     start:0x804D9848 end:0x804D9860
+
+melee/ft/chara/ftPopo/ftPp_SpecialHi.c:
+	.text       start:0x80120E68 end:0x80122898
+	.sdata2     start:0x804D9860 end:0x804D9890
+
+melee/ft/chara/ftPopo/ftPp_SpecialLw.c:
+	.text       start:0x80122898 end:0x80122EDC
+	.sdata2     start:0x804D9890 end:0x804D9898
 
 melee/ft/chara/ftNana/ftNn_Init.c:
 	.text       start:0x80122EDC end:0x801243AC
diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 740355ca9b..db0eefd64b 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -21477,7 +21477,7 @@ ftPp_Init_AnimDatFilename = .data:0x803CD738; // type:object size:0xB scope:glob
 @231 = .data:0x803CD798; // type:object size:0x1B scope:local data:string
 ftPp_Init_DemoMotionFilenames = .data:0x803CD7B4; // type:object size:0x10 scope:global
 ftPp_Init_CostumeStrings = .data:0x803CD7C4; // type:object size:0x30 scope:global
-ftNn_Init_803CD7F8 = .data:0x803CD7F8; // type:object size:0x25 scope:global data:string
+@265 = .data:0x803CD7F8; // type:object size:0x25 scope:local data:string
 ftNn_Init_803CD820 = .data:0x803CD820; // type:object size:0x18 scope:global data:float
 ftNn_Init_MotionStateTable = .data:0x803CD838; // type:object size:0x340 scope:global
 ftNn_Init_DatFilename = .data:0x803CDB78; // type:object size:0x9 scope:global data:string
@@ -26283,8 +26283,8 @@ ftKb_Init_804D3DF0 = .sdata:0x804D3DF0; // type:object size:0x5 scope:global dat
 ftKb_Init_804D3DF8 = .sdata:0x804D3DF8; // type:object size:0x6 scope:global data:string
 ftKb_Init_804D3E00 = .sdata:0x804D3E00; // type:object size:0x7 scope:global data:string
 ftKb_Init_804D3E08 = .sdata:0x804D3E08; // type:object size:0x5 scope:global data:string
-ftPp_Unk0_804D3E10 = .sdata:0x804D3E10; // type:object size:0x7 scope:global data:string
-ftPp_Unk0_804D3E18 = .sdata:0x804D3E18; // type:object size:0x5 scope:global data:string
+@263 = .sdata:0x804D3E10; // type:object size:0x7 scope:local data:string
+@264 = .sdata:0x804D3E18; // type:object size:0x5 scope:local data:string
 @351 = .sdata:0x804D3E20; // type:object size:0x7 scope:local data:string
 @352 = .sdata:0x804D3E28; // type:object size:0x5 scope:local data:string
 @306 = .sdata:0x804D3E30; // type:object size:0x7 scope:local data:string
@@ -30342,20 +30342,20 @@ ftKb_Init_804D95A8 = .sdata2:0x804D95A8; // type:object size:0x4 scope:global al
 @230 = .sdata2:0x804D9838; // type:object size:0x4 scope:local align:4 data:float
 @231 = .sdata2:0x804D983C; // type:object size:0x4 scope:local align:4 data:float
 @237 = .sdata2:0x804D9840; // type:object size:0x4 scope:local align:4 data:float
-ftPp_Init_804D9848 = .sdata2:0x804D9848; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D984C = .sdata2:0x804D984C; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D9850 = .sdata2:0x804D9850; // type:object size:0x8 scope:global align:8 data:double
-ftPp_Init_804D9858 = .sdata2:0x804D9858; // type:object size:0x8 scope:global align:8 data:double
-ftPp_Init_804D9860 = .sdata2:0x804D9860; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D9868 = .sdata2:0x804D9868; // type:object size:0x8 scope:global align:8 data:double
-ftPp_Init_804D9870 = .sdata2:0x804D9870; // type:object size:0x8 scope:global align:8 data:double
-ftPp_Init_804D9878 = .sdata2:0x804D9878; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D987C = .sdata2:0x804D987C; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D9880 = .sdata2:0x804D9880; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D9884 = .sdata2:0x804D9884; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D9888 = .sdata2:0x804D9888; // type:object size:0x8 scope:global align:8 data:double
-ftPp_Init_804D9890 = .sdata2:0x804D9890; // type:object size:0x4 scope:global align:4 data:float
-ftPp_Init_804D9894 = .sdata2:0x804D9894; // type:object size:0x4 scope:global align:4 data:float
+@201 = .sdata2:0x804D9848; // type:object size:0x4 scope:local align:4 data:float
+@295 = .sdata2:0x804D984C; // type:object size:0x4 scope:local align:4 data:float
+@363 = .sdata2:0x804D9850; // type:object size:0x8 scope:local align:8 data:double
+@498 = .sdata2:0x804D9858; // type:object size:0x8 scope:local align:8 data:double
+@194 = .sdata2:0x804D9860; // type:object size:0x4 scope:local align:4 data:float
+@195 = .sdata2:0x804D9868; // type:object size:0x8 scope:local align:8 data:double
+@196 = .sdata2:0x804D9870; // type:object size:0x8 scope:local align:8 data:double
+@197 = .sdata2:0x804D9878; // type:object size:0x4 scope:local align:4 data:float
+@198 = .sdata2:0x804D987C; // type:object size:0x4 scope:local align:4 data:float
+@199 = .sdata2:0x804D9880; // type:object size:0x4 scope:local align:4 data:float
+@200 = .sdata2:0x804D9884; // type:object size:0x4 scope:local align:4 data:float
+@369 = .sdata2:0x804D9888; // type:object size:0x8 scope:local align:8 data:double
+@190 = .sdata2:0x804D9890; // type:object size:0x4 scope:local align:4 data:float
+@196 = .sdata2:0x804D9894; // type:object size:0x4 scope:local align:4 data:float
 ftNn_Init_804D9898 = .sdata2:0x804D9898; // type:object size:0x4 scope:global align:4 data:float
 ftNn_Init_804D98A0 = .sdata2:0x804D98A0; // type:object size:0x8 scope:global align:8 data:double
 ftNn_Init_804D98A8 = .sdata2:0x804D98A8; // type:object size:0x4 scope:global align:4 data:float
diff --git a/configure.py b/configure.py
index 74e50beb13..f0b03a0ab6 100755
--- a/configure.py
+++ b/configure.py
@@ -769,7 +769,9 @@ def MatchingFor(*versions):
             # Popo
             Object(Matching, "melee/ft/chara/ftPopo/ftPp_Init.c"),
             Object(Matching, "melee/ft/chara/ftPopo/ftPp_SpecialN.c"),
-            Object(NonMatching, "melee/ft/chara/ftPopo/ftPp_SpecialS.c"),
+            Object(Matching, "melee/ft/chara/ftPopo/ftPp_SpecialS.c"),
+            Object(Matching, "melee/ft/chara/ftPopo/ftPp_SpecialHi.c"),
+            Object(Matching, "melee/ft/chara/ftPopo/ftPp_SpecialLw.c"),
             Object(NonMatching, "melee/ft/chara/ftPp_SpecialS.c"),
             # Nana
             Object(NonMatching, "melee/ft/chara/ftNana/ftNn_Init.c"),
diff --git a/src/melee/ft/chara/ftCommon/ftCo_0A01.c b/src/melee/ft/chara/ftCommon/ftCo_0A01.c
index 10086cf328..b862881db5 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_0A01.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_0A01.c
@@ -5,7 +5,8 @@
 #include <placeholder.h>
 #include <platform.h>
 
-#include "ft/chara/ftPopo/ftPp_1211.h"
+#include "ft/chara/ftPopo/ftPp_SpecialHi.h"
+#include "ft/chara/ftPopo/ftPp_SpecialLw.h"
 #include "ft/fighter.h"
 #include "ft/ft_0877.h"
 #include "ft/ftcpuattack.h"
diff --git a/src/melee/ft/chara/ftNana/ftNn_Init.c b/src/melee/ft/chara/ftNana/ftNn_Init.c
index 18a92c00d3..4a506b88b1 100644
--- a/src/melee/ft/chara/ftNana/ftNn_Init.c
+++ b/src/melee/ft/chara/ftNana/ftNn_Init.c
@@ -21,8 +21,9 @@
 #include "ftCommon/forward.h"
 #include "ftPopo/forward.h"
 
-#include "ftPopo/ftPp_1211.h"
 #include "ftPopo/ftPp_Init.h"
+#include "ftPopo/ftPp_SpecialHi.h"
+#include "ftPopo/ftPp_SpecialLw.h"
 #include "ftPopo/ftPp_SpecialN.h"
 #include "ftPopo/ftPp_SpecialS.h"
 #include "ftPopo/types.h"
diff --git a/src/melee/ft/chara/ftPopo/ftPp_Init.c b/src/melee/ft/chara/ftPopo/ftPp_Init.c
index dd601aadb0..b79ee2eef8 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_Init.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_Init.c
@@ -1,6 +1,7 @@
 #include "ftPp_Init.h"
 
-#include "ftPp_1211.h"
+#include "ftPp_SpecialHi.h"
+#include "ftPp_SpecialLw.h"
 #include "ftPp_SpecialN.h"
 #include "ftPp_SpecialS.h"
 #include "types.h"
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialHi.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialHi.c
new file mode 100644
index 0000000000..0e5f6685dc
--- /dev/null
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialHi.c
@@ -0,0 +1,932 @@
+
+
+#include "ft/chara/ftPopo/ftPp_SpecialHi.h"
+
+#include "ft/chara/ftCommon/ftCo_FallSpecial.h"
+#include "ft/chara/ftCommon/ftCo_Landing.h"
+#include "ft/chara/ftPopo/ftPp_SpecialS.h"
+#include "ft/fighter.h"
+#include "ft/ft_081B.h"
+#include "ft/ft_0892.h"
+#include "ft/ftcliffcommon.h"
+#include "ft/ftcommon.h"
+#include "ft/ftparts.h"
+#include "ftNana/ftNn_Init.h"
+#include "ftPopo/ftPp_Init.h"
+#include "it/items/itclimbersstring.h"
+#include "it/types.h"
+#include "lb/lb_00B0.h"
+#include "lb/lbvector.h"
+#include "pl/player.h"
+
+#include <math.h>
+#include <trigf.h>
+
+static float sdata2_ordering(void)
+{
+    float data_0 = 0.0f;
+    double data_1 = 0.5;
+    double data_2 = 3.0;
+    float data_3 = 3.0f;
+    float data_4 = 5.0f;
+    float data_5 = 1.0f;
+    float data_6 = -1.0f;
+    double data_7 = 0.5;
+    return data_7 + data_6 + data_5 + data_4 + data_3 + data_2 + data_1 +
+           data_0;
+}
+
+static inline float my_sqrtf(float x, volatile float* y)
+{
+    if (x > 0) {
+        double guess = __frsqrte(x);
+        guess = 0.5f * guess * (3.0f - guess * guess * x);
+        guess = 0.5f * guess * (3.0f - guess * guess * x);
+        guess = 0.5f * guess * (3.0f - guess * guess * x);
+        *y = (x * guess);
+        return *y;
+    }
+    return x;
+}
+
+void ftPp_SpecialS_80120E68(Fighter_GObj* gobj)
+{
+    u8 _pad[4];
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    Fighter_GObj* gobj2 = Player_GetEntityAtIndex(fp->player_id, 1);
+    volatile float y;
+    PAD_STACK(8);
+
+    if (gobj2 != NULL) {
+        Fighter* fp2 = GET_FIGHTER(gobj2);
+        f32 dx, dy;
+        f32 dist;
+        fp->self_vel.x = fp2->cur_pos.x - fp->cur_pos.x;
+        fp->self_vel.y = fp2->cur_pos.y - fp->cur_pos.y;
+        fp->self_vel.z = 0.0F;
+        fp->self_vel.x = -(3.0F * fp2->facing_dir - fp->self_vel.x);
+        fp->self_vel.y += 5.0F;
+        lbVector_Normalize(&fp->self_vel);
+        dx = SQ(fp->cur_pos.x - fp2->cur_pos.x);
+        dy = SQ(fp->cur_pos.y - fp2->cur_pos.y);
+        dist = my_sqrtf(dx + dy, &y) / da->x98;
+        fp->self_vel.x *= da->x94 + dist;
+        fp->self_vel.y *= da->x94 + dist;
+        if (fp->self_vel.x > 0.0F) {
+            fp->facing_dir = +1.0F;
+        } else {
+            fp->facing_dir = -1.0F;
+        }
+    }
+}
+
+bool ftPp_SpecialS_80120FE0(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    s32 cmd = fp->mv.pp.speciallw.x0;
+    PAD_STACK(16);
+
+    if (cmd > 8 && cmd <= 0x53) {
+        Item_GObj* item_gobj;
+        if ((item_gobj = fp->fv.pp.x2238) != NULL) {
+            Item_GObj* gobj = item_gobj;
+            Item* ip = item_gobj->user_data;
+            itClimbersStringAttributes* sa =
+                ip->xC4_article_data->x4_specialAttributes;
+            s32 ev0 = sa->x18;
+            s32 ev1 = sa->x1C;
+            s32 ev2 = sa->x20;
+            if (cmd == ev0) {
+                it_802C3950(gobj);
+            } else if (cmd == ev1) {
+                it_802C3810(gobj);
+            } else if (cmd == ev2) {
+                it_802C3864(gobj);
+            }
+            if (fp->mv.pp.speciallw.x0 == 0x53) {
+                it_802C2750(fp->fv.pp.x2238);
+            }
+        } else {
+            goto end;
+        }
+    } else if (fp->mv.pp.speciallw.x0 == 8) {
+        ftPp_SpecialS_801210C8(gobj);
+        if (fp->fv.pp.x2238 == NULL) {
+            ft_8008A2BC(gobj);
+            return true;
+        }
+    }
+    return false;
+end: {
+    // original code returns without a value
+}
+}
+
+void ftPp_SpecialS_801210C8(Fighter_GObj* arg0)
+{
+    Vec3 sp10;
+    Fighter* fp = GET_FIGHTER(arg0);
+    float dir;
+    lb_8000B1CC(fp->parts[FtPart_L4thNb].joint, NULL, &sp10);
+    dir = fp->facing_dir;
+    fp->fv.pp.x2238 = it_802C27D4(arg0, &sp10, fp->motion_id, dir);
+    fp->x1984_heldItemSpec = fp->fv.pp.x2238;
+    if (fp->fv.pp.x2238 != NULL) {
+        fp->death3_cb = ftPp_Init_8011F060;
+        fp->take_dmg_cb = ftPp_Init_8011F060;
+    }
+}
+
+void ftPp_SpecialS_8012114C(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->fv.pp.x2238 = NULL;
+    fp->death3_cb = NULL;
+    fp->take_dmg_cb = NULL;
+}
+
+void ftPp_SpecialS_80121164(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->fv.pp.x2238 != NULL) {
+        it_802C2750(fp->fv.pp.x2238);
+        ftPp_SpecialS_8012114C(gobj);
+    }
+}
+
+void ftPp_SpecialHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+
+    fp->gr_vel /= da->x84;
+
+    ftPp_SpecialHi_801218AC(gobj);
+
+    fp = GET_FIGHTER(gobj);
+    fp->cmd_vars[2] = 0;
+    fp->cmd_vars[1] = 0;
+    fp->cmd_vars[0] = 0;
+    fp = GET_FIGHTER(gobj);
+    fp->mv.pp.unk_80123954.x0 = 1;
+    fp->fv.pp.x223C = 0;
+    fp->fv.pp.x2240.z = 0.0f;
+    fp->fv.pp.x2240.y = 0.0f;
+    fp->fv.pp.x2240.x = 0.0f;
+}
+void ftPp_SpecialAirHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(16);
+
+    fp->self_vel.x /= da->x84;
+    fp->self_vel.y /= da->x88;
+
+    ftPp_SpecialHi_801218F8(gobj);
+
+    fp->x1968_jumpsUsed = fp->co_attrs.max_jumps;
+
+    fp = GET_FIGHTER(gobj);
+    fp->cmd_vars[2] = 0;
+    fp->cmd_vars[1] = 0;
+    fp->cmd_vars[0] = 0;
+    fp = GET_FIGHTER(gobj);
+    fp->mv.pp.unk_80123954.x0 = 1;
+    fp->fv.pp.x223C = 0;
+    fp->fv.pp.x2240.z = 0.0f;
+    fp->fv.pp.x2240.y = 0.0f;
+    fp->fv.pp.x2240.x = 0.0f;
+}
+
+static inline bool checkNanaInRange(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    Fighter_GObj* nana_gobj = Player_GetEntityAtIndex(fp->player_id, 1);
+    if (nana_gobj != NULL) {
+        Vec3* nana_pos = &GET_FIGHTER(nana_gobj)->cur_pos;
+        f32 dx = SQ(fp->cur_pos.x - nana_pos->x);
+        f32 dy = SQ(fp->cur_pos.y - nana_pos->y);
+        if (sqrtf__Ff(dx + dy) < da->x7C &&
+            ftNn_Init_8012300C(nana_gobj) == true)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+static inline void incrementMvAndCheck(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->mv.pp.unk_80123954.x0++;
+    ftPp_SpecialS_80120FE0(gobj);
+}
+
+void ftPp_SpecialHiStart_0_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    PAD_STACK(24);
+    if (fp->cmd_vars[2] != 0) {
+        fp->cmd_vars[2] = 0;
+        if (!checkNanaInRange(gobj)) {
+            ftPp_SpecialHi_80122098(gobj);
+            return;
+        }
+        fp->x2222_b2 = true;
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftPp_SpecialHi_80121DA0(gobj);
+    } else {
+        incrementMvAndCheck(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiStart_0_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    PAD_STACK(24);
+    if (fp->cmd_vars[2] != 0) {
+        fp->cmd_vars[2] = 0;
+        if (!checkNanaInRange(gobj)) {
+            ftPp_SpecialHi_801220D4(gobj);
+            return;
+        }
+        fp->x2222_b2 = true;
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftPp_SpecialHi_80121DD8(gobj);
+        return;
+    }
+    incrementMvAndCheck(gobj);
+}
+
+void ftPp_SpecialHiStart_0_IASA(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(16);
+    if (fp->cmd_vars[0] != 0) {
+        fp->cmd_vars[0] = 0;
+        if (ABS(fp->input.lstick.x) > da->x80) {
+            ftCommon_UpdateFacing(fp);
+            ftPartSetRotY(fp, 0, (float) (M_PI_2 * fp->facing_dir));
+        }
+    }
+}
+
+void ftPp_SpecialAirHiStart_0_IASA(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(16);
+    if (fp->cmd_vars[0] != 0) {
+        fp->cmd_vars[0] = 0;
+        if (ABS(fp->input.lstick.x) > da->x80) {
+            ftCommon_UpdateFacing(fp);
+            ftPartSetRotY(fp, 0, (float) (M_PI_2 * fp->facing_dir));
+        }
+    }
+}
+
+void ftPp_SpecialHiStart_0_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp;
+    Vec3 sp;
+    PAD_STACK(8);
+
+    ft_80084F3C(gobj);
+    fp = GET_FIGHTER(gobj);
+    sp.x = sp.y = sp.z = 0.0f;
+
+    {
+        Fighter_GObj* nn_gobj =
+            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+        if (nn_gobj != NULL) {
+            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+            {
+                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
+            }
+        }
+    }
+
+    fp->fv.pp.x2240 = sp;
+}
+
+void ftPp_SpecialAirHiStart_0_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    Vec3 sp;
+    PAD_STACK(16);
+
+    ftCommon_Fall(fp, da->x8C, da->x90);
+    ftCommon_8007CEF4(fp);
+
+    fp = GET_FIGHTER(gobj);
+    sp.x = sp.y = sp.z = 0.0f;
+
+    {
+        Fighter_GObj* nn_gobj =
+            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+        if (nn_gobj != NULL) {
+            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+            {
+                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
+            }
+        }
+    }
+
+    fp->fv.pp.x2240 = sp;
+}
+
+void ftPp_SpecialHiStart_0_Coll(Fighter_GObj* gobj)
+{
+    if (ft_800827A0(gobj) == 0) {
+        ftPp_SpecialHi_801217EC(gobj);
+    }
+}
+void ftPp_SpecialAirHiStart_0_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    s32 facingDir;
+    if (fp->facing_dir < 0.0f) {
+        facingDir = -1;
+    } else {
+        facingDir = 1;
+    }
+    if (ft_CheckGroundAndLedge(gobj, facingDir)) {
+        ftPp_SpecialHi_8012184C(gobj);
+        return;
+    }
+    if (ftCliffCommon_80081298(gobj)) {
+        return;
+    }
+}
+
+void ftPp_SpecialHi_801217EC(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D60C(fp);
+    Fighter_ChangeMotionState(gobj, 0x160, 0x0C4C508A, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_8012184C(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+
+    ftCommon_8007D7FC(fp);
+    Fighter_ChangeMotionState(gobj, 0x15B, 0x0C4C508AU, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_801218AC(Fighter_GObj* gobj)
+{
+    PAD_STACK(8);
+    Fighter_ChangeMotionState(gobj, 0x15B, Ft_MF_None, 0.0f, 1.0f, 0.0f, NULL);
+    ftAnim_8006EBA4(gobj);
+}
+
+void ftPp_SpecialHi_801218F8(Fighter_GObj* gobj)
+{
+    PAD_STACK(8);
+    Fighter_ChangeMotionState(gobj, 0x160, Ft_MF_None, 0.0f, 1.0f, 0.0f, NULL);
+    ftAnim_8006EBA4(gobj);
+}
+
+void ftPp_SpecialHiThrow_0_Anim(Fighter_GObj* gobj)
+{
+    PAD_STACK(40);
+
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ft_8008A2BC(gobj);
+    }
+
+    {
+        Fighter* fp = gobj->user_data;
+        if (fp->cmd_vars[1] != 0) {
+            fp->cmd_vars[1] = 0;
+            {
+                int found;
+                Fighter_GObj* nn_gobj =
+                    Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+                if (nn_gobj != NULL && ftNn_Init_8012309C(nn_gobj) == 1) {
+                    found = 1;
+                } else {
+                    found = 0;
+                }
+                if (found == 1) {
+                    ftPp_SpecialHi_8012280C(gobj);
+                    return;
+                }
+            }
+        }
+    }
+
+    {
+        Fighter* fp = GET_FIGHTER(gobj);
+        ++fp->mv.pp.unk_80123954.x0;
+        ftPp_SpecialS_80120FE0(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiThrow_0_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(48);
+
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
+    }
+
+    {
+        Fighter* fp = gobj->user_data;
+        if (fp->cmd_vars[1] != 0) {
+            fp->cmd_vars[1] = 0;
+            {
+                int found;
+                Fighter_GObj* nn_gobj =
+                    Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+                if (nn_gobj != NULL && ftNn_Init_8012309C(nn_gobj) == 1) {
+                    found = 1;
+                } else {
+                    found = 0;
+                }
+                if (found == 1) {
+                    ftPp_SpecialHi_8012280C(gobj);
+                    return;
+                }
+            }
+        }
+    }
+
+    {
+        Fighter* fp = GET_FIGHTER(gobj);
+        ++fp->mv.pp.unk_80123954.x0;
+        ftPp_SpecialS_80120FE0(gobj);
+    }
+}
+
+void ftPp_SpecialHiThrow_0_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialAirHiThrow_0_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialHiThrow_0_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp;
+    Vec3 sp;
+    PAD_STACK(8);
+
+    ft_80084F3C(gobj);
+    fp = GET_FIGHTER(gobj);
+    sp.x = sp.y = sp.z = 0.0f;
+
+    {
+        Fighter_GObj* nn_gobj =
+            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+        if (nn_gobj != NULL) {
+            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+            {
+                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
+            }
+        }
+    }
+
+    fp->fv.pp.x2240 = sp;
+}
+
+static inline void doFallPhys(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    ftCommon_Fall(fp, da->x8C, da->x90);
+    ftCommon_8007CEF4(fp);
+}
+
+void ftPp_SpecialAirHiThrow_0_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp;
+    Vec3 sp;
+    PAD_STACK(12);
+
+    doFallPhys(gobj);
+    fp = GET_FIGHTER(gobj);
+    sp.x = sp.y = sp.z = 0.0f;
+
+    {
+        Fighter_GObj* nn_gobj =
+            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+        if (nn_gobj != NULL) {
+            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+            {
+                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
+            }
+        }
+    }
+
+    fp->fv.pp.x2240 = sp;
+}
+
+void ftPp_SpecialHiThrow_0_Coll(Fighter_GObj* gobj)
+{
+    if (ft_800827A0(gobj) == 0) {
+        ftPp_SpecialHi_80121CE0(gobj);
+    }
+}
+void ftPp_SpecialAirHiThrow_0_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    s32 facingDir;
+
+    if (fp->facing_dir < 0.0f) {
+        facingDir = -1;
+    } else {
+        facingDir = 1;
+    }
+
+    if (ft_CheckGroundAndLedge(gobj, facingDir)) {
+        ftPp_SpecialHi_80121D40(gobj);
+        return;
+    }
+
+    if (ftCliffCommon_80081298(gobj)) {
+        return;
+    }
+}
+
+void ftPp_SpecialHi_80121CE0(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D60C(fp);
+    Fighter_ChangeMotionState(gobj, 0x161, 0x0C4C508A, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_80121D40(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D7FC(fp);
+    Fighter_ChangeMotionState(gobj, 0x15C, 0x0C4C508A, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_80121DA0(Fighter_GObj* gobj)
+{
+    Fighter_ChangeMotionState(gobj, 0x15C, 0, 0.0f, 1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_80121DD8(Fighter_GObj* gobj)
+{
+    Fighter_ChangeMotionState(gobj, 0x161, 0, 0.0f, 1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHiStart_1_Anim(Fighter_GObj* gobj)
+{
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftPp_SpecialHi_80122348(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiStart_1_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(8);
+    if (fp->cmd_vars[2] != 0) {
+        fp->cmd_vars[2] = 0;
+        fp->self_vel.y = da->xA4;
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftPp_SpecialHi_80122380(gobj);
+    }
+}
+
+void ftPp_SpecialHiStart_1_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialAirHiStart_1_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialHiStart_1_Phys(Fighter_GObj* gobj)
+{
+    ft_80084F3C(gobj);
+}
+
+void ftPp_SpecialAirHiStart_1_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(8);
+    ftCommon_Fall(fp, da->xA8, da->xAC);
+    if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+}
+
+void ftPp_SpecialHiStart_1_Coll(Fighter_GObj* gobj)
+{
+    if (!ft_800827A0(gobj)) {
+        ftPp_SpecialHi_80121FD8(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiStart_1_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    s32 facingDir;
+
+    if (fp->facing_dir < 0.0f) {
+        facingDir = -1;
+    } else {
+        facingDir = 1;
+    }
+
+    if (ft_CheckGroundAndLedge(gobj, facingDir)) {
+        ftPp_SpecialHi_80122038(gobj);
+        return;
+    }
+
+    if (ftCliffCommon_80081298(gobj)) {
+        return;
+    }
+}
+
+void ftPp_SpecialHi_80121FD8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D60C(fp);
+    Fighter_ChangeMotionState(gobj, 0x163, 0x0C4C508A, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_80122038(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D7FC(fp);
+    Fighter_ChangeMotionState(gobj, 0x15E, 0x0C4C508A, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_80122098(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter_ChangeMotionState(gobj, 0x15E, 0, fp->cur_anim_frame, 1.0f, 0.0f,
+                              NULL);
+}
+
+void ftPp_SpecialHi_801220D4(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter_ChangeMotionState(gobj, 0x163, 0, fp->cur_anim_frame, 1.0f, 0.0f,
+                              NULL);
+}
+
+void ftPp_SpecialHiThrow_1_Anim(Fighter_GObj* gobj)
+{
+    if (ftAnim_IsFramesRemaining(gobj) == 0) {
+        ft_8008A2BC(gobj);
+    }
+}
+void ftPp_SpecialAirHiThrow_1_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
+    }
+}
+
+void ftPp_SpecialHiThrow_1_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialAirHiThrow_1_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialHiThrow_1_Phys(Fighter_GObj* gobj)
+{
+    ft_80084F3C(gobj);
+}
+
+void ftPp_SpecialAirHiThrow_1_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    PAD_STACK(8);
+    ftCommon_Fall(fp, da->xA8, da->xAC);
+    if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+}
+
+void ftPp_SpecialHiThrow_1_Coll(Fighter_GObj* gobj)
+{
+    if (!ft_800827A0(gobj)) {
+        ftPp_SpecialHi_801222E8(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiThrow_1_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    s32 cliffCatchDir;
+    PAD_STACK(4);
+    if (fp->facing_dir < 0.0f) {
+        cliffCatchDir = -1;
+    } else {
+        cliffCatchDir = 1;
+    }
+    if (ft_CheckGroundAndLedge(gobj, cliffCatchDir)) {
+        ftCo_LandingFallSpecial_Enter(gobj, false, da->x78);
+        return;
+    }
+    if (ftCliffCommon_80081298(gobj)) {
+        return;
+    }
+}
+
+void ftPp_SpecialHi_801222E8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D60C(fp);
+    Fighter_ChangeMotionState(gobj, 0x164, 0x0C4C508A, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_80122348(Fighter_GObj* arg0)
+{
+    Fighter_ChangeMotionState(arg0, 0x15F, 0U, 0.0f, 1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_80122380(Fighter_GObj* gobj)
+{
+    Fighter_ChangeMotionState(gobj, 0x164, 0, 0.0f, 1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHiThrow2_Anim(Fighter_GObj* gobj)
+{
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ft_8008A2BC(gobj);
+    } else {
+        Fighter* fp = GET_FIGHTER(gobj);
+        ++fp->mv.pp.unk_80123954.x0;
+        ftPp_SpecialS_80120FE0(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiThrow2_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp;
+    ftIceClimberAttributes* da;
+
+    fp = GET_FIGHTER(gobj);
+    da = fp->dat_attrs;
+
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftCo_80096900(gobj, 0, 1, 0, da->x74, da->x78);
+    } else {
+        Fighter* fp2 = GET_FIGHTER(gobj);
+        ++fp2->mv.pp.unk_80123954.x0;
+        ftPp_SpecialS_80120FE0(gobj);
+    }
+}
+
+void ftPp_SpecialHiThrow2_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialAirHiThrow2_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialHiThrow2_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp;
+    Vec3 sp;
+    PAD_STACK(8);
+
+    ft_80084F3C(gobj);
+    fp = GET_FIGHTER(gobj);
+    sp.x = sp.y = sp.z = 0.0f;
+
+    {
+        Fighter_GObj* nn_gobj =
+            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+        if (nn_gobj != NULL) {
+            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+            {
+                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
+            }
+        }
+    }
+
+    fp->fv.pp.x2240 = sp;
+}
+
+static inline void doFallPhys2(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    ftCo_DatAttrs* co = &fp->co_attrs;
+    ftCommon_Fall(fp, da->x9C, da->xA0);
+    if (ABS(fp->input.lstick.x) > da->x80) {
+        ftCommon_8007D344(fp, 0.0f, co->air_drift_stick_mul * da->xB0,
+                          co->air_drift_max * da->xB4);
+    } else if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+}
+
+static inline void ftPp_SpecialAirHiThrow2_Phys_inline(Fighter_GObj* gobj,
+                                                       Vec3* sp)
+{
+    Fighter_GObj* nn_gobj =
+        Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+    if (nn_gobj != NULL) {
+        Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+        if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+            nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+        {
+            lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, sp);
+        }
+    }
+}
+
+void ftPp_SpecialAirHiThrow2_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Vec3 sp;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    ftCo_DatAttrs* co = &fp->co_attrs;
+    PAD_STACK(12);
+
+    ftCommon_Fall(fp, da->x9C, da->xA0);
+    if (ABS(fp->input.lstick.x) > da->x80) {
+        ftCommon_8007D344(fp, 0.0f, co->air_drift_stick_mul * da->xB0,
+                          co->air_drift_max * da->xB4);
+    } else if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+    {
+        fp = GET_FIGHTER(gobj);
+        sp.x = sp.y = sp.z = 0.0f;
+        ftPp_SpecialAirHiThrow2_Phys_inline(gobj, &sp);
+        fp->fv.pp.x2240 = sp;
+    }
+}
+
+void ftPp_SpecialHiThrow2_Coll(Fighter_GObj* gobj)
+{
+    if (!ft_800827A0(gobj)) {
+        ftPp_SpecialHi_801227AC(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiThrow2_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    CollData* cd = &fp->coll_data;
+    PAD_STACK(8);
+    if (ft_CheckGroundAndLedge(gobj, (fp->facing_dir < 0) ? -1 : 1)) {
+        ftCo_LandingFallSpecial_Enter(gobj, false, da->x78);
+    } else if (!ftCliffCommon_80081298(gobj)) {
+        if ((cd->env_flags & 0x3F) && fp->self_vel.x > 0.0f) {
+            fp->self_vel.x = 0.0f;
+        } else if ((cd->env_flags & 0xFC0) && fp->self_vel.x < 0.0f) {
+            fp->self_vel.x = 0.0f;
+        } else if (cd->env_flags & 0x6000) {
+            fp->self_vel.y = 0.0f;
+            ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
+        }
+    }
+}
+
+void ftPp_SpecialHi_801227AC(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D60C(fp);
+    Fighter_ChangeMotionState(gobj, 0x162, 0x0C4C508A, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialHi_8012280C(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    ftCo_DatAttrs* co = &fp->co_attrs;
+    PAD_STACK(8);
+    if (fp->ground_or_air == GA_Ground) {
+        ftCommon_8007D60C(fp);
+    } else {
+        fp->x1968_jumpsUsed = co->max_jumps;
+    }
+    ftPp_SpecialS_80120E68(gobj);
+    Fighter_ChangeMotionState(gobj, 0x162, 0, 0.0f, 1.0f, 0.0f, NULL);
+    fp->x21F8 = ftCommon_8007F76C;
+}
diff --git a/src/melee/ft/chara/ftPopo/ftPp_1211.h b/src/melee/ft/chara/ftPopo/ftPp_SpecialHi.h
similarity index 73%
rename from src/melee/ft/chara/ftPopo/ftPp_1211.h
rename to src/melee/ft/chara/ftPopo/ftPp_SpecialHi.h
index 56e587cfac..ec10736a82 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_1211.h
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialHi.h
@@ -14,6 +14,10 @@
 /* 121680 */ void ftPp_SpecialAirHiStart_0_Phys(Fighter_GObj* gobj);
 /* 121740 */ void ftPp_SpecialHiStart_0_Coll(Fighter_GObj* gobj);
 /* 12177C */ void ftPp_SpecialAirHiStart_0_Coll(Fighter_GObj* gobj);
+/* 1217EC */ void ftPp_SpecialHi_801217EC(Fighter_GObj* gobj);
+/* 12184C */ void ftPp_SpecialHi_8012184C(Fighter_GObj* gobj);
+/* 1218AC */ void ftPp_SpecialHi_801218AC(Fighter_GObj* gobj);
+/* 1218F8 */ void ftPp_SpecialHi_801218F8(Fighter_GObj* gobj);
 /* 121944 */ void ftPp_SpecialHiThrow_0_Anim(Fighter_GObj* gobj);
 /* 1219F4 */ void ftPp_SpecialAirHiThrow_0_Anim(Fighter_GObj* gobj);
 /* 121AC8 */ void ftPp_SpecialHiThrow_0_IASA(Fighter_GObj* gobj);
@@ -22,7 +26,10 @@
 /* 121B74 */ void ftPp_SpecialAirHiThrow_0_Phys(Fighter_GObj* gobj);
 /* 121C34 */ void ftPp_SpecialHiThrow_0_Coll(Fighter_GObj* gobj);
 /* 121C70 */ void ftPp_SpecialAirHiThrow_0_Coll(Fighter_GObj* gobj);
+/* 121CE0 */ void ftPp_SpecialHi_80121CE0(Fighter_GObj* gobj);
+/* 121D40 */ void ftPp_SpecialHi_80121D40(Fighter_GObj* gobj);
 /* 121DA0 */ void ftPp_SpecialHi_80121DA0(Fighter_GObj* gobj);
+/* 121DD8 */ void ftPp_SpecialHi_80121DD8(Fighter_GObj* gobj);
 /* 121E10 */ void ftPp_SpecialHiStart_1_Anim(Fighter_GObj* gobj);
 /* 121E4C */ void ftPp_SpecialAirHiStart_1_Anim(Fighter_GObj* gobj);
 /* 121EB0 */ void ftPp_SpecialHiStart_1_IASA(Fighter_GObj* gobj);
@@ -31,13 +38,21 @@
 /* 121ED8 */ void ftPp_SpecialAirHiStart_1_Phys(Fighter_GObj* gobj);
 /* 121F2C */ void ftPp_SpecialHiStart_1_Coll(Fighter_GObj* gobj);
 /* 121F68 */ void ftPp_SpecialAirHiStart_1_Coll(Fighter_GObj* gobj);
+/* 121FD8 */ void ftPp_SpecialHi_80121FD8(Fighter_GObj* gobj);
+/* 122038 */ void ftPp_SpecialHi_80122038(Fighter_GObj* gobj);
+/* 122098 */ void ftPp_SpecialHi_80122098(Fighter_GObj* gobj);
+/* 1220D4 */ void ftPp_SpecialHi_801220D4(Fighter_GObj* gobj);
 /* 122110 */ void ftPp_SpecialHiThrow_1_Anim(Fighter_GObj* gobj);
 /* 12214C */ void ftPp_SpecialAirHiThrow_1_Anim(Fighter_GObj* gobj);
 /* 1221AC */ void ftPp_SpecialHiThrow_1_IASA(Fighter_GObj* gobj);
 /* 1221B0 */ void ftPp_SpecialAirHiThrow_1_IASA(Fighter_GObj* gobj);
 /* 1221B4 */ void ftPp_SpecialHiThrow_1_Phys(Fighter_GObj* gobj);
 /* 1221D4 */ void ftPp_SpecialAirHiThrow_1_Phys(Fighter_GObj* gobj);
+/* 122228 */ void ftPp_SpecialHiThrow_1_Coll(Fighter_GObj* gobj);
 /* 122264 */ void ftPp_SpecialAirHiThrow_1_Coll(Fighter_GObj* gobj);
+/* 1222E8 */ void ftPp_SpecialHi_801222E8(Fighter_GObj* gobj);
+/* 122348 */ void ftPp_SpecialHi_80122348(Fighter_GObj* gobj);
+/* 122380 */ void ftPp_SpecialHi_80122380(Fighter_GObj* gobj);
 /* 1223B8 */ void ftPp_SpecialHiThrow2_Anim(Fighter_GObj* gobj);
 /* 122410 */ void ftPp_SpecialAirHiThrow2_Anim(Fighter_GObj* gobj);
 /* 12248C */ void ftPp_SpecialHiThrow2_IASA(Fighter_GObj* gobj);
@@ -46,16 +61,8 @@
 /* 122538 */ void ftPp_SpecialAirHiThrow2_Phys(Fighter_GObj* gobj);
 /* 122664 */ void ftPp_SpecialHiThrow2_Coll(Fighter_GObj* gobj);
 /* 1226A0 */ void ftPp_SpecialAirHiThrow2_Coll(Fighter_GObj* gobj);
+/* 1227AC */ void ftPp_SpecialHi_801227AC(Fighter_GObj* gobj);
+/* 12280C */ void ftPp_SpecialHi_8012280C(Fighter_GObj* gobj);
 /* 122898 */ void ftPp_SpecialHi_80122898(Fighter_GObj* gobj);
-/* 122904 */ void ftPp_SpecialLw_Enter(Fighter_GObj* gobj);
-/* 122988 */ void ftPp_SpecialAirLw_Enter(Fighter_GObj* gobj);
-/* 122A0C */ void ftPp_SpecialLw_Anim(Fighter_GObj* gobj);
-/* 122A8C */ void ftPp_SpecialAirLw_Anim(Fighter_GObj* gobj);
-/* 122B0C */ void ftPp_SpecialLw_IASA(Fighter_GObj* gobj);
-/* 122B10 */ void ftPp_SpecialAirLw_IASA(Fighter_GObj* gobj);
-/* 122B14 */ void ftPp_SpecialLw_Phys(Fighter_GObj* gobj);
-/* 122B34 */ void ftPp_SpecialAirLw_Phys(Fighter_GObj* gobj);
-/* 122C18 */ void ftPp_SpecialLw_Coll(Fighter_GObj* gobj);
-/* 122D04 */ void ftPp_SpecialAirLw_Coll(Fighter_GObj* gobj);
 
 #endif
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.c
new file mode 100644
index 0000000000..8d30ad6f87
--- /dev/null
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.c
@@ -0,0 +1,212 @@
+#include "ft/chara/ftPopo/ftPp_SpecialS.h"
+
+#include "ef/eflib.h"
+#include "ef/efsync.h"
+#include "ft/chara/ftCommon/ftCo_Fall.h"
+#include "ft/fighter.h"
+#include "ft/ft_081B.h"
+#include "ft/ft_0881.h"
+#include "ft/ft_0892.h"
+#include "ft/ftcommon.h"
+#include "ft/ftparts.h"
+#include "ftPopo/ftPp_SpecialHi.h"
+#include "ftPopo/ftPp_SpecialLw.h"
+#include "ftPopo/ftPp_Init.h"
+#include "it/items/itclimbersblizzard.h"
+#include "lb/lb_00B0.h"
+#include "pl/player.h"
+
+#include <trigf.h>
+
+void ftPp_SpecialHi_80122898(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    PAD_STACK(16);
+    if (fp->fv.pp.x2230_b0) {
+        efLib_DestroyAll(gobj);
+        fp = gobj->user_data;
+        fp->fv.pp.x2230_b0 = false;
+        fp->death2_cb = NULL;
+        fp->take_dmg_cb = NULL;
+        ftPartSetRotX(gobj->user_data, 0, 0.0f);
+    }
+}
+
+void ftPp_SpecialLw_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    PAD_STACK(8);
+    fp->throw_flags = 0;
+    fp->cmd_vars[0] = 0;
+    fp->cmd_vars[3] = 0;
+    fp->mv.pp.speciallw.x0 = 0;
+    fp->mv.pp.speciallw.x4_b0 = false;
+    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialLw, 0, 0.0f, 1.0f, 0.0f,
+                              NULL);
+    ftAnim_8006EBA4(gobj);
+    fp->accessory4_cb = fn_80122D2C;
+}
+
+void ftPp_SpecialAirLw_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    PAD_STACK(8);
+    fp->throw_flags = 0;
+    fp->cmd_vars[0] = 0;
+    fp->cmd_vars[3] = 0;
+    fp->mv.pp.speciallw.x0 = 0;
+    fp->mv.pp.speciallw.x4_b0 = false;
+    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialAirLw, 0, 0.0f, 1.0f, 0.0f,
+                              NULL);
+    ftAnim_8006EBA4(gobj);
+    fp->accessory4_cb = fn_80122D2C;
+}
+
+void ftPp_SpecialLw_Anim(Fighter_GObj* gobj)
+{
+    PAD_STACK(16);
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        Fighter* fp = gobj->user_data;
+        if (fp->fv.pp.x2230_b0) {
+            efLib_DestroyAll(gobj);
+            fp = gobj->user_data;
+            fp->fv.pp.x2230_b0 = false;
+            fp->death2_cb = NULL;
+            fp->take_dmg_cb = NULL;
+            ftPartSetRotX(gobj->user_data, 0, 0.0f);
+        }
+        ft_8008A2BC(gobj);
+    }
+}
+
+void ftPp_SpecialAirLw_Anim(Fighter_GObj* gobj)
+{
+    PAD_STACK(16);
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        Fighter* fp = gobj->user_data;
+        if (fp->fv.pp.x2230_b0) {
+            efLib_DestroyAll(gobj);
+            fp = gobj->user_data;
+            fp->fv.pp.x2230_b0 = false;
+            fp->death2_cb = NULL;
+            fp->take_dmg_cb = NULL;
+            ftPartSetRotX(gobj->user_data, 0, 0.0f);
+        }
+        ftCo_Fall_Enter(gobj);
+    }
+}
+
+void ftPp_SpecialLw_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialAirLw_IASA(Fighter_GObj* gobj) {}
+
+void ftPp_SpecialLw_Phys(Fighter_GObj* gobj)
+{
+    ft_80084F3C(gobj);
+}
+
+void ftPp_SpecialAirLw_Phys(Fighter_GObj* gobj)
+{
+    ft_80084EEC(gobj);
+}
+
+static inline void ftPp_set_cbs(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    fp->death2_cb = ftPp_Init_8011F060;
+    fp->take_dmg_cb = ftPp_Init_8011F060;
+}
+
+static inline void ftPp_SpecialLw_Coll_inline(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->cmd_vars[3]) {
+        ftPartSetRotX(fp, 0,
+                      fp->facing_dir * atan2f(fp->coll_data.floor.normal.x,
+                                              fp->coll_data.floor.normal.y));
+    } else {
+        ftPartSetRotX(fp, 0, 0.0F);
+    }
+}
+
+void fn_80122B54(Fighter_GObj* gobj)
+{
+    Fighter_GObj* new_var;
+    Fighter* fp = gobj->user_data;
+    Fighter* fp2;
+    new_var = gobj;
+    PAD_STACK(8);
+    ftCommon_8007D7FC(fp);
+    fp2 = fp;
+    Fighter_ChangeMotionState(new_var, ftPp_MS_SpecialLw, 0x0C4C5282,
+                              fp2->cur_anim_frame, 1.0F, 0.0F, 0L);
+    ftPp_set_cbs(new_var);
+    ftPp_SpecialLw_Coll_inline(gobj);
+    fp->accessory4_cb = fn_80122D2C;
+}
+
+static inline void ftPp_SpecialLw_Coll_Land(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D5D4(fp);
+    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialAirLw, 0x0C4C5282,
+                              fp->cur_anim_frame, 1.0f, 0.0f, NULL);
+    ftPp_set_cbs(gobj);
+    fp->accessory4_cb = fn_80122D2C;
+    ftCommon_ClampAirDrift(fp);
+}
+
+void ftPp_SpecialLw_Coll(Fighter_GObj* gobj)
+{
+    PAD_STACK(8);
+    if (ft_80082708(gobj) == GA_Ground) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        ftPartSetRotX(fp, 0, 0.0f);
+        ftPp_SpecialLw_Coll_Land(gobj);
+    } else {
+        ftPp_SpecialLw_Coll_inline(gobj);
+    }
+}
+
+void ftPp_SpecialAirLw_Coll(Fighter_GObj* gobj)
+{
+    ft_80082C74(gobj, fn_80122B54);
+}
+
+void fn_80122D2C(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Vec3 pos;
+    PAD_STACK(8);
+    if (fp->mv.pp.speciallw.x4_b0) {
+        if (fp->mv.pp.speciallw.x0 == 0) {
+            ftIceClimberAttributes* da = fp->dat_attrs;
+            lb_8000B1CC(fp->parts[FtPart_L3rdNa].joint, NULL, &pos);
+            pos.x += da->xBC * fp->facing_dir;
+            pos.y += da->xC0;
+            itClimbersBlizzard_Spawn(gobj, &pos, fp->facing_dir);
+            fp->mv.pp.speciallw.x0 = da->xB8;
+        }
+        fp->mv.pp.speciallw.x0--;
+    }
+    switch (fp->cmd_vars[0]) {
+    case 1:
+        efSync_Spawn(0x4EC, gobj, fp->parts[FtPart_L4thNb].joint);
+        fp->mv.pp.speciallw.x4_b0 = true;
+        fp->fv.pp.x2230_b0 = true;
+        ftPp_set_cbs(gobj);
+        fp->cmd_vars[0] = 0;
+        if (fp->kind == FTKIND_POPO) {
+            ft_800881D8(fp, 0x1FC42, 0x7F, 0x40);
+            ft_80088510(fp, 0x1FBEB, 0x7F, 0x40);
+        } else {
+            ft_800881D8(fp, 0x1FC0F, 0x7F, 0x40);
+            ft_80088510(fp, 0x1FBEE, 0x7F, 0x40);
+        }
+        break;
+    case 2:
+        fp->mv.pp.speciallw.x4_b0 = false;
+        fp->cmd_vars[0] = 0;
+        break;
+    }
+}
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.h b/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.h
new file mode 100644
index 0000000000..ccc10de11a
--- /dev/null
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialLw.h
@@ -0,0 +1,20 @@
+#ifndef GALE01_122904
+#define GALE01_122904
+
+#include "ft/forward.h"
+#include <baselib/forward.h>
+
+/* 122904 */ void ftPp_SpecialLw_Enter(Fighter_GObj* gobj);
+/* 122988 */ void ftPp_SpecialAirLw_Enter(Fighter_GObj* gobj);
+/* 122A0C */ void ftPp_SpecialLw_Anim(Fighter_GObj* gobj);
+/* 122A8C */ void ftPp_SpecialAirLw_Anim(Fighter_GObj* gobj);
+/* 122B0C */ void ftPp_SpecialLw_IASA(Fighter_GObj* gobj);
+/* 122B10 */ void ftPp_SpecialAirLw_IASA(Fighter_GObj* gobj);
+/* 122B14 */ void ftPp_SpecialLw_Phys(Fighter_GObj* gobj);
+/* 122B34 */ void ftPp_SpecialAirLw_Phys(Fighter_GObj* gobj);
+/* 122B54 */ void fn_80122B54(Fighter_GObj* gobj);
+/* 122C18 */ void ftPp_SpecialLw_Coll(Fighter_GObj* gobj);
+/* 122D04 */ void ftPp_SpecialAirLw_Coll(Fighter_GObj* gobj);
+/* 122D2C */ void fn_80122D2C(Fighter_GObj* gobj);
+
+#endif
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
index 2c0a5acb28..3315a02639 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
@@ -1,80 +1,21 @@
-#include "placeholder.h"
-
-#include <platform.h>
+#include "ft/chara/ftPopo/ftPp_SpecialS.h"
 
 #include "ef/eflib.h"
-#include "ft/chara/ftCommon/ftCo_Landing.h"
+#include "ft/chara/ftCommon/ftCo_Fall.h"
+#include "ft/chara/ftCommon/ftCo_FallSpecial.h"
 #include "ft/fighter.h"
-
-#include "ft/forward.h"
-
 #include "ft/ft_081B.h"
-#include "ft/ft_0881.h"
 #include "ft/ft_0892.h"
-#include "ft/ftcliffcommon.h"
 #include "ft/ftcommon.h"
 #include "ft/ftparts.h"
-#include "ft/inlines.h"
-#include "ft/types.h"
 #include "ftNana/ftNn_Init.h"
-
-#include "ftPopo/forward.h"
-
-#include "ftPopo/ftPp_1211.h"
 #include "ftPopo/ftPp_Init.h"
-#include "ftPopo/types.h"
 #include "pl/player.h"
 
-#include <common_structs.h>
 #include <math.h>
-#include <math_ppc.h>
 #include <trigf.h>
-#include <dolphin/pad.h>
-#include <melee/ef/efsync.h>
-#include <melee/ft/chara/ftCommon/ftCo_Fall.h>
-#include <melee/ft/chara/ftCommon/ftCo_FallSpecial.h>
-#include <melee/ft/chara/ftPopo/ftPp_SpecialS.h>
-#include <melee/it/itCharItems.h>
-#include <melee/it/items/itclimbersblizzard.h>
-#include <melee/it/items/itclimbersstring.h>
-#include <melee/it/types.h>
-#include <melee/lb/lb_00B0.h>
-#include <melee/lb/lbvector.h>
-
-/* 121AC8 */ void ftPp_SpecialHiThrow_0_IASA(Fighter_GObj* gobj);
-/* 121ACC */ void ftPp_SpecialAirHiThrow_0_IASA(Fighter_GObj* gobj);
-/* 121EB0 */ void ftPp_SpecialHiStart_1_IASA(Fighter_GObj* gobj);
-/* 121EB4 */ void ftPp_SpecialAirHiStart_1_IASA(Fighter_GObj* gobj);
-/* 1221B0 */ void ftPp_SpecialAirHiThrow_1_IASA(Fighter_GObj* gobj);
-/* 12248C */ void ftPp_SpecialHiThrow2_IASA(Fighter_GObj* gobj);
-/* 122490 */ void ftPp_SpecialAirHiThrow2_IASA(Fighter_GObj* gobj);
-/* 122B0C */ void ftPp_SpecialLw_IASA(Fighter_GObj* gobj);
-/* 122B10 */ void ftPp_SpecialAirLw_IASA(Fighter_GObj* gobj);
-
-/* 1218AC */ void ftPp_SpecialHi_801218AC(Fighter_GObj* gobj);
-/* 1218F8 */ void ftPp_SpecialHi_801218F8(Fighter_GObj* gobj);
-/* 12280C */ void ftPp_SpecialHi_8012280C(Fighter_GObj* gobj);
-
-static inline float my_sqrtf(float x)
-{
-    static const double _half = .5;
-    static const double _three = 3.0;
-
-    u8 _[8] = { 0 };
-
-    volatile float y;
-    if (x > 0) {
-        double guess = __frsqrte((double) x);
-        guess = _half * guess * (_three - guess * guess * x);
-        guess = _half * guess * (_three - guess * guess * x);
-        guess = _half * guess * (_three - guess * guess * x);
-        y = (float) (x * guess);
-        return y;
-    }
-    return x;
-}
 
-static void setRefGObjFlagAndClear(Fighter* fp)
+static inline void setRefGObjFlagAndClear(Fighter* fp)
 {
     Fighter_GObj* gobj = fp->x1A5C;
     Fighter* ref_fp;
@@ -578,18 +519,6 @@ static ftCollisionBox ftNn_Init_803CD820 = {
     12, 0, -6, 6, 6, 6,
 };
 
-static inline void ftPp_SpecialLw_Coll_inline(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    if (fp->cmd_vars[3]) {
-        ftPartSetRotX(fp, 0,
-                      fp->facing_dir * atan2f(fp->coll_data.floor.normal.x,
-                                              fp->coll_data.floor.normal.y));
-    } else {
-        ftPartSetRotX(fp, 0, 0.0F);
-    }
-}
-
 static inline void inline2(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
@@ -764,1076 +693,3 @@ void ftPp_SpecialAirS2_Coll(Fighter_GObj* gobj)
     ftPp_SpecialS_8011F720(gobj);
     inlineA0(gobj);
 }
-
-void ftPp_SpecialS_80120E68(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    Fighter_GObj* gobj2 = Player_GetEntityAtIndex(fp->player_id, 1);
-    PAD_STACK(8);
-
-    if (gobj2 != NULL) {
-        Fighter* fp2 = GET_FIGHTER(gobj2);
-        f32 dx, dy;
-        f32 dist;
-        fp->self_vel.x = fp2->cur_pos.x - fp->cur_pos.x;
-        fp->self_vel.y = fp2->cur_pos.y - fp->cur_pos.y;
-        fp->self_vel.z = 0.0F;
-        fp->self_vel.x = -(3.0F * fp2->facing_dir - fp->self_vel.x);
-        fp->self_vel.y += 5.0F;
-        lbVector_Normalize(&fp->self_vel);
-        dx = SQ(fp->cur_pos.x - fp2->cur_pos.x);
-        dy = SQ(fp->cur_pos.y - fp2->cur_pos.y);
-        dist = my_sqrtf(dx + dy) / da->x98;
-        fp->self_vel.x *= da->x94 + dist;
-        fp->self_vel.y *= da->x94 + dist;
-        if (fp->self_vel.x > 0.0F) {
-            fp->facing_dir = +1.0F;
-        } else {
-            fp->facing_dir = -1.0F;
-        }
-    }
-}
-
-bool ftPp_SpecialS_80120FE0(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    s32 cmd = fp->mv.pp.speciallw.x0;
-    PAD_STACK(16);
-
-    if (cmd > 8 && cmd <= 0x53) {
-        Item_GObj* item_gobj;
-        if ((item_gobj = fp->fv.pp.x2238) != NULL) {
-            Item_GObj* gobj = item_gobj;
-            Item* ip = item_gobj->user_data;
-            itClimbersStringAttributes* sa =
-                ip->xC4_article_data->x4_specialAttributes;
-            s32 ev0 = sa->x18;
-            s32 ev1 = sa->x1C;
-            s32 ev2 = sa->x20;
-            if (cmd == ev0) {
-                it_802C3950(gobj);
-            } else if (cmd == ev1) {
-                it_802C3810(gobj);
-            } else if (cmd == ev2) {
-                it_802C3864(gobj);
-            }
-            if (fp->mv.pp.speciallw.x0 == 0x53) {
-                it_802C2750(fp->fv.pp.x2238);
-            }
-        } else {
-            goto end;
-        }
-    } else if (fp->mv.pp.speciallw.x0 == 8) {
-        ftPp_SpecialS_801210C8(gobj);
-        if (fp->fv.pp.x2238 == NULL) {
-            ft_8008A2BC(gobj);
-            return true;
-        }
-    }
-    return false;
-end: {
-    // original code returns without a value
-}
-}
-
-void ftPp_SpecialS_801210C8(Fighter_GObj* arg0)
-{
-    Vec3 sp10;
-    Fighter* fp = GET_FIGHTER(arg0);
-    float dir;
-    lb_8000B1CC(fp->parts[FtPart_L4thNb].joint, NULL, &sp10);
-    dir = fp->facing_dir;
-    fp->fv.pp.x2238 = it_802C27D4(arg0, &sp10, fp->motion_id, dir);
-    fp->x1984_heldItemSpec = fp->fv.pp.x2238;
-    if (fp->fv.pp.x2238 != NULL) {
-        fp->death3_cb = ftPp_Init_8011F060;
-        fp->take_dmg_cb = ftPp_Init_8011F060;
-    }
-}
-
-void ftPp_SpecialS_8012114C(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    fp->fv.pp.x2238 = NULL;
-    fp->death3_cb = NULL;
-    fp->take_dmg_cb = NULL;
-}
-
-void ftPp_SpecialS_80121164(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    if (fp->fv.pp.x2238 != NULL) {
-        it_802C2750(fp->fv.pp.x2238);
-        ftPp_SpecialS_8012114C(gobj);
-    }
-}
-
-void ftPp_SpecialHi_Enter(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-
-    fp->gr_vel /= da->x84;
-
-    ftPp_SpecialHi_801218AC(gobj);
-
-    fp = GET_FIGHTER(gobj);
-    fp->cmd_vars[2] = 0;
-    fp->cmd_vars[1] = 0;
-    fp->cmd_vars[0] = 0;
-    fp = GET_FIGHTER(gobj);
-    fp->mv.pp.unk_80123954.x0 = 1;
-    fp->fv.pp.x223C = 0;
-    fp->fv.pp.x2240.z = 0.0f;
-    fp->fv.pp.x2240.y = 0.0f;
-    fp->fv.pp.x2240.x = 0.0f;
-}
-void ftPp_SpecialAirHi_Enter(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(16);
-
-    fp->self_vel.x /= da->x84;
-    fp->self_vel.y /= da->x88;
-
-    ftPp_SpecialHi_801218F8(gobj);
-
-    fp->x1968_jumpsUsed = fp->co_attrs.max_jumps;
-
-    fp = GET_FIGHTER(gobj);
-    fp->cmd_vars[2] = 0;
-    fp->cmd_vars[1] = 0;
-    fp->cmd_vars[0] = 0;
-    fp = GET_FIGHTER(gobj);
-    fp->mv.pp.unk_80123954.x0 = 1;
-    fp->fv.pp.x223C = 0;
-    fp->fv.pp.x2240.z = 0.0f;
-    fp->fv.pp.x2240.y = 0.0f;
-    fp->fv.pp.x2240.x = 0.0f;
-}
-
-static inline bool checkNanaInRange(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    Fighter_GObj* nana_gobj = Player_GetEntityAtIndex(fp->player_id, 1);
-    if (nana_gobj != NULL) {
-        Vec3* nana_pos = &((Fighter*) nana_gobj->user_data)->cur_pos;
-        f32 dx = SQ(fp->cur_pos.x - nana_pos->x);
-        f32 dy = SQ(fp->cur_pos.y - nana_pos->y);
-        if (sqrtf__Ff(dx + dy) < da->x7C && ftNn_Init_8012300C(nana_gobj) == 1)
-        {
-            return true;
-        }
-    }
-    return false;
-}
-
-static inline void incrementMvAndCheck(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    fp->mv.pp.unk_80123954.x0++;
-    ftPp_SpecialS_80120FE0(gobj);
-}
-
-void ftPp_SpecialHiStart_0_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    PAD_STACK(32);
-    if (fp->cmd_vars[2] != 0) {
-        fp->cmd_vars[2] = 0;
-        if (!checkNanaInRange(gobj)) {
-            ftPp_SpecialHi_80122098(gobj);
-            return;
-        }
-        fp->x2222_b2 = true;
-    }
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftPp_SpecialHi_80121DA0(gobj);
-    } else {
-        incrementMvAndCheck(gobj);
-    }
-}
-
-void ftPp_SpecialAirHiStart_0_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    PAD_STACK(32);
-    if (fp->cmd_vars[2] != 0) {
-        fp->cmd_vars[2] = 0;
-        if (!checkNanaInRange(gobj)) {
-            ftPp_SpecialHi_801220D4(gobj);
-            return;
-        }
-        fp->x2222_b2 = true;
-    }
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftPp_SpecialHi_80121DD8(gobj);
-        return;
-    }
-    incrementMvAndCheck(gobj);
-}
-
-void ftPp_SpecialHiStart_0_IASA(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(16);
-    if (fp->cmd_vars[0] != 0) {
-        fp->cmd_vars[0] = 0;
-        if (ABS(fp->input.lstick.x) > da->x80) {
-            ftCommon_UpdateFacing(fp);
-            ftPartSetRotY(fp, 0, (float) (M_PI_2 * fp->facing_dir));
-        }
-    }
-}
-
-void ftPp_SpecialAirHiStart_0_IASA(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(16);
-    if (fp->cmd_vars[0] != 0) {
-        fp->cmd_vars[0] = 0;
-        if (ABS(fp->input.lstick.x) > da->x80) {
-            ftCommon_UpdateFacing(fp);
-            ftPartSetRotY(fp, 0, (float) (M_PI_2 * fp->facing_dir));
-        }
-    }
-}
-
-void ftPp_SpecialHiStart_0_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp;
-    Vec3 sp;
-    PAD_STACK(8);
-
-    ft_80084F3C(gobj);
-    fp = GET_FIGHTER(gobj);
-    sp.x = sp.y = sp.z = 0.0f;
-
-    {
-        Fighter_GObj* nn_gobj =
-            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-        if (nn_gobj != NULL) {
-            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
-            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
-                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
-            {
-                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
-            }
-        }
-    }
-
-    fp->fv.pp.x2240 = sp;
-}
-
-void ftPp_SpecialAirHiStart_0_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    Vec3 sp;
-    PAD_STACK(16);
-
-    ftCommon_Fall(fp, da->x8C, da->x90);
-    ftCommon_8007CEF4(fp);
-
-    fp = GET_FIGHTER(gobj);
-    sp.x = sp.y = sp.z = 0.0f;
-
-    {
-        Fighter_GObj* nn_gobj =
-            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-        if (nn_gobj != NULL) {
-            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
-            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
-                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
-            {
-                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
-            }
-        }
-    }
-
-    fp->fv.pp.x2240 = sp;
-}
-
-void ftPp_SpecialHiStart_0_Coll(Fighter_GObj* gobj)
-{
-    if (ft_800827A0(gobj) == 0) {
-        ftPp_SpecialHi_801217EC(gobj);
-    }
-}
-void ftPp_SpecialAirHiStart_0_Coll(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    s32 facingDir;
-    if (fp->facing_dir < 0.0f) {
-        facingDir = -1;
-    } else {
-        facingDir = 1;
-    }
-    if (ft_CheckGroundAndLedge(gobj, facingDir)) {
-        ftPp_SpecialHi_8012184C(gobj);
-        return;
-    }
-    if (ftCliffCommon_80081298(gobj)) {
-        return;
-    }
-}
-
-void ftPp_SpecialHi_801217EC(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D60C(fp);
-    Fighter_ChangeMotionState(gobj, 0x160, 0x0C4C508A, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_8012184C(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-
-    ftCommon_8007D7FC(fp);
-    Fighter_ChangeMotionState(gobj, 0x15B, 0x0C4C508AU, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_801218AC(Fighter_GObj* gobj)
-{
-    PAD_STACK(8);
-    Fighter_ChangeMotionState(gobj, 0x15B, Ft_MF_None, 0.0f, 1.0f, 0.0f, NULL);
-    ftAnim_8006EBA4(gobj);
-}
-
-void ftPp_SpecialHi_801218F8(Fighter_GObj* gobj)
-{
-    PAD_STACK(8);
-    Fighter_ChangeMotionState(gobj, 0x160, Ft_MF_None, 0.0f, 1.0f, 0.0f, NULL);
-    ftAnim_8006EBA4(gobj);
-}
-
-void ftPp_SpecialHiThrow_0_Anim(Fighter_GObj* gobj)
-{
-    PAD_STACK(40);
-
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ft_8008A2BC(gobj);
-    }
-
-    {
-        Fighter* fp = gobj->user_data;
-        if (fp->cmd_vars[1] != 0) {
-            fp->cmd_vars[1] = 0;
-            {
-                int found;
-                Fighter_GObj* nn_gobj =
-                    Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-                if (nn_gobj != NULL && ftNn_Init_8012309C(nn_gobj) == 1) {
-                    found = 1;
-                } else {
-                    found = 0;
-                }
-                if (found == 1) {
-                    ftPp_SpecialHi_8012280C(gobj);
-                    return;
-                }
-            }
-        }
-    }
-
-    {
-        Fighter* fp = GET_FIGHTER(gobj);
-        ++fp->mv.pp.unk_80123954.x0;
-        ftPp_SpecialS_80120FE0(gobj);
-    }
-}
-
-void ftPp_SpecialAirHiThrow_0_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(48);
-
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
-    }
-
-    {
-        Fighter* fp = gobj->user_data;
-        if (fp->cmd_vars[1] != 0) {
-            fp->cmd_vars[1] = 0;
-            {
-                int found;
-                Fighter_GObj* nn_gobj =
-                    Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-                if (nn_gobj != NULL && ftNn_Init_8012309C(nn_gobj) == 1) {
-                    found = 1;
-                } else {
-                    found = 0;
-                }
-                if (found == 1) {
-                    ftPp_SpecialHi_8012280C(gobj);
-                    return;
-                }
-            }
-        }
-    }
-
-    {
-        Fighter* fp = GET_FIGHTER(gobj);
-        ++fp->mv.pp.unk_80123954.x0;
-        ftPp_SpecialS_80120FE0(gobj);
-    }
-}
-
-void ftPp_SpecialHiThrow_0_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialAirHiThrow_0_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialHiThrow_0_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp;
-    Vec3 sp;
-    PAD_STACK(8);
-
-    ft_80084F3C(gobj);
-    fp = GET_FIGHTER(gobj);
-    sp.x = sp.y = sp.z = 0.0f;
-
-    {
-        Fighter_GObj* nn_gobj =
-            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-        if (nn_gobj != NULL) {
-            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
-            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
-                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
-            {
-                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
-            }
-        }
-    }
-
-    fp->fv.pp.x2240 = sp;
-}
-
-static inline void doFallPhys(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    ftCommon_Fall(fp, da->x8C, da->x90);
-    ftCommon_8007CEF4(fp);
-}
-
-void ftPp_SpecialAirHiThrow_0_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp;
-    Vec3 sp;
-    PAD_STACK(12);
-
-    doFallPhys(gobj);
-    fp = GET_FIGHTER(gobj);
-    sp.x = sp.y = sp.z = 0.0f;
-
-    {
-        Fighter_GObj* nn_gobj =
-            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-        if (nn_gobj != NULL) {
-            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
-            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
-                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
-            {
-                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
-            }
-        }
-    }
-
-    fp->fv.pp.x2240 = sp;
-}
-
-void ftPp_SpecialHiThrow_0_Coll(Fighter_GObj* gobj)
-{
-    if (ft_800827A0(gobj) == 0) {
-        ftPp_SpecialHi_80121CE0(gobj);
-    }
-}
-void ftPp_SpecialAirHiThrow_0_Coll(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    s32 facingDir;
-
-    if (fp->facing_dir < 0.0f) {
-        facingDir = -1;
-    } else {
-        facingDir = 1;
-    }
-
-    if (ft_CheckGroundAndLedge(gobj, facingDir)) {
-        ftPp_SpecialHi_80121D40(gobj);
-        return;
-    }
-
-    if (ftCliffCommon_80081298(gobj)) {
-        return;
-    }
-}
-
-void ftPp_SpecialHi_80121CE0(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D60C(fp);
-    Fighter_ChangeMotionState(gobj, 0x161, 0x0C4C508A, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_80121D40(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D7FC(fp);
-    Fighter_ChangeMotionState(gobj, 0x15C, 0x0C4C508A, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_80121DA0(Fighter_GObj* gobj)
-{
-    Fighter_ChangeMotionState(gobj, 0x15C, 0, 0.0f, 1.0f, 0.0f, NULL);
-}
-void ftPp_SpecialHi_80121DD8(Fighter_GObj* gobj)
-{
-    Fighter_ChangeMotionState(gobj, 0x161, 0, 0.0f, 1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHiStart_1_Anim(Fighter_GObj* gobj)
-{
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftPp_SpecialHi_80122348(gobj);
-    }
-}
-
-void ftPp_SpecialAirHiStart_1_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(8);
-    if (fp->cmd_vars[2] != 0) {
-        fp->cmd_vars[2] = 0;
-        fp->self_vel.y = da->xA4;
-    }
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftPp_SpecialHi_80122380(gobj);
-    }
-}
-
-void ftPp_SpecialHiStart_1_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialAirHiStart_1_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialHiStart_1_Phys(Fighter_GObj* gobj)
-{
-    ft_80084F3C(gobj);
-}
-
-void ftPp_SpecialAirHiStart_1_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(8);
-    ftCommon_Fall(fp, da->xA8, da->xAC);
-    if (fp->self_vel.y < 0.0f) {
-        ftCommon_8007CEF4(fp);
-    }
-}
-
-void ftPp_SpecialHiStart_1_Coll(Fighter_GObj* gobj)
-{
-    if (!ft_800827A0(gobj)) {
-        ftPp_SpecialHi_80121FD8(gobj);
-    }
-}
-
-void ftPp_SpecialAirHiStart_1_Coll(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    s32 facingDir;
-
-    if (fp->facing_dir < 0.0f) {
-        facingDir = -1;
-    } else {
-        facingDir = 1;
-    }
-
-    if (ft_CheckGroundAndLedge(gobj, facingDir)) {
-        ftPp_SpecialHi_80122038(gobj);
-        return;
-    }
-
-    if (ftCliffCommon_80081298(gobj)) {
-        return;
-    }
-}
-
-void ftPp_SpecialHi_80121FD8(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D60C(fp);
-    Fighter_ChangeMotionState(gobj, 0x163, 0x0C4C508A, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_80122038(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D7FC(fp);
-    Fighter_ChangeMotionState(gobj, 0x15E, 0x0C4C508A, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_80122098(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    Fighter_ChangeMotionState(gobj, 0x15E, 0, fp->cur_anim_frame, 1.0f, 0.0f,
-                              NULL);
-}
-
-void ftPp_SpecialHi_801220D4(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    Fighter_ChangeMotionState(gobj, 0x163, 0, fp->cur_anim_frame, 1.0f, 0.0f,
-                              NULL);
-}
-
-void ftPp_SpecialHiThrow_1_Anim(Fighter_GObj* gobj)
-{
-    if (ftAnim_IsFramesRemaining(gobj) == 0) {
-        ft_8008A2BC(gobj);
-    }
-}
-void ftPp_SpecialAirHiThrow_1_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
-    }
-}
-
-void ftPp_SpecialHiThrow_1_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialAirHiThrow_1_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialHiThrow_1_Phys(Fighter_GObj* gobj)
-{
-    ft_80084F3C(gobj);
-}
-
-void ftPp_SpecialAirHiThrow_1_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(8);
-    ftCommon_Fall(fp, da->xA8, da->xAC);
-    if (fp->self_vel.y < 0.0f) {
-        ftCommon_8007CEF4(fp);
-    }
-}
-
-void ftPp_SpecialHiThrow_1_Coll(Fighter_GObj* gobj)
-{
-    if (!ft_800827A0(gobj)) {
-        ftPp_SpecialHi_801222E8(gobj);
-    }
-}
-
-void ftPp_SpecialAirHiThrow_1_Coll(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    s32 cliffCatchDir;
-    PAD_STACK(4);
-    if (fp->facing_dir < 0.0f) {
-        cliffCatchDir = -1;
-    } else {
-        cliffCatchDir = 1;
-    }
-    if (ft_CheckGroundAndLedge(gobj, cliffCatchDir)) {
-        ftCo_LandingFallSpecial_Enter(gobj, false, da->x78);
-        return;
-    }
-    if (ftCliffCommon_80081298(gobj)) {
-        return;
-    }
-}
-
-void ftPp_SpecialHi_801222E8(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D60C(fp);
-    Fighter_ChangeMotionState(gobj, 0x164, 0x0C4C508A, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_80122348(Fighter_GObj* arg0)
-{
-    Fighter_ChangeMotionState(arg0, 0x15F, 0U, 0.0f, 1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_80122380(Fighter_GObj* gobj)
-{
-    Fighter_ChangeMotionState(gobj, 0x164, 0, 0.0f, 1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHiThrow2_Anim(Fighter_GObj* gobj)
-{
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ft_8008A2BC(gobj);
-    } else {
-        Fighter* fp = GET_FIGHTER(gobj);
-        ++fp->mv.pp.unk_80123954.x0;
-        ftPp_SpecialS_80120FE0(gobj);
-    }
-}
-
-void ftPp_SpecialAirHiThrow2_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp;
-    ftIceClimberAttributes* da;
-
-    fp = GET_FIGHTER(gobj);
-    da = fp->dat_attrs;
-
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftCo_80096900(gobj, 0, 1, 0, da->x74, da->x78);
-    } else {
-        Fighter* fp2 = GET_FIGHTER(gobj);
-        ++fp2->mv.pp.unk_80123954.x0;
-        ftPp_SpecialS_80120FE0(gobj);
-    }
-}
-
-void ftPp_SpecialHiThrow2_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialAirHiThrow2_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialHiThrow2_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp;
-    Vec3 sp;
-    PAD_STACK(8);
-
-    ft_80084F3C(gobj);
-    fp = GET_FIGHTER(gobj);
-    sp.x = sp.y = sp.z = 0.0f;
-
-    {
-        Fighter_GObj* nn_gobj =
-            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-        if (nn_gobj != NULL) {
-            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
-            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
-                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
-            {
-                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
-            }
-        }
-    }
-
-    fp->fv.pp.x2240 = sp;
-}
-
-static inline void doFallPhys2(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    ftCo_DatAttrs* co = &fp->co_attrs;
-    ftCommon_Fall(fp, da->x9C, da->xA0);
-    if (ABS(fp->input.lstick.x) > da->x80) {
-        ftCommon_8007D344(fp, 0.0f, co->air_drift_stick_mul * da->xB0,
-                          co->air_drift_max * da->xB4);
-    } else if (fp->self_vel.y < 0.0f) {
-        ftCommon_8007CEF4(fp);
-    }
-}
-
-static inline void ftPp_SpecialAirHiThrow2_Phys_inline(Fighter_GObj* gobj,
-                                                       Vec3* sp)
-{
-    Fighter_GObj* nn_gobj =
-        Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
-    if (nn_gobj != NULL) {
-        Fighter* nn_fp = GET_FIGHTER(nn_gobj);
-        if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
-            nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
-        {
-            lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, sp);
-        }
-    }
-}
-
-void ftPp_SpecialAirHiThrow2_Phys(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    Vec3 sp;
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    ftCo_DatAttrs* co = &fp->co_attrs;
-    PAD_STACK(12);
-
-    ftCommon_Fall(fp, da->x9C, da->xA0);
-    if (ABS(fp->input.lstick.x) > da->x80) {
-        ftCommon_8007D344(fp, 0.0f, co->air_drift_stick_mul * da->xB0,
-                          co->air_drift_max * da->xB4);
-    } else if (fp->self_vel.y < 0.0f) {
-        ftCommon_8007CEF4(fp);
-    }
-    {
-        fp = GET_FIGHTER(gobj);
-        sp.x = sp.y = sp.z = 0.0f;
-        ftPp_SpecialAirHiThrow2_Phys_inline(gobj, &sp);
-        fp->fv.pp.x2240 = sp;
-    }
-}
-
-void ftPp_SpecialHiThrow2_Coll(Fighter_GObj* gobj)
-{
-    if (!ft_800827A0(gobj)) {
-        ftPp_SpecialHi_801227AC(gobj);
-    }
-}
-
-void ftPp_SpecialAirHiThrow2_Coll(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftIceClimberAttributes* da = fp->dat_attrs;
-    CollData* cd = &fp->coll_data;
-    PAD_STACK(8);
-    if (ft_CheckGroundAndLedge(gobj, (fp->facing_dir < 0) ? -1 : 1)) {
-        ftCo_LandingFallSpecial_Enter(gobj, false, da->x78);
-    } else if (!ftCliffCommon_80081298(gobj)) {
-        if ((cd->env_flags & 0x3F) && fp->self_vel.x > 0.0f) {
-            fp->self_vel.x = 0.0f;
-        } else if ((cd->env_flags & 0xFC0) && fp->self_vel.x < 0.0f) {
-            fp->self_vel.x = 0.0f;
-        } else if (cd->env_flags & 0x6000) {
-            fp->self_vel.y = 0.0f;
-            ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
-        }
-    }
-}
-
-void ftPp_SpecialHi_801227AC(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D60C(fp);
-    Fighter_ChangeMotionState(gobj, 0x162, 0x0C4C508A, fp->cur_anim_frame,
-                              1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialHi_8012280C(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    ftCo_DatAttrs* co = &fp->co_attrs;
-    PAD_STACK(8);
-    if (fp->ground_or_air == GA_Ground) {
-        ftCommon_8007D60C(fp);
-    } else {
-        fp->x1968_jumpsUsed = co->max_jumps;
-    }
-    ftPp_SpecialS_80120E68(gobj);
-    Fighter_ChangeMotionState(gobj, 0x162, 0, 0.0f, 1.0f, 0.0f, NULL);
-    fp->x21F8 = ftCommon_8007F76C;
-}
-
-void ftPp_SpecialHi_80122898(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    PAD_STACK(16);
-    if (fp->fv.pp.x2230_b0) {
-        efLib_DestroyAll(gobj);
-        fp = gobj->user_data;
-        fp->fv.pp.x2230_b0 = false;
-        fp->death2_cb = NULL;
-        fp->take_dmg_cb = NULL;
-        ftPartSetRotX(gobj->user_data, 0, 0.0f);
-    }
-}
-
-void ftPp_SpecialLw_Enter(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    PAD_STACK(8);
-    fp->throw_flags = 0;
-    fp->cmd_vars[0] = 0;
-    fp->cmd_vars[3] = 0;
-    fp->mv.pp.speciallw.x0 = 0;
-    fp->mv.pp.speciallw.x4_b0 = false;
-    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialLw, 0, 0.0f, 1.0f, 0.0f,
-                              NULL);
-    ftAnim_8006EBA4(gobj);
-    fp->accessory4_cb = fn_80122D2C;
-}
-
-void ftPp_SpecialAirLw_Enter(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    PAD_STACK(8);
-    fp->throw_flags = 0;
-    fp->cmd_vars[0] = 0;
-    fp->cmd_vars[3] = 0;
-    fp->mv.pp.speciallw.x0 = 0;
-    fp->mv.pp.speciallw.x4_b0 = false;
-    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialAirLw, 0, 0.0f, 1.0f, 0.0f,
-                              NULL);
-    ftAnim_8006EBA4(gobj);
-    fp->accessory4_cb = fn_80122D2C;
-}
-
-void ftPp_SpecialLw_Anim(Fighter_GObj* gobj)
-{
-    PAD_STACK(16);
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        Fighter* fp = gobj->user_data;
-        if (fp->fv.pp.x2230_b0) {
-            efLib_DestroyAll(gobj);
-            fp = gobj->user_data;
-            fp->fv.pp.x2230_b0 = false;
-            fp->death2_cb = NULL;
-            fp->take_dmg_cb = NULL;
-            ftPartSetRotX(gobj->user_data, 0, 0.0f);
-        }
-        ft_8008A2BC(gobj);
-    }
-}
-
-void ftPp_SpecialAirLw_Anim(Fighter_GObj* gobj)
-{
-    PAD_STACK(16);
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        Fighter* fp = gobj->user_data;
-        if (fp->fv.pp.x2230_b0) {
-            efLib_DestroyAll(gobj);
-            fp = gobj->user_data;
-            fp->fv.pp.x2230_b0 = false;
-            fp->death2_cb = NULL;
-            fp->take_dmg_cb = NULL;
-            ftPartSetRotX(gobj->user_data, 0, 0.0f);
-        }
-        ftCo_Fall_Enter(gobj);
-    }
-}
-
-void ftPp_SpecialLw_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialAirLw_IASA(Fighter_GObj* gobj) {}
-
-void ftPp_SpecialLw_Phys(Fighter_GObj* gobj)
-{
-    ft_80084F3C(gobj);
-}
-
-void ftPp_SpecialAirLw_Phys(Fighter_GObj* gobj)
-{
-    ft_80084EEC(gobj);
-}
-
-static inline void ftPp_set_cbs(Fighter_GObj* gobj)
-{
-    Fighter* fp = gobj->user_data;
-    fp->death2_cb = ftPp_Init_8011F060;
-    fp->take_dmg_cb = ftPp_Init_8011F060;
-}
-
-void fn_80122B54(Fighter_GObj* gobj)
-{
-    Fighter_GObj* new_var;
-    Fighter* fp = gobj->user_data;
-    Fighter* fp2;
-    new_var = gobj;
-    PAD_STACK(8);
-    ftCommon_8007D7FC(fp);
-    fp2 = fp;
-    Fighter_ChangeMotionState(new_var, ftPp_MS_SpecialLw, 0x0C4C5282,
-                              fp2->cur_anim_frame, 1.0F, 0.0F, 0L);
-    ftPp_set_cbs(new_var);
-    fp2 = gobj->user_data;
-    if (fp2->cmd_vars[3] != 0) {
-        ftPartSetRotX(fp2, 0,
-                      fp2->facing_dir * atan2f(fp2->coll_data.floor.normal.x,
-                                               fp2->coll_data.floor.normal.y));
-    } else {
-        ftPartSetRotX(fp2, 0, 0.0F);
-    }
-    fp->accessory4_cb = fn_80122D2C;
-}
-
-static inline void ftPp_SpecialLw_Coll_Land(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    ftCommon_8007D5D4(fp);
-    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialAirLw, 0x0C4C5282,
-                              fp->cur_anim_frame, 1.0f, 0.0f, NULL);
-}
-
-void ftPp_SpecialLw_Coll(Fighter_GObj* gobj)
-{
-    Fighter* fp;
-    PAD_STACK(8);
-    if (ft_80082708(gobj) == GA_Ground) {
-        ftPartSetRotX(GET_FIGHTER(gobj), 0, 0.0f);
-        fp = GET_FIGHTER(gobj);
-        ftCommon_8007D5D4(fp);
-        Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialAirLw, 0x0C4C5282,
-                                  fp->cur_anim_frame, 1.0f, 0.0f, NULL);
-        {
-            Fighter* temp_fp = gobj->user_data;
-            temp_fp->death2_cb = ftPp_Init_8011F060;
-            temp_fp->take_dmg_cb = ftPp_Init_8011F060;
-        }
-        fp->accessory4_cb = fn_80122D2C;
-        ftCommon_ClampAirDrift(fp);
-    } else {
-        ftPp_SpecialLw_Coll_inline(gobj);
-    }
-}
-
-void ftPp_SpecialAirLw_Coll(Fighter_GObj* gobj)
-{
-    ft_80082C74(gobj, fn_80122B54);
-}
-
-void fn_80122D2C(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    Vec3 pos;
-    PAD_STACK(8);
-    if (fp->mv.pp.speciallw.x4_b0) {
-        if (fp->mv.pp.speciallw.x0 == 0) {
-            ftIceClimberAttributes* da = fp->dat_attrs;
-            lb_8000B1CC(fp->parts[FtPart_L3rdNa].joint, NULL, &pos);
-            pos.x += da->xBC * fp->facing_dir;
-            pos.y += da->xC0;
-            itClimbersBlizzard_Spawn(gobj, &pos, fp->facing_dir);
-            fp->mv.pp.speciallw.x0 = da->xB8;
-        }
-        fp->mv.pp.speciallw.x0--;
-    }
-    switch (fp->cmd_vars[0]) {
-    case 1:
-        efSync_Spawn(0x4EC, gobj, fp->parts[FtPart_L4thNb].joint);
-        fp->mv.pp.speciallw.x4_b0 = true;
-        fp->fv.pp.x2230_b0 = true;
-        ftPp_set_cbs(gobj);
-        fp->cmd_vars[0] = 0;
-        if (fp->kind == FTKIND_POPO) {
-            ft_800881D8(fp, 0x1FC42, 0x7F, 0x40);
-            ft_80088510(fp, 0x1FBEB, 0x7F, 0x40);
-        } else {
-            ft_800881D8(fp, 0x1FC0F, 0x7F, 0x40);
-            ft_80088510(fp, 0x1FBEE, 0x7F, 0x40);
-        }
-        break;
-    case 2:
-        fp->mv.pp.speciallw.x4_b0 = false;
-        fp->cmd_vars[0] = 0;
-        break;
-    }
-}
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.h b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.h
index 3059160b5a..bc628cbb19 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.h
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.h
@@ -31,36 +31,5 @@
 /* 1210C8 */ void ftPp_SpecialS_801210C8(Fighter_GObj* gobj);
 /* 12114C */ void ftPp_SpecialS_8012114C(Fighter_GObj* gobj);
 /* 121164 */ void ftPp_SpecialS_80121164(Fighter_GObj* gobj);
-/* 1216E8 */ void ftPp_SpecialHiStart_0_Coll(Fighter_GObj* gobj);
-/* 121720 */ void ftPp_SpecialHiThrow_0_IASA(Fighter_GObj* gobj);
-/* 121728 */ void ftPp_SpecialAirHiThrow_0_IASA(Fighter_GObj* gobj);
-/* 121730 */ void ftPp_SpecialHiThrow_0_Coll(Fighter_GObj* gobj);
-/* 1217EC */ void ftPp_SpecialHi_801217EC(Fighter_GObj* gobj);
-/* 12184C */ void ftPp_SpecialHi_8012184C(Fighter_GObj* gobj);
-/* 121CE0 */ void ftPp_SpecialHi_80121CE0(Fighter_GObj* gobj);
-/* 121D40 */ void ftPp_SpecialHi_80121D40(Fighter_GObj* gobj);
-/* 121DD8 */ void ftPp_SpecialHi_80121DD8(Fighter_GObj* gobj);
-/* 121E90 */ void ftPp_SpecialHiStart_1_IASA(Fighter_GObj* gobj);
-/* 121E98 */ void ftPp_SpecialAirHiStart_1_IASA(Fighter_GObj* gobj);
-/* 121EB8 */ void ftPp_SpecialHiStart_1_Phys(Fighter_GObj* gobj);
-/* 121FD8 */ void ftPp_SpecialHi_80121FD8(Fighter_GObj* gobj);
-/* 122038 */ void ftPp_SpecialHi_80122038(Fighter_GObj* gobj);
-/* 122098 */ void ftPp_SpecialHi_80122098(Fighter_GObj* gobj);
-/* 1220D4 */ void ftPp_SpecialHi_801220D4(Fighter_GObj* gobj);
-/* 1220D4 */ void ftPp_SpecialHiThrow_1_Anim(Fighter_GObj* gobj);
-/* 1221A4 */ void ftPp_SpecialAirHiThrow_1_IASA(Fighter_GObj* gobj);
-/* 1221AC */ void ftPp_SpecialHiThrow_1_IASA(Fighter_GObj* gobj);
-/* 1221B4 */ void ftPp_SpecialHiThrow_1_Phys(Fighter_GObj* gobj);
-/* 122228 */ void ftPp_SpecialHiThrow_1_Coll(Fighter_GObj* gobj);
-/* 1222E8 */ void ftPp_SpecialHi_801222E8(Fighter_GObj* gobj);
-/* 122348 */ void ftPp_SpecialHi_80122348(Fighter_GObj* gobj);
-/* 122380 */ void ftPp_SpecialHi_80122380(Fighter_GObj* gobj);
-/* 1227AC */ void ftPp_SpecialHi_801227AC(Fighter_GObj* gobj);
-/* 122B14 */ void ftPp_SpecialLw_Phys(Fighter_GObj* gobj);
-/* 122B34 */ void ftPp_SpecialAirLw_Phys(Fighter_GObj* gobj);
-/* 122B54 */ void fn_80122B54(Fighter_GObj* gobj);
-/// ftPp_SpecialLw_Coll
-/* ??? */ void ftPp_SpecialAirLw_Coll(Fighter_GObj* gobj);
-/* 122D2C */ void fn_80122D2C(Fighter_GObj* gobj);
 
 #endif
diff --git a/src/melee/ft/ftdata.c b/src/melee/ft/ftdata.c
index f572674c6c..f70c8bfd91 100644
--- a/src/melee/ft/ftdata.c
+++ b/src/melee/ft/ftdata.c
@@ -209,8 +209,9 @@
 #include "ftPikachu/ftPk_SpecialLw.h"
 #include "ftPikachu/ftPk_SpecialN.h"
 #include "ftPikachu/ftPk_SpecialS.h"
-#include "ftPopo/ftPp_1211.h"
 #include "ftPopo/ftPp_Init.h"
+#include "ftPopo/ftPp_SpecialHi.h"
+#include "ftPopo/ftPp_SpecialLw.h"
 #include "ftPopo/ftPp_SpecialN.h"
 #include "ftPopo/ftPp_SpecialS.h"
 #include "ftPurin/ftPr_Init.h"
