diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index 9d016b18e7..cc1a7ace02 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -1301,7 +1301,7 @@ typedef struct itSScopeAttributes {
     /* +0 */ char pad_x0[0x4];
     /* +4 */ s32 x4;
     /* +8 */ s32 x8;
-    /* +C */ char pad_xC[0x28];
+    /* +C */ float xC[10];
     /* +34 */ Vec3 x34;
 } itSScopeAttributes;
 
diff --git a/src/melee/it/items/itsscope.c b/src/melee/it/items/itsscope.c
index ea4f89fccf..18732a2585 100644
--- a/src/melee/it/items/itsscope.c
+++ b/src/melee/it/items/itsscope.c
@@ -1,5 +1,7 @@
 #include "itsscope.h"
 
+#include "itsscopebeam.h"
+
 #include <placeholder.h>
 #include <platform.h>
 
@@ -68,9 +70,52 @@ s32 it_80291CF4(Item_GObj* gobj, s32 arg1)
 
 /// #it_80291DAC
 
-/// #it_80291F14
+void it_80291F14(Item_GObj* gobj, int charge_level)
+{
+    Item* ip = GET_ITEM(gobj);
+    itSScopeAttributes* attrs = GET_ATTRS(ip);
+    int cost = charge_level;
+
+    switch (charge_level) {
+    case 0:
+        cost = (int) attrs->xC[0];
+        break;
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+    case 8:
+        cost = (int) attrs->xC[charge_level];
+        break;
+    case 9:
+        cost = (int) attrs->xC[9];
+        break;
+    }
 
-/// #it_80291FA8
+    ip->xD4C -= cost;
+    if (ip->xD4C < 0) {
+        ip->xD4C = 0;
+    }
+
+    PAD_STACK(16);
+}
+
+void it_80291FA8(Item_GObj* gobj, Vec3* pos, int charge_level,
+                 float scale)
+{
+    Item* ip = GET_ITEM(gobj);
+    int cost = it_80291D38(gobj, charge_level);
+    ip->xD4C -= cost;
+    if (ip->xD4C < 0) {
+        ip->xD4C = 0;
+    }
+    it_80298DEC(ip->owner, pos, charge_level, scale);
+
+    PAD_STACK(8);
+}
 
 void it_80292030(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itsscope.h b/src/melee/it/items/itsscope.h
index 728e2165da..e30821720a 100644
--- a/src/melee/it/items/itsscope.h
+++ b/src/melee/it/items/itsscope.h
@@ -14,10 +14,10 @@
 /* 291C98 */ void itSScope_Logic21_Spawned(Item_GObj*);
 /* 291CCC */ itSScopeAttributes* it_80291CCC(Item_GObj* gobj, Vec3* pos);
 /* 291CF4 */ s32 it_80291CF4(Item_GObj*, s32);
-/* 291D38 */ UNK_RET it_80291D38(UNK_PARAMS);
+/* 291D38 */ int it_80291D38(Item_GObj*, int);
 /* 291DAC */ int it_80291DAC(Item_GObj*, int);
-/* 291F14 */ UNK_RET it_80291F14(UNK_PARAMS);
-/* 291FA8 */ UNK_RET it_80291FA8(UNK_PARAMS);
+/* 291F14 */ void it_80291F14(Item_GObj*, int);
+/* 291FA8 */ void it_80291FA8(Item_GObj*, Vec3*, int, float);
 /* 292030 */ void it_80292030(Item_GObj* gobj);
 /* 292080 */ bool itSscope_UnkMotion0_Anim(Item_GObj* gobj);
 /* 292088 */ void itSscope_UnkMotion0_Phys(Item_GObj* gobj);
