diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 99fb3a41a9..7cbc3dcb8f 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -621,7 +621,7 @@ fn_80025CBC = .text:0x80025CBC; // type:function size:0x17C
 fn_80025E38 = .text:0x80025E38; // type:function size:0x174
 fn_80025FAC = .text:0x80025FAC; // type:function size:0x2F4
 fn_800262A0 = .text:0x800262A0; // type:function size:0x114
-fn_800263B4 = .text:0x800263B4; // type:function size:0x34
+lbAudioAx_ObjFree = .text:0x800263B4; // type:function size:0x34
 lbAudioAx_800263E8 = .text:0x800263E8; // type:function size:0xFC scope:global
 lbAudioAx_800264E4 = .text:0x800264E4; // type:function size:0x2C scope:global
 lbAudioAx_80026510 = .text:0x80026510; // type:function size:0xB4 scope:global
diff --git a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
index 526134d21d..e361e795a8 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
@@ -1347,7 +1347,14 @@ void fn_800DC014(Fighter_GObj* gobj)
     }
 }
 
-/// #fn_800DC044
+bool fn_800DC044(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->input.lstick.y >= p_ftCommonData->tap_jump_threshold) {
+        return true;
+    }
+    return false;
+}
 
 /// #fn_800DC070
 
diff --git a/src/melee/ft/chara/ftCommon/ftCo_Attack100.h b/src/melee/ft/chara/ftCommon/ftCo_Attack100.h
index 4ebe3d4146..6fb33a72f1 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Attack100.h
+++ b/src/melee/ft/chara/ftCommon/ftCo_Attack100.h
@@ -172,9 +172,9 @@
 /* 0DBD50 */ void ftCo_CaptureWaitLw_Phys(Fighter_GObj* gobj);
 /* 0DBE9C */ void ftCo_CaptureWaitLw_Coll(Fighter_GObj* gobj);
 /* 0DBED4 */ UNK_RET fn_800DBED4(UNK_PARAMS);
-/* 0DC014 */ void fn_800DC014(Fighter_GObj*);
-/* 0DC044 */ bool fn_800DC044(Fighter_GObj*);
-/* 0DC070 */ void fn_800DC070(Fighter_GObj*);
+/* 0DC014 */ void fn_800DC014(Fighter_GObj* gobj);
+/* 0DC044 */ bool fn_800DC044(Fighter_GObj* gobj);
+/* 0DC070 */ void fn_800DC070(Fighter_GObj* gobj);
 /* 0DC110 */ void ftCo_CaptureJump_Anim(Fighter_GObj* gobj);
 /* 0DC160 */ void ftCo_CaptureJump_IASA(Fighter_GObj* gobj);
 /* 0DC224 */ void ftCo_CaptureJump_Phys(Fighter_GObj* gobj);
