diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index 966e47b971..9d016b18e7 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -508,20 +508,30 @@ typedef struct itTincle_ItemVars {
     /* +20 ip+DF4 */ s32 x20;
     /* +24 ip+DF8 */ u8 pad1[0x8];
     /* +2C ip+E00 */ s32 x2C;
-    /* +30 ip+E04 */ u8 pad1b[0x8];
+    /* +30 ip+E04 */ f32 x30;
+    /* +34 ip+E08 */ u8 pad1b[0x4];
     /* +38 ip+E0C */ f32 x38;
-    /* +3C ip+E10 */ u8 pad1c[0x10];
+    /* +3C ip+E10 */ u8 pad1c[0x4];
+    /* +40 ip+E14 */ f32 x40;
+    /* +44 ip+E18 */ f32 x44;
+    /* +48 ip+E1C */ f32 x48;
     /* +4C ip+E20 */ f32 x4C;
     /* +50 ip+E24 */ f32 x50;
     /* +54 ip+E28 */ f32 x54;
     /* +58 ip+E2C */ u8 pad2[0x4];
     /* +5C ip+E30 */ f32 x5C;
+    /* +60 ip+E34 */ u8 pad3[0x8];
+    /* +68 ip+E3C */ itECB x68;
 } itTincle_ItemVars;
 
 typedef struct itTincleAttributes {
-    /* 0x00 */ u8 pad[0x44];
+    /* 0x00 */ u8 pad[0x40];
+    /* 0x40 */ s32 x40;
     /* 0x44 */ f32 x44;
     /* 0x48 */ f32 x48;
+    /* 0x4C */ u8 pad2[0x8];
+    /* 0x54 */ s8 x54;
+    /* 0x55 */ s8 x55;
 } itTincleAttributes;
 
 typedef struct itGShell_HurtInit {
@@ -1389,7 +1399,9 @@ typedef struct itOldottosea_ItemVars {
 } itOldottosea_ItemVars;
 
 typedef struct itPatapata_ItemVars {
-    /* 0x00 */ u8 pad[0x20];
+    /* 0x00 */ u8 pad0[0x4];
+    /* 0x04 */ s32 x4;
+    /* 0x08 */ u8 pad[0x20 - 0x08];
     /* 0x20 */ s32 x20;
     /* 0x24 */ s32 x24;
     /* 0x28 */ s32 x28;
@@ -1397,6 +1409,7 @@ typedef struct itPatapata_ItemVars {
     /* 0x30 */ f32 x30;
     /* 0x34 */ u8 pad1[0x40 - 0x34];
     /* 0x40 */ s32 x40;
+    /* 0x44 */ s32 x44;
 } itPatapata_ItemVars;
 
 typedef struct itPatapataDatAttrs {
@@ -1415,6 +1428,8 @@ typedef struct itPatapataAttributes {
     /* 0x1C */ s32 x1C;
     /* 0x20 */ f32 x20;
     /* 0x24 */ s32 x24;
+    /* 0x28 */ s32 x28;
+    /* 0x2C */ s32 x2C;
 } itPatapataAttributes;
 
 typedef struct itOldottoseaAttributes {
@@ -1513,6 +1528,7 @@ typedef struct itUnknownAttributes {
     /* +18 */ s32 x18;
     /* +1C */ s32 x1C;
     /* +20 */ s32 x20;
+    /* +24 */ HSD_Joint* x24[26];
 } itUnknownAttributes;
 
 typedef struct itCrazyHandBombAttributes {
diff --git a/src/melee/it/items/itoldottosea.c b/src/melee/it/items/itoldottosea.c
index 111089c60d..92a119325e 100644
--- a/src/melee/it/items/itoldottosea.c
+++ b/src/melee/it/items/itoldottosea.c
@@ -55,7 +55,21 @@ bool itOldottosea_UnkMotion0_Coll(Item_GObj* gobj)
 
 /// #itOldottosea_UnkMotion2_Phys
 
-/// #itOldottosea_UnkMotion2_Coll
+bool itOldottosea_UnkMotion2_Coll(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (it_8026D8A4(gobj, it_802E3098)) {
+        if (ip->x40_vel.x != 0.0f) {
+            if (it_80276308(gobj) != 0) {
+                it_802E3098(gobj);
+            }
+        }
+    } else {
+        it_802E2BC0(gobj);
+    }
+    PAD_STACK(8);
+    return it_802E35CC(gobj);
+}
 
 void it_802E2BC0(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itpatapata.c b/src/melee/it/items/itpatapata.c
index 197584ff74..58501faf06 100644
--- a/src/melee/it/items/itpatapata.c
+++ b/src/melee/it/items/itpatapata.c
@@ -10,8 +10,14 @@
 #include "it/it_2725.h"
 #include "it/itCommonItems.h"
 #include "it/item.h"
+#include "it/items/itnokonoko.h"
+#include "lb/lb_00B0.h"
 #include "mp/mpcoll.h"
 
+#include <baselib/jobj.h>
+
+#include "cm/camera.h"
+
 void it_802E05A0(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -211,7 +217,32 @@ bool itPatapata_UnkMotion6_Coll(Item_GObj* gobj)
 
 /// #it_802E11E0
 
-/// #itPatapata_UnkMotion4_Anim
+bool itPatapata_UnkMotion4_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (ip->xDD4_itemVar.patapata.x28 == 0) {
+        itPatapataAttributes* attrs =
+            ip->xC4_article_data->x4_specialAttributes;
+        s32 toggle;
+        ip->xDD4_itemVar.patapata.x28 = attrs->x2C;
+        toggle = ip->xDD4_itemVar.patapata.x44 ^ 1;
+        ip->xDD4_itemVar.patapata.x44 = toggle;
+        if (toggle != 0) {
+            HSD_JObjSetFlagsAll(
+                ip->xBBC_dynamicBoneTable->bones[18], 0x10);
+            HSD_JObjSetFlagsAll(
+                ip->xBBC_dynamicBoneTable->bones[11], 0x10);
+        } else {
+            HSD_JObjClearFlagsAll(
+                ip->xBBC_dynamicBoneTable->bones[18], 0x10);
+            HSD_JObjClearFlagsAll(
+                ip->xBBC_dynamicBoneTable->bones[11], 0x10);
+        }
+    } else {
+        ip->xDD4_itemVar.patapata.x28 -= 1;
+    }
+    return false;
+}
 
 /// #itPatapata_UnkMotion4_Phys
 
@@ -226,7 +257,26 @@ bool itPatapata_UnkMotion4_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802E15B0
+void it_802E15B0(Item_GObj* gobj)
+{
+    Vec3 pos;
+    Item_GObj* new_gobj;
+    Item* new_ip;
+    HSD_JObj* jobj = gobj->hsd_obj;
+    Item* ip = GET_ITEM(gobj);
+
+    lb_8000B1CC(jobj, NULL, &pos);
+    Camera_80030E44(2, &ip->pos);
+    new_gobj = it_802DD7F0(ip->xDD4_itemVar.patapata.x40, &pos,
+                           &ip->x40_vel, (s32) ip->facing_dir);
+    new_ip = GET_ITEM(new_gobj);
+    new_ip->xDD4_itemVar.nokonoko.x34 = ip->xCC8_knockback;
+    new_ip->xDD4_itemVar.nokonoko.x3C = ip->xCAC_angle;
+    new_ip->xDD4_itemVar.nokonoko.x38 = ip->xCCC_incDamageDirection;
+    new_ip->xDD4_itemVar.nokonoko.x4 = ip->xDD4_itemVar.patapata.x4;
+    ip->xDD4_itemVar.patapata.x4 = -1;
+    it_802DCE00(new_gobj);
+}
 
 void it_802E1648(Item_GObj* gobj, int arg1, int arg2)
 {
diff --git a/src/melee/it/items/itpatapata.h b/src/melee/it/items/itpatapata.h
index 7a24436004..f12498c56a 100644
--- a/src/melee/it/items/itpatapata.h
+++ b/src/melee/it/items/itpatapata.h
@@ -33,7 +33,7 @@
 /* 2E13EC */ bool itPatapata_UnkMotion4_Anim(Item_GObj* gobj);
 /* 2E1498 */ void itPatapata_UnkMotion4_Phys(Item_GObj* gobj);
 /* 2E1578 */ bool itPatapata_UnkMotion4_Coll(Item_GObj* gobj);
-/* 2E15B0 */ UNK_RET it_802E15B0(UNK_PARAMS);
+/* 2E15B0 */ void it_802E15B0(Item_GObj*);
 /* 2E1648 */ void it_802E1648(Item_GObj*, int, int);
 /* 2E1694 */ void it_802E1694(Item_GObj*);
 /* 2E16D8 */ void it_802E16D8(Item_GObj*, Item_GObj*);
diff --git a/src/melee/it/items/ittincle.c b/src/melee/it/items/ittincle.c
index 0509c1d3f9..e5c36228bb 100644
--- a/src/melee/it/items/ittincle.c
+++ b/src/melee/it/items/ittincle.c
@@ -7,6 +7,8 @@
 #include "it/itCommonItems.h"
 #include "it/item.h"
 
+#include <baselib/random.h>
+
 /* 2EC870 */ static void it_802EC870(Item_GObj*, int);
 
 /// #it_802EB5C8
@@ -119,7 +121,25 @@ bool itTincle_UnkMotion3_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802EBE5C
+void it_802EBE5C(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itTincleAttributes* data = ip->xC4_article_data->x4_specialAttributes;
+
+    ip->xDD4_itemVar.tincle.x30 = (f32) ip->msid;
+    ip->xDD4_itemVar.tincle.x44 = ip->x40_vel.y;
+    ip->xDD4_itemVar.tincle.x48 = ip->xDD4_itemVar.tincle.x38;
+    ip->xDD4_itemVar.tincle.x40 = (f32) ip->xDD4_itemVar.tincle.x2C;
+
+    ip->x40_vel.z = 0.0f;
+    ip->x40_vel.y = 0.0f;
+    ip->x40_vel.x = 0.0f;
+
+    ip->xDD4_itemVar.tincle.x20 = data->x40;
+    ip->xDD4_itemVar.tincle.x4C = ip->pos.y;
+
+    Item_80268E5C(gobj, 4, 2);
+}
 
 bool itTincle_UnkMotion4_Anim(Item_GObj* gobj)
 {
@@ -206,7 +226,26 @@ void it_802EC1F4(Item_GObj* gobj)
     it_802EC9E8(gobj);
 }
 
-/// #itTincle_UnkMotion7_Anim
+bool itTincle_UnkMotion7_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(8);
+
+    if (ip->xDD4_itemVar.tincle.x20 == 0) {
+        ip->xDD4_itemVar.tincle.x20 = -1;
+        if (ip->msid == 6) {
+            Item_80268E5C(gobj, 7, 1);
+        }
+    } else if (ip->xDD4_itemVar.tincle.x20 > 0) {
+        ip->xDD4_itemVar.tincle.x20 -= 1;
+    }
+
+    if (!it_80272C6C(gobj)) {
+        it_802EC35C(gobj);
+    }
+
+    return false;
+}
 
 void itTincle_UnkMotion7_Phys(Item_GObj* gobj) {}
 
@@ -280,7 +319,23 @@ bool itTincle_UnkMotion9_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_802EC4D0
+void it_802EC4D0(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itTincleAttributes* data = ip->xC4_article_data->x4_specialAttributes;
+    s32 range;
+    PAD_STACK(8);
+
+    range = data->x55 - data->x54;
+    if (range < 0) {
+        range = -range;
+    }
+
+    ip->xDD4_itemVar.tincle.x20 = HSD_Randi(range);
+    ip->xDD4_itemVar.tincle.x20 += data->x54;
+
+    Item_80268E5C(gobj, 10, 2);
+}
 
 /// #itTincle_UnkMotion10_Anim
 
@@ -371,7 +426,26 @@ void it_802EC850(Item_GObj* gobj, Item_GObj* ref_gobj)
 
 /// #it_802EC870
 
-/// #it_802EC9E8
+void it_802EC9E8(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itECB ecb = ip->xDD4_itemVar.tincle.x68;
+
+    switch (ip->msid) {
+    case 6:
+    case 7:
+    case 8:
+    case 9:
+    case 10:
+    case 11:
+        ecb.top *= 0.5f;
+        ecb.right *= 0.5f;
+        ecb.left *= 0.5f;
+        break;
+    }
+
+    it_80275D5C(gobj, &ecb);
+}
 
 /// #it_802ECA70
 
diff --git a/src/melee/it/items/itunknown.c b/src/melee/it/items/itunknown.c
index 02936f28a1..e847d48349 100644
--- a/src/melee/it/items/itunknown.c
+++ b/src/melee/it/items/itunknown.c
@@ -2,13 +2,30 @@
 
 #include "baselib/random.h"
 #include "ef/eflib.h"
+#include "gr/stage.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/itCommonItems.h"
 #include "it/item.h"
 
-/// #it_802CE710
+void it_802CE710(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itUnknownAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+    it_80273318(gobj, attr->x24[HSD_Randi(26)]);
+    if (HSD_Randi(2) != 0) {
+        ip->xDD4_itemVar.unknown.x60 = 1.0F;
+    } else {
+        ip->xDD4_itemVar.unknown.x60 = -1.0F;
+    }
+    ip->xDD4_itemVar.unknown.x64 = 0;
+    ip->xDD4_itemVar.unknown.x68 = attr->x18;
+    ip->xDCC_flag.b3 = false;
+    ip->facing_dir = 0.0F;
+    it_80279CDC(gobj, attr->x0);
+    it_802CEC24(gobj);
+}
 
 void it_802CE7CC(Item_GObj* gobj) {}
 
@@ -17,7 +34,18 @@ void it_802CE7D0(Item_GObj* gobj, Item_GObj* ref_gobj)
     it_8026B894(gobj, ref_gobj);
 }
 
-/// #itUnknown_UnkMotion0_Anim
+bool itUnknown_UnkMotion0_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (ip->pos.x > Stage_GetBlastZoneRightOffset() ||
+        ip->pos.x < Stage_GetBlastZoneLeftOffset() ||
+        ip->pos.y > Stage_GetBlastZoneTopOffset() ||
+        ip->pos.y < Stage_GetBlastZoneBottomOffset())
+    {
+        it_802CE8D0(gobj);
+    }
+    return false;
+}
 
 void itUnknown_UnkMotion0_Phys(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itunknown.h b/src/melee/it/items/itunknown.h
index a2c22438b7..f4d0a5c18f 100644
--- a/src/melee/it/items/itunknown.h
+++ b/src/melee/it/items/itunknown.h
@@ -13,7 +13,7 @@
 /* 2CE7F0 */ bool itUnknown_UnkMotion0_Anim(Item_GObj* gobj);
 /* 2CE870 */ void itUnknown_UnkMotion0_Phys(Item_GObj* gobj);
 /* 2CE8C8 */ bool itUnknown_UnkMotion0_Coll(Item_GObj* gobj);
-/* 2CE8D0 */ UNK_RET it_802CE8D0(UNK_PARAMS);
+/* 2CE8D0 */ void it_802CE8D0(Item_GObj*);
 /* 2CEB9C */ bool itUnknown_UnkMotion1_Anim(Item_GObj* gobj);
 /* 2CEC18 */ void itUnknown_UnkMotion1_Phys(Item_GObj* gobj);
 /* 2CEC1C */ bool itUnknown_UnkMotion1_Coll(Item_GObj* gobj);
