diff --git a/src/melee/ft/chara/ftCommon/ftCo_DownBound.c b/src/melee/ft/chara/ftCommon/ftCo_DownBound.c
index 09cb96a977..f557e3907c 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_DownBound.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_DownBound.c
@@ -190,9 +190,9 @@ void ftCo_80097D40(Fighter_GObj* gobj)
 }
 #undef SOLUTION
 
-void ftCo_80097D88(Fighter_GObj* gobj)
+static void inlineA1(Fighter_GObj* gobj)
 {
-    Fighter* fp = gobj->user_data;
+    Fighter* fp = GET_FIGHTER(gobj);
     if (fp->x2228_b2) {
         ftCo_80097AF4(gobj);
     } else {
@@ -200,6 +200,11 @@ void ftCo_80097D88(Fighter_GObj* gobj)
     }
 }
 
+void ftCo_80097D88(Fighter_GObj* gobj)
+{
+    inlineA1(gobj);
+}
+
 void ftCo_DownBound_Anim(Fighter_GObj* gobj)
 {
     RETURN_IF(ftAnim_IsFramesRemaining(gobj));
