diff --git a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
index 549676245d..308cc6e5af 100644
--- a/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
+++ b/src/melee/ft/chara/ftCrazyHand/ftCh_Init.c
@@ -11,6 +11,7 @@
 #include "cm/camera.h"
 #include "ft/chara/ftCommon/ftCo_Attack100.h"
 #include "ft/chara/ftCommon/ftCo_CaptureCut.h"
+#include "ft/chara/ftCommon/ftCo_Lift.h"
 #include "ft/chara/ftCommon/ftCo_Throw.h"
 #include "ft/chara/ftCommon/ftCo_Thrown.h"
 #include "ft/fighter.h"
@@ -30,11 +31,12 @@
 #include "ftMasterHand/types.h"
 #include "gr/stage.h"
 #include "it/it_26B1.h"
+#include "it/items/itcrazyhandbomb.h"
 #include "it/items/itmasterhandlaser.h"
-#include "mp/mplib.h"
 #include "lb/lb_00B0.h"
 #include "lb/lbaudio_ax.h"
 #include "lb/lbvector.h"
+#include "mp/mplib.h"
 #include "pl/player.h"
 
 #include <common_structs.h>
@@ -93,9 +95,23 @@ static inline HSD_JObj* get_jobj(HSD_GObj* gobj)
 /* 15B670 */ static void ftCh_GrabUnk1_8015B670(HSD_GObj* gobj);
 /* 15B998 */ static void ftCh_GrabUnk1_8015B998(HSD_GObj* gobj);
 /* 156198 */ static void ftCh_Init_80156198(HSD_GObj* gobj);
+/* 156AD8 */ static void ftCh_Init_80156AD8(HSD_GObj* gobj);
 /* 15B4EC */ bool fn_8015B4EC(Vec3* vec);
 /* 1577B4 */ static void fn_801577B4(HSD_GObj* gobj);
 /* 159AA4 */ static void fn_80159AA4(HSD_GObj* gobj);
+/* 15BA34 */ void ftCh_GrabUnk1_8015BA34(HSD_GObj*, HSD_GObjEvent, Vec3*);
+
+typedef struct {
+    u8 a;
+    u8 b;
+} u8_pair;
+
+struct ftCh_Init_803D4878_t {
+    s32 states[18];
+    u8 x48[20];
+    u8_pair x5C[6];
+    u8 x68[32];
+};
 
 extern f32 ftCh_Init_804DA068;
 extern f32 ftCh_Init_804DA06C;
@@ -836,11 +852,224 @@ static void ftCh_Init_801560D8(HSD_GObj* gobj)
     fp->fv.mh.x2258 = 0x156;
 }
 
-/// #ftCh_Init_80156198
+static void ftCh_Init_80156198(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->cmd_vars[1] = 0;
+    if (fp->fv.mh.x2258 == 0x185) {
+        ftCh_Init_801560D8(gobj);
+    } else {
+        ftCh_Init_80156018(gobj);
+    }
+}
+
+static void ftCh_Init_80156310(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCrazyHand_DatAttrs* da = fp->ft_data->ext_attr;
+    fp->cmd_vars[1] = 0;
+    fp->fv.mh.x2258 = ftMh_MS_Wait1_0;
+    fp->motion_id = 0x183;
+    fp->cur_pos.x = da->x18;
+    fp->cur_pos.y = da->x1C;
+    ftCh_Init_80156018(gobj);
+}
+
+extern struct ftCh_Init_803D4878_t ftCh_Init_803D4878;
+
+inline void doAnim0(HSD_GObj* gobj)
+{
+    /// @todo #GET_FIGHTER
+    Fighter* fp = gobj->user_data;
+    if (fp->motion_id == 0x155 || fp->motion_id == 0x156) {
+        fp->cur_pos = fp->fv.mh.x2240_pos;
+    } else {
+        fp->fv.mh.x2240_pos = fp->cur_pos;
+    }
+    if (fp->fv.mh.x2258 == 0x185) {
+        Fighter_ChangeMotionState(gobj, 0x156, 0, fp->cur_anim_frame, 1.0f,
+                                  0.0f, NULL);
+    } else {
+        Fighter_ChangeMotionState(gobj, 0x156, 0, 0.0f, 1.0f, 0.0f, NULL);
+    }
+    fp->fv.mh.x2258 = 0x156;
+}
+
+inline void doAnim1(HSD_GObj* gobj)
+{
+    /// @todo #GET_FIGHTER
+    Fighter* fp = gobj->user_data;
+    fp->cmd_vars[1] = 0;
+    if (fp->fv.mh.x2258 == 0x185) {
+        ftCh_Init_801560D8(gobj);
+    } else {
+        ftCh_Init_80156018(gobj);
+    }
+}
+
+void ftCh_Wait1_0_Anim(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    struct ftCh_Init_803D4878_t* data = &ftCh_Init_803D4878;
+    u8 _[4];
+    s32 slot_type = Player_GetPlayerSlotType(fp->player_id);
+    if (fp->cmd_vars[1] != 0) {
+        ftBossLib_8015C5F8(gobj);
+        fp->cmd_vars[1] = 0;
+    }
+    if (slot_type == 0) {
+        ftCh_Init_80156AD8(gobj);
+        if (!ftAnim_IsFramesRemaining(gobj)) {
+            if (fp->motion_id == ftMh_MS_Wait1_0) {
+                doAnim0(gobj);
+            } else if (fp->motion_id == ftMh_MS_Wait2_0) {
+                doAnim1(gobj);
+            }
+        }
+    } else {
+        if (--fp->fv.mh.x223C < 0) {
+            ftCrazyHand_DatAttrs* da = fp->ft_data->ext_attr;
+            Vec3 vec;
+
+            u8 __[20];
 
-/// #ftCh_Init_80156310
+            ftBossLib_8015BD24(fp->x1A88.level, &fp->fv.mh.x223C,
+                               fp->fv.ch.x2238, da->x0, da->x8, da->x4);
+            {
+                s32 tmp;
+                s32 tmp2;
+                u8_pair* qwe;
+                float rand;
+                if (fp->fv.ch.x2254 == da->xC) {
+                    fp->fv.ch.x2254 = 0;
+                    tmp = 2;
+                } else {
+                    if (ftBossLib_8015C2E0()) {
+                        tmp = data->x68[fp->fv.ch.x224C * 5 + HSD_Randi(4)];
+                    } else {
+                        tmp = data->x68[fp->fv.ch.x224C * 5 + HSD_Randi(5)];
+                    }
+                    if (tmp == 2) {
+                        fp->fv.ch.x2254 = 0;
+                    } else {
+                        fp->fv.ch.x2254++;
+                    }
+                }
+                qwe = &((u8_pair*) data->x5C)[tmp];
+                tmp2 = data->states[data->x48[qwe->a + HSD_Randi(qwe->b)]];
 
-/// #ftCh_Wait1_0_Anim
+                fp->fv.ch.x224C = tmp;
+                fp->fv.ch.x2250 = tmp2;
+                if (fp->x221D_b4) {
+                    tmp2 = ftMh_MS_Wait1_0;
+                }
+                switch (tmp2) {
+                case ftMh_MS_Entry:
+                    mpFloorGetLeft(0, &vec);
+                    vec.y = da->x24;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_80156F6C, &vec);
+                    break;
+                case ftMh_MS_Damage2:
+                    mpFloorGetLeft(0, &vec);
+                    vec.y = da->x20;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_80157080, &vec);
+                    break;
+                case ftMh_MS_SweepLoop:
+                    vec.x = da->x2C;
+                    vec.y = da->x30_pos2.x;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_8015746C, &vec);
+                    break;
+                case ftMh_MS_Slap:
+                    vec.x = da->x30_pos2.y;
+                    vec.y = da->x38;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_8015755C, &vec);
+                    break;
+                case ftMh_MS_WalkLoop:
+                    vec.x = da->x6C;
+                    vec.y = da->x70;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_801578E8, &vec);
+                    break;
+                case ftMh_MS_Drill:
+                    vec.x = da->x80;
+                    vec.y = da->x84;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_80157C50, &vec);
+                    break;
+                case ftMh_MS_RockCrushUp:
+                    ftCh_Init_80157DF8(gobj);
+                    break;
+                case ftMh_MS_RockCrushDown:
+                    vec.x = da->x4C;
+                    vec.y = da->x50;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_80158144, &vec);
+                    break;
+                case ftMh_MS_Poke1:
+                    rand = HSD_Randf();
+                    vec.x = da->x90 * rand * 2 + da->x88_pos.x - da->x90;
+                    rand = HSD_Randf();
+                    vec.y = da->x94 * rand * 2 + da->x88_pos.y - da->x94;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_801582D8, &vec);
+                    break;
+                case ftMh_MS_FingerBeamStart:
+                    vec.x = da->x98;
+                    vec.y = da->x9C;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_801587B0, &vec);
+                    break;
+                case ftMh_MS_BackAirplane2:
+                    ftCh_Init_80158B3C(gobj);
+                    break;
+                case ftMh_MS_BackAirplane3:
+                    ftCh_Init_80158B3C(gobj);
+                    break;
+                case ftMh_MS_BackDisappear:
+                    ftCh_Init_801597F0(gobj, fn_80159908);
+                    break;
+                case ftMh_MS_Wait1_1:
+                    ftCh_Init_801597F0(gobj, fn_80159AA4);
+                    break;
+                case ftMh_MS_Grab:
+                    ftCh_Init_80159F40(gobj);
+                    break;
+                case ftMh_MS_Squeezing1:
+                    vec.x = da->x104;
+                    vec.y = da->x108;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_8015AAC8, &vec);
+                    break;
+                case ftMh_MS_Squeeze:
+                    vec.x = da->x10C;
+                    vec.y = da->x110_pos.x;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_8015AAC8, &vec);
+                    break;
+                case ftMh_MS_Throw:
+                    vec.x = da->x110_pos.y;
+                    vec.y = da->x118_pos.x;
+                    vec.z = 0;
+                    ftCh_GrabUnk1_8015BA34(gobj, fn_8015AAC8, &vec);
+                    break;
+                case ftMh_MS_Slam:
+                    vec.x = da->x118_pos.y;
+                    vec.y = da->x120;
+                    vec.z = 0;
+                    break;
+                }
+            }
+        } else if (!ftAnim_IsFramesRemaining(gobj)) {
+            if (fp->motion_id == ftMh_MS_Wait1_0) {
+                doAnim0(gobj);
+            } else if (fp->motion_id == ftMh_MS_Wait2_0) {
+                doAnim1(gobj);
+            }
+        }
+    }
+}
 
 void ftCh_Wait1_0_IASA(HSD_GObj* gobj)
 {
@@ -857,7 +1086,102 @@ void ftCh_Wait1_0_Phys(HSD_GObj* gobj)
 
 void ftCh_Wait1_0_Coll(HSD_GObj* gobj) {}
 
-/// #ftCh_Init_80156AD8
+static inline u32 btn_pressed(u32 mask)
+{
+    return HSD_PadMasterStatus[3].button & mask;
+}
+
+static void ftCh_Init_80156AD8(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCrazyHand_DatAttrs* da = fp->ft_data->ext_attr;
+    Vec3 pos;
+
+    if (btn_pressed(HSD_PAD_L) && btn_pressed(HSD_PAD_DPADUP)) {
+        mpFloorGetLeft(0, &pos);
+        pos.y = da->x24;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_80157080, &pos);
+    } else if (btn_pressed(HSD_PAD_L) && btn_pressed(HSD_PAD_DPADRIGHT)) {
+        mpFloorGetLeft(0, &pos);
+        pos.y = da->x20;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_80156F6C, &pos);
+    } else if (btn_pressed(HSD_PAD_L) && btn_pressed(HSD_PAD_DPADDOWN)) {
+        pos.x = da->x2C;
+        pos.y = da->x30_pos2.x;
+        pos.z = 0;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_8015746C, &pos);
+    } else if (btn_pressed(HSD_PAD_L) && btn_pressed(HSD_PAD_DPADLEFT)) {
+        pos.x = da->x30_pos2.y;
+        pos.y = da->x38;
+        pos.z = 0;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_8015755C, &pos);
+    } else if (btn_pressed(HSD_PAD_R) && btn_pressed(HSD_PAD_DPADUP)) {
+        pos.x = da->x6C;
+        pos.y = da->x70;
+        pos.z = 0;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_801578E8, &pos);
+    } else if (btn_pressed(HSD_PAD_R) && btn_pressed(HSD_PAD_DPADRIGHT)) {
+        pos.x = da->x80;
+        pos.y = da->x84;
+        pos.z = 0;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_80157C50, &pos);
+    } else if (btn_pressed(HSD_PAD_R) && btn_pressed(HSD_PAD_DPADDOWN)) {
+        ftCh_Init_80157DF8(gobj);
+    } else if (btn_pressed(HSD_PAD_R) && btn_pressed(HSD_PAD_DPADLEFT)) {
+        pos.x = da->x4C;
+        pos.y = da->x50;
+        pos.z = 0;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_80158144, &pos);
+    } else if (btn_pressed(HSD_PAD_B) && btn_pressed(HSD_PAD_DPADUP)) {
+        float rand = HSD_Randf();
+        pos.x = da->x90 * rand * 2 + da->x88_pos.x - da->x90;
+        rand = HSD_Randf();
+        pos.y = da->x94 * rand * 2 + da->x88_pos.y - da->x94;
+        pos.z = 0;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_801582D8, &pos);
+    } else if (btn_pressed(HSD_PAD_B) && btn_pressed(HSD_PAD_DPADRIGHT)) {
+        pos.x = da->x98;
+        pos.y = da->x9C;
+        pos.z = 0;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_801587B0, &pos);
+    } else if (btn_pressed(HSD_PAD_A) && btn_pressed(HSD_PAD_DPADUP)) {
+        fp->fv.ch.x2250 = ftMh_MS_BackAirplane2;
+        ftCh_Init_80158B3C(gobj);
+    } else if (btn_pressed(HSD_PAD_A) && btn_pressed(HSD_PAD_DPADRIGHT)) {
+        fp->fv.ch.x2250 = ftMh_MS_BackAirplane3;
+        ftCh_Init_80158B3C(gobj);
+    } else if (btn_pressed(HSD_PAD_Z) && btn_pressed(HSD_PAD_DPADUP)) {
+        ftCh_Init_801597F0(gobj, fn_80159908);
+    } else if (btn_pressed(HSD_PAD_Z) && btn_pressed(HSD_PAD_DPADRIGHT)) {
+        ftCh_Init_801597F0(gobj, fn_80159AA4);
+    } else if (btn_pressed(HSD_PAD_Z) && btn_pressed(HSD_PAD_DPADDOWN)) {
+        ftCh_Init_80159F40(gobj);
+    } else if (btn_pressed(HSD_PAD_Y) && btn_pressed(HSD_PAD_DPADUP)) {
+        pos.x = da->x104;
+        pos.y = da->x108;
+        pos.z = 0;
+        fp->fv.ch.x2250 = ftMh_MS_Squeezing1;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_8015AAC8, &pos);
+    } else if (btn_pressed(HSD_PAD_Y) && btn_pressed(HSD_PAD_DPADRIGHT)) {
+        pos.x = da->x10C;
+        pos.y = da->x110_pos.x;
+        pos.z = 0;
+        fp->fv.ch.x2250 = ftMh_MS_Squeeze;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_8015AAC8, &pos);
+    } else if (btn_pressed(HSD_PAD_Y) && btn_pressed(HSD_PAD_DPADDOWN)) {
+        pos.x = da->x110_pos.y;
+        pos.y = da->x118_pos.x;
+        pos.z = 0;
+        fp->fv.ch.x2250 = ftMh_MS_Throw;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_8015AAC8, &pos);
+    } else if (btn_pressed(HSD_PAD_Y) && btn_pressed(HSD_PAD_DPADLEFT)) {
+        pos.x = da->x118_pos.y;
+        pos.y = da->x120;
+        pos.z = 0;
+        fp->fv.ch.x2250 = ftMh_MS_Slam;
+        ftCh_GrabUnk1_8015BA34(gobj, fn_8015AAC8, &pos);
+    }
+}
 
 void fn_80156F6C(HSD_GObj* gobj)
 {
@@ -946,10 +1270,10 @@ void ftCh_Damage2_Anim(HSD_GObj* gobj)
             attrs = fp->ft_data->ext_attr;
             fp->mv.ch.unk0.xC.x = fp->cur_pos.x - attrs->x28;
             fp->mv.ch.unk0.xC.y = attrs->x24;
-            Fighter_ChangeMotionState(
-                gobj, ftMh_MS_Damage, 0,
-                fp->mv.ch.unk0.xC.z = ftCh_Init_804DA070,
-                ftCh_Init_804DA074, ftCh_Init_804DA070, NULL);
+            Fighter_ChangeMotionState(gobj, ftMh_MS_Damage, 0,
+                                      fp->mv.ch.unk0.xC.z = ftCh_Init_804DA070,
+                                      ftCh_Init_804DA074, ftCh_Init_804DA070,
+                                      NULL);
             ftAnim_8006EBA4(gobj);
             ft_PlaySFX(fp, 0x4E207, 127, 64);
         }
@@ -1584,7 +1908,48 @@ void ftCh_FingerBeamStart_Phys(HSD_GObj* gobj)
 
 void ftCh_FingerBeamStart_Coll(HSD_GObj* gobj) {}
 
-/// #fn_801588B8
+static void fn_801588B8(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCrazyHand_DatAttrs* da = fp->ft_data->ext_attr;
+    Vec3 pos, prev;
+    if (fp->cmd_vars[0] != 0) {
+        lb_8000B1CC(fp->parts[FtPart_RLegJA].joint, NULL, &pos);
+        prev = pos;
+        pos.x += da->xA0;
+        pos.y += da->xA4;
+        it_802F0F6C(gobj, &pos, &prev, It_Kind_CrazyHand_Bomb, fp->facing_dir);
+        fp->cmd_vars[0] = 0;
+        ft_PlaySFX(fp, 0x4E203, 0x7F, 0x40);
+    }
+    if (fp->cmd_vars[1] != 0) {
+        lb_8000B1CC(fp->parts[FtPart_BustN].joint, NULL, &pos);
+        prev = pos;
+        pos.x += da->xA8_pos.x;
+        pos.y += da->xA8_pos.y;
+        it_802F0F6C(gobj, &pos, &prev, It_Kind_CrazyHand_Bomb, fp->facing_dir);
+        fp->cmd_vars[1] = 0;
+        ft_PlaySFX(fp, 0x4E203, 0x7F, 0x40);
+    }
+    if (fp->cmd_vars[2] != 0) {
+        lb_8000B1CC(fp->parts[FtPart_LHandN].joint, NULL, &pos);
+        prev = pos;
+        pos.x += da->xB0;
+        pos.y += da->xB4;
+        it_802F0F6C(gobj, &pos, &prev, It_Kind_CrazyHand_Bomb, fp->facing_dir);
+        fp->cmd_vars[2] = 0;
+        ft_PlaySFX(fp, 0x4E203, 0x7F, 0x40);
+    }
+    if (fp->cmd_vars[3] != 0) {
+        lb_8000B1CC(fp->parts[FtPart_L3rdNa].joint, NULL, &pos);
+        prev = pos;
+        pos.x += da->xB8;
+        pos.y += da->xBC_pos.x;
+        it_802F0F6C(gobj, &pos, &prev, It_Kind_CrazyHand_Bomb, fp->facing_dir);
+        fp->cmd_vars[3] = 0;
+        ft_PlaySFX(fp, 0x4E203, 0x7F, 0x40);
+    }
+}
 
 void ftCh_Init_80158B3C(HSD_GObj* gobj)
 {
@@ -2798,7 +3163,29 @@ bool fn_8015B4EC(Vec3* vec)
     return false;
 }
 
-/// #fn_8015B548
+static void fn_8015B548(HSD_GObj* gobj, HSD_GObj* gobj2)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter* fp2 = GET_FIGHTER(gobj2);
+    ftCommon_8007DB58(gobj);
+    ftCo_8009750C(gobj);
+    ftCo_800DD168(gobj);
+    fp->x1A5C = gobj2;
+    fp->victim_gobj = gobj2;
+    fp->x221B_b5 = false;
+    fp->x221B_b7 = false;
+    fp->facing_dir = fp2->facing_dir;
+    ftCommon_InitGrab(fp, 0, ftCo_800DA824(fp2));
+    ftCo_800DB368(fp2, fp);
+    fp->accessory1_cb = ftCo_800DB464;
+    ftCommon_8007D5D4(fp);
+    Fighter_ChangeMotionState(gobj, 0x150, 0, 0.0f, 1.0f, 0.0f, NULL);
+    fp->invisible = true;
+    ftCommon_8007E2F4(fp, 0x1FF);
+    fp->x2220_b3 = true;
+    ftAnim_8006EBA4(gobj);
+    ftCommon_8007E2FC(gobj);
+}
 
 void ftCo_CaptureCrazyHand_Anim(HSD_GObj* gobj) {}
 
@@ -3000,18 +3387,15 @@ void ftCh_GrabUnk1_8015BC88(HSD_GObj* gobj)
     }
 }
 
-struct {
-    enum_t asids[18];
-    s8 bytes[64];
-} ftCh_Init_803D4878 = {
+struct ftCh_Init_803D4878_t ftCh_Init_803D4878 = {
     { ftMh_MS_Entry, ftMh_MS_Damage2, ftMh_MS_SweepLoop, ftMh_MS_Slap,
       ftMh_MS_WalkLoop, ftMh_MS_Drill, ftMh_MS_RockCrushUp,
       ftMh_MS_RockCrushDown, ftMh_MS_BackDisappear, ftMh_MS_Wait1_1,
       ftMh_MS_Grab, ftMh_MS_Poke1, ftMh_MS_FingerBeamStart,
       ftMh_MS_BackAirplane2, ftMh_MS_BackAirplane3, ftMh_MS_Squeezing1,
       ftMh_MS_Squeeze, ftMh_MS_Throw },
-    { 0,  1,  2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
-      16, 17, 0, 0, 0, 4, 4, 4, 8, 3, 11, 2,  13, 2,  15, 3,
-      1,  2,  3, 4, 5, 0, 2, 3, 4, 5, 0,  1,  3,  4,  5,  0,
-      1,  2,  4, 5, 0, 1, 2, 3, 5, 0, 1,  2,  3,  4,  0,  0 },
+    { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 0, 0 },
+    { { 0, 4 }, { 4, 4 }, { 8, 3 }, { 11, 2 }, { 13, 2 }, { 15, 3 } },
+    { 1, 2, 3, 4, 5, 0, 2, 3, 4, 5, 0, 1, 3, 4, 5, 0,
+      1, 2, 4, 5, 0, 1, 2, 3, 5, 0, 1, 2, 3, 4, 0, 0 }
 };
diff --git a/src/melee/ft/chara/ftCrazyHand/types.h b/src/melee/ft/chara/ftCrazyHand/types.h
index b4a0e42559..c350a3eef4 100644
--- a/src/melee/ft/chara/ftCrazyHand/types.h
+++ b/src/melee/ft/chara/ftCrazyHand/types.h
@@ -14,7 +14,7 @@ typedef struct _ftCrazyHandAttributes {
     float x14;
     float x18;
     float x1C;
-    s32 x20;
+    float x20;
     float x24;
     float x28;
     float x2C;
@@ -32,23 +32,23 @@ typedef struct _ftCrazyHandAttributes {
     float x60;
     s32 x64;
     float x68;
-    s32 x6C;
-    s32 x70;
+    float x6C;
+    float x70;
     s32 x74;
     s32 x78;
     float x7C;
     float x80;
-    s32 x84;
+    float x84;
     Vec2 x88_pos;
-    s32 x90;
-    s32 x94;
+    float x90;
+    float x94;
     float x98;
     float x9C;
-    s32 xA0;
+    float xA0;
     float xA4;
     Vec2 xA8_pos;
-    s32 xB0;
-    s32 xB4;
+    float xB0;
+    float xB4;
     float xB8;
     Vec2 xBC_pos;
     Vec2 xC4_pos;
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.c b/src/melee/ft/chara/ftKirby/ftKb_Init.c
index afcc4d33f2..f8bccb2007 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.c
@@ -3353,7 +3353,51 @@ void ftKb_SpecialN_800EF35C(Fighter_GObj* gobj, int arg1, u8* arg2)
 
 /// #ftKb_SpecialN_800EF69C
 
-/// #ftKb_UnkIntBoolFunc0
+void ftKb_UnkIntBoolFunc0(Fighter* fp, int arg1, bool arg2)
+{
+    if (fp->fv.kb.hat.x14.data != NULL) {
+        if (arg2) {
+            if (fp->fv.kb.hat.jobj != NULL) {
+                ftParts_80074CA0(&fp->fv.kb.hat.x24, arg1,
+                                 &fp->fv.kb.hat.x14);
+                return;
+            }
+            {
+                s32 var_r0;
+                s8 idx = fp->x5F4_arr[0].idx;
+                if (idx >= 2 && idx <= 6) {
+                    var_r0 = 1;
+                } else {
+                    var_r0 = 0;
+                }
+                if (var_r0 != 0) {
+                    ftParts_80074D7C(&fp->fv.kb.hat.x24, arg1,
+                                     &fp->fv.kb.hat.x14);
+                    return;
+                }
+            }
+            if (arg1 == 2) {
+                if (*fp->fv.kb.hat.x1C.data != NULL) {
+                    ftParts_80074D7C(&fp->x5AC, arg1, &fp->x203C);
+                    ftParts_80074CA0(&fp->fv.kb.hat.x24, arg1,
+                                     &fp->fv.kb.hat.x1C);
+                }
+            } else {
+                ftParts_80074D7C(&fp->x5AC, arg1, &fp->dobj_list);
+                ftParts_80074CA0(&fp->fv.kb.hat.x24, arg1,
+                                 &fp->fv.kb.hat.x14);
+            }
+        } else {
+            if (fp->fv.kb.hat.jobj == NULL && arg1 == 2) {
+                ftParts_80074D7C(&fp->fv.kb.hat.x24, arg1,
+                                 &fp->fv.kb.hat.x1C);
+                return;
+            }
+            ftParts_80074D7C(&fp->fv.kb.hat.x24, arg1,
+                             &fp->fv.kb.hat.x14);
+        }
+    }
+}
 
 void ftKb_Init_UnkCallbackPairs0_0(Fighter_GObj* gobj)
 {
@@ -4063,7 +4107,27 @@ void ftKb_Init_UnkMotionStates3(Fighter_GObj* gobj)
     }
 }
 
-/// #ftKb_SpecialN_800F1BAC
+void ftKb_SpecialN_800F1BAC(Fighter_GObj* gobj, s32 kind, bool arg2)
+{
+    Fighter* fp = gobj->user_data;
+    HSD_GObjEvent cb;
+
+    if (fp->fv.kb.hat.kind != kind) {
+        fp->fv.kb.hat.kind = kind;
+        ftKb_SpecialN_800F190C(gobj, fp->fv.kb.hat.kind);
+        if ((cb = ftKb_Init_803C9CC8[fp->fv.kb.hat.kind * 2]) != NULL) {
+            cb(gobj);
+        }
+        ftKb_SpecialN_800F16D0(gobj, fp->fv.kb.hat.kind);
+        if (arg2 == 1) {
+            ft_PlaySFX(fp, 0x222F3, 0x7F, 0x40);
+        }
+    } else if (arg2 == 1) {
+        ft_PlaySFX(fp, 0x222F6, 0x7F, 0x40);
+    }
+    fp->death3_cb = ftKb_Init_800EE74C;
+    fp->death1_cb = ftKb_Init_800EE7B8;
+}
 
 static bool fn_800F1CA0(HSD_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.h b/src/melee/ft/chara/ftKirby/ftKb_Init.h
index ea4bc02e14..80f50d4b7a 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.h
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.h
@@ -105,7 +105,8 @@
 /* 0F15D8 */ void ftKb_SpecialN_800F15D8(Fighter_GObj* gobj);
 /* 0F1630 */ void ftKb_SpecialN_Enter(Fighter_GObj* gobj);
 /* 0F1680 */ void ftKb_SpecialAirN_Enter(Fighter_GObj* gobj);
-/* 0F16D0 */ void ftKb_SpecialN_800F16D0(FighterKind victim_kind);
+/* 0F16D0 */ void ftKb_SpecialN_800F16D0(Fighter_GObj* gobj,
+                                         FighterKind victim_kind);
 /* 0F190C */ void ftKb_SpecialN_800F190C(Fighter_GObj* gobj,
                                          FighterKind victim_kind);
 /* 0F19AC */ void ftKb_SpecialN_800F19AC(Fighter_GObj* gobj);
@@ -796,14 +797,14 @@
 /* 1092AC */ float ftKb_SpecialNYs_801092AC(Fighter_GObj* gobj);
 /* 1092BC */ float ftKb_SpecialNYs_801092BC(Fighter_GObj* gobj);
 /* 1092CC */ float ftKb_SpecialNYs_801092CC(Fighter_GObj* gobj);
-/* 1092DC */ f32 ftKb_SpecialNYs_801092DC(Fighter_GObj* gobj);
-/* 1092F4 */ float ftKb_SpecialNYs_801092F4(Fighter_GObj* gobj);
-/* 10930C */ float ftKb_SpecialNYs_8010930C(Fighter_GObj* gobj);
-/* 109324 */ float ftKb_SpecialNYs_80109324(Fighter_GObj* gobj);
-/* 10933C */ int ftKb_SpecialNYs_8010933C(Fighter_GObj* gobj);
+/* 1092DC */ f32 ftKb_SpecialNYs_801092DC(void);
+/* 1092F4 */ float ftKb_SpecialNYs_801092F4(void);
+/* 10930C */ float ftKb_SpecialNYs_8010930C(void);
+/* 109324 */ float ftKb_SpecialNYs_80109324(void);
+/* 10933C */ int ftKb_SpecialNYs_8010933C(void);
 /* 109354 */ void ftKb_SpecialNYs_80109354(Vec3*);
-/* 109380 */ float ftKb_SpecialNYs_80109380(Fighter_GObj* gobj);
-/* 1093A0 */ ftDynamics* ftKb_SpecialNYs_801093A0(Fighter_GObj* gobj);
+/* 109380 */ float ftKb_SpecialNYs_80109380(void);
+/* 1093A0 */ ftDynamics* ftKb_SpecialNYs_801093A0(void);
 /* 1093B4 */ void ftKb_SpecialNYs_801093B4(Fighter_GObj* gobj);
 /* 10941C */ void ftKb_SpecialNYs_8010941C(Fighter_GObj* gobj);
 /* 1094FC */ void ftKb_SpecialNYs_801094FC(Fighter_GObj* gobj);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
index 9075f1fc92..648bfa8db7 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
@@ -3,6 +3,7 @@
 
 #include <placeholder.h>
 
+#include "ef/efasync.h"
 #include "ef/eflib.h"
 #include "ef/efsync.h"
 #include "ft/chara/ftCommon/ftCo_Bury.h"
@@ -26,6 +27,7 @@
 
 #include "ftKirby/forward.h"
 
+#include "it/items/itgamewatchchef.h"
 #include "it/items/itkirbygamewatchchefpan.h"
 #include "it/items/itkirbyyoshispecialn.h"
 #include "it/items/itpeachtoad.h"
@@ -88,31 +90,31 @@ f32 ftKb_SpecialNYs_801092CC(Fighter_GObj* gobj)
     return da->specialn_ys_base_duration;
 }
 
-f32 ftKb_SpecialNYs_801092DC(Fighter_GObj* gobj)
+f32 ftKb_SpecialNYs_801092DC(void)
 {
     ftKb_DatAttrs* ea = gFtDataList[FTKIND_KIRBY]->ext_attr;
     return ea->specialn_ys_egg_breakout_resistance;
 }
 
-f32 ftKb_SpecialNYs_801092F4(Fighter_GObj* gobj)
+f32 ftKb_SpecialNYs_801092F4(void)
 {
     ftKb_DatAttrs* ea = gFtDataList[FTKIND_KIRBY]->ext_attr;
     return ea->specialn_ys_frames_reduced_per_input;
 }
 
-f32 ftKb_SpecialNYs_8010930C(Fighter_GObj* gobj)
+f32 ftKb_SpecialNYs_8010930C(void)
 {
     ftKb_DatAttrs* ea = gFtDataList[FTKIND_KIRBY]->ext_attr;
     return ea->specialn_ys_unk1;
 }
 
-f32 ftKb_SpecialNYs_80109324(Fighter_GObj* gobj)
+f32 ftKb_SpecialNYs_80109324(void)
 {
     ftKb_DatAttrs* ea = gFtDataList[FTKIND_KIRBY]->ext_attr;
     return ea->specialn_ys_unk2;
 }
 
-int ftKb_SpecialNYs_8010933C(Fighter_GObj* gobj)
+int ftKb_SpecialNYs_8010933C(void)
 {
     ftKb_DatAttrs* ea = gFtDataList[FTKIND_KIRBY]->ext_attr;
     return ea->specialn_ys_iframes_on_release;
@@ -126,13 +128,13 @@ void ftKb_SpecialNYs_80109354(Vec3* v)
     v->z = 0.0f;
 }
 
-float ftKb_SpecialNYs_80109380(Fighter_GObj* gobj)
+float ftKb_SpecialNYs_80109380(void)
 {
     ftKb_DatAttrs* cd = gFtDataList[FTKIND_KIRBY]->ext_attr;
     return cd->specialn_ys_unk3 / cd->specialn_ys_damage_multiplier;
 }
 
-ftDynamics* ftKb_SpecialNYs_801093A0(Fighter_GObj* gobj)
+ftDynamics* ftKb_SpecialNYs_801093A0(void)
 {
     return ft_80459B88.hats[FTKIND_SAMUS]->hat_dynamics[0];
 }
@@ -417,13 +419,115 @@ void ftKb_YsSpecialAirNCapture2_Anim(Fighter_GObj* gobj)
     }
 }
 
-/// #ftKb_YsSpecialAirNCapture1_Anim
+void ftKb_YsSpecialAirNCapture1_Anim(Fighter_GObj* gobj)
+{
+    KirbyHatStruct* hat;
+    Fighter* fp2;
+    s32 new_var;
+    Fighter* fp = gobj->user_data;
+    new_var = 0;
+    do {
+        UNUSED unsigned char _[8];
+    } while (new_var);
+    if (fp->cmd_vars[0] != new_var && fp->target_item_gobj != NULL) {
+        fp->cmd_vars[0] = new_var;
+        fp->cmd_vars[1] = new_var;
+        fp2 = GET_FIGHTER(gobj);
+        hat = ft_80459B88.hats[0xD];
+        Fighter_ChangeMotionState(gobj, ftKb_MS_YsSpecialNCapture2_1, 0x80012,
+                                  0.0f, 1.0f, 0.0f, NULL);
+        HSD_JObjAddAnimAll(fp2->fv.kb.hat.jobj,
+                           (HSD_AnimJoint*) hat->hat_dynamics[2], NULL, NULL);
+        HSD_JObjReqAnimAll(fp2->fv.kb.hat.jobj, 0.0f);
+        fp2 = GET_FIGHTER(gobj);
+        fp2->death2_cb = ftKb_Init_800EE74C;
+        fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+        ftCommon_8007E2F4(fp, 0x1FF);
+    }
+}
 
-/// #ftKb_YsSpecialNCapture1_Anim
+void ftKb_YsSpecialNCapture1_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    KirbyHatStruct* hat;
+    Fighter* fp2;
+    s32 new_var;
+    new_var = 0;
+    do {
+        UNUSED unsigned char _[8];
+    } while (new_var);
+    HSD_JObjAnimAll(fp->fv.kb.hat.jobj);
+    if (fp->cmd_vars[0] != new_var && fp->victim_gobj != NULL) {
+        fp->cmd_vars[0] = new_var;
+        fp->cmd_vars[1] = new_var;
+        fp2 = GET_FIGHTER(gobj);
+        hat = ft_80459B88.hats[0xD];
+        Fighter_ChangeMotionState(gobj, ftKb_MS_YsSpecialNCapture2_0, 0x80012,
+                                  0.0f, 1.0f, 0.0f, NULL);
+        HSD_JObjAddAnimAll(fp2->fv.kb.hat.jobj,
+                           (HSD_AnimJoint*) hat->hat_dynamics[2], NULL, NULL);
+        HSD_JObjReqAnimAll(fp2->fv.kb.hat.jobj, 0.0f);
+        fp2 = GET_FIGHTER(gobj);
+        fp2->death2_cb = ftKb_Init_800EE74C;
+        fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+        ftCommon_8007E2F4(fp, 0x1FF);
+    }
+}
 
-/// #ftKb_YsSpecialAirCapture2_Anim
+void ftKb_YsSpecialAirCapture2_Anim(Fighter_GObj* gobj)
+{
+    KirbyHatStruct* hat;
+    Fighter* fp2;
+    s32 new_var;
+    Fighter* fp = gobj->user_data;
+    new_var = 0;
+    do {
+        UNUSED unsigned char _[8];
+    } while (new_var);
+    if (fp->cmd_vars[0] != new_var && fp->target_item_gobj != NULL) {
+        fp->cmd_vars[0] = new_var;
+        fp->cmd_vars[1] = new_var;
+        fp2 = GET_FIGHTER(gobj);
+        hat = ft_80459B88.hats[0xD];
+        Fighter_ChangeMotionState(gobj, ftKb_MS_YsSpecialAirN2_1, 0x80012,
+                                  0.0f, 1.0f, 0.0f, NULL);
+        HSD_JObjAddAnimAll(fp2->fv.kb.hat.jobj,
+                           (HSD_AnimJoint*) hat->hat_dynamics[4], NULL, NULL);
+        HSD_JObjReqAnimAll(fp2->fv.kb.hat.jobj, 0.0f);
+        fp2 = GET_FIGHTER(gobj);
+        fp2->death2_cb = ftKb_Init_800EE74C;
+        fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+        ftCommon_8007E2F4(fp, 0x1FF);
+    }
+}
 
-/// #ftKb_YsSpecialAirCapture1_Anim
+void ftKb_YsSpecialAirCapture1_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    KirbyHatStruct* hat;
+    Fighter* fp2;
+    s32 new_var;
+    new_var = 0;
+    do {
+        UNUSED unsigned char _[8];
+    } while (new_var);
+    HSD_JObjAnimAll(fp->fv.kb.hat.jobj);
+    if (fp->cmd_vars[0] != new_var && fp->victim_gobj != NULL) {
+        fp->cmd_vars[0] = new_var;
+        fp->cmd_vars[1] = new_var;
+        fp2 = GET_FIGHTER(gobj);
+        hat = ft_80459B88.hats[0xD];
+        Fighter_ChangeMotionState(gobj, ftKb_MS_YsSpecialAirN2_0, 0x80012,
+                                  0.0f, 1.0f, 0.0f, NULL);
+        HSD_JObjAddAnimAll(fp2->fv.kb.hat.jobj,
+                           (HSD_AnimJoint*) hat->hat_dynamics[4], NULL, NULL);
+        HSD_JObjReqAnimAll(fp2->fv.kb.hat.jobj, 0.0f);
+        fp2 = GET_FIGHTER(gobj);
+        fp2->death2_cb = ftKb_Init_800EE74C;
+        fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+        ftCommon_8007E2F4(fp, 0x1FF);
+    }
+}
 
 /// #ftKb_YsSpecialNCapture2_0_Anim
 
@@ -635,7 +739,41 @@ void ftKb_SpecialNYs_8010AA2C(Fighter_GObj* gobj)
 
 /// #ftKb_SpecialNYs_8010AC78
 
-/// #ftCo_KirbyYoshiEgg_Anim
+#pragma dont_inline on
+void ftCo_KirbyYoshiEgg_Anim(Fighter_GObj* gobj)
+{
+    f32 temp_ret;
+    Fighter* fp = gobj->user_data;
+    PAD_STACK(16);
+    temp_ret = ftKb_SpecialNYs_801092DC();
+    fp->grab_timer -= temp_ret;
+    fp->mv.kb.specialhi.x4 = ftCommon_GrabMash(fp, ftKb_SpecialNYs_801092F4());
+    if (fp->grab_timer <= 0.0f) {
+        Fighter* fp2;
+        ftCo_DatAttrs_xBC_t* xBCp;
+        ft_PlaySFX(fp, 0x44618, 0x7F, 0x40);
+        fp2 = gobj->user_data;
+        efAsync_Spawn(gobj, &fp2->x60C, 4, 0x4CF, fp->parts[0].joint,
+                      (xBCp = &fp->co_attrs.xBC));
+        ftKb_SpecialNYs_80109354(&fp->self_vel);
+        ftCommon_8007D5D4(fp);
+        ftColl_8007B760(gobj, ftKb_SpecialNYs_8010933C());
+        ftCo_Fall_Enter_YoshiEgg_Kirby(gobj);
+        return;
+    }
+    if (fp->mv.kb.specialhi.x8.f) {
+        fp->mv.kb.specialhi.x8.f--;
+        if (fp->mv.kb.specialhi.x8.f <= 0.0f && fp->mv.kb.specialhi.x4 == 0) {
+            ftAnim_SetAnimRate(gobj, 1.0f);
+            fp->mv.kb.specialhi.x8.f = 0.0f;
+        }
+    }
+    if (fp->mv.kb.specialhi.x8.f <= 0.0f && fp->mv.kb.specialhi.x4 != 0) {
+        fp->mv.kb.specialhi.x8.f = ftKb_SpecialNYs_8010930C();
+        ftAnim_SetAnimRate(gobj, ftKb_SpecialNYs_80109324());
+    }
+}
+#pragma dont_inline reset
 
 void ftCo_KirbyYoshiEgg_IASA(Fighter_GObj* gobj) {}
 
@@ -675,8 +813,7 @@ void fn_8010B148(Fighter_GObj* gobj)
 void fn_8010B16C(Fighter_GObj* gobj)
 {
     Fighter* fp = gobj->user_data;
-    fp->grab_timer -=
-        fp->dmg.x1838_percentTemp * ftKb_SpecialNYs_80109380(gobj);
+    fp->grab_timer -= fp->dmg.x1838_percentTemp * ftKb_SpecialNYs_80109380();
     if ((s32) fp->dmg.x18CC == 3 && ftCo_800C0C88(fp->dmg.x18D0) != 0) {
         fp->grab_timer = 0.0f;
     }
@@ -1120,18 +1257,14 @@ void ftKb_SpecialNPe_8010BF90(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     s32 msid;
-    if (fp->cmd_vars[0] == 0U) {
-        if ((s32) fp->fv.kb.hat.kind == FTKIND_MARS) {
-            msid = ftKb_MS_MsSpecialAirNEnd0;
-        } else {
-            msid = ftKb_MS_FeSpecialAirNEnd0;
-        }
+    PAD_STACK(16);
+    if (fp->cmd_vars[0] == 0) {
+        msid = (fp->fv.kb.hat.kind == FTKIND_MARS) ? ftKb_MS_MsSpecialAirNEnd0
+                                                   : ftKb_MS_FeSpecialAirNEnd0;
     } else {
-        if ((s32) fp->fv.kb.hat.kind == FTKIND_MARS) {
-            msid = ftKb_MS_MsSpecialAirNEnd0 + 1;
-        } else {
-            msid = ftKb_MS_FeSpecialAirNEnd0 + 1;
-        }
+        msid = (fp->fv.kb.hat.kind == FTKIND_MARS)
+                   ? ftKb_MS_MsSpecialAirNEnd0 + 1
+                   : ftKb_MS_FeSpecialAirNEnd0 + 1;
     }
     ftCommon_8007D5D4(fp);
     Fighter_ChangeMotionState(gobj, msid, 0x0C4C708E, fp->cur_anim_frame, 1.0F,
@@ -1146,18 +1279,14 @@ void ftKb_SpecialNPe_8010C06C(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     s32 msid;
-    if (fp->cmd_vars[0] == 0U) {
-        if ((s32) fp->fv.kb.hat.kind == FTKIND_MARS) {
-            msid = ftKb_MS_MsSpecialNEnd0;
-        } else {
-            msid = ftKb_MS_FeSpecialNEnd0;
-        }
+    PAD_STACK(16);
+    if (fp->cmd_vars[0] == 0) {
+        msid = (fp->fv.kb.hat.kind == FTKIND_MARS) ? ftKb_MS_MsSpecialNEnd0
+                                                   : ftKb_MS_FeSpecialNEnd0;
     } else {
-        if ((s32) fp->fv.kb.hat.kind == FTKIND_MARS) {
-            msid = ftKb_MS_MsSpecialNEnd0 + 1;
-        } else {
-            msid = ftKb_MS_FeSpecialNEnd0 + 1;
-        }
+        msid = (fp->fv.kb.hat.kind == FTKIND_MARS)
+                   ? ftKb_MS_MsSpecialNEnd0 + 1
+                   : ftKb_MS_FeSpecialNEnd0 + 1;
     }
     ftCommon_8007D7FC(fp);
     Fighter_ChangeMotionState(gobj, msid, 0x0C4C708E, fp->cur_anim_frame, 1.0F,
@@ -1579,6 +1708,8 @@ void fn_8010CD88(HSD_GObj* gobj)
     ftKb_SpecialNGw_8010CD44(gobj);
 }
 
+/// fn_8010CE5C: not yet matching
+
 void fn_8010CFB0(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
diff --git a/src/melee/ft/chara/ftNana/ftNn_Init.c b/src/melee/ft/chara/ftNana/ftNn_Init.c
index ba0db97cde..18a92c00d3 100644
--- a/src/melee/ft/chara/ftNana/ftNn_Init.c
+++ b/src/melee/ft/chara/ftNana/ftNn_Init.c
@@ -938,6 +938,7 @@ static inline void ftPp_SpecialS_0_Coll_inline3(Fighter_GObj* gobj)
 
 void ftPp_SpecialS_0_Coll(Fighter_GObj* nana_gobj)
 {
+    u8 _[16];
     Fighter* nana_fp = GET_FIGHTER(nana_gobj);
     Fighter_GObj* popo_gobj = Player_GetEntityAtIndex(nana_fp->player_id, 0);
     Fighter* popo_fp = GET_FIGHTER(popo_gobj);
@@ -981,6 +982,7 @@ static inline void ftPp_SpecialS_1_Coll_inline3(Fighter_GObj* gobj)
 
 void ftPp_SpecialS_1_Coll(Fighter_GObj* nana_gobj)
 {
+    u8 _[16];
     Fighter* nana_fp = GET_FIGHTER(nana_gobj);
     Fighter_GObj* popo_gobj = Player_GetEntityAtIndex(nana_fp->player_id, 0);
     Fighter* popo_fp = GET_FIGHTER(popo_gobj);
diff --git a/src/melee/ft/chara/ftNana/ftNn_Init.h b/src/melee/ft/chara/ftNana/ftNn_Init.h
index 95151e22d9..ddb2fe59da 100644
--- a/src/melee/ft/chara/ftNana/ftNn_Init.h
+++ b/src/melee/ft/chara/ftNana/ftNn_Init.h
@@ -10,6 +10,7 @@
 /* 122F28 */ void ftNn_Init_OnDeath(Fighter_GObj* gobj);
 /* 122FAC */ void ftNn_Init_80122FAC(Fighter_GObj* gobj);
 /* 122FCC */ void ftNn_Init_LoadSpecialAttrs(Fighter_GObj* gobj);
+/* 12300C */ bool ftNn_Init_8012300C(Fighter_GObj* gobj);
 /* 12309C */ bool ftNn_Init_8012309C(Fighter_GObj* gobj);
 /* 123348 */ void ftPp_SpecialHi_0_Anim(Fighter_GObj* gobj);
 /* 123384 */ void ftPp_SpecialHi_3_Anim(Fighter_GObj* gobj);
diff --git a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
index 75119a87e9..ca7267f5c0 100644
--- a/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
+++ b/src/melee/ft/chara/ftPopo/ftPp_SpecialS.c
@@ -9,6 +9,7 @@
 #include "ft/forward.h"
 
 #include "ft/ft_081B.h"
+#include "ft/ft_0881.h"
 #include "ft/ft_0892.h"
 #include "ft/ftcliffcommon.h"
 #include "ft/ftcommon.h"
@@ -29,10 +30,14 @@
 #include <math_ppc.h>
 #include <trigf.h>
 #include <dolphin/pad.h>
+#include <melee/ef/efsync.h>
 #include <melee/ft/chara/ftCommon/ftCo_Fall.h>
 #include <melee/ft/chara/ftCommon/ftCo_FallSpecial.h>
 #include <melee/ft/chara/ftPopo/ftPp_SpecialS.h>
+#include <melee/it/itCharItems.h>
+#include <melee/it/items/itclimbersblizzard.h>
 #include <melee/it/items/itclimbersstring.h>
+#include <melee/it/types.h>
 #include <melee/lb/lb_00B0.h>
 #include <melee/lb/lbvector.h>
 
@@ -50,6 +55,25 @@
 /* 1218F8 */ void ftPp_SpecialHi_801218F8(Fighter_GObj* gobj);
 /* 12280C */ void ftPp_SpecialHi_8012280C(Fighter_GObj* gobj);
 
+static inline float my_sqrtf(float x)
+{
+    static const double _half = .5;
+    static const double _three = 3.0;
+
+    u8 _[8] = { 0 };
+
+    volatile float y;
+    if (x > 0) {
+        double guess = __frsqrte((double) x);
+        guess = _half * guess * (_three - guess * guess * x);
+        guess = _half * guess * (_three - guess * guess * x);
+        guess = _half * guess * (_three - guess * guess * x);
+        y = (float) (x * guess);
+        return y;
+    }
+    return x;
+}
+
 static void setRefGObjFlagAndClear(Fighter* fp)
 {
     Fighter_GObj* gobj = fp->x1A5C;
@@ -554,6 +578,18 @@ static ftCollisionBox ftNn_Init_803CD820 = {
     12, 0, -6, 6, 6, 6,
 };
 
+static inline void ftPp_SpecialLw_Coll_inline(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->cmd_vars[3]) {
+        ftPartSetRotX(fp, 0,
+                      fp->facing_dir * atan2f(fp->coll_data.floor.normal.x,
+                                              fp->coll_data.floor.normal.y));
+    } else {
+        ftPartSetRotX(fp, 0, 0.0F);
+    }
+}
+
 static inline void inline2(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
@@ -731,40 +767,75 @@ void ftPp_SpecialAirS2_Coll(Fighter_GObj* gobj)
 
 void ftPp_SpecialS_80120E68(Fighter_GObj* gobj)
 {
-    Fighter* temp_r31;
-    HSD_GObj* temp_r3;
-    f32 temp_f1;
-    f32 temp_f2;
-    f32 temp_f3;
-    ftIceClimberAttributes* temp_r30;
-    Fighter* temp_r29;
-    PAD_STACK(0x8);
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    Fighter_GObj* gobj2 = Player_GetEntityAtIndex(fp->player_id, 1);
+    PAD_STACK(8);
 
-    temp_r31 = GET_FIGHTER(gobj);
-    temp_r30 = temp_r31->dat_attrs;
-    if (Player_GetEntityAtIndex(temp_r31->player_id, 1) != NULL) {
-        temp_r29 = GET_FIGHTER(gobj);
-        temp_r31->self_vel.x = temp_r29->cur_pos.x - temp_r31->cur_pos.x;
-        temp_r31->self_vel.y = temp_r29->cur_pos.y - temp_r31->cur_pos.y;
-        temp_r31->self_vel.z = 0.0F;
-        temp_r31->self_vel.x =
-            -(3.0F * temp_r29->facing_dir - temp_r31->self_vel.x);
-        temp_r31->self_vel.y += 5.0F;
-        lbVector_Normalize(&temp_r31->self_vel);
-        temp_f3 = SQ(temp_r31->cur_pos.x - temp_r29->cur_pos.x);
-        temp_f1 = SQ(temp_r31->cur_pos.y - temp_r29->cur_pos.y);
-        temp_f2 = sqrtf(temp_f3 + temp_f1) / temp_r30->x98;
-        temp_r31->self_vel.x *= temp_r30->x94 + temp_f2;
-        temp_r31->self_vel.y *= temp_r30->x94 + temp_f2;
-        if (temp_r31->self_vel.x > 0.0F) {
-            temp_r31->facing_dir = +1.0F;
+    if (gobj2 != NULL) {
+        Fighter* fp2 = GET_FIGHTER(gobj2);
+        f32 dx, dy;
+        f32 dist;
+        fp->self_vel.x = fp2->cur_pos.x - fp->cur_pos.x;
+        fp->self_vel.y = fp2->cur_pos.y - fp->cur_pos.y;
+        fp->self_vel.z = 0.0F;
+        fp->self_vel.x = -(3.0F * fp2->facing_dir - fp->self_vel.x);
+        fp->self_vel.y += 5.0F;
+        lbVector_Normalize(&fp->self_vel);
+        dx = SQ(fp->cur_pos.x - fp2->cur_pos.x);
+        dy = SQ(fp->cur_pos.y - fp2->cur_pos.y);
+        dist = my_sqrtf(dx + dy) / da->x98;
+        fp->self_vel.x *= da->x94 + dist;
+        fp->self_vel.y *= da->x94 + dist;
+        if (fp->self_vel.x > 0.0F) {
+            fp->facing_dir = +1.0F;
         } else {
-            temp_r31->facing_dir = -1.0F;
+            fp->facing_dir = -1.0F;
         }
     }
 }
 
-/// #ftPp_SpecialS_80120FE0
+bool ftPp_SpecialS_80120FE0(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    s32 cmd = fp->mv.pp.speciallw.x0;
+    PAD_STACK(16);
+
+    if (cmd > 8 && cmd <= 0x53) {
+        Item_GObj* item_gobj;
+        if ((item_gobj = fp->fv.pp.x2238) != NULL) {
+            Item_GObj* gobj = item_gobj;
+            Item* ip = item_gobj->user_data;
+            itClimbersStringAttributes* sa =
+                ip->xC4_article_data->x4_specialAttributes;
+            s32 ev0 = sa->x18;
+            s32 ev1 = sa->x1C;
+            s32 ev2 = sa->x20;
+            if (cmd == ev0) {
+                it_802C3950(gobj);
+            } else if (cmd == ev1) {
+                it_802C3810(gobj);
+            } else if (cmd == ev2) {
+                it_802C3864(gobj);
+            }
+            if (fp->mv.pp.speciallw.x0 == 0x53) {
+                it_802C2750(fp->fv.pp.x2238);
+            }
+        } else {
+            goto end;
+        }
+    } else if (fp->mv.pp.speciallw.x0 == 8) {
+        ftPp_SpecialS_801210C8(gobj);
+        if (fp->fv.pp.x2238 == NULL) {
+            ft_8008A2BC(gobj);
+            return true;
+        }
+    }
+    return false;
+end: {
+    // original code returns without a value
+}
+}
 
 void ftPp_SpecialS_801210C8(Fighter_GObj* arg0)
 {
@@ -843,9 +914,64 @@ void ftPp_SpecialAirHi_Enter(Fighter_GObj* gobj)
     fp->fv.pp.x2240.x = 0.0f;
 }
 
-/// #ftPp_SpecialHiStart_0_Anim
+static inline bool checkNanaInRange(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    Fighter_GObj* nana_gobj = Player_GetEntityAtIndex(fp->player_id, 1);
+    if (nana_gobj != NULL) {
+        Fighter* nana_fp = nana_gobj->user_data;
+        f32 dx = SQ(fp->cur_pos.x - nana_fp->cur_pos.x);
+        f32 dy = SQ(fp->cur_pos.y - nana_fp->cur_pos.y);
+        if (sqrtf__Ff(dx + dy) < da->x7C && ftNn_Init_8012300C(nana_gobj)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static inline void incrementMvAndCheck(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->mv.pp.unk_80123954.x0++;
+    ftPp_SpecialS_80120FE0(gobj);
+}
 
-/// #ftPp_SpecialAirHiStart_0_Anim
+void ftPp_SpecialHiStart_0_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->cmd_vars[2] != 0) {
+        fp->cmd_vars[2] = 0;
+        if (!checkNanaInRange(gobj)) {
+            ftPp_SpecialHi_80122098(gobj);
+            return;
+        }
+        fp->x2222_b2 = true;
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftPp_SpecialHi_80121DA0(gobj);
+    } else {
+        incrementMvAndCheck(gobj);
+    }
+}
+
+void ftPp_SpecialAirHiStart_0_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (fp->cmd_vars[2] != 0) {
+        fp->cmd_vars[2] = 0;
+        if (!checkNanaInRange(gobj)) {
+            ftPp_SpecialHi_80122098(gobj);
+            return;
+        }
+        fp->x2222_b2 = true;
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftPp_SpecialHi_80121DA0(gobj);
+        return;
+    }
+    incrementMvAndCheck(gobj);
+}
 
 void ftPp_SpecialHiStart_0_IASA(Fighter_GObj* gobj)
 {
@@ -967,8 +1093,8 @@ void ftPp_SpecialHi_8012184C(Fighter_GObj* gobj)
     Fighter* fp = GET_FIGHTER(gobj);
 
     ftCommon_8007D7FC(fp);
-    Fighter_ChangeMotionState(gobj, 0x15B, 0x0C4C508AU, fp->cur_anim_frame, 0,
-                              1, NULL);
+    Fighter_ChangeMotionState(gobj, 0x15B, 0x0C4C508AU, fp->cur_anim_frame,
+                              1.0f, 0.0f, NULL);
 }
 
 void ftPp_SpecialHi_801218AC(Fighter_GObj* gobj)
@@ -1025,7 +1151,7 @@ void ftPp_SpecialAirHiThrow_0_Anim(Fighter_GObj* gobj)
 {
     Fighter* fp = gobj->user_data;
     ftIceClimberAttributes* da = fp->dat_attrs;
-    PAD_STACK(40);
+    PAD_STACK(48);
 
     if (!ftAnim_IsFramesRemaining(gobj)) {
         ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
@@ -1038,11 +1164,8 @@ void ftPp_SpecialAirHiThrow_0_Anim(Fighter_GObj* gobj)
             {
                 int found;
                 Fighter_GObj* nn_gobj =
-                    Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id,
-                                            1);
-                if (nn_gobj != NULL &&
-                    ftNn_Init_8012309C(nn_gobj) == 1)
-                {
+                    Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+                if (nn_gobj != NULL && ftNn_Init_8012309C(nn_gobj) == 1) {
                     found = 1;
                 } else {
                     found = 0;
@@ -1092,7 +1215,39 @@ void ftPp_SpecialHiThrow_0_Phys(Fighter_GObj* gobj)
     fp->fv.pp.x2240 = sp;
 }
 
-/// #ftPp_SpecialAirHiThrow_0_Phys
+static inline void doFallPhys(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    ftCommon_Fall(fp, da->x8C, da->x90);
+    ftCommon_8007CEF4(fp);
+}
+
+void ftPp_SpecialAirHiThrow_0_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp;
+    Vec3 sp;
+    PAD_STACK(12);
+
+    doFallPhys(gobj);
+    fp = GET_FIGHTER(gobj);
+    sp.x = sp.y = sp.z = 0.0f;
+
+    {
+        Fighter_GObj* nn_gobj =
+            Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+        if (nn_gobj != NULL) {
+            Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+            if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+                nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+            {
+                lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, &sp);
+            }
+        }
+    }
+
+    fp->fv.pp.x2240 = sp;
+}
 
 void ftPp_SpecialHiThrow_0_Coll(Fighter_GObj* gobj)
 {
@@ -1383,7 +1538,57 @@ void ftPp_SpecialHiThrow2_Phys(Fighter_GObj* gobj)
     fp->fv.pp.x2240 = sp;
 }
 
-/// #ftPp_SpecialAirHiThrow2_Phys
+static inline void doFallPhys2(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    ftCo_DatAttrs* co = &fp->co_attrs;
+    ftCommon_Fall(fp, da->x9C, da->xA0);
+    if (ABS(fp->input.lstick.x) > da->x80) {
+        ftCommon_8007D344(fp, 0.0f, co->air_drift_stick_mul * da->xB0,
+                          co->air_drift_max * da->xB4);
+    } else if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+}
+
+static inline void ftPp_SpecialAirHiThrow2_Phys_inline(Fighter_GObj* gobj,
+                                                       Vec3* sp)
+{
+    Fighter_GObj* nn_gobj =
+        Player_GetEntityAtIndex(GET_FIGHTER(gobj)->player_id, 1);
+    if (nn_gobj != NULL) {
+        Fighter* nn_fp = GET_FIGHTER(nn_gobj);
+        if (nn_fp->motion_id >= ftPp_MS_SpecialHi_0 &&
+            nn_fp->motion_id <= ftPp_MS_SpecialHi_5)
+        {
+            lb_8000B1CC(nn_fp->parts[FtPart_L4thNb].joint, NULL, sp);
+        }
+    }
+}
+
+void ftPp_SpecialAirHiThrow2_Phys(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Vec3 sp;
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    ftCo_DatAttrs* co = &fp->co_attrs;
+    PAD_STACK(12);
+
+    ftCommon_Fall(fp, da->x9C, da->xA0);
+    if (ABS(fp->input.lstick.x) > da->x80) {
+        ftCommon_8007D344(fp, 0.0f, co->air_drift_stick_mul * da->xB0,
+                          co->air_drift_max * da->xB4);
+    } else if (fp->self_vel.y < 0.0f) {
+        ftCommon_8007CEF4(fp);
+    }
+    {
+        fp = GET_FIGHTER(gobj);
+        sp.x = sp.y = sp.z = 0.0f;
+        ftPp_SpecialAirHiThrow2_Phys_inline(gobj, &sp);
+        fp->fv.pp.x2240 = sp;
+    }
+}
 
 void ftPp_SpecialHiThrow2_Coll(Fighter_GObj* gobj)
 {
@@ -1392,7 +1597,25 @@ void ftPp_SpecialHiThrow2_Coll(Fighter_GObj* gobj)
     }
 }
 
-/// #ftPp_SpecialAirHiThrow2_Coll
+void ftPp_SpecialAirHiThrow2_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftIceClimberAttributes* da = fp->dat_attrs;
+    CollData* cd = &fp->coll_data;
+    PAD_STACK(8);
+    if (ft_CheckGroundAndLedge(gobj, (fp->facing_dir < 0) ? -1 : 1)) {
+        ftCo_LandingFallSpecial_Enter(gobj, false, da->x78);
+    } else if (!ftCliffCommon_80081298(gobj)) {
+        if ((cd->env_flags & 0x3F) && fp->self_vel.x > 0.0f) {
+            fp->self_vel.x = 0.0f;
+        } else if ((cd->env_flags & 0xFC0) && fp->self_vel.x < 0.0f) {
+            fp->self_vel.x = 0.0f;
+        } else if (cd->env_flags & 0x6000) {
+            fp->self_vel.y = 0.0f;
+            ftCo_80096900(gobj, 0, 1, false, da->x74, da->x78);
+        }
+    }
+}
 
 void ftPp_SpecialHi_801227AC(Fighter_GObj* gobj)
 {
@@ -1509,6 +1732,13 @@ void ftPp_SpecialAirLw_Phys(Fighter_GObj* gobj)
     ft_80084EEC(gobj);
 }
 
+static inline void ftPp_set_cbs(Fighter_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    fp->death2_cb = ftPp_Init_8011F060;
+    fp->take_dmg_cb = ftPp_Init_8011F060;
+}
+
 void fn_80122B54(Fighter_GObj* gobj)
 {
     Fighter* fp = gobj->user_data;
@@ -1516,11 +1746,7 @@ void fn_80122B54(Fighter_GObj* gobj)
     ftCommon_8007D7FC(fp);
     Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialLw, 0x0C4C5282,
                               fp->cur_anim_frame, 1.0F, 0.0F, NULL);
-    {
-        Fighter* fp2 = gobj->user_data;
-        fp2->death2_cb = ftPp_Init_8011F060;
-        fp2->take_dmg_cb = ftPp_Init_8011F060;
-    }
+    ftPp_set_cbs(gobj);
     fp = gobj->user_data;
     if (fp->cmd_vars[3] != 0) {
         ftPartSetRotX(fp, 0,
@@ -1532,11 +1758,67 @@ void fn_80122B54(Fighter_GObj* gobj)
     fp->accessory4_cb = fn_80122D2C;
 }
 
-/// #ftPp_SpecialLw_Coll
+static inline void ftPp_SpecialLw_Coll_Land(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftCommon_8007D5D4(fp);
+    Fighter_ChangeMotionState(gobj, ftPp_MS_SpecialAirLw, 0x0C4C5282,
+                              fp->cur_anim_frame, 1.0f, 0.0f, NULL);
+}
+
+void ftPp_SpecialLw_Coll(Fighter_GObj* gobj)
+{
+    if (ft_80082708(gobj) == GA_Ground) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        ftPartSetRotX(fp, 0, 0.0f);
+        ftPp_SpecialLw_Coll_Land(gobj);
+        ftPp_set_cbs(gobj);
+        fp->accessory4_cb = fn_80122D2C;
+        ftCommon_ClampAirDrift(fp);
+    } else {
+        ftPp_SpecialLw_Coll_inline(gobj);
+    }
+}
 
 void ftPp_SpecialAirLw_Coll(Fighter_GObj* gobj)
 {
     ft_80082C74(gobj, fn_80122B54);
 }
 
-/// #fn_80122D2C
+void fn_80122D2C(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Vec3 pos;
+    PAD_STACK(8);
+    if (fp->mv.pp.speciallw.x4_b0) {
+        if (fp->mv.pp.speciallw.x0 == 0) {
+            ftIceClimberAttributes* da = fp->dat_attrs;
+            lb_8000B1CC(fp->parts[FtPart_L3rdNa].joint, NULL, &pos);
+            pos.x += da->xBC * fp->facing_dir;
+            pos.y += da->xC0;
+            itClimbersBlizzard_Spawn(gobj, &pos, fp->facing_dir);
+            fp->mv.pp.speciallw.x0 = da->xB8;
+        }
+        fp->mv.pp.speciallw.x0--;
+    }
+    switch (fp->cmd_vars[0]) {
+    case 1:
+        efSync_Spawn(0x4EC, gobj, fp->parts[FtPart_L4thNb].joint);
+        fp->mv.pp.speciallw.x4_b0 = true;
+        fp->fv.pp.x2230_b0 = true;
+        ftPp_set_cbs(gobj);
+        fp->cmd_vars[0] = 0;
+        if (fp->kind == FTKIND_POPO) {
+            ft_800881D8(fp, 0x1FC42, 0x7F, 0x40);
+            ft_80088510(fp, 0x1FBEB, 0x7F, 0x40);
+        } else {
+            ft_800881D8(fp, 0x1FC0F, 0x7F, 0x40);
+            ft_80088510(fp, 0x1FBEE, 0x7F, 0x40);
+        }
+        break;
+    case 2:
+        fp->mv.pp.speciallw.x4_b0 = false;
+        fp->cmd_vars[0] = 0;
+        break;
+    }
+}
diff --git a/src/melee/ft/chara/ftPopo/types.h b/src/melee/ft/chara/ftPopo/types.h
index 38f1c09f2e..a85f8ae174 100644
--- a/src/melee/ft/chara/ftPopo/types.h
+++ b/src/melee/ft/chara/ftPopo/types.h
@@ -51,7 +51,7 @@ typedef struct ftIceClimberAttributes {
     float x70;
     float x74;
     float x78;
-    u8 _7C[0x80 - 0x7C];
+    float x7C;
     float x80;
     float x84;
     float x88;
@@ -59,13 +59,16 @@ typedef struct ftIceClimberAttributes {
     float x90;
     float x94;
     float x98;
-    u8 _9C[0xA4 - 0x9C];
+    float x9C;
+    float xA0;
     float xA4;
     float xA8;
     float xAC;
     float xB0;
     float xB4;
-    u8 _B8[0xC4 - 0xB8];
+    float xB8;
+    float xBC;
+    float xC0;
     float xC4;
     float xC8;
     u8 _CC[0xD0 - 0xCC];
diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index c6d5f49aa5..97141c8b3b 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -32,6 +32,15 @@ typedef struct itClimbersIce_ItemVars {
     /* +8 */ u8 x8_b0 : 1;
 } itClimbersIce_ItemVars;
 
+typedef struct itClimbersStringAttributes {
+    /* +00 */ u8 pad_00[0x18];
+    /* +18 */ s32 x18;
+    /* +1C */ s32 x1C;
+    /* +20 */ s32 x20;
+    /* +24 */ void* x24_joint;
+    /* +28 */ void* x28_joint;
+} itClimbersStringAttributes;
+
 typedef struct itClimbersIceAttributes {
     /* +00 */ u8 pad_00[0x4];
     /* +04 */ f32 x4;
diff --git a/src/melee/it/items/itclimbersstring.h b/src/melee/it/items/itclimbersstring.h
index 6d18c88b63..d19a1f7b9c 100644
--- a/src/melee/it/items/itclimbersstring.h
+++ b/src/melee/it/items/itclimbersstring.h
@@ -10,6 +10,8 @@
 /* 2C28B8 */ void fn_802C28B8(Item_GObj* gobj);
 /* 2C37BC */ void it_2725_Logic70_PickedUp(Item_GObj* gobj);
 /* 2C3810 */ void it_802C3810(Item_GObj* gobj);
+/* 2C3864 */ void it_802C3864(Item_GObj* gobj);
+/* 2C3950 */ void it_802C3950(Item_GObj* gobj);
 /* 2C3AA4 */ void it_2725_Logic70_EvtUnk(Item_GObj*, Item_GObj*);
 /* 3F76B8 */ extern ItemStateTable it_803F76B8[];
 
