diff --git a/configure.py b/configure.py
index cc0d43e802..e875609002 100755
--- a/configure.py
+++ b/configure.py
@@ -1152,7 +1152,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itevyoshiegg.c"),
             # Fighter-related items
             Object(Matching, "melee/it/items/itmariofireball.c"),
-            Object(NonMatching, "melee/it/items/itkirbycutterbeam.c"),
+            Object(Matching, "melee/it/items/itkirbycutterbeam.c"),
             Object(Matching, "melee/it/items/itfoxlaser.c"),
             Object(Matching, "melee/it/items/itfoxillusion.c"),
             Object(NonMatching, "melee/it/items/itlinkbomb.c"),
diff --git a/src/melee/it/itCharItems.h b/src/melee/it/itCharItems.h
index bc94ed2c4a..3cd3b3672e 100644
--- a/src/melee/it/itCharItems.h
+++ b/src/melee/it/itCharItems.h
@@ -37,7 +37,10 @@ typedef struct itClimbersIce_ItemVars {
 } itClimbersIce_ItemVars;
 
 typedef struct itClimbersStringAttributes {
-    /* +00 */ u8 pad_00[0x18];
+    /* +00 */ u8 pad_00[0x8];
+    /* +08 */ f32 x8;
+    /* +0C */ u8 pad_0C[0x8];
+    /* +14 */ f32 x14;
     /* +18 */ s32 x18;
     /* +1C */ s32 x1C;
     /* +20 */ s32 x20;
diff --git a/src/melee/it/items/itclimbersstring.c b/src/melee/it/items/itclimbersstring.c
index 2545752e22..68eb60d1a6 100644
--- a/src/melee/it/items/itclimbersstring.c
+++ b/src/melee/it/items/itclimbersstring.c
@@ -3,22 +3,18 @@
 #include "placeholder.h"
 
 #include "ft/chara/ftPopo/ftPp_SpecialS.h"
+#include "ft/ftlib.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/itCharItems.h"
 #include "it/item.h"
+#include "it/items/itlinkhookshot.h"
 
 #include <dolphin/mtx.h>
 #include <dolphin/os/OSError.h>
 #include <baselib/gobjplink.h>
 #include <baselib/jobj.h>
 
-/* 2C248C */ static void it_802C248C(Item* ip, HSD_JObj* jobj);
-/* 2C28DC */ static void fn_802C28DC(Item_GObj* gobj);
-/* 2C29E8 */ static void fn_802C29E8(Item_GObj* gobj);
-/* 2C2AF4 */ static void fn_802C2AF4(HSD_GObj* gobj);
-/* 2C33B8 */ static void it_802C33B8(Item* item);
-
 void it_802C2750(Item_GObj* gobj)
 {
     int unused[2];
@@ -75,6 +71,201 @@ void fn_802C28B8(Item_GObj* gobj)
     it_802C33B8(GET_ITEM(gobj));
 }
 
+static void fn_802C28DC(Item_GObj* gobj)
+{
+    Vec3 pos;
+    f32 pad[1];
+    Mtx m;
+    Item* ip = GET_ITEM(gobj);
+    itClimbersStringAttributes* attrs =
+        ip->xC4_article_data->x4_specialAttributes;
+    ItemLink* link = ip->xDD4_itemVar.climbersstring.x4;
+
+    link->x2C_b0 = true;
+    PSMTXIdentity(m);
+    m[0][3] = 0.0f;
+    m[1][3] = 0.0f;
+    m[2][3] = 0.0f;
+    HSD_JObjSetupMatrix(link->jobj);
+    PSMTXConcat(link->jobj->mtx, m, m);
+    pos.x = m[0][3];
+    pos.y = m[1][3];
+    pos.z = m[2][3];
+    it_802C2EC4(link, &pos, attrs, ip);
+    it_802C3520(ip, &pos);
+}
+
+static void fn_802C29E8(Item_GObj* gobj)
+{
+    Vec3 pos;
+    f32 pad[1];
+    Mtx m;
+    Item* ip = GET_ITEM(gobj);
+    itClimbersStringAttributes* attrs =
+        ip->xC4_article_data->x4_specialAttributes;
+    ItemLink* link = ip->xDD4_itemVar.climbersstring.x4;
+
+    link->x2C_b0 = true;
+    PSMTXIdentity(m);
+    m[0][3] = 0.0f;
+    m[1][3] = 0.0f;
+    m[2][3] = 0.0f;
+    HSD_JObjSetupMatrix(link->jobj);
+    PSMTXConcat(link->jobj->mtx, m, m);
+    pos.x = m[0][3];
+    pos.y = m[1][3];
+    pos.z = m[2][3];
+    it_802C30E8(link, &pos, attrs, ip);
+    it_802C3520(ip, &pos);
+}
+
+static void fn_802C2AF4(HSD_GObj* gobj)
+{
+    Vec3 pos;
+    Mtx m;
+    Item* ip = GET_ITEM(gobj);
+    itClimbersStringAttributes* attrs =
+        ip->xC4_article_data->x4_specialAttributes;
+    ItemLink* link = ip->xDD4_itemVar.climbersstring.x8;
+
+    PSMTXIdentity(m);
+    m[0][3] = 0.0f;
+    m[1][3] = 0.0f;
+    m[2][3] = 0.0f;
+    HSD_JObjSetupMatrix(link->jobj);
+    PSMTXConcat(link->jobj->mtx, m, m);
+    pos.x = m[0][3];
+    pos.y = m[1][3];
+    pos.z = m[2][3];
+    if (it_802C32D4(link, &pos, attrs, ip,
+                    ip->xDD4_itemVar.climbersstring.x0) != 0)
+    {
+        it_2725_Logic70_PickedUp(gobj);
+    }
+    it_802C3520(ip, &pos);
+}
+
+void it_802C2CA8(ItemLink* link, Vec3* target,
+                 itClimbersStringAttributes* attrs, f32 length)
+{
+    Vec3 dir;
+    ItemLink* cur = link;
+    ItemLink* prev = link->prev;
+
+    it_802A3C98(&cur->pos, target, &dir);
+    cur->pos.x = (dir.x * length) + target->x;
+    cur->pos.y = (dir.y * length) + target->y;
+    cur->pos.z = (dir.z * length) + target->z;
+    while (prev != NULL) {
+        if (prev->x2C_b0) {
+            if (it_802A3C98(&prev->pos, &cur->pos, &dir) > attrs->x8) {
+                prev->pos.x = (dir.x * attrs->x8) + cur->pos.x;
+                prev->pos.y = (dir.y * attrs->x8) + cur->pos.y;
+                prev->pos.z = (dir.z * attrs->x8) + cur->pos.z;
+            }
+        }
+        cur = prev;
+        prev = prev->prev;
+    }
+}
+
+void it_802C2DB0(ItemLink* cur, Vec3* target,
+                 itClimbersStringAttributes* attrs, f32 length)
+{
+    Vec3 dir;
+    ItemLink* prev = cur->prev;
+
+    it_802A3C98(&cur->pos, target, &dir);
+    cur->pos.x = (dir.x * length) + target->x;
+    cur->pos.y = (dir.y * length) + target->y;
+    cur->pos.z = (dir.z * length) + target->z;
+    while (prev != NULL) {
+        prev->vel.y -= attrs->x14;
+        it_802A4420(prev);
+        if (it_802A3C98(&prev->pos, &cur->pos, &dir) > attrs->x8) {
+            prev->pos.x = (dir.x * attrs->x8) + cur->pos.x;
+            prev->pos.y = (dir.y * attrs->x8) + cur->pos.y;
+            prev->pos.z = (dir.z * attrs->x8) + cur->pos.z;
+        }
+        cur = prev;
+        prev = prev->prev;
+    }
+}
+
+s32 it_802C32D4(ItemLink* cur, Vec3* pos, itClimbersStringAttributes* attrs,
+                Item* ip, f32 dist)
+{
+    Vec3 dir;
+    ItemLink* prev = cur->prev;
+    f32 len;
+    f32 step;
+
+    while (prev != NULL && !cur->x2C_b0) {
+        cur = prev;
+        prev = prev->prev;
+    }
+    len = it_802A3C98(&cur->pos, pos, &dir);
+    while (prev != NULL && dist > len) {
+        cur->x2C_b0 = false;
+        len = it_802A3C98(&prev->pos, pos, &dir);
+        cur = prev;
+        prev = prev->prev;
+    }
+    step = len - dist;
+    if (step > attrs->x8) {
+        step = attrs->x8;
+    }
+    it_802C2DB0(cur, pos, attrs, step);
+    if (prev != NULL) {
+        return 0;
+    }
+    return 1;
+}
+
+static inline void itClimbersstring_Cleanup(Item_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Item* ip = GET_ITEM(gobj);
+        if (ip->owner != NULL) {
+            ftPp_SpecialS_8012114C(ip->owner);
+        }
+        ip->owner = NULL;
+        ip->xDD4_itemVar.climbersstring.xC = NULL;
+        {
+            ItemLink* cur;
+            for (cur = ip->xDD4_itemVar.climbersstring.x4; cur != NULL;) {
+                HSD_GObj* cur_gobj = cur->gobj;
+                cur = cur->next;
+                HSD_GObjPLink_80390228(cur_gobj);
+            }
+        }
+    }
+}
+
+bool itClimbersstring_UnkMotion3_Anim(Item_GObj* gobj)
+{
+    s32 should_cleanup;
+    Item* ip = GET_ITEM(gobj);
+    HSD_GObj* owner = ip->owner;
+
+    if (owner != NULL) {
+        enum_t action = ftLib_80086C0C(owner);
+        if (action >= 0x15B && action <= 0x164) {
+            should_cleanup = 0;
+        } else {
+            should_cleanup = 1;
+        }
+    } else {
+        should_cleanup = 1;
+    }
+    if (should_cleanup != 0) {
+        itClimbersstring_Cleanup(gobj);
+        return true;
+    }
+    return false;
+    PAD_STACK(8);
+}
+
 void it_2725_Logic70_EvtUnk(Item_GObj* arg0, Item_GObj* arg1)
 {
     Item* ip = GET_ITEM((HSD_GObj*) arg0);
diff --git a/src/melee/it/items/itclimbersstring.h b/src/melee/it/items/itclimbersstring.h
index d19a1f7b9c..fd3aa3ba8b 100644
--- a/src/melee/it/items/itclimbersstring.h
+++ b/src/melee/it/items/itclimbersstring.h
@@ -3,12 +3,30 @@
 
 #include "it/forward.h"
 
+#include "it/itCharItems.h"
 #include "it/items/types.h"
 
+/* 2C248C */ static void it_802C248C(Item* ip, HSD_JObj* jobj);
+/* 2C28DC */ static void fn_802C28DC(Item_GObj* gobj);
+/* 2C29E8 */ static void fn_802C29E8(Item_GObj* gobj);
+/* 2C2AF4 */ static void fn_802C2AF4(HSD_GObj* gobj);
+/* 2C2DB0 */ void it_802C2DB0(ItemLink*, Vec3*, itClimbersStringAttributes*,
+                              f32);
+/* 2C2EC4 */ void it_802C2EC4(ItemLink*, Vec3*, itClimbersStringAttributes*,
+                              Item*);
+/* 2C30E8 */ void it_802C30E8(ItemLink*, Vec3*, itClimbersStringAttributes*,
+                              Item*);
+/* 2C32D4 */ s32 it_802C32D4(ItemLink*, Vec3*, itClimbersStringAttributes*,
+                             Item*, f32);
+/* 2C3520 */ void it_802C3520(Item*, Vec3*);
+/* 2C37BC */ void it_2725_Logic70_PickedUp(Item_GObj* gobj);
+/* 2C33B8 */ static void it_802C33B8(Item* item);
 /* 2C2750 */ void it_802C2750(Item_GObj*);
 /* 2C27D4 */ Item_GObj* it_802C27D4(Fighter_GObj*, Vec3*, int msid, float dir);
 /* 2C28B8 */ void fn_802C28B8(Item_GObj* gobj);
-/* 2C37BC */ void it_2725_Logic70_PickedUp(Item_GObj* gobj);
+/* 2C2BF0 */ bool itClimbersstring_UnkMotion3_Anim(Item_GObj* gobj);
+/* 2C2CA8 */ void it_802C2CA8(ItemLink* link, Vec3* target,
+                              itClimbersStringAttributes* attrs, f32 length);
 /* 2C3810 */ void it_802C3810(Item_GObj* gobj);
 /* 2C3864 */ void it_802C3864(Item_GObj* gobj);
 /* 2C3950 */ void it_802C3950(Item_GObj* gobj);
diff --git a/src/melee/it/items/itkirbycutterbeam.c b/src/melee/it/items/itkirbycutterbeam.c
index debad47636..5256c81dcb 100644
--- a/src/melee/it/items/itkirbycutterbeam.c
+++ b/src/melee/it/items/itkirbycutterbeam.c
@@ -1,25 +1,24 @@
 #include "itkirbycutterbeam.h"
 
-#include <placeholder.h>
-#include <platform.h>
-
 #include "db/db.h"
 #include "ft/chara/ftKirby/ftKb_Init.h"
-
-#include "it/forward.h"
-
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
-#include "it/types.h"
 #include "lb/lbvector.h"
 
 #include <math.h>
-#include <baselib/jobj.h>
 #include <MSL/trigf.h>
 
+ItemStateTable it_803F6798[] = {
+    NULL,
+    itKirbycutterbeam_UnkMotion0_Anim,
+    itKirbycutterbeam_UnkMotion0_Phys,
+    itKirbycutterbeam_UnkMotion0_Coll,
+};
+
 void it_8029BAB8(HSD_GObj* gobj, Vec3* pos, float facing_dir)
 {
     Item_GObj* spawned_gobj;
diff --git a/src/melee/it/items/itnessyoyo.c b/src/melee/it/items/itnessyoyo.c
index 8d0482f15c..65d1a131c7 100644
--- a/src/melee/it/items/itnessyoyo.c
+++ b/src/melee/it/items/itnessyoyo.c
@@ -3,14 +3,17 @@
 #include <placeholder.h>
 
 #include "ft/chara/ftNess/ftNs_AttackHi4.h"
-#include "ft/fighter.h"
+#include "ft/ftlib.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/item.h"
+#include "it/items/itlinkhookshot.h"
 #include "it/itYoyo.h"
 #include "lb/lb_00B0.h"
+#include "mp/mpcoll.h"
 
 #include <baselib/gobjplink.h>
+#include <MSL/math_ppc.h>
 
 void it_802BE598(Item_GObj* gobj)
 {
@@ -67,11 +70,51 @@ void itNessyoyo_UnkMotion0_Phys(Item_GObj* gobj)
     it_802BF900(GET_ITEM(gobj));
 }
 
-/// #itNessyoyo_UnkMotion1_Phys
+extern const f32 it_804DD150;
 
-/// #itNessyoyo_UnkMotion2_Phys
+void itNessyoyo_UnkMotion1_Phys(Item_GObj* gobj)
+{
+    Vec3 pos;
+    f32 pad[1];
+    Mtx m;
+    Item* ip = GET_ITEM(gobj);
+    itYoyoAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    ItemLink* link1 = (ItemLink*) ip->xDD4_itemVar.samusgrapple.x8;
 
-extern const f32 it_804DD150;
+    PSMTXIdentity(m);
+    m[0][3] = it_804DD150;
+    m[1][3] = it_804DD150;
+    m[2][3] = it_804DD150;
+    HSD_JObjSetupMatrix(link1->jobj);
+    PSMTXConcat(link1->jobj->mtx, m, m);
+    pos.x = m[0][3];
+    pos.y = m[1][3];
+    pos.z = m[2][3];
+    it_802BF28C(link1, &pos, attrs, ip);
+    it_802BFAFC(ip, &pos);
+}
+
+void itNessyoyo_UnkMotion2_Phys(Item_GObj* gobj)
+{
+    Vec3 pos;
+    f32 pad[1];
+    Mtx m;
+    Item* ip = GET_ITEM(gobj);
+    itYoyoAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    ItemLink* link1 = (ItemLink*) ip->xDD4_itemVar.samusgrapple.x8;
+
+    PSMTXIdentity(m);
+    m[0][3] = it_804DD150;
+    m[1][3] = it_804DD150;
+    m[2][3] = it_804DD150;
+    HSD_JObjSetupMatrix(link1->jobj);
+    PSMTXConcat(link1->jobj->mtx, m, m);
+    pos.x = m[0][3];
+    pos.y = m[1][3];
+    pos.z = m[2][3];
+    it_802BF4A0(link1, &pos, attrs, ip);
+    it_802BFAFC(ip, &pos);
+}
 
 void itNessyoyo_UnkMotion3_Phys(Item_GObj* gobj)
 {
@@ -100,10 +143,8 @@ void itNessyoyo_UnkMotion3_Phys(Item_GObj* gobj)
         zero_vec.x = it_804DD150;
         it_802C0010(gobj, &zero_vec);
         {
-            HSD_GObj* owner =
-                (HSD_GObj*) ip->xDD4_itemVar.samusgrapple.unk_10;
-            ItemLink* link1 =
-                (ItemLink*) ip->xDD4_itemVar.samusgrapple.x8;
+            HSD_GObj* owner = (HSD_GObj*) ip->xDD4_itemVar.samusgrapple.unk_10;
+            ItemLink* link1 = (ItemLink*) ip->xDD4_itemVar.samusgrapple.x8;
             Fighter* fp = owner->user_data;
             link1->pos = fp->fv.ns.yoyo_hitbox_pos;
         }
@@ -111,19 +152,184 @@ void itNessyoyo_UnkMotion3_Phys(Item_GObj* gobj)
     it_802BFAFC(ip, &pos);
 }
 
-/// #itNessyoyo_UnkMotion3_Anim
+static inline void itNessyoyo_UnkMotion3_Anim_Cleanup(Item_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Item* ip = GET_ITEM(gobj);
+        if (ip->owner != NULL) {
+            ftNs_AttackHi4_YoyoItemSetFlag(ip->owner);
+        }
+        ip->owner = NULL;
+        {
+            ItemLink* link = (ItemLink*) ip->xDD4_itemVar.samusgrapple.x8;
+            while (link != NULL) {
+                HSD_GObj* tmp = link->gobj;
+                link = link->next;
+                HSD_GObjPLink_80390228(tmp);
+            }
+        }
+    }
+}
+
+bool itNessyoyo_UnkMotion3_Anim(Item_GObj* gobj)
+{
+    s32 should_cleanup;
+    Item* ip = GET_ITEM(gobj);
+    HSD_JObj* jobj = ip->xDD4_itemVar.flipper.xDEC;
+    Fighter* fp =
+        ((HSD_GObj*) ip->xDD4_itemVar.samusgrapple.unk_10)->user_data;
+    HSD_JObj* child = HSD_JObjGetChild(jobj);
+
+    if (child != NULL) {
+        f32 rot = HSD_JObjGetRotationX(child) + fp->fv.ns.x223C;
+        HSD_JObjSetRotationX(child, rot);
+    }
+    {
+        HSD_GObj* owner = GET_ITEM(gobj)->owner;
+        if (owner != NULL) {
+            enum_t action = ftLib_80086C0C(owner);
+            if (action >= 0x156 && action <= 0x15B) {
+                should_cleanup = 0;
+            } else {
+                should_cleanup = 1;
+            }
+        } else {
+            should_cleanup = 1;
+        }
+    }
+    if (should_cleanup != 0) {
+        itNessyoyo_UnkMotion3_Anim_Cleanup(gobj);
+        return true;
+    }
+    return false;
+}
+
+s32 it_802BF030(ItemLink* link, s32 arg1, f32 offset)
+{
+    CollData* coll = &link->coll_data;
 
-/// #it_802BF030
+    link->coll_data.last_pos = link->coll_data.cur_pos;
+    link->coll_data.cur_pos = link->pos;
+    if (mpColl_800471F8(coll)) {
+        if (!link->x2C_b1 && arg1 != 0 && !link->x2C_b2) {
+            link->x2C_b2 = true;
+        }
+        link->x2C_b1 = true;
+    } else {
+        link->x2C_b1 = false;
+    }
+    {
+        s32 env = coll->env_flags;
+        if (env & 0x18000) {
+            coll->cur_pos.x = link->pos.x;
+        } else if (env & 0xFFF) {
+            if (link->next != NULL && link->next->x2C_b0 && link->prev != NULL)
+            {
+                coll->cur_pos.y += offset;
+            }
+        }
+    }
+    link->pos = coll->cur_pos;
+    return coll->env_flags & 0x18FFF;
+}
 
-/// #it_802BF180
+void it_802BF180(ItemLink* cur, Vec3* target, itYoyoAttributes* attrs,
+                 f32 length)
+{
+    f32 pad_top[4];
+    Vec3 dir;
+    f32 pad_bottom[1];
+    ItemLink* prev = cur->prev;
+
+    it_802A3C98(&cur->pos, target, &dir);
+    cur->pos.x = (dir.x * length) + target->x;
+    cur->pos.y = (dir.y * length) + target->y;
+    cur->pos.z = (dir.z * length) + target->z;
+    while (prev != NULL) {
+        prev->vel.y = 0.0f;
+        it_802A4420(prev);
+        it_802A43EC(prev);
+        it_802A3D90(prev);
+        if (it_802A3C98(&prev->pos, &cur->pos, &dir) > length) {
+            prev->pos.x = (dir.x * length) + cur->pos.x;
+            prev->pos.y = (dir.y * length) + cur->pos.y;
+            prev->pos.z = (dir.z * length) + cur->pos.z;
+        }
+        cur = prev;
+        prev = prev->prev;
+    }
+}
 
 /// #it_802BF28C
 
 /// #it_802BF4A0
 
-/// #it_802BF800
+s32 it_802BF800(ItemLink* cur, Vec3* pos, itYoyoAttributes* attrs, Item* ip,
+                f32 dist)
+{
+    Vec3 dir;
+    ItemLink* prev;
+    f32 size = attrs->xC_SIZE * ftLib_80086A0C(ip->owner);
+    f32 len;
+    f32 step;
+
+    prev = cur->prev;
 
-/// #it_802BF900
+    while (prev != NULL && !cur->x2C_b0) {
+        cur = prev;
+        prev = prev->prev;
+    }
+    len = it_802A3C98(&cur->pos, pos, &dir);
+    while (prev != NULL && dist > len) {
+        cur->x2C_b0 = false;
+        len = it_802A3C98(&prev->pos, pos, &dir);
+        cur = prev;
+        prev = prev->prev;
+    }
+    step = len - dist;
+    if (step > size) {
+        step = size;
+    }
+    it_802BF180(cur, pos, attrs, step);
+    if (prev != NULL) {
+        return 0;
+    }
+    return 1;
+}
+
+extern const Vec3 it_803B86A4;
+
+void it_802BF900(Item* ip)
+{
+    Mtx trans;
+    Mtx scale;
+    ItemLink* link1 = (ItemLink*) ip->xDD4_itemVar.samusgrapple.x8;
+    HSD_JObj* target = (HSD_JObj*) link1->gobj->hsd_obj;
+    Vec3 offset = it_803B86A4;
+    f32 scl = ftLib_80086A0C(ip->owner);
+
+    HSD_JObjSetupMatrix(link1->jobj);
+    PSMTXIdentity(trans);
+    trans[0][3] = offset.x * scl;
+    trans[1][3] = offset.y * scl;
+    trans[2][3] = offset.z * scl;
+    PSMTXConcat(link1->jobj->mtx, trans, trans);
+    PSMTXIdentity(scale);
+    scale[0][0] *= scl;
+    scale[1][0] *= scl;
+    scale[2][0] *= scl;
+    scale[0][1] *= scl;
+    scale[1][1] *= scl;
+    scale[2][1] *= scl;
+    scale[0][2] *= scl;
+    scale[1][2] *= scl;
+    scale[2][2] *= scl;
+    PSMTXConcat(trans, scale, trans);
+    HSD_JObjCopyMtx(target, trans);
+    target->flags |=
+        JOBJ_USER_DEF_MTX | JOBJ_MTX_INDEP_PARENT | JOBJ_MTX_INDEP_SRT;
+    HSD_JObjSetMtxDirty(target);
+}
 
 /// #it_802BFAFC
 
@@ -146,9 +352,57 @@ void it_802BFE5C(Item_GObj* gobj, Vec3* pos, float unused)
     Item_80268E5C(gobj, 2, ITEM_ANIM_UPDATE);
 }
 
-/// #it_802BFEC4
+void it_802BFEC4(Item_GObj* gobj)
+{
+    f32 pad_top[3];
+    Mtx m;
+    f32 pad_bottom[3];
+    Item* ip = GET_ITEM(gobj);
+    ItemLink* link2 = (ItemLink*) ip->xDD4_itemVar.samusgrapple.xC;
+    ItemLink* link1 = (ItemLink*) ip->xDD4_itemVar.samusgrapple.x8;
+
+    PSMTXIdentity(m);
+    m[0][3] = it_804DD150;
+    m[1][3] = it_804DD150;
+    m[2][3] = it_804DD150;
+    HSD_JObjSetupMatrix(link2->jobj);
+    PSMTXConcat(link2->jobj->mtx, m, m);
+    {
+        f32 dx = link1->pos.x - m[0][3];
+        f32 dy = link1->pos.y - m[1][3];
+        f32 dx2 = dx * dx;
+        f32 dy2 = dy * dy;
+        ip->xDD4_itemVar.foxillusion.xDD8 = sqrtf(dx2 + dy2);
+        ip->xDD4_itemVar.foxillusion.xDD8 *= 0.1f;
+    }
+    Item_80268E5C(gobj, 3, ITEM_ANIM_UPDATE);
+}
 
-/// #it_802C0010
+void it_802C0010(Item_GObj* gobj, Vec3* vel)
+{
+    f32 pad_top[2];
+    Vec3 pos;
+    f32 pad_mid[2];
+    Mtx m;
+    f32 pad_bottom[2];
+    ItemLink* link1 = (ItemLink*) GET_ITEM(gobj)->xDD4_itemVar.samusgrapple.x8;
+
+    PSMTXIdentity(m);
+    m[0][3] = it_804DD150;
+    m[1][3] = it_804DD150;
+    m[2][3] = it_804DD150;
+    HSD_JObjSetupMatrix(link1->jobj);
+    PSMTXConcat(link1->jobj->mtx, m, m);
+    pos.x = m[0][3];
+    pos.y = m[1][3];
+    pos.z = m[2][3];
+    link1->pos = pos;
+    link1->coll_data.cur_pos = link1->pos;
+    link1->coll_data.last_pos = link1->coll_data.cur_pos;
+    link1->x2C_b0 = true;
+    link1->vel = *vel;
+    Item_80268E5C(gobj, 1, ITEM_ANIM_UPDATE);
+}
 
 #ifdef MWERKS_GEKKO
 
diff --git a/src/melee/it/items/itnessyoyo.h b/src/melee/it/items/itnessyoyo.h
index d4b00de8b3..6f504b95e6 100644
--- a/src/melee/it/items/itnessyoyo.h
+++ b/src/melee/it/items/itnessyoyo.h
@@ -20,17 +20,17 @@
 /* 2BEC58 */ void itNessyoyo_UnkMotion2_Phys(Item_GObj* gobj);
 /* 2BED54 */ void itNessyoyo_UnkMotion3_Phys(Item_GObj* gobj);
 /* 2BEE88 */ bool itNessyoyo_UnkMotion3_Anim(Item_GObj* gobj);
-/* 2BF030 */ UNK_RET it_802BF030(UNK_PARAMS);
-/* 2BF180 */ UNK_RET it_802BF180(UNK_PARAMS);
-/* 2BF28C */ UNK_RET it_802BF28C(UNK_PARAMS);
+/* 2BF030 */ s32 it_802BF030(ItemLink*, s32, f32);
+/* 2BF180 */ void it_802BF180(ItemLink*, Vec3*, itYoyoAttributes*, f32);
+/* 2BF28C */ void it_802BF28C(ItemLink*, Vec3*, itYoyoAttributes*, Item*);
 /* 2BF4A0 */ void it_802BF4A0(ItemLink*, Vec3*, itYoyoAttributes*, Item*);
 /* 2BF800 */ s32 it_802BF800(ItemLink*, Vec3*, itYoyoAttributes*, Item*, f32);
 /* 2BF900 */ void it_802BF900(Item* item);
 /* 2BFAFC */ void it_802BFAFC(Item*, Vec3*);
 /* 2BFE34 */ void itNessYoyo_Logic59_PickedUp(Item_GObj*);
 /* 2BFE5C */ void it_802BFE5C(Item_GObj*, Vec3*, float);
-/* 2BFEC4 */ void it_802BFEC4(void*);
-/* 2C0010 */ void it_802C0010(Item_GObj*, void*);
+/* 2BFEC4 */ void it_802BFEC4(Item_GObj*);
+/* 2C0010 */ void it_802C0010(Item_GObj*, Vec3*);
 /* 2C016C */ void it_2725_Logic59_EvtUnk(Item_GObj*, Item_GObj*);
 /* 3F7558 */ extern ItemStateTable it_803F7558[];
 
