diff --git a/src/melee/ft/ft_0C31.c b/src/melee/ft/ft_0C31.c
index 50f1846644..3952bca503 100644
--- a/src/melee/ft/ft_0C31.c
+++ b/src/melee/ft/ft_0C31.c
@@ -652,7 +652,8 @@ void fn_800C7AE0(Fighter_GObj* gobj)
     it_8026B294(fp->mv.co.captureleadead.x0, &fp->cur_pos);
 }
 
-UNK_RET ftCo_800C7B0C(Fighter_GObj* arg0, void* arg1, void* arg2, lbColl_80008D30_arg1* arg3, f32 farg0)
+UNK_RET ftCo_800C7B0C(Fighter_GObj* arg0, void* arg1, void* arg2,
+                      lbColl_80008D30_arg1* arg3, f32 farg0)
 {
     NOT_IMPLEMENTED;
 }
diff --git a/src/melee/ft/ft_0C31.h b/src/melee/ft/ft_0C31.h
index 5201f3429c..5ff4346a46 100644
--- a/src/melee/ft/ft_0C31.h
+++ b/src/melee/ft/ft_0C31.h
@@ -53,7 +53,8 @@
 /* 0C7AD8 */ void ftCo_CaptureLikelike_Phys(Fighter_GObj* gobj);
 /* 0C7ADC */ void ftCo_CaptureLikelike_Coll(Fighter_GObj* gobj);
 /* 0C7AE0 */ void fn_800C7AE0(Fighter_GObj* gobj);
-/* 0C7B0C */ UNK_RET ftCo_800C7B0C(Fighter_GObj* arg0, void* arg1, void* arg2, lbColl_80008D30_arg1* arg3, f32 farg0);
+/* 0C7B0C */ UNK_RET ftCo_800C7B0C(Fighter_GObj* arg0, void* arg1, void* arg2,
+                                   lbColl_80008D30_arg1* arg3, f32 farg0);
 /* 0C7C60 */ void ftCo_800C7C60(Fighter_GObj* gobj, int damage_amount);
 /* 0C7CA0 */ bool ftCo_800C7CA0(Fighter_GObj* gobj);
 /* 0C7FC8 */ void ftCo_DownReflect_Anim(Fighter_GObj* gobj);
diff --git a/src/melee/gr/grinishie2.c b/src/melee/gr/grinishie2.c
index 6d904c5164..adfd7e2f3a 100644
--- a/src/melee/gr/grinishie2.c
+++ b/src/melee/gr/grinishie2.c
@@ -26,20 +26,9 @@
 
 #include <baselib/jobj.h>
 S16Vec3 grI2_803E4A60[] = {
-    {  0,  1,  1 },
-    { 12,  3,  1 },
-    { 13,  4,  1 },
-    { 14, 15,  0 },
-    {  3,  5,  0 },
-    {  4,  6,  0 },
-    {  5,  7,  0 },
-    {  6,  8,  0 },
-    {  7,  9,  0 },
-    {  8, 10,  0 },
-    {  9, 11,  0 },
-    { 10, 12,  0 },
-    { 11, 13,  0 },
-    {  2, 14,  0 }
+    { 0, 1, 1 },  { 12, 3, 1 },  { 13, 4, 1 },  { 14, 15, 0 }, { 3, 5, 0 },
+    { 4, 6, 0 },  { 5, 7, 0 },   { 6, 8, 0 },   { 7, 9, 0 },   { 8, 10, 0 },
+    { 9, 11, 0 }, { 10, 12, 0 }, { 11, 13, 0 }, { 2, 14, 0 },
 };
 StageCallbacks grI2_803E4AB4[] = {
     {
@@ -693,8 +682,7 @@ void grInishie2_801FD9EC(HSD_GObj* gobj)
                         gp->gv.inishie23.xC8_flags.b3 = (var_r3 == 2);
                         gp->gv.inishie23.xC8_flags.b0 = 1;
                     } else {
-                        temp_r0 = grI2_804D6A00->unk48 +
-                                  grI2_804D6A00->unk4A;
+                        temp_r0 = grI2_804D6A00->unk48 + grI2_804D6A00->unk4A;
 
                         var_r3_2 = test_random(temp_r0);
 
diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index 01940c39c0..64bc8b09b0 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -79,11 +79,11 @@ typedef struct itBombHei_ItemVars {
     /* +14 */ float xDE8;
     /* +18 */ float xDEC;
     /* +1C */ float xDF0;
-    /* +20 */ s32 xDF4;
+    /* +20 */ float xDF4;
     /* +24 */ float xDF8;
     /* +28 */ float xDFC;
     /* +2C */ s32 _E00;
-    /* +30 */ s32 _E04;
+    /* +30 */ f32 xE04;
     /* +34 */ s32 _E08;
     /* +38 */ Vec3 xE0C;
 } itBombHei_ItemVars;
@@ -92,11 +92,12 @@ typedef struct itBombHeiAttributes {
     /*  +0  */ float x0;
     /*  +4  */ float x4;
     /*  +8  */ float x8;
-    /*  +C  */ s32 xC;
+    /*  +C  */ float xC;
     /* +10  */ float x10;
     /* +14  */ float x14;
     /* +18  */ float x18;
     /* +1C  */ float x1C;
+    /* +20  */ Vec3 x20;
 } itBombHeiAttributes;
 
 typedef struct itBox_ItemVars {
diff --git a/src/melee/it/it_2725.c b/src/melee/it/it_2725.c
index f6e48b8c09..47d434df1c 100644
--- a/src/melee/it/it_2725.c
+++ b/src/melee/it/it_2725.c
@@ -5784,7 +5784,7 @@ void it_802742F4(Item_GObj* item_gobj, HSD_GObj* gobj, Fighter_Part ftpart)
         it_80279BBC(item);
     }
     if (item->kind < It_Kind_L_Gun_Ray) { // If a common item
-        it_80275158(item_gobj, it_804D6D28->x30);
+        it_80275158(item_gobj, it_804D6D28->x30_lifetime);
     }
     it_80274F48(item_gobj,
                 item->xC4_article_data->x10_modelDesc->x8_bone_attach_id, gobj,
diff --git a/src/melee/it/item.c b/src/melee/it/item.c
index 22d9febb1d..8c2c433c20 100644
--- a/src/melee/it/item.c
+++ b/src/melee/it/item.c
@@ -602,7 +602,7 @@ static void Item_80267AA8(HSD_GObj* gobj, SpawnItem* spawnItem)
     item_data->xCB8_outDamageDirection = 0.0f;
     item_data->xC68 = 0.0f;
     item_data->xCD0 = 0.0f;
-    it_80275158(gobj, it_804D6D28->x30);
+    it_80275158(gobj, it_804D6D28->x30_lifetime);
     item_data->xDD0_flag.b3 = false;
     item_data->spin_spd = item_data->xCC_item_attr->xC_spin_speed;
     item_data->xDC8_word.flags.x19 = item_data->xCC_item_attr->x1_3;
diff --git a/src/melee/it/items/it_27CF.c b/src/melee/it/items/it_27CF.c
index f9150e12ef..f0650f3f35 100644
--- a/src/melee/it/items/it_27CF.c
+++ b/src/melee/it/items/it_27CF.c
@@ -91,8 +91,7 @@ void it_27CF_UnkMotion1_Phys(HSD_GObj* item_gobj)
     ItemAttr* attr;
 
     attr = GET_ITEM(item_gobj)->xCC_item_attr;
-    it_80272860(item_gobj, attr->x10_fall_speed,
-                attr->x14_fall_speed_max);
+    it_80272860(item_gobj, attr->x10_fall_speed, attr->x14_fall_speed_max);
 }
 
 bool it_27CF_UnkMotion1_Coll(HSD_GObj* item_gobj)
diff --git a/src/melee/it/items/itbombhei.c b/src/melee/it/items/itbombhei.c
index ca2f7faff4..b97948d39c 100644
--- a/src/melee/it/items/itbombhei.c
+++ b/src/melee/it/items/itbombhei.c
@@ -1,6 +1,9 @@
 #include "itbombhei.h"
 
 #include "itbombhei.static.h"
+
+#include "math.h"
+
 #include <placeholder.h>
 #include <platform.h>
 
@@ -8,12 +11,19 @@
 
 #include "baselib/gobj.h"
 #include "baselib/jobj.h"
+
+#include "it/forward.h"
+
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/itCommonItems.h"
 #include "it/item.h"
+#include "lb/lb_00F9.h"
+
+#define GET_ATTRS(ip)                                                         \
+    ((itBombHeiAttributes*) ip->xC4_article_data->x4_specialAttributes)
 
 Item_GObj* it_8027D670(Vec3* pos)
 {
@@ -22,7 +32,7 @@ Item_GObj* it_8027D670(Vec3* pos)
 
     data.kind = It_Kind_BombHei;
     data.pos = data.prev_pos = *pos;
-    data.facing_dir = it_8026B684(&data.pos);
+    data.facing_dir = it_8026B684(&data.prev_pos);
     data.x3C_damage = 0;
     data.vel.x = data.vel.y = data.vel.z = 0.0f;
     data.x0_parent_gobj = NULL;
@@ -107,7 +117,7 @@ void fn_8027DAC8(Item_GObj* igp)
 
     it_8026B390(igp);
     itResetVelocity(ip);
-    itBombhei_UpdateStatePreserveBone(igp, 0);
+    itBombhei_UpdateStatePreserveBone(igp, 0, ITEM_ANIM_UPDATE);
 }
 
 bool itBombhei_UnkMotion0_Anim(Item_GObj* igp)
@@ -144,7 +154,7 @@ void it_8027DE18(Item_GObj* igp)
         ip->xDD4_itemVar.bombhei.xDD4 = (s32) ap->x14;
     }
 
-    itBombhei_UpdateStatePreserveBone(igp, 1);
+    itBombhei_UpdateStatePreserveBone(igp, 1, ITEM_ANIM_UPDATE);
 }
 
 bool itBombhei_UnkMotion1_Anim(Item_GObj* gobj)
@@ -183,10 +193,10 @@ void itBombhei_Logic6_PickedUp(Item_GObj* igp)
 
     if (ip->xDD4_itemVar.bombhei.xDE0 == 0) {
         ip->x5D0_animFrameSpeed = ap->x0;
-        itBombhei_UpdateStatePreserveBone(igp, 7);
+        itBombhei_UpdateStatePreserveBone(igp, 7, ITEM_ANIM_UPDATE);
     } else {
         ip->x5D0_animFrameSpeed = 1.0f;
-        itBombhei_UpdateStatePreserveBone2(igp, 8);
+        itBombhei_UpdateStatePreserveBone(igp, 8, ITEM_UNK_0x1);
     }
 }
 
@@ -206,11 +216,70 @@ bool itBombhei_UnkMotion8_Anim(Item_GObj* igp)
 
 void itBombhei_UnkMotion8_Phys(Item_GObj* gobj) {}
 
-/// #it_3F14_Logic6_Dropped
+void it_3F14_Logic6_Dropped(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* attrs;
+    ip = GET_ITEM(arg0);
+    attrs = GET_ATTRS(ip);
+    ip->xDC8_word.flags.x19 = 1;
+    ip->xDC8_word.flags.x17 = 1;
+    if (ip->xDD4_itemVar.bombhei.xDE0 == 0) {
+        ip->x5D0_animFrameSpeed = attrs->x0;
+        itBombhei_UpdateStatePreserveBone(arg0, 9, 6);
+    } else {
+        ip->x5D0_animFrameSpeed = 1.0f;
+        itBombhei_UpdateStatePreserveBone(arg0, 0xA, 5);
+    }
+}
+
+void it_8027E978(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* attrs;
+
+    ip = GET_ITEM(arg0);
+    attrs = GET_ATTRS(ip);
+    it_802762B0(ip);
+    ip->xDD4_itemVar.bombhei.xDD4 = (s32) attrs->x4;
+    ip->xDD4_itemVar.bombhei.xE04 = ((M_PI / 2 * (f64) ip->facing_dir) /
+                                     (f64) ip->xDD4_itemVar.bombhei.xDD4);
+    ip->xDD4_itemVar.bombhei.xDE4 = 1;
+
+    itBombhei_UpdateStatePreserveBone(arg0, 3, ITEM_ANIM_UPDATE);
+}
 
-/// #it_8027E978
+bool itBombhei_UnkMotion3_Anim(Item_GObj* gobj)
+{
+    HSD_JObj* temp_r0;
+    HSD_JObj* temp_r30_2;
+    Item* temp_r3;
+    f32 temp_f31;
+    PAD_STACK(8);
 
-/// #itBombhei_UnkMotion3_Anim
+    temp_r0 = gobj->hsd_obj;
+    temp_r3 = GET_ITEM(gobj);
+    temp_f31 = temp_r3->xDD4_itemVar.bombhei.xE04;
+    HSD_JObjAddRotationY(temp_r0, temp_f31);
+    temp_r3 = GET_ITEM(gobj);
+    temp_r3->xDD4_itemVar.bombhei.xDD4 -= 1;
+    temp_r3->xDD4_itemVar.bombhei.xDEC -= 1.0f;
+    if (temp_r3->xDD4_itemVar.bombhei.xDD4 <= 0) {
+        itBombHeiAttributes* attr = GET_ATTRS(temp_r3);
+        temp_r3->xDC8_word.flags.x19 = 1;
+        temp_r3->xDD4_itemVar.bombhei.xDD4 = attr->x14;
+        it_8027EE04(gobj);
+        Item_8026AE84(temp_r3, 0xF0, 0x7FU, 0x40U);
+    }
+    temp_r3 = GET_ITEM(gobj);
+    temp_f31 = temp_r3->xDD4_itemVar.bombhei.xDF8;
+    temp_r30_2 = temp_r3->xBBC_dynamicBoneTable->bones[0xB];
+    HSD_JObjAddTranslationY(temp_r30_2, temp_f31);
+    temp_f31 = temp_r3->xDD4_itemVar.bombhei.xDFC;
+    HSD_JObjAddRotationX(temp_r30_2, temp_f31);
+
+    return false;
+}
 
 void itBombhei_UnkMotion3_Phys(Item_GObj* gobj) {}
 
@@ -222,17 +291,130 @@ bool itBombhei_UnkMotion3_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_8027EE04
+void it_8027EE04(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* attrs;
 
-/// #itBombhei_UnkMotion2_Anim
+    ip = GET_ITEM(arg0);
+    attrs = GET_ATTRS(ip);
+    it_8026B3A8(arg0);
+    ip->x40_vel.x = attrs->xC * ip->facing_dir;
+    ip->xD5C = 0;
+    it_80275444(arg0);
+    itBombhei_UpdateStatePreserveBone(arg0, 2, 0x1A);
+}
+
+bool itBombhei_UnkMotion2_Anim(Item_GObj* gobj)
+{
+    Item* ip;
+    f32 temp_f30;
+    f32 temp_f31;
+    itBombHeiAttributes* attrs;
+    HSD_JObj* temp_r30_2;
+
+    ip = GET_ITEM(gobj);
+    ip->xDD4_itemVar.bombhei.xDD4 -= 1;
+    ip->xDD4_itemVar.bombhei.xDEC -= 1.0f;
+    if ((s32) ip->xDD4_itemVar.bombhei.xDD4 <= 0) {
+        it_8027F8E0(gobj);
+    }
+    if (it_80272C6C(gobj) == 0) {
+        ip = GET_ITEM(gobj);
+        attrs = GET_ATTRS(ip);
+        it_8026B3A8(gobj);
+        ip->x40_vel.x = attrs->xC * ip->facing_dir;
+        ip->xD5C = 0;
+        it_80275444(gobj);
+        itBombhei_UpdateStatePreserveBoneFake(gobj, 2, 0x1A);
+    }
+    ip = GET_ITEM(gobj);
+    temp_f31 = ip->xDD4_itemVar.bombhei.xDF8;
+    temp_r30_2 = ip->xBBC_dynamicBoneTable->bones[0xB];
+    HSD_JObjAddTranslationY(temp_r30_2, temp_f31);
+    temp_f30 = ip->xDD4_itemVar.bombhei.xDFC;
+    HSD_JObjAddRotationX(temp_r30_2, temp_f30);
+    return false;
+}
 
 void itBombhei_UnkMotion2_Phys(Item_GObj* gobj) {}
 
-/// #itBombhei_UnkMotion2_Coll
+bool itBombhei_UnkMotion2_Coll(Item_GObj* gobj)
+{
+    Item* ip;
+    Item_GObj* new_var;
 
-/// #it_8027F42C
+    s32 temp_r0;
+    itBombHeiAttributes* attr;
 
-/// #itBombhei_UnkMotion4_Anim
+    if (it_8026D8A4(gobj, (void (*)(HSD_GObj*)) it_8027F42C) != 0) {
+        if (it_80276308(gobj) != 0) {
+            it_8027F42C(gobj);
+        }
+        it_80276CB8(gobj);
+    } else {
+        ip = GET_ITEM(gobj);
+        temp_r0 = ip->msid;
+        attr = GET_ATTRS(ip);
+        if ((temp_r0 == 3) || (temp_r0 == 0)) {
+            ip->xDC8_word.flags.x19 = 1;
+            ip->xDC8_word.flags.x17 = 1;
+            it_80274740(gobj);
+        }
+        new_var = gobj;
+        if (GET_ITEM(gobj)->msid == 3) {
+            GET_ITEM(gobj)->xDD4_itemVar.bombhei.xDD4 = attr->x14;
+        }
+        itBombhei_UpdateStatePreserveBoneFake(new_var, 1, ITEM_ANIM_UPDATE);
+    }
+    return false;
+}
+
+void it_8027F42C(Item_GObj* arg0)
+{
+    HSD_JObj* temp_r30;
+    Item* temp_r31;
+    PAD_STACK(8);
+
+    temp_r31 = arg0->user_data;
+    it_802762B0(temp_r31);
+    itBombhei_UpdateStatePreserveBone(arg0, 4, 0x12);
+    temp_r31->x40_vel.x = 0.0f;
+    temp_r31->xDD4_itemVar.bombhei.xDF4 = -temp_r31->facing_dir;
+}
+
+bool itBombhei_UnkMotion4_Anim(Item_GObj* gobj)
+{
+    HSD_JObj* bone;
+    Item* ip;
+    f32 temp_f31;
+    itBombHeiAttributes* temp_r29;
+
+    ip = GET_ITEM(gobj);
+    if (it_80272C6C(gobj) == 0) {
+        ip->facing_dir = -ip->facing_dir;
+        ip = GET_ITEM(gobj);
+        temp_r29 = GET_ATTRS(ip);
+        it_8026B3A8(gobj);
+        ip->x40_vel.x = temp_r29->xC * ip->facing_dir;
+        ip->xD5C = 0;
+        it_80275444(gobj);
+        itBombhei_UpdateStatePreserveBoneFake(gobj, 2, 0x1A);
+    }
+    ip->xDD4_itemVar.bombhei.xDD4 -= 1;
+    ip->xDD4_itemVar.bombhei.xDEC -= 1.0f;
+    if (ip->xDD4_itemVar.bombhei.xDD4 <= 0) {
+        ip->facing_dir = ip->xDD4_itemVar.bombhei.xDF4;
+        it_8027F8E0(gobj);
+    }
+    ip = GET_ITEM(gobj);
+    temp_f31 = ip->xDD4_itemVar.bombhei.xDF8;
+    bone = ip->xBBC_dynamicBoneTable->bones[0xB];
+    HSD_JObjAddTranslationY(bone, temp_f31);
+    temp_f31 = ip->xDD4_itemVar.bombhei.xDFC;
+    HSD_JObjAddRotationX(bone, temp_f31);
+    return false;
+}
 
 void itBombhei_UnkMotion4_Phys(Item_GObj* gobj) {}
 
@@ -244,7 +426,26 @@ bool itBombhei_UnkMotion4_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_8027F8E0
+void it_8027F8E0(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* attr;
+
+    ip = GET_ITEM(arg0);
+    attr = GET_ATTRS(ip);
+    ip->x40_vel.x = 0.0f;
+    it_8026B390(arg0);
+    if (ip->xDD4_itemVar.bombhei.xDE0 == 0) {
+        ip->xDD4_itemVar.bombhei.xDD4 = attr->x18;
+        ip->xDD4_itemVar.bombhei.xDD8 = 1;
+        ip->xD44_lifeTimer = (f32) it_804D6D28->x30_lifetime;
+        ip->xDD4_itemVar.bombhei.xDF0 = attr->x8;
+        ip->xDD4_itemVar.bombhei.xDE0 = 1;
+        itBombhei_UpdateStatePreserveBone(arg0, 5, ITEM_ANIM_UPDATE);
+    } else {
+        itBombhei_UpdateStatePreserveBone(arg0, 5, ITEM_UNK_0x1);
+    }
+}
 
 bool itBombhei_UnkMotion5_Anim(Item_GObj* gobj)
 {
@@ -263,7 +464,26 @@ bool itBombhei_UnkMotion5_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #fn_8027FCA8
+void fn_8027FCA8(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* attrs;
+
+    ip = GET_ITEM(arg0);
+    attrs = GET_ATTRS(ip);
+    ip->x40_vel.x = 0.0f;
+    it_8026B390(arg0);
+    if (ip->xDD4_itemVar.bombhei.xDE0 == 0) {
+        ip->xDD4_itemVar.bombhei.xDD4 = attrs->x18;
+        ip->xDD4_itemVar.bombhei.xDD8 = 1;
+        ip->xD44_lifeTimer = (f32) it_804D6D28->x30_lifetime;
+        ip->xDD4_itemVar.bombhei.xDF0 = attrs->x8;
+        ip->xDD4_itemVar.bombhei.xDE0 = 1;
+        itBombhei_UpdateStatePreserveBone(arg0, 6, ITEM_ANIM_UPDATE);
+    } else {
+        itBombhei_UpdateStatePreserveBone(arg0, 6, ITEM_UNK_0x1);
+    }
+}
 
 bool itBombhei_UnkMotion6_Anim(Item_GObj* gobj)
 {
@@ -278,7 +498,33 @@ void itBombhei_UnkMotion6_Phys(Item_GObj* gobj)
     it_80274658(gobj, it_804D6D28->x68_float);
 }
 
-/// #fn_8028007C
+void fn_8028007C(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* attr;
+    PAD_STACK(8);
+
+    ip = GET_ITEM(arg0);
+    attr = GET_ATTRS(ip);
+    if (ABS(ip->xDD4_itemVar.bombhei.xE0C.x) > attr->x20.x ||
+        ABS(ip->xDD4_itemVar.bombhei.xE0C.y) > attr->x20.y)
+    {
+        it_80280DC0(arg0);
+        return;
+    }
+    ip->x40_vel.x = 0.0f;
+    it_8026B390(arg0);
+    if (ip->xDD4_itemVar.bombhei.xDE0 == 0) {
+        ip->xDD4_itemVar.bombhei.xDD4 = attr->x18;
+        ip->xDD4_itemVar.bombhei.xDD8 = 1;
+        ip->xD44_lifeTimer = it_804D6D28->x30_lifetime;
+        ip->xDD4_itemVar.bombhei.xDF0 = attr->x8;
+        ip->xDD4_itemVar.bombhei.xDE0 = 1;
+        itBombhei_UpdateStatePreserveBoneFake(arg0, 5, ITEM_ANIM_UPDATE);
+    } else {
+        itBombhei_UpdateStatePreserveBoneFake(arg0, 5, ITEM_UNK_0x1);
+    }
+}
 
 bool itBombhei_UnkMotion6_Coll(Item_GObj* gobj)
 {
@@ -288,9 +534,58 @@ bool itBombhei_UnkMotion6_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_3F14_Logic6_Thrown
+void it_3F14_Logic6_Thrown(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* temp_r3;
+    HSD_JObj* temp_r29;
+    PAD_STACK(8);
 
-/// #itBombhei_UnkMotion10_Anim
+    ip = GET_ITEM(arg0);
+    temp_r29 = arg0->hsd_obj;
+    temp_r3 = GET_ATTRS(ip);
+    if (ip->xDD4_itemVar.bombhei.xDE0 == 0) {
+        ip->x5D0_animFrameSpeed = temp_r3->x0;
+        itBombhei_UpdateStatePreserveBone(arg0, 9, 6);
+    } else {
+        ip->x5D0_animFrameSpeed = 1.0f;
+        itBombhei_UpdateStatePreserveBone(arg0, 0xA, 0x104);
+    }
+
+    it_80275474(arg0);
+    if (ip->xDD4_itemVar.bombhei.xDE8 != 1.0f) {
+        it_80274484(arg0, temp_r29, ip->xDD4_itemVar.bombhei.xDE8);
+    }
+}
+
+bool itBombhei_UnkMotion10_Anim(Item_GObj* gobj)
+{
+    HSD_JObj* temp_r31;
+    Item* temp_r30;
+    f32 temp_f1;
+    PAD_STACK(8);
+
+    temp_r30 = GET_ITEM(gobj);
+    if (it_80272C6C(gobj) == 0) {
+        temp_r31 = gobj->hsd_obj;
+        if (temp_r30->xDD4_itemVar.bombhei.xDE0 == 0) {
+            temp_r30->x5D0_animFrameSpeed = GET_ATTRS(temp_r30)->x0;
+            itBombhei_UpdateStatePreserveBoneFake(gobj, 9, 0x6);
+        } else {
+            temp_r30->x5D0_animFrameSpeed = 1.0f;
+            itBombhei_UpdateStatePreserveBoneFake(gobj, 0xA, 0x104);
+        }
+        it_80275474(gobj);
+        temp_f1 = temp_r30->xDD4_itemVar.bombhei.xDE8;
+        if (temp_f1 != 1.0f) {
+            it_80274484(gobj, temp_r31, temp_f1);
+        }
+    }
+    if (temp_r30->xDD4_itemVar.bombhei.xDE0 != 0) {
+        it_8027D820(gobj);
+    }
+    return false;
+}
 
 void itBombhei_UnkMotion10_Phys(Item_GObj* gobj)
 {
@@ -299,7 +594,26 @@ void itBombhei_UnkMotion10_Phys(Item_GObj* gobj)
     it_80274658(gobj, it_804D6D28->x68_float);
 }
 
-/// #fn_80280974
+void fn_80280974(Item_GObj* arg0)
+{
+    Item* ip;
+    itBombHeiAttributes* ap;
+    PAD_STACK(8);
+
+    ip = GET_ITEM(arg0);
+    ap = GET_ATTRS(ip);
+    if (ABS(ip->xDD4_itemVar.bombhei.xE0C.x) > ap->x20.x ||
+        ABS(ip->xDD4_itemVar.bombhei.xE0C.y) > ap->x20.y)
+    {
+        it_80280DC0(arg0);
+        return;
+    }
+    it_8026B3A8(arg0);
+    ip->x40_vel.x = ap->xC * ip->facing_dir;
+    ip->xD5C = 0;
+    it_80275444(arg0);
+    itBombhei_UpdateStatePreserveBoneFake(arg0, 2, ITEM_ANIM_UPDATE);
+}
 
 bool itBombhei_UnkMotion10_Coll(Item_GObj* gobj)
 {
@@ -308,8 +622,35 @@ bool itBombhei_UnkMotion10_Coll(Item_GObj* gobj)
     it_8026E414(gobj, fn_80280974);
     return false;
 }
+Vec3 it_803B8600;
+void it_80280B60(Item_GObj* gobj)
+{
+    Vec sp48;
+    Vec3 sp3C;
+    HSD_JObj* temp_r29;
+    Item* temp_r30;
+    PAD_STACK(8);
 
-/// #it_80280B60
+    temp_r30 = gobj->user_data;
+    temp_r29 = gobj->hsd_obj;
+    if (temp_r30->xDC8_word.flags.x13) {
+        sp3C = it_803B8600;
+        it_8027429C(gobj, &sp3C);
+    }
+    it_8026B3A8(gobj);
+    HSD_JObjSetFlagsAll(temp_r29, 0x10U);
+    it_8026BD24(gobj);
+    it_8027518C(gobj);
+    temp_r30->x5D0_animFrameSpeed = 1.0f;
+    it_80273454(gobj);
+    temp_r30->xDD4_itemVar.bombhei.xDDC = 1;
+    it_80272C08(gobj);
+    it_802756D0(gobj);
+    it_80275444(gobj);
+    sp48 = temp_r30->pos;
+    lb_800119DC(&sp48, 0x78, 1.0f, 0.02f, 1.0471976f);
+    itBombhei_UpdateStatePreserveBone(gobj, 0xB, 0x12);
+}
 
 void it_80280DC0(Item_GObj* gobj)
 {
@@ -351,7 +692,21 @@ bool itBombhei_UnkMotion11_Coll(Item_GObj* gobj)
     return false;
 }
 
-/// #it_3F14_Logic6_EnteredAir
+void it_3F14_Logic6_EnteredAir(Item_GObj* arg0)
+{
+    HSD_JObj* temp_r31;
+    f32 temp_f30;
+    f32 temp_f31;
+    DynamicBoneTable* new_var;
+    PAD_STACK(0x18);
+
+    temp_r31 = (new_var = GET_ITEM(arg0)->xBBC_dynamicBoneTable)->bones[0xB];
+    temp_f31 = HSD_JObjGetTranslationY(temp_r31);
+    temp_f30 = HSD_JObjGetRotationX(temp_r31);
+    Item_80268E5C((HSD_GObj*) arg0, 0xC, ITEM_ANIM_UPDATE);
+    HSD_JObjSetTranslateY(temp_r31, temp_f31);
+    HSD_JObjSetRotationX(temp_r31, temp_f30);
+}
 
 bool itBombhei_UnkMotion12_Anim(Item_GObj* gobj)
 {
diff --git a/src/melee/it/items/itbombhei.static.h b/src/melee/it/items/itbombhei.static.h
index 350e31d9d5..a7606a7ef0 100644
--- a/src/melee/it/items/itbombhei.static.h
+++ b/src/melee/it/items/itbombhei.static.h
@@ -35,7 +35,7 @@ static inline HSD_JObj* itGetDynamicBone(Item_GObj* igp, u32 bone_id)
 }
 
 static inline void itBombhei_UpdateStatePreserveBone(Item_GObj* igp,
-                                                     int anim_id)
+                                                     int anim_id, int flags)
 {
     Item* ip = GET_ITEM(igp);
     HSD_JObj* jobj = ip->xBBC_dynamicBoneTable->bones[0xB];
@@ -44,13 +44,13 @@ static inline void itBombhei_UpdateStatePreserveBone(Item_GObj* igp,
 
     y = HSD_JObjGetTranslationY(jobj);
     x = HSD_JObjGetRotationX(jobj);
-    Item_80268E5C((HSD_GObj*) igp, anim_id, ITEM_ANIM_UPDATE);
+    Item_80268E5C((HSD_GObj*) igp, anim_id, flags);
     HSD_JObjSetTranslateY(jobj, y);
     HSD_JObjSetRotationX(jobj, x);
 }
 
-static inline void itBombhei_UpdateStatePreserveBone2(Item_GObj* igp,
-                                                      int anim_id)
+static inline void
+itBombhei_UpdateStatePreserveBoneFake(Item_GObj* igp, int anim_id, int flags)
 {
     Item* ip = GET_ITEM(igp);
     HSD_JObj* jobj = ip->xBBC_dynamicBoneTable->bones[0xB];
@@ -59,37 +59,7 @@ static inline void itBombhei_UpdateStatePreserveBone2(Item_GObj* igp,
 
     y = HSD_JObjGetTranslationY(jobj);
     x = HSD_JObjGetRotationX(jobj);
-    Item_80268E5C((HSD_GObj*) igp, anim_id, ITEM_UNK_0x1);
-    HSD_JObjSetTranslateY(jobj, y);
-    HSD_JObjSetRotationX(jobj, x);
-}
-
-static inline void itBombhei_UpdateStatePreserveBone3(Item_GObj* igp,
-                                                      int anim_id)
-{
-    Item* ip = GET_ITEM(igp);
-    HSD_JObj* jobj = ip->xBBC_dynamicBoneTable->bones[0xB];
-    f32 x;
-    f32 y;
-
-    y = HSD_JObjGetTranslationY(jobj);
-    x = HSD_JObjGetRotationX(jobj);
-    Item_80268E5C((HSD_GObj*) igp, anim_id, ITEM_ANIM_UPDATE);
-    fake_HSD_JObjSetTranslateY(jobj, y);
-    fake_HSD_JObjSetRotationX(jobj, x);
-}
-
-static inline void itBombhei_UpdateStatePreserveBone4(Item_GObj* igp,
-                                                      int anim_id)
-{
-    Item* ip = GET_ITEM(igp);
-    HSD_JObj* jobj = ip->xBBC_dynamicBoneTable->bones[0xB];
-    f32 x;
-    f32 y;
-
-    y = HSD_JObjGetTranslationY(jobj);
-    x = HSD_JObjGetRotationX(jobj);
-    Item_80268E5C((HSD_GObj*) igp, anim_id, ITEM_UNK_0x1);
+    Item_80268E5C((HSD_GObj*) igp, anim_id, flags);
     fake_HSD_JObjSetTranslateY(jobj, y);
     fake_HSD_JObjSetRotationX(jobj, x);
 }
@@ -130,10 +100,10 @@ static inline void inline_UnkMotion8_Anim(Item_GObj* igp)
 
         if (ip->xDD4_itemVar.bombhei.xDE0 == 0) {
             ip->x5D0_animFrameSpeed = ap->x0;
-            itBombhei_UpdateStatePreserveBone3(igp, 7);
+            itBombhei_UpdateStatePreserveBoneFake(igp, 7, ITEM_ANIM_UPDATE);
         } else {
             ip->x5D0_animFrameSpeed = 1.0f;
-            itBombhei_UpdateStatePreserveBone4(igp, 8);
+            itBombhei_UpdateStatePreserveBoneFake(igp, 8, ITEM_UNK_0x1);
         }
     }
 }
diff --git a/src/melee/it/items/itfreeze.c b/src/melee/it/items/itfreeze.c
index eddccc03e1..4b2015a02d 100644
--- a/src/melee/it/items/itfreeze.c
+++ b/src/melee/it/items/itfreeze.c
@@ -15,7 +15,7 @@ void it_8028EC98(Item_GObj* gobj, f32 vel)
 {
     Item* ip = GET_ITEM(gobj);
     ip->xDD4_itemVar.freeze.x10 = vel;
-    it_80275158(gobj, (f32) it_804D6D28->x30);
+    it_80275158(gobj, (f32) it_804D6D28->x30_lifetime);
 }
 
 void it_8028ECE0(Item_GObj* gobj)
diff --git a/src/melee/it/items/itgshell.c b/src/melee/it/items/itgshell.c
index 99c81d19de..7f9444a8ff 100644
--- a/src/melee/it/items/itgshell.c
+++ b/src/melee/it/items/itgshell.c
@@ -384,7 +384,7 @@ void it_8028C3A8(Item_GObj* gobj)
     it_8026B3A8(gobj);
     it_80275474(gobj);
     it_80275414(gobj);
-    it_80275158(gobj, it_804D6D28->x30);
+    it_80275158(gobj, it_804D6D28->x30_lifetime);
     ip->xD5C = 0;
     ip->xDD4_itemVar.gshell.xDD8 = attrs->x24;
     ip->xDD4_itemVar.gshell.xDDC = attrs->x30;
@@ -473,7 +473,7 @@ void it_8028C898(Item_GObj* gobj)
     it_8026B3A8(gobj);
     it_80275474(gobj);
     it_80275414(gobj);
-    it_80275158(gobj, it_804D6D28->x30);
+    it_80275158(gobj, it_804D6D28->x30_lifetime);
     ip->xD5C = 0;
     ip->xDD4_itemVar.gshell.xDD8 = attrs->x24;
     ip->xDD4_itemVar.gshell.xDDC = attrs->x30;
diff --git a/src/melee/it/items/itlikelike.c b/src/melee/it/items/itlikelike.c
index d2edc13049..2794266dda 100644
--- a/src/melee/it/items/itlikelike.c
+++ b/src/melee/it/items/itlikelike.c
@@ -4,11 +4,13 @@
 
 #include <placeholder.h>
 #include <platform.h>
-#include "it/forward.h"
 
 #include "cm/camera.h"
 #include "ft/ft_0C31.h"
 #include "ft/ftlib.h"
+
+#include "it/forward.h"
+
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
@@ -26,13 +28,13 @@
     ((itLikelikeAttributes*) ip->xC4_article_data->x4_specialAttributes)
 
 // Can't inline this in inlines.h because of dynamic GET_ATTRS
-static inline void itSwapVelocity(Item* ip){
+static inline void itSwapVelocity(Item* ip)
+{
     itLikelikeAttributes* attr = GET_ATTRS(ip);
     ip->facing_dir = -ip->facing_dir;
     ip->x40_vel.x = ip->facing_dir * attr->x0->y;
 }
 
-
 bool it_802D9A2C(Item_GObj* gobj)
 {
     Item* ip = HSD_GObjGetUserData(gobj);
@@ -123,7 +125,7 @@ bool it_2725_Logic5_DmgReceived(Item_GObj* gobj)
         it_80275474(gobj);
         it_8027CE44(gobj);
         Camera_80030E44(2, &ip->pos);
-        if (HSD_Randf() < *(it_804D6D40+4)) {
+        if (HSD_Randf() < *(it_804D6D40 + 4)) {
             it_802DC3DC(gobj);
         } else {
             it_802DC310(gobj);
@@ -134,7 +136,6 @@ bool it_2725_Logic5_DmgReceived(Item_GObj* gobj)
     return false;
 }
 
-
 void it_802D9DDC(Item_GObj* gobj)
 {
     PAD_STACK(16);
@@ -192,7 +193,6 @@ bool itLikelike_UnkMotion0_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-
 bool itLikelike_UnkMotion7_Anim(Item_GObj* gobj)
 {
     if (it_80272C6C(gobj) == 0) {
@@ -239,7 +239,6 @@ bool itLikelike_UnkMotion7_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-
 void it_802DA104(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -256,7 +255,6 @@ void it_802DA104(Item_GObj* gobj)
     Item_80268E5C(gobj, 1, 1);
 }
 
-
 bool itLikelike_UnkMotion1_Anim(Item_GObj* gobj)
 {
     PAD_STACK(8);
@@ -326,7 +324,7 @@ bool itLikelike_UnkMotion1_Coll(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     Item* ip2;
-    Item_GObj *new_var;
+    Item_GObj* new_var;
     itLikelikeAttributes* attrs = GET_ATTRS(ip);
     f32 temp_f3;
     s32 temp_r3;
@@ -378,10 +376,9 @@ void it_802DA4C0(Item_GObj* gobj)
     ip->xDD4_itemVar.likelike.x4C = attr->x18;
     ip->xDD4_itemVar.likelike.x44 = 0x3c;
     ip->xDD4_itemVar.likelike.x48 = 0;
-    Item_80268E5C((HSD_GObj* ) gobj, 2, ITEM_UNK_0x1);
+    Item_80268E5C((HSD_GObj*) gobj, 2, ITEM_UNK_0x1);
 }
 
-
 bool itLikelike_UnkMotion2_Anim(Item_GObj* gobj)
 {
     PAD_STACK(8);
@@ -486,8 +483,8 @@ bool itLikelike_UnkMotion2_Coll(Item_GObj* gobj)
                 goto block_15;
             }
         } else {
-block_15:
-    ip->xDD4_itemVar.likelike.x48 = 0;
+        block_15:
+            ip->xDD4_itemVar.likelike.x48 = 0;
         }
     }
     ip->xDD4_itemVar.likelike.x20 = ip->pos;
@@ -509,7 +506,6 @@ void it_802DA8D8(Item_GObj* gobj)
     Item_80268E5C(gobj, 0, ITEM_ANIM_UPDATE);
 }
 
-
 void it_802DA960(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -530,7 +526,6 @@ void it_802DA960(Item_GObj* gobj)
     Item_80268E5C(gobj, 4, ITEM_ANIM_UPDATE);
 }
 
-
 void it_802DAA10(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -549,10 +544,9 @@ void it_802DAA10(Item_GObj* gobj)
     ip->xDD4_itemVar.likelike.x38 = 0;
     ip->xDD4_itemVar.likelike.x4C = 0;
     it_80273454(gobj);
-    Item_80268E5C((HSD_GObj* ) gobj, 5, ITEM_ANIM_UPDATE);
+    Item_80268E5C((HSD_GObj*) gobj, 5, ITEM_ANIM_UPDATE);
 }
 
-
 bool itLikelike_UnkMotion5_Anim(Item_GObj* gobj)
 {
     return false;
@@ -560,7 +554,6 @@ bool itLikelike_UnkMotion5_Anim(Item_GObj* gobj)
 
 void itLikelike_UnkMotion5_Phys(Item_GObj* gobj)
 {
-
     Item* ip = GET_ITEM(gobj);
     ItemAttr* temp_r4;
     PAD_STACK(8);
@@ -573,8 +566,6 @@ void itLikelike_UnkMotion5_Phys(Item_GObj* gobj)
     it_80272860(gobj, temp_r4->x10_fall_speed, temp_r4->x14_fall_speed_max);
 }
 
-
-
 bool itLikelike_UnkMotion4_Coll(Item_GObj* gobj)
 {
     PAD_STACK(8);
@@ -610,12 +601,11 @@ bool itLikelike_UnkMotion17_Anim(Item_GObj* gobj)
         it_8026D5CC(gobj);
         it_8027C0A8(gobj, 0.0f, 5.0f);
         it_80275270(gobj);
-        Item_80268E5C((HSD_GObj* ) gobj, 0, ITEM_ANIM_UPDATE);
+        Item_80268E5C((HSD_GObj*) gobj, 0, ITEM_ANIM_UPDATE);
     }
     return false;
 }
 
-
 void itLikelike_UnkMotion17_Phys(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -644,7 +634,6 @@ void it_802DAD18(HSD_GObj* gobj)
     Item_80268E5C(gobj, 3, ITEM_ANIM_UPDATE);
 }
 
-
 bool itLikelike_UnkMotion3_Anim(Item_GObj* gobj)
 {
     return false;
@@ -670,7 +659,6 @@ void it_802DAE6C(Item_GObj* gobj)
     Item_80268E5C(gobj, 0x10, ITEM_ANIM_UPDATE);
 }
 
-
 bool itLikelike_UnkMotion16_Anim(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -706,13 +694,12 @@ bool itLikelike_UnkMotion16_Anim(Item_GObj* gobj)
             ip3->xDD4_itemVar.likelike.x4C = attr->x18;
             ip3->xDD4_itemVar.likelike.x44 = 0x3C;
             ip3->xDD4_itemVar.likelike.x48 = 0;
-            Item_80268E5C( gobj, 2, ITEM_UNK_0x1);
+            Item_80268E5C(gobj, 2, ITEM_UNK_0x1);
         }
     }
     return false;
 }
 
-
 void itLikelike_UnkMotion16_Phys(Item_GObj* gobj) {}
 
 bool itLikelike_UnkMotion16_Coll(Item_GObj* gobj)
@@ -737,7 +724,6 @@ bool itLikelike_UnkMotion16_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-
 void it_802DB074(HSD_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -749,7 +735,6 @@ void it_802DB074(HSD_GObj* gobj)
     Item_80268E5C(gobj, 8, ITEM_ANIM_UPDATE);
 }
 
-
 bool itLikelike_UnkMotion8_Anim(Item_GObj* gobj)
 {
     PAD_STACK(8);
@@ -805,8 +790,7 @@ void itLikelike_UnkMotion8_Phys(Item_GObj* gobj)
         Item_80268E5C((HSD_GObj*) gobj, 0, ITEM_ANIM_UPDATE);
         return;
     }
-    it_8027C8D0(&ip->x40_vel, &ip->x378_itemColl.floor.normal,
-                ip->facing_dir);
+    it_8027C8D0(&ip->x40_vel, &ip->x378_itemColl.floor.normal, ip->facing_dir);
     it_8027C0CC(gobj, 0.0f, 5.0f);
 }
 
@@ -832,7 +816,6 @@ bool itLikelike_UnkMotion8_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-
 void it_802DB358(Item_GObj* gobj)
 {
     Item* ip = gobj->user_data;
@@ -855,10 +838,10 @@ void it_802DB398(Item_GObj* gobj)
     } else {
         Item_80268E5C(gobj, 0xC, ITEM_ANIM_UPDATE);
     }
-    it_80274F28(ip, 1, (void (*)(HSD_GObj*)) it_802DB358, (void (*)(HSD_GObj*, HSD_GObj*)) ftCo_800C78B0);
+    it_80274F28(ip, 1, (void (*)(HSD_GObj*)) it_802DB358,
+                (void (*)(HSD_GObj*, HSD_GObj*)) ftCo_800C78B0);
 }
 
-
 bool itLikelike_UnkMotion12_Anim(Item_GObj* gobj)
 {
     Item* ip;
@@ -887,15 +870,14 @@ void itLikelike_UnkMotion12_Phys(Item_GObj* gobj)
 
     if ((enum GroundOrAir) ip->ground_or_air == GA_Air) {
         temp_r4_2 = ip->xCC_item_attr;
-        it_80272860(gobj, temp_r4_2->x10_fall_speed, temp_r4_2->x14_fall_speed_max);
+        it_80272860(gobj, temp_r4_2->x10_fall_speed,
+                    temp_r4_2->x14_fall_speed_max);
         return;
     }
     it_8027C8D0(&ip->x40_vel, &ip->x378_itemColl.floor.normal, ip->facing_dir);
     it_8027C0A8(gobj, 0.0f, 5.0f);
 }
 
-
-
 bool itLikelike_UnkMotion12_Coll(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -919,7 +901,8 @@ void it_802DB5F0(Item_GObj* gobj)
     it_80273454(gobj);
     Item_80268E5C((HSD_GObj*) gobj, 9, ITEM_ANIM_UPDATE);
     Item_8026AE84(ip, 0x139, 0x7FU, 0x40U);
-    it_80274F28(ip, 1, (void (*)(HSD_GObj*)) it_802DB358, (void (*)(HSD_GObj*, HSD_GObj*)) ftCo_800C78B0);
+    it_80274F28(ip, 1, (void (*)(HSD_GObj*)) it_802DB358,
+                (void (*)(HSD_GObj*, HSD_GObj*)) ftCo_800C78B0);
     ip->xDD4_itemVar.likelike.x50 = NULL;
 }
 
@@ -937,13 +920,12 @@ bool itLikelike_UnkMotion9_Anim(Item_GObj* gobj)
             it_8026D5CC(gobj);
             it_8027C0A8(gobj, 0.0f, 5.0f);
             it_80275270(gobj);
-            Item_80268E5C((HSD_GObj* ) gobj, 0, ITEM_ANIM_UPDATE);
+            Item_80268E5C((HSD_GObj*) gobj, 0, ITEM_ANIM_UPDATE);
         }
     }
     return false;
 }
 
-
 void itLikelike_UnkMotion9_Phys(Item_GObj* gobj) {}
 
 bool itLikelike_UnkMotion9_Coll(Item_GObj* gobj)
@@ -959,9 +941,10 @@ void it_802DB74C(Item_GObj* gobj)
     PAD_STACK(8);
 
     it_80273454(gobj);
-    Item_80268E5C((HSD_GObj* ) gobj, 0xA, ITEM_ANIM_UPDATE);
+    Item_80268E5C((HSD_GObj*) gobj, 0xA, ITEM_ANIM_UPDATE);
     Item_8026AE84(ip, 0x139, 0x7FU, 0x40U);
-    it_80274F28(ip, 1, (void (*)(HSD_GObj*)) it_802DB358, (void (*)(HSD_GObj*, HSD_GObj*)) ftCo_800C78B0);
+    it_80274F28(ip, 1, (void (*)(HSD_GObj*)) it_802DB358,
+                (void (*)(HSD_GObj*, HSD_GObj*)) ftCo_800C78B0);
     ip->xDD4_itemVar.likelike.x50 = NULL;
 }
 
@@ -979,7 +962,7 @@ bool itLikelike_UnkMotion10_Anim(Item_GObj* gobj)
             it_8026D5CC(gobj);
             it_8027C0A8(gobj, 0.0f, 5.0f);
             it_80275270(gobj);
-            Item_80268E5C((HSD_GObj* ) gobj, 0, ITEM_ANIM_UPDATE);
+            Item_80268E5C((HSD_GObj*) gobj, 0, ITEM_ANIM_UPDATE);
         }
     }
     return false;
@@ -1002,10 +985,9 @@ void it_802DB8A8(Item_GObj* gobj)
     ip->xDD4_itemVar.likelike.x4C = 0.0f;
     Item_8026AE84(ip, 0x13B, 0x7FU, 0x40U);
     it_80275258(gobj);
-    Item_80268E5C( gobj, 0xD, ITEM_ANIM_UPDATE);
+    Item_80268E5C(gobj, 0xD, ITEM_ANIM_UPDATE);
 }
 
-
 bool itLikelike_UnkMotion13_Anim(Item_GObj* gobj)
 {
     PAD_STACK(8);
@@ -1023,16 +1005,12 @@ void itLikelike_UnkMotion13_Phys(Item_GObj* gobj)
     if (ip->xDD4_itemVar.likelike.x4C == 0) {
         ip->xDD4_itemVar.likelike.x4C = attr->x3D;
 
-        ftCo_800C7C60(
-            (Fighter_GObj*)ip->grab_victim,
-            attr->x3E
-        );
+        ftCo_800C7C60((Fighter_GObj*) ip->grab_victim, attr->x3E);
     } else {
         ip->xDD4_itemVar.likelike.x4C = ip->xDD4_itemVar.likelike.x4C - 1;
     }
 }
 
-
 bool itLikelike_UnkMotion13_Coll(Item_GObj* gobj)
 {
     PAD_STACK(8);
@@ -1053,7 +1031,6 @@ void it_802DB9F4(Item_GObj* gobj)
     Item_80268E5C(gobj, 0xF, ITEM_ANIM_UPDATE);
 }
 
-
 void it_802DBA68(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -1065,10 +1042,9 @@ void it_802DBA68(Item_GObj* gobj)
     it_8026D5CC(gobj);
     it_8027C0A8(gobj, 0.0f, 5.0f);
     it_80275270(gobj);
-    Item_80268E5C((HSD_GObj* ) gobj, 0, ITEM_ANIM_UPDATE);
+    Item_80268E5C((HSD_GObj*) gobj, 0, ITEM_ANIM_UPDATE);
 }
 
-
 void it_802DBAF0(Item_GObj* arg0, s32 arg1, s32 arg2)
 {
     Vec3 vec;
@@ -1082,7 +1058,8 @@ void it_802DBAF0(Item_GObj* arg0, s32 arg1, s32 arg2)
         temp_r4 = ip->grab_victim;
         if ((temp_r4 != NULL) && (temp_r4 == ip->xDD4_itemVar.likelike.x50)) {
             vec.x = 0.0f;
-            vec.y = 0.5f * (ip->x378_itemColl.ecb.top.y + ip->x378_itemColl.ecb.bottom.y);
+            vec.y = 0.5f * (ip->x378_itemColl.ecb.top.y +
+                            ip->x378_itemColl.ecb.bottom.y);
             vec.z = 0.0f;
             vec.x += ip->pos.x;
             vec.y += ip->pos.y;
@@ -1129,7 +1106,7 @@ bool itLikelike_UnkMotion14_Anim(Item_GObj* gobj)
         it_8026D5CC(gobj);
         it_8027C0A8(gobj, 0.0f, 5.0f);
         it_80275270(gobj);
-        Item_80268E5C((HSD_GObj* ) gobj, 0, ITEM_ANIM_UPDATE);
+        Item_80268E5C((HSD_GObj*) gobj, 0, ITEM_ANIM_UPDATE);
     } else {
         ip->xDD4_itemVar.likelike.x4C = temp_r3 + 1;
     }
@@ -1160,14 +1137,13 @@ bool itLikelike_UnkMotion15_Anim(Item_GObj* gobj)
         it_8026D5CC(gobj);
         it_8027C0A8(gobj, 0.0f, 5.0f);
         it_80275270(gobj);
-        Item_80268E5C((HSD_GObj* ) gobj, 0, ITEM_ANIM_UPDATE);
+        Item_80268E5C((HSD_GObj*) gobj, 0, ITEM_ANIM_UPDATE);
     } else {
         ip->xDD4_itemVar.likelike.x4C = temp_r3 + 1;
     }
     return false;
 }
 
-
 void itLikelike_UnkMotion15_Phys(Item_GObj* gobj) {}
 
 bool itLikelike_UnkMotion15_Coll(Item_GObj* gobj)
@@ -1217,10 +1193,9 @@ void it_2725_Logic5_Dropped(Item_GObj* gobj)
     }
     ip2->xDD4_itemVar.likelike.x38 = 0;
     it_80273454(gobj);
-    Item_80268E5C((HSD_GObj* ) gobj, 4, ITEM_ANIM_UPDATE);
+    Item_80268E5C((HSD_GObj*) gobj, 4, ITEM_ANIM_UPDATE);
 }
 
-
 void itLikeLike_Logic5_Thrown(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -1257,7 +1232,6 @@ void it_802DC0AC(Item_GObj* gobj)
     Item_80268E5C(gobj, 6, ITEM_ANIM_UPDATE);
 }
 
-
 bool itLikelike_UnkMotion6_Anim(Item_GObj* gobj)
 {
     return false;
@@ -1325,7 +1299,6 @@ bool itLikelike_UnkMotion6_Coll(Item_GObj* gobj)
     return it_8027C794(gobj);
 }
 
-
 void it_802DC310(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
diff --git a/src/melee/it/items/itmatadogas.c b/src/melee/it/items/itmatadogas.c
index 0f05168bf4..fe5bc9b3c0 100644
--- a/src/melee/it/items/itmatadogas.c
+++ b/src/melee/it/items/itmatadogas.c
@@ -2,7 +2,6 @@
 
 #include <placeholder.h>
 #include <platform.h>
-#include <baselib/random.h>
 
 #include "ef/eflib.h"
 #include "it/inlines.h"
@@ -11,6 +10,8 @@
 #include "it/it_2725.h"
 #include "it/item.h"
 
+#include <baselib/random.h>
+
 void it_802CAFD4(Item_GObj* gobj)
 {
     HSD_JObj* jobj;
@@ -66,7 +67,8 @@ void it_802CB2B0(Item_GObj* gobj)
     Item* ip = GET_ITEM(gobj);
     PAD_STACK(8);
     if (ip->xDAC_itcmd_var0 != 0) {
-        itMatadogasAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+        itMatadogasAttributes* attrs =
+            ip->xC4_article_data->x4_specialAttributes;
         if (--ip->xDD4_itemVar.matadogas.x60 == 0) {
             ip->xDD4_itemVar.matadogas.x60 = attrs->x4;
             if (HSD_Randi(2) == 0) {
diff --git a/src/melee/it/types.h b/src/melee/it/types.h
index 6948a707db..daec6427f4 100644
--- a/src/melee/it/types.h
+++ b/src/melee/it/types.h
@@ -695,7 +695,7 @@ struct ItemCommonData {
     u32 x24;
     u32 x28;
     u32 x2C;
-    u32 x30; // lifetime?
+    u32 x30_lifetime;
     u32 x34;
     u32 x38_float;
     s32 x3C_float;
