diff --git a/src/melee/gr/grdisplay.c b/src/melee/gr/grdisplay.c
index 2a8ed3b5cf..41d432bc4d 100644
--- a/src/melee/gr/grdisplay.c
+++ b/src/melee/gr/grdisplay.c
@@ -1,68 +1,140 @@
+#include "grdisplay.h"
+
+#include "inlines.h"
+
 #include "cm/camera.h"
 #include "ft/ftlib.h"
 #include "gr/ground.h"
+
+#include "lb/forward.h"
+
 #include "lb/lbshadow.h"
 
 #include <baselib/cobj.h>
 #include <baselib/displayfunc.h>
+#include <baselib/fog.h>
 #include <baselib/gobj.h>
 #include <baselib/jobj.h>
 #include <baselib/mtx.h>
+#include <baselib/state.h>
 
-/* 1C5B90 */ static void grDisplay_801C5B90(HSD_JObj* jobj, void* arg1,
-                                            u32 arg2, int arg3);
-/* 1C5F60 */ static void grDisplay_801C5F60(HSD_GObj* gobj, int code);
+/* 1C5B90 */ static void grDisplay_801C5B90(HSD_JObj* jobj, Mtx vmtx, u32 arg2,
+                                            u32 rendermode);
 /* 1C6620 */ static void grAnime_801C6620(HSD_PObj* arg0, HSD_ShapeAnim* arg1);
 
-void grDisplay_801C5B90(HSD_JObj* jobj, void* arg1, u32 arg2, int arg3)
+void grDisplay_801C5B90(HSD_JObj* jobj, Mtx vmtx, u32 flags, u32 rendermode)
 {
-    Mtx sp3C;
-    Vec3 sp30;
+    HSD_GObj* cur;
     HSD_CObj* cobj;
-    HSD_JObj* cur;
+    Mtx mtx;
+    Vec3 vec;
     u32 unused[6];
 
-    if (jobj == NULL) {
-        return;
+    if (jobj != NULL) {
+        if (jobj->flags & JOBJ_INSTANCE) {
+            if (!(jobj->flags & JOBJ_HIDDEN)) {
+                HSD_JObjSetupMatrix(jobj);
+                HSD_JObjSetupMatrix(jobj->child);
+
+                HSD_MtxInverseConcat(jobj->child->mtx, jobj->mtx, mtx);
+                cobj = HSD_CObjGetCurrent();
+                HSD_ASSERT(82, cobj);
+                PSMTXConcat(HSD_CObjGetViewingMtxPtrDirect(cobj), mtx, mtx);
+                grDisplay_801C5B90(jobj->child, mtx, flags, rendermode);
+            }
+        } else {
+            if (jobj->flags & (flags << 18)) {
+                for (cur = HSD_GObj_Entities->fighters; cur != NULL;
+                     cur = cur->next)
+                {
+                    LbShadow* shadow = ftLib_800872B0(cur);
+                    ftLib_80086644(cur, &vec);
+                    if (Camera_80030B24() || stage_info.x17C(&vec, -1, jobj)) {
+                        shadow->x0_b4 = 0;
+                    } else {
+                        shadow->x0_b4 = 1;
+                    }
+                }
+                lbShadow_8000EFEC();
+                HSD_JObjDisp(jobj, vmtx, flags, rendermode);
+            }
+            if (jobj->flags & (flags << 28)) {
+                for (jobj = jobj->child; jobj != NULL; jobj = jobj->next) {
+                    grDisplay_801C5B90(jobj, vmtx, flags, rendermode);
+                }
+            }
+        }
     }
+}
 
-    if (jobj->flags & 0x1000) {
-        if (!(jobj->flags & 0x10)) {
-            HSD_JObjSetupMatrix(jobj);
-            {
-                HSD_JObj* jtmp = jobj->child;
-                HSD_JObjSetupMatrix(jtmp);
+void grDisplay_801C5DB0(HSD_GObj* gobj, int code)
+{
+    Ground* gp;
+    HSD_GObj* camgobj;
+    HSD_GObj* foggobj;
+    HSD_Fog* fog;
+    int i;
+
+    gp = GET_GROUND(gobj);
+    if (gp->x11_flags.b012 == Camera_8003108C()) {
+        if (gp->x18 != NULL) {
+            if (((intptr_t) gp->x18 & ~0x7FFFFFFF) == 0) {
+                OSReport("oioi... %08x\n");
+            }
+            if (HSD_GObj_804D7818->hsd_obj == gp->x18->hsd_obj) {
+                goto skip;
+            }
+            if (gp->x10_flags.b3 == 0) {
+                return;
+            }
+            camgobj = Camera_80030A50();
+            if (camgobj == NULL) {
+                return;
+            }
+            if (GET_COBJ(camgobj) == NULL) {
+                return;
             }
 
-            HSD_MtxInverseConcat(jobj->child->mtx, jobj->mtx, sp3C);
-            cobj = HSD_CObjGetCurrent();
-            HSD_ASSERT(82, cobj);
-            {
-                MtxPtr tmp = cobj->view_mtx;
-                PSMTXConcat(tmp, sp3C, sp3C);
+            if (HSD_CObjGetCurrent() == GET_COBJ(camgobj)) {
+                goto skip;
             }
-            grDisplay_801C5B90(jobj->child, sp3C, arg2, arg3);
+            return;
         }
-    } else {
-        if (jobj->flags & (arg2 << 18)) {
-            HSD_GObj* cur;
-            for (cur = HSD_GObj_Entities->fighters; cur != NULL;
-                 cur = cur->next)
-            {
-                LbShadow* temp_r26 = ftLib_800872B0(cur);
-                ftLib_80086644(cur, &sp30);
-                if (Camera_80030B24() || stage_info.x17C(&sp30, -1, jobj)) {
-                    temp_r26->x0_b4 = 0;
-                } else {
-                    temp_r26->x0_b4 = 1;
+        if (Ground_801C2C8C(HSD_GObj_804D7818->hsd_obj) != false) {
+            return;
+        }
+
+    skip:
+        if (Camera_80030A78() == false && Camera_80030AC4() != false) {
+            if (gp->x10_flags.b2 == 0) {
+                HSD_FogSet(0);
+            }
+            HSD_StateInvalidate(-1);
+
+            if (gp->x10_flags.b5 != 0) {
+                grDisplay_801C5B90(GET_JOBJ(gobj), NULL,
+                                   HSD_GObj_80390EB8(code), 0);
+            } else {
+                HSD_GObj* fighter;
+                for (fighter = HSD_GObj_Entities->fighters; fighter != NULL;
+                     fighter = fighter->next)
+                {
+                    LbShadow* shadow = ftLib_800872B0(fighter);
+                    shadow->x0_b4 = 0;
+                    lbShadow_8000EEE0(fighter);
                 }
+                HSD_GObj_JObjCallback(gobj, code);
             }
-            lbShadow_8000EFEC();
-            HSD_JObjDisp(jobj, arg1, arg2, arg3);
-        }
-        if (jobj->flags & (arg2 << 28)) {
-            for (jobj = jobj->child; jobj != NULL; jobj = jobj->next) {
-                grDisplay_801C5B90(jobj, arg1, arg2, arg3);
+
+            if (gp->x10_flags.b2 == 0 && Camera_80030A78() == false &&
+                Ground_801C1E18() != 0)
+            {
+                foggobj = Ground_801C1E84();
+                if (foggobj != NULL) {
+                    if (foggobj->hsd_obj != NULL) {
+                        HSD_FogSet(GET_FOG(foggobj));
+                    }
+                }
             }
         }
     }
diff --git a/src/melee/gr/ground.c b/src/melee/gr/ground.c
index 9b1d77183a..9f352dc5f2 100644
--- a/src/melee/gr/ground.c
+++ b/src/melee/gr/ground.c
@@ -1046,7 +1046,7 @@ void Ground_801C1E2C(HSD_GObj* gobj, int code)
     }
 }
 
-void* Ground_801C1E84(void)
+HSD_GObj* Ground_801C1E84(void)
 {
     return stage_info.x12C;
 }
diff --git a/src/melee/gr/ground.h b/src/melee/gr/ground.h
index 56a6deaac9..3e2ebc93be 100644
--- a/src/melee/gr/ground.h
+++ b/src/melee/gr/ground.h
@@ -58,7 +58,7 @@
 /* 1C1DE4 */ void Ground_801C1DE4(s32*, s32*);
 /* 1C1E00 */ void Ground_801C1E00(s32);
 /* 1C1E18 */ s32 Ground_801C1E18(void);
-/* 1C1E84 */ void* Ground_801C1E84(void);
+/* 1C1E84 */ HSD_GObj* Ground_801C1E84(void);
 /* 1C1FFC */ void Ground_ApplyStageBackgroundColor(void);
 /* 1C205C */ void Ground_801C205C(GXColor* color);
 /* 1C2090 */ bool Ground_801C2090(GXColor* color);
diff --git a/src/sysdolphin/baselib/jobj.c b/src/sysdolphin/baselib/jobj.c
index d4a2f777ac..3b2a4b0cea 100644
--- a/src/sysdolphin/baselib/jobj.c
+++ b/src/sysdolphin/baselib/jobj.c
@@ -567,11 +567,6 @@ void HSD_JObjAnimAll(HSD_JObj* jobj)
     }
 }
 
-inline MtxPtr HSD_CObjGetViewMtx(HSD_CObj* cobj)
-{
-    return cobj->view_mtx;
-}
-
 void HSD_JObjDispAll(HSD_JObj* jobj, Mtx vmtx, u32 flags, u32 rendermode)
 {
     MtxPtr new_var = vmtx;
@@ -589,7 +584,7 @@ void HSD_JObjDispAll(HSD_JObj* jobj, Mtx vmtx, u32 flags, u32 rendermode)
                 PSMTXConcat(jobj->mtx, mtx, mtx);
                 cobj = HSD_CObjGetCurrent();
                 HSD_ASSERT(0x355, cobj);
-                PSMTXConcat(HSD_CObjGetViewMtx(cobj), mtx, mtx);
+                PSMTXConcat(HSD_CObjGetViewingMtxPtrDirect(cobj), mtx, mtx);
                 HSD_JObjDispAll(jobj->child, mtx, flags, rendermode);
             }
         } else {
