diff --git a/src/melee/gr/grkinokoroute.c b/src/melee/gr/grkinokoroute.c
index 655f49f38b..76b83b6374 100644
--- a/src/melee/gr/grkinokoroute.c
+++ b/src/melee/gr/grkinokoroute.c
@@ -360,7 +360,8 @@ void grKinokoRoute_80207C88(Ground_GObj* gobj)
         }
         Ground_801C3D44(NULL, 30.0f, 10000.0f);
         if (gp->gv.kinokoroute2.phase < 2) {
-            gp->gv.kinokoroute2.spawn_idx = Ground_801C3DB4(NULL, 60.0f, 10000.0f);
+            gp->gv.kinokoroute2.spawn_idx =
+                Ground_801C3DB4(NULL, 60.0f, 10000.0f);
             if (gp->gv.kinokoroute2.spawn_idx != -1) {
                 gp->gv.kinokoroute2.phase = 2;
                 gp->gv.kinokoroute2.cam_timer = 0x3C;
diff --git a/src/melee/gr/groldkongo.c b/src/melee/gr/groldkongo.c
index 0f53dad564..4cd575c451 100644
--- a/src/melee/gr/groldkongo.c
+++ b/src/melee/gr/groldkongo.c
@@ -306,8 +306,7 @@ void grOldKongo_8020F888(Ground_GObj* arg0)
     case 3:
     default:
         temp_f3 = temp_r31->gv.arwing.xE0.y;
-        temp_f31 =
-            0.5f * (temp_f3 * (temp_f3 / DegToRad(grOk_804D6A90->x1C)));
+        temp_f31 = 0.5f * (temp_f3 * (temp_f3 / DegToRad(grOk_804D6A90->x1C)));
         if (temp_f3 > 0.0f) {
             var_f30 = temp_r31->gv.oldkongo.xD8 - temp_r31->gv.arwing.xDC;
         } else if (temp_f3 < 0.0f) {
@@ -676,8 +675,8 @@ s32 grOldKongo_80210454(HSD_GObj* arg0, HSD_GObj* arg1)
     ftLib_80086644(arg1, &pos_ft);
 
     if (!((pos_gnd.x - pos_ft.x) * (pos_gnd.x - pos_ft.x) +
-          (pos_gnd.y - pos_ft.y) * (pos_gnd.y - pos_ft.y) +
-          (pos_gnd.z - pos_ft.z) * (pos_gnd.z - pos_ft.z) <
+              (pos_gnd.y - pos_ft.y) * (pos_gnd.y - pos_ft.y) +
+              (pos_gnd.z - pos_ft.z) * (pos_gnd.z - pos_ft.z) <
           grOk_804D6A90->x10 * grOk_804D6A90->x10))
     {
         goto done;
@@ -765,8 +764,7 @@ f32 grOldKongo_80210650(void)
                             if (temp_r3_7 < 0) {
                                 var_f31 = -1.5707964f;
                             } else {
-                                s32 temp_r3_8 =
-                                    temp_r3_7 - grOk_804D6A90->x38;
+                                s32 temp_r3_8 = temp_r3_7 - grOk_804D6A90->x38;
 
                                 if (temp_r3_8 < 0) {
                                     var_f31 = -2.3561945f;
diff --git a/src/melee/gr/groldpupupu.c b/src/melee/gr/groldpupupu.c
index 5de8092d9a..5fc15f746f 100644
--- a/src/melee/gr/groldpupupu.c
+++ b/src/melee/gr/groldpupupu.c
@@ -1,21 +1,31 @@
 #include "groldpupupu.h"
 
+#include "placeholder.h"
+
 #include "baselib/forward.h"
+
+#include "cm/camera.h"
+
 #include "forward.h"
 
+#include "ft/ftdevice.h"
 #include "ft/ftlib.h"
 #include "gr/granime.h"
+#include "gr/grdisplay.h"
 #include "gr/grmaterial.h"
 #include "gr/ground.h"
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
+#include "lb/lbaudio_ax.h"
 
+#include <baselib/debug.h>
 #include <baselib/gobj.h>
+#include <baselib/gobjgxlink.h>
+#include <baselib/gobjproc.h>
 #include <baselib/jobj.h>
 #include <baselib/random.h>
-#include "ft/ftdevice.h"
 
 static struct {
     s16 x0;
@@ -30,6 +40,8 @@ static struct {
     f32 x20;
     f32 x24;
     f32 x28;
+    f32 x2C;
+    f32 x30;
 }* grOp_804D6A98;
 
 static void* grOp_804D6A9C;
@@ -37,9 +49,18 @@ static void* grOp_804D6A9C;
 static int grOp_804D48A0[1][2] = { { 4, 0 } };
 static int grOp_804D48A8[1][2] = { { 5, 1 } };
 static int grOp_804D48B0[1][2] = { { 2, 3 } };
+static int grOp_804D48C8[1][2] = { { 4, 0 } };
+static int grOp_804D48D0[1][2] = { { 5, 1 } };
+static int grOp_804D48D8[1][2] = { { 2, 3 } };
 
 static int grOp_804D48E0[1][2] = { { 0, 0 } };
 
+typedef struct grOldPupupuSpawnDesc {
+    s8 x0;
+    s8 x1;
+    s16 x2;
+} grOldPupupuSpawnDesc;
+
 StageCallbacks grOp_803E6688[9] = {
     { grOldPupupu_8021099C, grOldPupupu_802109C8, grOldPupupu_802109D0,
       grOldPupupu_802109D4, 0 },
@@ -61,6 +82,40 @@ StageCallbacks grOp_803E6688[9] = {
       grOldPupupu_8021110C, 0 },
 };
 
+char grOp_803E673C[] = "/GrOp.dat";
+
+StageData grOp_803E6748 = {
+    OLDPUPUPU,
+    grOp_803E6688,
+    grOp_803E673C,
+    grOldPupupu_802107E0,
+    grOldPupupu_802107DC,
+    grOldPupupu_80210884,
+    grOldPupupu_80210888,
+    grOldPupupu_802108AC,
+    grOldPupupu_80211CA0,
+    grOldPupupu_80211CA8,
+    1,
+    0,
+    0,
+};
+
+static grOldPupupuSpawnDesc grOp_803E67B0[10] = {
+    { -1, 1, 1 },
+    { 1, 1, 3 },
+    { 1, 1, 5 },
+    { -1, 1, 7 },
+    { -1, 0, 9 },
+    { 1, 0, 11 },
+    { -1, 0, 13 },
+    { 1, 0, 15 },
+    { -1, 0, 17 },
+    { 1, 0, 19 },
+};
+
+static int grOp_803E67D8[3] = { 0, 1, 2 };
+static Vec3 grOp_803E67E4[2] = { { -1.0F, 0.0F, 0.0F }, { 1.0F, 0.0F, 0.0F } };
+
 void grOldPupupu_802107DC(bool arg) {}
 
 void grOldPupupu_802107E0(void)
@@ -94,6 +149,15 @@ bool grOldPupupu_802108AC(void)
 {
     return false;
 }
+#pragma push
+#pragma dont_inline on
+HSD_GObj* grOldPupupu_802108B4_noinline(int arg0);
+HSD_GObj* grOldPupupu_802108B4_noinline(int arg0)
+{
+    return grOldPupupu_802108B4(arg0);
+}
+#pragma dont_inline reset
+#pragma pop
 
 HSD_GObj* grOldPupupu_802108B4(int arg0)
 {
@@ -130,10 +194,10 @@ void grOldPupupu_802109D4(Ground_GObj* arg) {}
 void grOldPupupu_802109D8(Ground_GObj* gobj)
 {
     Ground* gp = GET_GROUND(gobj);
-    gp->gv.unk.xC4 = 0;
-    gp->gv.unk.xC8 = 0;
-    gp->gv.unk.xCC = 0;
-    gp->gv.unk.xD0 = 1;
+    gp->gv.oldpupupu.xC4 = 0;
+    gp->gv.oldpupupu.xC8 = 0;
+    gp->gv.oldpupupu.xCC = 0;
+    gp->gv.oldpupupu.xD0 = 1;
     grAnime_801C8138(gobj, gp->map_id, 0);
 }
 
@@ -146,23 +210,23 @@ void grOldPupupu_80210A24(Ground_GObj* gobj)
 {
     Ground* gp = GET_GROUND(gobj);
 
-    if ((int) grOp_804D6A9C == 0 && gp->gv.unk.xD0 == 1) {
-        gp->gv.unk.xD0 = 0;
-        switch (gp->gv.unk.xC8) {
+    if ((int) grOp_804D6A9C == 0 && gp->gv.oldpupupu.xD0 == 1) {
+        gp->gv.oldpupupu.xD0 = 0;
+        switch (gp->gv.oldpupupu.xC8) {
         case 0:
             grAnime_801C8138(
                 gobj, gp->map_id,
-                grOp_804D48A0[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+                grOp_804D48A0[gp->gv.oldpupupu.xCC][gp->gv.oldpupupu.xD4]);
             break;
         case 1:
             grAnime_801C8138(
                 gobj, gp->map_id,
-                grOp_804D48B0[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+                grOp_804D48B0[gp->gv.oldpupupu.xCC][gp->gv.oldpupupu.xD4]);
             break;
         case 2:
             grAnime_801C8138(
                 gobj, gp->map_id,
-                grOp_804D48A8[gp->gv.unk.xCC][gp->gv.unk.xD4]);
+                grOp_804D48A8[gp->gv.oldpupupu.xCC][gp->gv.oldpupupu.xD4]);
             break;
         }
     }
@@ -237,7 +301,7 @@ void grOldPupupu_80210C7C(Ground_GObj* gobj)
 
     x2 = grOp_804D6A98->x2;
     x0 = grOp_804D6A98->x0;
-    *(s16*) &gp->gv.unk.xC4 =
+    gp->gv.oldpupupu2.xC4 =
         x2 > x0
             ? x0 + (x2 - x0 != 0 ? HSD_Randi(x2 - x0) : 0)
         : x2 < x0
@@ -250,7 +314,90 @@ bool grOldPupupu_80210D08(Ground_GObj* gobj)
     return false;
 }
 
-/// #grOldPupupu_80210D10
+void grOldPupupu_80210D10(Ground_GObj* gobj)
+{
+    f32 cam_left;
+    f32 cam_right;
+    f32 cam_center;
+    Ground* gp;
+    HSD_GObj* spawn;
+    HSD_JObj* jobj;
+    f32 direction;
+    f32 y;
+    f32 x;
+    f32 step;
+    s16 timer;
+    s16 min_respawn;
+    s16 respawn;
+    s32 count;
+    s32 i;
+    s32 index;
+
+    PAD_STACK(0x10);
+    gp = GET_GROUND(gobj);
+    timer = gp->gv.oldpupupu2.xC4;
+    gp->gv.oldpupupu2.xC4 = timer - 1;
+    if (timer < 0) {
+        direction = Ground_801C0498();
+        index = HSD_Randi(10);
+        if (grOp_803E67B0[index].x1 != 0 && HSD_Randi(5) == 0) {
+            count = 3;
+        } else {
+            count = 1;
+        }
+        Camera_800307D0(&cam_left, &cam_center, &cam_right);
+        if (grOp_803E67B0[index].x0 == 1) {
+            step = -10.0F;
+            if (cam_right < 200.0F) {
+                cam_right = 200.0F;
+            }
+            x = -(((f32) count * -10.0F) - (50.0F + cam_right));
+        } else {
+            step = 10.0F;
+            if (cam_left > -200.0F) {
+                cam_left = -200.0F;
+            }
+            x = -(((f32) count * 10.0F) - (cam_left - 50.0F));
+        }
+        y = direction *
+            (((f32) grOp_804D6A98->x4 * ((2.0F * HSD_Randf()) - 1.0F)) +
+             20.0F);
+        for (i = 0; i < count; i++) {
+            spawn = grOldPupupu_802108B4_noinline(2);
+            if (spawn != NULL) {
+                jobj = spawn->hsd_obj;
+                HSD_ASSERT(0x216, jobj);
+                HSD_JObjSetTranslateX(jobj, direction * x);
+                HSD_JObjSetTranslateY(jobj, y);
+                HSD_JObjSetTranslateZ(jobj, -150.0F * direction);
+                grAnime_801C7FF8(spawn, 0, 7, 0, 40.0F * (f32) i, 1.0F);
+                HSD_JObjClearFlagsAll(
+                    Ground_801C3FA4(spawn, grOp_803E67B0[index].x2),
+                    JOBJ_HIDDEN);
+            }
+            x += step;
+        }
+
+        respawn = grOp_804D6A98->x2;
+        min_respawn = grOp_804D6A98->x0;
+        if (respawn > min_respawn) {
+            s32 range = respawn - min_respawn;
+
+            if (range != 0) {
+                respawn = min_respawn + HSD_Randi(range);
+            } else {
+                respawn = min_respawn;
+            }
+        } else if (respawn < min_respawn) {
+            s32 range = min_respawn - respawn;
+
+            if (range != 0) {
+                respawn += HSD_Randi(range);
+            }
+        }
+        gp->gv.oldpupupu2.xC4 = respawn;
+    }
+}
 
 void grOldPupupu_8021110C(Ground_GObj* arg) {}
 
@@ -281,9 +428,9 @@ void grOldPupupu_8021119C(Ground_GObj* gobj)
     s32 min;
 
     ftCo_800C06E8(gobj, 0xA, fn_802112F4);
-    gp->gv.unk.xC4 = 0;
-    gp->gv.unk.xC8 = 0;
-    gp->gv.unk.xCC = 0;
+    gp->gv.oldpupupu.xC4 = 0;
+    gp->gv.oldpupupu.xC8 = 0;
+    gp->gv.oldpupupu.xCC = 0;
 
     max = grOp_804D6A98->xC;
     min = grOp_804D6A98->x8;
@@ -302,10 +449,10 @@ void grOldPupupu_8021119C(Ground_GObj* gobj)
         }
     }
 
-    gp->gv.unk.xD0 = max;
-    gp->gv.unk.xD4 = 1;
-    gp->gv.unk.xE0 = 0;
-    gp->gv.unk.xD8 = 1;
+    gp->gv.oldpupupu.xD0 = max;
+    gp->gv.oldpupupu.xD4 = 1;
+    gp->gv.oldpupupu.xE0 = 0;
+    gp->gv.oldpupupu.xD8 = 1;
     grAnime_801C8138(gobj, gp->map_id, 0);
     gp->x11_flags.b012 = 1;
 }
@@ -341,17 +488,15 @@ bool grOldPupupu_8021128C(Vec* vec, f32 maxX, f32 minX, f32 maxY, f32 minY)
 
 bool fn_802112F4(Ground_GObj* gobj, HSD_GObj* fighter_gobj, Vec3* vel)
 {
-    Ground* gp;
-    void* new_var;
     Vec3 pos;
-    PAD_STACK(8);
+    Ground* gp;
 
     ftLib_80086644(fighter_gobj, &pos);
     vel->y = 0.0F;
     vel->z = 0.0F;
 
-    gp = (new_var = gobj->user_data);
-    switch (gp->gv.unk.xDC) {
+    gp = GET_GROUND(gobj);
+    switch (gp->gv.oldpupupu.xDC) {
     case 1:
         if (grOldPupupu_8021128C(&pos, grOp_804D6A98->x1C, grOp_804D6A98->x20,
                                  grOp_804D6A98->x24,
@@ -376,7 +521,216 @@ bool fn_802112F4(Ground_GObj* gobj, HSD_GObj* fighter_gobj, Vec3* vel)
     return false;
 }
 
-/// #grOldPupupu_802113E0
+void grOldPupupu_802113E0(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+
+    if (grOp_804D6A9C == 0) {
+        gp->gv.oldpupupu.xE0 += 1;
+        gp->gv.oldpupupu.xDC = 0;
+
+        switch (gp->gv.oldpupupu.xC8) {
+        case 0:
+            if (gp->gv.oldpupupu.xD4 != 0) {
+                Ground* other_gp;
+
+                gp->gv.oldpupupu.xD4 = 0;
+                gp->gv.oldpupupu.xCC = 0;
+                gp->gv.oldpupupu.xD0 =
+                    rand_range(grOp_804D6A98->xC, grOp_804D6A98->x8);
+                grAnime_801C8138(
+                    gobj, gp->map_id,
+                    grOp_804D48C8[gp->gv.oldpupupu.xCC][gp->gv.oldpupupu.xD8]);
+                other_gp = Ground_801C2BA4(1)->user_data;
+                other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                other_gp->gv.oldpupupu.xCC = 0;
+                other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                other_gp->gv.oldpupupu.xD0 = 1;
+                gp->gv.oldpupupu.xCC =
+                    rand_range((s32) grOp_804D6A98->x2C, (s32) grOp_804D6A98->x30);
+            }
+
+            gp->gv.oldpupupu.xD0 -= 1;
+            if (grAnime_801C83D0(gobj, 0, 7) != 0) {
+                if (gp->gv.oldpupupu.xD0 > 0) {
+                    Ground* other_gp;
+
+                    gp->gv.oldpupupu.xCC -= 1;
+                    if (gp->gv.oldpupupu.xCC <= 0) {
+                        gp->gv.oldpupupu.xCC = 0;
+                        grAnime_801C8138(
+                            gobj, gp->map_id,
+                            grOp_804D48C8[gp->gv.oldpupupu.xCC]
+                                         [gp->gv.oldpupupu.xD8]);
+                        gp->gv.oldpupupu.xCC =
+                            rand_range((s32) grOp_804D6A98->x2C,
+                                       (s32) grOp_804D6A98->x30);
+                    }
+                    other_gp = Ground_801C2BA4(1)->user_data;
+                    other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                    other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                    other_gp->gv.oldpupupu.xCC = 0;
+                    other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                    other_gp->gv.oldpupupu.xD0 = 1;
+                    return;
+                }
+
+                {
+                    Ground* other_gp;
+
+                    gp->gv.oldpupupu.xC4 += 1;
+                    if (gp->gv.oldpupupu.xC4 >= 3) {
+                        gp->gv.oldpupupu.xC4 = 0;
+                    }
+                    gp->gv.oldpupupu.xC8 =
+                        grOp_803E67D8[gp->gv.oldpupupu.xC4];
+                    gp->gv.oldpupupu.xD4 = 1;
+                    other_gp = Ground_801C2BA4(1)->user_data;
+                    other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                    other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                    other_gp->gv.oldpupupu.xCC = 0;
+                    other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                    other_gp->gv.oldpupupu.xD0 = 1;
+                    return;
+                }
+            }
+            break;
+
+        case 1:
+            if (gp->gv.oldpupupu.xD4 != 0) {
+                Vec3 pos;
+                s32 side;
+                Ground* other_gp;
+
+                gp->gv.oldpupupu.xCC = 0;
+                HSD_JObjGetTranslation(gobj->hsd_obj, &pos);
+                side = ftLib_800864A8(&pos, NULL) == 1.0F;
+                if ((f32) gp->gv.oldpupupu.xD8 == (f32) side) {
+                    gp->gv.oldpupupu.xC8 = 2;
+                    gp->gv.oldpupupu.xD4 = 0;
+                    gp->gv.oldpupupu.xCC = 0;
+                    gp->gv.oldpupupu.xD0 = 0;
+                    grAnime_801C8138(
+                        gobj, gp->map_id,
+                        grOp_804D48D0[gp->gv.oldpupupu.xCC]
+                                     [gp->gv.oldpupupu.xD8]);
+                    gp->gv.oldpupupu.xC4 += 1;
+                    if (gp->gv.oldpupupu.xC4 >= 3) {
+                        gp->gv.oldpupupu.xC4 = 0;
+                    }
+                    gp->gv.oldpupupu.xC8 =
+                        grOp_803E67D8[gp->gv.oldpupupu.xC4];
+                    gp->gv.oldpupupu.xD4 = 1;
+                    other_gp = Ground_801C2BA4(1)->user_data;
+                    other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                    other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                    other_gp->gv.oldpupupu.xCC = 0;
+                    other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                    other_gp->gv.oldpupupu.xD0 = 1;
+                } else {
+                    other_gp = Ground_801C2BA4(1)->user_data;
+                    gp->gv.oldpupupu.xD8 = side;
+                    gp->gv.oldpupupu.xD4 = 0;
+                    gp->gv.oldpupupu.xCC = 0;
+                    gp->gv.oldpupupu.xD0 = 0;
+                    grAnime_801C8138(
+                        gobj, gp->map_id,
+                        grOp_804D48D8[gp->gv.oldpupupu.xCC]
+                                     [gp->gv.oldpupupu.xD8]);
+                    other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                    other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                    other_gp->gv.oldpupupu.xCC = 0;
+                    other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                    other_gp->gv.oldpupupu.xD0 = 1;
+                }
+            }
+
+            if (gp->gv.oldpupupu.xC8 == 1 && grAnime_801C83D0(gobj, 0, 7) != 0)
+            {
+                Ground* other_gp;
+
+                gp->gv.oldpupupu.xC4 += 1;
+                if (gp->gv.oldpupupu.xC4 >= 3) {
+                    gp->gv.oldpupupu.xC4 = 0;
+                }
+                gp->gv.oldpupupu.xC8 = grOp_803E67D8[gp->gv.oldpupupu.xC4];
+                gp->gv.oldpupupu.xD4 = 1;
+                other_gp = Ground_801C2BA4(1)->user_data;
+                other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                other_gp->gv.oldpupupu.xCC = 0;
+                other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                other_gp->gv.oldpupupu.xD0 = 1;
+                return;
+            }
+            break;
+
+        case 2:
+            if (gp->gv.oldpupupu.xD4 != 0) {
+                Ground* other_gp;
+
+                gp->gv.oldpupupu.xD4 = 0;
+                gp->gv.oldpupupu.xCC = 0;
+                gp->gv.oldpupupu.xD0 = 0;
+                grAnime_801C8138(
+                    gobj, gp->map_id,
+                    grOp_804D48D0[gp->gv.oldpupupu.xCC][gp->gv.oldpupupu.xD8]);
+                other_gp = Ground_801C2BA4(1)->user_data;
+                other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                other_gp->gv.oldpupupu.xCC = 0;
+                other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                other_gp->gv.oldpupupu.xD0 = 1;
+            }
+
+            if (grAnime_801C83D0(gobj, 0, 7) == 0) {
+                gp->gv.oldpupupu.xD0 += 1;
+                if (gp->gv.oldpupupu.xD0 == 0x2D) {
+                    lbAudioAx_800237A8(
+                        gp->gv.oldpupupu.xD8 == 0 ? 0x72BF0 : 0x72BF1, 0x7F,
+                        0x40);
+                }
+                if (gp->gv.oldpupupu.xD0 > 0x2D &&
+                    gp->gv.oldpupupu.xD0 < 0x140)
+                {
+                    Camera_80030E44(1, NULL);
+                    gp->gv.oldpupupu.xDC = gp->gv.oldpupupu.xD8 + 1;
+                    if ((gp->gv.oldpupupu.xD0 % 10) == 0) {
+                        if (gp->gv.oldpupupu.xD8 == 0) {
+                            lb_80011A50(&grOp_803E67E4[0], 0xF, 0.5F, 0.0F,
+                                        0.0F, grOp_804D6A98->x20,
+                                        grOp_804D6A98->x24, grOp_804D6A98->x1C,
+                                        grOp_804D6A98->x28);
+                        } else {
+                            lb_80011A50(&grOp_803E67E4[1], 0xF, 0.5F, 0.0F,
+                                        0.0F, grOp_804D6A98->x14,
+                                        grOp_804D6A98->x24, grOp_804D6A98->x18,
+                                        grOp_804D6A98->x28);
+                        }
+                        return;
+                    }
+                }
+            } else {
+                Ground* other_gp;
+
+                gp->gv.oldpupupu.xC4 += 1;
+                if (gp->gv.oldpupupu.xC4 >= 3) {
+                    gp->gv.oldpupupu.xC4 = 0;
+                }
+                gp->gv.oldpupupu.xC8 = grOp_803E67D8[gp->gv.oldpupupu.xC4];
+                gp->gv.oldpupupu.xD4 = 1;
+                other_gp = Ground_801C2BA4(1)->user_data;
+                other_gp->gv.oldpupupu.xC4 = gp->gv.oldpupupu.xC4;
+                other_gp->gv.oldpupupu.xC8 = gp->gv.oldpupupu.xC8;
+                other_gp->gv.oldpupupu.xCC = 0;
+                other_gp->gv.oldpupupu.xD4 = gp->gv.oldpupupu.xD8;
+                other_gp->gv.oldpupupu.xD0 = 1;
+            }
+            break;
+        }
+    }
+}
 
 void grOldPupupu_80211C0C(Ground_GObj* arg) {}
 
@@ -393,10 +747,11 @@ void grOldPupupu_80211C1C(Ground_GObj* gobj)
     Ground* other_gp = Ground_801C2BA4(7)->user_data;
     PAD_STACK(8);
 
-    if (other_gp->gv.unk.xC8 == 2 && other_gp->gv.unk.xD4 != 0) {
+    if (other_gp->gv.oldpupupu.xC8 == 2 && other_gp->gv.oldpupupu.xD4 != 0) {
         grAnime_801C8138(
             gobj, gp->map_id,
-            grOp_804D48E0[other_gp->gv.unk.xCC][other_gp->gv.unk.xD8]);
+            grOp_804D48E0[other_gp->gv.oldpupupu.xCC]
+                         [other_gp->gv.oldpupupu.xD8]);
     }
 }
 
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index f61387f845..2230e38e84 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -742,6 +742,22 @@ struct grOldKongo_GroundVars {
     f32 xEC;
 };
 
+struct grOldPupupu_GroundVars {
+    s32 xC4;
+    s32 xC8;
+    s32 xCC;
+    s32 xD0;
+    s32 xD4;
+    s32 xD8;
+    s32 xDC;
+    s32 xE0;
+};
+
+struct grOldPupupu_GroundVars2 {
+    s16 xC4;
+    s16 xC6;
+};
+
 /// likely for Cathrine (Birdo)
 struct grInishie2_GroundVars2 {
     Item_GObj* xC4;
@@ -1600,6 +1616,8 @@ struct Ground {
             struct grInishie2_GroundVars2 inishie22;
             struct grInishie2_GroundVars3 inishie23;
             struct grOldKongo_GroundVars oldkongo;
+            struct grOldPupupu_GroundVars oldpupupu;
+            struct grOldPupupu_GroundVars2 oldpupupu2;
             struct grOldYoshi_Cloud_GroundVars oldyoshicloud;
             struct grOldYoshi_Guest_GroundVars oldyoshiguest;
             struct GroundVars_izumi izumi;
