diff --git a/src/melee/gr/gryorster.c b/src/melee/gr/gryorster.c
index d196bd1fca..08fd9f638d 100644
--- a/src/melee/gr/gryorster.c
+++ b/src/melee/gr/gryorster.c
@@ -6,15 +6,20 @@
 
 #include "ft/ftlib.h"
 #include "gr/grdisplay.h"
+#include "gr/grlib.h"
+#include "gr/grmaterial.h"
 #include "gr/ground.h"
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
+#include "it/it_26B1.h"
+#include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
 #include "mp/mplib.h"
 
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
+#include <baselib/jobj.h>
 
 StageCallbacks grYt_803E5198[2] = {
     { grYorster_80202124, grYorster_80202150, grYorster_80202158,
@@ -40,12 +45,19 @@ StageData grYt_803E51CC = {
 };
 
 typedef struct YorsterParams {
-    u8 pad_00[0x14];
-    int x14;
+    f32 x00;
+    f32 x04;
+    f32 x08;
+    f32 x0C;
+    s32 x10;
+    s32 x14;
+    s32 x18;
+    s32 x1C;
 } YorsterParams;
 
 typedef struct grYt_804D6A20_t {
     YorsterParams* x0;
+    u8 pad_04[4];
 } grYt_804D6A20_t;
 grYt_804D6A20_t grYt_804D6A20;
 
@@ -169,7 +181,35 @@ void grYorster_80202254(Ground_GObj* gobj)
 
 void grYorster_802022A0(HSD_GObj* gobj) {}
 
-void grYorster_802022A4(HSD_GObj* gobj) {}
+void grYorster_802022A4(HSD_GObj* gobj)
+{
+    s32 joints[9] = { 0xA, 0xB, 0xC, 0xD, 0xE, 0xF, 0x10, 0x11, 0x12 };
+    Ground* temp_r28;
+    int i;
+
+    temp_r28 = gobj->user_data;
+    for (i = 0; i < 9; i++) {
+        temp_r28->gv.yorster.elements[i].x00 = (s8) i;
+        temp_r28->gv.yorster.elements[i].x01 = 0;
+        temp_r28->gv.yorster.elements[i].x04 = 0.0f;
+        temp_r28->gv.yorster.elements[i].x08 = 0.0f;
+        temp_r28->gv.yorster.elements[i].x14 = joints[i];
+        temp_r28->gv.yorster.elements[i].x18 =
+            Ground_801C3FA4(gobj, joints[i]);
+        temp_r28->gv.yorster.elements[i].x1C = grMaterial_801C8CFC(
+            0, 0, temp_r28, temp_r28->gv.yorster.elements[i].x18, NULL,
+            grYorster_80202428, NULL);
+        grMaterial_801C8E68(temp_r28->gv.yorster.elements[i].x1C, GA_Ground);
+        mpJointSetCb2(Ground_801C32D4(temp_r28->map_id,
+                                      temp_r28->gv.yorster.elements[i].x14),
+                      temp_r28, grYorster_802024F0);
+        grMaterial_801C8DE0(temp_r28->gv.yorster.elements[i].x1C, 0.0f, 0.0f,
+                            0.0f, 0.0f, 0.0f, 0.0f, 5.0f);
+        grMaterial_801C8E08(temp_r28->gv.yorster.elements[i].x1C);
+        grAnime_801C7FF8(gobj, temp_r28->gv.yorster.elements[i].x14, 7, 1,
+                         0.0f, 0.0f);
+    }
+}
 
 void grYorster_80202428(HSD_GObj* item_gobj, Ground* gp, Vec3* pos,
                         HSD_GObj* fighter_gobj, f32 value)
@@ -177,20 +217,182 @@ void grYorster_80202428(HSD_GObj* item_gobj, Ground* gp, Vec3* pos,
     int i;
 
     if (ftLib_80086960(fighter_gobj)) {
-        ftLib_80086A4C(fighter_gobj, (f32) grYt_804D6A20.x0->x14);
+        ftLib_80086A4C(fighter_gobj, grYt_804D6A20.x0->x14);
+    }
+
+    for (i = 0; i < 9; i++) {
+        if (item_gobj == gp->gv.yorster.elements[i].x1C) {
+            gp->gv.yorster.elements[i].x04 += value;
+            break;
+        }
+    }
+}
+
+void grYorster_802024F0(Ground* gp, s32 joint_id, CollData* coll_data,
+                        s32 unused, mpLib_GroundEnum ground_kind, f32 value)
+{
+    HSD_GObj* gobj = coll_data->x0_gobj;
+    s32 env = coll_data->x34_flags.b1234;
+    s32 i;
+    PAD_STACK(8);
+
+    if (1 != env && env != 2 && env != 3) {
+        return;
+    }
+
+    if (env == 1 && ftLib_800873F4((0, coll_data->x0_gobj))) {
+        return;
+    }
+
+    switch (env) {
+    case 1: {
+        Vec3 pos;
+
+        ftLib_80086BEC(gobj, &pos);
+        value = pos.y;
+        if (value >= grYt_804D6A20.x0->x00) {
+            ftLib_80086A4C(gobj, (f32) grYt_804D6A20.x0->x10);
+        }
+        break;
+    }
+    case 2:
+    case 3:
+        if (value >= grYt_804D6A20.x0->x00) {
+            it_8026B718(gobj, (f32) grYt_804D6A20.x0->x10);
+        }
+        break;
     }
 
     for (i = 0; i < 9; i++) {
-        if (gp->gv.yorster.elements[i].x1C == item_gobj) {
-            gp->gv.yorster.elements[i].x4 += value;
+        if (joint_id ==
+            Ground_801C32D4(gp->map_id, gp->gv.yorster.elements[i].x14))
+        {
+            gp->gv.yorster.elements[i].x08 += value;
             break;
         }
     }
 }
 
-/// #grYorster_802024F0
+void grYorster_8020266C(HSD_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    s32 i;
 
-/// #grYorster_8020266C
+    for (i = 0; i < 9; i++) {
+        struct grYorster_TrackElement* entry = &gp->gv.yorster.elements[i];
+        s32 joint = Ground_801C32D4(gp->map_id, entry->x14);
+
+        switch ((s8) entry->x01) {
+        default: {
+            s32 advance = 0;
+
+            if (entry->x04 >= grYt_804D6A20.x0->x08) {
+                s32 frames_left;
+                f32 end_frame = (f32) grYt_804D6A20.x0->x1C;
+
+                grMaterial_801C8E28(entry->x1C);
+                entry->x04 -= grYt_804D6A20.x0->x08;
+                if (entry->x04 >= grYt_804D6A20.x0->x0C) {
+                    frames_left = 0;
+                } else {
+                    frames_left = (s32) (grYt_804D6A20.x0->x0C - entry->x04);
+                }
+                if (end_frame >= lbGetJObjEndFrame(entry->x18)) {
+                    end_frame = lbGetJObjEndFrame(entry->x18);
+                }
+                advance = 1;
+                entry->x10 =
+                    (s32) ((f32) frames_left *
+                           (end_frame / grYt_804D6A20.x0->x0C));
+            } else if (entry->x08 >= grYt_804D6A20.x0->x00) {
+                s32 frames_left;
+                f32 end_frame = (f32) grYt_804D6A20.x0->x1C;
+
+                grMaterial_801C8E28(entry->x1C);
+                entry->x08 -= grYt_804D6A20.x0->x00;
+                if (entry->x08 <= grYt_804D6A20.x0->x00) {
+                    entry->x08 = grYt_804D6A20.x0->x00;
+                }
+                if (entry->x08 >= grYt_804D6A20.x0->x04) {
+                    frames_left = 0;
+                } else {
+                    frames_left =
+                        (s32) (10.0f *
+                               (grYt_804D6A20.x0->x04 - entry->x08));
+                }
+                if (end_frame >= lbGetJObjEndFrame(entry->x18)) {
+                    end_frame = lbGetJObjEndFrame(entry->x18);
+                }
+                advance = 1;
+                entry->x10 =
+                    (s32) ((f32) frames_left /
+                           (10.0f * grYt_804D6A20.x0->x04) * end_frame);
+                entry->x08 = 0.0f;
+                Ground_801C53EC(0xE);
+            }
+            if (advance != 0) {
+                entry->x0C = grYt_804D6A20.x0->x18;
+                entry->x01 = 2;
+            }
+            break;
+        }
+        case 2:
+            if (entry->x0C <= 0) {
+                grAnime_801C7FF8(gobj, entry->x14, 7, 0, (f32) entry->x10,
+                                 1.0f);
+                mpLib_80057BC0(joint);
+                entry->x01 = 3;
+                entry->x0C = entry->x10;
+            } else {
+                entry->x0C--;
+            }
+            break;
+        case 3:
+            if (entry->x0C >= 0x143) {
+                Vec3 pos;
+                HSD_JObjGetTranslation(entry->x18, &pos);
+
+                if (grLib_801C9EE8(&pos, 10.0f * Ground_801C0498() - 2.0f)) {
+                    grAnime_801C7FF8(gobj, entry->x14, 7, 2, 0.0f, 0.3f);
+                    entry->x0C = 0;
+                    entry->x01 = 4;
+                } else {
+                    grAnime_801C7FF8(gobj, entry->x14, 7, 1, 0.0f, 1.0f);
+                    mpJointListAdd(joint);
+                    grMaterial_801C8E08(entry->x1C);
+                    entry->x04 = 0.0f;
+                    entry->x08 = 0.0f;
+                    entry->x01 = 1;
+                    OSReport("*** End Frame = %d\n", entry->x0C);
+                }
+            } else {
+                entry->x0C++;
+            }
+            break;
+        case 4:
+            if (grAnime_801C84A4(gobj, entry->x14, 7) != 0 ||
+                entry->x0C >= 0x19)
+            {
+                Vec3 pos;
+
+                entry->x0C = 0;
+                HSD_JObjGetTranslation(entry->x18, &pos);
+                if (!grLib_801C9EE8(&pos, 10.0f * Ground_801C0498() * 0.5f)) {
+                    grAnime_801C7FF8(gobj, entry->x14, 7, 1, 0.0f, 1.0f);
+                    mpJointListAdd(joint);
+                    grMaterial_801C8E08(entry->x1C);
+                    entry->x04 = 0.0f;
+                    entry->x08 = 0.0f;
+                    entry->x01 = 1;
+                }
+            } else {
+                entry->x0C++;
+            }
+            break;
+        }
+        entry->x08 = 0.0f;
+    }
+}
 
 DynamicsDesc* grYorster_80202B5C(enum_t unused)
 {
diff --git a/src/melee/gr/gryorster.h b/src/melee/gr/gryorster.h
index 6ded410598..d47630befb 100644
--- a/src/melee/gr/gryorster.h
+++ b/src/melee/gr/gryorster.h
@@ -6,6 +6,7 @@
 #include "baselib/forward.h"
 #include "gr/forward.h"
 #include "lb/forward.h"
+#include "mp/forward.h"
 
 #include <dolphin/mtx.h>
 
@@ -27,7 +28,8 @@
 /* 2022A4 */ void grYorster_802022A4(Ground_GObj*);
 /* 202428 */ void grYorster_80202428(HSD_GObj*, Ground*, Vec3*, HSD_GObj*,
                                      f32);
-/* 2024F0 */ UNK_RET grYorster_802024F0(UNK_PARAMS);
+/* 2024F0 */ void grYorster_802024F0(Ground*, s32, CollData*, s32,
+                                     mpLib_GroundEnum, f32);
 /* 20266C */ void grYorster_8020266C(HSD_GObj*);
 /* 202B5C */ DynamicsDesc* grYorster_80202B5C(enum_t);
 /* 202B64 */ bool grYorster_80202B64(Vec3*, int _, HSD_JObj*);
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index 2ac17bcc09..9be2512908 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -770,9 +770,15 @@ struct grStadium_type9_GroundVars {
 };
 
 struct grYorster_TrackElement {
-    /* 0x00 */ u8 pad_00[4];
-    /* 0x04 */ f32 x4;
-    /* 0x08 */ u8 pad_08[0x14];
+    /* 0x00 */ s8 x00;
+    /* 0x01 */ u8 x01;
+    /* 0x02 */ u8 pad_02[2];
+    /* 0x04 */ f32 x04;
+    /* 0x08 */ f32 x08;
+    /* 0x0C */ s32 x0C;
+    /* 0x10 */ s32 x10;
+    /* 0x14 */ s32 x14;
+    /* 0x18 */ HSD_JObj* x18;
     /* 0x1C */ HSD_GObj* x1C;
 };
 
