diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 07c8bb1efd..1ba175a8c4 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -14632,7 +14632,7 @@ itLipstickspore_UnkMotion0_Phys = .text:0x8029A524; // type:function size:0x68 s
 itLipstickspore_UnkMotion0_Coll = .text:0x8029A58C; // type:function size:0x34 scope:global
 itLipstickspore_UnkMotion1_Anim = .text:0x8029A5C0; // type:function size:0x20 scope:global
 itLipstickspore_UnkMotion1_Phys = .text:0x8029A5E0; // type:function size:0x18 scope:global
-itLipstickspore_UnkMotion1_Coll = .text:0x8029A5F8; // type:function size:0x8 scope:global
+itLipstickspore_UnkMotion1_Coll = .text:0x8029A5F8; // type:function size:0x8 scope:local
 itLipstickSpore_Logic37_DmgDealt = .text:0x8029A600; // type:function size:0x8 scope:global
 itLipstickSpore_Logic37_Clanked = .text:0x8029A608; // type:function size:0x8 scope:global
 itLipstickSpore_Logic37_HitShield = .text:0x8029A610; // type:function size:0x8 scope:global
@@ -15180,7 +15180,7 @@ it_802B43B0 = .text:0x802B43B0; // type:function size:0x20 scope:global
 it_802B43D0 = .text:0x802B43D0; // type:function size:0x80 scope:global
 itPikachutjoltair_UnkMotion0_Anim = .text:0x802B4450; // type:function size:0x180 scope:global
 itPikachutjoltair_UnkMotion0_Phys = .text:0x802B45D0; // type:function size:0x18 scope:global
-itPikachutjoltair_UnkMotion0_Coll = .text:0x802B45E8; // type:function size:0x8 scope:global
+itPikachutjoltair_UnkMotion0_Coll = .text:0x802B45E8; // type:function size:0x8 scope:local
 it_2725_Logic107_DmgDealt = .text:0x802B45F0; // type:function size:0x48 scope:global
 it_2725_Logic107_Clanked = .text:0x802B4638; // type:function size:0x48 scope:global
 it_2725_Logic107_Absorbed = .text:0x802B4680; // type:function size:0x48 scope:global
@@ -20180,7 +20180,7 @@ it_803B86A4 = .rodata:0x803B86A4; // type:object size:0xC scope:global data:4byt
 it_803B86B0 = .rodata:0x803B86B0; // type:object size:0xC scope:global data:4byte
 it_803B86BC = .rodata:0x803B86BC; // type:object size:0xC scope:global data:4byte
 it_803B86C8 = .rodata:0x803B86C8; // type:object size:0xC scope:local data:float
-it_803B86D8 = .rodata:0x803B86D8; // type:object size:0x10 scope:global data:4byte
+zero_vec = .rodata:0x803B86D8; // type:object size:0xC scope:global data:4byte
 it_803B86E8 = .rodata:0x803B86E8; // type:object size:0x10 scope:global data:4byte
 @211 = .rodata:0x803B86F8; // type:object size:0xC scope:local data:4byte
 it_803B8708 = .rodata:0x803B8708; // type:object size:0x10 scope:global data:4byte
@@ -23672,7 +23672,7 @@ jumptable_803F6608 = .data:0x803F6608; // type:object size:0x28 scope:local
 it_803F6630 = .data:0x803F6630; // type:object size:0x10 scope:global
 it_803F6640 = .data:0x803F6640; // type:object size:0x40 scope:global
 it_803F6680 = .data:0x803F6680; // type:object size:0x20 scope:global
-it_803F66A0 = .data:0x803F66A0; // type:object size:0x25 scope:global data:string
+@240 = .data:0x803F66A0; // type:object size:0x25 scope:local data:string
 it_803F66C8 = .data:0x803F66C8; // type:object size:0x50 scope:global
 @254 = .data:0x803F6718; // type:object size:0xA scope:local data:string
 it_803F6728 = .data:0x803F6728; // type:object size:0x60 scope:global
@@ -23732,8 +23732,9 @@ it_803F7158 = .data:0x803F7158; // type:object size:0x10 scope:global
 it_803F7190 = .data:0x803F7190; // type:object size:0x20 scope:global
 it_803F71B0 = .data:0x803F71B0; // type:object size:0x25 scope:global data:string
 it_803F71D8 = .data:0x803F71D8; // type:object size:0x10 scope:global
-it_803F71E8 = .data:0x803F71E8; // type:object size:0xA scope:global data:string
-it_803F71F4 = .data:0x803F71F4; // type:object size:0x25 scope:global data:string
+...data.0 = .data:0x803F71D8; // type:label scope:local
+@219 = .data:0x803F71E8; // type:object size:0xA scope:local data:string
+@221 = .data:0x803F71F4; // type:object size:0x25 scope:local data:string
 it_803F7220 = .data:0x803F7220; // type:object size:0x40 scope:global
 @246 = .data:0x803F7260; // type:object size:0x25 scope:local data:string
 it_803F7288 = .data:0x803F7288; // type:object size:0x90 scope:global
@@ -23811,7 +23812,7 @@ it_803F8058 = .data:0x803F8058; // type:object size:0x70 scope:global
 it_803F80C8 = .data:0x803F80C8; // type:object size:0x30 scope:global
 it_803F80F8 = .data:0x803F80F8; // type:object size:0x30 scope:global
 it_803F8128 = .data:0x803F8128; // type:object size:0x30 scope:global
-it_803F8158 = .data:0x803F8158; // type:object size:0x25 scope:global data:string
+@167 = .data:0x803F8158; // type:object size:0x25 scope:local data:string
 it_803F8180 = .data:0x803F8180; // type:object size:0x10 scope:global
 it_803F8190 = .data:0x803F8190; // type:object size:0x70 scope:global
 it_803F8200 = .data:0x803F8200; // type:object size:0x50 scope:global
@@ -23825,7 +23826,7 @@ it_803F83E0 = .data:0x803F83E0; // type:object size:0x10 scope:global
 it_803F83F0 = .data:0x803F83F0; // type:object size:0x50 scope:global
 @258 = .data:0x803F8440; // type:object size:0x25 scope:local data:string
 it_803F8468 = .data:0x803F8468; // type:object size:0x160 scope:global
-it_803F85C8 = .data:0x803F85C8; // type:object size:0x25 scope:global data:string
+@497 = .data:0x803F85C8; // type:object size:0x25 scope:local data:string
 it_803F85F0 = .data:0x803F85F0; // type:object size:0xB0 scope:global
 @551 = .data:0x803F86A0; // type:object size:0x25 scope:local data:string
 it_803F86C8 = .data:0x803F86C8; // type:object size:0x128 scope:global
@@ -27103,8 +27104,8 @@ it_804D52D8 = .sdata:0x804D52D8; // type:object size:0x5 scope:global data:strin
 @174 = .sdata:0x804D5318; // type:object size:0x5 scope:local data:string
 it_804D5320 = .sdata:0x804D5320; // type:object size:0x7 scope:global data:string
 it_804D5328 = .sdata:0x804D5328; // type:object size:0x5 scope:global data:string
-it_804D5330 = .sdata:0x804D5330; // type:object size:0x7 scope:global data:string
-it_804D5338 = .sdata:0x804D5338; // type:object size:0x5 scope:global data:string
+@238 = .sdata:0x804D5330; // type:object size:0x7 scope:local data:string
+@239 = .sdata:0x804D5338; // type:object size:0x5 scope:local data:string
 @207 = .sdata:0x804D5340; // type:object size:0x7 scope:local data:string
 @208 = .sdata:0x804D5348; // type:object size:0x5 scope:local data:string
 @237 = .sdata:0x804D5350; // type:object size:0x7 scope:local data:string
@@ -27152,8 +27153,8 @@ it_804D5490 = .sdata:0x804D5490; // type:object size:0x5 scope:global data:strin
 @239 = .sdata:0x804D54A0; // type:object size:0x5 scope:local data:string
 it_804D54A8 = .sdata:0x804D54A8; // type:object size:0x7 scope:global data:string
 it_804D54B0 = .sdata:0x804D54B0; // type:object size:0x5 scope:global data:string
-it_804D54B8 = .sdata:0x804D54B8; // type:object size:0x7 scope:global data:string
-it_804D54C0 = .sdata:0x804D54C0; // type:object size:0x5 scope:global data:string
+@217 = .sdata:0x804D54B8; // type:object size:0x7 scope:local data:string
+@218 = .sdata:0x804D54C0; // type:object size:0x5 scope:local data:string
 @244 = .sdata:0x804D54C8; // type:object size:0x7 scope:local data:string
 @245 = .sdata:0x804D54D0; // type:object size:0x5 scope:local data:string
 it_804D54D8 = .sdata:0x804D54D8; // type:object size:0x7 scope:global data:string
@@ -27197,14 +27198,14 @@ it_804D55E8 = .sdata:0x804D55E8; // type:object size:0x5 scope:global data:strin
 @297 = .sdata:0x804D5608; // type:object size:0x5 scope:local data:string
 it_804D5610 = .sdata:0x804D5610; // type:object size:0x7 scope:global data:string
 it_804D5618 = .sdata:0x804D5618; // type:object size:0x5 scope:global data:string
-it_804D5620 = .sdata:0x804D5620; // type:object size:0x7 scope:global data:string
-it_804D5628 = .sdata:0x804D5628; // type:object size:0x5 scope:global data:string
+@165 = .sdata:0x804D5620; // type:object size:0x7 scope:local data:string
+@166 = .sdata:0x804D5628; // type:object size:0x5 scope:local data:string
 @224 = .sdata:0x804D5630; // type:object size:0x7 scope:local data:string
 @225 = .sdata:0x804D5638; // type:object size:0x5 scope:local data:string
 @256 = .sdata:0x804D5640; // type:object size:0x7 scope:local data:string
 @257 = .sdata:0x804D5648; // type:object size:0x5 scope:local data:string
-it_804D5650 = .sdata:0x804D5650; // type:object size:0x7 scope:global data:string
-it_804D5658 = .sdata:0x804D5658; // type:object size:0x5 scope:global data:string
+@495 = .sdata:0x804D5650; // type:object size:0x7 scope:local data:string
+@496 = .sdata:0x804D5658; // type:object size:0x5 scope:local data:string
 @373 = .sdata:0x804D5660; // type:object size:0x7 scope:local data:string
 @374 = .sdata:0x804D5668; // type:object size:0x5 scope:local data:string
 it_804D5670 = .sdata:0x804D5670; // type:object size:0x7 scope:global data:string
@@ -33050,8 +33051,8 @@ it_804DCC5C = .sdata2:0x804DCC5C; // type:object size:0x4 scope:global align:4 d
 @152 = .sdata2:0x804DCC64; // type:object size:0x4 scope:local align:4 data:float
 @177 = .sdata2:0x804DCC68; // type:object size:0x4 scope:local align:4 data:float
 @179 = .sdata2:0x804DCC70; // type:object size:0x8 scope:local align:8 data:double
-it_804DCC78 = .sdata2:0x804DCC78; // type:object size:0x4 scope:global align:4 data:float
-it_804DCC7C = .sdata2:0x804DCC7C; // type:object size:0x4 scope:global align:4 data:float
+@164 = .sdata2:0x804DCC78; // type:object size:0x4 scope:local align:4 data:float
+@189 = .sdata2:0x804DCC7C; // type:object size:0x4 scope:local align:4 data:float
 @206 = .sdata2:0x804DCC80; // type:object size:0x4 scope:local align:4 data:float
 @153 = .sdata2:0x804DCC88; // type:object size:0x4 scope:local align:4 data:float
 @154 = .sdata2:0x804DCC8C; // type:object size:0x4 scope:local align:4 data:float
@@ -33202,11 +33203,11 @@ it_804DCFE0 = .sdata2:0x804DCFE0; // type:object size:0x8 scope:global align:8 d
 it_804DCFE8 = .sdata2:0x804DCFE8; // type:object size:0x4 scope:global align:4 data:float
 it_804DCFEC = .sdata2:0x804DCFEC; // type:object size:0x4 scope:global align:4 data:float
 it_804DCFF0 = .sdata2:0x804DCFF0; // type:object size:0x8 scope:global align:8 data:double
-it_804DCFF8 = .sdata2:0x804DCFF8; // type:object size:0x4 scope:global align:4 data:float
-it_804DCFFC = .sdata2:0x804DCFFC; // type:object size:0x4 scope:global align:4 data:float
-it_804DD000 = .sdata2:0x804DD000; // type:object size:0x8 scope:global align:8 data:double
-it_804DD008 = .sdata2:0x804DD008; // type:object size:0x8 scope:global align:8 data:double
-it_804DD010 = .sdata2:0x804DD010; // type:object size:0x8 scope:global align:8 data:double
+@173 = .sdata2:0x804DCFF8; // type:object size:0x4 scope:local align:4 data:float
+@220 = .sdata2:0x804DCFFC; // type:object size:0x4 scope:local align:4 data:float
+@244 = .sdata2:0x804DD000; // type:object size:0x8 scope:local align:8 data:double
+@391 = .sdata2:0x804DD008; // type:object size:0x8 scope:local align:8 data:double
+@392 = .sdata2:0x804DD010; // type:object size:0x8 scope:local align:8 data:double
 @165 = .sdata2:0x804DD018; // type:object size:0x4 scope:local align:4 data:float
 @190 = .sdata2:0x804DD01C; // type:object size:0x4 scope:local align:4 data:float
 @243 = .sdata2:0x804DD020; // type:object size:0x4 scope:local align:4 data:float
@@ -33422,14 +33423,14 @@ it_804DD4A4 = .sdata2:0x804DD4A4; // type:object size:0x4 scope:global align:4 d
 @207 = .sdata2:0x804DD4BC; // type:object size:0x4 scope:local align:4 data:float
 @200 = .sdata2:0x804DD4C0; // type:object size:0x4 scope:local align:4 data:float
 @206 = .sdata2:0x804DD4C4; // type:object size:0x4 scope:local align:4 data:float
-it_804DD4C8 = .sdata2:0x804DD4C8; // type:object size:0x4 scope:global align:4 data:float
-it_804DD4D0 = .sdata2:0x804DD4D0; // type:object size:0x8 scope:global align:8 data:double
-it_804DD4D8 = .sdata2:0x804DD4D8; // type:object size:0x4 scope:global align:4 data:float
-it_804DD4DC = .sdata2:0x804DD4DC; // type:object size:0x4 scope:global align:4 data:float
-it_804DD4E0 = .sdata2:0x804DD4E0; // type:object size:0x4 scope:global align:4 data:float
-it_804DD4E8 = .sdata2:0x804DD4E8; // type:object size:0x8 scope:global align:8 data:double
-it_804DD4F0 = .sdata2:0x804DD4F0; // type:object size:0x8 scope:global align:8 data:double
-it_804DD4F8 = .sdata2:0x804DD4F8; // type:object size:0x4 scope:global align:4 data:float
+@177 = .sdata2:0x804DD4C8; // type:object size:0x4 scope:local align:4 data:float
+@179 = .sdata2:0x804DD4D0; // type:object size:0x8 scope:local align:8 data:double
+@194 = .sdata2:0x804DD4D8; // type:object size:0x4 scope:local align:4 data:float
+@245 = .sdata2:0x804DD4DC; // type:object size:0x4 scope:local align:4 data:float
+@246 = .sdata2:0x804DD4E0; // type:object size:0x4 scope:local align:4 data:float
+@270 = .sdata2:0x804DD4E8; // type:object size:0x8 scope:local align:8 data:double
+@271 = .sdata2:0x804DD4F0; // type:object size:0x8 scope:local align:8 data:double
+@351 = .sdata2:0x804DD4F8; // type:object size:0x4 scope:local align:4 data:float
 @149 = .sdata2:0x804DD500; // type:object size:0x4 scope:local align:4 data:float
 @180 = .sdata2:0x804DD504; // type:object size:0x4 scope:local align:4 data:float
 @181 = .sdata2:0x804DD508; // type:object size:0x4 scope:local align:4 data:float
@@ -33475,16 +33476,16 @@ it_804DD5A8 = .sdata2:0x804DD5A8; // type:object size:0x8 scope:global align:8 d
 @452 = .sdata2:0x804DD5F0; // type:object size:0x4 scope:local align:4 data:float
 @566 = .sdata2:0x804DD5F4; // type:object size:0x4 scope:local align:4 data:float
 @608 = .sdata2:0x804DD5F8; // type:object size:0x4 scope:local align:4 data:float
-it_804DD600 = .sdata2:0x804DD600; // type:object size:0x4 scope:global align:4 data:float
-it_804DD608 = .sdata2:0x804DD608; // type:object size:0x8 scope:global align:8 data:double
-it_804DD610 = .sdata2:0x804DD610; // type:object size:0x4 scope:global align:4 data:float
-it_804DD614 = .sdata2:0x804DD614; // type:object size:0x4 scope:global align:4 data:float
-it_804DD618 = .sdata2:0x804DD618; // type:object size:0x4 scope:global align:4 data:float
-it_804DD61C = .sdata2:0x804DD61C; // type:object size:0x4 scope:global align:4 data:float
-it_804DD620 = .sdata2:0x804DD620; // type:object size:0x4 scope:global align:4 data:float
-it_804DD624 = .sdata2:0x804DD624; // type:object size:0x4 scope:global align:4 data:float
-it_804DD628 = .sdata2:0x804DD628; // type:object size:0x4 scope:global align:4 data:float
-it_804DD62C = .sdata2:0x804DD62C; // type:object size:0x4 scope:global align:4 data:float
+@180 = .sdata2:0x804DD600; // type:object size:0x4 scope:local align:4 data:float
+@182 = .sdata2:0x804DD608; // type:object size:0x8 scope:local align:8 data:double
+@197 = .sdata2:0x804DD610; // type:object size:0x4 scope:local align:4 data:float
+@218 = .sdata2:0x804DD614; // type:object size:0x4 scope:local align:4 data:float
+@303 = .sdata2:0x804DD618; // type:object size:0x4 scope:local align:4 data:float
+@343 = .sdata2:0x804DD61C; // type:object size:0x4 scope:local align:4 data:float
+@498 = .sdata2:0x804DD620; // type:object size:0x4 scope:local align:4 data:float
+@549 = .sdata2:0x804DD624; // type:object size:0x4 scope:local align:4 data:float
+@584 = .sdata2:0x804DD628; // type:object size:0x4 scope:local align:4 data:float
+@585 = .sdata2:0x804DD62C; // type:object size:0x4 scope:local align:4 data:float
 @204 = .sdata2:0x804DD630; // type:object size:0x4 scope:local align:4 data:float
 @206 = .sdata2:0x804DD634; // type:object size:0x4 scope:local align:4 data:float
 @220 = .sdata2:0x804DD638; // type:object size:0x8 scope:local align:8 data:double
diff --git a/configure.py b/configure.py
index 37be1b200e..a47088cb51 100755
--- a/configure.py
+++ b/configure.py
@@ -1147,7 +1147,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/itsscopebeam.c"),
             Object(NonMatching, "melee/it/items/itlgunbeam.c"),
             Object(Matching, "melee/it/items/ithammerhead.c"),
-            Object(NonMatching, "melee/it/items/itlipstickspore.c"),
+            Object(Matching, "melee/it/items/itlipstickspore.c"),
             Object(Matching, "melee/it/items/itfflowerflame.c"),
             Object(Matching, "melee/it/items/itevyoshiegg.c"),
             # Fighter-related items
@@ -1179,7 +1179,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/ityoshieggthrow.c"),
             Object(Matching, "melee/it/items/ityoshistar.c"),
             Object(NonMatching, "melee/it/items/itpikachutjoltground.c"),
-            Object(NonMatching, "melee/it/items/itpikachutjoltair.c"),
+            Object(Matching, "melee/it/items/itpikachutjoltair.c"),
             Object(Matching, "melee/it/items/itsamusbomb.c"),
             Object(NonMatching, "melee/it/items/itsamuschargeshot.c"),
             Object(NonMatching, "melee/it/items/itsamusmissile.c"),
@@ -1237,7 +1237,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/ittogepy.c"),
             Object(Matching, "melee/it/items/itmew.c"),
             Object(Matching, "melee/it/items/itcerebi.c"),
-            Object(NonMatching, "melee/it/items/ithitodeman.c"),
+            Object(Matching, "melee/it/items/ithitodeman.c"),
             Object(Matching, "melee/it/items/itlucky.c"),
             Object(Matching, "melee/it/items/itporygon2.c"),
             Object(NonMatching, "melee/it/items/ithinoarashi.c"),
@@ -1247,7 +1247,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/itoldkuri.c"),
             Object(Matching, "melee/it/items/itmato.c"),
             Object(Matching, "melee/it/items/itheiho.c"),
-            Object(NonMatching, "melee/it/items/itlikelike.c"),
+            Object(Matching, "melee/it/items/itlikelike.c"),
             Object(Matching, "melee/it/items/itnokonoko.c"),
             Object(NonMatching, "melee/it/items/itzgshell.c"),
             Object(Matching, "melee/it/items/itzrshell.c"),
diff --git a/src/melee/ft/ft_0C31.c b/src/melee/ft/ft_0C31.c
index 4a8abaf0cf..ffcc9d06e6 100644
--- a/src/melee/ft/ft_0C31.c
+++ b/src/melee/ft/ft_0C31.c
@@ -625,7 +625,7 @@ void fn_800C7890(Fighter_GObj* gobj)
     ftCo_800C7A30(gobj);
 }
 
-UNK_RET ftCo_800C78B0(UNK_PARAMS)
+UNK_RET ftCo_800C78B0(HSD_GObj* gobj, HSD_GObj* owner)
 {
     NOT_IMPLEMENTED;
 }
diff --git a/src/melee/ft/ft_0C31.h b/src/melee/ft/ft_0C31.h
index 5ff4346a46..dcc4372003 100644
--- a/src/melee/ft/ft_0C31.h
+++ b/src/melee/ft/ft_0C31.h
@@ -46,7 +46,7 @@
 /* 0C77B4 */ void ftCo_CaptureLeadead_Phys(Fighter_GObj* gobj);
 /* 0C77B8 */ void ftCo_CaptureLeadead_Coll(Fighter_GObj* gobj);
 /* 0C7800 */ void ftCo_800C7800(Fighter_GObj*);
-/* 0C78B0 */ UNK_RET ftCo_800C78B0(UNK_PARAMS);
+/* 0C78B0 */ UNK_RET ftCo_800C78B0(HSD_GObj*, HSD_GObj*);
 /* 0C7A30 */ void ftCo_800C7A30(Fighter_GObj* gobj);
 /* 0C7A58 */ void ftCo_CaptureLikelike_Anim(Fighter_GObj* gobj);
 /* 0C7AD4 */ void ftCo_CaptureLikelike_IASA(Fighter_GObj* gobj);
diff --git a/src/melee/it/items/ithitodeman.c b/src/melee/it/items/ithitodeman.c
index 7f07f83a97..0d7996a480 100644
--- a/src/melee/it/items/ithitodeman.c
+++ b/src/melee/it/items/ithitodeman.c
@@ -1,14 +1,8 @@
 #include "ithitodeman.h"
 
-#include <placeholder.h>
-#include <platform.h>
-
 #include "ef/eflib.h"
 #include "ef/efsync.h"
 #include "ft/ftlib.h"
-
-#include "it/forward.h"
-
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
@@ -17,15 +11,60 @@
 #include "lb/lbvector.h"
 
 #include <math.h>
-#include <baselib/jobj.h>
 #include <baselib/random.h>
-#include <MSL/math_ppc.h>
-#include <MSL/trigf.h>
+
+ItemStateTable it_803F8128[] = {
+    {
+        0,
+        it_802D4564,
+        it_802D472C,
+        it_802D48A8,
+    },
+    {
+        1,
+        itHitodeman_UnkMotion1_Anim,
+        itHitodeman_UnkMotion1_Phys,
+        itHitodeman_UnkMotion1_Coll,
+    },
+    {
+        -1,
+        itHitodeman_UnkMotion2_Anim,
+        itHitodeman_UnkMotion2_Phys,
+        itHitodeman_UnkMotion2_Coll,
+    },
+};
+
+static void ordering_func()
+{
+    HSD_JObjSetRotationY(NULL, 10);
+}
+
+ItemStateTable it_803F8180[] = { {
+    0x00000000,
+    it_802D4FFC,
+    it_802D5044,
+    it_802D5048,
+} };
+
+extern inline float sqrtf(float x)
+{
+    volatile float y;
+    if (x > 0.0f) {
+        double guess = __frsqrte((double) x); // returns an approximation to
+        guess = .5 * guess * (3.0 - guess * guess * x); // now have 12 sig bits
+        guess = .5 * guess * (3.0 - guess * guess * x); // now have 24 sig bits
+        guess = .5 * guess * (3.0 - guess * guess * x); // now have 32 sig bits
+        y = (float) (x * guess);
+        return y;
+    }
+    return x;
+}
 
 void it_2725_Logic24_Spawned(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
     itHitodemanAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    PAD_STACK(16);
 
     ip->facing_dir = 0.0f;
     ip->xDBC_itcmd_var4.flags.x0 = false;
@@ -35,7 +74,7 @@ void it_2725_Logic24_Spawned(Item_GObj* gobj)
     it_802D4494(gobj);
     it_802D4510(gobj);
 
-    ip->xDD4_itemVar.hitodeman.x8C = (f32) attrs->x3C;
+    ip->xDD4_itemVar.hitodeman.x8C = attrs->x3C;
 
     ip->xDD4_itemVar.hitodeman.x6C = 0.0f;
     ip->xDD4_itemVar.hitodeman.x70 = 0.0f;
@@ -51,7 +90,7 @@ void it_2725_Logic24_Spawned(Item_GObj* gobj)
     Item_8026AE84(ip, 0x2738, 0x7F, 0x40);
 }
 
-void it_802D43AC(void) {}
+void it_802D43AC(Item_GObj* gobj) {}
 
 void it_802D43B0(Item_GObj* gobj, Item_GObj* ref_gobj)
 {
@@ -151,15 +190,20 @@ bool it_802D4564(Item_GObj* gobj)
     return false;
 }
 
+inline Article* it_802D472C_inline(Item* ip)
+{
+    return ip->xC4_article_data;
+}
+
 void it_802D472C(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
-    itHitodemanAttributes* attrs = ip->xC4_article_data->x4_specialAttributes;
+    itHitodemanAttributes* attrs =
+        it_802D472C_inline(ip)->x4_specialAttributes;
     Vec3 target_pos;
     Vec3 dir;
     f32 speed;
     f32 dist;
-    PAD_STACK(4);
 
     if (ip->xDD4_itemVar.hitodeman.x90 != NULL) {
         ftLib_80086644(ip->xDD4_itemVar.hitodeman.x90, &target_pos);
@@ -193,20 +237,20 @@ f32 it_802D48B0(f32* value, f32 target, f32 max_val, f32 accel, f32 decel)
     f32 diff = target - current;
 
     if (ABS(diff) > ABS(accel)) {
-        if (diff > 0.0F) {
+        if (diff > 0.0f) {
             target = current + accel;
         } else {
             target = current - accel;
         }
     } else if (ABS(diff) < ABS(decel)) {
-        if (diff > 0.0F) {
+        if (diff > 0.0f) {
             target = current + decel;
         } else {
             target = current - decel;
         }
     }
 
-    if (target > 0.0F) {
+    if (target > 0.0f) {
         if (target > max_val) {
             target = max_val;
         }
@@ -240,10 +284,10 @@ bool itHitodeman_UnkMotion1_Anim(Item_GObj* gobj)
         return true;
     }
 
-    timer = ip->xDD4_itemVar.hitodeman.x8C - 1.0F;
+    timer = ip->xDD4_itemVar.hitodeman.x8C - 1.0f;
     ip->xDD4_itemVar.hitodeman.x8C = timer;
 
-    if (timer < 0.0F) {
+    if (timer < 0.0f) {
         if (--ip->xDD4_itemVar.hitodeman.x88 < 0) {
             return true;
         }
@@ -320,7 +364,7 @@ void itHitodeman_UnkMotion2_Phys(Item_GObj* gobj)
 
 bool itHitodeman_UnkMotion2_Coll(Item_GObj* gobj)
 {
-    return it_8027A118(gobj, (void (*)(HSD_GObj*)) it_802D43AC);
+    return it_8027A118(gobj, it_802D43AC);
 }
 
 void it_802D4C74(Item_GObj* gobj)
@@ -422,9 +466,9 @@ bool it_802D4FFC(Item_GObj* gobj)
     Item* ip = GET_ITEM(gobj);
     f32 timer;
 
-    timer = ip->xD44_lifeTimer - 1.0F;
+    timer = ip->xD44_lifeTimer - 1.0f;
     ip->xD44_lifeTimer = timer;
-    if (timer <= 0.0F) {
+    if (timer <= 0.0f) {
         return true;
     }
     if (ip->xDB4_itcmd_var2 != 0) {
diff --git a/src/melee/it/items/ithitodeman.h b/src/melee/it/items/ithitodeman.h
index 822e4b1518..ec0bac8978 100644
--- a/src/melee/it/items/ithitodeman.h
+++ b/src/melee/it/items/ithitodeman.h
@@ -8,7 +8,7 @@
 #include "it/items/types.h"
 
 /* 2D42C4 */ void it_2725_Logic24_Spawned(Item_GObj*);
-/* 2D43AC */ void it_802D43AC(void);
+/* 2D43AC */ void it_802D43AC(Item_GObj* gobj);
 /* 2D43B0 */ void it_802D43B0(Item_GObj*, Item_GObj*);
 /* 2D43EC */ void it_802D43EC(Item_GObj* gobj);
 /* 2D4494 */ void it_802D4494(Item_GObj* gobj);
diff --git a/src/melee/it/items/itlikelike.c b/src/melee/it/items/itlikelike.c
index 2ed169b99a..a204b99541 100644
--- a/src/melee/it/items/itlikelike.c
+++ b/src/melee/it/items/itlikelike.c
@@ -571,6 +571,24 @@ void it_802DA960(Item_GObj* gobj)
     Item_80268E5C(gobj, 4, ITEM_ANIM_UPDATE);
 }
 
+static inline void it_802DC4BC_ptr(Item_GObj* gobj, itECB* ecb)
+{
+    Item* ip = GET_ITEM(gobj);
+    it_802DBAF0(gobj, 1, 1);
+    it_802762BC(ip);
+    if (ip->xDD4_itemVar.likelike.x38 == 1) {
+        *ecb = ip->xC1C;
+        ecb->bottom = ecb->top;
+        ecb->top = ip->xC1C.bottom;
+        it_80275D5C(gobj, ecb);
+    }
+    it_8027C56C(gobj, 0.0f);
+    ip->xDD4_itemVar.likelike.x38 = 0;
+    ip->xDD4_itemVar.likelike.x4C = 0;
+    it_80273454(gobj);
+    Item_80268E5C(gobj, 5, ITEM_ANIM_UPDATE);
+}
+
 void it_802DAA10(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -1180,11 +1198,27 @@ void itLikelike_UnkMotion18_Phys(Item_GObj* gobj) {}
 void it_2725_Logic5_Dropped(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
+    Item* ip2;
     itLikelikeAttributes* attr = GET_ATTRS(ip);
+    itECB ecb;
+    PAD_STACK(12);
 
     it_8027CBA4(gobj);
     ip->xDD4_itemVar.likelike.x4C = attr->x4;
-    it_802DA960(gobj);
+
+    // permuterslop
+    ip = ip2 = GET_ITEM(gobj);
+    it_802762BC(ip2);
+    if (ip2->xDD4_itemVar.likelike.x38 == 1) {
+        ecb = ip2->xC1C;
+        ecb.bottom = ecb.top;
+        ecb.top = ip2->xC1C.bottom;
+        it_80275D5C(gobj, &ecb);
+        it_8027C56C(gobj, 0.0f);
+    }
+    ip2->xDD4_itemVar.likelike.x38 = 0;
+    it_80273454(gobj);
+    Item_80268E5C(gobj, 4, ITEM_ANIM_UPDATE);
 }
 
 void itLikeLike_Logic5_Thrown(Item_GObj* gobj)
@@ -1230,20 +1264,35 @@ bool itLikelike_UnkMotion6_Anim(Item_GObj* gobj)
 
 void itLikelike_UnkMotion6_Phys(Item_GObj* gobj)
 {
-    itLikelikeAttributes* attr;
     Item* ip = GET_ITEM(gobj);
-    s32 temp_r3;
-
-    temp_r3 = ip->xDD4_itemVar.likelike.x4C;
     if (ip->xDD4_itemVar.likelike.x4C == 0) {
         if (ip->xDD4_itemVar.likelike.x38 == 0) {
             it_802DA104(gobj);
             return;
         }
-        it_802DAA10(gobj);
+
+        {
+            Item* ip = GET_ITEM(gobj);
+            itECB ecb;
+            PAD_STACK(20);
+            it_802DBAF0(gobj, 1, 1);
+            it_802762BC(ip);
+            if (ip->xDD4_itemVar.likelike.x38 == 1) {
+                ecb = ip->xC1C;
+                ecb.bottom = ecb.top;
+                ecb.top = ip->xC1C.bottom;
+                it_80275D5C(gobj, &ecb);
+            }
+            it_8027C56C(gobj, 0.0f);
+            ip->xDD4_itemVar.likelike.x38 = 0;
+            ip->xDD4_itemVar.likelike.x4C = 0;
+            it_80273454(gobj);
+            Item_80268E5C(gobj, 5, ITEM_ANIM_UPDATE);
+        }
+
         return;
     }
-    ip->xDD4_itemVar.likelike.x4C = temp_r3 - 1;
+    ip->xDD4_itemVar.likelike.x4C--;
 }
 
 bool itLikelike_UnkMotion6_Coll(Item_GObj* gobj)
@@ -1325,19 +1374,16 @@ void itLikeLike_Logic5_Destroyed(Item_GObj* gobj)
 
 Item_GObj* it_802DC4BC(s32 arg0, Vec3* arg1, s32 arg2)
 {
-    // one r29-r30 regswap away from being solved
-    Vec3 sp50;
-    Item_GObj* gobj;
-
-    sp50 = zero_vec;
-    gobj = it_8027B5B0(It_Kind_Likelike, arg1, NULL, &sp50, 1);
+    Vec3 sp50 = zero_vec;
+    Item_GObj* gobj = it_8027B5B0(It_Kind_Likelike, arg1, NULL, &sp50, 1);
+    itECB ecb;
+    PAD_STACK(28);
     if (gobj != NULL) {
-        Item* item = GET_ITEM(gobj);
-        PAD_STACK(32);
-        item->xDD4_itemVar.likelike.x38 = arg0;
-        it_802762BC(item);
+        Item* ip = GET_ITEM(gobj);
+        ip->xDD4_itemVar.likelike.x38 = arg0;
+        it_802762BC(ip);
         if (arg0 == 0) {
-            it_802DAA10(gobj);
+            it_802DC4BC_ptr(gobj, &ecb);
         } else {
             it_802DAD18(gobj);
         }
diff --git a/src/melee/it/items/itlipstick.c b/src/melee/it/items/itlipstick.c
index 2c7104ba98..a0747ea432 100644
--- a/src/melee/it/items/itlipstick.c
+++ b/src/melee/it/items/itlipstick.c
@@ -47,7 +47,7 @@ void it_80295770(Item_GObj* gobj, Vec* arg1, float arg2, int arg3)
     if (ip->xD4C > 0) {
         ip->xD4C -= 1;
     }
-    it_8029A114(ip->owner, arg1, arg3, arg2);
+    it_8029A114(ip->owner, arg1, arg2, arg3);
 }
 
 void it_802957AC(Item_GObj* gobj)
diff --git a/src/melee/it/items/itlipstickspore.c b/src/melee/it/items/itlipstickspore.c
index 8df11a709d..8f855926ff 100644
--- a/src/melee/it/items/itlipstickspore.c
+++ b/src/melee/it/items/itlipstickspore.c
@@ -23,18 +23,32 @@
 
 /* 29A5F8 */ static bool itLipstickspore_UnkMotion1_Coll(Item_GObj* gobj);
 
-void it_8029A114(Fighter_GObj* gobj, Vec3* pos, int arg2, f32 facing_dir)
+ItemStateTable it_803F6680[] = {
+    {
+        0,
+        itLipstickspore_UnkMotion0_Anim,
+        itLipstickspore_UnkMotion0_Phys,
+        itLipstickspore_UnkMotion0_Coll,
+    },
+    {
+        1,
+        itLipstickspore_UnkMotion1_Anim,
+        itLipstickspore_UnkMotion1_Phys,
+        itLipstickspore_UnkMotion1_Coll,
+    },
+};
+
+void it_8029A114(Fighter_GObj* gobj, Vec3* pos, f32 facing_dir, s32 arg4)
 {
     Item_GObj* item_gobj;
     SpawnItem spawn;
-    u8 _pad[4];
-    f32 dir = facing_dir;
+    PAD_STACK(8);
 
     spawn.kind = 0x25;
     spawn.prev_pos = *pos;
     spawn.prev_pos.z = 0.0f;
     it_8026BB68(gobj, &spawn.pos);
-    spawn.facing_dir = dir;
+    spawn.facing_dir = facing_dir;
     spawn.x3C_damage = 0;
     spawn.vel.x = spawn.vel.y = spawn.vel.z = 0.0f;
     spawn.x0_parent_gobj = gobj;
@@ -44,10 +58,10 @@ void it_8029A114(Fighter_GObj* gobj, Vec3* pos, int arg2, f32 facing_dir)
     item_gobj = Item_80268B18(&spawn);
     if (item_gobj != NULL) {
         Item* ip = item_gobj->user_data;
-        ip->xDD4_itemVar.lipstickspore.xDD4 = arg2;
+        ip->xDD4_itemVar.lipstickspore.xDD4 = arg4;
         it_8029A31C(item_gobj, gobj, ip);
         db_80225DD8(item_gobj, gobj);
-        efSync_Spawn(0x440, item_gobj, pos, &dir);
+        efSync_Spawn(0x440, item_gobj, pos, &facing_dir);
     }
 }
 
diff --git a/src/melee/it/items/itlipstickspore.h b/src/melee/it/items/itlipstickspore.h
index f5ab15e451..3ecdb07f69 100644
--- a/src/melee/it/items/itlipstickspore.h
+++ b/src/melee/it/items/itlipstickspore.h
@@ -17,7 +17,7 @@ typedef struct itLipstickSporeAttributes {
     s32 x18;
 } itLipstickSporeAttributes;
 
-/* 29A114 */ void it_8029A114(Fighter_GObj*, Vec*, int, float);
+/* 29A114 */ void it_8029A114(Fighter_GObj*, Vec*, f32 facing_dir, s32 arg4);
 /* 29A218 */ void it_8029A218(HSD_GObj* owner, Vec3* pos, f32 facing_dir,
                               s32 arg4);
 /* 29A31C */ void it_8029A31C(Item_GObj*, Fighter_GObj*, Item*);
diff --git a/src/melee/it/items/itpikachutjoltair.c b/src/melee/it/items/itpikachutjoltair.c
index 2b922f90e2..c8c8977f95 100644
--- a/src/melee/it/items/itpikachutjoltair.c
+++ b/src/melee/it/items/itpikachutjoltair.c
@@ -3,11 +3,15 @@
 #include <placeholder.h>
 #include <platform.h>
 
+#include "baselib/gobj.h"
 #include "db/db.h"
 #include "it/inlines.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/item.h"
+
+#include "it/items/forward.h"
+
 #include "it/items/itpikachutjoltground.h"
 #include "lb/lb_00B0.h"
 
@@ -17,56 +21,52 @@
 
 /* 2B45E8 */ static bool itPikachutjoltair_UnkMotion0_Coll(Item_GObj* gobj);
 
-bool itPikachutjoltair_UnkMotion0_Coll(Item_GObj* gobj)
-{
-    return false;
-}
+ItemStateTable it_803F71D8[] = { {
+    0,
+    itPikachutjoltair_UnkMotion0_Anim,
+    itPikachutjoltair_UnkMotion0_Phys,
+    itPikachutjoltair_UnkMotion0_Coll,
+} };
 
 Item_GObj* it_802B3EFC(Item_GObj* gobj)
 {
-    Item* ip;
-
     if (gobj != NULL) {
-        ip = GET_ITEM(gobj);
+        Item* ip = GET_ITEM(gobj);
         if (ip != NULL) {
             return ip->xDD4_itemVar.pikachujoltair.xDD8;
         }
     }
-    return 0;
+    return NULL;
 }
 
-void it_802B3F20(Item_GObj* gobj, Vec3* arg1)
+void it_802B3F20(Item_GObj* gobj, Vec3* v)
 {
-    Item* ip = gobj->user_data;
-    PAD_STACK(8);
+    Item* ip = GET_ITEM(gobj);
+    PAD_STACK(4);
     if (ip->xDD4_itemVar.pikachujoltair.xDD8 != NULL) {
-        lb_8000B1CC(ip->xBBC_dynamicBoneTable->bones[6], NULL, arg1);
-        arg1->z = 0.0f;
+        lb_8000B1CC(ip->xBBC_dynamicBoneTable->bones[6], NULL, v);
+        v->z = 0.0f;
         return;
     }
-    arg1->z = 0.0f;
-    arg1->y = 0.0f;
-    arg1->x = 0.0f;
+    v->x = v->y = v->z = 0.0f;
 }
 
 void it_802B3F88(Item_GObj* gobj, Vec3* pos, CollData* coll, Vec3* vel)
 {
     Item* ip = GET_ITEM(gobj);
-    HSD_JObj* jobj = gobj->hsd_obj;
-    PAD_STACK(8);
+    HSD_JObj* jobj = GET_JOBJ(gobj);
+    PAD_STACK(4);
 
     if (gobj != NULL && ip != NULL) {
+        f32 angle;
         ip->pos = *pos;
         ip->x378_itemColl = *coll;
         ip->xDD4_itemVar.pikachujoltair.xDE8 = *vel;
         ip->xDD4_itemVar.pikachujoltair.xDF4 = 0;
         HSD_JObjSetTranslate(jobj, &ip->pos);
-        {
-            f32 angle = atan2f(ip->xDD4_itemVar.pikachujoltair.xDE8.x,
-                               ip->xDD4_itemVar.pikachujoltair.xDE8.y);
-            HSD_JObjSetRotationX(jobj,
-                                 (ip->facing_dir == 1.0f) ? angle : -angle);
-        }
+        angle = atan2f(ip->xDD4_itemVar.pikachujoltair.xDE8.x,
+                       ip->xDD4_itemVar.pikachujoltair.xDE8.y);
+        HSD_JObjSetRotationX(jobj, (ip->facing_dir == 1.0f) ? angle : -angle);
         Item_80268E5C(gobj, 0, 0x12);
         Item_802694CC(gobj);
         switch (ip->kind) {
@@ -110,7 +110,7 @@ Item_GObj* it_802B4224(HSD_GObj* owner, Item_GObj* gobj, Vec3* pos, s32 kind,
         ip->xDAC_itcmd_var0 = 0;
         ip->xDD4_itemVar.pikachujoltair.xDD8 = gobj;
         ip->xDD4_itemVar.pikachujoltair.xDF4 = 0;
-        HSD_JObjSetRotationY(jobj, (f32) (M_PI_2 * (f64) ip->facing_dir));
+        HSD_JObjSetRotationY(jobj, (M_PI_2 * ip->facing_dir));
         it_802B43D0(item_gobj, owner);
         ip->xDD4_itemVar.pikachujoltair.xDD4 = owner;
     }
@@ -127,15 +127,15 @@ void it_802B43B0(Item_GObj* gobj)
     }
 }
 
-void it_802B43D0(Item_GObj* gobj, HSD_GObj* arg1)
+void it_802B43D0(Item_GObj* gobj, HSD_GObj* owner)
 {
     Item* ip = GET_ITEM(gobj);
     it_8026B3A8(gobj);
     ip->xDC8_word.flags.x13 = 0;
     it_80272940(gobj);
-    Item_80268E5C((HSD_GObj*) gobj, 0, 2);
-    Item_802694CC((HSD_GObj*) gobj);
-    db_80225DD8(gobj, arg1);
+    Item_80268E5C(gobj, 0, 2);
+    Item_802694CC(gobj);
+    db_80225DD8(gobj, owner);
 }
 
 static inline void itPikachuTJoltAir_Destroy(Item_GObj* gobj)
@@ -201,7 +201,10 @@ void itPikachutjoltair_UnkMotion0_Phys(Item_GObj* gobj)
     itResetVelocity(ip);
 }
 
-/// #itPikachutjoltair_UnkMotion0_Coll
+bool itPikachutjoltair_UnkMotion0_Coll(Item_GObj* gobj)
+{
+    return false;
+}
 
 bool it_2725_Logic107_DmgDealt(Item_GObj* gobj)
 {
@@ -219,7 +222,7 @@ bool it_2725_Logic107_Clanked(Item_GObj* gobj)
     Item* ip = GET_ITEM(gobj);
     it_802725D4(gobj);
     if (ip->xDD4_itemVar.pikachujoltair.xDD8 != 0) {
-        it_802B3544((Item_GObj*) ip->xDD4_itemVar.pikachujoltair.xDD8);
+        it_802B3544(ip->xDD4_itemVar.pikachujoltair.xDD8);
         ip->xDD4_itemVar.pikachujoltair.xDD8 = 0;
     }
     return true;
@@ -230,7 +233,7 @@ bool it_2725_Logic107_Absorbed(Item_GObj* gobj)
     Item* ip = GET_ITEM(gobj);
     it_802725D4(gobj);
     if (ip->xDD4_itemVar.pikachujoltair.xDD8 != 0) {
-        it_802B3544((Item_GObj*) ip->xDD4_itemVar.pikachujoltair.xDD8);
+        it_802B3544(ip->xDD4_itemVar.pikachujoltair.xDD8);
         ip->xDD4_itemVar.pikachujoltair.xDD8 = 0;
     }
     return true;
@@ -285,7 +288,7 @@ bool it_2725_Logic107_HitShield(Item_GObj* gobj)
     Item* ip = GET_ITEM(gobj);
     it_802725D4(gobj);
     if (ip->xDD4_itemVar.pikachujoltair.xDD8 != 0) {
-        it_802B3544((Item_GObj*) ip->xDD4_itemVar.pikachujoltair.xDD8);
+        it_802B3544(ip->xDD4_itemVar.pikachujoltair.xDD8);
         ip->xDD4_itemVar.pikachujoltair.xDD8 = 0;
     }
     return true;
@@ -296,7 +299,7 @@ bool it_2725_Logic107_ShieldBounced(Item_GObj* gobj)
     Item* ip = GET_ITEM(gobj);
     it_802725D4(gobj);
     if (ip->xDD4_itemVar.pikachujoltair.xDD8 != 0) {
-        it_802B3544((Item_GObj*) ip->xDD4_itemVar.pikachujoltair.xDD8);
+        it_802B3544(ip->xDD4_itemVar.pikachujoltair.xDD8);
         ip->xDD4_itemVar.pikachujoltair.xDD8 = 0;
     }
     return true;
