diff --git a/src/melee/mn/mnname.c b/src/melee/mn/mnname.c
index b7ad18db45..6b368ee4f1 100644
--- a/src/melee/mn/mnname.c
+++ b/src/melee/mn/mnname.c
@@ -82,7 +82,23 @@ bool IsNameListFull(void)
 
 void fn_802377A4(void) {}
 
-/// #IsNameUnique
+bool IsNameUnique(s32 slot)
+{
+    int i;
+    char* namedata;
+    for (i = 0; i < 0x78; i++) {
+        namedata = !IsNameValid(i & 0xFF)
+                       ? (char*) NULL
+                       : GetPersistentNameData((u8) i)->namedata;
+        if (namedata == NULL) {
+            break;
+        }
+        if (CompareNameStrings(namedata, slot) == 0) {
+            return true;
+        }
+    }
+    return false;
+}
 
 /// #DeleteName
 
diff --git a/src/melee/mn/mnname.h b/src/melee/mn/mnname.h
index af030eaf0e..2a308f8591 100644
--- a/src/melee/mn/mnname.h
+++ b/src/melee/mn/mnname.h
@@ -22,7 +22,7 @@ typedef struct MnName_GObj {
 /* 2375EC */ bool IsNameListFull(void);
 /* 237654 */ s32 CompareNameStrings(char* str, s32 slot);
 /* 2377A4 */ void fn_802377A4(void);
-/* 2377A8 */ UNK_RET IsNameUnique(UNK_PARAMS);
+/* 2377A8 */ bool IsNameUnique(s32 slot);
 /* 237834 */ UNK_RET DeleteName(UNK_PARAMS);
 /* 2379BC */ bool IsNameValid(int slot);
 /* 237A04 */ void CreateNameAtIndex(s32 slot);
diff --git a/src/melee/ty/toy.c b/src/melee/ty/toy.c
index 9e73f3bab0..c8246914f7 100644
--- a/src/melee/ty/toy.c
+++ b/src/melee/ty/toy.c
@@ -664,7 +664,25 @@ s32 un_80305C44(void)
     return button;
 }
 
-/// #un_80305D00
+float un_80305D00(void)
+{
+    float ret;
+    float val;
+    int i;
+
+    for (i = 0; i < 4; i++) {
+        val = HSD_PadCopyStatus[(u8) i].nml_stickX;
+        ret = val;
+        if (val < 0.0F) {
+            val = -val;
+        }
+        if (val > 0.1F) {
+            break;
+        }
+    }
+
+    return ret;
+}
 
 float un_80305DB0(void)
 {
