diff --git a/src/melee/gm/gmregclear.c b/src/melee/gm/gmregclear.c
index ebcd810676..6f276422b8 100644
--- a/src/melee/gm/gmregclear.c
+++ b/src/melee/gm/gmregclear.c
@@ -58,7 +58,8 @@ struct lbl_80472E48_t {
     /* 0x00 */ char pad_0[4];
     /* 0x04 */ s32 unk_4;        /* inferred */
     /* 0x08 */ s32 unk_8;        /* inferred */
-    /* 0x0C */ char pad_C[0x74]; /* maybe part of unk_8[0x1E]? */
+    /* 0x0C */ char pad_C[0x8];
+    /* 0x14 */ s32 x14[0x1B];
 }; /* size = 0x80 */
 STATIC_ASSERT(sizeof(struct lbl_80472E48_t) == 0x80);
 
@@ -657,7 +658,33 @@ u8 fn_8017DD7C(PlayerInitData* arg0, Unk1PData_x24* arg1)
     return index - 1;
 }
 
-/// #fn_8017DE54
+s32 fn_8017DE54(u8 arg0, u8* arg1)
+{
+    u8* p;
+    s32 count;
+
+    if (arg0 & 0x20) {
+        return 0;
+    }
+    if (arg0 & 0x10) {
+        count = 0;
+        if ((s32) arg1[0] != 0x21) {
+            count = 1;
+        }
+        p = &arg1[1];
+        if ((s32) *p != 0x21) {
+            count += 1;
+        }
+        if ((s32) p[1] != 0x21) {
+            count += 1;
+        }
+        return 3 - count;
+    }
+    if (arg0 & 2) {
+        return 2;
+    }
+    return 0;
+}
 
 Unk1PData* fn_8017DEC8(int arg0)
 {
@@ -1074,7 +1101,16 @@ Fighter_GObj* gm_80180AF4(void)
 
 /// #gm_80180B18
 
-/// #gm_80180BA0
+void gm_80180BA0(void)
+{
+    int i;
+
+    for (i = 0; i < 0x1B; i++) {
+        u8 idx = gm_80164024((u8) i);
+        *gmMainLib_8015D06C(gm_80164024((u8) i)) =
+            lbl_80472E48.x14[idx] * 0xA;
+    }
+}
 
 void fn_80180C14(HSD_GObj* gobj)
 {
diff --git a/src/melee/gm/gmregclear.h b/src/melee/gm/gmregclear.h
index dc03d7907c..712ed8b480 100644
--- a/src/melee/gm/gmregclear.h
+++ b/src/melee/gm/gmregclear.h
@@ -30,7 +30,7 @@
 /* 17DB78 */ u8 gm_8017DB78(gm_8017DB6C_arg0_t* arg0, int index);
 /* 17DB88 */ UNK_RET gm_8017DB88(UNK_PARAMS);
 /* 17DD7C */ u8 fn_8017DD7C(PlayerInitData* arg0, Unk1PData_x24* arg1);
-/* 17DE54 */ UNK_RET fn_8017DE54(UNK_PARAMS);
+/* 17DE54 */ s32 fn_8017DE54(u8, u8*);
 /* 17DEC8 */ Unk1PData* fn_8017DEC8(int);
 /* 17DF28 */ Unk1PData* fn_8017DF28(void);
 /* 17DF90 */ u8 fn_8017DF90(void);
@@ -93,7 +93,7 @@
 /* 180AE4 */ s32 gm_80180AE4(void);
 /* 180AF4 */ Fighter_GObj* gm_80180AF4(void);
 /* 180B18 */ UNK_RET gm_80180B18(UNK_PARAMS);
-/* 180BA0 */ UNK_RET gm_80180BA0(UNK_PARAMS);
+/* 180BA0 */ void gm_80180BA0(void);
 /* 180C14 */ void fn_80180C14(HSD_GObj* gobj);
 /* 180C60 */ UNK_RET fn_80180C60(UNK_PARAMS);
 /* 181598 */ UNK_RET fn_80181598(UNK_PARAMS);
diff --git a/src/melee/gr/grcastle.c b/src/melee/gr/grcastle.c
index 2646167719..1559568ea5 100644
--- a/src/melee/gr/grcastle.c
+++ b/src/melee/gr/grcastle.c
@@ -472,7 +472,14 @@ void grCastle_801CF7B0(Ground_GObj* gobj)
 
 /// #grCastle_801CF868
 
-/// #fn_801CFAFC
+void fn_801CFAFC(Item_GObj* item, Ground* gp, Vec3* pos, HSD_GObj* gobj,
+                 f32 arg4)
+{
+    *(s16*) &gp->gv.castle2.xC4 = 4;
+    if (ftLib_80086960(gobj)) {
+        ftLib_80086A4C(gobj, (f32) grCs_804D6970->x4);
+    }
+}
 
 void fn_801CFB68(Item_GObj* item_gobj, Ground* gp, HSD_GObj* gobj)
 {
diff --git a/src/melee/gr/grcastle.h b/src/melee/gr/grcastle.h
index 8a4c44297c..714a83ed64 100644
--- a/src/melee/gr/grcastle.h
+++ b/src/melee/gr/grcastle.h
@@ -8,6 +8,7 @@
 #include "lb/forward.h"
 #include "mp/forward.h"
 #include <baselib/forward.h>
+#include <melee/it/forward.h>
 
 #include <dolphin/mtx.h>
 
@@ -69,7 +70,7 @@ struct unkCastle {
                                     mpLib_GroundEnum, f32);
 /* 1CF7B0 */ void grCastle_801CF7B0(Ground_GObj*);
 /* 1CF868 */ UNK_RET grCastle_801CF868(UNK_PARAMS);
-/* 1CFAFC */ UNK_RET fn_801CFAFC(UNK_PARAMS);
+/* 1CFAFC */ void fn_801CFAFC(Item_GObj*, Ground*, Vec3*, HSD_GObj*, f32);
 /* 1CFB68 */ void fn_801CFB68(Item_GObj*, Ground*, HSD_GObj*);
 /* 1CFBD4 */ UNK_RET grCastle_801CFBD4(UNK_PARAMS);
 /* 1D0298 */ UNK_RET grCastle_801D0298(UNK_PARAMS);
diff --git a/src/melee/gr/groldpupupu.c b/src/melee/gr/groldpupupu.c
index 5e674dcce6..c7c3a31ca9 100644
--- a/src/melee/gr/groldpupupu.c
+++ b/src/melee/gr/groldpupupu.c
@@ -20,6 +20,8 @@ static struct {
 
 static void* grOp_804D6A9C;
 
+static int grOp_804D48E0[1][2] = { { 0, 0 } };
+
 void grOldPupupu_802107DC(bool arg) {}
 
 void grOldPupupu_802107E0(void)
@@ -225,7 +227,18 @@ bool grOldPupupu_80211C14(Ground_GObj* gobj)
     return false;
 }
 
-/// #grOldPupupu_80211C1C
+void grOldPupupu_80211C1C(Ground_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    Ground* other_gp = Ground_801C2BA4(7)->user_data;
+    PAD_STACK(8);
+
+    if (other_gp->gv.unk.xC8 == 2 && other_gp->gv.unk.xD4 != 0) {
+        grAnime_801C8138(
+            gobj, gp->map_id,
+            grOp_804D48E0[other_gp->gv.unk.xCC][other_gp->gv.unk.xD8]);
+    }
+}
 
 void grOldPupupu_80211C9C(Ground_GObj* arg) {}
 
diff --git a/src/melee/ty/tydisplay.c b/src/melee/ty/tydisplay.c
index 32504870ad..5dc9d32e48 100644
--- a/src/melee/ty/tydisplay.c
+++ b/src/melee/ty/tydisplay.c
@@ -93,7 +93,60 @@ void* un_8031B9DC(s32 id)
 
 /// #un_8031BA78
 
-/// #un_8031BB34
+s32 un_8031BB34(s8 idx)
+{
+    char* table[] = {
+        "ToyDspQues_Top_joint",
+        "ToyDspMycharaCmA_Top_joint",
+        "ToyDspMycharaCmB_Top_joint",
+        "ToyDspMycharaCmC_Top_joint",
+        "ToyDspMycharaCmD_Top_joint",
+        "ToyDspMycharaCmE_Top_joint",
+        "ToyDspMycharaR1A_Top_joint",
+        "ToyDspMycharaR1B_Top_joint",
+        "ToyDspMycharaR1C_Top_joint",
+        "ToyDspMycharaR1D_Top_joint",
+        "ToyDspMycharaR1E_Top_joint",
+        "ToyDspMycharaR2A_Top_joint",
+        "ToyDspMycharaR2B_Top_joint",
+        "ToyDspMycharaR2C_Top_joint",
+        "ToyDspMycharaR2D_Top_joint",
+        "ToyDspMycharaR2E_Top_joint",
+        "ToyDspMapA_Top_joint",
+        "ToyDspMapB_Top_joint",
+        "ToyDspMapC_Top_joint",
+        "ToyDspMapD_Top_joint",
+        "ToyDspMapE_Top_joint",
+        "ToyDspFgseriesA_Top_joint",
+        "ToyDspFgseriesB_Top_joint",
+        "ToyDspFgseriesC_Top_joint",
+        "ToyDspFgseriesD_Top_joint",
+        "ToyDspFgseriesE_Top_joint",
+        "ToyDspFgetcA_Top_joint",
+        "ToyDspFgetcB_Top_joint",
+        "ToyDspFgetcC_Top_joint",
+        "ToyDspFgetcD_Top_joint",
+        "ToyDspFgetcE_Top_joint",
+        "ToyDspPokemonA_Top_joint",
+        "ToyDspPokemonB_Top_joint",
+        "ToyDspPokemonC_Top_joint",
+        "ToyDspPokemonD_Top_joint",
+        "ToyDspPokemonE_Top_joint",
+        "ToyDspItemA_Top_joint",
+        "ToyDspItemB_Top_joint",
+        "ToyDspItemC_Top_joint",
+        "ToyDspItemD_Top_joint",
+        "ToyDspItemE_Top_joint",
+        "ToyDspStand_Top_joint",
+        "ToyDspQues_Top_joint",
+    };
+
+    if (idx == -1) {
+        idx = 0;
+    }
+
+    return (s32) table[idx];
+}
 
 char* un_8031BB94(s8 idx)
 {
