diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 8ac3fbb527..a909d04733 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -14608,7 +14608,7 @@ itHammerHead_Logic40_PickedUp = .text:0x80299DD0; // type:function size:0x4C sco
 itHammerhead_UnkMotion1_Anim = .text:0x80299E1C; // type:function size:0x8 scope:global
 itHammerhead_UnkMotion1_Phys = .text:0x80299E24; // type:function size:0x4 scope:global
 itHammerHead_Logic40_Dropped = .text:0x80299E28; // type:function size:0x20 scope:global
-it_3F14_Logic40_Thrown = .text:0x80299E48; // type:function size:0x54 scope:global
+itHammerHead_Logic40_Thrown = .text:0x80299E48; // type:function size:0x54 scope:global
 itHammerhead_UnkMotion2_Anim = .text:0x80299E9C; // type:function size:0x80 scope:global
 itHammerhead_UnkMotion2_Phys = .text:0x80299F1C; // type:function size:0x4C scope:global
 itHammerhead_UnkMotion2_Coll = .text:0x80299F68; // type:function size:0x2C scope:global
@@ -32969,10 +32969,10 @@ it_804DCC48 = .sdata2:0x804DCC48; // type:object size:0x8 scope:global align:8 d
 it_804DCC50 = .sdata2:0x804DCC50; // type:object size:0x8 scope:global align:8 data:double
 it_804DCC58 = .sdata2:0x804DCC58; // type:object size:0x4 scope:global align:4 data:float
 it_804DCC5C = .sdata2:0x804DCC5C; // type:object size:0x4 scope:global align:4 data:float
-it_804DCC60 = .sdata2:0x804DCC60; // type:object size:0x4 scope:global align:4 data:float
-it_804DCC64 = .sdata2:0x804DCC64; // type:object size:0x4 scope:global align:4 data:float
-it_804DCC68 = .sdata2:0x804DCC68; // type:object size:0x4 scope:global align:4 data:float
-it_804DCC70 = .sdata2:0x804DCC70; // type:object size:0x8 scope:global align:8 data:double
+@151 = .sdata2:0x804DCC60; // type:object size:0x4 scope:local align:4 data:float
+@152 = .sdata2:0x804DCC64; // type:object size:0x4 scope:local align:4 data:float
+@177 = .sdata2:0x804DCC68; // type:object size:0x4 scope:local align:4 data:float
+@179 = .sdata2:0x804DCC70; // type:object size:0x8 scope:local align:8 data:double
 it_804DCC78 = .sdata2:0x804DCC78; // type:object size:0x4 scope:global align:4 data:float
 it_804DCC7C = .sdata2:0x804DCC7C; // type:object size:0x4 scope:global align:4 data:float
 @206 = .sdata2:0x804DCC80; // type:object size:0x4 scope:local align:4 data:float
diff --git a/configure.py b/configure.py
index a695d30cbf..c5bdbe9a5e 100755
--- a/configure.py
+++ b/configure.py
@@ -1144,7 +1144,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itstarrodstar.c"),
             Object(NonMatching, "melee/it/items/itsscopebeam.c"),
             Object(NonMatching, "melee/it/items/itlgunbeam.c"),
-            Object(NonMatching, "melee/it/items/ithammerhead.c"),
+            Object(Matching, "melee/it/items/ithammerhead.c"),
             Object(NonMatching, "melee/it/items/itlipstickspore.c"),
             Object(Matching, "melee/it/items/itfflowerflame.c"),
             Object(NonMatching, "melee/it/items/itevyoshiegg.c"),
diff --git a/src/melee/it/itCommonItems.h b/src/melee/it/itCommonItems.h
index 445eeb7e1c..3ebae7d716 100644
--- a/src/melee/it/itCommonItems.h
+++ b/src/melee/it/itCommonItems.h
@@ -458,7 +458,7 @@ typedef struct itLikelike_ItemVars {
 } itLikelike_ItemVars;
 
 typedef struct {
-    f32 x0;
+    f32 initial_velocity;
     f32 x4;
 } itHammerheadAttributes;
 
@@ -1225,6 +1225,11 @@ typedef struct itSScopeAttributes {
     /* +34 */ Vec3 x34;
 } itSScopeAttributes;
 
+typedef struct itScopeBeam_ItemVars {
+    int x0;
+    float x4;
+} itScopeBeam_ItemVars;
+
 typedef struct itsonansAttributes {
     f32 x0;
     f32 x4;
diff --git a/src/melee/it/it_2725.c b/src/melee/it/it_2725.c
index 0519f00139..b03b603b82 100644
--- a/src/melee/it/it_2725.c
+++ b/src/melee/it/it_2725.c
@@ -1007,7 +1007,7 @@ struct ItemLogicTable it_803F14C4[ARRAY_SIZE(it_803F1418)] = {
         NULL,
         itHammerHead_Logic40_PickedUp,
         itHammerHead_Logic40_Dropped,
-        it_3F14_Logic40_Thrown,
+        itHammerHead_Logic40_Thrown,
         itHammerHead_Logic40_DmgDealt,
         itHammerHead_Logic40_DmgReceived,
         NULL,
diff --git a/src/melee/it/it_3F14.c b/src/melee/it/it_3F14.c
index 9b2c6a6493..fb4d947087 100644
--- a/src/melee/it/it_3F14.c
+++ b/src/melee/it/it_3F14.c
@@ -788,7 +788,7 @@
 ///         NULL,
 ///         itHammerHead_Logic40_PickedUp,
 ///         itHammerHead_Logic40_Dropped,
-///         it_3F14_Logic40_Thrown,
+///         itHammerHead_Logic40_Thrown,
 ///         itHammerHead_Logic40_DmgDealt,
 ///         itHammerHead_Logic40_DmgReceived,
 ///         NULL,
diff --git a/src/melee/it/items/itfoods.c b/src/melee/it/items/itfoods.c
index 50c157a511..287f4ec9b7 100644
--- a/src/melee/it/items/itfoods.c
+++ b/src/melee/it/items/itfoods.c
@@ -73,16 +73,23 @@ HSD_GObj* it_8028FAF4(Item_GObj* arg0, Vec3* arg1)
     return gobj;
 }
 
-void itFoods_Logic18_Spawned(Item_GObj* arg0)
+static s32 getRandMax(Item* ip)
 {
-    Item* ip = GET_ITEM(arg0);
     itFoodsAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
-    s32 rand = HSD_Randi(*(s32*) ip->xC4_article_data->x4_specialAttributes);
-    PAD_STACK(12);
+    return attr->x0;
+}
+
+void itFoods_Logic18_Spawned(HSD_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    itFoodsAttributes* attr = ip->xC4_article_data->x4_specialAttributes;
+    s32 rand = HSD_Randi(getRandMax(ip));
+    PAD_STACK(8);
+
     ip->xDD4_itemVar.foods.heal_amount = attr[rand].x8;
     ip->xDD4_itemVar.foods.x0 = rand;
-    it_80273318(arg0, attr[rand].x4, rand);
-    it_8028FC5C((HSD_GObj*) arg0);
+    it_80273318(gobj, attr[rand].x4, rand);
+    it_8028FC5C(gobj);
 }
 
 void it_8028FC5C(HSD_GObj* arg0)
diff --git a/src/melee/it/items/ithammerhead.c b/src/melee/it/items/ithammerhead.c
index 3503131568..3fb21d9622 100644
--- a/src/melee/it/items/ithammerhead.c
+++ b/src/melee/it/items/ithammerhead.c
@@ -1,8 +1,11 @@
 #include "ithammerhead.h"
 
-#include <placeholder.h>
+#include "common_structs.h"
+
 #include <platform.h>
 
+#include "db/db.h"
+
 #include "it/forward.h"
 
 #include "it/inlines.h"
@@ -11,7 +14,65 @@
 #include "it/it_2725.h"
 #include "it/item.h"
 
-/// #it_80299C48
+#include "it/items/forward.h"
+
+ItemStateTable it_803F6640[] = {
+    {
+        -1,
+        itHammerhead_UnkMotion2_Anim,
+        itHammerhead_UnkMotion2_Phys,
+        itHammerhead_UnkMotion2_Coll,
+    },
+    {
+        -1,
+        itHammerhead_UnkMotion1_Anim,
+        itHammerhead_UnkMotion1_Phys,
+        NULL,
+    },
+    {
+        0,
+        itHammerhead_UnkMotion2_Anim,
+        itHammerhead_UnkMotion2_Phys,
+        itHammerhead_UnkMotion2_Coll,
+    },
+    {
+        -1,
+        itHammerhead_UnkMotion3_Anim,
+        itHammerhead_UnkMotion3_Phys,
+        itHammerhead_UnkMotion3_Coll,
+    },
+};
+
+void it_80299C48(Item_GObj* parent_gobj, Vec3* pos, Vec3* velocity,
+                 f32 facing_dir)
+{
+    SpawnItem spawn;
+    Item_GObj* gobj;
+    Item* it;
+    itHammerheadAttributes* attrs;
+
+    spawn.kind = It_Kind_Hammer_Head;
+    spawn.prev_pos = *pos;
+    spawn.prev_pos.z = 0.0f;
+    it_8026BB68(parent_gobj, &spawn.pos);
+    spawn.facing_dir = facing_dir;
+    spawn.x3C_damage = 0;
+    spawn.vel.x = spawn.vel.y = spawn.vel.z = 0.0f;
+    spawn.x0_parent_gobj = parent_gobj;
+    spawn.x4_parent_gobj2 = spawn.x0_parent_gobj;
+    spawn.x44_flag.b0 = 1;
+    spawn.x40 = 0;
+    gobj = Item_80268B18(&spawn);
+    if (gobj != NULL) {
+        it = GET_ITEM(gobj);
+        attrs = it->xC4_article_data->x4_specialAttributes;
+        it_80299D7C(gobj);
+        it->facing_dir = facing_dir;
+        it->x40_vel.x = velocity->x * attrs->initial_velocity * it->facing_dir;
+        it->x40_vel.y = velocity->y * attrs->initial_velocity + 0.5f;
+        db_80225DD8(gobj, parent_gobj);
+    }
+}
 
 void itHammerHead_Logic40_Spawned(Item_GObj* gobj)
 {
@@ -20,7 +81,15 @@ void itHammerHead_Logic40_Spawned(Item_GObj* gobj)
     it_80275158(gobj, attrs->x4);
 }
 
-/// #it_80299D7C
+void it_80299D7C(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (ip->ground_or_air != GA_Air) {
+        it_802762BC(ip);
+    }
+    it_8026B390(gobj);
+    Item_80268E5C(gobj, 0, ITEM_UNK_0x1 | ITEM_DROP_UPDATE);
+}
 
 void itHammerHead_Logic40_PickedUp(Item_GObj* gobj)
 {
@@ -39,10 +108,33 @@ void itHammerhead_UnkMotion1_Phys(Item_GObj* gobj) {}
 
 void itHammerHead_Logic40_Dropped(Item_GObj* gobj)
 {
-    it_3F14_Logic40_Thrown(gobj);
+    itHammerHead_Logic40_Thrown(gobj);
+}
+
+void itHammerHead_Logic40_Thrown(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (ip->ground_or_air != GA_Air) {
+        it_802762BC(ip);
+    }
+    it_8026B390(gobj);
+    Item_80268E5C(gobj, 2, ITEM_ANIM_UPDATE | ITEM_DROP_UPDATE);
 }
 
-/// #itHammerhead_UnkMotion2_Anim
+bool itHammerhead_UnkMotion2_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+
+    ip->xD44_lifeTimer--;
+    if (ip->xD44_lifeTimer <= 0.0f) {
+        return true;
+    }
+
+    if (ip->xD44_lifeTimer <= it_804D6D28->x34) {
+        it_802728C8(gobj);
+    }
+    return false;
+}
 
 void itHammerhead_UnkMotion2_Phys(Item_GObj* gobj)
 {
@@ -78,13 +170,26 @@ void it_80299FB4(Item_GObj* gobj)
     Item_80268E5C(gobj, 3, 1);
 }
 
-/// #itHammerhead_UnkMotion3_Anim
+bool itHammerhead_UnkMotion3_Anim(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+
+    ip->xD44_lifeTimer--;
+    if (ip->xD44_lifeTimer <= 0.0f) {
+        return true;
+    }
+
+    if (ip->xD44_lifeTimer <= it_804D6D28->x34) {
+        it_802728C8(gobj);
+    }
+    return false;
+}
 
 void itHammerhead_UnkMotion3_Phys(Item_GObj* gobj) {}
 
 bool itHammerhead_UnkMotion3_Coll(Item_GObj* gobj)
 {
-    it_8026D62C(gobj, it_3F14_Logic40_Thrown);
+    it_8026D62C(gobj, itHammerHead_Logic40_Thrown);
     return false;
 }
 
diff --git a/src/melee/it/items/ithammerhead.h b/src/melee/it/items/ithammerhead.h
index b429e1754b..c71e970de7 100644
--- a/src/melee/it/items/ithammerhead.h
+++ b/src/melee/it/items/ithammerhead.h
@@ -13,12 +13,12 @@
 
 /* 299C48 */ void it_80299C48(HSD_GObj*, Vec3*, Vec3*, f32);
 /* 299D4C */ void itHammerHead_Logic40_Spawned(Item_GObj*);
-/* 299D7C */ UNK_RET it_80299D7C(UNK_PARAMS);
+/* 299D7C */ void it_80299D7C(Item_GObj*);
 /* 299DD0 */ void itHammerHead_Logic40_PickedUp(Item_GObj*);
 /* 299E1C */ bool itHammerhead_UnkMotion1_Anim(Item_GObj* gobj);
 /* 299E24 */ void itHammerhead_UnkMotion1_Phys(Item_GObj* gobj);
 /* 299E28 */ void itHammerHead_Logic40_Dropped(Item_GObj* gobj);
-/* 299E48 */ void it_3F14_Logic40_Thrown(Item_GObj*);
+/* 299E48 */ void itHammerHead_Logic40_Thrown(Item_GObj*);
 /* 299E9C */ bool itHammerhead_UnkMotion2_Anim(Item_GObj* gobj);
 /* 299F1C */ void itHammerhead_UnkMotion2_Phys(Item_GObj* gobj);
 /* 299F68 */ bool itHammerhead_UnkMotion2_Coll(Item_GObj* gobj);
diff --git a/src/melee/it/items/itsscopebeam.c b/src/melee/it/items/itsscopebeam.c
index e850465a8d..068f39c496 100644
--- a/src/melee/it/items/itsscopebeam.c
+++ b/src/melee/it/items/itsscopebeam.c
@@ -3,6 +3,9 @@
 #include <placeholder.h>
 #include <platform.h>
 
+#include "baselib/forward.h"
+
+#include "baselib/gobj.h"
 #include "db/db.h"
 
 #include "it/forward.h"
@@ -13,6 +16,16 @@
 #include "it/item.h"
 #include "it/types.h"
 
+typedef struct BeamFloats {
+    float velocity;
+    float scale;
+    float lifetime;
+} BeamFloats;
+
+typedef struct ScopeBeamAttrs {
+    BeamFloats floats[9];
+} ScopeBeamAttrs;
+
 void it_80298DEC(Fighter_GObj* gobj, Vec* vec, int arg2, float arg3)
 {
     HSD_GObj* n;
@@ -31,24 +44,73 @@ void it_80298DEC(Fighter_GObj* gobj, Vec* vec, int arg2, float arg3)
     n = Item_80268B18(&si);
     if (n != NULL) {
         Item* ip = GET_ITEM(n);
-        ip->xDD4_itemVar.starrodstar.x0 = arg2;
+        ip->xDD4_itemVar.scopebeam.x0 = arg2;
         it_80298ED0(n, gobj);
         db_80225DD8(n, gobj);
     }
 }
 
-/// #it_80298ED0
+void it_80298ED0(HSD_GObj* projectile, HSD_GObj* owner)
+{
+    Item* it = GET_ITEM(projectile);
+    BeamFloats* data;
+    HSD_JObj* jobj = GET_JOBJ(projectile);
+    ScopeBeamAttrs* attrs = it->xC4_article_data->x4_specialAttributes;
+    Vec3 scale;
+
+    it->owner = owner;
+    data = &attrs->floats[it->xDD4_itemVar.scopebeam.x0];
+    it_8026B3A8(projectile);
+    switch (it->xDD4_itemVar.scopebeam.x0) {
+    case 0:
+        Item_80268E5C(projectile, 0, ITEM_ANIM_UPDATE);
+        break;
+    case 1:
+        Item_80268E5C(projectile, 1, ITEM_ANIM_UPDATE);
+        break;
+    case 2:
+        Item_80268E5C(projectile, 2, ITEM_ANIM_UPDATE);
+        break;
+    case 3:
+        Item_80268E5C(projectile, 3, ITEM_ANIM_UPDATE);
+        break;
+    case 4:
+        Item_80268E5C(projectile, 4, ITEM_ANIM_UPDATE);
+        break;
+    case 5:
+        Item_80268E5C(projectile, 5, ITEM_ANIM_UPDATE);
+        break;
+    case 6:
+        Item_80268E5C(projectile, 6, ITEM_ANIM_UPDATE);
+        break;
+    case 7:
+        Item_80268E5C(projectile, 7, ITEM_ANIM_UPDATE);
+        break;
+    case 8:
+        Item_80268E5C(projectile, 8, ITEM_ANIM_UPDATE);
+        break;
+    case 9:
+        Item_80268E5C(projectile, 9, ITEM_ANIM_UPDATE);
+        break;
+    }
+
+    it->x40_vel.x = data->velocity * it->facing_dir;
+    scale.x = scale.y = scale.z = data->scale;
+    HSD_JObjSetScale(jobj, &scale);
+    it_80275158(projectile, data->lifetime);
+}
 
 bool itSscopebeam_UnkMotion9_Anim(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
-    s32 index = ip->xDD4_itemVar.starrodstar.x0;
-    HSD_FObj* fobj = ip->xC4_article_data->x4_specialAttributes;
-    f32* frame_data = (f32*)((u8*)fobj + index * 0xC);
+    s32 index = ip->xDD4_itemVar.scopebeam.x0;
+    ScopeBeamAttrs* attrs = ip->xC4_article_data->x4_specialAttributes;
+    BeamFloats* data = &attrs->floats[index];
 
-    if (ip->xD44_lifeTimer == frame_data[2]) {
+    if (ip->xD44_lifeTimer == data->lifetime) {
         if (ip->x5D4_hitboxes[0].hit.state >= 1) {
-            ip->x5D4_hitboxes[0].hit.scale = ip->x5D4_hitboxes[0].hit.scale * frame_data[1];
+            ip->x5D4_hitboxes[0].hit.scale =
+                ip->x5D4_hitboxes[0].hit.scale * data->scale;
         }
     }
 
diff --git a/src/melee/it/types.h b/src/melee/it/types.h
index 588cbd038f..cd66f88a0a 100644
--- a/src/melee/it/types.h
+++ b/src/melee/it/types.h
@@ -616,6 +616,7 @@ struct Item {
         itSamusChargeshot_ItemVars samuschargeshot;
         itSamusGrapple_ItemVars samusgrapple;
         itSamusMissile_ItemVars samusmissile;
+        itScopeBeam_ItemVars scopebeam;
         itSeakChain_ItemVars seakchain;
         itSeakNeedleHeld_ItemVars seakneedleheld;
         itSeakNeedleThrown_ItemVars seakneedlethrown;
