diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 018db8c181..7a68c60709 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -11242,10 +11242,10 @@ grKinokoRoute_802086EC = .text:0x802086EC; // type:function size:0x68 scope:glob
 grKinokoRoute_80208754 = .text:0x80208754; // type:function size:0x8 scope:global
 grKinokoRoute_8020875C = .text:0x8020875C; // type:function size:0x54 scope:global
 grShrineRoute_802087B0 = .text:0x802087B0; // type:function size:0x8 scope:global
-grShrineRoute_802087B8 = .text:0x802087B8; // type:function size:0x4 scope:global
-grShrineRoute_802087BC = .text:0x802087BC; // type:function size:0x70 scope:global
-grShrineRoute_8020882C = .text:0x8020882C; // type:function size:0x24 scope:global
-grShrineRoute_80208850 = .text:0x80208850; // type:function size:0x68 scope:global
+grShrineRoute_OnDemoInit = .text:0x802087B8; // type:function size:0x4 scope:global
+grShrineRoute_OnInit = .text:0x802087BC; // type:function size:0x70 scope:global
+grShrineRoute_OnLoad = .text:0x8020882C; // type:function size:0x24 scope:global
+grShrineRoute_OnStart = .text:0x80208850; // type:function size:0x68 scope:global
 grShrineRoute_802088B8 = .text:0x802088B8; // type:function size:0x8 scope:global
 grShrineRoute_802088C0 = .text:0x802088C0; // type:function size:0xEC scope:global
 grShrineRoute_802089AC = .text:0x802089AC; // type:function size:0x2C scope:global
@@ -11278,14 +11278,14 @@ grShrineRoute_8020A8A4 = .text:0x8020A8A4; // type:function size:0x19C scope:glo
 grShrineRoute_8020AA40 = .text:0x8020AA40; // type:function size:0x118 scope:global
 grShrineRoute_8020AB58 = .text:0x8020AB58; // type:function size:0xEC scope:global
 grShrineRoute_8020AC44 = .text:0x8020AC44; // type:function size:0xE0 scope:global
-fn_8020AD24 = .text:0x8020AD24; // type:function size:0x34
-fn_8020AD58 = .text:0x8020AD58; // type:function size:0xB0
-fn_8020AE08 = .text:0x8020AE08; // type:function size:0x130
+grShrineRoute_8020AD24 = .text:0x8020AD24; // type:function size:0x34
+grShrineRoute_8020AD58 = .text:0x8020AD58; // type:function size:0xB0
+grShrineRoute_8020AE08 = .text:0x8020AE08; // type:function size:0x130
 grShrineRoute_8020AF38 = .text:0x8020AF38; // type:function size:0xE8 scope:global
 grShrineRoute_8020B020 = .text:0x8020B020; // type:function size:0x8C scope:global
 grShrineRoute_8020B0AC = .text:0x8020B0AC; // type:function size:0x58 scope:global
-grShrineRoute_8020B104 = .text:0x8020B104; // type:function size:0x8 scope:global
-grShrineRoute_8020B10C = .text:0x8020B10C; // type:function size:0x54 scope:global
+grShrineRoute_OnTouchLine = .text:0x8020B104; // type:function size:0x8 scope:global
+grShrineRoute_OnCheckShadowRender = .text:0x8020B10C; // type:function size:0x54 scope:global
 grZebesRoute_8020B160 = .text:0x8020B160; // type:function size:0x4 scope:global
 grZebesRoute_8020B164 = .text:0x8020B164; // type:function size:0x70 scope:global
 grZebesRoute_8020B1D4 = .text:0x8020B1D4; // type:function size:0x20 scope:global
diff --git a/src/melee/gr/grshrineroute.c b/src/melee/gr/grshrineroute.c
index efe1bfc7db..c4e8eac17d 100644
--- a/src/melee/gr/grshrineroute.c
+++ b/src/melee/gr/grshrineroute.c
@@ -12,9 +12,11 @@
 #include "lb/lb_00B0.h"
 #include "mp/mplib.h"
 
+#include <baselib/debug.h>
 #include <baselib/gobj.h>
 #include <baselib/gobjgxlink.h>
 #include <baselib/gobjproc.h>
+#include <baselib/lobj.h>
 #include <baselib/random.h>
 
 s16 grSh_Route_803E58E0[8] = {
@@ -77,13 +79,13 @@ StageData grSh_Route_803E5988 = {
     SHRINEROUTE,
     grSh_Route_803E58F0,
     "/GrNSr.dat",
-    grShrineRoute_802087BC,
-    grShrineRoute_802087B8,
-    grShrineRoute_8020882C,
-    grShrineRoute_80208850,
+    grShrineRoute_OnInit,
+    grShrineRoute_OnDemoInit,
+    grShrineRoute_OnLoad,
+    grShrineRoute_OnStart,
     grShrineRoute_802088B8,
-    grShrineRoute_8020B104,
-    grShrineRoute_8020B10C,
+    grShrineRoute_OnTouchLine,
+    grShrineRoute_OnCheckShadowRender,
     (1 << 0),
     NULL,
     0,
@@ -110,9 +112,9 @@ float grShrineRoute_802087B0(void)
     return grNKr_804DB868;
 }
 
-void grShrineRoute_802087B8(bool arg) {}
+void grShrineRoute_OnDemoInit(bool arg) {}
 
-void grShrineRoute_802087BC(void)
+void grShrineRoute_OnInit(void)
 {
     grSh_Route_804D6A58 = Ground_801C49F8();
     stage_info.unk8C.b4 = 0;
@@ -125,12 +127,12 @@ void grShrineRoute_802087BC(void)
     grShrineRoute_8020B0AC();
 }
 
-void grShrineRoute_8020882C(void)
+void grShrineRoute_OnLoad(void)
 {
     grShrineRoute_802088C0(6);
 }
 
-void grShrineRoute_80208850(void)
+void grShrineRoute_OnStart(void)
 {
     int val;
     grZakoGenerator_801CAE04(&grSh_Route_804D6A58->x28);
@@ -253,7 +255,38 @@ bool grShrineRoute_80209BE4(Ground_GObj* arg)
 
 void grShrineRoute_8020A100(Ground_GObj* arg) {}
 
-/// #grShrineRoute_8020A104
+void grShrineRoute_8020A104(Ground_GObj* gobj)
+{
+    Ground* gp = GET_GROUND(gobj);
+    HSD_GObj* lgobj = HSD_GObj_Entities->xC;
+    HSD_LObj* lobj;
+
+    while (lgobj) {
+        if (HSD_GObjGetClassifier(lgobj) != 0xC) {
+            lgobj = HSD_GObjGetNext(lgobj);
+        } else {
+            break;
+        }
+    }
+    HSD_ASSERT(1106, lgobj);
+    gp->gv.shrineroute2.xC4 = lgobj;
+    grShrineRoute_8020AA40(gp->gv.shrineroute2.xC4);
+
+    lobj = GET_LOBJ(gp->gv.shrineroute2.xC4);
+    gp->gv.shrineroute2.x168 = 0;
+    while (lobj != NULL) {
+        (&gp->gv.shrineroute2.xC8)[gp->gv.shrineroute2.x168] = lobj;
+        (&gp->gv.shrineroute2.x118)[gp->gv.shrineroute2.x168] =
+            HSD_LObjGetFlags(lobj);
+
+        lobj = HSD_LObjGetNext(lobj);
+        gp->gv.shrineroute2.x168 += 1;
+    }
+
+    gp->gv.shrineroute2.x16C = grShrineRoute_8020AB58(gp->gv.shrineroute2.xC4);
+    gp->gv.shrineroute2.x170 = grShrineRoute_8020AC44(gp->gv.shrineroute2.xC4);
+    grShrineRoute_8020A21C(gobj);
+}
 
 bool grShrineRoute_8020A214(Ground_GObj* arg)
 {
@@ -287,7 +320,7 @@ void grShrineRoute_8020A8A0(Ground_GObj* arg) {}
 
 /// #grShrineRoute_8020AC44
 
-bool fn_8020AD24(bool arg)
+bool grShrineRoute_8020AD24(bool arg)
 {
     if (arg == 1) {
         Ground_801C5740(arg);
@@ -296,9 +329,9 @@ bool fn_8020AD24(bool arg)
     return 0;
 }
 
-/// #fn_8020AD58
+/// #grShrineRoute_8020AD58
 
-/// #fn_8020AE08
+/// #grShrineRoute_8020AE08
 
 /// #grShrineRoute_8020AF38
 
@@ -313,12 +346,12 @@ void grShrineRoute_8020B0AC(void)
     }
 }
 
-DynamicsDesc* grShrineRoute_8020B104(enum_t arg)
+DynamicsDesc* grShrineRoute_OnTouchLine(enum_t arg)
 {
     return false;
 }
 
-bool grShrineRoute_8020B10C(Vec3* a, int b, HSD_JObj* jobj)
+bool grShrineRoute_OnCheckShadowRender(Vec3* a, int b, HSD_JObj* jobj)
 {
     Vec3 vec;
     lb_8000B1CC(jobj, NULL, &vec);
diff --git a/src/melee/gr/grshrineroute.h b/src/melee/gr/grshrineroute.h
index b1a576edfd..d03266abcb 100644
--- a/src/melee/gr/grshrineroute.h
+++ b/src/melee/gr/grshrineroute.h
@@ -3,13 +3,17 @@
 
 #include <placeholder.h>
 
+#include "forward.h"
+
 #include "it/inlines.h"
 
+#include <baselib/forward.h>
+
 /* 2087B0 */ float grShrineRoute_802087B0(void);
-/* 2087B8 */ void grShrineRoute_802087B8(bool);
-/* 2087BC */ void grShrineRoute_802087BC(void);
-/* 20882C */ void grShrineRoute_8020882C(void);
-/* 208850 */ void grShrineRoute_80208850(void);
+/* 2087B8 */ void grShrineRoute_OnDemoInit(bool);
+/* 2087BC */ void grShrineRoute_OnInit(void);
+/* 20882C */ void grShrineRoute_OnLoad(void);
+/* 208850 */ void grShrineRoute_OnStart(void);
 /* 2088B8 */ bool grShrineRoute_802088B8(void);
 /* 2088C0 */ HSD_GObj* grShrineRoute_802088C0(int);
 /* 2089AC */ void grShrineRoute_802089AC(Ground_GObj*);
@@ -39,16 +43,17 @@
 /* 20A89C */ void grShrineRoute_8020A89C(Ground_GObj*);
 /* 20A8A0 */ void grShrineRoute_8020A8A0(Ground_GObj*);
 /* 20A8A4 */ UNK_RET grShrineRoute_8020A8A4(UNK_PARAMS);
-/* 20AA40 */ UNK_RET grShrineRoute_8020AA40(UNK_PARAMS);
-/* 20AB58 */ UNK_RET grShrineRoute_8020AB58(UNK_PARAMS);
-/* 20AC44 */ UNK_RET grShrineRoute_8020AC44(UNK_PARAMS);
-/* 20AD24 */ bool fn_8020AD24(bool);
-/* 20AD58 */ UNK_RET fn_8020AD58(UNK_PARAMS);
-/* 20AE08 */ UNK_RET fn_8020AE08(UNK_PARAMS);
+/* 20AA40 */ void grShrineRoute_8020AA40(HSD_GObj*);
+/* 20AB58 */ HSD_LObj* grShrineRoute_8020AB58(Ground_GObj*);
+/* 20AC44 */ HSD_LObj* grShrineRoute_8020AC44(HSD_GObj*);
+/* 20AD24 */ bool grShrineRoute_8020AD24(bool);
+/* 20AD58 */ UNK_RET grShrineRoute_8020AD58(UNK_PARAMS);
+/* 20AE08 */ UNK_RET grShrineRoute_8020AE08(UNK_PARAMS);
 /* 20AF38 */ UNK_RET grShrineRoute_8020AF38(UNK_PARAMS);
 /* 20B020 */ void grShrineRoute_8020B020(HSD_GObj* gobj, int r4, bool hide);
 /* 20B0AC */ void grShrineRoute_8020B0AC(void);
-/* 20B104 */ DynamicsDesc* grShrineRoute_8020B104(enum_t);
-/* 20B10C */ bool grShrineRoute_8020B10C(Vec3*, int arg, HSD_JObj* jobj);
+/* 20B104 */ DynamicsDesc* grShrineRoute_OnTouchLine(enum_t);
+/* 20B10C */ bool grShrineRoute_OnCheckShadowRender(Vec3*, int arg,
+                                                    HSD_JObj* jobj);
 
 #endif
diff --git a/src/melee/gr/types.h b/src/melee/gr/types.h
index f6a5f8a301..a72d790dc9 100644
--- a/src/melee/gr/types.h
+++ b/src/melee/gr/types.h
@@ -9,12 +9,12 @@
 #include <melee/gr/forward.h>
 #include <melee/it/forward.h>
 #include <melee/lb/forward.h>
-#include <melee/lb/types.h>
 #include <melee/sc/forward.h>
 #include <sysdolphin/baselib/forward.h>
 
 #include <dolphin/gx.h>
 #include <dolphin/mtx.h>
+#include <melee/lb/types.h>
 
 typedef struct StageBlastZone {
     f32 left;   // 0x74
@@ -988,12 +988,23 @@ struct ShyGuys {
 };
 
 struct grShrineroute_GroundVars {
-    /* +0 gp+C4 */ u8 _[0x4];
+    /* +0 gp+C4 */ u32 xC4;
     /* +4 gp+C8 */ u16 xC8;
     /* +6 gp+CA */ u16 xCA;
     /* +8 gp+CA */ u16 xCC;
 };
 
+struct grShrineroute_GroundVars2 {
+    /* +0 gp+C4 */ HSD_GObj* xC4;
+    /* +4 gp+C8 */ HSD_LObj* xC8;
+    u8 _[0x118 - 0xCC];
+    u32 x118;
+    u8 _2[0x168 - 0x11C];
+    /* +0 gp+168 */ u32 x168;
+    /* +0 gp+16C */ HSD_LObj* x16C;
+    /* +0 gp+170 */ HSD_LObj* x170;
+};
+
 struct Battlefield {
     /* +0 gp+C4 */ int bg_state;
     /* +4 gp+C8 */ int curr_bg;
@@ -1144,6 +1155,7 @@ struct Ground {
             struct grPura_GroundVars2 pura2;
             struct grRCruise_GroundVars rcruise;
             struct grShrineroute_GroundVars shrineroute;
+            struct grShrineroute_GroundVars2 shrineroute2;
             struct grSmashTaunt_GroundVars smashtaunt;
             struct GroundVars_unk unk;
             struct grVenom_GroundVars venom;
