diff --git a/src/melee/gr/grmaterial.c b/src/melee/gr/grmaterial.c
index 7cb7300de1..b1d784e149 100644
--- a/src/melee/gr/grmaterial.c
+++ b/src/melee/gr/grmaterial.c
@@ -25,7 +25,7 @@ grMaterial_801C8D44(int arg0, int arg1, Ground* arg2, Vec3* arg3, int arg4,
 /* 1C8E48 */ static bool grMaterial_801C8E48(HSD_GObj* gobj);
 /* 1C8E74 */ static void grMaterial_801C8E74(void);
 /* 1C8EF8 */ static void fn_801C8EF8(HSD_MObj* mobj, u32 rendermode);
-/* 1C9490 */ void grMaterial_801C9490(HSD_JObj* jobj);
+/* 1C9490 */ void grMaterial_801C9490(Item_GObj* gobj, CommandInfo* cmd);
 /* 3E0A20 */ static HSD_MObjInfo grMaterial_803E0A20 = { 0 };
 /* 4D456C */ static ItCmd grMaterial_804D456C[1];
 
@@ -247,12 +247,18 @@ void grMaterial_801C8E74(void)
 
 /// #grMaterial_801C92C0
 
-void grMaterial_801C9470(HSD_JObj* jobj)
+void grMaterial_801C9470(Item_GObj* gobj, CommandInfo* cmd)
 {
-    grMaterial_801C9490(jobj);
+    grMaterial_801C9490(gobj, cmd);
 }
 
-/// #grMaterial_801C9490
+void grMaterial_801C9490(Item_GObj* gobj, CommandInfo* cmd)
+{
+    Ground* gp = gobj->user_data;
+    u32 val = (*(u16*) cmd->ptr[0] >> 2) & 0xFF;
+    *(f32*) ((u8*) gp + 0xC0) = (f32) val;
+    gp->x10_flags.b6 = 1;
+}
 
 void grMaterial_801C94D8(void* obj)
 {
@@ -310,7 +316,18 @@ void grMaterial_801C95C4(HSD_GObj* gobj)
     gp->x10_flags.b4 = 1;
 }
 
-/// #grMaterial_801C9604
+void grMaterial_801C9604(HSD_GObj* gobj, int arg1, bool arg2)
+{
+    Ground* gp = gobj->user_data;
+    ColorOverlay* co = (ColorOverlay*) ((u8*) gp + 0x40);
+    co->x4_pri = arg2;
+    co->x8_ptr1 = (union ColorOverlay_x8_t*) arg1;
+    co->x0_timer = 0;
+    co->xC_loop = 0;
+    co->x7C_color_enable = co->x7C_flag2 = 0;
+    gp->x10_flags.b4 = 0;
+    grMaterial_801C9698(gobj);
+}
 
 void fn_801C9664(Item_GObj* gobj, CommandInfo* cmd, int arg2)
 {
@@ -318,4 +335,10 @@ void fn_801C9664(Item_GObj* gobj, CommandInfo* cmd, int arg2)
     grMaterial_804D456C[idx](gobj, cmd);
 }
 
-/// #grMaterial_801C9698
+void grMaterial_801C9698(HSD_GObj* gobj)
+{
+    Ground* gp = gobj->user_data;
+    if (lb_80014258(gobj, (ColorOverlay*) ((u8*) gp + 0x40), fn_801C9664)) {
+        gp->x10_flags.b4 = 1;
+    }
+}
diff --git a/src/melee/gr/grmaterial.h b/src/melee/gr/grmaterial.h
index c9e9300039..aecf27a3bb 100644
--- a/src/melee/gr/grmaterial.h
+++ b/src/melee/gr/grmaterial.h
@@ -5,6 +5,7 @@
 #include <platform.h>
 
 #include "gr/forward.h"
+#include "lb/forward.h"
 #include <baselib/forward.h>
 #include <melee/it/forward.h>
 
@@ -28,7 +29,7 @@ grMaterial_801C8CFC(int, int, Ground*, HSD_JObj*,
 /* 1C8E28 */ void grMaterial_801C8E28(HSD_GObj*);
 /* 1C8E68 */ void grMaterial_801C8E68(HSD_GObj*, GroundOrAir);
 /* 1C92C0 */ void grMaterial_801C92C0(HSD_JObj*);
-/* 1C9470 */ void grMaterial_801C9470(HSD_JObj* jobj);
+/* 1C9470 */ void grMaterial_801C9470(Item_GObj*, CommandInfo*);
 /* 1C94D8 */ void grMaterial_801C94D8(UNK_T hsd_obj);
 /* 1C95C4 */ void grMaterial_801C95C4(HSD_GObj*);
 /* 1C9604 */ void grMaterial_801C9604(HSD_GObj* bg, int, bool);
