diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index f5c84eff3a..7fd24f0231 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -17,6 +17,7 @@
 #include "ft/ft_081B.h"
 #include "ft/ft_0877.h"
 #include "ft/ft_0892.h"
+#include "ft/ftanim.h"
 #include "ft/ftcliffcommon.h"
 #include "ft/ftcommon.h"
 #include "ft/ftdata.h"
@@ -25,6 +26,7 @@
 #include "ft/inlines.h"
 #include "ft/types.h"
 #include "ftCommon/ftCo_CaptureKirby.h"
+#include "ftCommon/ftCo_CaptureWaitKirby.h"
 #include "ftCommon/ftCo_Fall.h"
 #include "ftCommon/ftCo_Landing.h"
 
@@ -166,9 +168,45 @@ void ftKb_AttackDashAir_800F22D4(Fighter_GObj* gobj)
     ftPartSetRotX(ft, 0, 0.0F);
 }
 
-/// #ftKb_SpecialHi_Enter
+void ftKb_SpecialHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->cmd_vars[3] = 0;
+    fp->cmd_vars[2] = 0;
+    fp->cmd_vars[1] = 0;
+    fp->cmd_vars[0] = 0;
+    fp->mv.kb.specialhi.x0 = 0;
+    fp->mv.kb.specialhi.x4 = 0;
+    fp->mv.kb.specialhi.x8.i = 0;
+    fp->mv.kb.specialhi.xC = 0;
+    Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialHi1, 0, 0.0f, 1.0f, 0.0f,
+                              NULL);
+    ftAnim_8006EBA4(gobj);
+    fp = GET_FIGHTER(gobj);
+    efSync_Spawn(0x494, gobj);
+    fp->x2219_b0 = 1;
+    fp->pre_hitlag_cb = efLib_PauseAll;
+    fp->post_hitlag_cb = efLib_ResumeAll;
+}
 
-/// #ftKb_SpecialAirHi_Enter
+void ftKb_SpecialAirHi_Enter(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->cmd_vars[0] = fp->cmd_vars[1] = fp->cmd_vars[2] =
+        fp->cmd_vars[3] = 0;
+    fp->mv.kb.specialhi.x0 = 0;
+    fp->mv.kb.specialhi.x4 = 0;
+    fp->mv.kb.specialhi.x8.i = 0;
+    fp->mv.kb.specialhi.xC = 0;
+    Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialAirHi1, 0, 0.0f,
+                              1.0f, 0.0f, NULL);
+    ftAnim_8006EBA4(gobj);
+    fp = GET_FIGHTER(gobj);
+    efSync_Spawn(0x494, gobj);
+    fp->x2219_b0 = 1;
+    fp->pre_hitlag_cb = efLib_PauseAll;
+    fp->post_hitlag_cb = efLib_ResumeAll;
+}
 
 void ftKb_SpecialHi1_Anim(Fighter_GObj* gobj)
 {
@@ -1544,9 +1582,57 @@ void ftKb_SpecialN_800F5D04(Fighter_GObj* gobj, bool arg1)
     new_var->fv.kb.hat.kind = 4;
 }
 
-/// #ftKb_SpecialN_800F5DE8
+#pragma dont_inline on
+void ftKb_SpecialN_800F5DE8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    Vec3 pos;
+
+    pos = fp->cur_pos;
+    pos.x += da->specialn_x_offset_inhaled * fp->facing_dir;
+    pos.y += da->specialn_y_offset_inhaled;
+
+    {
+        f64 dist =
+            it_802F23AC((int*) fp->target_item_gobj, (float*) &pos);
+        if (dist <
+            da->specialn_inhale_velocity * da->specialn_inhale_velocity)
+        {
+            it_802F2810(fp->target_item_gobj);
+            if (fp->ground_or_air == GA_Air) {
+                ftKb_SpecialN_800F63EC(gobj);
+            } else {
+                ftKb_SpecialN_800F6388(gobj);
+            }
+        }
+    }
+}
+
+void ftKb_SpecialN_800F5EA8(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    Vec3 pos;
 
-/// #ftKb_SpecialN_800F5EA8
+    pos = fp->cur_pos;
+    pos.x += da->specialn_x_offset_inhaled * fp->facing_dir;
+    pos.y += da->specialn_y_offset_inhaled;
+
+    {
+        f32 dist = ftCo_800BD19C(fp->victim_gobj, &pos);
+        if (dist <
+            da->specialn_inhale_velocity * da->specialn_inhale_velocity)
+        {
+            ftCo_800BD620(fp->victim_gobj);
+            if (fp->ground_or_air == GA_Air) {
+                ftKb_SpecialN_800F63EC(gobj);
+            } else {
+                ftKb_SpecialN_800F6388(gobj);
+            }
+        }
+    }
+}
 
 void ftKb_SpecialN_800F5F68(HSD_GObj* gobj)
 {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
index 8c12684d8d..7f0f93a329 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
@@ -35,6 +35,7 @@
 
 /* 105A34 */ static void fn_80105A34(Fighter_GObj* gobj);
 
+extern ftCollisionBox ftKb_Init_803CB758;
 extern f32 ftKb_Init_804D94F0;
 extern f32 ftKb_Init_804D94F4;
 
@@ -1158,7 +1159,29 @@ void ftKb_PrSpecialNEnd_Coll(Fighter_GObj* gobj)
     }
 }
 
-/// #ftKb_PrSpecialAirNStart_Coll
+void ftKb_PrSpecialAirNStart_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (ft_80081D0C(gobj) != GA_Ground) {
+        FtMotionId msid;
+        ftCommon_8007D7FC(fp);
+        if (1.0F == fp->mv.pr.specialn.x34.x) {
+            msid = ftKb_MS_PrSpecialNStartL;
+        } else {
+            msid = ftKb_MS_PrSpecialNLoop;
+        }
+        Fighter_ChangeMotionState(gobj, msid, 0x0C4C5092U,
+                                  fp->cur_anim_frame, 1.0F, 0.0F,
+                                  NULL);
+        {
+            Fighter* fp2 = GET_FIGHTER(gobj);
+            fp2->death2_cb = ftKb_Init_800EE74C;
+            fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+            fp2->deal_dmg_cb = fn_80100E0C;
+            fp2->x21F8 = fn_80105978;
+        }
+    }
+}
 
 /// #ftKb_PrSpecialAirNLoop_Coll
 
@@ -1166,7 +1189,26 @@ void ftKb_PrSpecialNEnd_Coll(Fighter_GObj* gobj)
 
 /// #ftKb_PrSpecialAirN_Coll
 
-/// #ftKb_PrSpecialN_Coll
+void ftKb_PrSpecialN_Coll(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    if (ft_800824A0(gobj, &ftKb_Init_803CB758)) {
+        ftCommon_8007D7FC(fp);
+        Fighter_ChangeMotionState(gobj, ftKb_MS_PrSpecialNEndR,
+                                  0x0C4C5092U, fp->cur_anim_frame,
+                                  0.0F, 0.0F, NULL);
+        {
+            Fighter* fp2 = GET_FIGHTER(gobj);
+            fp2->death2_cb = ftKb_Init_800EE74C;
+            fp2->take_dmg_cb = ftKb_Init_800EE7B8;
+            fp2->deal_dmg_cb = fn_80100E0C;
+            fp2->x21F8 = fn_80105978;
+        }
+        fp->gr_vel = fp->self_vel.x;
+        fp->self_vel.y = 0;
+        fp->self_vel.z = 0;
+    }
+}
 
 void ftKb_PrSpecialAirNEndR_Coll(Fighter_GObj* gobj)
 {
@@ -1182,7 +1224,21 @@ void ftKb_PrSpecialAirNEndR_Coll(Fighter_GObj* gobj)
 
 /// #ftKb_PrSpecialNHit_Coll
 
-/// #fn_80105978
+void fn_80105978(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftPartSetRotY(fp, 0, M_PI_2 * fp->facing_dir);
+    fp->self_vel.x = -fp->self_vel.x;
+    fp->gr_vel = -fp->gr_vel;
+    fp->xE4_ground_accel_1 = -fp->xE4_ground_accel_1;
+    fp->mv.pr.specialn.x10 = -fp->mv.pr.specialn.x10;
+    fp->mv.pr.specialn.x14 = -fp->mv.pr.specialn.x14;
+    fp->mv.pr.specialn.x18 = -fp->mv.pr.specialn.x18;
+    fp->mv.pr.specialn.x1C = -fp->mv.pr.specialn.x1C;
+    fp->mv.pr.specialn.facing_dir = -fp->mv.pr.specialn.facing_dir;
+    fp->mv.pr.specialn.x34.x = -fp->mv.pr.specialn.x34.x;
+    fp->mv.pr.specialn.x34.y = -fp->mv.pr.specialn.x34.y;
+}
 
 void fn_80105A34(Fighter_GObj* gobj)
 {
diff --git a/src/melee/gm/gm_16F1.c b/src/melee/gm/gm_16F1.c
index 3258fd43e4..8ac10f46cd 100644
--- a/src/melee/gm/gm_16F1.c
+++ b/src/melee/gm/gm_16F1.c
@@ -163,7 +163,30 @@ void fn_80171AD4(void)
 
 /// #fn_8016FFD4
 
-/// #fn_80170110
+int fn_80170110(void* arg0, int idx, int mask, u8 player_id)
+{
+    struct lbl_803D5A4C_t* curr = lbl_803D5A4C;
+    int kind = lbl_803D5A4C[idx].kind;
+    u8 flags;
+
+    while (curr->kind != kind) {
+        if (curr->kind == 0x29A) {
+            flags = 0;
+            goto check;
+        }
+        curr++;
+    }
+    flags = curr->x4;
+
+check:
+    if ((u8) mask & flags) {
+        if (kind < 0xD7) {
+            return pl_80039418(player_id, kind);
+        }
+        return fn_801701C0(arg0, player_id, kind);
+    }
+    return 0;
+}
 
 void gm_801701A0(void)
 {
diff --git a/src/melee/gm/gm_16F1.h b/src/melee/gm/gm_16F1.h
index 9116283858..9368481a5a 100644
--- a/src/melee/gm/gm_16F1.h
+++ b/src/melee/gm/gm_16F1.h
@@ -28,11 +28,11 @@ struct lbl_804D65A8_t;
 /* 16F9A8 */ UNK_RET fn_8016F9A8(UNK_PARAMS);
 /* 16FAD4 */ int fn_8016FAD4(UNK_T, u16, int, u8);
 /* 16FFD4 */ UNK_RET fn_8016FFD4(UNK_PARAMS);
-/* 170110 */ int fn_80170110(UNK_T, int, int, u8);
+/* 170110 */ int fn_80170110(void*, int, int, u8);
 /* 1701A0 */ UNK_RET gm_801701A0(UNK_PARAMS);
 /* 1701AC */ UNK_RET fn_801701AC(UNK_PARAMS);
 /* 1701B8 */ int fn_801701B8(void);
-/* 1701C0 */ UNK_RET fn_801701C0(UNK_PARAMS);
+/* 1701C0 */ int fn_801701C0(void*, int, int);
 /* 171A88 */ int fn_80171A88(void);
 /* 171AD4 */ UNK_RET fn_80171AD4(UNK_PARAMS);
 /* 171B00 */ bool fn_80171B00(int);
diff --git a/src/melee/gm/gm_1BA8.c b/src/melee/gm/gm_1BA8.c
index 325b4d0346..380ffd4658 100644
--- a/src/melee/gm/gm_1BA8.c
+++ b/src/melee/gm/gm_1BA8.c
@@ -1430,7 +1430,23 @@ void gm_801BED3C(MinorScene* arg0)
     }
 }
 
-/// #gm_801BEDA8
+void gm_801BEDA8(MinorScene* arg0)
+{
+    CSSData* css = gm_801A427C(arg0);
+    VsModeData* vs = gm_801A5244();
+
+    if (css->pending_scene_change == 2) {
+        if (css->match_type != 0) {
+            css->match_type--;
+        } else {
+            css->match_type = 0x17;
+        }
+    } else {
+        css->match_type = (css->match_type + 1) % 24;
+    }
+
+    *vs = css->data;
+}
 
 void gm_801BEE58(MinorScene* arg0)
 {
diff --git a/src/melee/gm/gmregclear.c b/src/melee/gm/gmregclear.c
index bde07c53a1..80774cc3a8 100644
--- a/src/melee/gm/gmregclear.c
+++ b/src/melee/gm/gmregclear.c
@@ -1072,7 +1072,32 @@ int fn_8017F008(void)
     }
 }
 
-/// #fn_8017F09C
+s32 fn_8017F09C(void)
+{
+    struct lbl_80472D28_t* data = &lbl_80472D28;
+    lbl_8046B6A0_t* p = gm_8016AE38();
+    if (p->x24C8.x4_6) {
+        lbl_8046B6A0_t* p2 = gm_8016AE38();
+        int flag;
+        if (data->x118 != 0 ||
+            (data->x11A != 0 && data->x11B == 0) ||
+            p2->x24C8.x0_7)
+        {
+            flag = 0;
+        } else {
+            flag = 1;
+        }
+        if (flag != 0) {
+            s32 product = p->timer_seconds * data->x108;
+            s32 ret = product;
+            if (product < 0) {
+                ret = 0;
+            }
+            return ret;
+        }
+    }
+    return 0;
+}
 
 typedef struct fn_8017F14C_arg {
     /* 0x00 */ char pad_0[0x98];
diff --git a/src/melee/gm/gmregclear.h b/src/melee/gm/gmregclear.h
index b2920f2d3e..34e82c3c24 100644
--- a/src/melee/gm/gmregclear.h
+++ b/src/melee/gm/gmregclear.h
@@ -77,7 +77,7 @@
 /* 17EDDC */ bool fn_8017EDDC(void);
 /* 17EE40 */ void fn_8017EE40(int);
 /* 17F008 */ int fn_8017F008(void);
-/* 17F09C */ UNK_RET fn_8017F09C(UNK_PARAMS);
+/* 17F09C */ s32 fn_8017F09C(void);
 /* 17F14C */ s32 fn_8017F14C(void* arg0);
 /* 17F1B8 */ UNK_RET fn_8017F1B8(UNK_PARAMS);
 /* 17F2A4 */ UNK_RET fn_8017F2A4(UNK_PARAMS);
diff --git a/src/melee/gr/gricemt.c b/src/melee/gr/gricemt.c
index f0c2b06697..95e36e5198 100644
--- a/src/melee/gr/gricemt.c
+++ b/src/melee/gr/gricemt.c
@@ -29,10 +29,10 @@
 
 /* 1F8C64 */ static void fn_801F8C64(Item_GObj* gobj, Ground* u1, Vec3* u2,
                                      HSD_GObj* u3, f32 u4);
-/* 1F91EC */ static void FUN_801f91ec(HSD_GObj* param_1, s16* param_2,
-                                      int param_3, int param_4, int param_5,
-                                      int param_6, int param_7,
-                                      mpLib_Callback param_8);
+/* 1F91EC */ void grIceMt_801F91EC(HSD_GObj* param_1, s16* param_2,
+                                   int param_3, int param_4, int param_5,
+                                   int param_6, int param_7,
+                                   mpLib_Callback param_8);
 
 /// StageData grIm_803E8974 = {
 ///     45,
@@ -985,10 +985,9 @@ HSD_GObj* fn_801F91A8(HSD_GObj* gobj)
     return (HSD_GObj*) result;
 }
 
-/// #grIceMt_801F91EC
-void FUN_801f91ec(HSD_GObj* param_1, s16* param_2, int param_3, int param_4,
-                  int param_5, int param_6, int param_7,
-                  mpLib_Callback param_8)
+void grIceMt_801F91EC(HSD_GObj* param_1, s16* param_2, int param_3,
+                      int param_4, int param_5, int param_6, int param_7,
+                      mpLib_Callback param_8)
 {
     Ground* gp = GET_GROUND(param_1);
     param_2[0] = 0;
@@ -998,9 +997,9 @@ void FUN_801f91ec(HSD_GObj* param_1, s16* param_2, int param_3, int param_4,
     param_2[4] = param_5;
     param_2[5] = param_6;
     param_2[6] = param_7;
-    grAnime_801C7A04(param_1, param_3, 7, -20);
+    grAnime_801C7A04(param_1, param_3, 7, 0);
     if (param_4 != -1) {
-        grAnime_801C7A04(param_1, param_4, 7, -20);
+        grAnime_801C7A04(param_1, param_4, 7, 0);
     }
     mpJointSetCb1(param_5, gp, param_8);
 }
diff --git a/src/melee/gr/gricemt.h b/src/melee/gr/gricemt.h
index 2c46c1665f..0910cc2aa3 100644
--- a/src/melee/gr/gricemt.h
+++ b/src/melee/gr/gricemt.h
@@ -125,7 +125,8 @@ extern f32 grIm_804DB570;
 /* 1F9150 */ IceMountainParams* fn_801F9150(HSD_GObj*);
 /* 1F91A4 */ void fn_801F91A4(void);
 /* 1F91A8 */ HSD_GObj* fn_801F91A8(HSD_GObj*);
-/* 1F91EC */ UNK_RET grIceMt_801F91EC(HSD_GObj*);
+/* 1F91EC */ void grIceMt_801F91EC(HSD_GObj*, s16*, int, int, int, int, int,
+                                   mpLib_Callback);
 /* 1F929C */ void grIceMt_801F929C(HSD_GObj* arg0, void* arg1);
 /* 1F9338 */ void fn_801F9338(Ground*, int, CollData*, s32, mpLib_GroundEnum,
                               float);
diff --git a/src/melee/gr/grkongo.c b/src/melee/gr/grkongo.c
index 1ee2bc47c5..185ecbf0f4 100644
--- a/src/melee/gr/grkongo.c
+++ b/src/melee/gr/grkongo.c
@@ -13,6 +13,7 @@
 #include "gr/grzakogenerator.h"
 #include "gr/inlines.h"
 #include "gr/stage.h"
+#include "it/it_26B1.h"
 #include "it/items/itklap.h"
 #include "lb/lb_00B0.h"
 #include "lb/lb_00F9.h"
@@ -1433,7 +1434,37 @@ void grKongo_801D8058(Ground_GObj* arg)
     Ground_801C4A08(arg);
 }
 
-/// #grKongo_801D8078
+HSD_GObj* grKongo_801D8078(HSD_GObj* gobj)
+{
+    Vec3 pos;
+    f32 unk;
+    Vec3 item_pos;
+    HSD_GObj* cur;
+
+    Ground_801C4DA0(&pos, &unk);
+
+    for (cur = HSD_GObj_Entities->items; cur != NULL; cur = cur->next) {
+        if (itGetKind(cur) == 0xDA) {
+            f32 dx, dy, dz, dx2, dy2, dz2, r;
+
+            it_8026B294(cur, &item_pos);
+
+            dx = pos.x - item_pos.x;
+            dy = pos.y - item_pos.y;
+            dz = pos.z - item_pos.z;
+            dx2 = dx * dx;
+            dy2 = dy * dy;
+            dz2 = dz * dz;
+            r = grKg_804D6980->unk28;
+
+            if (dx2 + dy2 + dz2 < r * r) {
+                return cur;
+            }
+        }
+    }
+
+    return NULL;
+}
 
 /// #fn_801D8134
 
diff --git a/src/melee/mn/mnname.c b/src/melee/mn/mnname.c
index 8601c24cb3..4d838c9db2 100644
--- a/src/melee/mn/mnname.c
+++ b/src/melee/mn/mnname.c
@@ -53,7 +53,23 @@ bool IsNameListFull(void)
 
 void fn_802377A4(void) {}
 
-/// #IsNameUnique
+bool IsNameUnique(s32 slot)
+{
+    int i;
+    char* namedata;
+    for (i = 0; i < 0x78; i++) {
+        namedata = !IsNameValid(i & 0xFF)
+                       ? (char*) NULL
+                       : GetPersistentNameData((u8) i)->namedata;
+        if (namedata == NULL) {
+            break;
+        }
+        if (CompareNameStrings(namedata, slot) == 0) {
+            return true;
+        }
+    }
+    return false;
+}
 
 /// #DeleteName
 
diff --git a/src/melee/mn/mnname.h b/src/melee/mn/mnname.h
index 053c3c90c3..c7d23b0051 100644
--- a/src/melee/mn/mnname.h
+++ b/src/melee/mn/mnname.h
@@ -22,7 +22,7 @@ typedef struct MnName_GObj {
 /* 2375EC */ bool IsNameListFull(void);
 /* 237654 */ s32 CompareNameStrings(char* str, s32 slot);
 /* 2377A4 */ void fn_802377A4(void);
-/* 2377A8 */ UNK_RET IsNameUnique(UNK_PARAMS);
+/* 2377A8 */ bool IsNameUnique(s32 slot);
 /* 237834 */ UNK_RET DeleteName(UNK_PARAMS);
 /* 2379BC */ bool IsNameValid(int slot);
 /* 237A04 */ void CreateNameAtIndex(s32 slot);
diff --git a/src/melee/ty/toy.c b/src/melee/ty/toy.c
index 9e73f3bab0..c8246914f7 100644
--- a/src/melee/ty/toy.c
+++ b/src/melee/ty/toy.c
@@ -664,7 +664,25 @@ s32 un_80305C44(void)
     return button;
 }
 
-/// #un_80305D00
+float un_80305D00(void)
+{
+    float ret;
+    float val;
+    int i;
+
+    for (i = 0; i < 4; i++) {
+        val = HSD_PadCopyStatus[(u8) i].nml_stickX;
+        ret = val;
+        if (val < 0.0F) {
+            val = -val;
+        }
+        if (val > 0.1F) {
+            break;
+        }
+    }
+
+    return ret;
+}
 
 float un_80305DB0(void)
 {
diff --git a/src/melee/ty/toy.h b/src/melee/ty/toy.h
index 1757a7426d..a30d09e81f 100644
--- a/src/melee/ty/toy.h
+++ b/src/melee/ty/toy.h
@@ -15,7 +15,7 @@
 /* 305918 */ void un_80305918(signed char, long, long);
 /* 305B88 */ s32 un_80305B88(void);
 /* 305C44 */ s32 un_80305C44(void);
-/* 305D00 */ UNK_RET un_80305D00(UNK_PARAMS);
+/* 305D00 */ float un_80305D00(void);
 /* 305DB0 */ float un_80305DB0(void);
 /* 305EB4 */ float un_80305EB4(void);
 /* 305FB8 */ float un_80305FB8(void);
