diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 5e76049905..fe21e8d168 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -23739,7 +23739,7 @@ it_803F7C10 = .data:0x803F7C10; // type:object size:0x30 scope:global
 it_803F7C40 = .data:0x803F7C40; // type:object size:0x30 scope:global
 it_803F7C70 = .data:0x803F7C70; // type:object size:0x30 scope:global
 it_803F7CA0 = .data:0x803F7CA0; // type:object size:0x30 scope:global
-it_803F7CD0 = .data:0x803F7CD0; // type:object size:0x25 scope:global data:string
+@230 = .data:0x803F7CD0; // type:object size:0x25 scope:local data:string
 it_803F7CF8 = .data:0x803F7CF8; // type:object size:0x40 scope:global
 @299 = .data:0x803F7D38; // type:object size:0x25 scope:local data:string
 it_803F7D60 = .data:0x803F7D60; // type:object size:0x30 scope:global
@@ -27142,8 +27142,8 @@ it_804D55D0 = .sdata:0x804D55D0; // type:object size:0x7 scope:global data:strin
 it_804D55D8 = .sdata:0x804D55D8; // type:object size:0x5 scope:global data:string
 it_804D55E0 = .sdata:0x804D55E0; // type:object size:0x7 scope:global data:string
 it_804D55E8 = .sdata:0x804D55E8; // type:object size:0x5 scope:global data:string
-it_804D55F0 = .sdata:0x804D55F0; // type:object size:0x7 scope:global data:string
-it_804D55F8 = .sdata:0x804D55F8; // type:object size:0x5 scope:global data:string
+@228 = .sdata:0x804D55F0; // type:object size:0x7 scope:local data:string
+@229 = .sdata:0x804D55F8; // type:object size:0x5 scope:local data:string
 @296 = .sdata:0x804D5600; // type:object size:0x7 scope:local data:string
 @297 = .sdata:0x804D5608; // type:object size:0x5 scope:local data:string
 it_804D5610 = .sdata:0x804D5610; // type:object size:0x7 scope:global data:string
@@ -33325,12 +33325,12 @@ it_804DD378 = .sdata2:0x804DD378; // type:object size:0x4 scope:global align:4 d
 @175 = .sdata2:0x804DD3AC; // type:object size:0x4 scope:local align:4 data:float
 @145 = .sdata2:0x804DD3B0; // type:object size:0x4 scope:local align:4 data:float
 @161 = .sdata2:0x804DD3B4; // type:object size:0x4 scope:local align:4 data:float
-it_804DD3B8 = .sdata2:0x804DD3B8; // type:object size:0x4 scope:global align:4 data:float
-it_804DD3C0 = .sdata2:0x804DD3C0; // type:object size:0x8 scope:global align:8 data:double
-it_804DD3C8 = .sdata2:0x804DD3C8; // type:object size:0x8 scope:global align:8 data:double
-it_804DD3D0 = .sdata2:0x804DD3D0; // type:object size:0x4 scope:global align:4 data:float
-it_804DD3D4 = .sdata2:0x804DD3D4; // type:object size:0x4 scope:global align:4 data:float
-it_804DD3D8 = .sdata2:0x804DD3D8; // type:object size:0x4 scope:global align:4 data:float
+@159 = .sdata2:0x804DD3B8; // type:object size:0x4 scope:local align:4 data:float
+@161 = .sdata2:0x804DD3C0; // type:object size:0x8 scope:local align:8 data:double
+@226 = .sdata2:0x804DD3C8; // type:object size:0x8 scope:local align:8 data:double
+@227 = .sdata2:0x804DD3D0; // type:object size:0x4 scope:local align:4 data:float
+@233 = .sdata2:0x804DD3D4; // type:object size:0x4 scope:local align:4 data:float
+@259 = .sdata2:0x804DD3D8; // type:object size:0x4 scope:local align:4 data:float
 @252 = .sdata2:0x804DD3E0; // type:object size:0x4 scope:local align:4 data:float
 @257 = .sdata2:0x804DD3E4; // type:object size:0x4 scope:local align:4 data:float
 @261 = .sdata2:0x804DD3E8; // type:object size:0x8 scope:local align:8 data:double
diff --git a/configure.py b/configure.py
index 5330b62862..ba89980072 100755
--- a/configure.py
+++ b/configure.py
@@ -1221,7 +1221,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/it/items/itfire.c"),
             Object(Matching, "melee/it/items/itthunder.c"),
             Object(Matching, "melee/it/items/itfreezer.c"),
-            Object(NonMatching, "melee/it/items/itsonans.c"),
+            Object(Matching, "melee/it/items/itsonans.c"),
             Object(Matching, "melee/it/items/ithassam.c"),
             Object(NonMatching, "melee/it/items/itunknown.c"),
             Object(Matching, "melee/it/items/itentei.c"),
diff --git a/src/melee/it/items/itsonans.c b/src/melee/it/items/itsonans.c
index a4690d93a5..b65c3bc8a3 100644
--- a/src/melee/it/items/itsonans.c
+++ b/src/melee/it/items/itsonans.c
@@ -1,19 +1,22 @@
 #include "itsonans.h"
 
+#include "math.h"
+
 #include "ef/eflib.h"
 #include "it/inlines.h"
 #include "it/it_266F.h"
 #include "it/it_26B1.h"
 #include "it/it_2725.h"
 #include "it/itcoll.h"
+#include "it/itCommonItems.h"
 #include "it/item.h"
 
-ItemStateTable it_803F7C70[] = {
+ItemStateTable it_803F7CA0[] = {
     { 0, itSonans_UnkMotion0_Anim, itSonans_UnkMotion0_Phys,
       itSonans_UnkMotion0_Coll },
     { 1, itSonans_UnkMotion1_Anim, itSonans_UnkMotion1_Phys,
       itSonans_UnkMotion1_Coll },
-    { 2, itSonans_UnkMotion2_Anim, itSonans_UnkMotion2_Phys,
+    { -1, itSonans_UnkMotion2_Anim, itSonans_UnkMotion2_Phys,
       itSonans_UnkMotion2_Coll }
 };
 
@@ -39,7 +42,6 @@ void it_802CD4DC(Item_GObj* gobj, Item_GObj* ref_gobj)
     it_8026B894(gobj, ref_gobj);
 }
 
-/// 99.63%
 void it_802CD4FC(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -61,15 +63,9 @@ void it_802CD4FC(Item_GObj* gobj)
         } else if (angle <= 0.0f && ip->xDD4_itemVar.sonans.x60 > 0.0f) {
             ip->xDD4_itemVar.sonans.x60 += attrs->x10;
         }
-        angle = ip->xDD4_itemVar.sonans.x64;
-        if (angle < 0.0f) {
-            angle = -angle;
-        }
+        angle = ABS(ip->xDD4_itemVar.sonans.x64);
         if (angle < attrs->x1C) {
-            angle = ip->xDD4_itemVar.sonans.x60;
-            if (angle < 0.0f) {
-                angle = -angle;
-            }
+            angle = ABS(ip->xDD4_itemVar.sonans.x60);
             if (angle < attrs->x1C) {
                 ip->xDD4_itemVar.sonans.x64 = 0.0f;
                 ip->xDD4_itemVar.sonans.x60 = 0.0f;
@@ -84,16 +80,14 @@ void it_802CD4FC(Item_GObj* gobj)
         angle = ip->xDD4_itemVar.sonans.x64;
         max_angle = attrs->x18;
         if (angle > max_angle) {
-            ip->xDD4_itemVar.sonans.x60 =
-                (f32) (ip->xDD4_itemVar.sonans.x60 * it_804DD3C8);
+            ip->xDD4_itemVar.sonans.x60 *= -0.9;
             ip->xDD4_itemVar.sonans.x64 = attrs->x18;
         } else if (angle < -max_angle) {
-            ip->xDD4_itemVar.sonans.x60 =
-                (f32) (ip->xDD4_itemVar.sonans.x60 * it_804DD3C8);
+            ip->xDD4_itemVar.sonans.x60 *= -0.9;
             ip->xDD4_itemVar.sonans.x64 = -attrs->x18;
         }
         HSD_JObjSetRotationZ(ip->xBBC_dynamicBoneTable->bones[4],
-                             it_804DD3D0 * ip->xDD4_itemVar.sonans.x64);
+                             0.017453292f * ip->xDD4_itemVar.sonans.x64);
     }
     it_80272460(&ip->x5D4_hitboxes[0].hit, (u32) ip->xDD4_itemVar.sonans.x68,
                 gobj);
@@ -107,41 +101,26 @@ bool itSonans_Logic9_DmgDealt(Item_GObj* gobj)
     return false;
 }
 
-/// 99.04%
 bool it_802CD7D4(Item_GObj* gobj)
 {
     Item* item = GET_ITEM(gobj);
     itsonansAttributes* attrs = item->xC4_article_data->x4_specialAttributes;
-    f32 var_f1;
-    f32 var_f3;
-    f32 temp;
     if (item->xDAC_itcmd_var0 == 0) {
         item->xDD4_itemVar.sonans.x60 =
             attrs->x8 * (item->xCA0 * item->xCCC_incDamageDirection);
         item->xDAC_itcmd_var0 = 1;
     } else {
-        if ((attrs->x8 * (item->xCA0 * item->xCCC_incDamageDirection)) < 0.0f)
+        if (ABS(item->xDD4_itemVar.sonans.x60) <
+            ABS(attrs->x8 * (item->xCA0 * item->xCCC_incDamageDirection)))
         {
-            var_f1 =
-                -(attrs->x8 * (item->xCA0 * item->xCCC_incDamageDirection));
-        } else {
-            var_f1 =
-                (attrs->x8 * (item->xCA0 * item->xCCC_incDamageDirection));
-        }
-        var_f3 = item->xDD4_itemVar.sonans.x60;
-        if (var_f3 < 0.0f) {
-            var_f3 = -var_f3;
-        }
-        if (var_f3 < var_f1) {
             item->xDD4_itemVar.sonans.x60 =
                 (attrs->x8 * (item->xCA0 * item->xCCC_incDamageDirection));
         }
     }
-    temp = item->xCA0 * attrs->x4;
-    if (temp > attrs->xC) {
+    if (item->xCA0 * attrs->x4 > attrs->xC) {
         item->xDD4_itemVar.sonans.x68 = attrs->xC;
     } else {
-        item->xDD4_itemVar.sonans.x68 = temp;
+        item->xDD4_itemVar.sonans.x68 = item->xCA0 * attrs->x4;
     }
     return false;
 }
@@ -180,15 +159,15 @@ void it_802CD9C0(Item_GObj* gobj)
     ip->exited_hitlag = efLib_ResumeAll;
 }
 
-f32 it_804DD3B8 = 0.0f;
-
 bool itSonans_UnkMotion1_Anim(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
-    if (--ip->xD44_lifeTimer == it_804DD3B8) {
-        return true;
-    }
-    return false;
+    int zero = 0;
+
+    f32 lifetime = ip->xD44_lifeTimer - 1.0f;
+    ip->xD44_lifeTimer = lifetime;
+
+    return (lifetime == zero) ? true : false;
 }
 
 void itSonans_UnkMotion1_Phys(Item_GObj* gobj)
diff --git a/src/melee/it/items/itsonans.h b/src/melee/it/items/itsonans.h
index 4a791c5eaa..e6d34a3ab8 100644
--- a/src/melee/it/items/itsonans.h
+++ b/src/melee/it/items/itsonans.h
@@ -26,7 +26,4 @@
 /* 2CDBB8 */ bool itSonans_UnkMotion2_Coll(Item_GObj* gobj);
 /* 3F7CA0 */ extern ItemStateTable it_803F7CA0[];
 
-f32 it_804DD3D0 = 0.017453292f;
-f64 it_804DD3C8 = -0.9;
-
 #endif
