diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index d746c81d0c..2bf2ced617 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -8204,22 +8204,22 @@ GetPersistentNameData = .text:0x8015CC9C; // type:function size:0x48 scope:globa
 gmMainLib_8015CCE4 = .text:0x8015CCE4; // type:function size:0xC scope:global
 gmMainLib_8015CCF0 = .text:0x8015CCF0; // type:function size:0xC scope:global
 gmMainLib_8015CCFC = .text:0x8015CCFC; // type:function size:0xC scope:global
-gmMainLib_8015CD08 = .text:0x8015CD08; // type:function size:0xC scope:global
-gmMainLib_8015CD14 = .text:0x8015CD14; // type:function size:0xC scope:global
-gmMainLib_8015CD20 = .text:0x8015CD20; // type:function size:0xC scope:global
-gmMainLib_8015CD2C = .text:0x8015CD2C; // type:function size:0xC scope:global
-gmMainLib_8015CD38 = .text:0x8015CD38; // type:function size:0xC scope:global
-gmMainLib_8015CD44 = .text:0x8015CD44; // type:function size:0xC scope:global
-gmMainLib_8015CD50 = .text:0x8015CD50; // type:function size:0xC scope:global
+gmMainLib_GetVsPlayContestants = .text:0x8015CD08; // type:function size:0xC scope:global
+gmMainLib_GetVsPlayTime = .text:0x8015CD14; // type:function size:0xC scope:global
+gmMainLib_GetCombinedVSPlayTime = .text:0x8015CD20; // type:function size:0xC scope:global
+gmMainLib_GetTimeMatchTotal = .text:0x8015CD2C; // type:function size:0xC scope:global
+gmMainLib_GetStockMatchTotal = .text:0x8015CD38; // type:function size:0xC scope:global
+gmMainLib_GetCoinMatchTotal = .text:0x8015CD44; // type:function size:0xC scope:global
+gmMainLib_GetBonusMatchTotal = .text:0x8015CD50; // type:function size:0xC scope:global
 gmMainLib_8015CD5C = .text:0x8015CD5C; // type:function size:0xC scope:global
-gmMainLib_8015CD68 = .text:0x8015CD68; // type:function size:0xC scope:global
-gmMainLib_8015CD74 = .text:0x8015CD74; // type:function size:0xC scope:global
+gmMainLib_GetMatchResetCounter = .text:0x8015CD68; // type:function size:0xC scope:global
+gmMainLib_GetSingleplayerTime = .text:0x8015CD74; // type:function size:0xC scope:global
 gmMainLib_8015CD80 = .text:0x8015CD80; // type:function size:0xC scope:global
-gmMainLib_8015CD8C = .text:0x8015CD8C; // type:function size:0xC scope:global
-gmMainLib_8015CD98 = .text:0x8015CD98; // type:function size:0xC scope:global
-gmMainLib_8015CDA4 = .text:0x8015CDA4; // type:function size:0xC scope:global
-gmMainLib_8015CDB0 = .text:0x8015CDB0; // type:function size:0xC scope:global
-gmMainLib_8015CDBC = .text:0x8015CDBC; // type:function size:0xC scope:global
+gmMainLib_GetPowerCount = .text:0x8015CD8C; // type:function size:0xC scope:global
+gm_GetPowerTime = .text:0x8015CD98; // type:function size:0xC scope:global
+gmMainLib_GetTotalDamage = .text:0x8015CDA4; // type:function size:0xC scope:global
+gmMainLib_GetKOTotal = .text:0x8015CDB0; // type:function size:0xC scope:global
+gmMainLib_GetSelfDestructTotal = .text:0x8015CDBC; // type:function size:0xC scope:global
 gmMainLib_8015CDC8 = .text:0x8015CDC8; // type:function size:0xC scope:global
 gmMainLib_8015CDD4 = .text:0x8015CDD4; // type:function size:0xC scope:global
 gmMainLib_8015CDE0 = .text:0x8015CDE0; // type:function size:0xC scope:global
@@ -8389,16 +8389,16 @@ gm_801623FC = .text:0x801623FC; // type:function size:0x80 scope:global
 gm_8016247C = .text:0x8016247C; // type:function size:0xF8 scope:global
 gm_80162574 = .text:0x80162574; // type:function size:0x98 scope:global
 gm_8016260C = .text:0x8016260C; // type:function size:0x190 scope:global
-gm_8016279C = .text:0x8016279C; // type:function size:0x64 scope:global
+gm_GetVsPlayMatchTotal = .text:0x8016279C; // type:function size:0x64 scope:global
 gm_80162800 = .text:0x80162800; // type:function size:0xC4 scope:global
 gm_801628C4 = .text:0x801628C4; // type:function size:0xA4 scope:global
 gm_80162968 = .text:0x80162968; // type:function size:0x4C scope:global
 gm_801629B4 = .text:0x801629B4; // type:function size:0x4C scope:global
-gm_80162A00 = .text:0x80162A00; // type:function size:0x4C scope:global
+gm_GetPlayTime = .text:0x80162A00; // type:function size:0x4C scope:global
 gm_80162A4C = .text:0x80162A4C; // type:function size:0x4C scope:global
 gm_80162A98 = .text:0x80162A98; // type:function size:0xB4 scope:global
 gm_80162B4C = .text:0x80162B4C; // type:function size:0x4C scope:global
-gm_80162B98 = .text:0x80162B98; // type:function size:0x40 scope:global
+gm_IncrementPowerCount = .text:0x80162B98; // type:function size:0x40 scope:global
 gm_80162BD8 = .text:0x80162BD8; // type:function size:0x24 scope:global
 fn_80162BFC = .text:0x80162BFC; // type:function size:0x4C
 gm_80162C48 = .text:0x80162C48; // type:function size:0x84 scope:global
@@ -12858,27 +12858,27 @@ fn_8024FC48 = .text:0x8024FC48; // type:function size:0xF8
 fn_8024FD40 = .text:0x8024FD40; // type:function size:0x10C
 mnDataDel_8024FE4C = .text:0x8024FE4C; // type:function size:0x324 scope:global
 mnDataDel_80250170 = .text:0x80250170; // type:function size:0x15C scope:global
-mnCount_802502CC = .text:0x802502CC; // type:function size:0x24 scope:global
-fn_802502F0 = .text:0x802502F0; // type:function size:0x24
-fn_80250314 = .text:0x80250314; // type:function size:0x24
-fn_80250338 = .text:0x80250338; // type:function size:0x24
+mnCount_GetMatchTime = .text:0x802502CC; // type:function size:0x24 scope:global
+mnCount_GetKOKingpin = .text:0x802502F0; // type:function size:0x24
+mnCount_GetNoDefenseNelly = .text:0x80250314; // type:function size:0x24
+mnCount_GetDisasterMaster = .text:0x80250338; // type:function size:0x24
 mnCount_8025035C = .text:0x8025035C; // type:function size:0x340 scope:global
-fn_8025069C = .text:0x8025069C; // type:function size:0x24
-fn_802506C0 = .text:0x802506C0; // type:function size:0x24
-fn_802506E4 = .text:0x802506E4; // type:function size:0x24
-fn_80250708 = .text:0x80250708; // type:function size:0x24
+mnCount_GetSmashChamp = .text:0x8025069C; // type:function size:0x24
+mnCount_GetSmashSap = .text:0x802506C0; // type:function size:0x24
+mnCount_GetSlugMeister = .text:0x802506E4; // type:function size:0x24
+mnCount_GetPunchingBag = .text:0x80250708; // type:function size:0x24
 mnCount_8025072C = .text:0x8025072C; // type:function size:0x200 scope:global
 mnCount_8025092C = .text:0x8025092C; // type:function size:0x350 scope:global
-mnCount_80250C7C = .text:0x80250C7C; // type:function size:0x168 scope:global
-mnCount_80250DE4 = .text:0x80250DE4; // type:function size:0x214 scope:global
-mnCount_80250FF8 = .text:0x80250FF8; // type:function size:0x280 scope:global
-fn_80251278 = .text:0x80251278; // type:function size:0x17C
-mnCount_802513F4 = .text:0x802513F4; // type:function size:0xC4 scope:global
+mnCount_GetRowValue_Character = .text:0x80250C7C; // type:function size:0x168 scope:global
+mnCount_GetRowValue_Number = .text:0x80250DE4; // type:function size:0x214 scope:global
+mnCount_CreateRow = .text:0x80250FF8; // type:function size:0x280 scope:global
+mnCount_HandleUserInput = .text:0x80251278; // type:function size:0x17C
+mnCount_UpdateArrowIndicators = .text:0x802513F4; // type:function size:0xC4 scope:global
 fn_802514B8 = .text:0x802514B8; // type:function size:0x20
 fn_802514D8 = .text:0x802514D8; // type:function size:0x168
 fn_80251640 = .text:0x80251640; // type:function size:0x1A0
-mnCount_802517E0 = .text:0x802517E0; // type:function size:0x8C scope:global
-mnCount_8025186C = .text:0x8025186C; // type:function size:0x19C scope:global
+mnCount_InitUserData = .text:0x802517E0; // type:function size:0x8C scope:global
+mnCount_Create = .text:0x8025186C; // type:function size:0x19C scope:global
 mnInfo_80251A08 = .text:0x80251A08; // type:function size:0x9C scope:global
 mnInfo_80251AA4 = .text:0x80251AA4; // type:function size:0x58 scope:global
 mnInfo_80251AFC = .text:0x80251AFC; // type:function size:0x25C scope:global
@@ -17570,7 +17570,7 @@ un_80304510 = .text:0x80304510; // type:function size:0x90 scope:global
 un_803045A0 = .text:0x803045A0; // type:function size:0xF0 scope:global
 un_80304690 = .text:0x80304690; // type:function size:0xF0 scope:global
 un_80304780 = .text:0x80304780; // type:function size:0xF0 scope:global
-un_80304870 = .text:0x80304870; // type:function size:0x50 scope:global
+un_GetTrophyTotal = .text:0x80304870; // type:function size:0x50 scope:global
 un_803048C0 = .text:0x803048C0; // type:function size:0x64 scope:global
 un_80304924 = .text:0x80304924; // type:function size:0x64 scope:global
 un_80304988 = .text:0x80304988; // type:function size:0x6C scope:global
diff --git a/src/melee/gm/gm_1601.c b/src/melee/gm/gm_1601.c
index 06d62da0b2..5986f7495a 100644
--- a/src/melee/gm/gm_1601.c
+++ b/src/melee/gm/gm_1601.c
@@ -946,10 +946,12 @@ void gm_80162574(u8 arg0, u8 arg1)
 
 /// #gm_8016260C
 
-u32 gm_8016279C(void)
+u32 gm_GetVsPlayMatchTotal(void)
 {
-    u32 tmp = *(u32*) gmMainLib_8015CD2C() + *(u32*) gmMainLib_8015CD38() +
-              *(u32*) gmMainLib_8015CD44() + *(u32*) gmMainLib_8015CD50() +
+    u32 tmp = *(u32*) gmMainLib_GetTimeMatchTotal() +
+              *(u32*) gmMainLib_GetStockMatchTotal() +
+              *(u32*) gmMainLib_GetCoinMatchTotal() +
+              *(u32*) gmMainLib_GetBonusMatchTotal() +
               *(u32*) gmMainLib_8015CD5C();
     return tmp;
 }
@@ -961,7 +963,7 @@ u32 gm_80162800(MatchEnd* arg0)
     u32 sum;
     u32 i;
 
-    temp_r3 = gmMainLib_8015CD08();
+    temp_r3 = gmMainLib_GetVsPlayContestants();
     count = 0;
     for (i = 0; i < 6; i++) {
         if (arg0->player_standings[i].slot_type == 0) {
@@ -979,11 +981,11 @@ void gm_801628C4(u32 arg0, u32 arg1)
     u32 var_r4;
     struct gmm_retval_ED98* temp_r3_3;
 
-    temp_r3 = gmMainLib_8015CD14();
+    temp_r3 = gmMainLib_GetVsPlayTime();
     var_r4 = *temp_r3 + arg0;
     *temp_r3 = (var_r4 > (u32) -1) ? (u32) -1 : var_r4;
 
-    temp_r3 = gmMainLib_8015CD20();
+    temp_r3 = gmMainLib_GetCombinedVSPlayTime();
     var_r4 = (0, *temp_r3 + (arg0 * arg1));
     *temp_r3 = (var_r4 > (u32) -1) ? (u32) -1 : var_r4;
 
@@ -994,7 +996,7 @@ void gm_801628C4(u32 arg0, u32 arg1)
 
 long gm_80162968(u32 seconds)
 {
-    u32* ptr = gmMainLib_8015CD74();
+    u32* ptr = gmMainLib_GetSingleplayerTime();
 
     *ptr = ((*ptr + seconds) > -1) ? -1 : (*ptr + seconds);
 }
@@ -1006,13 +1008,13 @@ long gm_801629B4(s32 amount)
     *ptr = ((*ptr + amount) > -1) ? -1 : (*ptr + amount);
 }
 
-long gm_80162A00(void)
+long gm_GetPlayTime(void)
 {
     u32 a;
     u32 b;
 
-    a = *(u32*) gmMainLib_8015CD14();
-    b = *(u32*) gmMainLib_8015CD74();
+    a = *(u32*) gmMainLib_GetVsPlayTime();
+    b = *(u32*) gmMainLib_GetSingleplayerTime();
     return a + b > -1 ? -1 : a + b;
 }
 
@@ -1020,7 +1022,7 @@ long gm_80162A4C(s32 amount)
 {
     u32* ptr;
 
-    ptr = gmMainLib_8015CDA4();
+    ptr = gmMainLib_GetTotalDamage();
     *ptr = ((*ptr + amount) > -1) ? -1 : (*ptr + amount);
 }
 
@@ -1028,7 +1030,7 @@ struct gmm_x1868_1A8_t* gm_80162A98(s32 arg0)
 {
     struct gmm_x1868_1A8_t* temp_r3;
     u32 var_r31;
-    s32* temp_r3_2 = gmMainLib_8015CDB0();
+    s32* temp_r3_2 = gmMainLib_GetKOTotal();
 
     *temp_r3_2 = MAX(*temp_r3_2 + arg0, -1U);
 
@@ -1048,18 +1050,18 @@ void gm_80162B4C(s32 amount)
     u32* ptr;
     u32 sum;
 
-    ptr = gmMainLib_8015CDBC();
+    ptr = gmMainLib_GetSelfDestructTotal();
     sum = *ptr + amount;
     *ptr = (sum > (u32) -1) ? (u32) -1 : sum;
 }
 
-void gm_80162B98(void)
+void gm_IncrementPowerCount(void)
 {
     u32 result;
     u32 val;
     u32* ptr;
 
-    ptr = gmMainLib_8015CD8C();
+    ptr = gmMainLib_GetPowerCount();
     val = *ptr;
     result = (val + 1 > (u32) -1) ? (u32) -1 : val + 1;
     *ptr = result;
diff --git a/src/melee/gm/gm_1601.h b/src/melee/gm/gm_1601.h
index aec7ae0c0e..6b073b4498 100644
--- a/src/melee/gm/gm_1601.h
+++ b/src/melee/gm/gm_1601.h
@@ -64,16 +64,16 @@ struct lbl_8046B488_t;
 /* 16247C */ long gm_8016247C(long);
 /* 162574 */ void gm_80162574(u8, u8);
 /* 16260C */ void gm_8016260C(u8, u8);
-/* 16279C */ u32 gm_8016279C(void);
+/* 16279C */ u32 gm_GetVsPlayMatchTotal(void);
 /* 162800 */ u32 gm_80162800(struct MatchEnd*);
 /* 1628C4 */ void gm_801628C4(u32, u32);
 /* 162968 */ long gm_80162968(u32 seconds);
 /* 1629B4 */ long gm_801629B4(s32 amount);
-/* 162A00 */ long gm_80162A00(void);
+/* 162A00 */ long gm_GetPlayTime(void);
 /* 162A4C */ long gm_80162A4C(s32 amount);
 /* 162A98 */ struct gmm_x1868_1A8_t* gm_80162A98(s32);
 /* 162B4C */ void gm_80162B4C(s32 amount);
-/* 162B98 */ void gm_80162B98(void);
+/* 162B98 */ void gm_IncrementPowerCount(void);
 /* 162BD8 */ int gm_80162BD8(u8);
 /* 162BFC */ bool fn_80162BFC(s8 ckind, int);
 /* 162C48 */ long gm_80162C48(void);
diff --git a/src/melee/gm/gm_16F1.c b/src/melee/gm/gm_16F1.c
index 8ac10f46cd..fc78b13b9a 100644
--- a/src/melee/gm/gm_16F1.c
+++ b/src/melee/gm/gm_16F1.c
@@ -413,7 +413,7 @@ bool fn_801724C8(void)
 
 bool fn_801724D0(void)
 {
-    if (gm_8016279C() >= 0x64) {
+    if (gm_GetVsPlayMatchTotal() >= 0x64) {
         return true;
     }
     return false;
@@ -421,7 +421,7 @@ bool fn_801724D0(void)
 
 bool fn_80172504(void)
 {
-    if (gm_8016279C() >= 0x3E8) {
+    if (gm_GetVsPlayMatchTotal() >= 0x3E8) {
         return true;
     }
     return false;
@@ -429,7 +429,7 @@ bool fn_80172504(void)
 
 bool fn_80172538(void)
 {
-    if (gm_8016279C() >= 0x2710) {
+    if (gm_GetVsPlayMatchTotal() >= 0x2710) {
         return true;
     }
     return false;
@@ -437,7 +437,7 @@ bool fn_80172538(void)
 
 bool fn_8017256C(void)
 {
-    if (gm_8016279C() >= 0x186A0) {
+    if (gm_GetVsPlayMatchTotal() >= 0x186A0) {
         return true;
     }
     return false;
@@ -445,7 +445,7 @@ bool fn_8017256C(void)
 
 bool fn_801725A8(void)
 {
-    if (gm_8016279C() >= 0xF4240) {
+    if (gm_GetVsPlayMatchTotal() >= 0xF4240) {
         return true;
     }
     return false;
@@ -458,7 +458,9 @@ enum {
 
 bool fn_801725E4(void)
 {
-    if (lbLang_GetLanguageSetting() == LANG_JP && gm_8016279C() >= 0x1A90) {
+    if (lbLang_GetLanguageSetting() == LANG_JP &&
+        gm_GetVsPlayMatchTotal() >= 0x1A90)
+    {
         return true;
     }
     return false;
@@ -466,7 +468,9 @@ bool fn_801725E4(void)
 
 bool fn_80172624(void)
 {
-    if (lbLang_GetLanguageSetting() == LANG_US && gm_8016279C() >= 0x1383) {
+    if (lbLang_GetLanguageSetting() == LANG_US &&
+        gm_GetVsPlayMatchTotal() >= 0x1383)
+    {
         return true;
     }
     return false;
@@ -474,7 +478,7 @@ bool fn_80172624(void)
 
 bool fn_80172664(void)
 {
-    if (gm_8016279C() >= 0xC350) {
+    if (gm_GetVsPlayMatchTotal() >= 0xC350) {
         return true;
     }
     return false;
@@ -482,7 +486,7 @@ bool fn_80172664(void)
 
 bool fn_80172698(void)
 {
-    if (un_80304870() >= 0x32) {
+    if (un_GetTrophyTotal() >= 0x32) {
         return true;
     }
     return false;
@@ -490,7 +494,7 @@ bool fn_80172698(void)
 
 bool fn_801726CC(void)
 {
-    if (un_80304870() >= 0x64) {
+    if (un_GetTrophyTotal() >= 0x64) {
         return true;
     }
     return false;
@@ -498,7 +502,7 @@ bool fn_801726CC(void)
 
 bool fn_80172700(void)
 {
-    if (un_80304870() >= 0x96) {
+    if (un_GetTrophyTotal() >= 0x96) {
         return true;
     }
     return false;
@@ -506,7 +510,7 @@ bool fn_80172700(void)
 
 bool fn_80172734(void)
 {
-    if (un_80304870() >= 0xC8) {
+    if (un_GetTrophyTotal() >= 0xC8) {
         return true;
     }
     return false;
@@ -514,7 +518,7 @@ bool fn_80172734(void)
 
 bool fn_80172768(void)
 {
-    if (un_80304870() >= 0xFA) {
+    if (un_GetTrophyTotal() >= 0xFA) {
         return true;
     }
     return false;
@@ -1396,7 +1400,7 @@ void gm_80174180(void)
     if (fn_80162CCC() != 0 && gm_80162EC8() != 0) {
         un_80305918(4, 0, 0);
     }
-    if (gm_8016279C() >= 0xC8) {
+    if (gm_GetVsPlayMatchTotal() >= 0xC8) {
         un_80305918(5, 0, 0);
     }
     if (gm_80164ABC() != 0) {
diff --git a/src/melee/gm/gmmain_lib.c b/src/melee/gm/gmmain_lib.c
index 9b49401255..90534032bf 100644
--- a/src/melee/gm/gmmain_lib.c
+++ b/src/melee/gm/gmmain_lib.c
@@ -134,37 +134,37 @@ void* gmMainLib_8015CCFC(void)
     return &gmMainLib_804D3EE0->thing.x1A4C;
 }
 
-void* gmMainLib_8015CD08(void)
+void* gmMainLib_GetVsPlayContestants(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A30;
 }
 
-void* gmMainLib_8015CD14(void)
+void* gmMainLib_GetVsPlayTime(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A34;
 }
 
-void* gmMainLib_8015CD20(void)
+void* gmMainLib_GetCombinedVSPlayTime(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A38;
 }
 
-void* gmMainLib_8015CD2C(void)
+void* gmMainLib_GetTimeMatchTotal(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A18;
 }
 
-void* gmMainLib_8015CD38(void)
+void* gmMainLib_GetStockMatchTotal(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A1C;
 }
 
-void* gmMainLib_8015CD44(void)
+void* gmMainLib_GetCoinMatchTotal(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A20;
 }
 
-void* gmMainLib_8015CD50(void)
+void* gmMainLib_GetBonusMatchTotal(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A24;
 }
@@ -174,12 +174,12 @@ void* gmMainLib_8015CD5C(void)
     return &gmMainLib_804D3EE0->thing.x1A28;
 }
 
-void* gmMainLib_8015CD68(void)
+void* gmMainLib_GetMatchResetCounter(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A2C;
 }
 
-void* gmMainLib_8015CD74(void)
+void* gmMainLib_GetSingleplayerTime(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A3C;
 }
@@ -189,27 +189,27 @@ void* gmMainLib_8015CD80(void)
     return &gmMainLib_804D3EE0->thing.x1A44;
 }
 
-void* gmMainLib_8015CD8C(void)
+void* gmMainLib_GetPowerCount(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A50;
 }
 
-int* gmMainLib_8015CD98(void)
+int* gm_GetPowerTime(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A54;
 }
 
-void* gmMainLib_8015CDA4(void)
+void* gmMainLib_GetTotalDamage(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A58;
 }
 
-s32* gmMainLib_8015CDB0(void)
+s32* gmMainLib_GetKOTotal(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A5C;
 }
 
-void* gmMainLib_8015CDBC(void)
+void* gmMainLib_GetSelfDestructTotal(void)
 {
     return &gmMainLib_804D3EE0->thing.x1A60;
 }
@@ -1040,7 +1040,7 @@ void gmMainLib_8015FA34(int arg0)
         if ((arg0 != 0 && arg0 != 2) || lb_8001B6E0(i) != 0) {
             gmMainLib_8015F600(i, 0);
         } else if (i == 1 && !gmMainLib_8046B0F0.x0) {
-            gm_80162B98();
+            gm_IncrementPowerCount();
         }
     }
     if (g_debugLevel > 2 && db_804D6B20 != 0) {
diff --git a/src/melee/gm/gmmain_lib.h b/src/melee/gm/gmmain_lib.h
index 62b8f7d264..771f7f1e39 100644
--- a/src/melee/gm/gmmain_lib.h
+++ b/src/melee/gm/gmmain_lib.h
@@ -20,22 +20,22 @@
 /* 15CCE4 */ void* gmMainLib_8015CCE4(void);
 /* 15CCF0 */ u32* gmMainLib_8015CCF0(void);
 /* 15CCFC */ void* gmMainLib_8015CCFC(void);
-/* 15CD08 */ void* gmMainLib_8015CD08(void);
-/* 15CD14 */ void* gmMainLib_8015CD14(void);
-/* 15CD20 */ void* gmMainLib_8015CD20(void);
-/* 15CD2C */ void* gmMainLib_8015CD2C(void);
-/* 15CD38 */ void* gmMainLib_8015CD38(void);
-/* 15CD44 */ void* gmMainLib_8015CD44(void);
-/* 15CD50 */ void* gmMainLib_8015CD50(void);
+/* 15CD08 */ void* gmMainLib_GetVsPlayContestants(void);
+/* 15CD14 */ void* gmMainLib_GetVsPlayTime(void);
+/* 15CD20 */ void* gmMainLib_GetCombinedVSPlayTime(void);
+/* 15CD2C */ void* gmMainLib_GetTimeMatchTotal(void);
+/* 15CD38 */ void* gmMainLib_GetStockMatchTotal(void);
+/* 15CD44 */ void* gmMainLib_GetCoinMatchTotal(void);
+/* 15CD50 */ void* gmMainLib_GetBonusMatchTotal(void);
 /* 15CD5C */ void* gmMainLib_8015CD5C(void);
-/* 15CD68 */ void* gmMainLib_8015CD68(void);
-/* 15CD74 */ void* gmMainLib_8015CD74(void);
+/* 15CD68 */ void* gmMainLib_GetMatchResetCounter(void);
+/* 15CD74 */ void* gmMainLib_GetSingleplayerTime(void);
 /* 15CD80 */ void* gmMainLib_8015CD80(void);
-/* 15CD8C */ void* gmMainLib_8015CD8C(void);
-/* 15CD98 */ int* gmMainLib_8015CD98(void);
-/* 15CDA4 */ void* gmMainLib_8015CDA4(void);
-/* 15CDB0 */ s32* gmMainLib_8015CDB0(void);
-/* 15CDBC */ void* gmMainLib_8015CDBC(void);
+/* 15CD8C */ void* gmMainLib_GetPowerCount(void);
+/* 15CD98 */ int* gm_GetPowerTime(void);
+/* 15CDA4 */ void* gmMainLib_GetTotalDamage(void);
+/* 15CDB0 */ s32* gmMainLib_GetKOTotal(void);
+/* 15CDBC */ void* gmMainLib_GetSelfDestructTotal(void);
 /* 15CDC8 */ struct gmm_x0_528_t* gmMainLib_8015CDC8(void);
 /* 15CDD4 */ struct gmm_x0_528_t* gmMainLib_8015CDD4(void);
 /* 15CDE0 */ struct gmm_x0_528_t* gmMainLib_8015CDE0(void);
diff --git a/src/melee/gm/gmvsmelee.c b/src/melee/gm/gmvsmelee.c
index 859840e011..14045e2efd 100644
--- a/src/melee/gm/gmvsmelee.c
+++ b/src/melee/gm/gmvsmelee.c
@@ -291,7 +291,7 @@ void gm_801A5F64(MinorScene* minor_data, VsModeData* vs_data, u8 next_scene)
     if (gm_801A52D0(match_end)) {
         gm_8016247C(gm_801688AC(match_end));
         if (minor_data[1].idx != 0xFF) {
-            gm_8016279C();
+            gm_GetVsPlayMatchTotal();
             unk_bool = false;
             idx = gm_801A5360(match_end);
             unk = gm_80172DD4(gmMainLib_8015ED98()->x0);
diff --git a/src/melee/if/soundtest.c b/src/melee/if/soundtest.c
index 75582844c3..ab734cc887 100644
--- a/src/melee/if/soundtest.c
+++ b/src/melee/if/soundtest.c
@@ -803,11 +803,11 @@ void fn_80300CC8(int arg0)
         break;
     case 6:
         lbAudioAx_80024030(1);
-        r29 = gmMainLib_8015CD2C();
-        r30 = gmMainLib_8015CD38();
-        r31 = gmMainLib_8015CD44();
+        r29 = gmMainLib_GetTimeMatchTotal();
+        r30 = gmMainLib_GetStockMatchTotal();
+        r31 = gmMainLib_GetCoinMatchTotal();
         {
-            int* ptr = gmMainLib_8015CD50();
+            int* ptr = gmMainLib_GetBonusMatchTotal();
             *r29 = un_804D6DE0;
             *r30 = un_804D6DE4;
             *r31 = un_804D6DE8;
@@ -898,7 +898,7 @@ s32 un_80300F98(s32 arg0)
     s32* temp_r31;
 
     if (arg0 == 1) {
-        temp_r31 = gmMainLib_8015CDB0();
+        temp_r31 = gmMainLib_GetKOTotal();
         lbAudioAx_80024030(1);
         *temp_r31 = un_804D6DF4;
         gm_801A42F8(1);
diff --git a/src/melee/if/textlib.c b/src/melee/if/textlib.c
index b0bf92f76e..198b9d967c 100644
--- a/src/melee/if/textlib.c
+++ b/src/melee/if/textlib.c
@@ -984,7 +984,7 @@ bool un_80304780(void)
     }
 }
 
-int un_80304870(void)
+int un_GetTrophyTotal(void)
 {
     if (gm_8016B498() || gm_801A4310() == MJ_TOY_LOTTERY) {
         return (short) un_804A284C[0x258 / 2];
diff --git a/src/melee/if/textlib.h b/src/melee/if/textlib.h
index e13e16a6e3..5c2edff0cb 100644
--- a/src/melee/if/textlib.h
+++ b/src/melee/if/textlib.h
@@ -52,7 +52,7 @@
 /* 3045A0 */ int un_803045A0(void);
 /* 304690 */ int un_80304690(void);
 /* 304780 */ bool un_80304780(void);
-/* 304870 */ int un_80304870(void);
+/* 304870 */ int un_GetTrophyTotal(void);
 /* 3048C0 */ int un_803048C0(int);
 /* 304924 */ bool un_80304924(int);
 /* 304988 */ void un_80304988(int);
diff --git a/src/melee/lb/lbcardgame.c b/src/melee/lb/lbcardgame.c
index 9a2110a3f3..68220c8c49 100644
--- a/src/melee/lb/lbcardgame.c
+++ b/src/melee/lb/lbcardgame.c
@@ -227,7 +227,7 @@ void lb_8001CE00(void)
     if (!lb_80433318.enable) {
         __assert("lbcardgame.c", 0x2A3, "_p(enable)");
     }
-    *gmMainLib_8015CD98() += gmMainLib_8015FC74();
+    *gm_GetPowerTime() += gmMainLib_8015FC74();
     lb_80433318.xC = true;
 }
 
diff --git a/src/melee/mn/mncount.c b/src/melee/mn/mncount.c
index de4114f01f..7867667f47 100644
--- a/src/melee/mn/mncount.c
+++ b/src/melee/mn/mncount.c
@@ -1,69 +1,155 @@
-#include "mncount.h"
-
-#include "types.h"
-
-#include "gm/gm_1601.h"
-#include "gm/gmmain_lib.h"
-#include "mn/inlines.h"
-#include "mn/mnmain.h"
-#include "sysdolphin/baselib/gobjplink.h"
+#include "placeholder.h"
 
 #include <baselib/gobj.h>
-#include <baselib/jobj.h>
-
-extern s32 mnDiagram_GetFighterTotalKOs(s32);
-extern s32 mnDiagram_GetFighterTotalFalls(s32);
-
-/// --- Internal Structs ---
-
-typedef struct LocalFighterData {
-    char pad[0x34];
-    u16 unk34;
-    char pad1a[0xA];
-    u32 unk40;
-    u32 unk44;
-    char pad2[0x8];
-    u16 unk50;
-    u16 unk52;
-    u32 unk54;
-} LocalFighterData;
-
-/// --- Function Implementation ---
+#include <sysdolphin/baselib/debug.h>
+#include <sysdolphin/baselib/gobj.h>
+#include <sysdolphin/baselib/gobjgxlink.h>
+#include <sysdolphin/baselib/gobjobject.h>
+#include <sysdolphin/baselib/gobjplink.h>
+#include <sysdolphin/baselib/gobjproc.h>
+#include <sysdolphin/baselib/gobjuserdata.h>
+#include <sysdolphin/baselib/jobj.h>
+#include <sysdolphin/baselib/memory.h>
+#include <sysdolphin/baselib/sislib.h>
+#include <melee/gm/gm_1601.h>
+#include <melee/gm/gmmain_lib.h>
+#include <melee/gm/types.h>
+#include <melee/if/textlib.h>
+#include <melee/lb/lb_00F9.h>
+#include <melee/lb/lbarchive.h>
+#include <melee/lb/lbaudio_ax.h>
+#include <melee/mn/inlines.h>
+#include <melee/mn/mncount.h>
+#include <melee/mn/mndiagram.h>
+#include <melee/mn/mnmain.h>
+#include <melee/mn/mnname.h>
+#include <melee/mn/types.h>
+#include <melee/sc/types.h>
+
+// DATA / MELEE RECORDS / MISC RECORDS
+#define NUM_STAGES 29
+#define NUM_CHARACTERS 25
+#define MAX_SCROLL 20
+
+static AnimLoopSettings mnCount_803EFAA0 = { 0.0f, 199.0f, 0.0f };
+static StaticModelDesc model_desc;
+static HSD_GObj* menu_gobj;
+static HSD_JObj* menu_jobj;
+static u16 mnCount_sis_idx[30] = {
+    0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF, 0xD0, 0xD1, 0xD2,
+    0xD3, 0xD4, 0xD5, 0xD6, 0xD7, 0xD8, 0xD9, 0xDA, 0xDB, 0xDC,
+    0xDD, 0xDE, 0xDF, 0xE0, 0xE1, 0xE2, 0xE3, 0xE4, 0xE5, 0xE6,
+};
+extern GXColor mn_804D4B64;
+
+static inline bool inline_is_row_time(int row)
+{
+    switch (row) {
+    case POWER_TIME:
+    case PLAY_TIME:
+    case SINGLEPLAYER_TIME:
+    case VS_PLAY_TIME:
+    case COMBINED_VS_PLAY_TIME:
+        return true;
+    default:
+        return false;
+    }
+}
 
-s32 mnCount_802502CC(s32 arg0)
+static inline bool inline_is_row_char(int row)
 {
-    return ((LocalFighterData*) GetPersistentFighterData(arg0))->unk54;
+    switch (row) {
+    case LONGEST_TIME:
+    case SECOND_LONGEST_TIME:
+    case SHORTEST_TIME:
+    case SMASH_CHAMP:
+    case SMASH_SAP:
+    case SLUG_MEISTER:
+    case PUNCHING_BAG:
+    case KO_KINGPIN:
+    case NO_DEFENSE_NELLY:
+    case DISASTER_MASTER:
+        return true;
+    default:
+        return false;
+    }
 }
 
-s32 fn_802502F0(u8 arg0)
+static inline void inline_update_entries(HSD_GObj* gobj)
 {
-    return mnDiagram_GetFighterTotalKOs(arg0);
+    MnCountData* userdata = GET_MNCOUNT(gobj);
+    int i;
+    for (i = 0; i < MNCOUNT_VISIBLE_ROWS; i++) {
+        mnCount_CreateRow(gobj, i, userdata->scroll_pos + i);
+    }
 }
 
-s32 fn_80250314(u8 arg0)
+static inline void inline_free_text(HSD_GObj* gobj)
 {
-    return mnDiagram_GetFighterTotalFalls(arg0);
+    MnCountData* userdata = GET_MNCOUNT(gobj);
+    int i;
+
+    for (i = 0; i < MNCOUNT_VISIBLE_ROWS; i++) {
+        if (userdata->labels[i] != NULL) {
+            HSD_SisLib_803A5CC4(userdata->labels[i]);
+            userdata->labels[i] = NULL;
+        }
+        if (userdata->values[i] != NULL) {
+            HSD_SisLib_803A5CC4(userdata->values[i]);
+            userdata->values[i] = NULL;
+        }
+    }
+
+    HSD_SisLib_803A5CC4(userdata->title);
 }
 
-s32 fn_80250338(s32 arg0)
+static inline void mnCount_8025186C_inline(HSD_GObj* gobj)
 {
-    return ((LocalFighterData*) GetPersistentFighterData(arg0))->unk34;
+    MnCountData* userdata = GET_MNCOUNT(gobj);
+    HSD_Text* text;
+    if (userdata->title != NULL) {
+        HSD_SisLib_803A5CC4(userdata->title);
+    }
+    text =
+        HSD_SisLib_803A5ACC(0, 1, -9.5f, 9.1f, 17.0f, 364.68332f, 38.38772f);
+    userdata->title = text;
+    text->font_size.x = 0.0521f;
+    text->font_size.y = 0.0521f;
+    HSD_SisLib_803A6368(text, 166);
 }
 
-inline bool mnCount_8025035C_inline(void)
+static inline bool mnCount_8025035C_inline(void)
 {
     s32 i;
     for (i = 0; i < 25; i++) {
-        LocalFighterData* fdata =
-            (LocalFighterData*) GetPersistentFighterData(i);
-        if (fdata->unk54 != 0) {
+        if (GetPersistentFighterData(i)->play_time != 0) {
             return false;
         }
     }
     return true;
 }
 
-int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s8))
+u32 mnCount_GetMatchTime(s32 arg0)
+{
+    return GetPersistentFighterData(arg0)->play_time;
+}
+
+u32 mnCount_GetKOKingpin(s32 arg0)
+{
+    return mnDiagram_GetFighterTotalKOs(arg0);
+}
+
+u32 mnCount_GetNoDefenseNelly(s32 arg0)
+{
+    return mnDiagram_GetFighterTotalFalls(arg0);
+}
+
+u32 mnCount_GetDisasterMaster(s32 arg0)
+{
+    return GetPersistentFighterData(arg0)->sd_count;
+}
+
+int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s32))
 {
     PAD_STACK(4);
     {
@@ -80,7 +166,7 @@ int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s8))
 
         for (i = 0; i < arr_size; i++) {
             sp18[i].id = i;
-            sp18[i].val = get_val_func((s8) i);
+            sp18[i].val = get_val_func(i);
         }
 
         for (i = 0; i < arr_size; i++) {
@@ -111,8 +197,7 @@ int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s8))
                 j = i + 1;
                 while (j < arr_size) {
                     if (gm_80164840(gm_8016400C(sp18[j].id)) != 0 &&
-                        get_val_func((s8) sp18[i].id) ==
-                            get_val_func((s8) sp18[j].id))
+                        get_val_func(sp18[i].id) == get_val_func(sp18[j].id))
                     {
                         i++;
                         if (skip_count != 0) {
@@ -129,9 +214,7 @@ int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s8))
                         continue;
                     }
 
-                    if (get_val_func((s8) sp18[i].id) ==
-                        get_val_func((s8) sp18[j].id))
-                    {
+                    if (get_val_func(sp18[i].id) == get_val_func(sp18[j].id)) {
                         return arr_size;
                     }
                 }
@@ -143,128 +226,405 @@ int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s8))
     }
 }
 
-s32 fn_8025069C(s32 arg0)
+u32 mnCount_GetSmashChamp(s32 arg0)
 {
-    return ((LocalFighterData*) GetPersistentFighterData(arg0))->unk50;
+    return GetPersistentFighterData(arg0)->victories;
 }
 
-s32 fn_802506C0(s32 arg0)
+u32 mnCount_GetSmashSap(s32 arg0)
 {
-    return ((LocalFighterData*) GetPersistentFighterData(arg0))->unk52;
+    return GetPersistentFighterData(arg0)->losses;
 }
 
-s32 fn_802506E4(s32 arg0)
+u32 mnCount_GetSlugMeister(s32 arg0)
 {
-    return ((LocalFighterData*) GetPersistentFighterData(arg0))->unk40;
+    return GetPersistentFighterData(arg0)->damage_dealt;
 }
 
-s32 fn_80250708(s32 arg0)
+u32 mnCount_GetPunchingBag(s32 arg0)
 {
-    return ((LocalFighterData*) GetPersistentFighterData(arg0))->unk44;
+    return GetPersistentFighterData(arg0)->damage_taken;
 }
 
-s32 mnCount_8025072C(CountEntry* entries, s32 start_idx, s32 mode)
+#define GET_KOS(i) mnCount_GetKOKingpin(entries[i].id)
+#define GET_FALLS(i) mnCount_GetNoDefenseNelly(entries[i].id)
+s32 mnCount_8025072C(CountEntry* entries, s32 start_idx, bool mode)
 {
-    s32 i;
-    s32 best_idx = start_idx;
-    s32 found_tie = 0;
-    s32 is_invalid = 0;
-    s32 curr_stat, best_stat;
-
-    for (i = start_idx + 1; i < 25; i++) {
-        if (entries[i].val != entries[start_idx].val) {
-            break;
-        }
-
-        curr_stat = mnDiagram_GetFighterTotalKOs(entries[i].id);
-        best_stat = mnDiagram_GetFighterTotalKOs(entries[best_idx].id);
-
-        if (mode == 0) {
-            if (curr_stat > best_stat) {
-                best_idx = i;
+    int idx, best_idx;
+    bool tie = false;
+    best_idx = start_idx;
+    for (idx = start_idx + 1; idx < NUM_CHARACTERS; idx++) {
+        if (entries[start_idx].val == entries[idx].val) {
+            if (mode) {
+                if (GET_KOS(best_idx) > GET_KOS(idx)) {
+                    best_idx = idx;
+                }
+            } else {
+                if (GET_KOS(best_idx) < GET_KOS(idx)) {
+                    best_idx = idx;
+                }
             }
-        } else {
-            if (curr_stat < best_stat) {
-                best_idx = i;
+            if (GET_KOS(best_idx) == GET_KOS(idx)) {
+                tie = true;
+                break;
             }
         }
-
-        curr_stat = mnDiagram_GetFighterTotalKOs(entries[i].id);
-        best_stat = mnDiagram_GetFighterTotalKOs(entries[best_idx].id);
-
-        if (curr_stat == best_stat) {
-            found_tie = 1;
-        }
     }
-
-    if (found_tie == 0) {
+    if (!tie) {
         return entries[best_idx].id;
     }
-
-    for (i = start_idx + 1; i < 25; i++) {
-        if (i == best_idx) {
+    tie = false;
+    for (idx = start_idx + 1; idx < NUM_CHARACTERS; idx++) {
+        if (idx == best_idx) {
             continue;
         }
+        if (entries[start_idx].val == entries[idx].val) {
+            if (GET_KOS(start_idx) == GET_KOS(idx)) {
+                if (GET_FALLS(best_idx) == GET_FALLS(idx)) {
+                    tie = true;
+                    break;
+                }
+                if (mode) {
+                    if (GET_FALLS(best_idx) < GET_FALLS(idx)) {
+                        best_idx = idx;
+                    }
+                } else {
+                    if (GET_FALLS(best_idx) > GET_FALLS(idx)) {
+                        best_idx = idx;
+                    }
+                }
+            }
+        }
+    }
+    if (tie) {
+        return NUM_CHARACTERS;
+    } else {
+        return entries[best_idx].id;
+    }
+}
+#undef GET_KOS
+#undef GET_FALLS
 
-        if (entries[i].val != entries[start_idx].val) {
-            break;
+static inline bool mnCount_8025092C_inline(void)
+{
+    int i;
+    for (i = 0; i < NUM_CHARACTERS; i++) {
+        if (mnCount_GetMatchTime(i) != 0) {
+            return false;
         }
+    }
+    return true;
+}
 
-        curr_stat = mnDiagram_GetFighterTotalKOs(entries[i].id);
-        best_stat = mnDiagram_GetFighterTotalKOs(entries[best_idx].id);
-        if (curr_stat != best_stat) {
+s32 mnCount_8025092C(s32 rank, u32 (*getVal)(s32), bool mode)
+{
+    CountEntry entries[NUM_CHARACTERS];
+    int i, j, min;
+    if (mnCount_8025092C_inline()) {
+        return NUM_CHARACTERS;
+    }
+    for (i = 0; i < NUM_CHARACTERS; i++) {
+        entries[i].id = i;
+        entries[i].val = getVal(i);
+    }
+    for (i = 0; i < NUM_CHARACTERS; i++) {
+        min = i;
+        for (j = i + 1; j < NUM_CHARACTERS; j++) {
+            if (entries[j].val < entries[min].val) {
+                min = j;
+            }
+        }
+        if (min != i) {
+            CountEntry tmp = entries[min];
+            for (j = min; j > i; j--) {
+                entries[j] = entries[j - 1];
+            }
+            entries[i] = tmp;
+        }
+    }
+    for (i = 0; i < NUM_CHARACTERS; i++) {
+        if (!gm_80164840(gm_8016400C(entries[i].id))) {
             continue;
         }
+        if (rank != 0) {
+            rank--;
+            for (j = i + 1; j < NUM_CHARACTERS; j++) {
+                if (!gm_80164840(gm_8016400C(entries[j].id))) {
+                    continue;
+                }
+                if (getVal(entries[i].id) == getVal(entries[j].id)) {
+                    i++;
+                    if (rank != 0) {
+                        rank--;
+                    } else {
+                        return NUM_CHARACTERS;
+                    }
+                }
+            }
+        } else {
+            for (j = i + 1; j < NUM_CHARACTERS; j++) {
+                if (!gm_80164840(gm_8016400C(entries[j].id))) {
+                    continue;
+                }
+                if (getVal(entries[i].id) == getVal(entries[j].id)) {
+                    return mnCount_8025072C(entries, i, mode);
+                }
+            }
+            return entries[i].id;
+        }
+    }
+    return NUM_CHARACTERS;
+}
 
-        curr_stat = mnDiagram_GetFighterTotalFalls(entries[i].id);
-        best_stat = mnDiagram_GetFighterTotalFalls(entries[best_idx].id);
-
-        if (curr_stat == best_stat) {
-            is_invalid = 1;
-            break;
+int mnCount_GetRowValue_Character(mnCount_row row)
+{
+    int c;
+    int i;
+    switch (row) {
+    case LONGEST_TIME:
+        return mnCount_8025035C(0, mnCount_GetMatchTime);
+    case SECOND_LONGEST_TIME:
+        return mnCount_8025035C(1, mnCount_GetMatchTime);
+    case SHORTEST_TIME:
+        c = 0;
+        for (i = 0; i < NUM_CHARACTERS; i++) {
+            if (gm_80164840(gm_8016400C(i))) {
+                c += 1;
+            }
+        }
+        if (c < 3) {
+            return NUM_CHARACTERS;
         }
+        return mnCount_8025035C(c - 1, mnCount_GetMatchTime);
+    case SMASH_CHAMP:
+        return mnCount_8025092C(0, mnCount_GetSmashChamp, 0);
+    case SMASH_SAP:
+        return mnCount_8025092C(0, mnCount_GetSmashSap, 1);
+    case SLUG_MEISTER:
+        return mnCount_8025092C(0, mnCount_GetSlugMeister, 0);
+    case PUNCHING_BAG:
+        return mnCount_8025092C(0, mnCount_GetPunchingBag, 1);
+    case KO_KINGPIN:
+        return mnCount_8025035C(0, mnCount_GetKOKingpin);
+    case NO_DEFENSE_NELLY:
+        return mnCount_8025035C(0, mnCount_GetNoDefenseNelly);
+    case DISASTER_MASTER:
+        return mnCount_8025035C(0, mnCount_GetDisasterMaster);
+    default:
+        return NUM_CHARACTERS;
+    }
+}
 
-        if (mode == 0) {
-            if (curr_stat < best_stat) {
-                best_idx = i;
+unsigned int mnCount_GetRowValue_Number(int row)
+{
+    unsigned int ret;
+    switch (row) {
+    case POWER_COUNT:
+        ret = *(unsigned int*) gmMainLib_GetPowerCount();
+        break;
+    case POWER_TIME:
+        ret = *gm_GetPowerTime();
+        break;
+    case PLAY_TIME:
+        ret = gm_GetPlayTime();
+        break;
+    case SINGLEPLAYER_TIME:
+        ret = *(unsigned int*) gmMainLib_GetSingleplayerTime();
+        break;
+    case VS_PLAY_TIME:
+        ret = *(unsigned int*) gmMainLib_GetVsPlayTime();
+        break;
+    case COMBINED_VS_PLAY_TIME:
+        ret = *(unsigned int*) gmMainLib_GetCombinedVSPlayTime();
+        break;
+    case VS_PLAY_MATCH_TOTAL:
+        ret = gm_GetVsPlayMatchTotal();
+        break;
+    case TIME_MATCH_TOTAL:
+        ret = *(unsigned int*) gmMainLib_GetTimeMatchTotal();
+        break;
+    case STOCK_MATCH_TOTAL:
+        ret = *(unsigned int*) gmMainLib_GetStockMatchTotal();
+        break;
+    case COIN_MATCH_TOTAL:
+        ret = *(unsigned int*) gmMainLib_GetCoinMatchTotal();
+        break;
+    case BONUS_MATCH_TOTAL:
+        ret = *(unsigned int*) gmMainLib_GetBonusMatchTotal();
+        break;
+    case VS_PLAY_CONTESTANTS:
+        ret = *(unsigned int*) gmMainLib_GetVsPlayContestants();
+        break;
+    case MATCH_RESET_COUNTER:
+        ret = *(unsigned int*) gmMainLib_GetMatchResetCounter();
+        break;
+    case TOTAL_DAMAGE:
+        ret = *(unsigned int*) gmMainLib_GetTotalDamage();
+        break;
+    case KO_TOTAL:
+        ret = *gmMainLib_GetKOTotal();
+        break;
+    case SELFDESTRUCT_TOTAL:
+        ret = *(unsigned int*) gmMainLib_GetSelfDestructTotal();
+        break;
+    case AVAILABLE_CHARACTERS: {
+        int i, c;
+        c = 0;
+        for (i = 0; i < NUM_CHARACTERS; i++) {
+            if (gm_80164840(gm_8016400C(i))) {
+                c++;
             }
-        } else {
-            if (curr_stat > best_stat) {
-                best_idx = i;
+        }
+        ret = c;
+    } break;
+    case AVAILABLE_MAPS: {
+        int i, c;
+        c = 0;
+        for (i = 0; i < NUM_STAGES; i++) {
+            if (gm_80164430(gm_801641CC(i))) {
+                c++;
             }
         }
+        ret = c;
+    } break;
+    case TROPHY_TOTAL:
+        ret = un_GetTrophyTotal();
+        break;
+    case NAME_TOTAL:
+        ret = GetNameCount();
+        break;
+    default:
+        if (inline_is_row_time(row)) {
+            ret = 359940; // if in seconds, 99h 59m
+        } else {
+            ret = 9999;
+        }
+        break;
     }
 
-    if (is_invalid) {
-        return 25;
+    if (inline_is_row_time(row)) {
+        if (ret > 3599999940u) {
+            ret = 3599999940u; // if in seconds, 999999h 59m
+        }
+    } else if (ret > 999999999) {
+        ret = 999999999;
     }
-    return entries[best_idx].id;
+    return ret;
 }
 
-extern AnimLoopSettings mnCount_803EFAA0;
+void mnCount_CreateRow(HSD_GObj* gobj, int visible_row, mnCount_row data_row)
+{
+    MnCountData* userdata = GET_MNCOUNT(gobj);
+    HSD_Text* text;
+    unsigned int row_value;
+    int row_value_2;
+    float y;
+    char buf1[8];
+    char buf2[8];
+    static GXColor text_color = { 0xAA, 0xAA, 0xAA, 0xFF };
+    if (userdata->labels[visible_row] != NULL) {
+        HSD_SisLib_803A5CC4(userdata->labels[visible_row]);
+        userdata->labels[visible_row] = NULL;
+    }
+    y = (1.4f * visible_row) + -6.4f;
+    text = HSD_SisLib_803A5ACC(0, 1, -13.0f, y, 17.0f, 500.0f, 38.38772f);
+    userdata->labels[visible_row] = text;
+    text->font_size.x = 0.03f;
+    text->font_size.y = 0.03f;
+    text->text_color = text_color;
+    HSD_SisLib_803A6368(text, mnCount_sis_idx[data_row]);
+    if (userdata->values[visible_row] != NULL) {
+        HSD_SisLib_803A5CC4(userdata->values[visible_row]);
+        userdata->values[visible_row] = NULL;
+    }
+    text = HSD_SisLib_803A6754(0, 1);
+    userdata->values[visible_row] = text;
+    text->pos_x = 13.0f;
+    text->pos_y = y;
+    text->pos_z = 17.0f;
+    text->text_color = mn_804D4B64;
+    text->default_alignment = 2;
+    if (inline_is_row_time(data_row)) {
+        text->font_size.x = 0.03f;
+        text->font_size.y = 0.03f;
+        row_value = mnCount_GetRowValue_Number(data_row);
+        mn_8022EA78(buf1, 2, row_value / 60 / 60);
+        mn_8022EA78(buf2, 2, row_value / 60 % 60);
+        HSD_SisLib_803A6B98(text, 0.0f, 0.0f, "%u:%s", row_value / 60 / 60,
+                            buf2);
+    } else if (inline_is_row_char(data_row)) {
+        text->font_size.x = 0.03f;
+        text->font_size.y = 0.03f;
+        row_value_2 = mnCount_GetRowValue_Character(data_row);
+        if (row_value_2 == NUM_CHARACTERS) {
+            HSD_SisLib_803A6B98(text, 0.0f, 0.0f, "－");
+        } else {
+            gm_80160B40(text, gm_8016400C(row_value_2), 0);
+        }
+    } else {
+        text->font_size.x = 0.03f;
+        text->font_size.y = 0.03f;
+        HSD_SisLib_803A6B98(text, 0.0f, 0.0f, "%u",
+                            mnCount_GetRowValue_Number(data_row));
+    }
+}
 
-void mnCount_802513F4(HSD_GObj* gobj)
+void mnCount_HandleUserInput(HSD_GObj* gobj)
 {
-    HSD_JObj* jobj;
-    MnCountData* data = gobj->user_data;
-    HSD_JObj* tree = gobj->hsd_obj;
-    FORCE_PAD_STACK_4;
+    MnCountData* userdata = GET_MNCOUNT(menu_gobj);
+    long long x;
+    if (mn_804D6BC8.cooldown != 0) {
+        Menu_DecrementAnimTimer();
+    } else {
+        x = mn_804A04F0.buttons = mn_80229624(4);
+        if (x & MenuInput_Back) {
+            sfxBack();
+            mn_804A04F0.entering_menu = 0;
+            mn_80229894(MENU_KIND_RECORDS, 2, 3);
+        } else if (x & MenuInput_Up) {
+            if (userdata->scroll_pos != 0) {
+                userdata->scroll_pos -= 1;
+                sfxMove();
+                inline_update_entries(menu_gobj);
+            }
+        } else if (x & MenuInput_Down) {
+            if (userdata->scroll_pos < MAX_SCROLL) {
+                sfxMove();
+                userdata->scroll_pos += 1;
+                inline_update_entries(menu_gobj);
+            }
+        }
+    }
+}
 
-    lb_80011E24(tree, &jobj, 2, -1);
-    if (data->cursor != 0) {
-        HSD_JObjClearFlagsAll(jobj, JOBJ_HIDDEN);
+void mnCount_UpdateArrowIndicators(HSD_GObj* gobj)
+{
+    MnCountData* userdata = GET_MNCOUNT(gobj);
+    HSD_JObj* jobj = gobj->hsd_obj; // GET_JOBJ breaks this
+    HSD_JObj* child;
+
+    // up arrow
+    lb_80011E24(jobj, &child, 2, -1);
+    if (userdata->scroll_pos != 0) {
+        HSD_JObjClearFlagsAll(child, JOBJ_HIDDEN);
     } else {
-        HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
+        HSD_JObjSetFlagsAll(child, JOBJ_HIDDEN);
     }
 
-    lb_80011E24(tree, &jobj, 1, -1);
-    if (data->cursor < 0x14) {
-        HSD_JObjClearFlagsAll(jobj, JOBJ_HIDDEN);
+    // down arrow
+    lb_80011E24(jobj, &child, 1, -1);
+    if (userdata->scroll_pos < MAX_SCROLL) {
+        HSD_JObjClearFlagsAll(child, JOBJ_HIDDEN);
     } else {
-        HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
+        HSD_JObjSetFlagsAll(child, JOBJ_HIDDEN);
     }
 
-    mn_8022ED6C(tree, &mnCount_803EFAA0);
+    mn_8022ED6C(jobj, &mnCount_803EFAA0);
+}
+
+static void mnCount_UpdateArrowIndicators_noinline(HSD_GObj* gobj)
+{
+    mnCount_UpdateArrowIndicators(gobj);
 }
 
 void fn_802514B8(HSD_GObj* gobj)
@@ -272,15 +632,163 @@ void fn_802514B8(HSD_GObj* gobj)
     HSD_GObjPLink_80390228(gobj);
 }
 
-void mnCount_802517E0(MnCountData* data)
+void fn_802514D8(HSD_GObj* gobj)
 {
+    MnCountData* userdata = GET_MNCOUNT(gobj);
+    HSD_GObjProc* proc;
+    if (mn_804A04F0.cur_menu != MENU_KIND_RECORDS_MISC) {
+        HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+        proc = HSD_GObj_SetupProc(gobj, fn_802514B8, 0);
+        proc->flags_3 = HSD_GObj_804D783C;
+        {
+            // inline_free_text
+
+            MnCountData* userdata2 = GET_MNCOUNT(gobj);
+            MnCountData* userdata3 = userdata2;
+            int i;
+
+            for (i = 0; i < MNCOUNT_VISIBLE_ROWS; i++) {
+                if (userdata2->labels[i] != NULL) {
+                    HSD_SisLib_803A5CC4(userdata3->labels[i]);
+                    userdata2->labels[i] = NULL;
+                }
+                if (userdata2->values[i] != NULL) {
+                    HSD_SisLib_803A5CC4(userdata3->values[i]);
+                    userdata2->values[i] = NULL;
+                }
+            }
+
+            HSD_SisLib_803A5CC4(userdata->title);
+        }
+    } else {
+        // mnCount_UpdateArrowIndicators
+
+        HSD_JObj* jobj = gobj->hsd_obj;
+        HSD_JObj* child;
+
+        // up arrow
+        lb_80011E24(jobj, &child, 2, -1);
+        if (userdata->scroll_pos != 0) {
+            HSD_JObjClearFlagsAll(child, JOBJ_HIDDEN);
+        } else {
+            HSD_JObjSetFlagsAll(child, JOBJ_HIDDEN);
+        }
+
+        // down arrow
+        lb_80011E24(jobj, &child, 1, -1);
+        if (userdata->scroll_pos < MAX_SCROLL) {
+            HSD_JObjClearFlagsAll(child, JOBJ_HIDDEN);
+        } else {
+            HSD_JObjSetFlagsAll(child, JOBJ_HIDDEN);
+        }
+
+        mn_8022ED6C(jobj, &mnCount_803EFAA0);
+    }
+}
+
+void fn_80251640(HSD_GObj* gobj)
+{
+    MnCountData* userdata = GET_MNCOUNT(gobj);
+    HSD_GObjProc* proc;
     int i;
+    HSD_JObj* jobj;
+    StaticModelDesc* md;
+
+    if (mn_804A04F0.cur_menu != MENU_KIND_RECORDS_MISC) {
+        HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+        proc = HSD_GObj_SetupProc(gobj, fn_802514B8, 0);
+        proc->flags_3 = HSD_GObj_804D783C;
+        {
+            // inline_free_text
+
+            MnCountData* userdata2 = GET_MNCOUNT(gobj);
+            MnCountData* userdata3 = userdata2;
+            int i;
+
+            for (i = 0; i < MNCOUNT_VISIBLE_ROWS; i++) {
+                if (userdata2->labels[i] != NULL) {
+                    HSD_SisLib_803A5CC4(userdata3->labels[i]);
+                    userdata2->labels[i] = NULL;
+                }
+                if (userdata2->values[i] != NULL) {
+                    HSD_SisLib_803A5CC4(userdata3->values[i]);
+                    userdata2->values[i] = NULL;
+                }
+            }
+
+            HSD_SisLib_803A5CC4(userdata->title);
+        }
+        return;
+    }
 
-    data->cursor = 0;
-    data->page_size = 0xA;
+    if (userdata->expand_anim_duration != 0) {
+        userdata->expand_anim_duration--;
+        return;
+    }
 
-    for (i = 0; i < 0xA; i++) {
-        data->labels[i] = NULL;
-        data->values[i] = NULL;
+    for (i = 0; i < MNCOUNT_VISIBLE_ROWS; i++) {
+        mnCount_CreateRow(gobj, i, userdata->scroll_pos + i);
     }
+
+    md = &model_desc;
+    jobj = HSD_JObjLoadJoint(md->joint);
+    HSD_GObjObject_80390A70(gobj, HSD_GObj_804D7849, jobj);
+    GObj_SetupGXLink(gobj, HSD_GObj_JObjCallback, 4, 0x80);
+    HSD_JObjAddAnimAll(jobj, md->animjoint, md->matanim_joint,
+                       md->shapeanim_joint);
+    HSD_JObjReqAnimAll(jobj, 0.0f);
+    mnCount_UpdateArrowIndicators_noinline(gobj);
+    HSD_GObjProc_8038FE24(HSD_GObj_804D7838);
+    proc = HSD_GObj_SetupProc(gobj, fn_802514D8, 0);
+    proc->flags_3 = HSD_GObj_804D783C;
+}
+
+void mnCount_InitUserData(MnCountData* userdata)
+{
+    int i;
+    userdata->scroll_pos = 0;
+    userdata->expand_anim_duration = 10;
+    for (i = 0; i < MNCOUNT_VISIBLE_ROWS; i++) {
+        userdata->labels[i] = NULL;
+        userdata->values[i] = NULL;
+    }
+}
+
+static void mnCount_InitUserData_noinline(MnCountData* userdata)
+{
+    mnCount_InitUserData(userdata);
+}
+
+void mnCount_Create(void)
+{
+    HSD_GObj* gobj;
+    MnCountData* userdata;
+    HSD_GObjProc* proc;
+
+    mn_804D6BC8.cooldown = 5;
+    mn_804A04F0.prev_menu = mn_804A04F0.cur_menu;
+    mn_804A04F0.cur_menu = MENU_KIND_RECORDS_MISC;
+    mn_804A04F0.hovered_selection = 0;
+
+    lbArchive_LoadSections(
+        mn_804D6BB8, (void**) &model_desc.joint, "MenMainConCo_Top_joint",
+        &model_desc.animjoint, "MenMainConCo_Top_animjoint",
+        &model_desc.matanim_joint, "MenMainConCo_Top_matanim_joint",
+        &model_desc.shapeanim_joint, "MenMainConCo_Top_shapeanim_joint", 0);
+
+    gobj = GObj_Create(6, 7, 0x80);
+    menu_gobj = gobj;
+
+    userdata = HSD_MemAlloc(sizeof(MnCountData));
+    HSD_ASSERTREPORT(0x512, userdata, "Can't get user_data.\n");
+    mnCount_InitUserData_noinline(userdata);
+    GObj_InitUserData(gobj, 0, HSD_Free, userdata);
+
+    proc = HSD_GObj_SetupProc(gobj, fn_80251640, 0);
+    proc->flags_3 = HSD_GObj_804D783C;
+    mnCount_8025186C_inline(gobj);
+
+    gobj = GObj_Create(0, 1, 0x80);
+    proc = HSD_GObj_SetupProc(gobj, mnCount_HandleUserInput, 0);
+    proc->flags_3 = HSD_GObj_804D783C;
 }
diff --git a/src/melee/mn/mncount.h b/src/melee/mn/mncount.h
index 9a1907dc54..703d84eaea 100644
--- a/src/melee/mn/mncount.h
+++ b/src/melee/mn/mncount.h
@@ -4,40 +4,80 @@
 #include <placeholder.h>
 #include <platform.h>
 
-#include "mn/forward.h"
 #include <baselib/forward.h>
-#include <baselib/sislib.h>
+#include <melee/mn/forward.h>
+
+#define MNCOUNT_VISIBLE_ROWS 10
 
 typedef struct MnCountData {
-    /* 0x00 */ u8 cursor;
-    /* 0x01 */ u8 pad_01[3];
-    /* 0x04 */ s32 page_size;
-    /* 0x08 */ HSD_Text* labels[10];
-    /* 0x30 */ HSD_Text* values[10];
-    /* 0x58 */ HSD_Text* title;
-} MnCountData; // size 0x5C
-
-/* 2502CC */ s32 mnCount_802502CC(s32);
-/* 2502F0 */ s32 fn_802502F0(u8);
-/* 250314 */ s32 fn_80250314(u8);
-/* 250338 */ s32 fn_80250338(s32);
-/* 25035C */ int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s8));
-/* 25069C */ s32 fn_8025069C(s32);
-/* 2506C0 */ s32 fn_802506C0(s32);
-/* 2506E4 */ s32 fn_802506E4(s32);
-/* 250708 */ s32 fn_80250708(s32);
+    unsigned char scroll_pos;
+    int expand_anim_duration;
+    HSD_Text* labels[MNCOUNT_VISIBLE_ROWS];
+    HSD_Text* values[MNCOUNT_VISIBLE_ROWS];
+    HSD_Text* title;
+} MnCountData;
+
+#ifdef M2C
+#define GET_MNCOUNT(gobj)                                                     \
+    ((MnCountData*) HSD_GObjGetUserData((HSD_GObj*) gobj))
+#else
+#define GET_MNCOUNT(gobj) ((MnCountData*) HSD_GObjGetUserData(gobj))
+#endif
+
+typedef enum {
+    POWER_COUNT,           // Power Count
+    POWER_TIME,            // Power Time
+    PLAY_TIME,             // Play Time
+    SINGLEPLAYER_TIME,     // Single-Player Time
+    VS_PLAY_TIME,          // VS. Play Time
+    COMBINED_VS_PLAY_TIME, // Combined VS. Play Time
+    VS_PLAY_MATCH_TOTAL,   // VS. Play Match Total
+    TIME_MATCH_TOTAL,      // Time Match Total
+    STOCK_MATCH_TOTAL,     // Stock Match Total
+    COIN_MATCH_TOTAL,      // Coin Match Total
+    BONUS_MATCH_TOTAL,     // Bonus Match Total
+    VS_PLAY_CONTESTANTS,   // VS. Play Contestants
+    MATCH_RESET_COUNTER,   // Match Reset Counter
+    TOTAL_DAMAGE,          // Total Damage
+    KO_TOTAL,              // KO Total
+    SELFDESTRUCT_TOTAL,    // Self-Destruct Total
+    AVAILABLE_CHARACTERS,  // Available Characters
+    AVAILABLE_MAPS,        // Available Maps
+    TROPHY_TOTAL,          // Trophy Total
+    NAME_TOTAL,            // Name Total
+    LONGEST_TIME,          // Longest Time
+    SECOND_LONGEST_TIME,   // 2nd Longest Time
+    SHORTEST_TIME,         // Shortest Time
+    SMASH_CHAMP,           // Smash Champ
+    SMASH_SAP,             // Smash Sap
+    SLUG_MEISTER,          // Slug Meister
+    PUNCHING_BAG,          // Punching Bag
+    KO_KINGPIN,            // KO Kingpin
+    NO_DEFENSE_NELLY,      // No-Defense Nelly
+    DISASTER_MASTER,       // Disaster Master
+} mnCount_row;
+
+/* 2502CC */ u32 mnCount_GetMatchTime(s32);
+/* 2502F0 */ u32 mnCount_GetKOKingpin(s32);
+/* 250314 */ u32 mnCount_GetNoDefenseNelly(s32);
+/* 250338 */ u32 mnCount_GetDisasterMaster(s32);
+/* 25035C */ int mnCount_8025035C(s32 skip_count, u32 (*get_val_func)(s32));
+/* 25069C */ u32 mnCount_GetSmashChamp(s32);
+/* 2506C0 */ u32 mnCount_GetSmashSap(s32);
+/* 2506E4 */ u32 mnCount_GetSlugMeister(s32);
+/* 250708 */ u32 mnCount_GetPunchingBag(s32);
 /* 25072C */ s32 mnCount_8025072C(CountEntry* entries, s32 start_idx,
-                                  s32 mode);
-/* 25092C */ UNK_RET mnCount_8025092C(UNK_PARAMS);
-/* 250C7C */ UNK_RET mnCount_80250C7C(UNK_PARAMS);
-/* 250DE4 */ UNK_RET mnCount_80250DE4(UNK_PARAMS);
-/* 250FF8 */ UNK_RET mnCount_80250FF8(UNK_PARAMS);
-/* 251278 */ UNK_RET fn_80251278(UNK_PARAMS);
-/* 2513F4 */ void mnCount_802513F4(HSD_GObj*);
+                                  bool mode);
+/* 25092C */ s32 mnCount_8025092C(s32 rank, u32 (*getVal)(s32), bool mode);
+/* 250C7C */ int mnCount_GetRowValue_Character(mnCount_row row);
+/* 250DE4 */ unsigned int mnCount_GetRowValue_Number(int);
+/* 250FF8 */ void mnCount_CreateRow(HSD_GObj*, int, mnCount_row);
+/* 251278 */ void mnCount_HandleUserInput(HSD_GObj* gobj);
+/* 2513F4 */ void mnCount_UpdateArrowIndicators(HSD_GObj* gobj);
 /* 2514B8 */ void fn_802514B8(HSD_GObj* gobj);
-/* 2514D8 */ UNK_RET fn_802514D8(UNK_PARAMS);
-/* 251640 */ UNK_RET fn_80251640(UNK_PARAMS);
-/* 2517E0 */ void mnCount_802517E0(MnCountData*);
-/* 25186C */ void mnCount_8025186C(void);
+/* 2514D8 */ void fn_802514D8(HSD_GObj* gobj);
+/* 251640 */ void fn_80251640(HSD_GObj* gobj);
+/* 2517E0 */ void mnCount_InitUserData(MnCountData* arg0);
+/* 25186C */ void mnCount_Create(void);
 
 #endif
diff --git a/src/melee/mn/mnmain.c b/src/melee/mn/mnmain.c
index fd249d9e1a..18a25448f6 100644
--- a/src/melee/mn/mnmain.c
+++ b/src/melee/mn/mnmain.c
@@ -2045,7 +2045,7 @@ void mn_8022CA54(HSD_GObj* gp)
             return;
         case SEL_RECORDS_MISC:
             lbAudioAx_80024030(1);
-            mnCount_8025186C();
+            mnCount_Create();
             HSD_GObjPLink_80390228(gp);
             return;
         }
diff --git a/src/melee/ty/tydisplay.c b/src/melee/ty/tydisplay.c
index 59790a46aa..2cae85259a 100644
--- a/src/melee/ty/tydisplay.c
+++ b/src/melee/ty/tydisplay.c
@@ -1527,7 +1527,7 @@ void un_8031B460_OnEnter(void* arg0)
     memzero(un_804D6F18, 0x80);
     memzero(un_804D6ED4, 0xE4);
 
-    cfg->x08 = un_80304870();
+    cfg->x08 = un_GetTrophyTotal();
 
     if ((s32) un_804D6F20 != 0) {
         cfg->x08 = 1;
diff --git a/src/melee/ty/tylist.c b/src/melee/ty/tylist.c
index 6340f13cde..8b2efa93b7 100644
--- a/src/melee/ty/tylist.c
+++ b/src/melee/ty/tylist.c
@@ -251,7 +251,7 @@ void un_80312E88(TyListArg* arg, float delta)
     }
 skip_dirty:
 
-    if (arg->idx == un_80304870() - 1) {
+    if (arg->idx == un_GetTrophyTotal() - 1) {
         float pos;
         jobj = ((TyListState*) data)->jobj;
         pos = arg->x30;
@@ -507,7 +507,7 @@ void un_803147C4(void)
     }
 
     un_80307470(0);
-    if (un_80304870() != 0) {
+    if (un_GetTrophyTotal() != 0) {
         memzero(state, 0x2AC);
         un_80313774();
     }
@@ -521,7 +521,7 @@ void un_803148E4(s32 arg0)
     TyArchiveData* archive = un_804D6ED8;
     PAD_STACK(8);
 
-    if (un_80304870() != 0) {
+    if (un_GetTrophyTotal() != 0) {
         if (arg0 != 0) {
             un_804A284C[0x12A] = state->selectedIdx;
             un_804A284C[0x12B] = un_804D6EDC[state->selectedIdx];
@@ -529,7 +529,7 @@ void un_803148E4(s32 arg0)
             mode->x2 = state->x29C;
             mode->x3 = state->x2B8;
         } else {
-            if (un_80304870() != 0) {
+            if (un_GetTrophyTotal() != 0) {
                 s16 val = un_804D6EDC[state->selectedIdx];
                 un_803067BC(0, 0);
                 state->selectedIdx = un_803062BC(val);
