diff --git a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
index 1d7c5b4339..32724c4406 100644
--- a/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
+++ b/src/melee/ft/chara/ftCommon/ftCo_Attack100.c
@@ -1276,7 +1276,6 @@ void fn_800DB8A4(Fighter_GObj* gobj)
 
 void ftCo_CaptureWaitHi_Anim(Fighter_GObj* gobj)
 {
-    u8 _[8];
     Fighter* fp;
     FighterOverlay* fp_ovl;
     f32 dec;
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.c b/src/melee/ft/chara/ftKirby/ftKb_Init.c
index fbf2cfddff..fa8dd9cc1e 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.c
@@ -3972,7 +3972,7 @@ void ftKb_SpecialN_800F190C(Fighter_GObj* gobj, FighterKind kind)
     fp->fv.kb.ns_flash_gobj = NULL;
     fp->fv.kb.x80 = 0;
     fp->fv.kb.x84 = da->specialn_kp_max_fuel;
-    fp->fv.kb.x88 = da->specialn_kp_lowest_charge_graphic_size;
+    fp->fv.kb.x88 = da->specialn_kp_flame_scale;
     fp->fv.kb.x8C = 1.0F;
     fp->fv.kb.x90 = 1.0F;
     fp->fv.kb.x94 = 1.0F;
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index 4c1ff17c24..f9a3f1040f 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -713,7 +713,7 @@ void ftKb_SpecialHi_800F346C(Fighter_GObj* gobj)
         }
         break;
     case 2:
-        if (!fp->mv.kb.specialhi.x8.i && !fp->mv.kb.specialhi.xC) {
+        if (!fp->mv.kb.specialhi.x8.i && fp->mv.kb.specialhi.xC == 1) {
             fp->mv.kb.specialhi.x8.i = fp->cmd_vars[0];
             fp->mv.kb.specialhi.x14 = 0x16;
         }
@@ -1126,7 +1126,7 @@ void ftKb_SpecialLw_IASA(Fighter_GObj* gobj)
         Fighter_ChangeMotionState(gobj, ftKb_MS_SpecialLwEnd, 0x04440012U,
                                   0.0f, 1.0f, 0.0f, NULL);
         fp->x221C_b4 = temp_r28;
-        ftKirbyDmgInline(gobj);
+        ftKirbyDmgInline2(gobj);
         efSync_Spawn(0x49C, gobj, temp_r30);
     }
 }
@@ -1263,7 +1263,6 @@ void ftKb_SpecialAirS_Coll(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
-    PAD_STACK(8);
     if (ft_80081D0C(gobj)) {
         Fighter* fp2 = GET_FIGHTER(gobj);
         if (fp2->fv.kb.hat.x0 != NULL) {
@@ -1479,7 +1478,7 @@ void ftKb_SpecialN_800F5BA4(Fighter* fp)
 {
     ftKb_DatAttrs* da = fp->dat_attrs;
     if ((s32) fp->kind == FTKIND_KIRBY &&
-        (s32) fp->fv.kb.hat.kind != FTKIND_KIRBY && !fp->fv.kb.xF4_b0 &&
+        (s32) fp->fv.kb.hat.kind != FTKIND_KIRBY && !fp->fv.kb.hat.x8_b0 &&
         (u32) fp->victim_gobj == 0U && (u32) fp->dmg.x1860_element != 0xAU &&
         HSD_Randi((s32) da->specialn_odds_lose_ability_on_hit) == 0)
     {
@@ -1491,7 +1490,7 @@ void ftKb_SpecialN_800F5C34(Fighter* fp)
 {
     ftKb_DatAttrs* da = fp->dat_attrs;
     if ((s32) fp->kind == FTKIND_KIRBY &&
-        (s32) fp->fv.kb.hat.kind != FTKIND_KIRBY && !fp->fv.kb.xF4_b0 &&
+        (s32) fp->fv.kb.hat.kind != FTKIND_KIRBY && !fp->fv.kb.hat.x8_b0 &&
         (u32) fp->victim_gobj == 0U &&
         HSD_Randi((s32) da->specialn_odds_lose_ability_on_hit) == 0)
     {
@@ -1512,12 +1511,12 @@ void ftKb_SpecialN_800F5C34(Fighter* fp)
 
 void ftKb_SpecialN_800F5D04(Fighter_GObj* gobj, bool arg1)
 {
-    Vec3 pos;
     Vec3 vel;
-    int pad;
+    Vec3 pos;
     Fighter* new_var;
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da;
+    PAD_STACK(8);
     ftKb_SpecialN_800F19AC(gobj);
     ftKb_SpecialN_800F190C(gobj, fp->fv.kb.hat.kind);
     ftKb_SpecialN_800EEEC4(gobj, fp->fv.kb.hat.kind);
@@ -1545,7 +1544,7 @@ void ftKb_SpecialN_800F5F68(HSD_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
-    PAD_STACK(4 * 14);
+    PAD_STACK(4 * 12);
 
     fp->throw_flags = 0;
     fp->cmd_vars[0] = fp->cmd_vars[1] = 0;
@@ -1573,7 +1572,7 @@ void ftKb_SpecialN_800F6070(HSD_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
-    PAD_STACK(4 * 14);
+    PAD_STACK(4 * 12);
 
     fp->throw_flags = 0;
     fp->cmd_vars[0] = fp->cmd_vars[1] = 0;
@@ -2027,12 +2026,11 @@ void ftKb_SpecialAirNCaptured_Anim(Fighter_GObj* gobj)
 void ftKb_EatTurn_Anim(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
-    PAD_STACK(8);
     if (!ftAnim_IsFramesRemaining(gobj)) {
         fp->facing_dir = -fp->facing_dir;
         fp = GET_FIGHTER(gobj);
         Fighter_ChangeMotionState(gobj, ftKb_MS_EatWait, Ft_MF_SkipModel, 0.0f,
-                                  0.0f, 0.0f, NULL);
+                                  1.0f, 0.0f, NULL);
         ftAnim_8006EBA4(gobj);
         ftKb_SpecialN_800F9070(gobj);
         ftCommon_8007E2F4(fp, 0x1FF);
@@ -2475,7 +2473,6 @@ void ftKb_SpecialNMr_800F93CC(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     s32 msid = ftKb_MS_MrSpecialAirN;
-    PAD_STACK(8);
     fp->cmd_vars[0] = 0;
     fp->throw_flags = 0;
     if (GET_FIGHTER(gobj)->fv.kb.hat.kind != FTKIND_MARIO) {
@@ -2611,6 +2608,7 @@ void ftKb_SpecialNCa_800F99BC(Fighter_GObj* gobj)
 {
     ftKirby_MotionState msid = ftKb_MS_CaSpecialN;
     Fighter* fp = GET_FIGHTER(gobj);
+    PAD_STACK(8);
     fp->cmd_vars[1] = 0;
     fp->cmd_vars[0] = 0;
     fp->throw_flags = 0;
@@ -2628,6 +2626,7 @@ void ftKb_SpecialNCa_800F9A54(Fighter_GObj* gobj)
 {
     ftKirby_MotionState msid = ftKb_MS_CaSpecialAirN;
     Fighter* fp = GET_FIGHTER(gobj);
+    PAD_STACK(8);
     fp->cmd_vars[1] = 0;
     fp->cmd_vars[0] = 0;
     fp->throw_flags = 0;
@@ -2695,6 +2694,7 @@ void ftKb_CaSpecialAirN_Phys(Fighter_GObj* gobj)
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
     bool throw_b1;
+    PAD_STACK(16);
     if (fp->throw_flags_b1) {
         fp->throw_flags_b1 = false;
         throw_b1 = true;
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
index 08da84857e..3cf973b1e4 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
@@ -200,7 +200,6 @@ void fn_800FE0E0(Fighter_GObj* gobj)
 
 void ftKb_FxSpecialNEnd_Anim(Fighter_GObj* gobj)
 {
-    volatile short pad2;
     Fighter* fp = GET_FIGHTER(gobj);
 
     it_802ADDD0((Item_GObj*) fp->fv.kb.xB0, fp->cmd_vars[1]);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
index 57ceb43f44..fa99578df6 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNNs.c
@@ -353,7 +353,6 @@ void ftKb_NsSpecialAirNEnd_Coll(Fighter_GObj* gobj)
 static inline void ftKb_DkSpecialNStart_Coll_inline(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
-    fp = GET_FIGHTER(gobj);
     fp->death2_cb = (void (*)(HSD_GObj*)) ftKb_Init_800EE74C;
     fp->take_dmg_cb = (void (*)(HSD_GObj*)) ftKb_Init_800EE7B8;
     fp->pre_hitlag_cb = efLib_PauseAll;
@@ -380,8 +379,11 @@ void ftKb_DkSpecialNLoop_Anim(Fighter_GObj* gobj)
             Fighter* fighter2 = fp;
             ftCo_800BFFD0(fighter2, 58, 0);
             fp->fv.kb.xBC = da->specialn_dk_swings_to_full_charge;
-            fp->pre_hitlag_cb = NULL;
-            fp->post_hitlag_cb = NULL;
+            {
+                Fighter* fp2 = gobj->user_data;
+                fp2->pre_hitlag_cb = NULL;
+                fp2->post_hitlag_cb = NULL;
+            }
             ft_8008A2BC(gobj);
         }
     }
@@ -401,6 +403,7 @@ void ftKb_DkSpecialN_Anim(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
+    PAD_STACK(24);
     if (fp->cmd_vars[0]) {
         if (fp->mv.dk.specialn.x8 == 0) {
             fp->mv.dk.specialn.x8++;
@@ -450,8 +453,11 @@ void ftKb_DkSpecialN_Anim(Fighter_GObj* gobj)
     }
     if (!ftAnim_IsFramesRemaining(gobj)) {
         fp->fv.kb.xBC = 0;
-        fp->pre_hitlag_cb = NULL;
-        fp->post_hitlag_cb = NULL;
+        {
+            Fighter* fp2 = gobj->user_data;
+            fp2->pre_hitlag_cb = NULL;
+            fp2->post_hitlag_cb = NULL;
+        }
         ft_8008A2BC(gobj);
     }
 }
@@ -459,7 +465,7 @@ void ftKb_DkSpecialN_Anim(Fighter_GObj* gobj)
 void ftKb_DkSpecialNFull_Anim(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
-    u8 _[8];
+    u8 _[32];
     if (fp->cmd_vars[0]) {
         if (fp->mv.dk.specialn.x8 == 0) {
             fp->mv.dk.specialn.x8++;
@@ -493,8 +499,11 @@ void ftKb_DkSpecialNFull_Anim(Fighter_GObj* gobj)
     }
     if (!ftAnim_IsFramesRemaining(gobj)) {
         fp->fv.kb.xBC = 0;
-        fp->pre_hitlag_cb = NULL;
-        fp->post_hitlag_cb = NULL;
+        {
+            Fighter* fp2 = gobj->user_data;
+            fp2->pre_hitlag_cb = NULL;
+            fp2->post_hitlag_cb = NULL;
+        }
         ft_8008A2BC(gobj);
     }
 }
@@ -520,8 +529,11 @@ void ftKb_DkSpecialAirNLoop_Anim(Fighter_GObj* gobj)
             Fighter* fighter2 = fp;
             ftCo_800BFFD0(fighter2, 58, 0);
             fp->fv.kb.xBC = new_var->specialn_dk_swings_to_full_charge;
-            fp->pre_hitlag_cb = NULL;
-            fp->post_hitlag_cb = NULL;
+            {
+                Fighter* fp2 = gobj->user_data;
+                fp2->pre_hitlag_cb = NULL;
+                fp2->post_hitlag_cb = NULL;
+            }
             ftCo_Fall_Enter(gobj);
         }
     }
@@ -542,6 +554,7 @@ void ftKb_DkSpecialAirN_Anim(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
+    PAD_STACK(16);
     if (fp->cmd_vars[0]) {
         if (fp->mv.dk.specialn.x8 == 0) {
             fp->mv.dk.specialn.x8++;
@@ -577,8 +590,11 @@ void ftKb_DkSpecialAirN_Anim(Fighter_GObj* gobj)
     }
     if (!ftAnim_IsFramesRemaining(gobj)) {
         fp->fv.kb.xBC = 0;
-        fp->pre_hitlag_cb = NULL;
-        fp->post_hitlag_cb = NULL;
+        {
+            Fighter* fp2 = gobj->user_data;
+            fp2->pre_hitlag_cb = NULL;
+            fp2->post_hitlag_cb = NULL;
+        }
         if (da->specialn_dk_freefall_toggle == 0.0F) {
             ftCo_Fall_Enter(gobj);
         } else {
@@ -592,6 +608,7 @@ void ftKb_DkSpecialAirNFull_Anim(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
+    PAD_STACK(16);
     if (fp->cmd_vars[0]) {
         if (fp->mv.dk.specialn.x8 == 0) {
             fp->mv.dk.specialn.x8++;
@@ -609,8 +626,11 @@ void ftKb_DkSpecialAirNFull_Anim(Fighter_GObj* gobj)
     }
     if (!ftAnim_IsFramesRemaining(gobj)) {
         fp->fv.kb.xBC = 0;
-        fp->pre_hitlag_cb = NULL;
-        fp->post_hitlag_cb = NULL;
+        {
+            Fighter* fp2 = gobj->user_data;
+            fp2->pre_hitlag_cb = NULL;
+            fp2->post_hitlag_cb = NULL;
+        }
         if (da->specialn_dk_freefall_toggle == 0.0F) {
             ftCo_Fall_Enter(gobj);
         } else {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
index daa264e0bd..423696a72b 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNPk.c
@@ -282,6 +282,7 @@ void ftKb_SpecialNKp_800FA8B4(Fighter_GObj* gobj)
 {
     Fighter* fp;
     s32 msid = ftKb_MS_KpSpecialNStart;
+    PAD_STACK(8);
     switch ((s32) GET_FIGHTER(gobj)->fv.kb.hat.kind) {
     case FTKIND_KOOPA:
         break;
@@ -432,6 +433,7 @@ void ftKb_KpSpecialN_IASA(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
+    PAD_STACK(24);
     if ((s32) fp->mv.kb.specialn_kp.x0[3] >=
         (s32) da->specialn_kp_b_button_check_frequency)
     {
@@ -485,6 +487,7 @@ void ftKb_KpSpecialAirN_IASA(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
+    PAD_STACK(24);
     if ((s32) fp->mv.kb.specialn_kp.x0[3] >=
         (s32) da->specialn_kp_b_button_check_frequency)
     {
@@ -986,6 +989,7 @@ void ftKb_SpecialNSs_800FD020(Fighter_GObj* gobj)
 
 void ftKb_SsSpecialNCancel_Anim(Fighter_GObj* gobj)
 {
+    PAD_STACK(16);
     if (gobj != NULL) {
         Fighter* fp = GET_FIGHTER(gobj);
         if (fp->fv.kb.xA4 != 0) {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
index 1b03a69904..cbb5a3a1c2 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
@@ -294,7 +294,6 @@ static void fn_801097B8(HSD_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     int new_var;
-    long long pad;
     int new_var2;
     new_var2 = ftKb_MS_YsSpecialAirNCapture1_0;
     new_var = 0;
@@ -868,7 +867,7 @@ void ftKb_MsSpecialAirNLoop_Anim(Fighter_GObj* gobj)
     }
     fp->mv.kb.specialn_ms.cur_frame++;
     new_var = 1;
-    if (fp->mv.kb.specialn_ms.cur_frame > (*charge * 30)) {
+    if (fp->mv.kb.specialn_ms.cur_frame > (s32)(*charge * 30)) {
         fp->cmd_vars[0] = new_var;
         ftAnim_IsFramesRemaining(gobj);
     }
@@ -920,13 +919,11 @@ void ftKb_SpecialNMs_8010BB50(Fighter_GObj* gobj)
 {
     FtMotionId msid;
     Fighter* fp = GET_FIGHTER(gobj);
-    unsigned long long new_var;
     ftCommon_8007D5D4(fp);
     if ((s32) GET_FIGHTER(gobj)->fv.kb.hat.kind == FTKIND_MARS) {
         msid = ftKb_MS_MsSpecialAirNLoop;
     } else {
-        new_var = 0x217;
-        msid = new_var;
+        msid = ftKb_MS_FeSpecialAirNLoop;
     }
     Fighter_ChangeMotionState(gobj, msid, 0x0C4C7A86, fp->cur_anim_frame, 1.0f,
                               0.0f, NULL);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index b888635e27..c5aec20d76 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -292,14 +292,12 @@ s32 ftKb_SpecialNSk_80106020(Fighter_GObj* gobj)
 void ftKb_SpecialNSk_8010612C(Fighter_GObj* gobj)
 {
     int new_var;
-    int new_var2;
     Fighter* fp = GET_FIGHTER(gobj);
     Fighter_ChangeMotionState(gobj, ftKb_MS_SkSpecialNStart, 0, 0.0F, 1.0F,
                               0.0F, NULL);
     fp->throw_flags_b0 = false;
-    new_var2 = 3;
     new_var = 0;
-    fp->cmd_vars[new_var2] = new_var;
+    fp->cmd_vars[3] = new_var;
     fp->cmd_vars[2] = new_var;
     fp->cmd_vars[1] = new_var;
     fp->cmd_vars[new_var] = new_var;
@@ -653,7 +651,6 @@ void ftKb_SpecialNMt_80107040(Fighter_GObj* gobj)
 void ftKb_SpecialNMt_8010709C(Fighter_GObj* gobj)
 {
     Vec3 unused;
-    s32 unused2;
     Fighter* fp;
     Fighter* fp2;
     if (gobj == NULL) {
@@ -763,6 +760,7 @@ void ftKb_MtSpecialNLoopFull_Anim(Fighter_GObj* gobj)
 
 void ftKb_MtSpecialNCancel_Anim(Fighter_GObj* gobj)
 {
+    PAD_STACK(16);
     if (gobj != NULL) {
         Fighter* fp = GET_FIGHTER(gobj);
         if (fp->fv.kb.x98 != 0) {
@@ -1134,8 +1132,6 @@ void ftKb_PpSpecialAirN_Phys(Fighter_GObj* gobj)
 
 void ftKb_PpSpecialN_Coll(Fighter_GObj* gobj)
 {
-    void* new_var;
-    void* new_var2;
     if (!ft_800827A0(gobj)) {
         Fighter* fp = GET_FIGHTER(gobj);
         if (fp->fv.kb.xC0 != NULL) {
@@ -1143,9 +1139,7 @@ void ftKb_PpSpecialN_Coll(Fighter_GObj* gobj)
             {
                 Fighter* fp2 = GET_FIGHTER(gobj);
                 if (fp->fv.kb.xC0 == fp2->fv.kb.xC0) {
-                    new_var2 = NULL;
-                    new_var = new_var2;
-                    fp2->fv.kb.xC0 = new_var;
+                    fp2->fv.kb.xC0 = NULL;
                 }
             }
         }
diff --git a/src/melee/ft/ftchangeparam.c b/src/melee/ft/ftchangeparam.c
index b51e2f20f3..475774a844 100644
--- a/src/melee/ft/ftchangeparam.c
+++ b/src/melee/ft/ftchangeparam.c
@@ -188,8 +188,11 @@ void ftCo_800D105C(Fighter_GObj* fgp)
         fp->x294_itPickup.air_light_offset.y *= temp_f1;
         fp->x294_itPickup.air_light_offset.z *= temp_f1;
         fp->x294_itPickup.air_light_offset.w *= temp_f1;
-        fp->x2C4.x *= temp_f1;
-        fp->x2C4.y *= temp_f1;
+        {
+            f32 scale_y = fp->x34_scale.y;
+            fp->x2C4.x *= scale_y;
+            fp->x2C4.y *= scale_y;
+        }
     }
 
     /// if bunnyhood != NULL
diff --git a/src/melee/gm/gm_18A5.c b/src/melee/gm/gm_18A5.c
index 1e98057e6f..1010b63f5b 100644
--- a/src/melee/gm/gm_18A5.c
+++ b/src/melee/gm/gm_18A5.c
@@ -2609,22 +2609,22 @@ void fn_80198C60(void)
 {
     TmData* td;
     HSD_Text* text;
-    HSD_Text** arr;
+
+    PAD_STACK(24);
 
     td = gm_8018F634();
-    arr = (HSD_Text**) ((u8*) td + 0x524);
-    arr[2] = HSD_SisLib_803A6754(0, (s32) lbl_804D663C);
-    text = arr[2];
+    td->x524[2] = HSD_SisLib_803A6754(0, (s32) lbl_804D663C);
+    text = td->x524[2];
     text->font_size.x = 0.054945F;
     text->font_size.y = 0.08F;
-    arr[2]->default_alignment = 1;
-    arr[2]->default_kerning = 1;
-
-    arr[3] = HSD_SisLib_803A6754(0, (s32) lbl_804D663C);
-    arr[3]->default_alignment = 1;
-    arr[3]->default_kerning = 1;
-    HSD_SisLib_803A6B98(arr[3], 320.0F, 250.0F, "    ");
-    HSD_SisLib_803A7548(arr[3], 0, 1.5F, 1.5F);
+    td->x524[2]->default_alignment = 1;
+    td->x524[2]->default_kerning = 1;
+
+    td->x524[3] = HSD_SisLib_803A6754(0, (s32) lbl_804D663C);
+    td->x524[3]->default_alignment = 1;
+    td->x524[3]->default_kerning = 1;
+    HSD_SisLib_803A6B98(td->x524[3], 320.0F, 250.0F, "    ");
+    HSD_SisLib_803A7548(td->x524[3], 0, 1.5F, 1.5F);
 }
 
 extern SceneDesc* lbl_804D666C;
diff --git a/src/melee/gm/gm_1BA8.c b/src/melee/gm/gm_1BA8.c
index d4346bea20..325b4d0346 100644
--- a/src/melee/gm/gm_1BA8.c
+++ b/src/melee/gm/gm_1BA8.c
@@ -1472,7 +1472,7 @@ void gm_801BEF84(MinorScene* arg)
 
 #pragma push
 #pragma dont_inline on
-void gm_801BEFA4(s8 ckind)
+void gm_801BEFA4(int ckind)
 {
     gm_8049C178[0] = ckind;
 }
diff --git a/src/melee/gm/gm_1BA8.h b/src/melee/gm/gm_1BA8.h
index ab73ddd859..fc7c9824eb 100644
--- a/src/melee/gm/gm_1BA8.h
+++ b/src/melee/gm/gm_1BA8.h
@@ -74,7 +74,7 @@ struct EventData;
 /* 1BEE58 */ void gm_801BEE58(MinorScene*);
 /* 1BEE9C */ void gm_801BEE9C(MinorScene*);
 /* 1BEF84 */ void gm_801BEF84(MinorScene*);
-/* 1BEFA4 */ void gm_801BEFA4(s8 ckind);
+/* 1BEFA4 */ void gm_801BEFA4(int ckind);
 /* 1BEFB0 */ CharacterKind gm_801BEFB0(void);
 /* 1BEFC0 */ void gm_801BEFC0(int);
 /* 1BEFD0 */ int gm_801BEFD0(void);
diff --git a/src/melee/gm/types.h b/src/melee/gm/types.h
index c42224a53e..35019e0e38 100644
--- a/src/melee/gm/types.h
+++ b/src/melee/gm/types.h
@@ -806,7 +806,7 @@ struct TmData {
     HSD_Text* x4E8[6];
     HSD_Text* x500[6];
     HSD_Text* x518[3];
-    u8 pad_x524[0x534 - 0x524];
+    HSD_Text* x524[4];
     HSD_Text* x534[3];
     u8 pad_x540[0x56B - 0x540];
 };
diff --git a/src/melee/gr/grfourside.c b/src/melee/gr/grfourside.c
index 241805c8cf..ab90abc89c 100644
--- a/src/melee/gr/grfourside.c
+++ b/src/melee/gr/grfourside.c
@@ -181,6 +181,7 @@ void grFourside_801F2F34(Ground_GObj* gobj)
 {
     HSD_GObj* pHVar1;
     Ground* gp = GET_GROUND(gobj);
+    PAD_STACK(8);
     Ground_801C2ED0(gobj->hsd_obj, gp->map_id);
     grAnime_801C8138(gobj, gp->map_id, 0);
     Ground_801C4E70(Ground_801C3FA4(gobj, 7), Ground_801C3FA4(gobj, 4),
@@ -256,6 +257,7 @@ void grFourside_801F3154(Ground_GObj* gobj)
     Ground* gp = GET_GROUND(gobj);
     HSD_JObj* jobj = gobj->hsd_obj;
     HSD_JObj* crane_iron = Ground_801C3FA4(gobj, 4);
+    PAD_STACK(8);
     Ground_801C2ED0(jobj, gp->map_id);
     grAnime_801C7FF8(gobj, 0, 7, 0, 0.0f, 0.0f);
     gp->gv.foursideCrane.x1.b0 = 0;
diff --git a/src/melee/gr/gricemt.c b/src/melee/gr/gricemt.c
index 192069acbd..97eb7aa003 100644
--- a/src/melee/gr/gricemt.c
+++ b/src/melee/gr/gricemt.c
@@ -445,6 +445,7 @@ void grIceMt_801F75FC(Ground_GObj* arg0)
     u32 iVar1;
     s16 val;
     Ground* gp = GET_GROUND(arg0);
+    PAD_STACK(8);
     memzero(&gp->gv.icemt.xDC, 0x18);
     memzero(&gp->gv.icemt.xF4, 0x14);
     do {
@@ -529,7 +530,7 @@ void grIceMt_801F785C(Ground_GObj* gobj)
     gp->gv.icemt.xD0 = -1;
     gp->gv.icemt.xD4 = grIm_804DB574;
 
-    ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b7 = 1;
+    ((UnkFlagStruct*) &gp->gv.icemt.xD8)->b0 = 1;
 
     Ground_801C10B8(gobj, fn_801F77B0);
 }
@@ -811,10 +812,7 @@ void grIceMt_801F87FC(Ground_GObj* gobj)
 {
     u32 unused[2];
     int i;
-    Ground* gp;
-
-    gp = gobj->user_data;
-
+    Ground* gp = gobj->user_data;
     for (i = 0; i < 5; i++) {
         if (gp->gv.icemt.xF8[i] != NULL) {
             grMaterial_801C8CDC(gp->gv.icemt.xF8[i]);
diff --git a/src/melee/gr/grkongo.c b/src/melee/gr/grkongo.c
index a36e56e3af..4dc53dfd55 100644
--- a/src/melee/gr/grkongo.c
+++ b/src/melee/gr/grkongo.c
@@ -1245,12 +1245,13 @@ f32 grKongo_801D8314(void)
                                     temp_r3_7 - grKg_804D6980->unk50;
                                 if (temp_r3_8 < 0) {
                                     var_f31 = -2.3561945f;
-                                } else if ((temp_r3_8 - grKg_804D6980->unk52) <
-                                           0)
-                                {
-                                    var_f31 = -3.1415927f;
                                 } else {
-                                    __assert("grkongo.c", 1753, "0");
+                                    temp_r3_8 -= grKg_804D6980->unk52;
+                                    if (temp_r3_8 < 0) {
+                                        var_f31 = -3.1415927f;
+                                    } else {
+                                        __assert("grkongo.c", 1753, "0");
+                                    }
                                 }
                             }
                         }
diff --git a/src/melee/gr/grvenom.c b/src/melee/gr/grvenom.c
index 2ed2e668fb..bc740c83b5 100644
--- a/src/melee/gr/grvenom.c
+++ b/src/melee/gr/grvenom.c
@@ -280,6 +280,7 @@ void grVenom_80203FD4(Ground_GObj* gobj)
     HSD_JObj* jobj5;
     HSD_JObj* jobj6;
     Ground* gp = gobj->user_data;
+    PAD_STACK(8);
 
     grAnime_801C8138(gobj, gp->map_id, 0);
     gp->x11_flags.b012 = 2;
diff --git a/src/melee/it/items/it_2E5A.c b/src/melee/it/items/it_2E5A.c
index 8c2ba2f33b..c0b6ab9710 100644
--- a/src/melee/it/items/it_2E5A.c
+++ b/src/melee/it/items/it_2E5A.c
@@ -198,7 +198,7 @@ s32 it_802E5F8C(Item_GObj* item_gobj, Vec3* pos, s32 arg2, s32 arg3, f32 arg4,
     Vec3 vel;
 
     vel.x = vel.y = vel.z = 0.0f;
-    it_802E614C(NULL, item_gobj, &sp2C, pos, &vel);
+    it_802E614C(item_gobj, item_gobj, &sp2C, pos, &vel);
     angle = M_PI_2 + arg4;
     while (angle < 0.0f) {
         angle += M_TAU;
diff --git a/src/melee/ty/toy.c b/src/melee/ty/toy.c
index 45eb5171a3..a8d4aedc3f 100644
--- a/src/melee/ty/toy.c
+++ b/src/melee/ty/toy.c
@@ -834,7 +834,7 @@ s16 un_803062BC(s32 trophyId)
 
 void un_803062EC(s32 arg0, u32 arg1, f32 farg0)
 {
-    char sp14[80];
+    char sp14[72];
     TrophyData* td;
     s32 id;
 
diff --git a/src/melee/ty/tylist.c b/src/melee/ty/tylist.c
index f6c4f09f7b..f46b161b92 100644
--- a/src/melee/ty/tylist.c
+++ b/src/melee/ty/tylist.c
@@ -157,7 +157,7 @@ void un_803127D4(void)
 }
 
 /// Formats a number into a string buffer using digit glyphs from the font.
-void un_80312834(char* buf, u32 num)
+char* un_80312834(char* buf, u32 num)
 {
     u8* lookup = ((SisFontData*) HSD_SisLib_804D1124[0])->digits;
     u32 idx;
@@ -184,6 +184,7 @@ void un_80312834(char* buf, u32 num)
     *buf++ = lookup[idx];
     *buf++ = lookup[idx + 1];
     *buf = 0;
+    return buf;
 }
 
 /// #un_80312904
diff --git a/src/melee/ty/tylist.h b/src/melee/ty/tylist.h
index 97a7206a8a..48a6462ee1 100644
--- a/src/melee/ty/tylist.h
+++ b/src/melee/ty/tylist.h
@@ -12,7 +12,7 @@ struct TyListState;
 /* 3124BC */ void un_803124BC(void);
 /* 31263C */ void un_8031263C(void);
 /* 3127D4 */ void un_803127D4(void);
-/* 312834 */ void un_80312834(char* buf, u32 num);
+/* 312834 */ char* un_80312834(char* buf, u32 num);
 /* 312904 */ void un_80312904(void*, s8);
 /* 312E88 */ void un_80312E88(struct TyListArg* arg, float delta);
 /* 313358 */ void un_80313358(struct TyListState* state, s8 arg2, s8 arg3,
diff --git a/src/melee/ty/types.h b/src/melee/ty/types.h
index 283f086f08..47eeacc8e5 100644
--- a/src/melee/ty/types.h
+++ b/src/melee/ty/types.h
@@ -89,7 +89,7 @@ typedef struct TyListState {
     /* 0x29E */ u8 x29E;
     /* 0x29F */ u8 x29F;
     /* 0x2A0 */ u8 pad_2A0;
-    /* 0x2A1 */ u8 x2A1;
+    /* 0x2A1 */ s8 x2A1;
     /* 0x2A2 */ u8 pad_2A2[2];
     /* 0x2A4 */ float x2A4;
     /* 0x2A8 */ float x2A8;
diff --git a/src/sysdolphin/baselib/hsd_3AA7.c b/src/sysdolphin/baselib/hsd_3AA7.c
index 34115f4591..40bdeb9073 100644
--- a/src/sysdolphin/baselib/hsd_3AA7.c
+++ b/src/sysdolphin/baselib/hsd_3AA7.c
@@ -143,7 +143,7 @@ int hsd_803AC340(void* header)
         extra_size += 0x200;
     }
 
-    extra_size += base_size;
+    extra_size = base_size + extra_size;
     return extra_size + 0x40;
 }
 
diff --git a/src/sysdolphin/baselib/sislib.c b/src/sysdolphin/baselib/sislib.c
index befc2ed9b5..688e95b2cc 100644
--- a/src/sysdolphin/baselib/sislib.c
+++ b/src/sysdolphin/baselib/sislib.c
@@ -487,14 +487,12 @@ void HSD_SisLib_803A5FBC(void)
 /// compiler after dual global store
 void HSD_SisLib_803A6048(u32 size)
 {
-    sislib_UnkAllocData* alloc;
     int i;
 
     HSD_SisLib_804D7968 = size;
     HSD_SisLib_804D7974 = NULL;
-    alloc = HSD_MemAlloc(HSD_SisLib_804D7968);
-    HSD_SisLib_804D7970 = alloc;
-    HSD_SisLib_804D796C = alloc;
+    HSD_SisLib_804D796C = HSD_SisLib_804D7970 =
+        HSD_MemAlloc(HSD_SisLib_804D7968);
     HSD_SisLib_804D7970->data_0 = NULL;
     HSD_SisLib_804D7970->data_1 = (HSD_Text*) (HSD_SisLib_804D7970 + 1);
     HSD_SisLib_804D7970->size = HSD_SisLib_804D7968 - 0xC;
