diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 555cef7309..5810925865 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -33110,7 +33110,7 @@ it_804DCFA0 = .sdata2:0x804DCFA0; // type:object size:0x4 scope:global align:4 d
 it_804DCFA4 = .sdata2:0x804DCFA4; // type:object size:0x4 scope:global align:4 data:float
 it_804DCFA8 = .sdata2:0x804DCFA8; // type:object size:0x8 scope:global align:8 data:double
 it_804DCFB0 = .sdata2:0x804DCFB0; // type:object size:0x4 scope:global align:4 data:float
-it_804DCFB8 = .sdata2:0x804DCFB8; // type:object size:0x4 scope:global align:4 data:float
+@151 = .sdata2:0x804DCFB8; // type:object size:0x4 scope:local align:4 data:float
 it_804DCFC0 = .sdata2:0x804DCFC0; // type:object size:0x4 scope:global align:4 data:float
 it_804DCFC4 = .sdata2:0x804DCFC4; // type:object size:0x4 scope:global align:4 data:float
 @214 = .sdata2:0x804DCFC8; // type:object size:0x4 scope:local align:4 data:float
diff --git a/configure.py b/configure.py
index 5d21653449..a59f10ec7d 100755
--- a/configure.py
+++ b/configure.py
@@ -1175,7 +1175,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/itseakneedleheld.c"),
             Object(Matching, "melee/it/items/itseakvanish.c"),
             Object(NonMatching, "melee/it/items/itpikachuthunder.c"),
-            Object(NonMatching, "melee/it/items/itmariocape.c"),
+            Object(Matching, "melee/it/items/itmariocape.c"),
             Object(NonMatching, "melee/it/items/ityoshieggthrow.c"),
             Object(Matching, "melee/it/items/ityoshistar.c"),
             Object(NonMatching, "melee/it/items/itpikachutjoltground.c"),
diff --git a/src/melee/it/items/itmariocape.c b/src/melee/it/items/itmariocape.c
index c6adec0a8e..19fc7c0fb3 100644
--- a/src/melee/it/items/itmariocape.c
+++ b/src/melee/it/items/itmariocape.c
@@ -9,6 +9,11 @@
 #include "it/it_2725.h"
 #include "it/item.h"
 
+ItemStateTable it_803F70F8[] = {
+    { 0, itMariocape_UnkMotion1_Anim, NULL, NULL },
+    { 1, itMariocape_UnkMotion1_Anim, NULL, NULL },
+};
+
 Item_GObj* it_802B2560(Fighter_GObj* parent_gobj, float facing_dir, Vec3* pos,
                        Fighter_Part part, ItemKind kind)
 {
@@ -35,6 +40,14 @@ Item_GObj* it_802B2560(Fighter_GObj* parent_gobj, float facing_dir, Vec3* pos,
     return NULL;
 }
 
+void itMarioCape_Logic41_Destroyed(Item_GObj* gobj)
+{
+    Item* ip = GET_ITEM(gobj);
+    if (ip->owner != NULL) {
+        ftMr_SpecialS_Reset(ip->owner);
+    }
+}
+
 void it_802B2674(Item_GObj* gobj)
 {
     Item* ip = GET_ITEM(gobj);
@@ -123,11 +136,3 @@ void it_802B2870(Item_GObj* gobj, Item_GObj* ref_gobj)
 {
     it_8026B894(gobj, ref_gobj);
 }
-
-void itMarioCape_Logic41_Destroyed(Item_GObj* gobj)
-{
-    Item* ip = GET_ITEM(gobj);
-    if (ip->owner != NULL) {
-        ftMr_SpecialS_Reset(ip->owner);
-    }
-}
