diff --git a/config/GALE01/splits.txt b/config/GALE01/splits.txt
index eccd25a240..7de165391d 100644
--- a/config/GALE01/splits.txt
+++ b/config/GALE01/splits.txt
@@ -1195,7 +1195,7 @@ melee/ft/chara/ftKirby/ftKb_SpecialNPk.c:
 	.text       start:0x800FA064 end:0x800FDC70
 	.sdata2     start:0x804D93E4 end:0x804D944C
 
-melee/ft/chara/ftKirby/ftKb_SpecialNFx.c:
+melee/ft/chara/ftKirby/ftkirbyspecialfox.c:
 	.text       start:0x800FDC70 end:0x800FEDD0
 	.sdata2     start:0x804D944C end:0x804D9474
 
diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 338b4d6d64..8ac3fbb527 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -4927,14 +4927,14 @@ ftKb_SsSpecialAirNStart_Coll = .text:0x800FDAF0; // type:function size:0x88 scop
 ftKb_SsSpecialAirN_Coll = .text:0x800FDB78; // type:function size:0x88 scope:global
 ftKb_SpecialNFx_800FDC00 = .text:0x800FDC00; // type:function size:0x70 scope:global
 ftKb_SpecialNFx_800FDC70 = .text:0x800FDC70; // type:function size:0x70 scope:global
-fn_800FDCE0 = .text:0x800FDCE0; // type:function size:0x34
+ftKb_SpecialNFx_OnChangeAction = .text:0x800FDCE0; // type:function size:0x34
 ftKb_SpecialNFx_800FDD14 = .text:0x800FDD14; // type:function size:0x38 scope:global
 ftKb_SpecialNFx_800FDD4C = .text:0x800FDD4C; // type:function size:0xA8 scope:global
 ftKb_SpecialNFx_800FDDF4 = .text:0x800FDDF4; // type:function size:0xC0 scope:global
 ftKb_SpecialNFx_800FDEB4 = .text:0x800FDEB4; // type:function size:0x2C scope:global
 ftKb_SpecialNFx_800FDEE0 = .text:0x800FDEE0; // type:function size:0x50 scope:global
 ftKb_SpecialNFx_800FDF30 = .text:0x800FDF30; // type:function size:0x1B0 scope:global
-fn_800FE0E0 = .text:0x800FE0E0; // type:function size:0x20
+ftKb_SpecialNFx_CreateBlasterShot = .text:0x800FE0E0; // type:function size:0x20
 ftKb_SpecialNFx_800FE100 = .text:0x800FE100; // type:function size:0x140 scope:global
 ftKb_SpecialNFx_800FE240 = .text:0x800FE240; // type:function size:0x120 scope:global
 ftKb_FxSpecialNStart_Anim = .text:0x800FE360; // type:function size:0xF0 scope:global
diff --git a/configure.py b/configure.py
index 8a20e45afe..a695d30cbf 100755
--- a/configure.py
+++ b/configure.py
@@ -729,7 +729,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/ft/chara/ftKirby/ftKb_Init.c"),
             Object(NonMatching, "melee/ft/chara/ftKirby/ftKb_SpecialN.c"),
             Object(NonMatching, "melee/ft/chara/ftKirby/ftKb_SpecialNPk.c"),
-            Object(NonMatching, "melee/ft/chara/ftKirby/ftKb_SpecialNFx.c"),
+            Object(NonMatching, "melee/ft/chara/ftKirby/ftkirbyspecialfox.c"),
             Object(NonMatching, "melee/ft/chara/ftKirby/ftKb_SpecialNNs.c"),
             Object(NonMatching, "melee/ft/chara/ftKirby/ftKb_SpecialNZd.c"),
             Object(NonMatching, "melee/ft/chara/ftKirby/ftKb_SpecialNYs.c"),
diff --git a/src/melee/ft/chara/ftKirby/ftKb_Init.h b/src/melee/ft/chara/ftKirby/ftKb_Init.h
index ffc15c00eb..13b8818ad2 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_Init.h
+++ b/src/melee/ft/chara/ftKirby/ftKb_Init.h
@@ -507,14 +507,14 @@
 /* 0FDB78 */ void ftKb_SsSpecialAirN_Coll(Fighter_GObj* gobj);
 /* 0FDC00 */ void ftKb_SpecialNFx_800FDC00(Fighter_GObj* gobj, Vec3* vec);
 /* 0FDC70 */ void ftKb_SpecialNFx_800FDC70(Fighter_GObj* gobj, Vec3* vec);
-/* 0FDCE0 */ void fn_800FDCE0(Fighter_GObj* gobj);
+/* 0FDCE0 */ void ftKb_SpecialNFx_OnChangeAction(Fighter_GObj* gobj);
 /* 0FDD14 */ bool ftKb_SpecialNFx_800FDD14(Fighter_GObj* gobj);
 /* 0FDD4C */ ftFx_SpecialNIndex ftKb_SpecialNFx_800FDD4C(Fighter_GObj* gobj);
 /* 0FDDF4 */ bool ftKb_SpecialNFx_800FDDF4(Fighter_GObj* gobj);
 /* 0FDEB4 */ void ftKb_SpecialNFx_800FDEB4(Fighter_GObj* gobj);
 /* 0FDEE0 */ void ftKb_SpecialNFx_800FDEE0(Fighter_GObj* gobj);
 /* 0FDF30 */ void ftKb_SpecialNFx_800FDF30(Fighter_GObj* gobj);
-/* 0FE0E0 */ void fn_800FE0E0(Fighter_GObj* gobj);
+/* 0FE0E0 */ void ftKb_SpecialNFx_CreateBlasterShot(Fighter_GObj* gobj);
 /* 0FE100 */ void ftKb_SpecialNFx_800FE100(Fighter_GObj* gobj);
 /* 0FE240 */ void ftKb_SpecialNFx_800FE240(Fighter_GObj* gobj);
 /* 0FE360 */ void ftKb_FxSpecialNStart_Anim(Fighter_GObj* gobj);
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
deleted file mode 100644
index a99ff8fd4c..0000000000
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNFx.c
+++ /dev/null
@@ -1,372 +0,0 @@
-#include "ftKb_Init.h"
-#include "types.h"
-
-#include <placeholder.h>
-
-#include "ft/chara/ftCommon/ftCo_FallSpecial.h"
-#include "ft/chara/ftCommon/ftpickupitem.h"
-#include "ft/fighter.h"
-
-#include "ft/forward.h"
-
-#include "ft/ft_081B.h"
-#include "ft/ft_0881.h"
-#include "ft/ft_0892.h"
-#include "ft/inlines.h"
-#include "ft/types.h"
-#include "ftCommon/ftCo_Fall.h"
-
-#include "ftKirby/forward.h"
-
-#include "it/items/itfoxblaster.h"
-#include "it/items/itnesspkflush.h"
-#include "melee/lb/lbrefract.h"
-
-#include <common_structs.h>
-#include <stddef.h>
-#include <baselib/gobj.h>
-#include <baselib/random.h>
-#include <MSL/math.h>
-
-/// #ftKb_SpecialNFx_800FDC70
-
-void fn_800FDCE0(Fighter_GObj* gobj)
-{
-    FORCE_PAD_STACK_8;
-
-    ft_800892A0(gobj);
-    ft_80089824(gobj);
-}
-
-bool ftKb_SpecialNFx_800FDD14(Fighter_GObj* gobj)
-{
-    if (gobj != NULL) {
-        Fighter* ft = GET_FIGHTER(gobj);
-        if (ft != NULL) {
-            if (!ft->fv.kb.xB0) {
-                return true;
-            }
-            return false;
-        }
-    }
-    return true;
-}
-
-/// @todo Use enum values once ftKb_MS enum is fixed (#ftCo_MS_Count mismatch)
-ftFx_SpecialNIndex ftKb_SpecialNFx_800FDD4C(Fighter_GObj* gobj)
-{
-    s32 msid = 9;
-
-    if (gobj != NULL) {
-        Fighter* fp = GET_FIGHTER(gobj);
-        if (fp != NULL) {
-            switch (fp->fv.kb.hat.kind) {
-            case FTKIND_FOX: {
-                s32 motion_id = fp->motion_id;
-                switch (motion_id) {
-                case 0x1A7: /* ftKb_MS_FxSpecialNStart */
-                case 0x1A8: /* ftKb_MS_FxSpecialNLoop */
-                case 0x1A9: /* ftKb_MS_FxSpecialNEnd */
-                case 0x1AA: /* ftKb_MS_FxSpecialAirNStart */
-                case 0x1AB: /* ftKb_MS_FxSpecialAirNLoop */
-                case 0x1AC: /* ftKb_MS_FxSpecialAirNEnd */
-                    msid = motion_id - 0x1A7;
-                    break;
-                case 0xDC: /* ftCo_MS_ThrowB */
-                case 0xDD: /* ftCo_MS_ThrowHi */
-                case 0xDE: /* ftCo_MS_ThrowLw */
-                    msid = motion_id - 0xD6;
-                    break;
-                }
-                break;
-            }
-            case FTKIND_FALCO: {
-                s32 motion_id = fp->motion_id;
-                switch (motion_id) {
-                case 0x208: /* ftKb_MS_FcSpecialNStart */
-                case 0x209: /* ftKb_MS_FcSpecialNLoop */
-                case 0x20A: /* ftKb_MS_FcSpecialNEnd */
-                case 0x20B: /* ftKb_MS_FcSpecialAirNStart */
-                case 0x20C: /* ftKb_MS_FcSpecialAirNLoop */
-                case 0x20D: /* ftKb_MS_FcSpecialAirNEnd */
-                    msid = motion_id - 0x208;
-                    break;
-                case 0xDC: /* ftCo_MS_ThrowB */
-                case 0xDD: /* ftCo_MS_ThrowHi */
-                case 0xDE: /* ftCo_MS_ThrowLw */
-                    msid = motion_id - 0xD6;
-                    break;
-                }
-                break;
-            }
-            }
-        }
-    }
-    return msid;
-}
-
-/// #ftKb_SpecialNFx_800FDDF4
-
-void ftKb_SpecialNFx_800FDEB4(Fighter_GObj* gobj)
-{
-    if (gobj != NULL) {
-        Fighter* ft = GET_FIGHTER(gobj);
-        if ((ft != NULL) && (ft->fv.kb.xB0 != 0)) {
-            ft->fv.kb.xB0 = 0;
-        }
-    }
-}
-
-void ftKb_SpecialNFx_800FDEE0(Fighter_GObj* gobj)
-{
-    PAD_STACK(8);
-    if (gobj != NULL) {
-        Fighter* fp = GET_FIGHTER(gobj);
-        if (fp != NULL) {
-            if (fp->fv.kb.xB0 != 0) {
-                it_802AEAB4((Item_GObj*) fp->fv.kb.xB0);
-                fp->fv.kb.xB0 = 0;
-            }
-        }
-    }
-}
-
-/// #ftKb_SpecialNFx_800FDF30
-
-void fn_800FE0E0(Fighter_GObj* gobj)
-{
-    ftKb_SpecialNFx_800FDF30(gobj);
-}
-
-/// #ftKb_SpecialNFx_800FE100
-
-/// #ftKb_SpecialNFx_800FE240
-
-/// #ftKb_FxSpecialNStart_Anim
-
-/// #ftKb_FxSpecialNLoop_Anim
-
-void ftKb_FxSpecialNEnd_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-
-    it_802ADDD0((Item_GObj*) fp->fv.kb.xB0, fp->cmd_vars[1]);
-
-    if (fp->cmd_vars[1] == 2) {
-        ftpickupitem_80094818(gobj, 0);
-    }
-
-    if (fp->cmd_vars[3] == 2 && fp->fv.kb.xB0 != 0) {
-        unsigned long long pad;
-        fp->cmd_vars[3] = 0;
-        it_802AE608((Item_GObj*) fp->fv.kb.xB0);
-    }
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftKb_SpecialNFx_800FDEB4(gobj);
-        ft_8008A2BC(gobj);
-    }
-}
-
-/// #ftKb_FxSpecialAirNStart_Anim
-
-/// #ftKb_FxSpecialAirNLoop_Anim
-
-void ftKb_FxSpecialAirNEnd_Anim(Fighter_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    Fighter* new_var;
-    ftKb_DatAttrs* da = fp->dat_attrs;
-    volatile int pad;
-    int new_var2;
-    float landing_lag;
-
-    it_802ADDD0((Item_GObj*) fp->fv.kb.xB0, fp->cmd_vars[1]);
-
-    new_var = fp;
-    if (new_var->cmd_vars[1] == 2) {
-        ftpickupitem_80094818(gobj, 0);
-    }
-
-    if (fp->cmd_vars[3] == 2 && fp->fv.kb.xB0 != 0) {
-        new_var2 = 0;
-        fp->cmd_vars[3] = new_var2;
-        it_802AE608((Item_GObj*) fp->fv.kb.xB0);
-    }
-    if (!ftAnim_IsFramesRemaining(gobj)) {
-        ftKb_SpecialNFx_800FDEB4(gobj);
-        switch (fp->fv.kb.hat.kind) {
-        case FTKIND_FOX:
-            do {
-            } while (0);
-            landing_lag = da->specialn_fx_freefall_toggle;
-            break;
-        case FTKIND_FALCO:
-            landing_lag = da->specialn_fc_freefall_toggle;
-            break;
-        default:
-            break;
-        }
-        if (landing_lag == 0.0F) {
-            ftCo_Fall_Enter(gobj);
-            return;
-        }
-        ftCo_80096900(gobj, 1, 0, 1, 1.0F, landing_lag);
-    }
-}
-
-void ftKb_FxSpecialNStart_IASA(Fighter_GObj* gobj)
-{
-    Fighter* ft = GET_FIGHTER(gobj);
-    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
-        ft->mv.kb.specialhi.x0 = 1;
-    }
-}
-
-void ftKb_FxSpecialNLoop_IASA(Fighter_GObj* gobj)
-{
-    Fighter* ft = GET_FIGHTER(gobj);
-    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
-        ft->mv.kb.specialhi.x0 = 1;
-    }
-}
-
-void ftKb_FxSpecialNEnd_IASA(Fighter_GObj* gobj) {}
-
-void ftKb_FxSpecialAirNStart_IASA(Fighter_GObj* gobj)
-{
-    Fighter* ft = GET_FIGHTER(gobj);
-    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
-        ft->mv.kb.specialhi.x0 = 1;
-    }
-}
-
-void ftKb_FxSpecialAirNLoop_IASA(Fighter_GObj* gobj)
-{
-    Fighter* ft = GET_FIGHTER(gobj);
-    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
-        ft->mv.kb.specialhi.x0 = 1;
-    }
-}
-
-void ftKb_FxSpecialAirNEnd_IASA(Fighter_GObj* gobj) {}
-
-void ftKb_FxSpecialNStart_Phys(Fighter_GObj* gobj)
-{
-    ft_80084F3C(gobj);
-}
-
-void ftKb_FxSpecialNLoop_Phys(Fighter_GObj* gobj)
-{
-    ft_80084F3C(gobj);
-}
-
-void ftKb_FxSpecialNEnd_Phys(Fighter_GObj* gobj)
-{
-    ft_80084F3C(gobj);
-}
-
-void ftKb_FxSpecialAirNStart_Phys(Fighter_GObj* gobj)
-{
-    ft_80084DB0(gobj);
-}
-
-void ftKb_FxSpecialAirNLoop_Phys(Fighter_GObj* gobj)
-{
-    ft_80084DB0(gobj);
-}
-
-void ftKb_FxSpecialAirNEnd_Phys(Fighter_GObj* gobj)
-{
-    ft_80084DB0(gobj);
-}
-
-void ftKb_FxSpecialNStart_Coll(Fighter_GObj* gobj)
-{
-    ft_80083F88(gobj);
-}
-
-void ftKb_FxSpecialNLoop_Coll(Fighter_GObj* gobj)
-{
-    ft_80083F88(gobj);
-}
-
-void ftKb_FxSpecialNEnd_Coll(Fighter_GObj* gobj)
-{
-    ft_80083F88(gobj);
-}
-
-void ftKb_FxSpecialAirNStart_Coll(Fighter_GObj* gobj)
-{
-    ftCo_AirCatchHit_Coll(gobj);
-}
-
-void ftKb_FxSpecialAirNLoop_Coll(Fighter_GObj* gobj)
-{
-    ftCo_AirCatchHit_Coll(gobj);
-}
-
-void ftKb_FxSpecialAirNEnd_Coll(Fighter_GObj* gobj)
-{
-    ftCo_AirCatchHit_Coll(gobj);
-}
-
-bool ftKb_SpecialNNs_800FEC78(HSD_GObj* gobj)
-{
-    if (gobj != NULL) {
-        Fighter* fp = GET_FIGHTER(gobj);
-        if (fp != NULL && fp->fv.kb.ns_flash_gobj != NULL &&
-            (fp->motion_id == ftKb_MS_NsSpecialNHold0 ||
-             fp->motion_id == ftKb_MS_NsSpecialAirNHold0))
-        {
-            return true;
-        }
-    }
-    return false;
-}
-
-void ftKb_SpecialNNs_800FECBC(HSD_GObj* gobj)
-{
-    if (gobj != NULL) {
-        Fighter* fp = GET_FIGHTER(gobj);
-        if (fp != NULL && fp->fv.kb.ns_flash_gobj != NULL) {
-            fp->fv.kb.ns_flash_gobj = NULL;
-        }
-    }
-}
-
-void ftKb_SpecialNNs_800FECE8(HSD_GObj* gobj)
-{
-    Fighter* fp;
-    if (gobj != NULL) {
-        fp = GET_FIGHTER(gobj);
-        if (fp != NULL && fp->fv.kb.ns_flash_gobj != NULL) {
-            it_802AAA50(fp->fv.kb.ns_flash_gobj);
-            fp->fv.kb.ns_flash_gobj = NULL;
-        }
-    }
-}
-
-void ftKb_SpecialNNs_800FED38(HSD_GObj* gobj)
-{
-    Fighter* fp = GET_FIGHTER(gobj);
-    Fighter_ChangeMotionState(gobj, ftKb_MS_NsSpecialNStart, Ft_MF_None, 0, 1,
-                              0, NULL);
-    fp->cmd_vars[0] = fp->cmd_vars[1] = fp->cmd_vars[2] = fp->cmd_vars[3] = 0;
-    {
-        /// @todo @c da can't move below @c fp
-        ftKb_DatAttrs* da;
-        Fighter* fp = GET_FIGHTER(gobj);
-        PAD_STACK(4 * 4);
-        da = fp->dat_attrs;
-        fp->mv.kb.specialn_ns.frames_to_loop_charge_ground =
-            da->specialn_ns_frames_to_loop_charge_ground;
-        fp->mv.kb.specialn_ns.frames_to_loop_charge_air =
-            da->specialn_ns_frames_to_loop_charge_air;
-        fp->mv.kb.specialn_ns.falling_acceleration_delay =
-            da->specialn_ns_falling_acceleration_delay;
-        fp->fv.kb.ns_flash_gobj = NULL;
-        fp->mv.kb.specialn_ns.charge_release_delay =
-            da->specialn_ns_charge_release_delay;
-        ftAnim_8006EBA4(gobj);
-    }
-}
diff --git a/src/melee/ft/chara/ftKirby/ftkirbyspecialfox.c b/src/melee/ft/chara/ftKirby/ftkirbyspecialfox.c
new file mode 100644
index 0000000000..a79c1b5a06
--- /dev/null
+++ b/src/melee/ft/chara/ftKirby/ftkirbyspecialfox.c
@@ -0,0 +1,713 @@
+#include "ftKb_Init.h"
+#include "types.h"
+
+#include <placeholder.h>
+
+#include "baselib/debug.h"
+
+#include "baselib/forward.h"
+
+#include "ft/chara/ftCommon/ftCo_FallSpecial.h"
+#include "ft/chara/ftCommon/ftpickupitem.h"
+#include "ft/fighter.h"
+
+#include "ft/forward.h"
+
+#include "ft/ft_081B.h"
+#include "ft/ft_0881.h"
+#include "ft/ft_0892.h"
+#include "ft/ftcommon.h"
+#include "ft/inlines.h"
+#include "ft/types.h"
+#include "ftCommon/ftCo_Fall.h"
+
+#include "ftKirby/forward.h"
+
+#include "it/items/itfoxblaster.h"
+#include "it/items/itnesspkflush.h"
+#include "lb/lb_00B0.h"
+#include "lb/lbrefract.h"
+
+#include <common_structs.h>
+#include <stddef.h>
+#include <baselib/gobj.h>
+#include <baselib/random.h>
+#include <MSL/math.h>
+
+void ftKb_SpecialNFx_800FDC70(Fighter_GObj* gobj, Vec3* vec)
+{
+    Vec3 pos1;
+    if (gobj != NULL) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        if (fp != NULL) {
+            pos1.x = 0.0f;
+            pos1.y = 1.45f;
+            pos1.z = 0.016f;
+            lb_8000B1CC(fp->parts[FtPart_R3rdNa].joint, &pos1, vec);
+            return;
+        }
+    }
+    vec->x = vec->y = vec->z = 0.0f;
+}
+
+void ftKb_SpecialNFx_OnChangeAction(Fighter_GObj* gobj)
+{
+    FORCE_PAD_STACK_8;
+
+    ft_800892A0(gobj);
+    ft_80089824(gobj);
+}
+
+bool ftKb_SpecialNFx_800FDD14(Fighter_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Fighter* ft = GET_FIGHTER(gobj);
+        if (ft != NULL) {
+            if (!ft->fv.kb.xB0) {
+                return true;
+            }
+            return false;
+        }
+    }
+    return true;
+}
+
+ftFx_SpecialNIndex ftKb_SpecialNFx_800FDD4C(Fighter_GObj* gobj)
+{
+    s32 msid = 9;
+
+    if (gobj != NULL) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        if (fp != NULL) {
+            switch (fp->fv.kb.hat.kind) {
+            case FTKIND_FOX: {
+                s32 motion_id = fp->motion_id;
+                switch (motion_id) {
+                case ftKb_MS_FxSpecialNStart:
+                case ftKb_MS_FxSpecialNLoop:
+                case ftKb_MS_FxSpecialNEnd:
+                case ftKb_MS_FxSpecialAirNStart:
+                case ftKb_MS_FxSpecialAirNLoop:
+                case ftKb_MS_FxSpecialAirNEnd:
+                    msid = motion_id - 0x1A7;
+                    break;
+                case ftCo_MS_ThrowB:
+                case ftCo_MS_ThrowHi:
+                case ftCo_MS_ThrowLw:
+                    msid = motion_id - 0xD6;
+                    break;
+                }
+                break;
+            }
+            case FTKIND_FALCO: {
+                s32 motion_id = fp->motion_id;
+                switch (motion_id) {
+                case ftKb_MS_FcSpecialNStart:
+                case ftKb_MS_FcSpecialNLoop:
+                case ftKb_MS_FcSpecialNEnd:
+                case ftKb_MS_FcSpecialAirNStart:
+                case ftKb_MS_FcSpecialAirNLoop:
+                case ftKb_MS_FcSpecialAirNEnd:
+                    msid = motion_id - 0x208;
+                    break;
+                case ftCo_MS_ThrowB:
+                case ftCo_MS_ThrowHi:
+                case ftCo_MS_ThrowLw:
+                    msid = motion_id - 0xD6;
+                    break;
+                }
+                break;
+            }
+            }
+        }
+    }
+    return msid;
+}
+
+bool ftKb_SpecialNFx_800FDDF4(Fighter_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        if (fp != NULL) {
+            switch (fp->fv.kb.hat.kind) {
+            case FTKIND_FOX: {
+                s32 motion_id = fp->motion_id;
+                switch (motion_id) {
+                case ftKb_MS_FxSpecialNStart:
+                case ftKb_MS_FxSpecialNLoop:
+                case ftKb_MS_FxSpecialNEnd:
+                case ftKb_MS_FxSpecialAirNStart:
+                case ftKb_MS_FxSpecialAirNLoop:
+                case ftKb_MS_FxSpecialAirNEnd:
+                case ftCo_MS_ThrowB:
+                case ftCo_MS_ThrowHi:
+                case ftCo_MS_ThrowLw:
+                    if (fp->x2070.x2071_b6) {
+                        return true;
+                    }
+                    return false;
+                }
+                break;
+            }
+            case FTKIND_FALCO: {
+                s32 motion_id = fp->motion_id;
+                switch (motion_id) {
+                case ftKb_MS_FcSpecialNStart:
+                case ftKb_MS_FcSpecialNLoop:
+                case ftKb_MS_FcSpecialNEnd:
+                case ftKb_MS_FcSpecialAirNStart:
+                case ftKb_MS_FcSpecialAirNLoop:
+                case ftKb_MS_FcSpecialAirNEnd:
+                case ftCo_MS_ThrowB:
+                case ftCo_MS_ThrowHi:
+                case ftCo_MS_ThrowLw:
+                    if (fp->x2070.x2071_b6) {
+                        return true;
+                    }
+                    return false;
+                }
+                break;
+            }
+            }
+        }
+    }
+    return true;
+}
+
+void ftKb_SpecialNFx_800FDEB4(Fighter_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Fighter* ft = GET_FIGHTER(gobj);
+        if ((ft != NULL) && (ft->fv.kb.xB0 != 0)) {
+            ft->fv.kb.xB0 = 0;
+        }
+    }
+}
+
+void ftKb_SpecialNFx_800FDEE0(Fighter_GObj* gobj)
+{
+    PAD_STACK(8);
+    if (gobj != NULL) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        if (fp != NULL) {
+            if (fp->fv.kb.xB0 != 0) {
+                it_802AEAB4((Item_GObj*) fp->fv.kb.xB0);
+                fp->fv.kb.xB0 = 0;
+            }
+        }
+    }
+}
+
+inline void ftKb_SpecialNFx_SetCall(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    fp->death2_cb = ftKb_Init_800EE74C;
+    fp->take_dmg_cb = ftKb_Init_800EE7B8;
+}
+
+/// #ftKb_SpecialNFx_800FDF30
+
+void ftKb_SpecialNFx_CreateBlasterShot(Fighter_GObj* gobj)
+{
+    ftKb_SpecialNFx_800FDF30(gobj);
+}
+
+inline u32 ftKbGetBlasterId(Fighter* fp, ftKb_DatAttrs* da)
+{
+    u32 blaster_id;
+    switch (fp->fv.kb.hat.kind) {
+    case FTKIND_FOX:
+        blaster_id = da->specialn_fx_blaster_item_id;
+        break;
+    case FTKIND_FALCO:
+        blaster_id = da->specialn_fc_blaster_item_id;
+        break;
+    }
+    return blaster_id;
+}
+
+inline FtMotionId ftKbGetStartMotionId(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    FtMotionId msid = ftKb_MS_FxSpecialNStart;
+    switch (fp->fv.kb.hat.kind) {
+    case FTKIND_FALCO:
+        msid = ftKb_MS_FcSpecialNStart;
+        break;
+    case FTKIND_FOX:
+        msid = ftKb_MS_FxSpecialNStart;
+        break;
+    }
+    return msid;
+}
+
+void ftKb_SpecialNFx_800FE100(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    HSD_GObj* blasterGObj;
+
+    ftCommon_8007D7FC(fp);
+
+    Fighter_ChangeMotionState(gobj, ftKbGetStartMotionId(gobj), 0.0f, 0.0f,
+                              1.0f, 0.0f, NULL);
+
+    fp->cmd_vars[3] = 0;
+    fp->cmd_vars[2] = 0;
+    fp->cmd_vars[1] = 0;
+    fp->cmd_vars[0] = 0;
+
+    ftAnim_8006EBA4(gobj);
+
+    fp->gr_vel = 0.0f;
+    fp->self_vel.z = 0.0f;
+    fp->self_vel.y = 0.0f;
+    fp->self_vel.x = 0.0f;
+
+    fp->mv.fx.SpecialN.isBlasterLoop = false;
+    blasterGObj = it_802AE8A8(fp->facing_dir, gobj, &fp->cur_pos,
+                              FtPart_R3rdNa, ftKbGetBlasterId(fp, da));
+    fp->fv.kb.xB0 = blasterGObj;
+
+    if (blasterGObj != NULL) {
+        ftKb_SpecialNFx_SetCall(gobj);
+        return;
+    }
+
+    OSReport("ftToSpecialNFox::Caution!!!\n");
+    HSD_ASSERT(429, 0);
+}
+
+inline FtMotionId ftKbGetAirStartMotionId(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    FtMotionId msid = ftKb_MS_FxSpecialAirNStart;
+    switch (fp->fv.kb.hat.kind) {
+    case FTKIND_FALCO:
+        msid = ftKb_MS_FcSpecialAirNStart;
+        break;
+    case FTKIND_FOX:
+        msid = ftKb_MS_FxSpecialAirNStart;
+        break;
+    }
+    return msid;
+}
+
+void ftKb_SpecialNFx_800FE240(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    HSD_GObj* blasterGObj;
+
+    Fighter_ChangeMotionState(gobj, ftKbGetAirStartMotionId(gobj), 0.0f, 0.0f,
+                              1.0f, 0.0f, NULL);
+
+    fp->cmd_vars[3] = 0;
+    fp->cmd_vars[2] = 0;
+    fp->cmd_vars[1] = 0;
+    fp->cmd_vars[0] = 0;
+
+    ftAnim_8006EBA4(gobj);
+
+    fp->mv.fx.SpecialN.isBlasterLoop = false;
+    blasterGObj = it_802AE8A8(fp->facing_dir, gobj, &fp->cur_pos,
+                              FtPart_R3rdNa, ftKbGetBlasterId(fp, da));
+    fp->fv.kb.xB0 = blasterGObj;
+
+    if (blasterGObj != NULL) {
+        ftKb_SpecialNFx_SetCall(gobj);
+        return;
+    }
+
+    OSReport("ftToSpecialNFox::Caution!!!\n");
+    HSD_ASSERT(465, 0);
+}
+
+inline FtMotionId ftKbGetLoopMotionId(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    FtMotionId msid = ftKb_MS_FxSpecialNLoop;
+    switch (fp->fv.kb.hat.kind) {
+    case FTKIND_FALCO:
+        msid = ftKb_MS_FcSpecialNLoop;
+        break;
+    case FTKIND_FOX:
+        msid = ftKb_MS_FxSpecialNLoop;
+        break;
+    }
+    return msid;
+}
+
+void ftKb_FxSpecialNStart_Anim(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+
+    it_802ADDD0(fp->fv.kb.xB0, 1);
+    if (fp->cmd_vars[3] == 1 && fp->fv.kb.xB0 != NULL) {
+        fp->cmd_vars[3] = 0;
+        it_802AE538(fp->fv.kb.xB0);
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        Fighter_ChangeMotionState(gobj, ftKbGetLoopMotionId(gobj),
+                                  (Ft_MF_SkipModel | Ft_MF_KeepGfx), 0, 1, 0,
+                                  NULL);
+        ftKb_SpecialNFx_SetCall(gobj);
+        fp->accessory4_cb = ftKb_SpecialNFx_CreateBlasterShot;
+        it_802ADDD0(fp->fv.kb.xB0, 1);
+    }
+}
+
+inline FtMotionId ftKbGetEndMotionId(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    FtMotionId msid = ftKb_MS_FxSpecialNEnd;
+    switch (fp->fv.kb.hat.kind) {
+    case FTKIND_FALCO:
+        msid = ftKb_MS_FcSpecialNEnd;
+        break;
+    case FTKIND_FOX:
+        msid = ftKb_MS_FxSpecialNEnd;
+        break;
+    }
+    return msid;
+}
+
+void ftKb_FxSpecialNLoop_Anim(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    HSD_GObj* temp;
+
+    it_802ADDD0(fp->fv.kb.xB0, 1);
+    if (fp->cmd_vars[3] == 1 && fp->fv.kb.xB0 != NULL) {
+        fp->cmd_vars[3] = 0;
+        it_802AE538(fp->fv.kb.xB0);
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        if (fp->mv.fx.SpecialN.isBlasterLoop == true) {
+            fp->x21EC = ftKb_SpecialNFx_OnChangeAction;
+            Fighter_ChangeMotionState(
+                temp = gobj, ftKbGetLoopMotionId(gobj),
+                (Ft_MF_SkipAttackCount | Ft_MF_SkipModel | Ft_MF_KeepGfx), 0,
+                1, 0, NULL);
+            fp->accessory4_cb = ftKb_SpecialNFx_CreateBlasterShot;
+            fp->mv.fx.SpecialN.isBlasterLoop = false;
+            it_802ADDD0(fp->fv.kb.xB0, 1);
+        } else {
+            HSD_GObj* temp;
+            Fighter_ChangeMotionState(gobj, ftKbGetEndMotionId(gobj),
+                                      (Ft_MF_SkipModel | Ft_MF_KeepGfx), 0, 1,
+                                      0, NULL);
+            temp = fp->fv.kb.xB0;
+            fp->cmd_vars[1] = 1;
+            it_802ADDD0(temp, 1);
+        }
+        ftKb_SpecialNFx_SetCall(gobj);
+    }
+    ftKb_SpecialNFx_800FDF30(gobj);
+}
+
+void ftKb_FxSpecialNEnd_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+
+    it_802ADDD0(fp->fv.kb.xB0, fp->cmd_vars[1]);
+
+    if (fp->cmd_vars[1] == 2) {
+        ftpickupitem_80094818(gobj, 0);
+    }
+
+    if (fp->cmd_vars[3] == 2 && fp->fv.kb.xB0 != 0) {
+        unsigned long long pad;
+        fp->cmd_vars[3] = 0;
+        it_802AE608(fp->fv.kb.xB0);
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftKb_SpecialNFx_800FDEB4(gobj);
+        ft_8008A2BC(gobj);
+    }
+}
+
+inline FtMotionId ftKbGetAirLoopMotionId(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    FtMotionId msid = ftKb_MS_FxSpecialAirNLoop;
+    switch (fp->fv.kb.hat.kind) {
+    case FTKIND_FALCO:
+        msid = ftKb_MS_FcSpecialAirNLoop;
+        break;
+    case FTKIND_FOX:
+        msid = ftKb_MS_FxSpecialAirNLoop;
+        break;
+    }
+    return msid;
+}
+
+void ftKb_FxSpecialAirNStart_Anim(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+
+    it_802ADDD0(fp->fv.kb.xB0, 1);
+    if (fp->cmd_vars[3] == 1 && fp->fv.kb.xB0 != NULL) {
+        fp->cmd_vars[3] = 0;
+        it_802AE538(fp->fv.kb.xB0);
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        Fighter_ChangeMotionState(gobj, ftKbGetAirLoopMotionId(gobj),
+                                  (Ft_MF_SkipModel | Ft_MF_KeepGfx), 0, 1, 0,
+                                  NULL);
+        ftKb_SpecialNFx_SetCall(gobj);
+        fp->accessory4_cb = ftKb_SpecialNFx_CreateBlasterShot;
+        it_802ADDD0(fp->fv.kb.xB0, 1);
+    }
+}
+
+inline FtMotionId ftKbGetAirEndMotionId(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    FtMotionId msid = ftKb_MS_FxSpecialAirNEnd;
+    switch (fp->fv.kb.hat.kind) {
+    case FTKIND_FALCO:
+        msid = ftKb_MS_FcSpecialAirNEnd;
+        break;
+    case FTKIND_FOX:
+        msid = ftKb_MS_FxSpecialAirNEnd;
+        break;
+    }
+    return msid;
+}
+
+void ftKb_FxSpecialAirNLoop_Anim(HSD_GObj* gobj)
+{
+    Fighter* fp = gobj->user_data;
+    HSD_GObj* temp;
+    PAD_STACK(8);
+
+    it_802ADDD0(fp->fv.kb.xB0, 1);
+    if (fp->cmd_vars[3] == 1 && fp->fv.kb.xB0 != NULL) {
+        fp->cmd_vars[3] = 0;
+        it_802AE538(fp->fv.kb.xB0);
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        if (fp->mv.fx.SpecialN.isBlasterLoop == true) {
+            fp->x21EC = ftKb_SpecialNFx_OnChangeAction;
+            Fighter_ChangeMotionState(
+                temp = gobj, ftKbGetAirLoopMotionId(gobj),
+                (Ft_MF_SkipAttackCount | Ft_MF_SkipModel | Ft_MF_KeepGfx), 0,
+                1, 0, NULL);
+            ftKb_SpecialNFx_SetCall(gobj);
+            fp->accessory4_cb = ftKb_SpecialNFx_CreateBlasterShot;
+            fp->mv.fx.SpecialN.isBlasterLoop = false;
+            it_802ADDD0(fp->fv.kb.xB0, 1);
+        } else {
+            Fighter_ChangeMotionState(gobj, ftKbGetAirEndMotionId(gobj),
+                                      (Ft_MF_SkipModel | Ft_MF_KeepGfx), 0, 1,
+                                      0, NULL);
+            ftKb_SpecialNFx_SetCall(gobj);
+            temp = fp->fv.kb.xB0;
+            fp->cmd_vars[1] = 1;
+            it_802ADDD0(temp, 1);
+        }
+        ftKb_SpecialNFx_SetCall(gobj);
+    }
+    ftKb_SpecialNFx_800FDF30(gobj);
+}
+
+void ftKb_FxSpecialAirNEnd_Anim(Fighter_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter* new_var;
+    ftKb_DatAttrs* da = fp->dat_attrs;
+    volatile int pad;
+    int new_var2;
+    float landing_lag;
+
+    it_802ADDD0((Item_GObj*) fp->fv.kb.xB0, fp->cmd_vars[1]);
+
+    new_var = fp;
+    if (new_var->cmd_vars[1] == 2) {
+        ftpickupitem_80094818(gobj, 0);
+    }
+
+    if (fp->cmd_vars[3] == 2 && fp->fv.kb.xB0 != 0) {
+        new_var2 = 0;
+        fp->cmd_vars[3] = new_var2;
+        it_802AE608((Item_GObj*) fp->fv.kb.xB0);
+    }
+    if (!ftAnim_IsFramesRemaining(gobj)) {
+        ftKb_SpecialNFx_800FDEB4(gobj);
+        switch (fp->fv.kb.hat.kind) {
+        case FTKIND_FOX:
+            do {
+            } while (0);
+            landing_lag = da->specialn_fx_freefall_toggle;
+            break;
+        case FTKIND_FALCO:
+            landing_lag = da->specialn_fc_freefall_toggle;
+            break;
+        default:
+            break;
+        }
+        if (landing_lag == 0.0F) {
+            ftCo_Fall_Enter(gobj);
+            return;
+        }
+        ftCo_80096900(gobj, 1, 0, 1, 1.0F, landing_lag);
+    }
+}
+
+void ftKb_FxSpecialNStart_IASA(Fighter_GObj* gobj)
+{
+    Fighter* ft = GET_FIGHTER(gobj);
+    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
+        ft->mv.kb.specialhi.x0 = 1;
+    }
+}
+
+void ftKb_FxSpecialNLoop_IASA(Fighter_GObj* gobj)
+{
+    Fighter* ft = GET_FIGHTER(gobj);
+    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
+        ft->mv.kb.specialhi.x0 = 1;
+    }
+}
+
+void ftKb_FxSpecialNEnd_IASA(Fighter_GObj* gobj) {}
+
+void ftKb_FxSpecialAirNStart_IASA(Fighter_GObj* gobj)
+{
+    Fighter* ft = GET_FIGHTER(gobj);
+    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
+        ft->mv.kb.specialhi.x0 = 1;
+    }
+}
+
+void ftKb_FxSpecialAirNLoop_IASA(Fighter_GObj* gobj)
+{
+    Fighter* ft = GET_FIGHTER(gobj);
+    if ((ft->cmd_vars[0] != 0) && (ft->input.x668 & HSD_PAD_B)) {
+        ft->mv.kb.specialhi.x0 = 1;
+    }
+}
+
+void ftKb_FxSpecialAirNEnd_IASA(Fighter_GObj* gobj) {}
+
+void ftKb_FxSpecialNStart_Phys(Fighter_GObj* gobj)
+{
+    ft_80084F3C(gobj);
+}
+
+void ftKb_FxSpecialNLoop_Phys(Fighter_GObj* gobj)
+{
+    ft_80084F3C(gobj);
+}
+
+void ftKb_FxSpecialNEnd_Phys(Fighter_GObj* gobj)
+{
+    ft_80084F3C(gobj);
+}
+
+void ftKb_FxSpecialAirNStart_Phys(Fighter_GObj* gobj)
+{
+    ft_80084DB0(gobj);
+}
+
+void ftKb_FxSpecialAirNLoop_Phys(Fighter_GObj* gobj)
+{
+    ft_80084DB0(gobj);
+}
+
+void ftKb_FxSpecialAirNEnd_Phys(Fighter_GObj* gobj)
+{
+    ft_80084DB0(gobj);
+}
+
+void ftKb_FxSpecialNStart_Coll(Fighter_GObj* gobj)
+{
+    ft_80083F88(gobj);
+}
+
+void ftKb_FxSpecialNLoop_Coll(Fighter_GObj* gobj)
+{
+    ft_80083F88(gobj);
+}
+
+void ftKb_FxSpecialNEnd_Coll(Fighter_GObj* gobj)
+{
+    ft_80083F88(gobj);
+}
+
+void ftKb_FxSpecialAirNStart_Coll(Fighter_GObj* gobj)
+{
+    ftCo_AirCatchHit_Coll(gobj);
+}
+
+void ftKb_FxSpecialAirNLoop_Coll(Fighter_GObj* gobj)
+{
+    ftCo_AirCatchHit_Coll(gobj);
+}
+
+void ftKb_FxSpecialAirNEnd_Coll(Fighter_GObj* gobj)
+{
+    ftCo_AirCatchHit_Coll(gobj);
+}
+
+bool ftKb_SpecialNNs_800FEC78(HSD_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        if (fp != NULL && fp->fv.kb.ns_flash_gobj != NULL &&
+            (fp->motion_id == ftKb_MS_NsSpecialNHold0 ||
+             fp->motion_id == ftKb_MS_NsSpecialAirNHold0))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+void ftKb_SpecialNNs_800FECBC(HSD_GObj* gobj)
+{
+    if (gobj != NULL) {
+        Fighter* fp = GET_FIGHTER(gobj);
+        if (fp != NULL && fp->fv.kb.ns_flash_gobj != NULL) {
+            fp->fv.kb.ns_flash_gobj = NULL;
+        }
+    }
+}
+
+void ftKb_SpecialNNs_800FECE8(HSD_GObj* gobj)
+{
+    Fighter* fp;
+    if (gobj != NULL) {
+        fp = GET_FIGHTER(gobj);
+        if (fp != NULL && fp->fv.kb.ns_flash_gobj != NULL) {
+            it_802AAA50(fp->fv.kb.ns_flash_gobj);
+            fp->fv.kb.ns_flash_gobj = NULL;
+        }
+    }
+}
+
+void ftKb_SpecialNNs_800FED38(HSD_GObj* gobj)
+{
+    Fighter* fp = GET_FIGHTER(gobj);
+    Fighter_ChangeMotionState(gobj, ftKb_MS_NsSpecialNStart, Ft_MF_None, 0, 1,
+                              0, NULL);
+    fp->cmd_vars[0] = fp->cmd_vars[1] = fp->cmd_vars[2] = fp->cmd_vars[3] = 0;
+    {
+        /// @todo @c da can't move below @c fp
+        ftKb_DatAttrs* da;
+        Fighter* fp = GET_FIGHTER(gobj);
+        PAD_STACK(4 * 4);
+        da = fp->dat_attrs;
+        fp->mv.kb.specialn_ns.frames_to_loop_charge_ground =
+            da->specialn_ns_frames_to_loop_charge_ground;
+        fp->mv.kb.specialn_ns.frames_to_loop_charge_air =
+            da->specialn_ns_frames_to_loop_charge_air;
+        fp->mv.kb.specialn_ns.falling_acceleration_delay =
+            da->specialn_ns_falling_acceleration_delay;
+        fp->fv.kb.ns_flash_gobj = NULL;
+        fp->mv.kb.specialn_ns.charge_release_delay =
+            da->specialn_ns_charge_release_delay;
+        ftAnim_8006EBA4(gobj);
+    }
+}
diff --git a/src/melee/ft/chara/ftKirby/types.h b/src/melee/ft/chara/ftKirby/types.h
index 0ce39d719f..f31d52f7b0 100644
--- a/src/melee/ft/chara/ftKirby/types.h
+++ b/src/melee/ft/chara/ftKirby/types.h
@@ -51,7 +51,7 @@ struct ftKb_FighterVars {
     /* fp+22D0   */ UNK_T xA4;
     /* fp+22D4   */ int xA8;
     /* fp+22D8   */ UNK_T xAC;
-    /* fp+22DC   */ u32 xB0;
+    /* fp+22DC   */ HSD_GObj* xB0; ///< Fox/Falco blaster item
     /* fp+22E0   */ int xB4;
     /* fp+22E4   */ Item_GObj* xB8; ///< Sheik needle item
     /* fp+22E8   */ int xBC;
