diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 7528921f21..c3b1a1b39e 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -20135,6 +20135,7 @@ grSh_Route_803B8378 = .rodata:0x803B8378; // type:object size:0xC scope:global d
 grSh_Route_803B8384 = .rodata:0x803B8384; // type:object size:0xC scope:global data:4byte
 grSh_Route_803B8390 = .rodata:0x803B8390; // type:object size:0x10 scope:global data:4byte
 grZe_Route_803B83A0 = .rodata:0x803B83A0; // type:object size:0x40 scope:global data:4byte
+...rodata.0 = .rodata:0x803B83A0; // type:label scope:local
 grBb_Route_803B83E0 = .rodata:0x803B83E0; // type:object size:0x10 scope:global data:4byte
 grOy_803B83F0 = .rodata:0x803B83F0; // type:object size:0x18 scope:global data:4byte
 grOk_803B8408 = .rodata:0x803B8408; // type:object size:0x28 scope:global data:4byte
@@ -22821,9 +22822,10 @@ grSh_Route_803E5D64 = .data:0x803E5D64; // type:object size:0x10 scope:global da
 grSh_Route_803E5D74 = .data:0x803E5D74; // type:object size:0x1C scope:global
 grSh_Route_803E5D90 = .data:0x803E5D90; // type:object size:0x20 scope:global
 grZe_Route_803E5DB0 = .data:0x803E5DB0; // type:object size:0x50 scope:global
+...data.0 = .data:0x803E5DB0; // type:label scope:local
 grZe_Route_803E5E00 = .data:0x803E5E00; // type:object size:0xB scope:global data:string
 grZe_Route_803E5E0C = .data:0x803E5E0C; // type:object size:0x58 scope:global
-grZe_Route_803E5E64 = .data:0x803E5E64; // type:object size:0xF scope:global data:string
+@187 = .data:0x803E5E64; // type:object size:0xF scope:local data:string
 grBb_Route_803E5E78 = .data:0x803E5E78; // type:object size:0x2F8 scope:global
 grBb_Route_803E6170 = .data:0x803E6170; // type:object size:0xB scope:global data:string
 grBb_Route_803E617C = .data:0x803E617C; // type:object size:0x58 scope:global
@@ -26706,8 +26708,8 @@ grSh_Route_804D4824 = .sdata:0x804D4824; // type:object size:0x4 scope:global
 grSh_Route_804D4828 = .sdata:0x804D4828; // type:object size:0x7 scope:global data:string
 grSh_Route_804D4830 = .sdata:0x804D4830; // type:object size:0x5 scope:global data:string
 grSh_Route_804D4838 = .sdata:0x804D4838; // type:object size:0x8 scope:global
-grZe_Route_804D4840 = .sdata:0x804D4840; // type:object size:0x5 scope:global data:string
-grZe_Route_804D4848 = .sdata:0x804D4848; // type:object size:0x2 scope:global data:string
+@270 = .sdata:0x804D4840; // type:object size:0x5 scope:local data:string
+@274 = .sdata:0x804D4848; // type:object size:0x2 scope:local data:string
 grBb_Route_804D4850 = .sdata:0x804D4850; // type:object size:0x7 scope:global data:string
 grBb_Route_804D4858 = .sdata:0x804D4858; // type:object size:0x5 scope:global data:string
 grBb_Route_804D4860 = .sdata:0x804D4860; // type:object size:0x2 scope:global data:string
@@ -32119,18 +32121,18 @@ grSh_Route_804DB904 = .sdata2:0x804DB904; // type:object size:0x4 scope:global a
 grSh_Route_804DB908 = .sdata2:0x804DB908; // type:object size:0x4 scope:global align:4 data:float
 grSh_Route_804DB90C = .sdata2:0x804DB90C; // type:object size:0x4 scope:global align:4 data:float
 grSh_Route_804DB910 = .sdata2:0x804DB910; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB918 = .sdata2:0x804DB918; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB91C = .sdata2:0x804DB91C; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB920 = .sdata2:0x804DB920; // type:object size:0x4 scope:global data:4byte
-grZe_Route_804DB924 = .sdata2:0x804DB924; // type:object size:0x4 scope:global data:4byte
-grZe_Route_804DB928 = .sdata2:0x804DB928; // type:object size:0x4 scope:global data:4byte
-grZe_Route_804DB92C = .sdata2:0x804DB92C; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB930 = .sdata2:0x804DB930; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB934 = .sdata2:0x804DB934; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB938 = .sdata2:0x804DB938; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB93C = .sdata2:0x804DB93C; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB940 = .sdata2:0x804DB940; // type:object size:0x4 scope:global align:4 data:float
-grZe_Route_804DB944 = .sdata2:0x804DB944; // type:object size:0x4 scope:global align:4 data:float
+@211 = .sdata2:0x804DB918; // type:object size:0x4 scope:local align:4 data:float
+@212 = .sdata2:0x804DB91C; // type:object size:0x4 scope:local align:4 data:float
+@233 = .sdata2:0x804DB920; // type:object size:0x4 scope:local data:4byte
+@239 = .sdata2:0x804DB924; // type:object size:0x4 scope:local data:4byte
+@245 = .sdata2:0x804DB928; // type:object size:0x4 scope:local data:4byte
+@271 = .sdata2:0x804DB92C; // type:object size:0x4 scope:local align:4 data:float
+@272 = .sdata2:0x804DB930; // type:object size:0x4 scope:local align:4 data:float
+@273 = .sdata2:0x804DB934; // type:object size:0x4 scope:local align:4 data:float
+@275 = .sdata2:0x804DB938; // type:object size:0x4 scope:local align:4 data:float
+@276 = .sdata2:0x804DB93C; // type:object size:0x4 scope:local align:4 data:float
+@277 = .sdata2:0x804DB940; // type:object size:0x4 scope:local align:4 data:float
+@278 = .sdata2:0x804DB944; // type:object size:0x4 scope:local align:4 data:float
 grBb_Route_804DB948 = .sdata2:0x804DB948; // type:object size:0x4 scope:global align:4 data:float
 grBb_Route_804DB94C = .sdata2:0x804DB94C; // type:object size:0x4 scope:global align:4 data:float
 grBb_Route_804DB950 = .sdata2:0x804DB950; // type:object size:0x4 scope:global align:4 data:float
diff --git a/configure.py b/configure.py
index b3db31ad62..213d713213 100755
--- a/configure.py
+++ b/configure.py
@@ -1007,7 +1007,7 @@ def MatchingFor(*versions):
             Object(Matching, "melee/gr/grtest.c"),
             Object(NonMatching, "melee/gr/grkinokoroute.c"),
             Object(NonMatching, "melee/gr/grshrineroute.c"),
-            Object(NonMatching, "melee/gr/grzebesroute.c"),
+            Object(Matching, "melee/gr/grzebesroute.c"),
             Object(NonMatching, "melee/gr/grbigblueroute.c"),
             Object(Matching, "melee/gr/grfigure1.c"),
             Object(Matching, "melee/gr/grfigure2.c"),
diff --git a/src/melee/gr/grzebesroute.c b/src/melee/gr/grzebesroute.c
index af41758840..09267001a7 100644
--- a/src/melee/gr/grzebesroute.c
+++ b/src/melee/gr/grzebesroute.c
@@ -1,9 +1,15 @@
 #include "grzebesroute.h"
 
+#include "placeholder.h"
+
 #include <platform.h>
 
 #include "baselib/debug.h"
+
+#include "baselib/forward.h"
+
 #include "cm/camera.h"
+#include "dolphin/mtx.h"
 #include "ft/ftlib.h"
 #include "gr/grdisplay.h"
 #include "gr/ground.h"
@@ -87,29 +93,10 @@ struct {
     "%s:%d: couldn t get gobj(id=%d)\n",
 };
 
-char grZe_Route_803E5E64[] = "grzebesroute.c";
-
-SDATA char grZe_Route_804D4840[] = "gobj";
-SDATA char grZe_Route_804D4848[] = "0";
-
 grZebesRoute_ParamStore grZe_Route_804D6A60;
 
-const f32 grZe_Route_804DB918 = -50.0F;
-const f32 grZe_Route_804DB91C = 0.0F;
-const GXColor grZe_Route_804DB920 = { 0xFF, 0xFF, 0xCC, 0xFF };
-const GXColor grZe_Route_804DB924 = { 0xFD, 0xFD, 0xBF, 0xFF };
-const GXColor grZe_Route_804DB928 = { 0x00, 0x00, 0xF7, 0xFF };
-const f32 grZe_Route_804DB92C = 600.0F;
-const f32 grZe_Route_804DB930 = 1000.0F;
-const f32 grZe_Route_804DB934 = 400.0F;
-const f32 grZe_Route_804DB938 = 45.0F;
-const f32 grZe_Route_804DB93C = 0.99F;
-const f32 grZe_Route_804DB940 = 500.0F;
-const f32 grZe_Route_804DB944 = 0.03F;
-
 void grZebesRoute_8020B160(bool arg) {}
 
-/// #grZebesRoute_8020B164
 void grZebesRoute_8020B164(void)
 {
     grZe_Route_804D6A60.params = Ground_801C49F8();
@@ -169,7 +156,7 @@ HSD_GObj* grZebesRoute_8020B260(int gobj_id)
             HSD_GObj_SetupProc(gobj, callbacks->callback2, 4);
         }
     } else {
-        OSReport(grZe_Route_803E5E0C.fmt, grZe_Route_803E5E64, 0xC5, gobj_id);
+        OSReport(grZe_Route_803E5E0C.fmt, __FILE__, 0xC5, gobj_id);
     }
 
     return gobj;
@@ -230,10 +217,10 @@ void grZebesRoute_8020B42C(Ground_GObj* gobj)
     fighter = Ground_801C57A4();
     if (fighter != NULL) {
         ftLib_80086644(fighter, &pos);
-        if (pos.y < grZe_Route_804DB918) {
-            pos.y = grZe_Route_804DB918;
+        if (pos.y < -50.0f) {
+            pos.y = -50.0f;
         }
-        pos.x = grZe_Route_804DB91C;
+        pos.x = 0.0f;
         Ground_801C38BC(pos.x, pos.y);
     }
 
@@ -262,10 +249,24 @@ void fn_8020B4D8(Ground* gp, s32 arg1, CollData* coll, s32 arg3,
     }
 }
 
-/// #grZebesRoute_8020B548
+inline HSD_LObj* grZebesRoute_8020B548_inline1(HSD_GObj* light_gobj)
+{
+    return (HSD_LObj*) light_gobj->hsd_obj;
+}
+
+inline HSD_LObj* grZebesRoute_8020B548_inline2(HSD_LObj* arg0)
+{
+    return HSD_LObjGetNext(arg0);
+}
+
+inline int grZebesRoute_8020B548_inline3(HSD_LObj* lobj)
+{
+    return lobj->flags & LOBJ_TYPE_MASK;
+}
+
 void grZebesRoute_8020B548(void)
 {
-    HSD_GObj* light_gobj = HSD_GObjGXLinkHead[4];
+    HSD_GObj* gobj = HSD_GObjGXLinkHead[4];
     HSD_LObj* first_lobj;
     HSD_LObj* lobj;
     f32 scale;
@@ -275,34 +276,32 @@ void grZebesRoute_8020B548(void)
     f32 split_y;
     s32 i;
 
-    if (light_gobj == NULL) {
-        __assert(grZe_Route_803E5E64, 0x169U, grZe_Route_804D4840);
-    }
+    HSD_ASSERT(0x169, gobj);
 
-    first_lobj = light_gobj->hsd_obj;
+    first_lobj = grZebesRoute_8020B548_inline1(gobj);
     scale = Ground_801C0498();
-    lobj = HSD_LObjGetNext(first_lobj);
-    spot_ref_dist = grZe_Route_804DB92C * scale;
-    upper_ref_dist = grZe_Route_804DB930 * scale;
-    lower_ref_dist = grZe_Route_804DB934 * scale;
-    split_y = grZe_Route_804DB940;
-
-    i = 0;
-    while (i < 3 && lobj != NULL) {
+    lobj = grZebesRoute_8020B548_inline2(first_lobj);
+    spot_ref_dist = 600.0f * scale;
+    upper_ref_dist = 1000.0f * scale;
+    lower_ref_dist = 400.0f * scale;
+    split_y = 500.0f;
+
+    for (i = 0; i < 3 && lobj != NULL;
+         i++, lobj = grZebesRoute_8020B548_inline2(lobj))
+    {
         Vec3 current_pos;
         u32 light_type;
 
         if (!HSD_LObjGetPosition(lobj, &current_pos)) {
-            __assert(grZe_Route_803E5E64, 0x174U, grZe_Route_804D4848);
+            HSD_ASSERT(0x174, 0);
         }
 
-        light_type = lobj->flags & LOBJ_TYPE_MASK;
+        light_type = grZebesRoute_8020B548_inline3(lobj);
         if (light_type == LOBJ_SPOT) {
-            GXColor color;
+            GXColor color = { 0xFF, 0xFF, 0xCC, 0xFF };
             Vec3 pos = grZe_Route_803B83A0.spot_pos;
             Vec3 interest = grZe_Route_803B83A0.spot_interest;
 
-            color = grZe_Route_804DB920;
             HSD_LObjSetColor(lobj, color);
             pos.x *= scale;
             pos.y *= scale;
@@ -312,35 +311,29 @@ void grZebesRoute_8020B548(void)
             interest.y *= scale;
             interest.z *= scale;
             HSD_LObjSetInterest(lobj, &interest);
-            HSD_LObjSetSpot(lobj, grZe_Route_804DB938, 3);
-            HSD_LObjSetDistAttn(lobj, spot_ref_dist, grZe_Route_804DB93C, 3);
+            HSD_LObjSetSpot(lobj, 45.0f, 3);
+            HSD_LObjSetDistAttn(lobj, spot_ref_dist, 0.99f, 3);
         } else if (light_type == LOBJ_POINT && current_pos.y > split_y) {
-            GXColor color;
+            GXColor color = { 0xFD, 0xFD, 0xBF, 0xFF };
             Vec3 pos = grZe_Route_803B83A0.upper_point_pos;
 
-            color = grZe_Route_804DB924;
             HSD_LObjSetColor(lobj, color);
             pos.x *= scale;
             pos.y *= scale;
             pos.z *= scale;
             HSD_LObjSetPosition(lobj, &pos);
-            HSD_LObjSetDistAttn(lobj, upper_ref_dist, grZe_Route_804DB944, 3);
+            HSD_LObjSetDistAttn(lobj, upper_ref_dist, 0.03f, 3);
         } else if (light_type == LOBJ_POINT && current_pos.y < split_y) {
-            GXColor color;
+            GXColor color = { 0x00, 0x00, 0xF7, 0xFF };
             Vec3 pos = grZe_Route_803B83A0.lower_point_pos;
-            int pad[3];
 
-            color = grZe_Route_804DB928;
             HSD_LObjSetColor(lobj, color);
             pos.x *= scale;
             pos.y *= scale;
             pos.z *= scale;
             HSD_LObjSetPosition(lobj, &pos);
-            HSD_LObjSetDistAttn(lobj, lower_ref_dist, grZe_Route_804DB944, 3);
+            HSD_LObjSetDistAttn(lobj, lower_ref_dist, 0.03f, 3);
         }
-
-        i++;
-        lobj = HSD_LObjGetNext(lobj);
     }
 }
 
