diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
index a1e57d13a6..c276b0d26b 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialN.c
@@ -1324,7 +1324,7 @@ void ftKb_SpecialN_800F598C(Fighter_GObj* gobj, int arg1)
     ftKb_DatAttrs* da = fp->dat_attrs;
     if (fp->ground_or_air == GA_Ground) {
         new_var = (f32) arg1;
-        mpUpdateFloorSkip(&fp->coll_data);
+        ftKb_SpecialS_800F6BB0(gobj);
         fp->self_vel.y = da->specialn_stop_momentum * new_var;
     }
 }
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
index 2a9ee5241f..cfe8bf893b 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNYs.c
@@ -796,9 +796,9 @@ void ftKb_MsSpecialAirNStart_Coll(Fighter_GObj* gobj)
 
 void ftKb_MsSpecialNLoop_Anim(Fighter_GObj* gobj)
 {
+    u32* charge;
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
-    u32* charge;
     if (fp->fv.kb.hat.kind == FTKIND_MARS) {
         charge = &da->ms.charge_iterations;
     } else {
@@ -937,6 +937,7 @@ void ftKb_MsSpecialNEnd_Anim(Fighter_GObj* gobj)
     struct ftKb_SpecialNMs_DatAttrs* ms_da;
     s32 i;
     HitCapsule* hit;
+    PAD_STACK(24);
     if ((s32) fp->fv.kb.hat.kind == FTKIND_MARS) {
         ms_da = &da->ms;
     } else {
@@ -968,6 +969,7 @@ void ftKb_MsSpecialAirNEnd_Anim(Fighter_GObj* gobj)
     ftKb_DatAttrs* da = fp->dat_attrs;
     struct ftKb_SpecialNMs_DatAttrs* ms_da;
     HitCapsule* hit;
+    PAD_STACK(24);
     if ((s32) fp->fv.kb.hat.kind == FTKIND_MARS) {
         ms_da = &da->ms;
     } else {
diff --git a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
index a2e96b8f7e..097bc4a198 100644
--- a/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
+++ b/src/melee/ft/chara/ftKirby/ftKb_SpecialNZd.c
@@ -400,22 +400,21 @@ void ftKb_SkSpecialAirNStart_Anim(Fighter_GObj* gobj)
 
 void ftKb_SkSpecialAirNLoop_Anim(Fighter_GObj* gobj)
 {
-    volatile unsigned char pad;
     Fighter* fp = GET_FIGHTER(gobj);
-    Fighter* new_var;
+    Fighter* new_var = fp;
+    PAD_STACK(8);
     switch (fp->mv.kb.specialhi.x8.i) {
     case 0:
-        ft_PlaySFX(fp, 270134, 127, 64);
+        ft_PlaySFX(new_var, 270134, 127, 64);
         break;
     }
     ++fp->mv.kb.specialhi.x8.i;
     if (fp->cur_anim_frame == 0) {
-        new_var = fp;
         ++fp->fv.kb.xB4;
         fp->mv.kb.specialhi.x8.i = 0;
         if (fp->fv.kb.xB4 > 6) {
-            new_var->fv.kb.xB4 = 6;
-            new_var->mv.kb.specialhi.x8.i = 100;
+            fp->fv.kb.xB4 = 6;
+            fp->mv.kb.specialhi.x8.i = 100;
             ftCo_800BFFD0(fp, 87, 0);
         }
     }
@@ -524,11 +523,12 @@ void ftKb_SkSpecialNCancel_Coll(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
+    PAD_STACK(8);
     if (ft_80082708(gobj) == GA_Ground) {
         if (da->specialn_sk_freefall_toggle == 0.0F) {
             ftCo_Fall_Enter(gobj);
         } else {
-            ftCo_80096900(gobj, 1, 0, 1, 0.0F,
+            ftCo_80096900(gobj, 1, 0, 1, 1.0F,
                           da->specialn_sk_freefall_toggle);
         }
     }
@@ -538,13 +538,14 @@ void ftKb_SkSpecialNEnd_Coll(Fighter_GObj* gobj)
 {
     Fighter* fp = GET_FIGHTER(gobj);
     ftKb_DatAttrs* da = fp->dat_attrs;
+    PAD_STACK(16);
     if (ft_80082708(gobj) == GA_Ground) {
         fp->fv.kb.xB4 = 0;
         fp->mv.kb.specialhi.x4 = 0;
         if (da->specialn_sk_freefall_toggle == 0.0F) {
             ftCo_Fall_Enter(gobj);
         } else {
-            ftCo_80096900(gobj, 1, 0, 1, 0.0F,
+            ftCo_80096900(gobj, 1, 0, 1, 1.0F,
                           da->specialn_sk_freefall_toggle);
         }
     }
diff --git a/src/melee/ft/ftcoll.c b/src/melee/ft/ftcoll.c
index 171dbb2c80..2c1f08afa6 100644
--- a/src/melee/ft/ftcoll.c
+++ b/src/melee/ft/ftcoll.c
@@ -1627,7 +1627,7 @@ float ftColl_8007BBCC(UNUSED Fighter_GObj* gobj)
     if (dmg_log0_idx != 0) {
         for (i = 0; i < dmg_log0_idx; i++) {
             DmgLogEntry* entry = &dmg_log0[i];
-            switch (entry->kind) {
+            switch (entry->x0) {
             case EntityKind_Fighter: {
                 HitCapsule* hit = entry->hit0;
                 if (hit->element == HitElement_Lipstick) {
diff --git a/src/melee/gm/gm_1601.c b/src/melee/gm/gm_1601.c
index 2fae3255f4..344b365d0f 100644
--- a/src/melee/gm/gm_1601.c
+++ b/src/melee/gm/gm_1601.c
@@ -1668,14 +1668,14 @@ bool fn_801642A0(void)
 }
 
 // RandomStageSwitch
-bool gm_80164330(u8 arg0)
+bool gm_80164330(s32 arg0)
 {
     s32 total_stages_on;
     struct gmm_x1CB0* temp_ret;
     s32 i;
     u8 var_r0;
 
-    if (gm_80164430(lbl_803B7808[arg0]) == 0) {
+    if (gm_80164430(lbl_803B7808[(u8) arg0]) == 0) {
         return false;
     }
     if (gmMainLib_8015EE44() == 0) {
@@ -1686,7 +1686,7 @@ bool gm_80164330(u8 arg0)
     i = 0;
     for (i = 0; i < 0x1D; i++) {
         temp_ret = gmMainLib_8015CC58();
-        if ((1 << i) & temp_ret->stage_mask) {
+        if (temp_ret->stage_mask & (1 << (u16) i)) {
             var_r0 = 1;
         } else {
             var_r0 = 0;
@@ -1698,7 +1698,7 @@ bool gm_80164330(u8 arg0)
     if (total_stages_on == 0) {
         OSReport("RandomStageSwitch All-Off!!!!!!!!!!!!!!!!!!!!!!!!!!\n");
     }
-    if ((1 << arg0) & gmMainLib_8015CC58()->stage_mask) {
+    if ((1 << (u16) arg0) & gmMainLib_8015CC58()->stage_mask) {
         return true;
     }
     return false;
diff --git a/src/melee/gm/gm_1601.h b/src/melee/gm/gm_1601.h
index ff2ec1bb44..6477233232 100644
--- a/src/melee/gm/gm_1601.h
+++ b/src/melee/gm/gm_1601.h
@@ -128,7 +128,7 @@ struct lbl_8046B488_t;
 /* 1641E4 */ void gm_801641E4(u8 stage, u8 enable);
 /* 164250 */ long gm_80164250(u16);
 /* 1642A0 */ bool fn_801642A0(void);
-/* 164330 */ bool gm_80164330(u8);
+/* 164330 */ bool gm_80164330(s32);
 /* 164430 */ bool gm_80164430(u16);
 /* 164504 */ int gm_80164504(u16);
 /* 164600 */ bool gm_80164600(void);
diff --git a/src/melee/gr/grbigblueroute.c b/src/melee/gr/grbigblueroute.c
index 279f71edf8..d219a8bcf3 100644
--- a/src/melee/gr/grbigblueroute.c
+++ b/src/melee/gr/grbigblueroute.c
@@ -361,14 +361,20 @@ void grBigBlueRoute_8020C210(Ground_GObj* gobj)
 /// @todo Currently 96.39% match - r30/r31 register swap
 void grBigBlueRoute_8020C238(Ground_GObj* gobj)
 {
-    Ground* gp = GET_GROUND(gobj);
     HSD_JObj* root_jobj = gobj->hsd_obj;
+    HSD_JObj* jobj;
+    Ground* gp = gobj->user_data;
     Vec3 spline_pt;
     PAD_STACK(8);
 
     if (root_jobj != NULL) {
         f32 val;
-        HSD_JObj* jobj = root_jobj->child;
+
+        if (root_jobj == NULL) {
+            jobj = NULL;
+        } else {
+            jobj = root_jobj->child;
+        }
 
         while (jobj != NULL) {
             HSD_JObjSetFlagsAll(jobj, JOBJ_HIDDEN);
@@ -382,10 +388,10 @@ void grBigBlueRoute_8020C238(Ground_GObj* gobj)
             val = grBb_Route_804D6A68->x4;
             HSD_JObjSetScaleZ(jobj, val);
 
-            if (jobj != NULL) {
-                jobj = jobj->next;
-            } else {
+            if (jobj == NULL) {
                 jobj = NULL;
+            } else {
+                jobj = jobj->next;
             }
         }
     }
diff --git a/src/melee/gr/grpura.c b/src/melee/gr/grpura.c
index 060bcedea1..29e8f9d5de 100644
--- a/src/melee/gr/grpura.c
+++ b/src/melee/gr/grpura.c
@@ -95,14 +95,14 @@ const f32 grPu_804DBA74 = 2.0;
 const f32 grPu_804DBA78 = 30.0;
 const f32 grPu_804DBA7C = -30.0;
 
+/* 4D6AA0 */ static HSD_GObj* grPu_804D6AA0;
+
 void grPura_80211D00(void)
 {
-    HSD_GObj* r3;
-
     Vec3 cam_offset;
     f32 fVar1;
 
-    Ground_801C49F8();
+    grPu_804D6AA0 = Ground_801C49F8();
     stage_info.unk8C.b4 = 0;
     stage_info.unk8C.b5 = 1;
     grPura_80211E08(0);
@@ -120,7 +120,7 @@ void grPura_80211D00(void)
     fVar1 = Stage_GetCamBoundsLeftOffset();
     Ground_801C38A0(grPu_804DBA58 * (fVar1 - cam_offset.x));
     fVar1 = Stage_GetCamBoundsRightOffset();
-    Ground_801C38AC(grPu_804DBA58 * ((fVar1 - cam_offset.x)));
+    Ground_801C38AC(grPu_804DBA58 * (fVar1 - cam_offset.x));
 }
 
 void grPura_80211DD8(void) {}
diff --git a/src/melee/if/soundtest.c b/src/melee/if/soundtest.c
index 05d0e202e7..b9bb558a42 100644
--- a/src/melee/if/soundtest.c
+++ b/src/melee/if/soundtest.c
@@ -12,6 +12,7 @@
 #include "lb/lbcardnew.h"
 #include "lb/lblanguage.h"
 #include "lb/lbsnap.h"
+#include "placeholder.h"
 #include "ty/toy.h"
 #include "ty/tylist.h"
 
@@ -1166,6 +1167,7 @@ int un_80301634(void)
     HSD_GObj* gobj;
     void* r31;
     void* r3;
+    PAD_STACK(8);
 
     OSReport("<Init>\n");
     lb_8001C550();
@@ -1177,7 +1179,6 @@ int un_80301634(void)
     gobj = GObj_Create(0x13, 0x14, 0);
     HSD_SObjLib_803A55DC(gobj, 0x280, 0x1E0, 0xC);
     gobj->gxlink_prios = 0x40000;
-    gobj->user_data = NULL;
     un_804D6E08 = 0;
     return 0;
 }
diff --git a/src/melee/it/items/itlinkboomerang.c b/src/melee/it/items/itlinkboomerang.c
index 34ff33e073..90dbb4f485 100644
--- a/src/melee/it/items/itlinkboomerang.c
+++ b/src/melee/it/items/itlinkboomerang.c
@@ -70,8 +70,8 @@ static inline void clamp_tau(f32* angle)
 
 static inline void norm_xF74(Item* ip)
 {
-    f32 angle;
-    if (ip->facing_dir != 1.0f) {
+    f64 angle;
+    if (ip->facing_dir == 1.0f) {
         angle = ip->xDD4_itemVar.linkboomerang.xF74;
     } else {
         angle = ip->xDD4_itemVar.linkboomerang.xF74 - M_PI;
diff --git a/src/melee/lb/lbcardnew.c b/src/melee/lb/lbcardnew.c
index d4e3bc15b4..a7802b484e 100644
--- a/src/melee/lb/lbcardnew.c
+++ b/src/melee/lb/lbcardnew.c
@@ -670,13 +670,13 @@ int lb_8001B6F8(void)
     return result;
 }
 
-void lb_8001B760(int result)
+int lb_8001B760(int result)
 {
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
+    return result;
 }
 
 inline struct CardTask* setup_task(int a, int b)
@@ -709,8 +709,7 @@ u32 lb_8001B7E0(int chan, char* filename, void* file_entries, void* save_data,
     result = lb_80019CB0(0x10);
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
     return result;
 }
@@ -719,7 +718,7 @@ int lb_8001B8C8(int chan)
 {
     int unused_1;
     int unused_2;
-    int result;
+    s32 result;
     u8 _[0x18];
 
     lb_80019EF0(chan, 0, 0, 0);
@@ -730,8 +729,7 @@ int lb_8001B8C8(int chan)
     result = lb_80019CB0(0x10);
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
     return result;
 }
@@ -754,8 +752,7 @@ int lb_8001BA44(int chan, const char* filename, UNK_T status_out)
     s32 result = lb_8001B99C(chan, filename, status_out);
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
     return result;
 }
@@ -784,8 +781,7 @@ int lb_8001BC18(int chan, char* filename, void** file_entries, void* save_data,
                              write_buf, write_offset, write_len, status_out);
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
     return result;
 }
@@ -810,8 +806,7 @@ int lb_8001BD34(int chan, const char* filename, UNK_T file_entries,
     result = lb_80019CB0(0x10);
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
     return result;
 }
@@ -867,8 +862,7 @@ int lb_8001BFD8(int chan, UNK_T snap_buf, UNK_T snap_count, UNK_T snap_max)
     result = lb_80019CB0(0x10);
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
     return result;
 }
@@ -936,8 +930,7 @@ int lb_8001C2D8(int chan, const char* company, const char* game_name,
     result = lb_80019CB0(0x10);
     if (result == 0xB) {
         do {
-            result = lb_8001B6F8();
-        } while (result == 0xB);
+        } while ((result = lb_8001B6F8()) == 0xB);
     }
     return result;
 }
diff --git a/src/melee/lb/lbcardnew.h b/src/melee/lb/lbcardnew.h
index bc522d7a7a..2274bd9eaf 100644
--- a/src/melee/lb/lbcardnew.h
+++ b/src/melee/lb/lbcardnew.h
@@ -30,7 +30,7 @@
 /* 01B614 */ int lb_8001B614(const char* filename);
 /* 01B6E0 */ s32 lb_8001B6E0(s32 file_idx);
 /* 01B6F8 */ enum_t lb_8001B6F8(void);
-/* 01B760 */ void lb_8001B760(int result);
+/* 01B760 */ int lb_8001B760(int result);
 /* 01B7E0 */ u32 lb_8001B7E0(int chan, char* filename, void* file_entries,
                              void* save_data, int* status_out);
 /* 01B8C8 */ bool lb_8001B8C8(int chan);
