diff --git a/config/GALE01/symbols.txt b/config/GALE01/symbols.txt
index 10426b50d4..99fb3a41a9 100644
--- a/config/GALE01/symbols.txt
+++ b/config/GALE01/symbols.txt
@@ -33289,8 +33289,8 @@ it_804DD378 = .sdata2:0x804DD378; // type:object size:0x4 scope:global align:4 d
 @291 = .sdata2:0x804DD39C; // type:object size:0x4 scope:local align:4 data:float
 @145 = .sdata2:0x804DD3A0; // type:object size:0x4 scope:local align:4 data:float
 @160 = .sdata2:0x804DD3A4; // type:object size:0x4 scope:local align:4 data:float
-it_804DD3A8 = .sdata2:0x804DD3A8; // type:object size:0x4 scope:global align:4 data:float
-it_804DD3AC = .sdata2:0x804DD3AC; // type:object size:0x4 scope:global align:4 data:float
+@149 = .sdata2:0x804DD3A8; // type:object size:0x4 scope:local align:4 data:float
+@175 = .sdata2:0x804DD3AC; // type:object size:0x4 scope:local align:4 data:float
 @145 = .sdata2:0x804DD3B0; // type:object size:0x4 scope:local align:4 data:float
 @161 = .sdata2:0x804DD3B4; // type:object size:0x4 scope:local align:4 data:float
 it_804DD3B8 = .sdata2:0x804DD3B8; // type:object size:0x4 scope:global align:4 data:float
diff --git a/configure.py b/configure.py
index 79f8514d1d..8582e861b8 100755
--- a/configure.py
+++ b/configure.py
@@ -1220,7 +1220,7 @@ def MatchingFor(*versions):
             Object(NonMatching, "melee/it/items/itmatadogas.c"),
             Object(Matching, "melee/it/items/itlizardon.c"),
             Object(Matching, "melee/it/items/itfire.c"),
-            Object(NonMatching, "melee/it/items/itthunder.c"),
+            Object(Matching, "melee/it/items/itthunder.c"),
             Object(Matching, "melee/it/items/itfreezer.c"),
             Object(NonMatching, "melee/it/items/itsonans.c"),
             Object(Matching, "melee/it/items/ithassam.c"),
diff --git a/src/melee/it/items/itthunder.c b/src/melee/it/items/itthunder.c
index 599a64f12c..151a8d5d9e 100644
--- a/src/melee/it/items/itthunder.c
+++ b/src/melee/it/items/itthunder.c
@@ -94,7 +94,7 @@ void it_802CCC68(HSD_GObj* gobj)
     Item_80268E5C(gobj, 2, ITEM_ANIM_UPDATE);
     item->entered_hitlag = efLib_PauseAll;
     item->exited_hitlag = efLib_ResumeAll;
-    item->xDD4_itemVar.thunder.x68 = it_804DD3A8;
+    item->xDD4_itemVar.thunder.x68 = 0.0f;
     item->xDB0_itcmd_var1 = 0;
 }
 
diff --git a/src/melee/it/items/itthunder.h b/src/melee/it/items/itthunder.h
index 8c04b78c24..883365c403 100644
--- a/src/melee/it/items/itthunder.h
+++ b/src/melee/it/items/itthunder.h
@@ -22,7 +22,4 @@
 /* 2CCF74 */ bool itThunder_UnkMotion0_Coll(Item_GObj* gobj);
 /* 3F7C40 */ extern ItemStateTable it_803F7C40[];
 
-f32 it_804DD3AC = 1.0f;
-f32 it_804DD3A8 = 0.0f;
-
 #endif
